{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17664"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.getpid()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber\n",
      "/users/lserrano/mweber\n",
      "20\n"
     ]
    }
   ],
   "source": [
    "import matplotlib\n",
    "%matplotlib inline\n",
    "import seaborn\n",
    "seaborn.set_style(\"darkgrid\")\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "\n",
    "import pandas as pd\n",
    "idx = pd.IndexSlice\n",
    "pd.options.display.max_columns = 35\n",
    "from xlsxwriter.utility import xl_col_to_name, xl_range, xl_rowcol_to_cell\n",
    "import numpy as np\n",
    "import os.path\n",
    "from pathlib import Path\n",
    "import glob\n",
    "import re\n",
    "import lxml.etree as etree\n",
    "#import xml.etree.ElementTree as ET\n",
    "import json\n",
    "import scipy\n",
    "from math import log2\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "# Use autoreload to keep up-to-date version of custom packages\n",
    "\n",
    "import mwTools\n",
    "import cterminal\n",
    "import analysis_COG_enrichment\n",
    "from cterminal import full_analysis_multispecies\n",
    "from cterminal import simplify_prefix\n",
    "from analysis_COG_enrichment import enrichment_analysis_cterminal_subset\n",
    "from mwTools.paths import p\n",
    "\n",
    "p = p()\n",
    "print(p.rootPath)\n",
    "rootPath = p.rootPath\n",
    "cterminalPath = p.cterminalPath\n",
    "refSeqPath = p.refSeqPath\n",
    "taxonomyPath = p.taxonomyPath\n",
    "phylaPath = p.phylaPath\n",
    "entrezGenePath = p.entrezGenePath\n",
    "OMAPath = p.OMAPath\n",
    "eggNOGPath = p.eggNOGPath\n",
    "NCBI_COG_path = p.NCBI_COG_path\n",
    "analysisEggnogPath = p.analysisEggnogPath\n",
    "analysisCtermDataPath = p.analysisCtermDataPath\n",
    "analysisCtermPlotsPath = p.analysisCtermPlotsPath\n",
    "\n",
    "from cterminal import aaTable\n",
    "print(len(aaTable))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/users/lserrano/mweber/Databases/OMA')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OMAPath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EggNOG database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### taxonomy table EggNOG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax id</th>\n",
       "      <th>nog prefix</th>\n",
       "      <th>level name</th>\n",
       "      <th>species count</th>\n",
       "      <th>core species count</th>\n",
       "      <th>periphery species count</th>\n",
       "      <th>adherent species count</th>\n",
       "      <th>members tax ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NOG</td>\n",
       "      <td>LUCA</td>\n",
       "      <td>3686</td>\n",
       "      <td>848</td>\n",
       "      <td>1183</td>\n",
       "      <td>1655</td>\n",
       "      <td>394 882 883 1140 1148 2110 2850 3055 3218 3694...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>57723</td>\n",
       "      <td>aciNOG</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>204669 234267 240015 401053 682795 940615 981222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>204432</td>\n",
       "      <td>acidNOG</td>\n",
       "      <td>Acidobacteriia</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>240015 401053 682795 940615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>422676</td>\n",
       "      <td>acoNOG</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>5821 5825 5833 5850 5855 333668 352914 353154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>201174</td>\n",
       "      <td>actNOG</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>397</td>\n",
       "      <td>75</td>\n",
       "      <td>106</td>\n",
       "      <td>216</td>\n",
       "      <td>31964 83331 83332 100226 101510 106370 134676 ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax id nog prefix      level name  species count  core species count  \\\n",
       "0        1        NOG            LUCA           3686                 848   \n",
       "1    57723     aciNOG   Acidobacteria              7                   6   \n",
       "2   204432    acidNOG  Acidobacteriia              4                   3   \n",
       "3   422676     acoNOG    Aconoidasida              8                   1   \n",
       "4   201174     actNOG  Actinobacteria            397                  75   \n",
       "\n",
       "   periphery species count  adherent species count  \\\n",
       "0                     1183                    1655   \n",
       "1                        1                       0   \n",
       "2                        1                       0   \n",
       "3                        7                       0   \n",
       "4                      106                     216   \n",
       "\n",
       "                                     members tax ids  \n",
       "0  394 882 883 1140 1148 2110 2850 3055 3218 3694...  \n",
       "1   204669 234267 240015 401053 682795 940615 981222  \n",
       "2                        240015 401053 682795 940615  \n",
       "3      5821 5825 5833 5850 5855 333668 352914 353154  \n",
       "4  31964 83331 83332 100226 101510 106370 134676 ...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggTaxonomicDf = pd.read_csv(str(eggNOGPath / 'eggnog4.taxonomic_levels.txt'), sep='\\t', header=0)\n",
    "eggTaxonomicDf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use the database to obtain the COG category of other type of annotation for every protein with a Uniprot ID,\n",
    "```\n",
    "UniprotID\n",
    "  -- id_mappings/uniprot/uniprot-15-May-2015.Bacteria.tsv.gz -->\n",
    "    orthogroup ID (ENOG, COG, KOG)\n",
    "      -- bactNOG.annotations -->\n",
    "        COG category (annotation)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List of proteins in orthogroups\n",
    "\n",
    "**bactNOG.members.tsv**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "eggMemberDf = pd.read_csv(str(eggNOGPath / 'data/bactNOG/bactNOG.members.tsv'), sep='\\t', header=None,\n",
    "                         names='TaxonomicLevel|GroupName|ProteinCount|SpeciesCount|COGFunctionalCategory|ProteinIDs'.split('|'))\n",
    "eggMemberDf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EggNOG README:\n",
    "\n",
    "** members files **\n",
    "\n",
    "Tab delimited files, each line representing:\n",
    "\n",
    "TaxonomicLevel|GroupName|ProteinCount|SpeciesCount|COGFunctionalCategory|ProteinIDs\n",
    "\n",
    "ProteinIds format = Taxid.SeqID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Stack the orthogroup database\n",
    "eggMemberStackedDf = split_and_stack_column(eggMemberDf, col='ProteinIDs', colFieldSep=',', chunkSize=2000, copy_df=True)\n",
    "\n",
    "# Write stacked orthogroup database to file\n",
    "eggMemberStackedDf.to_csv(str(eggNOGPath / 'data/bactNOG/bactNOG.members_stacked.tsv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the script `eggNOG_stack_members.py` with 16G of memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "eggMemberStackedDf = pd.read_csv(str(eggNOGPath / 'data/bactNOG/bactNOG.members_stacked.tsv'),\n",
    "                                 sep='\\t').drop('Unnamed: 0', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at EggNOG phylogenetic and taxonomic tree..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax id</th>\n",
       "      <th>nog prefix</th>\n",
       "      <th>level name</th>\n",
       "      <th>species count</th>\n",
       "      <th>core species count</th>\n",
       "      <th>periphery species count</th>\n",
       "      <th>adherent species count</th>\n",
       "      <th>members tax ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NOG</td>\n",
       "      <td>LUCA</td>\n",
       "      <td>3686</td>\n",
       "      <td>848</td>\n",
       "      <td>1183</td>\n",
       "      <td>1655</td>\n",
       "      <td>394 882 883 1140 1148 2110 2850 3055 3218 3694...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>57723</td>\n",
       "      <td>aciNOG</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>204669 234267 240015 401053 682795 940615 981222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>204432</td>\n",
       "      <td>acidNOG</td>\n",
       "      <td>Acidobacteriia</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>240015 401053 682795 940615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>422676</td>\n",
       "      <td>acoNOG</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>5821 5825 5833 5850 5855 333668 352914 353154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>201174</td>\n",
       "      <td>actNOG</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>397</td>\n",
       "      <td>75</td>\n",
       "      <td>106</td>\n",
       "      <td>216</td>\n",
       "      <td>31964 83331 83332 100226 101510 106370 134676 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5338</td>\n",
       "      <td>agaNOG</td>\n",
       "      <td>Agaricales</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>240176 486041 578458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>155619</td>\n",
       "      <td>agarNOG</td>\n",
       "      <td>Agaricomycetes</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5306 240176 486041 578458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5794</td>\n",
       "      <td>apiNOG</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>5811 5821 5825 5833 5850 5855 333668 352914 35...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>28211</td>\n",
       "      <td>aproNOG</td>\n",
       "      <td>Proteobacteria_alpha</td>\n",
       "      <td>290</td>\n",
       "      <td>90</td>\n",
       "      <td>102</td>\n",
       "      <td>98</td>\n",
       "      <td>394 52598 66084 78245 82789 89187 114615 13772...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>200783</td>\n",
       "      <td>aquNOG</td>\n",
       "      <td>Aquificae</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>123214 204536 224324 380749 436114 608538 6383...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2157</td>\n",
       "      <td>arNOG</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>130</td>\n",
       "      <td>66</td>\n",
       "      <td>49</td>\n",
       "      <td>15</td>\n",
       "      <td>64091 69014 70601 79929 178306 186497 187420 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>183980</td>\n",
       "      <td>arcNOG</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>224325 572546 589924 693661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>6656</td>\n",
       "      <td>artNOG</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>22</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>6669 6945 7029 7070 7091 7159 7165 7176 7217 7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>34384</td>\n",
       "      <td>arthNOG</td>\n",
       "      <td>Arthrodermataceae</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>535722 554155 559305 663202 663331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4890</td>\n",
       "      <td>ascNOG</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>71</td>\n",
       "      <td>38</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>4896 4929 4932 4950 4952 4956 5022 5057 5059 5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>8782</td>\n",
       "      <td>aveNOG</td>\n",
       "      <td>Aves</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9031 9103 59729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>91061</td>\n",
       "      <td>bacNOG</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>692</td>\n",
       "      <td>37</td>\n",
       "      <td>164</td>\n",
       "      <td>491</td>\n",
       "      <td>40041 66692 93061 93062 158878 158879 160490 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>bactNOG</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>3318</td>\n",
       "      <td>667</td>\n",
       "      <td>1011</td>\n",
       "      <td>1640</td>\n",
       "      <td>394 882 883 1140 1148 2110 31964 32049 32051 3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>200643</td>\n",
       "      <td>bacteNOG</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>102</td>\n",
       "      <td>13</td>\n",
       "      <td>51</td>\n",
       "      <td>38</td>\n",
       "      <td>203275 226186 242619 264731 272559 295405 4114...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>5204</td>\n",
       "      <td>basNOG</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5270 5297 5306 208348 214684 240176 283643 367...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>976</td>\n",
       "      <td>bctoNOG</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>176</td>\n",
       "      <td>54</td>\n",
       "      <td>79</td>\n",
       "      <td>43</td>\n",
       "      <td>50743 156586 203275 216432 226186 242619 26473...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>33213</td>\n",
       "      <td>biNOG</td>\n",
       "      <td>Bilateria</td>\n",
       "      <td>85</td>\n",
       "      <td>40</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>6183 6238 6239 6334 6669 6945 7029 7070 7091 7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>28216</td>\n",
       "      <td>bproNOG</td>\n",
       "      <td>Proteobacteria_beta</td>\n",
       "      <td>199</td>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>88</td>\n",
       "      <td>62928 75379 76114 85643 122586 122587 153948 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>3699</td>\n",
       "      <td>braNOG</td>\n",
       "      <td>Brassicales</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3702 3711 59689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>33554</td>\n",
       "      <td>carNOG</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9615 9646 9669 9685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>35718</td>\n",
       "      <td>chaNOG</td>\n",
       "      <td>Chaetomiaceae</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>306901 573729 578455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1090</td>\n",
       "      <td>chlNOG</td>\n",
       "      <td>Chlorobi</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>194439 290315 290317 290318 290512 319225 3249...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>204428</td>\n",
       "      <td>chlaNOG</td>\n",
       "      <td>Chlamydiae</td>\n",
       "      <td>47</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>34</td>\n",
       "      <td>115713 138677 159254 182082 218497 227941 2431...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>200795</td>\n",
       "      <td>chloNOG</td>\n",
       "      <td>Chloroflexi</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>216389 243164 255470 309801 311424 316274 3246...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>32061</td>\n",
       "      <td>chlorNOG</td>\n",
       "      <td>Chloroflexi</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>316274 324602 326427 357808 383372 480224 765420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>6231</td>\n",
       "      <td>nemNOG</td>\n",
       "      <td>Nematodes</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6238 6239 6334 31234 54126 135651 281687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>33183</td>\n",
       "      <td>onyNOG</td>\n",
       "      <td>Onygenales</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>222929 246410 336963 339724 502779 535722 5541...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>33154</td>\n",
       "      <td>opiNOG</td>\n",
       "      <td>Opisthokonts</td>\n",
       "      <td>175</td>\n",
       "      <td>88</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>4896 4929 4932 4950 4952 4956 5022 5057 5059 5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>4776</td>\n",
       "      <td>perNOG</td>\n",
       "      <td>Peronosporales</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4787 67593 164328 272952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>203682</td>\n",
       "      <td>plaNOG</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>243090 314230 344747 521674 530564 575540 756272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>92860</td>\n",
       "      <td>pleNOG</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5022 13684 426418 861557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>38820</td>\n",
       "      <td>poaNOG</td>\n",
       "      <td>Poales</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>4513 4533 4538 4555 4558 4577 15368 39946 39947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>9443</td>\n",
       "      <td>prNOG</td>\n",
       "      <td>Primates</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>9478 9483 9544 9593 9598 9601 9606 30608 30611...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>1224</td>\n",
       "      <td>proNOG</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1378</td>\n",
       "      <td>297</td>\n",
       "      <td>394</td>\n",
       "      <td>687</td>\n",
       "      <td>394 882 883 36870 52598 56780 58051 59196 6048...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>6236</td>\n",
       "      <td>rhaNOG</td>\n",
       "      <td>Rhabditida</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>6238 6239 31234 135651 281687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>9989</td>\n",
       "      <td>roNOG</td>\n",
       "      <td>Rodents</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>10020 10090 10116 10141 43179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>4893</td>\n",
       "      <td>sacNOG</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4932 4950 4956 33169 284590 284593 381046 4369...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>4891</td>\n",
       "      <td>saccNOG</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>4929 4932 4950 4952 4956 33169 284590 284593 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>5139</td>\n",
       "      <td>sorNOG</td>\n",
       "      <td>Sordariales</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5141 306901 515849 573729 578455 771870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>147550</td>\n",
       "      <td>sordNOG</td>\n",
       "      <td>Sordariomycetes</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>5141 5507 5518 29850 29875 51453 117187 140110...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>117747</td>\n",
       "      <td>sphNOG</td>\n",
       "      <td>Sphingobacteriia</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>391596 485917 485918 525372 525373 700598 7149...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>203691</td>\n",
       "      <td>spiNOG</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>61</td>\n",
       "      <td>7</td>\n",
       "      <td>24</td>\n",
       "      <td>30</td>\n",
       "      <td>158189 158190 189518 224326 243275 243276 2676...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>314146</td>\n",
       "      <td>spriNOG</td>\n",
       "      <td>Supraprimates</td>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>9478 9483 9544 9593 9598 9601 9606 9986 10020 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>35493</td>\n",
       "      <td>strNOG</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>3218 3694 3702 3711 3847 4081 4113 4513 4533 4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>508458</td>\n",
       "      <td>synNOG</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>352165 469381 525903 572547 580340 584708 5920...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>544448</td>\n",
       "      <td>tenNOG</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>65</td>\n",
       "      <td>5</td>\n",
       "      <td>28</td>\n",
       "      <td>32</td>\n",
       "      <td>2110 37692 59748 243272 243273 262719 262722 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>651137</td>\n",
       "      <td>thaNOG</td>\n",
       "      <td>Thaumarchaeota</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>436308 886738 1001994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>183967</td>\n",
       "      <td>theNOG</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>263820 273075 273116 333146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>200918</td>\n",
       "      <td>therNOG</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>126740 243274 309803 381764 390874 391009 4038...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>183968</td>\n",
       "      <td>thermNOG</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>69014 70601 186497 246969 272844 342949 391623...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>5234</td>\n",
       "      <td>treNOG</td>\n",
       "      <td>Tremellales</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>214684 283643 367775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>7742</td>\n",
       "      <td>veNOG</td>\n",
       "      <td>Vertebrates</td>\n",
       "      <td>51</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>7757 7897 7955 8049 8083 8090 8128 8364 9031 9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>74201</td>\n",
       "      <td>verNOG</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>240016 320771 349741 382464 452637 481448 4979...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>203494</td>\n",
       "      <td>verrNOG</td>\n",
       "      <td>Verrucomicrobiae</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>240016 320771 349741 382464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>33090</td>\n",
       "      <td>virNOG</td>\n",
       "      <td>Viridiplantae</td>\n",
       "      <td>23</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>3055 3218 3694 3702 3711 3847 4081 4113 4513 4...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #tax id nog prefix            level name  species count  \\\n",
       "0          1        NOG                  LUCA           3686   \n",
       "1      57723     aciNOG         Acidobacteria              7   \n",
       "2     204432    acidNOG        Acidobacteriia              4   \n",
       "3     422676     acoNOG          Aconoidasida              8   \n",
       "4     201174     actNOG        Actinobacteria            397   \n",
       "5       5338     agaNOG            Agaricales              3   \n",
       "6     155619    agarNOG        Agaricomycetes              4   \n",
       "7       5794     apiNOG           Apicomplexa             13   \n",
       "8      28211    aproNOG  Proteobacteria_alpha            290   \n",
       "9     200783     aquNOG             Aquificae              9   \n",
       "10      2157      arNOG               Archaea            130   \n",
       "11    183980     arcNOG          Archaeoglobi              4   \n",
       "12      6656     artNOG            Arthropoda             22   \n",
       "13     34384    arthNOG     Arthrodermataceae              5   \n",
       "14      4890     ascNOG            Ascomycota             71   \n",
       "15      8782     aveNOG                  Aves              3   \n",
       "16     91061     bacNOG               Bacilli            692   \n",
       "17         2    bactNOG              Bacteria           3318   \n",
       "18    200643   bacteNOG           Bacteroidia            102   \n",
       "19      5204     basNOG         Basidiomycota             11   \n",
       "20       976    bctoNOG         Bacteroidetes            176   \n",
       "21     33213      biNOG             Bilateria             85   \n",
       "22     28216    bproNOG   Proteobacteria_beta            199   \n",
       "23      3699     braNOG           Brassicales              3   \n",
       "24     33554     carNOG             Carnivora              4   \n",
       "25     35718     chaNOG         Chaetomiaceae              3   \n",
       "26      1090     chlNOG              Chlorobi             12   \n",
       "27    204428    chlaNOG            Chlamydiae             47   \n",
       "28    200795    chloNOG           Chloroflexi             17   \n",
       "29     32061   chlorNOG           Chloroflexi              7   \n",
       "..       ...        ...                   ...            ...   \n",
       "77      6231     nemNOG             Nematodes              7   \n",
       "78     33183     onyNOG            Onygenales             11   \n",
       "79     33154     opiNOG          Opisthokonts            175   \n",
       "80      4776     perNOG        Peronosporales              4   \n",
       "81    203682     plaNOG        Planctomycetes              7   \n",
       "82     92860     pleNOG          Pleosporales              4   \n",
       "83     38820     poaNOG                Poales              9   \n",
       "84      9443      prNOG              Primates             10   \n",
       "85      1224     proNOG        Proteobacteria           1378   \n",
       "86      6236     rhaNOG            Rhabditida              5   \n",
       "87      9989      roNOG               Rodents              5   \n",
       "88      4893     sacNOG    Saccharomycetaceae             13   \n",
       "89      4891    saccNOG       Saccharomycetes             21   \n",
       "90      5139     sorNOG           Sordariales              6   \n",
       "91    147550    sordNOG       Sordariomycetes             18   \n",
       "92    117747     sphNOG      Sphingobacteriia             10   \n",
       "93    203691     spiNOG          Spirochaetes             61   \n",
       "94    314146    spriNOG         Supraprimates             17   \n",
       "95     35493     strNOG          Streptophyta             20   \n",
       "96    508458     synNOG         Synergistetes              8   \n",
       "97    544448     tenNOG           Tenericutes             65   \n",
       "98    651137     thaNOG        Thaumarchaeota              3   \n",
       "99    183967     theNOG        Thermoplasmata              4   \n",
       "100   200918    therNOG           Thermotogae             14   \n",
       "101   183968   thermNOG           Thermococci             12   \n",
       "102     5234     treNOG           Tremellales              3   \n",
       "103     7742      veNOG           Vertebrates             51   \n",
       "104    74201     verNOG       Verrucomicrobia              8   \n",
       "105   203494    verrNOG      Verrucomicrobiae              4   \n",
       "106    33090     virNOG         Viridiplantae             23   \n",
       "\n",
       "     core species count  periphery species count  adherent species count  \\\n",
       "0                   848                     1183                    1655   \n",
       "1                     6                        1                       0   \n",
       "2                     3                        1                       0   \n",
       "3                     1                        7                       0   \n",
       "4                    75                      106                     216   \n",
       "5                     3                        0                       0   \n",
       "6                     3                        1                       0   \n",
       "7                     2                       11                       0   \n",
       "8                    90                      102                      98   \n",
       "9                     8                        1                       0   \n",
       "10                   66                       49                      15   \n",
       "11                    2                        2                       0   \n",
       "12                   10                       12                       0   \n",
       "13                    2                        3                       0   \n",
       "14                   38                       33                       0   \n",
       "15                    2                        1                       0   \n",
       "16                   37                      164                     491   \n",
       "17                  667                     1011                    1640   \n",
       "18                   13                       51                      38   \n",
       "19                    6                        5                       0   \n",
       "20                   54                       79                      43   \n",
       "21                   40                       45                       0   \n",
       "22                   57                       54                      88   \n",
       "23                    2                        1                       0   \n",
       "24                    3                        1                       0   \n",
       "25                    2                        1                       0   \n",
       "26                    5                        7                       0   \n",
       "27                    7                        6                      34   \n",
       "28                   10                        4                       3   \n",
       "29                    4                        2                       1   \n",
       "..                  ...                      ...                     ...   \n",
       "77                    1                        6                       0   \n",
       "78                    5                        6                       0   \n",
       "79                   88                       87                       0   \n",
       "80                    2                        2                       0   \n",
       "81                    5                        2                       0   \n",
       "82                    2                        2                       0   \n",
       "83                    4                        5                       0   \n",
       "84                    5                        5                       0   \n",
       "85                  297                      394                     687   \n",
       "86                    1                        4                       0   \n",
       "87                    2                        3                       0   \n",
       "88                    9                        4                       0   \n",
       "89                   12                        9                       0   \n",
       "90                    5                        1                       0   \n",
       "91                   12                        6                       0   \n",
       "92                    6                        3                       1   \n",
       "93                    7                       24                      30   \n",
       "94                    7                       10                       0   \n",
       "95                   13                        7                       0   \n",
       "96                    8                        0                       0   \n",
       "97                    5                       28                      32   \n",
       "98                    2                        1                       0   \n",
       "99                    3                        1                       0   \n",
       "100                   7                        5                       2   \n",
       "101                   2                       10                       0   \n",
       "102                   1                        2                       0   \n",
       "103                  26                       25                       0   \n",
       "104                   7                        1                       0   \n",
       "105                   3                        1                       0   \n",
       "106                  15                        8                       0   \n",
       "\n",
       "                                       members tax ids  \n",
       "0    394 882 883 1140 1148 2110 2850 3055 3218 3694...  \n",
       "1     204669 234267 240015 401053 682795 940615 981222  \n",
       "2                          240015 401053 682795 940615  \n",
       "3        5821 5825 5833 5850 5855 333668 352914 353154  \n",
       "4    31964 83331 83332 100226 101510 106370 134676 ...  \n",
       "5                                 240176 486041 578458  \n",
       "6                            5306 240176 486041 578458  \n",
       "7    5811 5821 5825 5833 5850 5855 333668 352914 35...  \n",
       "8    394 52598 66084 78245 82789 89187 114615 13772...  \n",
       "9    123214 204536 224324 380749 436114 608538 6383...  \n",
       "10   64091 69014 70601 79929 178306 186497 187420 1...  \n",
       "11                         224325 572546 589924 693661  \n",
       "12   6669 6945 7029 7070 7091 7159 7165 7176 7217 7...  \n",
       "13                  535722 554155 559305 663202 663331  \n",
       "14   4896 4929 4932 4950 4952 4956 5022 5057 5059 5...  \n",
       "15                                     9031 9103 59729  \n",
       "16   40041 66692 93061 93062 158878 158879 160490 1...  \n",
       "17   394 882 883 1140 1148 2110 31964 32049 32051 3...  \n",
       "18   203275 226186 242619 264731 272559 295405 4114...  \n",
       "19   5270 5297 5306 208348 214684 240176 283643 367...  \n",
       "20   50743 156586 203275 216432 226186 242619 26473...  \n",
       "21   6183 6238 6239 6334 6669 6945 7029 7070 7091 7...  \n",
       "22   62928 75379 76114 85643 122586 122587 153948 1...  \n",
       "23                                     3702 3711 59689  \n",
       "24                                 9615 9646 9669 9685  \n",
       "25                                306901 573729 578455  \n",
       "26   194439 290315 290317 290318 290512 319225 3249...  \n",
       "27   115713 138677 159254 182082 218497 227941 2431...  \n",
       "28   216389 243164 255470 309801 311424 316274 3246...  \n",
       "29    316274 324602 326427 357808 383372 480224 765420  \n",
       "..                                                 ...  \n",
       "77            6238 6239 6334 31234 54126 135651 281687  \n",
       "78   222929 246410 336963 339724 502779 535722 5541...  \n",
       "79   4896 4929 4932 4950 4952 4956 5022 5057 5059 5...  \n",
       "80                            4787 67593 164328 272952  \n",
       "81    243090 314230 344747 521674 530564 575540 756272  \n",
       "82                            5022 13684 426418 861557  \n",
       "83     4513 4533 4538 4555 4558 4577 15368 39946 39947  \n",
       "84   9478 9483 9544 9593 9598 9601 9606 30608 30611...  \n",
       "85   394 882 883 36870 52598 56780 58051 59196 6048...  \n",
       "86                       6238 6239 31234 135651 281687  \n",
       "87                       10020 10090 10116 10141 43179  \n",
       "88   4932 4950 4956 33169 284590 284593 381046 4369...  \n",
       "89   4929 4932 4950 4952 4956 33169 284590 284593 2...  \n",
       "90             5141 306901 515849 573729 578455 771870  \n",
       "91   5141 5507 5518 29850 29875 51453 117187 140110...  \n",
       "92   391596 485917 485918 525372 525373 700598 7149...  \n",
       "93   158189 158190 189518 224326 243275 243276 2676...  \n",
       "94   9478 9483 9544 9593 9598 9601 9606 9986 10020 ...  \n",
       "95   3218 3694 3702 3711 3847 4081 4113 4513 4533 4...  \n",
       "96   352165 469381 525903 572547 580340 584708 5920...  \n",
       "97   2110 37692 59748 243272 243273 262719 262722 2...  \n",
       "98                               436308 886738 1001994  \n",
       "99                         263820 273075 273116 333146  \n",
       "100  126740 243274 309803 381764 390874 391009 4038...  \n",
       "101  69014 70601 186497 246969 272844 342949 391623...  \n",
       "102                               214684 283643 367775  \n",
       "103  7757 7897 7955 8049 8083 8090 8128 8364 9031 9...  \n",
       "104  240016 320771 349741 382464 452637 481448 4979...  \n",
       "105                        240016 320771 349741 382464  \n",
       "106  3055 3218 3694 3702 3711 3847 4081 4113 4513 4...  \n",
       "\n",
       "[107 rows x 8 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggSpeciesDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build list of all RefSeq and EggNOG species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Importing list of genome assemblies of the RefSeq database with their taxonomy\n",
    "# Note: I have included both subspecies and species in the different taxa groups\n",
    "assemblySummaryRepTaxonomyDf = pd.read_csv(str(taxonomyPath / 'assemblySummaryRepTaxonomyDf.csv'))\n",
    "assemblySummaryRepTaxonomyDf.set_index('assembly_accession', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# All species included in the EggNOG database\n",
    "eggSpeciesDf = pd.read_csv(str(eggNOGPath / 'eggnog4_species_by_taxonomic_level.tsv'), sep='\\t', header=0)\n",
    "\n",
    "bactNOGSpeciesTaxidSet = eggSpeciesDf[eggSpeciesDf['nog prefix'] == 'bactNOG']['members tax ids'].iloc[0].split()\n",
    "bactNOGSpeciesTaxidSet = set([int(x) for x in bactNOGSpeciesTaxidSet])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1576\n"
     ]
    }
   ],
   "source": [
    "# write list of protein fasta files\n",
    "proteinSeqFastaPath = refSeqPath / 'Protein_sequences_fasta'\n",
    "proteinFastaFilenameList = list(assemblySummaryRepTaxonomyDf['compressedGenomeFilename']\\\n",
    "                                .map(lambda x: re.sub(r'(GCF.+)_genomic\\.gbff.*', r'\\1_protein.faa', x)))\n",
    "print(len(proteinFastaFilenameList))\n",
    "with open(str(proteinSeqFastaPath / 'protein_sequences_fasta_file_list.txt'), 'w') as f:\n",
    "    f.write(\"\\n\".join(proteinFastaFilenameList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "123899.0\n",
      "123899\n"
     ]
    }
   ],
   "source": [
    "# checking if the genome species was in EggNOG database\n",
    "print(assemblySummaryRepTaxonomyDf.loc['GCF_900078695.1', 'species_taxid'])\n",
    "print(assemblySummaryRepTaxonomyDf.loc['GCF_900078695.1', 'taxid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "123899 in bactNOGSpeciesTaxidSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>release_type</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>...</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>subclass_name</th>\n",
       "      <th>subclass_taxid</th>\n",
       "      <th>subfamily_name</th>\n",
       "      <th>subfamily_taxid</th>\n",
       "      <th>subgenus_name</th>\n",
       "      <th>subgenus_taxid</th>\n",
       "      <th>suborder_name</th>\n",
       "      <th>suborder_taxid</th>\n",
       "      <th>subphylum_name</th>\n",
       "      <th>subphylum_taxid</th>\n",
       "      <th>subspecies_name</th>\n",
       "      <th>subspecies_taxid</th>\n",
       "      <th>superkingdom_name</th>\n",
       "      <th>superkingdom_taxid</th>\n",
       "      <th>tribe_name</th>\n",
       "      <th>tribe_taxid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000016265.1</th>\n",
       "      <td>280</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02602977</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>311403</td>\n",
       "      <td>Agrobacterium radiobacter K84</td>\n",
       "      <td>strain=K84</td>\n",
       "      <td>ATCC BAA-868</td>\n",
       "      <td>latest</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2009/01/27</td>\n",
       "      <td>ASM1626v1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_000016265.1</td>\n",
       "      <td>...</td>\n",
       "      <td>358.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0   bioproject     biosample  wgs_master  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000016265.1            280  PRJNA224116  SAMN02602977         NaN   \n",
       "\n",
       "                          refseq_category   taxid  \\\n",
       "assembly_accession                                  \n",
       "GCF_000016265.1     representative genome  311403   \n",
       "\n",
       "                                    organism_name infraspecific_name  \\\n",
       "assembly_accession                                                     \n",
       "GCF_000016265.1     Agrobacterium radiobacter K84         strain=K84   \n",
       "\n",
       "                         isolate version_status   assembly_level release_type  \\\n",
       "assembly_accession                                                              \n",
       "GCF_000016265.1     ATCC BAA-868         latest  Complete Genome        Major   \n",
       "\n",
       "                   genome_rep seq_rel_date   asm_name submitter  \\\n",
       "assembly_accession                                                \n",
       "GCF_000016265.1          Full   2009/01/27  ASM1626v1       NaN   \n",
       "\n",
       "                    gbrs_paired_asm     ...     species_taxid subclass_name  \\\n",
       "assembly_accession                      ...                                   \n",
       "GCF_000016265.1     GCA_000016265.1     ...             358.0           NaN   \n",
       "\n",
       "                    subclass_taxid subfamily_name subfamily_taxid  \\\n",
       "assembly_accession                                                  \n",
       "GCF_000016265.1                NaN            NaN             NaN   \n",
       "\n",
       "                    subgenus_name subgenus_taxid  suborder_name  \\\n",
       "assembly_accession                                                \n",
       "GCF_000016265.1               NaN            NaN            NaN   \n",
       "\n",
       "                   suborder_taxid  subphylum_name subphylum_taxid  \\\n",
       "assembly_accession                                                  \n",
       "GCF_000016265.1               NaN             NaN             NaN   \n",
       "\n",
       "                    subspecies_name subspecies_taxid  superkingdom_name  \\\n",
       "assembly_accession                                                        \n",
       "GCF_000016265.1                 NaN              NaN           Bacteria   \n",
       "\n",
       "                   superkingdom_taxid  tribe_name tribe_taxid  \n",
       "assembly_accession                                             \n",
       "GCF_000016265.1                   2.0         NaN         NaN  \n",
       "\n",
       "[1 rows x 53 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['taxid'] == 311403]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build list of common species from RefSeq and eggNOG databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb of RefSeq assemblies: 1576\n",
      "Nb of RefSeq unique taxid at rank species: 1516\n"
     ]
    }
   ],
   "source": [
    "refSeqRankSpeciesSet = set(assemblySummaryRepTaxonomyDf['species_taxid'].unique())\n",
    "\n",
    "print(\"Nb of RefSeq assemblies:\", len(assemblySummaryRepTaxonomyDf))\n",
    "print(\"Nb of RefSeq unique taxid at rank species:\", len(refSeqRankSpeciesSet))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each genome in RefSeq, we will include it in the common set if the organism is in the same **species** as the organism in the EggNOG database. This way, we will be able to run the eggmapper only for organisms in the RefSeq database that are very close in taxonomy to at least one organism in the EggNOG database.\n",
    "\n",
    "We first build a list of **species** taxid for the EggNOG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from ete3 import NCBITaxa\n",
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rank: {1: 'no rank'} rank name: {1: 'root'}\n",
      "rank: {131567: 'no rank'} rank name: {131567: 'cellular organisms'}\n",
      "rank: {2: 'superkingdom'} rank name: {2: 'Bacteria'}\n",
      "rank: {1783272: 'no rank'} rank name: {1783272: 'Terrabacteria group'}\n",
      "rank: {544448: 'phylum'} rank name: {544448: 'Tenericutes'}\n",
      "rank: {31969: 'class'} rank name: {31969: 'Mollicutes'}\n",
      "rank: {186328: 'order'} rank name: {186328: 'Entomoplasmatales'}\n",
      "rank: {2131: 'family'} rank name: {2131: 'Spiroplasmataceae'}\n",
      "rank: {2132: 'genus'} rank name: {2132: 'Spiroplasma'}\n",
      "2145\n",
      "rank: {2145: 'species'} rank name: {2145: 'Spiroplasma taiwanense'}\n",
      "rank: {1276220: 'no rank'} rank name: {1276220: 'Spiroplasma taiwanense CT-1'}\n"
     ]
    }
   ],
   "source": [
    "# example\n",
    "for node in ncbi.get_lineage(1276220):\n",
    "    if ncbi.get_rank([node])[node] == 'species':\n",
    "        speciesTaxId = node\n",
    "        print(speciesTaxId)\n",
    "    print(\"rank:\", ncbi.get_rank([node]), \"rank name:\", ncbi.get_taxid_translator([node]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 508765 was translated into 935198\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 525327 was translated into 1423757\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 525364 was translated into 1423799\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 525366 was translated into 1423814\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 1027843 was translated into 1112267\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 1027844 was translated into 1112250\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 1027845 was translated into 1112254\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 1027846 was translated into 1112252\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 1126229 was translated into 1003195\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 439483 was translated into 929558\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 269483 was translated into 482957\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/home/mweber/local/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:224: UserWarning: taxid 245018 was translated into 649756\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "bactNOGRankSpeciesTaxidList = []\n",
    "\n",
    "for taxid in list(bactNOGSpeciesTaxidSet):\n",
    "    for node in ncbi.get_lineage(taxid):\n",
    "        if ncbi.get_rank([node])[node] == 'species':\n",
    "            bactNOGRankSpeciesTaxidList.append((taxid, node))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have a list of pairs **(taxid in EggNOG, corresponding taxid at the species rank)**. In some cases they are the same and in other cases not, because the EggNOG database uses both ranks in their species list."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Seems that in most cases EggNOG uses the subspecies taxid..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bactNOGRankSpeciesTaxidList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bactNOGRankSpeciesTaxidList = [i[1] for i in bactNOGRankSpeciesTaxidList]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1713\n"
     ]
    }
   ],
   "source": [
    "print(len(bactNOGRankSpeciesTaxidList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1713\n"
     ]
    }
   ],
   "source": [
    "bactNOGRankSpeciesTaxidList = set(bactNOGRankSpeciesTaxidList)\n",
    "print(len(bactNOGRankSpeciesTaxidList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with (analysisEggnogPath / 'bactNOGRankSpeciesTaxidList.json').open('w') as f:\n",
    "    json.dump(list(bactNOGRankSpeciesTaxidList), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with (analysisEggnogPath / 'bactNOGRankSpeciesTaxidList.json').open('r') as f:\n",
    "    bactNOGRankSpeciesTaxidList = set(json.load(f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1713"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bactNOGRankSpeciesTaxidList)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We choose all genomes in the RefSeq database that are in the set of **species** taxid from EggNOG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb of taxid at rank species in RefSeq: 1516\n",
      "Nb of taxid at rank species in EggNOG: 1713\n",
      "Nb of taxid at rank species in common 893\n"
     ]
    }
   ],
   "source": [
    "# Define the taxa ids that are common to the EggNOG bacterial database and the full RefSeq database\n",
    "eggRefSeqCommonRankSpeciesTaxidSet = bactNOGRankSpeciesTaxidList & refSeqRankSpeciesSet\n",
    "\n",
    "print(\"Nb of taxid at rank species in RefSeq:\", len(refSeqRankSpeciesSet))\n",
    "print(\"Nb of taxid at rank species in EggNOG:\", len(bactNOGRankSpeciesTaxidList))\n",
    "print(\"Nb of taxid at rank species in common\", len(eggRefSeqCommonRankSpeciesTaxidSet))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with (analysisEggnogPath / 'eggRefSeqCommonRankSpeciesTaxidSet.json').open('w') as f:\n",
    "    json.dump([int(j) for j in list(eggRefSeqCommonRankSpeciesTaxidSet)], f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with (analysisEggnogPath / 'eggRefSeqCommonRankSpeciesTaxidSet.json').open('r') as f:\n",
    "    eggRefSeqCommonRankSpeciesTaxidSet = set(json.load(f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "949\n"
     ]
    }
   ],
   "source": [
    "# Extract the corresponding genome accession for all species on the list\n",
    "assemblySummaryCommonSpeciesDf = \\\n",
    "    assemblySummaryRepTaxonomyDf[\\\n",
    "        assemblySummaryRepTaxonomyDf.apply(lambda x: x['species_taxid'] in eggRefSeqCommonRankSpeciesTaxidSet,\n",
    "                                           axis=1)\n",
    "                                 ]\n",
    "print(len(assemblySummaryCommonSpeciesDf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "assemblySummaryCommonSpeciesDf.to_csv(str(analysisEggnogPath / 'assemblySummaryCommonSpeciesDf.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>release_type</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>...</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>subclass_name</th>\n",
       "      <th>subclass_taxid</th>\n",
       "      <th>subfamily_name</th>\n",
       "      <th>subfamily_taxid</th>\n",
       "      <th>subgenus_name</th>\n",
       "      <th>subgenus_taxid</th>\n",
       "      <th>suborder_name</th>\n",
       "      <th>suborder_taxid</th>\n",
       "      <th>subphylum_name</th>\n",
       "      <th>subphylum_taxid</th>\n",
       "      <th>subspecies_name</th>\n",
       "      <th>subspecies_taxid</th>\n",
       "      <th>superkingdom_name</th>\n",
       "      <th>superkingdom_taxid</th>\n",
       "      <th>tribe_name</th>\n",
       "      <th>tribe_taxid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000005825.2</th>\n",
       "      <td>0</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>398511</td>\n",
       "      <td>Bacillus pseudofirmus OF4</td>\n",
       "      <td>strain=OF4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2010/12/15</td>\n",
       "      <td>ASM582v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_000005825.2</td>\n",
       "      <td>...</td>\n",
       "      <td>79885.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000005845.2</th>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA57779</td>\n",
       "      <td>SAMN02604091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>511145</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>strain=K-12 substr. MG1655</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2013/09/26</td>\n",
       "      <td>ASM584v2</td>\n",
       "      <td>Univ. Wisconsin</td>\n",
       "      <td>GCA_000005845.2</td>\n",
       "      <td>...</td>\n",
       "      <td>562.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006605.1</th>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA3283089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>306537</td>\n",
       "      <td>Corynebacterium jeikeium K411</td>\n",
       "      <td>strain=K411 = NCTC 11915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2005/06/27</td>\n",
       "      <td>ASM660v1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_000006605.1</td>\n",
       "      <td>...</td>\n",
       "      <td>38289.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006685.1</th>\n",
       "      <td>3</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603994</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>243161</td>\n",
       "      <td>Chlamydia muridarum str. Nigg</td>\n",
       "      <td>strain=Nigg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2003/03/26</td>\n",
       "      <td>ASM668v1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_000006685.1</td>\n",
       "      <td>...</td>\n",
       "      <td>83560.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006725.1</th>\n",
       "      <td>4</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603773</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>160492</td>\n",
       "      <td>Xylella fastidiosa 9a5c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2004/06/04</td>\n",
       "      <td>ASM672v1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_000006725.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2371.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0   bioproject     biosample  wgs_master  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000005825.2              0  PRJNA224116  SAMN02603086         NaN   \n",
       "GCF_000005845.2              1   PRJNA57779  SAMN02604091         NaN   \n",
       "GCF_000006605.1              2  PRJNA224116  SAMEA3283089         NaN   \n",
       "GCF_000006685.1              3  PRJNA224116  SAMN02603994         NaN   \n",
       "GCF_000006725.1              4  PRJNA224116  SAMN02603773         NaN   \n",
       "\n",
       "                          refseq_category   taxid  \\\n",
       "assembly_accession                                  \n",
       "GCF_000005825.2     representative genome  398511   \n",
       "GCF_000005845.2          reference genome  511145   \n",
       "GCF_000006605.1     representative genome  306537   \n",
       "GCF_000006685.1     representative genome  243161   \n",
       "GCF_000006725.1     representative genome  160492   \n",
       "\n",
       "                                                organism_name  \\\n",
       "assembly_accession                                              \n",
       "GCF_000005825.2                     Bacillus pseudofirmus OF4   \n",
       "GCF_000005845.2     Escherichia coli str. K-12 substr. MG1655   \n",
       "GCF_000006605.1                 Corynebacterium jeikeium K411   \n",
       "GCF_000006685.1                 Chlamydia muridarum str. Nigg   \n",
       "GCF_000006725.1                       Xylella fastidiosa 9a5c   \n",
       "\n",
       "                            infraspecific_name isolate version_status  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000005825.2                     strain=OF4     NaN         latest   \n",
       "GCF_000005845.2     strain=K-12 substr. MG1655     NaN         latest   \n",
       "GCF_000006605.1       strain=K411 = NCTC 11915     NaN         latest   \n",
       "GCF_000006685.1                    strain=Nigg     NaN         latest   \n",
       "GCF_000006725.1                            NaN     NaN         latest   \n",
       "\n",
       "                     assembly_level release_type genome_rep seq_rel_date  \\\n",
       "assembly_accession                                                         \n",
       "GCF_000005825.2     Complete Genome        Major       Full   2010/12/15   \n",
       "GCF_000005845.2     Complete Genome        Major       Full   2013/09/26   \n",
       "GCF_000006605.1     Complete Genome        Major       Full   2005/06/27   \n",
       "GCF_000006685.1     Complete Genome        Major       Full   2003/03/26   \n",
       "GCF_000006725.1     Complete Genome        Major       Full   2004/06/04   \n",
       "\n",
       "                    asm_name        submitter  gbrs_paired_asm     ...      \\\n",
       "assembly_accession                                                 ...       \n",
       "GCF_000005825.2     ASM582v2              NaN  GCA_000005825.2     ...       \n",
       "GCF_000005845.2     ASM584v2  Univ. Wisconsin  GCA_000005845.2     ...       \n",
       "GCF_000006605.1     ASM660v1              NaN  GCA_000006605.1     ...       \n",
       "GCF_000006685.1     ASM668v1              NaN  GCA_000006685.1     ...       \n",
       "GCF_000006725.1     ASM672v1              NaN  GCA_000006725.1     ...       \n",
       "\n",
       "                   species_taxid subclass_name  subclass_taxid subfamily_name  \\\n",
       "assembly_accession                                                              \n",
       "GCF_000005825.2          79885.0           NaN             NaN            NaN   \n",
       "GCF_000005845.2            562.0           NaN             NaN            NaN   \n",
       "GCF_000006605.1          38289.0           NaN             NaN            NaN   \n",
       "GCF_000006685.1          83560.0           NaN             NaN            NaN   \n",
       "GCF_000006725.1           2371.0           NaN             NaN            NaN   \n",
       "\n",
       "                   subfamily_taxid  subgenus_name subgenus_taxid  \\\n",
       "assembly_accession                                                 \n",
       "GCF_000005825.2                NaN            NaN            NaN   \n",
       "GCF_000005845.2                NaN            NaN            NaN   \n",
       "GCF_000006605.1                NaN            NaN            NaN   \n",
       "GCF_000006685.1                NaN            NaN            NaN   \n",
       "GCF_000006725.1                NaN            NaN            NaN   \n",
       "\n",
       "                    suborder_name suborder_taxid  subphylum_name  \\\n",
       "assembly_accession                                                 \n",
       "GCF_000005825.2               NaN            NaN             NaN   \n",
       "GCF_000005845.2               NaN            NaN             NaN   \n",
       "GCF_000006605.1               NaN            NaN             NaN   \n",
       "GCF_000006685.1               NaN            NaN             NaN   \n",
       "GCF_000006725.1               NaN            NaN             NaN   \n",
       "\n",
       "                   subphylum_taxid  subspecies_name subspecies_taxid  \\\n",
       "assembly_accession                                                     \n",
       "GCF_000005825.2                NaN              NaN              NaN   \n",
       "GCF_000005845.2                NaN              NaN              NaN   \n",
       "GCF_000006605.1                NaN              NaN              NaN   \n",
       "GCF_000006685.1                NaN              NaN              NaN   \n",
       "GCF_000006725.1                NaN              NaN              NaN   \n",
       "\n",
       "                    superkingdom_name superkingdom_taxid  tribe_name  \\\n",
       "assembly_accession                                                     \n",
       "GCF_000005825.2              Bacteria                2.0         NaN   \n",
       "GCF_000005845.2              Bacteria                2.0         NaN   \n",
       "GCF_000006605.1              Bacteria                2.0         NaN   \n",
       "GCF_000006685.1              Bacteria                2.0         NaN   \n",
       "GCF_000006725.1              Bacteria                2.0         NaN   \n",
       "\n",
       "                   tribe_taxid  \n",
       "assembly_accession              \n",
       "GCF_000005825.2            NaN  \n",
       "GCF_000005845.2            NaN  \n",
       "GCF_000006605.1            NaN  \n",
       "GCF_000006685.1            NaN  \n",
       "GCF_000006725.1            NaN  \n",
       "\n",
       "[5 rows x 53 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblySummaryCommonSpeciesDf = pd.read_csv(str(analysisEggnogPath / 'assemblySummaryCommonSpeciesDf.csv'))\n",
    "assemblySummaryCommonSpeciesDf.set_index('assembly_accession', inplace=True)\n",
    "assemblySummaryCommonSpeciesDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb of protein fasta files: 949\n"
     ]
    }
   ],
   "source": [
    "# write list of protein fasta files\n",
    "proteinSeqFastaPath = refSeqPath / 'Protein_sequences_fasta'\n",
    "\n",
    "assemblySummaryCommonSpeciesDf['protein_fasta_filename'] = \\\n",
    "    assemblySummaryCommonSpeciesDf['compressedGenomeFilename']\\\n",
    "    .map(lambda x: re.sub(r'(GCF.+)_genomic\\.gbff.*', r'\\1_protein.faa', x))\n",
    "\n",
    "proteinFastaFilenameList = list(assemblySummaryCommonSpeciesDf['protein_fasta_filename'])\n",
    "print(\"Nb of protein fasta files:\", len(proteinFastaFilenameList))\n",
    "with open(str(proteinSeqFastaPath / 'protein_sequences_fasta_file_list_EggNOG_RefSeq_common_species.txt'), 'w') as f:\n",
    "    f.write(\"\\n\".join(proteinFastaFilenameList))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We the new definition of genomes, we have 6 more genomes in the common species list."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define taxa groups that contain at least 1 genome from common species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0            0  superkingdom       2.0             Bacteria   \n",
       "4            1        phylum    1224.0       Proteobacteria   \n",
       "5            1        phylum    1239.0           Firmicutes   \n",
       "32           3         class    1236.0  Gammaproteobacteria   \n",
       "23           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "    n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                  1582                                        955  \n",
       "4                   697                                        413  \n",
       "5                   259                                        170  \n",
       "32                  292                                        166  \n",
       "23                  226                                        126  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf = pd.read_csv(str(taxonomyPath / 'taxaDf.csv'), index_col=0)\n",
    "taxaDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2859"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(taxaDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2133"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaRefSeqFamilyLevel = taxaDf[(taxaDf['n_genomes_in_refseq'] > 0) &\n",
    "                               (taxaDf['rank_name'] != 'genus')\n",
    "                              ]\n",
    "len(taxaRefSeqFamilyLevel)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the new version, all taxa in taxaDf have at least 1 genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2859"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaRefSeqFamilyLevel = taxaDf\n",
    "len(taxaRefSeqFamilyLevel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0            0  superkingdom       2.0             Bacteria   \n",
       "4            1        phylum    1224.0       Proteobacteria   \n",
       "5            1        phylum    1239.0           Firmicutes   \n",
       "32           3         class    1236.0  Gammaproteobacteria   \n",
       "23           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "    n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                  1582                                        949  \n",
       "4                   697                                        407  \n",
       "5                   259                                        170  \n",
       "32                  292                                        162  \n",
       "23                  226                                        126  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test how many taxon groups have at least 1 genome in the RefSeq database\n",
    "def find_n_genomes_in_refseq_eggnog_common(taxonRow):\n",
    "    taxonRank = taxonRow['rank_name']\n",
    "    taxonId = taxonRow['taxon_id']\n",
    "    nGenomes = len(assemblySummaryCommonSpeciesDf[assemblySummaryCommonSpeciesDf[taxonRank + '_taxid'] == taxonId])\n",
    "    return nGenomes\n",
    "\n",
    "taxaDf['n_genomes_in_refseq_eggnog_common_species'] = taxaDf.apply(find_n_genomes_in_refseq_eggnog_common, axis=1)\n",
    "taxaDf.sort_values('n_genomes_in_refseq_eggnog_common_species', ascending=False, inplace=True)\n",
    "taxaDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxaDf.to_csv(str(taxonomyPath / 'taxaDf.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2859\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rank_index     rank_name  taxon_id      taxon_name  n_genomes_in_refseq  \\\n",
       "0           0  superkingdom       2.0        Bacteria                 1582   \n",
       "4           1        phylum    1224.0  Proteobacteria                  697   \n",
       "5           1        phylum    1239.0      Firmicutes                  259   \n",
       "\n",
       "   n_genomes_in_refseq_eggnog_common_species  \n",
       "0                                        949  \n",
       "4                                        407  \n",
       "5                                        170  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf = pd.read_csv(str(taxonomyPath / 'taxaDf.csv'), index_col=0)\n",
    "print(len(taxaDf))\n",
    "taxaDf.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2064</th>\n",
       "      <td>14</td>\n",
       "      <td>species</td>\n",
       "      <td>83455.0</td>\n",
       "      <td>Myxococcus stipitatus</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2772</th>\n",
       "      <td>14</td>\n",
       "      <td>species</td>\n",
       "      <td>1637999.0</td>\n",
       "      <td>Verrucomicrobia bacterium IMCC26134</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      rank_index rank_name   taxon_id                           taxon_name  \\\n",
       "2064          14   species    83455.0                Myxococcus stipitatus   \n",
       "2772          14   species  1637999.0  Verrucomicrobia bacterium IMCC26134   \n",
       "\n",
       "      n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "2064                    1                                          0  \n",
       "2772                    1                                          0  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf.tail(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1429\n"
     ]
    }
   ],
   "source": [
    "taxaEggnogDf = taxaDf[(taxaDf['n_genomes_in_refseq_eggnog_common_species'] > 0) &\n",
    "                      (taxaDf['rank_name'] != 'genus')\n",
    "                     ]\n",
    "taxaEggnogDf.to_csv(str(analysisEggnogPath / 'taxaEggnogDf.csv'))\n",
    "print(len(taxaEggnogDf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1429\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0            0  superkingdom       2.0             Bacteria   \n",
       "4            1        phylum    1224.0       Proteobacteria   \n",
       "5            1        phylum    1239.0           Firmicutes   \n",
       "32           3         class    1236.0  Gammaproteobacteria   \n",
       "23           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "    n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                  1582                                        949  \n",
       "4                   697                                        407  \n",
       "5                   259                                        170  \n",
       "32                  292                                        162  \n",
       "23                  226                                        126  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaEggnogDf = pd.read_csv(str(analysisEggnogPath / 'taxaEggnogDf.csv'), index_col=0)\n",
    "print(len(taxaEggnogDf))\n",
    "taxaEggnogDf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### General ID mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EggNOG mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "950\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['GCF_000005825.2',\n",
       " 'GCF_000005845.2',\n",
       " 'GCF_000006605.1',\n",
       " 'GCF_000006685.1',\n",
       " 'GCF_000006725.1',\n",
       " 'GCF_000006745.1',\n",
       " 'GCF_000006765.1',\n",
       " 'GCF_000006785.2',\n",
       " 'GCF_000006845.1',\n",
       " 'GCF_000006865.1',\n",
       " 'GCF_000006905.1',\n",
       " 'GCF_000006925.2',\n",
       " 'GCF_000006945.1',\n",
       " 'GCF_000006965.1',\n",
       " 'GCF_000006985.1',\n",
       " 'GCF_000007025.1',\n",
       " 'GCF_000007045.1',\n",
       " 'GCF_000007085.1',\n",
       " 'GCF_000007125.1',\n",
       " 'GCF_000007145.1',\n",
       " 'GCF_000007265.1',\n",
       " 'GCF_000007325.1',\n",
       " 'GCF_000007365.1',\n",
       " 'GCF_000007385.1',\n",
       " 'GCF_000007465.2',\n",
       " 'GCF_000007485.1',\n",
       " 'GCF_000007525.1',\n",
       " 'GCF_000007565.2',\n",
       " 'GCF_000007605.1',\n",
       " 'GCF_000007625.1',\n",
       " 'GCF_000007645.1',\n",
       " 'GCF_000007705.1',\n",
       " 'GCF_000007725.1',\n",
       " 'GCF_000007745.1',\n",
       " 'GCF_000007765.1',\n",
       " 'GCF_000007785.1',\n",
       " 'GCF_000007805.1',\n",
       " 'GCF_000007825.1',\n",
       " 'GCF_000007845.1',\n",
       " 'GCF_000007865.1',\n",
       " 'GCF_000007905.1',\n",
       " 'GCF_000007925.1',\n",
       " 'GCF_000007945.1',\n",
       " 'GCF_000007985.2',\n",
       " 'GCF_000008025.1',\n",
       " 'GCF_000008045.1',\n",
       " 'GCF_000008165.1',\n",
       " 'GCF_000008185.1',\n",
       " 'GCF_000008205.1',\n",
       " 'GCF_000008305.1',\n",
       " 'GCF_000008325.1',\n",
       " 'GCF_000008345.1',\n",
       " 'GCF_000008365.1',\n",
       " 'GCF_000008385.1',\n",
       " 'GCF_000008465.1',\n",
       " 'GCF_000008485.1',\n",
       " 'GCF_000008505.1',\n",
       " 'GCF_000008525.1',\n",
       " 'GCF_000008545.1',\n",
       " 'GCF_000008565.1',\n",
       " 'GCF_000008605.1',\n",
       " 'GCF_000008625.1',\n",
       " 'GCF_000008685.2',\n",
       " 'GCF_000008725.1',\n",
       " 'GCF_000008745.1',\n",
       " 'GCF_000008765.1',\n",
       " 'GCF_000008805.1',\n",
       " 'GCF_000008865.1',\n",
       " 'GCF_000008885.1',\n",
       " 'GCF_000008925.1',\n",
       " 'GCF_000008985.1',\n",
       " 'GCF_000009045.1',\n",
       " 'GCF_000009065.1',\n",
       " 'GCF_000009085.1',\n",
       " 'GCF_000009125.1',\n",
       " 'GCF_000009145.1',\n",
       " 'GCF_000009205.1',\n",
       " 'GCF_000009225.2',\n",
       " 'GCF_000009305.1',\n",
       " 'GCF_000009345.1',\n",
       " 'GCF_000009365.1',\n",
       " 'GCF_000009405.1',\n",
       " 'GCF_000009605.1',\n",
       " 'GCF_000009625.1',\n",
       " 'GCF_000009705.1',\n",
       " 'GCF_000009725.1',\n",
       " 'GCF_000009745.1',\n",
       " 'GCF_000009765.2',\n",
       " 'GCF_000009785.1',\n",
       " 'GCF_000009825.1',\n",
       " 'GCF_000009845.1',\n",
       " 'GCF_000009865.1',\n",
       " 'GCF_000009905.1',\n",
       " 'GCF_000009925.1',\n",
       " 'GCF_000009945.1',\n",
       " 'GCF_000009985.1',\n",
       " 'GCF_000010065.1',\n",
       " 'GCF_000010085.1',\n",
       " 'GCF_000010105.1',\n",
       " 'GCF_000010125.1',\n",
       " 'GCF_000010145.1',\n",
       " 'GCF_000010165.1',\n",
       " 'GCF_000010185.1',\n",
       " 'GCF_000010285.1',\n",
       " 'GCF_000010305.1',\n",
       " 'GCF_000010325.1',\n",
       " 'GCF_000010405.1',\n",
       " 'GCF_000010505.1',\n",
       " 'GCF_000010525.1',\n",
       " 'GCF_000010585.1',\n",
       " 'GCF_000010605.1',\n",
       " 'GCF_000010625.1',\n",
       " 'GCF_000010645.1',\n",
       " 'GCF_000010665.1',\n",
       " 'GCF_000010785.1',\n",
       " 'GCF_000010825.1',\n",
       " 'GCF_000010985.1',\n",
       " 'GCF_000011025.1',\n",
       " 'GCF_000011065.1',\n",
       " 'GCF_000011145.1',\n",
       " 'GCF_000011225.1',\n",
       " 'GCF_000011245.1',\n",
       " 'GCF_000011305.1',\n",
       " 'GCF_000011325.1',\n",
       " 'GCF_000011345.1',\n",
       " 'GCF_000011365.1',\n",
       " 'GCF_000011385.1',\n",
       " 'GCF_000011445.1',\n",
       " 'GCF_000011465.1',\n",
       " 'GCF_000011485.1',\n",
       " 'GCF_000011545.1',\n",
       " 'GCF_000011645.1',\n",
       " 'GCF_000011685.1',\n",
       " 'GCF_000011705.1',\n",
       " 'GCF_000011745.1',\n",
       " 'GCF_000011805.1',\n",
       " 'GCF_000011825.1',\n",
       " 'GCF_000011905.1',\n",
       " 'GCF_000011965.2',\n",
       " 'GCF_000011985.1',\n",
       " 'GCF_000012005.1',\n",
       " 'GCF_000012145.1',\n",
       " 'GCF_000012225.1',\n",
       " 'GCF_000012245.1',\n",
       " 'GCF_000012305.1',\n",
       " 'GCF_000012325.1',\n",
       " 'GCF_000012345.1',\n",
       " 'GCF_000012385.1',\n",
       " 'GCF_000012405.1',\n",
       " 'GCF_000012425.1',\n",
       " 'GCF_000012445.1',\n",
       " 'GCF_000012465.1',\n",
       " 'GCF_000012485.1',\n",
       " 'GCF_000012505.1',\n",
       " 'GCF_000012565.1',\n",
       " 'GCF_000012585.1',\n",
       " 'GCF_000012605.1',\n",
       " 'GCF_000012645.1',\n",
       " 'GCF_000012665.1',\n",
       " 'GCF_000012685.1',\n",
       " 'GCF_000012725.1',\n",
       " 'GCF_000012745.1',\n",
       " 'GCF_000012765.1',\n",
       " 'GCF_000012805.1',\n",
       " 'GCF_000012825.1',\n",
       " 'GCF_000012845.1',\n",
       " 'GCF_000012865.1',\n",
       " 'GCF_000012885.1',\n",
       " 'GCF_000012925.1',\n",
       " 'GCF_000012965.1',\n",
       " 'GCF_000012985.1',\n",
       " 'GCF_000013005.1',\n",
       " 'GCF_000013025.1',\n",
       " 'GCF_000013045.1',\n",
       " 'GCF_000013085.1',\n",
       " 'GCF_000013105.1',\n",
       " 'GCF_000013125.1',\n",
       " 'GCF_000013145.1',\n",
       " 'GCF_000013165.1',\n",
       " 'GCF_000013185.1',\n",
       " 'GCF_000013205.1',\n",
       " 'GCF_000013225.1',\n",
       " 'GCF_000013285.1',\n",
       " 'GCF_000013325.1',\n",
       " 'GCF_000013345.1',\n",
       " 'GCF_000013365.1',\n",
       " 'GCF_000013405.1',\n",
       " 'GCF_000013425.1',\n",
       " 'GCF_000013565.1',\n",
       " 'GCF_000013605.1',\n",
       " 'GCF_000013625.1',\n",
       " 'GCF_000013645.1',\n",
       " 'GCF_000013665.1',\n",
       " 'GCF_000013705.1',\n",
       " 'GCF_000013745.1',\n",
       " 'GCF_000013765.1',\n",
       " 'GCF_000013865.1',\n",
       " 'GCF_000013885.1',\n",
       " 'GCF_000013945.1',\n",
       " 'GCF_000013985.1',\n",
       " 'GCF_000014005.1',\n",
       " 'GCF_000014025.1',\n",
       " 'GCF_000014045.1',\n",
       " 'GCF_000014125.1',\n",
       " 'GCF_000014145.1',\n",
       " 'GCF_000014185.1',\n",
       " 'GCF_000014205.1',\n",
       " 'GCF_000014225.1',\n",
       " 'GCF_000014245.1',\n",
       " 'GCF_000014265.1',\n",
       " 'GCF_000014285.1',\n",
       " 'GCF_000014345.1',\n",
       " 'GCF_000014385.1',\n",
       " 'GCF_000014425.1',\n",
       " 'GCF_000014445.1',\n",
       " 'GCF_000014465.1',\n",
       " 'GCF_000014505.1',\n",
       " 'GCF_000014525.1',\n",
       " 'GCF_000014565.1',\n",
       " 'GCF_000014585.1',\n",
       " 'GCF_000014705.1',\n",
       " 'GCF_000014725.1',\n",
       " 'GCF_000014745.1',\n",
       " 'GCF_000014765.1',\n",
       " 'GCF_000014785.1',\n",
       " 'GCF_000014805.1',\n",
       " 'GCF_000014825.1',\n",
       " 'GCF_000014865.1',\n",
       " 'GCF_000014885.1',\n",
       " 'GCF_000014905.1',\n",
       " 'GCF_000014925.1',\n",
       " 'GCF_000014965.1',\n",
       " 'GCF_000015005.1',\n",
       " 'GCF_000015025.1',\n",
       " 'GCF_000015045.1',\n",
       " 'GCF_000015105.1',\n",
       " 'GCF_000015125.1',\n",
       " 'GCF_000015245.1',\n",
       " 'GCF_000015265.1',\n",
       " 'GCF_000015285.1',\n",
       " 'GCF_000015305.1',\n",
       " 'GCF_000015345.1',\n",
       " 'GCF_000015405.1',\n",
       " 'GCF_000015445.1',\n",
       " 'GCF_000015505.1',\n",
       " 'GCF_000015565.1',\n",
       " 'GCF_000015585.1',\n",
       " 'GCF_000015665.1',\n",
       " 'GCF_000015725.1',\n",
       " 'GCF_000015745.1',\n",
       " 'GCF_000015865.1',\n",
       " 'GCF_000015885.1',\n",
       " 'GCF_000016065.1',\n",
       " 'GCF_000016085.1',\n",
       " 'GCF_000016165.1',\n",
       " 'GCF_000016185.1',\n",
       " 'GCF_000016265.1',\n",
       " 'GCF_000016285.1',\n",
       " 'GCF_000016325.1',\n",
       " 'GCF_000016345.1',\n",
       " 'GCF_000016405.1',\n",
       " 'GCF_000016425.1',\n",
       " 'GCF_000016505.1',\n",
       " 'GCF_000016545.1',\n",
       " 'GCF_000016565.1',\n",
       " 'GCF_000016625.1',\n",
       " 'GCF_000016645.1',\n",
       " 'GCF_000016665.1',\n",
       " 'GCF_000016745.1',\n",
       " 'GCF_000016765.1',\n",
       " 'GCF_000016785.1',\n",
       " 'GCF_000016825.1',\n",
       " 'GCF_000016905.1',\n",
       " 'GCF_000016965.1',\n",
       " 'GCF_000016985.1',\n",
       " 'GCF_000017005.1',\n",
       " 'GCF_000017045.1',\n",
       " 'GCF_000017145.1',\n",
       " 'GCF_000017245.1',\n",
       " 'GCF_000017265.1',\n",
       " 'GCF_000017285.1',\n",
       " 'GCF_000017305.1',\n",
       " 'GCF_000017405.1',\n",
       " 'GCF_000017425.1',\n",
       " 'GCF_000017465.2',\n",
       " 'GCF_000017505.1',\n",
       " 'GCF_000017545.1',\n",
       " 'GCF_000017565.1',\n",
       " 'GCF_000017585.1',\n",
       " 'GCF_000017645.1',\n",
       " 'GCF_000017685.1',\n",
       " 'GCF_000017705.1',\n",
       " 'GCF_000017805.1',\n",
       " 'GCF_000017845.1',\n",
       " 'GCF_000017865.1',\n",
       " 'GCF_000018005.1',\n",
       " 'GCF_000018025.1',\n",
       " 'GCF_000018065.1',\n",
       " 'GCF_000018105.1',\n",
       " 'GCF_000018145.1',\n",
       " 'GCF_000018205.1',\n",
       " 'GCF_000018265.1',\n",
       " 'GCF_000018285.1',\n",
       " 'GCF_000018325.1',\n",
       " 'GCF_000018345.1',\n",
       " 'GCF_000018405.1',\n",
       " 'GCF_000018425.1',\n",
       " 'GCF_000018545.1',\n",
       " 'GCF_000018605.1',\n",
       " 'GCF_000018665.1',\n",
       " 'GCF_000018685.1',\n",
       " 'GCF_000018785.1',\n",
       " 'GCF_000018865.1',\n",
       " 'GCF_000018885.1',\n",
       " 'GCF_000018945.1',\n",
       " 'GCF_000019145.1',\n",
       " 'GCF_000019165.1',\n",
       " 'GCF_000019205.1',\n",
       " 'GCF_000019225.1',\n",
       " 'GCF_000019285.1',\n",
       " 'GCF_000019345.1',\n",
       " 'GCF_000019365.1',\n",
       " 'GCF_000019405.1',\n",
       " 'GCF_000019485.1',\n",
       " 'GCF_000019525.1',\n",
       " 'GCF_000019665.1',\n",
       " 'GCF_000019725.1',\n",
       " 'GCF_000019745.1',\n",
       " 'GCF_000019785.1',\n",
       " 'GCF_000019845.1',\n",
       " 'GCF_000019905.1',\n",
       " 'GCF_000019945.1',\n",
       " 'GCF_000019965.1',\n",
       " 'GCF_000020005.1',\n",
       " 'GCF_000020025.1',\n",
       " 'GCF_000020045.1',\n",
       " 'GCF_000020065.1',\n",
       " 'GCF_000020145.1',\n",
       " 'GCF_000020165.1',\n",
       " 'GCF_000020205.1',\n",
       " 'GCF_000020225.1',\n",
       " 'GCF_000020305.1',\n",
       " 'GCF_000020365.1',\n",
       " 'GCF_000020385.1',\n",
       " 'GCF_000020465.1',\n",
       " 'GCF_000020485.1',\n",
       " 'GCF_000020505.1',\n",
       " 'GCF_000020525.1',\n",
       " 'GCF_000020545.1',\n",
       " 'GCF_000020565.1',\n",
       " 'GCF_000020605.1',\n",
       " 'GCF_000020625.1',\n",
       " 'GCF_000020645.1',\n",
       " 'GCF_000020685.1',\n",
       " 'GCF_000020725.1',\n",
       " 'GCF_000020785.1',\n",
       " 'GCF_000020945.1',\n",
       " 'GCF_000020965.1',\n",
       " 'GCF_000020985.1',\n",
       " 'GCF_000021025.1',\n",
       " 'GCF_000021045.1',\n",
       " 'GCF_000021285.1',\n",
       " 'GCF_000021325.1',\n",
       " 'GCF_000021385.1',\n",
       " 'GCF_000021485.1',\n",
       " 'GCF_000021545.1',\n",
       " 'GCF_000021565.1',\n",
       " 'GCF_000021645.1',\n",
       " 'GCF_000021685.1',\n",
       " 'GCF_000021725.1',\n",
       " 'GCF_000021745.1',\n",
       " 'GCF_000021805.1',\n",
       " 'GCF_000021825.1',\n",
       " 'GCF_000021865.1',\n",
       " 'GCF_000021885.1',\n",
       " 'GCF_000021905.1',\n",
       " 'GCF_000021925.1',\n",
       " 'GCF_000021945.1',\n",
       " 'GCF_000021985.1',\n",
       " 'GCF_000022005.1',\n",
       " 'GCF_000022025.1',\n",
       " 'GCF_000022045.1',\n",
       " 'GCF_000022065.1',\n",
       " 'GCF_000022085.1',\n",
       " 'GCF_000022105.1',\n",
       " 'GCF_000022125.1',\n",
       " 'GCF_000022145.1',\n",
       " 'GCF_000022265.1',\n",
       " 'GCF_000022305.1',\n",
       " 'GCF_000022325.1',\n",
       " 'GCF_000022525.1',\n",
       " 'GCF_000022565.1',\n",
       " 'GCF_000022605.2',\n",
       " 'GCF_000022685.1',\n",
       " 'GCF_000022725.1',\n",
       " 'GCF_000022885.2',\n",
       " 'GCF_000022905.1',\n",
       " 'GCF_000022965.1',\n",
       " 'GCF_000023025.1',\n",
       " 'GCF_000023065.1',\n",
       " 'GCF_000023105.1',\n",
       " 'GCF_000023125.1',\n",
       " 'GCF_000023145.1',\n",
       " 'GCF_000023205.1',\n",
       " 'GCF_000023225.1',\n",
       " 'GCF_000023245.1',\n",
       " 'GCF_000023265.1',\n",
       " 'GCF_000023285.1',\n",
       " 'GCF_000023325.1',\n",
       " 'GCF_000023405.1',\n",
       " 'GCF_000023445.1',\n",
       " 'GCF_000023465.1',\n",
       " 'GCF_000023585.1',\n",
       " 'GCF_000023605.1',\n",
       " 'GCF_000023705.1',\n",
       " 'GCF_000023725.1',\n",
       " 'GCF_000023745.1',\n",
       " 'GCF_000023785.1',\n",
       " 'GCF_000023825.1',\n",
       " 'GCF_000023845.1',\n",
       " 'GCF_000023865.1',\n",
       " 'GCF_000023885.1',\n",
       " 'GCF_000023905.1',\n",
       " 'GCF_000023925.1',\n",
       " 'GCF_000024005.1',\n",
       " 'GCF_000024025.1',\n",
       " 'GCF_000024065.1',\n",
       " 'GCF_000024085.1',\n",
       " 'GCF_000024105.1',\n",
       " 'GCF_000024125.1',\n",
       " 'GCF_000024165.1',\n",
       " 'GCF_000024205.1',\n",
       " 'GCF_000024225.1',\n",
       " 'GCF_000024265.1',\n",
       " 'GCF_000024285.1',\n",
       " 'GCF_000024325.1',\n",
       " 'GCF_000024345.1',\n",
       " 'GCF_000024365.1',\n",
       " 'GCF_000024385.1',\n",
       " 'GCF_000024405.1',\n",
       " 'GCF_000024465.1',\n",
       " 'GCF_000024505.1',\n",
       " 'GCF_000024545.1',\n",
       " 'GCF_000024565.1',\n",
       " 'GCF_000024605.1',\n",
       " 'GCF_000024725.1',\n",
       " 'GCF_000024765.1',\n",
       " 'GCF_000024785.1',\n",
       " 'GCF_000024805.1',\n",
       " 'GCF_000024845.1',\n",
       " 'GCF_000024865.1',\n",
       " 'GCF_000024885.1',\n",
       " 'GCF_000024905.1',\n",
       " 'GCF_000024925.1',\n",
       " 'GCF_000024945.1',\n",
       " 'GCF_000024965.1',\n",
       " 'GCF_000024985.1',\n",
       " 'GCF_000025005.1',\n",
       " 'GCF_000025085.1',\n",
       " 'GCF_000025125.1',\n",
       " 'GCF_000025185.1',\n",
       " 'GCF_000025205.1',\n",
       " 'GCF_000025225.2',\n",
       " 'GCF_000025265.1',\n",
       " 'GCF_000025305.1',\n",
       " 'GCF_000025345.1',\n",
       " 'GCF_000025485.1',\n",
       " 'GCF_000025545.1',\n",
       " 'GCF_000025565.1',\n",
       " 'GCF_000025605.1',\n",
       " 'GCF_000025645.1',\n",
       " 'GCF_000025705.1',\n",
       " 'GCF_000025725.1',\n",
       " 'GCF_000025845.1',\n",
       " 'GCF_000025885.1',\n",
       " 'GCF_000025905.1',\n",
       " 'GCF_000025925.1',\n",
       " 'GCF_000025945.1',\n",
       " 'GCF_000025965.1',\n",
       " 'GCF_000026005.1',\n",
       " 'GCF_000026065.1',\n",
       " 'GCF_000026085.1',\n",
       " 'GCF_000026105.1',\n",
       " 'GCF_000026125.1',\n",
       " 'GCF_000026145.1',\n",
       " 'GCF_000026205.1',\n",
       " 'GCF_000026345.1',\n",
       " 'GCF_000026405.1',\n",
       " 'GCF_000026505.1',\n",
       " 'GCF_000026605.1',\n",
       " 'GCF_000026745.1',\n",
       " 'GCF_000027165.1',\n",
       " 'GCF_000027225.1',\n",
       " 'GCF_000027305.1',\n",
       " 'GCF_000027325.1',\n",
       " 'GCF_000027345.1',\n",
       " 'GCF_000043285.1',\n",
       " 'GCF_000046705.1',\n",
       " 'GCF_000055785.1',\n",
       " 'GCF_000055945.1',\n",
       " 'GCF_000056065.1',\n",
       " 'GCF_000058485.1',\n",
       " 'GCF_000060345.1',\n",
       " 'GCF_000061505.1',\n",
       " 'GCF_000062885.1',\n",
       " 'GCF_000063505.1',\n",
       " 'GCF_000063525.1',\n",
       " 'GCF_000063545.1',\n",
       " 'GCF_000063585.1',\n",
       " 'GCF_000063605.1',\n",
       " 'GCF_000064305.2',\n",
       " 'GCF_000067165.1',\n",
       " 'GCF_000067205.1',\n",
       " 'GCF_000068585.1',\n",
       " 'GCF_000069185.1',\n",
       " 'GCF_000069225.1',\n",
       " 'GCF_000069925.1',\n",
       " 'GCF_000069945.1',\n",
       " 'GCF_000069965.1',\n",
       " 'GCF_000072485.1',\n",
       " 'GCF_000085865.1',\n",
       " 'GCF_000090965.1',\n",
       " 'GCF_000091125.1',\n",
       " 'GCF_000091305.1',\n",
       " 'GCF_000091325.1',\n",
       " 'GCF_000091465.1',\n",
       " 'GCF_000091545.1',\n",
       " 'GCF_000091565.1',\n",
       " 'GCF_000091785.1',\n",
       " 'GCF_000091985.1',\n",
       " 'GCF_000092025.1',\n",
       " 'GCF_000092045.1',\n",
       " 'GCF_000092105.1',\n",
       " 'GCF_000092125.1',\n",
       " 'GCF_000092205.1',\n",
       " 'GCF_000092225.1',\n",
       " 'GCF_000092245.1',\n",
       " 'GCF_000092265.1',\n",
       " 'GCF_000092365.1',\n",
       " 'GCF_000092385.1',\n",
       " 'GCF_000092405.1',\n",
       " 'GCF_000092425.1',\n",
       " 'GCF_000092505.1',\n",
       " 'GCF_000092565.1',\n",
       " 'GCF_000092585.1',\n",
       " 'GCF_000092645.1',\n",
       " 'GCF_000092785.1',\n",
       " 'GCF_000092825.1',\n",
       " 'GCF_000092845.1',\n",
       " 'GCF_000092865.1',\n",
       " 'GCF_000092905.1',\n",
       " 'GCF_000092925.1',\n",
       " 'GCF_000092945.1',\n",
       " 'GCF_000092985.1',\n",
       " 'GCF_000093025.1',\n",
       " 'GCF_000093065.1',\n",
       " 'GCF_000093085.1',\n",
       " 'GCF_000093145.3',\n",
       " 'GCF_000093165.1',\n",
       " 'GCF_000143085.1',\n",
       " 'GCF_000143145.1',\n",
       " 'GCF_000143165.1',\n",
       " 'GCF_000143845.1',\n",
       " 'GCF_000143965.1',\n",
       " 'GCF_000143985.1',\n",
       " 'GCF_000144405.1',\n",
       " 'GCF_000144605.1',\n",
       " 'GCF_000144625.1',\n",
       " 'GCF_000144645.1',\n",
       " 'GCF_000144675.1',\n",
       " 'GCF_000144695.1',\n",
       " 'GCF_000145035.1',\n",
       " 'GCF_000145235.1',\n",
       " 'GCF_000145255.1',\n",
       " 'GCF_000145275.1',\n",
       " 'GCF_000145615.1',\n",
       " 'GCF_000146165.2',\n",
       " 'GCF_000146185.1',\n",
       " 'GCF_000146505.1',\n",
       " 'GCF_000147055.1',\n",
       " 'GCF_000147335.1',\n",
       " 'GCF_000147355.1',\n",
       " 'GCF_000147695.2',\n",
       " 'GCF_000147715.2',\n",
       " 'GCF_000147815.2',\n",
       " 'GCF_000147835.2',\n",
       " 'GCF_000148645.1',\n",
       " 'GCF_000152245.2',\n",
       " 'GCF_000152825.2',\n",
       " 'GCF_000153165.2',\n",
       " 'GCF_000153405.2',\n",
       " 'GCF_000153485.2',\n",
       " 'GCF_000154785.2',\n",
       " 'GCF_000155735.2',\n",
       " 'GCF_000157355.2',\n",
       " 'GCF_000157895.3',\n",
       " 'GCF_000158275.2',\n",
       " 'GCF_000159155.2',\n",
       " 'GCF_000162235.2',\n",
       " 'GCF_000163895.2',\n",
       " 'GCF_000164675.2',\n",
       " 'GCF_000164695.2',\n",
       " 'GCF_000164865.1',\n",
       " 'GCF_000164985.3',\n",
       " 'GCF_000165465.1',\n",
       " 'GCF_000165485.1',\n",
       " 'GCF_000165505.1',\n",
       " 'GCF_000165715.2',\n",
       " 'GCF_000165905.1',\n",
       " 'GCF_000166055.1',\n",
       " 'GCF_000166135.1',\n",
       " 'GCF_000166295.1',\n",
       " 'GCF_000166395.1',\n",
       " 'GCF_000166415.1',\n",
       " 'GCF_000174395.2',\n",
       " 'GCF_000175095.2',\n",
       " 'GCF_000175115.2',\n",
       " 'GCF_000175215.2',\n",
       " 'GCF_000175255.2',\n",
       " 'GCF_000175575.2',\n",
       " 'GCF_000176115.2',\n",
       " 'GCF_000176855.2',\n",
       " 'GCF_000176915.2',\n",
       " 'GCF_000177235.2',\n",
       " 'GCF_000177535.2',\n",
       " 'GCF_000177615.2',\n",
       " 'GCF_000177635.2',\n",
       " 'GCF_000178115.2',\n",
       " 'GCF_000178835.2',\n",
       " 'GCF_000178875.2',\n",
       " 'GCF_000178955.2',\n",
       " 'GCF_000178975.2',\n",
       " 'GCF_000179035.2',\n",
       " 'GCF_000179395.2',\n",
       " 'GCF_000179635.2',\n",
       " 'GCF_000179915.2',\n",
       " 'GCF_000180175.2',\n",
       " 'GCF_000182745.2',\n",
       " 'GCF_000183135.1',\n",
       " 'GCF_000183345.1',\n",
       " 'GCF_000183385.1',\n",
       " 'GCF_000183405.1',\n",
       " 'GCF_000183425.1',\n",
       " 'GCF_000183665.1',\n",
       " 'GCF_000183725.1',\n",
       " 'GCF_000183745.1',\n",
       " 'GCF_000184345.1',\n",
       " 'GCF_000184435.1',\n",
       " 'GCF_000184685.1',\n",
       " 'GCF_000184705.1',\n",
       " 'GCF_000184745.1',\n",
       " 'GCF_000185805.1',\n",
       " 'GCF_000185885.1',\n",
       " 'GCF_000185905.1',\n",
       " 'GCF_000185965.1',\n",
       " 'GCF_000185985.2',\n",
       " 'GCF_000186245.1',\n",
       " 'GCF_000186265.1',\n",
       " 'GCF_000186345.1',\n",
       " 'GCF_000186385.1',\n",
       " 'GCF_000186885.1',\n",
       " 'GCF_000186985.1',\n",
       " 'GCF_000187005.1',\n",
       " 'GCF_000189295.2',\n",
       " 'GCF_000189415.1',\n",
       " 'GCF_000189535.1',\n",
       " 'GCF_000189775.2',\n",
       " 'GCF_000190435.1',\n",
       " 'GCF_000190535.1',\n",
       " 'GCF_000190555.1',\n",
       " 'GCF_000190575.1',\n",
       " 'GCF_000190595.1',\n",
       " 'GCF_000190635.1',\n",
       " 'GCF_000190735.1',\n",
       " 'GCF_000191045.1',\n",
       " 'GCF_000191145.1',\n",
       " 'GCF_000191545.1',\n",
       " 'GCF_000192745.1',\n",
       " 'GCF_000192845.1',\n",
       " 'GCF_000192865.1',\n",
       " 'GCF_000193395.1',\n",
       " 'GCF_000194135.1',\n",
       " 'GCF_000194605.1',\n",
       " 'GCF_000195085.1',\n",
       " 'GCF_000195275.1',\n",
       " 'GCF_000195295.1',\n",
       " 'GCF_000195315.1',\n",
       " 'GCF_000195335.1',\n",
       " 'GCF_000195575.1',\n",
       " 'GCF_000195715.1',\n",
       " 'GCF_000195735.1',\n",
       " 'GCF_000195755.1',\n",
       " 'GCF_000195775.1',\n",
       " 'GCF_000195835.1',\n",
       " 'GCF_000195855.1',\n",
       " 'GCF_000195875.1',\n",
       " 'GCF_000195955.2',\n",
       " 'GCF_000195975.1',\n",
       " 'GCF_000195995.1',\n",
       " 'GCF_000196015.1',\n",
       " 'GCF_000196035.1',\n",
       " 'GCF_000196095.1',\n",
       " 'GCF_000196115.1',\n",
       " 'GCF_000196135.1',\n",
       " 'GCF_000196175.1',\n",
       " 'GCF_000196235.1',\n",
       " 'GCF_000196255.1',\n",
       " 'GCF_000196275.1',\n",
       " 'GCF_000196295.1',\n",
       " 'GCF_000196315.1',\n",
       " 'GCF_000196355.1',\n",
       " 'GCF_000196395.1',\n",
       " 'GCF_000196435.1',\n",
       " 'GCF_000196455.1',\n",
       " 'GCF_000196495.1',\n",
       " 'GCF_000196515.1',\n",
       " 'GCF_000196535.1',\n",
       " 'GCF_000196615.1',\n",
       " 'GCF_000196675.1',\n",
       " 'GCF_000196695.1',\n",
       " 'GCF_000196735.1',\n",
       " 'GCF_000196815.1',\n",
       " 'GCF_000196835.1',\n",
       " 'GCF_000196855.1',\n",
       " 'GCF_000196875.1',\n",
       " 'GCF_000197735.1',\n",
       " 'GCF_000198515.1',\n",
       " 'GCF_000198775.1',\n",
       " 'GCF_000199675.1',\n",
       " 'GCF_000200595.1',\n",
       " 'GCF_000202635.1',\n",
       " 'GCF_000202835.1',\n",
       " 'GCF_000203835.1',\n",
       " 'GCF_000203855.3',\n",
       " 'GCF_000203895.1',\n",
       " 'GCF_000204135.1',\n",
       " 'GCF_000204155.1',\n",
       " 'GCF_000204255.1',\n",
       " 'GCF_000204565.1',\n",
       " 'GCF_000204645.1',\n",
       " 'GCF_000208385.1',\n",
       " 'GCF_000208405.1',\n",
       " 'GCF_000209655.1',\n",
       " 'GCF_000209675.1',\n",
       " 'GCF_000210915.2',\n",
       " 'GCF_000211855.2',\n",
       " 'GCF_000212355.1',\n",
       " 'GCF_000212375.1',\n",
       " 'GCF_000212395.1',\n",
       " 'GCF_000212415.1',\n",
       " 'GCF_000212675.2',\n",
       " 'GCF_000212695.1',\n",
       " 'GCF_000212735.1',\n",
       " 'GCF_000213255.1',\n",
       " 'GCF_000213805.1',\n",
       " 'GCF_000213825.1',\n",
       " 'GCF_000214095.2',\n",
       " 'GCF_000214155.1',\n",
       " 'GCF_000214175.1',\n",
       " 'GCF_000214215.1',\n",
       " 'GCF_000214355.1',\n",
       " 'GCF_000214375.1',\n",
       " 'GCF_000214435.1',\n",
       " 'GCF_000214665.1',\n",
       " 'GCF_000214785.1',\n",
       " 'GCF_000214825.1',\n",
       " 'GCF_000215105.1',\n",
       " 'GCF_000215705.1',\n",
       " 'GCF_000215745.1',\n",
       " 'GCF_000215975.1',\n",
       " 'GCF_000217635.1',\n",
       " 'GCF_000217675.1',\n",
       " 'GCF_000217795.1',\n",
       " 'GCF_000217815.1',\n",
       " 'GCF_000218545.1',\n",
       " 'GCF_000218565.1',\n",
       " 'GCF_000218625.1',\n",
       " 'GCF_000218875.1',\n",
       " 'GCF_000218895.1',\n",
       " 'GCF_000219045.1',\n",
       " 'GCF_000219105.1',\n",
       " 'GCF_000219175.1',\n",
       " 'GCF_000219355.1',\n",
       " 'GCF_000219535.2',\n",
       " 'GCF_000219585.1',\n",
       " 'GCF_000219605.1',\n",
       " 'GCF_000219725.1',\n",
       " 'GCF_000219805.1',\n",
       " 'GCF_000220625.1',\n",
       " 'GCF_000222305.1',\n",
       " 'GCF_000222485.1',\n",
       " 'GCF_000223375.1',\n",
       " 'GCF_000224005.2',\n",
       " 'GCF_000224085.1',\n",
       " 'GCF_000224105.1',\n",
       " 'GCF_000224985.1',\n",
       " 'GCF_000225325.1',\n",
       " 'GCF_000225345.1',\n",
       " 'GCF_000225445.1',\n",
       " 'GCF_000225465.1',\n",
       " 'GCF_000226295.1',\n",
       " 'GCF_000226315.1',\n",
       " 'GCF_000226565.1',\n",
       " 'GCF_000226625.1',\n",
       " 'GCF_000230275.1',\n",
       " 'GCF_000230555.1',\n",
       " 'GCF_000230895.2',\n",
       " 'GCF_000233435.1',\n",
       " 'GCF_000233775.1',\n",
       " 'GCF_000233915.3',\n",
       " 'GCF_000235605.1',\n",
       " 'GCF_000236405.1',\n",
       " 'GCF_000236585.1',\n",
       " 'GCF_000236665.1',\n",
       " 'GCF_000236685.1',\n",
       " 'GCF_000236705.1',\n",
       " 'GCF_000237065.1',\n",
       " 'GCF_000237085.1',\n",
       " 'GCF_000237145.1',\n",
       " 'GCF_000237205.1',\n",
       " 'GCF_000237995.1',\n",
       " 'GCF_000238215.1',\n",
       " 'GCF_000238995.1',\n",
       " 'GCF_000240075.2',\n",
       " 'GCF_000240165.1',\n",
       " 'GCF_000240185.1',\n",
       " 'GCF_000242335.1',\n",
       " 'GCF_000243075.1',\n",
       " 'GCF_000245355.1',\n",
       " 'GCF_000246855.1',\n",
       " 'GCF_000247565.1',\n",
       " 'GCF_000247605.1',\n",
       " 'GCF_000247715.1',\n",
       " 'GCF_000250655.1',\n",
       " 'GCF_000252855.1',\n",
       " 'GCF_000253015.1',\n",
       " 'GCF_000253035.1',\n",
       " 'GCF_000253275.1',\n",
       " 'GCF_000253295.1',\n",
       " 'GCF_000255135.1',\n",
       " 'GCF_000259175.1',\n",
       " 'GCF_000261045.2',\n",
       " 'GCF_000269945.1',\n",
       " 'GCF_000269985.1',\n",
       " 'GCF_000270085.1',\n",
       " 'GCF_000270145.1',\n",
       " 'GCF_000270205.1',\n",
       " 'GCF_000270245.1',\n",
       " 'GCF_000270285.1',\n",
       " 'GCF_000270305.1',\n",
       " 'GCF_000276685.1',\n",
       " 'GCF_000279165.1',\n",
       " 'GCF_000283515.1',\n",
       " 'GCF_000283555.1',\n",
       " 'GCF_000283575.1',\n",
       " 'GCF_000283595.1',\n",
       " 'GCF_000283615.1',\n",
       " 'GCF_000283655.1',\n",
       " 'GCF_000283875.1',\n",
       " 'GCF_000284035.1',\n",
       " 'GCF_000284235.1',\n",
       " 'GCF_000284375.1',\n",
       " 'GCF_000284615.1',\n",
       " 'GCF_000294715.1',\n",
       " 'GCF_000298115.2',\n",
       " 'GCF_000299455.1',\n",
       " 'GCF_000300045.1',\n",
       " 'GCF_000304735.1',\n",
       " 'GCF_000311765.1',\n",
       " 'GCF_000313635.1',\n",
       " 'GCF_000317855.1',\n",
       " 'GCF_000317935.1',\n",
       " 'GCF_000318015.1',\n",
       " 'GCF_000325665.1',\n",
       " 'GCF_000328765.2',\n",
       " 'GCF_000332735.1',\n",
       " 'GCF_000346595.1',\n",
       " 'GCF_000349975.1',\n",
       " 'GCF_000354175.2',\n",
       " 'GCF_000359525.1',\n",
       " 'GCF_000376645.1',\n",
       " 'GCF_000385925.1',\n",
       " 'GCF_000397205.1',\n",
       " 'GCF_000422165.1',\n",
       " 'GCF_000463505.1',\n",
       " 'GCF_000470775.1',\n",
       " 'GCF_000495505.1',\n",
       " 'GCF_000497525.1',\n",
       " 'GCF_000517605.1',\n",
       " 'GCF_000521565.1',\n",
       " 'GCF_000525675.1',\n",
       " 'GCF_000590475.1',\n",
       " 'GCF_000612075.2',\n",
       " 'GCF_000724775.3',\n",
       " 'GCF_000739475.1',\n",
       " 'GCF_000746645.1',\n",
       " 'GCF_000754265.1',\n",
       " 'GCF_000754275.1',\n",
       " 'GCF_000763535.1',\n",
       " 'GCF_000772105.1',\n",
       " 'GCF_000785515.1',\n",
       " 'GCF_000800295.1',\n",
       " 'GCF_000818035.1',\n",
       " 'GCF_000827005.1',\n",
       " 'GCF_000830885.1',\n",
       " 'GCF_000832605.1',\n",
       " 'GCF_000832905.1',\n",
       " 'GCF_000832985.1',\n",
       " 'GCF_000940845.1',\n",
       " 'GCF_000959725.1',\n",
       " 'GCF_000969765.1',\n",
       " 'GCF_000971565.1',\n",
       " 'GCF_000972865.1',\n",
       " 'GCF_000982825.1',\n",
       " 'GCF_001022195.1',\n",
       " 'GCF_001025155.1',\n",
       " 'GCF_001025175.1',\n",
       " 'GCF_001040945.1',\n",
       " 'GCF_001042595.1',\n",
       " 'GCF_001042675.1',\n",
       " 'GCF_001042695.1',\n",
       " 'GCF_001182745.1',\n",
       " 'GCF_001190745.1',\n",
       " 'GCF_001262035.1',\n",
       " 'GCF_001269425.1',\n",
       " 'GCF_001278075.1',\n",
       " 'GCF_001280225.1',\n",
       " 'GCF_001295365.1',\n",
       " 'GCF_001298465.1',\n",
       " 'GCF_001302565.1',\n",
       " 'GCF_001314975.1',\n",
       " 'GCF_001314995.1',\n",
       " 'GCF_001315015.1',\n",
       " 'GCF_001399775.1',\n",
       " 'GCF_001431145.1',\n",
       " 'GCF_001431805.1',\n",
       " 'GCF_001456065.2',\n",
       " 'GCF_001457455.1',\n",
       " 'GCF_001457475.1',\n",
       " 'GCF_001458475.1',\n",
       " 'GCF_001483845.1',\n",
       " 'GCF_001484935.1',\n",
       " 'GCF_001514435.1',\n",
       " 'GCF_001543175.1',\n",
       " 'GCF_001543285.1',\n",
       " 'GCF_001547775.1',\n",
       " 'GCF_001551895.1',\n",
       " 'GCF_001558255.1',\n",
       " 'GCF_001593245.1',\n",
       " 'GCF_001606005.1']"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggnogMapperAnnotationResultPath = analysisEggnogPath / 'Eggnog_mapper_results' / 'HMM_cluster' / 'results'\n",
    "\n",
    "with (eggnogMapperAnnotationResultPath / 'list_completed_annotations_genomes.txt').open('r') as f:\n",
    "    listCompletedGenomes = f.readlines()\n",
    "listCompletedGenomes = [re.search(r'(GCF_\\d+(\\.\\d)?)_*.*', f.strip('\\n')).group(1) for f in listCompletedGenomes]\n",
    "print(len(listCompletedGenomes))\n",
    "listCompletedGenomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>release_type</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>...</th>\n",
       "      <th>subclass_name</th>\n",
       "      <th>subclass_taxid</th>\n",
       "      <th>subfamily_name</th>\n",
       "      <th>subfamily_taxid</th>\n",
       "      <th>subgenus_name</th>\n",
       "      <th>subgenus_taxid</th>\n",
       "      <th>suborder_name</th>\n",
       "      <th>suborder_taxid</th>\n",
       "      <th>subphylum_name</th>\n",
       "      <th>subphylum_taxid</th>\n",
       "      <th>subspecies_name</th>\n",
       "      <th>subspecies_taxid</th>\n",
       "      <th>superkingdom_name</th>\n",
       "      <th>superkingdom_taxid</th>\n",
       "      <th>tribe_name</th>\n",
       "      <th>tribe_taxid</th>\n",
       "      <th>protein_fasta_filename</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Unnamed: 0, bioproject, biosample, wgs_master, refseq_category, taxid, organism_name, infraspecific_name, isolate, version_status, assembly_level, release_type, genome_rep, seq_rel_date, asm_name, submitter, gbrs_paired_asm, paired_asm_comp, ftp_path, excluded_from_refseq, compressedGenomeFilename, class_name, class_taxid, family_name, family_taxid, genus_name, genus_taxid, order_name, order_taxid, phylum_name, phylum_taxid, species group_name, species group_taxid, species subgroup_name, species subgroup_taxid, species_name, species_taxid, subclass_name, subclass_taxid, subfamily_name, subfamily_taxid, subgenus_name, subgenus_taxid, suborder_name, suborder_taxid, subphylum_name, subphylum_taxid, subspecies_name, subspecies_taxid, superkingdom_name, superkingdom_taxid, tribe_name, tribe_taxid, protein_fasta_filename]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 54 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblySummaryIncompleteDf = assemblySummaryCommonSpeciesDf[assemblySummaryCommonSpeciesDf.index.map(lambda x: x not in listCompletedGenomes)]\n",
    "assemblySummaryIncompleteDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb of protein fasta files: 0\n"
     ]
    }
   ],
   "source": [
    "proteinFastaFilenameList = list(assemblySummaryIncompleteDf['protein_fasta_filename'])\n",
    "print(\"Nb of protein fasta files:\", len(proteinFastaFilenameList))\n",
    "with open(str(proteinSeqFastaPath / 'protein_sequences_fasta_file_list_EggNOG_RefSeq_common_species_incomplete.txt'), 'w') as f:\n",
    "    f.write(\"\\n\".join(proteinFastaFilenameList))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define taxa groups that contain at least 1 genome from common species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0            0  superkingdom       2.0             Bacteria   \n",
       "4            1        phylum    1224.0       Proteobacteria   \n",
       "5            1        phylum    1239.0           Firmicutes   \n",
       "32           3         class    1236.0  Gammaproteobacteria   \n",
       "23           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "    n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                  1582                                        949  \n",
       "4                   697                                        407  \n",
       "5                   259                                        170  \n",
       "32                  292                                        162  \n",
       "23                  226                                        126  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf = pd.read_csv(str(taxonomyPath / 'taxaDf.csv'), index_col=0)\n",
    "taxaDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2859"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(taxaDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2133"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaRefSeqFamilyLevel = taxaDf[(taxaDf['n_genomes_in_refseq'] > 0) &\n",
    "                               (taxaDf['rank_name'] != 'genus')\n",
    "                              ]\n",
    "len(taxaRefSeqFamilyLevel)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the new version, all taxa in taxaDf have at least 1 genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2859"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaRefSeqFamilyLevel = taxaDf\n",
    "len(taxaRefSeqFamilyLevel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0            0  superkingdom       2.0             Bacteria   \n",
       "4            1        phylum    1224.0       Proteobacteria   \n",
       "5            1        phylum    1239.0           Firmicutes   \n",
       "32           3         class    1236.0  Gammaproteobacteria   \n",
       "23           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "    n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                  1582                                        949  \n",
       "4                   697                                        407  \n",
       "5                   259                                        170  \n",
       "32                  292                                        162  \n",
       "23                  226                                        126  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test how many taxon groups have at least 1 genome in the RefSeq database\n",
    "def find_n_genomes_in_refseq_eggnog_common(taxonRow):\n",
    "    taxonRank = taxonRow['rank_name']\n",
    "    taxonId = taxonRow['taxon_id']\n",
    "    nGenomes = len(assemblySummaryCommonSpeciesDf[assemblySummaryCommonSpeciesDf[taxonRank + '_taxid'] == taxonId])\n",
    "    return nGenomes\n",
    "\n",
    "taxaDf['n_genomes_in_refseq_eggnog_common_species'] = taxaDf.apply(find_n_genomes_in_refseq_eggnog_common, axis=1)\n",
    "taxaDf.sort_values('n_genomes_in_refseq_eggnog_common_species', ascending=False, inplace=True)\n",
    "taxaDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxaDf.to_csv(str(taxonomyPath / 'taxaDf.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2859\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rank_index     rank_name  taxon_id      taxon_name  n_genomes_in_refseq  \\\n",
       "0           0  superkingdom       2.0        Bacteria                 1582   \n",
       "4           1        phylum    1224.0  Proteobacteria                  697   \n",
       "5           1        phylum    1239.0      Firmicutes                  259   \n",
       "\n",
       "   n_genomes_in_refseq_eggnog_common_species  \n",
       "0                                        949  \n",
       "4                                        407  \n",
       "5                                        170  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf = pd.read_csv(str(taxonomyPath / 'taxaDf.csv'), index_col=0)\n",
    "print(len(taxaDf))\n",
    "taxaDf.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import annotations results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From [eggmapper documentation](https://github.com/jhcepas/eggnog-mapper/wiki/Results-Interpretation):\n",
    "\n",
    "<code>[project_name].emapper.annotations</code> file\n",
    "This file provides final annotations of each query. Tab-delimited columns in the file are:\n",
    "<ol>\n",
    " \t<li><code>query_name</code>: query sequence name</li>\n",
    " \t<li><code>seed_eggNOG_ortholog</code>: best protein match in eggNOG</li>\n",
    " \t<li><code>seed_ortholog_evalue</code>: best protein match (e-value)</li>\n",
    " \t<li><code>seed_ortholog_score</code>: best protein match (bit-score)</li>\n",
    " \t<li><code>predicted_gene_name</code>: Predicted gene name for query sequences</li>\n",
    " \t<li><code>GO_terms</code>: Comma delimited list of predicted Gene Ontology terms</li>\n",
    " \t<li><code>KEGG_pathways</code>: Comma delimited list of predicted KEGG pathways</li>\n",
    " \t<li><code>Annotation_tax_scope</code>: The taxonomic scope used to annotate this query sequence</li>\n",
    " \t<li><code>Matching_OGs</code>: Comma delimited list of matching eggNOG Orthologous Groups</li>\n",
    " \t<li><code>best_OG|evalue|score</code>: Best matching Orthologous Groups (only in HMM mode)</li>\n",
    " \t<li><code>COG functional categories</code>: COG functional category inferred from best matching OG</li>\n",
    " \t<li><code>eggNOG_HMM_model_annotation</code>: eggNOG functional description inferred from best matching OG</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/users/lserrano/mweber/Analysis_data/C-terminal/Eggnog')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysisEggnogPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### 1 / 627\n",
      "eggmapper annotations imported for genome GCF_000087965.2 nb of annotations: 2549\n",
      "### 2 / 627\n",
      "eggmapper annotations imported for genome GCF_000146065.2 nb of annotations: 1275\n",
      "### 3 / 627\n",
      "eggmapper annotations imported for genome GCF_000172635.2 nb of annotations: 3483\n",
      "### 4 / 627\n",
      "eggmapper annotations imported for genome GCF_000215085.1 nb of annotations: 3190\n",
      "### 5 / 627\n",
      "eggmapper annotations imported for genome GCF_000227665.2 nb of annotations: 1836\n",
      "### 6 / 627\n",
      "eggmapper annotations imported for genome GCF_000227685.2 nb of annotations: 2925\n",
      "### 7 / 627\n",
      "eggmapper annotations imported for genome GCF_000227705.2 nb of annotations: 3259\n",
      "### 8 / 627\n",
      "eggmapper annotations imported for genome GCF_000227745.2 nb of annotations: 3561\n",
      "### 9 / 627\n",
      "eggmapper annotations imported for genome GCF_000230695.2 nb of annotations: 3196\n",
      "### 10 / 627\n",
      "eggmapper annotations imported for genome GCF_000231385.2 nb of annotations: 3828\n",
      "### 11 / 627\n",
      "eggmapper annotations imported for genome GCF_000231405.2 nb of annotations: 2611\n",
      "### 12 / 627\n",
      "eggmapper annotations imported for genome GCF_000233715.2 nb of annotations: 3690\n",
      "### 13 / 627\n",
      "eggmapper annotations imported for genome GCF_000235405.2 nb of annotations: 1781\n",
      "### 14 / 627\n",
      "eggmapper annotations imported for genome GCF_000242255.2 nb of annotations: 2695\n",
      "### 15 / 627\n",
      "eggmapper annotations imported for genome GCF_000242455.2 nb of annotations: 5113\n",
      "### 16 / 627\n",
      "eggmapper annotations imported for genome GCF_000242595.2 nb of annotations: 2258\n",
      "### 17 / 627\n",
      "eggmapper annotations imported for genome GCF_000242635.2 nb of annotations: 3671\n",
      "### 18 / 627\n",
      "eggmapper annotations imported for genome GCF_000242895.2 nb of annotations: 3847\n",
      "### 19 / 627\n",
      "eggmapper annotations imported for genome GCF_000242935.2 nb of annotations: 4023\n",
      "### 20 / 627\n",
      "eggmapper annotations imported for genome GCF_000243115.2 nb of annotations: 3324\n",
      "### 21 / 627\n",
      "eggmapper annotations imported for genome GCF_000243135.2 nb of annotations: 2960\n",
      "### 22 / 627\n",
      "eggmapper annotations imported for genome GCF_000243155.2 nb of annotations: 3544\n",
      "### 23 / 627\n",
      "eggmapper annotations imported for genome GCF_000248095.2 nb of annotations: 1696\n",
      "### 24 / 627\n",
      "eggmapper annotations imported for genome GCF_000250635.1 nb of annotations: 2338\n",
      "### 25 / 627\n",
      "eggmapper annotations imported for genome GCF_000250675.2 nb of annotations: 7238\n",
      "### 26 / 627\n",
      "eggmapper annotations imported for genome GCF_000252445.1 nb of annotations: 2348\n",
      "### 27 / 627\n",
      "eggmapper annotations imported for genome GCF_000255115.2 nb of annotations: 3683\n",
      "### 28 / 627\n",
      "eggmapper annotations imported for genome GCF_000255295.1 nb of annotations: 6713\n",
      "### 29 / 627\n",
      "eggmapper annotations imported for genome GCF_000258405.1 nb of annotations: 2504\n",
      "### 30 / 627\n",
      "eggmapper annotations imported for genome GCF_000259255.1 nb of annotations: 1314\n",
      "### 31 / 627\n",
      "eggmapper annotations imported for genome GCF_000259275.1 nb of annotations: 1266\n",
      "### 32 / 627\n",
      "eggmapper annotations imported for genome GCF_000260965.1 nb of annotations: 2268\n",
      "### 33 / 627\n",
      "eggmapper annotations imported for genome GCF_000260985.3 nb of annotations: 2618\n",
      "### 34 / 627\n",
      "eggmapper annotations imported for genome GCF_000261025.1 nb of annotations: 366\n",
      "### 35 / 627\n",
      "eggmapper annotations imported for genome GCF_000262305.1 nb of annotations: 3569\n",
      "### 36 / 627\n",
      "eggmapper annotations imported for genome GCF_000262715.1 nb of annotations: 551\n",
      "### 37 / 627\n",
      "eggmapper annotations imported for genome GCF_000263195.1 nb of annotations: 3568\n",
      "### 38 / 627\n",
      "eggmapper annotations imported for genome GCF_000264455.2 nb of annotations: 3117\n",
      "### 39 / 627\n",
      "eggmapper annotations imported for genome GCF_000265295.1 nb of annotations: 2276\n",
      "### 40 / 627\n",
      "eggmapper annotations imported for genome GCF_000265365.1 nb of annotations: 2437\n",
      "### 41 / 627\n",
      "eggmapper annotations imported for genome GCF_000265385.1 nb of annotations: 2830\n",
      "### 42 / 627\n",
      "eggmapper annotations imported for genome GCF_000265405.1 nb of annotations: 3052\n",
      "### 43 / 627\n",
      "eggmapper annotations imported for genome GCF_000265425.1 nb of annotations: 3408\n",
      "### 44 / 627\n",
      "eggmapper annotations imported for genome GCF_000265465.1 nb of annotations: 1747\n",
      "### 45 / 627\n",
      "eggmapper annotations imported for genome GCF_000265505.1 nb of annotations: 2861\n",
      "### 46 / 627\n",
      "eggmapper annotations imported for genome GCF_000266885.1 nb of annotations: 2597\n",
      "### 47 / 627\n",
      "eggmapper annotations imported for genome GCF_000266905.1 nb of annotations: 4596\n",
      "### 48 / 627\n",
      "eggmapper annotations imported for genome GCF_000266925.1 nb of annotations: 1906\n",
      "### 49 / 627\n",
      "eggmapper annotations imported for genome GCF_000266945.1 nb of annotations: 3911\n",
      "### 50 / 627\n",
      "eggmapper annotations imported for genome GCF_000271665.2 nb of annotations: 4084\n",
      "### 51 / 627\n",
      "eggmapper annotations imported for genome GCF_000277125.1 nb of annotations: 4450\n",
      "### 52 / 627\n",
      "eggmapper annotations imported for genome GCF_000277715.1 nb of annotations: 2501\n",
      "### 53 / 627\n",
      "eggmapper annotations imported for genome GCF_000277795.1 nb of annotations: 256\n",
      "### 54 / 627\n",
      "eggmapper annotations imported for genome GCF_000279145.1 nb of annotations: 2368\n",
      "### 55 / 627\n",
      "eggmapper annotations imported for genome GCF_000281175.1 nb of annotations: 3390\n",
      "### 56 / 627\n",
      "eggmapper annotations imported for genome GCF_000284015.1 nb of annotations: 3879\n",
      "### 57 / 627\n",
      "eggmapper annotations imported for genome GCF_000284075.1 nb of annotations: 855\n",
      "### 58 / 627\n",
      "eggmapper annotations imported for genome GCF_000284095.1 nb of annotations: 2333\n",
      "### 59 / 627\n",
      "eggmapper annotations imported for genome GCF_000284115.1 nb of annotations: 2149\n",
      "### 60 / 627\n",
      "eggmapper annotations imported for genome GCF_000284155.1 nb of annotations: 864\n",
      "### 61 / 627\n",
      "eggmapper annotations imported for genome GCF_000284175.1 nb of annotations: 898\n",
      "### 62 / 627\n",
      "eggmapper annotations imported for genome GCF_000284255.1 nb of annotations: 4178\n",
      "### 63 / 627\n",
      "eggmapper annotations imported for genome GCF_000284295.1 nb of annotations: 6709\n",
      "### 64 / 627\n",
      "eggmapper annotations imported for genome GCF_000284315.1 nb of annotations: 1760\n",
      "### 65 / 627\n",
      "eggmapper annotations imported for genome GCF_000284335.1 nb of annotations: 1213\n",
      "### 66 / 627\n",
      "eggmapper annotations imported for genome GCF_000284415.1 nb of annotations: 2683\n",
      "### 67 / 627\n",
      "eggmapper annotations imported for genome GCF_000284515.1 nb of annotations: 3354\n",
      "### 68 / 627\n",
      "eggmapper annotations imported for genome GCF_000286435.2 nb of annotations: 3181\n",
      "### 69 / 627\n",
      "eggmapper annotations imported for genome GCF_000287335.1 nb of annotations: 660\n",
      "### 70 / 627\n",
      "eggmapper annotations imported for genome GCF_000287355.1 nb of annotations: 524\n",
      "### 71 / 627\n",
      "eggmapper annotations imported for genome GCF_000294515.1 nb of annotations: 4230\n",
      "### 72 / 627\n",
      "eggmapper annotations imported for genome GCF_000294775.2 nb of annotations: 4573\n",
      "### 73 / 627\n",
      "eggmapper annotations imported for genome GCF_000297055.2 nb of annotations: 2675\n",
      "### 74 / 627\n",
      "eggmapper annotations imported for genome GCF_000299095.1 nb of annotations: 1267\n",
      "### 75 / 627\n",
      "eggmapper annotations imported for genome GCF_000299115.1 nb of annotations: 1287\n",
      "### 76 / 627\n",
      "eggmapper annotations imported for genome GCF_000299235.1 nb of annotations: 1588\n",
      "### 77 / 627\n",
      "eggmapper annotations imported for genome GCF_000299335.2 nb of annotations: 3300\n",
      "### 78 / 627\n",
      "eggmapper annotations imported for genome GCF_000299355.1 nb of annotations: 2460\n",
      "### 79 / 627\n",
      "eggmapper annotations imported for genome GCF_000300005.1 nb of annotations: 3987\n",
      "### 80 / 627\n",
      "eggmapper annotations imported for genome GCF_000300075.1 nb of annotations: 226\n",
      "### 81 / 627\n",
      "eggmapper annotations imported for genome GCF_000300135.1 nb of annotations: 1409\n",
      "### 82 / 627\n",
      "eggmapper annotations imported for genome GCF_000300235.2 nb of annotations: 3650\n",
      "### 83 / 627\n",
      "eggmapper annotations imported for genome GCF_000300295.3 nb of annotations: 1931\n",
      "### 84 / 627\n",
      "eggmapper annotations imported for genome GCF_000300455.3 nb of annotations: 4260\n",
      "### 85 / 627\n",
      "eggmapper annotations imported for genome GCF_000304215.1 nb of annotations: 1466\n",
      "### 86 / 627\n",
      "eggmapper annotations imported for genome GCF_000304315.1 nb of annotations: 3028\n",
      "### 87 / 627\n",
      "eggmapper annotations imported for genome GCF_000304455.1 nb of annotations: 591\n",
      "### 88 / 627\n",
      "eggmapper annotations imported for genome GCF_000305785.2 nb of annotations: 3316\n",
      "### 89 / 627\n",
      "eggmapper annotations imported for genome GCF_000305935.1 nb of annotations: 2361\n",
      "### 90 / 627\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eggmapper annotations imported for genome GCF_000306785.1 nb of annotations: 4497\n",
      "### 91 / 627\n",
      "eggmapper annotations imported for genome GCF_000306885.1 nb of annotations: 583\n",
      "### 92 / 627\n",
      "eggmapper annotations imported for genome GCF_000307105.1 nb of annotations: 3661\n",
      "### 93 / 627\n",
      "eggmapper annotations imported for genome GCF_000307165.1 nb of annotations: 2068\n",
      "### 94 / 627\n",
      "eggmapper annotations imported for genome GCF_000309885.1 nb of annotations: 1943\n",
      "### 95 / 627\n",
      "eggmapper annotations imported for genome GCF_000312705.1 nb of annotations: 3092\n",
      "### 96 / 627\n",
      "eggmapper annotations imported for genome GCF_000313175.2 nb of annotations: 2089\n",
      "### 97 / 627\n",
      "eggmapper annotations imported for genome GCF_000316515.1 nb of annotations: 2726\n",
      "### 98 / 627\n",
      "eggmapper annotations imported for genome GCF_000316575.1 nb of annotations: 5064\n",
      "### 99 / 627\n",
      "eggmapper annotations imported for genome GCF_000316605.1 nb of annotations: 3610\n",
      "### 100 / 627\n",
      "eggmapper annotations imported for genome GCF_000316625.1 nb of annotations: 4584\n",
      "### 101 / 627\n",
      "eggmapper annotations imported for genome GCF_000316645.1 nb of annotations: 4658\n",
      "### 102 / 627\n",
      "eggmapper annotations imported for genome GCF_000316665.1 nb of annotations: 5636\n",
      "### 103 / 627\n",
      "eggmapper annotations imported for genome GCF_000316685.1 nb of annotations: 2949\n",
      "### 104 / 627\n",
      "eggmapper annotations imported for genome GCF_000317025.1 nb of annotations: 3861\n",
      "### 105 / 627\n",
      "eggmapper annotations imported for genome GCF_000317045.1 nb of annotations: 3368\n",
      "### 106 / 627\n",
      "eggmapper annotations imported for genome GCF_000317065.1 nb of annotations: 2958\n",
      "### 107 / 627\n",
      "eggmapper annotations imported for genome GCF_000317085.1 nb of annotations: 2818\n",
      "### 108 / 627\n",
      "eggmapper annotations imported for genome GCF_000317105.1 nb of annotations: 4794\n",
      "### 109 / 627\n",
      "eggmapper annotations imported for genome GCF_000317125.1 nb of annotations: 4732\n",
      "### 110 / 627\n",
      "eggmapper annotations imported for genome GCF_000317305.3 nb of annotations: 4495\n",
      "### 111 / 627\n",
      "eggmapper annotations imported for genome GCF_000317435.1 nb of annotations: 4576\n",
      "### 112 / 627\n",
      "eggmapper annotations imported for genome GCF_000317475.1 nb of annotations: 4926\n",
      "### 113 / 627\n",
      "eggmapper annotations imported for genome GCF_000317495.1 nb of annotations: 3846\n",
      "### 114 / 627\n",
      "eggmapper annotations imported for genome GCF_000317515.1 nb of annotations: 5161\n",
      "### 115 / 627\n",
      "eggmapper annotations imported for genome GCF_000317555.1 nb of annotations: 4253\n",
      "### 116 / 627\n",
      "eggmapper annotations imported for genome GCF_000317575.1 nb of annotations: 3875\n",
      "### 117 / 627\n",
      "eggmapper annotations imported for genome GCF_000317615.1 nb of annotations: 2963\n",
      "### 118 / 627\n",
      "eggmapper annotations imported for genome GCF_000317635.1 nb of annotations: 3324\n",
      "### 119 / 627\n",
      "eggmapper annotations imported for genome GCF_000317675.1 nb of annotations: 3030\n",
      "### 120 / 627\n",
      "eggmapper annotations imported for genome GCF_000317695.1 nb of annotations: 4618\n",
      "### 121 / 627\n",
      "eggmapper annotations imported for genome GCF_000317835.1 nb of annotations: 2755\n",
      "### 122 / 627\n",
      "eggmapper annotations imported for genome GCF_000317975.2 nb of annotations: 2731\n",
      "### 123 / 627\n",
      "eggmapper annotations imported for genome GCF_000319245.1 nb of annotations: 705\n",
      "### 124 / 627\n",
      "eggmapper annotations imported for genome GCF_000319385.1 nb of annotations: 754\n",
      "### 125 / 627\n",
      "eggmapper annotations imported for genome GCF_000319575.2 nb of annotations: 1565\n",
      "### 126 / 627\n",
      "eggmapper annotations imported for genome GCF_000321415.2 nb of annotations: 3095\n",
      "### 127 / 627\n",
      "eggmapper annotations imported for genome GCF_000325705.1 nb of annotations: 3857\n",
      "### 128 / 627\n",
      "eggmapper annotations imported for genome GCF_000325745.1 nb of annotations: 1122\n",
      "### 129 / 627\n",
      "eggmapper annotations imported for genome GCF_000327045.1 nb of annotations: 2987\n",
      "### 130 / 627\n",
      "eggmapper annotations imported for genome GCF_000328565.1 nb of annotations: 5310\n",
      "### 131 / 627\n",
      "eggmapper annotations imported for genome GCF_000328625.1 nb of annotations: 2126\n",
      "### 132 / 627\n",
      "eggmapper annotations imported for genome GCF_000328705.1 nb of annotations: 6908\n",
      "### 133 / 627\n",
      "eggmapper annotations imported for genome GCF_000328725.1 nb of annotations: 536\n",
      "### 134 / 627\n",
      "eggmapper annotations imported for genome GCF_000330865.1 nb of annotations: 4072\n",
      "### 135 / 627\n",
      "eggmapper annotations imported for genome GCF_000330885.1 nb of annotations: 3381\n",
      "### 136 / 627\n",
      "eggmapper annotations imported for genome GCF_000331735.1 nb of annotations: 6719\n",
      "### 137 / 627\n",
      "eggmapper annotations imported for genome GCF_000331995.1 nb of annotations: 2292\n",
      "### 138 / 627\n",
      "eggmapper annotations imported for genome GCF_000332115.1 nb of annotations: 2878\n",
      "### 139 / 627\n",
      "eggmapper annotations imported for genome GCF_000334405.1 nb of annotations: 523\n",
      "### 140 / 627\n",
      "eggmapper annotations imported for genome GCF_000340435.2 nb of annotations: 3243\n",
      "### 141 / 627\n",
      "eggmapper annotations imported for genome GCF_000340795.1 nb of annotations: 711\n",
      "### 142 / 627\n",
      "eggmapper annotations imported for genome GCF_000340825.1 nb of annotations: 707\n",
      "### 143 / 627\n",
      "eggmapper annotations imported for genome GCF_000340865.1 nb of annotations: 676\n",
      "### 144 / 627\n",
      "eggmapper annotations imported for genome GCF_000340885.1 nb of annotations: 4856\n",
      "### 145 / 627\n",
      "eggmapper annotations imported for genome GCF_000340905.1 nb of annotations: 692\n",
      "### 146 / 627\n",
      "eggmapper annotations imported for genome GCF_000341345.1 nb of annotations: 2411\n",
      "### 147 / 627\n",
      "eggmapper annotations imported for genome GCF_000341355.1 nb of annotations: 1166\n",
      "### 148 / 627\n",
      "eggmapper annotations imported for genome GCF_000341385.1 nb of annotations: 1284\n",
      "### 149 / 627\n",
      "eggmapper annotations imported for genome GCF_000341395.1 nb of annotations: 2994\n",
      "### 150 / 627\n",
      "eggmapper annotations imported for genome GCF_000344785.1 nb of annotations: 2162\n",
      "### 151 / 627\n",
      "eggmapper annotations imported for genome GCF_000344805.1 nb of annotations: 6183\n",
      "### 152 / 627\n",
      "eggmapper annotations imported for genome GCF_000347595.1 nb of annotations: 1995\n",
      "### 153 / 627\n",
      "eggmapper annotations imported for genome GCF_000347635.1 nb of annotations: 3822\n",
      "### 154 / 627\n",
      "eggmapper annotations imported for genome GCF_000347675.2 nb of annotations: 3247\n",
      "### 155 / 627\n",
      "eggmapper annotations imported for genome GCF_000347695.1 nb of annotations: 1383\n",
      "### 156 / 627\n",
      "eggmapper annotations imported for genome GCF_000348725.1 nb of annotations: 1790\n",
      "### 157 / 627\n",
      "eggmapper annotations imported for genome GCF_000348785.1 nb of annotations: 3356\n",
      "### 158 / 627\n",
      "eggmapper annotations imported for genome GCF_000348805.1 nb of annotations: 552\n",
      "### 159 / 627\n",
      "eggmapper annotations imported for genome GCF_000349845.1 nb of annotations: 4621\n",
      "### 160 / 627\n",
      "eggmapper annotations imported for genome GCF_000349945.1 nb of annotations: 4149\n",
      "### 161 / 627\n",
      "eggmapper annotations imported for genome GCF_000355675.1 nb of annotations: 3063\n",
      "### 162 / 627\n",
      "eggmapper annotations imported for genome GCF_000367205.1 nb of annotations: 1949\n",
      "### 163 / 627\n",
      "eggmapper annotations imported for genome GCF_000376545.2 nb of annotations: 2840\n",
      "### 164 / 627\n",
      "eggmapper annotations imported for genome GCF_000376585.1 nb of annotations: 854\n",
      "### 165 / 627\n",
      "eggmapper annotations imported for genome GCF_000385945.1 nb of annotations: 5921\n",
      "### 166 / 627\n",
      "eggmapper annotations imported for genome GCF_000389635.1 nb of annotations: 3713\n",
      "### 167 / 627\n",
      "eggmapper annotations imported for genome GCF_000389965.1 nb of annotations: 6641\n",
      "### 168 / 627\n",
      "eggmapper annotations imported for genome GCF_000400935.1 nb of annotations: 627\n",
      "### 169 / 627\n",
      "eggmapper annotations imported for genome GCF_000400955.1 nb of annotations: 618\n",
      "### 170 / 627\n",
      "eggmapper annotations imported for genome GCF_000412695.1 nb of annotations: 5077\n",
      "### 171 / 627\n",
      "eggmapper annotations imported for genome GCF_000418365.1 nb of annotations: 2034\n",
      "### 172 / 627\n",
      "eggmapper annotations imported for genome GCF_000422085.1 nb of annotations: 4514\n",
      "### 173 / 627\n",
      "eggmapper annotations imported for genome GCF_000439435.1 nb of annotations: 624\n",
      "### 174 / 627\n",
      "eggmapper annotations imported for genome GCF_000439455.1 nb of annotations: 594\n",
      "### 175 / 627\n",
      "eggmapper annotations imported for genome GCF_000441555.1 nb of annotations: 336\n",
      "### 176 / 627\n",
      "eggmapper annotations imported for genome GCF_000442595.1 nb of annotations: 2241\n",
      "### 177 / 627\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eggmapper annotations imported for genome GCF_000442645.1 nb of annotations: 2201\n",
      "### 178 / 627\n",
      "eggmapper annotations imported for genome GCF_000444875.1 nb of annotations: 5644\n",
      "### 179 / 627\n",
      "eggmapper annotations imported for genome GCF_000444995.1 nb of annotations: 2983\n",
      "### 180 / 627\n",
      "eggmapper annotations imported for genome GCF_000445425.4 nb of annotations: 660\n",
      "### 181 / 627\n",
      "eggmapper annotations imported for genome GCF_000447675.1 nb of annotations: 2072\n",
      "### 182 / 627\n",
      "eggmapper annotations imported for genome GCF_000455605.1 nb of annotations: 2351\n",
      "### 183 / 627\n",
      "eggmapper annotations imported for genome GCF_000463355.1 nb of annotations: 1650\n",
      "### 184 / 627\n",
      "eggmapper annotations imported for genome GCF_000468615.2 nb of annotations: 3563\n",
      "### 185 / 627\n",
      "eggmapper annotations imported for genome GCF_000471025.2 nb of annotations: 715\n",
      "### 186 / 627\n",
      "eggmapper annotations imported for genome GCF_000471965.1 nb of annotations: 516\n",
      "### 187 / 627\n",
      "eggmapper annotations imported for genome GCF_000473245.1 nb of annotations: 4174\n",
      "### 188 / 627\n",
      "eggmapper annotations imported for genome GCF_000473305.1 nb of annotations: 1980\n",
      "### 189 / 627\n",
      "eggmapper annotations imported for genome GCF_000473995.1 nb of annotations: 3539\n",
      "### 190 / 627\n",
      "eggmapper annotations imported for genome GCF_000477415.1 nb of annotations: 242\n",
      "### 191 / 627\n",
      "eggmapper annotations imported for genome GCF_000477435.1 nb of annotations: 2037\n",
      "### 192 / 627\n",
      "eggmapper annotations imported for genome GCF_000478885.1 nb of annotations: 1937\n",
      "### 193 / 627\n",
      "eggmapper annotations imported for genome GCF_000478905.1 nb of annotations: 746\n",
      "### 194 / 627\n",
      "eggmapper annotations imported for genome GCF_000484505.1 nb of annotations: 3394\n",
      "### 195 / 627\n",
      "eggmapper annotations imported for genome GCF_000484535.1 nb of annotations: 3402\n",
      "### 196 / 627\n",
      "eggmapper annotations imported for genome GCF_000485905.1 nb of annotations: 1708\n",
      "### 197 / 627\n",
      "eggmapper annotations imported for genome GCF_000493735.1 nb of annotations: 1952\n",
      "### 198 / 627\n",
      "eggmapper annotations imported for genome GCF_000494755.1 nb of annotations: 7394\n",
      "### 199 / 627\n",
      "eggmapper annotations imported for genome GCF_000495935.2 nb of annotations: 3397\n",
      "### 200 / 627\n",
      "eggmapper annotations imported for genome GCF_000496595.1 nb of annotations: 851\n",
      "### 201 / 627\n",
      "eggmapper annotations imported for genome GCF_000499665.2 nb of annotations: 853\n",
      "### 202 / 627\n",
      "eggmapper annotations imported for genome GCF_000500935.1 nb of annotations: 643\n",
      "### 203 / 627\n",
      "eggmapper annotations imported for genome GCF_000503895.1 nb of annotations: 2662\n",
      "### 204 / 627\n",
      "eggmapper annotations imported for genome GCF_000504105.1 nb of annotations: 1971\n",
      "### 205 / 627\n",
      "eggmapper annotations imported for genome GCF_000504125.1 nb of annotations: 2209\n",
      "### 206 / 627\n",
      "eggmapper annotations imported for genome GCF_000507245.1 nb of annotations: 2560\n",
      "### 207 / 627\n",
      "eggmapper annotations imported for genome GCF_000508225.1 nb of annotations: 752\n",
      "### 208 / 627\n",
      "eggmapper annotations imported for genome GCF_000508245.1 nb of annotations: 247\n",
      "### 209 / 627\n",
      "eggmapper annotations imported for genome GCF_000510265.1 nb of annotations: 747\n",
      "### 210 / 627\n",
      "eggmapper annotations imported for genome GCF_000511305.1 nb of annotations: 3965\n",
      "### 211 / 627\n",
      "eggmapper annotations imported for genome GCF_000511355.1 nb of annotations: 3536\n",
      "### 212 / 627\n",
      "eggmapper annotations imported for genome GCF_000511385.1 nb of annotations: 3371\n",
      "### 213 / 627\n",
      "eggmapper annotations imported for genome GCF_000511405.1 nb of annotations: 2833\n",
      "### 214 / 627\n",
      "eggmapper annotations imported for genome GCF_000512205.2 nb of annotations: 4629\n",
      "### 215 / 627\n",
      "eggmapper annotations imported for genome GCF_000512355.1 nb of annotations: 1707\n",
      "### 216 / 627\n",
      "eggmapper annotations imported for genome GCF_000512735.1 nb of annotations: 1487\n",
      "### 217 / 627\n",
      "eggmapper annotations imported for genome GCF_000512895.1 nb of annotations: 2301\n",
      "### 218 / 627\n",
      "eggmapper annotations imported for genome GCF_000512915.1 nb of annotations: 2154\n",
      "### 219 / 627\n",
      "eggmapper annotations imported for genome GCF_000513215.1 nb of annotations: 4423\n",
      "### 220 / 627\n",
      "eggmapper annotations imported for genome GCF_000513295.1 nb of annotations: 3350\n",
      "### 221 / 627\n",
      "eggmapper annotations imported for genome GCF_000513475.1 nb of annotations: 542\n",
      "### 222 / 627\n",
      "eggmapper annotations imported for genome GCF_000517365.1 nb of annotations: 544\n",
      "### 223 / 627\n",
      "eggmapper annotations imported for genome GCF_000517425.1 nb of annotations: 3691\n",
      "### 224 / 627\n",
      "eggmapper annotations imported for genome GCF_000517565.1 nb of annotations: 1551\n",
      "### 225 / 627\n",
      "eggmapper annotations imported for genome GCF_000520015.2 nb of annotations: 4393\n",
      "### 226 / 627\n",
      "eggmapper annotations imported for genome GCF_000521505.1 nb of annotations: 3854\n",
      "### 227 / 627\n",
      "eggmapper annotations imported for genome GCF_000523235.1 nb of annotations: 6266\n",
      "### 228 / 627\n",
      "eggmapper annotations imported for genome GCF_000524555.1 nb of annotations: 478\n",
      "### 229 / 627\n",
      "eggmapper annotations imported for genome GCF_000525635.1 nb of annotations: 7461\n",
      "### 230 / 627\n",
      "eggmapper annotations imported for genome GCF_000525655.1 nb of annotations: 1880\n",
      "### 231 / 627\n",
      "eggmapper annotations imported for genome GCF_000550765.1 nb of annotations: 2519\n",
      "### 232 / 627\n",
      "eggmapper annotations imported for genome GCF_000550785.1 nb of annotations: 1\n",
      "### 233 / 627\n",
      "eggmapper annotations imported for genome GCF_000550805.1 nb of annotations: 2208\n",
      "### 234 / 627\n",
      "eggmapper annotations imported for genome GCF_000565175.1 nb of annotations: 671\n",
      "### 235 / 627\n",
      "eggmapper annotations imported for genome GCF_000565215.1 nb of annotations: 608\n",
      "### 236 / 627\n",
      "eggmapper annotations imported for genome GCF_000568815.1 nb of annotations: 2737\n",
      "### 237 / 627\n",
      "eggmapper annotations imported for genome GCF_000576555.1 nb of annotations: 3289\n",
      "### 238 / 627\n",
      "eggmapper annotations imported for genome GCF_000577895.1 nb of annotations: 2268\n",
      "### 239 / 627\n",
      "eggmapper annotations imported for genome GCF_000583875.1 nb of annotations: 672\n",
      "### 240 / 627\n",
      "eggmapper annotations imported for genome GCF_000590555.1 nb of annotations: 1537\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000590925.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000597865.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000597885.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000599985.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000600005.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000600105.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000612055.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000612485.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000612685.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000619905.2\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000626635.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000626675.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000632805.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000632845.1\n",
      "### 241 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_000632985.1\n",
      "### 241 / 627\n",
      "eggmapper annotations imported for genome GCF_000648515.1 nb of annotations: 4522\n",
      "### 242 / 627\n",
      "eggmapper annotations imported for genome GCF_000661895.1 nb of annotations: 2204\n",
      "### 243 / 627\n",
      "eggmapper annotations imported for genome GCF_000689415.1 nb of annotations: 4991\n",
      "### 244 / 627\n",
      "eggmapper annotations imported for genome GCF_000695235.1 nb of annotations: 6524\n",
      "### 245 / 627\n",
      "eggmapper annotations imported for genome GCF_000695835.1 nb of annotations: 480\n",
      "### 246 / 627\n",
      "eggmapper annotations imported for genome GCF_000696485.1 nb of annotations: 3340\n",
      "### 247 / 627\n",
      "eggmapper annotations imported for genome GCF_000697965.2 nb of annotations: 3541\n",
      "### 248 / 627\n",
      "eggmapper annotations imported for genome GCF_000699505.1 nb of annotations: 1216\n",
      "### 249 / 627\n",
      "eggmapper annotations imported for genome GCF_000706685.1 nb of annotations: 2267\n",
      "### 250 / 627\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eggmapper annotations imported for genome GCF_000706725.1 nb of annotations: 3335\n",
      "### 251 / 627\n",
      "eggmapper annotations imported for genome GCF_000709555.1 nb of annotations: 236\n",
      "### 252 / 627\n",
      "eggmapper annotations imported for genome GCF_000723165.1 nb of annotations: 4422\n",
      "### 253 / 627\n",
      "eggmapper annotations imported for genome GCF_000723465.1 nb of annotations: 1533\n",
      "### 254 / 627\n",
      "eggmapper annotations imported for genome GCF_000723505.1 nb of annotations: 2079\n",
      "### 255 / 627\n",
      "eggmapper annotations imported for genome GCF_000724485.1 nb of annotations: 2749\n",
      "### 256 / 627\n",
      "eggmapper annotations imported for genome GCF_000724605.1 nb of annotations: 1443\n",
      "### 257 / 627\n",
      "eggmapper annotations imported for genome GCF_000724625.1 nb of annotations: 3116\n",
      "### 258 / 627\n",
      "eggmapper annotations imported for genome GCF_000725345.1 nb of annotations: 4077\n",
      "### 259 / 627\n",
      "eggmapper annotations imported for genome GCF_000725365.1 nb of annotations: 3095\n",
      "### 260 / 627\n",
      "eggmapper annotations imported for genome GCF_000725405.1 nb of annotations: 2296\n",
      "### 261 / 627\n",
      "eggmapper annotations imported for genome GCF_000730245.1 nb of annotations: 1064\n",
      "### 262 / 627\n",
      "eggmapper annotations imported for genome GCF_000730385.1 nb of annotations: 3378\n",
      "### 263 / 627\n",
      "eggmapper annotations imported for genome GCF_000731315.1 nb of annotations: 4048\n",
      "### 264 / 627\n",
      "eggmapper annotations imported for genome GCF_000732925.1 nb of annotations: 7499\n",
      "### 265 / 627\n",
      "eggmapper annotations imported for genome GCF_000732945.1 nb of annotations: 1649\n",
      "### 266 / 627\n",
      "eggmapper annotations imported for genome GCF_000734015.1 nb of annotations: 1439\n",
      "### 267 / 627\n",
      "eggmapper annotations imported for genome GCF_000734895.2 nb of annotations: 4422\n",
      "### 268 / 627\n",
      "eggmapper annotations imported for genome GCF_000736415.1 nb of annotations: 1557\n",
      "### 269 / 627\n",
      "eggmapper annotations imported for genome GCF_000737325.1 nb of annotations: 5621\n",
      "### 270 / 627\n",
      "eggmapper annotations imported for genome GCF_000737515.1 nb of annotations: 3178\n",
      "### 271 / 627\n",
      "eggmapper annotations imported for genome GCF_000737535.1 nb of annotations: 2206\n",
      "### 272 / 627\n",
      "eggmapper annotations imported for genome GCF_000737575.1 nb of annotations: 2129\n",
      "### 273 / 627\n",
      "eggmapper annotations imported for genome GCF_000737595.1 nb of annotations: 2115\n",
      "### 274 / 627\n",
      "eggmapper annotations imported for genome GCF_000737865.1 nb of annotations: 1203\n",
      "### 275 / 627\n",
      "eggmapper annotations imported for genome GCF_000738435.1 nb of annotations: 2824\n",
      "### 276 / 627\n",
      "eggmapper annotations imported for genome GCF_000739085.1 nb of annotations: 6038\n",
      "### 277 / 627\n",
      "eggmapper annotations imported for genome GCF_000739455.1 nb of annotations: 1917\n",
      "### 278 / 627\n",
      "eggmapper annotations imported for genome GCF_000742835.1 nb of annotations: 1506\n",
      "### 279 / 627\n",
      "eggmapper annotations imported for genome GCF_000743945.1 nb of annotations: 1468\n",
      "### 280 / 627\n",
      "eggmapper annotations imported for genome GCF_000746525.1 nb of annotations: 4560\n",
      "### 281 / 627\n",
      "eggmapper annotations imported for genome GCF_000746585.1 nb of annotations: 1132\n",
      "### 282 / 627\n",
      "eggmapper annotations imported for genome GCF_000747315.1 nb of annotations: 1786\n",
      "### 283 / 627\n",
      "eggmapper annotations imported for genome GCF_000747345.1 nb of annotations: 2781\n",
      "### 284 / 627\n",
      "eggmapper annotations imported for genome GCF_000747525.1 nb of annotations: 3653\n",
      "### 285 / 627\n",
      "eggmapper annotations imported for genome GCF_000750535.1 nb of annotations: 1067\n",
      "### 286 / 627\n",
      "eggmapper annotations imported for genome GCF_000755145.1 nb of annotations: 1990\n",
      "### 287 / 627\n",
      "eggmapper annotations imported for genome GCF_000755705.1 nb of annotations: 1385\n",
      "### 288 / 627\n",
      "eggmapper annotations imported for genome GCF_000756615.1 nb of annotations: 4548\n",
      "### 289 / 627\n",
      "eggmapper annotations imported for genome GCF_000757785.1 nb of annotations: 4591\n",
      "### 290 / 627\n",
      "eggmapper annotations imported for genome GCF_000757825.1 nb of annotations: 4182\n",
      "### 291 / 627\n",
      "eggmapper annotations imported for genome GCF_000757865.1 nb of annotations: 1578\n",
      "### 292 / 627\n",
      "eggmapper annotations imported for genome GCF_000758625.1 nb of annotations: 5465\n",
      "### 293 / 627\n",
      "eggmapper annotations imported for genome GCF_000758665.1 nb of annotations: 5993\n",
      "### 294 / 627\n",
      "eggmapper annotations imported for genome GCF_000758685.1 nb of annotations: 4367\n",
      "### 295 / 627\n",
      "eggmapper annotations imported for genome GCF_000758725.1 nb of annotations: 5252\n",
      "### 296 / 627\n",
      "eggmapper annotations imported for genome GCF_000759475.1 nb of annotations: 3668\n",
      "### 297 / 627\n",
      "eggmapper annotations imported for genome GCF_000759535.1 nb of annotations: 3727\n",
      "### 298 / 627\n",
      "eggmapper annotations imported for genome GCF_000761155.1 nb of annotations: 3734\n",
      "### 299 / 627\n",
      "eggmapper annotations imported for genome GCF_000763575.1 nb of annotations: 2055\n",
      "### 300 / 627\n",
      "eggmapper annotations imported for genome GCF_000764535.1 nb of annotations: 2988\n",
      "### 301 / 627\n",
      "eggmapper annotations imported for genome GCF_000764555.1 nb of annotations: 1447\n",
      "### 302 / 627\n",
      "eggmapper annotations imported for genome GCF_000767055.1 nb of annotations: 2123\n",
      "### 303 / 627\n",
      "eggmapper annotations imported for genome GCF_000767275.2 nb of annotations: 1778\n",
      "### 304 / 627\n",
      "eggmapper annotations imported for genome GCF_000767465.1 nb of annotations: 135\n",
      "### 305 / 627\n",
      "eggmapper annotations imported for genome GCF_000767615.3 nb of annotations: 4178\n",
      "### 306 / 627\n",
      "eggmapper annotations imported for genome GCF_000784965.1 nb of annotations: 3580\n",
      "### 307 / 627\n",
      "eggmapper annotations imported for genome GCF_000785495.1 nb of annotations: 4541\n",
      "### 308 / 627\n",
      "eggmapper annotations imported for genome GCF_000785535.1 nb of annotations: 3384\n",
      "### 309 / 627\n",
      "eggmapper annotations imported for genome GCF_000785705.2 nb of annotations: 3800\n",
      "### 310 / 627\n",
      "eggmapper annotations imported for genome GCF_000789395.1 nb of annotations: 2536\n",
      "### 311 / 627\n",
      "eggmapper annotations imported for genome GCF_000800395.1 nb of annotations: 2575\n",
      "### 312 / 627\n",
      "eggmapper annotations imported for genome GCF_000801275.2 nb of annotations: 3238\n",
      "### 313 / 627\n",
      "eggmapper annotations imported for genome GCF_000801295.1 nb of annotations: 2064\n",
      "### 314 / 627\n",
      "eggmapper annotations imported for genome GCF_000801315.1 nb of annotations: 2725\n",
      "### 315 / 627\n",
      "eggmapper annotations imported for genome GCF_000803645.1 nb of annotations: 3868\n",
      "### 316 / 627\n",
      "eggmapper annotations imported for genome GCF_000807255.1 nb of annotations: 3310\n",
      "### 317 / 627\n",
      "eggmapper annotations imported for genome GCF_000807275.1 nb of annotations: 1985\n",
      "### 318 / 627\n",
      "eggmapper annotations imported for genome GCF_000812665.2 nb of annotations: 3432\n",
      "### 319 / 627\n",
      "eggmapper annotations imported for genome GCF_000814825.1 nb of annotations: 1496\n",
      "### 320 / 627\n",
      "eggmapper annotations imported for genome GCF_000815025.1 nb of annotations: 540\n",
      "### 321 / 627\n",
      "eggmapper annotations imported for genome GCF_000815065.1 nb of annotations: 488\n",
      "### 322 / 627\n",
      "eggmapper annotations imported for genome GCF_000815185.1 nb of annotations: 3033\n",
      "### 323 / 627\n",
      "eggmapper annotations imported for genome GCF_000815225.1 nb of annotations: 1161\n",
      "### 324 / 627\n",
      "eggmapper annotations imported for genome GCF_000816085.1 nb of annotations: 1937\n",
      "### 325 / 627\n",
      "eggmapper annotations imported for genome GCF_000816185.1 nb of annotations: 1361\n",
      "### 326 / 627\n",
      "eggmapper annotations imported for genome GCF_000816305.1 nb of annotations: 1544\n",
      "### 327 / 627\n",
      "eggmapper annotations imported for genome GCF_000816845.1 nb of annotations: 3529\n",
      "### 328 / 627\n",
      "eggmapper annotations imported for genome GCF_000817955.1 nb of annotations: 2838\n",
      "### 329 / 627\n",
      "eggmapper annotations imported for genome GCF_000818015.1 nb of annotations: 3669\n",
      "### 330 / 627\n",
      "eggmapper annotations imported for genome GCF_000818095.1 nb of annotations: 3032\n",
      "### 331 / 627\n",
      "eggmapper annotations imported for genome GCF_000819445.1 nb of annotations: 2129\n",
      "### 332 / 627\n",
      "eggmapper annotations imported for genome GCF_000819565.1 nb of annotations: 3870\n",
      "### 333 / 627\n",
      "eggmapper annotations imported for genome GCF_000827125.1 nb of annotations: 2764\n",
      "### 334 / 627\n",
      "eggmapper annotations imported for genome GCF_000828475.1 nb of annotations: 2629\n",
      "### 335 / 627\n",
      "eggmapper annotations imported for genome GCF_000828515.1 nb of annotations: 581\n",
      "### 336 / 627\n",
      "eggmapper annotations imported for genome GCF_000828615.1 nb of annotations: 2524\n",
      "### 337 / 627\n",
      "eggmapper annotations imported for genome GCF_000828635.1 nb of annotations: 3196\n",
      "### 338 / 627\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eggmapper annotations imported for genome GCF_000828655.1 nb of annotations: 1856\n",
      "### 339 / 627\n",
      "eggmapper annotations imported for genome GCF_000828675.1 nb of annotations: 2002\n",
      "### 340 / 627\n",
      "eggmapper annotations imported for genome GCF_000828715.1 nb of annotations: 2992\n",
      "### 341 / 627\n",
      "eggmapper annotations imported for genome GCF_000828815.1 nb of annotations: 587\n",
      "### 342 / 627\n",
      "eggmapper annotations imported for genome GCF_000828835.1 nb of annotations: 3036\n",
      "### 343 / 627\n",
      "eggmapper annotations imported for genome GCF_000828855.1 nb of annotations: 405\n",
      "### 344 / 627\n",
      "eggmapper annotations imported for genome GCF_000828895.1 nb of annotations: 2165\n",
      "### 345 / 627\n",
      "eggmapper annotations imported for genome GCF_000828915.1 nb of annotations: 2145\n",
      "### 346 / 627\n",
      "eggmapper annotations imported for genome GCF_000828975.1 nb of annotations: 2481\n",
      "### 347 / 627\n",
      "eggmapper annotations imported for genome GCF_000829195.1 nb of annotations: 3061\n",
      "### 348 / 627\n",
      "eggmapper annotations imported for genome GCF_000829235.1 nb of annotations: 1506\n",
      "### 349 / 627\n",
      "eggmapper annotations imported for genome GCF_000829315.1 nb of annotations: 786\n",
      "### 350 / 627\n",
      "eggmapper annotations imported for genome GCF_000829395.1 nb of annotations: 1889\n",
      "### 351 / 627\n",
      "eggmapper annotations imported for genome GCF_000829415.1 nb of annotations: 4956\n",
      "### 352 / 627\n",
      "eggmapper annotations imported for genome GCF_000830005.1 nb of annotations: 6692\n",
      "### 353 / 627\n",
      "eggmapper annotations imported for genome GCF_000830985.1 nb of annotations: 2284\n",
      "### 354 / 627\n",
      "eggmapper annotations imported for genome GCF_000831005.1 nb of annotations: 2699\n",
      "### 355 / 627\n",
      "eggmapper annotations imported for genome GCF_000831485.1 nb of annotations: 1712\n",
      "### 356 / 627\n",
      "eggmapper annotations imported for genome GCF_000831645.3 nb of annotations: 2033\n",
      "### 357 / 627\n",
      "eggmapper annotations imported for genome GCF_000832305.1 nb of annotations: 3696\n",
      "### 358 / 627\n",
      "eggmapper annotations imported for genome GCF_000833575.1 nb of annotations: 2145\n",
      "### 359 / 627\n",
      "eggmapper annotations imported for genome GCF_000835165.1 nb of annotations: 2105\n",
      "### 360 / 627\n",
      "eggmapper annotations imported for genome GCF_000875755.1 nb of annotations: 626\n",
      "### 361 / 627\n",
      "eggmapper annotations imported for genome GCF_000935025.1 nb of annotations: 3890\n",
      "### 362 / 627\n",
      "eggmapper annotations imported for genome GCF_000940805.1 nb of annotations: 4249\n",
      "### 363 / 627\n",
      "eggmapper annotations imported for genome GCF_000940995.1 nb of annotations: 1465\n",
      "### 364 / 627\n",
      "eggmapper annotations imported for genome GCF_000941055.1 nb of annotations: 2529\n",
      "### 365 / 627\n",
      "eggmapper annotations imported for genome GCF_000941075.1 nb of annotations: 558\n",
      "### 366 / 627\n",
      "eggmapper annotations imported for genome GCF_000949425.1 nb of annotations: 4341\n",
      "### 367 / 627\n",
      "eggmapper annotations imported for genome GCF_000950575.1 nb of annotations: 2916\n",
      "### 368 / 627\n",
      "eggmapper annotations imported for genome GCF_000952035.1 nb of annotations: 6580\n",
      "### 369 / 627\n",
      "eggmapper annotations imported for genome GCF_000952975.1 nb of annotations: 1348\n",
      "### 370 / 627\n",
      "eggmapper annotations imported for genome GCF_000953015.1 nb of annotations: 1588\n",
      "### 371 / 627\n",
      "eggmapper annotations imported for genome GCF_000953135.1 nb of annotations: 2550\n",
      "### 372 / 627\n",
      "eggmapper annotations imported for genome GCF_000953195.1 nb of annotations: 1108\n",
      "### 373 / 627\n",
      "eggmapper annotations imported for genome GCF_000953215.1 nb of annotations: 1736\n",
      "### 374 / 627\n",
      "eggmapper annotations imported for genome GCF_000953435.1 nb of annotations: 304\n",
      "### 375 / 627\n",
      "eggmapper annotations imported for genome GCF_000953475.1 nb of annotations: 1655\n",
      "### 376 / 627\n",
      "eggmapper annotations imported for genome GCF_000953535.1 nb of annotations: 2140\n",
      "### 377 / 627\n",
      "eggmapper annotations imported for genome GCF_000953635.1 nb of annotations: 2165\n",
      "### 378 / 627\n",
      "eggmapper annotations imported for genome GCF_000953655.1 nb of annotations: 2211\n",
      "### 379 / 627\n",
      "eggmapper annotations imported for genome GCF_000953695.1 nb of annotations: 2420\n",
      "### 380 / 627\n",
      "eggmapper annotations imported for genome GCF_000953735.1 nb of annotations: 3801\n",
      "### 381 / 627\n",
      "eggmapper annotations imported for genome GCF_000954135.1 nb of annotations: 2994\n",
      "### 382 / 627\n",
      "eggmapper annotations imported for genome GCF_000959245.1 nb of annotations: 2899\n",
      "### 383 / 627\n",
      "eggmapper annotations imported for genome GCF_000960975.1 nb of annotations: 4029\n",
      "### 384 / 627\n",
      "eggmapper annotations imported for genome GCF_000961095.1 nb of annotations: 4488\n",
      "### 385 / 627\n",
      "eggmapper annotations imported for genome GCF_000963865.1 nb of annotations: 2726\n",
      "### 386 / 627\n",
      "eggmapper annotations imported for genome GCF_000964565.1 nb of annotations: 2995\n",
      "### 387 / 627\n",
      "eggmapper annotations imported for genome GCF_000967305.2 nb of annotations: 3535\n",
      "### 388 / 627\n",
      "eggmapper annotations imported for genome GCF_000967425.1 nb of annotations: 3663\n",
      "### 389 / 627\n",
      "eggmapper annotations imported for genome GCF_000967895.1 nb of annotations: 3303\n",
      "### 390 / 627\n",
      "eggmapper annotations imported for genome GCF_000967915.1 nb of annotations: 1332\n",
      "### 391 / 627\n",
      "eggmapper annotations imported for genome GCF_000968055.1 nb of annotations: 1100\n",
      "### 392 / 627\n",
      "eggmapper annotations imported for genome GCF_000968135.1 nb of annotations: 3023\n",
      "### 393 / 627\n",
      "eggmapper annotations imported for genome GCF_000968195.1 nb of annotations: 3131\n",
      "### 394 / 627\n",
      "eggmapper annotations imported for genome GCF_000968375.1 nb of annotations: 4468\n",
      "### 395 / 627\n",
      "eggmapper annotations imported for genome GCF_000968535.2 nb of annotations: 3321\n",
      "### 396 / 627\n",
      "eggmapper annotations imported for genome GCF_000969645.2 nb of annotations: 3485\n",
      "### 397 / 627\n",
      "eggmapper annotations imported for genome GCF_000972245.3 nb of annotations: 4076\n",
      "### 398 / 627\n",
      "eggmapper annotations imported for genome GCF_000972765.1 nb of annotations: 2409\n",
      "### 399 / 627\n",
      "eggmapper annotations imported for genome GCF_000973085.1 nb of annotations: 1054\n",
      "### 400 / 627\n",
      "eggmapper annotations imported for genome GCF_000973105.1 nb of annotations: 3991\n",
      "### 401 / 627\n",
      "eggmapper annotations imported for genome GCF_000973505.1 nb of annotations: 572\n",
      "### 402 / 627\n",
      "eggmapper annotations imported for genome GCF_000973545.1 nb of annotations: 567\n",
      "### 403 / 627\n",
      "eggmapper annotations imported for genome GCF_000973625.1 nb of annotations: 1835\n",
      "### 404 / 627\n",
      "eggmapper annotations imported for genome GCF_000973725.1 nb of annotations: 3776\n",
      "### 405 / 627\n",
      "eggmapper annotations imported for genome GCF_000974425.1 nb of annotations: 4865\n",
      "### 406 / 627\n",
      "eggmapper annotations imported for genome GCF_000980815.1 nb of annotations: 1924\n",
      "### 407 / 627\n",
      "eggmapper annotations imported for genome GCF_000980835.1 nb of annotations: 1713\n",
      "### 408 / 627\n",
      "eggmapper annotations imported for genome GCF_000981505.1 nb of annotations: 1259\n",
      "### 409 / 627\n",
      "eggmapper annotations imported for genome GCF_000981525.1 nb of annotations: 1883\n",
      "### 410 / 627\n",
      "eggmapper annotations imported for genome GCF_000981545.1 nb of annotations: 3038\n",
      "### 411 / 627\n",
      "eggmapper annotations imported for genome GCF_000981765.1 nb of annotations: 2009\n",
      "### 412 / 627\n",
      "eggmapper annotations imported for genome GCF_000982715.1 nb of annotations: 4712\n",
      "### 413 / 627\n",
      "eggmapper annotations imported for genome GCF_000987835.1 nb of annotations: 1929\n",
      "### 414 / 627\n",
      "eggmapper annotations imported for genome GCF_000993785.2 nb of annotations: 4548\n",
      "### 415 / 627\n",
      "eggmapper annotations imported for genome GCF_001005905.1 nb of annotations: 2376\n",
      "### 416 / 627\n",
      "eggmapper annotations imported for genome GCF_001006005.1 nb of annotations: 4890\n",
      "### 417 / 627\n",
      "eggmapper annotations imported for genome GCF_001007875.1 nb of annotations: 3231\n",
      "### 418 / 627\n",
      "eggmapper annotations imported for genome GCF_001007935.1 nb of annotations: 2749\n",
      "### 419 / 627\n",
      "eggmapper annotations imported for genome GCF_001007995.1 nb of annotations: 2466\n",
      "### 420 / 627\n",
      "eggmapper annotations imported for genome GCF_001008165.2 nb of annotations: 2712\n",
      "### 421 / 627\n",
      "eggmapper annotations imported for genome GCF_001010285.1 nb of annotations: 3661\n",
      "### 422 / 627\n",
      "eggmapper annotations imported for genome GCF_001010805.1 nb of annotations: 2762\n",
      "### 423 / 627\n",
      "eggmapper annotations imported for genome GCF_001011035.1 nb of annotations: 5774\n",
      "### 424 / 627\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eggmapper annotations imported for genome GCF_001011155.1 nb of annotations: 4054\n",
      "### 425 / 627\n",
      "eggmapper annotations imported for genome GCF_001013905.1 nb of annotations: 5619\n",
      "### 426 / 627\n",
      "eggmapper annotations imported for genome GCF_001017435.1 nb of annotations: 4590\n",
      "### 427 / 627\n",
      "eggmapper annotations imported for genome GCF_001017775.2 nb of annotations: 3757\n",
      "### 428 / 627\n",
      "eggmapper annotations imported for genome GCF_001020955.1 nb of annotations: 2449\n",
      "### 429 / 627\n",
      "eggmapper annotations imported for genome GCF_001020985.1 nb of annotations: 2256\n",
      "### 430 / 627\n",
      "eggmapper annotations imported for genome GCF_001021025.1 nb of annotations: 2029\n",
      "### 431 / 627\n",
      "eggmapper annotations imported for genome GCF_001021045.1 nb of annotations: 2211\n",
      "### 432 / 627\n",
      "eggmapper annotations imported for genome GCF_001021065.1 nb of annotations: 1836\n",
      "### 433 / 627\n",
      "eggmapper annotations imported for genome GCF_001021085.1 nb of annotations: 851\n",
      "### 434 / 627\n",
      "eggmapper annotations imported for genome GCF_001021385.1 nb of annotations: 5069\n",
      "### 435 / 627\n",
      "eggmapper annotations imported for genome GCF_001021935.1 nb of annotations: 3145\n",
      "### 436 / 627\n",
      "eggmapper annotations imported for genome GCF_001021975.1 nb of annotations: 3466\n",
      "### 437 / 627\n",
      "eggmapper annotations imported for genome GCF_001022135.1 nb of annotations: 4687\n",
      "### 438 / 627\n",
      "eggmapper annotations imported for genome GCF_001022195.1 nb of annotations: 5320\n",
      "### 439 / 627\n",
      "eggmapper annotations imported for genome GCF_001023575.1 nb of annotations: 1796\n",
      "### 440 / 627\n",
      "eggmapper annotations imported for genome GCF_001026985.1 nb of annotations: 3143\n",
      "### 441 / 627\n",
      "eggmapper annotations imported for genome GCF_001027025.1 nb of annotations: 1775\n",
      "### 442 / 627\n",
      "eggmapper annotations imported for genome GCF_001027285.1 nb of annotations: 7806\n",
      "### 443 / 627\n",
      "eggmapper annotations imported for genome GCF_001027545.1 nb of annotations: 1083\n",
      "### 444 / 627\n",
      "eggmapper annotations imported for genome GCF_001028625.1 nb of annotations: 2327\n",
      "### 445 / 627\n",
      "eggmapper annotations imported for genome GCF_001028705.1 nb of annotations: 2786\n",
      "### 446 / 627\n",
      "eggmapper annotations imported for genome GCF_001029265.1 nb of annotations: 692\n",
      "### 447 / 627\n",
      "eggmapper annotations imported for genome GCF_001042405.1 nb of annotations: 1289\n",
      "### 448 / 627\n",
      "eggmapper annotations imported for genome GCF_001042615.1 nb of annotations: 1610\n",
      "### 449 / 627\n",
      "eggmapper annotations imported for genome GCF_001042635.1 nb of annotations: 2041\n",
      "### 450 / 627\n",
      "eggmapper annotations imported for genome GCF_001042715.1 nb of annotations: 3274\n",
      "### 451 / 627\n",
      "eggmapper annotations imported for genome GCF_001043175.1 nb of annotations: 3296\n",
      "### 452 / 627\n",
      "eggmapper annotations imported for genome GCF_001046955.1 nb of annotations: 3258\n",
      "### 453 / 627\n",
      "eggmapper annotations imported for genome GCF_001050115.1 nb of annotations: 2591\n",
      "### 454 / 627\n",
      "eggmapper annotations imported for genome GCF_001050435.1 nb of annotations: 2161\n",
      "### 455 / 627\n",
      "eggmapper annotations imported for genome GCF_001050475.1 nb of annotations: 1978\n",
      "### 456 / 627\n",
      "eggmapper annotations imported for genome GCF_001051995.2 nb of annotations: 3451\n",
      "### 457 / 627\n",
      "eggmapper annotations imported for genome GCF_001077715.1 nb of annotations: 800\n",
      "### 458 / 627\n",
      "eggmapper annotations imported for genome GCF_001077815.2 nb of annotations: 2406\n",
      "### 459 / 627\n",
      "eggmapper annotations imported for genome GCF_001078055.1 nb of annotations: 3254\n",
      "### 460 / 627\n",
      "eggmapper annotations imported for genome GCF_001078275.1 nb of annotations: 4282\n",
      "### 461 / 627\n",
      "eggmapper annotations imported for genome GCF_001185205.1 nb of annotations: 3934\n",
      "### 462 / 627\n",
      "eggmapper annotations imported for genome GCF_001187595.1 nb of annotations: 2635\n",
      "### 463 / 627\n",
      "eggmapper annotations imported for genome GCF_001187785.1 nb of annotations: 2576\n",
      "### 464 / 627\n",
      "eggmapper annotations imported for genome GCF_001187845.1 nb of annotations: 2962\n",
      "### 465 / 627\n",
      "eggmapper annotations imported for genome GCF_001188915.1 nb of annotations: 2031\n",
      "### 466 / 627\n",
      "eggmapper annotations imported for genome GCF_001189295.1 nb of annotations: 5565\n",
      "### 467 / 627\n",
      "eggmapper annotations imported for genome GCF_001189495.1 nb of annotations: 1408\n",
      "### 468 / 627\n",
      "eggmapper annotations imported for genome GCF_001189515.2 nb of annotations: 2039\n",
      "### 469 / 627\n",
      "eggmapper annotations imported for genome GCF_001189535.1 nb of annotations: 2756\n",
      "### 470 / 627\n",
      "eggmapper annotations imported for genome GCF_001190755.1 nb of annotations: 1415\n",
      "### 471 / 627\n",
      "eggmapper annotations imported for genome GCF_001190925.1 nb of annotations: 4216\n",
      "### 472 / 627\n",
      "eggmapper annotations imported for genome GCF_001190945.1 nb of annotations: 4368\n",
      "### 473 / 627\n",
      "eggmapper annotations imported for genome GCF_001191005.1 nb of annotations: 4531\n",
      "### 474 / 627\n",
      "eggmapper annotations imported for genome GCF_001262015.1 nb of annotations: 1199\n",
      "### 475 / 627\n",
      "eggmapper annotations imported for genome GCF_001262055.1 nb of annotations: 1487\n",
      "### 476 / 627\n",
      "eggmapper annotations imported for genome GCF_001262075.1 nb of annotations: 1989\n",
      "### 477 / 627\n",
      "eggmapper annotations imported for genome GCF_001263175.1 nb of annotations: 2776\n",
      "### 478 / 627\n",
      "eggmapper annotations imported for genome GCF_001263395.1 nb of annotations: 1195\n",
      "### 479 / 627\n",
      "eggmapper annotations imported for genome GCF_001267175.1 nb of annotations: 2146\n",
      "### 480 / 627\n",
      "eggmapper annotations imported for genome GCF_001267865.1 nb of annotations: 2003\n",
      "### 481 / 627\n",
      "eggmapper annotations imported for genome GCF_001267925.1 nb of annotations: 3995\n",
      "### 482 / 627\n",
      "eggmapper annotations imported for genome GCF_001273775.1 nb of annotations: 3084\n",
      "### 483 / 627\n",
      "eggmapper annotations imported for genome GCF_001273795.1 nb of annotations: 890\n",
      "### 484 / 627\n",
      "eggmapper annotations imported for genome GCF_001274535.1 nb of annotations: 1927\n",
      "### 485 / 627\n",
      "eggmapper annotations imported for genome GCF_001274875.1 nb of annotations: 521\n",
      "### 486 / 627\n",
      "eggmapper annotations imported for genome GCF_001274895.1 nb of annotations: 1977\n",
      "### 487 / 627\n",
      "eggmapper annotations imported for genome GCF_001275345.1 nb of annotations: 2128\n",
      "### 488 / 627\n",
      "eggmapper annotations imported for genome GCF_001277995.1 nb of annotations: 2235\n",
      "### 489 / 627\n",
      "eggmapper annotations imported for genome GCF_001278035.1 nb of annotations: 3486\n",
      "### 490 / 627\n",
      "eggmapper annotations imported for genome GCF_001278055.1 nb of annotations: 2976\n",
      "### 491 / 627\n",
      "eggmapper annotations imported for genome GCF_001278715.1 nb of annotations: 2288\n",
      "### 492 / 627\n",
      "eggmapper annotations imported for genome GCF_001281045.1 nb of annotations: 664\n",
      "### 493 / 627\n",
      "eggmapper annotations imported for genome GCF_001281115.1 nb of annotations: 2949\n",
      "### 494 / 627\n",
      "eggmapper annotations imported for genome GCF_001281315.1 nb of annotations: 3035\n",
      "### 495 / 627\n",
      "eggmapper annotations imported for genome GCF_001281385.1 nb of annotations: 1536\n",
      "### 496 / 627\n",
      "eggmapper annotations imported for genome GCF_001281405.1 nb of annotations: 1082\n",
      "### 497 / 627\n",
      "eggmapper annotations imported for genome GCF_001281465.1 nb of annotations: 2562\n",
      "### 498 / 627\n",
      "eggmapper annotations imported for genome GCF_001281485.1 nb of annotations: 2323\n",
      "### 499 / 627\n",
      "eggmapper annotations imported for genome GCF_001281505.1 nb of annotations: 1333\n",
      "### 500 / 627\n",
      "eggmapper annotations imported for genome GCF_001293165.1 nb of annotations: 1374\n",
      "### 501 / 627\n",
      "eggmapper annotations imported for genome GCF_001294645.1 nb of annotations: 4474\n",
      "### 502 / 627\n",
      "eggmapper annotations imported for genome GCF_001298525.1 nb of annotations: 1985\n",
      "### 503 / 627\n",
      "eggmapper annotations imported for genome GCF_001302585.1 nb of annotations: 8777\n",
      "### 504 / 627\n",
      "eggmapper annotations imported for genome GCF_001305575.2 nb of annotations: 2890\n",
      "### 505 / 627\n",
      "eggmapper annotations imported for genome GCF_001305595.1 nb of annotations: 2324\n",
      "### 506 / 627\n",
      "eggmapper annotations imported for genome GCF_001305615.1 nb of annotations: 3271\n",
      "### 507 / 627\n",
      "eggmapper annotations imported for genome GCF_001305675.1 nb of annotations: 3594\n",
      "### 508 / 627\n",
      "eggmapper annotations imported for genome GCF_001307195.1 nb of annotations: 2316\n",
      "### 509 / 627\n",
      "eggmapper annotations imported for genome GCF_001307545.1 nb of annotations: 5137\n",
      "### 510 / 627\n",
      "eggmapper annotations imported for genome GCF_001307805.1 nb of annotations: 3113\n",
      "### 511 / 627\n",
      "eggmapper annotations imported for genome GCF_001308105.1 nb of annotations: 2418\n",
      "### 512 / 627\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eggmapper annotations imported for genome GCF_001308265.1 nb of annotations: 2740\n",
      "### 513 / 627\n",
      "eggmapper annotations imported for genome GCF_001310085.1 nb of annotations: 3411\n",
      "### 514 / 627\n",
      "eggmapper annotations imported for genome GCF_001310225.1 nb of annotations: 3039\n",
      "### 515 / 627\n",
      "eggmapper annotations imported for genome GCF_001310255.1 nb of annotations: 2840\n",
      "### 516 / 627\n",
      "eggmapper annotations imported for genome GCF_001314225.1 nb of annotations: 3470\n",
      "### 517 / 627\n",
      "eggmapper annotations imported for genome GCF_001314305.1 nb of annotations: 3133\n",
      "### 518 / 627\n",
      "eggmapper annotations imported for genome GCF_001314325.1 nb of annotations: 4006\n",
      "### 519 / 627\n",
      "eggmapper annotations imported for genome GCF_001314945.1 nb of annotations: 1221\n",
      "### 520 / 627\n",
      "eggmapper annotations imported for genome GCF_001318295.1 nb of annotations: 683\n",
      "### 521 / 627\n",
      "eggmapper annotations imported for genome GCF_001318345.1 nb of annotations: 4497\n",
      "### 522 / 627\n",
      "eggmapper annotations imported for genome GCF_001402875.1 nb of annotations: 2833\n",
      "### 523 / 627\n",
      "eggmapper annotations imported for genome GCF_001402915.1 nb of annotations: 4161\n",
      "### 524 / 627\n",
      "eggmapper annotations imported for genome GCF_001411495.1 nb of annotations: 3094\n",
      "### 525 / 627\n",
      "eggmapper annotations imported for genome GCF_001412575.1 nb of annotations: 2608\n",
      "### 526 / 627\n",
      "eggmapper annotations imported for genome GCF_001412595.1 nb of annotations: 4603\n",
      "### 527 / 627\n",
      "eggmapper annotations imported for genome GCF_001420915.1 nb of annotations: 3394\n",
      "### 528 / 627\n",
      "eggmapper annotations imported for genome GCF_001420995.1 nb of annotations: 4867\n",
      "### 529 / 627\n",
      "eggmapper annotations imported for genome GCF_001421015.1 nb of annotations: 3919\n",
      "### 530 / 627\n",
      "eggmapper annotations imported for genome GCF_001430825.1 nb of annotations: 2901\n",
      "### 531 / 627\n",
      "eggmapper annotations imported for genome GCF_001431725.1 nb of annotations: 3723\n",
      "### 532 / 627\n",
      "eggmapper annotations imported for genome GCF_001432245.1 nb of annotations: 2310\n",
      "### 533 / 627\n",
      "eggmapper annotations imported for genome GCF_001439585.2 nb of annotations: 2904\n",
      "### 534 / 627\n",
      "eggmapper annotations imported for genome GCF_001441165.1 nb of annotations: 2747\n",
      "### 535 / 627\n",
      "eggmapper annotations imported for genome GCF_001442535.1 nb of annotations: 3644\n",
      "### 536 / 627\n",
      "eggmapper annotations imported for genome GCF_001442785.1 nb of annotations: 3923\n",
      "### 537 / 627\n",
      "eggmapper annotations imported for genome GCF_001443605.1 nb of annotations: 2982\n",
      "### 538 / 627\n",
      "eggmapper annotations imported for genome GCF_001444365.1 nb of annotations: 3131\n",
      "### 539 / 627\n",
      "eggmapper annotations imported for genome GCF_001444405.1 nb of annotations: 2957\n",
      "### 540 / 627\n",
      "eggmapper annotations imported for genome GCF_001444425.1 nb of annotations: 2678\n",
      "### 541 / 627\n",
      "eggmapper annotations imported for genome GCF_001444445.1 nb of annotations: 1974\n",
      "### 542 / 627\n",
      "eggmapper annotations imported for genome GCF_001447335.1 nb of annotations: 3517\n",
      "### 543 / 627\n",
      "eggmapper annotations imported for genome GCF_001454945.1 nb of annotations: 2696\n",
      "### 544 / 627\n",
      "eggmapper annotations imported for genome GCF_001455205.1 nb of annotations: 2429\n",
      "### 545 / 627\n",
      "eggmapper annotations imported for genome GCF_001455325.1 nb of annotations: 2849\n",
      "### 546 / 627\n",
      "eggmapper annotations imported for genome GCF_001458695.1 nb of annotations: 2280\n",
      "### 547 / 627\n",
      "eggmapper annotations imported for genome GCF_001460635.1 nb of annotations: 1505\n",
      "### 548 / 627\n",
      "eggmapper annotations imported for genome GCF_001465255.1 nb of annotations: 4269\n",
      "### 549 / 627\n",
      "eggmapper annotations imported for genome GCF_001465275.1 nb of annotations: 4197\n",
      "### 550 / 627\n",
      "eggmapper annotations imported for genome GCF_001465835.2 nb of annotations: 3014\n",
      "### 551 / 627\n",
      "eggmapper annotations imported for genome GCF_001465855.1 nb of annotations: 1640\n",
      "### 552 / 627\n",
      "eggmapper annotations imported for genome GCF_001466725.1 nb of annotations: 3436\n",
      "### 553 / 627\n",
      "eggmapper annotations imported for genome GCF_001477625.1 nb of annotations: 3790\n",
      "### 554 / 627\n",
      "eggmapper annotations imported for genome GCF_001481655.1 nb of annotations: 3121\n",
      "### 555 / 627\n",
      "eggmapper annotations imported for genome GCF_001482365.1 nb of annotations: 2603\n",
      "### 556 / 627\n",
      "eggmapper annotations imported for genome GCF_001482385.1 nb of annotations: 3194\n",
      "### 557 / 627\n",
      "eggmapper annotations imported for genome GCF_001483385.1 nb of annotations: 2098\n",
      "### 558 / 627\n",
      "eggmapper annotations imported for genome GCF_001483865.1 nb of annotations: 3916\n",
      "### 559 / 627\n",
      "eggmapper annotations imported for genome GCF_001483965.1 nb of annotations: 2096\n",
      "### 560 / 627\n",
      "eggmapper annotations imported for genome GCF_001484065.1 nb of annotations: 4384\n",
      "### 561 / 627\n",
      "eggmapper annotations imported for genome GCF_001484565.1 nb of annotations: 7495\n",
      "### 562 / 627\n",
      "eggmapper annotations imported for genome GCF_001484605.1 nb of annotations: 3890\n",
      "### 563 / 627\n",
      "eggmapper annotations imported for genome GCF_001499615.1 nb of annotations: 2759\n",
      "### 564 / 627\n",
      "eggmapper annotations imported for genome GCF_001499655.1 nb of annotations: 1706\n",
      "### 565 / 627\n",
      "eggmapper annotations imported for genome GCF_001507665.1 nb of annotations: 2416\n",
      "### 566 / 627\n",
      "eggmapper annotations imported for genome GCF_001518815.1 nb of annotations: 3990\n",
      "### 567 / 627\n",
      "eggmapper annotations imported for genome GCF_001518835.1 nb of annotations: 4153\n",
      "### 568 / 627\n",
      "eggmapper annotations imported for genome GCF_001542565.1 nb of annotations: 3517\n",
      "### 569 / 627\n",
      "eggmapper annotations imported for genome GCF_001543265.1 nb of annotations: 4009\n",
      "### 570 / 627\n",
      "eggmapper annotations imported for genome GCF_001547975.1 nb of annotations: 435\n",
      "### 571 / 627\n",
      "eggmapper annotations imported for genome GCF_001547995.1 nb of annotations: 2300\n",
      "### 572 / 627\n",
      "eggmapper annotations imported for genome GCF_001548015.1 nb of annotations: 3922\n",
      "### 573 / 627\n",
      "eggmapper annotations imported for genome GCF_001548055.1 nb of annotations: 2381\n",
      "### 574 / 627\n",
      "eggmapper annotations imported for genome GCF_001548075.1 nb of annotations: 3436\n",
      "### 575 / 627\n",
      "eggmapper annotations imported for genome GCF_001553195.1 nb of annotations: 490\n",
      "### 576 / 627\n",
      "eggmapper annotations imported for genome GCF_001553605.1 nb of annotations: 2584\n",
      "### 577 / 627\n",
      "eggmapper annotations imported for genome GCF_001553935.1 nb of annotations: 2090\n",
      "### 578 / 627\n",
      "eggmapper annotations imported for genome GCF_001553955.1 nb of annotations: 1828\n",
      "### 579 / 627\n",
      "eggmapper annotations imported for genome GCF_001558415.1 nb of annotations: 1382\n",
      "### 580 / 627\n",
      "eggmapper annotations imported for genome GCF_001558735.1 nb of annotations: 3412\n",
      "### 581 / 627\n",
      "eggmapper annotations imported for genome GCF_001559115.1 nb of annotations: 2119\n",
      "### 582 / 627\n",
      "eggmapper annotations imported for genome GCF_001562115.1 nb of annotations: 3562\n",
      "### 583 / 627\n",
      "eggmapper annotations imported for genome GCF_001578105.1 nb of annotations: 3767\n",
      "### 584 / 627\n",
      "eggmapper annotations imported for genome GCF_001578185.1 nb of annotations: 4120\n",
      "### 585 / 627\n",
      "eggmapper annotations imported for genome GCF_001583415.1 nb of annotations: 4452\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001584185.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001586165.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001586195.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001586215.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001586255.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001597285.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001606025.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001610975.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001611135.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001618685.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001620305.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001632865.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001636015.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001641285.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001642655.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001644565.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001648175.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001652465.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001652485.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001652565.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001653755.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001654495.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_001655245.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_900011245.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_900078695.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_900086555.1\n",
      "### 586 / 627\n",
      "ERROR: eggmapper annotation file not found for genome GCF_900087055.1\n"
     ]
    }
   ],
   "source": [
    "eggnogMapperAnnotationResultList = []\n",
    "eggnogMapperAnnotationErrorList = []\n",
    "eggnogMapperAnnotationResultPath = analysisEggnogPath / 'Eggnog_mapper_results' / 'HMM_cluster/results2'\n",
    "\n",
    "#genomeAccession = 'GCF_000006685.1'\n",
    "\n",
    "# For each genome in the RefSeq database (common species), import eggmapper results\n",
    "count = 1\n",
    "for genomeAccession, row in assemblySummaryRepRemainingDf.iterrows():\n",
    "\n",
    "    print(\"###\", count, \"/\", len(assemblySummaryRepRemainingDf))\n",
    "    \n",
    "    genomeBaseName = re.search(r'^(GCF_.+)_genomic\\.gbff\\.gz$', row['compressedGenomeFilename']).group(1)\n",
    "    annotationResultsFilename = genomeBaseName + '.emapper.annotations'\n",
    "#     print(\"annotationResultsFilename\", annotationResultsFilename)\n",
    "    annotationResultsFilePath =  eggnogMapperAnnotationResultPath / annotationResultsFilename\n",
    "    \n",
    "    if annotationResultsFilePath.is_file():    \n",
    "        eggnogMapperAnnotationResultDf = \\\n",
    "            pd.read_csv(str(annotationResultsFilePath),\n",
    "                        sep='\\t', header=2).rename(columns={'#query_name':'query_name'})\n",
    "\n",
    "        # Drop comment lines at the end of the file\n",
    "        eggnogMapperAnnotationResultDf = eggnogMapperAnnotationResultDf[~eggnogMapperAnnotationResultDf['query_name'].str.contains(r'^#')]\n",
    "\n",
    "#         eggnogMapperAnnotationResultDf = eggnogMapperAnnotationResultDf[['query_name', 'seed_eggNOG_ortholog', 'seed_ortholog_evalue', 'GO_terms', 'COG cat', 'eggNOG annot']]\n",
    "        eggnogMapperAnnotationResultDf.rename(columns={'query_name':'refSeqProteinId'}, inplace=True)\n",
    "        eggnogMapperAnnotationResultDf\n",
    "\n",
    "        # Find the taxid of the genome\n",
    "        taxid = assemblySummaryRepRemainingDf.loc[genomeAccession]['taxid']\n",
    "#         print(\"taxid\", taxid)\n",
    "\n",
    "        eggnogMapperAnnotationResultDf['genome_accession'] = genomeAccession\n",
    "        eggnogMapperAnnotationResultDf['taxid'] = taxid\n",
    "        #print(eggnogMapperAnnotationResultDf.head(2))\n",
    "\n",
    "        eggnogMapperAnnotationResultList.append(eggnogMapperAnnotationResultDf)\n",
    "        count += 1\n",
    "        print(\"eggmapper annotations imported for genome\", genomeAccession,\n",
    "              \"nb of annotations:\", len(eggnogMapperAnnotationResultDf))\n",
    "    else:\n",
    "        print(\"ERROR: eggmapper annotation file not found for genome\", genomeAccession)\n",
    "        eggnogMapperAnnotationErrorList.append(genomeAccession)\n",
    "\n",
    "\n",
    "eggnogMapperAnnotationResultDf = pd.concat(eggnogMapperAnnotationResultList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(eggnogMapperAnnotationErrorList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1639599\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>seed_eggNOG_ortholog</th>\n",
       "      <th>seed_ortholog_evalue</th>\n",
       "      <th>seed_ortholog_score</th>\n",
       "      <th>predicted_gene_name</th>\n",
       "      <th>GO_terms</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>Annotation_tax_scope</th>\n",
       "      <th>OGs</th>\n",
       "      <th>bestOG|evalue|score</th>\n",
       "      <th>COG cat</th>\n",
       "      <th>eggNOG annot</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WP_012914566.1</td>\n",
       "      <td>190485.XCC0001</td>\n",
       "      <td>4.800000e-265</td>\n",
       "      <td>875.7</td>\n",
       "      <td>DNAA</td>\n",
       "      <td>GO:0000166,GO:0001882,GO:0001883,GO:0003674,GO...</td>\n",
       "      <td>map02020,map04112</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>05CI4@bactNOG,0QIVT@gproNOG,16QDA@proNOG,COG05...</td>\n",
       "      <td>05CI4|2.97840944354e-179|600.451416016</td>\n",
       "      <td>L</td>\n",
       "      <td>it binds specifically double-stranded DNA at a...</td>\n",
       "      <td>GCF_000087965.2</td>\n",
       "      <td>380358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WP_012914567.1</td>\n",
       "      <td>190486.XAC0002</td>\n",
       "      <td>6.500000e-211</td>\n",
       "      <td>695.8</td>\n",
       "      <td>DNAN</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0003887,GO:0005575,GO...</td>\n",
       "      <td>map00230,map00240,map01100,map03030,map03430,m...</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>05CZ8@bactNOG,0QHTG@gproNOG,16QFW@proNOG,COG05...</td>\n",
       "      <td>05CZ8|5.69544324799e-132|443.953308105</td>\n",
       "      <td>L</td>\n",
       "      <td>DNA polymerase III is a complex, multichain en...</td>\n",
       "      <td>GCF_000087965.2</td>\n",
       "      <td>380358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WP_012914568.1</td>\n",
       "      <td>522373.Smlt0004</td>\n",
       "      <td>2.400000e-188</td>\n",
       "      <td>622.0</td>\n",
       "      <td>RECF</td>\n",
       "      <td>GO:0000731,GO:0003674,GO:0003676,GO:0003677,GO...</td>\n",
       "      <td>map03440</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>05C3X@bactNOG,0QJ20@gproNOG,16Q55@proNOG,COG11...</td>\n",
       "      <td>05C3X|2.57774202378e-97|329.962615967</td>\n",
       "      <td>L</td>\n",
       "      <td>it is required for DNA replication and normal ...</td>\n",
       "      <td>GCF_000087965.2</td>\n",
       "      <td>380358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refSeqProteinId seed_eggNOG_ortholog  seed_ortholog_evalue  \\\n",
       "0  WP_012914566.1       190485.XCC0001         4.800000e-265   \n",
       "1  WP_012914567.1       190486.XAC0002         6.500000e-211   \n",
       "2  WP_012914568.1      522373.Smlt0004         2.400000e-188   \n",
       "\n",
       "   seed_ortholog_score predicted_gene_name  \\\n",
       "0                875.7                DNAA   \n",
       "1                695.8                DNAN   \n",
       "2                622.0                RECF   \n",
       "\n",
       "                                            GO_terms  \\\n",
       "0  GO:0000166,GO:0001882,GO:0001883,GO:0003674,GO...   \n",
       "1  GO:0003674,GO:0003824,GO:0003887,GO:0005575,GO...   \n",
       "2  GO:0000731,GO:0003674,GO:0003676,GO:0003677,GO...   \n",
       "\n",
       "                                       KEGG_pathways Annotation_tax_scope  \\\n",
       "0                                  map02020,map04112          bactNOG[38]   \n",
       "1  map00230,map00240,map01100,map03030,map03430,m...          bactNOG[38]   \n",
       "2                                           map03440          bactNOG[38]   \n",
       "\n",
       "                                                 OGs  \\\n",
       "0  05CI4@bactNOG,0QIVT@gproNOG,16QDA@proNOG,COG05...   \n",
       "1  05CZ8@bactNOG,0QHTG@gproNOG,16QFW@proNOG,COG05...   \n",
       "2  05C3X@bactNOG,0QJ20@gproNOG,16Q55@proNOG,COG11...   \n",
       "\n",
       "                      bestOG|evalue|score COG cat  \\\n",
       "0  05CI4|2.97840944354e-179|600.451416016       L   \n",
       "1  05CZ8|5.69544324799e-132|443.953308105       L   \n",
       "2   05C3X|2.57774202378e-97|329.962615967       L   \n",
       "\n",
       "                                        eggNOG annot genome_accession   taxid  \n",
       "0  it binds specifically double-stranded DNA at a...  GCF_000087965.2  380358  \n",
       "1  DNA polymerase III is a complex, multichain en...  GCF_000087965.2  380358  \n",
       "2  it is required for DNA replication and normal ...  GCF_000087965.2  380358  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(eggnogMapperAnnotationResultDf))\n",
    "eggnogMapperAnnotationResultDf.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperAnnotationResultDf = \\\n",
    "    pd.concat([\n",
    "            eggnogMapperAnnotationResultDf[:],\n",
    "            eggnogMapperAnnotationResultDf[:]['bestOG|evalue|score'].str.split('|', expand=True)\\\n",
    "                .rename(columns={0:'bestOG', 1:'bestOG_evalue', 2:'bestOG_score'})\n",
    "        ], axis=1).drop(['bestOG|evalue|score'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperAnnotationResultDf2 = eggnogMapperAnnotationResultDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperAnnotationResultDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperAnnotationResultDf2.to_csv(str(analysisEggnogPath / 'eggnogMapperAnnotationResultDf2.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>seed_eggNOG_ortholog</th>\n",
       "      <th>seed_ortholog_evalue</th>\n",
       "      <th>seed_ortholog_score</th>\n",
       "      <th>predicted_gene_name</th>\n",
       "      <th>GO_terms</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>Annotation_tax_scope</th>\n",
       "      <th>OGs</th>\n",
       "      <th>COG cat</th>\n",
       "      <th>eggNOG annot</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>taxid</th>\n",
       "      <th>bestOG</th>\n",
       "      <th>bestOG_evalue</th>\n",
       "      <th>bestOG_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WP_012957018.1</td>\n",
       "      <td>398511.BpOF4_07955</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1008.6</td>\n",
       "      <td>DNAA</td>\n",
       "      <td>GO:0000166,GO:0001882,GO:0001883,GO:0003674,GO...</td>\n",
       "      <td>map02020,map04112</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>04RET@bacNOG,05CI4@bactNOG,0NDJT@firmNOG,COG05...</td>\n",
       "      <td>L</td>\n",
       "      <td>it binds specifically double-stranded DNA at a...</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>398511</td>\n",
       "      <td>05CI4</td>\n",
       "      <td>6.694999e-207</td>\n",
       "      <td>691.734497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WP_012957019.1</td>\n",
       "      <td>398511.BpOF4_07960</td>\n",
       "      <td>5.000000e-247</td>\n",
       "      <td>814.7</td>\n",
       "      <td>DNAN</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0003887,GO:0005575,GO...</td>\n",
       "      <td>map00230,map00240,map01100,map03030,map03430,m...</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>05CZ8@bactNOG,0ND7B@firmNOG,COG0592@NOG</td>\n",
       "      <td>L</td>\n",
       "      <td>DNA polymerase III is a complex, multichain en...</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>398511</td>\n",
       "      <td>05CZ8</td>\n",
       "      <td>1.862819e-120</td>\n",
       "      <td>405.965668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WP_012957020.1</td>\n",
       "      <td>398511.BpOF4_07965</td>\n",
       "      <td>8.700000e-46</td>\n",
       "      <td>145.8</td>\n",
       "      <td>YAAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>04XIN@bacNOG,0848D@bactNOG,0NR2T@firmNOG,COG25...</td>\n",
       "      <td>S</td>\n",
       "      <td>S4 domain protein YaaA</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>398511</td>\n",
       "      <td>0848D</td>\n",
       "      <td>3.257817e-30</td>\n",
       "      <td>107.957611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WP_012957021.1</td>\n",
       "      <td>398511.BpOF4_07970</td>\n",
       "      <td>2.500000e-246</td>\n",
       "      <td>812.6</td>\n",
       "      <td>RECF</td>\n",
       "      <td>GO:0000731,GO:0005575,GO:0005622,GO:0005623,GO...</td>\n",
       "      <td>map03440</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>04R5Y@bacNOG,05C3X@bactNOG,0NDA9@firmNOG,COG11...</td>\n",
       "      <td>L</td>\n",
       "      <td>it is required for DNA replication and normal ...</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>398511</td>\n",
       "      <td>05C3X</td>\n",
       "      <td>8.569240e-133</td>\n",
       "      <td>447.072968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WP_012957022.1</td>\n",
       "      <td>398511.BpOF4_07975</td>\n",
       "      <td>1.100000e-56</td>\n",
       "      <td>179.9</td>\n",
       "      <td>YAAB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bactNOG[38]</td>\n",
       "      <td>05WC4@bactNOG,0NNN5@firmNOG,1235W@NOG</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>398511</td>\n",
       "      <td>05WC4</td>\n",
       "      <td>1.301090e-27</td>\n",
       "      <td>98.584099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refSeqProteinId seed_eggNOG_ortholog  seed_ortholog_evalue  \\\n",
       "0  WP_012957018.1   398511.BpOF4_07955          0.000000e+00   \n",
       "1  WP_012957019.1   398511.BpOF4_07960         5.000000e-247   \n",
       "2  WP_012957020.1   398511.BpOF4_07965          8.700000e-46   \n",
       "3  WP_012957021.1   398511.BpOF4_07970         2.500000e-246   \n",
       "4  WP_012957022.1   398511.BpOF4_07975          1.100000e-56   \n",
       "\n",
       "   seed_ortholog_score predicted_gene_name  \\\n",
       "0               1008.6                DNAA   \n",
       "1                814.7                DNAN   \n",
       "2                145.8                YAAA   \n",
       "3                812.6                RECF   \n",
       "4                179.9                YAAB   \n",
       "\n",
       "                                            GO_terms  \\\n",
       "0  GO:0000166,GO:0001882,GO:0001883,GO:0003674,GO...   \n",
       "1  GO:0003674,GO:0003824,GO:0003887,GO:0005575,GO...   \n",
       "2                                                NaN   \n",
       "3  GO:0000731,GO:0005575,GO:0005622,GO:0005623,GO...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                       KEGG_pathways Annotation_tax_scope  \\\n",
       "0                                  map02020,map04112          bactNOG[38]   \n",
       "1  map00230,map00240,map01100,map03030,map03430,m...          bactNOG[38]   \n",
       "2                                                NaN          bactNOG[38]   \n",
       "3                                           map03440          bactNOG[38]   \n",
       "4                                                NaN          bactNOG[38]   \n",
       "\n",
       "                                                 OGs COG cat  \\\n",
       "0  04RET@bacNOG,05CI4@bactNOG,0NDJT@firmNOG,COG05...       L   \n",
       "1            05CZ8@bactNOG,0ND7B@firmNOG,COG0592@NOG       L   \n",
       "2  04XIN@bacNOG,0848D@bactNOG,0NR2T@firmNOG,COG25...       S   \n",
       "3  04R5Y@bacNOG,05C3X@bactNOG,0NDA9@firmNOG,COG11...       L   \n",
       "4              05WC4@bactNOG,0NNN5@firmNOG,1235W@NOG       S   \n",
       "\n",
       "                                        eggNOG annot genome_accession   taxid  \\\n",
       "0  it binds specifically double-stranded DNA at a...  GCF_000005825.2  398511   \n",
       "1  DNA polymerase III is a complex, multichain en...  GCF_000005825.2  398511   \n",
       "2                             S4 domain protein YaaA  GCF_000005825.2  398511   \n",
       "3  it is required for DNA replication and normal ...  GCF_000005825.2  398511   \n",
       "4                                                NaN  GCF_000005825.2  398511   \n",
       "\n",
       "  bestOG  bestOG_evalue  bestOG_score  \n",
       "0  05CI4  6.694999e-207    691.734497  \n",
       "1  05CZ8  1.862819e-120    405.965668  \n",
       "2  0848D   3.257817e-30    107.957611  \n",
       "3  05C3X  8.569240e-133    447.072968  \n",
       "4  05WC4   1.301090e-27     98.584099  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggnogMapperAnnotationResultDf = pd.read_csv(str(analysisEggnogPath / 'eggnogMapperAnnotationResultDf.csv'))\n",
    "eggnogMapperAnnotationResultDf.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "eggnogMapperAnnotationResultDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2574354"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(eggnogMapperAnnotationResultDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1639599"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(eggnogMapperAnnotationResultDf2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperAnnotationResultDf3 = pd.concat([eggnogMapperAnnotationResultDf, eggnogMapperAnnotationResultDf2])\n",
    "eggnogMapperAnnotationResultDf3.to_csv(str(analysisEggnogPath / 'eggnogMapperAnnotationResultDf3.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperAnnotationResultDf3 = pd.read_csv(str(analysisEggnogPath / 'eggnogMapperAnnotationResultDf3.csv'),\n",
    "                                              index_col=0, nrows=None)\n",
    "eggnogMapperAnnotationResultDf3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a first indication of the quality of the DIAMOND mapping, we look at the maximum expected value for the seed orthologs, among all sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8.7e-46"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggnogMapperAnnotationResultDf3['seed_ortholog_evalue'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.30109005478e-27"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggnogMapperAnnotationResultDf3['bestOG_evalue'].max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find and group annotations together for each best orthogroup."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unique_value_or_join_multiple_strings(x):\n",
    "    xUniqueValues = x.unique()\n",
    "    if len(xUniqueValues) == 1:\n",
    "        return xUniqueValues[0]\n",
    "    else:\n",
    "        return \"|\".join([x1 for x1 in xUniqueValues if not pd.isnull(x1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperOrthogroupAnnotationDf = \\\n",
    "    eggnogMapperAnnotationResultDf3.groupby(['bestOG'])[['predicted_gene_name', 'KEGG_pathways',\n",
    "                                                        'COG cat', 'eggNOG annot']]\\\n",
    "                                                      .agg(unique_value_or_join_multiple_strings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnogMapperOrthogroupAnnotationDf.to_csv(str(analysisEggnogPath / 'eggnogMapperOrthogroupAnnotationDf.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted_gene_name</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>COG cat</th>\n",
       "      <th>eggNOG annot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bestOG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0000B</th>\n",
       "      <td>ARGD</td>\n",
       "      <td>map00260,map00300,map00330,map01100,map01110,m...</td>\n",
       "      <td>E</td>\n",
       "      <td>acetylornithine and succinylornithine aminotra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0000E</th>\n",
       "      <td>ACCC</td>\n",
       "      <td>map00061,map00253,map00620,map00640,map00720,m...</td>\n",
       "      <td>I</td>\n",
       "      <td>acetyl-CoA carboxylase, biotin carboxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0000I</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>Dienelactone hydrolase family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0000K</th>\n",
       "      <td>CLPP</td>\n",
       "      <td>map04112</td>\n",
       "      <td>O</td>\n",
       "      <td>Cleaves peptides in various proteins in a proc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0000P</th>\n",
       "      <td>RLUB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>J</td>\n",
       "      <td>Pseudouridine synthase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       predicted_gene_name                                      KEGG_pathways  \\\n",
       "bestOG                                                                          \n",
       "0000B                 ARGD  map00260,map00300,map00330,map01100,map01110,m...   \n",
       "0000E                 ACCC  map00061,map00253,map00620,map00640,map00720,m...   \n",
       "0000I                  NaN                                                NaN   \n",
       "0000K                 CLPP                                           map04112   \n",
       "0000P                 RLUB                                                NaN   \n",
       "\n",
       "       COG cat                                       eggNOG annot  \n",
       "bestOG                                                             \n",
       "0000B        E  acetylornithine and succinylornithine aminotra...  \n",
       "0000E        I         acetyl-CoA carboxylase, biotin carboxylase  \n",
       "0000I        E                      Dienelactone hydrolase family  \n",
       "0000K        O  Cleaves peptides in various proteins in a proc...  \n",
       "0000P        J                             Pseudouridine synthase  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggnogMapperOrthogroupAnnotationDf = pd.read_csv(str(analysisEggnogPath / 'eggnogMapperOrthogroupAnnotationDf.csv'),\n",
    "                                                 index_col=0, nrows=10)\n",
    "eggnogMapperOrthogroupAnnotationDf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functional enrichment analysis, COG\n",
    "\n",
    "If we consider clusters of orthologous groups (COG), we can perform an enrichment analysis for protein subsets.\n",
    "\n",
    "With the EggNOG database we will be able to perform this analysis only at the specific taxonomic levels of the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "COG cat\n",
       "C          655\n",
       "C, O         2\n",
       "C, P         1\n",
       "D          109\n",
       "E          570\n",
       "E, F         1\n",
       "E, G, P     14\n",
       "E, P         7\n",
       "E, T         5\n",
       "F          629\n",
       "dtype: int64"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonAllCDSDf.groupby('COG cat').size().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I should check how to handle **multiple COG categories** for the enrichment test. This is a common counfounding factor in gene set analysis. See [Multifunctionality | ermineJ](http://erminej.chibi.ubc.ca/help/tutorials/multifunctionality/) for software that correct for genes that have more functions than others. Also, see\n",
    "\n",
    "[Park2015] Park, J., Taslim, C., & Lin, S. (2015). BOG: R-package for Bacterium and virus analysis of Orthologous Groups. Computational and Structural Biotechnology Journal, 13, 366–369. http://doi.org/10.1016/j.csbj.2015.05.002\n",
    "\n",
    "In the following I will assume that most sequences fall only into one COG category and that the influence of sequences with multiple categories is small enough such that we can neglect this effect."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "During the COG enrichment analysis, we consider all members of category `Y`, even if such elements can be members of multiple categories. In general, if some elements are members of many categories (annotations), they could bias the enrichment analysis (see references above)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# False discovery rate for the multiple tests correction\n",
    "alpha_family_wise_error_rate = 0.01\n",
    "enrichmentTestcolumnName = 'reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_' + str(alpha_family_wise_error_rate)\n",
    "groupCol = 'COG cat'\n",
    "groupName = 'COGGroup'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['U',\n",
       " 'J',\n",
       " 'T',\n",
       " 'E',\n",
       " 'N',\n",
       " 'M',\n",
       " 'F',\n",
       " 'W',\n",
       " 'V',\n",
       " 'S',\n",
       " 'K',\n",
       " 'G',\n",
       " 'D',\n",
       " 'O',\n",
       " 'L',\n",
       " 'P',\n",
       " 'C',\n",
       " 'I',\n",
       " 'Q',\n",
       " 'H']"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract list of COG categories\n",
    "COGList = list(set([s for smult in list(taxonAllCDSDf[groupCol].dropna().unique()) for s in smult.split(', ')]))\n",
    "COGList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "####### R\n",
      "####### H\n"
     ]
    }
   ],
   "source": [
    "taxonGroupEnrichDf = \\\n",
    "    enrichment_analysis_cterminal_subset(taxonAllCDSDf, taxonId, groupCol, groupName, COGList, aaTable[:2],\n",
    "                                         alpha_family_wise_error_rate, enrichmentTestcolumnName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MultiIndex(levels=[[1937961], ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'S', 'T', 'U', 'V', 'Z']],\n",
       "           labels=[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]],\n",
       "           names=['taxon_id', 'COGGroup_index'])"
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonGroupEnrichDf.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MultiIndex(levels=[['H', 'R'], ['count_NOT_cterm_NOT_in_group', 'count_NOT_cterm_in_group', 'count_cterm_NOT_in_group', 'count_cterm_in_group', 'odds_ratio', 'pvalue_fisher', 'reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01']],\n",
       "           labels=[[0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1], [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6]],\n",
       "           names=['cterm_aa', 'observable'])"
      ]
     },
     "execution_count": 292,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonGroupEnrichDf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxonGroupEnrichFilename = '{}_enrichment_COG.csv'.format(taxonSuffix)\n",
    "taxonGroupEnrichDf.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / taxonGroupEnrichFilename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(str(eggNOGPath / 'COG_functional_categories.txt')) as f:\n",
    "    text = f.read()\n",
    "    COGCategoriesDict = dict(re.findall(r'\\[([A-Z])\\]\\s+(.*)\\s+[$\\n]', text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'J': 'Translation, ribosomal structure and biogenesis',\n",
       " 'A': 'RNA processing and modification',\n",
       " 'K': 'Transcription',\n",
       " 'L': 'Replication, recombination and repair',\n",
       " 'B': 'Chromatin structure and dynamics ',\n",
       " 'D': 'Cell cycle control, cell division, chromosome partitioning',\n",
       " 'Y': 'Nuclear structure',\n",
       " 'V': 'Defense mechanisms',\n",
       " 'T': 'Signal transduction mechanisms',\n",
       " 'M': 'Cell wall/membrane/envelope biogenesis',\n",
       " 'N': 'Cell motility',\n",
       " 'Z': 'Cytoskeleton',\n",
       " 'W': 'Extracellular structures',\n",
       " 'U': 'Intracellular trafficking, secretion, and vesicular transport',\n",
       " 'O': 'Posttranslational modification, protein turnover, chaperones ',\n",
       " 'C': 'Energy production and conversion',\n",
       " 'G': 'Carbohydrate transport and metabolism',\n",
       " 'E': 'Amino acid transport and metabolism',\n",
       " 'F': 'Nucleotide transport and metabolism',\n",
       " 'H': 'Coenzyme transport and metabolism',\n",
       " 'I': 'Lipid transport and metabolism',\n",
       " 'P': 'Inorganic ion transport and metabolism',\n",
       " 'Q': 'Secondary metabolites biosynthesis, transport and catabolism ',\n",
       " 'R': 'General function prediction only',\n",
       " 'S': 'Function unknown'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COGCategoriesDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Stack the dataframe for c-terminal AA columns\n",
    "taxonGroupEnrichStackedDf = taxonGroupEnrichDf.copy()\n",
    "taxonGroupEnrichStackedDf = taxonGroupEnrichStackedDf.stack(level=['cterm_aa']).reset_index()\n",
    "taxonGroupEnrichStackedDf.columns.name = ''\n",
    "taxonGroupEnrichStackedDf.sort_values(['cterm_aa', groupName + '_index'], inplace=True)\n",
    "\n",
    "# Add COG category description\n",
    "taxonGroupEnrichStackedDf['COG_cat_description'] = \\\n",
    "    taxonGroupEnrichStackedDf.apply(lambda x:\n",
    "                                    COGCategoriesDict.get(x['COGGroup_index'], None),\n",
    "                                    axis=1)\n",
    "\n",
    "# Select only enriched results\n",
    "taxonGroupEnrichStackedDf = taxonGroupEnrichStackedDf[taxonGroupEnrichStackedDf[enrichmentTestcolumnName] == True]\n",
    "\n",
    "# Change column order\n",
    "taxonGroupEnrichStackedDf = taxonGroupEnrichStackedDf[['taxon_id', 'COGGroup_index', 'COG_cat_description', 'cterm_aa',\n",
    "       'count_NOT_cterm_NOT_in_group', 'count_NOT_cterm_in_group',\n",
    "       'count_cterm_NOT_in_group', 'count_cterm_in_group', 'odds_ratio',\n",
    "       'pvalue_fisher',\n",
    "       enrichmentTestcolumnName]]\n",
    "\n",
    "taxonGroupEnrichStackedFilename = '{}_enrichment_COG_stacked.csv'.format(taxonSuffix)\n",
    "taxonGroupEnrichStackedDf.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / taxonGroupEnrichStackedFilename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>COGGroup_index</th>\n",
       "      <th>COG_cat_description</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>544448</td>\n",
       "      <td>J</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>H</td>\n",
       "      <td>17165</td>\n",
       "      <td>216</td>\n",
       "      <td>2720</td>\n",
       "      <td>14</td>\n",
       "      <td>0.409024</td>\n",
       "      <td>0.000439059</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>544448</td>\n",
       "      <td>E</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>R</td>\n",
       "      <td>18751</td>\n",
       "      <td>767</td>\n",
       "      <td>552</td>\n",
       "      <td>45</td>\n",
       "      <td>1.99298</td>\n",
       "      <td>7.44082e-05</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>544448</td>\n",
       "      <td>J</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>R</td>\n",
       "      <td>16723</td>\n",
       "      <td>658</td>\n",
       "      <td>2580</td>\n",
       "      <td>154</td>\n",
       "      <td>1.51701</td>\n",
       "      <td>1.3368e-05</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    taxon_id COGGroup_index                              COG_cat_description  \\\n",
       "14    544448              J  Translation, ribosomal structure and biogenesis   \n",
       "5     544448              E              Amino acid transport and metabolism   \n",
       "15    544448              J  Translation, ribosomal structure and biogenesis   \n",
       "\n",
       "   cterm_aa count_NOT_cterm_NOT_in_group count_NOT_cterm_in_group  \\\n",
       "14        H                        17165                      216   \n",
       "5         R                        18751                      767   \n",
       "15        R                        16723                      658   \n",
       "\n",
       "   count_cterm_NOT_in_group count_cterm_in_group odds_ratio pvalue_fisher  \\\n",
       "14                     2720                   14   0.409024   0.000439059   \n",
       "5                       552                   45    1.99298   7.44082e-05   \n",
       "15                     2580                  154    1.51701    1.3368e-05   \n",
       "\n",
       "   reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01  \n",
       "14                                               True                 \n",
       "5                                                True                 \n",
       "15                                               True                 "
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonGroupEnrichStackedDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# taxonGroupEnrichFilename = 'enrichment_{}_taxonId_{:d}_{}_{}'\\\n",
    "#                            .format(groupName, int(taxonId), taxonRank, taxonName)\n",
    "# taxonGroupEnrichDf = pd.read_csv(str(eggNOGPath / 'Analysis_COG_enrichment' / (taxonGroupEnrichFilename + '.csv')),\n",
    "#                                  header=[0,1], index_col=[0,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def enrichment_analysis_cterminal_add_protein_list(taxonAllCDSDf, taxonGroupEnrichDf, groupCol, groupName):\n",
    "   \n",
    "    # Convert the enrichment dataframe to a stacked version and\n",
    "    # Include list of proteins for positive (enriched) orthoGroups into the dataframe.\n",
    "\n",
    "    # We will include list of proteins for specific orthoGroups into the dataframe,\n",
    "    # and these will differ for each ctermAA.\n",
    "    # Therefore, the best solution is to stack all the columns for ctermAA.\n",
    "    taxonGroupEnrichStackedDf = taxonGroupEnrichDf.copy()\n",
    "    taxonGroupEnrichStackedDf = taxonGroupEnrichStackedDf.stack(level=['cterm_aa']).reset_index()\n",
    "    taxonGroupEnrichStackedDf.columns.name = ''\n",
    "    taxonGroupEnrichStackedDf.sort_values(['cterm_aa', groupName + '_index'], inplace=True)\n",
    "\n",
    "    taxonAllCDSPositiveDfList = []\n",
    "\n",
    "    for ctermAA in aaTable:\n",
    "\n",
    "        # Get list of positive (enriched) orthoGroups\n",
    "        enrichmentTestDf = pd.DataFrame(taxonGroupEnrichDf.loc[(taxonId), (ctermAA, enrichmentTestcolumnName)])\n",
    "        enrichmentTestDf.columns = ['reject']\n",
    "\n",
    "        # Select proteins that are in a positive (enriched) orthoGroup\n",
    "        taxonAllCDSPositiveDf = pd.merge(taxonAllCDSDf, enrichmentTestDf,\n",
    "                                         left_on=groupCol, right_index=True, how='left')\n",
    "        taxonAllCDSPositiveDf = taxonAllCDSPositiveDf[ taxonAllCDSPositiveDf['reject'] == True ]\n",
    "        taxonAllCDSPositiveDf.drop('reject', axis=1, inplace=True)\n",
    "        taxonAllCDSPositiveDf.rename(columns={'taxon_taxid':'taxon_id'}, inplace=True)\n",
    "\n",
    "        taxonAllCDSPositiveDf = taxonAllCDSPositiveDf.reset_index(drop=True)\n",
    "        taxonAllCDSPositiveDf['cterm_aa'] = ctermAA\n",
    "        # In order to be able to have a unique multiindex (later), we assign an increasing\n",
    "        # index to proteins inside groups.\n",
    "        taxonAllCDSPositiveDf['protein_index'] = taxonAllCDSPositiveDf.groupby(groupCol).cumcount()\n",
    "        taxonAllCDSPositiveDf.sort_values([groupCol, 'protein_index'], inplace=True)\n",
    "\n",
    "        taxonAllCDSPositiveDfList.append(taxonAllCDSPositiveDf)\n",
    "\n",
    "\n",
    "    taxonAllCDSPositiveDf = pd.concat(taxonAllCDSPositiveDfList)\n",
    "\n",
    "    # Merge the list of proteins with the enrichment dataframe\n",
    "    print(taxonGroupEnrichStackedDf.columns)\n",
    "    print(taxonAllCDSPositiveDf.columns)\n",
    "    taxonGroupEnrichStackedDf = pd.merge(taxonGroupEnrichStackedDf,\n",
    "                                         taxonAllCDSPositiveDf.rename(columns={groupCol:groupName + '_index'}),\n",
    "                                         on=['taxon_id', 'cterm_aa', groupName + '_index'], how='outer'\n",
    "                                         ).sort_values(['taxon_id', 'cterm_aa', groupName + '_index', 'protein_index'])\n",
    "    return taxonGroupEnrichStackedDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# taxonGroupEnrichStackedDf = enrichment_analysis_cterminal_add_protein_list(taxonAllCDSDf, taxonGroupEnrichDf,\n",
    "#                                                                            groupCol, groupName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# taxonGroupEnrichStackedFilename = 'enrich_{}_taxonRank_{}_taxonName_{}_taxonId_{:0>10d}_stacked'\\\n",
    "#                                   .format(groupName, taxonRank, taxonName, int(taxonId))\n",
    "# taxonGroupEnrichStackedDf.to_csv(str(OMAFolder / 'TaxonOrthoGroupEnrichment' / (taxonGroupEnrichStackedFilename + '.csv')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functional enrichment analysis, orthologous groups"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will consider the \"Best matching Orthologous Groups\" as given by the HMM search and fine-grained orthology assignment by the eggmapper software."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 806,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7428 / 46366 = 0.160203597464\n"
     ]
    }
   ],
   "source": [
    "print(taxonAllCDSDf[groupCol].isnull().sum(), \"/\", len(taxonAllCDSDf), \"=\", taxonAllCDSDf[groupCol].isnull().sum()/len(taxonAllCDSDf))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "About 33% of protein do not have any OG annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 492,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "groupCol = 'bestOG'\n",
    "groupName = 'orthologous_group'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 807,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12665"
      ]
     },
     "execution_count": 807,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract list of COG categories\n",
    "orthogroupList = list(taxonAllCDSDf[groupCol].dropna().unique())\n",
    "len(orthogroupList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 821,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber\n",
      "\n",
      "\n",
      "######## END ANALYSIS JOB ########\n",
      "####### R\n",
      "nProteinSubset 7690 nProteinNotSubset 38676\n",
      "groupIndex 05C0W NjA 5 NjNotA 7685\n",
      "groupIndex 05CX5 NjA 1 NjNotA 7689\n",
      "####### H\n",
      "nProteinSubset 1016 nProteinNotSubset 45350\n",
      "groupIndex 05C0W NjA 2 NjNotA 1014\n",
      "groupIndex 05CX5 NjA 0 NjNotA 1016\n"
     ]
    }
   ],
   "source": [
    "taxonEnrichOrthoGroupDf = \\\n",
    "    enrichment_analysis_cterminal_subset(taxonAllCDSDf, taxonId, groupCol, groupName, orthogroupList[:2], aaTable[:2],\n",
    "                                         alpha_family_wise_error_rate, enrichmentTestcolumnName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 822,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th colspan=\"7\" halign=\"left\">H</th>\n",
       "      <th colspan=\"7\" halign=\"left\">R</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>observable</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxon_id</th>\n",
       "      <th>orthologous_group_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">1937961</th>\n",
       "      <th>05C0W</th>\n",
       "      <td>45316</td>\n",
       "      <td>34</td>\n",
       "      <td>1014</td>\n",
       "      <td>2</td>\n",
       "      <td>2.62884</td>\n",
       "      <td>0.186312</td>\n",
       "      <td>False</td>\n",
       "      <td>38645</td>\n",
       "      <td>31</td>\n",
       "      <td>7685</td>\n",
       "      <td>5</td>\n",
       "      <td>0.811069</td>\n",
       "      <td>0.824074</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>05CX5</th>\n",
       "      <td>45342</td>\n",
       "      <td>8</td>\n",
       "      <td>1016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>38669</td>\n",
       "      <td>7</td>\n",
       "      <td>7689</td>\n",
       "      <td>1</td>\n",
       "      <td>0.718448</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "cterm_aa                                                    H  \\\n",
       "observable                       count_NOT_cterm_NOT_in_group   \n",
       "taxon_id orthologous_group_index                                \n",
       "1937961  05C0W                                          45316   \n",
       "         05CX5                                          45342   \n",
       "\n",
       "cterm_aa                                                   \\\n",
       "observable                       count_NOT_cterm_in_group   \n",
       "taxon_id orthologous_group_index                            \n",
       "1937961  05C0W                                         34   \n",
       "         05CX5                                          8   \n",
       "\n",
       "cterm_aa                                                   \\\n",
       "observable                       count_cterm_NOT_in_group   \n",
       "taxon_id orthologous_group_index                            \n",
       "1937961  05C0W                                       1014   \n",
       "         05CX5                                       1016   \n",
       "\n",
       "cterm_aa                                                          \\\n",
       "observable                       count_cterm_in_group odds_ratio   \n",
       "taxon_id orthologous_group_index                                   \n",
       "1937961  05C0W                                      2    2.62884   \n",
       "         05CX5                                      0          0   \n",
       "\n",
       "cterm_aa                                        \\\n",
       "observable                       pvalue_fisher   \n",
       "taxon_id orthologous_group_index                 \n",
       "1937961  05C0W                        0.186312   \n",
       "         05CX5                               1   \n",
       "\n",
       "cterm_aa                                                                                            \\\n",
       "observable                       reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01   \n",
       "taxon_id orthologous_group_index                                                                     \n",
       "1937961  05C0W                                                                False                  \n",
       "         05CX5                                                                False                  \n",
       "\n",
       "cterm_aa                                                    R  \\\n",
       "observable                       count_NOT_cterm_NOT_in_group   \n",
       "taxon_id orthologous_group_index                                \n",
       "1937961  05C0W                                          38645   \n",
       "         05CX5                                          38669   \n",
       "\n",
       "cterm_aa                                                   \\\n",
       "observable                       count_NOT_cterm_in_group   \n",
       "taxon_id orthologous_group_index                            \n",
       "1937961  05C0W                                         31   \n",
       "         05CX5                                          7   \n",
       "\n",
       "cterm_aa                                                   \\\n",
       "observable                       count_cterm_NOT_in_group   \n",
       "taxon_id orthologous_group_index                            \n",
       "1937961  05C0W                                       7685   \n",
       "         05CX5                                       7689   \n",
       "\n",
       "cterm_aa                                                          \\\n",
       "observable                       count_cterm_in_group odds_ratio   \n",
       "taxon_id orthologous_group_index                                   \n",
       "1937961  05C0W                                      5   0.811069   \n",
       "         05CX5                                      1   0.718448   \n",
       "\n",
       "cterm_aa                                        \\\n",
       "observable                       pvalue_fisher   \n",
       "taxon_id orthologous_group_index                 \n",
       "1937961  05C0W                        0.824074   \n",
       "         05CX5                               1   \n",
       "\n",
       "cterm_aa                                                                                            \n",
       "observable                       reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01  \n",
       "taxon_id orthologous_group_index                                                                    \n",
       "1937961  05C0W                                                                False                 \n",
       "         05CX5                                                                False                 "
      ]
     },
     "execution_count": 822,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonEnrichOrthoGroupDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxonEnrichOrthoGroupFilename = '{}_enrichment_orthogroup.csv'.format(taxonSuffix)\n",
    "taxonEnrichOrthoGroupDf.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / taxonEnrichOrthoGroupFilename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bestOG</th>\n",
       "      <th>predicted_gene_name</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>COG cat</th>\n",
       "      <th>eggNOG annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0000B</td>\n",
       "      <td>ARGD</td>\n",
       "      <td>map00260,map00300,map00330,map01100,map01110,m...</td>\n",
       "      <td>E</td>\n",
       "      <td>acetylornithine and succinylornithine aminotra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0000E</td>\n",
       "      <td>ACCC</td>\n",
       "      <td>map00061,map00253,map00620,map00640,map00720,m...</td>\n",
       "      <td>I</td>\n",
       "      <td>acetyl-CoA carboxylase, biotin carboxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0000I</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>Dienelactone hydrolase family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0000K</td>\n",
       "      <td>CLPP</td>\n",
       "      <td>map04112</td>\n",
       "      <td>O</td>\n",
       "      <td>Cleaves peptides in various proteins in a proc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0000P</td>\n",
       "      <td>RLUB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>J</td>\n",
       "      <td>Pseudouridine synthase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  bestOG predicted_gene_name  \\\n",
       "0  0000B                ARGD   \n",
       "1  0000E                ACCC   \n",
       "2  0000I                 NaN   \n",
       "3  0000K                CLPP   \n",
       "4  0000P                RLUB   \n",
       "\n",
       "                                       KEGG_pathways COG cat  \\\n",
       "0  map00260,map00300,map00330,map01100,map01110,m...       E   \n",
       "1  map00061,map00253,map00620,map00640,map00720,m...       I   \n",
       "2                                                NaN       E   \n",
       "3                                           map04112       O   \n",
       "4                                                NaN       J   \n",
       "\n",
       "                                        eggNOG annot  \n",
       "0  acetylornithine and succinylornithine aminotra...  \n",
       "1         acetyl-CoA carboxylase, biotin carboxylase  \n",
       "2                      Dienelactone hydrolase family  \n",
       "3  Cleaves peptides in various proteins in a proc...  \n",
       "4                             Pseudouridine synthase  "
      ]
     },
     "execution_count": 496,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggnogMapperOrthogroupAnnotationDf = pd.read_csv(str(analysisEggnogPath / 'eggnogMapperOrthogroupAnnotationDf.csv'))\n",
    "eggnogMapperOrthogroupAnnotationDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th>orthologous_group_index</th>\n",
       "      <th>predicted_gene_name</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>COG cat</th>\n",
       "      <th>eggNOG annot</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1937961</td>\n",
       "      <td>H</td>\n",
       "      <td>07RBZ</td>\n",
       "      <td>SULP|YBAR|RV1739C|HVST</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>sulfate transporter</td>\n",
       "      <td>45350</td>\n",
       "      <td>1012</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>inf</td>\n",
       "      <td>2.29227e-07</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    taxon_id cterm_aa orthologous_group_index     predicted_gene_name  \\\n",
       "24   1937961        H                   07RBZ  SULP|YBAR|RV1739C|HVST   \n",
       "\n",
       "   KEGG_pathways COG cat         eggNOG annot count_NOT_cterm_NOT_in_group  \\\n",
       "24           NaN       P  sulfate transporter                        45350   \n",
       "\n",
       "   count_NOT_cterm_in_group count_cterm_NOT_in_group count_cterm_in_group  \\\n",
       "24                     1012                        0                    4   \n",
       "\n",
       "   odds_ratio pvalue_fisher  \\\n",
       "24        inf   2.29227e-07   \n",
       "\n",
       "   reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01  \n",
       "24                                               True                 "
      ]
     },
     "execution_count": 499,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stack the dataframe for c-terminal AA columns\n",
    "taxonEnrichOrthoGroupStackedDf = taxonEnrichOrthoGroupDf.copy()\n",
    "taxonEnrichOrthoGroupStackedDf = taxonEnrichOrthoGroupStackedDf.stack(level=['cterm_aa']).reset_index()\n",
    "taxonEnrichOrthoGroupStackedDf.columns.name = ''\n",
    "taxonEnrichOrthoGroupStackedDf.sort_values(['cterm_aa', groupName + '_index'], inplace=True)\n",
    "\n",
    "# Add orthogroup annotation\n",
    "taxonEnrichOrthoGroupStackedDf = \\\n",
    "    pd.merge(taxonEnrichOrthoGroupStackedDf, eggnogMapperOrthogroupAnnotationDf, left_on='orthologous_group_index',\n",
    "             right_on='bestOG', how='left')\n",
    "\n",
    "# Select only enriched results\n",
    "taxonEnrichOrthoGroupStackedDf = taxonEnrichOrthoGroupStackedDf[taxonEnrichOrthoGroupStackedDf[enrichmentTestcolumnName] == True]\n",
    "\n",
    "# Change column order\n",
    "taxonEnrichOrthoGroupStackedDf = taxonEnrichOrthoGroupStackedDf[\n",
    "    ['taxon_id', 'cterm_aa', 'orthologous_group_index', 'predicted_gene_name', 'KEGG_pathways', 'COG cat', 'eggNOG annot',\n",
    "     'count_NOT_cterm_NOT_in_group', 'count_NOT_cterm_in_group',\n",
    "     'count_cterm_NOT_in_group', 'count_cterm_in_group', 'odds_ratio',\n",
    "     'pvalue_fisher',\n",
    "     enrichmentTestcolumnName]]\n",
    "\n",
    "taxonEnrichOrthoGroupStackedDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxonEnrichOrthoGroupStackedFilename = '{}_enrichment_orthogroup_stacked.csv'.format(taxonSuffix)\n",
    "taxonEnrichOrthoGroupStackedDf.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / taxonEnrichOrthoGroupStackedFilename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th>orthologous_group_index</th>\n",
       "      <th>predicted_gene_name</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>COG cat</th>\n",
       "      <th>eggNOG annot</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2299</th>\n",
       "      <td>544448</td>\n",
       "      <td>R</td>\n",
       "      <td>07RPZ</td>\n",
       "      <td>OPPC|OPPC-1</td>\n",
       "      <td>map02010</td>\n",
       "      <td>E</td>\n",
       "      <td>transport system permease protein</td>\n",
       "      <td>19298</td>\n",
       "      <td>800</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>57.894</td>\n",
       "      <td>8.8859e-14</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>544448</td>\n",
       "      <td>R</td>\n",
       "      <td>07SI6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>map02010</td>\n",
       "      <td>E</td>\n",
       "      <td>transport system permease protein</td>\n",
       "      <td>19302</td>\n",
       "      <td>808</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>95.5545</td>\n",
       "      <td>1.27597e-05</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2713</th>\n",
       "      <td>544448</td>\n",
       "      <td>R</td>\n",
       "      <td>080EB</td>\n",
       "      <td>RBSC-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>ABC transporter, permease</td>\n",
       "      <td>19302</td>\n",
       "      <td>795</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>412.747</td>\n",
       "      <td>2.95835e-23</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2761</th>\n",
       "      <td>544448</td>\n",
       "      <td>R</td>\n",
       "      <td>0810X</td>\n",
       "      <td>CODA</td>\n",
       "      <td>map00240,map00330,map00791,map01100,map01120</td>\n",
       "      <td>F, J</td>\n",
       "      <td>deaminase</td>\n",
       "      <td>19301</td>\n",
       "      <td>807</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>59.7924</td>\n",
       "      <td>2.07839e-06</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2917</th>\n",
       "      <td>544448</td>\n",
       "      <td>R</td>\n",
       "      <td>084BG</td>\n",
       "      <td>RPSU</td>\n",
       "      <td>map03010</td>\n",
       "      <td>J</td>\n",
       "      <td>30S ribosomal protein S21</td>\n",
       "      <td>19303</td>\n",
       "      <td>805</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.7039e-10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      taxon_id cterm_aa orthologous_group_index predicted_gene_name  \\\n",
       "2299    544448        R                   07RPZ         OPPC|OPPC-1   \n",
       "2377    544448        R                   07SI6                 NaN   \n",
       "2713    544448        R                   080EB              RBSC-2   \n",
       "2761    544448        R                   0810X                CODA   \n",
       "2917    544448        R                   084BG                RPSU   \n",
       "\n",
       "                                     KEGG_pathways COG cat  \\\n",
       "2299                                      map02010       E   \n",
       "2377                                      map02010       E   \n",
       "2713                                           NaN       G   \n",
       "2761  map00240,map00330,map00791,map01100,map01120    F, J   \n",
       "2917                                      map03010       J   \n",
       "\n",
       "                           eggNOG annot count_NOT_cterm_NOT_in_group  \\\n",
       "2299  transport system permease protein                        19298   \n",
       "2377  transport system permease protein                        19302   \n",
       "2713          ABC transporter, permease                        19302   \n",
       "2761                          deaminase                        19301   \n",
       "2917          30S ribosomal protein S21                        19303   \n",
       "\n",
       "     count_NOT_cterm_in_group count_cterm_NOT_in_group count_cterm_in_group  \\\n",
       "2299                      800                        5                   12   \n",
       "2377                      808                        1                    4   \n",
       "2713                      795                        1                   17   \n",
       "2761                      807                        2                    5   \n",
       "2917                      805                        0                    7   \n",
       "\n",
       "     odds_ratio pvalue_fisher  \\\n",
       "2299     57.894    8.8859e-14   \n",
       "2377    95.5545   1.27597e-05   \n",
       "2713    412.747   2.95835e-23   \n",
       "2761    59.7924   2.07839e-06   \n",
       "2917        inf    1.7039e-10   \n",
       "\n",
       "     reject_null_multiple_tests_correction_Benjamini-Hochberg_FDR_0.01  \n",
       "2299                                               True                 \n",
       "2377                                               True                 \n",
       "2713                                               True                 \n",
       "2761                                               True                 \n",
       "2917                                               True                 "
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonEnrichOrthoGroupStackedDf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run enrichment computation for all taxa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichment analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### C-terminal biases for COG groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  rank_index rank_name  taxon_id  taxon_name  n_genomes_in_refseq  \\\n",
       "2      5           1    phylum    1239.0  Firmicutes                  259   \n",
       "\n",
       "   n_genomes_in_refseq_eggnog_common_species  \n",
       "2                                        170  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaEggnogDf = pd.read_csv(str(analysisEggnogPath / 'taxaEggnogDf.csv'), index_col=0)\n",
    "taxaEggnogDf.reset_index(inplace=True)\n",
    "taxaEggnogDf[taxaEggnogDf['taxon_name'] == 'Firmicutes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index                                                   0\n",
       "rank_index                                              0\n",
       "rank_name                                    superkingdom\n",
       "taxon_id                                                2\n",
       "taxon_name                                       Bacteria\n",
       "n_genomes_in_refseq                                  1582\n",
       "n_genomes_in_refseq_eggnog_common_species             949\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaEggnogDf[taxaEggnogDf['taxon_name'] == 'Bacteria']\n",
    "taxaEggnogDf.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index                                                 5\n",
       "rank_index                                            1\n",
       "rank_name                                        phylum\n",
       "taxon_id                                           1239\n",
       "taxon_name                                   Firmicutes\n",
       "n_genomes_in_refseq                                 259\n",
       "n_genomes_in_refseq_eggnog_common_species           170\n",
       "Name: 2, dtype: object"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaEggnogDf.iloc[2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import enrichment results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ Import results for each taxon\n",
    "+ Concatenate dataframes\n",
    "+ Write to disk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1429\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0           0  superkingdom       2.0             Bacteria   \n",
       "1           1        phylum    1224.0       Proteobacteria   \n",
       "2           1        phylum    1239.0           Firmicutes   \n",
       "3           3         class    1236.0  Gammaproteobacteria   \n",
       "4           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "   n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                 1582                                        949  \n",
       "1                  697                                        407  \n",
       "2                  259                                        170  \n",
       "3                  292                                        162  \n",
       "4                  226                                        126  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaEggnogDf = pd.read_csv(str(analysisEggnogPath / 'taxaEggnogDf.csv'), index_col=0).reset_index(drop=True)\n",
    "print(len(taxaEggnogDf))\n",
    "taxaEggnogDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "taxonGroupEnrichDfList = []\n",
    "taxonGroupEnrichStackedDfList = []\n",
    "taxonEnrichOrthoGroupDfList = []\n",
    "taxonEnrichOrthoGroupStackedDfList = []\n",
    "OGerrorList = []\n",
    "OGStackederrorList = []\n",
    "\n",
    "for i, taxaGroup in taxaEggnogDf.iterrows():\n",
    "    \n",
    "    print(\"### \", i, \"/\", len(taxaEggnogDf), \"taxonGroup:\", taxaGroup)\n",
    "    iTaxon = i\n",
    "    taxonRank = taxonGroup['rank_name']\n",
    "    taxonName = taxonGroup['taxon_name']\n",
    "    taxonId = int(taxonGroup['taxon_id'])\n",
    "    \n",
    "    taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "    \n",
    "    filePath = analysisEggnogPath / 'Analysis_COG_enrichment' / '{}_enrichment_COG.csv'.format(taxonSuffix)\n",
    "    if filePath.is_file():\n",
    "        taxonGroupEnrichDf = pd.read_csv(str(filePath), header=[0,1], index_col=[0,1])\n",
    "        taxonGroupEnrichDfList.append(taxonGroupEnrichDf)\n",
    "    else:\n",
    "        print(\"ERROR: file\", str(filePath.name), \"not found.\")\n",
    "        \n",
    "    filePath = analysisEggnogPath / 'Analysis_COG_enrichment' / '{}_enrichment_COG_stacked.csv'.format(taxonSuffix)\n",
    "    if filePath.is_file():\n",
    "        taxonGroupEnrichStackedDf = pd.read_csv(str(filePath), index_col=0)\n",
    "        taxonGroupEnrichStackedDfList.append(taxonGroupEnrichStackedDf)\n",
    "    else:\n",
    "        print(\"ERROR: file\", str(filePath.name), \"not found.\")\n",
    "\n",
    "    filePath = analysisEggnogPath / 'Analysis_COG_enrichment' / '{}_enrichment_orthogroup.csv'.format(taxonSuffix)\n",
    "    if filePath.is_file():\n",
    "        taxonEnrichOrthoGroupDf = pd.read_csv(str(filePath), header=[0,1], index_col=[0,1])\n",
    "        taxonEnrichOrthoGroupDfList.append(taxonEnrichOrthoGroupDf)\n",
    "    else:\n",
    "        print(\"ERROR: file\", str(filePath.name), \"not found.\")\n",
    "        OGerrorList.append(taxaGroup)\n",
    "        \n",
    "    filePath = analysisEggnogPath / 'Analysis_COG_enrichment' / '{}_enrichment_orthogroup_stacked.csv'.format(taxonSuffix)\n",
    "    if filePath.is_file():\n",
    "        taxonEnrichOrthoGroupStackedDf = pd.read_csv(str(filePath), index_col=0)\n",
    "        taxonEnrichOrthoGroupStackedDfList.append(taxonEnrichOrthoGroupStackedDf)\n",
    "    else:\n",
    "        print(\"ERROR: file\", str(filePath.name), \"not found.\")\n",
    "        OGStackederrorList.append(taxaGroup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxonGroupEnrichDfList: 434 / 434\n",
      "taxonGroupEnrichStackedDfList: 434 / 434\n",
      "taxonEnrichOrthoGroupDfList: 433 / 434\n",
      "taxonEnrichOrthoGroupStackedDfList: 433 / 434\n"
     ]
    }
   ],
   "source": [
    "taxonGroupEnrichDfFull = pd.concat(taxonGroupEnrichDfList)\n",
    "print(\"taxonGroupEnrichDfList:\", len(taxonGroupEnrichDfList), \"/\", len(taxaGroupCommonEggNOGRefSeqList))\n",
    "\n",
    "taxonGroupEnrichStackedDfFull = pd.concat(taxonGroupEnrichStackedDfList)\n",
    "print(\"taxonGroupEnrichStackedDfList:\", len(taxonGroupEnrichStackedDfList), \"/\", len(taxaGroupCommonEggNOGRefSeqList))\n",
    "\n",
    "taxonEnrichOrthoGroupDfFull = pd.concat(taxonEnrichOrthoGroupDfList)\n",
    "print(\"taxonEnrichOrthoGroupDfList:\", len(taxonEnrichOrthoGroupDfList), \"/\", len(taxaGroupCommonEggNOGRefSeqList))\n",
    "\n",
    "taxonEnrichOrthoGroupStackedDfFull = pd.concat(taxonEnrichOrthoGroupStackedDfList)\n",
    "print(\"taxonEnrichOrthoGroupStackedDfList:\", len(taxonEnrichOrthoGroupStackedDfList), \"/\", len(taxaGroupCommonEggNOGRefSeqList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxonGroupEnrichDfFull.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / 'taxonGroupEnrichDfFull.csv'))\n",
    "taxonGroupEnrichStackedDfFull.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / 'taxonGroupEnrichStackedDfFull.csv'))\n",
    "taxonEnrichOrthoGroupDfFull.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / 'taxonEnrichOrthoGroupDfFull.csv'))\n",
    "taxonEnrichOrthoGroupStackedDfFull.to_csv(str(analysisEggnogPath / 'Analysis_COG_enrichment' / 'taxonEnrichOrthoGroupStackedDfFull.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2017.02.24 09:50\n",
    "\n",
    "We get, at the moment,\n",
    "```\n",
    "taxonGroupEnrichDfList: 434 / 434\n",
    "taxonGroupEnrichStackedDfList: 434 / 434\n",
    "taxonEnrichOrthoGroupDfList: 421 / 434\n",
    "taxonEnrichOrthoGroupStackedDfList: 421 / 434\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import c-terminal bias results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%aimport analysis_summarize_results\n",
    "from analysis_summarize_results import gather_cterm_bias_taxon_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "499\n",
      "499\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[[3, 0, 'phylum', 'Proteobacteria', 1224.0], 691],\n",
       " [[143, 1, 'class', 'Gammaproteobacteria', 1236.0], 288],\n",
       " [[4, 0, 'phylum', 'Firmicutes', 1239.0], 259],\n",
       " [[32, 0, 'phylum', 'Actinobacteria', 201174.0], 226],\n",
       " [[144, 1, 'class', 'Actinobacteria', 1760.0], 211]]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(str(taxonomyPath / 'taxaGroupsInRefSeqNGenomes.json'), 'r') as f:\n",
    "    taxaGroupsInRefSeqNGenomes = json.load(f)\n",
    "\n",
    "taxaGroupsInRefSeqNGenomes = [g for g in taxaGroupsInRefSeqNGenomes if g[1] > 0 and g[0][2] != 'genus']\n",
    "taxaGroupsInRefSeq = [g[0] for g in taxaGroupsInRefSeqNGenomes if g[1] > 0 and g[0][2] != 'genus']\n",
    "print(len(taxaGroupsInRefSeq))\n",
    "print(len(taxaGroupsInRefSeqNGenomes))\n",
    "taxaGroupsInRefSeqNGenomes[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing C-terminal bias results for all taxon groups.\n",
      "###  1 / 499 taxonGroup: [3, 0, 'phylum', 'Proteobacteria', 1224.0]\n",
      "###  2 / 499 taxonGroup: [143, 1, 'class', 'Gammaproteobacteria', 1236.0]\n",
      "###  3 / 499 taxonGroup: [4, 0, 'phylum', 'Firmicutes', 1239.0]\n",
      "###  4 / 499 taxonGroup: [32, 0, 'phylum', 'Actinobacteria', 201174.0]\n",
      "###  5 / 499 taxonGroup: [144, 1, 'class', 'Actinobacteria', 1760.0]\n",
      "###  6 / 499 taxonGroup: [145, 1, 'class', 'Alphaproteobacteria', 28211.0]\n",
      "###  7 / 499 taxonGroup: [156, 1, 'class', 'Bacilli', 91061.0]\n",
      "###  8 / 499 taxonGroup: [0, 0, 'phylum', 'Bacteroidetes', 976.0]\n",
      "###  9 / 499 taxonGroup: [146, 1, 'class', 'Betaproteobacteria', 28216.0]\n",
      "###  10 / 499 taxonGroup: [161, 1, 'class', 'Clostridia', 186801.0]\n",
      "###  11 / 499 taxonGroup: [231, 2, 'order', 'Bacillales', 1385.0]\n",
      "###  12 / 499 taxonGroup: [263, 2, 'order', 'Enterobacterales', 91347.0]\n",
      "###  13 / 499 taxonGroup: [226, 2, 'order', 'Rhizobiales', 356.0]\n",
      "###  14 / 499 taxonGroup: [255, 2, 'order', 'Corynebacteriales', 85007.0]\n",
      "###  15 / 499 taxonGroup: [280, 2, 'order', 'Clostridiales', 186802.0]\n",
      "###  16 / 499 taxonGroup: [281, 2, 'order', 'Lactobacillales', 186826.0]\n",
      "###  17 / 499 taxonGroup: [248, 2, 'order', 'Burkholderiales', 80840.0]\n",
      "###  18 / 499 taxonGroup: [2, 0, 'phylum', 'Cyanobacteria', 1117.0]\n",
      "###  19 / 499 taxonGroup: [147, 1, 'class', 'Deltaproteobacteria', 28221.0]\n",
      "###  20 / 499 taxonGroup: [44, 0, 'phylum', 'Tenericutes', 544448.0]\n",
      "###  21 / 499 taxonGroup: [149, 1, 'class', 'Mollicutes', 31969.0]\n",
      "###  22 / 499 taxonGroup: [157, 1, 'class', 'Flavobacteriia', 117743.0]\n",
      "###  23 / 499 taxonGroup: [289, 2, 'order', 'Flavobacteriales', 200644.0]\n",
      "###  24 / 499 taxonGroup: [254, 2, 'order', 'Micrococcales', 85006.0]\n",
      "###  25 / 499 taxonGroup: [566, 3, 'family', 'Bacillaceae', 186817.0]\n",
      "###  26 / 499 taxonGroup: [247, 2, 'order', 'Pseudomonadales', 72274.0]\n",
      "###  27 / 499 taxonGroup: [479, 3, 'family', 'Flavobacteriaceae', 49546.0]\n",
      "###  28 / 499 taxonGroup: [148, 1, 'class', 'Epsilonproteobacteria', 29547.0]\n",
      "###  29 / 499 taxonGroup: [272, 2, 'order', 'Alteromonadales', 135622.0]\n",
      "###  30 / 499 taxonGroup: [428, 3, 'family', 'Enterobacteriaceae', 543.0]\n",
      "###  31 / 499 taxonGroup: [447, 3, 'family', 'Corynebacteriaceae', 1653.0]\n",
      "###  32 / 499 taxonGroup: [227, 2, 'order', 'Rickettsiales', 766.0]\n",
      "###  33 / 499 taxonGroup: [233, 2, 'order', 'Mycoplasmatales', 2085.0]\n",
      "###  34 / 499 taxonGroup: [306, 2, 'order', 'Campylobacterales', 213849.0]\n",
      "###  35 / 499 taxonGroup: [454, 3, 'family', 'Mycoplasmataceae', 2092.0]\n",
      "###  36 / 499 taxonGroup: [34, 0, 'phylum', 'Spirochaetes', 203691.0]\n",
      "###  37 / 499 taxonGroup: [172, 1, 'class', 'Spirochaetia', 203692.0]\n",
      "###  38 / 499 taxonGroup: [406, 2, 'order', 'Synechococcales', 1890424.0]\n",
      "###  39 / 499 taxonGroup: [167, 1, 'class', 'Bacteroidia', 200643.0]\n",
      "###  40 / 499 taxonGroup: [276, 2, 'order', 'Bacteroidales', 171549.0]\n",
      "###  41 / 499 taxonGroup: [295, 2, 'order', 'Rhodobacterales', 204455.0]\n",
      "###  42 / 499 taxonGroup: [462, 3, 'family', 'Clostridiaceae', 31979.0]\n",
      "###  43 / 499 taxonGroup: [470, 3, 'family', 'Lactobacillaceae', 33958.0]\n",
      "###  44 / 499 taxonGroup: [549, 3, 'family', 'Pseudomonadaceae', 135621.0]\n",
      "###  45 / 499 taxonGroup: [259, 2, 'order', 'Streptomycetales', 85011.0]\n",
      "###  46 / 499 taxonGroup: [452, 3, 'family', 'Streptomycetaceae', 2062.0]\n",
      "###  47 / 499 taxonGroup: [464, 3, 'family', 'Rhodobacteraceae', 31989.0]\n",
      "###  48 / 499 taxonGroup: [294, 2, 'order', 'Rhodospirillales', 204441.0]\n",
      "###  49 / 499 taxonGroup: [296, 2, 'order', 'Sphingomonadales', 204457.0]\n",
      "###  50 / 499 taxonGroup: [446, 3, 'family', 'Streptococcaceae', 1300.0]\n",
      "###  51 / 499 taxonGroup: [448, 3, 'family', 'Mycobacteriaceae', 1762.0]\n",
      "###  52 / 499 taxonGroup: [225, 2, 'order', 'Spirochaetales', 136.0]\n",
      "###  53 / 499 taxonGroup: [195, 1, 'class', 'Cytophagia', 768503.0]\n",
      "###  54 / 499 taxonGroup: [348, 2, 'order', 'Cytophagales', 768507.0]\n",
      "###  55 / 499 taxonGroup: [496, 3, 'family', 'Rhizobiaceae', 82115.0]\n",
      "###  56 / 499 taxonGroup: [5, 0, 'phylum', 'Deinococcus-Thermus', 1297.0]\n",
      "###  57 / 499 taxonGroup: [164, 1, 'class', 'Deinococci', 188787.0]\n",
      "###  58 / 499 taxonGroup: [271, 2, 'order', 'Oceanospirillales', 135619.0]\n",
      "###  59 / 499 taxonGroup: [445, 3, 'family', 'Micrococcaceae', 1268.0]\n",
      "###  60 / 499 taxonGroup: [488, 3, 'family', 'Campylobacteraceae', 72294.0]\n",
      "###  61 / 499 taxonGroup: [538, 3, 'family', 'Burkholderiaceae', 119060.0]\n",
      "###  62 / 499 taxonGroup: [28, 0, 'phylum', 'Thermotogae', 200918.0]\n",
      "###  63 / 499 taxonGroup: [163, 1, 'class', 'Thermotogae', 188708.0]\n",
      "###  64 / 499 taxonGroup: [267, 2, 'order', 'Chromatiales', 135613.0]\n",
      "###  65 / 499 taxonGroup: [570, 3, 'family', 'Paenibacillaceae', 186822.0]\n",
      "###  66 / 499 taxonGroup: [848, 3, 'family', 'Erwiniaceae', 1903409.0]\n",
      "###  67 / 499 taxonGroup: [243, 2, 'order', 'Thermoanaerobacterales', 68295.0]\n",
      "###  68 / 499 taxonGroup: [268, 2, 'order', 'Xanthomonadales', 135614.0]\n",
      "###  69 / 499 taxonGroup: [273, 2, 'order', 'Vibrionales', 135623.0]\n",
      "###  70 / 499 taxonGroup: [429, 3, 'family', 'Vibrionaceae', 641.0]\n",
      "###  71 / 499 taxonGroup: [475, 3, 'family', 'Sphingomonadaceae', 41297.0]\n",
      "###  72 / 499 taxonGroup: [810, 3, 'family', 'Synechococcaceae', 1890426.0]\n",
      "###  73 / 499 taxonGroup: [253, 2, 'order', 'Bifidobacteriales', 85004.0]\n",
      "###  74 / 499 taxonGroup: [420, 3, 'family', 'Spirochaetaceae', 137.0]\n",
      "###  75 / 499 taxonGroup: [436, 3, 'family', 'Anaplasmataceae', 942.0]\n",
      "###  76 / 499 taxonGroup: [459, 3, 'family', 'Bifidobacteriaceae', 31953.0]\n",
      "###  77 / 499 taxonGroup: [493, 3, 'family', 'Comamonadaceae', 80864.0]\n",
      "###  78 / 499 taxonGroup: [564, 3, 'family', 'Peptococcaceae', 186807.0]\n",
      "###  79 / 499 taxonGroup: [427, 3, 'family', 'Alcaligenaceae', 506.0]\n",
      "###  80 / 499 taxonGroup: [432, 3, 'family', 'Rickettsiaceae', 775.0]\n",
      "###  81 / 499 taxonGroup: [528, 3, 'family', 'Staphylococcaceae', 90964.0]\n",
      "###  82 / 499 taxonGroup: [35, 0, 'phylum', 'Chlamydiae', 204428.0]\n",
      "###  83 / 499 taxonGroup: [174, 1, 'class', 'Chlamydiia', 204429.0]\n",
      "###  84 / 499 taxonGroup: [239, 2, 'order', 'Chlamydiales', 51291.0]\n",
      "###  85 / 499 taxonGroup: [246, 2, 'order', 'Thiotrichales', 72273.0]\n",
      "###  86 / 499 taxonGroup: [223, 2, 'order', 'Myxococcales', 29.0]\n",
      "###  87 / 499 taxonGroup: [234, 2, 'order', 'Thermotogales', 2419.0]\n",
      "###  88 / 499 taxonGroup: [275, 2, 'order', 'Pasteurellales', 135625.0]\n",
      "###  89 / 499 taxonGroup: [302, 2, 'order', 'Desulfovibrionales', 213115.0]\n",
      "###  90 / 499 taxonGroup: [430, 3, 'family', 'Pasteurellaceae', 712.0]\n",
      "###  91 / 499 taxonGroup: [468, 3, 'family', 'Xanthomonadaceae', 32033.0]\n",
      "###  92 / 499 taxonGroup: [474, 3, 'family', 'Rhodospirillaceae', 41295.0]\n",
      "###  93 / 499 taxonGroup: [487, 3, 'family', 'Helicobacteraceae', 72293.0]\n",
      "###  94 / 499 taxonGroup: [245, 2, 'order', 'Desulfuromonadales', 69541.0]\n",
      "###  95 / 499 taxonGroup: [258, 2, 'order', 'Pseudonocardiales', 85010.0]\n",
      "###  96 / 499 taxonGroup: [264, 2, 'order', 'Deinococcales', 118964.0]\n",
      "###  97 / 499 taxonGroup: [277, 2, 'order', 'Entomoplasmatales', 186328.0]\n",
      "###  98 / 499 taxonGroup: [453, 3, 'family', 'Pseudonocardiaceae', 2070.0]\n",
      "###  99 / 499 taxonGroup: [473, 3, 'family', 'Bradyrhizobiaceae', 41294.0]\n",
      "###  100 / 499 taxonGroup: [485, 3, 'family', 'Alteromonadaceae', 72275.0]\n",
      "###  101 / 499 taxonGroup: [1, 0, 'phylum', 'Chlorobi', 1090.0]\n",
      "###  102 / 499 taxonGroup: [27, 0, 'phylum', 'Chloroflexi', 200795.0]\n",
      "###  103 / 499 taxonGroup: [166, 1, 'class', 'Chlorobia', 191410.0]\n",
      "###  104 / 499 taxonGroup: [229, 2, 'order', 'Oscillatoriales', 1150.0]\n",
      "###  105 / 499 taxonGroup: [230, 2, 'order', 'Nostocales', 1161.0]\n",
      "###  106 / 499 taxonGroup: [288, 2, 'order', 'Chlorobiales', 191411.0]\n",
      "###  107 / 499 taxonGroup: [425, 3, 'family', 'Moraxellaceae', 468.0]\n",
      "###  108 / 499 taxonGroup: [455, 3, 'family', 'Spiroplasmataceae', 2131.0]\n",
      "###  109 / 499 taxonGroup: [559, 3, 'family', 'Deinococcaceae', 183710.0]\n",
      "###  110 / 499 taxonGroup: [584, 3, 'family', 'Chlorobiaceae', 191412.0]\n",
      "###  111 / 499 taxonGroup: [585, 3, 'family', 'Desulfovibrionaceae', 194924.0]\n",
      "###  112 / 499 taxonGroup: [626, 3, 'family', 'Shewanellaceae', 267890.0]\n",
      "###  113 / 499 taxonGroup: [155, 1, 'class', 'Coriobacteriia', 84998.0]\n",
      "###  114 / 499 taxonGroup: [423, 3, 'family', 'Acetobacteraceae', 433.0]\n",
      "###  115 / 499 taxonGroup: [433, 3, 'family', 'Chlamydiaceae', 809.0]\n",
      "###  116 / 499 taxonGroup: [515, 3, 'family', 'Nocardiaceae', 85025.0]\n",
      "###  117 / 499 taxonGroup: [6, 0, 'phylum', 'Fusobacteria', 32066.0]\n",
      "###  118 / 499 taxonGroup: [26, 0, 'phylum', 'Aquificae', 200783.0]\n",
      "###  119 / 499 taxonGroup: [162, 1, 'class', 'Aquificae', 187857.0]\n",
      "###  120 / 499 taxonGroup: [169, 1, 'class', 'Fusobacteriia', 203490.0]\n",
      "###  121 / 499 taxonGroup: [257, 2, 'order', 'Propionibacteriales', 85009.0]\n",
      "###  122 / 499 taxonGroup: [265, 2, 'order', 'Legionellales', 118969.0]\n",
      "###  123 / 499 taxonGroup: [292, 2, 'order', 'Fusobacteriales', 203491.0]\n",
      "###  124 / 499 taxonGroup: [299, 2, 'order', 'Neisseriales', 206351.0]\n",
      "###  125 / 499 taxonGroup: [431, 3, 'family', 'Bartonellaceae', 772.0]\n",
      "###  126 / 499 taxonGroup: [477, 3, 'family', 'Hyphomicrobiaceae', 45401.0]\n",
      "###  127 / 499 taxonGroup: [486, 3, 'family', 'Ectothiorhodospiraceae', 72276.0]\n",
      "###  128 / 499 taxonGroup: [490, 3, 'family', 'Oxalobacteraceae', 75682.0]\n",
      "###  129 / 499 taxonGroup: [565, 3, 'family', 'Thermoanaerobacteraceae', 186814.0]\n",
      "###  130 / 499 taxonGroup: [600, 3, 'family', 'Geobacteraceae', 213422.0]\n",
      "###  131 / 499 taxonGroup: [664, 3, 'family', 'Ruminococcaceae', 541000.0]\n",
      "###  132 / 499 taxonGroup: [850, 3, 'family', 'Yersiniaceae', 1903411.0]\n",
      "###  133 / 499 taxonGroup: [8, 0, 'phylum', 'Acidobacteria', 57723.0]\n",
      "###  134 / 499 taxonGroup: [21, 0, 'phylum', 'Verrucomicrobia', 74201.0]\n",
      "###  135 / 499 taxonGroup: [235, 2, 'order', 'Nitrosomonadales', 32003.0]\n",
      "###  136 / 499 taxonGroup: [244, 2, 'order', 'Thermales', 68933.0]\n",
      "###  137 / 499 taxonGroup: [256, 2, 'order', 'Micromonosporales', 85008.0]\n",
      "###  138 / 499 taxonGroup: [278, 2, 'order', 'Acholeplasmatales', 186329.0]\n",
      "###  139 / 499 taxonGroup: [300, 2, 'order', 'Rhodocyclales', 206389.0]\n",
      "###  140 / 499 taxonGroup: [303, 2, 'order', 'Desulfobacterales', 213118.0]\n",
      "###  141 / 499 taxonGroup: [444, 3, 'family', 'Prochloraceae', 1213.0]\n",
      "###  142 / 499 taxonGroup: [456, 3, 'family', 'Acholeplasmataceae', 2146.0]\n",
      "###  143 / 499 taxonGroup: [457, 3, 'family', 'Micromonosporaceae', 28056.0]\n",
      "###  144 / 499 taxonGroup: [472, 3, 'family', 'Blattabacteriaceae', 39782.0]\n",
      "###  145 / 499 taxonGroup: [491, 3, 'family', 'Rhodocyclaceae', 75787.0]\n",
      "###  146 / 499 taxonGroup: [494, 3, 'family', 'Leuconostocaceae', 81850.0]\n",
      "###  147 / 499 taxonGroup: [575, 3, 'family', 'Thermotogaceae', 188709.0]\n",
      "###  148 / 499 taxonGroup: [577, 3, 'family', 'Thermaceae', 188786.0]\n",
      "###  149 / 499 taxonGroup: [593, 3, 'family', 'Nitrosomonadaceae', 206379.0]\n",
      "###  150 / 499 taxonGroup: [33, 0, 'phylum', 'Planctomycetes', 203682.0]\n",
      "###  151 / 499 taxonGroup: [198, 1, 'class', 'Negativicutes', 909932.0]\n",
      "###  152 / 499 taxonGroup: [237, 2, 'order', 'Aquificales', 32069.0]\n",
      "###  153 / 499 taxonGroup: [274, 2, 'order', 'Aeromonadales', 135624.0]\n",
      "###  154 / 499 taxonGroup: [297, 2, 'order', 'Caulobacterales', 204458.0]\n",
      "###  155 / 499 taxonGroup: [426, 3, 'family', 'Neisseriaceae', 481.0]\n",
      "###  156 / 499 taxonGroup: [434, 3, 'family', 'Bacteroidaceae', 815.0]\n",
      "###  157 / 499 taxonGroup: [438, 3, 'family', 'Chromatiaceae', 1046.0]\n",
      "###  158 / 499 taxonGroup: [440, 3, 'family', 'Nostocaceae', 1162.0]\n",
      "###  159 / 499 taxonGroup: [458, 3, 'family', 'Halomonadaceae', 28256.0]\n",
      "###  160 / 499 taxonGroup: [492, 3, 'family', 'Caulobacteraceae', 76892.0]\n",
      "###  161 / 499 taxonGroup: [502, 3, 'family', 'Aeromonadaceae', 84642.0]\n",
      "###  162 / 499 taxonGroup: [514, 3, 'family', 'Microbacteriaceae', 85023.0]\n",
      "###  163 / 499 taxonGroup: [546, 3, 'family', 'Piscirickettsiaceae', 135616.0]\n",
      "###  164 / 499 taxonGroup: [556, 3, 'family', 'Porphyromonadaceae', 171551.0]\n",
      "###  165 / 499 taxonGroup: [557, 3, 'family', 'Prevotellaceae', 171552.0]\n",
      "###  166 / 499 taxonGroup: [7, 0, 'phylum', 'Nitrospirae', 40117.0]\n",
      "###  167 / 499 taxonGroup: [171, 1, 'class', 'Planctomycetia', 203683.0]\n",
      "###  168 / 499 taxonGroup: [173, 1, 'class', 'Nitrospira', 203693.0]\n",
      "###  169 / 499 taxonGroup: [176, 1, 'class', 'Acidobacteriia', 204432.0]\n",
      "###  170 / 499 taxonGroup: [224, 2, 'order', 'Planctomycetales', 112.0]\n",
      "###  171 / 499 taxonGroup: [232, 2, 'order', 'Actinomycetales', 2037.0]\n",
      "###  172 / 499 taxonGroup: [286, 2, 'order', 'Nitrospirales', 189778.0]\n",
      "###  173 / 499 taxonGroup: [293, 2, 'order', 'Acidobacteriales', 204433.0]\n",
      "###  174 / 499 taxonGroup: [298, 2, 'order', 'Methylophilales', 206350.0]\n",
      "###  175 / 499 taxonGroup: [387, 2, 'order', 'Cellvibrionales', 1706369.0]\n",
      "###  176 / 499 taxonGroup: [424, 3, 'family', 'Legionellaceae', 444.0]\n",
      "###  177 / 499 taxonGroup: [451, 3, 'family', 'Actinomycetaceae', 2049.0]\n",
      "###  178 / 499 taxonGroup: [460, 3, 'family', 'Propionibacteriaceae', 31957.0]\n",
      "###  179 / 499 taxonGroup: [467, 3, 'family', 'Methylophilaceae', 32011.0]\n",
      "###  180 / 499 taxonGroup: [495, 3, 'family', 'Enterococcaceae', 81852.0]\n",
      "###  181 / 499 taxonGroup: [524, 3, 'family', 'Cytophagaceae', 89373.0]\n",
      "###  182 / 499 taxonGroup: [537, 3, 'family', 'Methylobacteriaceae', 119045.0]\n",
      "###  183 / 499 taxonGroup: [561, 3, 'family', 'Lachnospiraceae', 186803.0]\n",
      "###  184 / 499 taxonGroup: [581, 3, 'family', 'Nitrospiraceae', 189779.0]\n",
      "###  185 / 499 taxonGroup: [591, 3, 'family', 'Acidobacteriaceae', 204434.0]\n",
      "###  186 / 499 taxonGroup: [639, 3, 'family', 'Erythrobacteraceae', 335929.0]\n",
      "###  187 / 499 taxonGroup: [753, 3, 'family', 'Borreliaceae', 1643685.0]\n",
      "###  188 / 499 taxonGroup: [805, 3, 'family', 'Hymenobacteraceae', 1853232.0]\n",
      "###  189 / 499 taxonGroup: [852, 3, 'family', 'Morganellaceae', 1903414.0]\n",
      "###  190 / 499 taxonGroup: [158, 1, 'class', 'Sphingobacteriia', 117747.0]\n",
      "###  191 / 499 taxonGroup: [228, 2, 'order', 'Chroococcales', 1118.0]\n",
      "###  192 / 499 taxonGroup: [241, 2, 'order', 'Halanaerobiales', 53433.0]\n",
      "###  193 / 499 taxonGroup: [251, 2, 'order', 'Coriobacteriales', 84999.0]\n",
      "###  194 / 499 taxonGroup: [260, 2, 'order', 'Streptosporangiales', 85012.0]\n",
      "###  195 / 499 taxonGroup: [261, 2, 'order', 'Frankiales', 85013.0]\n",
      "###  196 / 499 taxonGroup: [290, 2, 'order', 'Sphingobacteriales', 200666.0]\n",
      "###  197 / 499 taxonGroup: [381, 2, 'order', 'Eggerthellales', 1643822.0]\n",
      "###  198 / 499 taxonGroup: [421, 3, 'family', 'Leptospiraceae', 170.0]\n",
      "###  199 / 499 taxonGroup: [471, 3, 'family', 'Francisellaceae', 34064.0]\n",
      "###  200 / 499 taxonGroup: [481, 3, 'family', 'Aquificaceae', 64898.0]\n",
      "###  201 / 499 taxonGroup: [489, 3, 'family', 'Frankiaceae', 74712.0]\n",
      "###  202 / 499 taxonGroup: [501, 3, 'family', 'Sphingobacteriaceae', 84566.0]\n",
      "###  203 / 499 taxonGroup: [548, 3, 'family', 'Oceanospirillaceae', 135620.0]\n",
      "###  204 / 499 taxonGroup: [624, 3, 'family', 'Pseudoalteromonadaceae', 267888.0]\n",
      "###  205 / 499 taxonGroup: [671, 3, 'family', 'Thermoanaerobacterales Family III. Incertae Sedis', 543371.0]\n",
      "###  206 / 499 taxonGroup: [712, 3, 'family', 'Leptotrichiaceae', 1129771.0]\n",
      "###  207 / 499 taxonGroup: [755, 3, 'family', 'Eggerthellaceae', 1643826.0]\n",
      "###  208 / 499 taxonGroup: [757, 3, 'family', 'Fervidobacteriaceae', 1643950.0]\n",
      "###  209 / 499 taxonGroup: [833, 3, 'family', 'Cyanothecaceae', 1892249.0]\n",
      "###  210 / 499 taxonGroup: [29, 0, 'phylum', 'Deferribacteres', 200930.0]\n",
      "###  211 / 499 taxonGroup: [43, 0, 'phylum', 'Synergistetes', 508458.0]\n",
      "###  212 / 499 taxonGroup: [150, 1, 'class', 'Chloroflexia', 32061.0]\n",
      "###  213 / 499 taxonGroup: [152, 1, 'class', 'Deferribacteres', 68337.0]\n",
      "###  214 / 499 taxonGroup: [192, 1, 'class', 'Synergistia', 649775.0]\n",
      "###  215 / 499 taxonGroup: [211, 1, 'class', 'Tissierellia', 1737404.0]\n",
      "###  216 / 499 taxonGroup: [236, 2, 'order', 'Chloroflexales', 32064.0]\n",
      "###  217 / 499 taxonGroup: [242, 2, 'order', 'Pelagibacterales', 54526.0]\n",
      "###  218 / 499 taxonGroup: [270, 2, 'order', 'Methylococcales', 135618.0]\n",
      "###  219 / 499 taxonGroup: [287, 2, 'order', 'Deferribacterales', 191393.0]\n",
      "###  220 / 499 taxonGroup: [304, 2, 'order', 'Syntrophobacterales', 213462.0]\n",
      "###  221 / 499 taxonGroup: [305, 2, 'order', 'Bdellovibrionales', 213481.0]\n",
      "###  222 / 499 taxonGroup: [341, 2, 'order', 'Synergistales', 649776.0]\n",
      "###  223 / 499 taxonGroup: [352, 2, 'order', 'Selenomonadales', 909929.0]\n",
      "###  224 / 499 taxonGroup: [388, 2, 'order', 'Tissierellales', 1737405.0]\n",
      "###  225 / 499 taxonGroup: [416, 3, 'family', 'Myxococcaceae', 31.0]\n",
      "###  226 / 499 taxonGroup: [419, 3, 'family', 'Planctomycetaceae', 126.0]\n",
      "###  227 / 499 taxonGroup: [422, 3, 'family', 'Methylococcaceae', 403.0]\n",
      "###  228 / 499 taxonGroup: [442, 3, 'family', 'Rivulariaceae', 1185.0]\n",
      "###  229 / 499 taxonGroup: [483, 3, 'family', 'Phyllobacteriaceae', 69277.0]\n",
      "###  230 / 499 taxonGroup: [562, 3, 'family', 'Peptostreptococcaceae', 186804.0]\n",
      "###  231 / 499 taxonGroup: [574, 3, 'family', 'Carnobacteriaceae', 186828.0]\n",
      "###  232 / 499 taxonGroup: [583, 3, 'family', 'Deferribacteraceae', 191394.0]\n",
      "###  233 / 499 taxonGroup: [588, 3, 'family', 'Fusobacteriaceae', 203492.0]\n",
      "###  234 / 499 taxonGroup: [596, 3, 'family', 'Desulfobacteraceae', 213119.0]\n",
      "###  235 / 499 taxonGroup: [597, 3, 'family', 'Desulfobulbaceae', 213121.0]\n",
      "###  236 / 499 taxonGroup: [688, 3, 'family', 'Synergistaceae', 649777.0]\n",
      "###  237 / 499 taxonGroup: [750, 3, 'family', 'Peptoniphilaceae', 1570339.0]\n",
      "###  238 / 499 taxonGroup: [762, 3, 'family', 'Pelagibacteraceae', 1655514.0]\n",
      "###  239 / 499 taxonGroup: [770, 3, 'family', 'Cellvibrionaceae', 1706371.0]\n",
      "###  240 / 499 taxonGroup: [788, 3, 'family', 'Rhodanobacteraceae', 1775411.0]\n",
      "###  241 / 499 taxonGroup: [849, 3, 'family', 'Pectobacteriaceae', 1903410.0]\n",
      "###  242 / 499 taxonGroup: [31, 0, 'phylum', 'Thermodesulfobacteria', 200940.0]\n",
      "###  243 / 499 taxonGroup: [151, 1, 'class', 'Thermodesulfobacteria', 67799.0]\n",
      "###  244 / 499 taxonGroup: [180, 1, 'class', 'Dehalococcoidia', 301297.0]\n",
      "###  245 / 499 taxonGroup: [184, 1, 'class', 'Opitutae', 414999.0]\n",
      "###  246 / 499 taxonGroup: [214, 1, 'class', 'Acidithiobacillia', 1807140.0]\n",
      "###  247 / 499 taxonGroup: [219, 1, 'class', 'Chitinophagia', 1853228.0]\n",
      "###  248 / 499 taxonGroup: [282, 2, 'order', 'Thermodesulfobacteriales', 188710.0]\n",
      "###  249 / 499 taxonGroup: [310, 2, 'order', 'Acidithiobacillales', 225057.0]\n",
      "###  250 / 499 taxonGroup: [376, 2, 'order', 'Geodermatophilales', 1643682.0]\n",
      "###  251 / 499 taxonGroup: [379, 2, 'order', 'Brachyspirales', 1643686.0]\n",
      "###  252 / 499 taxonGroup: [382, 2, 'order', 'Kosmotogales', 1643946.0]\n",
      "###  253 / 499 taxonGroup: [404, 2, 'order', 'Chitinophagales', 1853229.0]\n",
      "###  254 / 499 taxonGroup: [437, 3, 'family', 'Halanaerobiaceae', 972.0]\n",
      "###  255 / 499 taxonGroup: [484, 3, 'family', 'Hyphomonadaceae', 69657.0]\n",
      "###  256 / 499 taxonGroup: [497, 3, 'family', 'Nocardiopsaceae', 83676.0]\n",
      "###  257 / 499 taxonGroup: [506, 3, 'family', 'Nocardioidaceae', 85015.0]\n",
      "###  258 / 499 taxonGroup: [507, 3, 'family', 'Cellulomonadaceae', 85016.0]\n",
      "###  259 / 499 taxonGroup: [516, 3, 'family', 'Gordoniaceae', 85026.0]\n",
      "###  260 / 499 taxonGroup: [519, 3, 'family', 'Geodermatophilaceae', 85030.0]\n",
      "###  261 / 499 taxonGroup: [534, 3, 'family', 'Brucellaceae', 118882.0]\n",
      "###  262 / 499 taxonGroup: [535, 3, 'family', 'Coxiellaceae', 118968.0]\n",
      "###  263 / 499 taxonGroup: [550, 3, 'family', 'Brachyspiraceae', 143786.0]\n",
      "###  264 / 499 taxonGroup: [551, 3, 'family', 'Dermacoccaceae', 145357.0]\n",
      "###  265 / 499 taxonGroup: [563, 3, 'family', 'Eubacteriaceae', 186806.0]\n",
      "###  266 / 499 taxonGroup: [567, 3, 'family', 'Planococcaceae', 186818.0]\n",
      "###  267 / 499 taxonGroup: [576, 3, 'family', 'Thermodesulfobacteriaceae', 188711.0]\n",
      "###  268 / 499 taxonGroup: [599, 3, 'family', 'Desulfuromonadaceae', 213421.0]\n",
      "###  269 / 499 taxonGroup: [602, 3, 'family', 'Syntrophaceae', 213468.0]\n",
      "###  270 / 499 taxonGroup: [603, 3, 'family', 'Bdellovibrionaceae', 213483.0]\n",
      "###  271 / 499 taxonGroup: [607, 3, 'family', 'Alcanivoracaceae', 224372.0]\n",
      "###  272 / 499 taxonGroup: [614, 3, 'family', 'Acidithiobacillaceae', 225058.0]\n",
      "###  273 / 499 taxonGroup: [638, 3, 'family', 'Xanthobacteraceae', 335928.0]\n",
      "###  274 / 499 taxonGroup: [675, 3, 'family', 'Cyclobacteriaceae', 563798.0]\n",
      "###  275 / 499 taxonGroup: [676, 3, 'family', 'Chitinophagaceae', 563835.0]\n",
      "###  276 / 499 taxonGroup: [754, 3, 'family', 'Atopobiaceae', 1643824.0]\n",
      "###  277 / 499 taxonGroup: [756, 3, 'family', 'Kosmotogaceae', 1643948.0]\n",
      "###  278 / 499 taxonGroup: [18, 0, 'phylum', 'Dictyoglomi', 68297.0]\n",
      "###  279 / 499 taxonGroup: [20, 0, 'phylum', 'Elusimicrobia', 74152.0]\n",
      "###  280 / 499 taxonGroup: [49, 0, 'phylum', 'Ignavibacteriae', 1134404.0]\n",
      "###  281 / 499 taxonGroup: [153, 1, 'class', 'Acidimicrobiia', 84992.0]\n",
      "###  282 / 499 taxonGroup: [154, 1, 'class', 'Rubrobacteria', 84995.0]\n",
      "###  283 / 499 taxonGroup: [165, 1, 'class', 'Thermomicrobia', 189775.0]\n",
      "###  284 / 499 taxonGroup: [168, 1, 'class', 'Dictyoglomia', 203486.0]\n",
      "###  285 / 499 taxonGroup: [181, 1, 'class', 'Gloeobacteria', 307596.0]\n",
      "###  286 / 499 taxonGroup: [196, 1, 'class', 'Ignavibacteria', 795747.0]\n",
      "###  287 / 499 taxonGroup: [222, 1, 'class', 'Saprospiria', 1937959.0]\n",
      "###  288 / 499 taxonGroup: [240, 2, 'order', 'Pleurocapsales', 52604.0]\n",
      "###  289 / 499 taxonGroup: [249, 2, 'order', 'Acidimicrobiales', 84993.0]\n",
      "###  290 / 499 taxonGroup: [250, 2, 'order', 'Rubrobacterales', 84996.0]\n",
      "###  291 / 499 taxonGroup: [291, 2, 'order', 'Dictyoglomales', 203487.0]\n",
      "###  292 / 499 taxonGroup: [301, 2, 'order', 'Desulfurellales', 213113.0]\n",
      "###  293 / 499 taxonGroup: [316, 2, 'order', 'Gloeobacterales', 307595.0]\n",
      "###  294 / 499 taxonGroup: [322, 2, 'order', 'Opitutales', 415000.0]\n",
      "###  295 / 499 taxonGroup: [345, 2, 'order', 'Gallionellales', 713636.0]\n",
      "###  296 / 499 taxonGroup: [346, 2, 'order', 'Methylacidiphilales', 717963.0]\n",
      "###  297 / 499 taxonGroup: [350, 2, 'order', 'Ignavibacteriales', 795748.0]\n",
      "###  298 / 499 taxonGroup: [355, 2, 'order', 'Bacteroidetes Order II. Incertae sedis', 1100069.0]\n",
      "###  299 / 499 taxonGroup: [360, 2, 'order', 'Dehalococcoidales', 1202465.0]\n",
      "###  300 / 499 taxonGroup: [362, 2, 'order', 'Orbales', 1240482.0]\n",
      "###  301 / 499 taxonGroup: [368, 2, 'order', 'Desulfurobacteriales', 1485951.0]\n",
      "###  302 / 499 taxonGroup: [383, 2, 'order', 'Petrotogales', 1643947.0]\n",
      "###  303 / 499 taxonGroup: [400, 2, 'order', 'Acidaminococcales', 1843488.0]\n",
      "###  304 / 499 taxonGroup: [411, 2, 'order', 'Holosporales', 1921002.0]\n",
      "###  305 / 499 taxonGroup: [415, 2, 'order', 'Saprospirales', 1936988.0]\n",
      "###  306 / 499 taxonGroup: [417, 3, 'family', 'Archangiaceae', 39.0]\n",
      "###  307 / 499 taxonGroup: [418, 3, 'family', 'Polyangiaceae', 49.0]\n",
      "###  308 / 499 taxonGroup: [439, 3, 'family', 'Chloroflexaceae', 1106.0]\n",
      "###  309 / 499 taxonGroup: [478, 3, 'family', 'Beijerinckiaceae', 45404.0]\n",
      "###  310 / 499 taxonGroup: [480, 3, 'family', 'Halobacteroidaceae', 53434.0]\n",
      "###  311 / 499 taxonGroup: [482, 3, 'family', 'Syntrophomonadaceae', 68298.0]\n",
      "###  312 / 499 taxonGroup: [500, 3, 'family', 'Coriobacteriaceae', 84107.0]\n",
      "###  313 / 499 taxonGroup: [503, 3, 'family', 'Acidimicrobiaceae', 84994.0]\n",
      "###  314 / 499 taxonGroup: [504, 3, 'family', 'Rubrobacteraceae', 84997.0]\n",
      "###  315 / 499 taxonGroup: [508, 3, 'family', 'Promicromonosporaceae', 85017.0]\n",
      "###  316 / 499 taxonGroup: [512, 3, 'family', 'Intrasporangiaceae', 85021.0]\n",
      "###  317 / 499 taxonGroup: [527, 3, 'family', 'Gallionellaceae', 90627.0]\n",
      "###  318 / 499 taxonGroup: [530, 3, 'family', 'Parachlamydiaceae', 92713.0]\n",
      "###  319 / 499 taxonGroup: [533, 3, 'family', 'Desulfurellaceae', 117942.0]\n",
      "###  320 / 499 taxonGroup: [544, 3, 'family', 'Opitutaceae', 134623.0]\n",
      "###  321 / 499 taxonGroup: [547, 3, 'family', 'Thiotrichaceae', 135617.0]\n",
      "###  322 / 499 taxonGroup: [555, 3, 'family', 'Rikenellaceae', 171550.0]\n",
      "###  323 / 499 taxonGroup: [571, 3, 'family', 'Alicyclobacillaceae', 186823.0]\n",
      "###  324 / 499 taxonGroup: [573, 3, 'family', 'Aerococcaceae', 186827.0]\n",
      "###  325 / 499 taxonGroup: [586, 3, 'family', 'Flammeovirgaceae', 200667.0]\n",
      "###  326 / 499 taxonGroup: [587, 3, 'family', 'Dictyoglomaceae', 203488.0]\n",
      "###  327 / 499 taxonGroup: [606, 3, 'family', 'Hydrogenothermaceae', 224027.0]\n",
      "###  328 / 499 taxonGroup: [616, 3, 'family', 'Thermodesulfobiaceae', 227387.0]\n",
      "###  329 / 499 taxonGroup: [619, 3, 'family', 'Aurantimonadaceae', 255475.0]\n",
      "###  330 / 499 taxonGroup: [625, 3, 'family', 'Colwelliaceae', 267889.0]\n",
      "###  331 / 499 taxonGroup: [630, 3, 'family', 'Psychromonadaceae', 267894.0]\n",
      "###  332 / 499 taxonGroup: [673, 3, 'family', 'Desulfurobacteriaceae', 558314.0]\n",
      "###  333 / 499 taxonGroup: [677, 3, 'family', 'Rhodothermaceae', 563843.0]\n",
      "###  334 / 499 taxonGroup: [696, 3, 'family', 'Methylacidiphilaceae', 717964.0]\n",
      "###  335 / 499 taxonGroup: [702, 3, 'family', 'Acidaminococcaceae', 909930.0]\n",
      "###  336 / 499 taxonGroup: [716, 3, 'family', 'Dehalococcoidaceae', 1202464.0]\n",
      "###  337 / 499 taxonGroup: [718, 3, 'family', 'Orbaceae', 1240483.0]\n",
      "###  338 / 499 taxonGroup: [736, 3, 'family', 'Chromobacteriaceae', 1499392.0]\n",
      "###  339 / 499 taxonGroup: [737, 3, 'family', 'Amoebophilaceae', 1501348.0]\n",
      "###  340 / 499 taxonGroup: [740, 3, 'family', 'Roseiflexaceae', 1508635.0]\n",
      "###  341 / 499 taxonGroup: [744, 3, 'family', 'Anaeromyxobacteraceae', 1524215.0]\n",
      "###  342 / 499 taxonGroup: [774, 3, 'family', 'Spongiibacteraceae', 1706375.0]\n",
      "###  343 / 499 taxonGroup: [783, 3, 'family', 'Isosphaeraceae', 1763524.0]\n",
      "###  344 / 499 taxonGroup: [799, 3, 'family', 'Sporomusaceae', 1843490.0]\n",
      "###  345 / 499 taxonGroup: [800, 3, 'family', 'Selenomonadaceae', 1843491.0]\n",
      "###  346 / 499 taxonGroup: [809, 3, 'family', 'Gloeobacteraceae', 1890422.0]\n",
      "###  347 / 499 taxonGroup: [822, 3, 'family', 'Aphanothecaceae', 1890450.0]\n",
      "###  348 / 499 taxonGroup: [835, 3, 'family', 'Microcoleaceae', 1892252.0]\n",
      "###  349 / 499 taxonGroup: [836, 3, 'family', 'Oscillatoriaceae', 1892254.0]\n",
      "###  350 / 499 taxonGroup: [851, 3, 'family', 'Hafniaceae', 1903412.0]\n",
      "###  351 / 499 taxonGroup: [857, 3, 'family', 'Kangiellaceae', 1920240.0]\n",
      "###  352 / 499 taxonGroup: [10, 0, 'phylum', 'Fibrobacteres', 65842.0]\n",
      "###  353 / 499 taxonGroup: [13, 0, 'phylum', 'Caldiserica', 67814.0]\n",
      "###  354 / 499 taxonGroup: [16, 0, 'phylum', 'Armatimonadetes', 67819.0]\n",
      "###  355 / 499 taxonGroup: [24, 0, 'phylum', 'Gemmatimonadetes', 142182.0]\n",
      "###  356 / 499 taxonGroup: [30, 0, 'phylum', 'Chrysiogenetes', 200938.0]\n",
      "###  357 / 499 taxonGroup: [42, 0, 'phylum', 'Candidatus Cloacimonetes', 456828.0]\n",
      "###  358 / 499 taxonGroup: [159, 1, 'class', 'Chrysiogenetes', 118001.0]\n",
      "###  359 / 499 taxonGroup: [160, 1, 'class', 'Spartobacteria', 134549.0]\n",
      "###  360 / 499 taxonGroup: [170, 1, 'class', 'Verrucomicrobiae', 203494.0]\n",
      "###  361 / 499 taxonGroup: [175, 1, 'class', 'Fibrobacteria', 204430.0]\n",
      "###  362 / 499 taxonGroup: [177, 1, 'class', 'Gemmatimonadetes', 219685.0]\n",
      "###  363 / 499 taxonGroup: [178, 1, 'class', 'Anaerolineae', 292625.0]\n",
      "###  364 / 499 taxonGroup: [182, 1, 'class', 'Solibacteres', 332159.0]\n",
      "###  365 / 499 taxonGroup: [185, 1, 'class', 'Endomicrobia', 447830.0]\n",
      "###  366 / 499 taxonGroup: [187, 1, 'class', 'Caldilineae', 475962.0]\n",
      "###  367 / 499 taxonGroup: [188, 1, 'class', 'Erysipelotrichia', 526524.0]\n",
      "###  368 / 499 taxonGroup: [191, 1, 'class', 'Elusimicrobia', 641853.0]\n",
      "###  369 / 499 taxonGroup: [193, 1, 'class', 'Phycisphaerae', 666505.0]\n",
      "###  370 / 499 taxonGroup: [194, 1, 'class', 'Caldisericia', 693071.0]\n",
      "###  371 / 499 taxonGroup: [206, 1, 'class', 'Thermoleophilia', 1497346.0]\n",
      "###  372 / 499 taxonGroup: [208, 1, 'class', 'Blastocatellia', 1562566.0]\n",
      "###  373 / 499 taxonGroup: [209, 1, 'class', 'Fimbriimonadia', 1663419.0]\n",
      "###  374 / 499 taxonGroup: [238, 2, 'order', 'Verrucomicrobiales', 48461.0]\n",
      "###  375 / 499 taxonGroup: [252, 2, 'order', 'Sphaerobacterales', 85001.0]\n",
      "###  376 / 499 taxonGroup: [262, 2, 'order', 'Glycomycetales', 85014.0]\n",
      "###  377 / 499 taxonGroup: [266, 2, 'order', 'Hydrogenophilales', 119069.0]\n",
      "###  378 / 499 taxonGroup: [269, 2, 'order', 'Cardiobacteriales', 135615.0]\n",
      "###  379 / 499 taxonGroup: [283, 2, 'order', 'Chrysiogenales', 189769.0]\n",
      "###  380 / 499 taxonGroup: [285, 2, 'order', 'Thermomicrobiales', 189776.0]\n",
      "###  381 / 499 taxonGroup: [307, 2, 'order', 'Fibrobacterales', 218872.0]\n",
      "###  382 / 499 taxonGroup: [308, 2, 'order', 'Gemmatimonadales', 219686.0]\n",
      "###  383 / 499 taxonGroup: [311, 2, 'order', 'Nautiliales', 235899.0]\n",
      "###  384 / 499 taxonGroup: [312, 2, 'order', 'Parvularculales', 255473.0]\n",
      "###  385 / 499 taxonGroup: [315, 2, 'order', 'Anaerolineales', 292629.0]\n",
      "###  386 / 499 taxonGroup: [317, 2, 'order', 'Solibacterales', 332160.0]\n",
      "###  387 / 499 taxonGroup: [321, 2, 'order', 'Catenulisporales', 414714.0]\n",
      "###  388 / 499 taxonGroup: [323, 2, 'order', 'Puniceicoccales', 415001.0]\n",
      "###  389 / 499 taxonGroup: [324, 2, 'order', 'Desulfarculales', 453227.0]\n",
      "###  390 / 499 taxonGroup: [327, 2, 'order', 'Caldilineales', 475963.0]\n",
      "###  391 / 499 taxonGroup: [328, 2, 'order', 'Natranaerobiales', 485256.0]\n",
      "###  392 / 499 taxonGroup: [330, 2, 'order', 'Erysipelotrichales', 526525.0]\n",
      "###  393 / 499 taxonGroup: [334, 2, 'order', 'Solirubrobacterales', 588673.0]\n",
      "###  394 / 499 taxonGroup: [339, 2, 'order', 'Kineosporiales', 622452.0]\n",
      "###  395 / 499 taxonGroup: [340, 2, 'order', 'Elusimicrobiales', 641854.0]\n",
      "###  396 / 499 taxonGroup: [342, 2, 'order', 'Phycisphaerales', 666506.0]\n",
      "###  397 / 499 taxonGroup: [343, 2, 'order', 'Caldisericales', 693072.0]\n",
      "###  398 / 499 taxonGroup: [359, 2, 'order', 'Magnetococcales', 1191478.0]\n",
      "###  399 / 499 taxonGroup: [374, 2, 'order', 'Sulfuricellales', 1559979.0]\n",
      "###  400 / 499 taxonGroup: [377, 2, 'order', 'Acidothermales', 1643683.0]\n",
      "###  401 / 499 taxonGroup: [378, 2, 'order', 'Nakamurellales', 1643684.0]\n",
      "###  402 / 499 taxonGroup: [384, 2, 'order', 'Fimbriimonadales', 1663425.0]\n",
      "###  403 / 499 taxonGroup: [395, 2, 'order', 'Endomicrobiales', 1783344.0]\n",
      "###  404 / 499 taxonGroup: [401, 2, 'order', 'Veillonellales', 1843489.0]\n",
      "###  405 / 499 taxonGroup: [408, 2, 'order', 'Chroococcidiopsidales', 1890505.0]\n",
      "###  406 / 499 taxonGroup: [435, 3, 'family', 'Cardiobacteriaceae', 868.0]\n",
      "###  407 / 499 taxonGroup: [449, 3, 'family', 'Streptosporangiaceae', 2004.0]\n",
      "###  408 / 499 taxonGroup: [450, 3, 'family', 'Thermomonosporaceae', 2012.0]\n",
      "###  409 / 499 taxonGroup: [461, 3, 'family', 'Veillonellaceae', 31977.0]\n",
      "###  410 / 499 taxonGroup: [463, 3, 'family', 'Heliobacteriaceae', 31984.0]\n",
      "###  411 / 499 taxonGroup: [465, 3, 'family', 'Methylocystaceae', 31993.0]\n",
      "###  412 / 499 taxonGroup: [469, 3, 'family', 'Entomoplasmataceae', 33925.0]\n",
      "###  413 / 499 taxonGroup: [476, 3, 'family', 'Holosporaceae', 44746.0]\n",
      "###  414 / 499 taxonGroup: [499, 3, 'family', 'Kineosporiaceae', 83778.0]\n",
      "###  415 / 499 taxonGroup: [505, 3, 'family', 'Sphaerobacteraceae', 85002.0]\n",
      "###  416 / 499 taxonGroup: [510, 3, 'family', 'Brevibacteriaceae', 85019.0]\n",
      "###  417 / 499 taxonGroup: [511, 3, 'family', 'Dermabacteraceae', 85020.0]\n",
      "###  418 / 499 taxonGroup: [513, 3, 'family', 'Jonesiaceae', 85022.0]\n",
      "###  419 / 499 taxonGroup: [517, 3, 'family', 'Tsukamurellaceae', 85028.0]\n",
      "###  420 / 499 taxonGroup: [520, 3, 'family', 'Nakamurellaceae', 85031.0]\n",
      "###  421 / 499 taxonGroup: [521, 3, 'family', 'Acidothermaceae', 85032.0]\n",
      "###  422 / 499 taxonGroup: [523, 3, 'family', 'Glycomycetaceae', 85034.0]\n",
      "###  423 / 499 taxonGroup: [525, 3, 'family', 'Saprospiraceae', 89374.0]\n",
      "###  424 / 499 taxonGroup: [529, 3, 'family', 'Simkaniaceae', 92712.0]\n",
      "###  425 / 499 taxonGroup: [531, 3, 'family', 'Waddliaceae', 92714.0]\n",
      "###  426 / 499 taxonGroup: [536, 3, 'family', 'Rhodobiaceae', 119043.0]\n",
      "###  427 / 499 taxonGroup: [541, 3, 'family', 'Beutenbergiaceae', 125316.0]\n",
      "###  428 / 499 taxonGroup: [542, 3, 'family', 'Erysipelotrichaceae', 128827.0]\n",
      "###  429 / 499 taxonGroup: [554, 3, 'family', 'Sanguibacteraceae', 145360.0]\n",
      "###  430 / 499 taxonGroup: [568, 3, 'family', 'Listeriaceae', 186820.0]\n",
      "###  431 / 499 taxonGroup: [569, 3, 'family', 'Sporolactobacillaceae', 186821.0]\n",
      "###  432 / 499 taxonGroup: [578, 3, 'family', 'Chrysiogenaceae', 189770.0]\n",
      "###  433 / 499 taxonGroup: [580, 3, 'family', 'Thermomicrobiaceae', 189777.0]\n",
      "###  434 / 499 taxonGroup: [590, 3, 'family', 'Fibrobacteraceae', 204431.0]\n",
      "###  435 / 499 taxonGroup: [592, 3, 'family', 'Hydrogenophilaceae', 206349.0]\n",
      "###  436 / 499 taxonGroup: [594, 3, 'family', 'Desulfomicrobiaceae', 213116.0]\n",
      "###  437 / 499 taxonGroup: [595, 3, 'family', 'Desulfohalobiaceae', 213117.0]\n",
      "###  438 / 499 taxonGroup: [601, 3, 'family', 'Syntrophobacteraceae', 213465.0]\n",
      "###  439 / 499 taxonGroup: [604, 3, 'family', 'Oscillospiraceae', 216572.0]\n",
      "###  440 / 499 taxonGroup: [605, 3, 'family', 'Gemmatimonadaceae', 219687.0]\n",
      "###  441 / 499 taxonGroup: [608, 3, 'family', 'Hahellaceae', 224379.0]\n",
      "###  442 / 499 taxonGroup: [611, 3, 'family', 'Kofleriaceae', 224464.0]\n",
      "###  443 / 499 taxonGroup: [612, 3, 'family', 'Nautiliaceae', 224467.0]\n",
      "###  444 / 499 taxonGroup: [617, 3, 'family', 'Cryomorphaceae', 246874.0]\n",
      "###  445 / 499 taxonGroup: [618, 3, 'family', 'Parvularculaceae', 255474.0]\n",
      "###  446 / 499 taxonGroup: [620, 3, 'family', 'Halothiobacillaceae', 255526.0]\n",
      "###  447 / 499 taxonGroup: [621, 3, 'family', 'Saccharospirillaceae', 255527.0]\n",
      "###  448 / 499 taxonGroup: [627, 3, 'family', 'Moritellaceae', 267891.0]\n",
      "###  449 / 499 taxonGroup: [628, 3, 'family', 'Ferrimonadaceae', 267892.0]\n",
      "###  450 / 499 taxonGroup: [629, 3, 'family', 'Idiomarinaceae', 267893.0]\n",
      "###  451 / 499 taxonGroup: [631, 3, 'family', 'Anaerolineaceae', 292628.0]\n",
      "###  452 / 499 taxonGroup: [632, 3, 'family', 'Segniliparaceae', 316606.0]\n",
      "###  453 / 499 taxonGroup: [633, 3, 'family', 'Conexibacteraceae', 320583.0]\n",
      "###  454 / 499 taxonGroup: [636, 3, 'family', 'Solibacteraceae', 332161.0]\n",
      "###  455 / 499 taxonGroup: [637, 3, 'family', 'Trueperaceae', 332247.0]\n",
      "###  456 / 499 taxonGroup: [647, 3, 'family', 'Catenulisporaceae', 414877.0]\n",
      "###  457 / 499 taxonGroup: [648, 3, 'family', 'Puniceicoccaceae', 415002.0]\n",
      "###  458 / 499 taxonGroup: [652, 3, 'family', 'Desulfarculaceae', 453228.0]\n",
      "###  459 / 499 taxonGroup: [655, 3, 'family', 'Caldilineaceae', 475964.0]\n",
      "###  460 / 499 taxonGroup: [656, 3, 'family', 'Natranaerobiaceae', 485255.0]\n",
      "###  461 / 499 taxonGroup: [661, 3, 'family', 'Clostridiales Family XVII. Incertae Sedis', 539000.0]\n",
      "###  462 / 499 taxonGroup: [667, 3, 'family', 'Clostridiales Family XIII. Incertae Sedis', 543314.0]\n",
      "###  463 / 499 taxonGroup: [670, 3, 'family', 'Symbiobacteriaceae', 543349.0]\n",
      "###  464 / 499 taxonGroup: [672, 3, 'family', 'Thermoanaerobacterales Family IV. Incertae Sedis', 543372.0]\n",
      "###  465 / 499 taxonGroup: [687, 3, 'family', 'Elusimicrobiaceae', 641876.0]\n",
      "###  466 / 499 taxonGroup: [691, 3, 'family', 'Phycisphaeraceae', 666507.0]\n",
      "###  467 / 499 taxonGroup: [693, 3, 'family', 'Caldisericaceae', 693073.0]\n",
      "###  468 / 499 taxonGroup: [700, 3, 'family', 'Ignavibacteriaceae', 795749.0]\n",
      "###  469 / 499 taxonGroup: [707, 3, 'family', 'Sandaracinaceae', 1055686.0]\n",
      "###  470 / 499 taxonGroup: [715, 3, 'family', 'Magnetococcaceae', 1191479.0]\n",
      "###  471 / 499 taxonGroup: [722, 3, 'family', 'Candidatus Midichloriaceae', 1328881.0]\n",
      "###  472 / 499 taxonGroup: [726, 3, 'family', 'Melioribacteraceae', 1334117.0]\n",
      "###  473 / 499 taxonGroup: [732, 3, 'family', 'Prolixibacteraceae', 1471398.0]\n",
      "###  474 / 499 taxonGroup: [743, 3, 'family', 'Vulgatibacteraceae', 1524213.0]\n",
      "###  475 / 499 taxonGroup: [749, 3, 'family', 'Sulfuricellaceae', 1559980.0]\n",
      "###  476 / 499 taxonGroup: [759, 3, 'family', 'Akkermansiaceae', 1647988.0]\n",
      "###  477 / 499 taxonGroup: [761, 3, 'family', 'Halobacteriovoraceae', 1652132.0]\n",
      "###  478 / 499 taxonGroup: [763, 3, 'family', 'Fimbriimonadaceae', 1663426.0]\n",
      "###  479 / 499 taxonGroup: [764, 3, 'family', 'Wenzhouxiangellaceae', 1676141.0]\n",
      "###  480 / 499 taxonGroup: [789, 3, 'family', 'Candidatus Paracaedibacteraceae', 1777752.0]\n",
      "###  481 / 499 taxonGroup: [791, 3, 'family', 'Endomicrobiaceae', 1783343.0]\n",
      "###  482 / 499 taxonGroup: [803, 3, 'family', 'Crocinitomicaceae', 1853230.0]\n",
      "###  483 / 499 taxonGroup: [804, 3, 'family', 'Odoribacteraceae', 1853231.0]\n",
      "###  484 / 499 taxonGroup: [807, 3, 'family', 'Persicobacteraceae', 1853234.0]\n",
      "###  485 / 499 taxonGroup: [811, 3, 'family', 'Merismopediaceae', 1890428.0]\n",
      "###  486 / 499 taxonGroup: [812, 3, 'family', 'Acaryochloridaceae', 1890429.0]\n",
      "###  487 / 499 taxonGroup: [816, 3, 'family', 'Pseudanabaenaceae', 1890436.0]\n",
      "###  488 / 499 taxonGroup: [817, 3, 'family', 'Leptolyngbyaceae', 1890438.0]\n",
      "###  489 / 499 taxonGroup: [821, 3, 'family', 'Microcystaceae', 1890449.0]\n",
      "###  490 / 499 taxonGroup: [823, 3, 'family', 'Cyanobacteriaceae', 1890452.0]\n",
      "###  491 / 499 taxonGroup: [826, 3, 'family', 'Chroococcaceae', 1890464.0]\n",
      "###  492 / 499 taxonGroup: [829, 3, 'family', 'Dermocarpellaceae', 1890498.0]\n",
      "###  493 / 499 taxonGroup: [831, 3, 'family', 'Hyellaceae', 1890500.0]\n",
      "###  494 / 499 taxonGroup: [832, 3, 'family', 'Chroococcidiopsidaceae', 1890528.0]\n",
      "###  495 / 499 taxonGroup: [834, 3, 'family', 'Coleofasciculaceae', 1892251.0]\n",
      "###  496 / 499 taxonGroup: [837, 3, 'family', 'Gomontiellaceae', 1892255.0]\n",
      "###  497 / 499 taxonGroup: [853, 3, 'family', 'Budviciaceae', 1903416.0]\n",
      "###  498 / 499 taxonGroup: [863, 3, 'family', 'Haliscomenobacteraceae', 1937961.0]\n",
      "###  499 / 499 taxonGroup: [865, 3, 'family', 'Bernardetiaceae', 1937968.0]\n"
     ]
    }
   ],
   "source": [
    "oddsRatioCtermTaxaDf = gather_cterm_bias_taxon_results(taxaGroupsInRefSeq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gather all results by taxon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I gather all results by taxon and c-terminal AA.\n",
    "\n",
    "taxon, bias in c-terminal aa (-1 position), enrichment in COG, enrichment in orthogroup."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "taxon_id  taxon_name      taxon_rank  n_seq      n_genomes  observable   \n",
       "1224      Proteobacteria  phylum      2165400.0  691        count       R    208748.0\n",
       "                                                                        H     66152.0\n",
       "                                                                        K    172536.0\n",
       "                                                                        D    106100.0\n",
       "                                                                        E    139547.0\n",
       "                                                                        S    147598.0\n",
       "                                                                        T     53226.0\n",
       "                                                                        N     78816.0\n",
       "                                                                        Q     95791.0\n",
       "                                                                        C     19087.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioCtermTaxaStackedDf = \\\n",
    "    oddsRatioCtermTaxaDf.drop(['position from terminus', 'terminus'], axis=1)\\\n",
    "    .rename(columns={'taxaid':'taxon_id'})\\\n",
    "    .set_index(['taxon_id', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes', 'observable'])\\\n",
    "    .stack()\n",
    "    \n",
    "oddsRatioCtermTaxaStackedDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observable</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>taxon_rank</th>\n",
       "      <th>cterm_bias_n_seq</th>\n",
       "      <th>cterm_bias_n_genomes</th>\n",
       "      <th>cterm_bias_count</th>\n",
       "      <th>cterm_bias_log2OddsRatio</th>\n",
       "      <th>cterm_bias_oddsRatio</th>\n",
       "      <th>cterm_bias_pvalue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxon_id</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">29</th>\n",
       "      <th>R</th>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>order</td>\n",
       "      <td>91404.0</td>\n",
       "      <td>13</td>\n",
       "      <td>15288.0</td>\n",
       "      <td>1.176471</td>\n",
       "      <td>2.260232</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>order</td>\n",
       "      <td>91404.0</td>\n",
       "      <td>13</td>\n",
       "      <td>2076.0</td>\n",
       "      <td>0.086519</td>\n",
       "      <td>1.061805</td>\n",
       "      <td>7.429050e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>order</td>\n",
       "      <td>91404.0</td>\n",
       "      <td>13</td>\n",
       "      <td>3962.0</td>\n",
       "      <td>0.807761</td>\n",
       "      <td>1.750493</td>\n",
       "      <td>2.293645e-220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>order</td>\n",
       "      <td>91404.0</td>\n",
       "      <td>13</td>\n",
       "      <td>3398.0</td>\n",
       "      <td>-0.596094</td>\n",
       "      <td>0.661543</td>\n",
       "      <td>6.121232e-140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>order</td>\n",
       "      <td>91404.0</td>\n",
       "      <td>13</td>\n",
       "      <td>5420.0</td>\n",
       "      <td>-0.153837</td>\n",
       "      <td>0.898857</td>\n",
       "      <td>1.232448e-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "observable           taxon_name taxon_rank  cterm_bias_n_seq  \\\n",
       "taxon_id cterm_aa                                              \n",
       "29       R         Myxococcales      order           91404.0   \n",
       "         H         Myxococcales      order           91404.0   \n",
       "         K         Myxococcales      order           91404.0   \n",
       "         D         Myxococcales      order           91404.0   \n",
       "         E         Myxococcales      order           91404.0   \n",
       "\n",
       "observable         cterm_bias_n_genomes  cterm_bias_count  \\\n",
       "taxon_id cterm_aa                                           \n",
       "29       R                           13           15288.0   \n",
       "         H                           13            2076.0   \n",
       "         K                           13            3962.0   \n",
       "         D                           13            3398.0   \n",
       "         E                           13            5420.0   \n",
       "\n",
       "observable         cterm_bias_log2OddsRatio  cterm_bias_oddsRatio  \\\n",
       "taxon_id cterm_aa                                                   \n",
       "29       R                         1.176471              2.260232   \n",
       "         H                         0.086519              1.061805   \n",
       "         K                         0.807761              1.750493   \n",
       "         D                        -0.596094              0.661543   \n",
       "         E                        -0.153837              0.898857   \n",
       "\n",
       "observable         cterm_bias_pvalue  \n",
       "taxon_id cterm_aa                     \n",
       "29       R              0.000000e+00  \n",
       "         H              7.429050e-03  \n",
       "         K             2.293645e-220  \n",
       "         D             6.121232e-140  \n",
       "         E              1.232448e-14  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioCtermTaxaStackedDf.index.names = ['taxon_id', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes', 'observable', 'cterm_aa']\n",
    "oddsRatioCtermTaxaStackedDf.name = 'value'\n",
    "oddsRatioCtermTaxaStackedDf = \\\n",
    "    oddsRatioCtermTaxaStackedDf.unstack(level=['observable'])\\\n",
    "                               .reset_index(level=['n_seq', 'n_genomes'])\\\n",
    "                               .rename(columns=(lambda x: 'cterm_bias_' + x))\\\n",
    "                               .reset_index(level=['taxon_name', 'taxon_rank'])\n",
    "#                                .drop('count', axis=1)\\\n",
    "                \n",
    "oddsRatioCtermTaxaStackedDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroupEnrichStackedDfFull2 = \\\n",
    "    taxonGroupEnrichStackedDfFull\\\n",
    "        .drop(enrichmentTestcolumnName, axis=1)\\\n",
    "        .set_index(['taxon_id', 'cterm_aa'])\\\n",
    "        .rename(columns={'COGGroup_index':'group_index'})\n",
    "#         .rename(columns=(lambda x: 'COG_enrichment_' + x))\n",
    "\n",
    "taxonGroupEnrichStackedDfFull2['enrichment_analysis'] = 'COG'\n",
    "\n",
    "\n",
    "# CORRECT MISTAKE IN THE SCRIPT, SWAP COUNT COLUMNS\n",
    "# cols\n",
    "# cols = taxonGroupEnrichStackedDfFull2.columns.tolist()\n",
    "# cols[4], cols[5] = cols[5], cols[4]\n",
    "# taxonGroupEnrichStackedDfFull2 = taxonGroupEnrichStackedDfFull2[cols]\n",
    "# taxonGroupEnrichStackedDfFull2.rename(columns={'count_NOT_cterm_in_group':'count_cterm_NOT_in_group',\n",
    "#                                                'count_cterm_NOT_in_group':'count_NOT_cterm_in_group',\n",
    "#                                               })\n",
    "\n",
    "\n",
    "cols = taxonGroupEnrichStackedDfFull2.columns.tolist()\n",
    "cols = cols[-1:] + cols[:-1]\n",
    "taxonGroupEnrichStackedDfFull2 = taxonGroupEnrichStackedDfFull2[cols]\n",
    "taxonGroupEnrichStackedDfFull2.reset_index(inplace=True)\n",
    "taxonGroupEnrichStackedDfFull2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxonEnrichOrthoGroupStackedDfFull2 = \\\n",
    "    taxonEnrichOrthoGroupStackedDfFull\\\n",
    "        .drop(enrichmentTestcolumnName, axis=1)\\\n",
    "        .set_index(['taxon_id', 'cterm_aa'])\\\n",
    "        .rename(columns={'orthologous_group_index':'group_index',\n",
    "                         'predicted_gene_name':'OG_predicted_gene_name',\n",
    "                         'KEGG_pathways':'OG_KEGG_pathways',\n",
    "                         'COG cat':'OG_COG_cat',\n",
    "                         'eggNOG annot':'OG_eggNOG_annot'\n",
    "                        })\n",
    "\n",
    "        \n",
    "# CORRECT MISTAKE IN THE SCRIPT, SWAP COUNT COLUMNS\n",
    "# cols = taxonEnrichOrthoGroupStackedDfFull2.columns.tolist()\n",
    "# cols\n",
    "# print(cols[6])\n",
    "# print(cols[7])\n",
    "# cols[6], cols[7] = cols[7], cols[6]\n",
    "# taxonEnrichOrthoGroupStackedDfFull2 = taxonEnrichOrthoGroupStackedDfFull2[cols]\n",
    "# taxonEnrichOrthoGroupStackedDfFull2.rename(columns={'count_NOT_cterm_in_group':'count_cterm_NOT_in_group',\n",
    "#                                                'count_cterm_NOT_in_group':'count_NOT_cterm_in_group',\n",
    "#                                               })\n",
    "        \n",
    "\n",
    "taxonEnrichOrthoGroupStackedDfFull2['enrichment_analysis'] = 'OG'\n",
    "cols = taxonEnrichOrthoGroupStackedDfFull2.columns.tolist()\n",
    "cols = cols[-1:] + cols[:-1]\n",
    "taxonEnrichOrthoGroupStackedDfFull2 = taxonEnrichOrthoGroupStackedDfFull2[cols]\n",
    "taxonEnrichOrthoGroupStackedDfFull2.reset_index(inplace=True)\n",
    "taxonEnrichOrthoGroupStackedDfFull2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "enrichGatheredDf = pd.merge(taxonGroupEnrichStackedDfFull2,\n",
    "                            taxonEnrichOrthoGroupStackedDfFull2,\n",
    "                            on=['taxon_id', 'cterm_aa', 'enrichment_analysis', 'group_index',\n",
    "                                'count_NOT_cterm_NOT_in_group', 'count_NOT_cterm_in_group',\n",
    "                                'count_cterm_NOT_in_group', 'count_cterm_in_group', 'odds_ratio',\n",
    "                                'pvalue_fisher'],\n",
    "                            how='outer')\n",
    "\n",
    "# Add log2 odds ratio for easier plotting\n",
    "enrichGatheredDf['log2_odds_ratio'] = np.log2(enrichGatheredDf['odds_ratio'])\n",
    "\n",
    "enrichGatheredDf = enrichGatheredDf[[\n",
    "    'taxon_id',\n",
    "    'cterm_aa',\n",
    "    'enrichment_analysis',\n",
    "    'group_index',\n",
    "    'count_NOT_cterm_NOT_in_group',\n",
    "    'count_NOT_cterm_in_group',\n",
    "    'count_cterm_NOT_in_group',\n",
    "    'count_cterm_in_group',\n",
    "    'odds_ratio',\n",
    "    'log2_odds_ratio',\n",
    "    'pvalue_fisher',\n",
    "    'COG_cat_description',\n",
    "    'OG_predicted_gene_name',\n",
    "    'OG_KEGG_pathways',\n",
    "    'OG_COG_cat',\n",
    "    'OG_eggNOG_annot']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ctermGatheredDf = pd.merge(oddsRatioCtermTaxaStackedDf.reset_index(), enrichGatheredDf,\n",
    "                           on=['taxon_id', 'cterm_aa'], how='outer')\n",
    "\n",
    "# Add proportion\n",
    "ctermGatheredDf['cterm_in_group_proportion'] = \\\n",
    "    ctermGatheredDf['count_cterm_in_group'] / (ctermGatheredDf['count_cterm_in_group'] + ctermGatheredDf['count_cterm_NOT_in_group'])\n",
    "\n",
    "ctermGatheredDf = ctermGatheredDf[[\n",
    "    'taxon_id',\n",
    "    'taxon_name',\n",
    "    'taxon_rank',\n",
    "    'cterm_aa',\n",
    "    'cterm_bias_n_seq',\n",
    "    'cterm_bias_n_genomes',\n",
    "    'cterm_bias_count',\n",
    "    'cterm_bias_log2OddsRatio',\n",
    "    'cterm_bias_oddsRatio',\n",
    "    'cterm_bias_pvalue',\n",
    "    'enrichment_analysis',\n",
    "    'group_index',\n",
    "    'count_NOT_cterm_NOT_in_group',\n",
    "    'count_NOT_cterm_in_group',\n",
    "    'count_cterm_NOT_in_group',\n",
    "    'count_cterm_in_group',\n",
    "    'cterm_in_group_proportion',\n",
    "    'odds_ratio',\n",
    "    'log2_odds_ratio',\n",
    "    'pvalue_fisher',\n",
    "    'COG_cat_description',\n",
    "    'OG_predicted_gene_name',\n",
    "    'OG_KEGG_pathways',\n",
    "    'OG_COG_cat',\n",
    "    'OG_eggNOG_annot']]\n",
    "ctermGatheredDf.sort_values(['cterm_aa', 'cterm_bias_n_seq', 'taxon_rank', 'taxon_name', 'enrichment_analysis',\n",
    "                             'odds_ratio'],\n",
    "                            ascending=[True, False, True, True, True, True],\n",
    "                            inplace=True)\n",
    "print(len(ctermGatheredDf))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ctermGatheredDf.to_csv(str(analysisEggnogPath / 'ctermGatheredDf.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>taxon_rank</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th>cterm_bias_n_seq</th>\n",
       "      <th>cterm_bias_n_genomes</th>\n",
       "      <th>cterm_bias_count</th>\n",
       "      <th>cterm_bias_log2OddsRatio</th>\n",
       "      <th>cterm_bias_oddsRatio</th>\n",
       "      <th>cterm_bias_pvalue</th>\n",
       "      <th>enrichment_analysis</th>\n",
       "      <th>group_index</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>cterm_in_group_proportion</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>log2_odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>COG_cat_description</th>\n",
       "      <th>OG_predicted_gene_name</th>\n",
       "      <th>OG_KEGG_pathways</th>\n",
       "      <th>OG_COG_cat</th>\n",
       "      <th>OG_eggNOG_annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>W</td>\n",
       "      <td>1110051.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>165788.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.072778</td>\n",
       "      <td>-3.780347</td>\n",
       "      <td>4.702118e-09</td>\n",
       "      <td>Extracellular structures</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>Z</td>\n",
       "      <td>1110145.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>165787.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.223207</td>\n",
       "      <td>-2.163545</td>\n",
       "      <td>2.855893e-03</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>U</td>\n",
       "      <td>1085557.0</td>\n",
       "      <td>24678.0</td>\n",
       "      <td>163848.0</td>\n",
       "      <td>1942.0</td>\n",
       "      <td>0.011714</td>\n",
       "      <td>0.521376</td>\n",
       "      <td>-0.939605</td>\n",
       "      <td>1.033719e-198</td>\n",
       "      <td>Intracellular trafficking, secretion, and vesi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>M</td>\n",
       "      <td>1045675.0</td>\n",
       "      <td>64560.0</td>\n",
       "      <td>159418.0</td>\n",
       "      <td>6372.0</td>\n",
       "      <td>0.038434</td>\n",
       "      <td>0.647398</td>\n",
       "      <td>-0.627274</td>\n",
       "      <td>5.106927e-257</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>N</td>\n",
       "      <td>1094780.0</td>\n",
       "      <td>15455.0</td>\n",
       "      <td>163881.0</td>\n",
       "      <td>1909.0</td>\n",
       "      <td>0.011515</td>\n",
       "      <td>0.825154</td>\n",
       "      <td>-0.277264</td>\n",
       "      <td>8.300436e-16</td>\n",
       "      <td>Cell motility</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxon_id      taxon_name taxon_rank cterm_aa  cterm_bias_n_seq  \\\n",
       "0      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "1      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "2      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "3      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "4      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "\n",
       "   cterm_bias_n_genomes  cterm_bias_count  cterm_bias_log2OddsRatio  \\\n",
       "0                   691          282036.0                  0.288529   \n",
       "1                   691          282036.0                  0.288529   \n",
       "2                   691          282036.0                  0.288529   \n",
       "3                   691          282036.0                  0.288529   \n",
       "4                   691          282036.0                  0.288529   \n",
       "\n",
       "   cterm_bias_oddsRatio  cterm_bias_pvalue enrichment_analysis group_index  \\\n",
       "0              1.221394                0.0                 COG           W   \n",
       "1              1.221394                0.0                 COG           Z   \n",
       "2              1.221394                0.0                 COG           U   \n",
       "3              1.221394                0.0                 COG           M   \n",
       "4              1.221394                0.0                 COG           N   \n",
       "\n",
       "   count_NOT_cterm_NOT_in_group  count_NOT_cterm_in_group  \\\n",
       "0                     1110051.0                     184.0   \n",
       "1                     1110145.0                      90.0   \n",
       "2                     1085557.0                   24678.0   \n",
       "3                     1045675.0                   64560.0   \n",
       "4                     1094780.0                   15455.0   \n",
       "\n",
       "   count_cterm_NOT_in_group  count_cterm_in_group  cterm_in_group_proportion  \\\n",
       "0                  165788.0                   2.0                   0.000012   \n",
       "1                  165787.0                   3.0                   0.000018   \n",
       "2                  163848.0                1942.0                   0.011714   \n",
       "3                  159418.0                6372.0                   0.038434   \n",
       "4                  163881.0                1909.0                   0.011515   \n",
       "\n",
       "   odds_ratio  log2_odds_ratio  pvalue_fisher  \\\n",
       "0    0.072778        -3.780347   4.702118e-09   \n",
       "1    0.223207        -2.163545   2.855893e-03   \n",
       "2    0.521376        -0.939605  1.033719e-198   \n",
       "3    0.647398        -0.627274  5.106927e-257   \n",
       "4    0.825154        -0.277264   8.300436e-16   \n",
       "\n",
       "                                 COG_cat_description OG_predicted_gene_name  \\\n",
       "0                           Extracellular structures                    NaN   \n",
       "1                                       Cytoskeleton                    NaN   \n",
       "2  Intracellular trafficking, secretion, and vesi...                    NaN   \n",
       "3             Cell wall/membrane/envelope biogenesis                    NaN   \n",
       "4                                      Cell motility                    NaN   \n",
       "\n",
       "  OG_KEGG_pathways OG_COG_cat OG_eggNOG_annot  \n",
       "0              NaN        NaN             NaN  \n",
       "1              NaN        NaN             NaN  \n",
       "2              NaN        NaN             NaN  \n",
       "3              NaN        NaN             NaN  \n",
       "4              NaN        NaN             NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctermGatheredDf = pd.read_csv(str(analysisEggnogPath / 'ctermGatheredDf.csv'))\n",
    "ctermGatheredDf.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "ctermGatheredDf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# HTML rendering with Pandas style\n",
    "\n",
    "#     ctermGatheredDfFocusStyled = \\\n",
    "#         ctermGatheredDfFocus\\\n",
    "#             .style.apply(lambda x: background_gradient(x, cmap=colormap, vmin=-3, vmax=+3),\n",
    "#                          subset='log2 odds ratio')\n",
    "#     ctermGatheredDfFocusStyled\n",
    "\n",
    "#     with (analysisEggnogPath / 'ctermGatheredDfFocus.html').open('w') as f:\n",
    "#         f.write(ctermGatheredDfFocusStyled.render())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mwTools.plot\n",
    "colormap = mwTools.plot.get_divergent_color_map()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from xlsxpandasformatter.xlsxpandasformatter import FormatedWorksheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for ctermAA in ['T', 'P', 'M', 'K', 'R']:\n",
    "# for ctermAA in aaTable:\n",
    "# for ctermAA in ['T']:\n",
    "\n",
    "    ### Looking at COG enrichment\n",
    "    \n",
    "    ctermGatheredDfFocus = \\\n",
    "        ctermGatheredDf[\n",
    "            (ctermGatheredDf['cterm_aa'] == ctermAA)\n",
    "            & (ctermGatheredDf['enrichment_analysis'] == 'COG')\n",
    "    #             & (ctermGatheredDf['cterm_bias_oddsRatio'] < 0.6) &\n",
    "    #             (ctermGatheredDf['cterm_bias_n_seq'] > 20e3) &\n",
    "    #             & ((ctermGatheredDf['odds_ratio'] > 2) |\n",
    "    #              (ctermGatheredDf['odds_ratio'] < 0.5))\n",
    "        ]\\\n",
    "        .rename(columns=(lambda x: re.sub(r'_', r' ', x)))\n",
    "\n",
    "    df = ctermGatheredDfFocus\n",
    "    nRows = len(df)\n",
    "    nCols = len(df.columns)\n",
    "    writer = pd.ExcelWriter(str(analysisEggnogPath / 'ctermGatheredDf_cterm_{}_COG_enrichment.xlsx'.format(ctermAA)),\n",
    "                            engine='xlsxwriter', options={'nan_inf_to_errors': True})\n",
    "    df.to_excel(writer, index=False, sheet_name='sheet1')\n",
    "    workbook = writer.book\n",
    "    worksheet = writer.sheets['sheet1']\n",
    "\n",
    "    formattedWorksheet = FormatedWorksheet(worksheet, workbook, df, hasIndex=False)\n",
    "\n",
    "    headerFormat={'valign':'bottom', 'bold':True, 'bottom':2, 'text_wrap':True}\n",
    "    headerHeight = 80\n",
    "    formattedWorksheet.format_header(headerFormat=headerFormat, rowHeight=headerHeight)\n",
    "\n",
    "    formattedWorksheet.format_background_colormap('cterm bias log2OddsRatio', colormap, vmin=-3, vmax=+3)\n",
    "    formattedWorksheet.format_background_colormap('log2 odds ratio', colormap, vmin=-3, vmax=+3)\n",
    "    formattedWorksheet.format_add_separation_border_between_groups('taxon id')\n",
    "\n",
    "    numFormatScientific = '0.0E+00'\n",
    "    numFormatFloat2digits = 0x02\n",
    "    numFormatInteger = 0x01\n",
    "    colPatternFormatList = [\n",
    "        (r'(pvalue)', numFormatScientific),\n",
    "        (r'(proportion)', numFormatScientific),\n",
    "        (r'([oO]ddsRatio)|(odds ratio)', numFormatFloat2digits),\n",
    "        (r'(cterm bias n seq)|(count)', numFormatInteger)\n",
    "    ]\n",
    "    formattedWorksheet.format_numeric_cols(colPatternFormatList)\n",
    "\n",
    "    # Apply the color to the column\n",
    "    colIndexCOGdescription = formattedWorksheet.df.columns.get_loc('COG cat description')\n",
    "    colIndexCOGdescription\n",
    "\n",
    "    def format_COG_cat_description(row):\n",
    "\n",
    "        formatSeries = pd.Series(data=[dict() for _ in range(len(row))], index=row.index)\n",
    "\n",
    "        if row['group index'] in ['J', 'C']:\n",
    "            # Translation, ribosomal structure and biogenesis\n",
    "            # Energy production and conversion\n",
    "            formatSeries[colIndexCOGdescription]['font_color'] = '#0e4179'\n",
    "        elif row['group index'] == 'B':\n",
    "            # 'Chromatin structure and dynamics '\n",
    "            formatSeries[colIndexCOGdescription]['font_color'] = '#7c0722'\n",
    "\n",
    "        return formatSeries\n",
    "\n",
    "    formattedWorksheet.format_pandas_apply(format_COG_cat_description, axis=1)\n",
    "\n",
    "    formattedWorksheet.apply_format_table()\n",
    "    writer.save()\n",
    "    \n",
    "    ### Looking at OG enrichment\n",
    "\n",
    "#     ctermGatheredDfFocus = \\\n",
    "#         ctermGatheredDf[\n",
    "#             (ctermGatheredDf['cterm_aa'] == ctermAA)\n",
    "#             & (ctermGatheredDf['enrichment_analysis'] == 'OG')\n",
    "# #             & (ctermGatheredDf['cterm_bias_oddsRatio'] < 0.6)\n",
    "# #             & (ctermGatheredDf['cterm_bias_n_seq'] > 10e3)\n",
    "#     #         ((ctermGatheredDf['odds_ratio'] > 2) |\n",
    "#     #          (ctermGatheredDf['odds_ratio'] < 0.5))\n",
    "#         ]\\\n",
    "#         .rename(columns=(lambda x: re.sub(r'_', r' ', x)))\n",
    "    \n",
    "#     df = ctermGatheredDfFocus\n",
    "#     nRows = len(df)\n",
    "#     nCols = len(df.columns)\n",
    "#     writer = pd.ExcelWriter(str(analysisEggnogPath / 'ctermGatheredDf_cterm_{}_OG_enrichment.xlsx'.format(ctermAA)),\n",
    "#                             engine='xlsxwriter', options={'nan_inf_to_errors': True})\n",
    "#     df.to_excel(writer, index=False, sheet_name='sheet1')\n",
    "#     workbook = writer.book\n",
    "#     worksheet = writer.sheets['sheet1']\n",
    "    \n",
    "#     formattedWorksheet = FormatedWorksheet(worksheet, workbook, df, hasIndex=False)\n",
    "#     formattedWorksheet.format_header(headerFormat=headerFormat, rowHeight=headerHeight)\n",
    "#     formattedWorksheet.format_background_colormap('cterm bias log2OddsRatio', colormap, vmin=-3, vmax=+3)\n",
    "#     formattedWorksheet.format_add_separation_border_between_groups('taxon id')\n",
    "#     formattedWorksheet.format_numeric_cols(colPatternFormatList)\n",
    "#     formattedWorksheet.apply_format_table()\n",
    "#     writer.save()\n",
    "#     print(\"ctermAA\", ctermAA, \" done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>taxon_rank</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th>cterm_bias_n_seq</th>\n",
       "      <th>cterm_bias_n_genomes</th>\n",
       "      <th>cterm_bias_count</th>\n",
       "      <th>cterm_bias_log2OddsRatio</th>\n",
       "      <th>cterm_bias_oddsRatio</th>\n",
       "      <th>cterm_bias_pvalue</th>\n",
       "      <th>enrichment_analysis</th>\n",
       "      <th>group_index</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>cterm_in_group_proportion</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>log2_odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>COG_cat_description</th>\n",
       "      <th>OG_predicted_gene_name</th>\n",
       "      <th>OG_KEGG_pathways</th>\n",
       "      <th>OG_COG_cat</th>\n",
       "      <th>OG_eggNOG_annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>103071</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>S</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>147598.0</td>\n",
       "      <td>0.260771</td>\n",
       "      <td>1.198119</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>COG</td>\n",
       "      <td>A</td>\n",
       "      <td>1189455.0</td>\n",
       "      <td>284.0</td>\n",
       "      <td>86282.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.194164</td>\n",
       "      <td>-2.364650</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>RNA processing and modification</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103088</th>\n",
       "      <td>1236</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>class</td>\n",
       "      <td>S</td>\n",
       "      <td>907108.0</td>\n",
       "      <td>288</td>\n",
       "      <td>62501.0</td>\n",
       "      <td>0.217833</td>\n",
       "      <td>1.162985</td>\n",
       "      <td>2.670244e-277</td>\n",
       "      <td>COG</td>\n",
       "      <td>A</td>\n",
       "      <td>471226.0</td>\n",
       "      <td>185.0</td>\n",
       "      <td>34269.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.222986</td>\n",
       "      <td>-2.164976</td>\n",
       "      <td>0.001976</td>\n",
       "      <td>RNA processing and modification</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106544</th>\n",
       "      <td>85023</td>\n",
       "      <td>Microbacteriaceae</td>\n",
       "      <td>family</td>\n",
       "      <td>S</td>\n",
       "      <td>18534.0</td>\n",
       "      <td>7</td>\n",
       "      <td>1725.0</td>\n",
       "      <td>0.801149</td>\n",
       "      <td>1.742488</td>\n",
       "      <td>3.928653e-92</td>\n",
       "      <td>COG</td>\n",
       "      <td>D</td>\n",
       "      <td>7558.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>789.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.021092</td>\n",
       "      <td>2.856959</td>\n",
       "      <td>1.514480</td>\n",
       "      <td>0.000489</td>\n",
       "      <td>Cell cycle control, cell division, chromosome ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106667</th>\n",
       "      <td>775</td>\n",
       "      <td>Rickettsiaceae</td>\n",
       "      <td>family</td>\n",
       "      <td>S</td>\n",
       "      <td>15896.0</td>\n",
       "      <td>15</td>\n",
       "      <td>1231.0</td>\n",
       "      <td>0.199970</td>\n",
       "      <td>1.148674</td>\n",
       "      <td>4.766884e-06</td>\n",
       "      <td>COG</td>\n",
       "      <td>F</td>\n",
       "      <td>9648.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>842.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.004728</td>\n",
       "      <td>0.206458</td>\n",
       "      <td>-2.276078</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106716</th>\n",
       "      <td>81852</td>\n",
       "      <td>Enterococcaceae</td>\n",
       "      <td>family</td>\n",
       "      <td>S</td>\n",
       "      <td>15758.0</td>\n",
       "      <td>6</td>\n",
       "      <td>879.0</td>\n",
       "      <td>-0.104689</td>\n",
       "      <td>0.930006</td>\n",
       "      <td>3.675565e-02</td>\n",
       "      <td>COG</td>\n",
       "      <td>J</td>\n",
       "      <td>9655.0</td>\n",
       "      <td>612.0</td>\n",
       "      <td>579.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.018644</td>\n",
       "      <td>0.299719</td>\n",
       "      <td>-1.738315</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106719</th>\n",
       "      <td>1903410</td>\n",
       "      <td>Pectobacteriaceae</td>\n",
       "      <td>family</td>\n",
       "      <td>S</td>\n",
       "      <td>15451.0</td>\n",
       "      <td>4</td>\n",
       "      <td>949.0</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>1.044274</td>\n",
       "      <td>1.994609e-01</td>\n",
       "      <td>COG</td>\n",
       "      <td>J</td>\n",
       "      <td>10275.0</td>\n",
       "      <td>614.0</td>\n",
       "      <td>694.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.019774</td>\n",
       "      <td>0.337584</td>\n",
       "      <td>-1.566681</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106839</th>\n",
       "      <td>1763524</td>\n",
       "      <td>Isosphaeraceae</td>\n",
       "      <td>family</td>\n",
       "      <td>S</td>\n",
       "      <td>10771.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>0.898511</td>\n",
       "      <td>1.864141</td>\n",
       "      <td>3.583733e-71</td>\n",
       "      <td>COG</td>\n",
       "      <td>I</td>\n",
       "      <td>3281.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>383.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.047264</td>\n",
       "      <td>3.191471</td>\n",
       "      <td>1.674221</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>Lipid transport and metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        taxon_id           taxon_name taxon_rank cterm_aa  cterm_bias_n_seq  \\\n",
       "103071      1224       Proteobacteria     phylum        S         2165400.0   \n",
       "103088      1236  Gammaproteobacteria      class        S          907108.0   \n",
       "106544     85023    Microbacteriaceae     family        S           18534.0   \n",
       "106667       775       Rickettsiaceae     family        S           15896.0   \n",
       "106716     81852      Enterococcaceae     family        S           15758.0   \n",
       "106719   1903410    Pectobacteriaceae     family        S           15451.0   \n",
       "106839   1763524       Isosphaeraceae     family        S           10771.0   \n",
       "\n",
       "        cterm_bias_n_genomes  cterm_bias_count  cterm_bias_log2OddsRatio  \\\n",
       "103071                   691          147598.0                  0.260771   \n",
       "103088                   288           62501.0                  0.217833   \n",
       "106544                     7            1725.0                  0.801149   \n",
       "106667                    15            1231.0                  0.199970   \n",
       "106716                     6             879.0                 -0.104689   \n",
       "106719                     4             949.0                  0.062500   \n",
       "106839                     2            1077.0                  0.898511   \n",
       "\n",
       "        cterm_bias_oddsRatio  cterm_bias_pvalue enrichment_analysis  \\\n",
       "103071              1.198119       0.000000e+00                 COG   \n",
       "103088              1.162985      2.670244e-277                 COG   \n",
       "106544              1.742488       3.928653e-92                 COG   \n",
       "106667              1.148674       4.766884e-06                 COG   \n",
       "106716              0.930006       3.675565e-02                 COG   \n",
       "106719              1.044274       1.994609e-01                 COG   \n",
       "106839              1.864141       3.583733e-71                 COG   \n",
       "\n",
       "       group_index  count_NOT_cterm_NOT_in_group  count_NOT_cterm_in_group  \\\n",
       "103071           A                     1189455.0                     284.0   \n",
       "103088           A                      471226.0                     185.0   \n",
       "106544           D                        7558.0                      57.0   \n",
       "106667           F                        9648.0                     222.0   \n",
       "106716           J                        9655.0                     612.0   \n",
       "106719           J                       10275.0                     614.0   \n",
       "106839           I                        3281.0                      51.0   \n",
       "\n",
       "        count_cterm_NOT_in_group  count_cterm_in_group  \\\n",
       "103071                   86282.0                   4.0   \n",
       "103088                   34269.0                   3.0   \n",
       "106544                     789.0                  17.0   \n",
       "106667                     842.0                   4.0   \n",
       "106716                     579.0                  11.0   \n",
       "106719                     694.0                  14.0   \n",
       "106839                     383.0                  19.0   \n",
       "\n",
       "        cterm_in_group_proportion  odds_ratio  log2_odds_ratio  pvalue_fisher  \\\n",
       "103071                   0.000046    0.194164        -2.364650       0.000030   \n",
       "103088                   0.000088    0.222986        -2.164976       0.001976   \n",
       "106544                   0.021092    2.856959         1.514480       0.000489   \n",
       "106667                   0.004728    0.206458        -2.276078       0.000090   \n",
       "106716                   0.018644    0.299719        -1.738315       0.000004   \n",
       "106719                   0.019774    0.337584        -1.566681       0.000004   \n",
       "106839                   0.047264    3.191471         1.674221       0.000092   \n",
       "\n",
       "                                      COG_cat_description  \\\n",
       "103071                    RNA processing and modification   \n",
       "103088                    RNA processing and modification   \n",
       "106544  Cell cycle control, cell division, chromosome ...   \n",
       "106667                Nucleotide transport and metabolism   \n",
       "106716    Translation, ribosomal structure and biogenesis   \n",
       "106719    Translation, ribosomal structure and biogenesis   \n",
       "106839                     Lipid transport and metabolism   \n",
       "\n",
       "       OG_predicted_gene_name OG_KEGG_pathways OG_COG_cat OG_eggNOG_annot  \n",
       "103071                    NaN              NaN        NaN             NaN  \n",
       "103088                    NaN              NaN        NaN             NaN  \n",
       "106544                    NaN              NaN        NaN             NaN  \n",
       "106667                    NaN              NaN        NaN             NaN  \n",
       "106716                    NaN              NaN        NaN             NaN  \n",
       "106719                    NaN              NaN        NaN             NaN  \n",
       "106839                    NaN              NaN        NaN             NaN  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = ctermGatheredDf[(ctermGatheredDf['enrichment_analysis'] == 'COG') &\n",
    "                     (ctermGatheredDf['cterm_bias_n_genomes'] >= 1) &\n",
    "                     (ctermGatheredDf['cterm_aa'] == 'S') &\n",
    "#                      (ctermGatheredDf['taxon_rank'] == 'order') &\n",
    "                     (ctermGatheredDf['cterm_bias_n_seq'] > 1e4) &\n",
    "                     (ctermGatheredDf['log2_odds_ratio'].map(lambda x: x < -1.5 or x > +1.5))\n",
    "                    ]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## C-terminal bias for each COG category group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final set of EggNOG annotated sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In our pipeline, we gather the data in the following way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/users/lserrano/mweber/Analysis_data/C-terminal/Eggnog')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysisEggnogPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1429\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0           0  superkingdom       2.0             Bacteria   \n",
       "1           1        phylum    1224.0       Proteobacteria   \n",
       "2           1        phylum    1239.0           Firmicutes   \n",
       "3           3         class    1236.0  Gammaproteobacteria   \n",
       "4           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "   n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                 1582                                        949  \n",
       "1                  697                                        407  \n",
       "2                  259                                        170  \n",
       "3                  292                                        162  \n",
       "4                  226                                        126  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaEggnogDf = pd.read_csv(str(analysisEggnogPath / 'taxaEggnogDf.csv'), index_col=0).reset_index(drop=True)\n",
    "print(len(taxaEggnogDf))\n",
    "taxaEggnogDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonId = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxonSuffix taxaId_2\n",
      "len(multispeciesAllCDSDf_nr) 4871983\n"
     ]
    }
   ],
   "source": [
    "# Importing list of genome assemblies with their taxonomy in the RefSeq - EggNOG common set of species\n",
    "assemblySummaryCommonSpeciesDf = pd.read_csv(str(analysisEggnogPath / 'assemblySummaryCommonSpeciesDf.csv'))\n",
    "\n",
    "taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "print(\"taxonSuffix\", taxonSuffix)\n",
    "\n",
    "# Import all protein sequences in taxon group from file\n",
    "multispeciesAllCDSDf_nr = pd.read_csv(str(analysisCtermDataPath / 'Taxonomy' / taxonSuffix / (taxonSuffix + '_multispeciesAllCDSDf_nr.csv')))\n",
    "print(\"len(multispeciesAllCDSDf_nr)\", len(multispeciesAllCDSDf_nr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nSeqTaxonRefSeqEggNOGCommon: 2866659 nSeqTaxonRefSeq 4871983 ratio 0.5883967575420522\n",
      "Nb of subspecies in taxon group, RefSeq: 946\n",
      "Nb of taxid at rank species in taxon group, RefSeq: 893\n",
      "Nb of subspecies in taxon group, common: 946\n",
      "Nb of taxid at rank species, common: 893\n",
      "taxonAllCDSDf:\n",
      "   refSeqProteinId  proteinGI               species_name genome_accession  \\\n",
      "0  WP_012960546.1  502725562  Bacillus pseudofirmus OF4  GCF_000005825.2   \n",
      "1  WP_012960083.1  502725099  Bacillus pseudofirmus OF4  GCF_000005825.2   \n",
      "2  WP_012957129.1  502722145  Bacillus pseudofirmus OF4  GCF_000005825.2   \n",
      "3  WP_012957437.1  502722453  Bacillus pseudofirmus OF4  GCF_000005825.2   \n",
      "4  WP_012958269.1  502723285  Bacillus pseudofirmus OF4  GCF_000005825.2   \n",
      "\n",
      "                                              DNASeq stopCodon  \\\n",
      "0  ATGTCTAGAAAGAAGTACAACCCTTTCATTATGACGATCATGATTT...       TAA   \n",
      "1  TTGACAATCTATCAAGATATTATACGTATATTTGAAGAGAATTCTT...       TAA   \n",
      "2  TTGACAGGTCAACTTGTTCAGTATGGACGCCACCGCCAACGTAGAA...       TAA   \n",
      "3  ATGAAAACGATCGCTGCATTTGCCTCAAACCGTAAAGGGGTGTGGA...       TGA   \n",
      "4  ATGATTAATGAAAAACGTTTAGTAGATGAATTTTTAGAGCTTGTAC...       TAA   \n",
      "\n",
      "                       DNA3UTRSeq  codonTableBio  \\\n",
      "0  TACTTTGAAAAGGGCCCCAATGAATTTATT             11   \n",
      "1  GCCTTATTAAAAATAGTAGTTAAACTTACG             11   \n",
      "2  CGAATAAAAGTATAGAAAAAAGCGATGCGG             11   \n",
      "3  ATAAAACAAAAGTATTATTAAGATTCTCAG             11   \n",
      "4  TTTAGGGGTAAAGAGACCGGTGGACTTAAG             11   \n",
      "\n",
      "                                          proteinSeq  species_taxid     taxid  \\\n",
      "0  MSRKKYNPFIMTIMIFFIVFSSFAPAVAATASQTAGNEKLKIDIDE...        79885.0  398511.0   \n",
      "1  MTIYQDIIRIFEENSSYKNPTQAVNQASSLNALSGDLYTDANRFIY...        79885.0  398511.0   \n",
      "2  MTGQLVQYGRHRQRRSYARINEVLELPNLIEIQTASYQWFLDEGLR...        79885.0  398511.0   \n",
      "3  MKTIAAFASNRKGVWIILLAWVIIAGALGMAPSANDYTVNTGENDL...        79885.0  398511.0   \n",
      "4  MINEKRLVDEFLELVQIDSETKFEAEIAPILTEKFTSLGVEVVEDD...        79885.0  398511.0   \n",
      "\n",
      "  seed_eggNOG_ortholog  seed_ortholog_evalue  seed_ortholog_score  \\\n",
      "0   398511.BpOF4_06055          0.000000e+00               5924.2   \n",
      "1   398511.BpOF4_03710          0.000000e+00               3795.1   \n",
      "2   398511.BpOF4_08510          0.000000e+00               2645.4   \n",
      "3   398511.BpOF4_10080          0.000000e+00               1588.8   \n",
      "4     525903.Taci_0027         3.800000e-104                339.3   \n",
      "\n",
      "  predicted_gene_name                                           GO_terms  \\\n",
      "0                 NaN                                                NaN   \n",
      "1                 NaN                                                NaN   \n",
      "2                RPOB  GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...   \n",
      "3                 NaN  GO:0005575,GO:0005623,GO:0005886,GO:0016020,GO...   \n",
      "4                 NaN                                                NaN   \n",
      "\n",
      "                         KEGG_pathways Annotation_tax_scope  \\\n",
      "0                                  NaN          bactNOG[38]   \n",
      "1                                  NaN          bactNOG[38]   \n",
      "2  map00230,map00240,map01100,map03020          bactNOG[38]   \n",
      "3                                  NaN          bactNOG[38]   \n",
      "4                                  NaN          bactNOG[38]   \n",
      "\n",
      "                                                 OGs COG cat  \\\n",
      "0              07I90@bactNOG,0P1K8@firmNOG,11MJ3@NOG       S   \n",
      "1              08NWM@bactNOG,0NDUX@firmNOG,0XPG4@NOG       S   \n",
      "2  054MH@bacNOG,08IIJ@bactNOG,0P68T@firmNOG,COG00...       K   \n",
      "3                          05EYF@bactNOG,COG2409@NOG       S   \n",
      "4             07TGY@bactNOG,1C8R8@synNOG,COG2195@NOG       E   \n",
      "\n",
      "                                        eggNOG annot bestOG  bestOG_evalue  \\\n",
      "0           Inherit from NOG: s-layer domain protein  07I90   0.000000e+00   \n",
      "1                                                NaN  08NWM  3.494923e-234   \n",
      "2  DNA-dependent RNA polymerase catalyzes the tra...  08IIJ   0.000000e+00   \n",
      "3                                MMPL domain protein  05EYF  4.047655e-214   \n",
      "4                                          peptidase  07TGY  7.410892e-137   \n",
      "\n",
      "   bestOG_score  \n",
      "0   6053.847656  \n",
      "1    783.924255  \n",
      "2   1947.114868  \n",
      "3    716.653442  \n",
      "4    458.899597   \n",
      "\n",
      "Nb of protein sequences with EggNOG orthogroup annotation: 2561231 / 2877631 = 0.8900484461002818\n"
     ]
    }
   ],
   "source": [
    "# Choose protein sequences from genomes that are in the common set of species RefSeq - EggNOG\n",
    "multispeciesAllCDSDf_nr2 = pd.merge(multispeciesAllCDSDf_nr,\n",
    "                                    assemblySummaryCommonSpeciesDf.set_index('assembly_accession')[['species_taxid', 'taxid']],\n",
    "                                    left_on='genome_accession', right_index=True, how='left')\n",
    "taxonAllCDSEggNOGSpeciesDf = multispeciesAllCDSDf_nr2.dropna(subset=['taxid'], axis=0)\n",
    "nSeqTaxonRefSeq = len(multispeciesAllCDSDf_nr2)\n",
    "nSeqTaxonRefSeqEggNOGCommon = len(taxonAllCDSEggNOGSpeciesDf)\n",
    "print(\"nSeqTaxonRefSeqEggNOGCommon:\", nSeqTaxonRefSeqEggNOGCommon, \"nSeqTaxonRefSeq\", nSeqTaxonRefSeq,\n",
    "      \"ratio\", nSeqTaxonRefSeqEggNOGCommon/nSeqTaxonRefSeq)\n",
    "print(\"Nb of subspecies in taxon group, RefSeq:\", len(set(multispeciesAllCDSDf_nr2['taxid'].dropna())))\n",
    "print(\"Nb of taxid at rank species in taxon group, RefSeq:\", len(set(multispeciesAllCDSDf_nr2['species_taxid'].dropna())))\n",
    "print(\"Nb of subspecies in taxon group, common:\", len(set(taxonAllCDSEggNOGSpeciesDf['taxid'])))\n",
    "print(\"Nb of taxid at rank species, common:\", len(set(taxonAllCDSEggNOGSpeciesDf['species_taxid'])))\n",
    "\n",
    "# Import eggnog-mapper protein annotations\n",
    "eggnogMapperAnnotationResultDf = pd.read_csv(str(analysisEggnogPath / 'eggnogMapperAnnotationResultDf.csv'))\n",
    "eggnogMapperAnnotationResultDf.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "eggnogMapperOrthogroupAnnotationDf = pd.read_csv(str(analysisEggnogPath / 'eggnogMapperOrthogroupAnnotationDf.csv'))\n",
    "\n",
    "# Merging eggnog-mapper annotations with the taxonAllCDS\n",
    "taxonAllCDSDf = pd.merge(taxonAllCDSEggNOGSpeciesDf.reset_index(),\n",
    "                         eggnogMapperAnnotationResultDf,\n",
    "                         on='refSeqProteinId', how='left', suffixes=['_refseq', '_eggnog'])\n",
    "taxonAllCDSDf.drop('index', axis=1, inplace=True)\n",
    "taxonAllCDSDf.drop('genome_accession_eggnog', axis=1, inplace=True)\n",
    "taxonAllCDSDf.rename(columns={'genome_accession_refseq':'genome_accession'}, inplace=True)\n",
    "taxonAllCDSDf.drop('taxid_eggnog', axis=1, inplace=True)\n",
    "taxonAllCDSDf.rename(columns={'taxid_refseq':'taxid'}, inplace=True)\n",
    "print(\"taxonAllCDSDf:\\n\", taxonAllCDSDf.head(), \"\\n\")\n",
    "\n",
    "print(\"Nb of protein sequences with EggNOG orthogroup annotation:\",\n",
    "      taxonAllCDSDf['COG cat'].notnull().sum(), \"/\", len(taxonAllCDSDf), \"=\",\n",
    "      taxonAllCDSDf['COG cat'].notnull().sum()/len(taxonAllCDSDf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonAllCDSDf.to_csv(analysisEggnogPath / 'taxonAllCDSDf_bacteria.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>taxon_rank</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th>cterm_bias_n_seq</th>\n",
       "      <th>cterm_bias_n_genomes</th>\n",
       "      <th>cterm_bias_count</th>\n",
       "      <th>cterm_bias_log2OddsRatio</th>\n",
       "      <th>cterm_bias_oddsRatio</th>\n",
       "      <th>cterm_bias_pvalue</th>\n",
       "      <th>enrichment_analysis</th>\n",
       "      <th>group_index</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>cterm_in_group_proportion</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>log2_odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>COG_cat_description</th>\n",
       "      <th>OG_predicted_gene_name</th>\n",
       "      <th>OG_KEGG_pathways</th>\n",
       "      <th>OG_COG_cat</th>\n",
       "      <th>OG_eggNOG_annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>W</td>\n",
       "      <td>1110051.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>165788.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.072778</td>\n",
       "      <td>-3.780347</td>\n",
       "      <td>4.702118e-09</td>\n",
       "      <td>Extracellular structures</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>Z</td>\n",
       "      <td>1110145.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>165787.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.223207</td>\n",
       "      <td>-2.163545</td>\n",
       "      <td>2.855893e-03</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>U</td>\n",
       "      <td>1085557.0</td>\n",
       "      <td>24678.0</td>\n",
       "      <td>163848.0</td>\n",
       "      <td>1942.0</td>\n",
       "      <td>0.011714</td>\n",
       "      <td>0.521376</td>\n",
       "      <td>-0.939605</td>\n",
       "      <td>1.033719e-198</td>\n",
       "      <td>Intracellular trafficking, secretion, and vesi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>M</td>\n",
       "      <td>1045675.0</td>\n",
       "      <td>64560.0</td>\n",
       "      <td>159418.0</td>\n",
       "      <td>6372.0</td>\n",
       "      <td>0.038434</td>\n",
       "      <td>0.647398</td>\n",
       "      <td>-0.627274</td>\n",
       "      <td>5.106927e-257</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>COG</td>\n",
       "      <td>N</td>\n",
       "      <td>1094780.0</td>\n",
       "      <td>15455.0</td>\n",
       "      <td>163881.0</td>\n",
       "      <td>1909.0</td>\n",
       "      <td>0.011515</td>\n",
       "      <td>0.825154</td>\n",
       "      <td>-0.277264</td>\n",
       "      <td>8.300436e-16</td>\n",
       "      <td>Cell motility</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxon_id      taxon_name taxon_rank cterm_aa  cterm_bias_n_seq  \\\n",
       "0      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "1      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "2      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "3      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "4      1224  Proteobacteria     phylum        A         2165400.0   \n",
       "\n",
       "   cterm_bias_n_genomes  cterm_bias_count  cterm_bias_log2OddsRatio  \\\n",
       "0                   691          282036.0                  0.288529   \n",
       "1                   691          282036.0                  0.288529   \n",
       "2                   691          282036.0                  0.288529   \n",
       "3                   691          282036.0                  0.288529   \n",
       "4                   691          282036.0                  0.288529   \n",
       "\n",
       "   cterm_bias_oddsRatio  cterm_bias_pvalue enrichment_analysis group_index  \\\n",
       "0              1.221394                0.0                 COG           W   \n",
       "1              1.221394                0.0                 COG           Z   \n",
       "2              1.221394                0.0                 COG           U   \n",
       "3              1.221394                0.0                 COG           M   \n",
       "4              1.221394                0.0                 COG           N   \n",
       "\n",
       "   count_NOT_cterm_NOT_in_group  count_NOT_cterm_in_group  \\\n",
       "0                     1110051.0                     184.0   \n",
       "1                     1110145.0                      90.0   \n",
       "2                     1085557.0                   24678.0   \n",
       "3                     1045675.0                   64560.0   \n",
       "4                     1094780.0                   15455.0   \n",
       "\n",
       "   count_cterm_NOT_in_group  count_cterm_in_group  cterm_in_group_proportion  \\\n",
       "0                  165788.0                   2.0                   0.000012   \n",
       "1                  165787.0                   3.0                   0.000018   \n",
       "2                  163848.0                1942.0                   0.011714   \n",
       "3                  159418.0                6372.0                   0.038434   \n",
       "4                  163881.0                1909.0                   0.011515   \n",
       "\n",
       "   odds_ratio  log2_odds_ratio  pvalue_fisher  \\\n",
       "0    0.072778        -3.780347   4.702118e-09   \n",
       "1    0.223207        -2.163545   2.855893e-03   \n",
       "2    0.521376        -0.939605  1.033719e-198   \n",
       "3    0.647398        -0.627274  5.106927e-257   \n",
       "4    0.825154        -0.277264   8.300436e-16   \n",
       "\n",
       "                                 COG_cat_description OG_predicted_gene_name  \\\n",
       "0                           Extracellular structures                    NaN   \n",
       "1                                       Cytoskeleton                    NaN   \n",
       "2  Intracellular trafficking, secretion, and vesi...                    NaN   \n",
       "3             Cell wall/membrane/envelope biogenesis                    NaN   \n",
       "4                                      Cell motility                    NaN   \n",
       "\n",
       "  OG_KEGG_pathways OG_COG_cat OG_eggNOG_annot  \n",
       "0              NaN        NaN             NaN  \n",
       "1              NaN        NaN             NaN  \n",
       "2              NaN        NaN             NaN  \n",
       "3              NaN        NaN             NaN  \n",
       "4              NaN        NaN             NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctermGatheredDf = pd.read_csv(str(analysisEggnogPath / 'ctermGatheredDf.csv'))\n",
    "ctermGatheredDf.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "ctermGatheredDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>taxon_rank</th>\n",
       "      <th>cterm_aa</th>\n",
       "      <th>cterm_bias_n_seq</th>\n",
       "      <th>cterm_bias_n_genomes</th>\n",
       "      <th>cterm_bias_count</th>\n",
       "      <th>cterm_bias_log2OddsRatio</th>\n",
       "      <th>cterm_bias_oddsRatio</th>\n",
       "      <th>cterm_bias_pvalue</th>\n",
       "      <th>enrichment_analysis</th>\n",
       "      <th>group_index</th>\n",
       "      <th>count_NOT_cterm_NOT_in_group</th>\n",
       "      <th>count_NOT_cterm_in_group</th>\n",
       "      <th>count_cterm_NOT_in_group</th>\n",
       "      <th>count_cterm_in_group</th>\n",
       "      <th>cterm_in_group_proportion</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>log2_odds_ratio</th>\n",
       "      <th>pvalue_fisher</th>\n",
       "      <th>COG_cat_description</th>\n",
       "      <th>OG_predicted_gene_name</th>\n",
       "      <th>OG_KEGG_pathways</th>\n",
       "      <th>OG_COG_cat</th>\n",
       "      <th>OG_eggNOG_annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "      <td>A</td>\n",
       "      <td>2165400.0</td>\n",
       "      <td>691</td>\n",
       "      <td>282036.0</td>\n",
       "      <td>0.288529</td>\n",
       "      <td>1.221394</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>COG</td>\n",
       "      <td>W</td>\n",
       "      <td>1110051.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>165788.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.072778</td>\n",
       "      <td>-3.780347</td>\n",
       "      <td>4.702118e-09</td>\n",
       "      <td>Extracellular structures</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1236</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>class</td>\n",
       "      <td>A</td>\n",
       "      <td>907108.0</td>\n",
       "      <td>288</td>\n",
       "      <td>97855.0</td>\n",
       "      <td>0.127226</td>\n",
       "      <td>1.092192</td>\n",
       "      <td>8.434666e-146</td>\n",
       "      <td>COG</td>\n",
       "      <td>W</td>\n",
       "      <td>452391.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>53201.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-inf</td>\n",
       "      <td>9.264406e-05</td>\n",
       "      <td>Extracellular structures</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>28216</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>class</td>\n",
       "      <td>A</td>\n",
       "      <td>350719.0</td>\n",
       "      <td>108</td>\n",
       "      <td>57119.0</td>\n",
       "      <td>0.493002</td>\n",
       "      <td>1.407370</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>COG</td>\n",
       "      <td>W</td>\n",
       "      <td>167203.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>32959.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.110284</td>\n",
       "      <td>-3.180706</td>\n",
       "      <td>4.709665e-03</td>\n",
       "      <td>Extracellular structures</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>80840</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>order</td>\n",
       "      <td>A</td>\n",
       "      <td>244299.0</td>\n",
       "      <td>66</td>\n",
       "      <td>43671.0</td>\n",
       "      <td>0.582566</td>\n",
       "      <td>1.497510</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>COG</td>\n",
       "      <td>W</td>\n",
       "      <td>108457.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>24033.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-inf</td>\n",
       "      <td>9.661445e-04</td>\n",
       "      <td>Extracellular structures</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1443</th>\n",
       "      <td>28221</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>class</td>\n",
       "      <td>A</td>\n",
       "      <td>237872.0</td>\n",
       "      <td>58</td>\n",
       "      <td>24943.0</td>\n",
       "      <td>-0.072842</td>\n",
       "      <td>0.950763</td>\n",
       "      <td>3.515301e-14</td>\n",
       "      <td>COG</td>\n",
       "      <td>Z</td>\n",
       "      <td>144836.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>17234.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-inf</td>\n",
       "      <td>1.042534e-03</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4500</th>\n",
       "      <td>29</td>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>order</td>\n",
       "      <td>A</td>\n",
       "      <td>91404.0</td>\n",
       "      <td>13</td>\n",
       "      <td>11604.0</td>\n",
       "      <td>-0.021559</td>\n",
       "      <td>0.985167</td>\n",
       "      <td>1.338548e-01</td>\n",
       "      <td>COG</td>\n",
       "      <td>Z</td>\n",
       "      <td>40153.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>6175.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-inf</td>\n",
       "      <td>7.264395e-03</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6009</th>\n",
       "      <td>1161</td>\n",
       "      <td>Nostocales</td>\n",
       "      <td>order</td>\n",
       "      <td>A</td>\n",
       "      <td>55974.0</td>\n",
       "      <td>11</td>\n",
       "      <td>3276.0</td>\n",
       "      <td>-0.498777</td>\n",
       "      <td>0.707707</td>\n",
       "      <td>1.655817e-90</td>\n",
       "      <td>COG</td>\n",
       "      <td>U</td>\n",
       "      <td>13790.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>916.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-inf</td>\n",
       "      <td>7.727454e-05</td>\n",
       "      <td>Intracellular trafficking, secretion, and vesi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7341</th>\n",
       "      <td>1162</td>\n",
       "      <td>Nostocaceae</td>\n",
       "      <td>family</td>\n",
       "      <td>A</td>\n",
       "      <td>33780.0</td>\n",
       "      <td>7</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>-0.494139</td>\n",
       "      <td>0.709985</td>\n",
       "      <td>9.906988e-55</td>\n",
       "      <td>COG</td>\n",
       "      <td>U</td>\n",
       "      <td>13790.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>916.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-inf</td>\n",
       "      <td>7.727454e-05</td>\n",
       "      <td>Intracellular trafficking, secretion, and vesi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8331</th>\n",
       "      <td>2092</td>\n",
       "      <td>Mycoplasmataceae</td>\n",
       "      <td>family</td>\n",
       "      <td>A</td>\n",
       "      <td>23456.0</td>\n",
       "      <td>33</td>\n",
       "      <td>775.0</td>\n",
       "      <td>-0.673758</td>\n",
       "      <td>0.626872</td>\n",
       "      <td>6.314348e-43</td>\n",
       "      <td>COG</td>\n",
       "      <td>O</td>\n",
       "      <td>14920.0</td>\n",
       "      <td>414.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.003546</td>\n",
       "      <td>0.128251</td>\n",
       "      <td>-2.962953</td>\n",
       "      <td>6.936636e-05</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8336</th>\n",
       "      <td>2085</td>\n",
       "      <td>Mycoplasmatales</td>\n",
       "      <td>order</td>\n",
       "      <td>A</td>\n",
       "      <td>23456.0</td>\n",
       "      <td>33</td>\n",
       "      <td>775.0</td>\n",
       "      <td>-0.673758</td>\n",
       "      <td>0.626872</td>\n",
       "      <td>6.314348e-43</td>\n",
       "      <td>COG</td>\n",
       "      <td>O</td>\n",
       "      <td>14920.0</td>\n",
       "      <td>414.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.003546</td>\n",
       "      <td>0.128251</td>\n",
       "      <td>-2.962953</td>\n",
       "      <td>6.936636e-05</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8932</th>\n",
       "      <td>772</td>\n",
       "      <td>Bartonellaceae</td>\n",
       "      <td>family</td>\n",
       "      <td>A</td>\n",
       "      <td>12356.0</td>\n",
       "      <td>9</td>\n",
       "      <td>815.0</td>\n",
       "      <td>-0.397371</td>\n",
       "      <td>0.759240</td>\n",
       "      <td>3.371209e-15</td>\n",
       "      <td>COG</td>\n",
       "      <td>H</td>\n",
       "      <td>7664.0</td>\n",
       "      <td>320.0</td>\n",
       "      <td>564.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.005291</td>\n",
       "      <td>0.127394</td>\n",
       "      <td>-2.972635</td>\n",
       "      <td>9.280297e-07</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9346</th>\n",
       "      <td>1643685</td>\n",
       "      <td>Borreliaceae</td>\n",
       "      <td>family</td>\n",
       "      <td>A</td>\n",
       "      <td>3939.0</td>\n",
       "      <td>6</td>\n",
       "      <td>66.0</td>\n",
       "      <td>-1.493063</td>\n",
       "      <td>0.355258</td>\n",
       "      <td>5.221169e-23</td>\n",
       "      <td>COG</td>\n",
       "      <td>S</td>\n",
       "      <td>2596.0</td>\n",
       "      <td>566.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.016393</td>\n",
       "      <td>0.076443</td>\n",
       "      <td>-3.709474</td>\n",
       "      <td>1.375300e-04</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      taxon_id           taxon_name taxon_rank cterm_aa  cterm_bias_n_seq  \\\n",
       "0         1224       Proteobacteria     phylum        A         2165400.0   \n",
       "33        1236  Gammaproteobacteria      class        A          907108.0   \n",
       "107      28216   Betaproteobacteria      class        A          350719.0   \n",
       "1179     80840      Burkholderiales      order        A          244299.0   \n",
       "1443     28221  Deltaproteobacteria      class        A          237872.0   \n",
       "4500        29         Myxococcales      order        A           91404.0   \n",
       "6009      1161           Nostocales      order        A           55974.0   \n",
       "7341      1162          Nostocaceae     family        A           33780.0   \n",
       "8331      2092     Mycoplasmataceae     family        A           23456.0   \n",
       "8336      2085      Mycoplasmatales      order        A           23456.0   \n",
       "8932       772       Bartonellaceae     family        A           12356.0   \n",
       "9346   1643685         Borreliaceae     family        A            3939.0   \n",
       "\n",
       "      cterm_bias_n_genomes  cterm_bias_count  cterm_bias_log2OddsRatio  \\\n",
       "0                      691          282036.0                  0.288529   \n",
       "33                     288           97855.0                  0.127226   \n",
       "107                    108           57119.0                  0.493002   \n",
       "1179                    66           43671.0                  0.582566   \n",
       "1443                    58           24943.0                 -0.072842   \n",
       "4500                    13           11604.0                 -0.021559   \n",
       "6009                    11            3276.0                 -0.498777   \n",
       "7341                     7            1996.0                 -0.494139   \n",
       "8331                    33             775.0                 -0.673758   \n",
       "8336                    33             775.0                 -0.673758   \n",
       "8932                     9             815.0                 -0.397371   \n",
       "9346                     6              66.0                 -1.493063   \n",
       "\n",
       "      cterm_bias_oddsRatio  cterm_bias_pvalue enrichment_analysis group_index  \\\n",
       "0                 1.221394       0.000000e+00                 COG           W   \n",
       "33                1.092192      8.434666e-146                 COG           W   \n",
       "107               1.407370       0.000000e+00                 COG           W   \n",
       "1179              1.497510       0.000000e+00                 COG           W   \n",
       "1443              0.950763       3.515301e-14                 COG           Z   \n",
       "4500              0.985167       1.338548e-01                 COG           Z   \n",
       "6009              0.707707       1.655817e-90                 COG           U   \n",
       "7341              0.709985       9.906988e-55                 COG           U   \n",
       "8331              0.626872       6.314348e-43                 COG           O   \n",
       "8336              0.626872       6.314348e-43                 COG           O   \n",
       "8932              0.759240       3.371209e-15                 COG           H   \n",
       "9346              0.355258       5.221169e-23                 COG           S   \n",
       "\n",
       "      count_NOT_cterm_NOT_in_group  count_NOT_cterm_in_group  \\\n",
       "0                        1110051.0                     184.0   \n",
       "33                        452391.0                      91.0   \n",
       "107                       167203.0                      46.0   \n",
       "1179                      108457.0                      37.0   \n",
       "1443                      144836.0                      67.0   \n",
       "4500                       40153.0                      38.0   \n",
       "6009                       13790.0                     158.0   \n",
       "7341                       13790.0                     158.0   \n",
       "8331                       14920.0                     414.0   \n",
       "8336                       14920.0                     414.0   \n",
       "8932                        7664.0                     320.0   \n",
       "9346                        2596.0                     566.0   \n",
       "\n",
       "      count_cterm_NOT_in_group  count_cterm_in_group  \\\n",
       "0                     165788.0                   2.0   \n",
       "33                     53201.0                   0.0   \n",
       "107                    32959.0                   1.0   \n",
       "1179                   24033.0                   0.0   \n",
       "1443                   17234.0                   0.0   \n",
       "4500                    6175.0                   0.0   \n",
       "6009                     916.0                   0.0   \n",
       "7341                     916.0                   0.0   \n",
       "8331                     562.0                   2.0   \n",
       "8336                     562.0                   2.0   \n",
       "8932                     564.0                   3.0   \n",
       "9346                      60.0                   1.0   \n",
       "\n",
       "      cterm_in_group_proportion  odds_ratio  log2_odds_ratio  pvalue_fisher  \\\n",
       "0                      0.000012    0.072778        -3.780347   4.702118e-09   \n",
       "33                     0.000000    0.000000             -inf   9.264406e-05   \n",
       "107                    0.000030    0.110284        -3.180706   4.709665e-03   \n",
       "1179                   0.000000    0.000000             -inf   9.661445e-04   \n",
       "1443                   0.000000    0.000000             -inf   1.042534e-03   \n",
       "4500                   0.000000    0.000000             -inf   7.264395e-03   \n",
       "6009                   0.000000    0.000000             -inf   7.727454e-05   \n",
       "7341                   0.000000    0.000000             -inf   7.727454e-05   \n",
       "8331                   0.003546    0.128251        -2.962953   6.936636e-05   \n",
       "8336                   0.003546    0.128251        -2.962953   6.936636e-05   \n",
       "8932                   0.005291    0.127394        -2.972635   9.280297e-07   \n",
       "9346                   0.016393    0.076443        -3.709474   1.375300e-04   \n",
       "\n",
       "                                    COG_cat_description  \\\n",
       "0                              Extracellular structures   \n",
       "33                             Extracellular structures   \n",
       "107                            Extracellular structures   \n",
       "1179                           Extracellular structures   \n",
       "1443                                       Cytoskeleton   \n",
       "4500                                       Cytoskeleton   \n",
       "6009  Intracellular trafficking, secretion, and vesi...   \n",
       "7341  Intracellular trafficking, secretion, and vesi...   \n",
       "8331  Posttranslational modification, protein turnov...   \n",
       "8336  Posttranslational modification, protein turnov...   \n",
       "8932                  Coenzyme transport and metabolism   \n",
       "9346                                   Function unknown   \n",
       "\n",
       "     OG_predicted_gene_name OG_KEGG_pathways OG_COG_cat OG_eggNOG_annot  \n",
       "0                       NaN              NaN        NaN             NaN  \n",
       "33                      NaN              NaN        NaN             NaN  \n",
       "107                     NaN              NaN        NaN             NaN  \n",
       "1179                    NaN              NaN        NaN             NaN  \n",
       "1443                    NaN              NaN        NaN             NaN  \n",
       "4500                    NaN              NaN        NaN             NaN  \n",
       "6009                    NaN              NaN        NaN             NaN  \n",
       "7341                    NaN              NaN        NaN             NaN  \n",
       "8331                    NaN              NaN        NaN             NaN  \n",
       "8336                    NaN              NaN        NaN             NaN  \n",
       "8932                    NaN              NaN        NaN             NaN  \n",
       "9346                    NaN              NaN        NaN             NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = ctermGatheredDf[ctermGatheredDf['cterm_aa'] == 'A']\n",
    "df = df[df['enrichment_analysis'] == 'COG']\n",
    "df[df['log2_odds_ratio'] < -2.5]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "512px",
    "width": "252px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "931px",
    "left": "0px",
    "right": "1082px",
    "top": "107px",
    "width": "339px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
