{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "833"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.getpid()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### from IPython.display import Image\n",
    "\n",
    "import matplotlib_options\n",
    "matplotlib_options.define_matplotlib_backend_options(backend='agg')\n",
    "\n",
    "%matplotlib inline\n",
    "# Use autoreload to keep up-to-date version of custom packages\n",
    "# %load_ext autoreload\n",
    "import mwTools\n",
    "import cterminal\n",
    "# %autoreload 2\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from cterminal import colorBackground1, get_line_width, vmax, cmap, colorHeatmapLine\n",
    "from matplotlib.font_manager import FontProperties\n",
    "from matplotlib import transforms\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib import gridspec\n",
    "from matplotlib import colors\n",
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "import numpy as np\n",
    "from numpy import exp, log2, log10, log\n",
    "import pandas as pd\n",
    "idx = pd.IndexSlice\n",
    "import collections\n",
    "from collections import Counter\n",
    "import json\n",
    "import argparse\n",
    "import subprocess\n",
    "import shlex\n",
    "import itertools\n",
    "import os.path\n",
    "import re\n",
    "import gzip\n",
    "import urllib\n",
    "import gc\n",
    "import time\n",
    "from pathlib import Path\n",
    "import random\n",
    "import math\n",
    "\n",
    "import Bio\n",
    "from Bio import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation, ExactPosition\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "from importlib import reload\n",
    "import mwTools.bio\n",
    "reload(mwTools.bio)\n",
    "from mwTools.bio import sort_codon_index\n",
    "from mwTools.bio import convert_df_to_fasta\n",
    "from statannot import add_stat_annotation\n",
    "from mwTools.id import *\n",
    "from mwTools.paths import p\n",
    "from cterminal import *\n",
    "\n",
    "p = p()\n",
    "print(p.rootPath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "rootPath = p.rootPath\n",
    "cterminalPath = p.cterminalPath\n",
    "refSeqPath = p.refSeqPath\n",
    "mycoplasmaDataPath = p.mycoplasmaDataPath\n",
    "mpnAnnotationPath = p.mpnAnnotationPath\n",
    "analysisCtermDataPath = p.analysisCtermDataPath\n",
    "analysisCtermPlotsPath = p.analysisCtermPlotsPath\n",
    "analysisCtermPath = p.analysisCtermPath\n",
    "ctermELISAPath = p.ctermELISAPath\n",
    "ctermELMseqPath = p.ctermELMseqPath\n",
    "ctermAmiGOPath = analysisCtermPath / 'AmiGO'\n",
    "analysisAbundancePath = p.analysisAbundancePath\n",
    "taxonomyPath = p.taxonomyPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "importExistingDataframe = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Colors and styles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib_options import set_fontsize_paper, set_fontsize_screen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "colorAxis = '0.8'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_fontsize_paper():\n",
    "    sns.set_style('whitegrid')\n",
    "    matplotlib.rcParams['font.size'] = 7\n",
    "    matplotlib.rcParams['font.monospace'] = ['Courier New']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0, 'ABCDEFGHIJKLMNOPK')"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbUAAAD3CAYAAACaTcoOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3XlYU1f+BvA3AWQpijvgru1EUR4FrXXXYsUdW6kbohYf61K1rYpYO6P9WcfHupUytlato1J3XKtQqgLaFq3LKNY1LqCiCLgAWlAgkdzfHzZBIGSBBPT4fp5nninJuTdfDse8uTf3nCuTJEkCERGRAOSVXQAREZGlMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhGE01B4/fozg4GAkJSUBAPbs2YPNmzdj586dVi+OiIjIHLbGGty/fx+enp66nxMTExESEoLQ0FCD250+fRpyOQ8EAUCj0bAv/sa+KMS+KMS+KCSTyeDl5VXZZby0jIZakyZN4OLiUriB7bNNNBqNwe3kcjm8vb3LWZ4YlEolPDw8KruMFwL7ohD7ohD7opBSqazsEl5qRkNNKyMjA7m5uWjQoAE2bdqEpk2bWrMuIiIis5kUav7+/rr/fv5UJBER0YuEJ7GJiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhKGraEn4+LikJ6eDrVajaCgIOzYsQMqlQq5ubn48MMPK6pGIiIikxg8UktISEBgYCAyMzMBAI8fP0Zqairkch7gERHRi8fgkZqt7bOnNRoNAOCvv/5CSEgIFi5caHTHGo0GSqXSAiW+/PLy8tgXf2NfFGJfFGJfkKUYDDVPT0+Eh4fDwcEBFy5cgJ2dHdauXQt3d3ejO5bL5fDw8LBYoS8zpVLJvvgb+6IQ+6IQ+6IQw718DIaar69vkZ89PT2tWgwREVF58MsxIiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJh2Bp6Mi4uDunp6VCr1QgKCkJ0dDTu3bsHOzs7BAYGVlSNREREJjEYagkJCQgJCUFoaCgAID4+Ho0bN0bDhg2N7lij0UCpVFqmypdcXl4e++Jv7ItC7ItC7AuyFIOhZmv77GmNRgMAUKlUmDRpEhYsWIB33nnH4I7lcjk8PDwsVObLTalUsi/+xr4oxL4oxL4oxHAvH4PfqXl6eiI8PBwODg64cOECOnbsiPXr16NBgwYVVR8REZHJDB6p+fr6FvnZ09PTqsUQERGVB69+JCIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGLaGnoyLi0N6ejrUajWCgoIAAEuXLkWHDh3QvXv3iqiPiIjIZAaP1BISEhAYGIjMzEwAQHR0NBo3blwhhREREZnL4JGare2zpzUaDVQqFU6fPg1JknD79m2jR2oajQZKpdJylb7E8vLy2Bd/Y18UYl8UYl+QpRgMNU9PT4SHh8PBwQFXr17F3LlzceLECeTn5xvdsVwuh4eHh8UKfZkplUr2xd/YF4XYF4XYF4UY7uVjMNR8fX1LPNahQwerFUNERFQevPqRiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYDDUiIhIGQ42IiITBUCMiImEw1IiISBgMNSIiEgZDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJhMNSIiEgYtoaejIuLQ3p6OtRqNYKCgrBq1SrY2NjAzs4OQUFBFVQiERGRaQyGWkJCAkJCQhAaGgoA8Pf3h7OzM7799lujO9ZoNFAqlZap8iWXl5fHvvgb+6IQ+6IQ+4IsxWCo2do+e1qj0eh+Xr58OSZNmmR0x3K5HB4eHhYo8eWnVCrZF39jXxRiXxRiXxRiuJePwe/UPD09ER4eDgcHB1y4cAHTpk2Do6MjDh06VFH1ERERmczgkZqvr2+Rnzds2GDVYoiIiMqDVz8SEZEwGGpERCQMhhoREQmDoUZERMJgqBERkTAYakREJAyGGhERCYOhRkREwmCoERGRMBhqREQkDIYaEREJg6FGRETCYKgREZEwGGpERCQMhhoREQnD4P3UyLDMzEwkJCTg9u3bUKvVqFmzJpo1a4Y2bdrAxsamsssTXm5uLv78808kJSUhOzsbzs7OqF27Nry8vODu7l7Z5RFZVUZGBk6dOoW0tDSo1WrUq1cPb775JlxdXSu7tEpVoaE2btw4HDlyxGCbDz/8ECEhISbtLykpCUOHDsXjx49Nau/g4IA6deqgffv2CAgIQOvWrU3arrhr165h+fLlOHToEJ4+fVri+erVq2PYsGEYP348qlWrhlWrVkGlUmHlypV69/fJJ5/gwIEDRl/3tddeg7u7O9q1awd/f394eXkZbG9u/xhy8eJF2NrqHy4fffSRxe6Gvn79enTu3Nlgm9TUVKxcuRJRUVF48uSJ3jbNmzfHmDFjMHjwYCxduhRJSUlYs2ZNiXZ5eXno27cv0tLSSjzXokUL7N27t8hjixYtQnh4OCRJKvK4o6Mj1q1bh7Zt2+qtJyYmBtOnT4darTb4uxVnY2OD4OBgjBs3zqzttLX++OOP0Gg0Jm9jZ2eHJUuWoH///khPT4efnx/++usvvW1lMhl8fHxKjOvSxrObmxsOHjwIe3t73WPm9Iu+7UuTm5uL3r174969e0Ue37dvH5o3b250ewBQqVTo0aMHMjMzdY85ODhg1apV6NChA/r164ebN28W2Wbo0KFYsGBBqftcuHAhNmzYUGL8AM/6/ptvvkGDBg0M1nX9+nUsWbIE8fHxJd5/ZDIZunXrhuDgYLRo0cKE39K8v4G9vT1q164Nb29vDBo0CD169Ci1bUWMH30qNNS6desGBweHIn9QjUaDw4cPw93dHQUFBYiKikJwcDDkcuNnRmvVqoVevXohJycHABAXFwc3Nze0atVKb3uVSoXbt29j9+7d2L17N0aPHo3Zs2eX+matz/r16/H1119DrVbDy8sL/fv3R5MmTVCjRg1kZWXh/PnzOHDgAH744Qfs2LEDYWFhSE5OLvUPCwCdOnVCQUEBUlNTcenSJXh7e6NmzZol2uXn5yMlJQURERGIiIhA7969MXfuXNStW9ci/aPPpUuX9L7hP69bt26Qy+WQJEnXvmfPnpDJZCXaZmdno2rVqiUeT0tLw6VLl4zWs2fPHsyfPx9PnjxBixYt0LdvX7Rs2RI1a9aESqXCnTt3kJCQgIMHD+Jf//oXduzYgcePHyM7O1vv/mxtbdGrVy+kpaUhPj4ejo6OaNeuHWQyGTw8PEq09/b2RkpKCjQaDVJTU6FUKtG2bVvUq1cP9evXL7Xuxo0b45133inyxqHti9OnTyMvLw9dunTRW1/Lli2N9os+Xl5eulqLy8rKQkJCAjw8PFCvXj3d43Z2dmjatCkAwMXFBb169cKjR48AlBw/MpkMnTp1KrHvjh07oqCgAI8fP8axY8cAAG+99RYUCgWqVKlSpK22XzIzM3Xj4vr167hx4wY6d+4MR0dHXVt3d/cS25fGwcEBvXr1Qnp6Og4dOgQ3NzfcvXsXW7ZswZdffmnSPg4ePIjMzEzUr18fd+7cgbe3N+rXr48GDRpALpejV69euHHjBgDg9OnTePjwIXbt2oVRo0aVGijt2rVDamqq7m+SlJSEmzdvokuXLnB2dkbjxo1RUFBQak2bNm3C4sWL8fTpU3Tt2hV9+vRBw4YNIZPJcOfOHcTGxuLXX3/F0aNHMW3aNEyYMMHo7/n82CxtXGip1WqkpaXh559/RlRUFLp27Yqvv/4a1atXL9G2IsaPXpKVJCQkmNQuPj5eUigU0tatW6Xw8HBJoVBIx44dK9NrKhQKKTg42Gi7K1euSCNHjpQUCoX0xRdfmLz/b775RlIoFFL//v2N/n4HDhyQunTpIrVs2VLy8vKSfHx8jO5/165dkkKhkI4ePWq0/mnTpkkKhULy8fGRbt26ZVL9pvbP8z777DNJoVBIarXaIu0vXbqk93FTfvf169dLCoVC6tChg3TgwAGDdeTn50vr16+XWrduLSkUCqlbt25Ga+/WrZs0atQoo+20tm/fLikUCun48eMmb/M8bV+MGDHCpPFhScePH5cUCoW0fft2k7cxZ/w8fPhQGjp0qKRQKKTQ0FCj7Z8fF8uXL5cUCoV08+ZNk2srjVqtlhQKhfTZZ59JY8eOlby8vKTs7GyTth05cqTUoUMHaevWrUbH5ogRIyQvLy+pVatW0gcffGByfaGhoZJCoZBu376te6y0fyOrVq2SFAqF1Lt3b+ns2bOl7jMxMVEaOHCgpFAopEWLFplciySZPi5SUlKkGTNmSAqFQhowYIBJfWrN8fO8Sr9QJCoqCra2tujduzcGDBgAGxsb7Nu3z6qvqVAoEB4ejq5du2Lbtm34448/jG4TExODlStXonnz5tiyZQu8vb0Ntu/duzd27dqFBg0alHqKrKwUCgW++eYbLFq0CGlpaZg6darZp7VeNvHx8Vi8eDFq1qyJzZs3o3fv3gbbV6lSBUFBQVi+fHkFVUha9+/fx6hRo3D27FkEBwdj+vTplV0SAGDkyJF48uQJ9uzZY7TttWvXcOrUKfj7+5t8dFijRg0EBgbi2LFjFjsdrxUdHY3Q0FAoFAps377d4Fcnr7/+OiIiItCqVSusW7cOu3fvtmgtAFC/fn18/fXX+Pjjj3Ht2jX83//9n8X2Xd7xU6mhlp+fj5iYGHTq1Ak1a9ZE7dq10alTJxw8eBD5+flWfW07Ozt89dVXsLe3x8aNGw22zc3Nxb///W/Y29sjNDQULi4uJr2Gq6srlixZYoly9Ro8eDA+/fRTXL58Gd9//73VXqey5efn6/7RhIaG4vXXXzd52x49euCNN97QeyqULC8lJQUjR47UvdGZcvqrovj4+KBevXrYsmWL0bZbt26FTCbDiBEjzHqNKVOmoHr16liyZIne79vL4u7du5g3bx5q1KiBtWvXmvT+4+TkhFWrVqFmzZpYsGABMjIyLFJLcVOnTsXbb7+NqKgoKJXKcu/PEuOnUkPt8OHDyMnJwcCBA3WPDRw4ENnZ2Rb/pKNP3bp10b59eyQkJBhsFxERgbt372LSpEl44403zHqNNm3aoFatWuUp06Dx48ejdevWWLNmDe7evWu116lMu3btwp07dzBixAi95+CN6dixIzw9Pa1QGT0vMTERAQEBuHPnDhYtWoSRI0dWdklF2NjYYPjw4bh+/bruuxp9njx5gr1796JLly5o1KiRWa9RrVo1TJ06FTdu3DApPE2xcuVKPHr0CP/85z9L/f5cn7p162LGjBl4/PgxVq1aZZFa9JkyZQoAYP/+/eXaj6XGT6WGWlRUFOzt7dGrVy/dY76+vnBwcLD6KUit1157TXchRWm2b9+OatWqYcyYMWV6DW9vb5OvRDKXjY0NZsyYAbVajfDwcIvv38HBAVWqVLH6FAUHB4ci//+8iIgIODk54dNPPy3TvufOnYsVK1aUqz4y7Pz58wgMDERWVhbCwsLw3nvvVXZJeg0dOhR2dnbYvHlzqW0iIyORk5Nj9lGaVkBAAJo2bYoVK1boLpIoq0ePHuGnn35C8+bNMWjQILO3HzJkCJo1a4bdu3db/GsQrdatW8PJyQnXrl0r8z4sOX4qLdSys7Px22+/oUePHnB2dtY97uzsjLfffhvx8fHIysqyag2SJEGpVMLNza3UNsnJyUhKSsKAAQOK1GmOqVOnWvX0YKdOndCqVSv89NNPFt/3uHHjsHz5cqufvvPx8cFXX32FNm3aFHk8NTUVly9fRr9+/fReYUWV78SJE/jggw+Ql5eHlStXGv2+szLVqlULffr0waFDh5Cenq63zdatW+Hq6gofH58yvYatrS1mzZqFhw8flvvD1OHDh5Gbm1vmoxaZTIbhw4cjJycHv/76a7lqMcTe3h65ubll2tbS46fSQm3//v1QqVTo379/ief8/PygVqvxyy+/WLWGmJgY3Lp1y+DgPXfuHACU6bRXRerZsycyMzORlJRUpu3379+Pbt26lfg017BhwzL/49bnk08+0XtE6ejoCH9//xJHhOfPnwcAvZe7U+U7fPgwxo8fD5lMhrVr16Jbt26VXZJRgYGBKCgowLZt20o89+eff0KpVGLo0KFmTfUprmfPnujUqRO2bNlSYi6bOU6dOgUA5fo36OvrCwAGT7mWx8OHD5GVlWXWqVEta4yfSgu1qKgoODk56f1jde/eHS4uLlY9BXnu3DnMmTMHLi4uGDt2bKnttHO0XvQVKpo1awYAuHPnTpm2T0xMxL179/DgwQNLllVEQUEBbt26hStXrpi8TWpqKgCgSZMmVqqKyioyMhJTp06Fo6MjwsPD8eabb1Z2SSZp27YtPDw8sGPHDqhUqiLPbd26FTY2Nhg2bFi5X2f27NkoKCjA0qVLy7yPCxcuoH79+uVaJaR+/fqoW7cuLl++XOZ9GKI9Q2TuB39rjZ9KCbV79+7h5MmTeOedd/R+h1KlShX06dMHZ86cwe3bty3ymgUFBbh79y5+//13zJ49GwEBAVCr1fjPf/5jcNKsdtBXq1bNInVYi/aKqLKew5f+nhAv6VnpwFLK8hra1VBee+01q9REZbNlyxaEhITg6dOncHNzKxEOL7qRI0fiwYMHOHjwoO6xhw8f4pdffoGPj49Flppq0aIF/P39ERsbixMnTpRpHxkZGQbfn0zVpEkTpKSklHs/xSUnJ2PFihWoXbu27ojQFNYcP5Wy9mN0dDQ0Gg0GDBhQahs/Pz9s374d+/bt011dY4rIyEhERkYabWdnZ4fVq1fjrbfeMthO+wb8ol8Srq3P2Hw1U/unvMxZtcSQiu7/kydPmryM0qtq9erVCA0NRcOGDdGiRQscPnwYI0eOhL+/P+bNm2fSMlaVzc/PD0uXLsXmzZt1V1/v3r0b+fn5CAgIsNjrTJs2DdHR0Vi0aBF27dpl0kpJz3v06JFFvkt2cXExekGcuc6cOYPg4GBkZ2cjLCzM5A+e1h4/lRJq+/btg4uLC7p27Vpqm/bt28Pd3R2RkZFmhZqxZbKysrKQnJyM7OxsjBs3DgEBAfj0009fmSOB0vpHuzSRpehbJkuSpAqZqlEe1atXR7t27Uxqq10m61Vy+PBhREZGQqFQYN26dahTpw5u3LiB+fPnY/fu3UhMTMT69evLfFFVRXF0dMR7772HDRs24PLly2jevDkiIiLQqFEji35/W6dOHUyYMAFhYWHYs2cP3n//fbP3YYkPdDKZrNxnYSRJwr1793DlyhXs3r0bMTExkMvlmD9/Pvr27WvSPipi/FR4qN28eRMXL17UXVpbGplMhv79+2Pt2rU4d+6cyYsPt2/fHsuWLTPYRpIknD9/HqtXr8aPP/6Io0ePIjw8HHXq1NFbBwCLTaS0Fu16cca+3C6tf7799lt89913Fqvn22+/LVHL06dPzT6C0/a/sX+QiYmJ+O677wyexrC1tUVAQIDBc/8KhcLkK1V37NiBOXPmmNTWmn777TfEx8dXSC05OTnw9vbGDz/8oDsl37RpU6xduxaLFy9GeHg4Zs6ciZUrV77wZzdGjhyJjRs3YvPmzejbty9u3ryJkJAQi9c9duxYREREICwsDP369YOTk5PJ27q4uODhw4flruHhw4dmBcWcOXOMjqf27dtj7ty5Zp3ZqIjxU+Ghpj31dfHiRUyePNlgW+1FC/v27Svzivr6yGQytG7dGitWrMDOnTsxZ84cfPzxx9iyZUuJ0wPagWDt6QXldf/+fQAwebWT4rQDyJpvRGV5De3AN7QgNADcuHEDsbGxBk+/2tjYoG3bti/8lazmOnDgAHbv3o1//vOfZp/eMlerVq2wfv36IgsNA4BcLsfnn3+OnJwc7Ny5Ezt37sTQoUOtWsvz1q1bh9q1a5s1l6tp06bo1KkTIiMjcfv2bVSpUgX+/v4Wr83BwQEzZsxASEgI1qxZY9Z8y1q1apU69cAct27dQsOGDU1ur29BY7lcDmdnZzRq1Ajdu3cv04IGFTF+KjzUoqKiADxb/d2UVdmBZ9/BmbuavqmGDBmCy5cvY+PGjYiNjS0xR0L7h71x48YLfXVXcnIyAOhWWDdX8+bNUadOHdSuXduSZRVhY2ODJk2amPXJTvt9gja0S+Pr64sLFy7ofW706NFITk7G77//bnqxLxGNRgNJkqDRaKweas2aNSvxhvS8uXPn4vjx4wgNDUW/fv0q7DTkxo0b0ahRI7MnKAcGBuKPP/7AsWPH4Ofnp/fuGJbg5+eHjRs3Yv369Rg+fLjBubHP8/T0xI4dO5CZmVnm2tLS0pCenm7W5fKBgYFW+VBSEeOnQq9+PH/+PG7evImJEyfiypUrJv3vs88+Q0ZGBo4ePWq1usaPHw/g2fne4ry9vSGTyfDbb79Z7fUtITY2Fq6urmYv66Pl6+uLI0eO6D01cvbsWb19UxZhYWEICgoyub12nUdjS5nRi8HBwQHBwcHIzMzEhg0bKux1CwoKzLpvnJaPjw/eeusttGjRAqNGjbJCZc/IZDLMnj0bubm5CA0NNXk77Qfp+Pj4Mr92TEwMABi9T+GLwBLjp0JDTXuUZuiqx+IGDhwIuVxu1Tlrrq6uqFatWokbCgLPvujt3r07fv311zLPAbO2c+fOITEx0WorOYSFhRm88aE1tWzZErVq1cLhw4etOt3gVaM9tW+NqSr9+vXD66+/js2bN7/wl/rb2Nhg48aN2Lt3r9Gb7pZXu3bt0LdvX+zbt0+3qIAxPj4+cHR0NLislyGSJGH79u1wcnIyeEPPF0l5x0+FhZpGo0F0dDQUCoVZp5/q1q2Ljh07Ii4uziJ3cNYnPz8fOTk5pS48PGXKFBQUFJj1Cau4vXv36g3N8pIkCQsWLICNjY3VFpB9+vRppd3aRiaT4d1330VSUpLuEyeVn3YCfFmP7A2RyWQICgoqMQ+MgJkzZ8LOzg6LFi0yqb2Liwvee+89nD17FnFxcWa/3t69e3Ht2jUMHjz4pbnCu7zjp8JC7cSJE7h3716RFflNNWjQIOTm5lrtTe3nn3+GRqMp9WKUNm3aYMyYMYiKikJERITZ+9+2bRtmzZpllbXXvv/+e5w9exbDhw/XrSoimokTJ8LFxQULFiyw6oonrwqVSoWdO3fC1dXVagttDxo0CNWqVcP27dutsv+XVcOGDTF69GicOnXK5FXtP/roI7i4uGDu3Llmjf+MjAwsW7YMTk5OmDRpUllLrhTlGT8VFmqRkZGQyWRmnXrU8vX1haOjo1UmDaekpGDJkiVwdnaGn59fqe2Cg4PRpk0bzJ8/36SbDGqtWbMG27ZtQ6tWrfDuu+9aomSdlStXYvny5WjRogVmzZpl0X2/SKpXr46FCxfi3r17mDhxolWOeF9m2lV5QkJCTLoP4ddff42MjAxMmDDBale7Ojg44N1338XJkydx69Ytq7zGy2ry5MmoWbMmli1bZtIZEFdXV8ybNw8ZGRkYO3YsMjMzjW6Tl5eHjz76CPfv38fnn39epnUZK1N5xk+FhJpKpUJMTAy8vLzQoEEDs7d3dnZGz549cezYMaNXwZmqoKAABw4cwLBhw5CVlYWvvvrK4OXwVapUwYoVK+Du7o7Zs2fjyy+/NLgkVWZmJmbOnIlly5ahWbNmWLt2rcVWWkhMTMTkyZMRFhamm8Ro6IoiEfTq1QszZ87ExYsX4e/vj8jISN3cvNIUFBRg//79Fp1U/iKaOXMmunbtiujoaAwfPly3CLc+q1evRnh4ODp27FjmW6uY6v3334ckSdi1a5dVX+dl4+zsjI8//hi3b982+QNy//79MX36dFy9ehXDhg3D//73v1LbXr9+HQEBATh79ixGjx5tkXUsK0NZx49MMvDte1xcHNLT06FWqxEUFIQ9e/bgyZMnsLe3x5AhQwzu+MyZM1Cr1di0aROysrJw8uRJNG7cWHeTTblcjv79++tdpR94tlzRpk2bdJOeU1JScOXKFbRu3Rqurq7o378/OnfujIULF+qWf4mLizO4oohWdnY2rl+/jgcPHqBq1apYtGhRkXu6GXL37l1MmTIF58+fR7Vq1eDr64suXbrA1dUVGo0GqampOHr0KA4cOID8/Hz06dMHY8aMMTgdYOvWrThy5AhSU1Nx6dIleHt76718V61W4/bt27o36YEDB2IXN466AAAGVUlEQVT+/Pmlnit/+PBhmfqnuNOnT8Pe3r7US+K3bNmCo0ePQpIkXLp0CWlpaXpXFJHL5WjTpo3uatOyiI6OxhdffIHs7GzUr18fPXr0gLe3N+rUqQNnZ2fk5+cjJSUFZ86cQVxcHO7fvw8nJyeMHTsWn3zySYn9aftekiQcOXIEjo6OaNeuHWQyGTw8PDB16tQi7Q8cOICoqCgUFBToVhRp27YtatSoAXd3d8yZM0fv0c+1a9fw3XffFflknp2djapVq+L06dPIy8vTu5KFKRPGgWcBHhYWhv/+978Ant3VoG/fvmjSpAlUKhWuXr2KnTt34tq1a2jbti1WrVpldE5jbm4u5s+fr/vwVnz8yOVydOzY0eBVg4MGDcLdu3dL7VNtv2RmZqJq1aoACle36dy5s8kf1o4cOYI2bdro7mKv3a9KpcKhQ4fg7u6Oli1b6ur28/NDnz599O7r6tWrWLFihe5vlZaWpvt3Wbt2bd22z/dP8b9haeNHq6CgAIMGDUJiYqKub7Uf+JVKJTw8PPRut2nTJixevBhqtRqdO3dGv3790LhxY8hkMty5cweHDh3CoUOHUFBQgKlTp2Ly5MlGj8af/32zsrKQkJCgd54a8Gxh988//9yk6VUVMX70MVhZQkICQkJCdBdIJCYmFvnZmMuXLyM2Nlb3iTo5OVk3n0omk8HNza3UUCu+rda5c+d02zZv3hyxsbFFLiBJT083OllRO4EwICAAAQEBZt2Z2tXVFREREdi7dy82bdqEXbt2lfgkIZPJ0K5dO0yYMAE9evQwupTSsWPHEBsbq/v5zJkzettVqVIFdevWxeDBgzF06FCjyzllZGSUqX/0MbTA6/Hjx4vUD0DvclgymazcR5TaDzNr1qxBZGQktmzZUuodhhUKBcaMGYMhQ4aUOseneN/n5+frvpBPS0sr8Q/ozJkziImJKXIlpna6gZubG2bNmqX3iPzmzZuIi4szeLpJ34UANjY28PLyMhpqNjY2CA4Ohq+vL7777jscOXKkxGXgTk5OmDJlCiZNmoQqVaoY3B/wbN3B2NjYIhPfi4+fgoICg29Kfn5+WLZsme53S09PL9Knhvrljz/+MFpjaYrvNy0tTXfHDZlMhgYNGpQaaklJSXprOnPmTJFt9fXP83/D4r/r82xsbPDZZ5+Z/QFv1KhR6NSpE5YuXYr4+PgSU51kMhm6dOmCGTNmmPzhNTk5ucTvq1Qq9b5vubm5ISQkxKRQq4jxo4/BI7VvvvkG06dPx7JlyzBz5swSPxvy559/vhQLm5bXgwcPoFQqkZGRAblcjtq1a+Mf//iH3iW3yPKSk5Nx69YtPHr0CLm5uXByckKtWrWgUCisNpH2ZZCVlYVLly7h/v37kMvlcHNzQ5s2bV6Jf5Mvu/z8fJOmFzx48ACnTp1CWloanj59Cnd3d7z55psmT+wWlcFQi4mJwZ07d5CTk4O3334bSqUS+fn5cHR0LNPCnERERNZkMNSIiIheJpV252siIiJLY6gREZEwGGpERCQMhhoREQnDojcoK89kbZEU74dVq1bBxsYGdnZ2Zt12RQTF+wIAli5dig4dOqB79+6VW1wFK94X0dHRuHfvHuzs7BAYGFjZ5VWo4n2xY8cOqFQq5Obm4sMPP6zs8irc48eP8cUXX2Dy5Ml4/fXXX9n3Tkuw6JFaQkICAgMDdWuTJSYmIjAw8JVb+614P/j7+yMwMBB3796t5MoqXvG+iI6ORuPGjSu5qspRvC/i4+ORl5f3Ss4rKt4Xjx8/RmpqqtVvcvqiun//fpE7Sb+q752WYNERpJ1lrr1ZX/GfXxX6+mH58uWYOHFiZZZVKZ7vC5VKhdOnT+Py5csG164TVfFxoVKpMGnSJBw7dqwyy6oUxfvir7/+QkhISJlWuxFBkyZNiixb9qq+d1qCzbx58+ZZamc5OTmIj4+HRqOBk5MTVCoVzpw5gzp16ujWXXsVFO+HuXPnolGjRsjMzCx1TTdRPd8XVatWxdChQ+Hg4ABXV9dX7oit+LhwcHDAyZMn4eLiYvUbVL5oivdFSkoKzp07h2rVqsHb27uyy6sUSqUSdnZ2yMrKemXfOy2Bk6+JiEgYr+YJbCIiEhJDjYiIhMFQIyIiYTDUiIhIGAw1IiISBkONiIiEwVAjIiJh/D+EGJP4blFKEQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set_fontsize_paper()\n",
    "fig, ax = plt.subplots()\n",
    "ax.annotate(\"ABCDEFGHIJKLMNOPK\", xy=(0,0), fontsize=40, fontfamily='monospace')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkMAAABLCAYAAABz9YPfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAAohJREFUeJzt3K1uFFEcxuH/fpRtAIfhFgCzN4ACRAVXgEFgcFgEtVgcCYpUYiBBVBEEwfCRVJBwA6gSBEkrNm1nuIS2hJND9n0efcQ7Ocnkl0wyk3EcxwIACDXtPQAAoCcxBABEE0MAQDQxBABEE0MAQDQxBABEE0MAQDQxBABEE0MAQLT5eQ5/+vyljob1/WH15nSo6XDce0YTw3ReqzrpPaOZxTivOl7Pu6uqGjYu1Li+11eT6VDjsL4POJttVA29VzQyrTpe67ubVNVR7xnNnNRmre+bs+rSfFbL5fLUc+eKoaNhrHuv9/961P9u987vuvbhYe8ZTXy/+bzu/3jae0YzOxe3a/PJdu8Zzew/e1nf3hz2ntHMjbsH9XXnRe8Zzdx+sF1X3q16z2ji161Fvf2423tGM1tb1+vg8HHvGc38vPyqHh3Oes9o5v3VxZnO+UwGAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBANDEEAEQTQwBAtMk4juNZD+/t7dVisWi5BwDgn1itVrVcLk89d64YAgBYNz6TAQDRxBAAEE0MAQDRxBAAEE0MAQDRxBAAEE0MAQDRxBAAEE0MAQDR/gDVflK/Qav2jgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.palplot(sns.color_palette())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASwAAABLCAYAAADK+7ojAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAAcpJREFUeJzt1zFOG1EYhdHnscFhBVkAJc3QpaSgQmJNhDo7QKKhpqBAYg9EllzQeheksAzzsgRspNHjmnPqv7jTfE8zqbXWAhCgaz0AYFuCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBizXY4Xf59LN2zG2tLcZnpQ3upb6xmjmE1mZfq+v+/T0JVS31uvGM+kG0od9vcDD38clb7vP7zbKVjdsCmnTxefHvXV3Z/fluvVdesZo7g6viq/Hn+2njGa1dlheXn413rGaE4uX8vi7qb1jNFc/P6z1d3+PrnA3hEsIIZgATEEC4ghWEAMwQJiCBYQQ7CAGIIFxBAsIIZgATEEC4ghWEAMwQJiCBYQQ7CAGIIFxBAsIIZgATEEC4ghWEAMwQJiCBYQQ7CAGIIFxBAsIIZgATEEC4ghWEAMwQJiCBYQQ7CAGIIFxBAsIIZgATEEC4ghWEAMwQJiCBYQQ7CAGIIFxBAsIIZgATEEC4ghWEAMwQJiCBYQQ7CAGIIFxBAsIIZgATEEC4gxqbXWbY+Xy2WZz+dj7gG+ofV6Xfq+//Bup2ABtOSXEIghWEAMwQJiCBYQQ7CAGIIFxBAsIIZgATEEC4jxH3WsNK8faLNPAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 360x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "paletteCat = [sns.color_palette()[i] for i in [1,2,6,4,5]]\n",
    "sns.palplot(paletteCat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**I changed the colours in order not to repeat the colors from both bias color code and abundance color code.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZwAAABLCAYAAABEDTEaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAAgxJREFUeJzt2DFuE1EUheFrT8BITkuB2ACiMkV6KJOeljXQsoK0LIOOdHQ07kBILmjZRSzhBM9jC3akd4aE76tvcar5n2bWWmsFAJ3Npx4AwP9BcACIEBwAIgQHgAjBASBCcACIEBwAIgQHgAjBASDi5JjjH9+/1Xy87bVlcjfDk/oz9YhOTqpqqN9Tz+joUc32D/f9NM6r2n7qFf3MhlbjOE49o5thmFXVw/x2tnZTZ2dvDro9Kjjz8bZefbm406j74NPF13q/Haae0cXH5b6eXr+bekY3p8vLev752dQzuvn1+nH9vNpOPaObF29brdfrqWd0c37+sq63H6ae0cXp8vLg24f7JATgnyI4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAhOAAECE4AEQIDgARggNAxKy11g493mw2tVgseu4B4B7Z7Xa1Wq0Ouj0qOABwV36pARAhOABECA4AEYIDQITgABAhOABECA4AEYIDQITgABDxF2xWPG64RsuvAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 504x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "paletteCat = [sns.color_palette()[i] for i in [1,9,8,6,4,7,8]]\n",
    "sns.palplot(paletteCat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**I changed the colours in order not to repeat the colors from both bias color code and abundance color code.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAALwAAABLCAYAAADK3u0KAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAAWJJREFUeJzt2TFOW0EYRtFxQLig9TJSuPI6IrHKdFlHRCQXrIISlOLJQsMSwEgvJtxz6r/4iluMNJs55xwQ8e3SA+BfEjwpgidF8KQInhTBkyJ4UgRPiuBJuT7n+Pf9n3Faa8kncDPnmC+XXrGOzdUYY3zdT/U55zgcDm/enRX8aYxx93z10U2f3s9lGQ+//l56xiq+/7gdT6fHS89YzW63e9edJw0pgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInpTNnHO+9/h4PI7tdrvmHviQZVnGfr9/8+6s4OF/50lDiuBJETwpgidF8KQInhTBkyJ4UgRPyitQcyl+7oL9KAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 216x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "groupList = ['membrane proteins', 'non membrane proteins', 'all']\n",
    "\n",
    "cMembraneTrue = paletteCat[1]\n",
    "cMembraneFalse = paletteCat[4]\n",
    "cAll = colors.to_rgb('0.7')\n",
    "paletteMembraneCat = {'membrane':cMembraneTrue, 'cytoplasmic':cMembraneFalse, 'all':cAll}\n",
    "sns.palplot(paletteMembraneCat.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWQAAABLCAYAAABHlv13AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAAe1JREFUeJzt2KFuFFEYhuF/tpQlgeBrCNzBmF4CniC4D7gmFAKBIyRcADVrEdQ0iEpwm6VzEFxAuyQn5+vkefQvvkkm72x2aq21AmC4zegBAPwjyAAhBBkghCADhBBkgBCCDBBCkAFCCDJACEEGCPHgmONvFxd1WJZeW4Z7eHJSa326TVVtNiv+/rZW0zR6RD9tWapW+3ZWtVbVbv6MntHN6aPHNc/zrXdHBfmwLPX604f/HpXu/as39eXwe/SMLl6ePq2z589Gz+hm+nlVL86Oep3vle8/ruvJr8+jZ3RzfXNelx/fjp7Rzfm7r3e6W/FPJoD7RZABQggyQAhBBgghyAAhBBkghCADhBBkgBCCDBBCkAFCCDJACEEGCCHIACEEGSCEIAOEEGSAEIIMEEKQAUIIMkAIQQYIIcgAIQQZIIQgA4QQZIAQggwQQpABQggyQAhBBgghyAAhBBkghCADhBBkgBCCDBBCkAFCCDJACEEGCCHIACEEGSCEIAOEEGSAEIIMEEKQAUIIMkAIQQYIIcgAIQQZIIQgA4QQZIAQggwQQpABQggyQAhBBggxtdbaXY93u11tt9ueewBWZ7/f1zzPt94dFWQA+vGXBUAIQQYIIcgAIQQZIIQgA4QQZIAQggwQQpABQggyQIi/hQE5q64vKusAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAALwAAABLCAYAAADK3u0KAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAAWtJREFUeJzt2bFtFFEYRtE3GGkb2AqoYDMSImIiy325AwpwhCgDJ5vTwjYwsjVDCfZaGg9wz4n/4Atu8KQ3reu6Doj4sPcAeE+CJ0XwpAieFMGTInhSBE+K4EkRPCkfrzn+9fg4npZlqy27O9zcjP/143maprE+z3vP2MzyPI/PX76+eHdV8E/LMm5/Prx51N/ux7e7cblc9p6xiePxOH5/v9t7xmY+3d6/6s6ThhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPiuBJETwpgidF8KQInhTBkyJ4UgRPyrSu6/ra4/P5PA6Hw5Z74E3meR6n0+nFu6uCh3+dJw0pgidF8KQInhTBkyJ4UgRPiuBJETwpfwD3+y1R/upgIAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 216x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cmapName = \"BrBG_r\"\n",
    "paletteAbundance = sns.color_palette(cmapName)\n",
    "sns.palplot(sns.color_palette(cmapName))\n",
    "cLow = paletteAbundance[0]\n",
    "cMedium = colors.to_rgb('0.7')\n",
    "cHigh = paletteAbundance[-1]\n",
    "sns.palplot([cLow, cMedium, cHigh])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prolog"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import and analyze aa and codon frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mollicutes\n",
    "oddsratioDf = pd.read_csv(os.path.join(analysisDataFolder,'setMollicutes-subset-all_oddsRatioCodonDf.csv'))\n",
    "countSeries = oddsratioDf[(oddsratioDf['terminus']=='bulk') & (oddsratioDf['observable']=='count')].iloc[0,3:]\n",
    "bulkAAfreq = countSeries / countSeries.sum()\n",
    "bulkAAfreq.to_csv(os.path.join(root_folder,'setMollicutes_codon_freq_bulk.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "countSeries = oddsratioDf[(oddsratioDf['terminus']=='C') \n",
    "            & (oddsratioDf['position from terminus']==-1) \n",
    "            & (oddsratioDf['observable']=='count')].iloc[0,3:]\n",
    "codonFreq = countSeries / countSeries.sum()\n",
    "codonFreq.to_csv(os.path.join(root_folder,'setMollicutes_codon_freq_cterminal.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "countSeries = oddsratioDf[(oddsratioDf['terminus']=='N') \n",
    "            & (oddsratioDf['position from terminus']==+1) \n",
    "            & (oddsratioDf['observable']=='count')].iloc[0,3:]\n",
    "codonFreq = countSeries / countSeries.sum()\n",
    "codonFreq.to_csv(os.path.join(root_folder,'setMollicutes_codon_freq_nterminal.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsratioDf = pd.read_csv(os.path.join(analysisDataFolder,'setPhylum-index00-PhylumId-00000976-Bacteroidetes-subset-all_oddsRatioAADf.csv'))\n",
    "countSeries = oddsratioDf[(oddsratioDf['terminus']=='bulk') & (oddsratioDf['observable']=='count')].iloc[0,3:]\n",
    "bulkAAfreq = countSeries / countSeries.sum()\n",
    "bulkAAfreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsratioCterm = oddsratioDf[(oddsratioDf['position from terminus']==-1) & (oddsratioDf['observable']=='oddsRatio')].iloc[0,3:]\n",
    "oddsratioCterm.name = 'cterm'\n",
    "oddsratioCterm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mycoplasma pneumoniae"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_accession = 'GCF_000027345.1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GCF_000027345.1_Mycoplasma_pneumoniae_M129_countStatistics.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with (analysisCtermDataPath / 'GCF_000027345.1_Mycoplasma_pneumoniae_M129_countStatistics.json').open() as f:\n",
    "    oddsratioDf = (str())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "countSeries = oddsratioDf[(oddsratioDf['terminus']=='bulk') & (oddsratioDf['observable']=='count')].iloc[0,3:]\n",
    "bulkAAfreq = countSeries / countSeries.sum()\n",
    "bulkAAfreq.to_csv(os.path.join(root_folder,'setMollicutes_codon_freq_bulk.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pow(20, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pow(20, 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonDf = pd.read_csv(str(analysisCtermDataPath / 'GCF_000027345.1' / 'GCF_000027345.1_subset_all_oddsRatioCodonDf.csv'))\n",
    "oddsRatioCodonDf = oddsRatioCodonDf[(oddsRatioCodonDf['terminus'] == 'bulk') &\n",
    "                                    (oddsRatioCodonDf['observable'] == 'count')\n",
    "                                   ].drop(['position from terminus', 'observable', 'terminus'], axis=1).T\n",
    "oddsRatioCodonDf.columns = ['count']\n",
    "oddsRatioCodonDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "refCodonTableDf = build_refCodonTable(codonTableBio)\n",
    "refCodonTableDf['aa'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonDf['freq'] = oddsRatioCodonDf['count']/(oddsRatioCodonDf['count'].sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonDf['aa'] = refCodonTableDf['aa']\n",
    "oddsRatioCodonDf.sort_values(['aa', 'freq'], ascending=[True, False], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonDf.to_csv(str(analysisCtermDataPath / '..' / 'mpn_bulk_codon_freq.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AA pairs frequency"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default, pandas read_csv will consider a list of strings that evaluate to NaN:\n",
    "\n",
    "na_values : str or list-like or dict, default None\n",
    "Additional strings to recognize as NA/NaN. If dict passed, specific per-column NA values. By default the following values are interpreted as NaN: ‘’, ‘#N/A’, ‘#N/A N/A’, ‘#NA’, ‘-1.#IND’, ‘-1.#QNAN’, ‘-NaN’, ‘-nan’, ‘1.#IND’, ‘1.#QNAN’, ‘N/A’, ‘NA’, ‘NULL’, ‘NaN’, ‘nan’.\n",
    "\n",
    "**In our case, the AA pairs NA will be read as a NaN value!!!**\n",
    "\n",
    "We can use the options `keep_default_na=False` and `na_values=['','NaN','etc']` to specifiy manually the list of strings to be treated as NaN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mollicutes\n",
    "subseqOddsRatioDf = pd.read_csv(os.path.join(analysisDataFolder,'setMollicutes-subset-all_oddsRatioAAPairListDf.csv'),\n",
    "                                keep_default_na=False, na_values=[''])\n",
    "subseqOddsRatioDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf.query(\"subseq == 'NA'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf = subseqOddsRatioDf[(subseqOddsRatioDf['observable']=='count') &\n",
    "                                      (subseqOddsRatioDf['statistical_test']=='contingency')].drop('observable', axis=1)\n",
    "subseqOddsRatioDf = subseqOddsRatioDf.drop('statistical_test', axis=1)\n",
    "subseqOddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf.subseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped = subseqOddsRatioDf.groupby('terminus')\n",
    "grouped.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf['freq'] = subseqOddsRatioDf.groupby(['terminus']).value.transform(lambda x: x/x.sum())\n",
    "subseqOddsRatioDf.rename(columns={'value':'count'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf.groupby('terminus').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf.groupby('terminus').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf.to_csv(str(Path(root_folder) / '..' / 'results' / 'setMollicutes_AApairs_freq.csv.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find aa at position -2 that is the most neutral in terms of biases at position -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# row index is -2 position, column index is -1 position\n",
    "subseqOddsRatioPairsDf = pd.read_csv(os.path.join(analysisDataFolder,'setPhylum-index00-PhylumId-00000976-Bacteroidetes-subset-all_oddsRatioAAPairDf.csv'))\n",
    "subseqOddsRatioPairsDf = subseqOddsRatioPairsDf[(subseqOddsRatioPairsDf['terminus']=='C')\n",
    "                                                & (subseqOddsRatioPairsDf['observable']=='oddsRatio')\n",
    "                                                & (subseqOddsRatioPairsDf['statistical_test']=='contingency')\n",
    "                                               ]\n",
    "subseqOddsRatioPairsDf.drop(['terminus', 'statistical_test', 'observable'], inplace=True, axis=1)\n",
    "subseqOddsRatioPairsDf.set_index('aa0', drop=True, inplace=True)\n",
    "subseqOddsRatioPairsDf.dropna(axis=0, how='all', inplace=True)\n",
    "subseqOddsRatioPairsDf.dropna(axis=1, how='all', inplace=True)\n",
    "subseqOddsRatioPairsDf = subseqOddsRatioPairsDf.append(oddsratioCterm)\n",
    "subseqOddsRatioPairsDf = subseqOddsRatioPairsDf.transpose()\n",
    "subseqOddsRatioPairsDf\n",
    "# row index is -1 position, column index is -2 position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# row index is -1 position, column index is -2 position\n",
    "\n",
    "oddsRatioEuclidianDist = subseqOddsRatioPairsDf.drop('cterm', axis=1).copy()\n",
    "\n",
    "# Compute square difference for each (-2,-1) compared to cterm\n",
    "# We weight the difference by the bulk frequency of aa at C-terminal\n",
    "for index, row in subseqOddsRatioPairsDf.iterrows():\n",
    "    delta2 = row[:-1].copy()\n",
    "    oddsratioCterm = row[-1]\n",
    "    for aa, oddsratio in row[:-1].iteritems():\n",
    "        delta2[aa] = bulkAAfreq[index]*(oddsratio - oddsratioCterm)**2\n",
    "    oddsRatioEuclidianDist[index] = delta2\n",
    "\n",
    "# Sum square differences over -1 positions\n",
    "oddsRatioEuclidianDist.sum(axis=0).sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# row index is -1 position, column index is -2 position\n",
    "\n",
    "oddsRatioEuclidianDist = subseqOddsRatioPairsDf.drop('cterm', axis=1).copy()\n",
    "\n",
    "# Compute square difference for each (-2,-1) compared to cterm\n",
    "# We weight the difference by the bulk frequency of aa at C-terminal\n",
    "for index, row in subseqOddsRatioPairsDf.iterrows():\n",
    "    delta2 = row[:-1].copy()\n",
    "    oddsratioCterm = row[-1]\n",
    "    for aa, oddsratio in row[:-1].iteritems():\n",
    "        delta2[aa] = (oddsratio - oddsratioCterm)**2\n",
    "    oddsRatioEuclidianDist[index] = delta2\n",
    "\n",
    "# Sum square differences over -1 positions\n",
    "oddsRatioEuclidianDist.sum(axis=0).sort_values()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example of enrichment analysis of a.a. at the C-terminal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RefSeq bacterial genomes database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import RefSeq assembly summary file from web server\n",
    "if False:\n",
    "    os.chdir(str(refSeqPath))\n",
    "    with urllib.request.urlopen('ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt') as response:\n",
    "        assembly_summary = response.read().decode('utf-8')\n",
    "        with open('assembly_summary.txt', 'w') as file:\n",
    "            file.write(assembly_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "assemblySummaryRepDf = read_assemblySummary_file(str(refSeqPath / 'assembly_summary_refseq.txt'))\n",
    "assemblySummaryRepDf.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepDf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    # Build the list of ftp paths to download\n",
    "    downloadList = assemblySummaryRepDf['ftp_path'].map(\\\n",
    "                       lambda path: re.sub(r'(ftp://ftp\\.ncbi\\.nlm\\.nih\\.gov.*/)(GCF_.+)', \\\n",
    "                                           r'\\1\\2/\\2_genomic.gbff.gz', path) \\\n",
    "                                                    )\n",
    "    downloadList.to_csv(str(refSeqPath / 'refseq_download_list.txt'), sep='\\n', index=False)\n",
    "    downloadList[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We downloaded all genomes (~ total 12 GB) with the bash script `download_script.sh` located at `/users/lserrano/mweber/Databases/RefSeq/Release_2016_10/Genomes_compressed`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import genome file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomesCompressedPath = refSeqPath / 'Genomes_compressed'\n",
    "genomesExtractedPath = refSeqPath / 'Genomes_extracted'\n",
    "\n",
    "species_search_name = 'Bacillus subtilis subsp. subtilis str. 168'\n",
    "species_search_name = 'Escherichia coli str. K-12 substr. MG1655'\n",
    "species_search_name = 'Pseudomonas putida KT2440'\n",
    "species_search_name = 'Mycoplasma pneumoniae'\n",
    "\n",
    "species_search_name = 'Mycoplasma pneumoniae'\n",
    "\n",
    "compressedGenomeFilename, species_name, genome_accession = import_genome_gbff_file(assemblySummaryRepDf,\n",
    "                                                                                   species_name=species_search_name)\n",
    "# compressedGenomeFilename, species_name, genome_accession = import_genome_gbff_file(assemblySummaryRepDf,\n",
    "#                                                                                    genome_accession='GCF_000016665.1')\n",
    "print(compressedGenomeFilename)\n",
    "print(species_name)\n",
    "print(genome_accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not importExistingDataframe:\n",
    "    genomeFilename = extract_compressed_genome_file(compressedGenomeFilename,\n",
    "                                                    str(genomesCompressedPath), str(genomesExtractedPath))\n",
    "    print(genomeFilename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genomeFilename = re.match(r'(.+)\\.gz', compressedGenomeFilename).group(1)\n",
    "genomeBio = next(Bio.SeqIO.parse(str(genomesExtractedPath / genomeFilename), format=\"genbank\"))\n",
    "print(genomeBio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[feat.location for feat in genomeBio.features if feat.type == 'CDS']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if not importExistingDataframe:\n",
    "    allCDSDf = prepare_all_CDS(genomeBio, genome_accession, os.path.join('Genome_extracted', genomeFilename), verbose=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allCDSDf if not importExistingDataframe else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve the codon table for a specific sequence\n",
    "print(int(allCDSDf.codonTableBio[0]))\n",
    "codonTableBio = Bio.Data.CodonTable.unambiguous_dna_by_id[ int(allCDSDf.codonTableBio[0]) ]\n",
    "print(codonTableBio.start_codons)\n",
    "print(codonTableBio.stop_codons)\n",
    "print(codonTableBio)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate fasta file for protein sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(generate_protein_seq_fasta(genomeBio, verbose=False)[:2000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(str(refSeqPath))\n",
    "# write the fasta file\n",
    "proteinSeqFastaFolder = 'Protein_sequences_fasta'\n",
    "proteinSeqFastaFilename = re.sub(r'(GCF.+)_genomic.gbff', r'\\1_protein.faa', genomeFilename)\n",
    "allProteinSeqFastaFile = os.path.join(proteinSeqFastaFolder,proteinSeqFastaFilename)\n",
    "\n",
    "if not importExistingDataframe:\n",
    "    with open(allProteinSeqFastaFile, 'w') as file:\n",
    "        file.write(generate_protein_seq_fasta(genomeBio, verbose=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import all protein sequences from fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(str(refSeqPath))\n",
    "# Import protein sequences fasta file\n",
    "allProteinSeqBio = list(Bio.SeqIO.parse(allProteinSeqFastaFile, \"fasta\"))\n",
    "allProteinSeqBio[:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allProteinSeqDf = prepare_allSeq(allProteinSeqBio, species_name)\n",
    "allProteinSeqDf[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cluster analysis of protein sequences in MPN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will remove protein sequences that have both a high overall identity and a high identity at the C-terminal. The resulting list of sequences will form the non-redundant database for our analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform a cluster analysis of all protein sequences in MPN using CD-HIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "skipClustering = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "identity_threshold = 0.8\n",
    "print('Identity threshold for sequence clustering: ',identity_threshold)\n",
    "cterm_identity_threshold = 0.85\n",
    "print('Identity threshold for c-terminal sequence clustering: ',cterm_identity_threshold,\\\n",
    "      ' max nb of different residues: ',floor((1-cterm_identity_threshold)*cTerminalSize),\\\n",
    "     ' of ',cTerminalSize)\n",
    "\n",
    "if not skipClustering:\n",
    "    # We write the output of the clustering in a folder like \"Clustering_GCF_000027345.1_ASM2734v1_protein\",\n",
    "    # relative to the path of the protein sequence fasta file.\n",
    "    clusteringOutputFolder = os.path.join(os.path.dirname(allProteinSeqFastaFile),\\\n",
    "                                          \"Clustering_\" + re.sub(r'.faa$', r'', os.path.basename(allProteinSeqFastaFile)) )\n",
    "    clusteringFolder, clusteringOutputFile = clustering(allProteinSeqFastaFile, \\\n",
    "                                                        clusteringOutputFolder, \\\n",
    "                                                        identity_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to parse the list of cluster and sequences, we have to parse the CD-HIT output file. Following solution is adapted from a small wrapper script ([source code on Github](https://github.com/Y-Lammers/CD-HIT-Filter/blob/master/CD-HIT-Filter.py), author: Youri Lammers)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if not skipClustering:\n",
    "    # Parse cluster output file and create dictionary of clusters\n",
    "    cluster_dic = parse_cluster_file(clusteringOutputFile + '.clstr')\n",
    "\n",
    "    # Print cluster with more than 1 sequence\n",
    "    for key, cluster in cluster_dic.items():\n",
    "        if len(cluster) > 1:\n",
    "            print('Nb of seq: ',len(cluster),key,cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mark duplicated proteins\n",
    "# allProteinSeqDf['duplicated'] = False\n",
    "# allProteinSeqDf['duplicated_cluster'] = None\n",
    "# verbose = True\n",
    "# for key, cluster in cluster_dic.items():\n",
    "#     if verbose: print('Nb of seq: ',len(cluster),key,cluster)\n",
    "#     if len(cluster) > 1:\n",
    "#         seqlist = extract_seqIdList_from_cluster(key, cluster_dic, extract_refseq_accession_id,\n",
    "#                                                  allProteinSeqDf, 'refSeqProteinId', match_method=\"contains\")\n",
    "#         allProteinSeqDf.loc[seqlist,'duplicated'] = True\n",
    "#         allProteinSeqDf.loc[seqlist,'duplicated_cluster'] = key\n",
    "\n",
    "# mpnRefSeqDf = pd.read_csv(\"/users/lserrano/mweber/Research_Dropbox/Mycoplasma_pneumoniae_experimental_data/Genome/table_MPN_locusTag_proteinId.txt\", sep='|', header=None)\n",
    "# mpnRefSeqDf.columns = ['mpnId', 'refSeqProteinId']\n",
    "# allProteinSeqDf2 = pd.merge(allProteinSeqDf, mpnRefSeqDf, on='refSeqProteinId', how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add the clusters to the dataframe (use multi index to group the rows by cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not skipClustering:\n",
    "    allProteinSeqDf = add_clusters_to_dataframe(allProteinSeqDf, cluster_dic)\n",
    "allProteinSeqDf[:2] if not skipClustering else None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In each cluster, perform a second clustering analysis on the C-terminal part of sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if not skipClustering:\n",
    "    allProteinSeqDf = cterm_clustering(allProteinSeqDf, clusteringFolder, cterm_identity_threshold, cTerminalSize, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read allCDSDf from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if importExistingDataframe:\n",
    "    outputFilePrefix = genome_accession + '_' + re.sub(r'\\s', '_', species_name)\n",
    "    outputFilePrefix = re.sub(r'[=:/\\\\]', r'_', outputFilePrefix)\n",
    "    \n",
    "    allProteinSeqDf = pd.read_csv(str(analysisCtermDataPath / (outputFilePrefix + \"_allProteinSeqDf.csv\")))\n",
    "    allProteinSeqDf.set_index('refSeqProteinId', inplace=True, drop=True)\n",
    "    allProteinSeqDf.drop('refSeqProteinId.1', axis=1, inplace=True, errors='ignore')\n",
    "\n",
    "\n",
    "    allCDSDf = pd.read_csv(str(analysisCtermDataPath / (outputFilePrefix + \"_allCDSDf.csv\")))\n",
    "    allCDSDf.set_index('refSeqProteinId', inplace=True, drop=True)\n",
    "    allCDSDf.drop('refSeqProteinId.1', axis=1, inplace=True)\n",
    "    print(allCDSDf.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if importExistingDataframe:\n",
    "    n = len(allCDSDf)\n",
    "    allCDSDf.groupby('stopCodon').size() / n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allProteinSeqDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allCDSDf_nr['DNASeq'].head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter the protein sequence and CDS dataframes for the non-redundant sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, our non-redundant database will be"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "allProteinSeqDf_nr = allProteinSeqDf.copy()\n",
    "# Filter for proteins that are less than 50 a.a. long\n",
    "allProteinSeqDf_nr = allProteinSeqDf_nr[\n",
    "    allProteinSeqDf_nr.apply(lambda row: len(row['proteinSeq']) >= cTerminalSize + nTerminalSize + 10, axis=1 )\n",
    "    ]\n",
    "allProteinSeqDf_nr.dropna(subset=['non-redundant'], inplace=True)\n",
    "print(allProteinSeqDf_nr['non-redundant'])\n",
    "allProteinSeqDf_nr = allProteinSeqDf_nr[ allProteinSeqDf_nr['non-redundant'] ]\n",
    "#allProteinSeqDf_nr.reset_index(inplace=True)\n",
    "#allProteinSeqDf_nr.set_index('refSeqProteinId', drop=True, inplace=True)\n",
    "if 'cluster' in allProteinSeqDf_nr.columns:\n",
    "    allProteinSeqDf_nr.drop('cluster', axis=1, inplace=True)\n",
    "\n",
    "print(allProteinSeqDf_nr[:5])\n",
    "\n",
    "# Filter the coding sequences for the non-redundant ones using the index of allProteinSeqDf_nr\n",
    "allCDSDf_nr = allCDSDf.loc[ allProteinSeqDf_nr.index ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allCDSDf_nr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allProteinSeqDf_nr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform a multiple sequence alignment (MSA) of all sequences in the cluster, and look at the C-terminal similarity. Take the largest cluster as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allProteinSeqDf[ allProteinSeqDf['non-redundant'].map(lambda y: y is False) ].sort_index(axis=0, level=1);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment analysis of a.a."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Count the frequency of amino acids in all MPN protein sequences excluding the C-termini."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(root_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "bulkSeqAAFreq, bulkSeqAARelativeFreq = count_bulk_aa_frequency(allCDSDf, cTerminalSize, nTerminalSize)\n",
    "print(bulkSeqAAFreq)\n",
    "print()\n",
    "print(bulkSeqAARelativeFreq)\n",
    "bulkSeqAAFreqDf = convert_bulk_aa_freq_to_Df(bulkSeqAAFreq)\n",
    "print(bulkSeqAAFreqDf.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Computing the frequency of amino acids at the C-terminal and comparing to the reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctermSeqAAFreq, ntermSeqAAFreq = count_termina_aa_frequency(allProteinSeqDf_nr, cTerminalSize, nTerminalSize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "oddsRatioDf = compute_odds_ratio(bulkSeqAAFreq, ctermSeqAAFreq, cTerminalSize, ntermSeqAAFreq, nTerminalSize)    \n",
    "oddsRatioDf.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf.loc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot the results on a color mesh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonSuffix = 'taxaId_1224_phylum_Proteobacteria'\n",
    "filename = taxonSuffix + '_subset_all_oddsRatioAADf.csv'\n",
    "\n",
    "outputFilePrefix = re.sub(r'(.+)_.+\\.csv', r'\\1', filename)\n",
    "plotTitle = re.sub(r'_', r' ', outputFilePrefix)\n",
    "print(\"outputFilePrefix\", outputFilePrefix)\n",
    "print(\"plotTitle\", plotTitle)\n",
    "\n",
    "oddsRatioDf = pd.read_csv(analysisDataPath / taxonSuffix / filename)\n",
    "oddsRatioDf.set_index(['position from terminus', 'observable'], inplace=True)\n",
    "oddsRatioDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminus = 'C'\n",
    "plotData = compute_oddsratio_plot_data(oddsRatioDf, terminus)\n",
    "plotData[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskData = compute_oddsratio_mask_data(oddsRatioDf, pvalueThresholds, terminus)\n",
    "maskData[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot with pvalue as text annotation in each square"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvalAnnotTable = compute_pvalueAnnotation(oddsRatioDf, pvalueThresholds, terminus)\n",
    "pvalAnnotTable[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskDataSmallN = compute_smallN_mask_data(oddsRatioDf, pvalueThresholds, terminus)\n",
    "maskDataSmallN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "filenamePlot = str(analysisOutputPath / taxonSuffix / \n",
    "                   (outputFilePrefix + \"_composition_bias_aa_\" + terminus + \"terminal\" + \".png\"))\n",
    "fig = plot_aa_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus, plotTitle)\n",
    "fig.savefig(filenamePlot, dpi=300, bbox_inches=\"tight\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment analysis of codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "refCodonTableDf = build_refCodonTable(codonTableBio)\n",
    "refCodonTableDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "bulkCodonFreq, bulkCodonRelativeFreq = count_bulk_codon_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize)\n",
    "bulkCodonFreqDf = convert_bulk_codon_freq_to_Df(bulkCodonFreq)\n",
    "print(bulkCodonFreq,'\\n',len(bulkCodonFreq),'\\n\\n')\n",
    "# Applying the sorted function convert the dict to a list\n",
    "bulkCodonRelativeFreqDf = pd.DataFrame(sorted(bulkCodonRelativeFreq.items(), key=lambda x: x[1]))\n",
    "bulkCodonRelativeFreqDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulkCodonRelativeFreqDf.to_csv(str(cterminalPath / 'mpn_bulk_codon_frequency.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctermCodonFreq, ntermCodonFreq = count_termina_codon_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize)\n",
    "print(ctermCodonFreq, ntermCodonFreq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonDf = compute_odds_ratio_codons(bulkCodonFreq, \\\n",
    "                                             ctermCodonFreq, cTerminalSize, \\\n",
    "                                             ntermCodonFreq, nTerminalSize,\\\n",
    "                                             verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "oddsRatioCodonDf.loc[0] if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Select only C terminus\n",
    "oddsRatioCodonDf[ oddsRatioCodonDf['terminus']=='C' ].drop('terminus', axis=1) if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Draw codon biases table using HTML\n",
    "\n",
    "Using pandas dataframe styling. See [Pandas Style documentation](http://pandas.pydata.org/pandas-docs/stable/style.html)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`  |  U      |  C      |  A      |  G      |\n",
    "--+---------+---------+---------+---------+--\n",
    "U | UUU F   | UCU S   | UAU Y   | UGU C   | U\n",
    "U | UUC F   | UCC S   | UAC Y   | UGC C   | C\n",
    "U | UUA L   | UCA S   | UAA Stop| UGA Stop| A\n",
    "U | UUG L(s)| UCG S   | UAG Stop| UGG W   | G\n",
    "--+---------+---------+---------+---------+--\n",
    "C | CUU L   | CCU P   | CAU H   | CGU R   | U\n",
    "C | CUC L   | CCC P   | CAC H   | CGC R   | C\n",
    "C | CUA L   | CCA P   | CAA Q   | CGA R   | A\n",
    "C | CUG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
    "--+---------+---------+---------+---------+--\n",
    "A | AUU I(s)| ACU T   | AAU N   | AGU S   | U\n",
    "A | AUC I(s)| ACC T   | AAC N   | AGC S   | C\n",
    "A | AUA I(s)| ACA T   | AAA K   | AGA R   | A\n",
    "A | AUG M(s)| ACG T   | AAG K   | AGG R   | G\n",
    "--+---------+---------+---------+---------+--\n",
    "G | GUU V   | GCU A   | GAU D   | GGU G   | U\n",
    "G | GUC V   | GCC A   | GAC D   | GGC G   | C\n",
    "G | GUA V   | GCA A   | GAA E   | GGA G   | A\n",
    "G | GUG V(s)| GCG A   | GAG E   | GGG G   | G`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = compute_codon_oddsratio_plot_data(oddsRatioCodonDf, 'C')\n",
    "plotData[:5] if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvalAnnotTable = compute_codon_pvalueAnnotation(oddsRatioCodonDf, pvalueThresholds, 'C')\n",
    "pvalAnnotTable[:5] if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define global colormap for codon table\n",
    "colormapDiverging = seaborn.blend_palette(seaborn.color_palette(\"RdBu_r\"), as_cmap=True, input='rgb')\n",
    "# Normalize colormap\n",
    "vmax = 3\n",
    "vmin = -vmax\n",
    "\n",
    "codonTableColorDf = plot_codon_table_dataframe_styling(plotData[-2], pvalAnnotTable[-2], species_name, \n",
    "                                                           codonTableBio, colormapDiverging, vmin, vmax)\n",
    "codonTableColorDf if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "display(HTML(plot_codon_table_html_table_formatting(codonTableColorDf))) if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "positionDfIndex = -1\n",
    "fig = plot_codon_table(plotData, pvalAnnotTable, 'C', positionDfIndex, species_name, codonTableBio, analysisOutputFolder)\n",
    "filenamePlot = species_name + '_bias_codon_table_' + 'C' + 'terminal' + '_pos' + '{:03d}'.format(positionDfIndex)\n",
    "fig.savefig(os.path.join(analysisOutputFolder, filenamePlot + '.png'), dpi=400, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Draw codon biases table using heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminus = 'N'\n",
    "plotData = compute_codon_oddsratio_plot_data(oddsRatioCodonDf, terminus)\n",
    "plotData[:5] if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskData = compute_codon_oddsratio_mask_data(oddsRatioCodonDf, pvalueThresholds, terminus)\n",
    "maskData[:5] if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pvalAnnotTable = compute_codon_pvalueAnnotation(oddsRatioCodonDf, pvalueThresholds, terminus)\n",
    "pvalAnnotTable[:5] if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskDataSmallN = compute_codon_smallN_mask_data(oddsRatioCodonDf, pvalueThresholds, terminus)\n",
    "maskDataSmallN.head() if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plot_codon_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus, species_name, refCodonTableDf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example of table using matplotlib\n",
    "\n",
    "Not very flexible, especially for cell size manipulation, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if False:\n",
    "    data = [\n",
    "             [\"\" for x in range(10)],\n",
    "             [\"x\" for x in range(10)]\n",
    "           ]\n",
    "    cellColours = [\n",
    "             [str(0.9) for x in range(10)],\n",
    "             [str(0.7) for x in range(10)]\n",
    "           ]\n",
    "\n",
    "    columns = [\"head\" for x in range(10)]\n",
    "    rows = [1, 2]\n",
    "\n",
    "    fig = plt.figure(figsize=(15,8))\n",
    "    ax = fig.add_subplot(111, frameon=True, xticks=[], yticks=[])\n",
    "\n",
    "    the_table = plt.table(cellText=data,\n",
    "                          cellColours=cellColours,\n",
    "                          loc=(0,0), cellLoc='center',\n",
    "                          rowLabels=rows,\n",
    "                          colLabels=columns,\n",
    "                          rowColours=['r','g']\n",
    "                          )\n",
    "\n",
    "    table_props = the_table.properties()\n",
    "    table_cells = table_props['child_artists']\n",
    "    for cell in table_cells:\n",
    "        cell.set_height(0.1)\n",
    "    table_cells[5].set_linewidth(0.0)\n",
    "\n",
    "    cellDict=the_table.get_celld()\n",
    "    cellDict[(1,0)].set_linewidth(4)\n",
    "    #cellDict[(2,3)].set_background('blue')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pairs of amino acids at N-terminal and C-terminal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "subseqSize = 2\n",
    "bulkSubseqFreq, bulkSubseqRelativeFreq = count_bulk_subsequence_frequency(allProteinSeqDf_nr, cTerminalSize,\n",
    "                                                                          nTerminalSize, subseqSize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctermSubseqFreq, ntermSubseqFreq = count_termina_subseq_frequency(allProteinSeqDf_nr, cTerminalSize, nTerminalSize, subseqSize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "subseqOddsRatioDf, subseqOddsRatioPairsDf = compute_odds_ratio_subseq(bulkSubseqFreq, bulkSeqAAFreq,\\\n",
    "                                                                      ctermSubseqFreq, cTerminalSize,\\\n",
    "                                                                      ntermSubseqFreq, nTerminalSize, subseqSize)\n",
    "subseqOddsRatioDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "terminus = 'C'\n",
    "statistical_test = 'theoretical'\n",
    "statistical_test = 'contingency'\n",
    "subseqPlotData = compute_subseq_oddsratio_plot_data(subseqOddsRatioPairsDf, terminus, statistical_test)\n",
    "subseqPlotData if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqMaskData = compute_subseq_oddsratio_mask_data(subseqOddsRatioPairsDf, pvalueThresholds, terminus, statistical_test)\n",
    "subseqMaskData if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqMaskDataSmallN = compute_subseq_smallN_mask_data(subseqOddsRatioPairsDf, pvalueThresholds, 'C', statistical_test)\n",
    "subseqMaskDataSmallN.head() if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqpvalAnnotTable = compute_subseq_pvalueAnnotation(subseqOddsRatioPairsDf, pvalueThresholds, terminus, statistical_test)\n",
    "subseqpvalAnnotTable if runExampleAnalysis else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig = plot_aapair_composition_map(subseqPlotData, subseqMaskData, subseqpvalAnnotTable, subseqMaskDataSmallN, terminus, statistical_test, species_name)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge CDS dataframe with the GO database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allCDSDf_merged2 = merge_CDS_with_GO_database(allCDSDf_nr, amiGODf, membraneGOTermsDf)\n",
    "allCDSDf_merged2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nRandomGenomes=0\n",
    "skipClustering = False\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False\n",
    "speciesList = []\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True\n",
    "method = 'iterative'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'speciesSetEco'\n",
    "assemblyAccessionList=['GCF_000005845.2']\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'speciesSetEcoBsu'\n",
    "assemblyAccessionList=['GCF_000005845.2','GCF_000009045.1']\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyAccessionList=['GCF_000027345.1']\n",
    "speciesSetName = 'speciesSetMyco'\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'speciesSetTest3'\n",
    "assemblyAccessionList=['GCF_000005845.2','GCF_000009045.1','GCF_000007565.2']\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyAccessionList=['GCF_000027345.1','GCF_000008245.1','GCF_000008205.1']\n",
    "speciesSetName = 'speciesSetMyco3'\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyAccessionList=['GCF_000027345.1' for i in range(0,5)]\n",
    "speciesSetName = 'memTestDf'\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyAccessionList=['GCF_000005845.2' for i in range(0,5)]\n",
    "speciesSetName = 'memTestDf'\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Heliobacterium modesticaldum Ice1, with ~2500 seqs\n",
    "assemblyAccessionList=['GCF_000019165.1' for i in range(0,5)]\n",
    "speciesSetName = 'memTestDf'\n",
    "chooseRandomGenomes = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'speciesSetRandom120'\n",
    "chooseRandomGenomes = True\n",
    "nRandomGenomes = 120\n",
    "skipClustering = False\n",
    "GOanalysis = False\n",
    "method = 'iterative'\n",
    "multispeciesAllCDSDf_nr = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyAccessionList=['GCF_000027345.1','GCF_000008245.1','GCF_000008205.1']\n",
    "assemblyAccessionList=['GCF_000027345.1']\n",
    "speciesSetName = 'setMyco3MethodDataframe'\n",
    "speciesSetName = 'setMyco3MethodIterative'\n",
    "method = 'allSeqDataframe'\n",
    "method = 'iterative'\n",
    "chooseRandomGenomes = False\n",
    "nRandomGenomes = 0\n",
    "skipClustering = False\n",
    "GOanalysis = False\n",
    "multispeciesAllCDSDf_nr = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO membrane / not membrane proteins analysis\n",
    "\n",
    "# amiGOFolder = os.path.join('AmiGO','Bacterial_species_annotations')\n",
    "# amiGOSpeciesList = ['Bacillus subtilis subsp. subtilis str. 168',\n",
    "#                     'Escherichia coli str. K-12',\n",
    "#                     'Pseudomonas aeruginosa PAO1',\n",
    "#                     'Pseudomonas protegens Pf-5',\n",
    "#                     'Shewanella oneidensis MR-1']\n",
    "\n",
    "# amiGOSpeciesList = ['Escherichia coli str. K-12']\n",
    "# amiGODf = pd.read_csv(os.path.abspath(os.path.join(refSeqFolder, amiGOFolder, os.pardir, 'amiGODf.csv')), index_col=0)\n",
    "\n",
    "# AmiGOFolder = 'AmiGO'\n",
    "# membraneGOTermsFilename = 'GO_integral_component_membrane_descendants.csv'\n",
    "# indexName = 'GOid'\n",
    "# colNames = [indexName,'name','term_type']\n",
    "# membraneGOTermsDf = pd.read_csv(os.path.join(refSeqFolder, AmiGOFolder, membraneGOTermsFilename), sep='\\t', \n",
    "#                                 header=None, comment='#', names=colNames)\n",
    "# # Add the parent term \"integral component of membrane\"\n",
    "# membraneGOTermsDf = membraneGOTermsDf.append({'GOid':'GO:0016021','name':'integral component of membrane',\n",
    "#                                               'term_type':'cellular_component'}, ignore_index=True)\n",
    "\n",
    "# speciesSetName = 'speciesSetGO'\n",
    "# assemblyAccessionList = []\n",
    "# speciesList = amiGOSpeciesList\n",
    "# chooseRandomGenomes = False\n",
    "# nRandomGenomes = 0\n",
    "# skipClustering = False\n",
    "# GOanalysis = True\n",
    "# method = 'allSeqDataframe'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyAccessionList=['GCF_000027345.1','GCF_000008245.1']\n",
    "speciesSetName = 'speciesSetMyco2'\n",
    "method = 'iterative'\n",
    "chooseRandomGenomes = False\n",
    "skipClustering = True\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False\n",
    "speciesList = []\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyAccessionList=['GCF_000012245.1']\n",
    "speciesSetName = 'test1Species'\n",
    "method = 'iterative'\n",
    "chooseRandomGenomes = False\n",
    "skipClustering = True\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False\n",
    "speciesList = []\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The first genome file is missing on drive. Test the skipping.\n",
    "assemblyAccessionList=['GCF_000972245.3','GCF_000012245.1']\n",
    "speciesSetName = 'testMissing'\n",
    "method = 'iterative'\n",
    "chooseRandomGenomes = False\n",
    "skipClustering = False\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False\n",
    "speciesList = []\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'testEco'\n",
    "assemblyAccessionList=['GCF_000005845.2']\n",
    "speciesList = []\n",
    "chooseRandomGenomes = False\n",
    "nRandomGenomes = 0\n",
    "skipClustering = False\n",
    "method = 'iterative'\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# arg_parser = argparse.ArgumentParser(description='Takes the index of phylum as argument.')\n",
    "# arg_parser.add_argument('phylumIndex', type=int)\n",
    "# arg_namespace = arg_parser.parse_args()\n",
    "# phylumIndex = arg_namespace.phylumIndex - 1\n",
    "\n",
    "# # Phyla analysis\n",
    "# assemblySummaryRepPhylaDf = pd.read_csv(os.path.abspath(os.path.join(taxonomyFolder, 'assemblySummaryRepPhylaDf.csv')))\n",
    "# assemblySummaryRepPhylaDf.set_index('assembly_accession', inplace=True)\n",
    "# phylaGroups = assemblySummaryRepPhylaDf.groupby(['phylum_taxid','phylum_name'])\n",
    "# print(\"phylumIndex:\", phylumIndex, \"/ Nb of phyla in refseq database: \", len(phylaGroups),\" (0-based index)\\n\")\n",
    "# print(phylaGroups.size(),\"\\n\")\n",
    "# assemblyAccessionPhylumDf = list(phylaGroups)[phylumIndex][1]\n",
    "# assemblyAccessionList = assemblyAccessionPhylumDf.index.unique()\n",
    "# print(\"len(assemblyAccessionList):\",len(assemblyAccessionList))\n",
    "# phylumId = int(list(phylaGroups)[phylumIndex][0][0])\n",
    "# print(\"phylumId:\",phylumId)\n",
    "# phylumName = list(phylaGroups)[phylumIndex][0][1]\n",
    "# print(\"phylumName:\",phylumName)\n",
    "\n",
    "# speciesSetName = 'setPhylum_index{:0>2d}_PhylumId_{:0>8d}_{}'.format(phylumIndex, phylumId, phylumName)\n",
    "# print(\"speciesSetName:\",speciesSetName)\n",
    "# chooseRandomGenomes = False\n",
    "# nRandomGenomes = 0\n",
    "# skipClustering = False\n",
    "# GOanalysis = False\n",
    "# multispeciesAllCDSDf_nr = None\n",
    "# method = 'iterative'\n",
    "# skipAnalysisExistingDataFiles = True\n",
    "# amiGODf = None\n",
    "# membraneGOTermsDf = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'speciesSetAllRefSeq'\n",
    "assemblyAccessionList = list(assemblySummaryRepDf.index)\n",
    "speciesList = []\n",
    "chooseRandomGenomes = False\n",
    "nRandomGenomes = 0\n",
    "skipClustering = False\n",
    "method = 'iterative'\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'speciesSetRandom500'\n",
    "speciesList = []\n",
    "chooseRandomGenomes = True\n",
    "nRandomGenomes = 500\n",
    "skipClustering = False\n",
    "method = 'iterative'\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'memTestRandom50'\n",
    "assemblyAccessionList=['GCF_000027345.1' for i in range(0,6)]\n",
    "assemblyAccessionList=['GCF_000027345.1' for i in range(0,4)]\n",
    "assemblyAccessionList=[]\n",
    "speciesList = []\n",
    "chooseRandomGenomes = True\n",
    "nRandomGenomes = 50\n",
    "skipClustering = False\n",
    "method = 'iterative'\n",
    "multispeciesAllCDSDf_nr = None\n",
    "skipAnalysisExistingDataFiles = True\n",
    "amiGODf = None\n",
    "membraneGOTermsDf = None\n",
    "GOanalysis = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mollicutes class analysis\n",
    "assemblyAccessionList = pd.read_csv(os.path.join(refSeqFolder, taxonomyFolder, 'assembly_summary_mollicutes.csv'))\\\n",
    "    ['assembly_accession'].unique()\n",
    "speciesSetName = 'setMollicutes'\n",
    "chooseRandomGenomes = False\n",
    "nRandomGenomes = 0\n",
    "skipClustering = False\n",
    "GOanalysis = False\n",
    "multispeciesAllCDSDf_nr = None\n",
    "method = 'iterative'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "os.chdir(root_folder)\n",
    "full_analysis_multispecies(refSeqFolder=refSeqFolder,\n",
    "                           speciesSetName=speciesSetName,\n",
    "                           chooseRandomGenomes=chooseRandomGenomes,\n",
    "                           nRandomGenomes=nRandomGenomes,\n",
    "                           assemblyAccessionList=assemblyAccessionList,\n",
    "                           speciesList=speciesList,\n",
    "                           dfFormatOutput='csv',\n",
    "                           skipAnalysisExistingDataFiles=skipAnalysisExistingDataFiles,\n",
    "                           skipClustering=skipClustering,\n",
    "                           multispeciesAllCDSDf_nr=multispeciesAllCDSDf_nr, \n",
    "                           GOanalysis=GOanalysis,\n",
    "                           amiGODf=amiGODf,\n",
    "                           membraneGOTermsDf=membraneGOTermsDf,\n",
    "                           method=method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mycoplasma pneumoniae\n",
    "full_analysis_multispecies(str(refSeqPath), str(analysisCtermDataPath), str(analysisCtermPlotsPath), genome_accession,\n",
    "                           chooseRandomGenomes=False,\n",
    "                           nRandomGenomes=0,\n",
    "                           assemblyAccessionList=[genome_accession],\n",
    "                           speciesList=[],\n",
    "                           multispeciesAllCDSDf_nr=None,\n",
    "                           dfFormatOutput='csv',\n",
    "                           skipAnalysisExistingDataFiles=True,\n",
    "                           skipStatisticsAnalysis=False,\n",
    "                           method='iterative',\n",
    "                           skipClustering=False,\n",
    "                           GOanalysis=False, amiGODf=None, membraneGOTermsDf=None\n",
    "                           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Run this cell just ONCE if the captured cell output was not written to file\n",
    "\n",
    "# %capture cell_output --no-stderr\n",
    "\n",
    "# print(\"Hello world!\")\n",
    "\n",
    "# with open(os.path.join(analysisDataFolder, speciesSetName + \"_analysis_output.txt\"), 'w') as f:\n",
    "#     f.write(cell_output.stdout)\n",
    "# del cell_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Taxon enrichment pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## import taxon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpnAnnotationPath = p.mpnAnnotationPath\n",
    "rootPath = p.rootPath\n",
    "cterminalPath = p.cterminalPath\n",
    "refSeqPath = p.refSeqPath\n",
    "taxonomyPath = p.taxonomyPath\n",
    "phylaPath = p.phylaPath\n",
    "entrezGenePath = p.entrezGenePath\n",
    "OMAPath = p.OMAPath\n",
    "eggNOGPath = p.eggNOGPath\n",
    "NCBI_COG_path = p.NCBI_COG_path\n",
    "analysisEggnogPath = p.analysisEggnogPath\n",
    "analysisCtermDataPath = p.analysisCtermDataPath\n",
    "analysisCtermPlotsPath = p.analysisCtermPlotsPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing list of genome assemblies of the RefSeq database with their taxonomy\n",
    "assemblySummaryRepTaxonomyDf = pd.read_csv(str(taxonomyPath / 'assemblySummaryRepTaxonomyDf.csv'))\n",
    "assemblySummaryRepTaxonomyDf.set_index('assembly_accession', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>...</th>\n",
       "      <th>suborder_name</th>\n",
       "      <th>suborder_taxid</th>\n",
       "      <th>subphylum_name</th>\n",
       "      <th>subphylum_taxid</th>\n",
       "      <th>subspecies_name</th>\n",
       "      <th>subspecies_taxid</th>\n",
       "      <th>superkingdom_name</th>\n",
       "      <th>superkingdom_taxid</th>\n",
       "      <th>tribe_name</th>\n",
       "      <th>tribe_taxid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000005825.2</th>\n",
       "      <td>0</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>398511</td>\n",
       "      <td>Bacillus pseudofirmus OF4</td>\n",
       "      <td>strain=OF4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000005845.2</th>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA57779</td>\n",
       "      <td>SAMN02604091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>511145</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>strain=K-12 substr. MG1655</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006605.1</th>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA3283089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>306537</td>\n",
       "      <td>Corynebacterium jeikeium K411</td>\n",
       "      <td>strain=K411 = NCTC 11915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006685.1</th>\n",
       "      <td>3</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603994</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>243161</td>\n",
       "      <td>Chlamydia muridarum str. Nigg</td>\n",
       "      <td>strain=Nigg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006725.1</th>\n",
       "      <td>4</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603773</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>160492</td>\n",
       "      <td>Xylella fastidiosa 9a5c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006745.1</th>\n",
       "      <td>5</td>\n",
       "      <td>PRJNA57623</td>\n",
       "      <td>SAMN02603969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>243277</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>strain=N16961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006765.1</th>\n",
       "      <td>6</td>\n",
       "      <td>PRJNA57945</td>\n",
       "      <td>SAMN02603714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>208964</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>strain=PAO1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006785.2</th>\n",
       "      <td>7</td>\n",
       "      <td>PRJNA57845</td>\n",
       "      <td>SAMN02604089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>strain=SF370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006845.1</th>\n",
       "      <td>8</td>\n",
       "      <td>PRJNA57611</td>\n",
       "      <td>SAMN02604088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>242231</td>\n",
       "      <td>Neisseria gonorrhoeae FA 1090</td>\n",
       "      <td>strain=FA 1090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006865.1</th>\n",
       "      <td>9</td>\n",
       "      <td>PRJNA57671</td>\n",
       "      <td>SAMN02603339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>272623</td>\n",
       "      <td>Lactococcus lactis subsp. lactis Il1403</td>\n",
       "      <td>strain=IL1403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lactococcus lactis subsp. lactis</td>\n",
       "      <td>1360.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006905.1</th>\n",
       "      <td>10</td>\n",
       "      <td>PRJNA57891</td>\n",
       "      <td>SAMN02604005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>190650</td>\n",
       "      <td>Caulobacter crescentus CB15</td>\n",
       "      <td>strain=CB15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006925.2</th>\n",
       "      <td>11</td>\n",
       "      <td>PRJNA62907</td>\n",
       "      <td>SAMN02603599</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>198214</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>strain=301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006945.1</th>\n",
       "      <td>12</td>\n",
       "      <td>PRJNA57799</td>\n",
       "      <td>SAMN02604315</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>99287</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>strain=LT2; SGSC 1412; ATCC 700720</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Salmonella enterica subsp. enterica</td>\n",
       "      <td>59201.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006965.1</th>\n",
       "      <td>13</td>\n",
       "      <td>PRJNA57603</td>\n",
       "      <td>SAMEA3283068</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>266834</td>\n",
       "      <td>Sinorhizobium meliloti 1021</td>\n",
       "      <td>strain=1021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000006985.1</th>\n",
       "      <td>14</td>\n",
       "      <td>PRJNA57897</td>\n",
       "      <td>SAMN02604006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>194439</td>\n",
       "      <td>Chlorobium tepidum TLS</td>\n",
       "      <td>strain=TLS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007025.1</th>\n",
       "      <td>15</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603141</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>272944</td>\n",
       "      <td>Rickettsia conorii str. Malish 7</td>\n",
       "      <td>strain=Malish 7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Rickettsieae</td>\n",
       "      <td>33988.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007045.1</th>\n",
       "      <td>16</td>\n",
       "      <td>PRJNA57859</td>\n",
       "      <td>SAMN02603218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>171101</td>\n",
       "      <td>Streptococcus pneumoniae R6</td>\n",
       "      <td>strain=R6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007085.1</th>\n",
       "      <td>17</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>273068</td>\n",
       "      <td>Caldanaerobacter subterraneus subsp. tengconge...</td>\n",
       "      <td>strain=MB4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Caldanaerobacter subterraneus subsp. tengconge...</td>\n",
       "      <td>119072.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007125.1</th>\n",
       "      <td>18</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>224914</td>\n",
       "      <td>Brucella melitensis bv. 1 str. 16M</td>\n",
       "      <td>strain=16M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007145.1</th>\n",
       "      <td>19</td>\n",
       "      <td>PRJNA57887</td>\n",
       "      <td>SAMN02603845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>190485</td>\n",
       "      <td>Xanthomonas campestris pv. campestris str. ATC...</td>\n",
       "      <td>strain=ATCC 33913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007265.1</th>\n",
       "      <td>20</td>\n",
       "      <td>PRJNA57943</td>\n",
       "      <td>SAMN02604013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>208435</td>\n",
       "      <td>Streptococcus agalactiae 2603V/R</td>\n",
       "      <td>strain=2603V/R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007325.1</th>\n",
       "      <td>21</td>\n",
       "      <td>PRJNA57885</td>\n",
       "      <td>SAMN02603417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>190304</td>\n",
       "      <td>Fusobacterium nucleatum subsp. nucleatum ATCC ...</td>\n",
       "      <td>strain=ATCC 25586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fusobacterium nucleatum subsp. nucleatum</td>\n",
       "      <td>76856.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007365.1</th>\n",
       "      <td>22</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02604269</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>198804</td>\n",
       "      <td>Buchnera aphidicola str. Sg (Schizaphis graminum)</td>\n",
       "      <td>strain=Sg (Schizaphis graminum)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007385.1</th>\n",
       "      <td>23</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>291331</td>\n",
       "      <td>Xanthomonas oryzae pv. oryzae KACC 10331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007465.2</th>\n",
       "      <td>24</td>\n",
       "      <td>PRJNA57947</td>\n",
       "      <td>SAMN02604090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>210007</td>\n",
       "      <td>Streptococcus mutans UA159</td>\n",
       "      <td>strain=UA159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007485.1</th>\n",
       "      <td>25</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603248</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>203267</td>\n",
       "      <td>Tropheryma whipplei str. Twist</td>\n",
       "      <td>strain=Twist</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007525.1</th>\n",
       "      <td>26</td>\n",
       "      <td>PRJNA57939</td>\n",
       "      <td>SAMN02603675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>206672</td>\n",
       "      <td>Bifidobacterium longum NCC2705</td>\n",
       "      <td>strain=NCC2705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007565.2</th>\n",
       "      <td>27</td>\n",
       "      <td>PRJNA57843</td>\n",
       "      <td>SAMN02603999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>160488</td>\n",
       "      <td>Pseudomonas putida KT2440</td>\n",
       "      <td>strain=KT2440</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007605.1</th>\n",
       "      <td>28</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603993</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>227941</td>\n",
       "      <td>Chlamydophila caviae GPIC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPIC</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000007625.1</th>\n",
       "      <td>29</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>212717</td>\n",
       "      <td>Clostridium tetani E88</td>\n",
       "      <td>strain=Massachusetts substr. E88</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001583415.1</th>\n",
       "      <td>1546</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04481318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1771</td>\n",
       "      <td>Mycobacterium phlei</td>\n",
       "      <td>strain=CCUG21000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001584185.1</th>\n",
       "      <td>1547</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04195677</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>279113</td>\n",
       "      <td>Collimonas pratensis</td>\n",
       "      <td>strain=Ter91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001586165.1</th>\n",
       "      <td>1548</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04487151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>82983</td>\n",
       "      <td>Obesumbacterium proteus</td>\n",
       "      <td>strain=DSM 2777</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001586195.1</th>\n",
       "      <td>1549</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04111444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>76853</td>\n",
       "      <td>Solibacillus silvestris</td>\n",
       "      <td>strain=DSM 12223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001586215.1</th>\n",
       "      <td>1550</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN03455925</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>146827</td>\n",
       "      <td>Corynebacterium simulans</td>\n",
       "      <td>strain=PES1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001586255.1</th>\n",
       "      <td>1551</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04361929</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1470434</td>\n",
       "      <td>Zhongshania aliphaticivorans</td>\n",
       "      <td>strain=SM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001593245.1</th>\n",
       "      <td>1552</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04012505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>654</td>\n",
       "      <td>Aeromonas veronii</td>\n",
       "      <td>strain=TH0426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001597285.1</th>\n",
       "      <td>1553</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04539664</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>43263</td>\n",
       "      <td>Pseudomonas alcaligenes</td>\n",
       "      <td>strain=NEB 585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001606005.1</th>\n",
       "      <td>1554</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04569174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1703</td>\n",
       "      <td>Brevibacterium linens</td>\n",
       "      <td>strain=BS258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001606025.1</th>\n",
       "      <td>1555</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04549283</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>261164</td>\n",
       "      <td>Psychrobacter alimentarius</td>\n",
       "      <td>strain=PAMC 27889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001610975.1</th>\n",
       "      <td>1556</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04046241</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1219058</td>\n",
       "      <td>Sphingopyxis terrae NBRC 15098</td>\n",
       "      <td>strain=203-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001611135.1</th>\n",
       "      <td>1557</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN03876484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>51663</td>\n",
       "      <td>Pediococcus damnosus</td>\n",
       "      <td>strain=TMW 2.1535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001618685.1</th>\n",
       "      <td>1558</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04445793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>129338</td>\n",
       "      <td>Geobacillus subterraneus</td>\n",
       "      <td>strain=KCTC3922(T)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001620305.1</th>\n",
       "      <td>1559</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04521220</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1051973</td>\n",
       "      <td>Rhodococcus fascians D188</td>\n",
       "      <td>strain=D188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001632865.1</th>\n",
       "      <td>1560</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04160834</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1889</td>\n",
       "      <td>Streptomyces ambofaciens</td>\n",
       "      <td>strain=DSM 40697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001636015.1</th>\n",
       "      <td>1561</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN03455375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>29433</td>\n",
       "      <td>Moraxella ovis</td>\n",
       "      <td>strain=199/55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001641285.1</th>\n",
       "      <td>1562</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04433089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>445710</td>\n",
       "      <td>Dyella thiooxydans</td>\n",
       "      <td>strain=ATSB10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001642655.1</th>\n",
       "      <td>1563</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04415495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1519</td>\n",
       "      <td>Clostridium tyrobutyricum</td>\n",
       "      <td>strain=KCTC 5387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001644565.1</th>\n",
       "      <td>1564</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN03285365</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1182568</td>\n",
       "      <td>Deinococcus puniceus</td>\n",
       "      <td>strain=DY1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001648175.1</th>\n",
       "      <td>1565</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04817514</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>376489</td>\n",
       "      <td>Halotalea alkalilenta</td>\n",
       "      <td>strain=IHB B 13600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001652465.1</th>\n",
       "      <td>1566</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN05003931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>84292</td>\n",
       "      <td>Microbacterium chocolatum</td>\n",
       "      <td>strain=SIT 101</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001652485.1</th>\n",
       "      <td>1567</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04994838</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>419475</td>\n",
       "      <td>Ochrobactrum pseudogrignonense</td>\n",
       "      <td>strain=K8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001652565.1</th>\n",
       "      <td>1568</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04992853</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>879274</td>\n",
       "      <td>Shinella sp. HZN7</td>\n",
       "      <td>strain=HZN7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001653755.1</th>\n",
       "      <td>1569</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04572126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1300343</td>\n",
       "      <td>Dokdonia donghaensis DSW-1</td>\n",
       "      <td>strain=DSW-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001654495.1</th>\n",
       "      <td>1570</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN05161780</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1849967</td>\n",
       "      <td>Streptomyces sp. SAT1</td>\n",
       "      <td>strain=SAT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001655245.1</th>\n",
       "      <td>1571</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN03702434</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1354275</td>\n",
       "      <td>Mycobacterium vaccae 95051</td>\n",
       "      <td>strain=95051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_900011245.1</th>\n",
       "      <td>1572</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA3647810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>376</td>\n",
       "      <td>Bradyrhizobium sp.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_900078695.1</th>\n",
       "      <td>1573</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA3906487</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>123899</td>\n",
       "      <td>Bordetella trematum</td>\n",
       "      <td>strain=H044680328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_900086555.1</th>\n",
       "      <td>1574</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA3174300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>28091</td>\n",
       "      <td>Neisseria weaveri</td>\n",
       "      <td>strain=NCTC13585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_900087055.1</th>\n",
       "      <td>1575</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA2665130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>703</td>\n",
       "      <td>Plesiomonas shigelloides</td>\n",
       "      <td>strain=NCTC10360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1576 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0   bioproject     biosample  wgs_master  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000005825.2              0  PRJNA224116  SAMN02603086         NaN   \n",
       "GCF_000005845.2              1   PRJNA57779  SAMN02604091         NaN   \n",
       "GCF_000006605.1              2  PRJNA224116  SAMEA3283089         NaN   \n",
       "GCF_000006685.1              3  PRJNA224116  SAMN02603994         NaN   \n",
       "GCF_000006725.1              4  PRJNA224116  SAMN02603773         NaN   \n",
       "GCF_000006745.1              5   PRJNA57623  SAMN02603969         NaN   \n",
       "GCF_000006765.1              6   PRJNA57945  SAMN02603714         NaN   \n",
       "GCF_000006785.2              7   PRJNA57845  SAMN02604089         NaN   \n",
       "GCF_000006845.1              8   PRJNA57611  SAMN02604088         NaN   \n",
       "GCF_000006865.1              9   PRJNA57671  SAMN02603339         NaN   \n",
       "GCF_000006905.1             10   PRJNA57891  SAMN02604005         NaN   \n",
       "GCF_000006925.2             11   PRJNA62907  SAMN02603599         NaN   \n",
       "GCF_000006945.1             12   PRJNA57799  SAMN02604315         NaN   \n",
       "GCF_000006965.1             13   PRJNA57603  SAMEA3283068         NaN   \n",
       "GCF_000006985.1             14   PRJNA57897  SAMN02604006         NaN   \n",
       "GCF_000007025.1             15  PRJNA224116  SAMN02603141         NaN   \n",
       "GCF_000007045.1             16   PRJNA57859  SAMN02603218         NaN   \n",
       "GCF_000007085.1             17  PRJNA224116  SAMN02603009         NaN   \n",
       "GCF_000007125.1             18  PRJNA224116  SAMN02603416         NaN   \n",
       "GCF_000007145.1             19   PRJNA57887  SAMN02603845         NaN   \n",
       "GCF_000007265.1             20   PRJNA57943  SAMN02604013         NaN   \n",
       "GCF_000007325.1             21   PRJNA57885  SAMN02603417         NaN   \n",
       "GCF_000007365.1             22  PRJNA224116  SAMN02604269         NaN   \n",
       "GCF_000007385.1             23  PRJNA224116  SAMN02603680         NaN   \n",
       "GCF_000007465.2             24   PRJNA57947  SAMN02604090         NaN   \n",
       "GCF_000007485.1             25  PRJNA224116  SAMN02603248         NaN   \n",
       "GCF_000007525.1             26   PRJNA57939  SAMN02603675         NaN   \n",
       "GCF_000007565.2             27   PRJNA57843  SAMN02603999         NaN   \n",
       "GCF_000007605.1             28  PRJNA224116  SAMN02603993         NaN   \n",
       "GCF_000007625.1             29  PRJNA224116  SAMN02603289         NaN   \n",
       "...                        ...          ...           ...         ...   \n",
       "GCF_001583415.1           1546  PRJNA224116  SAMN04481318         NaN   \n",
       "GCF_001584185.1           1547  PRJNA224116  SAMN04195677         NaN   \n",
       "GCF_001586165.1           1548  PRJNA224116  SAMN04487151         NaN   \n",
       "GCF_001586195.1           1549  PRJNA224116  SAMN04111444         NaN   \n",
       "GCF_001586215.1           1550  PRJNA224116  SAMN03455925         NaN   \n",
       "GCF_001586255.1           1551  PRJNA224116  SAMN04361929         NaN   \n",
       "GCF_001593245.1           1552  PRJNA224116  SAMN04012505         NaN   \n",
       "GCF_001597285.1           1553  PRJNA224116  SAMN04539664         NaN   \n",
       "GCF_001606005.1           1554  PRJNA224116  SAMN04569174         NaN   \n",
       "GCF_001606025.1           1555  PRJNA224116  SAMN04549283         NaN   \n",
       "GCF_001610975.1           1556  PRJNA224116  SAMN04046241         NaN   \n",
       "GCF_001611135.1           1557  PRJNA224116  SAMN03876484         NaN   \n",
       "GCF_001618685.1           1558  PRJNA224116  SAMN04445793         NaN   \n",
       "GCF_001620305.1           1559  PRJNA224116  SAMN04521220         NaN   \n",
       "GCF_001632865.1           1560  PRJNA224116  SAMN04160834         NaN   \n",
       "GCF_001636015.1           1561  PRJNA224116  SAMN03455375         NaN   \n",
       "GCF_001641285.1           1562  PRJNA224116  SAMN04433089         NaN   \n",
       "GCF_001642655.1           1563  PRJNA224116  SAMN04415495         NaN   \n",
       "GCF_001644565.1           1564  PRJNA224116  SAMN03285365         NaN   \n",
       "GCF_001648175.1           1565  PRJNA224116  SAMN04817514         NaN   \n",
       "GCF_001652465.1           1566  PRJNA224116  SAMN05003931         NaN   \n",
       "GCF_001652485.1           1567  PRJNA224116  SAMN04994838         NaN   \n",
       "GCF_001652565.1           1568  PRJNA224116  SAMN04992853         NaN   \n",
       "GCF_001653755.1           1569  PRJNA224116  SAMN04572126         NaN   \n",
       "GCF_001654495.1           1570  PRJNA224116  SAMN05161780         NaN   \n",
       "GCF_001655245.1           1571  PRJNA224116  SAMN03702434         NaN   \n",
       "GCF_900011245.1           1572  PRJNA224116  SAMEA3647810         NaN   \n",
       "GCF_900078695.1           1573  PRJNA224116  SAMEA3906487         NaN   \n",
       "GCF_900086555.1           1574  PRJNA224116  SAMEA3174300         NaN   \n",
       "GCF_900087055.1           1575  PRJNA224116  SAMEA2665130         NaN   \n",
       "\n",
       "                          refseq_category    taxid  \\\n",
       "assembly_accession                                   \n",
       "GCF_000005825.2     representative genome   398511   \n",
       "GCF_000005845.2          reference genome   511145   \n",
       "GCF_000006605.1     representative genome   306537   \n",
       "GCF_000006685.1     representative genome   243161   \n",
       "GCF_000006725.1     representative genome   160492   \n",
       "GCF_000006745.1          reference genome   243277   \n",
       "GCF_000006765.1          reference genome   208964   \n",
       "GCF_000006785.2          reference genome   160490   \n",
       "GCF_000006845.1          reference genome   242231   \n",
       "GCF_000006865.1          reference genome   272623   \n",
       "GCF_000006905.1          reference genome   190650   \n",
       "GCF_000006925.2          reference genome   198214   \n",
       "GCF_000006945.1          reference genome    99287   \n",
       "GCF_000006965.1          reference genome   266834   \n",
       "GCF_000006985.1          reference genome   194439   \n",
       "GCF_000007025.1     representative genome   272944   \n",
       "GCF_000007045.1          reference genome   171101   \n",
       "GCF_000007085.1     representative genome   273068   \n",
       "GCF_000007125.1     representative genome   224914   \n",
       "GCF_000007145.1          reference genome   190485   \n",
       "GCF_000007265.1          reference genome   208435   \n",
       "GCF_000007325.1          reference genome   190304   \n",
       "GCF_000007365.1     representative genome   198804   \n",
       "GCF_000007385.1     representative genome   291331   \n",
       "GCF_000007465.2          reference genome   210007   \n",
       "GCF_000007485.1     representative genome   203267   \n",
       "GCF_000007525.1          reference genome   206672   \n",
       "GCF_000007565.2          reference genome   160488   \n",
       "GCF_000007605.1     representative genome   227941   \n",
       "GCF_000007625.1     representative genome   212717   \n",
       "...                                   ...      ...   \n",
       "GCF_001583415.1     representative genome     1771   \n",
       "GCF_001584185.1     representative genome   279113   \n",
       "GCF_001586165.1     representative genome    82983   \n",
       "GCF_001586195.1     representative genome    76853   \n",
       "GCF_001586215.1     representative genome   146827   \n",
       "GCF_001586255.1     representative genome  1470434   \n",
       "GCF_001593245.1     representative genome      654   \n",
       "GCF_001597285.1     representative genome    43263   \n",
       "GCF_001606005.1     representative genome     1703   \n",
       "GCF_001606025.1     representative genome   261164   \n",
       "GCF_001610975.1     representative genome  1219058   \n",
       "GCF_001611135.1     representative genome    51663   \n",
       "GCF_001618685.1     representative genome   129338   \n",
       "GCF_001620305.1     representative genome  1051973   \n",
       "GCF_001632865.1     representative genome     1889   \n",
       "GCF_001636015.1     representative genome    29433   \n",
       "GCF_001641285.1     representative genome   445710   \n",
       "GCF_001642655.1     representative genome     1519   \n",
       "GCF_001644565.1     representative genome  1182568   \n",
       "GCF_001648175.1     representative genome   376489   \n",
       "GCF_001652465.1     representative genome    84292   \n",
       "GCF_001652485.1     representative genome   419475   \n",
       "GCF_001652565.1     representative genome   879274   \n",
       "GCF_001653755.1     representative genome  1300343   \n",
       "GCF_001654495.1     representative genome  1849967   \n",
       "GCF_001655245.1     representative genome  1354275   \n",
       "GCF_900011245.1     representative genome      376   \n",
       "GCF_900078695.1     representative genome   123899   \n",
       "GCF_900086555.1     representative genome    28091   \n",
       "GCF_900087055.1     representative genome      703   \n",
       "\n",
       "                                                        organism_name  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000005825.2                             Bacillus pseudofirmus OF4   \n",
       "GCF_000005845.2             Escherichia coli str. K-12 substr. MG1655   \n",
       "GCF_000006605.1                         Corynebacterium jeikeium K411   \n",
       "GCF_000006685.1                         Chlamydia muridarum str. Nigg   \n",
       "GCF_000006725.1                               Xylella fastidiosa 9a5c   \n",
       "GCF_000006745.1          Vibrio cholerae O1 biovar El Tor str. N16961   \n",
       "GCF_000006765.1                           Pseudomonas aeruginosa PAO1   \n",
       "GCF_000006785.2                         Streptococcus pyogenes M1 GAS   \n",
       "GCF_000006845.1                         Neisseria gonorrhoeae FA 1090   \n",
       "GCF_000006865.1               Lactococcus lactis subsp. lactis Il1403   \n",
       "GCF_000006905.1                           Caulobacter crescentus CB15   \n",
       "GCF_000006925.2                         Shigella flexneri 2a str. 301   \n",
       "GCF_000006945.1     Salmonella enterica subsp. enterica serovar Ty...   \n",
       "GCF_000006965.1                           Sinorhizobium meliloti 1021   \n",
       "GCF_000006985.1                                Chlorobium tepidum TLS   \n",
       "GCF_000007025.1                      Rickettsia conorii str. Malish 7   \n",
       "GCF_000007045.1                           Streptococcus pneumoniae R6   \n",
       "GCF_000007085.1     Caldanaerobacter subterraneus subsp. tengconge...   \n",
       "GCF_000007125.1                    Brucella melitensis bv. 1 str. 16M   \n",
       "GCF_000007145.1     Xanthomonas campestris pv. campestris str. ATC...   \n",
       "GCF_000007265.1                      Streptococcus agalactiae 2603V/R   \n",
       "GCF_000007325.1     Fusobacterium nucleatum subsp. nucleatum ATCC ...   \n",
       "GCF_000007365.1     Buchnera aphidicola str. Sg (Schizaphis graminum)   \n",
       "GCF_000007385.1              Xanthomonas oryzae pv. oryzae KACC 10331   \n",
       "GCF_000007465.2                            Streptococcus mutans UA159   \n",
       "GCF_000007485.1                        Tropheryma whipplei str. Twist   \n",
       "GCF_000007525.1                        Bifidobacterium longum NCC2705   \n",
       "GCF_000007565.2                             Pseudomonas putida KT2440   \n",
       "GCF_000007605.1                             Chlamydophila caviae GPIC   \n",
       "GCF_000007625.1                                Clostridium tetani E88   \n",
       "...                                                               ...   \n",
       "GCF_001583415.1                                   Mycobacterium phlei   \n",
       "GCF_001584185.1                                  Collimonas pratensis   \n",
       "GCF_001586165.1                               Obesumbacterium proteus   \n",
       "GCF_001586195.1                               Solibacillus silvestris   \n",
       "GCF_001586215.1                              Corynebacterium simulans   \n",
       "GCF_001586255.1                          Zhongshania aliphaticivorans   \n",
       "GCF_001593245.1                                     Aeromonas veronii   \n",
       "GCF_001597285.1                               Pseudomonas alcaligenes   \n",
       "GCF_001606005.1                                 Brevibacterium linens   \n",
       "GCF_001606025.1                            Psychrobacter alimentarius   \n",
       "GCF_001610975.1                        Sphingopyxis terrae NBRC 15098   \n",
       "GCF_001611135.1                                  Pediococcus damnosus   \n",
       "GCF_001618685.1                              Geobacillus subterraneus   \n",
       "GCF_001620305.1                             Rhodococcus fascians D188   \n",
       "GCF_001632865.1                              Streptomyces ambofaciens   \n",
       "GCF_001636015.1                                        Moraxella ovis   \n",
       "GCF_001641285.1                                    Dyella thiooxydans   \n",
       "GCF_001642655.1                             Clostridium tyrobutyricum   \n",
       "GCF_001644565.1                                  Deinococcus puniceus   \n",
       "GCF_001648175.1                                 Halotalea alkalilenta   \n",
       "GCF_001652465.1                             Microbacterium chocolatum   \n",
       "GCF_001652485.1                        Ochrobactrum pseudogrignonense   \n",
       "GCF_001652565.1                                     Shinella sp. HZN7   \n",
       "GCF_001653755.1                            Dokdonia donghaensis DSW-1   \n",
       "GCF_001654495.1                                 Streptomyces sp. SAT1   \n",
       "GCF_001655245.1                            Mycobacterium vaccae 95051   \n",
       "GCF_900011245.1                                    Bradyrhizobium sp.   \n",
       "GCF_900078695.1                                   Bordetella trematum   \n",
       "GCF_900086555.1                                     Neisseria weaveri   \n",
       "GCF_900087055.1                              Plesiomonas shigelloides   \n",
       "\n",
       "                                    infraspecific_name isolate version_status  \\\n",
       "assembly_accession                                                              \n",
       "GCF_000005825.2                             strain=OF4     NaN         latest   \n",
       "GCF_000005845.2             strain=K-12 substr. MG1655     NaN         latest   \n",
       "GCF_000006605.1               strain=K411 = NCTC 11915     NaN         latest   \n",
       "GCF_000006685.1                            strain=Nigg     NaN         latest   \n",
       "GCF_000006725.1                                    NaN     NaN         latest   \n",
       "GCF_000006745.1                          strain=N16961     NaN         latest   \n",
       "GCF_000006765.1                            strain=PAO1     NaN         latest   \n",
       "GCF_000006785.2                           strain=SF370     NaN         latest   \n",
       "GCF_000006845.1                         strain=FA 1090     NaN         latest   \n",
       "GCF_000006865.1                          strain=IL1403     NaN         latest   \n",
       "GCF_000006905.1                            strain=CB15     NaN         latest   \n",
       "GCF_000006925.2                             strain=301     NaN         latest   \n",
       "GCF_000006945.1     strain=LT2; SGSC 1412; ATCC 700720     NaN         latest   \n",
       "GCF_000006965.1                            strain=1021     NaN         latest   \n",
       "GCF_000006985.1                             strain=TLS     NaN         latest   \n",
       "GCF_000007025.1                        strain=Malish 7     NaN         latest   \n",
       "GCF_000007045.1                              strain=R6     NaN         latest   \n",
       "GCF_000007085.1                             strain=MB4     NaN         latest   \n",
       "GCF_000007125.1                             strain=16M     NaN         latest   \n",
       "GCF_000007145.1                      strain=ATCC 33913     NaN         latest   \n",
       "GCF_000007265.1                         strain=2603V/R     NaN         latest   \n",
       "GCF_000007325.1                      strain=ATCC 25586     NaN         latest   \n",
       "GCF_000007365.1        strain=Sg (Schizaphis graminum)     NaN         latest   \n",
       "GCF_000007385.1                                    NaN     NaN         latest   \n",
       "GCF_000007465.2                           strain=UA159     NaN         latest   \n",
       "GCF_000007485.1                           strain=Twist     NaN         latest   \n",
       "GCF_000007525.1                         strain=NCC2705     NaN         latest   \n",
       "GCF_000007565.2                          strain=KT2440     NaN         latest   \n",
       "GCF_000007605.1                                    NaN    GPIC         latest   \n",
       "GCF_000007625.1       strain=Massachusetts substr. E88     NaN         latest   \n",
       "...                                                ...     ...            ...   \n",
       "GCF_001583415.1                       strain=CCUG21000     NaN         latest   \n",
       "GCF_001584185.1                           strain=Ter91     NaN         latest   \n",
       "GCF_001586165.1                        strain=DSM 2777     NaN         latest   \n",
       "GCF_001586195.1                       strain=DSM 12223     NaN         latest   \n",
       "GCF_001586215.1                            strain=PES1     NaN         latest   \n",
       "GCF_001586255.1                             strain=SM2     NaN         latest   \n",
       "GCF_001593245.1                          strain=TH0426     NaN         latest   \n",
       "GCF_001597285.1                         strain=NEB 585     NaN         latest   \n",
       "GCF_001606005.1                           strain=BS258     NaN         latest   \n",
       "GCF_001606025.1                      strain=PAMC 27889     NaN         latest   \n",
       "GCF_001610975.1                           strain=203-1     NaN         latest   \n",
       "GCF_001611135.1                      strain=TMW 2.1535     NaN         latest   \n",
       "GCF_001618685.1                     strain=KCTC3922(T)     NaN         latest   \n",
       "GCF_001620305.1                            strain=D188     NaN         latest   \n",
       "GCF_001632865.1                       strain=DSM 40697     NaN         latest   \n",
       "GCF_001636015.1                          strain=199/55     NaN         latest   \n",
       "GCF_001641285.1                          strain=ATSB10     NaN         latest   \n",
       "GCF_001642655.1                       strain=KCTC 5387     NaN         latest   \n",
       "GCF_001644565.1                             strain=DY1     NaN         latest   \n",
       "GCF_001648175.1                     strain=IHB B 13600     NaN         latest   \n",
       "GCF_001652465.1                         strain=SIT 101     NaN         latest   \n",
       "GCF_001652485.1                              strain=K8     NaN         latest   \n",
       "GCF_001652565.1                            strain=HZN7     NaN         latest   \n",
       "GCF_001653755.1                           strain=DSW-1     NaN         latest   \n",
       "GCF_001654495.1                            strain=SAT1     NaN         latest   \n",
       "GCF_001655245.1                           strain=95051     NaN         latest   \n",
       "GCF_900011245.1                                    NaN     NaN         latest   \n",
       "GCF_900078695.1                      strain=H044680328     NaN         latest   \n",
       "GCF_900086555.1                       strain=NCTC13585     NaN         latest   \n",
       "GCF_900087055.1                       strain=NCTC10360     NaN         latest   \n",
       "\n",
       "                       ...      suborder_name suborder_taxid subphylum_name  \\\n",
       "assembly_accession     ...                                                    \n",
       "GCF_000005825.2        ...                NaN            NaN            NaN   \n",
       "GCF_000005845.2        ...                NaN            NaN            NaN   \n",
       "GCF_000006605.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006685.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006725.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006745.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006765.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006785.2        ...                NaN            NaN            NaN   \n",
       "GCF_000006845.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006865.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006905.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006925.2        ...                NaN            NaN            NaN   \n",
       "GCF_000006945.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006965.1        ...                NaN            NaN            NaN   \n",
       "GCF_000006985.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007025.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007045.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007085.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007125.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007145.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007265.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007325.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007365.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007385.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007465.2        ...                NaN            NaN            NaN   \n",
       "GCF_000007485.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007525.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007565.2        ...                NaN            NaN            NaN   \n",
       "GCF_000007605.1        ...                NaN            NaN            NaN   \n",
       "GCF_000007625.1        ...                NaN            NaN            NaN   \n",
       "...                    ...                ...            ...            ...   \n",
       "GCF_001583415.1        ...                NaN            NaN            NaN   \n",
       "GCF_001584185.1        ...                NaN            NaN            NaN   \n",
       "GCF_001586165.1        ...                NaN            NaN            NaN   \n",
       "GCF_001586195.1        ...                NaN            NaN            NaN   \n",
       "GCF_001586215.1        ...                NaN            NaN            NaN   \n",
       "GCF_001586255.1        ...                NaN            NaN            NaN   \n",
       "GCF_001593245.1        ...                NaN            NaN            NaN   \n",
       "GCF_001597285.1        ...                NaN            NaN            NaN   \n",
       "GCF_001606005.1        ...                NaN            NaN            NaN   \n",
       "GCF_001606025.1        ...                NaN            NaN            NaN   \n",
       "GCF_001610975.1        ...                NaN            NaN            NaN   \n",
       "GCF_001611135.1        ...                NaN            NaN            NaN   \n",
       "GCF_001618685.1        ...                NaN            NaN            NaN   \n",
       "GCF_001620305.1        ...                NaN            NaN            NaN   \n",
       "GCF_001632865.1        ...                NaN            NaN            NaN   \n",
       "GCF_001636015.1        ...                NaN            NaN            NaN   \n",
       "GCF_001641285.1        ...                NaN            NaN            NaN   \n",
       "GCF_001642655.1        ...                NaN            NaN            NaN   \n",
       "GCF_001644565.1        ...                NaN            NaN            NaN   \n",
       "GCF_001648175.1        ...                NaN            NaN            NaN   \n",
       "GCF_001652465.1        ...                NaN            NaN            NaN   \n",
       "GCF_001652485.1        ...                NaN            NaN            NaN   \n",
       "GCF_001652565.1        ...                NaN            NaN            NaN   \n",
       "GCF_001653755.1        ...                NaN            NaN            NaN   \n",
       "GCF_001654495.1        ...                NaN            NaN            NaN   \n",
       "GCF_001655245.1        ...                NaN            NaN            NaN   \n",
       "GCF_900011245.1        ...                NaN            NaN            NaN   \n",
       "GCF_900078695.1        ...                NaN            NaN            NaN   \n",
       "GCF_900086555.1        ...                NaN            NaN            NaN   \n",
       "GCF_900087055.1        ...                NaN            NaN            NaN   \n",
       "\n",
       "                   subphylum_taxid  \\\n",
       "assembly_accession                   \n",
       "GCF_000005825.2                NaN   \n",
       "GCF_000005845.2                NaN   \n",
       "GCF_000006605.1                NaN   \n",
       "GCF_000006685.1                NaN   \n",
       "GCF_000006725.1                NaN   \n",
       "GCF_000006745.1                NaN   \n",
       "GCF_000006765.1                NaN   \n",
       "GCF_000006785.2                NaN   \n",
       "GCF_000006845.1                NaN   \n",
       "GCF_000006865.1                NaN   \n",
       "GCF_000006905.1                NaN   \n",
       "GCF_000006925.2                NaN   \n",
       "GCF_000006945.1                NaN   \n",
       "GCF_000006965.1                NaN   \n",
       "GCF_000006985.1                NaN   \n",
       "GCF_000007025.1                NaN   \n",
       "GCF_000007045.1                NaN   \n",
       "GCF_000007085.1                NaN   \n",
       "GCF_000007125.1                NaN   \n",
       "GCF_000007145.1                NaN   \n",
       "GCF_000007265.1                NaN   \n",
       "GCF_000007325.1                NaN   \n",
       "GCF_000007365.1                NaN   \n",
       "GCF_000007385.1                NaN   \n",
       "GCF_000007465.2                NaN   \n",
       "GCF_000007485.1                NaN   \n",
       "GCF_000007525.1                NaN   \n",
       "GCF_000007565.2                NaN   \n",
       "GCF_000007605.1                NaN   \n",
       "GCF_000007625.1                NaN   \n",
       "...                            ...   \n",
       "GCF_001583415.1                NaN   \n",
       "GCF_001584185.1                NaN   \n",
       "GCF_001586165.1                NaN   \n",
       "GCF_001586195.1                NaN   \n",
       "GCF_001586215.1                NaN   \n",
       "GCF_001586255.1                NaN   \n",
       "GCF_001593245.1                NaN   \n",
       "GCF_001597285.1                NaN   \n",
       "GCF_001606005.1                NaN   \n",
       "GCF_001606025.1                NaN   \n",
       "GCF_001610975.1                NaN   \n",
       "GCF_001611135.1                NaN   \n",
       "GCF_001618685.1                NaN   \n",
       "GCF_001620305.1                NaN   \n",
       "GCF_001632865.1                NaN   \n",
       "GCF_001636015.1                NaN   \n",
       "GCF_001641285.1                NaN   \n",
       "GCF_001642655.1                NaN   \n",
       "GCF_001644565.1                NaN   \n",
       "GCF_001648175.1                NaN   \n",
       "GCF_001652465.1                NaN   \n",
       "GCF_001652485.1                NaN   \n",
       "GCF_001652565.1                NaN   \n",
       "GCF_001653755.1                NaN   \n",
       "GCF_001654495.1                NaN   \n",
       "GCF_001655245.1                NaN   \n",
       "GCF_900011245.1                NaN   \n",
       "GCF_900078695.1                NaN   \n",
       "GCF_900086555.1                NaN   \n",
       "GCF_900087055.1                NaN   \n",
       "\n",
       "                                                      subspecies_name  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000005825.2                                                   NaN   \n",
       "GCF_000005845.2                                                   NaN   \n",
       "GCF_000006605.1                                                   NaN   \n",
       "GCF_000006685.1                                                   NaN   \n",
       "GCF_000006725.1                                                   NaN   \n",
       "GCF_000006745.1                                                   NaN   \n",
       "GCF_000006765.1                                                   NaN   \n",
       "GCF_000006785.2                                                   NaN   \n",
       "GCF_000006845.1                                                   NaN   \n",
       "GCF_000006865.1                      Lactococcus lactis subsp. lactis   \n",
       "GCF_000006905.1                                                   NaN   \n",
       "GCF_000006925.2                                                   NaN   \n",
       "GCF_000006945.1                   Salmonella enterica subsp. enterica   \n",
       "GCF_000006965.1                                                   NaN   \n",
       "GCF_000006985.1                                                   NaN   \n",
       "GCF_000007025.1                                                   NaN   \n",
       "GCF_000007045.1                                                   NaN   \n",
       "GCF_000007085.1     Caldanaerobacter subterraneus subsp. tengconge...   \n",
       "GCF_000007125.1                                                   NaN   \n",
       "GCF_000007145.1                                                   NaN   \n",
       "GCF_000007265.1                                                   NaN   \n",
       "GCF_000007325.1              Fusobacterium nucleatum subsp. nucleatum   \n",
       "GCF_000007365.1                                                   NaN   \n",
       "GCF_000007385.1                                                   NaN   \n",
       "GCF_000007465.2                                                   NaN   \n",
       "GCF_000007485.1                                                   NaN   \n",
       "GCF_000007525.1                                                   NaN   \n",
       "GCF_000007565.2                                                   NaN   \n",
       "GCF_000007605.1                                                   NaN   \n",
       "GCF_000007625.1                                                   NaN   \n",
       "...                                                               ...   \n",
       "GCF_001583415.1                                                   NaN   \n",
       "GCF_001584185.1                                                   NaN   \n",
       "GCF_001586165.1                                                   NaN   \n",
       "GCF_001586195.1                                                   NaN   \n",
       "GCF_001586215.1                                                   NaN   \n",
       "GCF_001586255.1                                                   NaN   \n",
       "GCF_001593245.1                                                   NaN   \n",
       "GCF_001597285.1                                                   NaN   \n",
       "GCF_001606005.1                                                   NaN   \n",
       "GCF_001606025.1                                                   NaN   \n",
       "GCF_001610975.1                                                   NaN   \n",
       "GCF_001611135.1                                                   NaN   \n",
       "GCF_001618685.1                                                   NaN   \n",
       "GCF_001620305.1                                                   NaN   \n",
       "GCF_001632865.1                                                   NaN   \n",
       "GCF_001636015.1                                                   NaN   \n",
       "GCF_001641285.1                                                   NaN   \n",
       "GCF_001642655.1                                                   NaN   \n",
       "GCF_001644565.1                                                   NaN   \n",
       "GCF_001648175.1                                                   NaN   \n",
       "GCF_001652465.1                                                   NaN   \n",
       "GCF_001652485.1                                                   NaN   \n",
       "GCF_001652565.1                                                   NaN   \n",
       "GCF_001653755.1                                                   NaN   \n",
       "GCF_001654495.1                                                   NaN   \n",
       "GCF_001655245.1                                                   NaN   \n",
       "GCF_900011245.1                                                   NaN   \n",
       "GCF_900078695.1                                                   NaN   \n",
       "GCF_900086555.1                                                   NaN   \n",
       "GCF_900087055.1                                                   NaN   \n",
       "\n",
       "                   subspecies_taxid superkingdom_name superkingdom_taxid  \\\n",
       "assembly_accession                                                         \n",
       "GCF_000005825.2                 NaN          Bacteria                2.0   \n",
       "GCF_000005845.2                 NaN          Bacteria                2.0   \n",
       "GCF_000006605.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006685.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006725.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006745.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006765.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006785.2                 NaN          Bacteria                2.0   \n",
       "GCF_000006845.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006865.1              1360.0          Bacteria                2.0   \n",
       "GCF_000006905.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006925.2                 NaN          Bacteria                2.0   \n",
       "GCF_000006945.1             59201.0          Bacteria                2.0   \n",
       "GCF_000006965.1                 NaN          Bacteria                2.0   \n",
       "GCF_000006985.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007025.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007045.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007085.1            119072.0          Bacteria                2.0   \n",
       "GCF_000007125.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007145.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007265.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007325.1             76856.0          Bacteria                2.0   \n",
       "GCF_000007365.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007385.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007465.2                 NaN          Bacteria                2.0   \n",
       "GCF_000007485.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007525.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007565.2                 NaN          Bacteria                2.0   \n",
       "GCF_000007605.1                 NaN          Bacteria                2.0   \n",
       "GCF_000007625.1                 NaN          Bacteria                2.0   \n",
       "...                             ...               ...                ...   \n",
       "GCF_001583415.1                 NaN          Bacteria                2.0   \n",
       "GCF_001584185.1                 NaN          Bacteria                2.0   \n",
       "GCF_001586165.1                 NaN          Bacteria                2.0   \n",
       "GCF_001586195.1                 NaN          Bacteria                2.0   \n",
       "GCF_001586215.1                 NaN          Bacteria                2.0   \n",
       "GCF_001586255.1                 NaN          Bacteria                2.0   \n",
       "GCF_001593245.1                 NaN          Bacteria                2.0   \n",
       "GCF_001597285.1                 NaN          Bacteria                2.0   \n",
       "GCF_001606005.1                 NaN          Bacteria                2.0   \n",
       "GCF_001606025.1                 NaN          Bacteria                2.0   \n",
       "GCF_001610975.1                 NaN          Bacteria                2.0   \n",
       "GCF_001611135.1                 NaN          Bacteria                2.0   \n",
       "GCF_001618685.1                 NaN          Bacteria                2.0   \n",
       "GCF_001620305.1                 NaN          Bacteria                2.0   \n",
       "GCF_001632865.1                 NaN          Bacteria                2.0   \n",
       "GCF_001636015.1                 NaN          Bacteria                2.0   \n",
       "GCF_001641285.1                 NaN          Bacteria                2.0   \n",
       "GCF_001642655.1                 NaN          Bacteria                2.0   \n",
       "GCF_001644565.1                 NaN          Bacteria                2.0   \n",
       "GCF_001648175.1                 NaN          Bacteria                2.0   \n",
       "GCF_001652465.1                 NaN          Bacteria                2.0   \n",
       "GCF_001652485.1                 NaN          Bacteria                2.0   \n",
       "GCF_001652565.1                 NaN          Bacteria                2.0   \n",
       "GCF_001653755.1                 NaN          Bacteria                2.0   \n",
       "GCF_001654495.1                 NaN          Bacteria                2.0   \n",
       "GCF_001655245.1                 NaN          Bacteria                2.0   \n",
       "GCF_900011245.1                 NaN          Bacteria                2.0   \n",
       "GCF_900078695.1                 NaN          Bacteria                2.0   \n",
       "GCF_900086555.1                 NaN          Bacteria                2.0   \n",
       "GCF_900087055.1                 NaN          Bacteria                2.0   \n",
       "\n",
       "                      tribe_name  tribe_taxid  \n",
       "assembly_accession                             \n",
       "GCF_000005825.2              NaN          NaN  \n",
       "GCF_000005845.2              NaN          NaN  \n",
       "GCF_000006605.1              NaN          NaN  \n",
       "GCF_000006685.1              NaN          NaN  \n",
       "GCF_000006725.1              NaN          NaN  \n",
       "GCF_000006745.1              NaN          NaN  \n",
       "GCF_000006765.1              NaN          NaN  \n",
       "GCF_000006785.2              NaN          NaN  \n",
       "GCF_000006845.1              NaN          NaN  \n",
       "GCF_000006865.1              NaN          NaN  \n",
       "GCF_000006905.1              NaN          NaN  \n",
       "GCF_000006925.2              NaN          NaN  \n",
       "GCF_000006945.1              NaN          NaN  \n",
       "GCF_000006965.1              NaN          NaN  \n",
       "GCF_000006985.1              NaN          NaN  \n",
       "GCF_000007025.1     Rickettsieae      33988.0  \n",
       "GCF_000007045.1              NaN          NaN  \n",
       "GCF_000007085.1              NaN          NaN  \n",
       "GCF_000007125.1              NaN          NaN  \n",
       "GCF_000007145.1              NaN          NaN  \n",
       "GCF_000007265.1              NaN          NaN  \n",
       "GCF_000007325.1              NaN          NaN  \n",
       "GCF_000007365.1              NaN          NaN  \n",
       "GCF_000007385.1              NaN          NaN  \n",
       "GCF_000007465.2              NaN          NaN  \n",
       "GCF_000007485.1              NaN          NaN  \n",
       "GCF_000007525.1              NaN          NaN  \n",
       "GCF_000007565.2              NaN          NaN  \n",
       "GCF_000007605.1              NaN          NaN  \n",
       "GCF_000007625.1              NaN          NaN  \n",
       "...                          ...          ...  \n",
       "GCF_001583415.1              NaN          NaN  \n",
       "GCF_001584185.1              NaN          NaN  \n",
       "GCF_001586165.1              NaN          NaN  \n",
       "GCF_001586195.1              NaN          NaN  \n",
       "GCF_001586215.1              NaN          NaN  \n",
       "GCF_001586255.1              NaN          NaN  \n",
       "GCF_001593245.1              NaN          NaN  \n",
       "GCF_001597285.1              NaN          NaN  \n",
       "GCF_001606005.1              NaN          NaN  \n",
       "GCF_001606025.1              NaN          NaN  \n",
       "GCF_001610975.1              NaN          NaN  \n",
       "GCF_001611135.1              NaN          NaN  \n",
       "GCF_001618685.1              NaN          NaN  \n",
       "GCF_001620305.1              NaN          NaN  \n",
       "GCF_001632865.1              NaN          NaN  \n",
       "GCF_001636015.1              NaN          NaN  \n",
       "GCF_001641285.1              NaN          NaN  \n",
       "GCF_001642655.1              NaN          NaN  \n",
       "GCF_001644565.1              NaN          NaN  \n",
       "GCF_001648175.1              NaN          NaN  \n",
       "GCF_001652465.1              NaN          NaN  \n",
       "GCF_001652485.1              NaN          NaN  \n",
       "GCF_001652565.1              NaN          NaN  \n",
       "GCF_001653755.1              NaN          NaN  \n",
       "GCF_001654495.1              NaN          NaN  \n",
       "GCF_001655245.1              NaN          NaN  \n",
       "GCF_900011245.1              NaN          NaN  \n",
       "GCF_900078695.1              NaN          NaN  \n",
       "GCF_900086555.1              NaN          NaN  \n",
       "GCF_900087055.1              NaN          NaN  \n",
       "\n",
       "[1576 rows x 53 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblySummaryRepTaxonomyDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1582</td>\n",
       "      <td>949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3</td>\n",
       "      <td>class</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>292</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank_index     rank_name  taxon_id           taxon_name  \\\n",
       "0            0  superkingdom       2.0             Bacteria   \n",
       "4            1        phylum    1224.0       Proteobacteria   \n",
       "5            1        phylum    1239.0           Firmicutes   \n",
       "32           3         class    1236.0  Gammaproteobacteria   \n",
       "23           1        phylum  201174.0       Actinobacteria   \n",
       "\n",
       "    n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "0                  1582                                        949  \n",
       "4                   697                                        407  \n",
       "5                   259                                        170  \n",
       "32                  292                                        162  \n",
       "23                  226                                        126  "
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf = pd.read_csv(str(taxonomyPath / 'taxaDf.csv'), index_col=0)\n",
    "taxaDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>697</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>259</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>226</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>976.0</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>108</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1117.0</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>62</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>544448.0</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>54</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>203691.0</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>Deinococcus-Thermus</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>200918.0</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>Chlorobi</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>204428.0</td>\n",
       "      <td>Chlamydiae</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>200795.0</td>\n",
       "      <td>Chloroflexi</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>200783.0</td>\n",
       "      <td>Aquificae</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>57723.0</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>32066.0</td>\n",
       "      <td>Fusobacteria</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>203682.0</td>\n",
       "      <td>Planctomycetes</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>74201.0</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>200930.0</td>\n",
       "      <td>Deferribacteres</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>508458.0</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>200940.0</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>40117.0</td>\n",
       "      <td>Nitrospirae</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>68297.0</td>\n",
       "      <td>Dictyoglomi</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>74152.0</td>\n",
       "      <td>Elusimicrobia</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>200938.0</td>\n",
       "      <td>Chrysiogenetes</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>142182.0</td>\n",
       "      <td>Gemmatimonadetes</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>65842.0</td>\n",
       "      <td>Fibrobacteres</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>67814.0</td>\n",
       "      <td>Caldiserica</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>67819.0</td>\n",
       "      <td>Armatimonadetes</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>456828.0</td>\n",
       "      <td>Candidatus Cloacimonetes</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1134404.0</td>\n",
       "      <td>Ignavibacteriae</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank_index rank_name   taxon_id                taxon_name  \\\n",
       "4            1    phylum     1224.0            Proteobacteria   \n",
       "5            1    phylum     1239.0                Firmicutes   \n",
       "23           1    phylum   201174.0            Actinobacteria   \n",
       "1            1    phylum      976.0             Bacteroidetes   \n",
       "3            1    phylum     1117.0             Cyanobacteria   \n",
       "29           1    phylum   544448.0               Tenericutes   \n",
       "25           1    phylum   203691.0              Spirochaetes   \n",
       "6            1    phylum     1297.0       Deinococcus-Thermus   \n",
       "19           1    phylum   200918.0               Thermotogae   \n",
       "2            1    phylum     1090.0                  Chlorobi   \n",
       "26           1    phylum   204428.0                Chlamydiae   \n",
       "18           1    phylum   200795.0               Chloroflexi   \n",
       "17           1    phylum   200783.0                 Aquificae   \n",
       "9            1    phylum    57723.0             Acidobacteria   \n",
       "7            1    phylum    32066.0              Fusobacteria   \n",
       "24           1    phylum   203682.0            Planctomycetes   \n",
       "15           1    phylum    74201.0           Verrucomicrobia   \n",
       "20           1    phylum   200930.0           Deferribacteres   \n",
       "28           1    phylum   508458.0             Synergistetes   \n",
       "22           1    phylum   200940.0     Thermodesulfobacteria   \n",
       "8            1    phylum    40117.0               Nitrospirae   \n",
       "13           1    phylum    68297.0               Dictyoglomi   \n",
       "14           1    phylum    74152.0             Elusimicrobia   \n",
       "21           1    phylum   200938.0            Chrysiogenetes   \n",
       "16           1    phylum   142182.0          Gemmatimonadetes   \n",
       "10           1    phylum    65842.0             Fibrobacteres   \n",
       "11           1    phylum    67814.0               Caldiserica   \n",
       "12           1    phylum    67819.0           Armatimonadetes   \n",
       "27           1    phylum   456828.0  Candidatus Cloacimonetes   \n",
       "30           1    phylum  1134404.0           Ignavibacteriae   \n",
       "\n",
       "    n_genomes_in_refseq  n_genomes_in_refseq_eggnog_common_species  \n",
       "4                   697                                        407  \n",
       "5                   259                                        170  \n",
       "23                  226                                        126  \n",
       "1                   108                                         59  \n",
       "3                    62                                         31  \n",
       "29                   54                                         27  \n",
       "25                   30                                         23  \n",
       "6                    20                                         13  \n",
       "19                   18                                         12  \n",
       "2                    11                                         11  \n",
       "26                   14                                         11  \n",
       "18                   11                                         10  \n",
       "17                    9                                          8  \n",
       "9                     8                                          7  \n",
       "7                     9                                          7  \n",
       "24                    7                                          5  \n",
       "15                    8                                          4  \n",
       "20                    4                                          4  \n",
       "28                    4                                          3  \n",
       "22                    3                                          2  \n",
       "8                     6                                          2  \n",
       "13                    2                                          2  \n",
       "14                    2                                          1  \n",
       "21                    1                                          1  \n",
       "16                    1                                          1  \n",
       "10                    1                                          1  \n",
       "11                    1                                          0  \n",
       "12                    1                                          0  \n",
       "27                    1                                          0  \n",
       "30                    2                                          0  "
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf[taxaDf['rank_name'] == 'phylum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1224.0,\n",
       " 1239.0,\n",
       " 201174.0,\n",
       " 976.0,\n",
       " 1117.0,\n",
       " 544448.0,\n",
       " 203691.0,\n",
       " 1297.0,\n",
       " 200918.0,\n",
       " 1090.0,\n",
       " 204428.0,\n",
       " 200795.0,\n",
       " 200783.0,\n",
       " 57723.0,\n",
       " 32066.0,\n",
       " 203682.0,\n",
       " 74201.0,\n",
       " 200930.0,\n",
       " 508458.0,\n",
       " 200940.0,\n",
       " 40117.0,\n",
       " 68297.0,\n",
       " 74152.0,\n",
       " 200938.0,\n",
       " 142182.0,\n",
       " 65842.0,\n",
       " 67814.0,\n",
       " 67819.0,\n",
       " 456828.0,\n",
       " 1134404.0]"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylaList = list(taxaDf[taxaDf['rank_name'] == 'phylum']['taxon_id'])\n",
    "list(phylaList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxaDf[taxaDf['n_genomes_in_refseq'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_id'] == 544448].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Enterobacterales'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Gammaproteobacteria'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Alteromonas'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rank_index                                             1\n",
       "rank_name                                         phylum\n",
       "taxon_id                                          544448\n",
       "taxon_name                                   Tenericutes\n",
       "n_genomes_in_refseq                                   54\n",
       "n_genomes_in_refseq_eggnog_common_species             27\n",
       "Name: 29, dtype: object"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Tenericutes'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Firmicutes'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Mycoplasma'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Ureaplasma'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rank_index                                            3\n",
       "rank_name                                         class\n",
       "taxon_id                                          31969\n",
       "taxon_name                                   Mollicutes\n",
       "n_genomes_in_refseq                                  53\n",
       "n_genomes_in_refseq_eggnog_common_species            27\n",
       "Name: 38, dtype: object"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Mollicutes'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rank_index                                                      14\n",
       "rank_name                                                  species\n",
       "taxon_id                                                      2104\n",
       "taxon_name                                   Mycoplasma pneumoniae\n",
       "n_genomes_in_refseq                                              1\n",
       "n_genomes_in_refseq_eggnog_common_species                        1\n",
       "Name: 1693, dtype: object"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Mycoplasma pneumoniae'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Mesoplasma'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Staphylococcus aureus'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Bacillus anthracis'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Lactococcus lactis'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rank_index                                              0\n",
       "rank_name                                    superkingdom\n",
       "taxon_id                                                2\n",
       "taxon_name                                       Bacteria\n",
       "n_genomes_in_refseq                                  1582\n",
       "n_genomes_in_refseq_eggnog_common_species             949\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Bacteria'].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxaDf[taxaDf['rank_name'] == 'species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_id'] == 1117].iloc[0]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Searching for all organisms that appear in the list of @Prat2012 Fig. 1A bacterial species that encode the Pyl decoding genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['family_name'] == 'Desulfovibrionaceae']['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'].str.contains('Bilophila', re.I)]\n",
    "taxonGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Bilophila', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Thermincola', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Desulfosporosinus orientis', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Desulfosporosinus meridiei', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Desulfosporosinus yougiae', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Desulfitobacterium hafniense', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Desulfitobacterium dehalogenans', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Desulfitobacterium acetoxidans', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Desulfitobacterium gibsoniae', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Acetohalobium arabaticum', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanococcoides burtonii', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanohalophilus mahii', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanohalobium evestigatum', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanosalsum zhilinae', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanosarcina thermophila', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanosarcina acetivorans', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanosarcina mazei', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf['species_name'].str.contains('Methanosarcina barkeri', re.I)]['organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxonGroupRef: rank_index                                              0\n",
      "rank_name                                    superkingdom\n",
      "taxon_id                                                2\n",
      "taxon_name                                       Bacteria\n",
      "n_genomes_in_refseq                                  1582\n",
      "n_genomes_in_refseq_eggnog_common_species             949\n",
      "Name: 0, dtype: object\n",
      "output path: /users/lserrano/mweber/Analysis_data/C-terminal/Data/taxaId_2\n"
     ]
    }
   ],
   "source": [
    "print(\"taxonGroupRef:\", taxonGroup)\n",
    "taxonRank = taxonGroup['rank_name']\n",
    "taxonName = taxonGroup['taxon_name']\n",
    "taxonId = int(taxonGroup['taxon_id'])\n",
    "\n",
    "# Select genomes in the taxon group\n",
    "taxonAssemblyDf = assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf[taxonRank + '_taxid'] == taxonId]\n",
    "taxonAssemblyAccessionList = list(taxonAssemblyDf.index)\n",
    "\n",
    "taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "print(\"output path:\", str(analysisCtermDataPath / taxonSuffix))\n",
    "suffix = taxonSuffix\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## import multispeciesAllCDSDf_nr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the multispeciesAllCDSDf_nr dataframe\n",
    "filePath = analysisCtermDataPath / 'Taxonomy' / taxonSuffix / (taxonSuffix + '_multispeciesAllCDSDf_nr.csv')\n",
    "print(filePath)\n",
    "if filePath.exists():\n",
    "    multispeciesAllCDSDf_nr = pd.read_csv(str(filePath))\n",
    "else:\n",
    "    print(\"file does not exists.\")\n",
    "\n",
    "len(multispeciesAllCDSDf_nr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## excluding overlapping start/stop codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multispeciesAllCDSDf_nr[['genome_accession', 'refSeqProteinId']].duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3249: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nb of genomes in RefSeq database: 75064\n",
      "Nb of genomes in RefSeq database, representative or reference genomes: 6165\n",
      "Nb of genomes in RefSeq database, after filtering out contigs: 1773\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>...</th>\n",
       "      <th>release_type</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>compressedGenomeFilename</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000002515.2</th>\n",
       "      <td>PRJNA12377</td>\n",
       "      <td>SAMEA3138170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>28985</td>\n",
       "      <td>28985</td>\n",
       "      <td>Kluyveromyces lactis</td>\n",
       "      <td>strain=NRRL Y-1140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2004/07/02</td>\n",
       "      <td>ASM251v1</td>\n",
       "      <td>Genolevures Consortium</td>\n",
       "      <td>GCA_000002515.1</td>\n",
       "      <td>different</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000002515.2_ASM251v1_genomic.gbff.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000002725.2</th>\n",
       "      <td>PRJNA15564</td>\n",
       "      <td>SAMEA3138173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>347515</td>\n",
       "      <td>5664</td>\n",
       "      <td>Leishmania major strain Friedlin</td>\n",
       "      <td>strain=Friedlin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2011/02/14</td>\n",
       "      <td>ASM272v2</td>\n",
       "      <td>Friedlin Consortium</td>\n",
       "      <td>GCA_000002725.2</td>\n",
       "      <td>identical</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000002725.2_ASM272v2_genomic.gbff.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000002985.6</th>\n",
       "      <td>PRJNA158</td>\n",
       "      <td>SAMEA3138177</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>6239</td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>strain=Bristol N2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Major</td>\n",
       "      <td>Full</td>\n",
       "      <td>2013/02/07</td>\n",
       "      <td>WBcel235</td>\n",
       "      <td>C. elegans Sequencing Consortium</td>\n",
       "      <td>GCA_000002985.3</td>\n",
       "      <td>different</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000002985.6_WBcel235_genomic.gbff.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    bioproject     biosample wgs_master  \\\n",
       "assembly_accession                                        \n",
       "GCF_000002515.2     PRJNA12377  SAMEA3138170        NaN   \n",
       "GCF_000002725.2     PRJNA15564  SAMEA3138173        NaN   \n",
       "GCF_000002985.6       PRJNA158  SAMEA3138177        NaN   \n",
       "\n",
       "                          refseq_category   taxid  species_taxid  \\\n",
       "assembly_accession                                                 \n",
       "GCF_000002515.2     representative genome   28985          28985   \n",
       "GCF_000002725.2     representative genome  347515           5664   \n",
       "GCF_000002985.6          reference genome    6239           6239   \n",
       "\n",
       "                                       organism_name  infraspecific_name  \\\n",
       "assembly_accession                                                         \n",
       "GCF_000002515.2                 Kluyveromyces lactis  strain=NRRL Y-1140   \n",
       "GCF_000002725.2     Leishmania major strain Friedlin     strain=Friedlin   \n",
       "GCF_000002985.6               Caenorhabditis elegans   strain=Bristol N2   \n",
       "\n",
       "                   isolate version_status  \\\n",
       "assembly_accession                          \n",
       "GCF_000002515.2        NaN         latest   \n",
       "GCF_000002725.2        NaN         latest   \n",
       "GCF_000002985.6        NaN         latest   \n",
       "\n",
       "                                      ...                    release_type  \\\n",
       "assembly_accession                    ...                                   \n",
       "GCF_000002515.2                       ...                           Major   \n",
       "GCF_000002725.2                       ...                           Major   \n",
       "GCF_000002985.6                       ...                           Major   \n",
       "\n",
       "                   genome_rep seq_rel_date  asm_name  \\\n",
       "assembly_accession                                     \n",
       "GCF_000002515.2          Full   2004/07/02  ASM251v1   \n",
       "GCF_000002725.2          Full   2011/02/14  ASM272v2   \n",
       "GCF_000002985.6          Full   2013/02/07  WBcel235   \n",
       "\n",
       "                                           submitter  gbrs_paired_asm  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000002515.2               Genolevures Consortium  GCA_000002515.1   \n",
       "GCF_000002725.2                  Friedlin Consortium  GCA_000002725.2   \n",
       "GCF_000002985.6     C. elegans Sequencing Consortium  GCA_000002985.3   \n",
       "\n",
       "                   paired_asm_comp  \\\n",
       "assembly_accession                   \n",
       "GCF_000002515.2          different   \n",
       "GCF_000002725.2          identical   \n",
       "GCF_000002985.6          different   \n",
       "\n",
       "                                                             ftp_path  \\\n",
       "assembly_accession                                                      \n",
       "GCF_000002515.2     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...   \n",
       "GCF_000002725.2     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...   \n",
       "GCF_000002985.6     ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...   \n",
       "\n",
       "                   excluded_from_refseq  \\\n",
       "assembly_accession                        \n",
       "GCF_000002515.2                     NaN   \n",
       "GCF_000002725.2                     NaN   \n",
       "GCF_000002985.6                     NaN   \n",
       "\n",
       "                                    compressedGenomeFilename  \n",
       "assembly_accession                                            \n",
       "GCF_000002515.2     GCF_000002515.2_ASM251v1_genomic.gbff.gz  \n",
       "GCF_000002725.2     GCF_000002725.2_ASM272v2_genomic.gbff.gz  \n",
       "GCF_000002985.6     GCF_000002985.6_WBcel235_genomic.gbff.gz  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblySummaryRepDf = read_assemblySummary_file(str(refSeqPath / 'assembly_summary_refseq.txt'))\n",
    "assemblySummaryRepDf.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000002515.2\n",
      "GCF_000002725.2\n",
      "GCF_000002985.6\n",
      "GCF_000005825.2\n",
      "GCF_000005845.2\n",
      "GCF_000006175.1\n",
      "GCF_000006605.1\n",
      "GCF_000006685.1\n",
      "GCF_000006725.1\n",
      "GCF_000006745.1\n",
      "GCF_000006765.1\n",
      "GCF_000006785.2\n",
      "GCF_000006845.1\n",
      "GCF_000006865.1\n",
      "GCF_000006905.1\n",
      "GCF_000006925.2\n",
      "GCF_000006945.1\n",
      "GCF_000006965.1\n",
      "GCF_000006985.1\n",
      "GCF_000007005.1\n",
      "GCF_000007025.1\n",
      "GCF_000007045.1\n",
      "GCF_000007085.1\n",
      "GCF_000007125.1\n",
      "GCF_000007145.1\n",
      "GCF_000007265.1\n",
      "GCF_000007305.1\n",
      "GCF_000007325.1\n",
      "GCF_000007345.1\n",
      "GCF_000007365.1\n",
      "GCF_000007385.1\n",
      "GCF_000007465.2\n",
      "GCF_000007485.1\n",
      "GCF_000007525.1\n",
      "GCF_000007565.2\n",
      "GCF_000007605.1\n",
      "GCF_000007625.1\n",
      "GCF_000007645.1\n",
      "GCF_000007705.1\n",
      "GCF_000007725.1\n",
      "GCF_000007745.1\n",
      "GCF_000007765.1\n",
      "GCF_000007785.1\n",
      "GCF_000007805.1\n",
      "GCF_000007825.1\n",
      "GCF_000007845.1\n",
      "GCF_000007865.1\n",
      "GCF_000007905.1\n",
      "GCF_000007925.1\n",
      "GCF_000007945.1\n",
      "GCF_000007985.2\n",
      "GCF_000008025.1\n",
      "GCF_000008045.1\n",
      "GCF_000008165.1\n",
      "GCF_000008185.1\n",
      "GCF_000008205.1\n",
      "GCF_000008265.1\n",
      "GCF_000008305.1\n",
      "GCF_000008325.1\n",
      "GCF_000008345.1\n",
      "GCF_000008365.1\n",
      "GCF_000008385.1\n",
      "GCF_000008465.1\n",
      "GCF_000008485.1\n",
      "GCF_000008505.1\n",
      "GCF_000008525.1\n",
      "GCF_000008545.1\n",
      "GCF_000008565.1\n",
      "GCF_000008605.1\n",
      "GCF_000008625.1\n",
      "GCF_000008645.1\n",
      "GCF_000008665.1\n",
      "GCF_000008685.2\n",
      "GCF_000008725.1\n",
      "GCF_000008745.1\n",
      "GCF_000008765.1\n",
      "GCF_000008805.1\n",
      "GCF_000008865.1\n",
      "GCF_000008885.1\n",
      "GCF_000008925.1\n",
      "GCF_000008985.1\n",
      "GCF_000009045.1\n",
      "GCF_000009065.1\n",
      "GCF_000009085.1\n",
      "GCF_000009125.1\n",
      "GCF_000009145.1\n",
      "GCF_000009205.1\n",
      "GCF_000009225.2\n",
      "GCF_000009305.1\n",
      "GCF_000009345.1\n",
      "GCF_000009365.1\n",
      "GCF_000009405.1\n",
      "GCF_000009605.1\n",
      "GCF_000009625.1\n",
      "GCF_000009705.1\n",
      "GCF_000009725.1\n",
      "GCF_000009745.1\n",
      "GCF_000009765.2\n",
      "GCF_000009785.1\n",
      "GCF_000009825.1\n",
      "GCF_000009845.1\n",
      "GCF_000009865.1\n",
      "GCF_000009905.1\n",
      "GCF_000009925.1\n",
      "GCF_000009945.1\n",
      "GCF_000009965.1\n",
      "GCF_000009985.1\n",
      "GCF_000010065.1\n",
      "GCF_000010085.1\n",
      "GCF_000010105.1\n",
      "GCF_000010125.1\n",
      "GCF_000010145.1\n",
      "GCF_000010165.1\n",
      "GCF_000010185.1\n",
      "GCF_000010285.1\n",
      "GCF_000010305.1\n",
      "GCF_000010325.1\n",
      "GCF_000010405.1\n",
      "GCF_000010505.1\n",
      "GCF_000010525.1\n",
      "GCF_000010585.1\n",
      "GCF_000010605.1\n",
      "GCF_000010625.1\n",
      "GCF_000010645.1\n",
      "GCF_000010665.1\n",
      "GCF_000010785.1\n",
      "GCF_000010825.1\n",
      "GCF_000010985.1\n",
      "GCF_000011025.1\n",
      "GCF_000011065.1\n",
      "GCF_000011105.1\n",
      "GCF_000011145.1\n",
      "GCF_000011185.1\n",
      "GCF_000011205.1\n",
      "GCF_000011225.1\n",
      "GCF_000011245.1\n",
      "GCF_000011305.1\n",
      "GCF_000011325.1\n",
      "GCF_000011345.1\n",
      "GCF_000011365.1\n",
      "GCF_000011385.1\n",
      "GCF_000011445.1\n",
      "GCF_000011465.1\n",
      "GCF_000011485.1\n",
      "GCF_000011545.1\n",
      "GCF_000011645.1\n",
      "GCF_000011685.1\n",
      "GCF_000011705.1\n",
      "GCF_000011745.1\n",
      "GCF_000011805.1\n",
      "GCF_000011825.1\n",
      "GCF_000011905.1\n",
      "GCF_000011965.2\n",
      "GCF_000011985.1\n",
      "GCF_000012005.1\n",
      "GCF_000012145.1\n",
      "GCF_000012225.1\n",
      "GCF_000012245.1\n",
      "GCF_000012285.1\n",
      "GCF_000012305.1\n",
      "GCF_000012325.1\n",
      "GCF_000012345.1\n",
      "GCF_000012385.1\n",
      "GCF_000012405.1\n",
      "GCF_000012425.1\n",
      "GCF_000012445.1\n",
      "GCF_000012465.1\n",
      "GCF_000012485.1\n",
      "GCF_000012505.1\n",
      "GCF_000012545.1\n",
      "GCF_000012565.1\n",
      "GCF_000012585.1\n",
      "GCF_000012605.1\n",
      "GCF_000012645.1\n",
      "GCF_000012665.1\n",
      "GCF_000012685.1\n",
      "GCF_000012725.1\n",
      "GCF_000012745.1\n",
      "GCF_000012765.1\n",
      "GCF_000012805.1\n",
      "GCF_000012825.1\n",
      "GCF_000012845.1\n",
      "GCF_000012865.1\n",
      "GCF_000012885.1\n",
      "GCF_000012925.1\n",
      "GCF_000012965.1\n",
      "GCF_000012985.1\n",
      "GCF_000013005.1\n",
      "GCF_000013025.1\n",
      "GCF_000013045.1\n",
      "GCF_000013085.1\n",
      "GCF_000013105.1\n",
      "GCF_000013125.1\n",
      "GCF_000013145.1\n",
      "GCF_000013165.1\n",
      "GCF_000013185.1\n",
      "GCF_000013205.1\n",
      "GCF_000013225.1\n",
      "GCF_000013285.1\n",
      "GCF_000013325.1\n",
      "GCF_000013345.1\n",
      "GCF_000013365.1\n",
      "GCF_000013405.1\n",
      "GCF_000013425.1\n",
      "GCF_000013445.1\n",
      "GCF_000013565.1\n",
      "GCF_000013605.1\n",
      "GCF_000013625.1\n",
      "GCF_000013645.1\n",
      "GCF_000013665.1\n",
      "GCF_000013705.1\n",
      "GCF_000013725.1\n",
      "GCF_000013745.1\n",
      "GCF_000013765.1\n",
      "GCF_000013865.1\n",
      "GCF_000013885.1\n",
      "GCF_000013945.1\n",
      "GCF_000013985.1\n",
      "GCF_000014005.1\n",
      "GCF_000014025.1\n",
      "GCF_000014045.1\n",
      "GCF_000014125.1\n",
      "GCF_000014145.1\n",
      "GCF_000014185.1\n",
      "GCF_000014205.1\n",
      "GCF_000014225.1\n",
      "GCF_000014245.1\n",
      "GCF_000014265.1\n",
      "GCF_000014285.1\n",
      "GCF_000014345.1\n",
      "GCF_000014385.1\n",
      "GCF_000014425.1\n",
      "GCF_000014445.1\n",
      "GCF_000014465.1\n",
      "GCF_000014505.1\n",
      "GCF_000014525.1\n",
      "GCF_000014565.1\n",
      "GCF_000014585.1\n",
      "GCF_000014705.1\n",
      "GCF_000014725.1\n",
      "GCF_000014745.1\n",
      "GCF_000014765.1\n",
      "GCF_000014785.1\n",
      "GCF_000014805.1\n",
      "GCF_000014825.1\n",
      "GCF_000014865.1\n",
      "GCF_000014885.1\n",
      "GCF_000014905.1\n",
      "GCF_000014925.1\n",
      "GCF_000014965.1\n",
      "GCF_000015005.1\n",
      "GCF_000015025.1\n",
      "GCF_000015045.1\n",
      "GCF_000015105.1\n",
      "GCF_000015125.1\n",
      "GCF_000015145.1\n",
      "GCF_000015205.1\n",
      "GCF_000015225.1\n",
      "GCF_000015245.1\n",
      "GCF_000015265.1\n",
      "GCF_000015285.1\n",
      "GCF_000015305.1\n",
      "GCF_000015345.1\n",
      "GCF_000015405.1\n",
      "GCF_000015445.1\n",
      "GCF_000015505.1\n",
      "GCF_000015565.1\n",
      "GCF_000015585.1\n",
      "GCF_000015665.1\n",
      "GCF_000015725.1\n",
      "GCF_000015745.1\n",
      "GCF_000015765.1\n",
      "GCF_000015825.1\n",
      "GCF_000015865.1\n",
      "GCF_000015885.1\n",
      "GCF_000015945.1\n",
      "GCF_000016065.1\n",
      "GCF_000016085.1\n",
      "GCF_000016165.1\n",
      "GCF_000016185.1\n",
      "GCF_000016265.1\n",
      "GCF_000016285.1\n",
      "GCF_000016325.1\n",
      "GCF_000016345.1\n",
      "GCF_000016405.1\n",
      "GCF_000016425.1\n",
      "GCF_000016505.1\n",
      "GCF_000016525.1\n",
      "GCF_000016545.1\n",
      "GCF_000016565.1\n",
      "GCF_000016605.1\n",
      "GCF_000016625.1\n",
      "GCF_000016645.1\n",
      "GCF_000016665.1\n",
      "GCF_000016745.1\n",
      "GCF_000016765.1\n",
      "GCF_000016785.1\n",
      "GCF_000016825.1\n",
      "GCF_000016905.1\n",
      "GCF_000016965.1\n",
      "GCF_000016985.1\n",
      "GCF_000017005.1\n",
      "GCF_000017045.1\n",
      "GCF_000017145.1\n",
      "GCF_000017165.1\n",
      "GCF_000017185.1\n",
      "GCF_000017225.1\n",
      "GCF_000017245.1\n",
      "GCF_000017265.1\n",
      "GCF_000017285.1\n",
      "GCF_000017305.1\n",
      "GCF_000017405.1\n",
      "GCF_000017425.1\n",
      "GCF_000017465.2\n",
      "GCF_000017505.1\n",
      "GCF_000017545.1\n",
      "GCF_000017565.1\n",
      "GCF_000017585.1\n",
      "GCF_000017625.1\n",
      "GCF_000017645.1\n",
      "GCF_000017685.1\n",
      "GCF_000017705.1\n",
      "GCF_000017805.1\n",
      "GCF_000017845.1\n",
      "GCF_000017865.1\n",
      "GCF_000017945.1\n",
      "GCF_000018005.1\n",
      "GCF_000018025.1\n",
      "GCF_000018065.1\n",
      "GCF_000018105.1\n",
      "GCF_000018145.1\n",
      "GCF_000018205.1\n",
      "GCF_000018265.1\n",
      "GCF_000018285.1\n",
      "GCF_000018305.1\n",
      "GCF_000018325.1\n",
      "GCF_000018345.1\n",
      "GCF_000018365.1\n",
      "GCF_000018405.1\n",
      "GCF_000018425.1\n",
      "GCF_000018545.1\n",
      "GCF_000018605.1\n",
      "GCF_000018665.1\n",
      "GCF_000018685.1\n",
      "GCF_000018785.1\n",
      "GCF_000018865.1\n",
      "GCF_000018885.1\n",
      "GCF_000018945.1\n",
      "GCF_000019145.1\n",
      "GCF_000019165.1\n",
      "GCF_000019205.1\n",
      "GCF_000019225.1\n",
      "GCF_000019285.1\n",
      "GCF_000019345.1\n",
      "GCF_000019365.1\n",
      "GCF_000019405.1\n",
      "GCF_000019485.1\n",
      "GCF_000019525.1\n",
      "GCF_000019605.1\n",
      "GCF_000019665.1\n",
      "GCF_000019725.1\n",
      "GCF_000019745.1\n",
      "GCF_000019785.1\n",
      "GCF_000019845.1\n",
      "GCF_000019905.1\n",
      "GCF_000019945.1\n",
      "GCF_000019965.1\n",
      "GCF_000020005.1\n",
      "GCF_000020025.1\n",
      "GCF_000020045.1\n",
      "GCF_000020065.1\n",
      "GCF_000020145.1\n",
      "GCF_000020165.1\n",
      "GCF_000020205.1\n",
      "GCF_000020225.1\n",
      "GCF_000020305.1\n",
      "GCF_000020365.1\n",
      "GCF_000020385.1\n",
      "GCF_000020465.1\n",
      "GCF_000020485.1\n",
      "GCF_000020505.1\n",
      "GCF_000020525.1\n",
      "GCF_000020545.1\n",
      "GCF_000020565.1\n",
      "GCF_000020605.1\n",
      "GCF_000020625.1\n",
      "GCF_000020645.1\n",
      "GCF_000020685.1\n",
      "GCF_000020725.1\n",
      "GCF_000020785.1\n",
      "GCF_000020905.1\n",
      "GCF_000020945.1\n",
      "GCF_000020965.1\n",
      "GCF_000020985.1\n",
      "GCF_000021025.1\n",
      "GCF_000021045.1\n",
      "GCF_000021285.1\n",
      "GCF_000021325.1\n",
      "GCF_000021385.1\n",
      "GCF_000021485.1\n",
      "GCF_000021545.1\n",
      "GCF_000021565.1\n",
      "GCF_000021645.1\n",
      "GCF_000021685.1\n",
      "GCF_000021725.1\n",
      "GCF_000021745.1\n",
      "GCF_000021805.1\n",
      "GCF_000021825.1\n",
      "GCF_000021865.1\n",
      "GCF_000021885.1\n",
      "GCF_000021905.1\n",
      "GCF_000021925.1\n",
      "GCF_000021945.1\n",
      "GCF_000021965.1\n",
      "GCF_000021985.1\n",
      "GCF_000022005.1\n",
      "GCF_000022025.1\n",
      "GCF_000022045.1\n",
      "GCF_000022065.1\n",
      "GCF_000022085.1\n",
      "GCF_000022105.1\n",
      "GCF_000022125.1\n",
      "GCF_000022145.1\n",
      "GCF_000022205.1\n",
      "GCF_000022265.1\n",
      "GCF_000022305.1\n",
      "GCF_000022325.1\n",
      "GCF_000022365.1\n",
      "GCF_000022385.1\n",
      "GCF_000022525.1\n",
      "GCF_000022545.1\n",
      "GCF_000022565.1\n",
      "GCF_000022605.2\n",
      "GCF_000022685.1\n",
      "GCF_000022725.1\n",
      "GCF_000022885.2\n",
      "GCF_000022905.1\n",
      "GCF_000022965.1\n",
      "GCF_000023025.1\n",
      "GCF_000023065.1\n",
      "GCF_000023105.1\n",
      "GCF_000023125.1\n",
      "GCF_000023145.1\n",
      "GCF_000023205.1\n",
      "GCF_000023225.1\n",
      "GCF_000023245.1\n",
      "GCF_000023265.1\n",
      "GCF_000023285.1\n",
      "GCF_000023325.1\n",
      "GCF_000023405.1\n",
      "GCF_000023445.1\n",
      "GCF_000023465.1\n",
      "GCF_000023585.1\n",
      "GCF_000023605.1\n",
      "GCF_000023705.1\n",
      "GCF_000023725.1\n",
      "GCF_000023745.1\n",
      "GCF_000023785.1\n",
      "GCF_000023825.1\n",
      "GCF_000023845.1\n",
      "GCF_000023865.1\n",
      "GCF_000023885.1\n",
      "GCF_000023905.1\n",
      "GCF_000023925.1\n",
      "GCF_000023945.1\n",
      "GCF_000023965.1\n",
      "GCF_000023985.1\n",
      "GCF_000024005.1\n",
      "GCF_000024025.1\n",
      "GCF_000024065.1\n",
      "GCF_000024085.1\n",
      "GCF_000024105.1\n",
      "GCF_000024125.1\n",
      "GCF_000024165.1\n",
      "GCF_000024185.1\n",
      "GCF_000024205.1\n",
      "GCF_000024225.1\n",
      "GCF_000024265.1\n",
      "GCF_000024285.1\n",
      "GCF_000024325.1\n",
      "GCF_000024345.1\n",
      "GCF_000024365.1\n",
      "GCF_000024385.1\n",
      "GCF_000024405.1\n",
      "GCF_000024465.1\n",
      "GCF_000024505.1\n",
      "GCF_000024545.1\n",
      "GCF_000024565.1\n",
      "GCF_000024605.1\n",
      "GCF_000024625.1\n",
      "GCF_000024725.1\n",
      "GCF_000024765.1\n",
      "GCF_000024785.1\n",
      "GCF_000024805.1\n",
      "GCF_000024845.1\n",
      "GCF_000024865.1\n",
      "GCF_000024885.1\n",
      "GCF_000024905.1\n",
      "GCF_000024925.1\n",
      "GCF_000024945.1\n",
      "GCF_000024965.1\n",
      "GCF_000024985.1\n",
      "GCF_000025005.1\n",
      "GCF_000025085.1\n",
      "GCF_000025125.1\n",
      "GCF_000025185.1\n",
      "GCF_000025205.1\n",
      "GCF_000025225.2\n",
      "GCF_000025265.1\n",
      "GCF_000025285.1\n",
      "GCF_000025305.1\n",
      "GCF_000025325.1\n",
      "GCF_000025345.1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000025485.1\n",
      "GCF_000025505.1\n",
      "GCF_000025525.1\n",
      "GCF_000025545.1\n",
      "GCF_000025565.1\n",
      "GCF_000025605.1\n",
      "GCF_000025625.1\n",
      "GCF_000025645.1\n",
      "GCF_000025665.1\n",
      "GCF_000025685.1\n",
      "GCF_000025705.1\n",
      "GCF_000025725.1\n",
      "GCF_000025845.1\n",
      "GCF_000025865.1\n",
      "GCF_000025885.1\n",
      "GCF_000025905.1\n",
      "GCF_000025925.1\n",
      "GCF_000025945.1\n",
      "GCF_000025965.1\n",
      "GCF_000026005.1\n",
      "GCF_000026045.1\n",
      "GCF_000026065.1\n",
      "GCF_000026085.1\n",
      "GCF_000026105.1\n",
      "GCF_000026125.1\n",
      "GCF_000026145.1\n",
      "GCF_000026205.1\n",
      "GCF_000026345.1\n",
      "GCF_000026405.1\n",
      "GCF_000026505.1\n",
      "GCF_000026605.1\n",
      "GCF_000026745.1\n",
      "GCF_000026945.1\n",
      "GCF_000027165.1\n",
      "GCF_000027225.1\n",
      "GCF_000027305.1\n",
      "GCF_000027325.1\n",
      "GCF_000027345.1\n",
      "GCF_000043285.1\n",
      "GCF_000046705.1\n",
      "GCF_000055785.1\n",
      "GCF_000055945.1\n",
      "GCF_000056065.1\n",
      "GCF_000058485.1\n",
      "GCF_000060345.1\n",
      "GCF_000061505.1\n",
      "GCF_000062885.1\n",
      "GCF_000063445.1\n",
      "GCF_000063505.1\n",
      "GCF_000063525.1\n",
      "GCF_000063545.1\n",
      "GCF_000063585.1\n",
      "GCF_000063605.1\n",
      "GCF_000064305.2\n",
      "GCF_000067165.1\n",
      "GCF_000067205.1\n",
      "GCF_000068585.1\n",
      "GCF_000069025.1\n",
      "GCF_000069185.1\n",
      "GCF_000069225.1\n",
      "GCF_000069925.1\n",
      "GCF_000069945.1\n",
      "GCF_000069965.1\n",
      "GCF_000072485.1\n",
      "GCF_000085865.1\n",
      "GCF_000087965.2\n",
      "GCF_000090965.1\n",
      "GCF_000090985.2\n",
      "GCF_000091025.4\n",
      "GCF_000091125.1\n",
      "GCF_000091205.1\n",
      "GCF_000091305.1\n",
      "GCF_000091325.1\n",
      "GCF_000091465.1\n",
      "GCF_000091545.1\n",
      "GCF_000091565.1\n",
      "GCF_000091785.1\n",
      "GCF_000091985.1\n",
      "GCF_000092025.1\n",
      "GCF_000092045.1\n",
      "GCF_000092065.1\n",
      "GCF_000092105.1\n",
      "GCF_000092125.1\n",
      "GCF_000092185.1\n",
      "GCF_000092205.1\n",
      "GCF_000092225.1\n",
      "GCF_000092245.1\n",
      "GCF_000092265.1\n",
      "GCF_000092305.1\n",
      "GCF_000092365.1\n",
      "GCF_000092385.1\n",
      "GCF_000092405.1\n",
      "GCF_000092425.1\n",
      "GCF_000092465.1\n",
      "GCF_000092505.1\n",
      "GCF_000092565.1\n",
      "GCF_000092585.1\n",
      "GCF_000092645.1\n",
      "GCF_000092785.1\n",
      "GCF_000092825.1\n",
      "GCF_000092845.1\n",
      "GCF_000092865.1\n",
      "GCF_000092905.1\n",
      "GCF_000092925.1\n",
      "GCF_000092945.1\n",
      "GCF_000092985.1\n",
      "GCF_000093025.1\n",
      "GCF_000093065.1\n",
      "GCF_000093085.1\n",
      "GCF_000093145.3\n",
      "GCF_000093165.1\n",
      "GCF_000143085.1\n",
      "GCF_000143145.1\n",
      "GCF_000143165.1\n",
      "GCF_000143845.1\n",
      "GCF_000143965.1\n",
      "GCF_000143985.1\n",
      "GCF_000144405.1\n",
      "GCF_000144605.1\n",
      "GCF_000144625.1\n",
      "GCF_000144645.1\n",
      "GCF_000144675.1\n",
      "GCF_000144695.1\n",
      "GCF_000144915.1\n",
      "GCF_000145035.1\n",
      "GCF_000145235.1\n",
      "GCF_000145255.1\n",
      "GCF_000145275.1\n",
      "GCF_000145295.1\n",
      "GCF_000145615.1\n",
      "GCF_000146045.2\n",
      "GCF_000146065.2\n",
      "GCF_000146165.2\n",
      "GCF_000146185.1\n",
      "GCF_000146505.1\n",
      "GCF_000147055.1\n",
      "GCF_000147335.1\n",
      "GCF_000147355.1\n",
      "GCF_000147695.2\n",
      "GCF_000147715.2\n",
      "GCF_000147815.2\n",
      "GCF_000147835.2\n",
      "GCF_000147875.1\n",
      "GCF_000148385.1\n",
      "GCF_000148645.1\n",
      "GCF_000151105.2\n",
      "GCF_000151205.2\n",
      "GCF_000152245.2\n",
      "GCF_000152265.2\n",
      "GCF_000152825.2\n",
      "GCF_000153165.2\n",
      "GCF_000153405.2\n",
      "GCF_000153485.2\n",
      "GCF_000154785.2\n",
      "GCF_000155735.2\n",
      "GCF_000157355.2\n",
      "GCF_000157895.3\n",
      "GCF_000158275.2\n",
      "GCF_000159155.2\n",
      "GCF_000162235.2\n",
      "GCF_000163895.2\n",
      "GCF_000164675.2\n",
      "GCF_000164695.2\n",
      "GCF_000164865.1\n",
      "GCF_000164985.3\n",
      "GCF_000165465.1\n",
      "GCF_000165485.1\n",
      "GCF_000165505.1\n",
      "GCF_000165715.2\n",
      "GCF_000165905.1\n",
      "GCF_000166055.1\n",
      "GCF_000166095.1\n",
      "GCF_000166135.1\n",
      "GCF_000166295.1\n",
      "GCF_000166395.1\n",
      "GCF_000166415.1\n",
      "GCF_000172635.2\n",
      "GCF_000172995.2\n",
      "GCF_000174395.2\n",
      "GCF_000175095.2\n",
      "GCF_000175115.2\n",
      "GCF_000175215.2\n",
      "GCF_000175255.2\n",
      "GCF_000175575.2\n",
      "GCF_000176115.2\n",
      "GCF_000176855.2\n",
      "GCF_000176915.2\n",
      "GCF_000177235.2\n",
      "GCF_000177535.2\n",
      "GCF_000177615.2\n",
      "GCF_000177635.2\n",
      "GCF_000178115.2\n",
      "GCF_000178835.2\n",
      "GCF_000178875.2\n",
      "GCF_000178955.2\n",
      "GCF_000178975.2\n",
      "GCF_000179035.2\n",
      "GCF_000179395.2\n",
      "GCF_000179575.2\n",
      "GCF_000179635.2\n",
      "GCF_000179915.2\n",
      "GCF_000180175.2\n",
      "GCF_000182745.2\n",
      "GCF_000183135.1\n",
      "GCF_000183345.1\n",
      "GCF_000183385.1\n",
      "GCF_000183405.1\n",
      "GCF_000183425.1\n",
      "GCF_000183665.1\n",
      "GCF_000183725.1\n",
      "GCF_000183745.1\n",
      "GCF_000184345.1\n",
      "GCF_000184435.1\n",
      "GCF_000184685.1\n",
      "GCF_000184705.1\n",
      "GCF_000184745.1\n",
      "GCF_000185805.1\n",
      "GCF_000185885.1\n",
      "GCF_000185905.1\n",
      "GCF_000185965.1\n",
      "GCF_000185985.2\n",
      "GCF_000186245.1\n",
      "GCF_000186265.1\n",
      "GCF_000186345.1\n",
      "GCF_000186365.1\n",
      "GCF_000186385.1\n",
      "GCF_000186885.1\n",
      "GCF_000186985.1\n",
      "GCF_000187005.1\n",
      "GCF_000189295.2\n",
      "GCF_000189415.1\n",
      "GCF_000189535.1\n",
      "GCF_000189775.2\n",
      "GCF_000190315.1\n",
      "GCF_000190435.1\n",
      "GCF_000190535.1\n",
      "GCF_000190555.1\n",
      "GCF_000190575.1\n",
      "GCF_000190595.1\n",
      "GCF_000190635.1\n",
      "GCF_000190735.1\n",
      "GCF_000191045.1\n",
      "GCF_000191145.1\n",
      "GCF_000191545.1\n",
      "GCF_000191585.1\n",
      "GCF_000192745.1\n",
      "GCF_000192845.1\n",
      "GCF_000192865.1\n",
      "GCF_000193375.1\n",
      "GCF_000193395.1\n",
      "GCF_000194135.1\n",
      "GCF_000194605.1\n",
      "GCF_000194625.1\n",
      "GCF_000195085.1\n",
      "GCF_000195275.1\n",
      "GCF_000195295.1\n",
      "GCF_000195315.1\n",
      "GCF_000195335.1\n",
      "GCF_000195575.1\n",
      "GCF_000195715.1\n",
      "GCF_000195735.1\n",
      "GCF_000195755.1\n",
      "GCF_000195775.1\n",
      "GCF_000195835.1\n",
      "GCF_000195855.1\n",
      "GCF_000195875.1\n",
      "GCF_000195915.1\n",
      "GCF_000195935.2\n",
      "GCF_000195955.2\n",
      "GCF_000195975.1\n",
      "GCF_000195995.1\n",
      "GCF_000196015.1\n",
      "GCF_000196035.1\n",
      "GCF_000196095.1\n",
      "GCF_000196115.1\n",
      "GCF_000196135.1\n",
      "GCF_000196175.1\n",
      "GCF_000196235.1\n",
      "GCF_000196255.1\n",
      "GCF_000196275.1\n",
      "GCF_000196295.1\n",
      "GCF_000196315.1\n",
      "GCF_000196355.1\n",
      "GCF_000196395.1\n",
      "GCF_000196435.1\n",
      "GCF_000196455.1\n",
      "GCF_000196495.1\n",
      "GCF_000196515.1\n",
      "GCF_000196535.1\n",
      "GCF_000196615.1\n",
      "GCF_000196655.1\n",
      "GCF_000196675.1\n",
      "GCF_000196695.1\n",
      "GCF_000196735.1\n",
      "GCF_000196815.1\n",
      "GCF_000196835.1\n",
      "GCF_000196855.1\n",
      "GCF_000196875.1\n",
      "GCF_000196895.1\n",
      "GCF_000197735.1\n",
      "GCF_000198515.1\n",
      "GCF_000198775.1\n",
      "GCF_000199675.1\n",
      "GCF_000200595.1\n",
      "GCF_000202635.1\n",
      "GCF_000202835.1\n",
      "GCF_000203835.1\n",
      "GCF_000203855.3\n",
      "GCF_000203895.1\n",
      "GCF_000204135.1\n",
      "GCF_000204155.1\n",
      "GCF_000204255.1\n",
      "GCF_000204415.1\n",
      "GCF_000204565.1\n",
      "GCF_000204645.1\n",
      "GCF_000204925.1\n",
      "GCF_000208385.1\n",
      "GCF_000208405.1\n",
      "GCF_000209655.1\n",
      "GCF_000209675.1\n",
      "GCF_000210915.2\n",
      "GCF_000211475.1\n",
      "GCF_000211855.2\n",
      "GCF_000212355.1\n",
      "GCF_000212375.1\n",
      "GCF_000212395.1\n",
      "GCF_000212415.1\n",
      "GCF_000212675.2\n",
      "GCF_000212695.1\n",
      "GCF_000212735.1\n",
      "GCF_000213215.1\n",
      "GCF_000213255.1\n",
      "GCF_000213805.1\n",
      "GCF_000213825.1\n",
      "GCF_000214095.2\n",
      "GCF_000214155.1\n",
      "GCF_000214175.1\n",
      "GCF_000214215.1\n",
      "GCF_000214355.1\n",
      "GCF_000214375.1\n",
      "GCF_000214415.1\n",
      "GCF_000214435.1\n",
      "GCF_000214665.1\n",
      "GCF_000214725.1\n",
      "GCF_000214785.1\n",
      "GCF_000214825.1\n",
      "GCF_000215085.1\n",
      "GCF_000215105.1\n",
      "GCF_000215705.1\n",
      "GCF_000215745.1\n",
      "GCF_000215975.1\n",
      "GCF_000215995.1\n",
      "GCF_000217635.1\n",
      "GCF_000217675.1\n",
      "GCF_000217715.1\n",
      "GCF_000217795.1\n",
      "GCF_000217815.1\n",
      "GCF_000217995.1\n",
      "GCF_000218545.1\n",
      "GCF_000218565.1\n",
      "GCF_000218625.1\n",
      "GCF_000218875.1\n",
      "GCF_000218895.1\n",
      "GCF_000219045.1\n",
      "GCF_000219105.1\n",
      "GCF_000219175.1\n",
      "GCF_000219355.1\n",
      "GCF_000219535.2\n",
      "GCF_000219585.1\n",
      "GCF_000219605.1\n",
      "GCF_000219725.1\n",
      "GCF_000219805.1\n",
      "GCF_000220625.1\n",
      "GCF_000220645.1\n",
      "GCF_000222305.1\n",
      "GCF_000222485.1\n",
      "GCF_000223375.1\n",
      "GCF_000223395.1\n",
      "GCF_000223905.1\n",
      "GCF_000224005.2\n",
      "GCF_000224085.1\n",
      "GCF_000224105.1\n",
      "GCF_000224475.1\n",
      "GCF_000224985.1\n",
      "GCF_000225325.1\n",
      "GCF_000225345.1\n",
      "GCF_000225445.1\n",
      "GCF_000225465.1\n",
      "GCF_000226095.1\n",
      "GCF_000226115.1\n",
      "GCF_000226295.1\n",
      "GCF_000226315.1\n",
      "GCF_000226565.1\n",
      "GCF_000226625.1\n",
      "GCF_000227665.2\n",
      "GCF_000227685.2\n",
      "GCF_000227705.2\n",
      "GCF_000227745.2\n",
      "GCF_000230275.1\n",
      "GCF_000230555.1\n",
      "GCF_000230695.2\n",
      "GCF_000230715.2\n",
      "GCF_000230735.2\n",
      "GCF_000230895.2\n",
      "GCF_000230955.2\n",
      "GCF_000231015.2\n",
      "GCF_000231385.2\n",
      "GCF_000231405.2\n",
      "GCF_000233435.1\n",
      "GCF_000233715.2\n",
      "GCF_000233775.1\n",
      "GCF_000233915.3\n",
      "GCF_000235405.2\n",
      "GCF_000235565.1\n",
      "GCF_000235605.1\n",
      "GCF_000236405.1\n",
      "GCF_000236585.1\n",
      "GCF_000236665.1\n",
      "GCF_000236685.1\n",
      "GCF_000236705.1\n",
      "GCF_000237065.1\n",
      "GCF_000237085.1\n",
      "GCF_000237145.1\n",
      "GCF_000237205.1\n",
      "GCF_000237865.1\n",
      "GCF_000237995.1\n",
      "GCF_000238215.1\n",
      "GCF_000238995.1\n",
      "GCF_000240075.2\n",
      "GCF_000240165.1\n",
      "GCF_000240185.1\n",
      "GCF_000242255.2\n",
      "GCF_000242335.1\n",
      "GCF_000242455.2\n",
      "GCF_000242595.2\n",
      "GCF_000242635.2\n",
      "GCF_000242895.2\n",
      "GCF_000242935.2\n",
      "GCF_000243075.1\n",
      "GCF_000243115.2\n",
      "GCF_000243135.2\n",
      "GCF_000243155.2\n",
      "GCF_000245355.1\n",
      "GCF_000246855.1\n",
      "GCF_000246985.2\n",
      "GCF_000247565.1\n",
      "GCF_000247605.1\n",
      "GCF_000247715.1\n",
      "GCF_000248095.2\n",
      "GCF_000250635.1\n",
      "GCF_000250655.1\n",
      "GCF_000250675.2\n",
      "GCF_000251105.1\n",
      "GCF_000252445.1\n",
      "GCF_000252855.1\n",
      "GCF_000253015.1\n",
      "GCF_000253035.1\n",
      "GCF_000253055.1\n",
      "GCF_000253275.1\n",
      "GCF_000253295.1\n",
      "GCF_000255115.2\n",
      "GCF_000255135.1\n",
      "GCF_000255295.1\n",
      "GCF_000258405.1\n",
      "GCF_000259175.1\n",
      "GCF_000259255.1\n",
      "GCF_000259275.1\n",
      "GCF_000260965.1\n",
      "GCF_000260985.3\n",
      "GCF_000261025.1\n",
      "GCF_000261045.2\n",
      "GCF_000262305.1\n",
      "GCF_000262715.1\n",
      "GCF_000263195.1\n",
      "GCF_000263735.1\n",
      "GCF_000264455.2\n",
      "GCF_000264495.1\n",
      "GCF_000265295.1\n",
      "GCF_000265365.1\n",
      "GCF_000265385.1\n",
      "GCF_000265405.1\n",
      "GCF_000265425.1\n",
      "GCF_000265465.1\n",
      "GCF_000265505.1\n",
      "GCF_000265525.1\n",
      "GCF_000266885.1\n",
      "GCF_000266905.1\n",
      "GCF_000266925.1\n",
      "GCF_000266945.1\n",
      "GCF_000269945.1\n",
      "GCF_000269985.1\n",
      "GCF_000270085.1\n",
      "GCF_000270145.1\n",
      "GCF_000270205.1\n",
      "GCF_000270245.1\n",
      "GCF_000270285.1\n",
      "GCF_000270305.1\n",
      "GCF_000271665.2\n",
      "GCF_000276685.1\n",
      "GCF_000277125.1\n",
      "GCF_000277715.1\n",
      "GCF_000277795.1\n",
      "GCF_000279145.1\n",
      "GCF_000279165.1\n",
      "GCF_000281175.1\n",
      "GCF_000283515.1\n",
      "GCF_000283555.1\n",
      "GCF_000283575.1\n",
      "GCF_000283595.1\n",
      "GCF_000283615.1\n",
      "GCF_000283655.1\n",
      "GCF_000283875.1\n",
      "GCF_000284015.1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000284035.1\n",
      "GCF_000284075.1\n",
      "GCF_000284095.1\n",
      "GCF_000284115.1\n",
      "GCF_000284155.1\n",
      "GCF_000284175.1\n",
      "GCF_000284235.1\n",
      "GCF_000284255.1\n",
      "GCF_000284295.1\n",
      "GCF_000284315.1\n",
      "GCF_000284335.1\n",
      "GCF_000284375.1\n",
      "GCF_000284415.1\n",
      "GCF_000284515.1\n",
      "GCF_000284615.1\n",
      "GCF_000286435.2\n",
      "GCF_000287335.1\n",
      "GCF_000287355.1\n",
      "GCF_000294515.1\n",
      "GCF_000294715.1\n",
      "GCF_000294775.2\n",
      "GCF_000297055.2\n",
      "GCF_000298115.2\n",
      "GCF_000299095.1\n",
      "GCF_000299115.1\n",
      "GCF_000299235.1\n",
      "GCF_000299335.2\n",
      "GCF_000299355.1\n",
      "GCF_000299395.1\n",
      "GCF_000299455.1\n",
      "GCF_000300005.1\n",
      "GCF_000300045.1\n",
      "GCF_000300075.1\n",
      "GCF_000300135.1\n",
      "GCF_000300235.2\n",
      "GCF_000300255.2\n",
      "GCF_000300295.3\n",
      "GCF_000300455.3\n",
      "GCF_000304215.1\n",
      "GCF_000304315.1\n",
      "GCF_000304355.2\n",
      "GCF_000304455.1\n",
      "GCF_000304735.1\n",
      "GCF_000305785.2\n",
      "GCF_000305935.1\n",
      "GCF_000306725.1\n",
      "GCF_000306765.2\n",
      "GCF_000306785.1\n",
      "GCF_000306885.1\n",
      "GCF_000307105.1\n",
      "GCF_000307165.1\n",
      "GCF_000309885.1\n",
      "GCF_000311765.1\n",
      "GCF_000312705.1\n",
      "GCF_000313175.2\n",
      "GCF_000313635.1\n",
      "GCF_000316515.1\n",
      "GCF_000316575.1\n",
      "GCF_000316605.1\n",
      "GCF_000316625.1\n",
      "GCF_000316645.1\n",
      "GCF_000316665.1\n",
      "GCF_000316685.1\n",
      "GCF_000317025.1\n",
      "GCF_000317045.1\n",
      "GCF_000317065.1\n",
      "GCF_000317085.1\n",
      "GCF_000317105.1\n",
      "GCF_000317125.1\n",
      "GCF_000317305.3\n",
      "GCF_000317435.1\n",
      "GCF_000317475.1\n",
      "GCF_000317495.1\n",
      "GCF_000317515.1\n",
      "GCF_000317555.1\n",
      "GCF_000317575.1\n",
      "GCF_000317615.1\n",
      "GCF_000317635.1\n",
      "GCF_000317675.1\n",
      "GCF_000317695.1\n",
      "GCF_000317795.1\n",
      "GCF_000317835.1\n",
      "GCF_000317855.1\n",
      "GCF_000317935.1\n",
      "GCF_000317975.2\n",
      "GCF_000318015.1\n",
      "GCF_000319245.1\n",
      "GCF_000319385.1\n",
      "GCF_000319575.2\n",
      "GCF_000321415.2\n",
      "GCF_000325665.1\n",
      "GCF_000325705.1\n",
      "GCF_000325745.1\n",
      "GCF_000327045.1\n",
      "GCF_000327485.1\n",
      "GCF_000327505.1\n",
      "GCF_000328565.1\n",
      "GCF_000328625.1\n",
      "GCF_000328665.1\n",
      "GCF_000328685.1\n",
      "GCF_000328705.1\n",
      "GCF_000328725.1\n",
      "GCF_000328765.2\n",
      "GCF_000330865.1\n",
      "GCF_000330885.1\n",
      "GCF_000331735.1\n",
      "GCF_000331995.1\n",
      "GCF_000332115.1\n",
      "GCF_000332735.1\n",
      "GCF_000334405.1\n",
      "GCF_000340435.2\n",
      "GCF_000340795.1\n",
      "GCF_000340825.1\n",
      "GCF_000340865.1\n",
      "GCF_000340885.1\n",
      "GCF_000340905.1\n",
      "GCF_000341345.1\n",
      "GCF_000341355.1\n",
      "GCF_000341385.1\n",
      "GCF_000341395.1\n",
      "GCF_000344785.1\n",
      "GCF_000344805.1\n",
      "GCF_000346595.1\n",
      "GCF_000347595.1\n",
      "GCF_000347635.1\n",
      "GCF_000347675.2\n",
      "GCF_000347695.1\n",
      "GCF_000348725.1\n",
      "GCF_000348785.1\n",
      "GCF_000348805.1\n",
      "GCF_000349845.1\n",
      "GCF_000349945.1\n",
      "GCF_000349975.1\n",
      "GCF_000350305.1\n",
      "GCF_000354175.2\n",
      "GCF_000355675.1\n",
      "GCF_000359525.1\n",
      "GCF_000367205.1\n",
      "GCF_000376545.2\n",
      "GCF_000376585.1\n",
      "GCF_000376645.1\n",
      "GCF_000385565.1\n",
      "GCF_000385925.1\n",
      "GCF_000385945.1\n",
      "GCF_000389635.1\n",
      "GCF_000389965.1\n",
      "GCF_000397205.1\n",
      "GCF_000400935.1\n",
      "GCF_000400955.1\n",
      "GCF_000403645.1\n",
      "GCF_000404225.1\n",
      "GCF_000412695.1\n",
      "GCF_000418365.1\n",
      "GCF_000422085.1\n",
      "GCF_000422165.1\n",
      "GCF_000439435.1\n",
      "GCF_000439455.1\n",
      "GCF_000441555.1\n",
      "GCF_000442595.1\n",
      "GCF_000442645.1\n",
      "GCF_000444875.1\n",
      "GCF_000444995.1\n",
      "GCF_000445425.4\n",
      "GCF_000447675.1\n",
      "GCF_000455605.1\n",
      "GCF_000463355.1\n",
      "GCF_000463505.1\n",
      "GCF_000468615.2\n",
      "GCF_000470655.1\n",
      "GCF_000470775.1\n",
      "GCF_000471025.2\n",
      "GCF_000471965.1\n",
      "GCF_000473245.1\n",
      "GCF_000473305.1\n",
      "GCF_000473995.1\n",
      "GCF_000477415.1\n",
      "GCF_000477435.1\n",
      "GCF_000478885.1\n",
      "GCF_000478905.1\n",
      "GCF_000484505.1\n",
      "GCF_000484535.1\n",
      "GCF_000485905.1\n",
      "GCF_000493735.1\n",
      "GCF_000494755.1\n",
      "GCF_000495505.1\n",
      "GCF_000495935.2\n",
      "GCF_000496595.1\n",
      "GCF_000497525.1\n",
      "GCF_000499665.2\n",
      "GCF_000499765.1\n",
      "GCF_000500935.1\n",
      "GCF_000503895.1\n",
      "GCF_000504105.1\n",
      "GCF_000504125.1\n",
      "GCF_000507245.1\n",
      "GCF_000508225.1\n",
      "GCF_000508245.1\n",
      "GCF_000510265.1\n",
      "GCF_000511305.1\n",
      "GCF_000511355.1\n",
      "GCF_000511385.1\n",
      "GCF_000511405.1\n",
      "GCF_000512205.2\n",
      "GCF_000512355.1\n",
      "GCF_000512735.1\n",
      "GCF_000512895.1\n",
      "GCF_000512915.1\n",
      "GCF_000513215.1\n",
      "GCF_000513295.1\n",
      "GCF_000513475.1\n",
      "GCF_000517365.1\n",
      "GCF_000517425.1\n",
      "GCF_000517445.1\n",
      "GCF_000517565.1\n",
      "GCF_000517605.1\n",
      "GCF_000517625.1\n",
      "GCF_000520015.2\n",
      "GCF_000521505.1\n",
      "GCF_000521565.1\n",
      "GCF_000523235.1\n",
      "GCF_000524555.1\n",
      "GCF_000525635.1\n",
      "GCF_000525655.1\n",
      "GCF_000525675.1\n",
      "GCF_000550765.1\n",
      "GCF_000550785.1\n",
      "GCF_000550805.1\n",
      "GCF_000565175.1\n",
      "GCF_000565215.1\n",
      "GCF_000568815.1\n",
      "GCF_000576555.1\n",
      "GCF_000577895.1\n",
      "GCF_000583875.1\n",
      "GCF_000585495.1\n",
      "GCF_000590475.1\n",
      "GCF_000590555.1\n",
      "GCF_000590925.1\n",
      "GCF_000591055.1\n",
      "GCF_000597865.1\n",
      "GCF_000597885.1\n",
      "GCF_000599985.1\n",
      "GCF_000600005.1\n",
      "GCF_000600105.1\n",
      "GCF_000612055.1\n",
      "GCF_000612075.2\n",
      "GCF_000612485.1\n",
      "GCF_000612685.1\n",
      "GCF_000619905.2\n",
      "GCF_000626635.1\n",
      "GCF_000626675.1\n",
      "GCF_000632805.1\n",
      "GCF_000632845.1\n",
      "GCF_000632985.1\n",
      "GCF_000648515.1\n",
      "GCF_000661895.1\n",
      "GCF_000689415.1\n",
      "GCF_000695235.1\n",
      "GCF_000695835.1\n",
      "GCF_000696485.1\n",
      "GCF_000697965.2\n",
      "GCF_000699505.1\n",
      "GCF_000706685.1\n",
      "GCF_000706725.1\n",
      "GCF_000709555.1\n",
      "GCF_000723165.1\n",
      "GCF_000723465.1\n",
      "GCF_000723505.1\n",
      "GCF_000724485.1\n",
      "GCF_000724605.1\n",
      "GCF_000724625.1\n",
      "GCF_000724775.3\n",
      "GCF_000725345.1\n",
      "GCF_000725365.1\n",
      "GCF_000725405.1\n",
      "GCF_000725425.1\n",
      "GCF_000730245.1\n",
      "GCF_000730385.1\n",
      "GCF_000731315.1\n",
      "GCF_000732925.1\n",
      "GCF_000732945.1\n",
      "GCF_000734015.1\n",
      "GCF_000734895.2\n",
      "GCF_000736415.1\n",
      "GCF_000737325.1\n",
      "GCF_000737515.1\n",
      "GCF_000737535.1\n",
      "GCF_000737575.1\n",
      "GCF_000737595.1\n",
      "GCF_000737865.1\n",
      "GCF_000738435.1\n",
      "GCF_000739065.1\n",
      "GCF_000739085.1\n",
      "GCF_000739455.1\n",
      "GCF_000739475.1\n",
      "GCF_000740895.1\n",
      "GCF_000742835.1\n",
      "GCF_000743945.1\n",
      "GCF_000746525.1\n",
      "GCF_000746585.1\n",
      "GCF_000746645.1\n",
      "GCF_000747315.1\n",
      "GCF_000747345.1\n",
      "GCF_000747525.1\n",
      "GCF_000750535.1\n",
      "GCF_000754265.1\n",
      "GCF_000754275.1\n",
      "GCF_000755145.1\n",
      "GCF_000755705.1\n",
      "GCF_000756615.1\n",
      "GCF_000757785.1\n",
      "GCF_000757825.1\n",
      "GCF_000757865.1\n",
      "GCF_000758625.1\n",
      "GCF_000758665.1\n",
      "GCF_000758685.1\n",
      "GCF_000758725.1\n",
      "GCF_000759475.1\n",
      "GCF_000759535.1\n",
      "GCF_000761155.1\n",
      "GCF_000762265.1\n",
      "GCF_000763535.1\n",
      "GCF_000763575.1\n",
      "GCF_000764535.1\n",
      "GCF_000764555.1\n",
      "GCF_000767055.1\n",
      "GCF_000767275.2\n",
      "GCF_000767465.1\n",
      "GCF_000767615.3\n",
      "GCF_000769655.1\n",
      "GCF_000772105.1\n",
      "GCF_000784965.1\n",
      "GCF_000785495.1\n",
      "GCF_000785515.1\n",
      "GCF_000785535.1\n",
      "GCF_000785705.2\n",
      "GCF_000789255.1\n",
      "GCF_000789395.1\n",
      "GCF_000800295.1\n",
      "GCF_000800395.1\n",
      "GCF_000800805.1\n",
      "GCF_000801275.2\n",
      "GCF_000801295.1\n",
      "GCF_000801315.1\n",
      "GCF_000803645.1\n",
      "GCF_000807255.1\n",
      "GCF_000807275.1\n",
      "GCF_000812185.1\n",
      "GCF_000812665.2\n",
      "GCF_000813245.1\n",
      "GCF_000814825.1\n",
      "GCF_000815025.1\n",
      "GCF_000815065.1\n",
      "GCF_000815185.1\n",
      "GCF_000815225.1\n",
      "GCF_000816085.1\n",
      "GCF_000816185.1\n",
      "GCF_000816305.1\n",
      "GCF_000816845.1\n",
      "GCF_000817955.1\n",
      "GCF_000818015.1\n",
      "GCF_000818035.1\n",
      "GCF_000818095.1\n",
      "GCF_000819445.1\n",
      "GCF_000819565.1\n",
      "GCF_000827005.1\n",
      "GCF_000827125.1\n",
      "GCF_000828475.1\n",
      "GCF_000828515.1\n",
      "GCF_000828615.1\n",
      "GCF_000828635.1\n",
      "GCF_000828655.1\n",
      "GCF_000828675.1\n",
      "GCF_000828715.1\n",
      "GCF_000828815.1\n",
      "GCF_000828835.1\n",
      "GCF_000828855.1\n",
      "GCF_000828895.1\n",
      "GCF_000828915.1\n",
      "GCF_000828975.1\n",
      "GCF_000829195.1\n",
      "GCF_000829235.1\n",
      "GCF_000829315.1\n",
      "GCF_000829395.1\n",
      "GCF_000829415.1\n",
      "GCF_000830005.1\n",
      "GCF_000830885.1\n",
      "GCF_000830985.1\n",
      "GCF_000831005.1\n",
      "GCF_000831485.1\n",
      "GCF_000831645.3\n",
      "GCF_000832305.1\n",
      "GCF_000832605.1\n",
      "GCF_000832905.1\n",
      "GCF_000832985.1\n",
      "GCF_000833575.1\n",
      "GCF_000835165.1\n",
      "GCF_000836805.1\n",
      "GCF_000836945.1\n",
      "GCF_000839085.1\n",
      "GCF_000844925.1\n",
      "GCF_000846145.1\n",
      "GCF_000846805.1\n",
      "GCF_000847425.1\n",
      "GCF_000849665.1\n",
      "GCF_000851485.1\n",
      "GCF_000853245.1\n",
      "GCF_000854845.1\n",
      "GCF_000855585.1\n",
      "GCF_000856665.2\n",
      "GCF_000857085.1\n",
      "GCF_000860865.1\n",
      "GCF_000861085.1\n",
      "GCF_000861165.1\n",
      "GCF_000861845.1\n",
      "GCF_000862445.1\n",
      "GCF_000863085.1\n",
      "GCF_000863925.1\n",
      "GCF_000864005.1\n",
      "GCF_000864765.1\n",
      "GCF_000864885.1\n",
      "GCF_000865085.1\n",
      "GCF_000866645.1\n",
      "GCF_000868425.1\n",
      "GCF_000871845.1\n",
      "GCF_000874105.1\n",
      "GCF_000875755.1\n",
      "GCF_000880355.1\n",
      "GCF_000880515.1\n",
      "GCF_000888475.1\n",
      "GCF_000901155.1\n",
      "GCF_000907255.1\n",
      "GCF_000928555.1\n",
      "GCF_000935025.1\n",
      "GCF_000940805.1\n",
      "GCF_000940845.1\n",
      "GCF_000940995.1\n",
      "GCF_000941055.1\n",
      "GCF_000941075.1\n",
      "GCF_000949425.1\n",
      "GCF_000950575.1\n",
      "GCF_000952035.1\n",
      "GCF_000952975.1\n",
      "GCF_000953015.1\n",
      "GCF_000953135.1\n",
      "GCF_000953195.1\n",
      "GCF_000953215.1\n",
      "GCF_000953435.1\n",
      "GCF_000953475.1\n",
      "GCF_000953535.1\n",
      "GCF_000953635.1\n",
      "GCF_000953655.1\n",
      "GCF_000953695.1\n",
      "GCF_000953735.1\n",
      "GCF_000954135.1\n",
      "GCF_000955905.1\n",
      "GCF_000959245.1\n",
      "GCF_000959725.1\n",
      "GCF_000960975.1\n",
      "GCF_000961095.1\n",
      "GCF_000963865.1\n",
      "GCF_000964565.1\n",
      "GCF_000967305.2\n",
      "GCF_000967425.1\n",
      "GCF_000967895.1\n",
      "GCF_000967915.1\n",
      "GCF_000968055.1\n",
      "GCF_000968135.1\n",
      "GCF_000968195.1\n",
      "GCF_000968375.1\n",
      "GCF_000968535.2\n",
      "GCF_000969645.2\n",
      "GCF_000969765.1\n",
      "GCF_000969885.1\n",
      "GCF_000969905.1\n",
      "GCF_000969965.1\n",
      "GCF_000970025.1\n",
      "GCF_000970045.1\n",
      "GCF_000970085.1\n",
      "GCF_000970205.1\n",
      "GCF_000970265.1\n",
      "GCF_000970285.1\n",
      "GCF_000970305.1\n",
      "GCF_000970325.1\n",
      "GCF_000971565.1\n",
      "GCF_000972245.3\n",
      "GCF_000972765.1\n",
      "GCF_000972865.1\n",
      "GCF_000973085.1\n",
      "GCF_000973105.1\n",
      "GCF_000973505.1\n",
      "GCF_000973545.1\n",
      "GCF_000973625.1\n",
      "GCF_000973725.1\n",
      "GCF_000974425.1\n",
      "GCF_000980815.1\n",
      "GCF_000980835.1\n",
      "GCF_000981505.1\n",
      "GCF_000981525.1\n",
      "GCF_000981545.1\n",
      "GCF_000981765.1\n",
      "GCF_000982715.1\n",
      "GCF_000982825.1\n",
      "GCF_000987835.1\n",
      "GCF_000993785.2\n",
      "GCF_000993805.1\n",
      "GCF_001005905.1\n",
      "GCF_001006005.1\n",
      "GCF_001006045.1\n",
      "GCF_001007875.1\n",
      "GCF_001007935.1\n",
      "GCF_001007995.1\n",
      "GCF_001008165.2\n",
      "GCF_001010285.1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_001010805.1\n",
      "GCF_001011035.1\n",
      "GCF_001011155.1\n",
      "GCF_001013905.1\n",
      "GCF_001017435.1\n",
      "GCF_001017655.1\n",
      "GCF_001017775.2\n",
      "GCF_001020955.1\n",
      "GCF_001020985.1\n",
      "GCF_001021025.1\n",
      "GCF_001021045.1\n",
      "GCF_001021065.1\n",
      "GCF_001021085.1\n",
      "GCF_001021385.1\n",
      "GCF_001021935.1\n",
      "GCF_001021975.1\n",
      "GCF_001022135.1\n",
      "GCF_001022195.1\n",
      "GCF_001023575.1\n",
      "GCF_001025155.1\n",
      "GCF_001025175.1\n",
      "GCF_001026985.1\n",
      "GCF_001027025.1\n",
      "GCF_001027285.1\n",
      "GCF_001027545.1\n",
      "GCF_001028625.1\n",
      "GCF_001028705.1\n",
      "GCF_001029265.1\n",
      "GCF_001040945.1\n",
      "GCF_001042405.1\n",
      "GCF_001042595.1\n",
      "GCF_001042615.1\n",
      "GCF_001042635.1\n",
      "GCF_001042675.1\n",
      "GCF_001042695.1\n",
      "GCF_001042715.1\n",
      "GCF_001043175.1\n",
      "GCF_001046955.1\n",
      "GCF_001050115.1\n",
      "GCF_001050435.1\n",
      "GCF_001050475.1\n",
      "GCF_001051995.2\n",
      "GCF_001077715.1\n",
      "GCF_001077815.2\n",
      "GCF_001078055.1\n",
      "GCF_001078275.1\n",
      "GCF_001182745.1\n",
      "GCF_001185205.1\n",
      "GCF_001187595.1\n",
      "GCF_001187785.1\n",
      "GCF_001187845.1\n",
      "GCF_001188915.1\n",
      "GCF_001189295.1\n",
      "GCF_001189495.1\n",
      "GCF_001189515.2\n",
      "GCF_001189535.1\n",
      "GCF_001190745.1\n",
      "GCF_001190755.1\n",
      "GCF_001190925.1\n",
      "GCF_001190945.1\n",
      "GCF_001191005.1\n",
      "GCF_001262015.1\n",
      "GCF_001262035.1\n",
      "GCF_001262055.1\n",
      "GCF_001262075.1\n",
      "GCF_001263175.1\n",
      "GCF_001263395.1\n",
      "GCF_001267175.1\n",
      "GCF_001267865.1\n",
      "GCF_001267925.1\n",
      "GCF_001269425.1\n",
      "GCF_001273775.1\n",
      "GCF_001273795.1\n",
      "GCF_001274535.1\n",
      "GCF_001274875.1\n",
      "GCF_001274895.1\n",
      "GCF_001275345.1\n",
      "GCF_001277995.1\n",
      "GCF_001278035.1\n",
      "GCF_001278055.1\n",
      "GCF_001278075.1\n",
      "GCF_001278715.1\n",
      "GCF_001280225.1\n",
      "GCF_001281045.1\n",
      "GCF_001281115.1\n",
      "GCF_001281315.1\n",
      "GCF_001281385.1\n",
      "GCF_001281405.1\n",
      "GCF_001281465.1\n",
      "GCF_001281485.1\n",
      "GCF_001281505.1\n",
      "GCF_001293165.1\n",
      "GCF_001294645.1\n",
      "GCF_001295365.1\n",
      "GCF_001298465.1\n",
      "GCF_001298525.1\n",
      "GCF_001302565.1\n",
      "GCF_001302585.1\n",
      "GCF_001305575.2\n",
      "GCF_001305595.1\n",
      "GCF_001305615.1\n",
      "GCF_001305655.1\n",
      "GCF_001305675.1\n",
      "GCF_001307195.1\n",
      "GCF_001307545.1\n",
      "GCF_001307805.1\n",
      "GCF_001308105.1\n",
      "GCF_001308265.1\n",
      "GCF_001310085.1\n",
      "GCF_001310225.1\n",
      "GCF_001310255.1\n",
      "GCF_001314225.1\n",
      "GCF_001314305.1\n",
      "GCF_001314325.1\n",
      "GCF_001314945.1\n",
      "GCF_001314975.1\n",
      "GCF_001314995.1\n",
      "GCF_001315015.1\n",
      "GCF_001318295.1\n",
      "GCF_001318345.1\n",
      "GCF_001343785.1\n",
      "GCF_001399775.1\n",
      "GCF_001402875.1\n",
      "GCF_001402915.1\n",
      "GCF_001411495.1\n",
      "GCF_001412575.1\n",
      "GCF_001412595.1\n",
      "GCF_001412615.1\n",
      "GCF_001420915.1\n",
      "GCF_001420995.1\n",
      "GCF_001421015.1\n",
      "GCF_001430825.1\n",
      "GCF_001431145.1\n",
      "GCF_001431725.1\n",
      "GCF_001431805.1\n",
      "GCF_001432245.1\n",
      "GCF_001439585.2\n",
      "GCF_001441165.1\n",
      "GCF_001442535.1\n",
      "GCF_001442785.1\n",
      "GCF_001443605.1\n",
      "GCF_001444365.1\n",
      "GCF_001444405.1\n",
      "GCF_001444425.1\n",
      "GCF_001444445.1\n",
      "GCF_001447335.1\n",
      "GCF_001454945.1\n",
      "GCF_001455205.1\n",
      "GCF_001455325.1\n",
      "GCF_001456065.2\n",
      "GCF_001457455.1\n",
      "GCF_001457475.1\n",
      "GCF_001458475.1\n",
      "GCF_001458695.1\n",
      "GCF_001460635.1\n",
      "GCF_001465255.1\n",
      "GCF_001465275.1\n",
      "GCF_001465835.2\n",
      "GCF_001465855.1\n",
      "GCF_001466725.1\n",
      "GCF_001477625.1\n",
      "GCF_001477655.1\n",
      "GCF_001481655.1\n",
      "GCF_001482365.1\n",
      "GCF_001482385.1\n",
      "GCF_001483385.1\n",
      "GCF_001483845.1\n",
      "GCF_001483865.1\n",
      "GCF_001483965.1\n",
      "GCF_001484065.1\n",
      "GCF_001484565.1\n",
      "GCF_001484605.1\n",
      "GCF_001484935.1\n",
      "GCF_001488575.1\n",
      "GCF_001499615.1\n",
      "GCF_001499655.1\n",
      "GCF_001507665.1\n",
      "GCF_001514435.1\n",
      "GCF_001518815.1\n",
      "GCF_001518835.1\n",
      "GCF_001542565.1\n",
      "GCF_001543175.1\n",
      "GCF_001543265.1\n",
      "GCF_001543285.1\n",
      "GCF_001547775.1\n",
      "GCF_001547975.1\n",
      "GCF_001547995.1\n",
      "GCF_001548015.1\n",
      "GCF_001548055.1\n",
      "GCF_001548075.1\n",
      "GCF_001551895.1\n",
      "GCF_001553195.1\n",
      "GCF_001553605.1\n",
      "GCF_001553935.1\n",
      "GCF_001553955.1\n",
      "GCF_001558255.1\n",
      "GCF_001558415.1\n",
      "GCF_001558735.1\n",
      "GCF_001559115.1\n",
      "GCF_001562115.1\n",
      "GCF_001578105.1\n",
      "GCF_001578185.1\n",
      "GCF_001583415.1\n",
      "GCF_001584185.1\n",
      "GCF_001586165.1\n",
      "GCF_001586195.1\n",
      "GCF_001586215.1\n",
      "GCF_001586255.1\n",
      "GCF_001593245.1\n",
      "GCF_001597285.1\n",
      "GCF_001606005.1\n",
      "GCF_001606025.1\n",
      "GCF_001610725.1\n",
      "GCF_001610975.1\n",
      "GCF_001611135.1\n",
      "GCF_001618685.1\n",
      "GCF_001620305.1\n",
      "GCF_001632865.1\n",
      "GCF_001636015.1\n",
      "GCF_001641285.1\n",
      "GCF_001642655.1\n",
      "GCF_001644565.1\n",
      "GCF_001648175.1\n",
      "GCF_001652465.1\n",
      "GCF_001652485.1\n",
      "GCF_001652565.1\n",
      "GCF_001653755.1\n",
      "GCF_001654495.1\n",
      "GCF_001655245.1\n",
      "GCF_001712785.1\n",
      "GCF_900011245.1\n",
      "GCF_900078695.1\n",
      "GCF_900086555.1\n",
      "GCF_900087055.1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3834</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451004.1</td>\n",
       "      <td>2787</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6151</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451005.1</td>\n",
       "      <td>4627</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7056</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_002999344.1</td>\n",
       "      <td>6837</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8579</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451006.2</td>\n",
       "      <td>8138</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11511</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451007.1</td>\n",
       "      <td>10101</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15211</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451008.1</td>\n",
       "      <td>14236</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>18067</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451009.1</td>\n",
       "      <td>17581</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>19105</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451010.1</td>\n",
       "      <td>18271</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>21338</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451011.1</td>\n",
       "      <td>20633</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>25698</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451012.2</td>\n",
       "      <td>22023</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>27238</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451013.1</td>\n",
       "      <td>26581</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>29091</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451014.1</td>\n",
       "      <td>28197</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>32755</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451015.1</td>\n",
       "      <td>29716</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>35012</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451016.1</td>\n",
       "      <td>33083</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>38201</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451017.1</td>\n",
       "      <td>37133</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>39827</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451018.1</td>\n",
       "      <td>38765</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>41961</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451019.1</td>\n",
       "      <td>40329</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>44488</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451020.1</td>\n",
       "      <td>43948</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>48276</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451021.1</td>\n",
       "      <td>44586</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>49774</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_002999345.1</td>\n",
       "      <td>49432</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>51241</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451022.1</td>\n",
       "      <td>50098</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>53081</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451023.1</td>\n",
       "      <td>51264</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>58550</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451024.1</td>\n",
       "      <td>53303</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>60105</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451025.2</td>\n",
       "      <td>59179</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>63845</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451026.1</td>\n",
       "      <td>60377</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>65971</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451027.1</td>\n",
       "      <td>64396</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>69811</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451028.1</td>\n",
       "      <td>67684</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>71636</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451030.1</td>\n",
       "      <td>69860</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>73981</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451031.1</td>\n",
       "      <td>72232</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>75329</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451032.1</td>\n",
       "      <td>74186</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298293</th>\n",
       "      <td>3376197</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863017.1</td>\n",
       "      <td>3374820</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298294</th>\n",
       "      <td>3377091</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863016.1</td>\n",
       "      <td>3376227</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298295</th>\n",
       "      <td>3378684</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863015.1</td>\n",
       "      <td>3377142</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298296</th>\n",
       "      <td>3379230</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863014.1</td>\n",
       "      <td>3378696</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298297</th>\n",
       "      <td>3379715</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863013.1</td>\n",
       "      <td>3379244</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298298</th>\n",
       "      <td>3380010</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863012.1</td>\n",
       "      <td>3379752</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298299</th>\n",
       "      <td>3380867</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863011.1</td>\n",
       "      <td>3380066</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298300</th>\n",
       "      <td>3381256</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863010.1</td>\n",
       "      <td>3380875</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298301</th>\n",
       "      <td>3382601</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863009.1</td>\n",
       "      <td>3381974</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298302</th>\n",
       "      <td>3384522</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978535.1</td>\n",
       "      <td>3382632</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298303</th>\n",
       "      <td>3385371</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978372.1</td>\n",
       "      <td>3384924</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298304</th>\n",
       "      <td>3387910</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978373.1</td>\n",
       "      <td>3385858</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298305</th>\n",
       "      <td>3389703</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978374.1</td>\n",
       "      <td>3388383</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298306</th>\n",
       "      <td>3389957</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_017410478.1</td>\n",
       "      <td>3389702</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298307</th>\n",
       "      <td>3391113</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978536.1</td>\n",
       "      <td>3390495</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298308</th>\n",
       "      <td>3391433</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978375.1</td>\n",
       "      <td>3391109</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298309</th>\n",
       "      <td>3392282</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978376.1</td>\n",
       "      <td>3392030</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298310</th>\n",
       "      <td>3392821</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978377.1</td>\n",
       "      <td>3392596</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298311</th>\n",
       "      <td>3393303</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978378.1</td>\n",
       "      <td>3393066</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298312</th>\n",
       "      <td>3394300</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978379.1</td>\n",
       "      <td>3393337</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298313</th>\n",
       "      <td>3396317</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978380.1</td>\n",
       "      <td>3394301</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298314</th>\n",
       "      <td>3397705</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978381.1</td>\n",
       "      <td>3396316</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298315</th>\n",
       "      <td>3398233</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978382.1</td>\n",
       "      <td>3397897</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298316</th>\n",
       "      <td>3399092</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978537.1</td>\n",
       "      <td>3398387</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298317</th>\n",
       "      <td>3399918</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978383.1</td>\n",
       "      <td>3399117</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298318</th>\n",
       "      <td>3401259</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978384.1</td>\n",
       "      <td>3400119</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298319</th>\n",
       "      <td>3402780</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863005.1</td>\n",
       "      <td>3401421</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298320</th>\n",
       "      <td>3404471</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863004.1</td>\n",
       "      <td>3402848</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298321</th>\n",
       "      <td>3405054</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863002.1</td>\n",
       "      <td>3404697</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298322</th>\n",
       "      <td>3405693</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978385.1</td>\n",
       "      <td>3405408</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5298323 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             end genome_accession refSeqProteinId    start  strand\n",
       "0           3834  GCF_000002515.2     XP_451004.1     2787       1\n",
       "1           6151  GCF_000002515.2     XP_451005.1     4627       1\n",
       "2           7056  GCF_000002515.2  XP_002999344.1     6837      -1\n",
       "3           8579  GCF_000002515.2     XP_451006.2     8138       1\n",
       "4          11511  GCF_000002515.2     XP_451007.1    10101      -1\n",
       "5          15211  GCF_000002515.2     XP_451008.1    14236      -1\n",
       "6          18067  GCF_000002515.2     XP_451009.1    17581       1\n",
       "7          19105  GCF_000002515.2     XP_451010.1    18271      -1\n",
       "8          21338  GCF_000002515.2     XP_451011.1    20633       1\n",
       "9          25698  GCF_000002515.2     XP_451012.2    22023      -1\n",
       "10         27238  GCF_000002515.2     XP_451013.1    26581      -1\n",
       "11         29091  GCF_000002515.2     XP_451014.1    28197       1\n",
       "12         32755  GCF_000002515.2     XP_451015.1    29716       1\n",
       "13         35012  GCF_000002515.2     XP_451016.1    33083      -1\n",
       "14         38201  GCF_000002515.2     XP_451017.1    37133      -1\n",
       "15         39827  GCF_000002515.2     XP_451018.1    38765       1\n",
       "16         41961  GCF_000002515.2     XP_451019.1    40329      -1\n",
       "17         44488  GCF_000002515.2     XP_451020.1    43948       1\n",
       "18         48276  GCF_000002515.2     XP_451021.1    44586      -1\n",
       "19         49774  GCF_000002515.2  XP_002999345.1    49432      -1\n",
       "20         51241  GCF_000002515.2     XP_451022.1    50098       1\n",
       "21         53081  GCF_000002515.2     XP_451023.1    51264      -1\n",
       "22         58550  GCF_000002515.2     XP_451024.1    53303      -1\n",
       "23         60105  GCF_000002515.2     XP_451025.2    59179       1\n",
       "24         63845  GCF_000002515.2     XP_451026.1    60377      -1\n",
       "25         65971  GCF_000002515.2     XP_451027.1    64396      -1\n",
       "26         69811  GCF_000002515.2     XP_451028.1    67684       1\n",
       "27         71636  GCF_000002515.2     XP_451030.1    69860      -1\n",
       "28         73981  GCF_000002515.2     XP_451031.1    72232       1\n",
       "29         75329  GCF_000002515.2     XP_451032.1    74186       1\n",
       "...          ...              ...             ...      ...     ...\n",
       "5298293  3376197  GCF_900087055.1  WP_010863017.1  3374820      -1\n",
       "5298294  3377091  GCF_900087055.1  WP_010863016.1  3376227      -1\n",
       "5298295  3378684  GCF_900087055.1  WP_010863015.1  3377142      -1\n",
       "5298296  3379230  GCF_900087055.1  WP_010863014.1  3378696      -1\n",
       "5298297  3379715  GCF_900087055.1  WP_010863013.1  3379244      -1\n",
       "5298298  3380010  GCF_900087055.1  WP_010863012.1  3379752      -1\n",
       "5298299  3380867  GCF_900087055.1  WP_010863011.1  3380066      -1\n",
       "5298300  3381256  GCF_900087055.1  WP_010863010.1  3380875      -1\n",
       "5298301  3382601  GCF_900087055.1  WP_010863009.1  3381974      -1\n",
       "5298302  3384522  GCF_900087055.1  WP_064978535.1  3382632      -1\n",
       "5298303  3385371  GCF_900087055.1  WP_064978372.1  3384924      -1\n",
       "5298304  3387910  GCF_900087055.1  WP_064978373.1  3385858       1\n",
       "5298305  3389703  GCF_900087055.1  WP_064978374.1  3388383      -1\n",
       "5298306  3389957  GCF_900087055.1  WP_017410478.1  3389702      -1\n",
       "5298307  3391113  GCF_900087055.1  WP_064978536.1  3390495       1\n",
       "5298308  3391433  GCF_900087055.1  WP_064978375.1  3391109      -1\n",
       "5298309  3392282  GCF_900087055.1  WP_064978376.1  3392030      -1\n",
       "5298310  3392821  GCF_900087055.1  WP_064978377.1  3392596       1\n",
       "5298311  3393303  GCF_900087055.1  WP_064978378.1  3393066       1\n",
       "5298312  3394300  GCF_900087055.1  WP_064978379.1  3393337       1\n",
       "5298313  3396317  GCF_900087055.1  WP_064978380.1  3394301       1\n",
       "5298314  3397705  GCF_900087055.1  WP_064978381.1  3396316       1\n",
       "5298315  3398233  GCF_900087055.1  WP_064978382.1  3397897       1\n",
       "5298316  3399092  GCF_900087055.1  WP_064978537.1  3398387       1\n",
       "5298317  3399918  GCF_900087055.1  WP_064978383.1  3399117       1\n",
       "5298318  3401259  GCF_900087055.1  WP_064978384.1  3400119      -1\n",
       "5298319  3402780  GCF_900087055.1  WP_010863005.1  3401421      -1\n",
       "5298320  3404471  GCF_900087055.1  WP_010863004.1  3402848      -1\n",
       "5298321  3405054  GCF_900087055.1  WP_010863002.1  3404697      -1\n",
       "5298322  3405693  GCF_900087055.1  WP_064978385.1  3405408      -1\n",
       "\n",
       "[5298323 rows x 5 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomesCompressedPath = refSeqPath / 'Genomes_compressed'\n",
    "genomesExtractedPath = refSeqPath / 'Genomes_extracted'\n",
    "\n",
    "dfList = []\n",
    "\n",
    "for genome_accession in assemblySummaryRepDf.index:\n",
    "    print(genome_accession)\n",
    "    compressedGenomeFilename, species_name, genome_accession = \\\n",
    "        import_genome_gbff_file(assemblySummaryRepDf, genome_accession=genome_accession)\n",
    "    genomeFilename = re.match(r'(.+)\\.gz', compressedGenomeFilename).group(1)\n",
    "    genomeBio = next(Bio.SeqIO.parse(str(genomesExtractedPath / genomeFilename), format=\"genbank\"))\n",
    "\n",
    "    verbose = 1\n",
    "    for feature in genomeBio.features:\n",
    "        if feature.type == 'CDS':\n",
    "            if verbose >= 2:\n",
    "                print(feature.qualifiers,'\\n\\n',feature.type,'\\n',feature.location,'\\n\\n',\n",
    "                      feature.extract(genomeBio),'\\n')\n",
    "\n",
    "            refSeqProteinId = feature.qualifiers.get('protein_id')\n",
    "            refSeqProteinId = refSeqProteinId[0] if refSeqProteinId is not None else None\n",
    "\n",
    "            if refSeqProteinId is None:\n",
    "                None\n",
    "            else:\n",
    "                dfList.append({'refSeqProteinId':refSeqProteinId, 'genome_accession':genome_accession,\n",
    "                 'strand':feature.location.strand, 'start':feature.location.start, 'end':feature.location.end})\n",
    "\n",
    "locationDf = pd.DataFrame(dfList)\n",
    "locationDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "locationDf.to_csv(analysisCtermDataPath / 'locationDf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/numpy/lib/arraysetops.py:472: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    }
   ],
   "source": [
    "locationDf = pd.read_csv(analysisCtermDataPath / 'locationDf.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3834</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451004.1</td>\n",
       "      <td>2787</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6151</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451005.1</td>\n",
       "      <td>4627</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7056</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_002999344.1</td>\n",
       "      <td>6837</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8579</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451006.2</td>\n",
       "      <td>8138</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11511</td>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>XP_451007.1</td>\n",
       "      <td>10101</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     end genome_accession refSeqProteinId  start  strand\n",
       "0   3834  GCF_000002515.2     XP_451004.1   2787       1\n",
       "1   6151  GCF_000002515.2     XP_451005.1   4627       1\n",
       "2   7056  GCF_000002515.2  XP_002999344.1   6837      -1\n",
       "3   8579  GCF_000002515.2     XP_451006.2   8138       1\n",
       "4  11511  GCF_000002515.2     XP_451007.1  10101      -1"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "locationDf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "does the coordinates include the stop codon? Yes it does.\n",
    "\n",
    "then the overlapping NNA-TGA-NNN cases will be those for which the\n",
    "\n",
    "`start2 == end1 - 4`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: CDS\n",
      "location: [2383:3460](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GI:1039817707']\n",
      "    Key: inference, Value: ['EXISTENCE: similar to AA sequence:RefSeq:WP_006121376.1']\n",
      "    Key: locus_tag, Value: ['BAY29_RS00015']\n",
      "    Key: note, Value: ['Derived by automated computational analysis using gene prediction method: Protein Homology.']\n",
      "    Key: old_locus_tag, Value: ['SAMEA2665130_00003']\n",
      "    Key: product, Value: ['DNA replication/repair protein RecF']\n",
      "    Key: protein_id, Value: ['WP_064977015.1']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ['MALSRLLVQDFRNIAQADLAFDSGFNFLVGANGSGKTSLLEAIYYLGHGRSFKSQLAGRVIQQDKPHFILHGRLDEGGHPVSLGLQKTRQGDTRVRIDGQDGQKVAALAQLLPMQLITPEGFTLLSGGPKYRRAFLDWGLFHDEPGFYAAWSNLRRLLKQRNAALKQVRRYAELRPWDRELVQLALQVSEWRAQYAAAIAQEIQQTCALFLPEIPVSVSFHRGWEKGADYAQVLERDFERDRALTYTAVGPHKADFRFRAAGTPVEDVLSRGQLKLLMCALRLAQGQYLTQRSARQCIYLIDDFASELDRGRRQLLAEKLKQTGAQVFVSAISPDQVSDMTGEKGRMFRVEHGKISAQ']\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Seq('CAATGA', IUPACAmbiguousDNA())"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feat = genomeBio.features[6]\n",
    "print(feat)\n",
    "feat.extract(genomeBio).seq[-6:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "strand = +1\n",
    "df2 = df[df['strand'] == strand]\n",
    "# gene1_stop_codon overlapping gene2_start_codon\n",
    "gene1_stop_codon_end_plus = set(df2['start'] + 4) & set(df2['end'])\n",
    "gene1_stop_codon = df2[df2['end'].isin(gene1_stop_codon_end_plus)]\n",
    "gene1_stop_codonDf = gene1_stop_codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5298076</th>\n",
       "      <td>3099832</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010864837.1</td>\n",
       "      <td>3099520</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298137</th>\n",
       "      <td>3174095</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978526.1</td>\n",
       "      <td>3173153</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298139</th>\n",
       "      <td>3176402</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010864906.1</td>\n",
       "      <td>3175295</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298143</th>\n",
       "      <td>3181032</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978295.1</td>\n",
       "      <td>3179901</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298153</th>\n",
       "      <td>3189887</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978300.1</td>\n",
       "      <td>3188663</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             end genome_accession refSeqProteinId    start  strand\n",
       "5298076  3099832  GCF_900087055.1  WP_010864837.1  3099520       1\n",
       "5298137  3174095  GCF_900087055.1  WP_064978526.1  3173153       1\n",
       "5298139  3176402  GCF_900087055.1  WP_010864906.1  3175295       1\n",
       "5298143  3181032  GCF_900087055.1  WP_064978295.1  3179901       1\n",
       "5298153  3189887  GCF_900087055.1  WP_064978300.1  3188663       1"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene1_stop_codonDf.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "strand = -1\n",
    "df2 = df[df['strand'] == strand]\n",
    "# gene1_stop_codon overlapping gene2_start_codon\n",
    "gene1_stop_codon_start_minus = set(df2['start']) & set(df2['end'] - 4)\n",
    "gene1_stop_codon = df2[df2['start'].isin(gene1_stop_codon_start_minus)]\n",
    "gene1_stop_codonDf = pd.concat([gene1_stop_codonDf, gene1_stop_codon])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5298247</th>\n",
       "      <td>3319111</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978348.1</td>\n",
       "      <td>3317848</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298248</th>\n",
       "      <td>3320241</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863063.1</td>\n",
       "      <td>3319107</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298259</th>\n",
       "      <td>3330958</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978353.1</td>\n",
       "      <td>3330361</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298262</th>\n",
       "      <td>3333558</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978356.1</td>\n",
       "      <td>3332952</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298268</th>\n",
       "      <td>3339944</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_036769360.1</td>\n",
       "      <td>3339182</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             end genome_accession refSeqProteinId    start  strand\n",
       "5298247  3319111  GCF_900087055.1  WP_064978348.1  3317848      -1\n",
       "5298248  3320241  GCF_900087055.1  WP_010863063.1  3319107      -1\n",
       "5298259  3330958  GCF_900087055.1  WP_064978353.1  3330361      -1\n",
       "5298262  3333558  GCF_900087055.1  WP_064978356.1  3332952      -1\n",
       "5298268  3339944  GCF_900087055.1  WP_036769360.1  3339182      -1"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene1_stop_codonDf.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For all genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfList = []\n",
    "\n",
    "for genome_accession, df in locationDf.groupby('genome_accession'):\n",
    "\n",
    "    strand = +1\n",
    "    df2 = df[df['strand'] == strand]\n",
    "    # gene1_stop_codon overlapping gene2_start_codon\n",
    "    gene1_stop_codon_end_plus = set(df2['start'] + 4) & set(df2['end'])\n",
    "    gene1_stop_codon1 = df2[df2['end'].isin(gene1_stop_codon_end_plus)]\n",
    "\n",
    "    strand = -1\n",
    "    df2 = df[df['strand'] == strand]\n",
    "    # gene1_stop_codon overlapping gene2_start_codon\n",
    "    gene1_stop_codon_start_minus = set(df2['start']) & set(df2['end'] - 4)\n",
    "    gene1_stop_codon2 = df2[df2['start'].isin(gene1_stop_codon_start_minus)]\n",
    "    \n",
    "    gene1_stop_codonDf = pd.concat([gene1_stop_codon1, gene1_stop_codon2])\n",
    "\n",
    "    gene1_stop_codonDf['genome_accession'] = genome_accession\n",
    "    dfList.append(gene1_stop_codonDf)\n",
    "    \n",
    "gene1_stop_codonDf = pd.concat(dfList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1066</th>\n",
       "      <td>2085377</td>\n",
       "      <td>GCF_000002985.6</td>\n",
       "      <td>NP_001293346.1</td>\n",
       "      <td>2084485</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3473</th>\n",
       "      <td>10391316</td>\n",
       "      <td>GCF_000002985.6</td>\n",
       "      <td>NP_001300397.1</td>\n",
       "      <td>10384218</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3474</th>\n",
       "      <td>10391316</td>\n",
       "      <td>GCF_000002985.6</td>\n",
       "      <td>NP_492720.3</td>\n",
       "      <td>10384218</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           end genome_accession refSeqProteinId     start  strand\n",
       "1066   2085377  GCF_000002985.6  NP_001293346.1   2084485       1\n",
       "3473  10391316  GCF_000002985.6  NP_001300397.1  10384218      -1\n",
       "3474  10391316  GCF_000002985.6     NP_492720.3  10384218      -1"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene1_stop_codonDf.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5298067</th>\n",
       "      <td>3073521</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978263.1</td>\n",
       "      <td>3072480</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298241</th>\n",
       "      <td>3312827</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978344.1</td>\n",
       "      <td>3311714</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298242</th>\n",
       "      <td>3314074</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978345.1</td>\n",
       "      <td>3312823</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298247</th>\n",
       "      <td>3319111</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978348.1</td>\n",
       "      <td>3317848</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298248</th>\n",
       "      <td>3320241</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_010863063.1</td>\n",
       "      <td>3319107</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298259</th>\n",
       "      <td>3330958</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978353.1</td>\n",
       "      <td>3330361</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298262</th>\n",
       "      <td>3333558</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_064978356.1</td>\n",
       "      <td>3332952</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298268</th>\n",
       "      <td>3339944</td>\n",
       "      <td>GCF_900087055.1</td>\n",
       "      <td>WP_036769360.1</td>\n",
       "      <td>3339182</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             end genome_accession refSeqProteinId    start  strand\n",
       "5298067  3073521  GCF_900087055.1  WP_064978263.1  3072480      -1\n",
       "5298241  3312827  GCF_900087055.1  WP_064978344.1  3311714      -1\n",
       "5298242  3314074  GCF_900087055.1  WP_064978345.1  3312823      -1\n",
       "5298247  3319111  GCF_900087055.1  WP_064978348.1  3317848      -1\n",
       "5298248  3320241  GCF_900087055.1  WP_010863063.1  3319107      -1\n",
       "5298259  3330958  GCF_900087055.1  WP_064978353.1  3330361      -1\n",
       "5298262  3333558  GCF_900087055.1  WP_064978356.1  3332952      -1\n",
       "5298268  3339944  GCF_900087055.1  WP_036769360.1  3339182      -1"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene1_stop_codonDf.tail(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1641942</th>\n",
       "      <td>1834</td>\n",
       "      <td>GCF_000027345.1</td>\n",
       "      <td>NP_109689.1</td>\n",
       "      <td>691</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641943</th>\n",
       "      <td>2767</td>\n",
       "      <td>GCF_000027345.1</td>\n",
       "      <td>NP_109690.1</td>\n",
       "      <td>1837</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641944</th>\n",
       "      <td>4821</td>\n",
       "      <td>GCF_000027345.1</td>\n",
       "      <td>NP_109691.1</td>\n",
       "      <td>2868</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641945</th>\n",
       "      <td>7340</td>\n",
       "      <td>GCF_000027345.1</td>\n",
       "      <td>NP_109692.1</td>\n",
       "      <td>4820</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641946</th>\n",
       "      <td>8574</td>\n",
       "      <td>GCF_000027345.1</td>\n",
       "      <td>NP_109693.1</td>\n",
       "      <td>7311</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          end genome_accession refSeqProteinId  start  strand\n",
       "1641942  1834  GCF_000027345.1     NP_109689.1    691       1\n",
       "1641943  2767  GCF_000027345.1     NP_109690.1   1837       1\n",
       "1641944  4821  GCF_000027345.1     NP_109691.1   2868       1\n",
       "1641945  7340  GCF_000027345.1     NP_109692.1   4820       1\n",
       "1641946  8574  GCF_000027345.1     NP_109693.1   7311       1"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_accession = 'GCF_000027345.1'\n",
    "df = locationDf[locationDf['genome_accession'] == genome_accession]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [end, genome_accession, refSeqProteinId, start, strand]\n",
       "Index: []"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfStop = gene1_stop_codonDf[gene1_stop_codonDf['genome_accession'] == genome_accession]\n",
    "dfStop.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome_accession\n",
       "GCF_000002985.6      3\n",
       "GCF_000005825.2    147\n",
       "GCF_000005845.2    253\n",
       "GCF_000006175.1     13\n",
       "GCF_000006605.1    119\n",
       "GCF_000006685.1     45\n",
       "GCF_000006725.1    238\n",
       "GCF_000006745.1    119\n",
       "GCF_000006765.1    484\n",
       "GCF_000006785.2     83\n",
       "GCF_000006845.1     59\n",
       "GCF_000006865.1    114\n",
       "GCF_000006905.1    325\n",
       "GCF_000006925.2    211\n",
       "GCF_000006945.1    296\n",
       "GCF_000006965.1    263\n",
       "GCF_000006985.1    153\n",
       "GCF_000007005.1    194\n",
       "GCF_000007025.1     45\n",
       "GCF_000007045.1     85\n",
       "GCF_000007085.1    116\n",
       "GCF_000007125.1    132\n",
       "GCF_000007145.1    440\n",
       "GCF_000007265.1    104\n",
       "GCF_000007305.1    180\n",
       "GCF_000007325.1     63\n",
       "GCF_000007345.1    188\n",
       "GCF_000007365.1     13\n",
       "GCF_000007385.1    388\n",
       "GCF_000007465.2     96\n",
       "                  ... \n",
       "GCF_001583415.1    664\n",
       "GCF_001584185.1    270\n",
       "GCF_001586165.1    274\n",
       "GCF_001586195.1    161\n",
       "GCF_001586215.1    130\n",
       "GCF_001586255.1    209\n",
       "GCF_001593245.1    301\n",
       "GCF_001597285.1    357\n",
       "GCF_001606005.1    345\n",
       "GCF_001606025.1     44\n",
       "GCF_001610975.1    423\n",
       "GCF_001611135.1     41\n",
       "GCF_001618685.1    204\n",
       "GCF_001620305.1    593\n",
       "GCF_001632865.1    654\n",
       "GCF_001636015.1     64\n",
       "GCF_001641285.1    377\n",
       "GCF_001642655.1     68\n",
       "GCF_001644565.1    227\n",
       "GCF_001648175.1    351\n",
       "GCF_001652465.1    404\n",
       "GCF_001652485.1    264\n",
       "GCF_001652565.1    381\n",
       "GCF_001653755.1     79\n",
       "GCF_001654495.1    578\n",
       "GCF_001655245.1    757\n",
       "GCF_900011245.1    464\n",
       "GCF_900078695.1    370\n",
       "GCF_900086555.1     66\n",
       "GCF_900087055.1    122\n",
       "Name: start, Length: 1709, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene1_stop_codonDf.groupby('genome_accession')['start'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2575"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene1_stop_codonDf.duplicated(subset=['genome_accession', 'refSeqProteinId']).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "374988"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene1_stop_codonDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene1_stop_codonDf2 = gene1_stop_codonDf.set_index(['genome_accession', 'refSeqProteinId'])\n",
    "gene1_stop_codonDf2 = gene1_stop_codonDf2.loc[~gene1_stop_codonDf2.index.duplicated()]\n",
    "gene1_stop_codonDf2.index.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multispeciesAllCDSDf_nr2 = multispeciesAllCDSDf_nr.set_index(['genome_accession', 'refSeqProteinId'])\n",
    "multispeciesAllCDSDf_nr2.index.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4871983\n",
      "4523129\n"
     ]
    }
   ],
   "source": [
    "print(len(multispeciesAllCDSDf_nr2))\n",
    "multispeciesAllCDSDf_nr3 = multispeciesAllCDSDf_nr2.loc[\n",
    "    multispeciesAllCDSDf_nr2.index.difference(gene1_stop_codonDf2.index)]\n",
    "print(len(multispeciesAllCDSDf_nr3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.07160410863502603"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(len(multispeciesAllCDSDf_nr3) - len(multispeciesAllCDSDf_nr2))/len(multispeciesAllCDSDf_nr2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "overall, the 4nt overlapping stop codons are 7.16% of the total of all CDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1836672\n",
      "1836105\n",
      "0.00030871053731967385\n"
     ]
    }
   ],
   "source": [
    "df = multispeciesAllCDSDf_nr2[multispeciesAllCDSDf_nr2['stopCodon'] == 'TAA']\n",
    "n_tot = len(df)\n",
    "n_nonoverlap = len(df.loc[df.index.difference(gene1_stop_codonDf2.index)])\n",
    "print(n_tot)\n",
    "print(n_nonoverlap)\n",
    "print((n_tot - n_nonoverlap)/n_tot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "936585\n",
      "936376\n",
      "0.00022315112883507636\n"
     ]
    }
   ],
   "source": [
    "df = multispeciesAllCDSDf_nr2[multispeciesAllCDSDf_nr2['stopCodon'] == 'TAG']\n",
    "n_tot = len(df)\n",
    "n_nonoverlap = len(df.loc[df.index.difference(gene1_stop_codonDf2.index)])\n",
    "print(n_tot)\n",
    "print(n_nonoverlap)\n",
    "print((n_tot - n_nonoverlap)/n_tot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2098684\n",
      "1750606\n",
      "0.16585536459991118\n"
     ]
    }
   ],
   "source": [
    "df = multispeciesAllCDSDf_nr2[multispeciesAllCDSDf_nr2['stopCodon'] == 'TGA']\n",
    "n_tot = len(df)\n",
    "n_nonoverlap = len(df.loc[df.index.difference(gene1_stop_codonDf2.index)])\n",
    "print(n_tot)\n",
    "print(n_nonoverlap)\n",
    "print((n_tot - n_nonoverlap)/n_tot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1750606\n"
     ]
    }
   ],
   "source": [
    "df = multispeciesAllCDSDf_nr2[multispeciesAllCDSDf_nr2['stopCodon'] == 'TGA']\n",
    "multispeciesAllCDSDf_nr3 = df.loc[df.index.difference(gene1_stop_codonDf2.index)]\n",
    "print(len(multispeciesAllCDSDf_nr3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Analysis_data/C-terminal/Data/Taxonomy/taxaId_2/taxaId_2_multispeciesAllCDSDf_nr3_filtered_overlapping_start_codons.csv\n"
     ]
    }
   ],
   "source": [
    "filePath = analysisCtermDataPath / 'Taxonomy' / taxonSuffix / (taxonSuffix + '_multispeciesAllCDSDf_nr3_filtered_overlapping_start_codons.csv')\n",
    "print(filePath)\n",
    "multispeciesAllCDSDf_nr3.to_csv(str(filePath))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Analysis_data/C-terminal/Data/Taxonomy/taxaId_2/taxaId_2_multispeciesAllCDSDf_nr3_filtered_overlapping_start_codons.csv.gz\n"
     ]
    }
   ],
   "source": [
    "filePath = analysisCtermDataPath / 'Taxonomy' / taxonSuffix / (taxonSuffix + '_multispeciesAllCDSDf_nr3_filtered_overlapping_start_codons.csv.gz')\n",
    "print(filePath)\n",
    "multispeciesAllCDSDf_nr3 = pd.read_csv(str(filePath), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1750606\n"
     ]
    }
   ],
   "source": [
    "print(len(multispeciesAllCDSDf_nr3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "multispeciesAllCDSDf_nr4 = multispeciesAllCDSDf_nr3.sample(frac=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "175061\n"
     ]
    }
   ],
   "source": [
    "print(len(multispeciesAllCDSDf_nr4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "version 1.2\n",
      "stopCodonList ['TAA', 'TAG', 'TGA']\n",
      "Taking as input the multispeciesAllCDSDf_nr dataframe. Skipping first part of analysis.\n",
      "len(multispeciesAllCDSDf_nr): 175061\n",
      "statisticsNameList: ['all', 'stopCodon_TAA', 'stopCodon_TAG', 'stopCodon_TGA']\n",
      "\n",
      "multispeciesStatisticsSummary:      genome_accession                                       species_name  nSeq\n",
      "0     GCF_000005825.2                          Bacillus pseudofirmus OF4    42\n",
      "1     GCF_000005845.2          Escherichia coli str. K-12 substr. MG1655    89\n",
      "2     GCF_000006605.1                      Corynebacterium jeikeium K411    19\n",
      "3     GCF_000006685.1                      Chlamydia muridarum str. Nigg    10\n",
      "4     GCF_000006725.1                            Xylella fastidiosa 9a5c   122\n",
      "5     GCF_000006745.1       Vibrio cholerae O1 biovar El Tor str. N16961    37\n",
      "6     GCF_000006765.1                        Pseudomonas aeruginosa PAO1   367\n",
      "7     GCF_000006785.2                      Streptococcus pyogenes M1 GAS    14\n",
      "8     GCF_000006845.1                      Neisseria gonorrhoeae FA 1090    56\n",
      "9     GCF_000006865.1            Lactococcus lactis subsp. lactis Il1403    27\n",
      "10    GCF_000006905.1                        Caulobacter crescentus CB15   139\n",
      "11    GCF_000006925.2                      Shigella flexneri 2a str. 301    73\n",
      "12    GCF_000006945.1  Salmonella enterica subsp. enterica serovar Ty...    99\n",
      "13    GCF_000006965.1                        Sinorhizobium meliloti 1021   199\n",
      "14    GCF_000006985.1                             Chlorobium tepidum TLS   111\n",
      "15    GCF_000007025.1                   Rickettsia conorii str. Malish 7    13\n",
      "16    GCF_000007045.1                        Streptococcus pneumoniae R6    16\n",
      "17    GCF_000007085.1  Caldanaerobacter subterraneus subsp. tengconge...    63\n",
      "18    GCF_000007125.1                 Brucella melitensis bv. 1 str. 16M   104\n",
      "19    GCF_000007145.1  Xanthomonas campestris pv. campestris str. ATC...   225\n",
      "20    GCF_000007265.1                   Streptococcus agalactiae 2603V/R    23\n",
      "21    GCF_000007325.1  Fusobacterium nucleatum subsp. nucleatum ATCC ...     8\n",
      "22    GCF_000007365.1  Buchnera aphidicola str. Sg (Schizaphis graminum)     8\n",
      "23    GCF_000007385.1           Xanthomonas oryzae pv. oryzae KACC 10331   190\n",
      "24    GCF_000007465.2                         Streptococcus mutans UA159    29\n",
      "25    GCF_000007485.1                     Tropheryma whipplei str. Twist    26\n",
      "26    GCF_000007525.1                     Bifidobacterium longum NCC2705    83\n",
      "27    GCF_000007565.2                          Pseudomonas putida KT2440   335\n",
      "28    GCF_000007605.1                          Chlamydophila caviae GPIC     2\n",
      "29    GCF_000007625.1                             Clostridium tetani E88    17\n",
      "...               ...                                                ...   ...\n",
      "1498  GCF_001583415.1                                Mycobacterium phlei   270\n",
      "1499  GCF_001584185.1                               Collimonas pratensis   176\n",
      "1500  GCF_001586165.1                            Obesumbacterium proteus    80\n",
      "1501  GCF_001586195.1                            Solibacillus silvestris    24\n",
      "1502  GCF_001586215.1                           Corynebacterium simulans    15\n",
      "1503  GCF_001586255.1                       Zhongshania aliphaticivorans    69\n",
      "1504  GCF_001593245.1                                  Aeromonas veronii   187\n",
      "1505  GCF_001597285.1                            Pseudomonas alcaligenes   221\n",
      "1506  GCF_001606005.1                              Brevibacterium linens   233\n",
      "1507  GCF_001606025.1                         Psychrobacter alimentarius    36\n",
      "1508  GCF_001610975.1                     Sphingopyxis terrae NBRC 15098   194\n",
      "1509  GCF_001611135.1                               Pediococcus damnosus    16\n",
      "1510  GCF_001618685.1                           Geobacillus subterraneus    76\n",
      "1511  GCF_001620305.1                          Rhodococcus fascians D188   271\n",
      "1512  GCF_001632865.1                           Streptomyces ambofaciens   471\n",
      "1513  GCF_001636015.1                                     Moraxella ovis    38\n",
      "1514  GCF_001641285.1                                 Dyella thiooxydans   252\n",
      "1515  GCF_001642655.1                          Clostridium tyrobutyricum    42\n",
      "1516  GCF_001644565.1                               Deinococcus puniceus    95\n",
      "1517  GCF_001648175.1                              Halotalea alkalilenta   261\n",
      "1518  GCF_001652465.1                          Microbacterium chocolatum   192\n",
      "1519  GCF_001652485.1                     Ochrobactrum pseudogrignonense   118\n",
      "1520  GCF_001652565.1                                  Shinella sp. HZN7   286\n",
      "1521  GCF_001653755.1                         Dokdonia donghaensis DSW-1     6\n",
      "1522  GCF_001654495.1                              Streptomyces sp. SAT1   417\n",
      "1523  GCF_001655245.1                         Mycobacterium vaccae 95051   333\n",
      "1524  GCF_900011245.1                                 Bradyrhizobium sp.   345\n",
      "1525  GCF_900078695.1                                Bordetella trematum   237\n",
      "1526  GCF_900086555.1                                  Neisseria weaveri    38\n",
      "1527  GCF_900087055.1                           Plesiomonas shigelloides    68\n",
      "\n",
      "[1528 rows x 3 columns]\n",
      "\n",
      "#### Compute statistics for stat  all , start\n",
      "Total nb of aa:  49997365\n",
      "\n",
      "#### Amino acid frequency bulk:\n",
      "\n",
      "Counter({'A': 5854079, 'L': 5295411, 'G': 4265977, 'V': 3887022, 'R': 3415139, 'E': 2943709, 'D': 2862091, 'S': 2745745, 'T': 2701127, 'P': 2589886, 'I': 2434117, 'F': 1783915, 'Q': 1740147, 'K': 1602123, 'N': 1378609, 'Y': 1262569, 'H': 1101488, 'M': 976293, 'W': 706293, 'C': 451608, 'X': 14, 'U': 3}) \n",
      "\n",
      "[('A', 0.11708775052445264), ('L', 0.10591380165734735), ('G', 0.08532403657672759), ('V', 0.07774453713710712), ('R', 0.06830637974621262), ('E', 0.058877282832805286), ('D', 0.05724483680289951), ('S', 0.05491779416775264), ('T', 0.05402538713790217), ('P', 0.05180044988370887), ('I', 0.0486849056945301), ('F', 0.03568018034550421), ('Q', 0.03480477421160095), ('K', 0.0320441487266379), ('N', 0.027573633130465975), ('Y', 0.0252527108178601), ('H', 0.022030921029538256), ('M', 0.019526889067053833), ('W', 0.014126604472055677), ('C', 0.00903263601991825), ('X', 2.800147567776822e-07), ('U', 6.000316216664618e-08)]\n",
      "ctermFreq.shape:  (20, 175061)\n",
      "ntermFreq.shape:  (20, 175061)\n",
      "Total nb of aa at pos 0:  175061\n",
      "Total nb of aa at pos 0:  175061\n",
      "\n",
      "#### Amino acid frequency c-terminal:\n",
      " Counter({'A': 25403, 'R': 19952, 'G': 19487, 'S': 14875, 'D': 11264, 'L': 9843, 'P': 9418, 'V': 9222, 'E': 7510, 'F': 6373, 'N': 6307, 'H': 5932, 'K': 5858, 'I': 5599, 'T': 5428, 'Y': 3778, 'Q': 3685, 'C': 2110, 'W': 1770, 'M': 1246, 'X': 1})\n",
      "\n",
      "#### Amino acid frequency n-terminal:\n",
      " Counter({'A': 15392, 'T': 15166, 'R': 14677, 'L': 13964, 'S': 12335, 'K': 11911, 'D': 11780, 'E': 11328, 'I': 10517, 'V': 9996, 'G': 8021, 'P': 7791, 'Q': 7043, 'N': 7035, 'F': 5507, 'H': 4187, 'Y': 3542, 'M': 2247, 'W': 1706, 'C': 916})\n",
      "Total nb of codons:  50172426\n",
      "{'GTR', 'TAC', 'CAT', 'GCT', 'SCC', 'SGC', 'CGC', 'GTA', 'AGA', 'CTC', 'AAC', 'TGC', 'TCM', 'CMC', 'GTC', 'TTC', 'TGT', 'GTW', 'NNA', 'GGC', 'GGG', 'CGA', 'GAC', 'GAT', 'CAG', 'GTG', 'ATT', 'TTM', 'TCT', 'TCC', 'WRA', 'TGA', 'CTG', 'ACT', 'ACG', 'GCG', 'ACC', 'ARA', 'ATA', 'CAR', 'GCR', 'TGG', 'AGY', 'CGT', 'GGS', 'CGG', 'GMC', 'CCC', 'ACM', 'TTG', 'GCM', 'TTA', 'CGS', 'CCG', 'CCA', 'CCM', 'AAA', 'TTT', 'ATG', 'CAA', 'GGT', 'GNN', 'GAA', 'CTT', 'TNC', 'AAG', 'ACA', 'TKG', 'AAR', 'CAC', 'AGC', 'CTA', 'SCG', 'GGA', 'AGT', 'GAG', 'AAT', 'TCG', 'AGG', 'TAT', 'GCA', 'GTT', 'CCT', 'ATC', 'RCC', 'GCC', 'TCA'}\n",
      "ctermFreq.shape:  (21, 175061)\n",
      "ntermFreq.shape:  (21, 175061)\n",
      "Total nb of codons at pos 0:  175061\n",
      "Total nb of codons at pos 0:  175061\n",
      "Bulk codon frequency:\n",
      " Counter({'GCC': 2733811, 'CTG': 2668481, 'GGC': 2425045, 'GCG': 2137962, 'GAC': 1914509, 'GAG': 1808581, 'GTG': 1703975, 'ATC': 1671044, 'CGC': 1647485, 'GTC': 1539765, 'ACC': 1443888, 'CAG': 1345521, 'CTC': 1337680, 'CCG': 1331078, 'TTC': 1272142, 'GAA': 1146410, 'AAG': 1023014, 'ATG': 979639, 'GAT': 957811, 'CGG': 916395, 'AAC': 902706, 'CCC': 831267, 'ACG': 829420, 'TAC': 775025, 'AGC': 774660, 'GGG': 771525, 'TCG': 757826, 'CAC': 713295, 'TGG': 708752, 'TCC': 643227, 'GGT': 639971, 'AAA': 585425, 'ATT': 567781, 'GCA': 537959, 'TTG': 532385, 'TTT': 517720, 'TAT': 491795, 'AAT': 480405, 'GCT': 464822, 'GGA': 442941, 'CGT': 405814, 'CTT': 401907, 'CAA': 400966, 'GTT': 394749, 'CAT': 392018, 'TGC': 343568, 'GTA': 261881, 'TTA': 246692, 'CCT': 237536, 'ACA': 231097, 'AGT': 218834, 'ACT': 205837, 'ATA': 203761, 'CCA': 199212, 'TCT': 182422, 'TCA': 178076, 'CGA': 176648, 'AGG': 158911, 'CTA': 126606, 'AGA': 122911, 'TGT': 109808, 'TGA': 3, 'SCC': 2, 'AAR': 1, 'ACM': 1, 'GCR': 1, 'TCM': 1, 'TTM': 1, 'TKG': 1, 'AGY': 1, 'CMC': 1, 'CGS': 1, 'WRA': 1, 'GGS': 1, 'RCC': 1, 'GMC': 1, 'SCG': 1, 'CCM': 1, 'ARA': 1, 'GTR': 1, 'CAR': 1, 'GTW': 1, 'GCM': 1, 'NNA': 1, 'GNN': 1, 'SGC': 1, 'TNC': 1, 'TAA': 0, 'TAG': 0})\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bulk amino acid pairs frequency:\n",
      " Counter({'AA': 791698, 'LA': 673865, 'AL': 651728, 'LL': 550518, 'AG': 524506, 'VA': 456480, 'AV': 456130, 'GA': 445941, 'GL': 442145, 'LG': 436192, 'VL': 418114, 'AR': 407551, 'LV': 399071, 'RL': 390077, 'RA': 371567, 'LR': 365475, 'GG': 364942, 'AE': 362232, 'EA': 343072, 'GV': 336161, 'AD': 335885, 'VV': 335545, 'DA': 332337, 'EL': 325512, 'LS': 322324, 'LD': 317975, 'PA': 315669, 'TA': 311936, 'IA': 301548, 'SA': 301287, 'TL': 301103, 'DL': 300912, 'LT': 298973, 'AS': 297475, 'GR': 293801, 'LP': 293727, 'VG': 293655, 'AT': 292504, 'SL': 290466, 'LE': 288873, 'SG': 277519, 'AP': 268456, 'RR': 268427, 'AI': 263280, 'DG': 262652, 'TG': 262338, 'VR': 262074, 'PL': 254311, 'RV': 250873, 'GE': 246099, 'GT': 243937, 'ER': 243179, 'GS': 242547, 'PG': 242380, 'LI': 240646, 'RG': 238640, 'TV': 234816, 'VT': 230795, 'VE': 230161, 'GD': 228026, 'VD': 227370, 'VS': 224907, 'IG': 222830, 'GI': 222737, 'PV': 220641, 'EV': 219619, 'DV': 218709, 'RE': 215716, 'QA': 213955, 'IL': 209593, 'SV': 209287, 'EG': 208274, 'AQ': 206294, 'VP': 204168, 'FA': 198917, 'DR': 197703, 'VI': 192002, 'RD': 191824, 'AF': 191508, 'PE': 189872, 'QL': 188322, 'IV': 186875, 'DE': 185078, 'PD': 180585, 'LF': 179981, 'RI': 179130, 'EE': 178053, 'GP': 177066, 'DP': 176129, 'RT': 175835, 'RS': 175496, 'SR': 175121, 'LK': 174533, 'KA': 172075, 'RP': 169727, 'FG': 169271, 'GF': 168050, 'LQ': 167658, 'TP': 167487, 'FL': 167386, 'AK': 167331, 'ID': 165705, 'SS': 163499, 'EI': 163181, 'DD': 162588, 'KL': 162462, 'IE': 161938, 'TR': 161786, 'GK': 158092, 'TT': 152523, 'ET': 152448, 'TD': 147940, 'IS': 147230, 'ST': 147052, 'SD': 146977, 'TS': 146534, 'GQ': 145940, 'SE': 144811, 'PR': 144768, 'IR': 144415, 'NL': 144336, 'NA': 144332, 'LN': 143675, 'SP': 143013, 'ED': 141321, 'DI': 141146, 'ES': 140271, 'TE': 140101, 'IT': 139158, 'DT': 138895, 'PS': 138882, 'VF': 138388, 'QV': 138081, 'EP': 137806, 'YL': 136483, 'FV': 135120, 'SI': 134777, 'QR': 133958, 'EQ': 133228, 'AN': 132957, 'RF': 132191, 'PP': 132151, 'QG': 131606, 'RQ': 131149, 'DS': 130741, 'PT': 130425, 'YA': 130109, 'TI': 129852, 'AY': 128757, 'GY': 125334, 'AM': 123241, 'FD': 122695, 'KV': 122476, 'IP': 121396, 'NG': 121281, 'KG': 119829, 'EK': 119062, 'HA': 118297, 'HL': 117998, 'AH': 117741, 'VQ': 116905, 'LY': 116309, 'MA': 116000, 'VK': 115862, 'FS': 115705, 'GN': 115067, 'II': 114260, 'YG': 112132, 'LH': 112075, 'ML': 110451, 'SF': 109656, 'VN': 108989, 'DF': 105408, 'FT': 104898, 'PI': 102474, 'HG': 102347, 'FE': 101994, 'RK': 100861, 'LM': 100326, 'KR': 100107, 'FR': 100085, 'KE': 99994, 'NV': 99536, 'GH': 98532, 'RY': 96845, 'YR': 96618, 'GM': 95198, 'TF': 95104, 'EF': 93619, 'KT': 92459, 'PF': 91889, 'KI': 91600, 'QI': 91187, 'QP': 90813, 'QE': 90787, 'NP': 90685, 'SQ': 90208, 'WL': 89740, 'QS': 89686, 'KS': 89212, 'PQ': 89190, 'KD': 88981, 'YV': 88645, 'RN': 88288, 'QT': 87734, 'DK': 87056, 'DQ': 87030, 'EN': 86741, 'NR': 86509, 'VY': 86288, 'QQ': 86057, 'KP': 85893, 'RH': 85670, 'FI': 85257, 'IK': 84325, 'SK': 82345, 'IF': 81640, 'IN': 81627, 'VH': 80907, 'DY': 80877, 'TQ': 80794, 'HR': 80154, 'AW': 79821, 'QD': 79379, 'VM': 78998, 'MG': 78795, 'FP': 78225, 'HV': 77875, 'KK': 77702, 'NI': 76942, 'YD': 76838, 'MV': 76697, 'SN': 75874, 'HP': 75510, 'SY': 75277, 'NT': 73377, 'ND': 73228, 'EH': 73215, 'GW': 72756, 'NS': 72417, 'PK': 71908, 'FF': 71811, 'DN': 71607, 'YS': 70405, 'NE': 70398, 'IQ': 69739, 'TK': 69627, 'YE': 69386, 'TN': 68590, 'RM': 68217, 'EY': 67357, 'TY': 66403, 'YT': 66400, 'MR': 65459, 'WA': 65440, 'PN': 64365, 'LW': 64118, 'YP': 63655, 'HD': 63607, 'PY': 63356, 'DH': 61903, 'KQ': 61441, 'IY': 60702, 'HE': 59460, 'EM': 59445, 'QK': 58960, 'KN': 58698, 'MT': 57828, 'SH': 57756, 'FN': 57377, 'WR': 56776, 'QF': 56090, 'TH': 56025, 'PH': 55712, 'RW': 55578, 'MP': 55576, 'AC': 54981, 'MS': 54683, 'SM': 54552, 'HS': 54181, 'HT': 53809, 'FK': 53719, 'FQ': 52869, 'YI': 52831, 'YF': 52465, 'MI': 51467, 'NF': 51267, 'FY': 51247, 'DM': 50571, 'ME': 50275, 'CA': 49553, 'IH': 49259, 'YQ': 49233, 'NQ': 49179, 'WG': 49104, 'LC': 49094, 'TM': 49075, 'MD': 49039, 'VW': 48295, 'HI': 48008, 'WV': 47732, 'CG': 47684, 'NK': 47389, 'QN': 47388, 'NN': 47061, 'DW': 46574, 'KF': 46377, 'PM': 45911, 'CL': 43753, 'GC': 43604, 'WS': 43179, 'HF': 42765, 'NY': 41903, 'KY': 41777, 'SW': 41386, 'QY': 41048, 'IM': 40511, 'QH': 39640, 'FH': 39073, 'YK': 38958, 'YY': 38581, 'HQ': 38290, 'WT': 38287, 'YN': 38044, 'PW': 37714, 'EW': 37338, 'VC': 37117, 'TW': 36791, 'MQ': 36322, 'QM': 35974, 'WD': 35137, 'WQ': 34786, 'WP': 33801, 'WE': 33657, 'WI': 33430, 'CV': 33132, 'MK': 32847, 'HY': 32591, 'HH': 32457, 'FM': 32221, 'KH': 32216, 'MF': 31282, 'CR': 31170, 'KM': 30911, 'NH': 29943, 'MN': 29100, 'RC': 29028, 'YH': 28285, 'WF': 27759, 'HK': 27682, 'HN': 27589, 'CD': 26991, 'QW': 26762, 'IW': 26729, 'CS': 26370, 'FW': 25850, 'NM': 25123, 'SC': 24882, 'CE': 24823, 'IC': 24635, 'CP': 24599, 'MM': 24523, 'DC': 24175, 'CT': 23788, 'WN': 23108, 'WK': 22323, 'TC': 22301, 'EC': 20968, 'NW': 20501, 'FC': 20194, 'YM': 20127, 'MH': 20045, 'CI': 19909, 'YW': 19854, 'MY': 18589, 'HW': 18258, 'WH': 18085, 'KW': 18014, 'HM': 17865, 'PC': 17683, 'WY': 17096, 'CF': 16665, 'WM': 15974, 'CQ': 13835, 'YC': 13520, 'WW': 13492, 'NC': 13201, 'CH': 12946, 'HC': 12744, 'QC': 12719, 'CN': 12464, 'CY': 12232, 'CK': 11540, 'MW': 9957, 'KC': 9899, 'CM': 7530, 'WC': 7387, 'MC': 7357, 'CW': 6505, 'CC': 6119, 'XR': 3, 'UH': 3, 'XP': 3, 'LX': 3, 'AX': 3, 'GX': 2, 'XT': 2, 'IX': 1, 'XA': 1, 'FX': 1, 'VU': 1, 'QX': 1, 'XE': 1, 'TX': 1, 'XK': 1, 'NU': 1, 'IU': 1, 'MX': 1, 'XL': 1, 'HX': 1, 'XV': 1, 'XS': 1})\n",
      "Amino acid pairs frequency c-terminal:\n",
      " Counter({'AA': 4608, 'AG': 2621, 'RR': 2397, 'AR': 2268, 'GR': 2189, 'RG': 2138, 'GA': 2088, 'PA': 2081, 'GG': 2009, 'RA': 1935, 'AS': 1897, 'SA': 1883, 'LA': 1752, 'LG': 1673, 'TA': 1561, 'RS': 1537, 'LR': 1490, 'RP': 1412, 'PR': 1391, 'SR': 1366, 'EA': 1365, 'VA': 1363, 'GS': 1349, 'EG': 1299, 'LS': 1276, 'KG': 1255, 'KA': 1250, 'VR': 1243, 'DA': 1232, 'AP': 1229, 'AD': 1225, 'TR': 1102, 'AV': 1094, 'VG': 1076, 'ER': 1022, 'LD': 1016, 'KR': 1016, 'RD': 1013, 'AL': 1012, 'LP': 991, 'SS': 988, 'PG': 975, 'ED': 969, 'ES': 959, 'GL': 952, 'KS': 952, 'QA': 932, 'TG': 893, 'AE': 875, 'VS': 873, 'RL': 863, 'QG': 854, 'RV': 852, 'LL': 851, 'TP': 850, 'GV': 841, 'DD': 841, 'GD': 829, 'GE': 829, 'DR': 811, 'PS': 810, 'AH': 802, 'SG': 802, 'KK': 748, 'LV': 741, 'QR': 715, 'RT': 706, 'TS': 699, 'NA': 697, 'EE': 693, 'GN': 689, 'SL': 686, 'LF': 680, 'DG': 667, 'HA': 667, 'KP': 660, 'NG': 656, 'KD': 655, 'IG': 654, 'HG': 653, 'IA': 640, 'GH': 634, 'EV': 632, 'QS': 627, 'VL': 624, 'PV': 620, 'LN': 616, 'KL': 603, 'TL': 591, 'VD': 588, 'SP': 588, 'RF': 584, 'EL': 583, 'AT': 576, 'RH': 574, 'TV': 571, 'SV': 571, 'QP': 569, 'AN': 566, 'GK': 561, 'DS': 553, 'IR': 553, 'PD': 545, 'LI': 543, 'NR': 538, 'LT': 537, 'KV': 536, 'RI': 533, 'PL': 530, 'GI': 530, 'LK': 529, 'DE': 525, 'KN': 525, 'VV': 523, 'GF': 522, 'GT': 515, 'IS': 515, 'SD': 515, 'TD': 513, 'FA': 507, 'FG': 504, 'SE': 501, 'KE': 489, 'RK': 486, 'NS': 471, 'AF': 471, 'VP': 469, 'AK': 468, 'RN': 465, 'SN': 462, 'SF': 461, 'AI': 461, 'SH': 460, 'LH': 459, 'DL': 451, 'QD': 447, 'AQ': 442, 'PF': 439, 'ND': 439, 'VE': 436, 'EP': 435, 'DV': 432, 'RY': 426, 'ID': 421, 'EN': 415, 'VK': 412, 'RE': 404, 'FS': 401, 'KI': 396, 'VI': 395, 'HR': 394, 'EI': 394, 'PE': 394, 'GP': 393, 'GY': 392, 'TF': 392, 'DF': 385, 'FR': 383, 'LE': 380, 'EK': 380, 'GQ': 379, 'ST': 371, 'TT': 370, 'SI': 369, 'YR': 366, 'FD': 365, 'TH': 363, 'IL': 358, 'QL': 352, 'VN': 349, 'SK': 348, 'KF': 345, 'AY': 338, 'QV': 331, 'EF': 326, 'HD': 324, 'EH': 323, 'TN': 318, 'NV': 318, 'ET': 315, 'HS': 312, 'KH': 312, 'NP': 311, 'VF': 310, 'PT': 307, 'NL': 304, 'VH': 293, 'DI': 290, 'KT': 287, 'LY': 286, 'NE': 283, 'RQ': 283, 'TE': 282, 'YA': 281, 'NN': 278, 'WR': 273, 'IK': 273, 'IP': 270, 'IV': 264, 'HP': 263, 'PH': 262, 'PN': 260, 'QE': 259, 'IE': 258, 'HV': 254, 'QN': 254, 'EQ': 253, 'SY': 249, 'TI': 247, 'VT': 247, 'MR': 247, 'QF': 246, 'SQ': 246, 'TK': 245, 'LQ': 242, 'II': 239, 'PP': 239, 'MA': 237, 'HH': 236, 'VY': 236, 'HE': 236, 'QH': 233, 'IF': 231, 'DN': 231, 'IN': 230, 'FF': 230, 'NF': 229, 'DK': 228, 'PQ': 227, 'FL': 226, 'PI': 226, 'KQ': 225, 'KY': 221, 'DH': 218, 'NK': 218, 'GC': 217, 'RW': 216, 'QK': 213, 'AC': 213, 'YL': 210, 'FV': 209, 'YG': 208, 'HL': 205, 'MG': 203, 'VQ': 202, 'TY': 199, 'NI': 199, 'PK': 191, 'FN': 190, 'FE': 189, 'NT': 189, 'YE': 189, 'CR': 188, 'WL': 188, 'CG': 184, 'DT': 183, 'NH': 182, 'EY': 179, 'YD': 178, 'ML': 178, 'FI': 175, 'MS': 174, 'IT': 174, 'GW': 173, 'QT': 171, 'YF': 171, 'YS': 170, 'YV': 169, 'SC': 168, 'DY': 167, 'QI': 166, 'WA': 166, 'IY': 164, 'WG': 163, 'SW': 163, 'DQ': 159, 'CS': 159, 'TQ': 158, 'QQ': 158, 'CA': 158, 'MP': 154, 'IH': 154, 'PC': 153, 'WS': 153, 'AW': 151, 'PY': 150, 'NY': 147, 'GM': 146, 'IQ': 146, 'NQ': 145, 'QY': 144, 'DP': 142, 'TC': 141, 'WD': 139, 'FK': 139, 'EC': 138, 'MD': 137, 'FH': 136, 'RC': 134, 'RM': 130, 'CP': 128, 'WE': 127, 'HF': 126, 'LC': 126, 'PW': 125, 'MV': 123, 'HY': 121, 'TW': 119, 'LW': 117, 'YP': 116, 'FT': 116, 'VC': 115, 'MK': 115, 'DC': 109, 'MN': 109, 'FP': 108, 'CD': 105, 'IC': 104, 'MI': 103, 'FY': 102, 'YI': 101, 'HN': 100, 'EW': 97, 'HI': 97, 'LM': 96, 'ME': 96, 'WI': 96, 'DW': 94, 'WN': 94, 'AM': 94, 'YY': 93, 'WT': 93, 'WK': 93, 'WP': 91, 'CC': 90, 'FQ': 89, 'HT': 89, 'MF': 89, 'YN': 88, 'SM': 87, 'YQ': 86, 'KM': 86, 'WF': 83, 'VM': 81, 'KC': 80, 'YK': 80, 'CH': 79, 'HK': 78, 'WV': 78, 'MH': 78, 'KW': 76, 'CL': 76, 'YH': 73, 'VW': 72, 'MT': 70, 'EM': 70, 'NC': 70, 'CN': 68, 'CE': 65, 'QW': 65, 'YT': 64, 'HQ': 64, 'FC': 63, 'CQ': 63, 'CV': 63, 'MY': 63, 'WY': 61, 'WH': 61, 'WQ': 61, 'NW': 60, 'IW': 58, 'MQ': 57, 'PM': 56, 'NM': 56, 'TM': 54, 'DM': 53, 'CK': 53, 'CF': 53, 'WW': 52, 'QC': 51, 'HC': 50, 'YC': 48, 'CT': 48, 'QM': 42, 'FM': 42, 'CY': 40, 'CI': 39, 'MM': 38, 'HW': 38, 'HM': 36, 'IM': 35, 'YW': 27, 'FW': 26, 'MW': 26, 'MC': 23, 'YM': 21, 'WC': 17, 'WM': 16, 'CW': 15, 'CM': 7, 'EX': 1})\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bulk frequency hexamers:\n",
      " Counter({'CGCCGC': 315830, 'GGCGGC': 300312, 'GCGGCG': 295725, 'CGGCGC': 284880, 'GCCGCC': 275593, 'CGGCGG': 275401, 'CCGCCG': 271223, 'GCGCCG': 251445, 'CGCCGA': 250404, 'TCGCCG': 248379, 'CCGGCG': 247892, 'GCCGGC': 245942, 'CGCGGC': 245236, 'CGCCGG': 239978, 'GCGCTG': 238077, 'GCCGCG': 234621, 'GCGCGC': 231896, 'CGGCCG': 227272, 'GCTGGC': 222331, 'CCGCGC': 220167, 'TCGGCG': 220027, 'GGCCGC': 219428, 'CGGCGA': 219217, 'GGCGCG': 218856, 'CGCTGG': 210075, 'CTGCTG': 204306, 'CGCGCC': 202719, 'GCGCGG': 202493, 'GCGGCC': 197822, 'CTGGCC': 196585, 'CGCGCG': 193893, 'GGCCGG': 190511, 'TGGCCG': 190357, 'GCCGAC': 188725, 'TGCTGG': 182654, 'CCGGCC': 180928, 'GGCGCC': 178318, 'CGCTGC': 177191, 'CTGCGC': 176041, 'TCGACG': 175297, 'CCCGGC': 172598, 'GCCGAG': 171462, 'CGACGA': 169956, 'GACGGC': 168738, 'CGACGC': 168518, 'CGGGCG': 168460, 'GCGGGC': 168100, 'CTGGCG': 167703, 'CGCGCT': 167095, 'GGCCGA': 166802, 'CGCCCG': 165265, 'GCCCGC': 164897, 'GCTGCT': 164527, 'CTCGGC': 162532, 'CCGACG': 162475, 'ACCGGC': 162438, 'GGCGCT': 161558, 'GCCGGG': 160295, 'ACGCCG': 160212, 'GTGCTG': 159572, 'TGCTGC': 157669, 'ACGGCG': 155871, 'GCTGGA': 155689, 'CCGGGC': 155145, 'GCTGCG': 154298, 'CGACGG': 149811, 'GGCGAC': 149741, 'GCGCGA': 149449, 'TGCGCG': 148223, 'CGGCTG': 147933, 'GGCCTG': 147702, 'GCGGTG': 146036, 'GGCTGG': 145910, 'CCGCGG': 145853, 'GCCCGG': 145386, 'CTCGCC': 144470, 'TGGCGC': 144213, 'GCGACG': 143804, 'TGGCGG': 143636, 'ATCGCC': 143026, 'TCGTCG': 142424, 'GACGCC': 141741, 'CCTGGC': 141332, 'GTCGGC': 141177, 'GGGCGC': 140251, 'CTGGGC': 140214, 'CGGCGT': 140017, 'CGGCCT': 139059, 'CGAGGC': 138925, 'CCGGCA': 138809, 'CGCTCG': 138116, 'GGGCGG': 137600, 'CGACCG': 137421, 'CCGCTG': 136805, 'GCCTGG': 136550, 'CTGCCG': 135340, 'GCGGCA': 135189, 'CCGCGA': 135085, 'GCTGGT': 134988, 'CGGCAC': 134422, 'TCGGCC': 134088, 'CGCGAC': 134000, 'CGGTGC': 133974, 'CTGGTG': 133538, 'TCACCG': 132830, 'GCTCGC': 132393, 'CACCGC': 132035, 'GGCGGG': 131505, 'TGCTCG': 130887, 'ATCGGC': 130693, 'TCGCGC': 130537, 'CCGAGG': 130456, 'GTCGCC': 130390, 'TGCCGG': 130008, 'GCCTGC': 129756, 'CTCGAC': 129263, 'CGTCGC': 128901, 'CACCGG': 128581, 'CGTCGA': 128488, 'CGGTGG': 128344, 'GGTGGC': 128043, 'CGGCCA': 127113, 'GGCCCG': 125946, 'TGCCGC': 125591, 'GCTCGA': 125368, 'CGCCGT': 125046, 'CGCCTG': 124565, 'GCGATC': 124462, 'CCGACC': 124073, 'CGGCAT': 123864, 'GCTCGG': 123791, 'GGCGAG': 123439, 'ACCGCC': 123410, 'CCTGCT': 123299, 'GGGCCG': 123146, 'CCCGCG': 123094, 'CCGCCC': 123010, 'GATCGC': 122598, 'GCACCG': 122592, 'GCCGCT': 122196, 'TCGCGG': 122017, 'CCGGCT': 121674, 'ACCGCG': 121632, 'GCCCTG': 121488, 'AGGCCG': 121467, 'GCTGGG': 121108, 'CGACCT': 121085, 'GGCGGT': 120983, 'CCGGTG': 120761, 'GCCGGT': 120746, 'CGGCAA': 120742, 'GCGGCT': 120472, 'GCTGCC': 120423, 'GACCGC': 120330, 'TGGGCG': 120187, 'CGGCCC': 119786, 'GACCTG': 119751, 'GGTGCT': 119421, 'TGGTGG': 119115, 'GTGGCG': 118992, 'CCCGCC': 118817, 'GACCGG': 118804, 'GCGCTC': 118428, 'TCGACC': 117813, 'GGCGAT': 117629, 'TGATCG': 117247, 'GGCATC': 116758, 'CGAGCG': 116248, 'CGATCG': 115600, 'CGGGCC': 115393, 'GGCTGC': 115281, 'CAGCGC': 115254, 'CTGCTC': 115210, 'GAGCTG': 115138, 'CCGAGC': 115088, 'GCGCCC': 114826, 'GCGACC': 114751, 'CGCGCA': 114750, 'CGCGGG': 114567, 'GCCACC': 114465, 'GCAGCG': 114261, 'GCGCAG': 113930, 'CGCGGT': 113692, 'GCCGGA': 113652, 'CGAGGA': 113423, 'GCATCG': 113014, 'GCCGTC': 112992, 'GGCGTG': 112957, 'CATCGC': 112781, 'CCACCG': 112769, 'CCGTCG': 112567, 'GGCACC': 111514, 'CGGTCG': 111457, 'CACCGA': 111415, 'TCGAGG': 111316, 'TCGGCA': 111303, 'TGGTCG': 111121, 'GACGAC': 110293, 'CATCGA': 110274, 'CTGATC': 110185, 'GCAGGC': 110107, 'GAGGCG': 110073, 'CCTGCG': 109838, 'GCCGCA': 109646, 'GACGCG': 109637, 'GCGTCG': 109212, 'CTGGTC': 108489, 'GAGCGC': 108447, 'TCCGGC': 108325, 'GGGCGA': 107861, 'CCGCCA': 107377, 'GTGGTG': 107158, 'GGCGCA': 106715, 'GTGCGC': 106572, 'GTGCCG': 106265, 'GATCGA': 106078, 'GCTGAT': 105318, 'GCCGTG': 105262, 'GCCGAT': 105250, 'CGCCAC': 105159, 'GTGGCC': 105090, 'GCGCCA': 105066, 'CTGGAC': 104592, 'TCGCCC': 104561, 'CGTCGG': 104033, 'ACGACG': 103657, 'ACCTGG': 103596, 'AGCGGC': 103583, 'GGTCGA': 103472, 'GGCAGC': 103118, 'TTCGGC': 103087, 'GGCGTC': 103023, 'GACGAG': 102811, 'GGGCTG': 102810, 'CGCAGC': 101910, 'ACCTGC': 101853, 'CGTGGC': 101764, 'TCATCG': 101746, 'GCGAGC': 101667, 'GCCAGC': 101561, 'TGGTGC': 101471, 'GGTCGC': 101434, 'CTGGAG': 101316, 'GGTGCG': 101195, 'CCTCGC': 101193, 'CGCACC': 101040, 'GTCGAC': 101025, 'TGCGCC': 100918, 'GTCGCG': 100517, 'AGCGCG': 100450, 'ACGGCC': 100393, 'CGGCTC': 100376, 'CCTGGA': 100161, 'GGCTCG': 100157, 'TGCGGC': 100077, 'CGTGCT': 100071, 'GGCGGA': 99888, 'CGCGAG': 99761, 'ATCGAC': 99646, 'CACGGC': 98830, 'GAGGCC': 98724, 'TGCCCG': 98666, 'GCCTCG': 98424, 'CCCTGG': 98294, 'GCGCCT': 97967, 'CGGCAG': 97941, 'CGCCCT': 97680, 'GCCATC': 97548, 'CCTGCC': 97307, 'CGACGT': 96948, 'GCGAGG': 96926, 'TTCGCC': 96847, 'CGGGCA': 96672, 'AGGCGC': 96103, 'CGCGAT': 96092, 'AGGCGG': 96050, 'GGCAAG': 95934, 'ACGCGC': 95857, 'CGCTGA': 95648, 'GCGGTC': 95601, 'CGGTGA': 95561, 'CCGCCT': 95552, 'AGCTGG': 95477, 'CCTCGG': 95295, 'CGGGCT': 95270, 'CGAGCT': 95031, 'CTGACC': 94434, 'CGATCC': 94185, 'CGATGC': 94074, 'TCGCCA': 93707, 'TCCGCG': 93703, 'TGCCGA': 93179, 'GTCGAG': 93109, 'CCATCG': 93049, 'ATCGCG': 93026, 'TGGACG': 92964, 'GCGATG': 92762, 'CTGCGG': 92731, 'GCGCAC': 92536, 'CAGGCC': 92448, 'GGTGGT': 92248, 'CGGGGC': 92089, 'TCGAGC': 91985, 'CGGACG': 91952, 'GGACGC': 91808, 'ACGGCA': 91668, 'TGACCG': 91345, 'CAGGCG': 91299, 'ACGCTG': 91173, 'GCCGAA': 91140, 'CTCGCG': 91041, 'GACCCG': 90851, 'CCGATC': 90547, 'CCTGGT': 90266, 'CCTCGA': 90259, 'GGCCAG': 89978, 'GCGTGC': 89867, 'GATCGG': 89815, 'GCTGCA': 89757, 'CCGGAC': 89268, 'GCCCGA': 89252, 'ACCCGG': 89250, 'CGCATC': 89172, 'GCGTGG': 89122, 'CGAGGT': 89071, 'ATCCGC': 88895, 'CGCCAG': 88780, 'CGCCAA': 88719, 'CGAGCA': 88610, 'GGCACG': 88452, 'CCAGCG': 88413, 'GCACGC': 88361, 'ACCCGC': 88289, 'CCTGGG': 88097, 'CATCGG': 87815, 'CGCCAT': 87796, 'CCCTGC': 87758, 'GGTGGA': 87722, 'ACGCGG': 87666, 'CGACCC': 87626, 'CAGCGG': 87530, 'CTGCCC': 87473, 'AGCGCC': 87349, 'CAAGGC': 87035, 'GGTCGG': 86943, 'GGGGCG': 86747, 'TGCTGA': 86549, 'CGCAGG': 86534, 'ACCTCG': 86496, 'CCCCGG': 86154, 'CCGTGG': 85865, 'GAGCGG': 85759, 'CCGCAC': 85752, 'CACGCC': 85725, 'ATGCCG': 85581, 'GGCCGT': 85501, 'GATCCG': 85262, 'TCGGGC': 85149, 'GGACGG': 85013, 'CGCGGA': 84534, 'ATGGCG': 84495, 'GCTGAC': 84490, 'GAAGGC': 84402, 'GAGCAG': 84278, 'AGGGCG': 84254, 'GATGGC': 84157, 'TCCTCG': 83814, 'GCCCCG': 83779, 'CCGGTC': 83756, 'CCGCAG': 83644, 'CAGCCG': 83568, 'GGTGCC': 83549, 'CGACCA': 83503, 'AGATCG': 83353, 'ATCGAG': 83039, 'TCCTGC': 83001, 'CTGTTC': 82910, 'CCCGAC': 82764, 'GTCGTC': 82522, 'GGCCAC': 82490, 'GCGGAC': 82346, 'CCGGGG': 82270, 'GCCTTC': 82222, 'TCGATC': 82129, 'GAGATC': 82022, 'GGCTTC': 81972, 'CCGTGC': 81962, 'TCGCTG': 81818, 'TGGCGA': 81763, 'ATCGTC': 81606, 'TCCTGG': 81416, 'GCCAAG': 81232, 'AGCCGC': 81207, 'ACCGAC': 81159, 'GGTGAT': 81079, 'GGACGA': 81046, 'CGCCTC': 80975, 'ACCGGG': 80882, 'GCTTCG': 80704, 'CGGGTG': 80523, 'CGCCCA': 80444, 'GTCACC': 80341, 'AGCCGG': 80138, 'CGTGGT': 79845, 'TCGGTG': 79742, 'TGGCTG': 79590, 'GCAGCC': 79549, 'GCACGG': 79430, 'TCCGCC': 79342, 'AAGGCG': 79242, 'TGCAGG': 79153, 'CTGGAA': 79131, 'TGGCCA': 78955, 'AGCTGC': 78864, 'CGTCGT': 78778, 'GTGATC': 78682, 'CTTCGA': 78432, 'CCGATG': 77781, 'TGCGCA': 77760, 'GCGCGT': 77585, 'GGCGAA': 77515, 'CGCGTC': 77430, 'CTTCGC': 77407, 'CACCCG': 77398, 'GCCACG': 77348, 'GAGCCG': 77327, 'ATCACC': 77288, 'GTGCTC': 77072, 'TTCGAC': 76995, 'GACATC': 76878, 'TGGCCC': 76549, 'GCAGCA': 76301, 'CCTTCG': 76290, 'ATCCTG': 76271, 'CAGGGC': 76260, 'GCAAGG': 76084, 'CGATGG': 76061, 'ACGTCG': 75997, 'AACGGC': 75867, 'TGTTCG': 75843, 'CCGCTC': 75777, 'CCCGGG': 75707, 'TGGAGC': 75431, 'GATGCC': 75411, 'CGTGCG': 75346, 'GCGCAA': 75310, 'CCAGGC': 75303, 'ACGCCC': 75299, 'CGTGCC': 75196, 'ACGAGG': 75145, 'CAGCAG': 75087, 'GTGGGC': 74964, 'CGCACG': 74919, 'TCAAGG': 74915, 'GCGGGG': 74751, 'GAGGGC': 74614, 'CCCCGC': 74509, 'GACGTG': 74430, 'CCCGCA': 74425, 'CGCCCC': 74308, 'CGACAT': 74278, 'GGGCGT': 73936, 'CGATCA': 73914, 'GTGGTC': 73911, 'GACGTC': 73910, 'CGAGCC': 73547, 'GCCCAG': 73486, 'CTGCAG': 73427, 'TCGTGC': 73400, 'GCAGGA': 73369, 'GGAGCG': 73332, 'GGACCG': 73298, 'CGATCT': 73247, 'ACACCG': 73211, 'TCGACA': 73125, 'CCCGCT': 73085, 'AGCAGG': 73069, 'ACGACC': 72998, 'GGAGGC': 72887, 'GCGCAT': 72874, 'CCCGGA': 72826, 'GAGGAG': 72823, 'GGCAAC': 72735, 'ACCACC': 72675, 'CGGCTT': 72636, 'GAGGTG': 72635, 'CGAGAT': 72596, 'CGAGGG': 72424, 'TGGGCC': 72153, 'GATGCG': 72102, 'TCGGCT': 72071, 'CCGGGT': 72042, 'CCGCGT': 71945, 'CCTGCA': 71925, 'GGCCTC': 71857, 'GGATCG': 71604, 'GGCATG': 71592, 'TTGCCG': 71514, 'CTGATG': 71461, 'GGCCAT': 71444, 'TTCCTG': 71114, 'GAACTG': 70972, 'CCTGAT': 70932, 'ACCCTG': 70908, 'CAAGGA': 70874, 'GGCCAA': 70837, 'AGGTCG': 70715, 'ACATCG': 70702, 'TGCGGG': 70697, 'ACCCCG': 70582, 'CGCTTC': 70439, 'AAGGCC': 70376, 'TCGTGG': 70192, 'GTGCGG': 70059, 'CTCACC': 70049, 'CCAAGG': 70003, 'ATGCTG': 69888, 'CGAAGC': 69839, 'CTTCGG': 69759, 'GCGGGT': 69706, 'AACGCC': 69696, 'CAACGC': 69685, 'CATGGC': 69601, 'CCGGAG': 69505, 'TCAACG': 69498, 'CCACGC': 69455, 'CGCGTG': 69409, 'CACCGT': 69371, 'GATCGT': 69313, 'CGGATC': 69242, 'GACACC': 69228, 'ATCCGG': 69206, 'TCGATG': 69182, 'GCCATG': 69168, 'GACCTC': 69158, 'ATGGCC': 69101, 'GCGGAG': 68822, 'ACGGTG': 68819, 'AAGCCG': 68792, 'CCGAAG': 68649, 'GCCAGG': 68644, 'GCGAAG': 68580, 'TCTTCG': 68571, 'ATCATC': 68468, 'GTTCGC': 68360, 'ATGCGC': 68213, 'CTGCAC': 68179, 'ATCTGG': 68155, 'AGGACG': 68143, 'GCTGTT': 68125, 'CTGGCT': 68098, 'TCGCCT': 68086, 'TCCCGG': 68007, 'GTCCGC': 67980, 'AAGCTG': 67978, 'GAAGCC': 67969, 'GATCTG': 67929, 'AGCAGC': 67914, 'TGCGCT': 67906, 'AGCTCG': 67892, 'TGGAGG': 67886, 'CCGACA': 67795, 'TTCGCG': 67766, 'GACCGA': 67709, 'CCGGGA': 67674, 'CCCGGT': 67666, 'CATCGT': 67641, 'CTCGTC': 67515, 'GGTGAC': 67462, 'ACGGGC': 67401, 'CGACAA': 67278, 'GAGGAC': 67270, 'CCCCGA': 67003, 'CGACAC': 66944, 'GCTGAA': 66925, 'GCAGCT': 66905, 'CCAGCC': 66902, 'ACCGAG': 66865, 'GGCAGG': 66776, 'TCTGGC': 66609, 'TGATGG': 66548, 'CACCAC': 66532, 'CGAAGG': 66474, 'CCGCAA': 66216, 'CCCGAG': 66212, 'GCCAAC': 66190, 'CCACGG': 66130, 'GTTCGA': 65992, 'CGGTCA': 65951, 'CGATGA': 65926, 'CCGAAC': 65904, 'CACGCT': 65833, 'ACCAGC': 65743, 'GCAACG': 65739, 'GTCCGG': 65703, 'AGGTGC': 65656, 'GTGGAC': 65576, 'CGTGGA': 65575, 'TTCCGC': 65522, 'GGGCCT': 65467, 'CACCTG': 65465, 'CAGCTG': 65427, 'CGTCCG': 65421, 'GCAAGC': 65410, 'CGCAAG': 65237, 'CGTTCG': 65213, 'CGCCTT': 65027, 'GGGCAG': 64988, 'CGCGAA': 64984, 'CGCTGT': 64959, 'GAACGC': 64913, 'GAAGAA': 64718, 'ACGAGC': 64696, 'TGGGGC': 64588, 'CCGGAT': 64584, 'GTCATC': 64561, 'ACCGTC': 64530, 'CAGGTG': 64496, 'GTGCAG': 64481, 'GGGCAT': 64474, 'ACGCCA': 64304, 'GGCCCT': 64303, 'CAGCGA': 64285, 'GGGCAC': 64234, 'GTTCGG': 64233, 'TCAGCG': 64143, 'GCGTTC': 64128, 'TGGGCA': 64101, 'TGCAGC': 64074, 'CGTCAC': 64062, 'GGGCAA': 63989, 'GACGCT': 63935, 'GAAGCG': 63841, 'GCACCC': 63834, 'ATCCCG': 63732, 'CGGACC': 63544, 'GCATCC': 63515, 'TCGCGA': 63511, 'TGGACC': 63415, 'GCTCAA': 63345, 'CTGGAT': 63309, 'GCGGAT': 63214, 'ACTGGC': 63196, 'CCCTCG': 63140, 'GCGGGA': 63139, 'GAGGTC': 63095, 'GGGTGC': 62908, 'CTGTCG': 62483, 'GGGTGG': 62475, 'TACGGC': 62320, 'CATCCG': 62293, 'CAACGG': 62290, 'GGTGGG': 62213, 'GCAGGT': 62120, 'GACGGG': 62017, 'CCGCAT': 61855, 'CTCGAT': 61844, 'TTCACC': 61834, 'TCACCC': 61820, 'CCAGCA': 61770, 'GCATGG': 61765, 'TCCCGC': 61725, 'CTCCGG': 61715, 'GACCAG': 61471, 'GTCGAT': 61427, 'CTGGGG': 61360, 'CGCAAC': 61341, 'AGCCCG': 61331, 'CCACCC': 61330, 'AAGGGC': 61289, 'CGGAGC': 61211, 'CAAGCT': 61180, 'CTGGCA': 61082, 'TGGTGA': 60938, 'GACGGT': 60884, 'GCCTGA': 60845, 'GCGACA': 60699, 'AAGCGC': 60658, 'GCTGTC': 60418, 'CGAAGA': 60386, 'ACCGTG': 60341, 'GTGGAG': 60336, 'TCGTCA': 60335, 'CAACGA': 60318, 'AGGTGG': 60292, 'AACTGG': 60267, 'TCCGGG': 60247, 'GATCCT': 60169, 'GCAGGG': 59936, 'GGTGCA': 59844, 'ACGATC': 59808, 'CAGATC': 59769, 'TGGAAG': 59700, 'CGAACT': 59598, 'CGAACG': 59512, 'AGGCGA': 59483, 'GGAGGA': 59458, 'CCACCA': 59062, 'TGACGG': 59060, 'GGAGCT': 59028, 'CGGGTC': 58957, 'AGCGGG': 58927, 'GACGAT': 58861, 'TGATCC': 58810, 'GCCCTC': 58769, 'ACCGCA': 58757, 'GGCTGA': 58736, 'ATCGTG': 58708, 'TGGCCT': 58639, 'GCACCA': 58443, 'GCAGAT': 58402, 'ACCATC': 58387, 'ATGGGC': 58348, 'TGCTCA': 58310, 'AGCCTG': 58309, 'TGATGC': 58306, 'CGGAGG': 58235, 'TTCGTC': 58095, 'TTCGAG': 58080, 'GGCGTT': 57874, 'GATGCT': 57824, 'ACGGCT': 57823, 'TGCTGT': 57648, 'CCTGAC': 57607, 'CGGTGT': 57479, 'GTGATG': 57449, 'CAGCTC': 57395, 'CCAGCT': 57329, 'GCGAAC': 57323, 'AACCTG': 57309, 'CACGCG': 57249, 'GGCTAC': 57121, 'GCATCA': 57085, 'CACCCT': 57080, 'AAGAAG': 56969, 'TGGTCA': 56945, 'ACGTGC': 56936, 'TGTCGG': 56888, 'AGACCG': 56832, 'GAAAAA': 56810, 'GTGACC': 56756, 'CCGAGA': 56604, 'CATCCT': 56507, 'TCGAAG': 56445, 'CAAGGT': 56435, 'GACCAC': 56387, 'TTCCGG': 56306, 'CATCAC': 56305, 'GCACCT': 56304, 'CCGTCC': 56238, 'GATCAC': 56134, 'GGGACG': 56101, 'TGGCAG': 55975, 'CAAGGG': 55970, 'AGGCCA': 55954, 'GATCAA': 55915, 'GATGAA': 55911, 'CCCAGG': 55911, 'CGCTCA': 55873, 'CATCAA': 55856, 'GCACGA': 55831, 'GATCAT': 55802, 'GTGACG': 55801, 'TGAAGG': 55774, 'GTCGTG': 55773, 'GGTCGT': 55614, 'GCGCTT': 55544, 'GATGAT': 55513, 'GCTCGT': 55500, 'TCACCA': 55498, 'TGATGA': 55485, 'CTCGGT': 55454, 'AGCGCA': 55425, 'CGAGAA': 55423, 'CTGAAG': 55315, 'CTGACG': 55256, 'ATCTCG': 54971, 'CTCGGG': 54937, 'GACAAG': 54824, 'GGATGC': 54674, 'CGGGAC': 54651, 'ATCTGC': 54639, 'TCCGCA': 54639, 'CCAACG': 54624, 'GGGGCC': 54600, 'GACTGG': 54585, 'GGACCT': 54516, 'TCTCGG': 54470, 'GGTCAA': 54412, 'CACGAC': 54364, 'CGTGAT': 54249, 'GGGTCG': 54170, 'ACGTGG': 54168, 'CTACGA': 54144, 'CCAGGA': 54096, 'TGACGC': 54060, 'ACCACG': 54048, 'GGTGAA': 54008, 'TACGCC': 53917, 'AAAAAA': 53907, 'GGTCCG': 53884, 'TTCATC': 53866, 'CCCAGC': 53795, 'TCGAGA': 53746, 'GATGGA': 53730, 'TGGATG': 53673, 'TCAAGC': 53646, 'CGAGAC': 53601, 'CTTCCT': 53596, 'GTCGGG': 53583, 'CGGATG': 53572, 'TTCCCG': 53535, 'GGAGAT': 53464, 'GGGCTC': 53432, 'ATCAGC': 53403, 'GGGGCT': 53376, 'TCCAGC': 53315, 'TGATCA': 53244, 'CCGTCA': 53149, 'CACGGT': 53131, 'CAGGAC': 53090, 'GACTTC': 53079, 'GTCCTG': 53069, 'TGCACG': 53042, 'GCGAGA': 53020, 'GCGTGA': 52924, 'CGGTTC': 52878, 'GTGTTC': 52860, 'CGGGGT': 52795, 'CCGTTC': 52774, 'GGGCCA': 52727, 'TTCTGG': 52691, 'GCGTCC': 52669, 'TCCCCG': 52648, 'GCCCAC': 52647, 'GGTTCG': 52547, 'CCGGAA': 52495, 'CTCAAG': 52491, 'CTCCGC': 52488, 'CAGCCC': 52423, 'GCTCAC': 52422, 'TGTCGC': 52333, 'GGCCTT': 52332, 'AGGGCA': 52324, 'TGAAGA': 52306, 'GAGAAG': 52255, 'GGTCAC': 52225, 'GCTCCG': 52216, 'TCATCA': 52178, 'AACGCG': 52166, 'CGTGGG': 52163, 'GTGCCC': 52097, 'GACCGT': 52053, 'CAGCCT': 52050, 'CCATGG': 51997, 'CGCTCC': 51913, 'CAGGTC': 51853, 'CGTCAA': 51837, 'TCCAGG': 51800, 'CAACCT': 51793, 'CAAGCG': 51781, 'GTCAGC': 51778, 'GCCTAC': 51778, 'CTCGCT': 51754, 'CAAGCC': 51703, 'GCCCGT': 51673, 'GCGTCA': 51589, 'CATCAT': 51561, 'CGTGCA': 51552, 'ATCAAG': 51478, 'ACCTGA': 51458, 'GGAGCA': 51394, 'ATGCGG': 51375, 'GGAAGC': 51334, 'GGACGT': 51293, 'ACGCCT': 51263, 'GACAGC': 51197, 'CGGTCT': 51180, 'GGCCCC': 51175, 'TCTCGC': 51130, 'AGGAGG': 51112, 'CGGCTA': 51095, 'CGTCAT': 51079, 'CCACCT': 51069, 'CTGGGA': 51056, 'ATCAAC': 50950, 'GCTACG': 50935, 'CACCAT': 50916, 'CGGTCC': 50910, 'GCTTCC': 50810, 'TTGCTG': 50754, 'AACCGC': 50733, 'GCAACC': 50659, 'GAACGG': 50605, 'AAGATC': 50593, 'GGAAGA': 50517, 'GAGACC': 50458, 'GAAGAT': 50455, 'AAGCGG': 50455, 'GTCTGG': 50435, 'TACCTG': 50356, 'CCATCC': 50351, 'TCGGTC': 50294, 'TTCTTC': 50183, 'TCTCCG': 50137, 'TCGTCC': 50127, 'GATCTC': 50109, 'GCCGTT': 50065, 'AGGAGC': 50021, 'CCACGA': 49992, 'CCCACC': 49935, 'TGCCGT': 49922, 'ACGGTC': 49847, 'GAGTTC': 49788, 'ACCTTC': 49771, 'AGGCCC': 49759, 'CTCGAG': 49749, 'GATCCC': 49747, 'CACCCC': 49724, 'ATGGTG': 49723, 'TCTGGA': 49712, 'CGTCCT': 49689, 'CCAAGC': 49606, 'CAGGAG': 49601, 'GTCGGT': 49597, 'CGCTAC': 49544, 'GGAAAA': 49459, 'TCAACC': 49385, 'ACTTCG': 49353, 'CTTCGT': 49320, 'ACAAGG': 49312, 'TCAGCC': 49273, 'GCAAGA': 49224, 'CCAGGG': 49211, 'CAACCG': 49159, 'GGCCCA': 49131, 'CCCGAT': 49126, 'TCACGC': 49096, 'CTACGC': 49025, 'TCGAAC': 48996, 'AGAAGG': 48978, 'GGGGTG': 48915, 'AGCGCT': 48912, 'ACCAGG': 48878, 'CAGCAC': 48878, 'AGCTGA': 48873, 'GTCGCT': 48794, 'TACCGC': 48791, 'TGGCGT': 48669, 'CGCAGA': 48653, 'GCAGAA': 48612, 'CCTGTT': 48602, 'CGGGGA': 48539, 'AAGGTG': 48509, 'TGGAGA': 48439, 'GACGAA': 48408, 'CGTGAC': 48386, 'ATGATC': 48368, 'GGTGTT': 48361, 'CCATGC': 48330, 'GGAGAA': 48262, 'GAGCAC': 48243, 'CCCCCG': 48224, 'ATCCTC': 48119, 'ACCGCT': 48090, 'GCCTCC': 48088, 'TCATGG': 47983, 'TGAAGC': 47976, 'TGGAAC': 47868, 'CCATCA': 47843, 'CACCTC': 47825, 'ACCGGT': 47816, 'TCATCC': 47749, 'CCTGAA': 47747, 'CCCGTC': 47742, 'TGGATC': 47715, 'CTTCCG': 47705, 'AACCCG': 47647, 'GGCTCC': 47644, 'ACAGCG': 47637, 'CGCATG': 47636, 'GATGAC': 47603, 'TCTGGG': 47599, 'GGTCTG': 47583, 'GTCAAC': 47560, 'ACGACA': 47450, 'ATCTTC': 47442, 'GAAGAC': 47432, 'CAGCGT': 47385, 'TCTACG': 47374, 'AGTTCG': 47255, 'TCCGCT': 47201, 'ACGCTC': 47186, 'TCCACC': 47154, 'CACCAA': 47143, 'CGGGAG': 47098, 'TTGGCG': 47091, 'AACCGG': 47090, 'CCTTCC': 47085, 'CTCAAC': 47079, 'ACCGGA': 46961, 'TGGGTG': 46902, 'CAGCAA': 46845, 'CTGGGT': 46802, 'GCATGC': 46794, 'ACGCGA': 46792, 'GGGGGC': 46769, 'TGAGCG': 46743, 'CCGCTT': 46739, 'GTTCCG': 46696, 'CCAGGT': 46623, 'CAAGAA': 46619, 'CGCGTT': 46576, 'CGGACA': 46419, 'AAGGAC': 46401, 'GAAGGA': 46383, 'GCCTGT': 46358, 'GATGGT': 46337, 'CTACGG': 46310, 'CCGGTT': 46299, 'TTCGTG': 46277, 'CGATGT': 46242, 'GAAGCT': 46225, 'GCGGAA': 46218, 'TGGAAA': 46197, 'AGCCGA': 46192, 'CGGGAT': 46183, 'CCTCCG': 46168, 'GACCCC': 46136, 'TCACGG': 46028, 'AAGCAG': 45957, 'GAACCG': 45952, 'AGATCA': 45919, 'GAGCTC': 45877, 'TCCCGA': 45739, 'AAGACC': 45728, 'TCTGCG': 45653, 'GAAGAG': 45642, 'CATGGA': 45583, 'TACGAC': 45573, 'CCTGTC': 45520, 'GGATGG': 45464, 'GTGAAG': 45450, 'CATGCC': 45447, 'GATCAG': 45395, 'AAAAAG': 45354, 'GCTTCA': 45313, 'TGACCA': 45262, 'CCTACG': 45237, 'CTGTCC': 45223, 'ATGAAG': 45221, 'CTCGAA': 45214, 'CGGAGA': 45198, 'CGTTCC': 45176, 'CCTCGT': 45127, 'CATGCT': 45106, 'ATCCAG': 45060, 'TGTCGA': 45042, 'AGAAGA': 45038, 'TGCACC': 44946, 'CAGACC': 44938, 'TCGGGG': 44908, 'GTCTTC': 44853, 'CGACTG': 44846, 'GCACTG': 44825, 'TCCTGA': 44804, 'TGCAGA': 44756, 'TTCCTC': 44729, 'GGGTGA': 44721, 'GATGGG': 44715, 'CAGCCA': 44671, 'GGGGCA': 44588, 'GACAAC': 44579, 'GGGCTT': 44547, 'GGGCCC': 44461, 'CCGTGA': 44424, 'CGGGGG': 44372, 'AGAAGC': 44359, 'CTGAAC': 44357, 'GGTGTC': 44353, 'GTGTCG': 44305, 'GTTCCT': 44305, 'GGAGCC': 44190, 'TGACCC': 44098, 'GGTCAT': 44083, 'TTGCGC': 43981, 'CACCAG': 43922, 'CTGTGG': 43850, 'GTCGAA': 43825, 'CGAACC': 43795, 'CATCCC': 43785, 'CCCTGA': 43725, 'CTTCAC': 43706, 'CTCATC': 43697, 'AAGGTC': 43691, 'AGCACC': 43673, 'CCAACC': 43669, 'AGGGCC': 43654, 'GTGGAA': 43636, 'CCCGAA': 43631, 'AAAAAT': 43626, 'TACCGG': 43622, 'GTCCCG': 43607, 'ATGACC': 43593, 'CAACAA': 43552, 'CGAACA': 43507, 'CTACCG': 43448, 'ACAACG': 43431, 'TGTCCG': 43343, 'TCGCTC': 43334, 'CAAGAT': 43303, 'GAGATG': 43280, 'TGACGA': 43279, 'ATGATG': 43271, 'TCTCGA': 43218, 'ATGTCG': 43143, 'TGGACA': 43105, 'CAAGCA': 42967, 'TGCCTG': 42933, 'AGATCC': 42900, 'GCGGTT': 42863, 'ATTGCC': 42854, 'GGAACG': 42830, 'ACTGGA': 42786, 'AACGAC': 42740, 'GGACAT': 42723, 'GCAGAC': 42708, 'TGTTCC': 42696, 'CATGGG': 42691, 'TGCCCA': 42681, 'AGCACG': 42678, 'ATGACG': 42634, 'ACTGGG': 42628, 'AACTGC': 42619, 'TCCGAC': 42615, 'GGGATC': 42601, 'ACGATG': 42578, 'GCTGTG': 42564, 'TCTTCC': 42518, 'CGGAAG': 42434, 'TCCGGA': 42413, 'AGCGAC': 42405, 'TGTGGC': 42399, 'ATCGAA': 42384, 'GTCAAG': 42379, 'AGCAGA': 42336, 'CGCCTA': 42301, 'GGACAA': 42278, 'GTTCGT': 42273, 'CACCCA': 42255, 'TGAAAA': 42167, 'CTTCAA': 42107, 'GAAATC': 42107, 'CCTCAA': 42068, 'CTACCT': 42046, 'TTCCAG': 42032, 'TTCTCG': 42027, 'GGATGA': 41948, 'TCCGGT': 41945, 'TCTGGT': 41937, 'ATGCAG': 41903, 'TGAACG': 41855, 'CAGGAA': 41805, 'TCAACA': 41790, 'GGAAGG': 41786, 'CTGAGC': 41754, 'ATCGGG': 41686, 'GACCAT': 41565, 'GTGGCT': 41517, 'GAAGGT': 41500, 'TGGGCT': 41488, 'CGAGTT': 41411, 'CTTCAT': 41346, 'CAACCC': 41336, 'CCCGTG': 41318, 'CAGTTC': 41288, 'GGGAGC': 41281, 'GGAGGT': 41270, 'GGCAGA': 41229, 'TCCACG': 41191, 'CTGCGT': 41190, 'GCAACA': 41186, 'GGCTGT': 41163, 'GAACAG': 41125, 'CAAGAC': 41110, 'AACATC': 41107, 'GTCTCG': 41053, 'CCCACG': 41047, 'GATTTC': 41033, 'AAGAAA': 41009, 'CAACGT': 40999, 'GTGGAT': 40989, 'GGTCAG': 40927, 'CCGACT': 40920, 'TTCAGC': 40872, 'AGATGG': 40866, 'AGGTGA': 40854, 'TTTGCC': 40834, 'CCTTCA': 40799, 'GCCCCC': 40786, 'ACGAGA': 40746, 'GAAGGG': 40728, 'TCGACT': 40694, 'ATCGAT': 40679, 'GAGGAA': 40657, 'GAGAAC': 40586, 'CATGCG': 40574, 'CGACAG': 40569, 'ACATCC': 40556, 'GCTACC': 40526, 'AAGCTC': 40481, 'TGCGGT': 40467, 'CACTGG': 40463, 'CACGGG': 40431, 'GACATG': 40414, 'TGCGGA': 40389, 'AGCGTG': 40362, 'AGGTCA': 40334, 'CAGAAG': 40319, 'ACATCA': 40284, 'AGGACC': 40275, 'CGTCCC': 40275, 'GAGACG': 40264, 'AGGAGA': 40242, 'GAGCCC': 40213, 'TGGCAC': 40203, 'GAACTC': 40203, 'GGATCA': 40169, 'CCAGAT': 40118, 'CATGGT': 40109, 'AGGAAG': 40099, 'AACACC': 40049, 'GGAACT': 40037, 'TTCAAC': 40018, 'CGTCTG': 39973, 'TTCAAG': 39972, 'CGACTT': 39971, 'GGCTTT': 39956, 'ATGCCC': 39920, 'GCCAGA': 39865, 'GGTGTG': 39852, 'ATCATG': 39852, 'CTCGTG': 39827, 'GGCATT': 39823, 'CTCCGA': 39803, 'GCGTTG': 39769, 'GATGTG': 39768, 'GTCACG': 39747, 'ATGGTC': 39745, 'ACCCGA': 39739, 'GAGCGA': 39714, 'ACGTCA': 39629, 'ATGGAC': 39627, 'CATCTG': 39616, 'TCAAGA': 39611, 'GGACTG': 39587, 'CATCCA': 39586, 'CCTCAC': 39584, 'GGTGAG': 39525, 'TGTTCA': 39511, 'ACACCC': 39472, 'GAAAAG': 39437, 'TCACCT': 39364, 'TCTGCT': 39362, 'GTGGCA': 39326, 'GAAACC': 39310, 'GGCAAT': 39293, 'GGAGGG': 39255, 'GGACCC': 39233, 'GACTAC': 39197, 'TGCCCT': 39157, 'TGGCTC': 39091, 'GTCCTC': 39081, 'AACAGC': 39069, 'GACCCT': 39022, 'CAGATG': 39005, 'AAAGAA': 38991, 'TCGCAG': 38980, 'TCAGCA': 38969, 'ACAAGC': 38969, 'GACGGA': 38909, 'GCAGTT': 38898, 'CGTTGC': 38884, 'CGCTCT': 38823, 'CGGAAC': 38805, 'TGGTCC': 38781, 'GTGGGG': 38738, 'TGGTGT': 38737, 'ATGCTC': 38722, 'GCATCT': 38721, 'CGTCTC': 38701, 'GAAGTG': 38697, 'AAAAGA': 38618, 'CGGGAA': 38605, 'TTTCCG': 38571, 'AGAAAA': 38524, 'CTCCTG': 38482, 'CTGCAA': 38459, 'TCTGCC': 38457, 'CAACTG': 38412, 'CTCCCG': 38392, 'GTGCAC': 38361, 'ACGAAG': 38359, 'TACGCG': 38324, 'ACCAAC': 38315, 'ACCGAT': 38314, 'GGATCT': 38313, 'GTCTGC': 38308, 'TTGATG': 38296, 'GCGACT': 38292, 'GGAGAC': 38292, 'GGCAAA': 38268, 'TGATCT': 38243, 'ATCACG': 38217, 'AAGAAC': 38152, 'GATGTC': 38121, 'CAGCAT': 38117, 'CGTCCA': 38109, 'CACGAT': 38075, 'CACCTT': 38073, 'CTTGCC': 38006, 'GCCTTG': 37970, 'AGAACG': 37919, 'AGGCCT': 37896, 'ATGGAG': 37756, 'CGTCAG': 37724, 'GATGCA': 37703, 'AGGACA': 37606, 'GCATGA': 37573, 'AAGGAG': 37489, 'TTGCGG': 37465, 'AAGGAA': 37465, 'CTGGTT': 37464, 'TGGCAA': 37464, 'CAAAAA': 37458, 'CAACAT': 37410, 'AAGATG': 37388, 'GAAAGC': 37349, 'AAAAGC': 37282, 'GATGAG': 37256, 'TGACCT': 37255, 'TCTTCA': 37236, 'CGGACT': 37207, 'AGCGTC': 37178, 'CGGAAA': 37165, 'ACCCAG': 37137, 'GATCCA': 37058, 'TTCGGG': 37047, 'GGAAAT': 36996, 'ACAGCC': 36988, 'TCATGC': 36921, 'AAGCCC': 36874, 'AGCGAG': 36857, 'TCGTCT': 36852, 'CTCAGC': 36842, 'TGCGTG': 36794, 'CAGCTT': 36780, 'AAGACG': 36777, 'AACTCG': 36760, 'CTTCCC': 36744, 'AGCATC': 36740, 'GCTCTC': 36729, 'CGTGAA': 36711, 'GGACAC': 36692, 'ATTTCG': 36688, 'GATCTT': 36680, 'GTACGC': 36671, 'AGACGG': 36663, 'GGCTCA': 36632, 'ACCAAG': 36624, 'GACACG': 36567, 'GAGCGT': 36495, 'GTTGCC': 36483, 'CTCCTC': 36460, 'TGGGAC': 36447, 'GCAAAA': 36441, 'AACGTC': 36422, 'TGGTTC': 36367, 'GATTTT': 36362, 'GGATCC': 36341, 'GTGAGC': 36328, 'CACGTC': 36291, 'TCGCGT': 36285, 'ACATGG': 36239, 'ATGGCA': 36174, 'GCTTGC': 36165, 'CGTTCA': 36164, 'AAGAGC': 36159, 'ACTGGT': 36059, 'AGCCAG': 36051, 'CGAAAA': 36034, 'CAACCA': 35991, 'TTTCGC': 35991, 'AGATGC': 35987, 'ACCTGT': 35974, 'CTTCCA': 35959, 'AATGGC': 35954, 'GGGATG': 35939, 'ACGGGG': 35866, 'AGGGGC': 35865, 'GCAACT': 35843, 'ATGTGG': 35819, 'GCTCCT': 35812, 'AGGGCT': 35810, 'TCGTGA': 35810, 'AGCGGA': 35789, 'AGCTTC': 35772, 'AGCGGT': 35763, 'GCTTCT': 35729, 'GAACGA': 35718, 'ACCGAA': 35676, 'GCCTAT': 35658, 'GCGTCT': 35653, 'ATGAGC': 35643, 'GGGAGG': 35637, 'AGGCGT': 35610, 'CACACC': 35607, 'TTGAAG': 35603, 'CAACAC': 35586, 'GCCCAA': 35526, 'ACACCA': 35510, 'AGACGC': 35505, 'CCTTGC': 35489, 'GCTCAT': 35483, 'ACTACG': 35482, 'GAAGCA': 35473, 'GCACGT': 35469, 'CGGGTT': 35454, 'ACTGCG': 35388, 'CATGAT': 35373, 'CGAAGT': 35334, 'GGCTAT': 35299, 'CACGAG': 35277, 'GACCAA': 35270, 'GCTCCA': 35268, 'CTGCGA': 35215, 'TATCGC': 35203, 'TCGGAC': 35191, 'GTTCAC': 35154, 'GAAAAT': 35130, 'GCGATT': 35103, 'GGACCA': 35075, 'CCGTCT': 35073, 'TTCGAT': 35045, 'CGATTT': 35026, 'GTTGCG': 35018, 'GCTTTC': 34999, 'ATCGGT': 34975, 'GCGAAA': 34952, 'GTTCAA': 34914, 'ATTCCG': 34911, 'GGGACC': 34888, 'AAACCG': 34841, 'ACCATG': 34829, 'GAAAAC': 34807, 'CAGTGG': 34784, 'GGTCTC': 34783, 'AAAAAC': 34779, 'GTGTGG': 34733, 'CTCTAC': 34731, 'ACACGC': 34713, 'ACGCAG': 34698, 'CTCCAG': 34654, 'GAACCT': 34600, 'TCCTCA': 34592, 'AGCTCA': 34579, 'GACGCA': 34578, 'CCCCTG': 34578, 'CTCTGG': 34576, 'ACAACC': 34566, 'ACCTCA': 34529, 'GCCCAT': 34522, 'ATGGAA': 34518, 'GTACCG': 34512, 'CATGAC': 34498, 'ACGTCC': 34460, 'TTCTGC': 34428, 'TGGGGG': 34422, 'GGTCCT': 34421, 'GAACAA': 34399, 'CATTGC': 34373, 'GCCAAT': 34364, 'TCAGGC': 34325, 'GTTGGC': 34313, 'CCAGAA': 34282, 'CCGAAA': 34199, 'ATCAGG': 34194, 'ACGAAC': 34189, 'CGCAGT': 34155, 'GCTCAG': 34144, 'AGAGCG': 34092, 'GGATTT': 34081, 'CAGAAC': 34075, 'GACCTT': 34049, 'GGTTCC': 34023, 'CGGTTT': 33997, 'TTTCGG': 33990, 'TTGCAG': 33985, 'ACCCTC': 33970, 'CAGGCA': 33931, 'AACAAC': 33876, 'AAAGCC': 33869, 'TATCTG': 33837, 'GTACGG': 33837, 'CATCTC': 33833, 'CCAAGA': 33819, 'GCCATT': 33809, 'GATTGC': 33783, 'CCTACC': 33778, 'CCATGA': 33776, 'CTGCCT': 33762, 'GCAGTG': 33751, 'TGGTCT': 33751, 'CAATGC': 33750, 'AACGAG': 33744, 'TGTGGG': 33743, 'TTGGCC': 33724, 'GCAATG': 33713, 'TGCTCT': 33687, 'GCTCTG': 33681, 'CCGAGT': 33669, 'CATCAG': 33668, 'TCCATC': 33657, 'GGTTTC': 33598, 'ACTTCC': 33591, 'GTCATG': 33581, 'GCCAGT': 33550, 'AACAAG': 33539, 'CTGCAT': 33532, 'TTCATG': 33500, 'CCAACA': 33472, 'GTTCAT': 33469, 'GGCAGT': 33458, 'AAGGCA': 33453, 'TGTGGA': 33433, 'GCTTTG': 33426, 'GCGCTA': 33422, 'TCTACC': 33418, 'GTCCGA': 33417, 'GTGAAC': 33417, 'CTCTTC': 33400, 'AGGAAA': 33385, 'ATGAAC': 33378, 'GAAGTC': 33253, 'AACCAG': 33244, 'GGCACT': 33232, 'ACTGCT': 33215, 'CTGAAA': 33211, 'GGAGTT': 33123, 'GCTGAG': 33073, 'CTCTCC': 33042, 'GGGTTC': 32949, 'CGCTTT': 32929, 'AGATGA': 32924, 'TACGAG': 32911, 'ATGAAA': 32899, 'GCCTCA': 32872, 'ACTCGC': 32871, 'GGGGAG': 32863, 'CTTGGC': 32846, 'GAGCAA': 32826, 'TCGTTC': 32821, 'CACGCA': 32818, 'GGGGAC': 32816, 'GCATTG': 32741, 'TGGGAG': 32723, 'AGTGGC': 32719, 'TTGCCC': 32626, 'GTCCAG': 32616, 'ATCCCC': 32605, 'ATATCG': 32600, 'ACCTCC': 32592, 'AACGTG': 32587, 'TGTTGC': 32568, 'CCCATC': 32559, 'GTTGCT': 32518, 'AATCCG': 32513, 'CGGTTG': 32501, 'GGGGAT': 32499, 'CTCGGA': 32485, 'GCCTTT': 32433, 'GAACCC': 32388, 'CCCTTC': 32383, 'ACGGAC': 32317, 'TGCTCC': 32307, 'CTTTGC': 32306, 'AGCCGT': 32301, 'CTACAA': 32287, 'CGACTA': 32237, 'CTGTTG': 32230, 'AACCTC': 32213, 'TACACC': 32201, 'CTTCTC': 32179, 'ATTTTT': 32171, 'TTTGGC': 32168, 'CGCACT': 32121, 'GTTTCG': 32118, 'CCAGAC': 32073, 'CATGAA': 32065, 'GGTTGC': 32061, 'AGCTGT': 32059, 'ATCGCT': 32040, 'TCCCTG': 32032, 'GGACAG': 32025, 'ATTGGC': 31936, 'GCAATC': 31926, 'TGAACC': 31925, 'TTGAAA': 31916, 'AAAGCG': 31912, 'GTCTCC': 31903, 'GCTCCC': 31890, 'ACAGCA': 31887, 'TGAACA': 31882, 'GTCGGA': 31865, 'ACTCGG': 31851, 'CGGTAC': 31846, 'ACTGCC': 31845, 'GTCTAC': 31811, 'AATGCC': 31785, 'TCGCAC': 31776, 'TCCGTC': 31762, 'ATCTCC': 31743, 'CTTCTT': 31721, 'CGACTC': 31717, 'GAGAGC': 31648, 'TGCCCC': 31612, 'AGCAAC': 31601, 'ATGCCA': 31551, 'ATTCGC': 31537, 'GTTCTG': 31526, 'TGCGTC': 31492, 'CCTCCA': 31479, 'CTCTCG': 31463, 'CTGCCA': 31456, 'CACTGC': 31428, 'TGAGCA': 31419, 'TTCGGT': 31408, 'TGCGAC': 31400, 'CTGTGC': 31368, 'ACAAGA': 31341, 'GCGTAC': 31330, 'CGTACG': 31306, 'CGGATT': 31300, 'TCCGAG': 31278, 'TTGCCA': 31262, 'CCGCTA': 31251, 'TGAGCC': 31244, 'ATGTTC': 31240, 'AGCAAG': 31231, 'GAGCCT': 31217, 'CTCCAC': 31190, 'ACTTCA': 31173, 'CTTCTG': 31173, 'GGGAGA': 31145, 'AAAAGG': 31138, 'TTCTCC': 31106, 'AAAATT': 31092, 'GTCGCA': 31091, 'CGAAAC': 31079, 'TGTGCG': 31043, 'CGCACA': 31002, 'GGAACA': 30969, 'GAGGGG': 30962, 'GATGTT': 30923, 'GGTACG': 30917, 'CCATCT': 30908, 'GAACGT': 30880, 'GCGAGT': 30874, 'GGGAAG': 30860, 'AACTTC': 30859, 'CGTGTT': 30858, 'AGACCA': 30853, 'ACACGG': 30846, 'TTTCGA': 30838, 'AAAACC': 30825, 'AGCGAT': 30820, 'AAATCG': 30815, 'GAAATT': 30782, 'CGATTC': 30761, 'AGGCTG': 30757, 'CGTTCT': 30739, 'GTCCAC': 30723, 'CCCAAG': 30721, 'GATATC': 30711, 'ATTTCC': 30656, 'TTTCTG': 30651, 'GGAAAC': 30628, 'GCTTTT': 30622, 'GACTGC': 30583, 'GCATTC': 30581, 'GAGGAT': 30579, 'GTGGGA': 30573, 'CAGGAT': 30565, 'CCTCCT': 30536, 'TCAGCT': 30510, 'AAGTGG': 30501, 'GAACAT': 30479, 'CCAGTT': 30443, 'TATGGC': 30433, 'ATGGCT': 30375, 'ACCTAC': 30332, 'CAGGCT': 30328, 'CTCGCA': 30322, 'GAGCAT': 30312, 'TTGATC': 30309, 'TTCACG': 30301, 'CGTTTC': 30295, 'ACGTGA': 30288, 'TACCCG': 30270, 'CAATGG': 30253, 'GGGGTC': 30247, 'ACCAGA': 30228, 'GGAACC': 30212, 'TATTTT': 30207, 'AGACCC': 30188, 'CAGACG': 30174, 'TGATGT': 30138, 'CGTCTT': 30112, 'GGCTTG': 30108, 'TCCAGA': 30088, 'CTGTAC': 30084, 'CGAAAT': 30082, 'ATGTGC': 30079, 'TGCCAA': 30043, 'CACGGA': 30042, 'TGGCAT': 30037, 'GCAGTC': 30036, 'AAATGG': 30014, 'GAGTGG': 30000, 'CAGGGG': 29982, 'TTTTTG': 29980, 'GAAATG': 29947, 'TCGAAA': 29937, 'GGCACA': 29927, 'GGTCTT': 29908, 'ATCCAC': 29894, 'CCGTAC': 29847, 'TTGGTG': 29842, 'CGCTTG': 29809, 'ATCGCA': 29801, 'GTTCCA': 29800, 'CAGAGC': 29797, 'TATGCC': 29757, 'AAGTCG': 29753, 'AAAATC': 29745, 'GCTACA': 29742, 'CGAGTC': 29730, 'TCCTGT': 29715, 'AAAGGC': 29700, 'AGGAAC': 29677, 'GAGCTT': 29668, 'GTTCCC': 29667, 'TGCATC': 29665, 'ACAACA': 29647, 'GGGGGA': 29614, 'GAACAC': 29588, 'TACCAG': 29584, 'ATCTAC': 29572, 'AACGGG': 29555, 'CTACGT': 29549, 'CAACAG': 29547, 'AACAGG': 29541, 'CTACAC': 29535, 'TGCCAT': 29531, 'GCCCCT': 29528, 'ATTTTG': 29523, 'TAAAAA': 29491, 'GGAAGT': 29456, 'TGATTT': 29444, 'GGATGT': 29441, 'GCCCTT': 29438, 'TCCGTG': 29421, 'TACTGG': 29354, 'TTTTGC': 29349, 'CCTGTG': 29335, 'CGCAAT': 29312, 'TTTTTC': 29283, 'TCGGGA': 29282, 'CTATGC': 29262, 'CCGTTG': 29248, 'GAGTCG': 29237, 'AGAACA': 29233, 'CCCTGT': 29229, 'GCTATC': 29206, 'AGAACC': 29195, 'GAATGG': 29193, 'TTGTCG': 29187, 'GCCAAA': 29176, 'GGCCTA': 29169, 'AAGCCA': 29147, 'GATTCC': 29139, 'TGCCAC': 29136, 'GGAATT': 29114, 'CAGTCG': 29107, 'GCAAAG': 29093, 'ACTACC': 29081, 'GCGTTT': 29056, 'TCTGCA': 29049, 'CCGATT': 29048, 'GAATTC': 29028, 'ACGCAC': 29018, 'AGCAGT': 28972, 'TTCCAC': 28964, 'GTTGAT': 28964, 'CAATCC': 28921, 'TCATCT': 28919, 'CACATC': 28912, 'CCTTCT': 28895, 'ACACCT': 28892, 'ACCCGT': 28888, 'AAAATG': 28867, 'ACGACT': 28865, 'CTCACG': 28855, 'TCCTCC': 28813, 'TGCCAG': 28811, 'TTTATT': 28791, 'CCTATG': 28736, 'TTTTGA': 28728, 'TTCCCC': 28724, 'TCCTTC': 28693, 'GACTCG': 28666, 'TGTTCT': 28648, 'CATGCA': 28619, 'CGCATT': 28599, 'CAAAGC': 28599, 'GAGTAC': 28584, 'AGTTCC': 28573, 'ATTGAA': 28557, 'TCTCCA': 28551, 'GTTCTT': 28522, 'TTTTTT': 28511, 'TATCCG': 28450, 'GGGTCA': 28417, 'GGCTCT': 28405, 'GCACAC': 28403, 'CTATCT': 28393, 'TCATGA': 28383, 'GCACAG': 28371, 'AGTCGG': 28358, 'TTTGAA': 28355, 'AGGATG': 28343, 'GTACGA': 28321, 'CGCAAA': 28316, 'ACAGGC': 28271, 'TGAAAT': 28260, 'TTCTAC': 28258, 'TTTTAT': 28250, 'CATTCC': 28224, 'ATGGAT': 28214, 'GGGTGT': 28212, 'GATATT': 28190, 'CCTCCC': 28178, 'TTCAGG': 28154, 'TCAGGA': 28136, 'AAAGAT': 28124, 'AGTCGC': 28124, 'TGTGGT': 28116, 'CGATTG': 28095, 'TGGCTT': 28087, 'GGATTC': 28074, 'CCACGT': 28057, 'TCAATG': 28025, 'CGTGTC': 28023, 'ATTGAT': 28000, 'GGACTT': 27998, 'ATTGCG': 27978, 'CATCTT': 27957, 'GGTTTT': 27948, 'ATGGGG': 27943, 'CCCAAC': 27940, 'GCTATG': 27928, 'GAAAGA': 27920, 'TGGATT': 27914, 'TCGGGT': 27914, 'AAAATA': 27899, 'TCGGAG': 27898, 'AGTACG': 27874, 'CACAGC': 27857, 'GGGAAA': 27826, 'GTGGGT': 27816, 'CTACCA': 27802, 'TATCGA': 27783, 'TGGGTC': 27782, 'ATTATT': 27766, 'TGCAAC': 27750, 'TCTACA': 27748, 'CCCATG': 27738, 'CTGCTT': 27736, 'GTTCTC': 27707, 'TACGTC': 27703, 'ACATGC': 27701, 'GGGAAC': 27698, 'GATTCG': 27692, 'TTATTT': 27680, 'CGATAT': 27640, 'GCTTGG': 27640, 'CCCGTT': 27621, 'GCGTGT': 27609, 'AAGTTC': 27598, 'TGTACG': 27581, 'TCAAAA': 27575, 'ACGTTC': 27569, 'CCCCGT': 27567, 'TCTTCT': 27553, 'ATTTTC': 27551, 'TTTTCC': 27532, 'CCAATG': 27503, 'GCACAA': 27502, 'GGAAAG': 27490, 'TTTGAT': 27398, 'GGATAT': 27398, 'CGTACC': 27367, 'TATTTC': 27363, 'CTTCAG': 27362, 'CCGGTA': 27357, 'CGTTGG': 27341, 'GCCCCA': 27340, 'AATCGC': 27311, 'TGCGAT': 27310, 'GAAACG': 27309, 'AAAACA': 27298, 'AGACCT': 27286, 'CAGTTG': 27258, 'TATCGG': 27254, 'GCAGAG': 27252, 'TGAGCT': 27239, 'ACTCGA': 27225, 'CTCCCC': 27206, 'AGTCCG': 27200, 'AGAAAT': 27177, 'GTGCGT': 27129, 'ACTCCG': 27126, 'TTGGCA': 27097, 'ACGCGT': 27061, 'CTTTCC': 27060, 'AAACGC': 27033, 'CAGAAA': 27017, 'TGAAAG': 27007, 'TACATC': 26993, 'AGTTCA': 26991, 'TGGAAT': 26990, 'TTTTGG': 26982, 'TCACGA': 26933, 'TCGAGT': 26930, 'GTTTTT': 26901, 'TCGATT': 26881, 'AGGATC': 26869, 'CCCTCC': 26868, 'ACAGCT': 26834, 'TGGAGT': 26831, 'GTGGTT': 26823, 'AAGAGG': 26789, 'GCGAAT': 26782, 'ATTTCA': 26781, 'GCCACA': 26752, 'GTTTCC': 26719, 'TACAAC': 26718, 'TGATTG': 26714, 'CCCTCA': 26708, 'AGCGAA': 26692, 'TGCACA': 26676, 'CTGATT': 26673, 'TTTATC': 26651, 'AGCCTC': 26638, 'CCTCAT': 26630, 'CGCTAT': 26613, 'TGGTTT': 26603, 'TTTTCG': 26590, 'GGGACA': 26588, 'ATCAAT': 26584, 'CTTGCG': 26578, 'TTCGCT': 26574, 'TGTGCT': 26566, 'GCTCTT': 26538, 'ACCCAC': 26525, 'GGTTCA': 26514, 'AAATAT': 26492, 'CACGTG': 26485, 'CAAAGA': 26482, 'TGAACT': 26478, 'CGAATT': 26474, 'ATCAAA': 26449, 'ACGGGA': 26442, 'ACCCCC': 26406, 'GCGGTA': 26405, 'CGGAGT': 26389, 'GTATCG': 26375, 'CCTTTG': 26367, 'TGGGAA': 26360, 'TCTCCC': 26355, 'ATTCTG': 26342, 'AAATCA': 26237, 'TACTTC': 26233, 'CGGTAT': 26205, 'ATACCG': 26197, 'GATTAT': 26189, 'CTGTTT': 26158, 'CCGAAT': 26158, 'TGTGCC': 26120, 'GCTTGA': 26110, 'GAGAAA': 26109, 'AGACGA': 26092, 'GCAATT': 26079, 'AAAACG': 26064, 'AGCATG': 26039, 'TGGTTG': 26011, 'CACTTC': 26008, 'TGCCTT': 25975, 'TTTCAA': 25939, 'CTTTTT': 25908, 'GGTTGG': 25905, 'GTCAGG': 25901, 'TTACCG': 25897, 'GAATTT': 25882, 'GAATCG': 25867, 'AGGTCC': 25867, 'AAGGGG': 25864, 'TGCAGT': 25813, 'AAATCC': 25799, 'GCATTT': 25792, 'CCTTGG': 25787, 'CCATTG': 25769, 'GCAAAC': 25765, 'GCTCTA': 25754, 'GGAATC': 25737, 'CCTACA': 25732, 'TTTTTA': 25717, 'CCCCAC': 25713, 'ATCCGA': 25712, 'AGGCAG': 25677, 'CTTTGG': 25673, 'TGGACT': 25653, 'GTTCAG': 25643, 'TACCTC': 25635, 'GGTCCC': 25634, 'CGAATC': 25624, 'CTACAT': 25599, 'TTCGAA': 25583, 'ACGGGT': 25554, 'AAAGCA': 25548, 'TGAAGT': 25533, 'AGCCCT': 25520, 'GCAAGT': 25476, 'AAGGAT': 25465, 'TCCGAT': 25435, 'GGTATC': 25434, 'CATTTC': 25370, 'AATCTG': 25352, 'ATGCCT': 25348, 'CTTTCG': 25342, 'TGGGGA': 25324, 'CAGTAC': 25318, 'TCAATC': 25312, 'GTGTGC': 25310, 'TCCATG': 25286, 'GGGATT': 25249, 'TACAGC': 25223, 'CAGTGC': 25214, 'AAGCCT': 25212, 'GATTGG': 25185, 'CCTATC': 25183, 'GTGCGA': 25180, 'GGTCCA': 25179, 'GGGCTA': 25155, 'CTCTGC': 25146, 'TACAAG': 25143, 'TTTCCC': 25142, 'GACCCA': 25136, 'TTTGCG': 25136, 'GAATAT': 25129, 'AAGCAC': 25128, 'GGATTG': 25109, 'CCCAGA': 25108, 'GGGGAA': 25103, 'TACGTG': 25085, 'AATATC': 25066, 'ATCTGA': 25037, 'AAAGAG': 25036, 'GCACTC': 25031, 'CAAAAG': 25026, 'GGAGAG': 25006, 'TTTTCA': 24999, 'AAGGCT': 24994, 'CGGAAT': 24989, 'CAACTC': 24960, 'TCAACT': 24934, 'GAATCC': 24911, 'TGCATG': 24908, 'CCTCTC': 24905, 'TTCCGA': 24859, 'TGTCGT': 24855, 'GAAGTT': 24851, 'AAACTG': 24841, 'ATTCGG': 24824, 'AAAGCT': 24819, 'GTTTGC': 24815, 'AACCCC': 24809, 'ACGGAG': 24808, 'GCACAT': 24801, 'AAACGG': 24770, 'AACGCT': 24761, 'GAGTTG': 24743, 'GCTGTA': 24722, 'GGTTCT': 24720, 'ATCCGT': 24705, 'CGTTGA': 24695, 'AGCCAC': 24689, 'AATTTT': 24687, 'TTGACG': 24675, 'GTTGAA': 24674, 'TGTTGA': 24670, 'AATATT': 24662, 'CAATCT': 24653, 'CGAAAG': 24609, 'AAATTT': 24604, 'TTTCAG': 24598, 'GACGTT': 24569, 'AGATCT': 24555, 'CATGTC': 24553, 'GTCCCC': 24550, 'CCTGAG': 24546, 'TTAAAA': 24466, 'TCCGAA': 24463, 'GTGTTG': 24459, 'GCCGTA': 24451, 'GTGTCC': 24408, 'GAATTG': 24405, 'TTGGGC': 24403, 'ATTCCC': 24378, 'GGCGTA': 24378, 'TGCTTC': 24373, 'CCGTTT': 24340, 'AGCCCC': 24337, 'AAACCC': 24333, 'TCGGTT': 24290, 'AAGCGA': 24288, 'CACCTA': 24266, 'AGGCAC': 24266, 'GACTCC': 24260, 'AGGTGT': 24240, 'ACGTCT': 24240, 'AGAGCA': 24236, 'ACCTTG': 24234, 'ATTGCT': 24208, 'TTTCCT': 24207, 'AGCTTT': 24207, 'GTCGTT': 24207, 'ATTTTA': 24204, 'CGAGTG': 24198, 'AGGTCT': 24197, 'AGCAAA': 24190, 'TTGCTC': 24187, 'ACGAGT': 24156, 'AGTTGC': 24152, 'TCAAAG': 24107, 'CGTGAG': 24100, 'CCGTGT': 24091, 'TGAAAC': 24084, 'ATAAAA': 24080, 'CCAACT': 24068, 'TTTCCA': 24047, 'GCATGT': 24043, 'ATTATC': 24040, 'AGGCAA': 24035, 'ATTTAT': 24028, 'CTATCG': 23987, 'CCAGTC': 23982, 'AGCCAT': 23981, 'AACGGT': 23973, 'AGTGGG': 23972, 'GAACTT': 23963, 'TGCAAA': 23946, 'TTTCTC': 23937, 'AGCCCA': 23925, 'TTATCG': 23905, 'AATGCG': 23895, 'CACGAA': 23892, 'TCAGGG': 23881, 'GGACTC': 23870, 'AACATG': 23855, 'GTGCCT': 23817, 'TGGGAT': 23812, 'TGTTGG': 23808, 'AGTGGA': 23806, 'GAATAC': 23795, 'TCCCGT': 23794, 'AATCGG': 23794, 'TGCAAG': 23792, 'CCAGTG': 23780, 'TTTTCT': 23771, 'GGGTCT': 23763, 'AATTGC': 23763, 'CTGGTA': 23759, 'CAATAT': 23746, 'GATTAC': 23739, 'CAGGGA': 23735, 'GGAGTC': 23715, 'CGATAC': 23700, 'CCCCAG': 23668, 'GTTGCA': 23646, 'GAGCCA': 23621, 'AAACAG': 23619, 'CAACTT': 23593, 'GTTTTC': 23591, 'TTCAAA': 23579, 'CAAACC': 23576, 'GGGTCC': 23560, 'CATTGA': 23546, 'CCATTC': 23517, 'ATTACC': 23508, 'CTACCC': 23507, 'AGAAGT': 23492, 'GTGAAA': 23492, 'AAATTG': 23489, 'GATACC': 23484, 'TTGGAA': 23470, 'CACTCG': 23426, 'AGGACT': 23425, 'GCAAAT': 23421, 'AGAGCC': 23397, 'TCCAAC': 23388, 'CATTGG': 23383, 'AACGAT': 23376, 'TGTCCA': 23373, 'ACCAGT': 23371, 'AACACG': 23367, 'TTGGCT': 23351, 'CTTGAT': 23343, 'AAATTC': 23326, 'AGCTCC': 23322, 'CCACAC': 23312, 'TATCAG': 23310, 'CAAAAT': 23294, 'TTGTGG': 23289, 'TATTGC': 23277, 'GGTATT': 23272, 'TACGGG': 23260, 'GTGCCA': 23252, 'CAATCG': 23235, 'AACTAC': 23235, 'CTTGCT': 23234, 'TCTGAT': 23221, 'TGCTTG': 23208, 'GGGTTG': 23196, 'AAGAAT': 23191, 'ACAAAA': 23185, 'TTGACC': 23177, 'CAAGTC': 23138, 'CTATGA': 23131, 'CCACTG': 23122, 'AATGAA': 23113, 'AAGTGC': 23110, 'TCTATG': 23110, 'CTTGAA': 23106, 'TTATTG': 23104, 'AATTCG': 23089, 'ACGGAT': 23078, 'GGGGGT': 23061, 'CCAATC': 23033, 'CGTCTA': 23032, 'CTCCAT': 23002, 'GGAATG': 22991, 'GAATGC': 22989, 'TGCTTT': 22967, 'CAAGAG': 22963, 'GTTTTG': 22934, 'AACGCA': 22928, 'TTGAGC': 22920, 'CCATGT': 22918, 'CTCCAA': 22912, 'GCCACT': 22909, 'GAGGCA': 22902, 'CCCCTC': 22877, 'GTACCT': 22873, 'TCTATC': 22870, 'ACTGCA': 22866, 'AGAACT': 22857, 'CATTCG': 22841, 'ATGATT': 22839, 'CGAATG': 22838, 'ATGCAC': 22836, 'CCTTGA': 22826, 'AGAAAG': 22819, 'AGAGCT': 22803, 'AGGAGT': 22790, 'CAGTCC': 22776, 'TTGCCT': 22775, 'CAATGA': 22769, 'TCTCCT': 22751, 'CACAAC': 22749, 'GTCCGT': 22748, 'ACGAAA': 22716, 'TCGCAA': 22708, 'TCCAAG': 22707, 'GCCTCT': 22704, 'CAAAAC': 22643, 'AATTTC': 22641, 'AACTGA': 22624, 'TGCCTC': 22618, 'CAAAGG': 22600, 'CCCCAA': 22592, 'CCTTTC': 22576, 'TTTACC': 22569, 'GAGTCC': 22559, 'ACGGTT': 22547, 'GGAGTG': 22531, 'AATGGA': 22497, 'CGAGAG': 22496, 'ACGGAA': 22492, 'GATTGA': 22482, 'CTATTT': 22468, 'TTTGCT': 22467, 'CTTTTC': 22467, 'TTTATG': 22456, 'TGAGGC': 22453, 'CTCATG': 22444, 'TGCGAG': 22430, 'TTCTTT': 22425, 'AAGCAA': 22406, 'AGCAAT': 22406, 'AGTCGA': 22384, 'AACGAA': 22358, 'ACATCT': 22348, 'GTGATT': 22330, 'AAACAA': 22303, 'ATTTCT': 22282, 'AACCAC': 22279, 'ATATTT': 22273, 'CGTTTT': 22271, 'GATTTG': 22268, 'TTGTTG': 22260, 'TATCAA': 22252, 'AGGAAT': 22223, 'CACAAG': 22201, 'CATTTT': 22194, 'TCGGAA': 22188, 'GAACCA': 22187, 'TGGTAC': 22179, 'TATGCG': 22163, 'AATGGG': 22160, 'GCTATT': 22141, 'CCAAAG': 22129, 'CTTTAT': 22127, 'CTTTGA': 22121, 'GTACTG': 22117, 'TTCGCA': 22108, 'ACATGA': 22103, 'TTTAAA': 22073, 'ACTTCT': 22062, 'ATCATT': 22058, 'TGCGTT': 22042, 'CAAGTT': 22042, 'GGGACT': 22034, 'AGCTTG': 22031, 'AGCTAC': 22019, 'AACAAA': 21997, 'TGGGGT': 21996, 'AGTGGT': 21995, 'AGGGTG': 21976, 'CAGCTA': 21976, 'ATATTG': 21973, 'TGACGT': 21967, 'AGGCAT': 21958, 'AAATGC': 21918, 'TCAGGT': 21918, 'CTATGG': 21912, 'TGAGGA': 21891, 'TGTTTG': 21869, 'ACTACA': 21851, 'ACCGTT': 21850, 'TCGCTT': 21843, 'GGTTGA': 21822, 'ACGATT': 21820, 'AAACCA': 21816, 'TATTGA': 21812, 'TTTCTT': 21802, 'AGCTCT': 21787, 'AAGATT': 21777, 'ATTTGC': 21719, 'GTTTCT': 21717, 'GGTTTG': 21687, 'GGTACC': 21680, 'TCGGAT': 21661, 'CTCCGT': 21641, 'TCGAAT': 21628, 'TGTTTT': 21616, 'AATCGA': 21613, 'GGGTTT': 21607, 'CAATTT': 21605, 'ACCCCA': 21603, 'TTGATT': 21587, 'CTTTTG': 21559, 'GTGCAT': 21555, 'CTTTCT': 21547, 'TCGTTG': 21520, 'AAAACT': 21516, 'TTTCAT': 21480, 'TCATTG': 21475, 'AAGAGA': 21475, 'TGGATA': 21472, 'AGCACA': 21472, 'GTTGGA': 21472, 'TGATTC': 21462, 'TACCGA': 21459, 'AATCAA': 21438, 'AACCGA': 21431, 'GGGGTT': 21404, 'GAGGCT': 21400, 'CTATCC': 21395, 'AAAGAC': 21393, 'AGCCAA': 21363, 'AGTGCG': 21356, 'GATAAA': 21327, 'ATCGGA': 21324, 'CTCCCT': 21288, 'ATTCAG': 21271, 'TTCGGA': 21264, 'GTTTCA': 21261, 'ATGTCC': 21239, 'CCAAAA': 21221, 'ACCTCT': 21213, 'TTCCGT': 21196, 'TCGTGT': 21183, 'ACGCTT': 21176, 'ATGAGG': 21170, 'TCCTCT': 21170, 'GATTCT': 21163, 'GTTGGT': 21149, 'GCACTT': 21148, 'AGATTT': 21134, 'TTATCA': 21123, 'AAATTA': 21120, 'GCTACT': 21113, 'ATGCGA': 21083, 'TGAATT': 21042, 'GGGAAT': 21033, 'CCACAA': 21021, 'AATATG': 21020, 'GCTTTA': 21012, 'CCCTAC': 21006, 'TTGAAC': 20988, 'TATGAA': 20971, 'GACAGG': 20969, 'TATCTC': 20965, 'CAGGGT': 20945, 'TCCCCA': 20858, 'ATTACG': 20853, 'AATTAT': 20849, 'GAGATT': 20837, 'GTCAAT': 20834, 'AGCCTT': 20829, 'TCCCCC': 20828, 'TCGATA': 20803, 'AAATGA': 20795, 'ACTGAT': 20769, 'CCGATA': 20749, 'TTTGGT': 20739, 'AATACC': 20735, 'TTTTAA': 20729, 'TCTTTG': 20715, 'TCTCGT': 20696, 'AGTACC': 20685, 'AAAAGT': 20678, 'AAAGGA': 20667, 'AGGGGG': 20650, 'CGTGTG': 20649, 'ATTCAA': 20621, 'TGTTTC': 20618, 'GGTCTA': 20617, 'CAGTTT': 20600, 'AGTGCC': 20566, 'ATGAAT': 20562, 'CGTATC': 20556, 'CAATCA': 20542, 'AGGATT': 20526, 'TAAAGA': 20524, 'ATGGGA': 20516, 'CATTCT': 20509, 'ACGCAT': 20496, 'TATTCC': 20492, 'TATTGG': 20458, 'GTTGTT': 20443, 'ATTCGA': 20441, 'ATTGGT': 20432, 'GTGAGG': 20425, 'CCTCAG': 20419, 'CATGTG': 20410, 'GTTGGG': 20397, 'TGCACT': 20357, 'GACAAA': 20352, 'ATATGG': 20351, 'ATTCCT': 20328, 'AACCGT': 20318, 'ACCACA': 20314, 'CTCAAT': 20313, 'TTTGGA': 20310, 'GAGTTT': 20306, 'GAAACA': 20305, 'CGAGTA': 20290, 'CGGGTA': 20282, 'ACAACT': 20278, 'AGTTGG': 20277, 'ATTCTC': 20264, 'AGAAAC': 20261, 'GTATTT': 20250, 'CTCAAA': 20234, 'CTCGTT': 20230, 'TGTCCT': 20229, 'ATCTTT': 20217, 'AATGAT': 20201, 'AATTGA': 20195, 'AAACCT': 20167, 'ATTGCA': 20155, 'CCCCCA': 20133, 'TTTCAC': 20128, 'TCAAAC': 20101, 'TACTAC': 20067, 'AAGTAC': 20064, 'TTATGG': 20058, 'TCTTTT': 20037, 'AACTCC': 20029, 'TTTTAC': 20023, 'GATTTA': 20022, 'TTCAAT': 20012, 'CTCAGG': 20011, 'AGCGTT': 20002, 'AATTCC': 19983, 'TCATTC': 19982, 'GTCCAT': 19974, 'ACGCAA': 19972, 'CCTCTG': 19963, 'TGTACC': 19960, 'CAGATT': 19950, 'AGGCTC': 19939, 'CTATCA': 19937, 'TTTCGT': 19918, 'TATTAT': 19917, 'GACAAT': 19914, 'CATGAG': 19899, 'AATTGG': 19890, 'GTTTGG': 19873, 'TCTTTC': 19871, 'TCGCAT': 19858, 'CTATGT': 19855, 'CAAACG': 19845, 'GTTTGA': 19819, 'CCACAT': 19815, 'AAGCTT': 19811, 'ATTATG': 19805, 'AGGGGA': 19804, 'AATACG': 19803, 'ATATCA': 19789, 'ATCGTT': 19766, 'AATCAG': 19757, 'GGGTAC': 19752, 'TTTGCA': 19752, 'CAATTG': 19721, 'AATCCC': 19718, 'TTCCAT': 19715, 'GCAGTA': 19713, 'GTGCTT': 19698, 'TCTTGC': 19697, 'TTGTTC': 19688, 'AGGCTT': 19686, 'ACCATT': 19679, 'GAAATA': 19679, 'CAATGT': 19676, 'CCCCCT': 19664, 'GACATT': 19653, 'ACCTAT': 19651, 'TATTCG': 19646, 'GATCTA': 19640, 'CCTCTT': 19637, 'TTATTA': 19631, 'CATGTT': 19626, 'GATTCA': 19601, 'TCTGAA': 19558, 'AGCATT': 19541, 'AATGGT': 19540, 'GCGATA': 19524, 'ATGGTT': 19524, 'CAAATC': 19507, 'GTGCAA': 19497, 'CATCTA': 19490, 'GATATG': 19469, 'AGAATT': 19464, 'CGCGTA': 19461, 'ACAGGA': 19460, 'GGATAC': 19450, 'ATATTC': 19421, 'ATTTGG': 19417, 'CACTAC': 19402, 'TTGCAA': 19400, 'ATCTCA': 19386, 'ACCTTT': 19374, 'AAGTCC': 19351, 'ACCAAT': 19344, 'AAGACA': 19341, 'GGTTAT': 19336, 'GAGTGC': 19313, 'AATTCA': 19313, 'GATACG': 19302, 'TGTGCA': 19276, 'ATATCC': 19270, 'CCCTTG': 19261, 'CGTTTG': 19243, 'CACGTT': 19237, 'TTATCT': 19208, 'AGTTCT': 19202, 'CCTTTT': 19187, 'TTGGAG': 19179, 'CTTCTA': 19166, 'TCCAGT': 19159, 'ATTAAA': 19143, 'GAAACT': 19128, 'TTATGA': 19119, 'CTTTCA': 19104, 'AAGCGT': 19086, 'GTCTAT': 19067, 'TCTCAA': 19049, 'CCATTT': 19032, 'AACAAT': 18997, 'AGATGT': 18973, 'GGGAGT': 18971, 'CGATTA': 18970, 'ATGTTG': 18942, 'TTTTGT': 18937, 'TTACGG': 18937, 'GTTTAT': 18933, 'AAATAC': 18915, 'CAAATT': 18912, 'ATCTGT': 18904, 'TTGAGG': 18901, 'CACAGG': 18898, 'TTTGTC': 18866, 'CATTCA': 18852, 'ACACGA': 18852, 'ATACGG': 18851, 'GTACGT': 18849, 'CTACTG': 18842, 'AGAGGC': 18834, 'AGTTTT': 18833, 'TCCCTC': 18833, 'ATTTGA': 18830, 'AACGGA': 18822, 'CCAGAG': 18816, 'TGTCCC': 18802, 'TCTGTT': 18793, 'TCCCCT': 18751, 'CTACTT': 18746, 'AAAGTG': 18745, 'TACCAC': 18738, 'TTTGGG': 18728, 'TGCAAT': 18703, 'AATTTG': 18702, 'ATGGGT': 18695, 'GTCAAA': 18689, 'TACTGC': 18683, 'TCACTG': 18673, 'TCCGTT': 18665, 'GACTAT': 18664, 'GTATTG': 18649, 'TGGTAT': 18648, 'TATGAC': 18643, 'TGAATG': 18612, 'AATTTA': 18589, 'TTCTGA': 18588, 'GGGGGG': 18559, 'ATCCAT': 18533, 'GTTGTC': 18517, 'TGGGTT': 18513, 'GTTGAC': 18509, 'GAAAGG': 18466, 'AGCACT': 18456, 'TTGGTC': 18453, 'CTTTAC': 18450, 'TGCGAA': 18446, 'CTCTAT': 18426, 'CGGATA': 18423, 'GCGTAT': 18421, 'CAATAC': 18412, 'TACGAA': 18408, 'ATTGGG': 18382, 'TTTGAC': 18376, 'TCCCAG': 18364, 'TCCACA': 18350, 'TATTAC': 18326, 'CCTCTA': 18315, 'CCAAAC': 18308, 'TCAATT': 18308, 'ATCTTG': 18305, 'TATGGA': 18305, 'TTGGGG': 18289, 'AACAGA': 18278, 'ATCTAT': 18275, 'AATAAA': 18274, 'TTGGAT': 18262, 'CACACG': 18235, 'ATTCCA': 18221, 'CTGAAT': 18207, 'CAGGTT': 18191, 'GTGTAC': 18191, 'GAAAGT': 18190, 'ATTGGA': 18166, 'AGATTG': 18163, 'GTCATT': 18159, 'CTCCTT': 18159, 'CAACTA': 18158, 'TCAAGT': 18148, 'AGTGCT': 18131, 'TTGCGA': 18108, 'CAAACT': 18100, 'ATGCGT': 18063, 'ATGACA': 18058, 'AAGCAT': 18056, 'AAAGGG': 18054, 'GCAATA': 17997, 'TTGGTT': 17979, 'TTTGTT': 17977, 'AATCCT': 17970, 'ATACGC': 17966, 'CAATTC': 17965, 'CCACAG': 17964, 'CCCCAT': 17961, 'TACTCG': 17946, 'TTACGA': 17919, 'CAAACA': 17908, 'TAAAAG': 17898, 'ATGTCA': 17897, 'ACAGGG': 17894, 'TACACG': 17894, 'CTTGCA': 17887, 'CCAAGT': 17856, 'ACCCCT': 17853, 'CCCAGT': 17841, 'ATAAAG': 17828, 'CCTGTA': 17819, 'ACGAAT': 17802, 'TTCATT': 17800, 'AAGATA': 17748, 'AAGGGA': 17735, 'CCCTCT': 17733, 'GGATTA': 17725, 'TTTACG': 17701, 'GCTTAT': 17688, 'TCAGAA': 17683, 'GTCCAA': 17677, 'TATTTG': 17655, 'CTACAG': 17652, 'ACAGGT': 17636, 'AAAGGT': 17610, 'TTATGC': 17582, 'CAAATG': 17577, 'TCTGAC': 17556, 'ACAAAG': 17552, 'TTATTC': 17548, 'TTAAAG': 17529, 'ACGTTG': 17525, 'AGCTAT': 17514, 'TTGATA': 17513, 'GGGTAT': 17493, 'GACTTT': 17489, 'AACCCT': 17460, 'TTACAA': 17456, 'TATTTA': 17447, 'CAAGTG': 17443, 'CGAATA': 17434, 'TACGGT': 17433, 'TATGTG': 17426, 'TGCATT': 17422, 'GTGTTT': 17411, 'ACCAAA': 17406, 'GGAATA': 17393, 'AACTCA': 17388, 'TCATTT': 17385, 'AATGCT': 17384, 'ATTGTT': 17380, 'ATTGTC': 17369, 'AGGGAA': 17349, 'TTCTTG': 17347, 'TTGTTT': 17344, 'TATCCC': 17333, 'TTACGC': 17329, 'ATCAGA': 17326, 'GGTATG': 17321, 'TTCCTT': 17321, 'CTATTC': 17318, 'TTCCAA': 17316, 'CACTCC': 17311, 'AGTTTC': 17304, 'AGATTC': 17298, 'GTTATT': 17295, 'TCCTAC': 17285, 'ATCCCT': 17254, 'TATGAT': 17250, 'CACTGA': 17249, 'GTTTTA': 17241, 'AGAGAA': 17231, 'GTATTC': 17227, 'ACAATG': 17207, 'ATCAGT': 17207, 'TTTGTG': 17204, 'TTGGGA': 17197, 'AAATCT': 17197, 'ACGTGT': 17195, 'CTTGAG': 17190, 'ATTACA': 17189, 'GACCTA': 17184, 'TTCCCT': 17170, 'TCAAAT': 17168, 'AACCAT': 17162, 'ATCCTT': 17160, 'GGTTAC': 17155, 'ATTCAT': 17154, 'CCCCTT': 17150, 'AATCCA': 17147, 'CGTTGT': 17142, 'ATTCTT': 17137, 'TCTTGA': 17121, 'ATGCTT': 17115, 'GTACAA': 17112, 'TCGGTA': 17108, 'CATTAC': 17100, 'TGTGAT': 17083, 'TTATCC': 17081, 'ACAATC': 17080, 'GTATGG': 17076, 'TGATAT': 17063, 'CACATG': 17037, 'GAGGGA': 17036, 'GAAGTA': 17014, 'GAGGGT': 17007, 'ATCCCA': 16992, 'GGTAAA': 16987, 'CATTAT': 16975, 'GCCCTA': 16965, 'GAGAAT': 16964, 'TCGTTT': 16946, 'ATTTAC': 16932, 'CAGTGA': 16929, 'TTTAAT': 16926, 'TTGCTT': 16922, 'CTTGGT': 16914, 'TTCTAT': 16906, 'TATGTC': 16903, 'TCAGTG': 16900, 'ACCACT': 16899, 'CTTTGT': 16894, 'AATCAT': 16887, 'TTCTCA': 16886, 'AGATAT': 16876, 'GGACTA': 16864, 'ACTGAA': 16860, 'TGTTGT': 16845, 'AGAGAT': 16844, 'ATGCAA': 16835, 'ATGTAC': 16812, 'TATTCA': 16789, 'GTCCCT': 16783, 'GAATTA': 16779, 'CGTACT': 16770, 'GGTACT': 16762, 'AAACGA': 16753, 'TTACCT': 16747, 'TACATG': 16738, 'TCCTTG': 16731, 'ACTGAC': 16725, 'TACGAT': 16711, 'AATCTC': 16706, 'CCCACA': 16696, 'ATGTGA': 16694, 'GATTGT': 16631, 'TATCGT': 16627, 'CGGTTA': 16623, 'ACAGAA': 16616, 'ATTTAA': 16611, 'TGAATC': 16601, 'ACAGAT': 16594, 'TGGCTA': 16573, 'AGAGGA': 16557, 'ACATTG': 16517, 'GTACCC': 16508, 'TTCCCA': 16502, 'GTCTTT': 16498, 'TCAGTT': 16490, 'AAGTGA': 16489, 'CCGTAT': 16488, 'ACTTTG': 16486, 'TACAAA': 16483, 'AGTTGA': 16477, 'CTGTAT': 16468, 'ATTCGT': 16467, 'GGTTGT': 16460, 'ACTCCA': 16460, 'ACTATC': 16453, 'GTTATC': 16437, 'ACTCCC': 16433, 'CCTATT': 16393, 'GTCCTT': 16387, 'GTATCC': 16366, 'AGTTTG': 16358, 'AAACTC': 16349, 'ACATTC': 16346, 'CTACTC': 16328, 'AAGGTT': 16327, 'AATGTC': 16326, 'TCTATT': 16324, 'CTTGGA': 16305, 'TCAGAT': 16304, 'TTCAGA': 16272, 'GCATTA': 16271, 'ATATTA': 16270, 'AAATAA': 16256, 'TTGTGC': 16248, 'AACCTT': 16237, 'CCACTC': 16225, 'TGTCAC': 16215, 'CGGTAA': 16211, 'AATTAC': 16208, 'ACCCAT': 16206, 'TTGAAT': 16185, 'TACCGT': 16182, 'AGTACA': 16173, 'CTTGGG': 16159, 'TGACAA': 16159, 'TGTCAT': 16149, 'AAGTTT': 16135, 'TATCAC': 16130, 'GAATCT': 16116, 'GGTGTA': 16102, 'GTAAAA': 16099, 'TCAATA': 16098, 'TTACCA': 16093, 'TCATGT': 16088, 'ACAAGT': 16085, 'ATATGC': 16083, 'CCACTT': 16045, 'TCCTTT': 16013, 'AGTGAT': 16002, 'GAATCA': 16002, 'GATGTA': 15982, 'CTCCCA': 15970, 'TATGGG': 15936, 'GTGAAT': 15929, 'ACTTTT': 15906, 'AAGGGT': 15904, 'ATTGAC': 15900, 'ATGTTT': 15883, 'AAAGTT': 15869, 'ACAAAT': 15857, 'GTATCT': 15854, 'TCCCAC': 15840, 'TATATC': 15840, 'TCTACT': 15832, 'CTTTTA': 15819, 'TGATTA': 15809, 'AACTTT': 15808, 'GTACCA': 15799, 'CAGAAT': 15782, 'TACCCC': 15775, 'TGATAA': 15759, 'TGTCTG': 15740, 'CCAGTA': 15716, 'AGGGAC': 15697, 'TCTGTC': 15685, 'ATATGA': 15672, 'ACTCAA': 15655, 'CCTACT': 15654, 'TTGCAC': 15645, 'TTTACA': 15641, 'TGAGTT': 15639, 'ACTCCT': 15631, 'TGAGGT': 15613, 'GGAGTA': 15612, 'AGGTTC': 15610, 'ACTTTC': 15607, 'TCTTGG': 15582, 'AATGTG': 15581, 'GTGACA': 15581, 'ATTGAG': 15578, 'CTCTTT': 15567, 'AATGAC': 15563, 'CGTATT': 15562, 'TCCCAA': 15562, 'GTTCTA': 15554, 'TAAAAT': 15553, 'GTTGAG': 15517, 'ATCTCT': 15513, 'ATTCAC': 15508, 'ACTACT': 15487, 'TATAAA': 15485, 'AATGCA': 15483, 'TTCGTT': 15441, 'TTTATA': 15440, 'AACATT': 15410, 'GTATGC': 15408, 'CGTACA': 15371, 'TTTGAG': 15369, 'AAGTCA': 15352, 'TTGGAC': 15342, 'CATATC': 15316, 'CTGACA': 15314, 'CAATTA': 15313, 'ATATCT': 15274, 'ACAAAC': 15255, 'CCCGTA': 15241, 'GCTTGT': 15241, 'TGACAG': 15224, 'CCAATT': 15221, 'AACCAA': 15214, 'CAAAGT': 15200, 'CATTGT': 15200, 'TTGCGT': 15156, 'TCGTAC': 15153, 'CTTGTC': 15148, 'GTATCA': 15131, 'AGGGAT': 15126, 'AATTAA': 15114, 'TTATAT': 15114, 'TGAGGG': 15112, 'GTACAC': 15076, 'TCACGT': 15063, 'AGGTTT': 15059, 'AGAATG': 15028, 'TATCTT': 15018, 'AACCCA': 15009, 'AATTCT': 15007, 'TGAGAA': 14998, 'CGTTAC': 14997, 'ACAATT': 14996, 'AATACA': 14993, 'GTCTGA': 14987, 'GTTACG': 14979, 'AATCGT': 14955, 'CTATAT': 14954, 'CCCTTT': 14950, 'AAACTT': 14948, 'AAACAT': 14948, 'TATATT': 14946, 'CAAGTA': 14941, 'GTTGTG': 14938, 'AGAATC': 14934, 'CATTTG': 14930, 'AAGTTG': 14922, 'TACGGA': 14922, 'CCCTAT': 14914, 'GGTACA': 14914, 'CTATTG': 14898, 'TACGCT': 14890, 'TTGTCC': 14877, 'ACTCGT': 14851, 'ATACCT': 14841, 'GAGGTT': 14835, 'AAAGTC': 14832, 'GTCAGT': 14830, 'TCCATT': 14812, 'GTTACC': 14812, 'CAGTAT': 14804, 'AATCTT': 14799, 'ATTGTG': 14797, 'TGAGAT': 14786, 'CAGAGG': 14781, 'AGGGGT': 14762, 'TATGGT': 14723, 'ACTGTT': 14712, 'TTGTCA': 14711, 'ATCCAA': 14701, 'TTACCC': 14701, 'AGTATT': 14683, 'CCCAAA': 14651, 'GTTTGT': 14650, 'TAGAAA': 14645, 'ACACTG': 14640, 'CCTTTA': 14640, 'AGTGAA': 14629, 'TGAATA': 14628, 'ATTAAT': 14626, 'TCTCTG': 14625, 'TATTCT': 14587, 'CTGAGG': 14573, 'AGGTTG': 14540, 'CTGTCA': 14524, 'TGTCAG': 14509, 'TATTAA': 14504, 'ACGTAC': 14484, 'CGTTAT': 14468, 'CAGTCA': 14448, 'CCAATA': 14438, 'CTTGAC': 14432, 'CAAATA': 14424, 'TCCAAA': 14399, 'TATCCA': 14397, 'ACTATG': 14386, 'CCCAAT': 14375, 'ACTTGC': 14354, 'AGGGAG': 14350, 'GACAGT': 14337, 'AGTCCA': 14329, 'AACTGT': 14317, 'TGTATC': 14316, 'GTGGTA': 14312, 'GTTATG': 14294, 'AGACGT': 14291, 'GCTTAC': 14289, 'TATCAT': 14288, 'GTACAG': 14274, 'TGACAT': 14261, 'TGTATT': 14244, 'GAATGA': 14224, 'GCATAT': 14221, 'ACGGTA': 14207, 'CTTATC': 14189, 'TTACAG': 14187, 'TGCCTA': 14176, 'AATATA': 14176, 'CTACTA': 14170, 'AACAGT': 14163, 'TGTATG': 14122, 'CGATAA': 14113, 'CTTATT': 14111, 'AATCAC': 14106, 'CCTTGT': 14094, 'TCTTTA': 14084, 'TTACTG': 14082, 'TCCACT': 14076, 'TTCTCT': 14068, 'AGGATA': 14066, 'TACCAA': 14062, 'TACTCC': 14050, 'ATACCC': 14048, 'TGACAC': 14039, 'ATGAGA': 14033, 'GGTTTA': 14033, 'GTCTTG': 14012, 'AAGAGT': 14009, 'TTAATG': 14009, 'ACCCAA': 14007, 'CTTGTT': 13998, 'TTCTGT': 13985, 'TTTTAG': 13979, 'CCAAAT': 13974, 'AGACAA': 13954, 'TATGAG': 13933, 'TATCCT': 13932, 'CCCCCC': 13919, 'AGATAA': 13901, 'TGGGTA': 13893, 'ATCACT': 13872, 'GAGTAT': 13871, 'TACGCA': 13860, 'ATGATA': 13859, 'AGTCGT': 13855, 'TACAGG': 13853, 'AAACAC': 13843, 'GAACTA': 13832, 'GTTTAC': 13832, 'GGGATA': 13831, 'AACTAT': 13830, 'TAAAGG': 13827, 'ATTACT': 13823, 'GGATAA': 13822, 'ACAGTT': 13792, 'GTACTT': 13790, 'TTGGGT': 13777, 'TGTACA': 13774, 'GTACAT': 13770, 'TCTCAC': 13769, 'CAGGTA': 13757, 'ATACCA': 13752, 'TCAGTC': 13751, 'ACCCTT': 13698, 'GATAAT': 13686, 'AATAAT': 13686, 'AGAGGT': 13642, 'AACTTG': 13638, 'TGCTAC': 13632, 'ATGTCT': 13624, 'GACTTG': 13612, 'ATTTGT': 13589, 'TTCAGT': 13576, 'GCGTTA': 13541, 'AGGGTT': 13535, 'GTACTC': 13487, 'TCTCTC': 13478, 'TAAAAC': 13474, 'TAATGA': 13462, 'GACAGA': 13453, 'TCCCTT': 13439, 'AAAGTA': 13435, 'AGTATC': 13420, 'GAGCTA': 13418, 'TATGCT': 13407, 'AGTCTG': 13370, 'AAGTAT': 13369, 'CAGACA': 13368, 'TTTACT': 13326, 'ACAGTG': 13317, 'AGAGTT': 13308, 'ATCACA': 13298, 'CACTTT': 13297, 'TTAATT': 13289, 'ATGCAT': 13239, 'ACTGTC': 13232, 'TGACTG': 13227, 'AGTGCA': 13222, 'ATACGA': 13219, 'ACTTGG': 13215, 'TACCTT': 13209, 'GCACTA': 13162, 'CTATTA': 13159, 'CTGCTA': 13151, 'TATTGT': 13077, 'GAATGT': 13063, 'TTTAGA': 13055, 'AGTCCT': 13055, 'CCCACT': 13047, 'ACGTTT': 13046, 'CCCATT': 13021, 'CACAAA': 13017, 'TATGCA': 13008, 'CCATTA': 13003, 'ACTATT': 12985, 'ACATTT': 12981, 'CTTATG': 12970, 'TTGGTA': 12960, 'AAATGT': 12944, 'ATAAAT': 12943, 'TGTGAA': 12940, 'TACTTT': 12938, 'GTCTGT': 12936, 'TTTCTA': 12927, 'AGGGTC': 12924, 'TTAAAT': 12921, 'GGCTTA': 12920, 'TAAAGC': 12918, 'GTATGA': 12875, 'AATGTT': 12858, 'CTCATT': 12842, 'AGAGGG': 12828, 'TTAGAA': 12822, 'ACGATA': 12817, 'TGGTTA': 12809, 'TTATGT': 12766, 'CTCTTG': 12702, 'TGACTT': 12698, 'CATTTA': 12696, 'CAGATA': 12694, 'TGTTTA': 12693, 'AGTCCC': 12684, 'CTGTCT': 12625, 'ATGAGT': 12615, 'AGAATA': 12605, 'CAGTCT': 12601, 'TCTCTT': 12571, 'CTTACC': 12549, 'CGTATG': 12546, 'TAATTT': 12528, 'AATTGT': 12520, 'TATACC': 12505, 'GAGAGG': 12489, 'TTATAA': 12473, 'GTCAGA': 12460, 'GTCCCA': 12455, 'ATGTAT': 12445, 'TGTCAA': 12441, 'AAGACT': 12418, 'TTGCAT': 12410, 'AAGGTA': 12385, 'AATGAG': 12373, 'TTACTT': 12365, 'AACGTT': 12357, 'TATATG': 12349, 'CTTTAA': 12334, 'TGTGTT': 12329, 'TAGAAG': 12324, 'ATTATA': 12319, 'TCTGTG': 12296, 'ATACAA': 12291, 'TGTACT': 12259, 'CAGAGA': 12255, 'TGGTAA': 12251, 'TCTCAG': 12234, 'GCCATA': 12227, 'TTGAGA': 12222, 'GTGTCA': 12182, 'AAACGT': 12175, 'GTCACT': 12171, 'TCGCTA': 12167, 'GGCATA': 12160, 'TCCCAT': 12153, 'AGACAG': 12139, 'CATATT': 12139, 'CGCTTA': 12125, 'TCTCAT': 12120, 'CACTGT': 12120, 'CATACC': 12114, 'GACACT': 12113, 'ATATAT': 12113, 'TCCAAT': 12112, 'TGTCTT': 12098, 'ATAGAA': 12087, 'CCTTAT': 12083, 'GCTATA': 12056, 'AGACAT': 12054, 'CACTCA': 12048, 'GGGGTA': 12048, 'GACTGA': 12044, 'AGATTA': 12041, 'ATGACT': 12022, 'TCATTA': 12017, 'ACTTGA': 12015, 'TCAGAC': 12010, 'ACATGT': 12001, 'CACAGA': 11992, 'TCACAG': 11982, 'ATAATG': 11976, 'TCCTAT': 11946, 'CTGATA': 11921, 'TGCTAT': 11905, 'ACTCAC': 11879, 'CCTATA': 11857, 'CCGTTA': 11839, 'AGTACT': 11821, 'AGTCAG': 11809, 'GCATAC': 11788, 'CTATAC': 11776, 'CTAAAA': 11771, 'CCATAT': 11767, 'TAAATT': 11753, 'GTAAAG': 11740, 'ATGGTA': 11717, 'TCACTC': 11706, 'TGTCTC': 11703, 'GCTAAA': 11684, 'TGTGAC': 11662, 'ACACGT': 11652, 'GTTAAA': 11650, 'GTATTA': 11635, 'CTCCTA': 11629, 'TAATAA': 11614, 'AGTATG': 11608, 'GACGTA': 11590, 'CTTGTG': 11585, 'CGTTTA': 11579, 'ATACAG': 11561, 'GAGATA': 11551, 'GGTAAT': 11542, 'CCACTA': 11539, 'GCCTTA': 11538, 'ACTTTA': 11533, 'ACAGAC': 11523, 'CAATAA': 11500, 'TTCACT': 11488, 'TACCCA': 11475, 'GATAAC': 11463, 'AGTTAT': 11375, 'CTGACT': 11373, 'GGTAGC': 11340, 'AGGTAT': 11319, 'CTTAAA': 11314, 'TATGTT': 11312, 'AACTCT': 11260, 'AGTGAC': 11259, 'ACTCTG': 11244, 'TACCCT': 11244, 'AATACT': 11218, 'AGTCAT': 11213, 'TGCGTA': 11210, 'AGTTGT': 11207, 'TTGTGA': 11206, 'ACCGTA': 11203, 'CCCCTA': 11191, 'ATTTAG': 11184, 'GATAGC': 11182, 'ATACTG': 11181, 'TGAGTG': 11169, 'ACAGTC': 11163, 'GGGTTA': 11154, 'ATAATA': 11148, 'ACAATA': 11134, 'TTGTCT': 11125, 'TCACAA': 11122, 'GATAAG': 11119, 'TTTAAC': 11115, 'TACCAT': 11062, 'TGTTAT': 11054, 'ATATAA': 11043, 'GTGAGA': 11022, 'GATACA': 11010, 'TCACTT': 10977, 'TACTGA': 10974, 'TGATAC': 10947, 'TTACAT': 10946, 'AAATAG': 10945, 'TACAGA': 10938, 'TGCTTA': 10922, 'CTTACG': 10916, 'GTCTCA': 10911, 'TTAATA': 10890, 'AGTTTA': 10888, 'CATTAA': 10884, 'GAGACA': 10882, 'TTCACA': 10873, 'ACTCAG': 10853, 'CTCAGT': 10830, 'CAGTGT': 10821, 'AAGCTA': 10821, 'CGGTAG': 10815, 'TCACAC': 10809, 'TACAAT': 10802, 'AGTCAA': 10772, 'CCTTAC': 10771, 'CACTTG': 10771, 'GATACT': 10761, 'TTGACA': 10730, 'GTGAGT': 10728, 'CACAAT': 10704, 'TAATGG': 10691, 'GTGACT': 10631, 'GTGTAT': 10627, 'GATATA': 10607, 'TAAATA': 10606, 'CTCTGA': 10601, 'GTCGTA': 10600, 'AGACTT': 10599, 'ATGTGT': 10593, 'ACTGTG': 10591, 'CTCTCT': 10588, 'GACTGT': 10580, 'TTGCTA': 10564, 'TAACGG': 10552, 'CGTGTA': 10548, 'TTGTTA': 10506, 'GAGAGA': 10502, 'AGGTAC': 10496, 'AAGTCT': 10478, 'CAGTTA': 10477, 'GGGTAA': 10475, 'GGTAAC': 10464, 'CATGTA': 10461, 'GTGTGA': 10460, 'AGATAC': 10460, 'CTCACT': 10447, 'TTTAGC': 10446, 'TCCGTA': 10442, 'CACATT': 10423, 'AACACA': 10410, 'TCTATA': 10370, 'GTCACA': 10368, 'AGACAC': 10331, 'TAGCGG': 10319, 'ATATGT': 10299, 'AGGTAA': 10293, 'AGGCTA': 10274, 'TTACGT': 10264, 'GTATGT': 10238, 'GGTAAG': 10237, 'CTGAGT': 10212, 'ACTCAT': 10197, 'ATAATT': 10192, 'ATAACG': 10188, 'TTAATC': 10168, 'GTTACA': 10159, 'TAGCCG': 10151, 'AGACTG': 10132, 'AGTCAC': 10125, 'GTTTAA': 10119, 'GATTAA': 10117, 'CCATAC': 10107, 'GACACA': 10098, 'ACACAG': 10070, 'ACTTAT': 10063, 'AGCTTA': 10030, 'TCAGAG': 10028, 'TTTGTA': 10023, 'AAGTTA': 10018, 'TACATT': 10015, 'CTCTCA': 9989, 'GACTCA': 9985, 'ACTCTC': 9970, 'CTCAGA': 9967, 'TCTTGT': 9966, 'TCTTAT': 9960, 'GTGATA': 9952, 'CACTAT': 9937, 'ACAGAG': 9928, 'TAATGC': 9890, 'GTCTCT': 9884, 'ACGCTA': 9884, 'CTGTGT': 9875, 'CTATAA': 9861, 'CAGACT': 9834, 'GAGTGA': 9817, 'GATAGA': 9814, 'TTGTAT': 9810, 'ATACTT': 9803, 'AGTGTT': 9744, 'TCGTAT': 9743, 'AACACT': 9710, 'CTGTGA': 9693, 'TTTAAG': 9679, 'CGCATA': 9676, 'TGGTAG': 9670, 'TGTGTC': 9659, 'ACATTA': 9657, 'CACAGT': 9640, 'TAACCG': 9620, 'CATACG': 9619, 'CACTCT': 9616, 'TTGAGT': 9611, 'GTAGCG': 9598, 'GTGTCT': 9586, 'AGTAAA': 9584, 'CATATG': 9567, 'ATAAAC': 9565, 'ATAGCG': 9557, 'TAATAT': 9557, 'GTAATG': 9521, 'TGTGTG': 9500, 'CACACT': 9499, 'TGTTAA': 9498, 'TTACAC': 9496, 'ATAACC': 9488, 'TGATAG': 9476, 'GTATAT': 9472, 'ATAGCC': 9443, 'AAACTA': 9440, 'ACACTC': 9439, 'TAACGA': 9429, 'GTAGAA': 9418, 'TAATCA': 9407, 'GGTAGA': 9404, 'TCTCTA': 9379, 'AGCGTA': 9346, 'CATAAA': 9330, 'GTTACT': 9324, 'TTAACG': 9323, 'ATAATC': 9317, 'AAGTGT': 9308, 'TCAGTA': 9302, 'TATAAT': 9296, 'TTATAC': 9294, 'TTACTC': 9290, 'TATACG': 9286, 'TAAATG': 9280, 'TTCCTA': 9272, 'AATAGA': 9268, 'TTAAAC': 9264, 'AATTAG': 9249, 'GCTTAA': 9244, 'TTCTTA': 9227, 'TAAACA': 9202, 'GAGTCA': 9197, 'CAGAGT': 9191, 'CTCTGT': 9190, 'GTAGCC': 9188, 'TACTCA': 9163, 'GAGAGT': 9110, 'CTTACA': 9096, 'GTTAAT': 9088, 'AGAGAG': 9086, 'TAGATG': 9077, 'ATTAAC': 9070, 'TGAGTC': 9055, 'TACTAT': 9052, 'AAGTAA': 9047, 'ACGTAT': 9046, 'GAGGTA': 9031, 'ACACAA': 9026, 'AGAGAC': 9020, 'TCATAT': 8994, 'TAAGAA': 8993, 'CTAAAG': 8985, 'TTAACC': 8984, 'CTGTTA': 8934, 'TCACAT': 8929, 'TACTTG': 8919, 'TATCTA': 8909, 'TGACTC': 8905, 'TAAGGA': 8896, 'ATAAGG': 8887, 'GAGTCT': 8886, 'TCTGAG': 8878, 'AGAGTG': 8863, 'GTCCTA': 8857, 'TATAGA': 8842, 'ACTCTT': 8837, 'AGTGTG': 8835, 'TGAGAC': 8827, 'TGCTAA': 8822, 'GTAACG': 8813, 'TAAAGT': 8813, 'GCGTAA': 8809, 'AGTCTT': 8804, 'TACGTT': 8796, 'ATACAT': 8765, 'GACTCT': 8759, 'TAATTG': 8753, 'TGCATA': 8741, 'GTAAAT': 8703, 'TAAATC': 8689, 'CGTAAA': 8654, 'ATGTTA': 8654, 'TAGCCA': 8645, 'GAGTTA': 8644, 'TTAGAT': 8638, 'TTGACT': 8635, 'AGTTAC': 8629, 'AGATAG': 8602, 'AGCCTA': 8602, 'CTCGTA': 8602, 'TGTAAA': 8594, 'CCGTAA': 8585, 'ATAGAT': 8573, 'ATTAGA': 8568, 'AGTCTC': 8565, 'GAATAA': 8562, 'ACTGAG': 8537, 'CTTATA': 8524, 'TATATA': 8518, 'AACTTA': 8515, 'AGTGAG': 8510, 'CTGAGA': 8495, 'TGTTAC': 8493, 'ATCTTA': 8490, 'TAGCGA': 8487, 'TTGTAC': 8487, 'TTACTA': 8487, 'ACACAC': 8465, 'AATAAC': 8462, 'ATCGTA': 8456, 'AGTGTC': 8446, 'CGATAG': 8440, 'TTTAGT': 8400, 'CTTTAG': 8358, 'TCGTTA': 8357, 'ACACTT': 8344, 'GTGTGT': 8333, 'AATGTA': 8311, 'ATAAGC': 8308, 'CTCACA': 8303, 'TTCATA': 8284, 'GTAACC': 8276, 'ATAGAG': 8274, 'TCTGTA': 8241, 'CACGTA': 8228, 'GTACTA': 8218, 'TAATTA': 8208, 'GTTATA': 8207, 'TTAAGC': 8195, 'TCCCTA': 8191, 'ACCTTA': 8189, 'AGTTAA': 8188, 'AATCTA': 8187, 'TTCGTA': 8165, 'ATCATA': 8160, 'TAACAA': 8155, 'GAGACT': 8136, 'GGTTAA': 8093, 'ATATAC': 8087, 'AGCATA': 8074, 'AGGTTA': 8040, 'ATGCTA': 7996, 'ATACGT': 7974, 'TAGAGA': 7949, 'TATTAG': 7948, 'TTTAGG': 7929, 'GTAATT': 7920, 'TAACGC': 7920, 'ATAACA': 7912, 'TATAAC': 7909, 'TAACCA': 7905, 'AGGGTA': 7883, 'ATTGTA': 7882, 'ATACAC': 7862, 'GTGCTA': 7847, 'TAGATA': 7846, 'TGACTA': 7831, 'ATACTC': 7810, 'ACCATA': 7810, 'GTAATC': 7807, 'TAGAGG': 7802, 'ATAAGA': 7800, 'TTGTGT': 7781, 'ATTAAG': 7780, 'TAGCGC': 7777, 'ACTTAC': 7767, 'ATTAGC': 7766, 'TAATCG': 7756, 'GTTGTA': 7756, 'CTTACT': 7751, 'TTAAGG': 7727, 'AACCTA': 7720, 'TAATCC': 7717, 'CTTAAT': 7716, 'AATAGC': 7707, 'TTAACA': 7670, 'TACAGT': 7660, 'ACTATA': 7659, 'CGTAGC': 7654, 'CACACA': 7648, 'GGTATA': 7638, 'TATACA': 7632, 'GAGTGT': 7631, 'TTATAG': 7609, 'TCCTTA': 7590, 'ATATAG': 7585, 'TCTTAA': 7583, 'ACATAT': 7567, 'TAAGCA': 7559, 'AGCTAA': 7531, 'ACAGTA': 7528, 'CGTAAC': 7526, 'CCCATA': 7524, 'ATTCTA': 7523, 'TATGTA': 7517, 'TAAACG': 7513, 'AATAAG': 7510, 'TCTTAC': 7485, 'TCCATA': 7469, 'TTAGCG': 7452, 'TAAACC': 7446, 'TGTCTA': 7419, 'TAATGT': 7393, 'AACGTA': 7387, 'AGAGTC': 7379, 'ACTTGT': 7356, 'TAGCAA': 7351, 'AACATA': 7330, 'AGTAAT': 7327, 'TTAGCC': 7323, 'TACCTA': 7307, 'TGTAGA': 7307, 'TGAGTA': 7303, 'GTAAAC': 7277, 'CGTTAA': 7247, 'ACACAT': 7220, 'TACTGT': 7211, 'TAGTGG': 7207, 'GGCTAA': 7187, 'GCTAAT': 7180, 'AGACTC': 7178, 'TAGTGA': 7165, 'TGAGAG': 7161, 'TTAGAG': 7160, 'GGTAGT': 7159, 'CCCTTA': 7131, 'GGATAG': 7122, 'TTAAGA': 7104, 'CAGTAA': 7104, 'GACTTA': 7086, 'CCTTAA': 7084, 'TAATCT': 7078, 'TAATTC': 7073, 'GTAGAG': 7072, 'GTAGTG': 7070, 'AGGTAG': 7068, 'TTAGCA': 7032, 'ATCCTA': 7028, 'GCATAA': 7017, 'GATAGT': 6997, 'AAGTAG': 6978, 'TAGCAG': 6969, 'TAAGCG': 6953, 'GTAAGC': 6939, 'CATAGA': 6930, 'TAAGGC': 6928, 'TAGAGC': 6924, 'TAACCT': 6919, 'TCTAAA': 6899, 'TAACCC': 6868, 'TAAGCT': 6867, 'ATGTAA': 6864, 'GTAGAT': 6852, 'TCATAC': 6845, 'CATACA': 6844, 'TAAACT': 6834, 'CCATAA': 6831, 'TCATAA': 6826, 'ATAGCA': 6824, 'TTGTAA': 6821, 'ACGTTA': 6819, 'TTAGCT': 6790, 'ACTTAA': 6768, 'TAACAG': 6760, 'TATAAG': 6750, 'TAGCTG': 6730, 'CATAGC': 6726, 'TTAGTG': 6688, 'TATAGC': 6687, 'CGTAAT': 6675, 'TGTATA': 6668, 'TAGGCG': 6632, 'TACACA': 6625, 'TCACTA': 6621, 'TACTTA': 6614, 'CGCTAA': 6608, 'TTAACT': 6600, 'GTGTTA': 6562, 'GTAAGG': 6540, 'ACTAAA': 6523, 'AGAGTA': 6519, 'TAAGCC': 6484, 'ATAGCT': 6478, 'TTGTAG': 6469, 'CACTTA': 6438, 'CCGTAG': 6417, 'TAGAAC': 6391, 'GTTAAC': 6389, 'GGGTAG': 6386, 'ATAGTG': 6385, 'ATTAGT': 6377, 'GACATA': 6371, 'TGTGAG': 6368, 'TTCTAA': 6357, 'TAGTTT': 6342, 'CATACT': 6337, 'ATAACT': 6327, 'CCATAG': 6325, 'TAACTT': 6316, 'GTTTAG': 6309, 'CCTAAA': 6300, 'TTAGTT': 6252, 'ACCCTA': 6230, 'AGTATA': 6227, 'GTCATA': 6223, 'AGTAGA': 6212, 'GTAATA': 6204, 'TAATAG': 6174, 'ACATAC': 6174, 'AATAGT': 6165, 'ATGTAG': 6150, 'TAGAAT': 6094, 'CATAAC': 6081, 'TAGATT': 6072, 'GTATAA': 6063, 'GTATAC': 6062, 'TAAGGG': 6040, 'GGTAGG': 6036, 'CAATAG': 6009, 'CTTAAC': 6003, 'TCGTAA': 6001, 'CATTAG': 5989, 'TTAAGT': 5978, 'CATAAT': 5965, 'CTAATG': 5930, 'CTTGTA': 5926, 'ATAGAC': 5920, 'CTCTTA': 5914, 'TAAGTT': 5911, 'TGTAAT': 5908, 'GTTAAG': 5899, 'TAGCCT': 5899, 'GTAGGC': 5854, 'GCGTAG': 5835, 'TATACT': 5796, 'ATAGGC': 5792, 'TAGCTT': 5781, 'ATAGTT': 5778, 'CTAAAT': 5761, 'TACTCT': 5756, 'ACTGTA': 5693, 'GCTAAC': 5663, 'TAGGCA': 5650, 'GATAGG': 5647, 'CTTAAG': 5646, 'TAGTGC': 5638, 'TAATAC': 5627, 'GCATAG': 5625, 'TAAGAT': 5624, 'GCTAAG': 5621, 'TGTTAG': 5611, 'AACTAA': 5605, 'TAGCCC': 5603, 'TAACTG': 5602, 'GTCTTA': 5559, 'GTTAGA': 5541, 'CATATA': 5527, 'GTAGCA': 5486, 'TAGGAA': 5457, 'TCATAG': 5449, 'CTCATA': 5447, 'GTAACA': 5441, 'CGTAAG': 5431, 'GTAAGA': 5430, 'TAGGTG': 5402, 'ACATAA': 5396, 'CTTAGA': 5387, 'TAAGGT': 5367, 'CTATAG': 5351, 'ACTCTA': 5340, 'TAGACG': 5336, 'TACATA': 5334, 'ATACTA': 5318, 'TAGTAA': 5315, 'GTTAGC': 5315, 'TTAGGA': 5286, 'TTAGGT': 5273, 'AGACTA': 5273, 'TACACT': 5249, 'CTGTAA': 5243, 'TGTAGC': 5230, 'TTAGTA': 5215, 'CGTAGA': 5199, 'TGCTAG': 5199, 'AGTAGC': 5194, 'TAGTTG': 5190, 'CTAATT': 5161, 'GTAGCT': 5146, 'CTAGAA': 5127, 'TGTGTA': 5094, 'ATAAGT': 5078, 'CTTAGC': 5070, 'TACGTA': 5052, 'CGTATA': 5034, 'CAGTAG': 5033, 'ATCTAA': 5031, 'AGTTAG': 5019, 'ATAGTA': 4998, 'TCCTAA': 4992, 'ATTAGG': 4975, 'GCTAGA': 4960, 'TAGCTA': 4935, 'TTAGGC': 4930, 'TCGTAG': 4926, 'TAGTCA': 4922, 'ACGTAA': 4891, 'CTAAAC': 4889, 'TAGACA': 4887, 'GCTAGC': 4874, 'TAACAT': 4865, 'TAAGAG': 4791, 'GTAGAC': 4780, 'GCTTAG': 4776, 'TAACTA': 4755, 'GTAACT': 4752, 'AGTAAC': 4749, 'CTAATA': 4730, 'TAACGT': 4699, 'CATAAG': 4694, 'TACTAA': 4680, 'CTGTAG': 4680, 'TATAGT': 4678, 'CTAATC': 4666, 'TAGTCG': 4624, 'ATAGGT': 4597, 'GTAGTT': 4594, 'ATAGGA': 4584, 'GAATAG': 4574, 'CTAACC': 4564, 'AATAGG': 4562, 'ACTAAT': 4561, 'TGTAAC': 4543, 'TCTAAT': 4533, 'TAAGTG': 4523, 'TAACAC': 4519, 'CCTTAG': 4510, 'GATTAG': 4500, 'TTAGGG': 4458, 'GAGTAA': 4446, 'GTAAGT': 4426, 'CTAAGC': 4421, 'CACATA': 4393, 'ATAGTC': 4358, 'CATAGG': 4355, 'TAGTAG': 4346, 'CTAGCC': 4344, 'TAGATC': 4343, 'TAAGAC': 4329, 'AGTAGT': 4327, 'TATAGG': 4308, 'ACATAG': 4291, 'TTAGTC': 4285, 'CTAACG': 4276, 'TAGGGG': 4273, 'AGTAAG': 4252, 'TAGGAG': 4252, 'CCTAAT': 4213, 'GTAGTC': 4190, 'CGTAGT': 4184, 'GCCTAA': 4178, 'TCTTAG': 4162, 'AGTCTA': 4158, 'ACACTA': 4148, 'TGTAGG': 4140, 'GTTAGT': 4086, 'AGCTAG': 4086, 'CTAAGG': 4077, 'TAGTTA': 4072, 'TAAGTA': 4068, 'GTAGGG': 4068, 'TGTAGT': 4061, 'ACTTAG': 4055, 'AGTGTA': 4045, 'GGTTAG': 4042, 'TAGACC': 4035, 'TAGCAC': 4030, 'TAGGGA': 4016, 'TAGGTA': 4003, 'CGTAGG': 3977, 'GTATAG': 3970, 'CGCTAG': 3959, 'TAGGCT': 3934, 'TTAGAC': 3931, 'CTAGAG': 3928, 'TAACTC': 3925, 'TAGCGT': 3922, 'CACTAA': 3915, 'TAGCAT': 3909, 'ATAGGG': 3899, 'CTAGCG': 3898, 'TAGGTT': 3896, 'TAGTAT': 3894, 'GTAGGT': 3880, 'TAGGGC': 3879, 'GTAGTA': 3877, 'ACCTAA': 3866, 'CCCTAA': 3862, 'CGTTAG': 3844, 'ACGTAG': 3836, 'TAGCTC': 3827, 'GTAGGA': 3798, 'TAGAGT': 3759, 'TAAGTC': 3711, 'CATAGT': 3679, 'TTCTAG': 3649, 'CTTAGT': 3577, 'GTTAGG': 3574, 'GTGTAA': 3532, 'TAGGCC': 3528, 'AACTAG': 3515, 'TCTAGA': 3490, 'GGCTAG': 3490, 'CTTAGG': 3469, 'GCTAGT': 3460, 'AGTAGG': 3447, 'GCTAGG': 3440, 'GTGTAG': 3439, 'TGTAAG': 3415, 'CTAAGA': 3403, 'CTCTAA': 3389, 'ATCTAG': 3348, 'TAGTTC': 3347, 'TAGTGT': 3310, 'CCTAGC': 3299, 'CTAACA': 3276, 'CTAGAT': 3275, 'CCTAAC': 3245, 'ACTAAC': 3197, 'TAGGAT': 3194, 'CCTAGA': 3193, 'CTAGTG': 3177, 'CTAGGC': 3168, 'CCTAAG': 3140, 'ACTAGA': 3108, 'TCTAGC': 3041, 'TCTAAG': 3032, 'TAGTCT': 2932, 'TCTAAC': 2930, 'GACTAA': 2920, 'CTAACT': 2919, 'CTAGCA': 2917, 'CTAGCT': 2859, 'ACTAAG': 2839, 'GAGTAG': 2823, 'ACTAGC': 2804, 'CCCTAG': 2795, 'TAGACT': 2794, 'CTAAGT': 2731, 'TAGGAC': 2688, 'GCCTAG': 2634, 'ACCTAG': 2628, 'GTCTAA': 2617, 'TAGTAC': 2587, 'TAGTCC': 2582, 'TCCTAG': 2563, 'TAGGTC': 2474, 'TAGGGT': 2462, 'CTAGAC': 2421, 'TACTAG': 2394, 'TCTAGT': 2332, 'CTAGTT': 2327, 'CTCTAG': 2242, 'CTAGTC': 2227, 'CACTAG': 2216, 'CTAGGG': 2205, 'CCTAGT': 2159, 'TCTAGG': 2107, 'CTAGTA': 2052, 'CTAGGT': 2025, 'CCTAGG': 1969, 'CTAGGA': 1923, 'ACTAGT': 1832, 'ACTAGG': 1805, 'GTCTAG': 1731, 'GACTAG': 1424, 'TCATGN': 2, 'CATGNN': 2, 'CAAAAR': 1, 'AAAARC': 1, 'AAARCT': 1, 'AARCTT': 1, 'ARCTTT': 1, 'RCTTTT': 1, 'CTGACM': 1, 'TGACMA': 1, 'GACMAT': 1, 'ACMATC': 1, 'CMATCT': 1, 'MATCTG': 1, 'GATGCR': 1, 'ATGCRG': 1, 'TGCRGA': 1, 'GCRGAA': 1, 'CRGAAG': 1, 'RGAAGG': 1, 'ACCTCM': 1, 'CCTCMT': 1, 'CTCMTT': 1, 'TCMTTC': 1, 'CMTTCA': 1, 'MTTCAT': 1, 'ATTTTM': 1, 'TTTTMG': 1, 'TTTMGC': 1, 'TTMGCA': 1, 'TMGCAA': 1, 'MGCAAT': 1, 'GTTTTK': 1, 'TTTTKG': 1, 'TTTKGC': 1, 'TTKGCG': 1, 'TKGCGC': 1, 'KGCGCG': 1, 'GACAGY': 1, 'ACAGYG': 1, 'CAGYGA': 1, 'AGYGAC': 1, 'GYGACG': 1, 'YGACGG': 1, 'CGGCCM': 1, 'GGCCMC': 1, 'GCCMCA': 1, 'CCMCAC': 1, 'CMCACC': 1, 'MCACCC': 1, 'ACCAGS': 1, 'CCAGSC': 1, 'CAGSCC': 1, 'AGSCCC': 1, 'GSCCCC': 1, 'SCCCCC': 1, 'TCCCGS': 1, 'CCCGST': 1, 'CCGSTA': 1, 'CGSTAT': 1, 'GSTATT': 1, 'STATTT': 1, 'TGCTTW': 1, 'GCTTWR': 1, 'CTTWRA': 1, 'TTWRAG': 1, 'TWRAGA': 1, 'WRAGAA': 1, 'RAGAAA': 1, 'AGCGGS': 1, 'GCGGSR': 1, 'CGGSRC': 1, 'GGSRCC': 1, 'GSRCCA': 1, 'SRCCAC': 1, 'RCCACC': 1, 'CGCCGM': 1, 'GCCGMC': 1, 'CCGMCC': 1, 'CGMCCG': 1, 'GMCCGT': 1, 'MCCGTC': 1, 'TCACTS': 1, 'CACTSC': 1, 'ACTSCG': 1, 'CTSCGC': 1, 'TSCGCC': 1, 'SCGCCC': 1, 'CTGCCM': 1, 'TGCCMG': 1, 'GCCMGG': 1, 'CCMGGG': 1, 'CMGGGC': 1, 'MGGGCC': 1, 'CTTGAR': 1, 'TTGARA': 1, 'TGARAA': 1, 'GARAAA': 1, 'ARAAAA': 1, 'RAAAAG': 1, 'CAAGTR': 1, 'AAGTRC': 1, 'AGTRCC': 1, 'GTRCCT': 1, 'TRCCTG': 1, 'RCCTGC': 1, 'ATTCAR': 1, 'TTCARG': 1, 'TCARGG': 1, 'CARGGC': 1, 'ARGGCG': 1, 'RGGCGG': 1, 'ACGGTW': 1, 'CGGTWA': 1, 'GGTWAG': 1, 'GTWAGG': 1, 'TWAGGA': 1, 'WAGGAT': 1, 'ATTGCM': 1, 'TTGCMA': 1, 'TGCMAT': 1, 'GCMATT': 1, 'CMATTT': 1, 'MATTTA': 1, 'ACGCCS': 1, 'CGCCSC': 1, 'GCCSCC': 1, 'CCSCCC': 1, 'CSCCCG': 1, 'SCCCGC': 1, 'ATGNNA': 1, 'TGNNAC': 1, 'GNNACT': 1, 'NNACTG': 1, 'NACTGC': 1, 'ATGNNG': 1, 'TGNNGT': 1, 'GNNGTG': 1, 'NNGTGG': 1, 'NGTGGA': 1, 'GTCTCS': 1, 'TCTCSG': 1, 'CTCSGC': 1, 'TCSGCC': 1, 'CSGCCC': 1, 'SGCCCC': 1, 'GGCCTN': 1, 'GCCTNC': 1, 'CCTNCT': 1, 'CTNCTC': 1, 'TNCTCG': 1, 'NCTCGG': 1})\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Frequency c-terminal hexamers:\n",
      " Counter({'GCCGCC': 1316, 'GCGGCC': 1115, 'GCCGGC': 817, 'GGCGGC': 743, 'CCGGCC': 686, 'CGCGGC': 662, 'GGCGCC': 643, 'CTGGCC': 609, 'CGCGCC': 604, 'CTGCGC': 599, 'GCGGGC': 578, 'GACGCC': 571, 'GCGCGC': 565, 'CGCCGC': 559, 'AAGGGC': 527, 'GAGGGC': 515, 'ACCGCC': 509, 'GGCCGC': 498, 'CCGCGC': 493, 'CTGGGC': 486, 'GCCCGC': 480, 'AAGGCC': 464, 'CCCGCC': 455, 'GAGGCC': 450, 'CGGCGC': 448, 'CTCGGC': 436, 'CAGGGC': 433, 'GAGGAC': 429, 'GCCGAC': 416, 'CAGGCC': 414, 'GTCGCC': 413, 'GTCGGC': 404, 'GACGAC': 401, 'AGCGCC': 389, 'AAGCGC': 384, 'CTCGCC': 362, 'GTGCGC': 351, 'GAGCGC': 350, 'ACCGGC': 344, 'ACGGCC': 341, 'CGCGAC': 341, 'GCGGCG': 337, 'ATCGGC': 336, 'ATCGCC': 327, 'CAGCGC': 326, 'GGGCGC': 321, 'CGGGCC': 320, 'GGCGAC': 318, 'GCTGCC': 316, 'CTGGAC': 310, 'TCGGCC': 309, 'CTGCCC': 304, 'GAAGCC': 303, 'GCGGAC': 301, 'GTGGCC': 298, 'AACGCC': 298, 'CGGGGC': 291, 'GCAGCC': 286, 'GCGCAC': 280, 'AAAAAA': 279, 'GAGGAG': 278, 'GCGCCC': 276, 'GAAGGC': 270, 'GACGGC': 270, 'ACCCGC': 269, 'GACCGC': 269, 'TCCGCC': 268, 'CTGTTC': 267, 'AAGGAC': 265, 'CAGGAC': 265, 'AAAGCC': 263, 'GCGAGC': 263, 'CCGGGC': 261, 'GCCGTC': 261, 'GAGGCG': 259, 'CGCCCC': 256, 'GGGGCC': 254, 'AACGGC': 253, 'GCGGTC': 250, 'GGCAAC': 249, 'GCCGCG': 249, 'TTCGGC': 245, 'CATGGC': 244, 'ACGCGC': 242, 'CTCGAC': 239, 'CCCGGC': 239, 'CACGCC': 235, 'CGGCCC': 235, 'TTCGCC': 235, 'GAAGAC': 233, 'GGAGCC': 232, 'GCCTCC': 229, 'GTCCGC': 225, 'GCGTCC': 221, 'AAGCCC': 219, 'GCGGAG': 215, 'GCGCTC': 214, 'CAGCCC': 213, 'AGCCGC': 213, 'GATGCC': 213, 'GAGGTC': 211, 'AAAGGC': 211, 'CGCTTC': 208, 'GCCGAG': 208, 'CTGCTC': 205, 'GGACGC': 205, 'GAGAGC': 205, 'CCGTTC': 203, 'AACCGC': 202, 'CTGAAC': 200, 'CGTGGC': 199, 'TTCGAC': 198, 'TCGCGC': 198, 'CTGAGC': 198, 'AAGAAC': 197, 'CCGGAC': 197, 'CGGGAC': 196, 'GGCAGC': 196, 'AAGAAG': 195, 'GCAGGC': 195, 'AAGGAG': 194, 'CCGGCG': 194, 'ACGCCC': 194, 'CACGGC': 193, 'ATCGAC': 193, 'CCGGTC': 192, 'CTGATC': 192, 'GCCCAC': 191, 'GATGAC': 191, 'CCCCGC': 190, 'CTGGTC': 190, 'GTCGAC': 189, 'ATCCGC': 185, 'GTGGGC': 183, 'GCCCCC': 183, 'CGCGTC': 183, 'ACCGAC': 183, 'GAGGAT': 181, 'GGAGGC': 181, 'GAGTCC': 180, 'GACTTC': 180, 'GCGAAC': 179, 'GCGGCT': 179, 'CGCAGC': 179, 'GCCAGC': 178, 'CTGCAC': 177, 'AAGAGC': 176, 'GATGAG': 176, 'AAGTTC': 175, 'GGGGAG': 174, 'GGGGGC': 173, 'GGCGTC': 173, 'CTGTCC': 172, 'CGCTCC': 171, 'AATGGC': 171, 'GGCGAG': 170, 'CACCGC': 167, 'GCGCGG': 167, 'GCGATC': 167, 'GCGGCA': 166, 'AAGGTC': 166, 'GAGGAA': 166, 'GGCCAC': 165, 'AAAGAC': 164, 'GCCGGG': 163, 'AACGAC': 163, 'GCGTTC': 161, 'GCACGC': 161, 'AAGCTC': 160, 'AAACGC': 160, 'GGCTCC': 159, 'GAAAGC': 159, 'GCCAAC': 158, 'CATGCC': 158, 'GGCGCG': 157, 'GAGAAC': 157, 'GTGATC': 157, 'GAGGGG': 156, 'GCGGGG': 156, 'AAATCC': 155, 'GTGAAG': 154, 'TTCCGC': 154, 'CAGAGC': 153, 'GAACGC': 153, 'GATCGC': 152, 'CGGGTC': 152, 'GCCACC': 152, 'GAGCAC': 151, 'TCCGGC': 151, 'AAGAAA': 151, 'GCTTCC': 151, 'CGCACC': 151, 'GCCGGT': 151, 'GGTCGC': 150, 'CTGACC': 149, 'GCGCCG': 149, 'GGCACC': 148, 'GGGAGC': 147, 'CCGTCC': 146, 'GAGTTC': 144, 'CGGAGC': 143, 'CAGTTC': 143, 'GCCGCT': 142, 'TCCCGC': 142, 'ACCGTC': 142, 'GGCTTC': 141, 'AAGCAC': 141, 'CAGCAC': 141, 'CAGGTC': 141, 'AAGTCC': 140, 'GTGGAG': 139, 'AAGGCG': 139, 'GGCCGG': 139, 'GGTGGC': 138, 'ACGGGC': 138, 'GCGCAG': 138, 'GTGCTC': 138, 'AAAAAC': 138, 'AAGGGG': 138, 'AAAAGC': 137, 'TACCGC': 137, 'GCGCTG': 137, 'GACGTC': 137, 'TGGCGC': 137, 'ACGCTC': 136, 'CGGCTC': 136, 'GACGAG': 136, 'GATGGC': 136, 'AATGCC': 135, 'ACCCCC': 135, 'GCGGTG': 134, 'GTGGAC': 134, 'GGCATC': 134, 'GACAGC': 133, 'AGCGGC': 133, 'GAGATC': 132, 'CCGGAG': 132, 'AAGGCT': 131, 'GGGGCG': 131, 'GACGAT': 130, 'GGGCAC': 130, 'GTGGCG': 130, 'CAGTCC': 130, 'GGCAAG': 129, 'CGCTAC': 129, 'CGGCCG': 129, 'GCCGCA': 129, 'CCGCTC': 129, 'GAAGAG': 129, 'GAGCCC': 128, 'GATGAA': 126, 'GAAGAT': 126, 'GATTTC': 126, 'GACATC': 126, 'CGGATC': 125, 'CAAGCC': 125, 'GGGCGG': 125, 'GGCTAC': 125, 'CCTGCC': 125, 'GTCAGC': 124, 'GAATCC': 124, 'ATGCGC': 124, 'CATGAC': 124, 'CGCCAC': 124, 'AAAAAG': 123, 'AATGAC': 123, 'CGGCAC': 123, 'CGGCGG': 123, 'GTCGTC': 122, 'AAGGAT': 122, 'CATGAG': 122, 'ACCTTC': 122, 'AAACCC': 121, 'CTGGAT': 121, 'CAGGAG': 121, 'GGCGGG': 120, 'GAAGAA': 120, 'GCCGAA': 120, 'GACGCG': 120, 'CAGCTC': 119, 'GCGAAG': 119, 'GTGCTG': 119, 'CTCACC': 119, 'CAGAAC': 119, 'GATGAT': 119, 'GACAAC': 119, 'TACGCC': 118, 'GGCCTC': 118, 'GTGAGC': 117, 'GGCGAT': 117, 'GCCTTC': 117, 'GAGAAG': 117, 'GCGACC': 117, 'CTGGCG': 116, 'ACGGAC': 116, 'ACGTTC': 116, 'ACCACC': 115, 'GGTGCC': 114, 'GGGGTC': 114, 'TTGCGC': 114, 'CGCCGG': 114, 'GGGCTC': 113, 'GTGCGG': 113, 'TTCTTC': 113, 'CACGAC': 113, 'GAGCGG': 113, 'AACAAC': 113, 'AACGTC': 113, 'ACGGTC': 112, 'GAAAAA': 112, 'CGACGC': 112, 'ATGGCC': 112, 'GCCAAG': 111, 'GTCTCC': 111, 'CTCGTC': 111, 'CGCAAC': 111, 'AAAAAT': 111, 'CCGAGC': 111, 'AGGCGC': 111, 'GCATCC': 111, 'CATGTC': 110, 'AAGATC': 110, 'GAAAAC': 110, 'GGCTGC': 110, 'GCTCGC': 110, 'CAAGGC': 110, 'CTGCCG': 110, 'CCGCAC': 109, 'CTGGAG': 109, 'CAGGCG': 109, 'CGTCGC': 109, 'AGCTTC': 109, 'GAGCAG': 109, 'CGGTCC': 109, 'GGGCTG': 109, 'CCGAAC': 109, 'CGGTTC': 108, 'CGCTGG': 108, 'ATGGGC': 107, 'CCCGAC': 107, 'GATGTC': 107, 'GCGGAT': 107, 'CTGAAG': 107, 'GTGCCC': 106, 'AAAGAG': 106, 'TTTGCC': 106, 'CTGCGG': 106, 'CGTCCC': 106, 'GAGGTG': 106, 'CGCGCG': 106, 'GCCCGG': 105, 'CGCCTC': 105, 'GCGCAT': 105, 'GTGAAC': 105, 'GGGAAC': 105, 'GAGACC': 104, 'ACGCAC': 104, 'AGCAAC': 104, 'GCCCTC': 104, 'AAAGAT': 104, 'GTGTCC': 104, 'GGGAAG': 104, 'CGGGCG': 103, 'CAGGCT': 103, 'GAGCTG': 103, 'AACAGC': 103, 'GCCGGA': 103, 'CTGGGG': 102, 'GCGGAA': 102, 'GGCGGT': 102, 'GGTTCC': 101, 'ACCTCC': 101, 'CCAGCC': 100, 'AGCGAC': 100, 'CTCCGC': 100, 'AAGGGT': 100, 'CGCATC': 100, 'GTGCAC': 100, 'CGTGCC': 100, 'CTGCTG': 100, 'CAGGGG': 99, 'GCAAGC': 99, 'AATGAG': 99, 'CCGACC': 99, 'GTGTTC': 99, 'CACCAC': 99, 'GCTGGC': 99, 'GCTGAC': 98, 'GGTGAG': 98, 'CGGGTG': 98, 'GAGCTC': 97, 'CGGGGG': 96, 'GGGTTC': 96, 'GTGGTC': 96, 'CTTGCC': 96, 'AAAGAA': 95, 'CGTTCC': 95, 'AAATTC': 95, 'TCGGCG': 95, 'GGCGCA': 95, 'GAGGCA': 95, 'CTCAAC': 95, 'ATCAGC': 94, 'GACCGG': 94, 'GCCGAT': 94, 'GGAAGC': 94, 'GTCAAC': 94, 'AAGGAA': 94, 'GGAGAG': 94, 'CGCGGG': 94, 'GCAGAC': 93, 'TTCAGC': 93, 'GTGGGG': 93, 'GCCTAC': 93, 'TCGCAC': 93, 'CTCTCC': 93, 'CAGCCG': 93, 'AATGAA': 93, 'GAAATC': 93, 'CAACGC': 93, 'ACGCCG': 92, 'CTCAGC': 92, 'GGTCAC': 92, 'GAAACC': 92, 'AACTTC': 92, 'GGGGAC': 92, 'CCCGCG': 92, 'CGGACC': 92, 'CTGGTG': 92, 'GTTCGC': 92, 'GACCAC': 92, 'GACCTC': 92, 'GACTCC': 92, 'TCGCCC': 91, 'TCGGGC': 91, 'TGGGCC': 91, 'ACCAAC': 91, 'GGGGTG': 91, 'CCGGCA': 91, 'CGTGAC': 90, 'ACCCAC': 90, 'GCGCGT': 90, 'CCGATC': 90, 'ACAGCC': 90, 'CCGCTG': 89, 'TACGGC': 89, 'GTCATC': 89, 'ATCAAC': 89, 'GTTGCC': 89, 'CCCGTC': 89, 'GGCGAA': 89, 'GCGCGA': 89, 'TCCTCC': 89, 'GATTCC': 89, 'GGCCTG': 88, 'GCCATC': 88, 'ACGGCG': 88, 'CACCCC': 88, 'CGCCCG': 88, 'TGCCGC': 88, 'ACCAGC': 87, 'CGCGGT': 87, 'GGCGCT': 87, 'ATCATC': 87, 'GAAAAT': 87, 'GACGAA': 87, 'CCGGCT': 86, 'CCGGTG': 86, 'ACTGCC': 86, 'ACCGCG': 86, 'GAGCCG': 86, 'AAAGTC': 86, 'CTCCCC': 86, 'GATCTC': 86, 'TCGTCC': 85, 'AAGGCA': 85, 'GAGGCT': 85, 'GAACTC': 85, 'AATGAT': 85, 'GATGCT': 85, 'AAAGGA': 85, 'GACACC': 84, 'AGCCAC': 84, 'GGGATC': 84, 'CCGCGG': 83, 'TGGGGC': 83, 'CTCTTC': 83, 'GCCTGC': 83, 'GAAAAG': 83, 'GGAAAC': 83, 'TGCGGC': 83, 'GAGGGT': 83, 'GGGCAG': 82, 'GAGGGA': 82, 'GCGGGT': 82, 'AATCGC': 82, 'TCGTTC': 82, 'GGCGGA': 82, 'GAAGCG': 82, 'AACCCC': 82, 'AAGTAC': 82, 'ACCCTC': 82, 'AAGGTG': 82, 'CCGCAG': 81, 'ATTGCC': 81, 'TTGGCC': 81, 'AGTGCC': 81, 'GCCCTG': 81, 'AAAACC': 81, 'CCCTCC': 81, 'AAGCCT': 81, 'TCGGTC': 80, 'TCTGCC': 80, 'GCACCC': 80, 'CAGGAT': 80, 'AGCGAG': 80, 'AAGACC': 80, 'GCGCCT': 80, 'TCGGAC': 80, 'GTGGAT': 79, 'CTTGGC': 79, 'AATAAA': 79, 'AGGGGC': 79, 'TTCGTC': 79, 'AAGAAT': 79, 'GCTCCC': 79, 'TTCGAG': 78, 'CCCTTC': 78, 'CATGCT': 78, 'ACGAAC': 78, 'GATGCG': 78, 'GCCCGA': 78, 'GGGGAA': 78, 'CAGGCA': 78, 'GAAGGG': 78, 'TGCGCC': 78, 'TGGGAG': 78, 'CCACGC': 78, 'AAGCCG': 78, 'TCGCTC': 77, 'TCGAAC': 77, 'GGCGTG': 77, 'ATCTTC': 77, 'GCAGCG': 77, 'ATCGAG': 77, 'TGGGAC': 77, 'GAAGTC': 77, 'GGTTTC': 77, 'ATTCGC': 77, 'CAGCAG': 77, 'GAATTC': 77, 'AAAATC': 76, 'AAGCGG': 76, 'TCGATC': 76, 'ACGACC': 76, 'CCGGGG': 76, 'AGGAGC': 76, 'TCAGCC': 75, 'ATCTCC': 75, 'ACGATC': 75, 'GATATC': 75, 'TGGAAC': 75, 'GCGTCG': 75, 'GCGGGA': 75, 'ATCGTC': 75, 'CGCAAG': 75, 'CAGCGG': 75, 'CGCCGT': 75, 'CCGCGT': 75, 'GGCTGG': 75, 'TATCGC': 75, 'AGCAGC': 74, 'CATCAC': 74, 'AAAATT': 74, 'ATTTCC': 74, 'TTTGGC': 74, 'GGGTCC': 74, 'ACGTCC': 74, 'CTGGGT': 74, 'ATGGAC': 74, 'ATGCTC': 74, 'AATAAT': 74, 'GGCCAG': 73, 'AATTTC': 73, 'GCGCAA': 73, 'AACTCC': 73, 'GCGTAC': 73, 'ACCCGG': 73, 'CGGCGT': 73, 'ATCACC': 73, 'AACACC': 73, 'TTGGGC': 73, 'GGAGAC': 73, 'AACCTC': 73, 'CAACCC': 73, 'ACGAGC': 73, 'GGTGGT': 72, 'GGCCGA': 72, 'GGATTC': 72, 'AGGGCC': 72, 'AATGTC': 72, 'GTGACC': 72, 'GGGACC': 72, 'GAACCC': 72, 'TTCAAC': 72, 'GACTAC': 72, 'TGGAGC': 72, 'GGAGAA': 72, 'CTTTCC': 72, 'CACTCC': 72, 'ACCTAC': 71, 'CAGATC': 71, 'GCTGCT': 71, 'ATGTTC': 71, 'AAGCTG': 71, 'GCAGAG': 71, 'AGCGTC': 71, 'GGAGCG': 71, 'AAGCAG': 71, 'CGCGAT': 71, 'CGAGGC': 70, 'GAGCGT': 70, 'CATGAA': 70, 'GGCAAT': 70, 'GTGCAG': 70, 'TCCGTC': 70, 'GGGCAT': 70, 'CGTCAC': 70, 'AAAGCT': 70, 'GCGCCA': 70, 'GAGCAT': 70, 'CGGCTG': 70, 'CATCGC': 70, 'GGTGAC': 69, 'GAGGTT': 69, 'GCAAAC': 69, 'CGGGAG': 69, 'GTGGCT': 69, 'GGGCGT': 69, 'TCCGAC': 69, 'GAACAC': 69, 'CTGAAA': 69, 'CAGGAA': 69, 'ATCCTC': 69, 'ATGAGC': 69, 'CCTTCC': 69, 'GTGCCG': 69, 'CCGGAA': 69, 'AACCAC': 69, 'GGAAAA': 68, 'GAGAAA': 68, 'AACGAT': 68, 'AGGGAC': 68, 'ACCTGC': 68, 'GACCTG': 68, 'CTCGAT': 68, 'GACGGG': 68, 'ACCGGG': 68, 'GTGAAA': 68, 'TCGGAG': 68, 'GCTGAG': 67, 'CTGCGT': 67, 'AGCTGG': 67, 'CTGCAT': 67, 'CCCGAG': 67, 'GGGGAT': 67, 'TATGAG': 67, 'CCCACC': 67, 'ACGGCA': 67, 'ACCTGG': 67, 'TGGTTC': 67, 'AGCGCG': 67, 'ATGAAG': 67, 'CCGGAT': 66, 'TTGAAG': 66, 'AAAATA': 66, 'AAACTC': 66, 'TTCCAC': 66, 'TACTTC': 66, 'CGGGCA': 66, 'GCCCCG': 66, 'TTGTTC': 66, 'CAGAAG': 66, 'ATCCCC': 66, 'ACCCCG': 66, 'GGTGCG': 66, 'AAGCAT': 66, 'GCAGCA': 65, 'AGCTAC': 65, 'AAGAGG': 65, 'TTTTCC': 65, 'AAGGGA': 65, 'TGGCTC': 65, 'GATAAA': 65, 'CAGGGT': 65, 'GGCAAA': 65, 'AAAGGT': 65, 'TTGAAC': 65, 'GGGCCC': 65, 'GATTTT': 65, 'GGCCCC': 65, 'TATGAC': 65, 'GCCCGT': 65, 'CCCCGG': 65, 'GCTCAC': 64, 'GACGCT': 64, 'ACCGAG': 64, 'CGCGCA': 64, 'GTGGTG': 64, 'AAAGGG': 64, 'ATGAAC': 64, 'AGCCTC': 64, 'GCACAC': 64, 'TTTTTC': 64, 'ATCCAC': 64, 'GCGCTT': 64, 'TTCCTC': 63, 'GAGCAA': 63, 'CGGATG': 63, 'ATTGAT': 63, 'CGGCAG': 63, 'GCCTCG': 63, 'ATGGAT': 63, 'CCAGGC': 63, 'AGCCGT': 63, 'CCTCGC': 63, 'CAGGTG': 63, 'AAGGTT': 63, 'CGGGAT': 63, 'CGAGCC': 63, 'ATGCCC': 63, 'GAGAGG': 63, 'GACAAG': 63, 'CGGAAC': 63, 'CACGTC': 63, 'GGAGGG': 63, 'TTGAGC': 62, 'TCCAAC': 62, 'ATAAAA': 62, 'GACTGG': 62, 'CAAGAC': 62, 'GGTAAC': 62, 'TGGGAT': 62, 'GTCACC': 62, 'ACCGAT': 62, 'TTTGAT': 62, 'TCGACC': 62, 'AAACAC': 62, 'CTCGCG': 62, 'CATGCG': 62, 'CTGTAC': 62, 'AAAAGT': 62, 'ATTCCC': 62, 'CAGTAC': 62, 'ACGAAG': 61, 'CAAAAA': 61, 'CGGGGT': 61, 'GTCCCC': 61, 'AATGCT': 61, 'TATGCC': 61, 'CGCCTG': 61, 'GGGAAA': 61, 'GAGAAT': 61, 'CAGACC': 61, 'TGGATC': 61, 'AGGAAG': 61, 'ATCGAT': 61, 'CCCGCA': 61, 'GCTGAT': 61, 'CGATCC': 61, 'GTCCAC': 60, 'ATGATC': 60, 'GAAATT': 60, 'GTCGAT': 60, 'GGAGGT': 60, 'TATTTC': 60, 'CTCATC': 60, 'GGATCC': 60, 'CGCGGA': 60, 'CGCCGA': 60, 'GGGATG': 60, 'TTAAAA': 60, 'AATCCC': 60, 'TCCAGC': 60, 'GGAAAG': 60, 'ACGGTG': 59, 'CGGAAG': 59, 'GTCGCG': 59, 'ATCGAA': 59, 'CAGGGA': 59, 'CAGCCA': 59, 'TTTCGC': 59, 'GCCCAG': 59, 'TTCTCC': 59, 'CTGCAG': 59, 'ATTGGC': 59, 'AGGCCC': 59, 'ATGGCG': 59, 'ACACGC': 59, 'CTCAAG': 59, 'GGACAC': 59, 'GACTGC': 58, 'GGCGTT': 58, 'CAATCC': 58, 'CGGCGA': 58, 'TACGAC': 58, 'CGCGCT': 58, 'GAAGGT': 58, 'GGAATC': 58, 'GACGCA': 58, 'CACTAC': 58, 'GCAGCT': 58, 'GTCTTC': 58, 'GTCTAC': 58, 'GCTGTC': 58, 'GATCAC': 57, 'AGCATC': 57, 'ACGCCA': 57, 'ATCAAG': 57, 'CTTCCC': 57, 'CGGCCA': 57, 'CAAAAC': 57, 'GGCAGG': 57, 'GTTCCC': 57, 'TGGCTG': 57, 'GCACTC': 57, 'GAAGGA': 57, 'CGTGTC': 57, 'GGGTCG': 57, 'CAGCCT': 57, 'GTCGAG': 57, 'GTTTCC': 57, 'GGAACC': 57, 'TCTTCC': 57, 'GGGCGA': 56, 'GTGGAA': 56, 'GGAGTC': 56, 'GATCTG': 56, 'GCGTCA': 56, 'AAGCTT': 56, 'CGGCAT': 56, 'CATCCC': 56, 'GGACGG': 56, 'GGGGCT': 56, 'CAAAGC': 56, 'GCCTGG': 56, 'ATGCAC': 56, 'TGCTGC': 56, 'CATGAT': 56, 'AAGCGT': 56, 'ATCGCG': 56, 'GGCCGT': 56, 'TTGGAT': 56, 'GGGAGG': 55, 'ACCGCT': 55, 'CTCGGG': 55, 'TCACGC': 55, 'GAGCTT': 55, 'GCTGCG': 55, 'TTCATC': 55, 'ATTTTC': 55, 'AATTCC': 55, 'GACCAG': 55, 'GCAGAT': 55, 'GAGTCG': 55, 'TTGAAA': 55, 'ACGCTG': 55, 'GCGAAA': 55, 'ACCGCA': 54, 'CACTTC': 54, 'ACGGCT': 54, 'GCCAGT': 54, 'AGCCTG': 54, 'AACGAG': 54, 'AGCCCC': 54, 'CTTCGC': 54, 'ATGGTC': 54, 'GGTGAA': 54, 'CTGCCT': 54, 'CGTCCG': 54, 'GCACAG': 54, 'CGCGAG': 54, 'AACGGT': 54, 'AGTGAG': 54, 'GAGATT': 53, 'GGGGGA': 53, 'GAGTGC': 53, 'GTCAAG': 53, 'CTGGCT': 53, 'ACCCGA': 53, 'CTGCCA': 53, 'TTCGAT': 53, 'GAGTTT': 53, 'AAGCAA': 53, 'GGTAGC': 53, 'GCGATG': 53, 'GTGATG': 53, 'GCAAAA': 53, 'AGGAAA': 53, 'CGACCC': 53, 'GATCAG': 52, 'CTGCTT': 52, 'CCGAAG': 52, 'AGAAAG': 52, 'CGGGGA': 52, 'CCGCGA': 52, 'ATCGGG': 52, 'GCCGTG': 52, 'CGGTAC': 52, 'CCGTCG': 52, 'AGAAAA': 52, 'ATCGGT': 52, 'GCATTC': 52, 'GTGGTT': 52, 'GAAGTG': 52, 'TGGTGG': 52, 'TCCACC': 52, 'GAATTT': 52, 'GTCCTC': 52, 'AGCGCA': 52, 'ACGGAG': 52, 'GGTTAC': 52, 'CTCGGT': 52, 'TTTGAC': 52, 'ATTAAA': 52, 'GCTGAA': 52, 'TGGAAG': 52, 'TCCGCG': 52, 'CATTCC': 52, 'AATTAT': 52, 'ACTCAC': 51, 'GGGAAT': 51, 'CTGGAA': 51, 'ACGCGG': 51, 'AACGCA': 51, 'GAGTGG': 51, 'TGCGAC': 51, 'ACGGGG': 51, 'ATGGAG': 51, 'AAACAA': 51, 'AACAAG': 51, 'TCGCGG': 51, 'CGTTTC': 51, 'GTCGGG': 51, 'AAATAT': 51, 'GACCGT': 51, 'GCGAGG': 51, 'GAATAT': 51, 'GAATAC': 51, 'GACAAA': 51, 'ATGGGG': 51, 'CTTGAT': 51, 'CCGGGT': 51, 'TATGAA': 51, 'CGGCCT': 51, 'AGGAAC': 50, 'ACCATC': 50, 'GAGCGA': 50, 'AAACAG': 50, 'GTCCGG': 50, 'TTGGAC': 50, 'AACAAT': 50, 'AAATAC': 50, 'ACCAAG': 50, 'CCCTGC': 50, 'CCCGAT': 50, 'GGAGGA': 50, 'TTGCCC': 50, 'CCCGCT': 50, 'GGAAAT': 50, 'GCACCG': 50, 'CCGGGA': 50, 'GTCGGT': 50, 'CTGGCA': 50, 'TTCACC': 50, 'CCCAAC': 50, 'GGAAGG': 50, 'AACATC': 50, 'GTGCGT': 49, 'GACGGT': 49, 'AAAGCA': 49, 'CTTTTC': 49, 'CCCAGC': 49, 'GGGGTT': 49, 'GCAGTC': 49, 'GATAAG': 49, 'GAAGCT': 49, 'GGAGAT': 49, 'CTGTCG': 49, 'AGGGAG': 49, 'TGGGAA': 49, 'GAAAGG': 49, 'CAGTCG': 49, 'GCGACG': 49, 'GAGCCT': 49, 'GATATT': 49, 'AACCGG': 49, 'GCAATC': 49, 'AAAGCG': 49, 'GAAGTT': 49, 'GCATCG': 48, 'GTTGAA': 48, 'AGCAAG': 48, 'GATGTT': 48, 'GGCCAT': 48, 'TCGGCA': 48, 'CCCTGG': 48, 'GACCGA': 48, 'ATGAAA': 48, 'ATTTTT': 48, 'CGTTAC': 48, 'GGGAGT': 48, 'AGCCGG': 48, 'AATATT': 48, 'TTGCTC': 48, 'GTTGGC': 48, 'CTCTAC': 48, 'GGGGGG': 48, 'AGTCGC': 48, 'AACGCG': 48, 'TCGGCT': 47, 'GCGGTT': 47, 'TCGAGC': 47, 'GCGAGT': 47, 'AATCAA': 47, 'GGGCAA': 47, 'GGATCG': 47, 'AAGCCA': 47, 'AATAAG': 47, 'GAGTAC': 47, 'GCTGCA': 47, 'GTCGCT': 47, 'CGGTCG': 47, 'TTTAAT': 47, 'AGAGCC': 47, 'CAGAAA': 47, 'AATGGA': 47, 'TACGTC': 47, 'AGCACC': 46, 'GAAAGT': 46, 'CGAAGC': 46, 'CGGGCT': 46, 'AATGTT': 46, 'ATCCGG': 46, 'GCCAAA': 46, 'TTTGAG': 46, 'ATCTAC': 46, 'AGCCAG': 46, 'GAATGG': 46, 'CACCTC': 46, 'CCGCAA': 46, 'CGAACC': 46, 'ATCAAT': 46, 'CCTGGC': 46, 'GCCAAT': 46, 'AAATCA': 46, 'TATGTC': 46, 'CCGGTT': 46, 'CGGACG': 46, 'CAGGTT': 46, 'TATGGC': 45, 'AATGGT': 45, 'CAGCAT': 45, 'GCATCA': 45, 'GCTCTC': 45, 'GCCCAT': 45, 'ACTCGC': 45, 'CTGCAA': 45, 'CTGGTT': 45, 'GGGCCG': 45, 'TGGACC': 45, 'AATCAC': 45, 'TTGATC': 45, 'AAACTT': 45, 'AAATCT': 45, 'GGTATC': 45, 'TCGTCG': 45, 'TGGTAC': 45, 'GGTCGG': 45, 'AAAATG': 45, 'CCGCTT': 45, 'AAACGG': 45, 'CCCATC': 45, 'CCCGGG': 45, 'ATGAAT': 45, 'TTAAAT': 45, 'CACAGC': 45, 'AGCGCT': 45, 'ATGGAA': 45, 'TCCTTC': 45, 'CAGCGA': 45, 'AAAGTT': 45, 'CTTGAC': 45, 'GCGAAT': 44, 'GATTAT': 44, 'ACACCC': 44, 'GGAGCA': 44, 'GGATGG': 44, 'ATTTAT': 44, 'GCCTAT': 44, 'TTTAAA': 44, 'GGACTG': 44, 'GAATGC': 44, 'ACTCCC': 44, 'GGTGAT': 44, 'CACAAC': 44, 'GGTGGG': 44, 'GGGCTT': 44, 'CCCCAC': 44, 'ATTGAC': 44, 'TTCGGT': 44, 'GTGTAC': 44, 'AGCCAT': 44, 'CTGGGA': 44, 'AGCGAA': 44, 'GGGATT': 44, 'GTGGCA': 44, 'CTGTGG': 44, 'GGCAGT': 44, 'AAATTA': 44, 'GGTGTC': 44, 'ATTGAA': 44, 'CGCTCG': 44, 'ATTCTC': 44, 'GTGAGG': 44, 'GGTCAG': 44, 'TCGCTG': 44, 'ACAGGC': 43, 'AAATCG': 43, 'GTCGCA': 43, 'GTCGAA': 43, 'GAACGG': 43, 'GACGTG': 43, 'TACCTC': 43, 'AAGATT': 43, 'GAACTT': 43, 'GAGATG': 43, 'GTGGGT': 43, 'AAACAT': 43, 'GCCGTT': 43, 'AGTTTC': 43, 'AATTTT': 43, 'CAGCTG': 43, 'GTTGAG': 43, 'TCTCGC': 43, 'GGCTAT': 43, 'GATGCA': 43, 'ATGGCT': 43, 'GGACGT': 43, 'TCCGAG': 43, 'CATGGA': 43, 'AGTGAA': 43, 'GATCGA': 43, 'AGGGAT': 43, 'AAGTCG': 43, 'CAGCGT': 42, 'GAAGCA': 42, 'GTTGAC': 42, 'GTTGTC': 42, 'ATCTGC': 42, 'AAGAGT': 42, 'AGGGCG': 42, 'GATCAT': 42, 'AATATC': 42, 'GGACTC': 42, 'ACCGGT': 42, 'GGGGCA': 42, 'GAAAGA': 42, 'AGGGTC': 42, 'TACGAG': 42, 'GGTCTC': 42, 'GGTCTG': 42, 'CCGCCC': 42, 'GTGCGA': 42, 'CCGCCG': 42, 'GTGCAT': 42, 'GGGTAC': 42, 'GAACAG': 42, 'TCATCC': 42, 'CATGGT': 41, 'ACCGGA': 41, 'TTGGCT': 41, 'CTTTTT': 41, 'GTTGAT': 41, 'TCCCGA': 41, 'AAAAGA': 41, 'TTGAAT': 41, 'CGCCCA': 41, 'CAAAAT': 41, 'TGCGGG': 41, 'AACAAA': 41, 'AATAAC': 41, 'AAGCGA': 41, 'ATGCTG': 41, 'CTGATG': 41, 'CGCTGC': 41, 'CGTACC': 41, 'ATCAAA': 41, 'TGGAAA': 41, 'AATAGT': 41, 'GGACAG': 41, 'GATAAC': 41, 'GGTTTT': 41, 'AGGACC': 41, 'GTAGCC': 41, 'ATGTCC': 41, 'AAACGA': 41, 'CCCGAA': 41, 'TTTATT': 41, 'GATCGG': 41, 'GGGTAT': 41, 'GAACTG': 41, 'AGAGGC': 41, 'ATGCCG': 41, 'TCCCGG': 41, 'AGGCTC': 41, 'TTCGCG': 41, 'ATTGAG': 41, 'TTCAAG': 41, 'TGCCCC': 41, 'CGTGGT': 41, 'CAGTTT': 41, 'AAACCT': 41, 'AAGATG': 41, 'CTGAAT': 40, 'CCTGAC': 40, 'GTTCAC': 40, 'GACCCC': 40, 'TGCCAC': 40, 'ACCGAA': 40, 'TTGGGG': 40, 'GCAGGT': 40, 'CCCGGT': 40, 'TTGCCT': 40, 'AACTGG': 40, 'CATCAT': 40, 'AGGTTC': 40, 'GGGGGT': 40, 'GCACGG': 40, 'ACGGAA': 40, 'GCTTAC': 40, 'TCCCAC': 40, 'TGCAGC': 40, 'GATGTG': 40, 'GCCCCT': 40, 'AGGAGT': 40, 'CGCCCT': 40, 'TGGCCC': 40, 'CTCCTC': 40, 'AACGCT': 40, 'CACCTG': 40, 'TCAAAA': 40, 'ACCCTG': 40, 'AACTAC': 40, 'GCGTGG': 40, 'CAACTC': 40, 'ACATCC': 40, 'GCAGAA': 40, 'GGTGGA': 40, 'CAAGAT': 40, 'AGAAGC': 40, 'TTGGAG': 40, 'CGCCAT': 40, 'ACGAGG': 39, 'CAAAAG': 39, 'CAAGAA': 39, 'AAACGT': 39, 'CGGGAA': 39, 'AATCAT': 39, 'AGGGGG': 39, 'AAATGG': 39, 'TCCGCA': 39, 'AACCTG': 39, 'CGCCAG': 39, 'CGACCG': 39, 'CTGCGA': 39, 'TTGATT': 39, 'AACGGG': 39, 'AGCCGA': 39, 'GGAGCT': 39, 'TTGTTT': 39, 'GAACAA': 39, 'TGGCAG': 39, 'CATTTC': 39, 'GCGTTG': 39, 'ACGCAG': 39, 'TCCGCT': 39, 'GTACGC': 39, 'CCCCGT': 39, 'GCTCAG': 39, 'GGTAAA': 39, 'GCGGTA': 39, 'AAATTT': 39, 'TCCGAA': 39, 'GTAAAA': 39, 'ACTGAC': 38, 'ATGATG': 38, 'TTGAGG': 38, 'AAACCA': 38, 'CGCGTG': 38, 'CAAGCT': 38, 'GGAGTG': 38, 'GGGACG': 38, 'GATGGT': 38, 'GCACGA': 38, 'TTTAAC': 38, 'TGGGCG': 38, 'CGTCGT': 38, 'TTGCGT': 38, 'ATGCAG': 38, 'GGCATG': 38, 'CCATCC': 38, 'ATCCAG': 38, 'AGCCCG': 38, 'GAACGT': 38, 'AGCTGC': 38, 'GGAATT': 38, 'TCCGGT': 38, 'AAGTAT': 38, 'GAATCG': 38, 'GGACGA': 38, 'GATCGT': 38, 'GCAACC': 38, 'CACGCG': 38, 'TTCCTG': 38, 'GGATTT': 38, 'CCTGCT': 37, 'CGTCAT': 37, 'CCGTGG': 37, 'CTCCAC': 37, 'TTGGCG': 37, 'GTGGGA': 37, 'TTCGGG': 37, 'CGACTC': 37, 'CACACC': 37, 'AACGAA': 37, 'TTAGAT': 37, 'CCGAAT': 37, 'AGGGAA': 37, 'ACGCGA': 37, 'GGGTTG': 37, 'TTGTCC': 37, 'ATGACC': 37, 'AGAGAT': 37, 'AGCGAT': 37, 'GCCATT': 37, 'AACCAG': 37, 'GCTTTC': 37, 'CGCGAA': 37, 'TTCCGT': 37, 'TCGAAG': 37, 'GGTCGT': 37, 'TGCTCC': 37, 'GCCCCA': 37, 'AAGTGG': 37, 'CCCCTG': 37, 'GACGGA': 37, 'AGCAAA': 37, 'GATGGG': 37, 'CGCAGG': 37, 'GCTGTT': 37, 'TTTTAT': 37, 'TCCCCC': 37, 'CGGAGT': 37, 'CAGAAT': 37, 'GCATGG': 37, 'GTGAAT': 37, 'ACCGTG': 36, 'ACTGTC': 36, 'AGCTCC': 36, 'TGGCGA': 36, 'GCAGGG': 36, 'TTTGTT': 36, 'GGCCCG': 36, 'TTCCGG': 36, 'AGAAAC': 36, 'CGTCGG': 36, 'TGGCGT': 36, 'CGGTGG': 36, 'CACGGG': 36, 'CATGCA': 36, 'CGCAAA': 36, 'TTTATC': 36, 'CACATC': 36, 'AAGAGA': 36, 'TTGGTC': 36, 'AATGCG': 36, 'GACAAT': 36, 'TGGCAC': 36, 'GGACAT': 36, 'CTGTTT': 36, 'TCCATC': 36, 'AAGTTT': 36, 'AATACT': 36, 'GTGTTT': 36, 'AATAGC': 36, 'GTGATT': 36, 'CAGATT': 36, 'GGTAAG': 35, 'CAGATG': 35, 'TTGACC': 35, 'AGTGAC': 35, 'GATACC': 35, 'CAATTC': 35, 'ATTCAC': 35, 'CAGTGG': 35, 'GATAAT': 35, 'ACGAAT': 35, 'GCAGGA': 35, 'CCGATG': 35, 'GATTAC': 35, 'CGGTCA': 35, 'ATCAGT': 35, 'TTGCAC': 35, 'TCGCCA': 35, 'CCGTTT': 35, 'TTTGAA': 35, 'CACGGT': 35, 'ACGCAT': 35, 'AAATTG': 35, 'GCTGGT': 35, 'ACAGAA': 35, 'AACCCG': 35, 'AAACCG': 35, 'CGAGAC': 35, 'AAAGTG': 35, 'AACCGT': 35, 'CCGTAC': 35, 'TACATC': 35, 'CGGAGG': 34, 'GCTCCG': 34, 'GTGCTT': 34, 'CCTTTC': 34, 'GATTCA': 34, 'GAAGTA': 34, 'GTCGGA': 34, 'CGTAAC': 34, 'TCGGAA': 34, 'AATCGT': 34, 'TCGCGT': 34, 'ATAAAG': 34, 'AGTAAT': 34, 'ATCTGG': 34, 'TCCGAT': 34, 'GAGGTA': 34, 'GAACGA': 34, 'CGCACG': 34, 'GGAGTT': 34, 'GGACAA': 34, 'CACCGG': 34, 'GATAGC': 34, 'AGGATC': 34, 'GTTAAA': 34, 'CGTGCG': 34, 'GGTCAT': 34, 'GCTTCT': 34, 'CGGAAT': 34, 'GCAAAG': 34, 'GAGTCA': 34, 'CAAACC': 34, 'GTGCAA': 34, 'ACTTTC': 34, 'CAAGCA': 34, 'ATTATT': 34, 'TATTCC': 34, 'AAAGTA': 34, 'ACGCTT': 34, 'GCATGC': 34, 'CGTATC': 34, 'TTTGGT': 34, 'TTGCAT': 34, 'TATAAA': 34, 'TACAAC': 34, 'AAACTG': 34, 'GCTCGA': 34, 'ACGCCT': 34, 'CGGAAA': 34, 'GTGTCG': 34, 'ACAAAA': 34, 'TCAGAC': 34, 'CCCGTG': 33, 'AATCCA': 33, 'GCAAAT': 33, 'GCGACA': 33, 'CAGCTT': 33, 'CACCCG': 33, 'CCAGAC': 33, 'TCGGTG': 33, 'CAAGGT': 33, 'AATCAG': 33, 'GGCATT': 33, 'CCTGAT': 33, 'GTTTAT': 33, 'GAAATA': 33, 'TGGGGG': 33, 'GTAGGC': 33, 'GCTGGA': 33, 'TTGCCG': 33, 'CCCCTC': 33, 'CGGGTT': 33, 'TTTTTT': 33, 'GGATTG': 33, 'GTAAAC': 33, 'ATCATT': 33, 'GCGTTT': 33, 'GCCAGG': 33, 'TCGCCG': 33, 'ACGGTT': 33, 'GAACAT': 33, 'ACATTC': 33, 'GGCTCG': 33, 'CGTCTC': 33, 'TTCAAT': 33, 'AATTCT': 33, 'CGGCAA': 33, 'ATAGAC': 33, 'TTCAAA': 33, 'AAGTCA': 33, 'CAGTCA': 33, 'AATTAC': 33, 'AATCTC': 33, 'CTGACG': 33, 'GGCCTT': 33, 'AGCAAT': 33, 'ACGTAC': 33, 'CGCTAT': 32, 'GACAGT': 32, 'AGCGGG': 32, 'GTCCGT': 32, 'TATTTT': 32, 'ACTTCC': 32, 'GGTCAA': 32, 'ATTCAA': 32, 'CACGCT': 32, 'GCCCTT': 32, 'GATCTT': 32, 'GGATAC': 32, 'CTTTAT': 32, 'CTCAAT': 32, 'ACGACG': 32, 'ACGGAT': 32, 'CAGAGT': 32, 'AGTTCC': 32, 'ATGCGT': 32, 'GAATTA': 32, 'AAAAGG': 32, 'TCGATG': 32, 'CGACGG': 32, 'ATTTCT': 32, 'AGGGTG': 32, 'GCGTCT': 32, 'GATGGA': 32, 'TGGCGG': 32, 'TGCGAG': 32, 'GGTTCG': 32, 'CAGCAA': 32, 'GTTGCT': 32, 'CTCGCA': 32, 'CCCTAC': 32, 'ACTGGC': 32, 'GAATCA': 32, 'GATTGG': 32, 'ACCCCA': 32, 'GGTGCT': 32, 'ATGTAC': 32, 'TTCTGC': 32, 'AAAACT': 32, 'AGAGAC': 32, 'GGGTGG': 32, 'TTATTT': 32, 'CGGTTG': 32, 'CAATCA': 32, 'AGAACC': 32, 'GGAATG': 31, 'ATTATC': 31, 'CAATAT': 31, 'AGGCAC': 31, 'AACAGT': 31, 'GCACAT': 31, 'AGGGGT': 31, 'GTTTAC': 31, 'AATACC': 31, 'AGGGGA': 31, 'GACAGG': 31, 'CACGAT': 31, 'CATCTC': 31, 'AGTAAG': 31, 'TCTTTT': 31, 'CGTCCT': 31, 'CCTGTC': 31, 'TGGTCC': 31, 'ATGTAT': 31, 'ATCGTT': 31, 'CTCGCT': 31, 'TTTGCT': 31, 'AGACCC': 31, 'CGAGGG': 31, 'GTGTTG': 31, 'AGACGC': 31, 'CATGGG': 31, 'GAATCT': 31, 'TGCAAC': 31, 'ACACCG': 31, 'AGGTCC': 31, 'AATCCG': 31, 'GCCTCA': 31, 'CCCCGA': 31, 'GCTATC': 31, 'ATTTTG': 31, 'GGCTTT': 31, 'ACAACC': 31, 'CAAGCG': 31, 'AGGAGG': 31, 'GATTTG': 31, 'TCCAAT': 31, 'GTGTAT': 31, 'GGTATT': 31, 'CCATTC': 31, 'TATGAT': 31, 'GGATAT': 31, 'ATGGTG': 31, 'CTTCTC': 31, 'CCGTCA': 31, 'GTCTGC': 30, 'TTAATT': 30, 'GTAGAT': 30, 'CCCAAG': 30, 'CAAGAG': 30, 'CGGCTT': 30, 'GTTGCA': 30, 'CGTGAG': 30, 'CTTGTC': 30, 'TCAGCT': 30, 'TATCCC': 30, 'GGTGTT': 30, 'GCAGTT': 30, 'TGGGTG': 30, 'TGGCCG': 30, 'AATCCT': 30, 'AGCGGT': 30, 'ACTGAT': 30, 'GGCACG': 30, 'TTGAGT': 30, 'TTCGAA': 30, 'AAGTCT': 30, 'ATCCTG': 30, 'CACGGA': 30, 'AGTAAA': 30, 'CCGAGG': 30, 'GCGTAT': 30, 'GTCCCG': 30, 'AACTGC': 30, 'GAGTTG': 30, 'GTTGGT': 30, 'CCGTGC': 30, 'GAACCG': 30, 'GGTCGA': 30, 'CATGTT': 30, 'CCAGCA': 30, 'CAATGG': 30, 'GCACAA': 30, 'CCGCAT': 30, 'ATTCGT': 30, 'AGGAAT': 30, 'TCAAGC': 30, 'CGTGGG': 30, 'ATGAGG': 30, 'GGAAGT': 30, 'ACCCGT': 30, 'GATAGT': 30, 'GTAGAG': 30, 'AATATG': 30, 'GAAACT': 30, 'CCGACG': 30, 'AGCAGG': 30, 'AGTGAT': 30, 'CTTTCT': 30, 'GAGTAT': 30, 'CTTCAC': 30, 'AACCAT': 29, 'GAGTCT': 29, 'CTCTGG': 29, 'TCGCCT': 29, 'CGTTCG': 29, 'CGCGTT': 29, 'TTGGCA': 29, 'AGCGTT': 29, 'CTGTAT': 29, 'CCATGC': 29, 'TTAAAG': 29, 'ATTGTC': 29, 'GTGCCT': 29, 'ATGATT': 29, 'CCAAAA': 29, 'ATCGCT': 29, 'CGTCGA': 29, 'TACTAC': 29, 'TACAGC': 29, 'AGAATC': 29, 'TATCTG': 29, 'GCTAAA': 29, 'AGGAGA': 29, 'TACGAA': 29, 'GCTCGT': 29, 'GTGTGC': 29, 'CTGAGT': 29, 'GGAAGA': 29, 'GATCCC': 29, 'GCTTTT': 29, 'CGTTTT': 29, 'CCTGAG': 29, 'ATTCAG': 29, 'CCCTCG': 29, 'TTCTAC': 29, 'AGAGGG': 29, 'AGATCC': 29, 'ACAAAG': 29, 'TTCCAG': 29, 'TGCCAG': 29, 'GTAAGC': 29, 'GACATG': 29, 'ACCAAT': 29, 'CTGTGC': 29, 'ACGGGT': 29, 'CAACCG': 29, 'GCGAGA': 29, 'CGAGGA': 29, 'CGCAAT': 29, 'GTCCGA': 29, 'AACCAA': 28, 'CGCTCA': 28, 'GGTCTT': 28, 'GCCTCT': 28, 'ATTAAT': 28, 'ACTCTC': 28, 'CGCATG': 28, 'ATCCGT': 28, 'AGGCGG': 28, 'TCACAC': 28, 'AGAGAA': 28, 'TATATC': 28, 'AGCTTT': 28, 'TCAATC': 28, 'CACGCA': 28, 'GGACTT': 28, 'TGGCTT': 28, 'GCTTCG': 28, 'TATCAA': 28, 'TTTTCA': 28, 'GAATTG': 28, 'GCCACG': 28, 'ACCCAG': 28, 'AGGCTG': 28, 'TCGGAT': 28, 'TGGGTC': 28, 'TCGGGG': 28, 'AGTGTC': 28, 'TTGCTG': 28, 'CAACAC': 28, 'GTTAAT': 28, 'TATTAT': 28, 'TTCGGA': 28, 'AGTCAC': 28, 'CAAGGG': 28, 'CATTAT': 28, 'GTGACG': 28, 'CGGACA': 28, 'TCCCTC': 28, 'TACCCC': 28, 'GGTTTG': 28, 'CTCGAG': 28, 'CCGAAA': 28, 'ACGATG': 28, 'ACAGAC': 28, 'TTTTCT': 28, 'ACCGTT': 28, 'TCATGC': 28, 'TTTGTC': 28, 'CTCCGG': 28, 'AGGCAG': 28, 'AGCGTG': 28, 'AATGCA': 28, 'CAACAA': 28, 'TACCAC': 28, 'CGTAGC': 28, 'CCTCGG': 28, 'TTTGCG': 28, 'TCTAAA': 27, 'TCTGAA': 27, 'TTGTTG': 27, 'TCGTAC': 27, 'ATTGCT': 27, 'GAGACG': 27, 'ATCGGA': 27, 'TTGGTG': 27, 'ATGCGG': 27, 'TTCCCC': 27, 'AGGTAC': 27, 'CATCCG': 27, 'CCTGCG': 27, 'ACGTGG': 27, 'TATCTC': 27, 'TCCAAG': 27, 'TCAGGC': 27, 'TTTCAA': 27, 'AGGCAT': 27, 'TCACCG': 27, 'GTCCTG': 27, 'ACCTAT': 27, 'ATTAAG': 27, 'TCAGCA': 27, 'TACCAG': 27, 'GATTCG': 27, 'CAAGGA': 27, 'TCTGCT': 27, 'AGGGCT': 27, 'TTAGAA': 27, 'TTGGGT': 27, 'ACAAAT': 27, 'TGCGGT': 27, 'GTTTTT': 27, 'CAGTAT': 27, 'TCGCAG': 27, 'ATTGGG': 27, 'GTTCTC': 27, 'GCACTG': 27, 'TCCGGG': 27, 'ATCCCG': 27, 'ACACAC': 27, 'CAATTT': 27, 'CCATGG': 27, 'ACTGCG': 27, 'GTGTCA': 27, 'CGAATC': 27, 'GAAATG': 27, 'AGGCGT': 27, 'CTTGGT': 27, 'GACCAT': 27, 'CAACCT': 27, 'ATAGAT': 27, 'CGCAGT': 27, 'GCTTAT': 27, 'GATTCT': 27, 'ACCCCT': 27, 'AAGATA': 27, 'AACCGA': 26, 'CAACGG': 26, 'GATTTA': 26, 'GCGTGC': 26, 'TCCTAC': 26, 'GGGTTT': 26, 'AGAAAT': 26, 'GCTGTG': 26, 'AGTTTT': 26, 'AACAGG': 26, 'GGCCAA': 26, 'AGTAAC': 26, 'CGTCAG': 26, 'TTGGTT': 26, 'GAGAGT': 26, 'CTGTTG': 26, 'AACATG': 26, 'CGATCA': 26, 'CTTGTT': 26, 'ATGTGG': 26, 'AGGCGA': 26, 'TACCGG': 26, 'GCCATG': 26, 'TTGCAG': 26, 'TTTCAC': 26, 'TACCTG': 26, 'GTTTTC': 26, 'TCTCAT': 26, 'GCGATT': 26, 'GTTTCT': 26, 'GATCAA': 26, 'GTGTGG': 26, 'TCGCAT': 26, 'CTCGGA': 26, 'GTGAGT': 26, 'TCCCAT': 26, 'TCCCGT': 26, 'ATGGGT': 26, 'CATTTT': 26, 'CGGTTT': 26, 'ATTCTT': 26, 'TCATTC': 26, 'TACCGT': 26, 'TTTCCC': 26, 'GTTCAG': 26, 'GCATCT': 26, 'GGCTCT': 26, 'CGCATT': 26, 'GGCTTG': 26, 'GCCCAA': 26, 'TCGGTT': 26, 'AATCGA': 26, 'ATAAAT': 26, 'CACGAG': 26, 'CCAAGC': 26, 'ATTAGC': 26, 'GACCAA': 26, 'ATTCCT': 26, 'GTCAAA': 26, 'AATTTA': 26, 'TCAGAA': 26, 'GGTTAT': 26, 'ACTGCT': 26, 'GCTACC': 26, 'AGCGGA': 26, 'TCAGAT': 26, 'AACTCT': 26, 'ATTTTA': 26, 'CTCAAA': 25, 'CACTGG': 25, 'ATGCCT': 25, 'CTAGCC': 25, 'CTTGAG': 25, 'TGGCAT': 25, 'CACTGC': 25, 'ACAGCT': 25, 'ACTTCA': 25, 'GCAAGT': 25, 'ACCCAT': 25, 'TCGAGG': 25, 'AAGGTA': 25, 'CACCGT': 25, 'ATAGCC': 25, 'AAGTTG': 25, 'ACGAAA': 25, 'AGATTT': 25, 'ATGCCA': 25, 'ATCAGG': 25, 'TCTTCT': 25, 'AGATAT': 25, 'GAACCT': 25, 'TTTAGT': 25, 'CGAGCG': 25, 'AATCTT': 25, 'CATAAA': 25, 'CACCGA': 25, 'TTTATG': 25, 'ACCAGG': 25, 'AGAGCG': 25, 'TTAAGC': 25, 'TCCTGG': 25, 'AGAATT': 25, 'ACACCA': 25, 'AGGCCG': 25, 'GACGTT': 25, 'ACCTTT': 25, 'ATCCAT': 25, 'CACAAG': 25, 'GAGCTA': 25, 'AATATA': 25, 'CCCCAT': 25, 'TTACCC': 25, 'TATGTT': 25, 'GCAGTG': 25, 'GAGTTA': 25, 'AGTGTT': 25, 'TCCCCG': 25, 'TCAGTT': 25, 'GTCGTG': 25, 'GGTACC': 25, 'AGTCAT': 25, 'GGTGCA': 25, 'AACATT': 25, 'TTAAGT': 25, 'AAGTTA': 25, 'CGATTC': 25, 'ACGGGA': 25, 'TCGACG': 25, 'TCTCCC': 25, 'CAGTCT': 25, 'AATTCG': 25, 'AGCTAT': 25, 'CTGAGG': 25, 'TTCAGT': 25, 'CGTTGG': 25, 'TTATTA': 25, 'AGCCAA': 25, 'GTCTGG': 25, 'GATATG': 24, 'CGTGAT': 24, 'TATTAC': 24, 'TTAAAC': 24, 'CTTTAC': 24, 'GGCACA': 24, 'TTTCAT': 24, 'GTTCGT': 24, 'ATGGTT': 24, 'CTTAAT': 24, 'GGGTGC': 24, 'TTATCC': 24, 'ACAGAT': 24, 'ATTGTG': 24, 'TGGACG': 24, 'AAGACT': 24, 'ACCAGT': 24, 'TCGCTT': 24, 'TCGTCA': 24, 'ATGCTT': 24, 'AAATGC': 24, 'AACCTT': 24, 'AGAGGT': 24, 'GTTGCG': 24, 'AACGGA': 24, 'TACTCC': 24, 'CACATG': 24, 'GCGCTA': 24, 'ACGCAA': 24, 'GTCCAG': 24, 'AGGCTT': 24, 'GCTCGG': 24, 'TCTTTC': 24, 'CTTTCG': 24, 'ACGCGT': 24, 'GGAATA': 24, 'TCTCAC': 24, 'ACAAAC': 24, 'TCAGCG': 24, 'GTATTC': 24, 'CCTCGA': 24, 'TCAGAG': 24, 'GCCTTT': 24, 'AGTGCT': 24, 'ATTAGT': 24, 'ATCGTG': 24, 'GCTCCT': 24, 'ACTTTT': 24, 'CGTTCA': 24, 'CTGTCA': 24, 'CAATAC': 24, 'GCACCT': 24, 'GCGACT': 24, 'TGGGGA': 24, 'GGACCG': 24, 'AAGTGC': 24, 'GACCCG': 24, 'TCCCTG': 24, 'ATATTC': 24, 'TACGAT': 24, 'CTACGC': 24, 'CTCCCG': 24, 'CGGTAT': 24, 'AGCCTT': 24, 'GCTTCA': 23, 'GAGATA': 23, 'TCAAAC': 23, 'TTGATG': 23, 'TGGGGT': 23, 'TATTTG': 23, 'TCAAAT': 23, 'TTGCGG': 23, 'ACAAGC': 23, 'CGTTAT': 23, 'CCGTTG': 23, 'GTTTTA': 23, 'ATTTAC': 23, 'CATCTG': 23, 'ACTCGA': 23, 'GGTTCT': 23, 'CGGGTA': 23, 'TATATT': 23, 'CGACGA': 23, 'CTTCCG': 23, 'AAGCTA': 23, 'CGATAC': 23, 'TTCGCA': 23, 'TTTAGC': 23, 'ATGGCA': 23, 'GTCAAT': 23, 'ATTAAC': 23, 'ACAGTC': 23, 'CCGTCT': 23, 'AACGTG': 23, 'GTTCCG': 23, 'GGGAGA': 23, 'CCACGA': 23, 'ATTCAT': 23, 'TTAGCC': 23, 'GCTGGG': 23, 'GGGTCT': 23, 'CGCCAA': 23, 'GAACCA': 23, 'TACAAG': 23, 'TGGATT': 23, 'CTCTAT': 23, 'GCTCAA': 23, 'CGTGCT': 23, 'TCTGGC': 23, 'GTCTCG': 23, 'GTTGTT': 23, 'TATAAT': 23, 'TTAATC': 23, 'GCTTGC': 23, 'GGTCCC': 23, 'CCTCAG': 23, 'CTCCGA': 23, 'CTTGAA': 23, 'ACGTCA': 23, 'AATGGG': 23, 'CGAGGT': 23, 'CACCAT': 23, 'GATTGC': 23, 'ATTGGT': 23, 'CAAGTT': 22, 'AGAAGA': 22, 'TCCGGA': 22, 'GCTCAT': 22, 'CTTAGC': 22, 'ATAGAA': 22, 'CCGGTA': 22, 'GAGCCA': 22, 'TTGCTT': 22, 'CCATTT': 22, 'TGCTTC': 22, 'TCGAAT': 22, 'TTTGCA': 22, 'TGCGCG': 22, 'AGTCCC': 22, 'CCCCAG': 22, 'GCTTTG': 22, 'GCCTTG': 22, 'ACCTCG': 22, 'AACCCT': 22, 'TGGCAA': 22, 'ATGCAT': 22, 'CGGATT': 22, 'TTAGGC': 22, 'AACTAT': 22, 'AGAGGA': 22, 'CTGATA': 22, 'GTCAGT': 22, 'TGTGCC': 22, 'TACGCG': 22, 'GCATTG': 22, 'CTACCC': 22, 'ATCTAT': 22, 'TGGGCA': 22, 'TCACCC': 22, 'TCCTGC': 22, 'AGGACG': 22, 'AGTTGG': 22, 'TCGAAA': 22, 'GGTGTG': 22, 'ACGTTT': 22, 'GACTCG': 22, 'AGTTAT': 22, 'GGTTGG': 22, 'CCCGGA': 22, 'ATATCC': 22, 'CTTAAC': 22, 'TATGCT': 22, 'TCGGGT': 22, 'AGTGCG': 22, 'CGTGCA': 22, 'CCACAC': 22, 'TGGAGG': 22, 'TCGCGA': 22, 'TTTGGG': 22, 'CAATCT': 22, 'AATCTG': 22, 'TCCAAA': 22, 'CTTCAT': 22, 'ATTGCG': 22, 'TGCGAT': 22, 'GTCGTT': 22, 'TCCTTT': 21, 'CATTCG': 21, 'CGAAAA': 21, 'CATCAG': 21, 'TGCGTC': 21, 'ATCGCA': 21, 'GCTCTG': 21, 'CTCGAA': 21, 'TCGGGA': 21, 'GTCTAT': 21, 'TGCAAG': 21, 'CTCGTT': 21, 'GTAGCG': 21, 'TCGCAA': 21, 'ATGTCG': 21, 'ATGCGA': 21, 'AGTGGC': 21, 'AACTTT': 21, 'TTCGCT': 21, 'CTGATT': 21, 'CAACAG': 21, 'GTAATC': 21, 'TCTGAT': 21, 'ATTACC': 21, 'ATATAT': 21, 'GGAACG': 21, 'GAAACA': 21, 'CGATCG': 21, 'CTATTT': 21, 'GAATGT': 21, 'TACACC': 21, 'TCTGAC': 21, 'ATTCTG': 21, 'ATTCGA': 21, 'CATAAC': 21, 'CAACTT': 21, 'CGACAC': 21, 'TCAATT': 21, 'ACGTGC': 21, 'CGCTCT': 21, 'CCTGTG': 21, 'GCTCCA': 21, 'CACCAG': 21, 'GCAAGG': 21, 'GTGTCT': 21, 'GTTCAA': 21, 'ACTCTT': 21, 'CCTCCC': 21, 'CGTCCA': 21, 'AATTCA': 21, 'TGCACC': 21, 'CAGAGG': 21, 'ACAGGT': 21, 'ATAAGC': 21, 'ACACCT': 21, 'CCACGG': 21, 'GTATCC': 21, 'AAGACA': 21, 'GCTAAG': 21, 'TTTAAG': 21, 'TTACGC': 21, 'CGAACG': 21, 'CTTATC': 21, 'TATCGT': 21, 'CTAAAA': 21, 'CCAATC': 21, 'CATATC': 21, 'CCCCCC': 21, 'CTTGCT': 20, 'TTTGGA': 20, 'GTCTTT': 20, 'TGCCGG': 20, 'TACTGC': 20, 'AATTGG': 20, 'TTCCAT': 20, 'ATACAA': 20, 'ACACTC': 20, 'CACCCT': 20, 'CATTCT': 20, 'GCCCTA': 20, 'TCATGG': 20, 'TTGGAA': 20, 'TTAGGT': 20, 'TCTTTG': 20, 'AGATAC': 20, 'TCACAA': 20, 'AGAGCT': 20, 'CGAGCA': 20, 'TCTCTT': 20, 'GTTTGC': 20, 'AAATGT': 20, 'CTTGCG': 20, 'CCTGGA': 20, 'AGAAGG': 20, 'AGGCCT': 20, 'TTCTTT': 20, 'GTTCCT': 20, 'GAGTGT': 20, 'AAAACA': 20, 'AATTGC': 20, 'TCAGGA': 20, 'CCTGAA': 20, 'GCAACG': 20, 'GGATGC': 20, 'GATTGT': 20, 'CTCCTG': 20, 'TCTTCA': 20, 'GACATT': 20, 'ACTTAC': 20, 'CTGACT': 20, 'TTACTT': 20, 'TTTTTG': 20, 'CACAAT': 20, 'TTGTCT': 20, 'TCCTCG': 20, 'GCACGT': 20, 'TCATCG': 20, 'ATACGC': 20, 'AGCCCT': 20, 'GCCACA': 20, 'AGGTCG': 20, 'CGTCTT': 20, 'CAACTG': 20, 'CAAGTC': 20, 'CCTTTT': 20, 'GTATTT': 20, 'TCCTCT': 20, 'CAATCG': 20, 'ATCCCT': 20, 'TCGTGG': 20, 'AGGATT': 20, 'TTCTAT': 20, 'CAAATT': 20, 'AGTCGT': 20, 'GGTAGT': 20, 'ACTTCT': 20, 'TTCATT': 20, 'GTAGAA': 20, 'CAACGT': 20, 'CAGAGA': 20, 'AGCAGT': 20, 'AGTCGA': 20, 'TGCAAA': 20, 'ATCCGA': 20, 'TCTTCG': 20, 'ACTAAT': 19, 'AGGCCA': 19, 'TTCCCG': 19, 'CCTCAC': 19, 'TTTCGG': 19, 'TTGTAT': 19, 'CTCTCG': 19, 'CATTCA': 19, 'TCAACC': 19, 'CACAAA': 19, 'AGTTTG': 19, 'CAACAT': 19, 'AGTTCG': 19, 'CGTAAA': 19, 'ATGCAA': 19, 'GGCACT': 19, 'GTTCTT': 19, 'TGGACA': 19, 'AAACTA': 19, 'TCTAAT': 19, 'AGCAGA': 19, 'GTTAAG': 19, 'ACAGCG': 19, 'AGTCTC': 19, 'CAAATC': 19, 'ATGTTG': 19, 'ATAAGT': 19, 'GGTAAT': 19, 'TATCAT': 19, 'ATTGGA': 19, 'TGTCCC': 19, 'CCACTG': 19, 'GTAAAG': 19, 'CCAGAT': 19, 'ACAGCA': 19, 'TGGAAT': 19, 'GCTATT': 19, 'CTGTCT': 19, 'TTAGTT': 19, 'AGTGGA': 19, 'CGCTTT': 19, 'TATTGC': 19, 'TATAAC': 19, 'TCCAGT': 19, 'ATGGGA': 19, 'CCGCTA': 19, 'AGATTC': 19, 'TATCCG': 19, 'CATTAC': 19, 'ACTGAA': 19, 'GCTAAC': 19, 'CGAGTC': 19, 'CATCGT': 19, 'AGGTTT': 19, 'ATTTGT': 19, 'CCTGTT': 19, 'GAACTA': 19, 'CGTGAA': 19, 'ACTCAT': 19, 'CGACGT': 19, 'ACATGG': 19, 'GGATCA': 19, 'CTCGTG': 19, 'GTTCAT': 19, 'TTTACC': 19, 'GTTCGG': 19, 'CCAGAG': 19, 'TCTGAG': 19, 'CATAGC': 19, 'GTTTTG': 19, 'ACGTCG': 19, 'GGCTGT': 19, 'AGTATT': 19, 'GGGCCT': 19, 'CTCTTT': 19, 'GTCAGG': 19, 'TCACCA': 19, 'GCTTGG': 18, 'CGAAAT': 18, 'AGACGA': 18, 'GATAGA': 18, 'TTAGAG': 18, 'ACTTAT': 18, 'GGCAGA': 18, 'GCCTGT': 18, 'GTAGCT': 18, 'ATTGCA': 18, 'CCATCA': 18, 'AGTCGG': 18, 'ATGTTT': 18, 'CGTGTT': 18, 'GGATTA': 18, 'GCCAGA': 18, 'AGCATG': 18, 'CTTTTG': 18, 'ACTCCA': 18, 'TTGCCA': 18, 'AATTTG': 18, 'TGTCAG': 18, 'ACCAAA': 18, 'TTATTC': 18, 'CAGACT': 18, 'ATTGTT': 18, 'CCAAAG': 18, 'CGAAGG': 18, 'TATCTT': 18, 'AACCCA': 18, 'GTCATG': 18, 'CTCTGC': 18, 'TTTCTC': 18, 'TTCAGG': 18, 'TATTTA': 18, 'TACGGG': 18, 'GTTTCG': 18, 'GCATTT': 18, 'TCCTCA': 18, 'TCACTC': 18, 'CGTTCT': 18, 'CCAGTC': 18, 'GTTAGC': 18, 'CCGATT': 18, 'AGAATA': 18, 'GCATAT': 18, 'ACACTT': 18, 'CACGAA': 18, 'TCCCAA': 18, 'GTATCT': 18, 'TTCTCG': 18, 'AATGTA': 18, 'ATAAAC': 18, 'TCAGTC': 18, 'CCCAGT': 18, 'TTTGTG': 18, 'TTAACC': 18, 'CAACCA': 18, 'CCAGCG': 18, 'CGGATA': 18, 'GATATA': 18, 'GCAATT': 18, 'CTTCTT': 18, 'AGTTAC': 18, 'GGCCCA': 18, 'CAGTGC': 18, 'CGTATT': 18, 'TCATCA': 18, 'GCTAAT': 18, 'ACAATC': 18, 'GGCTCA': 18, 'GACCTT': 17, 'ACTAAA': 17, 'ACCGTA': 17, 'ATCCAA': 17, 'AACACT': 17, 'GATGTA': 17, 'CCAACC': 17, 'CTTACC': 17, 'TACGCA': 17, 'CTCCAG': 17, 'CAAAGT': 17, 'AGGTTG': 17, 'GCTCTT': 17, 'AGAGTG': 17, 'GTTATC': 17, 'AATAGA': 17, 'GGATCT': 17, 'ACATGC': 17, 'TTTTTA': 17, 'CCAAAC': 17, 'ATAGGG': 17, 'CGATGG': 17, 'CTTATT': 17, 'CGCTTG': 17, 'CGAAAC': 17, 'TGCCGA': 17, 'CCGACA': 17, 'CTCAGT': 17, 'GGTATG': 17, 'TCTGTC': 17, 'TTGGGA': 17, 'TTCATG': 17, 'TTGTAC': 17, 'CCTGGT': 17, 'TGTTCC': 17, 'ATTAGA': 17, 'GCTAGC': 17, 'CGTCTG': 17, 'CCGCCA': 17, 'ACAGTT': 17, 'ACAGGG': 17, 'CGAGCT': 17, 'CGTAAT': 17, 'ACAATT': 17, 'TCTTAT': 17, 'GTTATT': 17, 'TCCGTT': 17, 'TTACAA': 17, 'GGTCCG': 17, 'GCACTT': 17, 'AGTCAA': 17, 'ATGAGT': 17, 'CTCACT': 17, 'TGCTAC': 17, 'TCTCCG': 17, 'GGAGTA': 17, 'TCACTT': 17, 'CAGCTA': 17, 'GTTAAC': 17, 'TTTCAG': 17, 'CCTCGT': 17, 'TCCCAG': 17, 'AGACAT': 17, 'TTATCT': 17, 'TCTGCA': 17, 'CCTCTT': 17, 'CCAGAA': 17, 'CTATCC': 17, 'TATCGG': 17, 'AGCTCG': 17, 'AATCTA': 17, 'CCCTCA': 17, 'TATCAC': 17, 'TGTCGC': 17, 'GAGAGA': 17, 'TCGACA': 17, 'CCCTAT': 17, 'GACACG': 17, 'GACTCA': 17, 'GCCGTA': 17, 'ACTGGT': 17, 'AGTAGG': 17, 'TTTTGT': 17, 'ACTCGT': 17, 'CCACGT': 17, 'GGACCC': 17, 'AGGTGG': 17, 'TTCGTG': 17, 'AACAGA': 17, 'TGGTAT': 16, 'CGGAGA': 16, 'GGTCCT': 16, 'CAAACT': 16, 'CACTCG': 16, 'ATATTG': 16, 'GATCCG': 16, 'ATGTGC': 16, 'TTCCAA': 16, 'CGACTG': 16, 'TTTCTT': 16, 'GTTGGG': 16, 'CTTCAG': 16, 'TTTTAC': 16, 'TACCGA': 16, 'TCTGTT': 16, 'GTTCTG': 16, 'AGTATC': 16, 'GTGCCA': 16, 'CCCCAA': 16, 'ACGATT': 16, 'GCACCA': 16, 'CCCAGG': 16, 'TGTGAT': 16, 'AGGTAT': 16, 'ATCTCT': 16, 'ACATTT': 16, 'TCGTGC': 16, 'GGGGTA': 16, 'GCAACT': 16, 'GACTAT': 16, 'TATGGT': 16, 'AGTCTG': 16, 'CTAGGC': 16, 'CATATT': 16, 'GTTTCA': 16, 'TATTCA': 16, 'TCGATT': 16, 'AACGTA': 16, 'GTAGGT': 16, 'GGAACT': 16, 'TTATTG': 16, 'CACTAT': 16, 'TGGATG': 16, 'GTTCGA': 16, 'CGCACA': 16, 'ACCTCA': 16, 'AATGTG': 16, 'AGGGTT': 16, 'CCCTTT': 16, 'TCTCGA': 16, 'TTAGCT': 16, 'CTCCGT': 16, 'TACCAA': 16, 'AGACAG': 16, 'GTACTT': 16, 'CCTTGC': 16, 'TTTCTG': 16, 'CGACCT': 16, 'AGAGCA': 16, 'TGTGAC': 16, 'CCCTCT': 16, 'CTGGTA': 16, 'GGTTTA': 16, 'CCCGTT': 16, 'CTAAAT': 16, 'TATACC': 16, 'CAAAGG': 16, 'ACCCTT': 16, 'ACTGAG': 16, 'CTCTCT': 16, 'ACACGT': 16, 'TTACCT': 16, 'ATCGTA': 16, 'GCCACT': 16, 'TGGTCG': 16, 'TGGTTT': 16, 'TGGTTG': 16, 'TCCATT': 16, 'CGGTGC': 16, 'CCTGCA': 16, 'TCACAT': 16, 'ATAGGC': 16, 'GTCATT': 16, 'ACACAG': 16, 'ACGTCT': 16, 'ACCATT': 16, 'ACAGAG': 16, 'TCTCGT': 16, 'TCTGTG': 16, 'GTTGGA': 16, 'TCTCAA': 16, 'CGCCTT': 16, 'TTAGAC': 16, 'ATATTT': 16, 'CTTTCA': 16, 'GACTCT': 16, 'TCCCCT': 16, 'GTGAGA': 16, 'ATCCTT': 16, 'ATTTCA': 16, 'CAACGA': 16, 'TCACGG': 16, 'CGAGAG': 16, 'CCCTTG': 16, 'AGATCA': 16, 'CGTGGA': 16, 'AGAGAG': 16, 'TTTTCG': 16, 'TGCAAT': 16, 'CATCGG': 16, 'ACACGA': 16, 'TTGTGG': 15, 'ACACGG': 15, 'AGCTTG': 15, 'GACAGA': 15, 'TGGGCT': 15, 'TCTGCG': 15, 'CTTGGG': 15, 'TTATAT': 15, 'TATCAG': 15, 'TTCTGG': 15, 'CGGACT': 15, 'CCTCTC': 15, 'TACGTG': 15, 'ACGAGT': 15, 'CTTCCT': 15, 'ATCTGT': 15, 'CCTATC': 15, 'GTACGG': 15, 'ATAGAG': 15, 'TCTATC': 15, 'TGCCCG': 15, 'CTTGGA': 15, 'CACGTG': 15, 'CCACCC': 15, 'AAAACG': 15, 'TGTGCT': 15, 'CCCACT': 15, 'TGGCCA': 15, 'ACTCCT': 15, 'CATAAT': 15, 'ACTGCA': 15, 'GGTTCA': 15, 'ATTATG': 15, 'CGACAG': 15, 'GGCGTA': 15, 'ATGTCT': 15, 'AGCACT': 15, 'CTTCGG': 15, 'TCGAGT': 15, 'GTCCTT': 15, 'CAAAGA': 15, 'TCCACG': 15, 'TATGCG': 15, 'AGTCTT': 15, 'ACTAAC': 15, 'GGTTGC': 15, 'CACACA': 15, 'AACTCG': 15, 'TCCAGG': 15, 'CATGTG': 15, 'AGACTC': 15, 'CTGCTA': 15, 'ACGACA': 15, 'TACTGG': 15, 'ATTCCG': 15, 'TTTATA': 15, 'GTTGTA': 15, 'CTCCAT': 15, 'TTGCGA': 15, 'GGGCTA': 15, 'CCTTAC': 15, 'GACGTA': 15, 'AGGGCA': 15, 'AGACAC': 15, 'ATTTCG': 15, 'TTGTCA': 15, 'TTACAG': 15, 'CCGCCT': 15, 'GAGACA': 15, 'ATTACT': 15, 'GCGTGT': 15, 'CTAAGC': 15, 'ACTCCG': 15, 'CGATTT': 15, 'TGGTGC': 15, 'GTCCAT': 15, 'ATGACG': 15, 'CAATGC': 15, 'ACCACA': 15, 'GGGATA': 15, 'GGGACT': 14, 'ACTGTA': 14, 'TTGACT': 14, 'CCTTAT': 14, 'ATGGTA': 14, 'ACATCG': 14, 'GTCTTG': 14, 'GAAACG': 14, 'TCTAAG': 14, 'ACCAGA': 14, 'TCATTT': 14, 'ACCCAA': 14, 'CCGAGT': 14, 'TTACAT': 14, 'CCATAC': 14, 'GACTTG': 14, 'TGTCAC': 14, 'AGGACA': 14, 'ATATGG': 14, 'CTTAAA': 14, 'TCTTAC': 14, 'CGGTCT': 14, 'TTAAGA': 14, 'TCACGA': 14, 'AGTCAG': 14, 'ACTATC': 14, 'TGCCAT': 14, 'CAGGTA': 14, 'CCACAG': 14, 'CGCACT': 14, 'TATAAG': 14, 'TGCTGT': 14, 'CGATGC': 14, 'CATCGA': 14, 'AAGACG': 14, 'TCCCCA': 14, 'TTGATA': 14, 'AGTAGC': 14, 'GTGATA': 14, 'CAATTA': 14, 'TCTCGG': 14, 'GGGTTA': 14, 'CCTCAA': 14, 'ACTCGG': 14, 'ACGTAT': 14, 'ATTCGG': 14, 'TATTCT': 14, 'GTATAT': 14, 'GTCAGA': 14, 'CTACAC': 14, 'ATAGGT': 14, 'TATGCA': 14, 'CATCCA': 14, 'CTCACA': 14, 'TTGAGA': 14, 'GACTTT': 14, 'ACTGTT': 14, 'AGTAGT': 14, 'AGACCG': 14, 'ATAGCT': 14, 'CTTTGC': 14, 'TACAGG': 14, 'CACTCA': 14, 'ATACGG': 14, 'TTGGTA': 14, 'ACATCA': 14, 'TTCACG': 14, 'CGTCAA': 14, 'CTGACA': 14, 'ATCACT': 14, 'TTACTC': 14, 'CGCAGA': 14, 'AGTGGT': 14, 'TCACGT': 14, 'AACGTT': 14, 'TGTAAT': 14, 'CCTTCA': 14, 'ATGATA': 14, 'AGTGTG': 14, 'CCAAGG': 14, 'AGAACG': 14, 'CCACTT': 13, 'GACACT': 13, 'AGTACC': 13, 'ACTATT': 13, 'GGAACA': 13, 'CCCAAA': 13, 'ACGCTA': 13, 'CATTGG': 13, 'AGGATG': 13, 'ATCCCA': 13, 'CTCAGG': 13, 'TTTCGT': 13, 'ACGGTA': 13, 'TCAGGT': 13, 'CCTTCG': 13, 'CTTATA': 13, 'CAGACA': 13, 'CAGTTA': 13, 'GTTGTG': 13, 'GGGTCA': 13, 'GTCCAA': 13, 'TCCAGA': 13, 'GTAACC': 13, 'AGACTG': 13, 'ACTCAG': 13, 'CGAACT': 13, 'TGTTGC': 13, 'CCTGTA': 13, 'CTTTGT': 13, 'GACACA': 13, 'GTACCC': 13, 'CGACAT': 13, 'TCTCTC': 13, 'GCAAGA': 13, 'CTTCAA': 13, 'TCGCTA': 13, 'CGACCA': 13, 'ATGAGA': 13, 'CAGACG': 13, 'CCAAAT': 13, 'CCAGCT': 13, 'CCAGGA': 13, 'GCTGTA': 13, 'CGAGAT': 13, 'ATAATA': 13, 'TGGGTT': 13, 'TTATCG': 13, 'TACGGT': 13, 'CAAGTA': 13, 'TCCGTG': 13, 'CTCCAA': 13, 'CTCCCT': 13, 'AGAACT': 13, 'AGATGG': 13, 'GTAGAC': 13, 'TACTTT': 13, 'CGAAAG': 13, 'CCACTC': 13, 'ATAGCG': 13, 'ACATCT': 13, 'ATATGC': 13, 'ACCTCT': 13, 'CTTGCA': 13, 'CCTTGG': 13, 'TTCACT': 13, 'GCATAC': 13, 'TCGTTG': 13, 'CGTTGC': 13, 'ACCACG': 13, 'CCAGTT': 13, 'AGAGTT': 13, 'ACTTTA': 13, 'GTAGTC': 13, 'GCAACA': 13, 'TCAAAG': 13, 'TATATG': 13, 'CTAAGT': 13, 'TCATAT': 13, 'TATGGA': 13, 'ACTACC': 13, 'TGCATC': 13, 'CACCCA': 13, 'CCCCCG': 13, 'CATTGC': 13, 'AGTTGC': 13, 'TGTTAT': 13, 'TCACAG': 13, 'GATCTA': 13, 'AGCTTA': 13, 'GGGACA': 13, 'GCAGTA': 13, 'ATCTCA': 13, 'CTAGGG': 13, 'GCTAGT': 13, 'TCTTTA': 13, 'GGCATA': 13, 'TCAAGT': 13, 'GTTAGT': 13, 'CTAGAC': 13, 'AGCATT': 13, 'TATGTG': 13, 'GTAAAT': 12, 'TCATTA': 12, 'ATCTTG': 12, 'TATCGA': 12, 'TTTTGC': 12, 'TGCGAA': 12, 'TCACTG': 12, 'ACCACT': 12, 'CACCTT': 12, 'TTGTGC': 12, 'TGTCCA': 12, 'AATAGG': 12, 'CATAGT': 12, 'CGTTTG': 12, 'CCATAT': 12, 'CATCAA': 12, 'TGGAGT': 12, 'TACGCT': 12, 'TGCCCT': 12, 'CTTTTA': 12, 'GTAGGG': 12, 'TTAATA': 12, 'AGGCAA': 12, 'ACCTTG': 12, 'ACTTTG': 12, 'CTATCT': 12, 'CTAAAC': 12, 'TATAGC': 12, 'TCGTTT': 12, 'GTACCG': 12, 'TCCACT': 12, 'CCGACT': 12, 'AACTCA': 12, 'TTAACT': 12, 'CATTTG': 12, 'TACAAT': 12, 'CGAACA': 12, 'ACCATG': 12, 'TCAGGG': 12, 'CTAAAG': 12, 'TACTAT': 12, 'AGAGTC': 12, 'CGAATT': 12, 'TGCTTT': 12, 'AGGTCA': 12, 'ATACCT': 12, 'TATTGG': 12, 'TGGATA': 12, 'TGGCTA': 12, 'AGAATG': 12, 'AGCTCT': 12, 'GGTTGT': 12, 'TCCTTG': 12, 'CACAGT': 12, 'TGTGAG': 12, 'ATAGGA': 12, 'TCTCCT': 12, 'ACACAT': 12, 'CAATTG': 12, 'CTTTGG': 12, 'GTTTGG': 12, 'AGATTA': 12, 'TCTCAG': 12, 'ACACAA': 12, 'TTGTTA': 12, 'TTTCGA': 12, 'GGCCCT': 12, 'CTTCGA': 12, 'CCCAAT': 12, 'AATCGG': 12, 'TTGCTA': 12, 'CATCCT': 12, 'GCTACT': 12, 'ACATTG': 12, 'ATATTA': 12, 'CTTAAG': 12, 'AAGTGT': 12, 'GGTACT': 12, 'TCGACT': 12, 'AGAAGT': 12, 'CTTAGT': 12, 'ACTCTG': 12, 'CCTAAA': 12, 'TCCATG': 12, 'CGCGTA': 12, 'CTCATT': 12, 'CAAGTG': 12, 'ACTCAA': 12, 'TTCTCA': 12, 'CATATG': 12, 'CTCTCA': 12, 'ACAAGA': 12, 'CGACTT': 12, 'GAGACT': 12, 'TTCCGA': 12, 'CCTAAT': 11, 'AGTTCT': 11, 'TATCCT': 11, 'GATCCA': 11, 'TCGAGA': 11, 'CCTGGG': 11, 'CAGTTG': 11, 'CACGTT': 11, 'AGTGGG': 11, 'ATTATA': 11, 'TGGTCT': 11, 'GTACTC': 11, 'AGATCT': 11, 'GTACTG': 11, 'TTGACG': 11, 'GGGCCA': 11, 'ATCAGA': 11, 'CGGTGT': 11, 'ATAGTC': 11, 'ATACTC': 11, 'TACCTT': 11, 'GGGTGT': 11, 'TATGTA': 11, 'CCTATT': 11, 'GGTCTA': 11, 'CTATGC': 11, 'GTCTCT': 11, 'TTAGGG': 11, 'TTATCA': 11, 'CGAGTG': 11, 'CGTAAG': 11, 'ACGACT': 11, 'ACTAGC': 11, 'CCAGGG': 11, 'ACTTCG': 11, 'TGGTCA': 11, 'CCGTAT': 11, 'GGTACG': 11, 'TGCCTC': 11, 'TGTCCT': 11, 'TATTCG': 11, 'CCTAAC': 11, 'GACCCT': 11, 'AGTCCG': 11, 'TTTTGG': 11, 'GATAGG': 11, 'CGATCT': 11, 'GGTATA': 11, 'ATACTT': 11, 'GTGACA': 11, 'AGTGCA': 11, 'TCTCTA': 11, 'CCACAA': 11, 'ACGTTG': 11, 'AGGCTA': 11, 'ATACCC': 11, 'ACCTGT': 11, 'GCACTA': 11, 'CTGAGA': 11, 'GTACGA': 11, 'TTTGTA': 11, 'GCGTTA': 11, 'AGTTCA': 11, 'ACTAAG': 11, 'TGTGTC': 11, 'TCAATG': 11, 'GTATCG': 11, 'CCCACG': 11, 'TTAATG': 11, 'ATGACA': 11, 'ATAAGG': 11, 'ATGTCA': 11, 'TGTCAT': 11, 'GGACCT': 11, 'TGTTTT': 11, 'ATTCTA': 11, 'ATCTTT': 11, 'GTTACC': 11, 'AGGTCT': 11, 'CTAGTC': 11, 'CAGATA': 11, 'CACATT': 11, 'ATAACC': 11, 'GTATTG': 11, 'ATAATT': 11, 'AGCTCA': 11, 'GTCACA': 10, 'TCTCTG': 10, 'CACCAA': 10, 'GTCTCA': 10, 'CTGTTA': 10, 'GTAAGT': 10, 'ATAATG': 10, 'GTAAGG': 10, 'CTACTG': 10, 'CCTTGT': 10, 'CTTACT': 10, 'ATCATG': 10, 'TACGGA': 10, 'TGTAGC': 10, 'ATAGTG': 10, 'TTCGTA': 10, 'TACGTT': 10, 'ACAAGG': 10, 'TGCCGT': 10, 'TCACCT': 10, 'GTAGGA': 10, 'AGTCCA': 10, 'ACTGTG': 10, 'AACACA': 10, 'AGGTTA': 10, 'CCAGGT': 10, 'ACAACT': 10, 'CTAGCT': 10, 'CGAGAA': 10, 'ACACTG': 10, 'ACATTA': 10, 'TTCTTA': 10, 'ATAGTA': 10, 'TCTATA': 10, 'CCGAGA': 10, 'ATCATA': 10, 'GTGGTA': 10, 'CTAGAT': 10, 'CCTTTG': 10, 'ATTTGG': 10, 'TTTCCT': 10, 'CTACAA': 10, 'CTAGGT': 10, 'TACCCG': 10, 'ATCACG': 10, 'GTACGT': 10, 'CACTCT': 10, 'GTTATA': 10, 'TTACAC': 10, 'ATGTTA': 10, 'CGACAA': 10, 'TGGTTA': 10, 'AGACGG': 10, 'GATACG': 10, 'TTACTG': 10, 'ATACAG': 10, 'TTATGG': 10, 'ATACGT': 10, 'TTGTCG': 10, 'TGGAGA': 10, 'ACGAGA': 10, 'ATGCTA': 10, 'CCATCG': 10, 'CCCCTT': 10, 'TTAGTC': 10, 'AATTGT': 10, 'AGCCTA': 10, 'CTACCG': 10, 'CCCAGA': 10, 'CACACG': 10, 'TGTTCG': 10, 'TCATGT': 10, 'CATACC': 10, 'GGACTA': 10, 'GTATGC': 10, 'TCTATT': 10, 'CATCTT': 10, 'TCTGGA': 10, 'ACAGTG': 10, 'TTCGTT': 10, 'CCACCT': 10, 'TGCGTT': 10, 'TTAGCG': 10, 'GTAAGA': 10, 'GCTTGT': 10, 'CAGTGT': 10, 'ATTCCA': 10, 'ACACTA': 10, 'TACAGT': 10, 'CTTGTG': 10, 'CGAGTA': 10, 'TCCTAT': 10, 'CGTAGT': 10, 'AGTATG': 10, 'GTAATT': 10, 'AACTGT': 10, 'TGTCTC': 10, 'ATTTGC': 10, 'GTACAT': 10, 'AGACCT': 9, 'CAAATA': 9, 'CCAATG': 9, 'TGTGAA': 9, 'GTATTA': 9, 'CCATCT': 9, 'CGTTTA': 9, 'ACGTTA': 9, 'GTATGG': 9, 'GTCTGT': 9, 'CCCCCT': 9, 'CAACTA': 9, 'CTCATG': 9, 'TGCCCA': 9, 'AGTTTA': 9, 'ATAATC': 9, 'CTCACG': 9, 'TGTGGC': 9, 'CTATTC': 9, 'TTTAGG': 9, 'TCTAGC': 9, 'CGCCTA': 9, 'TGTATT': 9, 'GACTTA': 9, 'CCTAAG': 9, 'TGCCAA': 9, 'GCAATG': 9, 'AGACTT': 9, 'ATTAGG': 9, 'CGAGTT': 9, 'CATAAG': 9, 'AGAACA': 9, 'ACATAC': 9, 'TTCCCA': 9, 'TGTCCG': 9, 'TTCCTT': 9, 'TACCAT': 9, 'GGACCA': 9, 'TACAAA': 9, 'TCCCTA': 9, 'CTAATC': 9, 'AACTTA': 9, 'CCCATG': 9, 'AACCTA': 9, 'TGCTGG': 9, 'TTCTTG': 9, 'TCCGTA': 9, 'ACTTGC': 9, 'GTCGTA': 9, 'CGTACT': 9, 'TTTAGA': 9, 'CTACGG': 9, 'CATCTA': 9, 'ATATCA': 9, 'AACTTG': 9, 'GTAGTT': 9, 'AGCACG': 9, 'TCTGGG': 9, 'TTAGGA': 9, 'TCCCTT': 9, 'AGTAGA': 9, 'TGCTCG': 9, 'CTATAC': 9, 'GCAATA': 9, 'AGCATA': 9, 'TCTTGG': 9, 'CTAGAG': 9, 'TTCTCT': 9, 'ATAGTT': 9, 'AATACG': 9, 'GTATGT': 9, 'GGTAGG': 9, 'TCTAGT': 9, 'TTAAGG': 9, 'TACAGA': 9, 'TGCGCT': 9, 'ACAAGT': 9, 'TCAACT': 9, 'TCGGTA': 9, 'CTAGAA': 9, 'AACATA': 9, 'ACAGGA': 9, 'TTACGG': 8, 'ATCTCG': 8, 'TGCCTG': 8, 'ACAATA': 8, 'TTGCAA': 8, 'GTGTTA': 8, 'TGTAGT': 8, 'TGTGTT': 8, 'ACAATG': 8, 'GCGATA': 8, 'TTAGTG': 8, 'CACATA': 8, 'GCTTTA': 8, 'ATATAC': 8, 'TACATG': 8, 'CCTTCT': 8, 'TGTATC': 8, 'AGATTG': 8, 'TTACGT': 8, 'CAATGT': 8, 'GCTACG': 8, 'TCTCCA': 8, 'TACTCT': 8, 'ACTGGA': 8, 'CCCGTA': 8, 'GGTAGA': 8, 'CGTACG': 8, 'TGTTTC': 8, 'CAAACA': 8, 'CATTGT': 8, 'TTCATA': 8, 'CATGTA': 8, 'TATCCA': 8, 'CCTCCG': 8, 'TGTCTG': 8, 'CGCTGT': 8, 'CGGCTA': 8, 'ATATCG': 8, 'TTGACA': 8, 'TCTACC': 8, 'ACGATA': 8, 'TGCAGT': 8, 'AGGGTA': 8, 'CTATCA': 8, 'GGCCTA': 8, 'GATACT': 8, 'CACTTG': 8, 'TCAACG': 8, 'AGTATA': 8, 'TGCAGG': 8, 'GTTCTA': 8, 'CTTGTA': 8, 'CTATTG': 8, 'TGCGGA': 8, 'ATACGA': 8, 'AGGACT': 8, 'CGAATA': 8, 'CGAAGT': 8, 'CCAAGT': 8, 'CAAACG': 8, 'TACATT': 8, 'TTAACG': 8, 'TCGATA': 8, 'CTAGCG': 8, 'ATATCT': 8, 'TCTTGC': 8, 'GACTGT': 8, 'CGTGTG': 8, 'CTTCTG': 8, 'TGTCGA': 8, 'TCTAAC': 8, 'GCATTA': 8, 'ATCCTA': 8, 'GTTAGA': 8, 'CTTCCA': 8, 'ACTGGG': 8, 'ATTACG': 8, 'AGGTGC': 7, 'GTACAG': 7, 'TGTCAA': 7, 'CGAAGA': 7, 'GATCCT': 7, 'TGGGTA': 7, 'TGTTTA': 7, 'TTACTA': 7, 'TACACA': 7, 'TTTCCA': 7, 'AATACA': 7, 'GTCCCT': 7, 'CACAGA': 7, 'GTCCCA': 7, 'ATAAGA': 7, 'ATGTGT': 7, 'TCAATA': 7, 'CCAACA': 7, 'TGCTCA': 7, 'TGTAAG': 7, 'CTTACA': 7, 'TATAGA': 7, 'GTCACG': 7, 'TCATCT': 7, 'TGTGGT': 7, 'ACATAT': 7, 'ATAGCA': 7, 'ACAGTA': 7, 'GTACAA': 7, 'AGATCG': 7, 'TGTAAC': 7, 'CAAATG': 7, 'AGCACA': 7, 'GTACCT': 7, 'TGCACA': 7, 'AGTTGT': 7, 'ACTAGA': 7, 'CCCATT': 7, 'CGATTG': 7, 'CGGTTA': 7, 'GTTCCA': 7, 'CCAGTG': 7, 'CCTCCT': 7, 'CACTTT': 7, 'ACCATA': 7, 'TCAAGG': 7, 'CTATGG': 7, 'TCGTGT': 7, 'TTAGTA': 7, 'AGCTGT': 7, 'CCGTGT': 7, 'CCTCAT': 7, 'AGCCCA': 7, 'ATTGTA': 7, 'GTAATG': 7, 'CGAATG': 7, 'CTACGA': 7, 'ACAACG': 7, 'CTACTC': 7, 'GTACAC': 7, 'AGAGTA': 7, 'CATACT': 7, 'CACACT': 7, 'TATAGT': 7, 'CCACCG': 7, 'TGTCGT': 7, 'TCACTA': 7, 'TCGTTA': 7, 'CCTACT': 7, 'TCAGTG': 7, 'CTTATG': 7, 'CCTACC': 7, 'TCCTGT': 7, 'TTTCCG': 7, 'ATGACT': 7, 'CTCCCA': 7, 'CGTACA': 7, 'GACCTA': 7, 'GGTGTA': 7, 'TCTACT': 7, 'TGTTGT': 7, 'CGTATG': 7, 'CCATTG': 7, 'CGATAT': 7, 'ACAACA': 7, 'GTTACG': 7, 'TATACT': 6, 'GCCATA': 6, 'CTAACC': 6, 'TTTACT': 6, 'CTCGTA': 6, 'TCATAC': 6, 'TCCACA': 6, 'CCAGTA': 6, 'AGTGTA': 6, 'CACAGG': 6, 'CTAACT': 6, 'TTAACA': 6, 'TGTTGG': 6, 'TCATTG': 6, 'AACACG': 6, 'CTAATT': 6, 'CGCATA': 6, 'AGTCCT': 6, 'CCCACA': 6, 'CTAGTT': 6, 'GTGCTA': 6, 'TTTCTA': 6, 'TATATA': 6, 'CTTCGT': 6, 'CCTAGA': 6, 'GCTATG': 6, 'ATACTA': 6, 'TTAGCA': 6, 'TTATGT': 6, 'TCGTAT': 6, 'ACCCTA': 6, 'CCATTA': 6, 'CCCTGT': 6, 'TGGCCT': 6, 'GTAATA': 6, 'GGATGT': 6, 'CTACCA': 6, 'TTACCG': 6, 'ACTTGG': 6, 'TTTACA': 6, 'TGCTTA': 6, 'ACGTGT': 6, 'CTATAT': 6, 'CTCTTG': 6, 'TTATGC': 6, 'TATCTA': 6, 'CTATCG': 6, 'ACTATG': 6, 'CTAAGG': 6, 'CTATGT': 6, 'ACTACA': 6, 'CCTTTA': 6, 'CGTCTA': 6, 'GTATCA': 6, 'TGCCTT': 6, 'TGTCTA': 6, 'TGTGGA': 6, 'CTCAGA': 6, 'TGTAGA': 6, 'TCAACA': 6, 'TGTGCG': 6, 'TATAGG': 6, 'CATTTA': 6, 'CATAGG': 6, 'CTAATA': 6, 'AGGTGT': 6, 'ACCTTA': 6, 'GTGACT': 6, 'AGGATA': 6, 'CTACCT': 6, 'CCACAT': 6, 'AGTACT': 5, 'GGCTTA': 5, 'CCAATT': 5, 'CCTAGT': 5, 'CGATTA': 5, 'ACTCTA': 5, 'TACCCT': 5, 'ATTACA': 5, 'GTAGCA': 5, 'CGTTGT': 5, 'TGTAAA': 5, 'ACTTGT': 5, 'ATCACA': 5, 'CGCTTA': 5, 'TCTACG': 5, 'GGTCCA': 5, 'TGCTAT': 5, 'GTACCA': 5, 'TTACCA': 5, 'TGTTTG': 5, 'ATAACT': 5, 'TCAGTA': 5, 'CTAAGA': 5, 'TGTATG': 5, 'CCATGT': 5, 'CCTAGC': 5, 'CCACCA': 5, 'CCAACG': 5, 'CCTCTG': 5, 'AGTCTA': 5, 'ATATGT': 5, 'TACTGT': 5, 'TACCCA': 5, 'TGTTAC': 5, 'CTTAGA': 5, 'TATACA': 5, 'AGACCA': 5, 'CCGATA': 5, 'AGTACA': 5, 'TTGTGT': 5, 'CGTGTA': 5, 'TCGTCT': 5, 'CATAGA': 5, 'CCAAGA': 5, 'ATACAC': 5, 'TGCACG': 5, 'TTACGA': 5, 'GTAACT': 5, 'CTCCTT': 5, 'TCTGTA': 5, 'GCCTTA': 5, 'GTTACT': 5, 'ACTATA': 5, 'CGTATA': 5, 'CTAATG': 5, 'TGTTCA': 5, 'TGGACT': 5, 'CGATGT': 5, 'GTTAGG': 5, 'TGCGCA': 5, 'CATATA': 5, 'TCTTGT': 5, 'TTCAGA': 5, 'TGCGTG': 5, 'CTCCTA': 5, 'CCACTA': 5, 'AGATGC': 5, 'GCTCTA': 4, 'ACTACT': 4, 'TTCCTA': 4, 'TCAAGA': 4, 'TACCTA': 4, 'ATACTG': 4, 'TGCTTG': 4, 'GCTACA': 4, 'ACATGT': 4, 'TGTCTT': 4, 'CACCTA': 4, 'TACTTG': 4, 'GACCCA': 4, 'CTACTT': 4, 'TTCACA': 4, 'TGCACT': 4, 'GTCCTA': 4, 'TACACT': 4, 'TGCATT': 4, 'TATGGG': 4, 'CGTAGA': 4, 'TACACG': 4, 'ACTAGT': 4, 'GTCACT': 4, 'TGTTCT': 4, 'CACTTA': 4, 'GTGTGT': 4, 'CTAGGA': 4, 'CCCCCA': 4, 'GTAGTA': 4, 'CCAATA': 4, 'TCTACA': 4, 'TATTGT': 4, 'CCCCTA': 4, 'GTATAC': 4, 'GCTAGA': 4, 'GCATGT': 4, 'AGACAA': 4, 'CCTATA': 4, 'ATACCG': 4, 'GTTTGT': 4, 'CTGTGT': 4, 'GCTAGG': 4, 'TCTATG': 4, 'TCCTTA': 4, 'AGTACG': 4, 'ATAACA': 4, 'CACTGT': 4, 'ATCTTA': 4, 'CTACAT': 4, 'TCCATA': 4, 'ATACCA': 4, 'TGCTCT': 4, 'TGTACT': 4, 'TTTACG': 4, 'AGCGTA': 4, 'GTAGTG': 4, 'CTACGT': 4, 'GTACTA': 4, 'CTTCTA': 4, 'CGTAGG': 3, 'CCAACT': 3, 'ACTACG': 3, 'AGATGT': 3, 'CTACAG': 3, 'TGCATA': 3, 'TACTCG': 3, 'CTAGTG': 3, 'TGTGTA': 3, 'TCTAGA': 3, 'ATAACG': 3, 'CTATTA': 3, 'GCTATA': 3, 'CCTCCA': 3, 'CTTAGG': 3, 'GTAACA': 3, 'CCGTTA': 3, 'TGTACA': 3, 'AGACGT': 3, 'TGTCGG': 3, 'CCCTTA': 3, 'GATACA': 3, 'CCTATG': 3, 'TGTGGG': 3, 'TGTGTG': 3, 'CCTCTA': 3, 'GTTATG': 3, 'CGACTA': 3, 'TTCCCT': 3, 'TTCTGT': 2, 'CATACA': 2, 'TGGTGT': 2, 'GTCTTA': 2, 'CCTACG': 2, 'GTCATA': 2, 'TGTAGG': 2, 'AGACTA': 2, 'GTTACA': 2, 'TCTGGT': 2, 'CTTACG': 2, 'GGTACA': 2, 'TGCATG': 2, 'TACTCA': 2, 'CTCTGT': 2, 'TGTATA': 2, 'CTAACG': 2, 'TGTACC': 2, 'CTAGCA': 2, 'CACGTA': 2, 'TACGTA': 2, 'TGCGTA': 2, 'GACATA': 2, 'TGTACG': 2, 'CCTACA': 2, 'TGCAGA': 2, 'CTCTTA': 2, 'TTATAC': 2, 'ATACAT': 2, 'CATACG': 1, 'TATACG': 1, 'GAGGMT': 1, 'CTACTA': 1, 'TCTAGG': 1, 'GTAACG': 1, 'CCTAGG': 1, 'CCCATA': 1, 'TACATA': 1, 'CTCATA': 1, 'TGTGCA': 1, 'TGCCTA': 1, 'TACTTA': 1, 'ACTAGG': 1})\n",
      "\n",
      "#### Compute statistics for stat  all , end\n",
      "\n",
      "#### Analysis for stat: all , start\n",
      "\n",
      "outputFilePrefix: TGA_non_overlapping_start_codon_subsample_subset_all\n",
      "\n",
      "#### Termini amino acid composition analysis\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#### Writing oddsRatioAADf files, start\n",
      "\n",
      "#### Writing oddsRatioAADf files, end\n",
      "\n",
      "#### Termini amino acid composition analysis end\n",
      "\n",
      "#### Termini codons composition analysis, start\n",
      "\n",
      "#### Writing oddsRatioDfCodon files, start\n",
      "\n",
      "#### Writing oddsRatioDfCodon files, end\n",
      "\n",
      "#### Termini codons composition analysis, end\n",
      "\n",
      "#### AA pairs bias analysis start\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/IPython/core/async_helpers.py:68: PerformanceWarning: indexing past lexsort depth may impact performance.\n",
      "  coro.send(None)\n",
      "/users/lserrano/mweber/Research_cloud/C-terminal/src/cterminal.py:1640: RuntimeWarning: divide by zero encountered in log2\n",
      "  log2OddsRatio = log2(oddsRatio)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#### Writing oddsRatioDf AA pair files, start\n",
      "\n",
      "#### Writing oddsRatioDf AA pair files, end\n",
      "\n",
      "#### AA pairs bias analysis end\n",
      "\n",
      "#### hexamer bias analysis start\n",
      "\n",
      "#### Writing oddsRatioDf hexamer files, start\n",
      "\n",
      "#### Writing oddsRatioDf hexamer files, end\n",
      "\n",
      "#### hexamer bias analysis end\n",
      "\n",
      "#### Analysis for stat: all , end\n",
      "\n",
      "#### Drawing plots, start\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/font_manager.py:1241: UserWarning: findfont: Font family ['cursive'] not found. Falling back to DejaVu Sans.\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n",
      "/users/lserrano/mweber/Research_cloud/Python_mwTools/mwTools/bio.py:847: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  plotDf['sort_index'] = plotDf.index.get_level_values(0).map(lambda x: sort_order.get(x))\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/users/lserrano/mweber/Research_cloud/Python_mwTools/mwTools/bio.py:847: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  plotDf['sort_index'] = plotDf.index.get_level_values(0).map(lambda x: sort_order.get(x))\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus N\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus N\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus C\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus C\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#### Drawing plots, end\n",
      "\n",
      "#### Statistics all total nb sequences:  175061\n",
      "\n",
      "multispeciesStatisticsSummary: Empty DataFrame\n",
      "Columns: [genome_accession, species_name, nSeq]\n",
      "Index: []\n",
      "\n",
      "#### Zero statistics for stat: stopCodon_TAA , skipping analysis\n",
      "\n",
      "multispeciesStatisticsSummary: Empty DataFrame\n",
      "Columns: [genome_accession, species_name, nSeq]\n",
      "Index: []\n",
      "\n",
      "#### Zero statistics for stat: stopCodon_TAG , skipping analysis\n",
      "\n",
      "multispeciesStatisticsSummary:      genome_accession                                       species_name  nSeq\n",
      "0     GCF_000005825.2                          Bacillus pseudofirmus OF4    42\n",
      "1     GCF_000005845.2          Escherichia coli str. K-12 substr. MG1655    89\n",
      "2     GCF_000006605.1                      Corynebacterium jeikeium K411    19\n",
      "3     GCF_000006685.1                      Chlamydia muridarum str. Nigg    10\n",
      "4     GCF_000006725.1                            Xylella fastidiosa 9a5c   122\n",
      "5     GCF_000006745.1       Vibrio cholerae O1 biovar El Tor str. N16961    37\n",
      "6     GCF_000006765.1                        Pseudomonas aeruginosa PAO1   367\n",
      "7     GCF_000006785.2                      Streptococcus pyogenes M1 GAS    14\n",
      "8     GCF_000006845.1                      Neisseria gonorrhoeae FA 1090    56\n",
      "9     GCF_000006865.1            Lactococcus lactis subsp. lactis Il1403    27\n",
      "10    GCF_000006905.1                        Caulobacter crescentus CB15   139\n",
      "11    GCF_000006925.2                      Shigella flexneri 2a str. 301    73\n",
      "12    GCF_000006945.1  Salmonella enterica subsp. enterica serovar Ty...    99\n",
      "13    GCF_000006965.1                        Sinorhizobium meliloti 1021   199\n",
      "14    GCF_000006985.1                             Chlorobium tepidum TLS   111\n",
      "15    GCF_000007025.1                   Rickettsia conorii str. Malish 7    13\n",
      "16    GCF_000007045.1                        Streptococcus pneumoniae R6    16\n",
      "17    GCF_000007085.1  Caldanaerobacter subterraneus subsp. tengconge...    63\n",
      "18    GCF_000007125.1                 Brucella melitensis bv. 1 str. 16M   104\n",
      "19    GCF_000007145.1  Xanthomonas campestris pv. campestris str. ATC...   225\n",
      "20    GCF_000007265.1                   Streptococcus agalactiae 2603V/R    23\n",
      "21    GCF_000007325.1  Fusobacterium nucleatum subsp. nucleatum ATCC ...     8\n",
      "22    GCF_000007365.1  Buchnera aphidicola str. Sg (Schizaphis graminum)     8\n",
      "23    GCF_000007385.1           Xanthomonas oryzae pv. oryzae KACC 10331   190\n",
      "24    GCF_000007465.2                         Streptococcus mutans UA159    29\n",
      "25    GCF_000007485.1                     Tropheryma whipplei str. Twist    26\n",
      "26    GCF_000007525.1                     Bifidobacterium longum NCC2705    83\n",
      "27    GCF_000007565.2                          Pseudomonas putida KT2440   335\n",
      "28    GCF_000007605.1                          Chlamydophila caviae GPIC     2\n",
      "29    GCF_000007625.1                             Clostridium tetani E88    17\n",
      "...               ...                                                ...   ...\n",
      "1498  GCF_001583415.1                                Mycobacterium phlei   270\n",
      "1499  GCF_001584185.1                               Collimonas pratensis   176\n",
      "1500  GCF_001586165.1                            Obesumbacterium proteus    80\n",
      "1501  GCF_001586195.1                            Solibacillus silvestris    24\n",
      "1502  GCF_001586215.1                           Corynebacterium simulans    15\n",
      "1503  GCF_001586255.1                       Zhongshania aliphaticivorans    69\n",
      "1504  GCF_001593245.1                                  Aeromonas veronii   187\n",
      "1505  GCF_001597285.1                            Pseudomonas alcaligenes   221\n",
      "1506  GCF_001606005.1                              Brevibacterium linens   233\n",
      "1507  GCF_001606025.1                         Psychrobacter alimentarius    36\n",
      "1508  GCF_001610975.1                     Sphingopyxis terrae NBRC 15098   194\n",
      "1509  GCF_001611135.1                               Pediococcus damnosus    16\n",
      "1510  GCF_001618685.1                           Geobacillus subterraneus    76\n",
      "1511  GCF_001620305.1                          Rhodococcus fascians D188   271\n",
      "1512  GCF_001632865.1                           Streptomyces ambofaciens   471\n",
      "1513  GCF_001636015.1                                     Moraxella ovis    38\n",
      "1514  GCF_001641285.1                                 Dyella thiooxydans   252\n",
      "1515  GCF_001642655.1                          Clostridium tyrobutyricum    42\n",
      "1516  GCF_001644565.1                               Deinococcus puniceus    95\n",
      "1517  GCF_001648175.1                              Halotalea alkalilenta   261\n",
      "1518  GCF_001652465.1                          Microbacterium chocolatum   192\n",
      "1519  GCF_001652485.1                     Ochrobactrum pseudogrignonense   118\n",
      "1520  GCF_001652565.1                                  Shinella sp. HZN7   286\n",
      "1521  GCF_001653755.1                         Dokdonia donghaensis DSW-1     6\n",
      "1522  GCF_001654495.1                              Streptomyces sp. SAT1   417\n",
      "1523  GCF_001655245.1                         Mycobacterium vaccae 95051   333\n",
      "1524  GCF_900011245.1                                 Bradyrhizobium sp.   345\n",
      "1525  GCF_900078695.1                                Bordetella trematum   237\n",
      "1526  GCF_900086555.1                                  Neisseria weaveri    38\n",
      "1527  GCF_900087055.1                           Plesiomonas shigelloides    68\n",
      "\n",
      "[1528 rows x 3 columns]\n",
      "\n",
      "#### Compute statistics for stat  stopCodon_TGA , start\n",
      "Total nb of aa:  49997365\n",
      "\n",
      "#### Amino acid frequency bulk:\n",
      "\n",
      "Counter({'A': 5854079, 'L': 5295411, 'G': 4265977, 'V': 3887022, 'R': 3415139, 'E': 2943709, 'D': 2862091, 'S': 2745745, 'T': 2701127, 'P': 2589886, 'I': 2434117, 'F': 1783915, 'Q': 1740147, 'K': 1602123, 'N': 1378609, 'Y': 1262569, 'H': 1101488, 'M': 976293, 'W': 706293, 'C': 451608, 'X': 14, 'U': 3}) \n",
      "\n",
      "[('A', 0.11708775052445264), ('L', 0.10591380165734735), ('G', 0.08532403657672759), ('V', 0.07774453713710712), ('R', 0.06830637974621262), ('E', 0.058877282832805286), ('D', 0.05724483680289951), ('S', 0.05491779416775264), ('T', 0.05402538713790217), ('P', 0.05180044988370887), ('I', 0.0486849056945301), ('F', 0.03568018034550421), ('Q', 0.03480477421160095), ('K', 0.0320441487266379), ('N', 0.027573633130465975), ('Y', 0.0252527108178601), ('H', 0.022030921029538256), ('M', 0.019526889067053833), ('W', 0.014126604472055677), ('C', 0.00903263601991825), ('X', 2.800147567776822e-07), ('U', 6.000316216664618e-08)]\n",
      "ctermFreq.shape:  (20, 175061)\n",
      "ntermFreq.shape:  (20, 175061)\n",
      "Total nb of aa at pos 0:  175061\n",
      "Total nb of aa at pos 0:  175061\n",
      "\n",
      "#### Amino acid frequency c-terminal:\n",
      " Counter({'A': 25403, 'R': 19952, 'G': 19487, 'S': 14875, 'D': 11264, 'L': 9843, 'P': 9418, 'V': 9222, 'E': 7510, 'F': 6373, 'N': 6307, 'H': 5932, 'K': 5858, 'I': 5599, 'T': 5428, 'Y': 3778, 'Q': 3685, 'C': 2110, 'W': 1770, 'M': 1246, 'X': 1})\n",
      "\n",
      "#### Amino acid frequency n-terminal:\n",
      " Counter({'A': 15392, 'T': 15166, 'R': 14677, 'L': 13964, 'S': 12335, 'K': 11911, 'D': 11780, 'E': 11328, 'I': 10517, 'V': 9996, 'G': 8021, 'P': 7791, 'Q': 7043, 'N': 7035, 'F': 5507, 'H': 4187, 'Y': 3542, 'M': 2247, 'W': 1706, 'C': 916})\n",
      "Total nb of codons:  50172426\n",
      "{'GTR', 'TAC', 'CAT', 'GCT', 'SCC', 'SGC', 'CGC', 'GTA', 'AGA', 'CTC', 'AAC', 'TGC', 'TCM', 'CMC', 'GTC', 'TTC', 'TGT', 'GTW', 'NNA', 'GGC', 'GGG', 'CGA', 'GAC', 'GAT', 'CAG', 'GTG', 'ATT', 'TTM', 'TCT', 'TCC', 'WRA', 'TGA', 'CTG', 'ACT', 'ACG', 'GCG', 'ACC', 'ARA', 'ATA', 'CAR', 'GCR', 'TGG', 'AGY', 'CGT', 'GGS', 'CGG', 'GMC', 'CCC', 'ACM', 'TTG', 'GCM', 'TTA', 'CGS', 'CCG', 'CCA', 'CCM', 'AAA', 'TTT', 'ATG', 'CAA', 'GGT', 'GNN', 'GAA', 'CTT', 'TNC', 'AAG', 'ACA', 'TKG', 'AAR', 'CAC', 'AGC', 'CTA', 'SCG', 'GGA', 'AGT', 'GAG', 'AAT', 'TCG', 'AGG', 'TAT', 'GCA', 'GTT', 'CCT', 'ATC', 'RCC', 'GCC', 'TCA'}\n",
      "ctermFreq.shape:  (21, 175061)\n",
      "ntermFreq.shape:  (21, 175061)\n",
      "Total nb of codons at pos 0:  175061\n",
      "Total nb of codons at pos 0:  175061\n",
      "Bulk codon frequency:\n",
      " Counter({'GCC': 2733811, 'CTG': 2668481, 'GGC': 2425045, 'GCG': 2137962, 'GAC': 1914509, 'GAG': 1808581, 'GTG': 1703975, 'ATC': 1671044, 'CGC': 1647485, 'GTC': 1539765, 'ACC': 1443888, 'CAG': 1345521, 'CTC': 1337680, 'CCG': 1331078, 'TTC': 1272142, 'GAA': 1146410, 'AAG': 1023014, 'ATG': 979639, 'GAT': 957811, 'CGG': 916395, 'AAC': 902706, 'CCC': 831267, 'ACG': 829420, 'TAC': 775025, 'AGC': 774660, 'GGG': 771525, 'TCG': 757826, 'CAC': 713295, 'TGG': 708752, 'TCC': 643227, 'GGT': 639971, 'AAA': 585425, 'ATT': 567781, 'GCA': 537959, 'TTG': 532385, 'TTT': 517720, 'TAT': 491795, 'AAT': 480405, 'GCT': 464822, 'GGA': 442941, 'CGT': 405814, 'CTT': 401907, 'CAA': 400966, 'GTT': 394749, 'CAT': 392018, 'TGC': 343568, 'GTA': 261881, 'TTA': 246692, 'CCT': 237536, 'ACA': 231097, 'AGT': 218834, 'ACT': 205837, 'ATA': 203761, 'CCA': 199212, 'TCT': 182422, 'TCA': 178076, 'CGA': 176648, 'AGG': 158911, 'CTA': 126606, 'AGA': 122911, 'TGT': 109808, 'TGA': 3, 'SCC': 2, 'AAR': 1, 'ACM': 1, 'GCR': 1, 'TCM': 1, 'TTM': 1, 'TKG': 1, 'AGY': 1, 'CMC': 1, 'CGS': 1, 'WRA': 1, 'GGS': 1, 'RCC': 1, 'GMC': 1, 'SCG': 1, 'CCM': 1, 'ARA': 1, 'GTR': 1, 'CAR': 1, 'GTW': 1, 'GCM': 1, 'NNA': 1, 'GNN': 1, 'SGC': 1, 'TNC': 1, 'TAA': 0, 'TAG': 0})\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bulk amino acid pairs frequency:\n",
      " Counter({'AA': 791698, 'LA': 673865, 'AL': 651728, 'LL': 550518, 'AG': 524506, 'VA': 456480, 'AV': 456130, 'GA': 445941, 'GL': 442145, 'LG': 436192, 'VL': 418114, 'AR': 407551, 'LV': 399071, 'RL': 390077, 'RA': 371567, 'LR': 365475, 'GG': 364942, 'AE': 362232, 'EA': 343072, 'GV': 336161, 'AD': 335885, 'VV': 335545, 'DA': 332337, 'EL': 325512, 'LS': 322324, 'LD': 317975, 'PA': 315669, 'TA': 311936, 'IA': 301548, 'SA': 301287, 'TL': 301103, 'DL': 300912, 'LT': 298973, 'AS': 297475, 'GR': 293801, 'LP': 293727, 'VG': 293655, 'AT': 292504, 'SL': 290466, 'LE': 288873, 'SG': 277519, 'AP': 268456, 'RR': 268427, 'AI': 263280, 'DG': 262652, 'TG': 262338, 'VR': 262074, 'PL': 254311, 'RV': 250873, 'GE': 246099, 'GT': 243937, 'ER': 243179, 'GS': 242547, 'PG': 242380, 'LI': 240646, 'RG': 238640, 'TV': 234816, 'VT': 230795, 'VE': 230161, 'GD': 228026, 'VD': 227370, 'VS': 224907, 'IG': 222830, 'GI': 222737, 'PV': 220641, 'EV': 219619, 'DV': 218709, 'RE': 215716, 'QA': 213955, 'IL': 209593, 'SV': 209287, 'EG': 208274, 'AQ': 206294, 'VP': 204168, 'FA': 198917, 'DR': 197703, 'VI': 192002, 'RD': 191824, 'AF': 191508, 'PE': 189872, 'QL': 188322, 'IV': 186875, 'DE': 185078, 'PD': 180585, 'LF': 179981, 'RI': 179130, 'EE': 178053, 'GP': 177066, 'DP': 176129, 'RT': 175835, 'RS': 175496, 'SR': 175121, 'LK': 174533, 'KA': 172075, 'RP': 169727, 'FG': 169271, 'GF': 168050, 'LQ': 167658, 'TP': 167487, 'FL': 167386, 'AK': 167331, 'ID': 165705, 'SS': 163499, 'EI': 163181, 'DD': 162588, 'KL': 162462, 'IE': 161938, 'TR': 161786, 'GK': 158092, 'TT': 152523, 'ET': 152448, 'TD': 147940, 'IS': 147230, 'ST': 147052, 'SD': 146977, 'TS': 146534, 'GQ': 145940, 'SE': 144811, 'PR': 144768, 'IR': 144415, 'NL': 144336, 'NA': 144332, 'LN': 143675, 'SP': 143013, 'ED': 141321, 'DI': 141146, 'ES': 140271, 'TE': 140101, 'IT': 139158, 'DT': 138895, 'PS': 138882, 'VF': 138388, 'QV': 138081, 'EP': 137806, 'YL': 136483, 'FV': 135120, 'SI': 134777, 'QR': 133958, 'EQ': 133228, 'AN': 132957, 'RF': 132191, 'PP': 132151, 'QG': 131606, 'RQ': 131149, 'DS': 130741, 'PT': 130425, 'YA': 130109, 'TI': 129852, 'AY': 128757, 'GY': 125334, 'AM': 123241, 'FD': 122695, 'KV': 122476, 'IP': 121396, 'NG': 121281, 'KG': 119829, 'EK': 119062, 'HA': 118297, 'HL': 117998, 'AH': 117741, 'VQ': 116905, 'LY': 116309, 'MA': 116000, 'VK': 115862, 'FS': 115705, 'GN': 115067, 'II': 114260, 'YG': 112132, 'LH': 112075, 'ML': 110451, 'SF': 109656, 'VN': 108989, 'DF': 105408, 'FT': 104898, 'PI': 102474, 'HG': 102347, 'FE': 101994, 'RK': 100861, 'LM': 100326, 'KR': 100107, 'FR': 100085, 'KE': 99994, 'NV': 99536, 'GH': 98532, 'RY': 96845, 'YR': 96618, 'GM': 95198, 'TF': 95104, 'EF': 93619, 'KT': 92459, 'PF': 91889, 'KI': 91600, 'QI': 91187, 'QP': 90813, 'QE': 90787, 'NP': 90685, 'SQ': 90208, 'WL': 89740, 'QS': 89686, 'KS': 89212, 'PQ': 89190, 'KD': 88981, 'YV': 88645, 'RN': 88288, 'QT': 87734, 'DK': 87056, 'DQ': 87030, 'EN': 86741, 'NR': 86509, 'VY': 86288, 'QQ': 86057, 'KP': 85893, 'RH': 85670, 'FI': 85257, 'IK': 84325, 'SK': 82345, 'IF': 81640, 'IN': 81627, 'VH': 80907, 'DY': 80877, 'TQ': 80794, 'HR': 80154, 'AW': 79821, 'QD': 79379, 'VM': 78998, 'MG': 78795, 'FP': 78225, 'HV': 77875, 'KK': 77702, 'NI': 76942, 'YD': 76838, 'MV': 76697, 'SN': 75874, 'HP': 75510, 'SY': 75277, 'NT': 73377, 'ND': 73228, 'EH': 73215, 'GW': 72756, 'NS': 72417, 'PK': 71908, 'FF': 71811, 'DN': 71607, 'YS': 70405, 'NE': 70398, 'IQ': 69739, 'TK': 69627, 'YE': 69386, 'TN': 68590, 'RM': 68217, 'EY': 67357, 'TY': 66403, 'YT': 66400, 'MR': 65459, 'WA': 65440, 'PN': 64365, 'LW': 64118, 'YP': 63655, 'HD': 63607, 'PY': 63356, 'DH': 61903, 'KQ': 61441, 'IY': 60702, 'HE': 59460, 'EM': 59445, 'QK': 58960, 'KN': 58698, 'MT': 57828, 'SH': 57756, 'FN': 57377, 'WR': 56776, 'QF': 56090, 'TH': 56025, 'PH': 55712, 'RW': 55578, 'MP': 55576, 'AC': 54981, 'MS': 54683, 'SM': 54552, 'HS': 54181, 'HT': 53809, 'FK': 53719, 'FQ': 52869, 'YI': 52831, 'YF': 52465, 'MI': 51467, 'NF': 51267, 'FY': 51247, 'DM': 50571, 'ME': 50275, 'CA': 49553, 'IH': 49259, 'YQ': 49233, 'NQ': 49179, 'WG': 49104, 'LC': 49094, 'TM': 49075, 'MD': 49039, 'VW': 48295, 'HI': 48008, 'WV': 47732, 'CG': 47684, 'NK': 47389, 'QN': 47388, 'NN': 47061, 'DW': 46574, 'KF': 46377, 'PM': 45911, 'CL': 43753, 'GC': 43604, 'WS': 43179, 'HF': 42765, 'NY': 41903, 'KY': 41777, 'SW': 41386, 'QY': 41048, 'IM': 40511, 'QH': 39640, 'FH': 39073, 'YK': 38958, 'YY': 38581, 'HQ': 38290, 'WT': 38287, 'YN': 38044, 'PW': 37714, 'EW': 37338, 'VC': 37117, 'TW': 36791, 'MQ': 36322, 'QM': 35974, 'WD': 35137, 'WQ': 34786, 'WP': 33801, 'WE': 33657, 'WI': 33430, 'CV': 33132, 'MK': 32847, 'HY': 32591, 'HH': 32457, 'FM': 32221, 'KH': 32216, 'MF': 31282, 'CR': 31170, 'KM': 30911, 'NH': 29943, 'MN': 29100, 'RC': 29028, 'YH': 28285, 'WF': 27759, 'HK': 27682, 'HN': 27589, 'CD': 26991, 'QW': 26762, 'IW': 26729, 'CS': 26370, 'FW': 25850, 'NM': 25123, 'SC': 24882, 'CE': 24823, 'IC': 24635, 'CP': 24599, 'MM': 24523, 'DC': 24175, 'CT': 23788, 'WN': 23108, 'WK': 22323, 'TC': 22301, 'EC': 20968, 'NW': 20501, 'FC': 20194, 'YM': 20127, 'MH': 20045, 'CI': 19909, 'YW': 19854, 'MY': 18589, 'HW': 18258, 'WH': 18085, 'KW': 18014, 'HM': 17865, 'PC': 17683, 'WY': 17096, 'CF': 16665, 'WM': 15974, 'CQ': 13835, 'YC': 13520, 'WW': 13492, 'NC': 13201, 'CH': 12946, 'HC': 12744, 'QC': 12719, 'CN': 12464, 'CY': 12232, 'CK': 11540, 'MW': 9957, 'KC': 9899, 'CM': 7530, 'WC': 7387, 'MC': 7357, 'CW': 6505, 'CC': 6119, 'XR': 3, 'UH': 3, 'XP': 3, 'LX': 3, 'AX': 3, 'GX': 2, 'XT': 2, 'IX': 1, 'XA': 1, 'FX': 1, 'VU': 1, 'QX': 1, 'XE': 1, 'TX': 1, 'XK': 1, 'NU': 1, 'IU': 1, 'MX': 1, 'XL': 1, 'HX': 1, 'XV': 1, 'XS': 1})\n",
      "Amino acid pairs frequency c-terminal:\n",
      " Counter({'AA': 4608, 'AG': 2621, 'RR': 2397, 'AR': 2268, 'GR': 2189, 'RG': 2138, 'GA': 2088, 'PA': 2081, 'GG': 2009, 'RA': 1935, 'AS': 1897, 'SA': 1883, 'LA': 1752, 'LG': 1673, 'TA': 1561, 'RS': 1537, 'LR': 1490, 'RP': 1412, 'PR': 1391, 'SR': 1366, 'EA': 1365, 'VA': 1363, 'GS': 1349, 'EG': 1299, 'LS': 1276, 'KG': 1255, 'KA': 1250, 'VR': 1243, 'DA': 1232, 'AP': 1229, 'AD': 1225, 'TR': 1102, 'AV': 1094, 'VG': 1076, 'ER': 1022, 'LD': 1016, 'KR': 1016, 'RD': 1013, 'AL': 1012, 'LP': 991, 'SS': 988, 'PG': 975, 'ED': 969, 'ES': 959, 'GL': 952, 'KS': 952, 'QA': 932, 'TG': 893, 'AE': 875, 'VS': 873, 'RL': 863, 'QG': 854, 'RV': 852, 'LL': 851, 'TP': 850, 'GV': 841, 'DD': 841, 'GD': 829, 'GE': 829, 'DR': 811, 'PS': 810, 'AH': 802, 'SG': 802, 'KK': 748, 'LV': 741, 'QR': 715, 'RT': 706, 'TS': 699, 'NA': 697, 'EE': 693, 'GN': 689, 'SL': 686, 'LF': 680, 'DG': 667, 'HA': 667, 'KP': 660, 'NG': 656, 'KD': 655, 'IG': 654, 'HG': 653, 'IA': 640, 'GH': 634, 'EV': 632, 'QS': 627, 'VL': 624, 'PV': 620, 'LN': 616, 'KL': 603, 'TL': 591, 'VD': 588, 'SP': 588, 'RF': 584, 'EL': 583, 'AT': 576, 'RH': 574, 'TV': 571, 'SV': 571, 'QP': 569, 'AN': 566, 'GK': 561, 'DS': 553, 'IR': 553, 'PD': 545, 'LI': 543, 'NR': 538, 'LT': 537, 'KV': 536, 'RI': 533, 'PL': 530, 'GI': 530, 'LK': 529, 'DE': 525, 'KN': 525, 'VV': 523, 'GF': 522, 'GT': 515, 'IS': 515, 'SD': 515, 'TD': 513, 'FA': 507, 'FG': 504, 'SE': 501, 'KE': 489, 'RK': 486, 'NS': 471, 'AF': 471, 'VP': 469, 'AK': 468, 'RN': 465, 'SN': 462, 'SF': 461, 'AI': 461, 'SH': 460, 'LH': 459, 'DL': 451, 'QD': 447, 'AQ': 442, 'PF': 439, 'ND': 439, 'VE': 436, 'EP': 435, 'DV': 432, 'RY': 426, 'ID': 421, 'EN': 415, 'VK': 412, 'RE': 404, 'FS': 401, 'KI': 396, 'VI': 395, 'HR': 394, 'EI': 394, 'PE': 394, 'GP': 393, 'GY': 392, 'TF': 392, 'DF': 385, 'FR': 383, 'LE': 380, 'EK': 380, 'GQ': 379, 'ST': 371, 'TT': 370, 'SI': 369, 'YR': 366, 'FD': 365, 'TH': 363, 'IL': 358, 'QL': 352, 'VN': 349, 'SK': 348, 'KF': 345, 'AY': 338, 'QV': 331, 'EF': 326, 'HD': 324, 'EH': 323, 'TN': 318, 'NV': 318, 'ET': 315, 'HS': 312, 'KH': 312, 'NP': 311, 'VF': 310, 'PT': 307, 'NL': 304, 'VH': 293, 'DI': 290, 'KT': 287, 'LY': 286, 'NE': 283, 'RQ': 283, 'TE': 282, 'YA': 281, 'NN': 278, 'WR': 273, 'IK': 273, 'IP': 270, 'IV': 264, 'HP': 263, 'PH': 262, 'PN': 260, 'QE': 259, 'IE': 258, 'HV': 254, 'QN': 254, 'EQ': 253, 'SY': 249, 'TI': 247, 'VT': 247, 'MR': 247, 'QF': 246, 'SQ': 246, 'TK': 245, 'LQ': 242, 'II': 239, 'PP': 239, 'MA': 237, 'HH': 236, 'VY': 236, 'HE': 236, 'QH': 233, 'IF': 231, 'DN': 231, 'IN': 230, 'FF': 230, 'NF': 229, 'DK': 228, 'PQ': 227, 'FL': 226, 'PI': 226, 'KQ': 225, 'KY': 221, 'DH': 218, 'NK': 218, 'GC': 217, 'RW': 216, 'QK': 213, 'AC': 213, 'YL': 210, 'FV': 209, 'YG': 208, 'HL': 205, 'MG': 203, 'VQ': 202, 'TY': 199, 'NI': 199, 'PK': 191, 'FN': 190, 'FE': 189, 'NT': 189, 'YE': 189, 'CR': 188, 'WL': 188, 'CG': 184, 'DT': 183, 'NH': 182, 'EY': 179, 'YD': 178, 'ML': 178, 'FI': 175, 'MS': 174, 'IT': 174, 'GW': 173, 'QT': 171, 'YF': 171, 'YS': 170, 'YV': 169, 'SC': 168, 'DY': 167, 'QI': 166, 'WA': 166, 'IY': 164, 'WG': 163, 'SW': 163, 'DQ': 159, 'CS': 159, 'TQ': 158, 'QQ': 158, 'CA': 158, 'MP': 154, 'IH': 154, 'PC': 153, 'WS': 153, 'AW': 151, 'PY': 150, 'NY': 147, 'GM': 146, 'IQ': 146, 'NQ': 145, 'QY': 144, 'DP': 142, 'TC': 141, 'WD': 139, 'FK': 139, 'EC': 138, 'MD': 137, 'FH': 136, 'RC': 134, 'RM': 130, 'CP': 128, 'WE': 127, 'HF': 126, 'LC': 126, 'PW': 125, 'MV': 123, 'HY': 121, 'TW': 119, 'LW': 117, 'YP': 116, 'FT': 116, 'VC': 115, 'MK': 115, 'DC': 109, 'MN': 109, 'FP': 108, 'CD': 105, 'IC': 104, 'MI': 103, 'FY': 102, 'YI': 101, 'HN': 100, 'EW': 97, 'HI': 97, 'LM': 96, 'ME': 96, 'WI': 96, 'DW': 94, 'WN': 94, 'AM': 94, 'YY': 93, 'WT': 93, 'WK': 93, 'WP': 91, 'CC': 90, 'FQ': 89, 'HT': 89, 'MF': 89, 'YN': 88, 'SM': 87, 'YQ': 86, 'KM': 86, 'WF': 83, 'VM': 81, 'KC': 80, 'YK': 80, 'CH': 79, 'HK': 78, 'WV': 78, 'MH': 78, 'KW': 76, 'CL': 76, 'YH': 73, 'VW': 72, 'MT': 70, 'EM': 70, 'NC': 70, 'CN': 68, 'CE': 65, 'QW': 65, 'YT': 64, 'HQ': 64, 'FC': 63, 'CQ': 63, 'CV': 63, 'MY': 63, 'WY': 61, 'WH': 61, 'WQ': 61, 'NW': 60, 'IW': 58, 'MQ': 57, 'PM': 56, 'NM': 56, 'TM': 54, 'DM': 53, 'CK': 53, 'CF': 53, 'WW': 52, 'QC': 51, 'HC': 50, 'YC': 48, 'CT': 48, 'QM': 42, 'FM': 42, 'CY': 40, 'CI': 39, 'MM': 38, 'HW': 38, 'HM': 36, 'IM': 35, 'YW': 27, 'FW': 26, 'MW': 26, 'MC': 23, 'YM': 21, 'WC': 17, 'WM': 16, 'CW': 15, 'CM': 7, 'EX': 1})\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bulk frequency hexamers:\n",
      " Counter({'CGCCGC': 315830, 'GGCGGC': 300312, 'GCGGCG': 295725, 'CGGCGC': 284880, 'GCCGCC': 275593, 'CGGCGG': 275401, 'CCGCCG': 271223, 'GCGCCG': 251445, 'CGCCGA': 250404, 'TCGCCG': 248379, 'CCGGCG': 247892, 'GCCGGC': 245942, 'CGCGGC': 245236, 'CGCCGG': 239978, 'GCGCTG': 238077, 'GCCGCG': 234621, 'GCGCGC': 231896, 'CGGCCG': 227272, 'GCTGGC': 222331, 'CCGCGC': 220167, 'TCGGCG': 220027, 'GGCCGC': 219428, 'CGGCGA': 219217, 'GGCGCG': 218856, 'CGCTGG': 210075, 'CTGCTG': 204306, 'CGCGCC': 202719, 'GCGCGG': 202493, 'GCGGCC': 197822, 'CTGGCC': 196585, 'CGCGCG': 193893, 'GGCCGG': 190511, 'TGGCCG': 190357, 'GCCGAC': 188725, 'TGCTGG': 182654, 'CCGGCC': 180928, 'GGCGCC': 178318, 'CGCTGC': 177191, 'CTGCGC': 176041, 'TCGACG': 175297, 'CCCGGC': 172598, 'GCCGAG': 171462, 'CGACGA': 169956, 'GACGGC': 168738, 'CGACGC': 168518, 'CGGGCG': 168460, 'GCGGGC': 168100, 'CTGGCG': 167703, 'CGCGCT': 167095, 'GGCCGA': 166802, 'CGCCCG': 165265, 'GCCCGC': 164897, 'GCTGCT': 164527, 'CTCGGC': 162532, 'CCGACG': 162475, 'ACCGGC': 162438, 'GGCGCT': 161558, 'GCCGGG': 160295, 'ACGCCG': 160212, 'GTGCTG': 159572, 'TGCTGC': 157669, 'ACGGCG': 155871, 'GCTGGA': 155689, 'CCGGGC': 155145, 'GCTGCG': 154298, 'CGACGG': 149811, 'GGCGAC': 149741, 'GCGCGA': 149449, 'TGCGCG': 148223, 'CGGCTG': 147933, 'GGCCTG': 147702, 'GCGGTG': 146036, 'GGCTGG': 145910, 'CCGCGG': 145853, 'GCCCGG': 145386, 'CTCGCC': 144470, 'TGGCGC': 144213, 'GCGACG': 143804, 'TGGCGG': 143636, 'ATCGCC': 143026, 'TCGTCG': 142424, 'GACGCC': 141741, 'CCTGGC': 141332, 'GTCGGC': 141177, 'GGGCGC': 140251, 'CTGGGC': 140214, 'CGGCGT': 140017, 'CGGCCT': 139059, 'CGAGGC': 138925, 'CCGGCA': 138809, 'CGCTCG': 138116, 'GGGCGG': 137600, 'CGACCG': 137421, 'CCGCTG': 136805, 'GCCTGG': 136550, 'CTGCCG': 135340, 'GCGGCA': 135189, 'CCGCGA': 135085, 'GCTGGT': 134988, 'CGGCAC': 134422, 'TCGGCC': 134088, 'CGCGAC': 134000, 'CGGTGC': 133974, 'CTGGTG': 133538, 'TCACCG': 132830, 'GCTCGC': 132393, 'CACCGC': 132035, 'GGCGGG': 131505, 'TGCTCG': 130887, 'ATCGGC': 130693, 'TCGCGC': 130537, 'CCGAGG': 130456, 'GTCGCC': 130390, 'TGCCGG': 130008, 'GCCTGC': 129756, 'CTCGAC': 129263, 'CGTCGC': 128901, 'CACCGG': 128581, 'CGTCGA': 128488, 'CGGTGG': 128344, 'GGTGGC': 128043, 'CGGCCA': 127113, 'GGCCCG': 125946, 'TGCCGC': 125591, 'GCTCGA': 125368, 'CGCCGT': 125046, 'CGCCTG': 124565, 'GCGATC': 124462, 'CCGACC': 124073, 'CGGCAT': 123864, 'GCTCGG': 123791, 'GGCGAG': 123439, 'ACCGCC': 123410, 'CCTGCT': 123299, 'GGGCCG': 123146, 'CCCGCG': 123094, 'CCGCCC': 123010, 'GATCGC': 122598, 'GCACCG': 122592, 'GCCGCT': 122196, 'TCGCGG': 122017, 'CCGGCT': 121674, 'ACCGCG': 121632, 'GCCCTG': 121488, 'AGGCCG': 121467, 'GCTGGG': 121108, 'CGACCT': 121085, 'GGCGGT': 120983, 'CCGGTG': 120761, 'GCCGGT': 120746, 'CGGCAA': 120742, 'GCGGCT': 120472, 'GCTGCC': 120423, 'GACCGC': 120330, 'TGGGCG': 120187, 'CGGCCC': 119786, 'GACCTG': 119751, 'GGTGCT': 119421, 'TGGTGG': 119115, 'GTGGCG': 118992, 'CCCGCC': 118817, 'GACCGG': 118804, 'GCGCTC': 118428, 'TCGACC': 117813, 'GGCGAT': 117629, 'TGATCG': 117247, 'GGCATC': 116758, 'CGAGCG': 116248, 'CGATCG': 115600, 'CGGGCC': 115393, 'GGCTGC': 115281, 'CAGCGC': 115254, 'CTGCTC': 115210, 'GAGCTG': 115138, 'CCGAGC': 115088, 'GCGCCC': 114826, 'GCGACC': 114751, 'CGCGCA': 114750, 'CGCGGG': 114567, 'GCCACC': 114465, 'GCAGCG': 114261, 'GCGCAG': 113930, 'CGCGGT': 113692, 'GCCGGA': 113652, 'CGAGGA': 113423, 'GCATCG': 113014, 'GCCGTC': 112992, 'GGCGTG': 112957, 'CATCGC': 112781, 'CCACCG': 112769, 'CCGTCG': 112567, 'GGCACC': 111514, 'CGGTCG': 111457, 'CACCGA': 111415, 'TCGAGG': 111316, 'TCGGCA': 111303, 'TGGTCG': 111121, 'GACGAC': 110293, 'CATCGA': 110274, 'CTGATC': 110185, 'GCAGGC': 110107, 'GAGGCG': 110073, 'CCTGCG': 109838, 'GCCGCA': 109646, 'GACGCG': 109637, 'GCGTCG': 109212, 'CTGGTC': 108489, 'GAGCGC': 108447, 'TCCGGC': 108325, 'GGGCGA': 107861, 'CCGCCA': 107377, 'GTGGTG': 107158, 'GGCGCA': 106715, 'GTGCGC': 106572, 'GTGCCG': 106265, 'GATCGA': 106078, 'GCTGAT': 105318, 'GCCGTG': 105262, 'GCCGAT': 105250, 'CGCCAC': 105159, 'GTGGCC': 105090, 'GCGCCA': 105066, 'CTGGAC': 104592, 'TCGCCC': 104561, 'CGTCGG': 104033, 'ACGACG': 103657, 'ACCTGG': 103596, 'AGCGGC': 103583, 'GGTCGA': 103472, 'GGCAGC': 103118, 'TTCGGC': 103087, 'GGCGTC': 103023, 'GACGAG': 102811, 'GGGCTG': 102810, 'CGCAGC': 101910, 'ACCTGC': 101853, 'CGTGGC': 101764, 'TCATCG': 101746, 'GCGAGC': 101667, 'GCCAGC': 101561, 'TGGTGC': 101471, 'GGTCGC': 101434, 'CTGGAG': 101316, 'GGTGCG': 101195, 'CCTCGC': 101193, 'CGCACC': 101040, 'GTCGAC': 101025, 'TGCGCC': 100918, 'GTCGCG': 100517, 'AGCGCG': 100450, 'ACGGCC': 100393, 'CGGCTC': 100376, 'CCTGGA': 100161, 'GGCTCG': 100157, 'TGCGGC': 100077, 'CGTGCT': 100071, 'GGCGGA': 99888, 'CGCGAG': 99761, 'ATCGAC': 99646, 'CACGGC': 98830, 'GAGGCC': 98724, 'TGCCCG': 98666, 'GCCTCG': 98424, 'CCCTGG': 98294, 'GCGCCT': 97967, 'CGGCAG': 97941, 'CGCCCT': 97680, 'GCCATC': 97548, 'CCTGCC': 97307, 'CGACGT': 96948, 'GCGAGG': 96926, 'TTCGCC': 96847, 'CGGGCA': 96672, 'AGGCGC': 96103, 'CGCGAT': 96092, 'AGGCGG': 96050, 'GGCAAG': 95934, 'ACGCGC': 95857, 'CGCTGA': 95648, 'GCGGTC': 95601, 'CGGTGA': 95561, 'CCGCCT': 95552, 'AGCTGG': 95477, 'CCTCGG': 95295, 'CGGGCT': 95270, 'CGAGCT': 95031, 'CTGACC': 94434, 'CGATCC': 94185, 'CGATGC': 94074, 'TCGCCA': 93707, 'TCCGCG': 93703, 'TGCCGA': 93179, 'GTCGAG': 93109, 'CCATCG': 93049, 'ATCGCG': 93026, 'TGGACG': 92964, 'GCGATG': 92762, 'CTGCGG': 92731, 'GCGCAC': 92536, 'CAGGCC': 92448, 'GGTGGT': 92248, 'CGGGGC': 92089, 'TCGAGC': 91985, 'CGGACG': 91952, 'GGACGC': 91808, 'ACGGCA': 91668, 'TGACCG': 91345, 'CAGGCG': 91299, 'ACGCTG': 91173, 'GCCGAA': 91140, 'CTCGCG': 91041, 'GACCCG': 90851, 'CCGATC': 90547, 'CCTGGT': 90266, 'CCTCGA': 90259, 'GGCCAG': 89978, 'GCGTGC': 89867, 'GATCGG': 89815, 'GCTGCA': 89757, 'CCGGAC': 89268, 'GCCCGA': 89252, 'ACCCGG': 89250, 'CGCATC': 89172, 'GCGTGG': 89122, 'CGAGGT': 89071, 'ATCCGC': 88895, 'CGCCAG': 88780, 'CGCCAA': 88719, 'CGAGCA': 88610, 'GGCACG': 88452, 'CCAGCG': 88413, 'GCACGC': 88361, 'ACCCGC': 88289, 'CCTGGG': 88097, 'CATCGG': 87815, 'CGCCAT': 87796, 'CCCTGC': 87758, 'GGTGGA': 87722, 'ACGCGG': 87666, 'CGACCC': 87626, 'CAGCGG': 87530, 'CTGCCC': 87473, 'AGCGCC': 87349, 'CAAGGC': 87035, 'GGTCGG': 86943, 'GGGGCG': 86747, 'TGCTGA': 86549, 'CGCAGG': 86534, 'ACCTCG': 86496, 'CCCCGG': 86154, 'CCGTGG': 85865, 'GAGCGG': 85759, 'CCGCAC': 85752, 'CACGCC': 85725, 'ATGCCG': 85581, 'GGCCGT': 85501, 'GATCCG': 85262, 'TCGGGC': 85149, 'GGACGG': 85013, 'CGCGGA': 84534, 'ATGGCG': 84495, 'GCTGAC': 84490, 'GAAGGC': 84402, 'GAGCAG': 84278, 'AGGGCG': 84254, 'GATGGC': 84157, 'TCCTCG': 83814, 'GCCCCG': 83779, 'CCGGTC': 83756, 'CCGCAG': 83644, 'CAGCCG': 83568, 'GGTGCC': 83549, 'CGACCA': 83503, 'AGATCG': 83353, 'ATCGAG': 83039, 'TCCTGC': 83001, 'CTGTTC': 82910, 'CCCGAC': 82764, 'GTCGTC': 82522, 'GGCCAC': 82490, 'GCGGAC': 82346, 'CCGGGG': 82270, 'GCCTTC': 82222, 'TCGATC': 82129, 'GAGATC': 82022, 'GGCTTC': 81972, 'CCGTGC': 81962, 'TCGCTG': 81818, 'TGGCGA': 81763, 'ATCGTC': 81606, 'TCCTGG': 81416, 'GCCAAG': 81232, 'AGCCGC': 81207, 'ACCGAC': 81159, 'GGTGAT': 81079, 'GGACGA': 81046, 'CGCCTC': 80975, 'ACCGGG': 80882, 'GCTTCG': 80704, 'CGGGTG': 80523, 'CGCCCA': 80444, 'GTCACC': 80341, 'AGCCGG': 80138, 'CGTGGT': 79845, 'TCGGTG': 79742, 'TGGCTG': 79590, 'GCAGCC': 79549, 'GCACGG': 79430, 'TCCGCC': 79342, 'AAGGCG': 79242, 'TGCAGG': 79153, 'CTGGAA': 79131, 'TGGCCA': 78955, 'AGCTGC': 78864, 'CGTCGT': 78778, 'GTGATC': 78682, 'CTTCGA': 78432, 'CCGATG': 77781, 'TGCGCA': 77760, 'GCGCGT': 77585, 'GGCGAA': 77515, 'CGCGTC': 77430, 'CTTCGC': 77407, 'CACCCG': 77398, 'GCCACG': 77348, 'GAGCCG': 77327, 'ATCACC': 77288, 'GTGCTC': 77072, 'TTCGAC': 76995, 'GACATC': 76878, 'TGGCCC': 76549, 'GCAGCA': 76301, 'CCTTCG': 76290, 'ATCCTG': 76271, 'CAGGGC': 76260, 'GCAAGG': 76084, 'CGATGG': 76061, 'ACGTCG': 75997, 'AACGGC': 75867, 'TGTTCG': 75843, 'CCGCTC': 75777, 'CCCGGG': 75707, 'TGGAGC': 75431, 'GATGCC': 75411, 'CGTGCG': 75346, 'GCGCAA': 75310, 'CCAGGC': 75303, 'ACGCCC': 75299, 'CGTGCC': 75196, 'ACGAGG': 75145, 'CAGCAG': 75087, 'GTGGGC': 74964, 'CGCACG': 74919, 'TCAAGG': 74915, 'GCGGGG': 74751, 'GAGGGC': 74614, 'CCCCGC': 74509, 'GACGTG': 74430, 'CCCGCA': 74425, 'CGCCCC': 74308, 'CGACAT': 74278, 'GGGCGT': 73936, 'CGATCA': 73914, 'GTGGTC': 73911, 'GACGTC': 73910, 'CGAGCC': 73547, 'GCCCAG': 73486, 'CTGCAG': 73427, 'TCGTGC': 73400, 'GCAGGA': 73369, 'GGAGCG': 73332, 'GGACCG': 73298, 'CGATCT': 73247, 'ACACCG': 73211, 'TCGACA': 73125, 'CCCGCT': 73085, 'AGCAGG': 73069, 'ACGACC': 72998, 'GGAGGC': 72887, 'GCGCAT': 72874, 'CCCGGA': 72826, 'GAGGAG': 72823, 'GGCAAC': 72735, 'ACCACC': 72675, 'CGGCTT': 72636, 'GAGGTG': 72635, 'CGAGAT': 72596, 'CGAGGG': 72424, 'TGGGCC': 72153, 'GATGCG': 72102, 'TCGGCT': 72071, 'CCGGGT': 72042, 'CCGCGT': 71945, 'CCTGCA': 71925, 'GGCCTC': 71857, 'GGATCG': 71604, 'GGCATG': 71592, 'TTGCCG': 71514, 'CTGATG': 71461, 'GGCCAT': 71444, 'TTCCTG': 71114, 'GAACTG': 70972, 'CCTGAT': 70932, 'ACCCTG': 70908, 'CAAGGA': 70874, 'GGCCAA': 70837, 'AGGTCG': 70715, 'ACATCG': 70702, 'TGCGGG': 70697, 'ACCCCG': 70582, 'CGCTTC': 70439, 'AAGGCC': 70376, 'TCGTGG': 70192, 'GTGCGG': 70059, 'CTCACC': 70049, 'CCAAGG': 70003, 'ATGCTG': 69888, 'CGAAGC': 69839, 'CTTCGG': 69759, 'GCGGGT': 69706, 'AACGCC': 69696, 'CAACGC': 69685, 'CATGGC': 69601, 'CCGGAG': 69505, 'TCAACG': 69498, 'CCACGC': 69455, 'CGCGTG': 69409, 'CACCGT': 69371, 'GATCGT': 69313, 'CGGATC': 69242, 'GACACC': 69228, 'ATCCGG': 69206, 'TCGATG': 69182, 'GCCATG': 69168, 'GACCTC': 69158, 'ATGGCC': 69101, 'GCGGAG': 68822, 'ACGGTG': 68819, 'AAGCCG': 68792, 'CCGAAG': 68649, 'GCCAGG': 68644, 'GCGAAG': 68580, 'TCTTCG': 68571, 'ATCATC': 68468, 'GTTCGC': 68360, 'ATGCGC': 68213, 'CTGCAC': 68179, 'ATCTGG': 68155, 'AGGACG': 68143, 'GCTGTT': 68125, 'CTGGCT': 68098, 'TCGCCT': 68086, 'TCCCGG': 68007, 'GTCCGC': 67980, 'AAGCTG': 67978, 'GAAGCC': 67969, 'GATCTG': 67929, 'AGCAGC': 67914, 'TGCGCT': 67906, 'AGCTCG': 67892, 'TGGAGG': 67886, 'CCGACA': 67795, 'TTCGCG': 67766, 'GACCGA': 67709, 'CCGGGA': 67674, 'CCCGGT': 67666, 'CATCGT': 67641, 'CTCGTC': 67515, 'GGTGAC': 67462, 'ACGGGC': 67401, 'CGACAA': 67278, 'GAGGAC': 67270, 'CCCCGA': 67003, 'CGACAC': 66944, 'GCTGAA': 66925, 'GCAGCT': 66905, 'CCAGCC': 66902, 'ACCGAG': 66865, 'GGCAGG': 66776, 'TCTGGC': 66609, 'TGATGG': 66548, 'CACCAC': 66532, 'CGAAGG': 66474, 'CCGCAA': 66216, 'CCCGAG': 66212, 'GCCAAC': 66190, 'CCACGG': 66130, 'GTTCGA': 65992, 'CGGTCA': 65951, 'CGATGA': 65926, 'CCGAAC': 65904, 'CACGCT': 65833, 'ACCAGC': 65743, 'GCAACG': 65739, 'GTCCGG': 65703, 'AGGTGC': 65656, 'GTGGAC': 65576, 'CGTGGA': 65575, 'TTCCGC': 65522, 'GGGCCT': 65467, 'CACCTG': 65465, 'CAGCTG': 65427, 'CGTCCG': 65421, 'GCAAGC': 65410, 'CGCAAG': 65237, 'CGTTCG': 65213, 'CGCCTT': 65027, 'GGGCAG': 64988, 'CGCGAA': 64984, 'CGCTGT': 64959, 'GAACGC': 64913, 'GAAGAA': 64718, 'ACGAGC': 64696, 'TGGGGC': 64588, 'CCGGAT': 64584, 'GTCATC': 64561, 'ACCGTC': 64530, 'CAGGTG': 64496, 'GTGCAG': 64481, 'GGGCAT': 64474, 'ACGCCA': 64304, 'GGCCCT': 64303, 'CAGCGA': 64285, 'GGGCAC': 64234, 'GTTCGG': 64233, 'TCAGCG': 64143, 'GCGTTC': 64128, 'TGGGCA': 64101, 'TGCAGC': 64074, 'CGTCAC': 64062, 'GGGCAA': 63989, 'GACGCT': 63935, 'GAAGCG': 63841, 'GCACCC': 63834, 'ATCCCG': 63732, 'CGGACC': 63544, 'GCATCC': 63515, 'TCGCGA': 63511, 'TGGACC': 63415, 'GCTCAA': 63345, 'CTGGAT': 63309, 'GCGGAT': 63214, 'ACTGGC': 63196, 'CCCTCG': 63140, 'GCGGGA': 63139, 'GAGGTC': 63095, 'GGGTGC': 62908, 'CTGTCG': 62483, 'GGGTGG': 62475, 'TACGGC': 62320, 'CATCCG': 62293, 'CAACGG': 62290, 'GGTGGG': 62213, 'GCAGGT': 62120, 'GACGGG': 62017, 'CCGCAT': 61855, 'CTCGAT': 61844, 'TTCACC': 61834, 'TCACCC': 61820, 'CCAGCA': 61770, 'GCATGG': 61765, 'TCCCGC': 61725, 'CTCCGG': 61715, 'GACCAG': 61471, 'GTCGAT': 61427, 'CTGGGG': 61360, 'CGCAAC': 61341, 'AGCCCG': 61331, 'CCACCC': 61330, 'AAGGGC': 61289, 'CGGAGC': 61211, 'CAAGCT': 61180, 'CTGGCA': 61082, 'TGGTGA': 60938, 'GACGGT': 60884, 'GCCTGA': 60845, 'GCGACA': 60699, 'AAGCGC': 60658, 'GCTGTC': 60418, 'CGAAGA': 60386, 'ACCGTG': 60341, 'GTGGAG': 60336, 'TCGTCA': 60335, 'CAACGA': 60318, 'AGGTGG': 60292, 'AACTGG': 60267, 'TCCGGG': 60247, 'GATCCT': 60169, 'GCAGGG': 59936, 'GGTGCA': 59844, 'ACGATC': 59808, 'CAGATC': 59769, 'TGGAAG': 59700, 'CGAACT': 59598, 'CGAACG': 59512, 'AGGCGA': 59483, 'GGAGGA': 59458, 'CCACCA': 59062, 'TGACGG': 59060, 'GGAGCT': 59028, 'CGGGTC': 58957, 'AGCGGG': 58927, 'GACGAT': 58861, 'TGATCC': 58810, 'GCCCTC': 58769, 'ACCGCA': 58757, 'GGCTGA': 58736, 'ATCGTG': 58708, 'TGGCCT': 58639, 'GCACCA': 58443, 'GCAGAT': 58402, 'ACCATC': 58387, 'ATGGGC': 58348, 'TGCTCA': 58310, 'AGCCTG': 58309, 'TGATGC': 58306, 'CGGAGG': 58235, 'TTCGTC': 58095, 'TTCGAG': 58080, 'GGCGTT': 57874, 'GATGCT': 57824, 'ACGGCT': 57823, 'TGCTGT': 57648, 'CCTGAC': 57607, 'CGGTGT': 57479, 'GTGATG': 57449, 'CAGCTC': 57395, 'CCAGCT': 57329, 'GCGAAC': 57323, 'AACCTG': 57309, 'CACGCG': 57249, 'GGCTAC': 57121, 'GCATCA': 57085, 'CACCCT': 57080, 'AAGAAG': 56969, 'TGGTCA': 56945, 'ACGTGC': 56936, 'TGTCGG': 56888, 'AGACCG': 56832, 'GAAAAA': 56810, 'GTGACC': 56756, 'CCGAGA': 56604, 'CATCCT': 56507, 'TCGAAG': 56445, 'CAAGGT': 56435, 'GACCAC': 56387, 'TTCCGG': 56306, 'CATCAC': 56305, 'GCACCT': 56304, 'CCGTCC': 56238, 'GATCAC': 56134, 'GGGACG': 56101, 'TGGCAG': 55975, 'CAAGGG': 55970, 'AGGCCA': 55954, 'GATCAA': 55915, 'GATGAA': 55911, 'CCCAGG': 55911, 'CGCTCA': 55873, 'CATCAA': 55856, 'GCACGA': 55831, 'GATCAT': 55802, 'GTGACG': 55801, 'TGAAGG': 55774, 'GTCGTG': 55773, 'GGTCGT': 55614, 'GCGCTT': 55544, 'GATGAT': 55513, 'GCTCGT': 55500, 'TCACCA': 55498, 'TGATGA': 55485, 'CTCGGT': 55454, 'AGCGCA': 55425, 'CGAGAA': 55423, 'CTGAAG': 55315, 'CTGACG': 55256, 'ATCTCG': 54971, 'CTCGGG': 54937, 'GACAAG': 54824, 'GGATGC': 54674, 'CGGGAC': 54651, 'ATCTGC': 54639, 'TCCGCA': 54639, 'CCAACG': 54624, 'GGGGCC': 54600, 'GACTGG': 54585, 'GGACCT': 54516, 'TCTCGG': 54470, 'GGTCAA': 54412, 'CACGAC': 54364, 'CGTGAT': 54249, 'GGGTCG': 54170, 'ACGTGG': 54168, 'CTACGA': 54144, 'CCAGGA': 54096, 'TGACGC': 54060, 'ACCACG': 54048, 'GGTGAA': 54008, 'TACGCC': 53917, 'AAAAAA': 53907, 'GGTCCG': 53884, 'TTCATC': 53866, 'CCCAGC': 53795, 'TCGAGA': 53746, 'GATGGA': 53730, 'TGGATG': 53673, 'TCAAGC': 53646, 'CGAGAC': 53601, 'CTTCCT': 53596, 'GTCGGG': 53583, 'CGGATG': 53572, 'TTCCCG': 53535, 'GGAGAT': 53464, 'GGGCTC': 53432, 'ATCAGC': 53403, 'GGGGCT': 53376, 'TCCAGC': 53315, 'TGATCA': 53244, 'CCGTCA': 53149, 'CACGGT': 53131, 'CAGGAC': 53090, 'GACTTC': 53079, 'GTCCTG': 53069, 'TGCACG': 53042, 'GCGAGA': 53020, 'GCGTGA': 52924, 'CGGTTC': 52878, 'GTGTTC': 52860, 'CGGGGT': 52795, 'CCGTTC': 52774, 'GGGCCA': 52727, 'TTCTGG': 52691, 'GCGTCC': 52669, 'TCCCCG': 52648, 'GCCCAC': 52647, 'GGTTCG': 52547, 'CCGGAA': 52495, 'CTCAAG': 52491, 'CTCCGC': 52488, 'CAGCCC': 52423, 'GCTCAC': 52422, 'TGTCGC': 52333, 'GGCCTT': 52332, 'AGGGCA': 52324, 'TGAAGA': 52306, 'GAGAAG': 52255, 'GGTCAC': 52225, 'GCTCCG': 52216, 'TCATCA': 52178, 'AACGCG': 52166, 'CGTGGG': 52163, 'GTGCCC': 52097, 'GACCGT': 52053, 'CAGCCT': 52050, 'CCATGG': 51997, 'CGCTCC': 51913, 'CAGGTC': 51853, 'CGTCAA': 51837, 'TCCAGG': 51800, 'CAACCT': 51793, 'CAAGCG': 51781, 'GTCAGC': 51778, 'GCCTAC': 51778, 'CTCGCT': 51754, 'CAAGCC': 51703, 'GCCCGT': 51673, 'GCGTCA': 51589, 'CATCAT': 51561, 'CGTGCA': 51552, 'ATCAAG': 51478, 'ACCTGA': 51458, 'GGAGCA': 51394, 'ATGCGG': 51375, 'GGAAGC': 51334, 'GGACGT': 51293, 'ACGCCT': 51263, 'GACAGC': 51197, 'CGGTCT': 51180, 'GGCCCC': 51175, 'TCTCGC': 51130, 'AGGAGG': 51112, 'CGGCTA': 51095, 'CGTCAT': 51079, 'CCACCT': 51069, 'CTGGGA': 51056, 'ATCAAC': 50950, 'GCTACG': 50935, 'CACCAT': 50916, 'CGGTCC': 50910, 'GCTTCC': 50810, 'TTGCTG': 50754, 'AACCGC': 50733, 'GCAACC': 50659, 'GAACGG': 50605, 'AAGATC': 50593, 'GGAAGA': 50517, 'GAGACC': 50458, 'GAAGAT': 50455, 'AAGCGG': 50455, 'GTCTGG': 50435, 'TACCTG': 50356, 'CCATCC': 50351, 'TCGGTC': 50294, 'TTCTTC': 50183, 'TCTCCG': 50137, 'TCGTCC': 50127, 'GATCTC': 50109, 'GCCGTT': 50065, 'AGGAGC': 50021, 'CCACGA': 49992, 'CCCACC': 49935, 'TGCCGT': 49922, 'ACGGTC': 49847, 'GAGTTC': 49788, 'ACCTTC': 49771, 'AGGCCC': 49759, 'CTCGAG': 49749, 'GATCCC': 49747, 'CACCCC': 49724, 'ATGGTG': 49723, 'TCTGGA': 49712, 'CGTCCT': 49689, 'CCAAGC': 49606, 'CAGGAG': 49601, 'GTCGGT': 49597, 'CGCTAC': 49544, 'GGAAAA': 49459, 'TCAACC': 49385, 'ACTTCG': 49353, 'CTTCGT': 49320, 'ACAAGG': 49312, 'TCAGCC': 49273, 'GCAAGA': 49224, 'CCAGGG': 49211, 'CAACCG': 49159, 'GGCCCA': 49131, 'CCCGAT': 49126, 'TCACGC': 49096, 'CTACGC': 49025, 'TCGAAC': 48996, 'AGAAGG': 48978, 'GGGGTG': 48915, 'AGCGCT': 48912, 'ACCAGG': 48878, 'CAGCAC': 48878, 'AGCTGA': 48873, 'GTCGCT': 48794, 'TACCGC': 48791, 'TGGCGT': 48669, 'CGCAGA': 48653, 'GCAGAA': 48612, 'CCTGTT': 48602, 'CGGGGA': 48539, 'AAGGTG': 48509, 'TGGAGA': 48439, 'GACGAA': 48408, 'CGTGAC': 48386, 'ATGATC': 48368, 'GGTGTT': 48361, 'CCATGC': 48330, 'GGAGAA': 48262, 'GAGCAC': 48243, 'CCCCCG': 48224, 'ATCCTC': 48119, 'ACCGCT': 48090, 'GCCTCC': 48088, 'TCATGG': 47983, 'TGAAGC': 47976, 'TGGAAC': 47868, 'CCATCA': 47843, 'CACCTC': 47825, 'ACCGGT': 47816, 'TCATCC': 47749, 'CCTGAA': 47747, 'CCCGTC': 47742, 'TGGATC': 47715, 'CTTCCG': 47705, 'AACCCG': 47647, 'GGCTCC': 47644, 'ACAGCG': 47637, 'CGCATG': 47636, 'GATGAC': 47603, 'TCTGGG': 47599, 'GGTCTG': 47583, 'GTCAAC': 47560, 'ACGACA': 47450, 'ATCTTC': 47442, 'GAAGAC': 47432, 'CAGCGT': 47385, 'TCTACG': 47374, 'AGTTCG': 47255, 'TCCGCT': 47201, 'ACGCTC': 47186, 'TCCACC': 47154, 'CACCAA': 47143, 'CGGGAG': 47098, 'TTGGCG': 47091, 'AACCGG': 47090, 'CCTTCC': 47085, 'CTCAAC': 47079, 'ACCGGA': 46961, 'TGGGTG': 46902, 'CAGCAA': 46845, 'CTGGGT': 46802, 'GCATGC': 46794, 'ACGCGA': 46792, 'GGGGGC': 46769, 'TGAGCG': 46743, 'CCGCTT': 46739, 'GTTCCG': 46696, 'CCAGGT': 46623, 'CAAGAA': 46619, 'CGCGTT': 46576, 'CGGACA': 46419, 'AAGGAC': 46401, 'GAAGGA': 46383, 'GCCTGT': 46358, 'GATGGT': 46337, 'CTACGG': 46310, 'CCGGTT': 46299, 'TTCGTG': 46277, 'CGATGT': 46242, 'GAAGCT': 46225, 'GCGGAA': 46218, 'TGGAAA': 46197, 'AGCCGA': 46192, 'CGGGAT': 46183, 'CCTCCG': 46168, 'GACCCC': 46136, 'TCACGG': 46028, 'AAGCAG': 45957, 'GAACCG': 45952, 'AGATCA': 45919, 'GAGCTC': 45877, 'TCCCGA': 45739, 'AAGACC': 45728, 'TCTGCG': 45653, 'GAAGAG': 45642, 'CATGGA': 45583, 'TACGAC': 45573, 'CCTGTC': 45520, 'GGATGG': 45464, 'GTGAAG': 45450, 'CATGCC': 45447, 'GATCAG': 45395, 'AAAAAG': 45354, 'GCTTCA': 45313, 'TGACCA': 45262, 'CCTACG': 45237, 'CTGTCC': 45223, 'ATGAAG': 45221, 'CTCGAA': 45214, 'CGGAGA': 45198, 'CGTTCC': 45176, 'CCTCGT': 45127, 'CATGCT': 45106, 'ATCCAG': 45060, 'TGTCGA': 45042, 'AGAAGA': 45038, 'TGCACC': 44946, 'CAGACC': 44938, 'TCGGGG': 44908, 'GTCTTC': 44853, 'CGACTG': 44846, 'GCACTG': 44825, 'TCCTGA': 44804, 'TGCAGA': 44756, 'TTCCTC': 44729, 'GGGTGA': 44721, 'GATGGG': 44715, 'CAGCCA': 44671, 'GGGGCA': 44588, 'GACAAC': 44579, 'GGGCTT': 44547, 'GGGCCC': 44461, 'CCGTGA': 44424, 'CGGGGG': 44372, 'AGAAGC': 44359, 'CTGAAC': 44357, 'GGTGTC': 44353, 'GTGTCG': 44305, 'GTTCCT': 44305, 'GGAGCC': 44190, 'TGACCC': 44098, 'GGTCAT': 44083, 'TTGCGC': 43981, 'CACCAG': 43922, 'CTGTGG': 43850, 'GTCGAA': 43825, 'CGAACC': 43795, 'CATCCC': 43785, 'CCCTGA': 43725, 'CTTCAC': 43706, 'CTCATC': 43697, 'AAGGTC': 43691, 'AGCACC': 43673, 'CCAACC': 43669, 'AGGGCC': 43654, 'GTGGAA': 43636, 'CCCGAA': 43631, 'AAAAAT': 43626, 'TACCGG': 43622, 'GTCCCG': 43607, 'ATGACC': 43593, 'CAACAA': 43552, 'CGAACA': 43507, 'CTACCG': 43448, 'ACAACG': 43431, 'TGTCCG': 43343, 'TCGCTC': 43334, 'CAAGAT': 43303, 'GAGATG': 43280, 'TGACGA': 43279, 'ATGATG': 43271, 'TCTCGA': 43218, 'ATGTCG': 43143, 'TGGACA': 43105, 'CAAGCA': 42967, 'TGCCTG': 42933, 'AGATCC': 42900, 'GCGGTT': 42863, 'ATTGCC': 42854, 'GGAACG': 42830, 'ACTGGA': 42786, 'AACGAC': 42740, 'GGACAT': 42723, 'GCAGAC': 42708, 'TGTTCC': 42696, 'CATGGG': 42691, 'TGCCCA': 42681, 'AGCACG': 42678, 'ATGACG': 42634, 'ACTGGG': 42628, 'AACTGC': 42619, 'TCCGAC': 42615, 'GGGATC': 42601, 'ACGATG': 42578, 'GCTGTG': 42564, 'TCTTCC': 42518, 'CGGAAG': 42434, 'TCCGGA': 42413, 'AGCGAC': 42405, 'TGTGGC': 42399, 'ATCGAA': 42384, 'GTCAAG': 42379, 'AGCAGA': 42336, 'CGCCTA': 42301, 'GGACAA': 42278, 'GTTCGT': 42273, 'CACCCA': 42255, 'TGAAAA': 42167, 'CTTCAA': 42107, 'GAAATC': 42107, 'CCTCAA': 42068, 'CTACCT': 42046, 'TTCCAG': 42032, 'TTCTCG': 42027, 'GGATGA': 41948, 'TCCGGT': 41945, 'TCTGGT': 41937, 'ATGCAG': 41903, 'TGAACG': 41855, 'CAGGAA': 41805, 'TCAACA': 41790, 'GGAAGG': 41786, 'CTGAGC': 41754, 'ATCGGG': 41686, 'GACCAT': 41565, 'GTGGCT': 41517, 'GAAGGT': 41500, 'TGGGCT': 41488, 'CGAGTT': 41411, 'CTTCAT': 41346, 'CAACCC': 41336, 'CCCGTG': 41318, 'CAGTTC': 41288, 'GGGAGC': 41281, 'GGAGGT': 41270, 'GGCAGA': 41229, 'TCCACG': 41191, 'CTGCGT': 41190, 'GCAACA': 41186, 'GGCTGT': 41163, 'GAACAG': 41125, 'CAAGAC': 41110, 'AACATC': 41107, 'GTCTCG': 41053, 'CCCACG': 41047, 'GATTTC': 41033, 'AAGAAA': 41009, 'CAACGT': 40999, 'GTGGAT': 40989, 'GGTCAG': 40927, 'CCGACT': 40920, 'TTCAGC': 40872, 'AGATGG': 40866, 'AGGTGA': 40854, 'TTTGCC': 40834, 'CCTTCA': 40799, 'GCCCCC': 40786, 'ACGAGA': 40746, 'GAAGGG': 40728, 'TCGACT': 40694, 'ATCGAT': 40679, 'GAGGAA': 40657, 'GAGAAC': 40586, 'CATGCG': 40574, 'CGACAG': 40569, 'ACATCC': 40556, 'GCTACC': 40526, 'AAGCTC': 40481, 'TGCGGT': 40467, 'CACTGG': 40463, 'CACGGG': 40431, 'GACATG': 40414, 'TGCGGA': 40389, 'AGCGTG': 40362, 'AGGTCA': 40334, 'CAGAAG': 40319, 'ACATCA': 40284, 'AGGACC': 40275, 'CGTCCC': 40275, 'GAGACG': 40264, 'AGGAGA': 40242, 'GAGCCC': 40213, 'TGGCAC': 40203, 'GAACTC': 40203, 'GGATCA': 40169, 'CCAGAT': 40118, 'CATGGT': 40109, 'AGGAAG': 40099, 'AACACC': 40049, 'GGAACT': 40037, 'TTCAAC': 40018, 'CGTCTG': 39973, 'TTCAAG': 39972, 'CGACTT': 39971, 'GGCTTT': 39956, 'ATGCCC': 39920, 'GCCAGA': 39865, 'GGTGTG': 39852, 'ATCATG': 39852, 'CTCGTG': 39827, 'GGCATT': 39823, 'CTCCGA': 39803, 'GCGTTG': 39769, 'GATGTG': 39768, 'GTCACG': 39747, 'ATGGTC': 39745, 'ACCCGA': 39739, 'GAGCGA': 39714, 'ACGTCA': 39629, 'ATGGAC': 39627, 'CATCTG': 39616, 'TCAAGA': 39611, 'GGACTG': 39587, 'CATCCA': 39586, 'CCTCAC': 39584, 'GGTGAG': 39525, 'TGTTCA': 39511, 'ACACCC': 39472, 'GAAAAG': 39437, 'TCACCT': 39364, 'TCTGCT': 39362, 'GTGGCA': 39326, 'GAAACC': 39310, 'GGCAAT': 39293, 'GGAGGG': 39255, 'GGACCC': 39233, 'GACTAC': 39197, 'TGCCCT': 39157, 'TGGCTC': 39091, 'GTCCTC': 39081, 'AACAGC': 39069, 'GACCCT': 39022, 'CAGATG': 39005, 'AAAGAA': 38991, 'TCGCAG': 38980, 'TCAGCA': 38969, 'ACAAGC': 38969, 'GACGGA': 38909, 'GCAGTT': 38898, 'CGTTGC': 38884, 'CGCTCT': 38823, 'CGGAAC': 38805, 'TGGTCC': 38781, 'GTGGGG': 38738, 'TGGTGT': 38737, 'ATGCTC': 38722, 'GCATCT': 38721, 'CGTCTC': 38701, 'GAAGTG': 38697, 'AAAAGA': 38618, 'CGGGAA': 38605, 'TTTCCG': 38571, 'AGAAAA': 38524, 'CTCCTG': 38482, 'CTGCAA': 38459, 'TCTGCC': 38457, 'CAACTG': 38412, 'CTCCCG': 38392, 'GTGCAC': 38361, 'ACGAAG': 38359, 'TACGCG': 38324, 'ACCAAC': 38315, 'ACCGAT': 38314, 'GGATCT': 38313, 'GTCTGC': 38308, 'TTGATG': 38296, 'GCGACT': 38292, 'GGAGAC': 38292, 'GGCAAA': 38268, 'TGATCT': 38243, 'ATCACG': 38217, 'AAGAAC': 38152, 'GATGTC': 38121, 'CAGCAT': 38117, 'CGTCCA': 38109, 'CACGAT': 38075, 'CACCTT': 38073, 'CTTGCC': 38006, 'GCCTTG': 37970, 'AGAACG': 37919, 'AGGCCT': 37896, 'ATGGAG': 37756, 'CGTCAG': 37724, 'GATGCA': 37703, 'AGGACA': 37606, 'GCATGA': 37573, 'AAGGAG': 37489, 'TTGCGG': 37465, 'AAGGAA': 37465, 'CTGGTT': 37464, 'TGGCAA': 37464, 'CAAAAA': 37458, 'CAACAT': 37410, 'AAGATG': 37388, 'GAAAGC': 37349, 'AAAAGC': 37282, 'GATGAG': 37256, 'TGACCT': 37255, 'TCTTCA': 37236, 'CGGACT': 37207, 'AGCGTC': 37178, 'CGGAAA': 37165, 'ACCCAG': 37137, 'GATCCA': 37058, 'TTCGGG': 37047, 'GGAAAT': 36996, 'ACAGCC': 36988, 'TCATGC': 36921, 'AAGCCC': 36874, 'AGCGAG': 36857, 'TCGTCT': 36852, 'CTCAGC': 36842, 'TGCGTG': 36794, 'CAGCTT': 36780, 'AAGACG': 36777, 'AACTCG': 36760, 'CTTCCC': 36744, 'AGCATC': 36740, 'GCTCTC': 36729, 'CGTGAA': 36711, 'GGACAC': 36692, 'ATTTCG': 36688, 'GATCTT': 36680, 'GTACGC': 36671, 'AGACGG': 36663, 'GGCTCA': 36632, 'ACCAAG': 36624, 'GACACG': 36567, 'GAGCGT': 36495, 'GTTGCC': 36483, 'CTCCTC': 36460, 'TGGGAC': 36447, 'GCAAAA': 36441, 'AACGTC': 36422, 'TGGTTC': 36367, 'GATTTT': 36362, 'GGATCC': 36341, 'GTGAGC': 36328, 'CACGTC': 36291, 'TCGCGT': 36285, 'ACATGG': 36239, 'ATGGCA': 36174, 'GCTTGC': 36165, 'CGTTCA': 36164, 'AAGAGC': 36159, 'ACTGGT': 36059, 'AGCCAG': 36051, 'CGAAAA': 36034, 'CAACCA': 35991, 'TTTCGC': 35991, 'AGATGC': 35987, 'ACCTGT': 35974, 'CTTCCA': 35959, 'AATGGC': 35954, 'GGGATG': 35939, 'ACGGGG': 35866, 'AGGGGC': 35865, 'GCAACT': 35843, 'ATGTGG': 35819, 'GCTCCT': 35812, 'AGGGCT': 35810, 'TCGTGA': 35810, 'AGCGGA': 35789, 'AGCTTC': 35772, 'AGCGGT': 35763, 'GCTTCT': 35729, 'GAACGA': 35718, 'ACCGAA': 35676, 'GCCTAT': 35658, 'GCGTCT': 35653, 'ATGAGC': 35643, 'GGGAGG': 35637, 'AGGCGT': 35610, 'CACACC': 35607, 'TTGAAG': 35603, 'CAACAC': 35586, 'GCCCAA': 35526, 'ACACCA': 35510, 'AGACGC': 35505, 'CCTTGC': 35489, 'GCTCAT': 35483, 'ACTACG': 35482, 'GAAGCA': 35473, 'GCACGT': 35469, 'CGGGTT': 35454, 'ACTGCG': 35388, 'CATGAT': 35373, 'CGAAGT': 35334, 'GGCTAT': 35299, 'CACGAG': 35277, 'GACCAA': 35270, 'GCTCCA': 35268, 'CTGCGA': 35215, 'TATCGC': 35203, 'TCGGAC': 35191, 'GTTCAC': 35154, 'GAAAAT': 35130, 'GCGATT': 35103, 'GGACCA': 35075, 'CCGTCT': 35073, 'TTCGAT': 35045, 'CGATTT': 35026, 'GTTGCG': 35018, 'GCTTTC': 34999, 'ATCGGT': 34975, 'GCGAAA': 34952, 'GTTCAA': 34914, 'ATTCCG': 34911, 'GGGACC': 34888, 'AAACCG': 34841, 'ACCATG': 34829, 'GAAAAC': 34807, 'CAGTGG': 34784, 'GGTCTC': 34783, 'AAAAAC': 34779, 'GTGTGG': 34733, 'CTCTAC': 34731, 'ACACGC': 34713, 'ACGCAG': 34698, 'CTCCAG': 34654, 'GAACCT': 34600, 'TCCTCA': 34592, 'AGCTCA': 34579, 'GACGCA': 34578, 'CCCCTG': 34578, 'CTCTGG': 34576, 'ACAACC': 34566, 'ACCTCA': 34529, 'GCCCAT': 34522, 'ATGGAA': 34518, 'GTACCG': 34512, 'CATGAC': 34498, 'ACGTCC': 34460, 'TTCTGC': 34428, 'TGGGGG': 34422, 'GGTCCT': 34421, 'GAACAA': 34399, 'CATTGC': 34373, 'GCCAAT': 34364, 'TCAGGC': 34325, 'GTTGGC': 34313, 'CCAGAA': 34282, 'CCGAAA': 34199, 'ATCAGG': 34194, 'ACGAAC': 34189, 'CGCAGT': 34155, 'GCTCAG': 34144, 'AGAGCG': 34092, 'GGATTT': 34081, 'CAGAAC': 34075, 'GACCTT': 34049, 'GGTTCC': 34023, 'CGGTTT': 33997, 'TTTCGG': 33990, 'TTGCAG': 33985, 'ACCCTC': 33970, 'CAGGCA': 33931, 'AACAAC': 33876, 'AAAGCC': 33869, 'TATCTG': 33837, 'GTACGG': 33837, 'CATCTC': 33833, 'CCAAGA': 33819, 'GCCATT': 33809, 'GATTGC': 33783, 'CCTACC': 33778, 'CCATGA': 33776, 'CTGCCT': 33762, 'GCAGTG': 33751, 'TGGTCT': 33751, 'CAATGC': 33750, 'AACGAG': 33744, 'TGTGGG': 33743, 'TTGGCC': 33724, 'GCAATG': 33713, 'TGCTCT': 33687, 'GCTCTG': 33681, 'CCGAGT': 33669, 'CATCAG': 33668, 'TCCATC': 33657, 'GGTTTC': 33598, 'ACTTCC': 33591, 'GTCATG': 33581, 'GCCAGT': 33550, 'AACAAG': 33539, 'CTGCAT': 33532, 'TTCATG': 33500, 'CCAACA': 33472, 'GTTCAT': 33469, 'GGCAGT': 33458, 'AAGGCA': 33453, 'TGTGGA': 33433, 'GCTTTG': 33426, 'GCGCTA': 33422, 'TCTACC': 33418, 'GTCCGA': 33417, 'GTGAAC': 33417, 'CTCTTC': 33400, 'AGGAAA': 33385, 'ATGAAC': 33378, 'GAAGTC': 33253, 'AACCAG': 33244, 'GGCACT': 33232, 'ACTGCT': 33215, 'CTGAAA': 33211, 'GGAGTT': 33123, 'GCTGAG': 33073, 'CTCTCC': 33042, 'GGGTTC': 32949, 'CGCTTT': 32929, 'AGATGA': 32924, 'TACGAG': 32911, 'ATGAAA': 32899, 'GCCTCA': 32872, 'ACTCGC': 32871, 'GGGGAG': 32863, 'CTTGGC': 32846, 'GAGCAA': 32826, 'TCGTTC': 32821, 'CACGCA': 32818, 'GGGGAC': 32816, 'GCATTG': 32741, 'TGGGAG': 32723, 'AGTGGC': 32719, 'TTGCCC': 32626, 'GTCCAG': 32616, 'ATCCCC': 32605, 'ATATCG': 32600, 'ACCTCC': 32592, 'AACGTG': 32587, 'TGTTGC': 32568, 'CCCATC': 32559, 'GTTGCT': 32518, 'AATCCG': 32513, 'CGGTTG': 32501, 'GGGGAT': 32499, 'CTCGGA': 32485, 'GCCTTT': 32433, 'GAACCC': 32388, 'CCCTTC': 32383, 'ACGGAC': 32317, 'TGCTCC': 32307, 'CTTTGC': 32306, 'AGCCGT': 32301, 'CTACAA': 32287, 'CGACTA': 32237, 'CTGTTG': 32230, 'AACCTC': 32213, 'TACACC': 32201, 'CTTCTC': 32179, 'ATTTTT': 32171, 'TTTGGC': 32168, 'CGCACT': 32121, 'GTTTCG': 32118, 'CCAGAC': 32073, 'CATGAA': 32065, 'GGTTGC': 32061, 'AGCTGT': 32059, 'ATCGCT': 32040, 'TCCCTG': 32032, 'GGACAG': 32025, 'ATTGGC': 31936, 'GCAATC': 31926, 'TGAACC': 31925, 'TTGAAA': 31916, 'AAAGCG': 31912, 'GTCTCC': 31903, 'GCTCCC': 31890, 'ACAGCA': 31887, 'TGAACA': 31882, 'GTCGGA': 31865, 'ACTCGG': 31851, 'CGGTAC': 31846, 'ACTGCC': 31845, 'GTCTAC': 31811, 'AATGCC': 31785, 'TCGCAC': 31776, 'TCCGTC': 31762, 'ATCTCC': 31743, 'CTTCTT': 31721, 'CGACTC': 31717, 'GAGAGC': 31648, 'TGCCCC': 31612, 'AGCAAC': 31601, 'ATGCCA': 31551, 'ATTCGC': 31537, 'GTTCTG': 31526, 'TGCGTC': 31492, 'CCTCCA': 31479, 'CTCTCG': 31463, 'CTGCCA': 31456, 'CACTGC': 31428, 'TGAGCA': 31419, 'TTCGGT': 31408, 'TGCGAC': 31400, 'CTGTGC': 31368, 'ACAAGA': 31341, 'GCGTAC': 31330, 'CGTACG': 31306, 'CGGATT': 31300, 'TCCGAG': 31278, 'TTGCCA': 31262, 'CCGCTA': 31251, 'TGAGCC': 31244, 'ATGTTC': 31240, 'AGCAAG': 31231, 'GAGCCT': 31217, 'CTCCAC': 31190, 'ACTTCA': 31173, 'CTTCTG': 31173, 'GGGAGA': 31145, 'AAAAGG': 31138, 'TTCTCC': 31106, 'AAAATT': 31092, 'GTCGCA': 31091, 'CGAAAC': 31079, 'TGTGCG': 31043, 'CGCACA': 31002, 'GGAACA': 30969, 'GAGGGG': 30962, 'GATGTT': 30923, 'GGTACG': 30917, 'CCATCT': 30908, 'GAACGT': 30880, 'GCGAGT': 30874, 'GGGAAG': 30860, 'AACTTC': 30859, 'CGTGTT': 30858, 'AGACCA': 30853, 'ACACGG': 30846, 'TTTCGA': 30838, 'AAAACC': 30825, 'AGCGAT': 30820, 'AAATCG': 30815, 'GAAATT': 30782, 'CGATTC': 30761, 'AGGCTG': 30757, 'CGTTCT': 30739, 'GTCCAC': 30723, 'CCCAAG': 30721, 'GATATC': 30711, 'ATTTCC': 30656, 'TTTCTG': 30651, 'GGAAAC': 30628, 'GCTTTT': 30622, 'GACTGC': 30583, 'GCATTC': 30581, 'GAGGAT': 30579, 'GTGGGA': 30573, 'CAGGAT': 30565, 'CCTCCT': 30536, 'TCAGCT': 30510, 'AAGTGG': 30501, 'GAACAT': 30479, 'CCAGTT': 30443, 'TATGGC': 30433, 'ATGGCT': 30375, 'ACCTAC': 30332, 'CAGGCT': 30328, 'CTCGCA': 30322, 'GAGCAT': 30312, 'TTGATC': 30309, 'TTCACG': 30301, 'CGTTTC': 30295, 'ACGTGA': 30288, 'TACCCG': 30270, 'CAATGG': 30253, 'GGGGTC': 30247, 'ACCAGA': 30228, 'GGAACC': 30212, 'TATTTT': 30207, 'AGACCC': 30188, 'CAGACG': 30174, 'TGATGT': 30138, 'CGTCTT': 30112, 'GGCTTG': 30108, 'TCCAGA': 30088, 'CTGTAC': 30084, 'CGAAAT': 30082, 'ATGTGC': 30079, 'TGCCAA': 30043, 'CACGGA': 30042, 'TGGCAT': 30037, 'GCAGTC': 30036, 'AAATGG': 30014, 'GAGTGG': 30000, 'CAGGGG': 29982, 'TTTTTG': 29980, 'GAAATG': 29947, 'TCGAAA': 29937, 'GGCACA': 29927, 'GGTCTT': 29908, 'ATCCAC': 29894, 'CCGTAC': 29847, 'TTGGTG': 29842, 'CGCTTG': 29809, 'ATCGCA': 29801, 'GTTCCA': 29800, 'CAGAGC': 29797, 'TATGCC': 29757, 'AAGTCG': 29753, 'AAAATC': 29745, 'GCTACA': 29742, 'CGAGTC': 29730, 'TCCTGT': 29715, 'AAAGGC': 29700, 'AGGAAC': 29677, 'GAGCTT': 29668, 'GTTCCC': 29667, 'TGCATC': 29665, 'ACAACA': 29647, 'GGGGGA': 29614, 'GAACAC': 29588, 'TACCAG': 29584, 'ATCTAC': 29572, 'AACGGG': 29555, 'CTACGT': 29549, 'CAACAG': 29547, 'AACAGG': 29541, 'CTACAC': 29535, 'TGCCAT': 29531, 'GCCCCT': 29528, 'ATTTTG': 29523, 'TAAAAA': 29491, 'GGAAGT': 29456, 'TGATTT': 29444, 'GGATGT': 29441, 'GCCCTT': 29438, 'TCCGTG': 29421, 'TACTGG': 29354, 'TTTTGC': 29349, 'CCTGTG': 29335, 'CGCAAT': 29312, 'TTTTTC': 29283, 'TCGGGA': 29282, 'CTATGC': 29262, 'CCGTTG': 29248, 'GAGTCG': 29237, 'AGAACA': 29233, 'CCCTGT': 29229, 'GCTATC': 29206, 'AGAACC': 29195, 'GAATGG': 29193, 'TTGTCG': 29187, 'GCCAAA': 29176, 'GGCCTA': 29169, 'AAGCCA': 29147, 'GATTCC': 29139, 'TGCCAC': 29136, 'GGAATT': 29114, 'CAGTCG': 29107, 'GCAAAG': 29093, 'ACTACC': 29081, 'GCGTTT': 29056, 'TCTGCA': 29049, 'CCGATT': 29048, 'GAATTC': 29028, 'ACGCAC': 29018, 'AGCAGT': 28972, 'TTCCAC': 28964, 'GTTGAT': 28964, 'CAATCC': 28921, 'TCATCT': 28919, 'CACATC': 28912, 'CCTTCT': 28895, 'ACACCT': 28892, 'ACCCGT': 28888, 'AAAATG': 28867, 'ACGACT': 28865, 'CTCACG': 28855, 'TCCTCC': 28813, 'TGCCAG': 28811, 'TTTATT': 28791, 'CCTATG': 28736, 'TTTTGA': 28728, 'TTCCCC': 28724, 'TCCTTC': 28693, 'GACTCG': 28666, 'TGTTCT': 28648, 'CATGCA': 28619, 'CGCATT': 28599, 'CAAAGC': 28599, 'GAGTAC': 28584, 'AGTTCC': 28573, 'ATTGAA': 28557, 'TCTCCA': 28551, 'GTTCTT': 28522, 'TTTTTT': 28511, 'TATCCG': 28450, 'GGGTCA': 28417, 'GGCTCT': 28405, 'GCACAC': 28403, 'CTATCT': 28393, 'TCATGA': 28383, 'GCACAG': 28371, 'AGTCGG': 28358, 'TTTGAA': 28355, 'AGGATG': 28343, 'GTACGA': 28321, 'CGCAAA': 28316, 'ACAGGC': 28271, 'TGAAAT': 28260, 'TTCTAC': 28258, 'TTTTAT': 28250, 'CATTCC': 28224, 'ATGGAT': 28214, 'GGGTGT': 28212, 'GATATT': 28190, 'CCTCCC': 28178, 'TTCAGG': 28154, 'TCAGGA': 28136, 'AAAGAT': 28124, 'AGTCGC': 28124, 'TGTGGT': 28116, 'CGATTG': 28095, 'TGGCTT': 28087, 'GGATTC': 28074, 'CCACGT': 28057, 'TCAATG': 28025, 'CGTGTC': 28023, 'ATTGAT': 28000, 'GGACTT': 27998, 'ATTGCG': 27978, 'CATCTT': 27957, 'GGTTTT': 27948, 'ATGGGG': 27943, 'CCCAAC': 27940, 'GCTATG': 27928, 'GAAAGA': 27920, 'TGGATT': 27914, 'TCGGGT': 27914, 'AAAATA': 27899, 'TCGGAG': 27898, 'AGTACG': 27874, 'CACAGC': 27857, 'GGGAAA': 27826, 'GTGGGT': 27816, 'CTACCA': 27802, 'TATCGA': 27783, 'TGGGTC': 27782, 'ATTATT': 27766, 'TGCAAC': 27750, 'TCTACA': 27748, 'CCCATG': 27738, 'CTGCTT': 27736, 'GTTCTC': 27707, 'TACGTC': 27703, 'ACATGC': 27701, 'GGGAAC': 27698, 'GATTCG': 27692, 'TTATTT': 27680, 'CGATAT': 27640, 'GCTTGG': 27640, 'CCCGTT': 27621, 'GCGTGT': 27609, 'AAGTTC': 27598, 'TGTACG': 27581, 'TCAAAA': 27575, 'ACGTTC': 27569, 'CCCCGT': 27567, 'TCTTCT': 27553, 'ATTTTC': 27551, 'TTTTCC': 27532, 'CCAATG': 27503, 'GCACAA': 27502, 'GGAAAG': 27490, 'TTTGAT': 27398, 'GGATAT': 27398, 'CGTACC': 27367, 'TATTTC': 27363, 'CTTCAG': 27362, 'CCGGTA': 27357, 'CGTTGG': 27341, 'GCCCCA': 27340, 'AATCGC': 27311, 'TGCGAT': 27310, 'GAAACG': 27309, 'AAAACA': 27298, 'AGACCT': 27286, 'CAGTTG': 27258, 'TATCGG': 27254, 'GCAGAG': 27252, 'TGAGCT': 27239, 'ACTCGA': 27225, 'CTCCCC': 27206, 'AGTCCG': 27200, 'AGAAAT': 27177, 'GTGCGT': 27129, 'ACTCCG': 27126, 'TTGGCA': 27097, 'ACGCGT': 27061, 'CTTTCC': 27060, 'AAACGC': 27033, 'CAGAAA': 27017, 'TGAAAG': 27007, 'TACATC': 26993, 'AGTTCA': 26991, 'TGGAAT': 26990, 'TTTTGG': 26982, 'TCACGA': 26933, 'TCGAGT': 26930, 'GTTTTT': 26901, 'TCGATT': 26881, 'AGGATC': 26869, 'CCCTCC': 26868, 'ACAGCT': 26834, 'TGGAGT': 26831, 'GTGGTT': 26823, 'AAGAGG': 26789, 'GCGAAT': 26782, 'ATTTCA': 26781, 'GCCACA': 26752, 'GTTTCC': 26719, 'TACAAC': 26718, 'TGATTG': 26714, 'CCCTCA': 26708, 'AGCGAA': 26692, 'TGCACA': 26676, 'CTGATT': 26673, 'TTTATC': 26651, 'AGCCTC': 26638, 'CCTCAT': 26630, 'CGCTAT': 26613, 'TGGTTT': 26603, 'TTTTCG': 26590, 'GGGACA': 26588, 'ATCAAT': 26584, 'CTTGCG': 26578, 'TTCGCT': 26574, 'TGTGCT': 26566, 'GCTCTT': 26538, 'ACCCAC': 26525, 'GGTTCA': 26514, 'AAATAT': 26492, 'CACGTG': 26485, 'CAAAGA': 26482, 'TGAACT': 26478, 'CGAATT': 26474, 'ATCAAA': 26449, 'ACGGGA': 26442, 'ACCCCC': 26406, 'GCGGTA': 26405, 'CGGAGT': 26389, 'GTATCG': 26375, 'CCTTTG': 26367, 'TGGGAA': 26360, 'TCTCCC': 26355, 'ATTCTG': 26342, 'AAATCA': 26237, 'TACTTC': 26233, 'CGGTAT': 26205, 'ATACCG': 26197, 'GATTAT': 26189, 'CTGTTT': 26158, 'CCGAAT': 26158, 'TGTGCC': 26120, 'GCTTGA': 26110, 'GAGAAA': 26109, 'AGACGA': 26092, 'GCAATT': 26079, 'AAAACG': 26064, 'AGCATG': 26039, 'TGGTTG': 26011, 'CACTTC': 26008, 'TGCCTT': 25975, 'TTTCAA': 25939, 'CTTTTT': 25908, 'GGTTGG': 25905, 'GTCAGG': 25901, 'TTACCG': 25897, 'GAATTT': 25882, 'GAATCG': 25867, 'AGGTCC': 25867, 'AAGGGG': 25864, 'TGCAGT': 25813, 'AAATCC': 25799, 'GCATTT': 25792, 'CCTTGG': 25787, 'CCATTG': 25769, 'GCAAAC': 25765, 'GCTCTA': 25754, 'GGAATC': 25737, 'CCTACA': 25732, 'TTTTTA': 25717, 'CCCCAC': 25713, 'ATCCGA': 25712, 'AGGCAG': 25677, 'CTTTGG': 25673, 'TGGACT': 25653, 'GTTCAG': 25643, 'TACCTC': 25635, 'GGTCCC': 25634, 'CGAATC': 25624, 'CTACAT': 25599, 'TTCGAA': 25583, 'ACGGGT': 25554, 'AAAGCA': 25548, 'TGAAGT': 25533, 'AGCCCT': 25520, 'GCAAGT': 25476, 'AAGGAT': 25465, 'TCCGAT': 25435, 'GGTATC': 25434, 'CATTTC': 25370, 'AATCTG': 25352, 'ATGCCT': 25348, 'CTTTCG': 25342, 'TGGGGA': 25324, 'CAGTAC': 25318, 'TCAATC': 25312, 'GTGTGC': 25310, 'TCCATG': 25286, 'GGGATT': 25249, 'TACAGC': 25223, 'CAGTGC': 25214, 'AAGCCT': 25212, 'GATTGG': 25185, 'CCTATC': 25183, 'GTGCGA': 25180, 'GGTCCA': 25179, 'GGGCTA': 25155, 'CTCTGC': 25146, 'TACAAG': 25143, 'TTTCCC': 25142, 'GACCCA': 25136, 'TTTGCG': 25136, 'GAATAT': 25129, 'AAGCAC': 25128, 'GGATTG': 25109, 'CCCAGA': 25108, 'GGGGAA': 25103, 'TACGTG': 25085, 'AATATC': 25066, 'ATCTGA': 25037, 'AAAGAG': 25036, 'GCACTC': 25031, 'CAAAAG': 25026, 'GGAGAG': 25006, 'TTTTCA': 24999, 'AAGGCT': 24994, 'CGGAAT': 24989, 'CAACTC': 24960, 'TCAACT': 24934, 'GAATCC': 24911, 'TGCATG': 24908, 'CCTCTC': 24905, 'TTCCGA': 24859, 'TGTCGT': 24855, 'GAAGTT': 24851, 'AAACTG': 24841, 'ATTCGG': 24824, 'AAAGCT': 24819, 'GTTTGC': 24815, 'AACCCC': 24809, 'ACGGAG': 24808, 'GCACAT': 24801, 'AAACGG': 24770, 'AACGCT': 24761, 'GAGTTG': 24743, 'GCTGTA': 24722, 'GGTTCT': 24720, 'ATCCGT': 24705, 'CGTTGA': 24695, 'AGCCAC': 24689, 'AATTTT': 24687, 'TTGACG': 24675, 'GTTGAA': 24674, 'TGTTGA': 24670, 'AATATT': 24662, 'CAATCT': 24653, 'CGAAAG': 24609, 'AAATTT': 24604, 'TTTCAG': 24598, 'GACGTT': 24569, 'AGATCT': 24555, 'CATGTC': 24553, 'GTCCCC': 24550, 'CCTGAG': 24546, 'TTAAAA': 24466, 'TCCGAA': 24463, 'GTGTTG': 24459, 'GCCGTA': 24451, 'GTGTCC': 24408, 'GAATTG': 24405, 'TTGGGC': 24403, 'ATTCCC': 24378, 'GGCGTA': 24378, 'TGCTTC': 24373, 'CCGTTT': 24340, 'AGCCCC': 24337, 'AAACCC': 24333, 'TCGGTT': 24290, 'AAGCGA': 24288, 'CACCTA': 24266, 'AGGCAC': 24266, 'GACTCC': 24260, 'AGGTGT': 24240, 'ACGTCT': 24240, 'AGAGCA': 24236, 'ACCTTG': 24234, 'ATTGCT': 24208, 'TTTCCT': 24207, 'AGCTTT': 24207, 'GTCGTT': 24207, 'ATTTTA': 24204, 'CGAGTG': 24198, 'AGGTCT': 24197, 'AGCAAA': 24190, 'TTGCTC': 24187, 'ACGAGT': 24156, 'AGTTGC': 24152, 'TCAAAG': 24107, 'CGTGAG': 24100, 'CCGTGT': 24091, 'TGAAAC': 24084, 'ATAAAA': 24080, 'CCAACT': 24068, 'TTTCCA': 24047, 'GCATGT': 24043, 'ATTATC': 24040, 'AGGCAA': 24035, 'ATTTAT': 24028, 'CTATCG': 23987, 'CCAGTC': 23982, 'AGCCAT': 23981, 'AACGGT': 23973, 'AGTGGG': 23972, 'GAACTT': 23963, 'TGCAAA': 23946, 'TTTCTC': 23937, 'AGCCCA': 23925, 'TTATCG': 23905, 'AATGCG': 23895, 'CACGAA': 23892, 'TCAGGG': 23881, 'GGACTC': 23870, 'AACATG': 23855, 'GTGCCT': 23817, 'TGGGAT': 23812, 'TGTTGG': 23808, 'AGTGGA': 23806, 'GAATAC': 23795, 'TCCCGT': 23794, 'AATCGG': 23794, 'TGCAAG': 23792, 'CCAGTG': 23780, 'TTTTCT': 23771, 'GGGTCT': 23763, 'AATTGC': 23763, 'CTGGTA': 23759, 'CAATAT': 23746, 'GATTAC': 23739, 'CAGGGA': 23735, 'GGAGTC': 23715, 'CGATAC': 23700, 'CCCCAG': 23668, 'GTTGCA': 23646, 'GAGCCA': 23621, 'AAACAG': 23619, 'CAACTT': 23593, 'GTTTTC': 23591, 'TTCAAA': 23579, 'CAAACC': 23576, 'GGGTCC': 23560, 'CATTGA': 23546, 'CCATTC': 23517, 'ATTACC': 23508, 'CTACCC': 23507, 'AGAAGT': 23492, 'GTGAAA': 23492, 'AAATTG': 23489, 'GATACC': 23484, 'TTGGAA': 23470, 'CACTCG': 23426, 'AGGACT': 23425, 'GCAAAT': 23421, 'AGAGCC': 23397, 'TCCAAC': 23388, 'CATTGG': 23383, 'AACGAT': 23376, 'TGTCCA': 23373, 'ACCAGT': 23371, 'AACACG': 23367, 'TTGGCT': 23351, 'CTTGAT': 23343, 'AAATTC': 23326, 'AGCTCC': 23322, 'CCACAC': 23312, 'TATCAG': 23310, 'CAAAAT': 23294, 'TTGTGG': 23289, 'TATTGC': 23277, 'GGTATT': 23272, 'TACGGG': 23260, 'GTGCCA': 23252, 'CAATCG': 23235, 'AACTAC': 23235, 'CTTGCT': 23234, 'TCTGAT': 23221, 'TGCTTG': 23208, 'GGGTTG': 23196, 'AAGAAT': 23191, 'ACAAAA': 23185, 'TTGACC': 23177, 'CAAGTC': 23138, 'CTATGA': 23131, 'CCACTG': 23122, 'AATGAA': 23113, 'AAGTGC': 23110, 'TCTATG': 23110, 'CTTGAA': 23106, 'TTATTG': 23104, 'AATTCG': 23089, 'ACGGAT': 23078, 'GGGGGT': 23061, 'CCAATC': 23033, 'CGTCTA': 23032, 'CTCCAT': 23002, 'GGAATG': 22991, 'GAATGC': 22989, 'TGCTTT': 22967, 'CAAGAG': 22963, 'GTTTTG': 22934, 'AACGCA': 22928, 'TTGAGC': 22920, 'CCATGT': 22918, 'CTCCAA': 22912, 'GCCACT': 22909, 'GAGGCA': 22902, 'CCCCTC': 22877, 'GTACCT': 22873, 'TCTATC': 22870, 'ACTGCA': 22866, 'AGAACT': 22857, 'CATTCG': 22841, 'ATGATT': 22839, 'CGAATG': 22838, 'ATGCAC': 22836, 'CCTTGA': 22826, 'AGAAAG': 22819, 'AGAGCT': 22803, 'AGGAGT': 22790, 'CAGTCC': 22776, 'TTGCCT': 22775, 'CAATGA': 22769, 'TCTCCT': 22751, 'CACAAC': 22749, 'GTCCGT': 22748, 'ACGAAA': 22716, 'TCGCAA': 22708, 'TCCAAG': 22707, 'GCCTCT': 22704, 'CAAAAC': 22643, 'AATTTC': 22641, 'AACTGA': 22624, 'TGCCTC': 22618, 'CAAAGG': 22600, 'CCCCAA': 22592, 'CCTTTC': 22576, 'TTTACC': 22569, 'GAGTCC': 22559, 'ACGGTT': 22547, 'GGAGTG': 22531, 'AATGGA': 22497, 'CGAGAG': 22496, 'ACGGAA': 22492, 'GATTGA': 22482, 'CTATTT': 22468, 'TTTGCT': 22467, 'CTTTTC': 22467, 'TTTATG': 22456, 'TGAGGC': 22453, 'CTCATG': 22444, 'TGCGAG': 22430, 'TTCTTT': 22425, 'AAGCAA': 22406, 'AGCAAT': 22406, 'AGTCGA': 22384, 'AACGAA': 22358, 'ACATCT': 22348, 'GTGATT': 22330, 'AAACAA': 22303, 'ATTTCT': 22282, 'AACCAC': 22279, 'ATATTT': 22273, 'CGTTTT': 22271, 'GATTTG': 22268, 'TTGTTG': 22260, 'TATCAA': 22252, 'AGGAAT': 22223, 'CACAAG': 22201, 'CATTTT': 22194, 'TCGGAA': 22188, 'GAACCA': 22187, 'TGGTAC': 22179, 'TATGCG': 22163, 'AATGGG': 22160, 'GCTATT': 22141, 'CCAAAG': 22129, 'CTTTAT': 22127, 'CTTTGA': 22121, 'GTACTG': 22117, 'TTCGCA': 22108, 'ACATGA': 22103, 'TTTAAA': 22073, 'ACTTCT': 22062, 'ATCATT': 22058, 'TGCGTT': 22042, 'CAAGTT': 22042, 'GGGACT': 22034, 'AGCTTG': 22031, 'AGCTAC': 22019, 'AACAAA': 21997, 'TGGGGT': 21996, 'AGTGGT': 21995, 'AGGGTG': 21976, 'CAGCTA': 21976, 'ATATTG': 21973, 'TGACGT': 21967, 'AGGCAT': 21958, 'AAATGC': 21918, 'TCAGGT': 21918, 'CTATGG': 21912, 'TGAGGA': 21891, 'TGTTTG': 21869, 'ACTACA': 21851, 'ACCGTT': 21850, 'TCGCTT': 21843, 'GGTTGA': 21822, 'ACGATT': 21820, 'AAACCA': 21816, 'TATTGA': 21812, 'TTTCTT': 21802, 'AGCTCT': 21787, 'AAGATT': 21777, 'ATTTGC': 21719, 'GTTTCT': 21717, 'GGTTTG': 21687, 'GGTACC': 21680, 'TCGGAT': 21661, 'CTCCGT': 21641, 'TCGAAT': 21628, 'TGTTTT': 21616, 'AATCGA': 21613, 'GGGTTT': 21607, 'CAATTT': 21605, 'ACCCCA': 21603, 'TTGATT': 21587, 'CTTTTG': 21559, 'GTGCAT': 21555, 'CTTTCT': 21547, 'TCGTTG': 21520, 'AAAACT': 21516, 'TTTCAT': 21480, 'TCATTG': 21475, 'AAGAGA': 21475, 'TGGATA': 21472, 'AGCACA': 21472, 'GTTGGA': 21472, 'TGATTC': 21462, 'TACCGA': 21459, 'AATCAA': 21438, 'AACCGA': 21431, 'GGGGTT': 21404, 'GAGGCT': 21400, 'CTATCC': 21395, 'AAAGAC': 21393, 'AGCCAA': 21363, 'AGTGCG': 21356, 'GATAAA': 21327, 'ATCGGA': 21324, 'CTCCCT': 21288, 'ATTCAG': 21271, 'TTCGGA': 21264, 'GTTTCA': 21261, 'ATGTCC': 21239, 'CCAAAA': 21221, 'ACCTCT': 21213, 'TTCCGT': 21196, 'TCGTGT': 21183, 'ACGCTT': 21176, 'ATGAGG': 21170, 'TCCTCT': 21170, 'GATTCT': 21163, 'GTTGGT': 21149, 'GCACTT': 21148, 'AGATTT': 21134, 'TTATCA': 21123, 'AAATTA': 21120, 'GCTACT': 21113, 'ATGCGA': 21083, 'TGAATT': 21042, 'GGGAAT': 21033, 'CCACAA': 21021, 'AATATG': 21020, 'GCTTTA': 21012, 'CCCTAC': 21006, 'TTGAAC': 20988, 'TATGAA': 20971, 'GACAGG': 20969, 'TATCTC': 20965, 'CAGGGT': 20945, 'TCCCCA': 20858, 'ATTACG': 20853, 'AATTAT': 20849, 'GAGATT': 20837, 'GTCAAT': 20834, 'AGCCTT': 20829, 'TCCCCC': 20828, 'TCGATA': 20803, 'AAATGA': 20795, 'ACTGAT': 20769, 'CCGATA': 20749, 'TTTGGT': 20739, 'AATACC': 20735, 'TTTTAA': 20729, 'TCTTTG': 20715, 'TCTCGT': 20696, 'AGTACC': 20685, 'AAAAGT': 20678, 'AAAGGA': 20667, 'AGGGGG': 20650, 'CGTGTG': 20649, 'ATTCAA': 20621, 'TGTTTC': 20618, 'GGTCTA': 20617, 'CAGTTT': 20600, 'AGTGCC': 20566, 'ATGAAT': 20562, 'CGTATC': 20556, 'CAATCA': 20542, 'AGGATT': 20526, 'TAAAGA': 20524, 'ATGGGA': 20516, 'CATTCT': 20509, 'ACGCAT': 20496, 'TATTCC': 20492, 'TATTGG': 20458, 'GTTGTT': 20443, 'ATTCGA': 20441, 'ATTGGT': 20432, 'GTGAGG': 20425, 'CCTCAG': 20419, 'CATGTG': 20410, 'GTTGGG': 20397, 'TGCACT': 20357, 'GACAAA': 20352, 'ATATGG': 20351, 'ATTCCT': 20328, 'AACCGT': 20318, 'ACCACA': 20314, 'CTCAAT': 20313, 'TTTGGA': 20310, 'GAGTTT': 20306, 'GAAACA': 20305, 'CGAGTA': 20290, 'CGGGTA': 20282, 'ACAACT': 20278, 'AGTTGG': 20277, 'ATTCTC': 20264, 'AGAAAC': 20261, 'GTATTT': 20250, 'CTCAAA': 20234, 'CTCGTT': 20230, 'TGTCCT': 20229, 'ATCTTT': 20217, 'AATGAT': 20201, 'AATTGA': 20195, 'AAACCT': 20167, 'ATTGCA': 20155, 'CCCCCA': 20133, 'TTTCAC': 20128, 'TCAAAC': 20101, 'TACTAC': 20067, 'AAGTAC': 20064, 'TTATGG': 20058, 'TCTTTT': 20037, 'AACTCC': 20029, 'TTTTAC': 20023, 'GATTTA': 20022, 'TTCAAT': 20012, 'CTCAGG': 20011, 'AGCGTT': 20002, 'AATTCC': 19983, 'TCATTC': 19982, 'GTCCAT': 19974, 'ACGCAA': 19972, 'CCTCTG': 19963, 'TGTACC': 19960, 'CAGATT': 19950, 'AGGCTC': 19939, 'CTATCA': 19937, 'TTTCGT': 19918, 'TATTAT': 19917, 'GACAAT': 19914, 'CATGAG': 19899, 'AATTGG': 19890, 'GTTTGG': 19873, 'TCTTTC': 19871, 'TCGCAT': 19858, 'CTATGT': 19855, 'CAAACG': 19845, 'GTTTGA': 19819, 'CCACAT': 19815, 'AAGCTT': 19811, 'ATTATG': 19805, 'AGGGGA': 19804, 'AATACG': 19803, 'ATATCA': 19789, 'ATCGTT': 19766, 'AATCAG': 19757, 'GGGTAC': 19752, 'TTTGCA': 19752, 'CAATTG': 19721, 'AATCCC': 19718, 'TTCCAT': 19715, 'GCAGTA': 19713, 'GTGCTT': 19698, 'TCTTGC': 19697, 'TTGTTC': 19688, 'AGGCTT': 19686, 'ACCATT': 19679, 'GAAATA': 19679, 'CAATGT': 19676, 'CCCCCT': 19664, 'GACATT': 19653, 'ACCTAT': 19651, 'TATTCG': 19646, 'GATCTA': 19640, 'CCTCTT': 19637, 'TTATTA': 19631, 'CATGTT': 19626, 'GATTCA': 19601, 'TCTGAA': 19558, 'AGCATT': 19541, 'AATGGT': 19540, 'GCGATA': 19524, 'ATGGTT': 19524, 'CAAATC': 19507, 'GTGCAA': 19497, 'CATCTA': 19490, 'GATATG': 19469, 'AGAATT': 19464, 'CGCGTA': 19461, 'ACAGGA': 19460, 'GGATAC': 19450, 'ATATTC': 19421, 'ATTTGG': 19417, 'CACTAC': 19402, 'TTGCAA': 19400, 'ATCTCA': 19386, 'ACCTTT': 19374, 'AAGTCC': 19351, 'ACCAAT': 19344, 'AAGACA': 19341, 'GGTTAT': 19336, 'GAGTGC': 19313, 'AATTCA': 19313, 'GATACG': 19302, 'TGTGCA': 19276, 'ATATCC': 19270, 'CCCTTG': 19261, 'CGTTTG': 19243, 'CACGTT': 19237, 'TTATCT': 19208, 'AGTTCT': 19202, 'CCTTTT': 19187, 'TTGGAG': 19179, 'CTTCTA': 19166, 'TCCAGT': 19159, 'ATTAAA': 19143, 'GAAACT': 19128, 'TTATGA': 19119, 'CTTTCA': 19104, 'AAGCGT': 19086, 'GTCTAT': 19067, 'TCTCAA': 19049, 'CCATTT': 19032, 'AACAAT': 18997, 'AGATGT': 18973, 'GGGAGT': 18971, 'CGATTA': 18970, 'ATGTTG': 18942, 'TTTTGT': 18937, 'TTACGG': 18937, 'GTTTAT': 18933, 'AAATAC': 18915, 'CAAATT': 18912, 'ATCTGT': 18904, 'TTGAGG': 18901, 'CACAGG': 18898, 'TTTGTC': 18866, 'CATTCA': 18852, 'ACACGA': 18852, 'ATACGG': 18851, 'GTACGT': 18849, 'CTACTG': 18842, 'AGAGGC': 18834, 'AGTTTT': 18833, 'TCCCTC': 18833, 'ATTTGA': 18830, 'AACGGA': 18822, 'CCAGAG': 18816, 'TGTCCC': 18802, 'TCTGTT': 18793, 'TCCCCT': 18751, 'CTACTT': 18746, 'AAAGTG': 18745, 'TACCAC': 18738, 'TTTGGG': 18728, 'TGCAAT': 18703, 'AATTTG': 18702, 'ATGGGT': 18695, 'GTCAAA': 18689, 'TACTGC': 18683, 'TCACTG': 18673, 'TCCGTT': 18665, 'GACTAT': 18664, 'GTATTG': 18649, 'TGGTAT': 18648, 'TATGAC': 18643, 'TGAATG': 18612, 'AATTTA': 18589, 'TTCTGA': 18588, 'GGGGGG': 18559, 'ATCCAT': 18533, 'GTTGTC': 18517, 'TGGGTT': 18513, 'GTTGAC': 18509, 'GAAAGG': 18466, 'AGCACT': 18456, 'TTGGTC': 18453, 'CTTTAC': 18450, 'TGCGAA': 18446, 'CTCTAT': 18426, 'CGGATA': 18423, 'GCGTAT': 18421, 'CAATAC': 18412, 'TACGAA': 18408, 'ATTGGG': 18382, 'TTTGAC': 18376, 'TCCCAG': 18364, 'TCCACA': 18350, 'TATTAC': 18326, 'CCTCTA': 18315, 'CCAAAC': 18308, 'TCAATT': 18308, 'ATCTTG': 18305, 'TATGGA': 18305, 'TTGGGG': 18289, 'AACAGA': 18278, 'ATCTAT': 18275, 'AATAAA': 18274, 'TTGGAT': 18262, 'CACACG': 18235, 'ATTCCA': 18221, 'CTGAAT': 18207, 'CAGGTT': 18191, 'GTGTAC': 18191, 'GAAAGT': 18190, 'ATTGGA': 18166, 'AGATTG': 18163, 'GTCATT': 18159, 'CTCCTT': 18159, 'CAACTA': 18158, 'TCAAGT': 18148, 'AGTGCT': 18131, 'TTGCGA': 18108, 'CAAACT': 18100, 'ATGCGT': 18063, 'ATGACA': 18058, 'AAGCAT': 18056, 'AAAGGG': 18054, 'GCAATA': 17997, 'TTGGTT': 17979, 'TTTGTT': 17977, 'AATCCT': 17970, 'ATACGC': 17966, 'CAATTC': 17965, 'CCACAG': 17964, 'CCCCAT': 17961, 'TACTCG': 17946, 'TTACGA': 17919, 'CAAACA': 17908, 'TAAAAG': 17898, 'ATGTCA': 17897, 'ACAGGG': 17894, 'TACACG': 17894, 'CTTGCA': 17887, 'CCAAGT': 17856, 'ACCCCT': 17853, 'CCCAGT': 17841, 'ATAAAG': 17828, 'CCTGTA': 17819, 'ACGAAT': 17802, 'TTCATT': 17800, 'AAGATA': 17748, 'AAGGGA': 17735, 'CCCTCT': 17733, 'GGATTA': 17725, 'TTTACG': 17701, 'GCTTAT': 17688, 'TCAGAA': 17683, 'GTCCAA': 17677, 'TATTTG': 17655, 'CTACAG': 17652, 'ACAGGT': 17636, 'AAAGGT': 17610, 'TTATGC': 17582, 'CAAATG': 17577, 'TCTGAC': 17556, 'ACAAAG': 17552, 'TTATTC': 17548, 'TTAAAG': 17529, 'ACGTTG': 17525, 'AGCTAT': 17514, 'TTGATA': 17513, 'GGGTAT': 17493, 'GACTTT': 17489, 'AACCCT': 17460, 'TTACAA': 17456, 'TATTTA': 17447, 'CAAGTG': 17443, 'CGAATA': 17434, 'TACGGT': 17433, 'TATGTG': 17426, 'TGCATT': 17422, 'GTGTTT': 17411, 'ACCAAA': 17406, 'GGAATA': 17393, 'AACTCA': 17388, 'TCATTT': 17385, 'AATGCT': 17384, 'ATTGTT': 17380, 'ATTGTC': 17369, 'AGGGAA': 17349, 'TTCTTG': 17347, 'TTGTTT': 17344, 'TATCCC': 17333, 'TTACGC': 17329, 'ATCAGA': 17326, 'GGTATG': 17321, 'TTCCTT': 17321, 'CTATTC': 17318, 'TTCCAA': 17316, 'CACTCC': 17311, 'AGTTTC': 17304, 'AGATTC': 17298, 'GTTATT': 17295, 'TCCTAC': 17285, 'ATCCCT': 17254, 'TATGAT': 17250, 'CACTGA': 17249, 'GTTTTA': 17241, 'AGAGAA': 17231, 'GTATTC': 17227, 'ACAATG': 17207, 'ATCAGT': 17207, 'TTTGTG': 17204, 'TTGGGA': 17197, 'AAATCT': 17197, 'ACGTGT': 17195, 'CTTGAG': 17190, 'ATTACA': 17189, 'GACCTA': 17184, 'TTCCCT': 17170, 'TCAAAT': 17168, 'AACCAT': 17162, 'ATCCTT': 17160, 'GGTTAC': 17155, 'ATTCAT': 17154, 'CCCCTT': 17150, 'AATCCA': 17147, 'CGTTGT': 17142, 'ATTCTT': 17137, 'TCTTGA': 17121, 'ATGCTT': 17115, 'GTACAA': 17112, 'TCGGTA': 17108, 'CATTAC': 17100, 'TGTGAT': 17083, 'TTATCC': 17081, 'ACAATC': 17080, 'GTATGG': 17076, 'TGATAT': 17063, 'CACATG': 17037, 'GAGGGA': 17036, 'GAAGTA': 17014, 'GAGGGT': 17007, 'ATCCCA': 16992, 'GGTAAA': 16987, 'CATTAT': 16975, 'GCCCTA': 16965, 'GAGAAT': 16964, 'TCGTTT': 16946, 'ATTTAC': 16932, 'CAGTGA': 16929, 'TTTAAT': 16926, 'TTGCTT': 16922, 'CTTGGT': 16914, 'TTCTAT': 16906, 'TATGTC': 16903, 'TCAGTG': 16900, 'ACCACT': 16899, 'CTTTGT': 16894, 'AATCAT': 16887, 'TTCTCA': 16886, 'AGATAT': 16876, 'GGACTA': 16864, 'ACTGAA': 16860, 'TGTTGT': 16845, 'AGAGAT': 16844, 'ATGCAA': 16835, 'ATGTAC': 16812, 'TATTCA': 16789, 'GTCCCT': 16783, 'GAATTA': 16779, 'CGTACT': 16770, 'GGTACT': 16762, 'AAACGA': 16753, 'TTACCT': 16747, 'TACATG': 16738, 'TCCTTG': 16731, 'ACTGAC': 16725, 'TACGAT': 16711, 'AATCTC': 16706, 'CCCACA': 16696, 'ATGTGA': 16694, 'GATTGT': 16631, 'TATCGT': 16627, 'CGGTTA': 16623, 'ACAGAA': 16616, 'ATTTAA': 16611, 'TGAATC': 16601, 'ACAGAT': 16594, 'TGGCTA': 16573, 'AGAGGA': 16557, 'ACATTG': 16517, 'GTACCC': 16508, 'TTCCCA': 16502, 'GTCTTT': 16498, 'TCAGTT': 16490, 'AAGTGA': 16489, 'CCGTAT': 16488, 'ACTTTG': 16486, 'TACAAA': 16483, 'AGTTGA': 16477, 'CTGTAT': 16468, 'ATTCGT': 16467, 'GGTTGT': 16460, 'ACTCCA': 16460, 'ACTATC': 16453, 'GTTATC': 16437, 'ACTCCC': 16433, 'CCTATT': 16393, 'GTCCTT': 16387, 'GTATCC': 16366, 'AGTTTG': 16358, 'AAACTC': 16349, 'ACATTC': 16346, 'CTACTC': 16328, 'AAGGTT': 16327, 'AATGTC': 16326, 'TCTATT': 16324, 'CTTGGA': 16305, 'TCAGAT': 16304, 'TTCAGA': 16272, 'GCATTA': 16271, 'ATATTA': 16270, 'AAATAA': 16256, 'TTGTGC': 16248, 'AACCTT': 16237, 'CCACTC': 16225, 'TGTCAC': 16215, 'CGGTAA': 16211, 'AATTAC': 16208, 'ACCCAT': 16206, 'TTGAAT': 16185, 'TACCGT': 16182, 'AGTACA': 16173, 'CTTGGG': 16159, 'TGACAA': 16159, 'TGTCAT': 16149, 'AAGTTT': 16135, 'TATCAC': 16130, 'GAATCT': 16116, 'GGTGTA': 16102, 'GTAAAA': 16099, 'TCAATA': 16098, 'TTACCA': 16093, 'TCATGT': 16088, 'ACAAGT': 16085, 'ATATGC': 16083, 'CCACTT': 16045, 'TCCTTT': 16013, 'AGTGAT': 16002, 'GAATCA': 16002, 'GATGTA': 15982, 'CTCCCA': 15970, 'TATGGG': 15936, 'GTGAAT': 15929, 'ACTTTT': 15906, 'AAGGGT': 15904, 'ATTGAC': 15900, 'ATGTTT': 15883, 'AAAGTT': 15869, 'ACAAAT': 15857, 'GTATCT': 15854, 'TCCCAC': 15840, 'TATATC': 15840, 'TCTACT': 15832, 'CTTTTA': 15819, 'TGATTA': 15809, 'AACTTT': 15808, 'GTACCA': 15799, 'CAGAAT': 15782, 'TACCCC': 15775, 'TGATAA': 15759, 'TGTCTG': 15740, 'CCAGTA': 15716, 'AGGGAC': 15697, 'TCTGTC': 15685, 'ATATGA': 15672, 'ACTCAA': 15655, 'CCTACT': 15654, 'TTGCAC': 15645, 'TTTACA': 15641, 'TGAGTT': 15639, 'ACTCCT': 15631, 'TGAGGT': 15613, 'GGAGTA': 15612, 'AGGTTC': 15610, 'ACTTTC': 15607, 'TCTTGG': 15582, 'AATGTG': 15581, 'GTGACA': 15581, 'ATTGAG': 15578, 'CTCTTT': 15567, 'AATGAC': 15563, 'CGTATT': 15562, 'TCCCAA': 15562, 'GTTCTA': 15554, 'TAAAAT': 15553, 'GTTGAG': 15517, 'ATCTCT': 15513, 'ATTCAC': 15508, 'ACTACT': 15487, 'TATAAA': 15485, 'AATGCA': 15483, 'TTCGTT': 15441, 'TTTATA': 15440, 'AACATT': 15410, 'GTATGC': 15408, 'CGTACA': 15371, 'TTTGAG': 15369, 'AAGTCA': 15352, 'TTGGAC': 15342, 'CATATC': 15316, 'CTGACA': 15314, 'CAATTA': 15313, 'ATATCT': 15274, 'ACAAAC': 15255, 'CCCGTA': 15241, 'GCTTGT': 15241, 'TGACAG': 15224, 'CCAATT': 15221, 'AACCAA': 15214, 'CAAAGT': 15200, 'CATTGT': 15200, 'TTGCGT': 15156, 'TCGTAC': 15153, 'CTTGTC': 15148, 'GTATCA': 15131, 'AGGGAT': 15126, 'AATTAA': 15114, 'TTATAT': 15114, 'TGAGGG': 15112, 'GTACAC': 15076, 'TCACGT': 15063, 'AGGTTT': 15059, 'AGAATG': 15028, 'TATCTT': 15018, 'AACCCA': 15009, 'AATTCT': 15007, 'TGAGAA': 14998, 'CGTTAC': 14997, 'ACAATT': 14996, 'AATACA': 14993, 'GTCTGA': 14987, 'GTTACG': 14979, 'AATCGT': 14955, 'CTATAT': 14954, 'CCCTTT': 14950, 'AAACTT': 14948, 'AAACAT': 14948, 'TATATT': 14946, 'CAAGTA': 14941, 'GTTGTG': 14938, 'AGAATC': 14934, 'CATTTG': 14930, 'AAGTTG': 14922, 'TACGGA': 14922, 'CCCTAT': 14914, 'GGTACA': 14914, 'CTATTG': 14898, 'TACGCT': 14890, 'TTGTCC': 14877, 'ACTCGT': 14851, 'ATACCT': 14841, 'GAGGTT': 14835, 'AAAGTC': 14832, 'GTCAGT': 14830, 'TCCATT': 14812, 'GTTACC': 14812, 'CAGTAT': 14804, 'AATCTT': 14799, 'ATTGTG': 14797, 'TGAGAT': 14786, 'CAGAGG': 14781, 'AGGGGT': 14762, 'TATGGT': 14723, 'ACTGTT': 14712, 'TTGTCA': 14711, 'ATCCAA': 14701, 'TTACCC': 14701, 'AGTATT': 14683, 'CCCAAA': 14651, 'GTTTGT': 14650, 'TAGAAA': 14645, 'ACACTG': 14640, 'CCTTTA': 14640, 'AGTGAA': 14629, 'TGAATA': 14628, 'ATTAAT': 14626, 'TCTCTG': 14625, 'TATTCT': 14587, 'CTGAGG': 14573, 'AGGTTG': 14540, 'CTGTCA': 14524, 'TGTCAG': 14509, 'TATTAA': 14504, 'ACGTAC': 14484, 'CGTTAT': 14468, 'CAGTCA': 14448, 'CCAATA': 14438, 'CTTGAC': 14432, 'CAAATA': 14424, 'TCCAAA': 14399, 'TATCCA': 14397, 'ACTATG': 14386, 'CCCAAT': 14375, 'ACTTGC': 14354, 'AGGGAG': 14350, 'GACAGT': 14337, 'AGTCCA': 14329, 'AACTGT': 14317, 'TGTATC': 14316, 'GTGGTA': 14312, 'GTTATG': 14294, 'AGACGT': 14291, 'GCTTAC': 14289, 'TATCAT': 14288, 'GTACAG': 14274, 'TGACAT': 14261, 'TGTATT': 14244, 'GAATGA': 14224, 'GCATAT': 14221, 'ACGGTA': 14207, 'CTTATC': 14189, 'TTACAG': 14187, 'TGCCTA': 14176, 'AATATA': 14176, 'CTACTA': 14170, 'AACAGT': 14163, 'TGTATG': 14122, 'CGATAA': 14113, 'CTTATT': 14111, 'AATCAC': 14106, 'CCTTGT': 14094, 'TCTTTA': 14084, 'TTACTG': 14082, 'TCCACT': 14076, 'TTCTCT': 14068, 'AGGATA': 14066, 'TACCAA': 14062, 'TACTCC': 14050, 'ATACCC': 14048, 'TGACAC': 14039, 'ATGAGA': 14033, 'GGTTTA': 14033, 'GTCTTG': 14012, 'AAGAGT': 14009, 'TTAATG': 14009, 'ACCCAA': 14007, 'CTTGTT': 13998, 'TTCTGT': 13985, 'TTTTAG': 13979, 'CCAAAT': 13974, 'AGACAA': 13954, 'TATGAG': 13933, 'TATCCT': 13932, 'CCCCCC': 13919, 'AGATAA': 13901, 'TGGGTA': 13893, 'ATCACT': 13872, 'GAGTAT': 13871, 'TACGCA': 13860, 'ATGATA': 13859, 'AGTCGT': 13855, 'TACAGG': 13853, 'AAACAC': 13843, 'GAACTA': 13832, 'GTTTAC': 13832, 'GGGATA': 13831, 'AACTAT': 13830, 'TAAAGG': 13827, 'ATTACT': 13823, 'GGATAA': 13822, 'ACAGTT': 13792, 'GTACTT': 13790, 'TTGGGT': 13777, 'TGTACA': 13774, 'GTACAT': 13770, 'TCTCAC': 13769, 'CAGGTA': 13757, 'ATACCA': 13752, 'TCAGTC': 13751, 'ACCCTT': 13698, 'GATAAT': 13686, 'AATAAT': 13686, 'AGAGGT': 13642, 'AACTTG': 13638, 'TGCTAC': 13632, 'ATGTCT': 13624, 'GACTTG': 13612, 'ATTTGT': 13589, 'TTCAGT': 13576, 'GCGTTA': 13541, 'AGGGTT': 13535, 'GTACTC': 13487, 'TCTCTC': 13478, 'TAAAAC': 13474, 'TAATGA': 13462, 'GACAGA': 13453, 'TCCCTT': 13439, 'AAAGTA': 13435, 'AGTATC': 13420, 'GAGCTA': 13418, 'TATGCT': 13407, 'AGTCTG': 13370, 'AAGTAT': 13369, 'CAGACA': 13368, 'TTTACT': 13326, 'ACAGTG': 13317, 'AGAGTT': 13308, 'ATCACA': 13298, 'CACTTT': 13297, 'TTAATT': 13289, 'ATGCAT': 13239, 'ACTGTC': 13232, 'TGACTG': 13227, 'AGTGCA': 13222, 'ATACGA': 13219, 'ACTTGG': 13215, 'TACCTT': 13209, 'GCACTA': 13162, 'CTATTA': 13159, 'CTGCTA': 13151, 'TATTGT': 13077, 'GAATGT': 13063, 'TTTAGA': 13055, 'AGTCCT': 13055, 'CCCACT': 13047, 'ACGTTT': 13046, 'CCCATT': 13021, 'CACAAA': 13017, 'TATGCA': 13008, 'CCATTA': 13003, 'ACTATT': 12985, 'ACATTT': 12981, 'CTTATG': 12970, 'TTGGTA': 12960, 'AAATGT': 12944, 'ATAAAT': 12943, 'TGTGAA': 12940, 'TACTTT': 12938, 'GTCTGT': 12936, 'TTTCTA': 12927, 'AGGGTC': 12924, 'TTAAAT': 12921, 'GGCTTA': 12920, 'TAAAGC': 12918, 'GTATGA': 12875, 'AATGTT': 12858, 'CTCATT': 12842, 'AGAGGG': 12828, 'TTAGAA': 12822, 'ACGATA': 12817, 'TGGTTA': 12809, 'TTATGT': 12766, 'CTCTTG': 12702, 'TGACTT': 12698, 'CATTTA': 12696, 'CAGATA': 12694, 'TGTTTA': 12693, 'AGTCCC': 12684, 'CTGTCT': 12625, 'ATGAGT': 12615, 'AGAATA': 12605, 'CAGTCT': 12601, 'TCTCTT': 12571, 'CTTACC': 12549, 'CGTATG': 12546, 'TAATTT': 12528, 'AATTGT': 12520, 'TATACC': 12505, 'GAGAGG': 12489, 'TTATAA': 12473, 'GTCAGA': 12460, 'GTCCCA': 12455, 'ATGTAT': 12445, 'TGTCAA': 12441, 'AAGACT': 12418, 'TTGCAT': 12410, 'AAGGTA': 12385, 'AATGAG': 12373, 'TTACTT': 12365, 'AACGTT': 12357, 'TATATG': 12349, 'CTTTAA': 12334, 'TGTGTT': 12329, 'TAGAAG': 12324, 'ATTATA': 12319, 'TCTGTG': 12296, 'ATACAA': 12291, 'TGTACT': 12259, 'CAGAGA': 12255, 'TGGTAA': 12251, 'TCTCAG': 12234, 'GCCATA': 12227, 'TTGAGA': 12222, 'GTGTCA': 12182, 'AAACGT': 12175, 'GTCACT': 12171, 'TCGCTA': 12167, 'GGCATA': 12160, 'TCCCAT': 12153, 'AGACAG': 12139, 'CATATT': 12139, 'CGCTTA': 12125, 'TCTCAT': 12120, 'CACTGT': 12120, 'CATACC': 12114, 'GACACT': 12113, 'ATATAT': 12113, 'TCCAAT': 12112, 'TGTCTT': 12098, 'ATAGAA': 12087, 'CCTTAT': 12083, 'GCTATA': 12056, 'AGACAT': 12054, 'CACTCA': 12048, 'GGGGTA': 12048, 'GACTGA': 12044, 'AGATTA': 12041, 'ATGACT': 12022, 'TCATTA': 12017, 'ACTTGA': 12015, 'TCAGAC': 12010, 'ACATGT': 12001, 'CACAGA': 11992, 'TCACAG': 11982, 'ATAATG': 11976, 'TCCTAT': 11946, 'CTGATA': 11921, 'TGCTAT': 11905, 'ACTCAC': 11879, 'CCTATA': 11857, 'CCGTTA': 11839, 'AGTACT': 11821, 'AGTCAG': 11809, 'GCATAC': 11788, 'CTATAC': 11776, 'CTAAAA': 11771, 'CCATAT': 11767, 'TAAATT': 11753, 'GTAAAG': 11740, 'ATGGTA': 11717, 'TCACTC': 11706, 'TGTCTC': 11703, 'GCTAAA': 11684, 'TGTGAC': 11662, 'ACACGT': 11652, 'GTTAAA': 11650, 'GTATTA': 11635, 'CTCCTA': 11629, 'TAATAA': 11614, 'AGTATG': 11608, 'GACGTA': 11590, 'CTTGTG': 11585, 'CGTTTA': 11579, 'ATACAG': 11561, 'GAGATA': 11551, 'GGTAAT': 11542, 'CCACTA': 11539, 'GCCTTA': 11538, 'ACTTTA': 11533, 'ACAGAC': 11523, 'CAATAA': 11500, 'TTCACT': 11488, 'TACCCA': 11475, 'GATAAC': 11463, 'AGTTAT': 11375, 'CTGACT': 11373, 'GGTAGC': 11340, 'AGGTAT': 11319, 'CTTAAA': 11314, 'TATGTT': 11312, 'AACTCT': 11260, 'AGTGAC': 11259, 'ACTCTG': 11244, 'TACCCT': 11244, 'AATACT': 11218, 'AGTCAT': 11213, 'TGCGTA': 11210, 'AGTTGT': 11207, 'TTGTGA': 11206, 'ACCGTA': 11203, 'CCCCTA': 11191, 'ATTTAG': 11184, 'GATAGC': 11182, 'ATACTG': 11181, 'TGAGTG': 11169, 'ACAGTC': 11163, 'GGGTTA': 11154, 'ATAATA': 11148, 'ACAATA': 11134, 'TTGTCT': 11125, 'TCACAA': 11122, 'GATAAG': 11119, 'TTTAAC': 11115, 'TACCAT': 11062, 'TGTTAT': 11054, 'ATATAA': 11043, 'GTGAGA': 11022, 'GATACA': 11010, 'TCACTT': 10977, 'TACTGA': 10974, 'TGATAC': 10947, 'TTACAT': 10946, 'AAATAG': 10945, 'TACAGA': 10938, 'TGCTTA': 10922, 'CTTACG': 10916, 'GTCTCA': 10911, 'TTAATA': 10890, 'AGTTTA': 10888, 'CATTAA': 10884, 'GAGACA': 10882, 'TTCACA': 10873, 'ACTCAG': 10853, 'CTCAGT': 10830, 'CAGTGT': 10821, 'AAGCTA': 10821, 'CGGTAG': 10815, 'TCACAC': 10809, 'TACAAT': 10802, 'AGTCAA': 10772, 'CCTTAC': 10771, 'CACTTG': 10771, 'GATACT': 10761, 'TTGACA': 10730, 'GTGAGT': 10728, 'CACAAT': 10704, 'TAATGG': 10691, 'GTGACT': 10631, 'GTGTAT': 10627, 'GATATA': 10607, 'TAAATA': 10606, 'CTCTGA': 10601, 'GTCGTA': 10600, 'AGACTT': 10599, 'ATGTGT': 10593, 'ACTGTG': 10591, 'CTCTCT': 10588, 'GACTGT': 10580, 'TTGCTA': 10564, 'TAACGG': 10552, 'CGTGTA': 10548, 'TTGTTA': 10506, 'GAGAGA': 10502, 'AGGTAC': 10496, 'AAGTCT': 10478, 'CAGTTA': 10477, 'GGGTAA': 10475, 'GGTAAC': 10464, 'CATGTA': 10461, 'GTGTGA': 10460, 'AGATAC': 10460, 'CTCACT': 10447, 'TTTAGC': 10446, 'TCCGTA': 10442, 'CACATT': 10423, 'AACACA': 10410, 'TCTATA': 10370, 'GTCACA': 10368, 'AGACAC': 10331, 'TAGCGG': 10319, 'ATATGT': 10299, 'AGGTAA': 10293, 'AGGCTA': 10274, 'TTACGT': 10264, 'GTATGT': 10238, 'GGTAAG': 10237, 'CTGAGT': 10212, 'ACTCAT': 10197, 'ATAATT': 10192, 'ATAACG': 10188, 'TTAATC': 10168, 'GTTACA': 10159, 'TAGCCG': 10151, 'AGACTG': 10132, 'AGTCAC': 10125, 'GTTTAA': 10119, 'GATTAA': 10117, 'CCATAC': 10107, 'GACACA': 10098, 'ACACAG': 10070, 'ACTTAT': 10063, 'AGCTTA': 10030, 'TCAGAG': 10028, 'TTTGTA': 10023, 'AAGTTA': 10018, 'TACATT': 10015, 'CTCTCA': 9989, 'GACTCA': 9985, 'ACTCTC': 9970, 'CTCAGA': 9967, 'TCTTGT': 9966, 'TCTTAT': 9960, 'GTGATA': 9952, 'CACTAT': 9937, 'ACAGAG': 9928, 'TAATGC': 9890, 'GTCTCT': 9884, 'ACGCTA': 9884, 'CTGTGT': 9875, 'CTATAA': 9861, 'CAGACT': 9834, 'GAGTGA': 9817, 'GATAGA': 9814, 'TTGTAT': 9810, 'ATACTT': 9803, 'AGTGTT': 9744, 'TCGTAT': 9743, 'AACACT': 9710, 'CTGTGA': 9693, 'TTTAAG': 9679, 'CGCATA': 9676, 'TGGTAG': 9670, 'TGTGTC': 9659, 'ACATTA': 9657, 'CACAGT': 9640, 'TAACCG': 9620, 'CATACG': 9619, 'CACTCT': 9616, 'TTGAGT': 9611, 'GTAGCG': 9598, 'GTGTCT': 9586, 'AGTAAA': 9584, 'CATATG': 9567, 'ATAAAC': 9565, 'ATAGCG': 9557, 'TAATAT': 9557, 'GTAATG': 9521, 'TGTGTG': 9500, 'CACACT': 9499, 'TGTTAA': 9498, 'TTACAC': 9496, 'ATAACC': 9488, 'TGATAG': 9476, 'GTATAT': 9472, 'ATAGCC': 9443, 'AAACTA': 9440, 'ACACTC': 9439, 'TAACGA': 9429, 'GTAGAA': 9418, 'TAATCA': 9407, 'GGTAGA': 9404, 'TCTCTA': 9379, 'AGCGTA': 9346, 'CATAAA': 9330, 'GTTACT': 9324, 'TTAACG': 9323, 'ATAATC': 9317, 'AAGTGT': 9308, 'TCAGTA': 9302, 'TATAAT': 9296, 'TTATAC': 9294, 'TTACTC': 9290, 'TATACG': 9286, 'TAAATG': 9280, 'TTCCTA': 9272, 'AATAGA': 9268, 'TTAAAC': 9264, 'AATTAG': 9249, 'GCTTAA': 9244, 'TTCTTA': 9227, 'TAAACA': 9202, 'GAGTCA': 9197, 'CAGAGT': 9191, 'CTCTGT': 9190, 'GTAGCC': 9188, 'TACTCA': 9163, 'GAGAGT': 9110, 'CTTACA': 9096, 'GTTAAT': 9088, 'AGAGAG': 9086, 'TAGATG': 9077, 'ATTAAC': 9070, 'TGAGTC': 9055, 'TACTAT': 9052, 'AAGTAA': 9047, 'ACGTAT': 9046, 'GAGGTA': 9031, 'ACACAA': 9026, 'AGAGAC': 9020, 'TCATAT': 8994, 'TAAGAA': 8993, 'CTAAAG': 8985, 'TTAACC': 8984, 'CTGTTA': 8934, 'TCACAT': 8929, 'TACTTG': 8919, 'TATCTA': 8909, 'TGACTC': 8905, 'TAAGGA': 8896, 'ATAAGG': 8887, 'GAGTCT': 8886, 'TCTGAG': 8878, 'AGAGTG': 8863, 'GTCCTA': 8857, 'TATAGA': 8842, 'ACTCTT': 8837, 'AGTGTG': 8835, 'TGAGAC': 8827, 'TGCTAA': 8822, 'GTAACG': 8813, 'TAAAGT': 8813, 'GCGTAA': 8809, 'AGTCTT': 8804, 'TACGTT': 8796, 'ATACAT': 8765, 'GACTCT': 8759, 'TAATTG': 8753, 'TGCATA': 8741, 'GTAAAT': 8703, 'TAAATC': 8689, 'CGTAAA': 8654, 'ATGTTA': 8654, 'TAGCCA': 8645, 'GAGTTA': 8644, 'TTAGAT': 8638, 'TTGACT': 8635, 'AGTTAC': 8629, 'AGATAG': 8602, 'AGCCTA': 8602, 'CTCGTA': 8602, 'TGTAAA': 8594, 'CCGTAA': 8585, 'ATAGAT': 8573, 'ATTAGA': 8568, 'AGTCTC': 8565, 'GAATAA': 8562, 'ACTGAG': 8537, 'CTTATA': 8524, 'TATATA': 8518, 'AACTTA': 8515, 'AGTGAG': 8510, 'CTGAGA': 8495, 'TGTTAC': 8493, 'ATCTTA': 8490, 'TAGCGA': 8487, 'TTGTAC': 8487, 'TTACTA': 8487, 'ACACAC': 8465, 'AATAAC': 8462, 'ATCGTA': 8456, 'AGTGTC': 8446, 'CGATAG': 8440, 'TTTAGT': 8400, 'CTTTAG': 8358, 'TCGTTA': 8357, 'ACACTT': 8344, 'GTGTGT': 8333, 'AATGTA': 8311, 'ATAAGC': 8308, 'CTCACA': 8303, 'TTCATA': 8284, 'GTAACC': 8276, 'ATAGAG': 8274, 'TCTGTA': 8241, 'CACGTA': 8228, 'GTACTA': 8218, 'TAATTA': 8208, 'GTTATA': 8207, 'TTAAGC': 8195, 'TCCCTA': 8191, 'ACCTTA': 8189, 'AGTTAA': 8188, 'AATCTA': 8187, 'TTCGTA': 8165, 'ATCATA': 8160, 'TAACAA': 8155, 'GAGACT': 8136, 'GGTTAA': 8093, 'ATATAC': 8087, 'AGCATA': 8074, 'AGGTTA': 8040, 'ATGCTA': 7996, 'ATACGT': 7974, 'TAGAGA': 7949, 'TATTAG': 7948, 'TTTAGG': 7929, 'GTAATT': 7920, 'TAACGC': 7920, 'ATAACA': 7912, 'TATAAC': 7909, 'TAACCA': 7905, 'AGGGTA': 7883, 'ATTGTA': 7882, 'ATACAC': 7862, 'GTGCTA': 7847, 'TAGATA': 7846, 'TGACTA': 7831, 'ATACTC': 7810, 'ACCATA': 7810, 'GTAATC': 7807, 'TAGAGG': 7802, 'ATAAGA': 7800, 'TTGTGT': 7781, 'ATTAAG': 7780, 'TAGCGC': 7777, 'ACTTAC': 7767, 'ATTAGC': 7766, 'TAATCG': 7756, 'GTTGTA': 7756, 'CTTACT': 7751, 'TTAAGG': 7727, 'AACCTA': 7720, 'TAATCC': 7717, 'CTTAAT': 7716, 'AATAGC': 7707, 'TTAACA': 7670, 'TACAGT': 7660, 'ACTATA': 7659, 'CGTAGC': 7654, 'CACACA': 7648, 'GGTATA': 7638, 'TATACA': 7632, 'GAGTGT': 7631, 'TTATAG': 7609, 'TCCTTA': 7590, 'ATATAG': 7585, 'TCTTAA': 7583, 'ACATAT': 7567, 'TAAGCA': 7559, 'AGCTAA': 7531, 'ACAGTA': 7528, 'CGTAAC': 7526, 'CCCATA': 7524, 'ATTCTA': 7523, 'TATGTA': 7517, 'TAAACG': 7513, 'AATAAG': 7510, 'TCTTAC': 7485, 'TCCATA': 7469, 'TTAGCG': 7452, 'TAAACC': 7446, 'TGTCTA': 7419, 'TAATGT': 7393, 'AACGTA': 7387, 'AGAGTC': 7379, 'ACTTGT': 7356, 'TAGCAA': 7351, 'AACATA': 7330, 'AGTAAT': 7327, 'TTAGCC': 7323, 'TACCTA': 7307, 'TGTAGA': 7307, 'TGAGTA': 7303, 'GTAAAC': 7277, 'CGTTAA': 7247, 'ACACAT': 7220, 'TACTGT': 7211, 'TAGTGG': 7207, 'GGCTAA': 7187, 'GCTAAT': 7180, 'AGACTC': 7178, 'TAGTGA': 7165, 'TGAGAG': 7161, 'TTAGAG': 7160, 'GGTAGT': 7159, 'CCCTTA': 7131, 'GGATAG': 7122, 'TTAAGA': 7104, 'CAGTAA': 7104, 'GACTTA': 7086, 'CCTTAA': 7084, 'TAATCT': 7078, 'TAATTC': 7073, 'GTAGAG': 7072, 'GTAGTG': 7070, 'AGGTAG': 7068, 'TTAGCA': 7032, 'ATCCTA': 7028, 'GCATAA': 7017, 'GATAGT': 6997, 'AAGTAG': 6978, 'TAGCAG': 6969, 'TAAGCG': 6953, 'GTAAGC': 6939, 'CATAGA': 6930, 'TAAGGC': 6928, 'TAGAGC': 6924, 'TAACCT': 6919, 'TCTAAA': 6899, 'TAACCC': 6868, 'TAAGCT': 6867, 'ATGTAA': 6864, 'GTAGAT': 6852, 'TCATAC': 6845, 'CATACA': 6844, 'TAAACT': 6834, 'CCATAA': 6831, 'TCATAA': 6826, 'ATAGCA': 6824, 'TTGTAA': 6821, 'ACGTTA': 6819, 'TTAGCT': 6790, 'ACTTAA': 6768, 'TAACAG': 6760, 'TATAAG': 6750, 'TAGCTG': 6730, 'CATAGC': 6726, 'TTAGTG': 6688, 'TATAGC': 6687, 'CGTAAT': 6675, 'TGTATA': 6668, 'TAGGCG': 6632, 'TACACA': 6625, 'TCACTA': 6621, 'TACTTA': 6614, 'CGCTAA': 6608, 'TTAACT': 6600, 'GTGTTA': 6562, 'GTAAGG': 6540, 'ACTAAA': 6523, 'AGAGTA': 6519, 'TAAGCC': 6484, 'ATAGCT': 6478, 'TTGTAG': 6469, 'CACTTA': 6438, 'CCGTAG': 6417, 'TAGAAC': 6391, 'GTTAAC': 6389, 'GGGTAG': 6386, 'ATAGTG': 6385, 'ATTAGT': 6377, 'GACATA': 6371, 'TGTGAG': 6368, 'TTCTAA': 6357, 'TAGTTT': 6342, 'CATACT': 6337, 'ATAACT': 6327, 'CCATAG': 6325, 'TAACTT': 6316, 'GTTTAG': 6309, 'CCTAAA': 6300, 'TTAGTT': 6252, 'ACCCTA': 6230, 'AGTATA': 6227, 'GTCATA': 6223, 'AGTAGA': 6212, 'GTAATA': 6204, 'TAATAG': 6174, 'ACATAC': 6174, 'AATAGT': 6165, 'ATGTAG': 6150, 'TAGAAT': 6094, 'CATAAC': 6081, 'TAGATT': 6072, 'GTATAA': 6063, 'GTATAC': 6062, 'TAAGGG': 6040, 'GGTAGG': 6036, 'CAATAG': 6009, 'CTTAAC': 6003, 'TCGTAA': 6001, 'CATTAG': 5989, 'TTAAGT': 5978, 'CATAAT': 5965, 'CTAATG': 5930, 'CTTGTA': 5926, 'ATAGAC': 5920, 'CTCTTA': 5914, 'TAAGTT': 5911, 'TGTAAT': 5908, 'GTTAAG': 5899, 'TAGCCT': 5899, 'GTAGGC': 5854, 'GCGTAG': 5835, 'TATACT': 5796, 'ATAGGC': 5792, 'TAGCTT': 5781, 'ATAGTT': 5778, 'CTAAAT': 5761, 'TACTCT': 5756, 'ACTGTA': 5693, 'GCTAAC': 5663, 'TAGGCA': 5650, 'GATAGG': 5647, 'CTTAAG': 5646, 'TAGTGC': 5638, 'TAATAC': 5627, 'GCATAG': 5625, 'TAAGAT': 5624, 'GCTAAG': 5621, 'TGTTAG': 5611, 'AACTAA': 5605, 'TAGCCC': 5603, 'TAACTG': 5602, 'GTCTTA': 5559, 'GTTAGA': 5541, 'CATATA': 5527, 'GTAGCA': 5486, 'TAGGAA': 5457, 'TCATAG': 5449, 'CTCATA': 5447, 'GTAACA': 5441, 'CGTAAG': 5431, 'GTAAGA': 5430, 'TAGGTG': 5402, 'ACATAA': 5396, 'CTTAGA': 5387, 'TAAGGT': 5367, 'CTATAG': 5351, 'ACTCTA': 5340, 'TAGACG': 5336, 'TACATA': 5334, 'ATACTA': 5318, 'TAGTAA': 5315, 'GTTAGC': 5315, 'TTAGGA': 5286, 'TTAGGT': 5273, 'AGACTA': 5273, 'TACACT': 5249, 'CTGTAA': 5243, 'TGTAGC': 5230, 'TTAGTA': 5215, 'CGTAGA': 5199, 'TGCTAG': 5199, 'AGTAGC': 5194, 'TAGTTG': 5190, 'CTAATT': 5161, 'GTAGCT': 5146, 'CTAGAA': 5127, 'TGTGTA': 5094, 'ATAAGT': 5078, 'CTTAGC': 5070, 'TACGTA': 5052, 'CGTATA': 5034, 'CAGTAG': 5033, 'ATCTAA': 5031, 'AGTTAG': 5019, 'ATAGTA': 4998, 'TCCTAA': 4992, 'ATTAGG': 4975, 'GCTAGA': 4960, 'TAGCTA': 4935, 'TTAGGC': 4930, 'TCGTAG': 4926, 'TAGTCA': 4922, 'ACGTAA': 4891, 'CTAAAC': 4889, 'TAGACA': 4887, 'GCTAGC': 4874, 'TAACAT': 4865, 'TAAGAG': 4791, 'GTAGAC': 4780, 'GCTTAG': 4776, 'TAACTA': 4755, 'GTAACT': 4752, 'AGTAAC': 4749, 'CTAATA': 4730, 'TAACGT': 4699, 'CATAAG': 4694, 'TACTAA': 4680, 'CTGTAG': 4680, 'TATAGT': 4678, 'CTAATC': 4666, 'TAGTCG': 4624, 'ATAGGT': 4597, 'GTAGTT': 4594, 'ATAGGA': 4584, 'GAATAG': 4574, 'CTAACC': 4564, 'AATAGG': 4562, 'ACTAAT': 4561, 'TGTAAC': 4543, 'TCTAAT': 4533, 'TAAGTG': 4523, 'TAACAC': 4519, 'CCTTAG': 4510, 'GATTAG': 4500, 'TTAGGG': 4458, 'GAGTAA': 4446, 'GTAAGT': 4426, 'CTAAGC': 4421, 'CACATA': 4393, 'ATAGTC': 4358, 'CATAGG': 4355, 'TAGTAG': 4346, 'CTAGCC': 4344, 'TAGATC': 4343, 'TAAGAC': 4329, 'AGTAGT': 4327, 'TATAGG': 4308, 'ACATAG': 4291, 'TTAGTC': 4285, 'CTAACG': 4276, 'TAGGGG': 4273, 'AGTAAG': 4252, 'TAGGAG': 4252, 'CCTAAT': 4213, 'GTAGTC': 4190, 'CGTAGT': 4184, 'GCCTAA': 4178, 'TCTTAG': 4162, 'AGTCTA': 4158, 'ACACTA': 4148, 'TGTAGG': 4140, 'GTTAGT': 4086, 'AGCTAG': 4086, 'CTAAGG': 4077, 'TAGTTA': 4072, 'TAAGTA': 4068, 'GTAGGG': 4068, 'TGTAGT': 4061, 'ACTTAG': 4055, 'AGTGTA': 4045, 'GGTTAG': 4042, 'TAGACC': 4035, 'TAGCAC': 4030, 'TAGGGA': 4016, 'TAGGTA': 4003, 'CGTAGG': 3977, 'GTATAG': 3970, 'CGCTAG': 3959, 'TAGGCT': 3934, 'TTAGAC': 3931, 'CTAGAG': 3928, 'TAACTC': 3925, 'TAGCGT': 3922, 'CACTAA': 3915, 'TAGCAT': 3909, 'ATAGGG': 3899, 'CTAGCG': 3898, 'TAGGTT': 3896, 'TAGTAT': 3894, 'GTAGGT': 3880, 'TAGGGC': 3879, 'GTAGTA': 3877, 'ACCTAA': 3866, 'CCCTAA': 3862, 'CGTTAG': 3844, 'ACGTAG': 3836, 'TAGCTC': 3827, 'GTAGGA': 3798, 'TAGAGT': 3759, 'TAAGTC': 3711, 'CATAGT': 3679, 'TTCTAG': 3649, 'CTTAGT': 3577, 'GTTAGG': 3574, 'GTGTAA': 3532, 'TAGGCC': 3528, 'AACTAG': 3515, 'TCTAGA': 3490, 'GGCTAG': 3490, 'CTTAGG': 3469, 'GCTAGT': 3460, 'AGTAGG': 3447, 'GCTAGG': 3440, 'GTGTAG': 3439, 'TGTAAG': 3415, 'CTAAGA': 3403, 'CTCTAA': 3389, 'ATCTAG': 3348, 'TAGTTC': 3347, 'TAGTGT': 3310, 'CCTAGC': 3299, 'CTAACA': 3276, 'CTAGAT': 3275, 'CCTAAC': 3245, 'ACTAAC': 3197, 'TAGGAT': 3194, 'CCTAGA': 3193, 'CTAGTG': 3177, 'CTAGGC': 3168, 'CCTAAG': 3140, 'ACTAGA': 3108, 'TCTAGC': 3041, 'TCTAAG': 3032, 'TAGTCT': 2932, 'TCTAAC': 2930, 'GACTAA': 2920, 'CTAACT': 2919, 'CTAGCA': 2917, 'CTAGCT': 2859, 'ACTAAG': 2839, 'GAGTAG': 2823, 'ACTAGC': 2804, 'CCCTAG': 2795, 'TAGACT': 2794, 'CTAAGT': 2731, 'TAGGAC': 2688, 'GCCTAG': 2634, 'ACCTAG': 2628, 'GTCTAA': 2617, 'TAGTAC': 2587, 'TAGTCC': 2582, 'TCCTAG': 2563, 'TAGGTC': 2474, 'TAGGGT': 2462, 'CTAGAC': 2421, 'TACTAG': 2394, 'TCTAGT': 2332, 'CTAGTT': 2327, 'CTCTAG': 2242, 'CTAGTC': 2227, 'CACTAG': 2216, 'CTAGGG': 2205, 'CCTAGT': 2159, 'TCTAGG': 2107, 'CTAGTA': 2052, 'CTAGGT': 2025, 'CCTAGG': 1969, 'CTAGGA': 1923, 'ACTAGT': 1832, 'ACTAGG': 1805, 'GTCTAG': 1731, 'GACTAG': 1424, 'TCATGN': 2, 'CATGNN': 2, 'CAAAAR': 1, 'AAAARC': 1, 'AAARCT': 1, 'AARCTT': 1, 'ARCTTT': 1, 'RCTTTT': 1, 'CTGACM': 1, 'TGACMA': 1, 'GACMAT': 1, 'ACMATC': 1, 'CMATCT': 1, 'MATCTG': 1, 'GATGCR': 1, 'ATGCRG': 1, 'TGCRGA': 1, 'GCRGAA': 1, 'CRGAAG': 1, 'RGAAGG': 1, 'ACCTCM': 1, 'CCTCMT': 1, 'CTCMTT': 1, 'TCMTTC': 1, 'CMTTCA': 1, 'MTTCAT': 1, 'ATTTTM': 1, 'TTTTMG': 1, 'TTTMGC': 1, 'TTMGCA': 1, 'TMGCAA': 1, 'MGCAAT': 1, 'GTTTTK': 1, 'TTTTKG': 1, 'TTTKGC': 1, 'TTKGCG': 1, 'TKGCGC': 1, 'KGCGCG': 1, 'GACAGY': 1, 'ACAGYG': 1, 'CAGYGA': 1, 'AGYGAC': 1, 'GYGACG': 1, 'YGACGG': 1, 'CGGCCM': 1, 'GGCCMC': 1, 'GCCMCA': 1, 'CCMCAC': 1, 'CMCACC': 1, 'MCACCC': 1, 'ACCAGS': 1, 'CCAGSC': 1, 'CAGSCC': 1, 'AGSCCC': 1, 'GSCCCC': 1, 'SCCCCC': 1, 'TCCCGS': 1, 'CCCGST': 1, 'CCGSTA': 1, 'CGSTAT': 1, 'GSTATT': 1, 'STATTT': 1, 'TGCTTW': 1, 'GCTTWR': 1, 'CTTWRA': 1, 'TTWRAG': 1, 'TWRAGA': 1, 'WRAGAA': 1, 'RAGAAA': 1, 'AGCGGS': 1, 'GCGGSR': 1, 'CGGSRC': 1, 'GGSRCC': 1, 'GSRCCA': 1, 'SRCCAC': 1, 'RCCACC': 1, 'CGCCGM': 1, 'GCCGMC': 1, 'CCGMCC': 1, 'CGMCCG': 1, 'GMCCGT': 1, 'MCCGTC': 1, 'TCACTS': 1, 'CACTSC': 1, 'ACTSCG': 1, 'CTSCGC': 1, 'TSCGCC': 1, 'SCGCCC': 1, 'CTGCCM': 1, 'TGCCMG': 1, 'GCCMGG': 1, 'CCMGGG': 1, 'CMGGGC': 1, 'MGGGCC': 1, 'CTTGAR': 1, 'TTGARA': 1, 'TGARAA': 1, 'GARAAA': 1, 'ARAAAA': 1, 'RAAAAG': 1, 'CAAGTR': 1, 'AAGTRC': 1, 'AGTRCC': 1, 'GTRCCT': 1, 'TRCCTG': 1, 'RCCTGC': 1, 'ATTCAR': 1, 'TTCARG': 1, 'TCARGG': 1, 'CARGGC': 1, 'ARGGCG': 1, 'RGGCGG': 1, 'ACGGTW': 1, 'CGGTWA': 1, 'GGTWAG': 1, 'GTWAGG': 1, 'TWAGGA': 1, 'WAGGAT': 1, 'ATTGCM': 1, 'TTGCMA': 1, 'TGCMAT': 1, 'GCMATT': 1, 'CMATTT': 1, 'MATTTA': 1, 'ACGCCS': 1, 'CGCCSC': 1, 'GCCSCC': 1, 'CCSCCC': 1, 'CSCCCG': 1, 'SCCCGC': 1, 'ATGNNA': 1, 'TGNNAC': 1, 'GNNACT': 1, 'NNACTG': 1, 'NACTGC': 1, 'ATGNNG': 1, 'TGNNGT': 1, 'GNNGTG': 1, 'NNGTGG': 1, 'NGTGGA': 1, 'GTCTCS': 1, 'TCTCSG': 1, 'CTCSGC': 1, 'TCSGCC': 1, 'CSGCCC': 1, 'SGCCCC': 1, 'GGCCTN': 1, 'GCCTNC': 1, 'CCTNCT': 1, 'CTNCTC': 1, 'TNCTCG': 1, 'NCTCGG': 1})\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Frequency c-terminal hexamers:\n",
      " Counter({'GCCGCC': 1316, 'GCGGCC': 1115, 'GCCGGC': 817, 'GGCGGC': 743, 'CCGGCC': 686, 'CGCGGC': 662, 'GGCGCC': 643, 'CTGGCC': 609, 'CGCGCC': 604, 'CTGCGC': 599, 'GCGGGC': 578, 'GACGCC': 571, 'GCGCGC': 565, 'CGCCGC': 559, 'AAGGGC': 527, 'GAGGGC': 515, 'ACCGCC': 509, 'GGCCGC': 498, 'CCGCGC': 493, 'CTGGGC': 486, 'GCCCGC': 480, 'AAGGCC': 464, 'CCCGCC': 455, 'GAGGCC': 450, 'CGGCGC': 448, 'CTCGGC': 436, 'CAGGGC': 433, 'GAGGAC': 429, 'GCCGAC': 416, 'CAGGCC': 414, 'GTCGCC': 413, 'GTCGGC': 404, 'GACGAC': 401, 'AGCGCC': 389, 'AAGCGC': 384, 'CTCGCC': 362, 'GTGCGC': 351, 'GAGCGC': 350, 'ACCGGC': 344, 'ACGGCC': 341, 'CGCGAC': 341, 'GCGGCG': 337, 'ATCGGC': 336, 'ATCGCC': 327, 'CAGCGC': 326, 'GGGCGC': 321, 'CGGGCC': 320, 'GGCGAC': 318, 'GCTGCC': 316, 'CTGGAC': 310, 'TCGGCC': 309, 'CTGCCC': 304, 'GAAGCC': 303, 'GCGGAC': 301, 'GTGGCC': 298, 'AACGCC': 298, 'CGGGGC': 291, 'GCAGCC': 286, 'GCGCAC': 280, 'AAAAAA': 279, 'GAGGAG': 278, 'GCGCCC': 276, 'GAAGGC': 270, 'GACGGC': 270, 'ACCCGC': 269, 'GACCGC': 269, 'TCCGCC': 268, 'CTGTTC': 267, 'AAGGAC': 265, 'CAGGAC': 265, 'AAAGCC': 263, 'GCGAGC': 263, 'CCGGGC': 261, 'GCCGTC': 261, 'GAGGCG': 259, 'CGCCCC': 256, 'GGGGCC': 254, 'AACGGC': 253, 'GCGGTC': 250, 'GGCAAC': 249, 'GCCGCG': 249, 'TTCGGC': 245, 'CATGGC': 244, 'ACGCGC': 242, 'CTCGAC': 239, 'CCCGGC': 239, 'CACGCC': 235, 'CGGCCC': 235, 'TTCGCC': 235, 'GAAGAC': 233, 'GGAGCC': 232, 'GCCTCC': 229, 'GTCCGC': 225, 'GCGTCC': 221, 'AAGCCC': 219, 'GCGGAG': 215, 'GCGCTC': 214, 'CAGCCC': 213, 'AGCCGC': 213, 'GATGCC': 213, 'GAGGTC': 211, 'AAAGGC': 211, 'CGCTTC': 208, 'GCCGAG': 208, 'CTGCTC': 205, 'GGACGC': 205, 'GAGAGC': 205, 'CCGTTC': 203, 'AACCGC': 202, 'CTGAAC': 200, 'CGTGGC': 199, 'TTCGAC': 198, 'TCGCGC': 198, 'CTGAGC': 198, 'AAGAAC': 197, 'CCGGAC': 197, 'CGGGAC': 196, 'GGCAGC': 196, 'AAGAAG': 195, 'GCAGGC': 195, 'AAGGAG': 194, 'CCGGCG': 194, 'ACGCCC': 194, 'CACGGC': 193, 'ATCGAC': 193, 'CCGGTC': 192, 'CTGATC': 192, 'GCCCAC': 191, 'GATGAC': 191, 'CCCCGC': 190, 'CTGGTC': 190, 'GTCGAC': 189, 'ATCCGC': 185, 'GTGGGC': 183, 'GCCCCC': 183, 'CGCGTC': 183, 'ACCGAC': 183, 'GAGGAT': 181, 'GGAGGC': 181, 'GAGTCC': 180, 'GACTTC': 180, 'GCGAAC': 179, 'GCGGCT': 179, 'CGCAGC': 179, 'GCCAGC': 178, 'CTGCAC': 177, 'AAGAGC': 176, 'GATGAG': 176, 'AAGTTC': 175, 'GGGGAG': 174, 'GGGGGC': 173, 'GGCGTC': 173, 'CTGTCC': 172, 'CGCTCC': 171, 'AATGGC': 171, 'GGCGAG': 170, 'CACCGC': 167, 'GCGCGG': 167, 'GCGATC': 167, 'GCGGCA': 166, 'AAGGTC': 166, 'GAGGAA': 166, 'GGCCAC': 165, 'AAAGAC': 164, 'GCCGGG': 163, 'AACGAC': 163, 'GCGTTC': 161, 'GCACGC': 161, 'AAGCTC': 160, 'AAACGC': 160, 'GGCTCC': 159, 'GAAAGC': 159, 'GCCAAC': 158, 'CATGCC': 158, 'GGCGCG': 157, 'GAGAAC': 157, 'GTGATC': 157, 'GAGGGG': 156, 'GCGGGG': 156, 'AAATCC': 155, 'GTGAAG': 154, 'TTCCGC': 154, 'CAGAGC': 153, 'GAACGC': 153, 'GATCGC': 152, 'CGGGTC': 152, 'GCCACC': 152, 'GAGCAC': 151, 'TCCGGC': 151, 'AAGAAA': 151, 'GCTTCC': 151, 'CGCACC': 151, 'GCCGGT': 151, 'GGTCGC': 150, 'CTGACC': 149, 'GCGCCG': 149, 'GGCACC': 148, 'GGGAGC': 147, 'CCGTCC': 146, 'GAGTTC': 144, 'CGGAGC': 143, 'CAGTTC': 143, 'GCCGCT': 142, 'TCCCGC': 142, 'ACCGTC': 142, 'GGCTTC': 141, 'AAGCAC': 141, 'CAGCAC': 141, 'CAGGTC': 141, 'AAGTCC': 140, 'GTGGAG': 139, 'AAGGCG': 139, 'GGCCGG': 139, 'GGTGGC': 138, 'ACGGGC': 138, 'GCGCAG': 138, 'GTGCTC': 138, 'AAAAAC': 138, 'AAGGGG': 138, 'AAAAGC': 137, 'TACCGC': 137, 'GCGCTG': 137, 'GACGTC': 137, 'TGGCGC': 137, 'ACGCTC': 136, 'CGGCTC': 136, 'GACGAG': 136, 'GATGGC': 136, 'AATGCC': 135, 'ACCCCC': 135, 'GCGGTG': 134, 'GTGGAC': 134, 'GGCATC': 134, 'GACAGC': 133, 'AGCGGC': 133, 'GAGATC': 132, 'CCGGAG': 132, 'AAGGCT': 131, 'GGGGCG': 131, 'GACGAT': 130, 'GGGCAC': 130, 'GTGGCG': 130, 'CAGTCC': 130, 'GGCAAG': 129, 'CGCTAC': 129, 'CGGCCG': 129, 'GCCGCA': 129, 'CCGCTC': 129, 'GAAGAG': 129, 'GAGCCC': 128, 'GATGAA': 126, 'GAAGAT': 126, 'GATTTC': 126, 'GACATC': 126, 'CGGATC': 125, 'CAAGCC': 125, 'GGGCGG': 125, 'GGCTAC': 125, 'CCTGCC': 125, 'GTCAGC': 124, 'GAATCC': 124, 'ATGCGC': 124, 'CATGAC': 124, 'CGCCAC': 124, 'AAAAAG': 123, 'AATGAC': 123, 'CGGCAC': 123, 'CGGCGG': 123, 'GTCGTC': 122, 'AAGGAT': 122, 'CATGAG': 122, 'ACCTTC': 122, 'AAACCC': 121, 'CTGGAT': 121, 'CAGGAG': 121, 'GGCGGG': 120, 'GAAGAA': 120, 'GCCGAA': 120, 'GACGCG': 120, 'CAGCTC': 119, 'GCGAAG': 119, 'GTGCTG': 119, 'CTCACC': 119, 'CAGAAC': 119, 'GATGAT': 119, 'GACAAC': 119, 'TACGCC': 118, 'GGCCTC': 118, 'GTGAGC': 117, 'GGCGAT': 117, 'GCCTTC': 117, 'GAGAAG': 117, 'GCGACC': 117, 'CTGGCG': 116, 'ACGGAC': 116, 'ACGTTC': 116, 'ACCACC': 115, 'GGTGCC': 114, 'GGGGTC': 114, 'TTGCGC': 114, 'CGCCGG': 114, 'GGGCTC': 113, 'GTGCGG': 113, 'TTCTTC': 113, 'CACGAC': 113, 'GAGCGG': 113, 'AACAAC': 113, 'AACGTC': 113, 'ACGGTC': 112, 'GAAAAA': 112, 'CGACGC': 112, 'ATGGCC': 112, 'GCCAAG': 111, 'GTCTCC': 111, 'CTCGTC': 111, 'CGCAAC': 111, 'AAAAAT': 111, 'CCGAGC': 111, 'AGGCGC': 111, 'GCATCC': 111, 'CATGTC': 110, 'AAGATC': 110, 'GAAAAC': 110, 'GGCTGC': 110, 'GCTCGC': 110, 'CAAGGC': 110, 'CTGCCG': 110, 'CCGCAC': 109, 'CTGGAG': 109, 'CAGGCG': 109, 'CGTCGC': 109, 'AGCTTC': 109, 'GAGCAG': 109, 'CGGTCC': 109, 'GGGCTG': 109, 'CCGAAC': 109, 'CGGTTC': 108, 'CGCTGG': 108, 'ATGGGC': 107, 'CCCGAC': 107, 'GATGTC': 107, 'GCGGAT': 107, 'CTGAAG': 107, 'GTGCCC': 106, 'AAAGAG': 106, 'TTTGCC': 106, 'CTGCGG': 106, 'CGTCCC': 106, 'GAGGTG': 106, 'CGCGCG': 106, 'GCCCGG': 105, 'CGCCTC': 105, 'GCGCAT': 105, 'GTGAAC': 105, 'GGGAAC': 105, 'GAGACC': 104, 'ACGCAC': 104, 'AGCAAC': 104, 'GCCCTC': 104, 'AAAGAT': 104, 'GTGTCC': 104, 'GGGAAG': 104, 'CGGGCG': 103, 'CAGGCT': 103, 'GAGCTG': 103, 'AACAGC': 103, 'GCCGGA': 103, 'CTGGGG': 102, 'GCGGAA': 102, 'GGCGGT': 102, 'GGTTCC': 101, 'ACCTCC': 101, 'CCAGCC': 100, 'AGCGAC': 100, 'CTCCGC': 100, 'AAGGGT': 100, 'CGCATC': 100, 'GTGCAC': 100, 'CGTGCC': 100, 'CTGCTG': 100, 'CAGGGG': 99, 'GCAAGC': 99, 'AATGAG': 99, 'CCGACC': 99, 'GTGTTC': 99, 'CACCAC': 99, 'GCTGGC': 99, 'GCTGAC': 98, 'GGTGAG': 98, 'CGGGTG': 98, 'GAGCTC': 97, 'CGGGGG': 96, 'GGGTTC': 96, 'GTGGTC': 96, 'CTTGCC': 96, 'AAAGAA': 95, 'CGTTCC': 95, 'AAATTC': 95, 'TCGGCG': 95, 'GGCGCA': 95, 'GAGGCA': 95, 'CTCAAC': 95, 'ATCAGC': 94, 'GACCGG': 94, 'GCCGAT': 94, 'GGAAGC': 94, 'GTCAAC': 94, 'AAGGAA': 94, 'GGAGAG': 94, 'CGCGGG': 94, 'GCAGAC': 93, 'TTCAGC': 93, 'GTGGGG': 93, 'GCCTAC': 93, 'TCGCAC': 93, 'CTCTCC': 93, 'CAGCCG': 93, 'AATGAA': 93, 'GAAATC': 93, 'CAACGC': 93, 'ACGCCG': 92, 'CTCAGC': 92, 'GGTCAC': 92, 'GAAACC': 92, 'AACTTC': 92, 'GGGGAC': 92, 'CCCGCG': 92, 'CGGACC': 92, 'CTGGTG': 92, 'GTTCGC': 92, 'GACCAC': 92, 'GACCTC': 92, 'GACTCC': 92, 'TCGCCC': 91, 'TCGGGC': 91, 'TGGGCC': 91, 'ACCAAC': 91, 'GGGGTG': 91, 'CCGGCA': 91, 'CGTGAC': 90, 'ACCCAC': 90, 'GCGCGT': 90, 'CCGATC': 90, 'ACAGCC': 90, 'CCGCTG': 89, 'TACGGC': 89, 'GTCATC': 89, 'ATCAAC': 89, 'GTTGCC': 89, 'CCCGTC': 89, 'GGCGAA': 89, 'GCGCGA': 89, 'TCCTCC': 89, 'GATTCC': 89, 'GGCCTG': 88, 'GCCATC': 88, 'ACGGCG': 88, 'CACCCC': 88, 'CGCCCG': 88, 'TGCCGC': 88, 'ACCAGC': 87, 'CGCGGT': 87, 'GGCGCT': 87, 'ATCATC': 87, 'GAAAAT': 87, 'GACGAA': 87, 'CCGGCT': 86, 'CCGGTG': 86, 'ACTGCC': 86, 'ACCGCG': 86, 'GAGCCG': 86, 'AAAGTC': 86, 'CTCCCC': 86, 'GATCTC': 86, 'TCGTCC': 85, 'AAGGCA': 85, 'GAGGCT': 85, 'GAACTC': 85, 'AATGAT': 85, 'GATGCT': 85, 'AAAGGA': 85, 'GACACC': 84, 'AGCCAC': 84, 'GGGATC': 84, 'CCGCGG': 83, 'TGGGGC': 83, 'CTCTTC': 83, 'GCCTGC': 83, 'GAAAAG': 83, 'GGAAAC': 83, 'TGCGGC': 83, 'GAGGGT': 83, 'GGGCAG': 82, 'GAGGGA': 82, 'GCGGGT': 82, 'AATCGC': 82, 'TCGTTC': 82, 'GGCGGA': 82, 'GAAGCG': 82, 'AACCCC': 82, 'AAGTAC': 82, 'ACCCTC': 82, 'AAGGTG': 82, 'CCGCAG': 81, 'ATTGCC': 81, 'TTGGCC': 81, 'AGTGCC': 81, 'GCCCTG': 81, 'AAAACC': 81, 'CCCTCC': 81, 'AAGCCT': 81, 'TCGGTC': 80, 'TCTGCC': 80, 'GCACCC': 80, 'CAGGAT': 80, 'AGCGAG': 80, 'AAGACC': 80, 'GCGCCT': 80, 'TCGGAC': 80, 'GTGGAT': 79, 'CTTGGC': 79, 'AATAAA': 79, 'AGGGGC': 79, 'TTCGTC': 79, 'AAGAAT': 79, 'GCTCCC': 79, 'TTCGAG': 78, 'CCCTTC': 78, 'CATGCT': 78, 'ACGAAC': 78, 'GATGCG': 78, 'GCCCGA': 78, 'GGGGAA': 78, 'CAGGCA': 78, 'GAAGGG': 78, 'TGCGCC': 78, 'TGGGAG': 78, 'CCACGC': 78, 'AAGCCG': 78, 'TCGCTC': 77, 'TCGAAC': 77, 'GGCGTG': 77, 'ATCTTC': 77, 'GCAGCG': 77, 'ATCGAG': 77, 'TGGGAC': 77, 'GAAGTC': 77, 'GGTTTC': 77, 'ATTCGC': 77, 'CAGCAG': 77, 'GAATTC': 77, 'AAAATC': 76, 'AAGCGG': 76, 'TCGATC': 76, 'ACGACC': 76, 'CCGGGG': 76, 'AGGAGC': 76, 'TCAGCC': 75, 'ATCTCC': 75, 'ACGATC': 75, 'GATATC': 75, 'TGGAAC': 75, 'GCGTCG': 75, 'GCGGGA': 75, 'ATCGTC': 75, 'CGCAAG': 75, 'CAGCGG': 75, 'CGCCGT': 75, 'CCGCGT': 75, 'GGCTGG': 75, 'TATCGC': 75, 'AGCAGC': 74, 'CATCAC': 74, 'AAAATT': 74, 'ATTTCC': 74, 'TTTGGC': 74, 'GGGTCC': 74, 'ACGTCC': 74, 'CTGGGT': 74, 'ATGGAC': 74, 'ATGCTC': 74, 'AATAAT': 74, 'GGCCAG': 73, 'AATTTC': 73, 'GCGCAA': 73, 'AACTCC': 73, 'GCGTAC': 73, 'ACCCGG': 73, 'CGGCGT': 73, 'ATCACC': 73, 'AACACC': 73, 'TTGGGC': 73, 'GGAGAC': 73, 'AACCTC': 73, 'CAACCC': 73, 'ACGAGC': 73, 'GGTGGT': 72, 'GGCCGA': 72, 'GGATTC': 72, 'AGGGCC': 72, 'AATGTC': 72, 'GTGACC': 72, 'GGGACC': 72, 'GAACCC': 72, 'TTCAAC': 72, 'GACTAC': 72, 'TGGAGC': 72, 'GGAGAA': 72, 'CTTTCC': 72, 'CACTCC': 72, 'ACCTAC': 71, 'CAGATC': 71, 'GCTGCT': 71, 'ATGTTC': 71, 'AAGCTG': 71, 'GCAGAG': 71, 'AGCGTC': 71, 'GGAGCG': 71, 'AAGCAG': 71, 'CGCGAT': 71, 'CGAGGC': 70, 'GAGCGT': 70, 'CATGAA': 70, 'GGCAAT': 70, 'GTGCAG': 70, 'TCCGTC': 70, 'GGGCAT': 70, 'CGTCAC': 70, 'AAAGCT': 70, 'GCGCCA': 70, 'GAGCAT': 70, 'CGGCTG': 70, 'CATCGC': 70, 'GGTGAC': 69, 'GAGGTT': 69, 'GCAAAC': 69, 'CGGGAG': 69, 'GTGGCT': 69, 'GGGCGT': 69, 'TCCGAC': 69, 'GAACAC': 69, 'CTGAAA': 69, 'CAGGAA': 69, 'ATCCTC': 69, 'ATGAGC': 69, 'CCTTCC': 69, 'GTGCCG': 69, 'CCGGAA': 69, 'AACCAC': 69, 'GGAAAA': 68, 'GAGAAA': 68, 'AACGAT': 68, 'AGGGAC': 68, 'ACCTGC': 68, 'GACCTG': 68, 'CTCGAT': 68, 'GACGGG': 68, 'ACCGGG': 68, 'GTGAAA': 68, 'TCGGAG': 68, 'GCTGAG': 67, 'CTGCGT': 67, 'AGCTGG': 67, 'CTGCAT': 67, 'CCCGAG': 67, 'GGGGAT': 67, 'TATGAG': 67, 'CCCACC': 67, 'ACGGCA': 67, 'ACCTGG': 67, 'TGGTTC': 67, 'AGCGCG': 67, 'ATGAAG': 67, 'CCGGAT': 66, 'TTGAAG': 66, 'AAAATA': 66, 'AAACTC': 66, 'TTCCAC': 66, 'TACTTC': 66, 'CGGGCA': 66, 'GCCCCG': 66, 'TTGTTC': 66, 'CAGAAG': 66, 'ATCCCC': 66, 'ACCCCG': 66, 'GGTGCG': 66, 'AAGCAT': 66, 'GCAGCA': 65, 'AGCTAC': 65, 'AAGAGG': 65, 'TTTTCC': 65, 'AAGGGA': 65, 'TGGCTC': 65, 'GATAAA': 65, 'CAGGGT': 65, 'GGCAAA': 65, 'AAAGGT': 65, 'TTGAAC': 65, 'GGGCCC': 65, 'GATTTT': 65, 'GGCCCC': 65, 'TATGAC': 65, 'GCCCGT': 65, 'CCCCGG': 65, 'GCTCAC': 64, 'GACGCT': 64, 'ACCGAG': 64, 'CGCGCA': 64, 'GTGGTG': 64, 'AAAGGG': 64, 'ATGAAC': 64, 'AGCCTC': 64, 'GCACAC': 64, 'TTTTTC': 64, 'ATCCAC': 64, 'GCGCTT': 64, 'TTCCTC': 63, 'GAGCAA': 63, 'CGGATG': 63, 'ATTGAT': 63, 'CGGCAG': 63, 'GCCTCG': 63, 'ATGGAT': 63, 'CCAGGC': 63, 'AGCCGT': 63, 'CCTCGC': 63, 'CAGGTG': 63, 'AAGGTT': 63, 'CGGGAT': 63, 'CGAGCC': 63, 'ATGCCC': 63, 'GAGAGG': 63, 'GACAAG': 63, 'CGGAAC': 63, 'CACGTC': 63, 'GGAGGG': 63, 'TTGAGC': 62, 'TCCAAC': 62, 'ATAAAA': 62, 'GACTGG': 62, 'CAAGAC': 62, 'GGTAAC': 62, 'TGGGAT': 62, 'GTCACC': 62, 'ACCGAT': 62, 'TTTGAT': 62, 'TCGACC': 62, 'AAACAC': 62, 'CTCGCG': 62, 'CATGCG': 62, 'CTGTAC': 62, 'AAAAGT': 62, 'ATTCCC': 62, 'CAGTAC': 62, 'ACGAAG': 61, 'CAAAAA': 61, 'CGGGGT': 61, 'GTCCCC': 61, 'AATGCT': 61, 'TATGCC': 61, 'CGCCTG': 61, 'GGGAAA': 61, 'GAGAAT': 61, 'CAGACC': 61, 'TGGATC': 61, 'AGGAAG': 61, 'ATCGAT': 61, 'CCCGCA': 61, 'GCTGAT': 61, 'CGATCC': 61, 'GTCCAC': 60, 'ATGATC': 60, 'GAAATT': 60, 'GTCGAT': 60, 'GGAGGT': 60, 'TATTTC': 60, 'CTCATC': 60, 'GGATCC': 60, 'CGCGGA': 60, 'CGCCGA': 60, 'GGGATG': 60, 'TTAAAA': 60, 'AATCCC': 60, 'TCCAGC': 60, 'GGAAAG': 60, 'ACGGTG': 59, 'CGGAAG': 59, 'GTCGCG': 59, 'ATCGAA': 59, 'CAGGGA': 59, 'CAGCCA': 59, 'TTTCGC': 59, 'GCCCAG': 59, 'TTCTCC': 59, 'CTGCAG': 59, 'ATTGGC': 59, 'AGGCCC': 59, 'ATGGCG': 59, 'ACACGC': 59, 'CTCAAG': 59, 'GGACAC': 59, 'GACTGC': 58, 'GGCGTT': 58, 'CAATCC': 58, 'CGGCGA': 58, 'TACGAC': 58, 'CGCGCT': 58, 'GAAGGT': 58, 'GGAATC': 58, 'GACGCA': 58, 'CACTAC': 58, 'GCAGCT': 58, 'GTCTTC': 58, 'GTCTAC': 58, 'GCTGTC': 58, 'GATCAC': 57, 'AGCATC': 57, 'ACGCCA': 57, 'ATCAAG': 57, 'CTTCCC': 57, 'CGGCCA': 57, 'CAAAAC': 57, 'GGCAGG': 57, 'GTTCCC': 57, 'TGGCTG': 57, 'GCACTC': 57, 'GAAGGA': 57, 'CGTGTC': 57, 'GGGTCG': 57, 'CAGCCT': 57, 'GTCGAG': 57, 'GTTTCC': 57, 'GGAACC': 57, 'TCTTCC': 57, 'GGGCGA': 56, 'GTGGAA': 56, 'GGAGTC': 56, 'GATCTG': 56, 'GCGTCA': 56, 'AAGCTT': 56, 'CGGCAT': 56, 'CATCCC': 56, 'GGACGG': 56, 'GGGGCT': 56, 'CAAAGC': 56, 'GCCTGG': 56, 'ATGCAC': 56, 'TGCTGC': 56, 'CATGAT': 56, 'AAGCGT': 56, 'ATCGCG': 56, 'GGCCGT': 56, 'TTGGAT': 56, 'GGGAGG': 55, 'ACCGCT': 55, 'CTCGGG': 55, 'TCACGC': 55, 'GAGCTT': 55, 'GCTGCG': 55, 'TTCATC': 55, 'ATTTTC': 55, 'AATTCC': 55, 'GACCAG': 55, 'GCAGAT': 55, 'GAGTCG': 55, 'TTGAAA': 55, 'ACGCTG': 55, 'GCGAAA': 55, 'ACCGCA': 54, 'CACTTC': 54, 'ACGGCT': 54, 'GCCAGT': 54, 'AGCCTG': 54, 'AACGAG': 54, 'AGCCCC': 54, 'CTTCGC': 54, 'ATGGTC': 54, 'GGTGAA': 54, 'CTGCCT': 54, 'CGTCCG': 54, 'GCACAG': 54, 'CGCGAG': 54, 'AACGGT': 54, 'AGTGAG': 54, 'GAGATT': 53, 'GGGGGA': 53, 'GAGTGC': 53, 'GTCAAG': 53, 'CTGGCT': 53, 'ACCCGA': 53, 'CTGCCA': 53, 'TTCGAT': 53, 'GAGTTT': 53, 'AAGCAA': 53, 'GGTAGC': 53, 'GCGATG': 53, 'GTGATG': 53, 'GCAAAA': 53, 'AGGAAA': 53, 'CGACCC': 53, 'GATCAG': 52, 'CTGCTT': 52, 'CCGAAG': 52, 'AGAAAG': 52, 'CGGGGA': 52, 'CCGCGA': 52, 'ATCGGG': 52, 'GCCGTG': 52, 'CGGTAC': 52, 'CCGTCG': 52, 'AGAAAA': 52, 'ATCGGT': 52, 'GCATTC': 52, 'GTGGTT': 52, 'GAAGTG': 52, 'TGGTGG': 52, 'TCCACC': 52, 'GAATTT': 52, 'GTCCTC': 52, 'AGCGCA': 52, 'ACGGAG': 52, 'GGTTAC': 52, 'CTCGGT': 52, 'TTTGAC': 52, 'ATTAAA': 52, 'GCTGAA': 52, 'TGGAAG': 52, 'TCCGCG': 52, 'CATTCC': 52, 'AATTAT': 52, 'ACTCAC': 51, 'GGGAAT': 51, 'CTGGAA': 51, 'ACGCGG': 51, 'AACGCA': 51, 'GAGTGG': 51, 'TGCGAC': 51, 'ACGGGG': 51, 'ATGGAG': 51, 'AAACAA': 51, 'AACAAG': 51, 'TCGCGG': 51, 'CGTTTC': 51, 'GTCGGG': 51, 'AAATAT': 51, 'GACCGT': 51, 'GCGAGG': 51, 'GAATAT': 51, 'GAATAC': 51, 'GACAAA': 51, 'ATGGGG': 51, 'CTTGAT': 51, 'CCGGGT': 51, 'TATGAA': 51, 'CGGCCT': 51, 'AGGAAC': 50, 'ACCATC': 50, 'GAGCGA': 50, 'AAACAG': 50, 'GTCCGG': 50, 'TTGGAC': 50, 'AACAAT': 50, 'AAATAC': 50, 'ACCAAG': 50, 'CCCTGC': 50, 'CCCGAT': 50, 'GGAGGA': 50, 'TTGCCC': 50, 'CCCGCT': 50, 'GGAAAT': 50, 'GCACCG': 50, 'CCGGGA': 50, 'GTCGGT': 50, 'CTGGCA': 50, 'TTCACC': 50, 'CCCAAC': 50, 'GGAAGG': 50, 'AACATC': 50, 'GTGCGT': 49, 'GACGGT': 49, 'AAAGCA': 49, 'CTTTTC': 49, 'CCCAGC': 49, 'GGGGTT': 49, 'GCAGTC': 49, 'GATAAG': 49, 'GAAGCT': 49, 'GGAGAT': 49, 'CTGTCG': 49, 'AGGGAG': 49, 'TGGGAA': 49, 'GAAAGG': 49, 'CAGTCG': 49, 'GCGACG': 49, 'GAGCCT': 49, 'GATATT': 49, 'AACCGG': 49, 'GCAATC': 49, 'AAAGCG': 49, 'GAAGTT': 49, 'GCATCG': 48, 'GTTGAA': 48, 'AGCAAG': 48, 'GATGTT': 48, 'GGCCAT': 48, 'TCGGCA': 48, 'CCCTGG': 48, 'GACCGA': 48, 'ATGAAA': 48, 'ATTTTT': 48, 'CGTTAC': 48, 'GGGAGT': 48, 'AGCCGG': 48, 'AATATT': 48, 'TTGCTC': 48, 'GTTGGC': 48, 'CTCTAC': 48, 'GGGGGG': 48, 'AGTCGC': 48, 'AACGCG': 48, 'TCGGCT': 47, 'GCGGTT': 47, 'TCGAGC': 47, 'GCGAGT': 47, 'AATCAA': 47, 'GGGCAA': 47, 'GGATCG': 47, 'AAGCCA': 47, 'AATAAG': 47, 'GAGTAC': 47, 'GCTGCA': 47, 'GTCGCT': 47, 'CGGTCG': 47, 'TTTAAT': 47, 'AGAGCC': 47, 'CAGAAA': 47, 'AATGGA': 47, 'TACGTC': 47, 'AGCACC': 46, 'GAAAGT': 46, 'CGAAGC': 46, 'CGGGCT': 46, 'AATGTT': 46, 'ATCCGG': 46, 'GCCAAA': 46, 'TTTGAG': 46, 'ATCTAC': 46, 'AGCCAG': 46, 'GAATGG': 46, 'CACCTC': 46, 'CCGCAA': 46, 'CGAACC': 46, 'ATCAAT': 46, 'CCTGGC': 46, 'GCCAAT': 46, 'AAATCA': 46, 'TATGTC': 46, 'CCGGTT': 46, 'CGGACG': 46, 'CAGGTT': 46, 'TATGGC': 45, 'AATGGT': 45, 'CAGCAT': 45, 'GCATCA': 45, 'GCTCTC': 45, 'GCCCAT': 45, 'ACTCGC': 45, 'CTGCAA': 45, 'CTGGTT': 45, 'GGGCCG': 45, 'TGGACC': 45, 'AATCAC': 45, 'TTGATC': 45, 'AAACTT': 45, 'AAATCT': 45, 'GGTATC': 45, 'TCGTCG': 45, 'TGGTAC': 45, 'GGTCGG': 45, 'AAAATG': 45, 'CCGCTT': 45, 'AAACGG': 45, 'CCCATC': 45, 'CCCGGG': 45, 'ATGAAT': 45, 'TTAAAT': 45, 'CACAGC': 45, 'AGCGCT': 45, 'ATGGAA': 45, 'TCCTTC': 45, 'CAGCGA': 45, 'AAAGTT': 45, 'CTTGAC': 45, 'GCGAAT': 44, 'GATTAT': 44, 'ACACCC': 44, 'GGAGCA': 44, 'GGATGG': 44, 'ATTTAT': 44, 'GCCTAT': 44, 'TTTAAA': 44, 'GGACTG': 44, 'GAATGC': 44, 'ACTCCC': 44, 'GGTGAT': 44, 'CACAAC': 44, 'GGTGGG': 44, 'GGGCTT': 44, 'CCCCAC': 44, 'ATTGAC': 44, 'TTCGGT': 44, 'GTGTAC': 44, 'AGCCAT': 44, 'CTGGGA': 44, 'AGCGAA': 44, 'GGGATT': 44, 'GTGGCA': 44, 'CTGTGG': 44, 'GGCAGT': 44, 'AAATTA': 44, 'GGTGTC': 44, 'ATTGAA': 44, 'CGCTCG': 44, 'ATTCTC': 44, 'GTGAGG': 44, 'GGTCAG': 44, 'TCGCTG': 44, 'ACAGGC': 43, 'AAATCG': 43, 'GTCGCA': 43, 'GTCGAA': 43, 'GAACGG': 43, 'GACGTG': 43, 'TACCTC': 43, 'AAGATT': 43, 'GAACTT': 43, 'GAGATG': 43, 'GTGGGT': 43, 'AAACAT': 43, 'GCCGTT': 43, 'AGTTTC': 43, 'AATTTT': 43, 'CAGCTG': 43, 'GTTGAG': 43, 'TCTCGC': 43, 'GGCTAT': 43, 'GATGCA': 43, 'ATGGCT': 43, 'GGACGT': 43, 'TCCGAG': 43, 'CATGGA': 43, 'AGTGAA': 43, 'GATCGA': 43, 'AGGGAT': 43, 'AAGTCG': 43, 'CAGCGT': 42, 'GAAGCA': 42, 'GTTGAC': 42, 'GTTGTC': 42, 'ATCTGC': 42, 'AAGAGT': 42, 'AGGGCG': 42, 'GATCAT': 42, 'AATATC': 42, 'GGACTC': 42, 'ACCGGT': 42, 'GGGGCA': 42, 'GAAAGA': 42, 'AGGGTC': 42, 'TACGAG': 42, 'GGTCTC': 42, 'GGTCTG': 42, 'CCGCCC': 42, 'GTGCGA': 42, 'CCGCCG': 42, 'GTGCAT': 42, 'GGGTAC': 42, 'GAACAG': 42, 'TCATCC': 42, 'CATGGT': 41, 'ACCGGA': 41, 'TTGGCT': 41, 'CTTTTT': 41, 'GTTGAT': 41, 'TCCCGA': 41, 'AAAAGA': 41, 'TTGAAT': 41, 'CGCCCA': 41, 'CAAAAT': 41, 'TGCGGG': 41, 'AACAAA': 41, 'AATAAC': 41, 'AAGCGA': 41, 'ATGCTG': 41, 'CTGATG': 41, 'CGCTGC': 41, 'CGTACC': 41, 'ATCAAA': 41, 'TGGAAA': 41, 'AATAGT': 41, 'GGACAG': 41, 'GATAAC': 41, 'GGTTTT': 41, 'AGGACC': 41, 'GTAGCC': 41, 'ATGTCC': 41, 'AAACGA': 41, 'CCCGAA': 41, 'TTTATT': 41, 'GATCGG': 41, 'GGGTAT': 41, 'GAACTG': 41, 'AGAGGC': 41, 'ATGCCG': 41, 'TCCCGG': 41, 'AGGCTC': 41, 'TTCGCG': 41, 'ATTGAG': 41, 'TTCAAG': 41, 'TGCCCC': 41, 'CGTGGT': 41, 'CAGTTT': 41, 'AAACCT': 41, 'AAGATG': 41, 'CTGAAT': 40, 'CCTGAC': 40, 'GTTCAC': 40, 'GACCCC': 40, 'TGCCAC': 40, 'ACCGAA': 40, 'TTGGGG': 40, 'GCAGGT': 40, 'CCCGGT': 40, 'TTGCCT': 40, 'AACTGG': 40, 'CATCAT': 40, 'AGGTTC': 40, 'GGGGGT': 40, 'GCACGG': 40, 'ACGGAA': 40, 'GCTTAC': 40, 'TCCCAC': 40, 'TGCAGC': 40, 'GATGTG': 40, 'GCCCCT': 40, 'AGGAGT': 40, 'CGCCCT': 40, 'TGGCCC': 40, 'CTCCTC': 40, 'AACGCT': 40, 'CACCTG': 40, 'TCAAAA': 40, 'ACCCTG': 40, 'AACTAC': 40, 'GCGTGG': 40, 'CAACTC': 40, 'ACATCC': 40, 'GCAGAA': 40, 'GGTGGA': 40, 'CAAGAT': 40, 'AGAAGC': 40, 'TTGGAG': 40, 'CGCCAT': 40, 'ACGAGG': 39, 'CAAAAG': 39, 'CAAGAA': 39, 'AAACGT': 39, 'CGGGAA': 39, 'AATCAT': 39, 'AGGGGG': 39, 'AAATGG': 39, 'TCCGCA': 39, 'AACCTG': 39, 'CGCCAG': 39, 'CGACCG': 39, 'CTGCGA': 39, 'TTGATT': 39, 'AACGGG': 39, 'AGCCGA': 39, 'GGAGCT': 39, 'TTGTTT': 39, 'GAACAA': 39, 'TGGCAG': 39, 'CATTTC': 39, 'GCGTTG': 39, 'ACGCAG': 39, 'TCCGCT': 39, 'GTACGC': 39, 'CCCCGT': 39, 'GCTCAG': 39, 'GGTAAA': 39, 'GCGGTA': 39, 'AAATTT': 39, 'TCCGAA': 39, 'GTAAAA': 39, 'ACTGAC': 38, 'ATGATG': 38, 'TTGAGG': 38, 'AAACCA': 38, 'CGCGTG': 38, 'CAAGCT': 38, 'GGAGTG': 38, 'GGGACG': 38, 'GATGGT': 38, 'GCACGA': 38, 'TTTAAC': 38, 'TGGGCG': 38, 'CGTCGT': 38, 'TTGCGT': 38, 'ATGCAG': 38, 'GGCATG': 38, 'CCATCC': 38, 'ATCCAG': 38, 'AGCCCG': 38, 'GAACGT': 38, 'AGCTGC': 38, 'GGAATT': 38, 'TCCGGT': 38, 'AAGTAT': 38, 'GAATCG': 38, 'GGACGA': 38, 'GATCGT': 38, 'GCAACC': 38, 'CACGCG': 38, 'TTCCTG': 38, 'GGATTT': 38, 'CCTGCT': 37, 'CGTCAT': 37, 'CCGTGG': 37, 'CTCCAC': 37, 'TTGGCG': 37, 'GTGGGA': 37, 'TTCGGG': 37, 'CGACTC': 37, 'CACACC': 37, 'AACGAA': 37, 'TTAGAT': 37, 'CCGAAT': 37, 'AGGGAA': 37, 'ACGCGA': 37, 'GGGTTG': 37, 'TTGTCC': 37, 'ATGACC': 37, 'AGAGAT': 37, 'AGCGAT': 37, 'GCCATT': 37, 'AACCAG': 37, 'GCTTTC': 37, 'CGCGAA': 37, 'TTCCGT': 37, 'TCGAAG': 37, 'GGTCGT': 37, 'TGCTCC': 37, 'GCCCCA': 37, 'AAGTGG': 37, 'CCCCTG': 37, 'GACGGA': 37, 'AGCAAA': 37, 'GATGGG': 37, 'CGCAGG': 37, 'GCTGTT': 37, 'TTTTAT': 37, 'TCCCCC': 37, 'CGGAGT': 37, 'CAGAAT': 37, 'GCATGG': 37, 'GTGAAT': 37, 'ACCGTG': 36, 'ACTGTC': 36, 'AGCTCC': 36, 'TGGCGA': 36, 'GCAGGG': 36, 'TTTGTT': 36, 'GGCCCG': 36, 'TTCCGG': 36, 'AGAAAC': 36, 'CGTCGG': 36, 'TGGCGT': 36, 'CGGTGG': 36, 'CACGGG': 36, 'CATGCA': 36, 'CGCAAA': 36, 'TTTATC': 36, 'CACATC': 36, 'AAGAGA': 36, 'TTGGTC': 36, 'AATGCG': 36, 'GACAAT': 36, 'TGGCAC': 36, 'GGACAT': 36, 'CTGTTT': 36, 'TCCATC': 36, 'AAGTTT': 36, 'AATACT': 36, 'GTGTTT': 36, 'AATAGC': 36, 'GTGATT': 36, 'CAGATT': 36, 'GGTAAG': 35, 'CAGATG': 35, 'TTGACC': 35, 'AGTGAC': 35, 'GATACC': 35, 'CAATTC': 35, 'ATTCAC': 35, 'CAGTGG': 35, 'GATAAT': 35, 'ACGAAT': 35, 'GCAGGA': 35, 'CCGATG': 35, 'GATTAC': 35, 'CGGTCA': 35, 'ATCAGT': 35, 'TTGCAC': 35, 'TCGCCA': 35, 'CCGTTT': 35, 'TTTGAA': 35, 'CACGGT': 35, 'ACGCAT': 35, 'AAATTG': 35, 'GCTGGT': 35, 'ACAGAA': 35, 'AACCCG': 35, 'AAACCG': 35, 'CGAGAC': 35, 'AAAGTG': 35, 'AACCGT': 35, 'CCGTAC': 35, 'TACATC': 35, 'CGGAGG': 34, 'GCTCCG': 34, 'GTGCTT': 34, 'CCTTTC': 34, 'GATTCA': 34, 'GAAGTA': 34, 'GTCGGA': 34, 'CGTAAC': 34, 'TCGGAA': 34, 'AATCGT': 34, 'TCGCGT': 34, 'ATAAAG': 34, 'AGTAAT': 34, 'ATCTGG': 34, 'TCCGAT': 34, 'GAGGTA': 34, 'GAACGA': 34, 'CGCACG': 34, 'GGAGTT': 34, 'GGACAA': 34, 'CACCGG': 34, 'GATAGC': 34, 'AGGATC': 34, 'GTTAAA': 34, 'CGTGCG': 34, 'GGTCAT': 34, 'GCTTCT': 34, 'CGGAAT': 34, 'GCAAAG': 34, 'GAGTCA': 34, 'CAAACC': 34, 'GTGCAA': 34, 'ACTTTC': 34, 'CAAGCA': 34, 'ATTATT': 34, 'TATTCC': 34, 'AAAGTA': 34, 'ACGCTT': 34, 'GCATGC': 34, 'CGTATC': 34, 'TTTGGT': 34, 'TTGCAT': 34, 'TATAAA': 34, 'TACAAC': 34, 'AAACTG': 34, 'GCTCGA': 34, 'ACGCCT': 34, 'CGGAAA': 34, 'GTGTCG': 34, 'ACAAAA': 34, 'TCAGAC': 34, 'CCCGTG': 33, 'AATCCA': 33, 'GCAAAT': 33, 'GCGACA': 33, 'CAGCTT': 33, 'CACCCG': 33, 'CCAGAC': 33, 'TCGGTG': 33, 'CAAGGT': 33, 'AATCAG': 33, 'GGCATT': 33, 'CCTGAT': 33, 'GTTTAT': 33, 'GAAATA': 33, 'TGGGGG': 33, 'GTAGGC': 33, 'GCTGGA': 33, 'TTGCCG': 33, 'CCCCTC': 33, 'CGGGTT': 33, 'TTTTTT': 33, 'GGATTG': 33, 'GTAAAC': 33, 'ATCATT': 33, 'GCGTTT': 33, 'GCCAGG': 33, 'TCGCCG': 33, 'ACGGTT': 33, 'GAACAT': 33, 'ACATTC': 33, 'GGCTCG': 33, 'CGTCTC': 33, 'TTCAAT': 33, 'AATTCT': 33, 'CGGCAA': 33, 'ATAGAC': 33, 'TTCAAA': 33, 'AAGTCA': 33, 'CAGTCA': 33, 'AATTAC': 33, 'AATCTC': 33, 'CTGACG': 33, 'GGCCTT': 33, 'AGCAAT': 33, 'ACGTAC': 33, 'CGCTAT': 32, 'GACAGT': 32, 'AGCGGG': 32, 'GTCCGT': 32, 'TATTTT': 32, 'ACTTCC': 32, 'GGTCAA': 32, 'ATTCAA': 32, 'CACGCT': 32, 'GCCCTT': 32, 'GATCTT': 32, 'GGATAC': 32, 'CTTTAT': 32, 'CTCAAT': 32, 'ACGACG': 32, 'ACGGAT': 32, 'CAGAGT': 32, 'AGTTCC': 32, 'ATGCGT': 32, 'GAATTA': 32, 'AAAAGG': 32, 'TCGATG': 32, 'CGACGG': 32, 'ATTTCT': 32, 'AGGGTG': 32, 'GCGTCT': 32, 'GATGGA': 32, 'TGGCGG': 32, 'TGCGAG': 32, 'GGTTCG': 32, 'CAGCAA': 32, 'GTTGCT': 32, 'CTCGCA': 32, 'CCCTAC': 32, 'ACTGGC': 32, 'GAATCA': 32, 'GATTGG': 32, 'ACCCCA': 32, 'GGTGCT': 32, 'ATGTAC': 32, 'TTCTGC': 32, 'AAAACT': 32, 'AGAGAC': 32, 'GGGTGG': 32, 'TTATTT': 32, 'CGGTTG': 32, 'CAATCA': 32, 'AGAACC': 32, 'GGAATG': 31, 'ATTATC': 31, 'CAATAT': 31, 'AGGCAC': 31, 'AACAGT': 31, 'GCACAT': 31, 'AGGGGT': 31, 'GTTTAC': 31, 'AATACC': 31, 'AGGGGA': 31, 'GACAGG': 31, 'CACGAT': 31, 'CATCTC': 31, 'AGTAAG': 31, 'TCTTTT': 31, 'CGTCCT': 31, 'CCTGTC': 31, 'TGGTCC': 31, 'ATGTAT': 31, 'ATCGTT': 31, 'CTCGCT': 31, 'TTTGCT': 31, 'AGACCC': 31, 'CGAGGG': 31, 'GTGTTG': 31, 'AGACGC': 31, 'CATGGG': 31, 'GAATCT': 31, 'TGCAAC': 31, 'ACACCG': 31, 'AGGTCC': 31, 'AATCCG': 31, 'GCCTCA': 31, 'CCCCGA': 31, 'GCTATC': 31, 'ATTTTG': 31, 'GGCTTT': 31, 'ACAACC': 31, 'CAAGCG': 31, 'AGGAGG': 31, 'GATTTG': 31, 'TCCAAT': 31, 'GTGTAT': 31, 'GGTATT': 31, 'CCATTC': 31, 'TATGAT': 31, 'GGATAT': 31, 'ATGGTG': 31, 'CTTCTC': 31, 'CCGTCA': 31, 'GTCTGC': 30, 'TTAATT': 30, 'GTAGAT': 30, 'CCCAAG': 30, 'CAAGAG': 30, 'CGGCTT': 30, 'GTTGCA': 30, 'CGTGAG': 30, 'CTTGTC': 30, 'TCAGCT': 30, 'TATCCC': 30, 'GGTGTT': 30, 'GCAGTT': 30, 'TGGGTG': 30, 'TGGCCG': 30, 'AATCCT': 30, 'AGCGGT': 30, 'ACTGAT': 30, 'GGCACG': 30, 'TTGAGT': 30, 'TTCGAA': 30, 'AAGTCT': 30, 'ATCCTG': 30, 'CACGGA': 30, 'AGTAAA': 30, 'CCGAGG': 30, 'GCGTAT': 30, 'GTCCCG': 30, 'AACTGC': 30, 'GAGTTG': 30, 'GTTGGT': 30, 'CCGTGC': 30, 'GAACCG': 30, 'GGTCGA': 30, 'CATGTT': 30, 'CCAGCA': 30, 'CAATGG': 30, 'GCACAA': 30, 'CCGCAT': 30, 'ATTCGT': 30, 'AGGAAT': 30, 'TCAAGC': 30, 'CGTGGG': 30, 'ATGAGG': 30, 'GGAAGT': 30, 'ACCCGT': 30, 'GATAGT': 30, 'GTAGAG': 30, 'AATATG': 30, 'GAAACT': 30, 'CCGACG': 30, 'AGCAGG': 30, 'AGTGAT': 30, 'CTTTCT': 30, 'GAGTAT': 30, 'CTTCAC': 30, 'AACCAT': 29, 'GAGTCT': 29, 'CTCTGG': 29, 'TCGCCT': 29, 'CGTTCG': 29, 'CGCGTT': 29, 'TTGGCA': 29, 'AGCGTT': 29, 'CTGTAT': 29, 'CCATGC': 29, 'TTAAAG': 29, 'ATTGTC': 29, 'GTGCCT': 29, 'ATGATT': 29, 'CCAAAA': 29, 'ATCGCT': 29, 'CGTCGA': 29, 'TACTAC': 29, 'TACAGC': 29, 'AGAATC': 29, 'TATCTG': 29, 'GCTAAA': 29, 'AGGAGA': 29, 'TACGAA': 29, 'GCTCGT': 29, 'GTGTGC': 29, 'CTGAGT': 29, 'GGAAGA': 29, 'GATCCC': 29, 'GCTTTT': 29, 'CGTTTT': 29, 'CCTGAG': 29, 'ATTCAG': 29, 'CCCTCG': 29, 'TTCTAC': 29, 'AGAGGG': 29, 'AGATCC': 29, 'ACAAAG': 29, 'TTCCAG': 29, 'TGCCAG': 29, 'GTAAGC': 29, 'GACATG': 29, 'ACCAAT': 29, 'CTGTGC': 29, 'ACGGGT': 29, 'CAACCG': 29, 'GCGAGA': 29, 'CGAGGA': 29, 'CGCAAT': 29, 'GTCCGA': 29, 'AACCAA': 28, 'CGCTCA': 28, 'GGTCTT': 28, 'GCCTCT': 28, 'ATTAAT': 28, 'ACTCTC': 28, 'CGCATG': 28, 'ATCCGT': 28, 'AGGCGG': 28, 'TCACAC': 28, 'AGAGAA': 28, 'TATATC': 28, 'AGCTTT': 28, 'TCAATC': 28, 'CACGCA': 28, 'GGACTT': 28, 'TGGCTT': 28, 'GCTTCG': 28, 'TATCAA': 28, 'TTTTCA': 28, 'GAATTG': 28, 'GCCACG': 28, 'ACCCAG': 28, 'AGGCTG': 28, 'TCGGAT': 28, 'TGGGTC': 28, 'TCGGGG': 28, 'AGTGTC': 28, 'TTGCTG': 28, 'CAACAC': 28, 'GTTAAT': 28, 'TATTAT': 28, 'TTCGGA': 28, 'AGTCAC': 28, 'CAAGGG': 28, 'CATTAT': 28, 'GTGACG': 28, 'CGGACA': 28, 'TCCCTC': 28, 'TACCCC': 28, 'GGTTTG': 28, 'CTCGAG': 28, 'CCGAAA': 28, 'ACGATG': 28, 'ACAGAC': 28, 'TTTTCT': 28, 'ACCGTT': 28, 'TCATGC': 28, 'TTTGTC': 28, 'CTCCGG': 28, 'AGGCAG': 28, 'AGCGTG': 28, 'AATGCA': 28, 'CAACAA': 28, 'TACCAC': 28, 'CGTAGC': 28, 'CCTCGG': 28, 'TTTGCG': 28, 'TCTAAA': 27, 'TCTGAA': 27, 'TTGTTG': 27, 'TCGTAC': 27, 'ATTGCT': 27, 'GAGACG': 27, 'ATCGGA': 27, 'TTGGTG': 27, 'ATGCGG': 27, 'TTCCCC': 27, 'AGGTAC': 27, 'CATCCG': 27, 'CCTGCG': 27, 'ACGTGG': 27, 'TATCTC': 27, 'TCCAAG': 27, 'TCAGGC': 27, 'TTTCAA': 27, 'AGGCAT': 27, 'TCACCG': 27, 'GTCCTG': 27, 'ACCTAT': 27, 'ATTAAG': 27, 'TCAGCA': 27, 'TACCAG': 27, 'GATTCG': 27, 'CAAGGA': 27, 'TCTGCT': 27, 'AGGGCT': 27, 'TTAGAA': 27, 'TTGGGT': 27, 'ACAAAT': 27, 'TGCGGT': 27, 'GTTTTT': 27, 'CAGTAT': 27, 'TCGCAG': 27, 'ATTGGG': 27, 'GTTCTC': 27, 'GCACTG': 27, 'TCCGGG': 27, 'ATCCCG': 27, 'ACACAC': 27, 'CAATTT': 27, 'CCATGG': 27, 'ACTGCG': 27, 'GTGTCA': 27, 'CGAATC': 27, 'GAAATG': 27, 'AGGCGT': 27, 'CTTGGT': 27, 'GACCAT': 27, 'CAACCT': 27, 'ATAGAT': 27, 'CGCAGT': 27, 'GCTTAT': 27, 'GATTCT': 27, 'ACCCCT': 27, 'AAGATA': 27, 'AACCGA': 26, 'CAACGG': 26, 'GATTTA': 26, 'GCGTGC': 26, 'TCCTAC': 26, 'GGGTTT': 26, 'AGAAAT': 26, 'GCTGTG': 26, 'AGTTTT': 26, 'AACAGG': 26, 'GGCCAA': 26, 'AGTAAC': 26, 'CGTCAG': 26, 'TTGGTT': 26, 'GAGAGT': 26, 'CTGTTG': 26, 'AACATG': 26, 'CGATCA': 26, 'CTTGTT': 26, 'ATGTGG': 26, 'AGGCGA': 26, 'TACCGG': 26, 'GCCATG': 26, 'TTGCAG': 26, 'TTTCAC': 26, 'TACCTG': 26, 'GTTTTC': 26, 'TCTCAT': 26, 'GCGATT': 26, 'GTTTCT': 26, 'GATCAA': 26, 'GTGTGG': 26, 'TCGCAT': 26, 'CTCGGA': 26, 'GTGAGT': 26, 'TCCCAT': 26, 'TCCCGT': 26, 'ATGGGT': 26, 'CATTTT': 26, 'CGGTTT': 26, 'ATTCTT': 26, 'TCATTC': 26, 'TACCGT': 26, 'TTTCCC': 26, 'GTTCAG': 26, 'GCATCT': 26, 'GGCTCT': 26, 'CGCATT': 26, 'GGCTTG': 26, 'GCCCAA': 26, 'TCGGTT': 26, 'AATCGA': 26, 'ATAAAT': 26, 'CACGAG': 26, 'CCAAGC': 26, 'ATTAGC': 26, 'GACCAA': 26, 'ATTCCT': 26, 'GTCAAA': 26, 'AATTTA': 26, 'TCAGAA': 26, 'GGTTAT': 26, 'ACTGCT': 26, 'GCTACC': 26, 'AGCGGA': 26, 'TCAGAT': 26, 'AACTCT': 26, 'ATTTTA': 26, 'CTCAAA': 25, 'CACTGG': 25, 'ATGCCT': 25, 'CTAGCC': 25, 'CTTGAG': 25, 'TGGCAT': 25, 'CACTGC': 25, 'ACAGCT': 25, 'ACTTCA': 25, 'GCAAGT': 25, 'ACCCAT': 25, 'TCGAGG': 25, 'AAGGTA': 25, 'CACCGT': 25, 'ATAGCC': 25, 'AAGTTG': 25, 'ACGAAA': 25, 'AGATTT': 25, 'ATGCCA': 25, 'ATCAGG': 25, 'TCTTCT': 25, 'AGATAT': 25, 'GAACCT': 25, 'TTTAGT': 25, 'CGAGCG': 25, 'AATCTT': 25, 'CATAAA': 25, 'CACCGA': 25, 'TTTATG': 25, 'ACCAGG': 25, 'AGAGCG': 25, 'TTAAGC': 25, 'TCCTGG': 25, 'AGAATT': 25, 'ACACCA': 25, 'AGGCCG': 25, 'GACGTT': 25, 'ACCTTT': 25, 'ATCCAT': 25, 'CACAAG': 25, 'GAGCTA': 25, 'AATATA': 25, 'CCCCAT': 25, 'TTACCC': 25, 'TATGTT': 25, 'GCAGTG': 25, 'GAGTTA': 25, 'AGTGTT': 25, 'TCCCCG': 25, 'TCAGTT': 25, 'GTCGTG': 25, 'GGTACC': 25, 'AGTCAT': 25, 'GGTGCA': 25, 'AACATT': 25, 'TTAAGT': 25, 'AAGTTA': 25, 'CGATTC': 25, 'ACGGGA': 25, 'TCGACG': 25, 'TCTCCC': 25, 'CAGTCT': 25, 'AATTCG': 25, 'AGCTAT': 25, 'CTGAGG': 25, 'TTCAGT': 25, 'CGTTGG': 25, 'TTATTA': 25, 'AGCCAA': 25, 'GTCTGG': 25, 'GATATG': 24, 'CGTGAT': 24, 'TATTAC': 24, 'TTAAAC': 24, 'CTTTAC': 24, 'GGCACA': 24, 'TTTCAT': 24, 'GTTCGT': 24, 'ATGGTT': 24, 'CTTAAT': 24, 'GGGTGC': 24, 'TTATCC': 24, 'ACAGAT': 24, 'ATTGTG': 24, 'TGGACG': 24, 'AAGACT': 24, 'ACCAGT': 24, 'TCGCTT': 24, 'TCGTCA': 24, 'ATGCTT': 24, 'AAATGC': 24, 'AACCTT': 24, 'AGAGGT': 24, 'GTTGCG': 24, 'AACGGA': 24, 'TACTCC': 24, 'CACATG': 24, 'GCGCTA': 24, 'ACGCAA': 24, 'GTCCAG': 24, 'AGGCTT': 24, 'GCTCGG': 24, 'TCTTTC': 24, 'CTTTCG': 24, 'ACGCGT': 24, 'GGAATA': 24, 'TCTCAC': 24, 'ACAAAC': 24, 'TCAGCG': 24, 'GTATTC': 24, 'CCTCGA': 24, 'TCAGAG': 24, 'GCCTTT': 24, 'AGTGCT': 24, 'ATTAGT': 24, 'ATCGTG': 24, 'GCTCCT': 24, 'ACTTTT': 24, 'CGTTCA': 24, 'CTGTCA': 24, 'CAATAC': 24, 'GCACCT': 24, 'GCGACT': 24, 'TGGGGA': 24, 'GGACCG': 24, 'AAGTGC': 24, 'GACCCG': 24, 'TCCCTG': 24, 'ATATTC': 24, 'TACGAT': 24, 'CTACGC': 24, 'CTCCCG': 24, 'CGGTAT': 24, 'AGCCTT': 24, 'GCTTCA': 23, 'GAGATA': 23, 'TCAAAC': 23, 'TTGATG': 23, 'TGGGGT': 23, 'TATTTG': 23, 'TCAAAT': 23, 'TTGCGG': 23, 'ACAAGC': 23, 'CGTTAT': 23, 'CCGTTG': 23, 'GTTTTA': 23, 'ATTTAC': 23, 'CATCTG': 23, 'ACTCGA': 23, 'GGTTCT': 23, 'CGGGTA': 23, 'TATATT': 23, 'CGACGA': 23, 'CTTCCG': 23, 'AAGCTA': 23, 'CGATAC': 23, 'TTCGCA': 23, 'TTTAGC': 23, 'ATGGCA': 23, 'GTCAAT': 23, 'ATTAAC': 23, 'ACAGTC': 23, 'CCGTCT': 23, 'AACGTG': 23, 'GTTCCG': 23, 'GGGAGA': 23, 'CCACGA': 23, 'ATTCAT': 23, 'TTAGCC': 23, 'GCTGGG': 23, 'GGGTCT': 23, 'CGCCAA': 23, 'GAACCA': 23, 'TACAAG': 23, 'TGGATT': 23, 'CTCTAT': 23, 'GCTCAA': 23, 'CGTGCT': 23, 'TCTGGC': 23, 'GTCTCG': 23, 'GTTGTT': 23, 'TATAAT': 23, 'TTAATC': 23, 'GCTTGC': 23, 'GGTCCC': 23, 'CCTCAG': 23, 'CTCCGA': 23, 'CTTGAA': 23, 'ACGTCA': 23, 'AATGGG': 23, 'CGAGGT': 23, 'CACCAT': 23, 'GATTGC': 23, 'ATTGGT': 23, 'CAAGTT': 22, 'AGAAGA': 22, 'TCCGGA': 22, 'GCTCAT': 22, 'CTTAGC': 22, 'ATAGAA': 22, 'CCGGTA': 22, 'GAGCCA': 22, 'TTGCTT': 22, 'CCATTT': 22, 'TGCTTC': 22, 'TCGAAT': 22, 'TTTGCA': 22, 'TGCGCG': 22, 'AGTCCC': 22, 'CCCCAG': 22, 'GCTTTG': 22, 'GCCTTG': 22, 'ACCTCG': 22, 'AACCCT': 22, 'TGGCAA': 22, 'ATGCAT': 22, 'CGGATT': 22, 'TTAGGC': 22, 'AACTAT': 22, 'AGAGGA': 22, 'CTGATA': 22, 'GTCAGT': 22, 'TGTGCC': 22, 'TACGCG': 22, 'GCATTG': 22, 'CTACCC': 22, 'ATCTAT': 22, 'TGGGCA': 22, 'TCACCC': 22, 'TCCTGC': 22, 'AGGACG': 22, 'AGTTGG': 22, 'TCGAAA': 22, 'GGTGTG': 22, 'ACGTTT': 22, 'GACTCG': 22, 'AGTTAT': 22, 'GGTTGG': 22, 'CCCGGA': 22, 'ATATCC': 22, 'CTTAAC': 22, 'TATGCT': 22, 'TCGGGT': 22, 'AGTGCG': 22, 'CGTGCA': 22, 'CCACAC': 22, 'TGGAGG': 22, 'TCGCGA': 22, 'TTTGGG': 22, 'CAATCT': 22, 'AATCTG': 22, 'TCCAAA': 22, 'CTTCAT': 22, 'ATTGCG': 22, 'TGCGAT': 22, 'GTCGTT': 22, 'TCCTTT': 21, 'CATTCG': 21, 'CGAAAA': 21, 'CATCAG': 21, 'TGCGTC': 21, 'ATCGCA': 21, 'GCTCTG': 21, 'CTCGAA': 21, 'TCGGGA': 21, 'GTCTAT': 21, 'TGCAAG': 21, 'CTCGTT': 21, 'GTAGCG': 21, 'TCGCAA': 21, 'ATGTCG': 21, 'ATGCGA': 21, 'AGTGGC': 21, 'AACTTT': 21, 'TTCGCT': 21, 'CTGATT': 21, 'CAACAG': 21, 'GTAATC': 21, 'TCTGAT': 21, 'ATTACC': 21, 'ATATAT': 21, 'GGAACG': 21, 'GAAACA': 21, 'CGATCG': 21, 'CTATTT': 21, 'GAATGT': 21, 'TACACC': 21, 'TCTGAC': 21, 'ATTCTG': 21, 'ATTCGA': 21, 'CATAAC': 21, 'CAACTT': 21, 'CGACAC': 21, 'TCAATT': 21, 'ACGTGC': 21, 'CGCTCT': 21, 'CCTGTG': 21, 'GCTCCA': 21, 'CACCAG': 21, 'GCAAGG': 21, 'GTGTCT': 21, 'GTTCAA': 21, 'ACTCTT': 21, 'CCTCCC': 21, 'CGTCCA': 21, 'AATTCA': 21, 'TGCACC': 21, 'CAGAGG': 21, 'ACAGGT': 21, 'ATAAGC': 21, 'ACACCT': 21, 'CCACGG': 21, 'GTATCC': 21, 'AAGACA': 21, 'GCTAAG': 21, 'TTTAAG': 21, 'TTACGC': 21, 'CGAACG': 21, 'CTTATC': 21, 'TATCGT': 21, 'CTAAAA': 21, 'CCAATC': 21, 'CATATC': 21, 'CCCCCC': 21, 'CTTGCT': 20, 'TTTGGA': 20, 'GTCTTT': 20, 'TGCCGG': 20, 'TACTGC': 20, 'AATTGG': 20, 'TTCCAT': 20, 'ATACAA': 20, 'ACACTC': 20, 'CACCCT': 20, 'CATTCT': 20, 'GCCCTA': 20, 'TCATGG': 20, 'TTGGAA': 20, 'TTAGGT': 20, 'TCTTTG': 20, 'AGATAC': 20, 'TCACAA': 20, 'AGAGCT': 20, 'CGAGCA': 20, 'TCTCTT': 20, 'GTTTGC': 20, 'AAATGT': 20, 'CTTGCG': 20, 'CCTGGA': 20, 'AGAAGG': 20, 'AGGCCT': 20, 'TTCTTT': 20, 'GTTCCT': 20, 'GAGTGT': 20, 'AAAACA': 20, 'AATTGC': 20, 'TCAGGA': 20, 'CCTGAA': 20, 'GCAACG': 20, 'GGATGC': 20, 'GATTGT': 20, 'CTCCTG': 20, 'TCTTCA': 20, 'GACATT': 20, 'ACTTAC': 20, 'CTGACT': 20, 'TTACTT': 20, 'TTTTTG': 20, 'CACAAT': 20, 'TTGTCT': 20, 'TCCTCG': 20, 'GCACGT': 20, 'TCATCG': 20, 'ATACGC': 20, 'AGCCCT': 20, 'GCCACA': 20, 'AGGTCG': 20, 'CGTCTT': 20, 'CAACTG': 20, 'CAAGTC': 20, 'CCTTTT': 20, 'GTATTT': 20, 'TCCTCT': 20, 'CAATCG': 20, 'ATCCCT': 20, 'TCGTGG': 20, 'AGGATT': 20, 'TTCTAT': 20, 'CAAATT': 20, 'AGTCGT': 20, 'GGTAGT': 20, 'ACTTCT': 20, 'TTCATT': 20, 'GTAGAA': 20, 'CAACGT': 20, 'CAGAGA': 20, 'AGCAGT': 20, 'AGTCGA': 20, 'TGCAAA': 20, 'ATCCGA': 20, 'TCTTCG': 20, 'ACTAAT': 19, 'AGGCCA': 19, 'TTCCCG': 19, 'CCTCAC': 19, 'TTTCGG': 19, 'TTGTAT': 19, 'CTCTCG': 19, 'CATTCA': 19, 'TCAACC': 19, 'CACAAA': 19, 'AGTTTG': 19, 'CAACAT': 19, 'AGTTCG': 19, 'CGTAAA': 19, 'ATGCAA': 19, 'GGCACT': 19, 'GTTCTT': 19, 'TGGACA': 19, 'AAACTA': 19, 'TCTAAT': 19, 'AGCAGA': 19, 'GTTAAG': 19, 'ACAGCG': 19, 'AGTCTC': 19, 'CAAATC': 19, 'ATGTTG': 19, 'ATAAGT': 19, 'GGTAAT': 19, 'TATCAT': 19, 'ATTGGA': 19, 'TGTCCC': 19, 'CCACTG': 19, 'GTAAAG': 19, 'CCAGAT': 19, 'ACAGCA': 19, 'TGGAAT': 19, 'GCTATT': 19, 'CTGTCT': 19, 'TTAGTT': 19, 'AGTGGA': 19, 'CGCTTT': 19, 'TATTGC': 19, 'TATAAC': 19, 'TCCAGT': 19, 'ATGGGA': 19, 'CCGCTA': 19, 'AGATTC': 19, 'TATCCG': 19, 'CATTAC': 19, 'ACTGAA': 19, 'GCTAAC': 19, 'CGAGTC': 19, 'CATCGT': 19, 'AGGTTT': 19, 'ATTTGT': 19, 'CCTGTT': 19, 'GAACTA': 19, 'CGTGAA': 19, 'ACTCAT': 19, 'CGACGT': 19, 'ACATGG': 19, 'GGATCA': 19, 'CTCGTG': 19, 'GTTCAT': 19, 'TTTACC': 19, 'GTTCGG': 19, 'CCAGAG': 19, 'TCTGAG': 19, 'CATAGC': 19, 'GTTTTG': 19, 'ACGTCG': 19, 'GGCTGT': 19, 'AGTATT': 19, 'GGGCCT': 19, 'CTCTTT': 19, 'GTCAGG': 19, 'TCACCA': 19, 'GCTTGG': 18, 'CGAAAT': 18, 'AGACGA': 18, 'GATAGA': 18, 'TTAGAG': 18, 'ACTTAT': 18, 'GGCAGA': 18, 'GCCTGT': 18, 'GTAGCT': 18, 'ATTGCA': 18, 'CCATCA': 18, 'AGTCGG': 18, 'ATGTTT': 18, 'CGTGTT': 18, 'GGATTA': 18, 'GCCAGA': 18, 'AGCATG': 18, 'CTTTTG': 18, 'ACTCCA': 18, 'TTGCCA': 18, 'AATTTG': 18, 'TGTCAG': 18, 'ACCAAA': 18, 'TTATTC': 18, 'CAGACT': 18, 'ATTGTT': 18, 'CCAAAG': 18, 'CGAAGG': 18, 'TATCTT': 18, 'AACCCA': 18, 'GTCATG': 18, 'CTCTGC': 18, 'TTTCTC': 18, 'TTCAGG': 18, 'TATTTA': 18, 'TACGGG': 18, 'GTTTCG': 18, 'GCATTT': 18, 'TCCTCA': 18, 'TCACTC': 18, 'CGTTCT': 18, 'CCAGTC': 18, 'GTTAGC': 18, 'CCGATT': 18, 'AGAATA': 18, 'GCATAT': 18, 'ACACTT': 18, 'CACGAA': 18, 'TCCCAA': 18, 'GTATCT': 18, 'TTCTCG': 18, 'AATGTA': 18, 'ATAAAC': 18, 'TCAGTC': 18, 'CCCAGT': 18, 'TTTGTG': 18, 'TTAACC': 18, 'CAACCA': 18, 'CCAGCG': 18, 'CGGATA': 18, 'GATATA': 18, 'GCAATT': 18, 'CTTCTT': 18, 'AGTTAC': 18, 'GGCCCA': 18, 'CAGTGC': 18, 'CGTATT': 18, 'TCATCA': 18, 'GCTAAT': 18, 'ACAATC': 18, 'GGCTCA': 18, 'GACCTT': 17, 'ACTAAA': 17, 'ACCGTA': 17, 'ATCCAA': 17, 'AACACT': 17, 'GATGTA': 17, 'CCAACC': 17, 'CTTACC': 17, 'TACGCA': 17, 'CTCCAG': 17, 'CAAAGT': 17, 'AGGTTG': 17, 'GCTCTT': 17, 'AGAGTG': 17, 'GTTATC': 17, 'AATAGA': 17, 'GGATCT': 17, 'ACATGC': 17, 'TTTTTA': 17, 'CCAAAC': 17, 'ATAGGG': 17, 'CGATGG': 17, 'CTTATT': 17, 'CGCTTG': 17, 'CGAAAC': 17, 'TGCCGA': 17, 'CCGACA': 17, 'CTCAGT': 17, 'GGTATG': 17, 'TCTGTC': 17, 'TTGGGA': 17, 'TTCATG': 17, 'TTGTAC': 17, 'CCTGGT': 17, 'TGTTCC': 17, 'ATTAGA': 17, 'GCTAGC': 17, 'CGTCTG': 17, 'CCGCCA': 17, 'ACAGTT': 17, 'ACAGGG': 17, 'CGAGCT': 17, 'CGTAAT': 17, 'ACAATT': 17, 'TCTTAT': 17, 'GTTATT': 17, 'TCCGTT': 17, 'TTACAA': 17, 'GGTCCG': 17, 'GCACTT': 17, 'AGTCAA': 17, 'ATGAGT': 17, 'CTCACT': 17, 'TGCTAC': 17, 'TCTCCG': 17, 'GGAGTA': 17, 'TCACTT': 17, 'CAGCTA': 17, 'GTTAAC': 17, 'TTTCAG': 17, 'CCTCGT': 17, 'TCCCAG': 17, 'AGACAT': 17, 'TTATCT': 17, 'TCTGCA': 17, 'CCTCTT': 17, 'CCAGAA': 17, 'CTATCC': 17, 'TATCGG': 17, 'AGCTCG': 17, 'AATCTA': 17, 'CCCTCA': 17, 'TATCAC': 17, 'TGTCGC': 17, 'GAGAGA': 17, 'TCGACA': 17, 'CCCTAT': 17, 'GACACG': 17, 'GACTCA': 17, 'GCCGTA': 17, 'ACTGGT': 17, 'AGTAGG': 17, 'TTTTGT': 17, 'ACTCGT': 17, 'CCACGT': 17, 'GGACCC': 17, 'AGGTGG': 17, 'TTCGTG': 17, 'AACAGA': 17, 'TGGTAT': 16, 'CGGAGA': 16, 'GGTCCT': 16, 'CAAACT': 16, 'CACTCG': 16, 'ATATTG': 16, 'GATCCG': 16, 'ATGTGC': 16, 'TTCCAA': 16, 'CGACTG': 16, 'TTTCTT': 16, 'GTTGGG': 16, 'CTTCAG': 16, 'TTTTAC': 16, 'TACCGA': 16, 'TCTGTT': 16, 'GTTCTG': 16, 'AGTATC': 16, 'GTGCCA': 16, 'CCCCAA': 16, 'ACGATT': 16, 'GCACCA': 16, 'CCCAGG': 16, 'TGTGAT': 16, 'AGGTAT': 16, 'ATCTCT': 16, 'ACATTT': 16, 'TCGTGC': 16, 'GGGGTA': 16, 'GCAACT': 16, 'GACTAT': 16, 'TATGGT': 16, 'AGTCTG': 16, 'CTAGGC': 16, 'CATATT': 16, 'GTTTCA': 16, 'TATTCA': 16, 'TCGATT': 16, 'AACGTA': 16, 'GTAGGT': 16, 'GGAACT': 16, 'TTATTG': 16, 'CACTAT': 16, 'TGGATG': 16, 'GTTCGA': 16, 'CGCACA': 16, 'ACCTCA': 16, 'AATGTG': 16, 'AGGGTT': 16, 'CCCTTT': 16, 'TCTCGA': 16, 'TTAGCT': 16, 'CTCCGT': 16, 'TACCAA': 16, 'AGACAG': 16, 'GTACTT': 16, 'CCTTGC': 16, 'TTTCTG': 16, 'CGACCT': 16, 'AGAGCA': 16, 'TGTGAC': 16, 'CCCTCT': 16, 'CTGGTA': 16, 'GGTTTA': 16, 'CCCGTT': 16, 'CTAAAT': 16, 'TATACC': 16, 'CAAAGG': 16, 'ACCCTT': 16, 'ACTGAG': 16, 'CTCTCT': 16, 'ACACGT': 16, 'TTACCT': 16, 'ATCGTA': 16, 'GCCACT': 16, 'TGGTCG': 16, 'TGGTTT': 16, 'TGGTTG': 16, 'TCCATT': 16, 'CGGTGC': 16, 'CCTGCA': 16, 'TCACAT': 16, 'ATAGGC': 16, 'GTCATT': 16, 'ACACAG': 16, 'ACGTCT': 16, 'ACCATT': 16, 'ACAGAG': 16, 'TCTCGT': 16, 'TCTGTG': 16, 'GTTGGA': 16, 'TCTCAA': 16, 'CGCCTT': 16, 'TTAGAC': 16, 'ATATTT': 16, 'CTTTCA': 16, 'GACTCT': 16, 'TCCCCT': 16, 'GTGAGA': 16, 'ATCCTT': 16, 'ATTTCA': 16, 'CAACGA': 16, 'TCACGG': 16, 'CGAGAG': 16, 'CCCTTG': 16, 'AGATCA': 16, 'CGTGGA': 16, 'AGAGAG': 16, 'TTTTCG': 16, 'TGCAAT': 16, 'CATCGG': 16, 'ACACGA': 16, 'TTGTGG': 15, 'ACACGG': 15, 'AGCTTG': 15, 'GACAGA': 15, 'TGGGCT': 15, 'TCTGCG': 15, 'CTTGGG': 15, 'TTATAT': 15, 'TATCAG': 15, 'TTCTGG': 15, 'CGGACT': 15, 'CCTCTC': 15, 'TACGTG': 15, 'ACGAGT': 15, 'CTTCCT': 15, 'ATCTGT': 15, 'CCTATC': 15, 'GTACGG': 15, 'ATAGAG': 15, 'TCTATC': 15, 'TGCCCG': 15, 'CTTGGA': 15, 'CACGTG': 15, 'CCACCC': 15, 'AAAACG': 15, 'TGTGCT': 15, 'CCCACT': 15, 'TGGCCA': 15, 'ACTCCT': 15, 'CATAAT': 15, 'ACTGCA': 15, 'GGTTCA': 15, 'ATTATG': 15, 'CGACAG': 15, 'GGCGTA': 15, 'ATGTCT': 15, 'AGCACT': 15, 'CTTCGG': 15, 'TCGAGT': 15, 'GTCCTT': 15, 'CAAAGA': 15, 'TCCACG': 15, 'TATGCG': 15, 'AGTCTT': 15, 'ACTAAC': 15, 'GGTTGC': 15, 'CACACA': 15, 'AACTCG': 15, 'TCCAGG': 15, 'CATGTG': 15, 'AGACTC': 15, 'CTGCTA': 15, 'ACGACA': 15, 'TACTGG': 15, 'ATTCCG': 15, 'TTTATA': 15, 'GTTGTA': 15, 'CTCCAT': 15, 'TTGCGA': 15, 'GGGCTA': 15, 'CCTTAC': 15, 'GACGTA': 15, 'AGGGCA': 15, 'AGACAC': 15, 'ATTTCG': 15, 'TTGTCA': 15, 'TTACAG': 15, 'CCGCCT': 15, 'GAGACA': 15, 'ATTACT': 15, 'GCGTGT': 15, 'CTAAGC': 15, 'ACTCCG': 15, 'CGATTT': 15, 'TGGTGC': 15, 'GTCCAT': 15, 'ATGACG': 15, 'CAATGC': 15, 'ACCACA': 15, 'GGGATA': 15, 'GGGACT': 14, 'ACTGTA': 14, 'TTGACT': 14, 'CCTTAT': 14, 'ATGGTA': 14, 'ACATCG': 14, 'GTCTTG': 14, 'GAAACG': 14, 'TCTAAG': 14, 'ACCAGA': 14, 'TCATTT': 14, 'ACCCAA': 14, 'CCGAGT': 14, 'TTACAT': 14, 'CCATAC': 14, 'GACTTG': 14, 'TGTCAC': 14, 'AGGACA': 14, 'ATATGG': 14, 'CTTAAA': 14, 'TCTTAC': 14, 'CGGTCT': 14, 'TTAAGA': 14, 'TCACGA': 14, 'AGTCAG': 14, 'ACTATC': 14, 'TGCCAT': 14, 'CAGGTA': 14, 'CCACAG': 14, 'CGCACT': 14, 'TATAAG': 14, 'TGCTGT': 14, 'CGATGC': 14, 'CATCGA': 14, 'AAGACG': 14, 'TCCCCA': 14, 'TTGATA': 14, 'AGTAGC': 14, 'GTGATA': 14, 'CAATTA': 14, 'TCTCGG': 14, 'GGGTTA': 14, 'CCTCAA': 14, 'ACTCGG': 14, 'ACGTAT': 14, 'ATTCGG': 14, 'TATTCT': 14, 'GTATAT': 14, 'GTCAGA': 14, 'CTACAC': 14, 'ATAGGT': 14, 'TATGCA': 14, 'CATCCA': 14, 'CTCACA': 14, 'TTGAGA': 14, 'GACTTT': 14, 'ACTGTT': 14, 'AGTAGT': 14, 'AGACCG': 14, 'ATAGCT': 14, 'CTTTGC': 14, 'TACAGG': 14, 'CACTCA': 14, 'ATACGG': 14, 'TTGGTA': 14, 'ACATCA': 14, 'TTCACG': 14, 'CGTCAA': 14, 'CTGACA': 14, 'ATCACT': 14, 'TTACTC': 14, 'CGCAGA': 14, 'AGTGGT': 14, 'TCACGT': 14, 'AACGTT': 14, 'TGTAAT': 14, 'CCTTCA': 14, 'ATGATA': 14, 'AGTGTG': 14, 'CCAAGG': 14, 'AGAACG': 14, 'CCACTT': 13, 'GACACT': 13, 'AGTACC': 13, 'ACTATT': 13, 'GGAACA': 13, 'CCCAAA': 13, 'ACGCTA': 13, 'CATTGG': 13, 'AGGATG': 13, 'ATCCCA': 13, 'CTCAGG': 13, 'TTTCGT': 13, 'ACGGTA': 13, 'TCAGGT': 13, 'CCTTCG': 13, 'CTTATA': 13, 'CAGACA': 13, 'CAGTTA': 13, 'GTTGTG': 13, 'GGGTCA': 13, 'GTCCAA': 13, 'TCCAGA': 13, 'GTAACC': 13, 'AGACTG': 13, 'ACTCAG': 13, 'CGAACT': 13, 'TGTTGC': 13, 'CCTGTA': 13, 'CTTTGT': 13, 'GACACA': 13, 'GTACCC': 13, 'CGACAT': 13, 'TCTCTC': 13, 'GCAAGA': 13, 'CTTCAA': 13, 'TCGCTA': 13, 'CGACCA': 13, 'ATGAGA': 13, 'CAGACG': 13, 'CCAAAT': 13, 'CCAGCT': 13, 'CCAGGA': 13, 'GCTGTA': 13, 'CGAGAT': 13, 'ATAATA': 13, 'TGGGTT': 13, 'TTATCG': 13, 'TACGGT': 13, 'CAAGTA': 13, 'TCCGTG': 13, 'CTCCAA': 13, 'CTCCCT': 13, 'AGAACT': 13, 'AGATGG': 13, 'GTAGAC': 13, 'TACTTT': 13, 'CGAAAG': 13, 'CCACTC': 13, 'ATAGCG': 13, 'ACATCT': 13, 'ATATGC': 13, 'ACCTCT': 13, 'CTTGCA': 13, 'CCTTGG': 13, 'TTCACT': 13, 'GCATAC': 13, 'TCGTTG': 13, 'CGTTGC': 13, 'ACCACG': 13, 'CCAGTT': 13, 'AGAGTT': 13, 'ACTTTA': 13, 'GTAGTC': 13, 'GCAACA': 13, 'TCAAAG': 13, 'TATATG': 13, 'CTAAGT': 13, 'TCATAT': 13, 'TATGGA': 13, 'ACTACC': 13, 'TGCATC': 13, 'CACCCA': 13, 'CCCCCG': 13, 'CATTGC': 13, 'AGTTGC': 13, 'TGTTAT': 13, 'TCACAG': 13, 'GATCTA': 13, 'AGCTTA': 13, 'GGGACA': 13, 'GCAGTA': 13, 'ATCTCA': 13, 'CTAGGG': 13, 'GCTAGT': 13, 'TCTTTA': 13, 'GGCATA': 13, 'TCAAGT': 13, 'GTTAGT': 13, 'CTAGAC': 13, 'AGCATT': 13, 'TATGTG': 13, 'GTAAAT': 12, 'TCATTA': 12, 'ATCTTG': 12, 'TATCGA': 12, 'TTTTGC': 12, 'TGCGAA': 12, 'TCACTG': 12, 'ACCACT': 12, 'CACCTT': 12, 'TTGTGC': 12, 'TGTCCA': 12, 'AATAGG': 12, 'CATAGT': 12, 'CGTTTG': 12, 'CCATAT': 12, 'CATCAA': 12, 'TGGAGT': 12, 'TACGCT': 12, 'TGCCCT': 12, 'CTTTTA': 12, 'GTAGGG': 12, 'TTAATA': 12, 'AGGCAA': 12, 'ACCTTG': 12, 'ACTTTG': 12, 'CTATCT': 12, 'CTAAAC': 12, 'TATAGC': 12, 'TCGTTT': 12, 'GTACCG': 12, 'TCCACT': 12, 'CCGACT': 12, 'AACTCA': 12, 'TTAACT': 12, 'CATTTG': 12, 'TACAAT': 12, 'CGAACA': 12, 'ACCATG': 12, 'TCAGGG': 12, 'CTAAAG': 12, 'TACTAT': 12, 'AGAGTC': 12, 'CGAATT': 12, 'TGCTTT': 12, 'AGGTCA': 12, 'ATACCT': 12, 'TATTGG': 12, 'TGGATA': 12, 'TGGCTA': 12, 'AGAATG': 12, 'AGCTCT': 12, 'GGTTGT': 12, 'TCCTTG': 12, 'CACAGT': 12, 'TGTGAG': 12, 'ATAGGA': 12, 'TCTCCT': 12, 'ACACAT': 12, 'CAATTG': 12, 'CTTTGG': 12, 'GTTTGG': 12, 'AGATTA': 12, 'TCTCAG': 12, 'ACACAA': 12, 'TTGTTA': 12, 'TTTCGA': 12, 'GGCCCT': 12, 'CTTCGA': 12, 'CCCAAT': 12, 'AATCGG': 12, 'TTGCTA': 12, 'CATCCT': 12, 'GCTACT': 12, 'ACATTG': 12, 'ATATTA': 12, 'CTTAAG': 12, 'AAGTGT': 12, 'GGTACT': 12, 'TCGACT': 12, 'AGAAGT': 12, 'CTTAGT': 12, 'ACTCTG': 12, 'CCTAAA': 12, 'TCCATG': 12, 'CGCGTA': 12, 'CTCATT': 12, 'CAAGTG': 12, 'ACTCAA': 12, 'TTCTCA': 12, 'CATATG': 12, 'CTCTCA': 12, 'ACAAGA': 12, 'CGACTT': 12, 'GAGACT': 12, 'TTCCGA': 12, 'CCTAAT': 11, 'AGTTCT': 11, 'TATCCT': 11, 'GATCCA': 11, 'TCGAGA': 11, 'CCTGGG': 11, 'CAGTTG': 11, 'CACGTT': 11, 'AGTGGG': 11, 'ATTATA': 11, 'TGGTCT': 11, 'GTACTC': 11, 'AGATCT': 11, 'GTACTG': 11, 'TTGACG': 11, 'GGGCCA': 11, 'ATCAGA': 11, 'CGGTGT': 11, 'ATAGTC': 11, 'ATACTC': 11, 'TACCTT': 11, 'GGGTGT': 11, 'TATGTA': 11, 'CCTATT': 11, 'GGTCTA': 11, 'CTATGC': 11, 'GTCTCT': 11, 'TTAGGG': 11, 'TTATCA': 11, 'CGAGTG': 11, 'CGTAAG': 11, 'ACGACT': 11, 'ACTAGC': 11, 'CCAGGG': 11, 'ACTTCG': 11, 'TGGTCA': 11, 'CCGTAT': 11, 'GGTACG': 11, 'TGCCTC': 11, 'TGTCCT': 11, 'TATTCG': 11, 'CCTAAC': 11, 'GACCCT': 11, 'AGTCCG': 11, 'TTTTGG': 11, 'GATAGG': 11, 'CGATCT': 11, 'GGTATA': 11, 'ATACTT': 11, 'GTGACA': 11, 'AGTGCA': 11, 'TCTCTA': 11, 'CCACAA': 11, 'ACGTTG': 11, 'AGGCTA': 11, 'ATACCC': 11, 'ACCTGT': 11, 'GCACTA': 11, 'CTGAGA': 11, 'GTACGA': 11, 'TTTGTA': 11, 'GCGTTA': 11, 'AGTTCA': 11, 'ACTAAG': 11, 'TGTGTC': 11, 'TCAATG': 11, 'GTATCG': 11, 'CCCACG': 11, 'TTAATG': 11, 'ATGACA': 11, 'ATAAGG': 11, 'ATGTCA': 11, 'TGTCAT': 11, 'GGACCT': 11, 'TGTTTT': 11, 'ATTCTA': 11, 'ATCTTT': 11, 'GTTACC': 11, 'AGGTCT': 11, 'CTAGTC': 11, 'CAGATA': 11, 'CACATT': 11, 'ATAACC': 11, 'GTATTG': 11, 'ATAATT': 11, 'AGCTCA': 11, 'GTCACA': 10, 'TCTCTG': 10, 'CACCAA': 10, 'GTCTCA': 10, 'CTGTTA': 10, 'GTAAGT': 10, 'ATAATG': 10, 'GTAAGG': 10, 'CTACTG': 10, 'CCTTGT': 10, 'CTTACT': 10, 'ATCATG': 10, 'TACGGA': 10, 'TGTAGC': 10, 'ATAGTG': 10, 'TTCGTA': 10, 'TACGTT': 10, 'ACAAGG': 10, 'TGCCGT': 10, 'TCACCT': 10, 'GTAGGA': 10, 'AGTCCA': 10, 'ACTGTG': 10, 'AACACA': 10, 'AGGTTA': 10, 'CCAGGT': 10, 'ACAACT': 10, 'CTAGCT': 10, 'CGAGAA': 10, 'ACACTG': 10, 'ACATTA': 10, 'TTCTTA': 10, 'ATAGTA': 10, 'TCTATA': 10, 'CCGAGA': 10, 'ATCATA': 10, 'GTGGTA': 10, 'CTAGAT': 10, 'CCTTTG': 10, 'ATTTGG': 10, 'TTTCCT': 10, 'CTACAA': 10, 'CTAGGT': 10, 'TACCCG': 10, 'ATCACG': 10, 'GTACGT': 10, 'CACTCT': 10, 'GTTATA': 10, 'TTACAC': 10, 'ATGTTA': 10, 'CGACAA': 10, 'TGGTTA': 10, 'AGACGG': 10, 'GATACG': 10, 'TTACTG': 10, 'ATACAG': 10, 'TTATGG': 10, 'ATACGT': 10, 'TTGTCG': 10, 'TGGAGA': 10, 'ACGAGA': 10, 'ATGCTA': 10, 'CCATCG': 10, 'CCCCTT': 10, 'TTAGTC': 10, 'AATTGT': 10, 'AGCCTA': 10, 'CTACCG': 10, 'CCCAGA': 10, 'CACACG': 10, 'TGTTCG': 10, 'TCATGT': 10, 'CATACC': 10, 'GGACTA': 10, 'GTATGC': 10, 'TCTATT': 10, 'CATCTT': 10, 'TCTGGA': 10, 'ACAGTG': 10, 'TTCGTT': 10, 'CCACCT': 10, 'TGCGTT': 10, 'TTAGCG': 10, 'GTAAGA': 10, 'GCTTGT': 10, 'CAGTGT': 10, 'ATTCCA': 10, 'ACACTA': 10, 'TACAGT': 10, 'CTTGTG': 10, 'CGAGTA': 10, 'TCCTAT': 10, 'CGTAGT': 10, 'AGTATG': 10, 'GTAATT': 10, 'AACTGT': 10, 'TGTCTC': 10, 'ATTTGC': 10, 'GTACAT': 10, 'AGACCT': 9, 'CAAATA': 9, 'CCAATG': 9, 'TGTGAA': 9, 'GTATTA': 9, 'CCATCT': 9, 'CGTTTA': 9, 'ACGTTA': 9, 'GTATGG': 9, 'GTCTGT': 9, 'CCCCCT': 9, 'CAACTA': 9, 'CTCATG': 9, 'TGCCCA': 9, 'AGTTTA': 9, 'ATAATC': 9, 'CTCACG': 9, 'TGTGGC': 9, 'CTATTC': 9, 'TTTAGG': 9, 'TCTAGC': 9, 'CGCCTA': 9, 'TGTATT': 9, 'GACTTA': 9, 'CCTAAG': 9, 'TGCCAA': 9, 'GCAATG': 9, 'AGACTT': 9, 'ATTAGG': 9, 'CGAGTT': 9, 'CATAAG': 9, 'AGAACA': 9, 'ACATAC': 9, 'TTCCCA': 9, 'TGTCCG': 9, 'TTCCTT': 9, 'TACCAT': 9, 'GGACCA': 9, 'TACAAA': 9, 'TCCCTA': 9, 'CTAATC': 9, 'AACTTA': 9, 'CCCATG': 9, 'AACCTA': 9, 'TGCTGG': 9, 'TTCTTG': 9, 'TCCGTA': 9, 'ACTTGC': 9, 'GTCGTA': 9, 'CGTACT': 9, 'TTTAGA': 9, 'CTACGG': 9, 'CATCTA': 9, 'ATATCA': 9, 'AACTTG': 9, 'GTAGTT': 9, 'AGCACG': 9, 'TCTGGG': 9, 'TTAGGA': 9, 'TCCCTT': 9, 'AGTAGA': 9, 'TGCTCG': 9, 'CTATAC': 9, 'GCAATA': 9, 'AGCATA': 9, 'TCTTGG': 9, 'CTAGAG': 9, 'TTCTCT': 9, 'ATAGTT': 9, 'AATACG': 9, 'GTATGT': 9, 'GGTAGG': 9, 'TCTAGT': 9, 'TTAAGG': 9, 'TACAGA': 9, 'TGCGCT': 9, 'ACAAGT': 9, 'TCAACT': 9, 'TCGGTA': 9, 'CTAGAA': 9, 'AACATA': 9, 'ACAGGA': 9, 'TTACGG': 8, 'ATCTCG': 8, 'TGCCTG': 8, 'ACAATA': 8, 'TTGCAA': 8, 'GTGTTA': 8, 'TGTAGT': 8, 'TGTGTT': 8, 'ACAATG': 8, 'GCGATA': 8, 'TTAGTG': 8, 'CACATA': 8, 'GCTTTA': 8, 'ATATAC': 8, 'TACATG': 8, 'CCTTCT': 8, 'TGTATC': 8, 'AGATTG': 8, 'TTACGT': 8, 'CAATGT': 8, 'GCTACG': 8, 'TCTCCA': 8, 'TACTCT': 8, 'ACTGGA': 8, 'CCCGTA': 8, 'GGTAGA': 8, 'CGTACG': 8, 'TGTTTC': 8, 'CAAACA': 8, 'CATTGT': 8, 'TTCATA': 8, 'CATGTA': 8, 'TATCCA': 8, 'CCTCCG': 8, 'TGTCTG': 8, 'CGCTGT': 8, 'CGGCTA': 8, 'ATATCG': 8, 'TTGACA': 8, 'TCTACC': 8, 'ACGATA': 8, 'TGCAGT': 8, 'AGGGTA': 8, 'CTATCA': 8, 'GGCCTA': 8, 'GATACT': 8, 'CACTTG': 8, 'TCAACG': 8, 'AGTATA': 8, 'TGCAGG': 8, 'GTTCTA': 8, 'CTTGTA': 8, 'CTATTG': 8, 'TGCGGA': 8, 'ATACGA': 8, 'AGGACT': 8, 'CGAATA': 8, 'CGAAGT': 8, 'CCAAGT': 8, 'CAAACG': 8, 'TACATT': 8, 'TTAACG': 8, 'TCGATA': 8, 'CTAGCG': 8, 'ATATCT': 8, 'TCTTGC': 8, 'GACTGT': 8, 'CGTGTG': 8, 'CTTCTG': 8, 'TGTCGA': 8, 'TCTAAC': 8, 'GCATTA': 8, 'ATCCTA': 8, 'GTTAGA': 8, 'CTTCCA': 8, 'ACTGGG': 8, 'ATTACG': 8, 'AGGTGC': 7, 'GTACAG': 7, 'TGTCAA': 7, 'CGAAGA': 7, 'GATCCT': 7, 'TGGGTA': 7, 'TGTTTA': 7, 'TTACTA': 7, 'TACACA': 7, 'TTTCCA': 7, 'AATACA': 7, 'GTCCCT': 7, 'CACAGA': 7, 'GTCCCA': 7, 'ATAAGA': 7, 'ATGTGT': 7, 'TCAATA': 7, 'CCAACA': 7, 'TGCTCA': 7, 'TGTAAG': 7, 'CTTACA': 7, 'TATAGA': 7, 'GTCACG': 7, 'TCATCT': 7, 'TGTGGT': 7, 'ACATAT': 7, 'ATAGCA': 7, 'ACAGTA': 7, 'GTACAA': 7, 'AGATCG': 7, 'TGTAAC': 7, 'CAAATG': 7, 'AGCACA': 7, 'GTACCT': 7, 'TGCACA': 7, 'AGTTGT': 7, 'ACTAGA': 7, 'CCCATT': 7, 'CGATTG': 7, 'CGGTTA': 7, 'GTTCCA': 7, 'CCAGTG': 7, 'CCTCCT': 7, 'CACTTT': 7, 'ACCATA': 7, 'TCAAGG': 7, 'CTATGG': 7, 'TCGTGT': 7, 'TTAGTA': 7, 'AGCTGT': 7, 'CCGTGT': 7, 'CCTCAT': 7, 'AGCCCA': 7, 'ATTGTA': 7, 'GTAATG': 7, 'CGAATG': 7, 'CTACGA': 7, 'ACAACG': 7, 'CTACTC': 7, 'GTACAC': 7, 'AGAGTA': 7, 'CATACT': 7, 'CACACT': 7, 'TATAGT': 7, 'CCACCG': 7, 'TGTCGT': 7, 'TCACTA': 7, 'TCGTTA': 7, 'CCTACT': 7, 'TCAGTG': 7, 'CTTATG': 7, 'CCTACC': 7, 'TCCTGT': 7, 'TTTCCG': 7, 'ATGACT': 7, 'CTCCCA': 7, 'CGTACA': 7, 'GACCTA': 7, 'GGTGTA': 7, 'TCTACT': 7, 'TGTTGT': 7, 'CGTATG': 7, 'CCATTG': 7, 'CGATAT': 7, 'ACAACA': 7, 'GTTACG': 7, 'TATACT': 6, 'GCCATA': 6, 'CTAACC': 6, 'TTTACT': 6, 'CTCGTA': 6, 'TCATAC': 6, 'TCCACA': 6, 'CCAGTA': 6, 'AGTGTA': 6, 'CACAGG': 6, 'CTAACT': 6, 'TTAACA': 6, 'TGTTGG': 6, 'TCATTG': 6, 'AACACG': 6, 'CTAATT': 6, 'CGCATA': 6, 'AGTCCT': 6, 'CCCACA': 6, 'CTAGTT': 6, 'GTGCTA': 6, 'TTTCTA': 6, 'TATATA': 6, 'CTTCGT': 6, 'CCTAGA': 6, 'GCTATG': 6, 'ATACTA': 6, 'TTAGCA': 6, 'TTATGT': 6, 'TCGTAT': 6, 'ACCCTA': 6, 'CCATTA': 6, 'CCCTGT': 6, 'TGGCCT': 6, 'GTAATA': 6, 'GGATGT': 6, 'CTACCA': 6, 'TTACCG': 6, 'ACTTGG': 6, 'TTTACA': 6, 'TGCTTA': 6, 'ACGTGT': 6, 'CTATAT': 6, 'CTCTTG': 6, 'TTATGC': 6, 'TATCTA': 6, 'CTATCG': 6, 'ACTATG': 6, 'CTAAGG': 6, 'CTATGT': 6, 'ACTACA': 6, 'CCTTTA': 6, 'CGTCTA': 6, 'GTATCA': 6, 'TGCCTT': 6, 'TGTCTA': 6, 'TGTGGA': 6, 'CTCAGA': 6, 'TGTAGA': 6, 'TCAACA': 6, 'TGTGCG': 6, 'TATAGG': 6, 'CATTTA': 6, 'CATAGG': 6, 'CTAATA': 6, 'AGGTGT': 6, 'ACCTTA': 6, 'GTGACT': 6, 'AGGATA': 6, 'CTACCT': 6, 'CCACAT': 6, 'AGTACT': 5, 'GGCTTA': 5, 'CCAATT': 5, 'CCTAGT': 5, 'CGATTA': 5, 'ACTCTA': 5, 'TACCCT': 5, 'ATTACA': 5, 'GTAGCA': 5, 'CGTTGT': 5, 'TGTAAA': 5, 'ACTTGT': 5, 'ATCACA': 5, 'CGCTTA': 5, 'TCTACG': 5, 'GGTCCA': 5, 'TGCTAT': 5, 'GTACCA': 5, 'TTACCA': 5, 'TGTTTG': 5, 'ATAACT': 5, 'TCAGTA': 5, 'CTAAGA': 5, 'TGTATG': 5, 'CCATGT': 5, 'CCTAGC': 5, 'CCACCA': 5, 'CCAACG': 5, 'CCTCTG': 5, 'AGTCTA': 5, 'ATATGT': 5, 'TACTGT': 5, 'TACCCA': 5, 'TGTTAC': 5, 'CTTAGA': 5, 'TATACA': 5, 'AGACCA': 5, 'CCGATA': 5, 'AGTACA': 5, 'TTGTGT': 5, 'CGTGTA': 5, 'TCGTCT': 5, 'CATAGA': 5, 'CCAAGA': 5, 'ATACAC': 5, 'TGCACG': 5, 'TTACGA': 5, 'GTAACT': 5, 'CTCCTT': 5, 'TCTGTA': 5, 'GCCTTA': 5, 'GTTACT': 5, 'ACTATA': 5, 'CGTATA': 5, 'CTAATG': 5, 'TGTTCA': 5, 'TGGACT': 5, 'CGATGT': 5, 'GTTAGG': 5, 'TGCGCA': 5, 'CATATA': 5, 'TCTTGT': 5, 'TTCAGA': 5, 'TGCGTG': 5, 'CTCCTA': 5, 'CCACTA': 5, 'AGATGC': 5, 'GCTCTA': 4, 'ACTACT': 4, 'TTCCTA': 4, 'TCAAGA': 4, 'TACCTA': 4, 'ATACTG': 4, 'TGCTTG': 4, 'GCTACA': 4, 'ACATGT': 4, 'TGTCTT': 4, 'CACCTA': 4, 'TACTTG': 4, 'GACCCA': 4, 'CTACTT': 4, 'TTCACA': 4, 'TGCACT': 4, 'GTCCTA': 4, 'TACACT': 4, 'TGCATT': 4, 'TATGGG': 4, 'CGTAGA': 4, 'TACACG': 4, 'ACTAGT': 4, 'GTCACT': 4, 'TGTTCT': 4, 'CACTTA': 4, 'GTGTGT': 4, 'CTAGGA': 4, 'CCCCCA': 4, 'GTAGTA': 4, 'CCAATA': 4, 'TCTACA': 4, 'TATTGT': 4, 'CCCCTA': 4, 'GTATAC': 4, 'GCTAGA': 4, 'GCATGT': 4, 'AGACAA': 4, 'CCTATA': 4, 'ATACCG': 4, 'GTTTGT': 4, 'CTGTGT': 4, 'GCTAGG': 4, 'TCTATG': 4, 'TCCTTA': 4, 'AGTACG': 4, 'ATAACA': 4, 'CACTGT': 4, 'ATCTTA': 4, 'CTACAT': 4, 'TCCATA': 4, 'ATACCA': 4, 'TGCTCT': 4, 'TGTACT': 4, 'TTTACG': 4, 'AGCGTA': 4, 'GTAGTG': 4, 'CTACGT': 4, 'GTACTA': 4, 'CTTCTA': 4, 'CGTAGG': 3, 'CCAACT': 3, 'ACTACG': 3, 'AGATGT': 3, 'CTACAG': 3, 'TGCATA': 3, 'TACTCG': 3, 'CTAGTG': 3, 'TGTGTA': 3, 'TCTAGA': 3, 'ATAACG': 3, 'CTATTA': 3, 'GCTATA': 3, 'CCTCCA': 3, 'CTTAGG': 3, 'GTAACA': 3, 'CCGTTA': 3, 'TGTACA': 3, 'AGACGT': 3, 'TGTCGG': 3, 'CCCTTA': 3, 'GATACA': 3, 'CCTATG': 3, 'TGTGGG': 3, 'TGTGTG': 3, 'CCTCTA': 3, 'GTTATG': 3, 'CGACTA': 3, 'TTCCCT': 3, 'TTCTGT': 2, 'CATACA': 2, 'TGGTGT': 2, 'GTCTTA': 2, 'CCTACG': 2, 'GTCATA': 2, 'TGTAGG': 2, 'AGACTA': 2, 'GTTACA': 2, 'TCTGGT': 2, 'CTTACG': 2, 'GGTACA': 2, 'TGCATG': 2, 'TACTCA': 2, 'CTCTGT': 2, 'TGTATA': 2, 'CTAACG': 2, 'TGTACC': 2, 'CTAGCA': 2, 'CACGTA': 2, 'TACGTA': 2, 'TGCGTA': 2, 'GACATA': 2, 'TGTACG': 2, 'CCTACA': 2, 'TGCAGA': 2, 'CTCTTA': 2, 'TTATAC': 2, 'ATACAT': 2, 'CATACG': 1, 'TATACG': 1, 'GAGGMT': 1, 'CTACTA': 1, 'TCTAGG': 1, 'GTAACG': 1, 'CCTAGG': 1, 'CCCATA': 1, 'TACATA': 1, 'CTCATA': 1, 'TGTGCA': 1, 'TGCCTA': 1, 'TACTTA': 1, 'ACTAGG': 1})\n",
      "\n",
      "#### Compute statistics for stat  stopCodon_TGA , end\n",
      "\n",
      "#### Analysis for stat: stopCodon_TGA , start\n",
      "\n",
      "outputFilePrefix: TGA_non_overlapping_start_codon_subsample_subset_stopCodon_TGA\n",
      "\n",
      "#### Termini amino acid composition analysis\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#### Writing oddsRatioAADf files, start\n",
      "\n",
      "#### Writing oddsRatioAADf files, end\n",
      "\n",
      "#### Termini amino acid composition analysis end\n",
      "\n",
      "#### Termini codons composition analysis, start\n",
      "\n",
      "#### Writing oddsRatioDfCodon files, start\n",
      "\n",
      "#### Writing oddsRatioDfCodon files, end\n",
      "\n",
      "#### Termini codons composition analysis, end\n",
      "\n",
      "#### AA pairs bias analysis start\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/IPython/core/async_helpers.py:68: PerformanceWarning: indexing past lexsort depth may impact performance.\n",
      "  coro.send(None)\n",
      "/users/lserrano/mweber/Research_cloud/C-terminal/src/cterminal.py:1640: RuntimeWarning: divide by zero encountered in log2\n",
      "  log2OddsRatio = log2(oddsRatio)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#### Writing oddsRatioDf AA pair files, start\n",
      "\n",
      "#### Writing oddsRatioDf AA pair files, end\n",
      "\n",
      "#### AA pairs bias analysis end\n",
      "\n",
      "#### hexamer bias analysis start\n",
      "\n",
      "#### Writing oddsRatioDf hexamer files, start\n",
      "\n",
      "#### Writing oddsRatioDf hexamer files, end\n",
      "\n",
      "#### hexamer bias analysis end\n",
      "\n",
      "#### Analysis for stat: stopCodon_TGA , end\n",
      "\n",
      "#### Drawing plots, start\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Research_cloud/Python_mwTools/mwTools/bio.py:847: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  plotDf['sort_index'] = plotDf.index.get_level_values(0).map(lambda x: sort_order.get(x))\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/users/lserrano/mweber/Research_cloud/Python_mwTools/mwTools/bio.py:847: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  plotDf['sort_index'] = plotDf.index.get_level_values(0).map(lambda x: sort_order.get(x))\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus N\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus N\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus C\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminus C\n",
      "oddsRatioDfTableAApair2.head()\n",
      ": aa1                                                    R         K         H  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests     True      True      True   \n",
      "                              NexpectedSubseq        NaN       NaN       NaN   \n",
      "                              count                 2397       486       574   \n",
      "                              log2OddsRatio      1.36282  0.461752  0.938519   \n",
      "                              oddsRatio          2.57187   1.37721   1.91656   \n",
      "\n",
      "aa1                                                     D         E        S  \\\n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True      True     True   \n",
      "                              NexpectedSubseq         NaN       NaN      NaN   \n",
      "                              count                  1013       404     1537   \n",
      "                              log2OddsRatio      0.595672 -0.905616  1.33032   \n",
      "                              oddsRatio           1.51118  0.533805  2.51459   \n",
      "\n",
      "aa1                                                     T         N         Q  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   706       465       283   \n",
      "                              log2OddsRatio      0.198259  0.590295 -0.699794   \n",
      "                              oddsRatio           1.14731   1.50555   0.61566   \n",
      "\n",
      "aa1                                                     C        G        P  \\\n",
      "terminus aa0 statistical_test observable                                      \n",
      "C        R   contingency      BH_multiple_tests      True     True     True   \n",
      "                              NexpectedSubseq         NaN      NaN      NaN   \n",
      "                              count                   134     2138     1412   \n",
      "                              log2OddsRatio      0.399049  1.36625   1.2553   \n",
      "                              oddsRatio           1.31864  2.57799  2.38716   \n",
      "\n",
      "aa1                                                     A          V  \\\n",
      "terminus aa0 statistical_test observable                               \n",
      "C        R   contingency      BH_multiple_tests      True      False   \n",
      "                              NexpectedSubseq         NaN        NaN   \n",
      "                              count                  1935        852   \n",
      "                              log2OddsRatio      0.577979 -0.0442563   \n",
      "                              oddsRatio           1.49276    0.96979   \n",
      "\n",
      "aa1                                                     I         L         M  \\\n",
      "terminus aa0 statistical_test observable                                        \n",
      "C        R   contingency      BH_multiple_tests      True      True      True   \n",
      "                              NexpectedSubseq         NaN       NaN       NaN   \n",
      "                              count                   533       863       130   \n",
      "                              log2OddsRatio     -0.235584 -0.666502 -0.878523   \n",
      "                              oddsRatio          0.849341  0.630032  0.543924   \n",
      "\n",
      "aa1                                                     F         W         Y  \n",
      "terminus aa0 statistical_test observable                                       \n",
      "C        R   contingency      BH_multiple_tests      True     False      True  \n",
      "                              NexpectedSubseq         NaN       NaN       NaN  \n",
      "                              count                   584       216       426  \n",
      "                              log2OddsRatio      0.336412  0.150688  0.329888  \n",
      "                              oddsRatio           1.26261    1.1101   1.25692  \n",
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#### Drawing plots, end\n",
      "\n",
      "#### Statistics stopCodon_TGA total nb sequences:  175061\n",
      "\n",
      "\n",
      "######## END ANALYSIS ########\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th>proteinGI</th>\n",
       "      <th>species_name</th>\n",
       "      <th>DNASeq</th>\n",
       "      <th>stopCodon</th>\n",
       "      <th>DNA3UTRSeq</th>\n",
       "      <th>codonTableBio</th>\n",
       "      <th>proteinSeq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000284615.1</th>\n",
       "      <td>WP_011783808.1</td>\n",
       "      <td>500107803</td>\n",
       "      <td>Marinobacter hydrocarbonoclasticus ATCC 49840</td>\n",
       "      <td>ATGAGTCTGCGTCTTGGAGATACTGCACCCGATTTTGAACAGGATT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCCACCGCCACAAACAAAAAGGGCAGCCAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSLRLGDTAPDFEQDSTEGRIRFHEWLGNGWGVLFSHPADFTPVCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000008325.1</th>\n",
       "      <td>WP_010960091.1</td>\n",
       "      <td>499262551</td>\n",
       "      <td>Methylococcus capsulatus str. Bath</td>\n",
       "      <td>GTGCCAATAGACTTGTCGAATTCTGAATGCCCCTTGGATCGCGATC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCGCTACACATTGATCGAGCGGGAGCTCGA</td>\n",
       "      <td>11</td>\n",
       "      <td>MPIDLSNSECPLDRDLPSSLMRATVDQAPVAISITDLKANILYVNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000143145.1</th>\n",
       "      <td>WP_013217028.1</td>\n",
       "      <td>502982052</td>\n",
       "      <td>Hyphomicrobium denitrificans ATCC 51888</td>\n",
       "      <td>ATGGCTCTCAATTTCGACGGTTTAGCGTCTAATGCTGCGGTCGCCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>ATCTGGAAGAACAATTTCCAAGCGCGTGCG</td>\n",
       "      <td>11</td>\n",
       "      <td>MALNFDGLASNAAVARSSGRRRPANVNFAETSWGRRGFASALDQFG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000063505.1</th>\n",
       "      <td>WP_011932283.1</td>\n",
       "      <td>500457779</td>\n",
       "      <td>Synechococcus sp. WH 7803</td>\n",
       "      <td>ATGGGCAGCAGTTTCGGCACCCTGTTTCGCATCAGCACCTTCGGCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>AGGATCTCAGGAGCGCTTAGCTCCGGGCCG</td>\n",
       "      <td>11</td>\n",
       "      <td>MGSSFGTLFRISTFGESHGGGVGVILEGCPPRLELDPLAIQAELDR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000063505.1</th>\n",
       "      <td>WP_011932424.1</td>\n",
       "      <td>500457920</td>\n",
       "      <td>Synechococcus sp. WH 7803</td>\n",
       "      <td>TTGCTGACTCCCCTCAAGGACTGCCTCAAGGATCTGCTCACCACCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTTAAGATCTCCATGTCGCGCAGACAACGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MLTPLKDCLKDLLTTLECPLCRQELERSPRRGALCMACWQRMDLAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000145035.1</th>\n",
       "      <td>WP_013280594.1</td>\n",
       "      <td>503045618</td>\n",
       "      <td>Butyrivibrio proteoclasticus B316</td>\n",
       "      <td>ATGGGTAATAAGACAAAAGTTGCCTGTGTGCAATTTGCCTGTGGCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGGAGAGAGTACTGCGAAGTGGGGAATGCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MGNKTKVACVQFACGAIASSDADQVKRNIETADRLTREAASGGAKI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000219605.1</th>\n",
       "      <td>WP_013982211.1</td>\n",
       "      <td>503748135</td>\n",
       "      <td>Pseudomonas stutzeri</td>\n",
       "      <td>ATGGCTTTGATCGTACAGAAGTTTGGGGGCACCTCGGTCGGCACCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTCGCAGAAGGCGCGCTAGTCCGCGCCTTC</td>\n",
       "      <td>11</td>\n",
       "      <td>MALIVQKFGGTSVGTVERIEQVAEKVKKFRDGGDDIVVVVSAMSGE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000317935.1</th>\n",
       "      <td>YP_006894825.1</td>\n",
       "      <td>410471544</td>\n",
       "      <td>Bordetella parapertussis Bpp5</td>\n",
       "      <td>GTGAGCCAGCAGGAGAGCCGCATGTTCGCACCCGCCTGCGTGTCCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>AGGCGCGATGGCGGCCCGGCCGGCCGAGCC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSQQESRMFAPACVSEFVRYDVDAGVARITLDRPDRRNAIDVPMRA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000145235.1</th>\n",
       "      <td>WP_013287839.1</td>\n",
       "      <td>503052863</td>\n",
       "      <td>Micromonospora aurantiaca ATCC 27029</td>\n",
       "      <td>ATGAGTCGTCCCGACGTCGTCGTCACCGGGCTCGGCGCGACGACCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGTTCCGCCCTGAGGTAAGGAAGGGCCCCT</td>\n",
       "      <td>11</td>\n",
       "      <td>MSRPDVVVTGLGATTPLGGDVASTWDAMLAGRSGVGALTQEWAAQL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001648175.1</th>\n",
       "      <td>WP_064122910.1</td>\n",
       "      <td>1031618553</td>\n",
       "      <td>Halotalea alkalilenta</td>\n",
       "      <td>ATGGCAAGACGCACCAAGGCAGAGGCCGAAGCCACTCGATCGATAC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CTCCCTGCGCATCAGCCCGGATGATGAACG</td>\n",
       "      <td>11</td>\n",
       "      <td>MARRTKAEAEATRSILLDAAEEIFFEKGVSTTSLEQIARRAGMTRG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000294515.1</th>\n",
       "      <td>WP_014909031.1</td>\n",
       "      <td>504721929</td>\n",
       "      <td>Nocardiopsis alba ATCC BAA-2165</td>\n",
       "      <td>ATGGGCGAGACGACCGAGACCGACACCGGCGGCGGAACCACCGAGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCTTCCCGGAGAGCCCGGTTCAGTCGGTGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MGETTETDTGGGTTEESDGGHVWSAGITAAGAALMFILLRLMAVTH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000219605.1</th>\n",
       "      <td>WP_013981774.1</td>\n",
       "      <td>503747698</td>\n",
       "      <td>Pseudomonas stutzeri</td>\n",
       "      <td>ATGAAACTCGCAACCCTCAACCAGGGCCGCGACGGCGTGCTCGTCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCTGACGCTCTACGGCTACTGGCGCTCCAG</td>\n",
       "      <td>11</td>\n",
       "      <td>MKLATLNQGRDGVLVVVSRDLAQAVKVPQIAATLQAALDDWNYCKP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000024385.1</th>\n",
       "      <td>WP_012855077.1</td>\n",
       "      <td>502618199</td>\n",
       "      <td>Thermomonospora curvata DSM 43183</td>\n",
       "      <td>ATGAACGTGATCGCGCTGACCCGCATCGAGCCGGGCATGGTCGAGC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CGGCCCGTCCGGGCATCGGCGCCGAGCCGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MNVIALTRIEPGMVELVGGKAAGLGGLIGAGERVPDGFCVTTAAHD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000196175.1</th>\n",
       "      <td>WP_011165490.1</td>\n",
       "      <td>499478850</td>\n",
       "      <td>Bdellovibrio bacteriovorus HD100</td>\n",
       "      <td>ATGAGTACTCACCGCTGCGACATTTTGATTATCGGCTCTGGCACTG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CTTGCCCGTCCGGGGCGAGTCACGGCAAGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSTHRCDILIIGSGTAGLALALKLSENGKVIVLAKESAGGTNSAMA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001189295.1</th>\n",
       "      <td>WP_050435835.1</td>\n",
       "      <td>913456426</td>\n",
       "      <td>Chondromyces crocatus</td>\n",
       "      <td>TTGAGCACGCTCATCACGCCGTGGATCCGCGGGCTCGCGCTCGCTT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGGCAGGTCGCCATGGCCGAGTCGCCCGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSTLITPWIRGLALASAVLVTGCASGLGDVDGPAPAGEAAAMPTFH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000019525.1</th>\n",
       "      <td>WP_041417799.1</td>\n",
       "      <td>752770688</td>\n",
       "      <td>Shewanella woodyi ATCC 51908</td>\n",
       "      <td>ATGACTATGAATAAGCTGTTTGATGGCGTAGTGATCTTCACCCAAG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TATAGGTCAACCTCTTGCAGGAACGGTAAC</td>\n",
       "      <td>11</td>\n",
       "      <td>MTMNKLFDGVVIFTQVVKSGGFSAAAEALGHSTSYVSKEVNKLESR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000279145.1</th>\n",
       "      <td>WP_014855372.1</td>\n",
       "      <td>504668270</td>\n",
       "      <td>Melioribacter roseus P3M-2</td>\n",
       "      <td>ATGCCATACAAAATCATTAAGAAAGAAGAATTCGACAAAGTACGGT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTTGAATTTCTTAAATAACTTGAATAAGTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MPYKIIKKEEFDKVRYHEGDWELRMRLYAEMCRYNTLVAVKKAGSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001310255.1</th>\n",
       "      <td>WP_054766197.1</td>\n",
       "      <td>938988888</td>\n",
       "      <td>Brevundimonas sp. DS20</td>\n",
       "      <td>ATGTCCCTGCGCCCCCTCTTCGTCACCCAGGTCTATGAAGCCTCCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGCGCCCCGTGGGCGACACGGACCTTCCCC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSLRPLFVTQVYEASLAEGRGWPAFNAQLIEVVRMMAEEDAAGRRW...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000255295.1</th>\n",
       "      <td>WP_014400608.1</td>\n",
       "      <td>504213506</td>\n",
       "      <td>Corallococcus coralloides DSM 2259</td>\n",
       "      <td>ATGCATGACGACCTGAGCCCGTTGTGCTCGCGGTTCCTGGCGGCGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TGTCCGGGGCGCGGGTGCTGCGGGCGGCCT</td>\n",
       "      <td>11</td>\n",
       "      <td>MHDDLSPLCSRFLAAADLLGRRWTGVILRILMDGPCRFGELTERIG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001022135.1</th>\n",
       "      <td>WP_047373182.1</td>\n",
       "      <td>829967038</td>\n",
       "      <td>Kluyvera intermedia</td>\n",
       "      <td>ATGTCAAAACCTGACTGGCAGCCCGACACGCCGCTCTATTTGCAGC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CCGAGTATCCAGGCAGGGCTAAACACATCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSKPDWQPDTPLYLQLYQRYREAITSGRLRPGDRVPSIRNLASELH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001310255.1</th>\n",
       "      <td>WP_054764825.1</td>\n",
       "      <td>938987516</td>\n",
       "      <td>Brevundimonas sp. DS20</td>\n",
       "      <td>ATGTCCATCATCGACAAGGCCTTGGCCGCCGTAACCCCCCAACCCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCCGATGAGCGACCGCAACCACAGCCGCGA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSIIDKALAAVTPQPGQEARAEATAQARAAAAPGDWLSLVLDHHDE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001420995.1</th>\n",
       "      <td>WP_010225305.1</td>\n",
       "      <td>497911149</td>\n",
       "      <td>Pseudonocardia sp. EC080619-01</td>\n",
       "      <td>ATGTGCGGGATCGTGGGGTACGTCGGGCCGCGGGAGGCCGCGGCGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CGGCGCGGGAGGACGCAGTGGAGTCACGGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MCGIVGYVGPREAAAVLLDGLARLEYRGYDSAGLAVVHRRRVTVRR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000147695.2</th>\n",
       "      <td>WP_014062581.1</td>\n",
       "      <td>503828587</td>\n",
       "      <td>Thermoanaerobacter wiegelii Rt8.B1</td>\n",
       "      <td>ATGAGTATGTTAAAATATTTAATAGGACTAATGTTGGACGGAGGTG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGGCATGTTTTTTAGAATGCCCTCATATTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MSMLKYLIGLMLDGGVILGNKYLYDKKTVCPVCKKEFTYTKVRSSQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001507665.1</th>\n",
       "      <td>WP_062157480.1</td>\n",
       "      <td>1011238412</td>\n",
       "      <td>Deinococcus actinosclerus</td>\n",
       "      <td>ATGAGCCACGAACGAACGAGCGGTCAGCCGAGTGTCACCCCGCACC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCGTCAGTCGGGGCCGTTGCGGACGCGGCG</td>\n",
       "      <td>11</td>\n",
       "      <td>MSHERTSGQPSVTPHLPVLPTLPPLWDELRAYALPGGLLARLGREQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000247715.1</th>\n",
       "      <td>WP_014360514.1</td>\n",
       "      <td>504127208</td>\n",
       "      <td>Gordonia polyisoprenivorans VH2</td>\n",
       "      <td>ATGTCCAAGTCGGGCAAGTCGGATTCGCAGAAGAGCGCTCCCGCCA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCCGACCGCAACTCGTCGATGGCGGGGCCC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSKSGKSDSQKSAPARTVWAAGGVLWRRSKGTVQIGVVHRPRYDDW...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000147055.1</th>\n",
       "      <td>WP_013318543.1</td>\n",
       "      <td>503083671</td>\n",
       "      <td>Dickeya dadantii 3937</td>\n",
       "      <td>ATGAAAAGCGCCATGATCATCGCTGCCATCATGATGACGCTTAGCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCTGAATAGTCTGCTGGATGAAACTGAAGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MKSAMIIAAIMMTLSAGVGVQSWRLHNTRQLTEQQAQTLSLQQTAL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000196515.1</th>\n",
       "      <td>WP_013190959.1</td>\n",
       "      <td>502955983</td>\n",
       "      <td>'Nostoc azollae' 0708</td>\n",
       "      <td>ATGAATCCAGTTCTGCAATTTGGCGATGCCTCCGGCGAGGAAAGAT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GTTTTCAGGAAGATGAGGCATAAAAAAACC</td>\n",
       "      <td>11</td>\n",
       "      <td>MNPVLQFGDASGEERSHLLTAPEVLELLGKYQLIPPLLKEVVIDQA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000014565.1</th>\n",
       "      <td>WP_011596118.1</td>\n",
       "      <td>499915384</td>\n",
       "      <td>Rhodococcus jostii RHA1</td>\n",
       "      <td>ATGCGCGACATCCTGACGGGCCTGGAAGCGTGGTTCGACCGCGGCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CCATCGTTTGACGGGTATTCGGGTCGAGAC</td>\n",
       "      <td>11</td>\n",
       "      <td>MRDILTGLEAWFDRGENFALATVVRTWKSSPRAPGAAMAVSQSGEV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001620305.1</th>\n",
       "      <td>WP_032381691.1</td>\n",
       "      <td>694034537</td>\n",
       "      <td>Rhodococcus fascians D188</td>\n",
       "      <td>TTGAGTACCGAACAGCTCGAATTCCAGGCGGAGACGCGTCAGCTCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGGCTTCGCCCCATGTGAGTGGAACTTCG</td>\n",
       "      <td>11</td>\n",
       "      <td>MSTEQLEFQAETRQLLDLMVHSIYSNKDSFLRELISNSSDALDKLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000067165.1</th>\n",
       "      <td>WP_044966841.1</td>\n",
       "      <td>769200837</td>\n",
       "      <td>Sorangium cellulosum So ce56</td>\n",
       "      <td>ATGAGCGCGCACGCCGAGGCCGCCGACCGGCCGAACAAGGCAGGCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGCCTAGAGGACAGCATGTCGACACACGCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSAHAEAADRPNKAGPILSPNFHVAHHFDSADTQFDAGRMGIWLFL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000270085.1</th>\n",
       "      <td>WP_013852844.1</td>\n",
       "      <td>503618768</td>\n",
       "      <td>Erysipelothrix rhusiopathiae str. Fujisawa</td>\n",
       "      <td>ATGAAAGTATTAGTAGGTTTATCTGGAGGCGTCGATAGCGCAGTCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCTTGATCAGGGACGGTTTGAATTTACACC</td>\n",
       "      <td>11</td>\n",
       "      <td>MKVLVGLSGGVDSAVAAHILKSQGHDVTCAFMRNWDSFANNDILGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000010605.1</th>\n",
       "      <td>WP_042497106.1</td>\n",
       "      <td>755147116</td>\n",
       "      <td>Streptomyces griseus subsp. griseus NBRC 13350</td>\n",
       "      <td>GTGGGGGCGTCGGGCCTGTTCGAGCAGGTGGCGCAGTTGTGCGCGC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGAGCGGCTCGTCGTCCGTGGTGCCGTACG</td>\n",
       "      <td>11</td>\n",
       "      <td>MGASGLFEQVAQLCAHGVAATHGTADDRFVDHRPDVGRLVPVEQLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000972865.1</th>\n",
       "      <td>WP_052631282.1</td>\n",
       "      <td>918903501</td>\n",
       "      <td>Pseudoxanthomonas suwonensis</td>\n",
       "      <td>ATGCCTTCCTTCGACATCGTCTCCGAAGTCGACACCCACGAGCTGA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCCGCGCCGCGATGTCCGCCGACCACGACA</td>\n",
       "      <td>11</td>\n",
       "      <td>MPSFDIVSEVDTHELTNAVDQASRELTTRFDFKGVDAKFVREDSVI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001618685.1</th>\n",
       "      <td>WP_063165902.1</td>\n",
       "      <td>1017203233</td>\n",
       "      <td>Geobacillus subterraneus</td>\n",
       "      <td>ATGAACAACGAAAAAGCATTACCGTTTATTTATGAATTGACAAATC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GTGTTGAAAAAGAAACTTCTTCTTGAATTA</td>\n",
       "      <td>11</td>\n",
       "      <td>MNNEKALPFIYELTNQFDKKEMEKGLVIELLKLHTLFFAMENEKYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000737325.1</th>\n",
       "      <td>WP_053231157.1</td>\n",
       "      <td>921295836</td>\n",
       "      <td>Sandaracinus amylolyticus</td>\n",
       "      <td>ATGGCGATCGAGCGCGCATCTCGGAGCAGGGCGTCGAGCTTCGTTT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCAGTCCCAGAGATCGTAGTCGTCGCGCGA</td>\n",
       "      <td>11</td>\n",
       "      <td>MAIERASRSRASSFVSDLRVACLLALATTTSCTFDAPRIEGRACPC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000023865.1</th>\n",
       "      <td>WP_012795617.1</td>\n",
       "      <td>502461869</td>\n",
       "      <td>Saccharomonospora viridis DSM 43017</td>\n",
       "      <td>ATGAGCGCCTCCGGCCGTGACGTGGACGACGACGAGCTCGACGCCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>ACAAGCCGGATGCCGCAAGCGGAGGACTCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSASGRDVDDDELDARLRALFADPRLDVASLPRTEEIVAGARRLRR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000952035.1</th>\n",
       "      <td>WP_046924274.1</td>\n",
       "      <td>822197830</td>\n",
       "      <td>Streptomyces lydicus A02</td>\n",
       "      <td>ATGCCCCCCGCACTGCGTTACGACGACCTGAGCGCCCTCTTCATCA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGTGCGAGCGGTGGCGGCGGCCGGTCTCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MPPALRYDDLSALFINCTLKRSPETSNTEGLIDKSRGIMERNGVST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000189415.1</th>\n",
       "      <td>WP_013597275.1</td>\n",
       "      <td>503362614</td>\n",
       "      <td>Weeksella virosa DSM 16922</td>\n",
       "      <td>ATGAAAATCATTCGTTTAAACCCTGAAAATATTCCTGCACCTGTTG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>ATATAAACAACGAACAGCTAACAAGGGTTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKIIRLNPENIPAPVGKYSHVTIIPKNSDLYTFSGQIGVDSDGNIP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001310255.1</th>\n",
       "      <td>WP_054766266.1</td>\n",
       "      <td>938988957</td>\n",
       "      <td>Brevundimonas sp. DS20</td>\n",
       "      <td>ATGCTGGACCTGACGCGGCTGATCCATCAGACGCCCGCCCTTCAGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCCAGGACGGCGGACGCCGAGAGCGGCTTA</td>\n",
       "      <td>11</td>\n",
       "      <td>MLDLTRLIHQTPALQAGDHAKYEGVERLLAQTLAERKGLAPDDVAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000245355.1</th>\n",
       "      <td>WP_014673618.1</td>\n",
       "      <td>504486516</td>\n",
       "      <td>Streptomyces hygroscopicus subsp. jinggangensi...</td>\n",
       "      <td>ATGGTGTTCGGTGCGTTGTACGGTCTGGCGCTGGTCACCGAAGGCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCCCAAGCGCAAGGCGACTGTCGGAGTGGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MVFGALYGLALVTEGDDVRDEVVRSANAAGWAVERADPAERTTDPH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000237145.1</th>\n",
       "      <td>WP_043512257.1</td>\n",
       "      <td>759815636</td>\n",
       "      <td>Actinoplanes sp. SE50/110</td>\n",
       "      <td>ATGACGCAGGAACCGATGGACCCCACGGTCGCCTTCCGCGAGCTGA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CCGCCGTCCGGCGGGCGCTGACCGCTGTCC</td>\n",
       "      <td>11</td>\n",
       "      <td>MTQEPMDPTVAFRELSRIKLGDTDLEGVLATVADLAKRSIPGAYEV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001402875.1</th>\n",
       "      <td>WP_055038834.1</td>\n",
       "      <td>940985927</td>\n",
       "      <td>Blastochloris viridis</td>\n",
       "      <td>ATGATGCAACCCACCGACGACATCCTGGCCATCGCCGAGCGCTGGC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TAGCCGTTTCGTAACCCGTAACGTCGCCGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MMQPTDDILAIAERWRGEGREVALATVVETWGSAPRPVGSHLVVDG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001021385.1</th>\n",
       "      <td>WP_047329554.1</td>\n",
       "      <td>829459486</td>\n",
       "      <td>Mycobacterium sp. EPa45</td>\n",
       "      <td>GTGACCGCCACCGCCGACGAGAGCGGCGTCTTCGACGGCGACGCCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGCCGTCCGAGCCGTGGCCCGCCCGCAGA</td>\n",
       "      <td>11</td>\n",
       "      <td>MTATADESGVFDGDAELDVAGVRCPVRVRLAGHLDPIDGRYHWQGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000237145.1</th>\n",
       "      <td>WP_014693151.1</td>\n",
       "      <td>504506049</td>\n",
       "      <td>Actinoplanes sp. SE50/110</td>\n",
       "      <td>GTGAGCGCCCCGGTCCGGTCGGCCGCCGCGCCGATCCGGTTGCACC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGCCACCTGGACCCGGCTGCGGGCACTGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSAPVRSAAAPIRLHLSENPIGPSPLARSAAHAALDDLHIYPDPAR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001483865.1</th>\n",
       "      <td>WP_058936831.1</td>\n",
       "      <td>971867943</td>\n",
       "      <td>Roseateles depolymerans</td>\n",
       "      <td>ATGTTGCGCGCCTTGGTGGTTCTGCTTCTGCTGGTCAACCTGCTGC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCTACAGGCGGCGCACCGCTTCCTGCCGCC</td>\n",
       "      <td>11</td>\n",
       "      <td>MLRALVVLLLLVNLLLLAWRGGWINTGATAPSRLEQQVRPEQLKLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000091305.1</th>\n",
       "      <td>WP_041671180.1</td>\n",
       "      <td>753986246</td>\n",
       "      <td>Streptomyces scabiei 87.22</td>\n",
       "      <td>CTGCTCCTCGGCGGGTGCGGTGACTCGGCCCCCTCGCTGCTCGCGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>ACCCCGGGAGGGACCGATGCAGTCAGAAGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MLLGGCGDSAPSLLAVKAVAAGVASLAPFFDEGEQLGRDEPRLTPL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000020965.1</th>\n",
       "      <td>WP_012547794.1</td>\n",
       "      <td>501542676</td>\n",
       "      <td>Dictyoglomus thermophilum H-6-12</td>\n",
       "      <td>ATGAGAGTAAAGGGTGGAACAATAACAAGAAAGAGACATAAGAAGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GATGAAATTTATCTCCAGTCCTTCTAATGA</td>\n",
       "      <td>11</td>\n",
       "      <td>MRVKGGTITRKRHKKVLKLAKGYKGRRSRIFKIANQSVIKALYRAY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000277125.1</th>\n",
       "      <td>WP_038538645.1</td>\n",
       "      <td>740753361</td>\n",
       "      <td>Mycobacterium intracellulare ATCC 13950</td>\n",
       "      <td>ATGGGATCTGGACCGAGGACAACGGTCGCCCTTTTCGGCGGCGCGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCAGCCACTGGCGTTGAGCCCGTTCTAGTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MGSGPRTTVALFGGAAMLVLAVGCGGGHKEGPSSTTTPTTTPSSSV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000025565.1</th>\n",
       "      <td>YP_003610929.1</td>\n",
       "      <td>296100783</td>\n",
       "      <td>Enterobacter cloacae subsp. cloacae ATCC 13047</td>\n",
       "      <td>ATGGCAGTACCTTCCAGACGTGTGGCAACGGGCGCTGCCCTCCTCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TATGGGCTATGCGCTGCCCGTCTTTGTTTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MAVPSRRVATGAALLACSVIQAGCSSAVSDPAPFANGRYQQPTRFD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000092865.1</th>\n",
       "      <td>WP_013117041.1</td>\n",
       "      <td>502882065</td>\n",
       "      <td>Cellulomonas flavigena DSM 20109</td>\n",
       "      <td>GTGGCGGGCCTGCTCGGGCTGCTCCCGGCAGCCCTCACCACGCGCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGACCCGGTCGGTCAGGCCGTGCGCAGGT</td>\n",
       "      <td>11</td>\n",
       "      <td>MAGLLGLLPAALTTRRYVMDGRDRWADAAVLATGDASAVLTTPMPG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001618685.1</th>\n",
       "      <td>WP_063165881.1</td>\n",
       "      <td>1017203212</td>\n",
       "      <td>Geobacillus subterraneus</td>\n",
       "      <td>ATGAAAACACATACGATCACCGTCAATCCGCGTTTTTGCGAAACCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CAACATTTGAATTGCATCATCCGCCAACAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MKTHTITVNPRFCETDALGHLSNISYFIYLEEARTRLFDELSYGGR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000689415.1</th>\n",
       "      <td>WP_043253536.1</td>\n",
       "      <td>759533522</td>\n",
       "      <td>Pseudomonas knackmussii B13</td>\n",
       "      <td>ATGCGTACCGAGCAACCGAAAGTCATCTACCTCAAGGATTATCAGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CCCTCGGCGAGCCTTGCCAAAGGTGTTACC</td>\n",
       "      <td>11</td>\n",
       "      <td>MRTEQPKVIYLKDYQAPEYLIDETNLTFELYEDHTLVHAQLVMRRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000261045.2</th>\n",
       "      <td>WP_006320502.1</td>\n",
       "      <td>493364076</td>\n",
       "      <td>Serratia plymuthica PRI-2C</td>\n",
       "      <td>ATGGACATCATTAAAGAACTGTTACATGCCCTATGGCAGCAGGATT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TGGCAAAACGCCCGCGCTGGCAGTATGTTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MDIIKELLHALWQQDFETLANPSLVWTLYVLLFLILFLENGLLPAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000187005.1</th>\n",
       "      <td>WP_049775675.1</td>\n",
       "      <td>908631642</td>\n",
       "      <td>Thermus scotoductus SA-01</td>\n",
       "      <td>ATGGTCTCCCTGGCGGAGGCCCCCAAGGGGGGAAGGATCCTCGAGC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>AGTAAACGCGAAGGACAAGCCCCACGGCAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MVSLAEAPKGGRILEPASAHAPFLRAFREAHGTGYRFHAVEVDPLA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001294645.1</th>\n",
       "      <td>WP_060712343.1</td>\n",
       "      <td>983561852</td>\n",
       "      <td>Pseudonocardia sp. HH130629-09</td>\n",
       "      <td>ATGAGCGTTCCCGCACCCGGACCCCGTCCGGGCGACCCGCGGCCCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CGATGGGCCCCACCCGTTCCCGGCTGGACG</td>\n",
       "      <td>11</td>\n",
       "      <td>MSVPAPGPRPGDPRPEQDRPVAVLRAEVDAAVAALDALDDRPVAEH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000024025.1</th>\n",
       "      <td>WP_012785574.1</td>\n",
       "      <td>502435693</td>\n",
       "      <td>Catenulispora acidiphila DSM 44928</td>\n",
       "      <td>ATGGTGTTCTGGGCGGCCCTCGCCCTCGTCGTCGGCGTGATCGTCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGCCGGCGGCATCCGTTACCGAGTCGTTGT</td>\n",
       "      <td>11</td>\n",
       "      <td>MVFWAALALVVGVIVAVGLARLFGRSERAVSGSHNPQALSVVGSAL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001477625.1</th>\n",
       "      <td>WP_058720004.1</td>\n",
       "      <td>970438825</td>\n",
       "      <td>Paucibacter sp. KCTC 42545</td>\n",
       "      <td>ATGGTTGTTATTCGTCTGGCTCGTGGCGGTTCTAAAAAGCGTCCTT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTGACCAAGAAATGACAAGAGCACCCACCG</td>\n",
       "      <td>11</td>\n",
       "      <td>MVVIRLARGGSKKRPFYNIVVANSQQRRDGRFIEKVGFYNPVASGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000019725.1</th>\n",
       "      <td>WP_046154906.1</td>\n",
       "      <td>806637200</td>\n",
       "      <td>Methylobacterium radiotolerans JCM 2831</td>\n",
       "      <td>CTGGACGACCCGACCCCCGCGGATCCGCGCCGCCGCTGGGACACCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGCCCCGGCCGGGATCCCGCACCGGACCCG</td>\n",
       "      <td>11</td>\n",
       "      <td>MDDPTPADPRRRWDTLAQEARDACYDNTRAVADSASQVAARDAASA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001190945.1</th>\n",
       "      <td>WP_052592688.1</td>\n",
       "      <td>918739271</td>\n",
       "      <td>Luteipulveratus mongoliensis</td>\n",
       "      <td>GTGTCTGCCCAGTCGCCTTCATCCGTCGCCGGGAGCGAGCGAACGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GGTCCCGGGAACTCTCACACGTTCCGCTCC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSAQSPSSVAGSERTVPEGTTGTDRFGDDKKVVAMRVDGVLQDLFR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000619905.2</th>\n",
       "      <td>WP_025041641.1</td>\n",
       "      <td>640613551</td>\n",
       "      <td>Nitrosospira briensis C-128</td>\n",
       "      <td>ATGACCATAAAGTCGGATAAGTGGATTCGCCGCATGGCGGCGGAAC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTTTTTGCAGCAATGGCCGCACCTCGGCCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MTIKSDKWIRRMAAEHRMIEPFEPNQVKHANGHRIVSYGTSSYGYD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175061 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 refSeqProteinId   proteinGI  \\\n",
       "genome_accession                               \n",
       "GCF_000284615.1   WP_011783808.1   500107803   \n",
       "GCF_000008325.1   WP_010960091.1   499262551   \n",
       "GCF_000143145.1   WP_013217028.1   502982052   \n",
       "GCF_000063505.1   WP_011932283.1   500457779   \n",
       "GCF_000063505.1   WP_011932424.1   500457920   \n",
       "GCF_000145035.1   WP_013280594.1   503045618   \n",
       "GCF_000219605.1   WP_013982211.1   503748135   \n",
       "GCF_000317935.1   YP_006894825.1   410471544   \n",
       "GCF_000145235.1   WP_013287839.1   503052863   \n",
       "GCF_001648175.1   WP_064122910.1  1031618553   \n",
       "GCF_000294515.1   WP_014909031.1   504721929   \n",
       "GCF_000219605.1   WP_013981774.1   503747698   \n",
       "GCF_000024385.1   WP_012855077.1   502618199   \n",
       "GCF_000196175.1   WP_011165490.1   499478850   \n",
       "GCF_001189295.1   WP_050435835.1   913456426   \n",
       "GCF_000019525.1   WP_041417799.1   752770688   \n",
       "GCF_000279145.1   WP_014855372.1   504668270   \n",
       "GCF_001310255.1   WP_054766197.1   938988888   \n",
       "GCF_000255295.1   WP_014400608.1   504213506   \n",
       "GCF_001022135.1   WP_047373182.1   829967038   \n",
       "GCF_001310255.1   WP_054764825.1   938987516   \n",
       "GCF_001420995.1   WP_010225305.1   497911149   \n",
       "GCF_000147695.2   WP_014062581.1   503828587   \n",
       "GCF_001507665.1   WP_062157480.1  1011238412   \n",
       "GCF_000247715.1   WP_014360514.1   504127208   \n",
       "GCF_000147055.1   WP_013318543.1   503083671   \n",
       "GCF_000196515.1   WP_013190959.1   502955983   \n",
       "GCF_000014565.1   WP_011596118.1   499915384   \n",
       "GCF_001620305.1   WP_032381691.1   694034537   \n",
       "GCF_000067165.1   WP_044966841.1   769200837   \n",
       "...                          ...         ...   \n",
       "GCF_000270085.1   WP_013852844.1   503618768   \n",
       "GCF_000010605.1   WP_042497106.1   755147116   \n",
       "GCF_000972865.1   WP_052631282.1   918903501   \n",
       "GCF_001618685.1   WP_063165902.1  1017203233   \n",
       "GCF_000737325.1   WP_053231157.1   921295836   \n",
       "GCF_000023865.1   WP_012795617.1   502461869   \n",
       "GCF_000952035.1   WP_046924274.1   822197830   \n",
       "GCF_000189415.1   WP_013597275.1   503362614   \n",
       "GCF_001310255.1   WP_054766266.1   938988957   \n",
       "GCF_000245355.1   WP_014673618.1   504486516   \n",
       "GCF_000237145.1   WP_043512257.1   759815636   \n",
       "GCF_001402875.1   WP_055038834.1   940985927   \n",
       "GCF_001021385.1   WP_047329554.1   829459486   \n",
       "GCF_000237145.1   WP_014693151.1   504506049   \n",
       "GCF_001483865.1   WP_058936831.1   971867943   \n",
       "GCF_000091305.1   WP_041671180.1   753986246   \n",
       "GCF_000020965.1   WP_012547794.1   501542676   \n",
       "GCF_000277125.1   WP_038538645.1   740753361   \n",
       "GCF_000025565.1   YP_003610929.1   296100783   \n",
       "GCF_000092865.1   WP_013117041.1   502882065   \n",
       "GCF_001618685.1   WP_063165881.1  1017203212   \n",
       "GCF_000689415.1   WP_043253536.1   759533522   \n",
       "GCF_000261045.2   WP_006320502.1   493364076   \n",
       "GCF_000187005.1   WP_049775675.1   908631642   \n",
       "GCF_001294645.1   WP_060712343.1   983561852   \n",
       "GCF_000024025.1   WP_012785574.1   502435693   \n",
       "GCF_001477625.1   WP_058720004.1   970438825   \n",
       "GCF_000019725.1   WP_046154906.1   806637200   \n",
       "GCF_001190945.1   WP_052592688.1   918739271   \n",
       "GCF_000619905.2   WP_025041641.1   640613551   \n",
       "\n",
       "                                                       species_name  \\\n",
       "genome_accession                                                      \n",
       "GCF_000284615.1       Marinobacter hydrocarbonoclasticus ATCC 49840   \n",
       "GCF_000008325.1                  Methylococcus capsulatus str. Bath   \n",
       "GCF_000143145.1             Hyphomicrobium denitrificans ATCC 51888   \n",
       "GCF_000063505.1                           Synechococcus sp. WH 7803   \n",
       "GCF_000063505.1                           Synechococcus sp. WH 7803   \n",
       "GCF_000145035.1                   Butyrivibrio proteoclasticus B316   \n",
       "GCF_000219605.1                                Pseudomonas stutzeri   \n",
       "GCF_000317935.1                       Bordetella parapertussis Bpp5   \n",
       "GCF_000145235.1                Micromonospora aurantiaca ATCC 27029   \n",
       "GCF_001648175.1                               Halotalea alkalilenta   \n",
       "GCF_000294515.1                     Nocardiopsis alba ATCC BAA-2165   \n",
       "GCF_000219605.1                                Pseudomonas stutzeri   \n",
       "GCF_000024385.1                   Thermomonospora curvata DSM 43183   \n",
       "GCF_000196175.1                    Bdellovibrio bacteriovorus HD100   \n",
       "GCF_001189295.1                               Chondromyces crocatus   \n",
       "GCF_000019525.1                        Shewanella woodyi ATCC 51908   \n",
       "GCF_000279145.1                          Melioribacter roseus P3M-2   \n",
       "GCF_001310255.1                              Brevundimonas sp. DS20   \n",
       "GCF_000255295.1                  Corallococcus coralloides DSM 2259   \n",
       "GCF_001022135.1                                 Kluyvera intermedia   \n",
       "GCF_001310255.1                              Brevundimonas sp. DS20   \n",
       "GCF_001420995.1                      Pseudonocardia sp. EC080619-01   \n",
       "GCF_000147695.2                  Thermoanaerobacter wiegelii Rt8.B1   \n",
       "GCF_001507665.1                           Deinococcus actinosclerus   \n",
       "GCF_000247715.1                     Gordonia polyisoprenivorans VH2   \n",
       "GCF_000147055.1                               Dickeya dadantii 3937   \n",
       "GCF_000196515.1                               'Nostoc azollae' 0708   \n",
       "GCF_000014565.1                             Rhodococcus jostii RHA1   \n",
       "GCF_001620305.1                           Rhodococcus fascians D188   \n",
       "GCF_000067165.1                        Sorangium cellulosum So ce56   \n",
       "...                                                             ...   \n",
       "GCF_000270085.1          Erysipelothrix rhusiopathiae str. Fujisawa   \n",
       "GCF_000010605.1      Streptomyces griseus subsp. griseus NBRC 13350   \n",
       "GCF_000972865.1                        Pseudoxanthomonas suwonensis   \n",
       "GCF_001618685.1                            Geobacillus subterraneus   \n",
       "GCF_000737325.1                           Sandaracinus amylolyticus   \n",
       "GCF_000023865.1                 Saccharomonospora viridis DSM 43017   \n",
       "GCF_000952035.1                            Streptomyces lydicus A02   \n",
       "GCF_000189415.1                          Weeksella virosa DSM 16922   \n",
       "GCF_001310255.1                              Brevundimonas sp. DS20   \n",
       "GCF_000245355.1   Streptomyces hygroscopicus subsp. jinggangensi...   \n",
       "GCF_000237145.1                           Actinoplanes sp. SE50/110   \n",
       "GCF_001402875.1                               Blastochloris viridis   \n",
       "GCF_001021385.1                             Mycobacterium sp. EPa45   \n",
       "GCF_000237145.1                           Actinoplanes sp. SE50/110   \n",
       "GCF_001483865.1                             Roseateles depolymerans   \n",
       "GCF_000091305.1                          Streptomyces scabiei 87.22   \n",
       "GCF_000020965.1                    Dictyoglomus thermophilum H-6-12   \n",
       "GCF_000277125.1             Mycobacterium intracellulare ATCC 13950   \n",
       "GCF_000025565.1      Enterobacter cloacae subsp. cloacae ATCC 13047   \n",
       "GCF_000092865.1                    Cellulomonas flavigena DSM 20109   \n",
       "GCF_001618685.1                            Geobacillus subterraneus   \n",
       "GCF_000689415.1                         Pseudomonas knackmussii B13   \n",
       "GCF_000261045.2                          Serratia plymuthica PRI-2C   \n",
       "GCF_000187005.1                           Thermus scotoductus SA-01   \n",
       "GCF_001294645.1                      Pseudonocardia sp. HH130629-09   \n",
       "GCF_000024025.1                  Catenulispora acidiphila DSM 44928   \n",
       "GCF_001477625.1                          Paucibacter sp. KCTC 42545   \n",
       "GCF_000019725.1             Methylobacterium radiotolerans JCM 2831   \n",
       "GCF_001190945.1                        Luteipulveratus mongoliensis   \n",
       "GCF_000619905.2                         Nitrosospira briensis C-128   \n",
       "\n",
       "                                                             DNASeq stopCodon  \\\n",
       "genome_accession                                                                \n",
       "GCF_000284615.1   ATGAGTCTGCGTCTTGGAGATACTGCACCCGATTTTGAACAGGATT...       TGA   \n",
       "GCF_000008325.1   GTGCCAATAGACTTGTCGAATTCTGAATGCCCCTTGGATCGCGATC...       TGA   \n",
       "GCF_000143145.1   ATGGCTCTCAATTTCGACGGTTTAGCGTCTAATGCTGCGGTCGCCC...       TGA   \n",
       "GCF_000063505.1   ATGGGCAGCAGTTTCGGCACCCTGTTTCGCATCAGCACCTTCGGCG...       TGA   \n",
       "GCF_000063505.1   TTGCTGACTCCCCTCAAGGACTGCCTCAAGGATCTGCTCACCACCC...       TGA   \n",
       "GCF_000145035.1   ATGGGTAATAAGACAAAAGTTGCCTGTGTGCAATTTGCCTGTGGCG...       TGA   \n",
       "GCF_000219605.1   ATGGCTTTGATCGTACAGAAGTTTGGGGGCACCTCGGTCGGCACCG...       TGA   \n",
       "GCF_000317935.1   GTGAGCCAGCAGGAGAGCCGCATGTTCGCACCCGCCTGCGTGTCCG...       TGA   \n",
       "GCF_000145235.1   ATGAGTCGTCCCGACGTCGTCGTCACCGGGCTCGGCGCGACGACCC...       TGA   \n",
       "GCF_001648175.1   ATGGCAAGACGCACCAAGGCAGAGGCCGAAGCCACTCGATCGATAC...       TGA   \n",
       "GCF_000294515.1   ATGGGCGAGACGACCGAGACCGACACCGGCGGCGGAACCACCGAGG...       TGA   \n",
       "GCF_000219605.1   ATGAAACTCGCAACCCTCAACCAGGGCCGCGACGGCGTGCTCGTCG...       TGA   \n",
       "GCF_000024385.1   ATGAACGTGATCGCGCTGACCCGCATCGAGCCGGGCATGGTCGAGC...       TGA   \n",
       "GCF_000196175.1   ATGAGTACTCACCGCTGCGACATTTTGATTATCGGCTCTGGCACTG...       TGA   \n",
       "GCF_001189295.1   TTGAGCACGCTCATCACGCCGTGGATCCGCGGGCTCGCGCTCGCTT...       TGA   \n",
       "GCF_000019525.1   ATGACTATGAATAAGCTGTTTGATGGCGTAGTGATCTTCACCCAAG...       TGA   \n",
       "GCF_000279145.1   ATGCCATACAAAATCATTAAGAAAGAAGAATTCGACAAAGTACGGT...       TGA   \n",
       "GCF_001310255.1   ATGTCCCTGCGCCCCCTCTTCGTCACCCAGGTCTATGAAGCCTCCC...       TGA   \n",
       "GCF_000255295.1   ATGCATGACGACCTGAGCCCGTTGTGCTCGCGGTTCCTGGCGGCGG...       TGA   \n",
       "GCF_001022135.1   ATGTCAAAACCTGACTGGCAGCCCGACACGCCGCTCTATTTGCAGC...       TGA   \n",
       "GCF_001310255.1   ATGTCCATCATCGACAAGGCCTTGGCCGCCGTAACCCCCCAACCCG...       TGA   \n",
       "GCF_001420995.1   ATGTGCGGGATCGTGGGGTACGTCGGGCCGCGGGAGGCCGCGGCGG...       TGA   \n",
       "GCF_000147695.2   ATGAGTATGTTAAAATATTTAATAGGACTAATGTTGGACGGAGGTG...       TGA   \n",
       "GCF_001507665.1   ATGAGCCACGAACGAACGAGCGGTCAGCCGAGTGTCACCCCGCACC...       TGA   \n",
       "GCF_000247715.1   ATGTCCAAGTCGGGCAAGTCGGATTCGCAGAAGAGCGCTCCCGCCA...       TGA   \n",
       "GCF_000147055.1   ATGAAAAGCGCCATGATCATCGCTGCCATCATGATGACGCTTAGCG...       TGA   \n",
       "GCF_000196515.1   ATGAATCCAGTTCTGCAATTTGGCGATGCCTCCGGCGAGGAAAGAT...       TGA   \n",
       "GCF_000014565.1   ATGCGCGACATCCTGACGGGCCTGGAAGCGTGGTTCGACCGCGGCG...       TGA   \n",
       "GCF_001620305.1   TTGAGTACCGAACAGCTCGAATTCCAGGCGGAGACGCGTCAGCTCC...       TGA   \n",
       "GCF_000067165.1   ATGAGCGCGCACGCCGAGGCCGCCGACCGGCCGAACAAGGCAGGCC...       TGA   \n",
       "...                                                             ...       ...   \n",
       "GCF_000270085.1   ATGAAAGTATTAGTAGGTTTATCTGGAGGCGTCGATAGCGCAGTCG...       TGA   \n",
       "GCF_000010605.1   GTGGGGGCGTCGGGCCTGTTCGAGCAGGTGGCGCAGTTGTGCGCGC...       TGA   \n",
       "GCF_000972865.1   ATGCCTTCCTTCGACATCGTCTCCGAAGTCGACACCCACGAGCTGA...       TGA   \n",
       "GCF_001618685.1   ATGAACAACGAAAAAGCATTACCGTTTATTTATGAATTGACAAATC...       TGA   \n",
       "GCF_000737325.1   ATGGCGATCGAGCGCGCATCTCGGAGCAGGGCGTCGAGCTTCGTTT...       TGA   \n",
       "GCF_000023865.1   ATGAGCGCCTCCGGCCGTGACGTGGACGACGACGAGCTCGACGCCC...       TGA   \n",
       "GCF_000952035.1   ATGCCCCCCGCACTGCGTTACGACGACCTGAGCGCCCTCTTCATCA...       TGA   \n",
       "GCF_000189415.1   ATGAAAATCATTCGTTTAAACCCTGAAAATATTCCTGCACCTGTTG...       TGA   \n",
       "GCF_001310255.1   ATGCTGGACCTGACGCGGCTGATCCATCAGACGCCCGCCCTTCAGG...       TGA   \n",
       "GCF_000245355.1   ATGGTGTTCGGTGCGTTGTACGGTCTGGCGCTGGTCACCGAAGGCG...       TGA   \n",
       "GCF_000237145.1   ATGACGCAGGAACCGATGGACCCCACGGTCGCCTTCCGCGAGCTGA...       TGA   \n",
       "GCF_001402875.1   ATGATGCAACCCACCGACGACATCCTGGCCATCGCCGAGCGCTGGC...       TGA   \n",
       "GCF_001021385.1   GTGACCGCCACCGCCGACGAGAGCGGCGTCTTCGACGGCGACGCCG...       TGA   \n",
       "GCF_000237145.1   GTGAGCGCCCCGGTCCGGTCGGCCGCCGCGCCGATCCGGTTGCACC...       TGA   \n",
       "GCF_001483865.1   ATGTTGCGCGCCTTGGTGGTTCTGCTTCTGCTGGTCAACCTGCTGC...       TGA   \n",
       "GCF_000091305.1   CTGCTCCTCGGCGGGTGCGGTGACTCGGCCCCCTCGCTGCTCGCGG...       TGA   \n",
       "GCF_000020965.1   ATGAGAGTAAAGGGTGGAACAATAACAAGAAAGAGACATAAGAAGG...       TGA   \n",
       "GCF_000277125.1   ATGGGATCTGGACCGAGGACAACGGTCGCCCTTTTCGGCGGCGCGG...       TGA   \n",
       "GCF_000025565.1   ATGGCAGTACCTTCCAGACGTGTGGCAACGGGCGCTGCCCTCCTCG...       TGA   \n",
       "GCF_000092865.1   GTGGCGGGCCTGCTCGGGCTGCTCCCGGCAGCCCTCACCACGCGCC...       TGA   \n",
       "GCF_001618685.1   ATGAAAACACATACGATCACCGTCAATCCGCGTTTTTGCGAAACCG...       TGA   \n",
       "GCF_000689415.1   ATGCGTACCGAGCAACCGAAAGTCATCTACCTCAAGGATTATCAGG...       TGA   \n",
       "GCF_000261045.2   ATGGACATCATTAAAGAACTGTTACATGCCCTATGGCAGCAGGATT...       TGA   \n",
       "GCF_000187005.1   ATGGTCTCCCTGGCGGAGGCCCCCAAGGGGGGAAGGATCCTCGAGC...       TGA   \n",
       "GCF_001294645.1   ATGAGCGTTCCCGCACCCGGACCCCGTCCGGGCGACCCGCGGCCCG...       TGA   \n",
       "GCF_000024025.1   ATGGTGTTCTGGGCGGCCCTCGCCCTCGTCGTCGGCGTGATCGTCG...       TGA   \n",
       "GCF_001477625.1   ATGGTTGTTATTCGTCTGGCTCGTGGCGGTTCTAAAAAGCGTCCTT...       TGA   \n",
       "GCF_000019725.1   CTGGACGACCCGACCCCCGCGGATCCGCGCCGCCGCTGGGACACCC...       TGA   \n",
       "GCF_001190945.1   GTGTCTGCCCAGTCGCCTTCATCCGTCGCCGGGAGCGAGCGAACGG...       TGA   \n",
       "GCF_000619905.2   ATGACCATAAAGTCGGATAAGTGGATTCGCCGCATGGCGGCGGAAC...       TGA   \n",
       "\n",
       "                                      DNA3UTRSeq  codonTableBio  \\\n",
       "genome_accession                                                  \n",
       "GCF_000284615.1   TCCACCGCCACAAACAAAAAGGGCAGCCAA             11   \n",
       "GCF_000008325.1   TCGCTACACATTGATCGAGCGGGAGCTCGA             11   \n",
       "GCF_000143145.1   ATCTGGAAGAACAATTTCCAAGCGCGTGCG             11   \n",
       "GCF_000063505.1   AGGATCTCAGGAGCGCTTAGCTCCGGGCCG             11   \n",
       "GCF_000063505.1   TTTAAGATCTCCATGTCGCGCAGACAACGC             11   \n",
       "GCF_000145035.1   GGGAGAGAGTACTGCGAAGTGGGGAATGCA             11   \n",
       "GCF_000219605.1   TTCGCAGAAGGCGCGCTAGTCCGCGCCTTC             11   \n",
       "GCF_000317935.1   AGGCGCGATGGCGGCCCGGCCGGCCGAGCC             11   \n",
       "GCF_000145235.1   GGTTCCGCCCTGAGGTAAGGAAGGGCCCCT             11   \n",
       "GCF_001648175.1   CTCCCTGCGCATCAGCCCGGATGATGAACG             11   \n",
       "GCF_000294515.1   TCTTCCCGGAGAGCCCGGTTCAGTCGGTGC             11   \n",
       "GCF_000219605.1   GCTGACGCTCTACGGCTACTGGCGCTCCAG             11   \n",
       "GCF_000024385.1   CGGCCCGTCCGGGCATCGGCGCCGAGCCGC             11   \n",
       "GCF_000196175.1   CTTGCCCGTCCGGGGCGAGTCACGGCAAGC             11   \n",
       "GCF_001189295.1   GCGGCAGGTCGCCATGGCCGAGTCGCCCGC             11   \n",
       "GCF_000019525.1   TATAGGTCAACCTCTTGCAGGAACGGTAAC             11   \n",
       "GCF_000279145.1   TTTGAATTTCTTAAATAACTTGAATAAGTT             11   \n",
       "GCF_001310255.1   GGCGCCCCGTGGGCGACACGGACCTTCCCC             11   \n",
       "GCF_000255295.1   TGTCCGGGGCGCGGGTGCTGCGGGCGGCCT             11   \n",
       "GCF_001022135.1   CCGAGTATCCAGGCAGGGCTAAACACATCA             11   \n",
       "GCF_001310255.1   GCCGATGAGCGACCGCAACCACAGCCGCGA             11   \n",
       "GCF_001420995.1   CGGCGCGGGAGGACGCAGTGGAGTCACGGC             11   \n",
       "GCF_000147695.2   GGGCATGTTTTTTAGAATGCCCTCATATTT             11   \n",
       "GCF_001507665.1   TCGTCAGTCGGGGCCGTTGCGGACGCGGCG             11   \n",
       "GCF_000247715.1   GCCGACCGCAACTCGTCGATGGCGGGGCCC             11   \n",
       "GCF_000147055.1   GCTGAATAGTCTGCTGGATGAAACTGAAGC             11   \n",
       "GCF_000196515.1   GTTTTCAGGAAGATGAGGCATAAAAAAACC             11   \n",
       "GCF_000014565.1   CCATCGTTTGACGGGTATTCGGGTCGAGAC             11   \n",
       "GCF_001620305.1   GCGGCTTCGCCCCATGTGAGTGGAACTTCG             11   \n",
       "GCF_000067165.1   GGCCTAGAGGACAGCATGTCGACACACGCA             11   \n",
       "...                                          ...            ...   \n",
       "GCF_000270085.1   TCTTGATCAGGGACGGTTTGAATTTACACC             11   \n",
       "GCF_000010605.1   GGAGCGGCTCGTCGTCCGTGGTGCCGTACG             11   \n",
       "GCF_000972865.1   GCCGCGCCGCGATGTCCGCCGACCACGACA             11   \n",
       "GCF_001618685.1   GTGTTGAAAAAGAAACTTCTTCTTGAATTA             11   \n",
       "GCF_000737325.1   TCAGTCCCAGAGATCGTAGTCGTCGCGCGA             11   \n",
       "GCF_000023865.1   ACAAGCCGGATGCCGCAAGCGGAGGACTCA             11   \n",
       "GCF_000952035.1   GCGTGCGAGCGGTGGCGGCGGCCGGTCTCA             11   \n",
       "GCF_000189415.1   ATATAAACAACGAACAGCTAACAAGGGTTT             11   \n",
       "GCF_001310255.1   TCCAGGACGGCGGACGCCGAGAGCGGCTTA             11   \n",
       "GCF_000245355.1   GCCCAAGCGCAAGGCGACTGTCGGAGTGGC             11   \n",
       "GCF_000237145.1   CCGCCGTCCGGCGGGCGCTGACCGCTGTCC             11   \n",
       "GCF_001402875.1   TAGCCGTTTCGTAACCCGTAACGTCGCCGC             11   \n",
       "GCF_001021385.1   GCGCCGTCCGAGCCGTGGCCCGCCCGCAGA             11   \n",
       "GCF_000237145.1   GCGCCACCTGGACCCGGCTGCGGGCACTGC             11   \n",
       "GCF_001483865.1   TCTACAGGCGGCGCACCGCTTCCTGCCGCC             11   \n",
       "GCF_000091305.1   ACCCCGGGAGGGACCGATGCAGTCAGAAGC             11   \n",
       "GCF_000020965.1   GATGAAATTTATCTCCAGTCCTTCTAATGA             11   \n",
       "GCF_000277125.1   TCAGCCACTGGCGTTGAGCCCGTTCTAGTT             11   \n",
       "GCF_000025565.1   TATGGGCTATGCGCTGCCCGTCTTTGTTTT             11   \n",
       "GCF_000092865.1   GCGACCCGGTCGGTCAGGCCGTGCGCAGGT             11   \n",
       "GCF_001618685.1   CAACATTTGAATTGCATCATCCGCCAACAA             11   \n",
       "GCF_000689415.1   CCCTCGGCGAGCCTTGCCAAAGGTGTTACC             11   \n",
       "GCF_000261045.2   TGGCAAAACGCCCGCGCTGGCAGTATGTTT             11   \n",
       "GCF_000187005.1   AGTAAACGCGAAGGACAAGCCCCACGGCAA             11   \n",
       "GCF_001294645.1   CGATGGGCCCCACCCGTTCCCGGCTGGACG             11   \n",
       "GCF_000024025.1   GGCCGGCGGCATCCGTTACCGAGTCGTTGT             11   \n",
       "GCF_001477625.1   TTGACCAAGAAATGACAAGAGCACCCACCG             11   \n",
       "GCF_000019725.1   GGCCCCGGCCGGGATCCCGCACCGGACCCG             11   \n",
       "GCF_001190945.1   GGTCCCGGGAACTCTCACACGTTCCGCTCC             11   \n",
       "GCF_000619905.2   TTTTTTGCAGCAATGGCCGCACCTCGGCCA             11   \n",
       "\n",
       "                                                         proteinSeq  \n",
       "genome_accession                                                     \n",
       "GCF_000284615.1   MSLRLGDTAPDFEQDSTEGRIRFHEWLGNGWGVLFSHPADFTPVCT...  \n",
       "GCF_000008325.1   MPIDLSNSECPLDRDLPSSLMRATVDQAPVAISITDLKANILYVNA...  \n",
       "GCF_000143145.1   MALNFDGLASNAAVARSSGRRRPANVNFAETSWGRRGFASALDQFG...  \n",
       "GCF_000063505.1   MGSSFGTLFRISTFGESHGGGVGVILEGCPPRLELDPLAIQAELDR...  \n",
       "GCF_000063505.1   MLTPLKDCLKDLLTTLECPLCRQELERSPRRGALCMACWQRMDLAA...  \n",
       "GCF_000145035.1   MGNKTKVACVQFACGAIASSDADQVKRNIETADRLTREAASGGAKI...  \n",
       "GCF_000219605.1   MALIVQKFGGTSVGTVERIEQVAEKVKKFRDGGDDIVVVVSAMSGE...  \n",
       "GCF_000317935.1   MSQQESRMFAPACVSEFVRYDVDAGVARITLDRPDRRNAIDVPMRA...  \n",
       "GCF_000145235.1   MSRPDVVVTGLGATTPLGGDVASTWDAMLAGRSGVGALTQEWAAQL...  \n",
       "GCF_001648175.1   MARRTKAEAEATRSILLDAAEEIFFEKGVSTTSLEQIARRAGMTRG...  \n",
       "GCF_000294515.1   MGETTETDTGGGTTEESDGGHVWSAGITAAGAALMFILLRLMAVTH...  \n",
       "GCF_000219605.1   MKLATLNQGRDGVLVVVSRDLAQAVKVPQIAATLQAALDDWNYCKP...  \n",
       "GCF_000024385.1   MNVIALTRIEPGMVELVGGKAAGLGGLIGAGERVPDGFCVTTAAHD...  \n",
       "GCF_000196175.1   MSTHRCDILIIGSGTAGLALALKLSENGKVIVLAKESAGGTNSAMA...  \n",
       "GCF_001189295.1   MSTLITPWIRGLALASAVLVTGCASGLGDVDGPAPAGEAAAMPTFH...  \n",
       "GCF_000019525.1   MTMNKLFDGVVIFTQVVKSGGFSAAAEALGHSTSYVSKEVNKLESR...  \n",
       "GCF_000279145.1   MPYKIIKKEEFDKVRYHEGDWELRMRLYAEMCRYNTLVAVKKAGSG...  \n",
       "GCF_001310255.1   MSLRPLFVTQVYEASLAEGRGWPAFNAQLIEVVRMMAEEDAAGRRW...  \n",
       "GCF_000255295.1   MHDDLSPLCSRFLAAADLLGRRWTGVILRILMDGPCRFGELTERIG...  \n",
       "GCF_001022135.1   MSKPDWQPDTPLYLQLYQRYREAITSGRLRPGDRVPSIRNLASELH...  \n",
       "GCF_001310255.1   MSIIDKALAAVTPQPGQEARAEATAQARAAAAPGDWLSLVLDHHDE...  \n",
       "GCF_001420995.1   MCGIVGYVGPREAAAVLLDGLARLEYRGYDSAGLAVVHRRRVTVRR...  \n",
       "GCF_000147695.2   MSMLKYLIGLMLDGGVILGNKYLYDKKTVCPVCKKEFTYTKVRSSQ...  \n",
       "GCF_001507665.1   MSHERTSGQPSVTPHLPVLPTLPPLWDELRAYALPGGLLARLGREQ...  \n",
       "GCF_000247715.1   MSKSGKSDSQKSAPARTVWAAGGVLWRRSKGTVQIGVVHRPRYDDW...  \n",
       "GCF_000147055.1   MKSAMIIAAIMMTLSAGVGVQSWRLHNTRQLTEQQAQTLSLQQTAL...  \n",
       "GCF_000196515.1   MNPVLQFGDASGEERSHLLTAPEVLELLGKYQLIPPLLKEVVIDQA...  \n",
       "GCF_000014565.1   MRDILTGLEAWFDRGENFALATVVRTWKSSPRAPGAAMAVSQSGEV...  \n",
       "GCF_001620305.1   MSTEQLEFQAETRQLLDLMVHSIYSNKDSFLRELISNSSDALDKLR...  \n",
       "GCF_000067165.1   MSAHAEAADRPNKAGPILSPNFHVAHHFDSADTQFDAGRMGIWLFL...  \n",
       "...                                                             ...  \n",
       "GCF_000270085.1   MKVLVGLSGGVDSAVAAHILKSQGHDVTCAFMRNWDSFANNDILGN...  \n",
       "GCF_000010605.1   MGASGLFEQVAQLCAHGVAATHGTADDRFVDHRPDVGRLVPVEQLL...  \n",
       "GCF_000972865.1   MPSFDIVSEVDTHELTNAVDQASRELTTRFDFKGVDAKFVREDSVI...  \n",
       "GCF_001618685.1   MNNEKALPFIYELTNQFDKKEMEKGLVIELLKLHTLFFAMENEKYG...  \n",
       "GCF_000737325.1   MAIERASRSRASSFVSDLRVACLLALATTTSCTFDAPRIEGRACPC...  \n",
       "GCF_000023865.1   MSASGRDVDDDELDARLRALFADPRLDVASLPRTEEIVAGARRLRR...  \n",
       "GCF_000952035.1   MPPALRYDDLSALFINCTLKRSPETSNTEGLIDKSRGIMERNGVST...  \n",
       "GCF_000189415.1   MKIIRLNPENIPAPVGKYSHVTIIPKNSDLYTFSGQIGVDSDGNIP...  \n",
       "GCF_001310255.1   MLDLTRLIHQTPALQAGDHAKYEGVERLLAQTLAERKGLAPDDVAC...  \n",
       "GCF_000245355.1   MVFGALYGLALVTEGDDVRDEVVRSANAAGWAVERADPAERTTDPH...  \n",
       "GCF_000237145.1   MTQEPMDPTVAFRELSRIKLGDTDLEGVLATVADLAKRSIPGAYEV...  \n",
       "GCF_001402875.1   MMQPTDDILAIAERWRGEGREVALATVVETWGSAPRPVGSHLVVDG...  \n",
       "GCF_001021385.1   MTATADESGVFDGDAELDVAGVRCPVRVRLAGHLDPIDGRYHWQGL...  \n",
       "GCF_000237145.1   MSAPVRSAAAPIRLHLSENPIGPSPLARSAAHAALDDLHIYPDPAR...  \n",
       "GCF_001483865.1   MLRALVVLLLLVNLLLLAWRGGWINTGATAPSRLEQQVRPEQLKLL...  \n",
       "GCF_000091305.1   MLLGGCGDSAPSLLAVKAVAAGVASLAPFFDEGEQLGRDEPRLTPL...  \n",
       "GCF_000020965.1   MRVKGGTITRKRHKKVLKLAKGYKGRRSRIFKIANQSVIKALYRAY...  \n",
       "GCF_000277125.1   MGSGPRTTVALFGGAAMLVLAVGCGGGHKEGPSSTTTPTTTPSSSV...  \n",
       "GCF_000025565.1   MAVPSRRVATGAALLACSVIQAGCSSAVSDPAPFANGRYQQPTRFD...  \n",
       "GCF_000092865.1   MAGLLGLLPAALTTRRYVMDGRDRWADAAVLATGDASAVLTTPMPG...  \n",
       "GCF_001618685.1   MKTHTITVNPRFCETDALGHLSNISYFIYLEEARTRLFDELSYGGR...  \n",
       "GCF_000689415.1   MRTEQPKVIYLKDYQAPEYLIDETNLTFELYEDHTLVHAQLVMRRN...  \n",
       "GCF_000261045.2   MDIIKELLHALWQQDFETLANPSLVWTLYVLLFLILFLENGLLPAA...  \n",
       "GCF_000187005.1   MVSLAEAPKGGRILEPASAHAPFLRAFREAHGTGYRFHAVEVDPLA...  \n",
       "GCF_001294645.1   MSVPAPGPRPGDPRPEQDRPVAVLRAEVDAAVAALDALDDRPVAEH...  \n",
       "GCF_000024025.1   MVFWAALALVVGVIVAVGLARLFGRSERAVSGSHNPQALSVVGSAL...  \n",
       "GCF_001477625.1   MVVIRLARGGSKKRPFYNIVVANSQQRRDGRFIEKVGFYNPVASGA...  \n",
       "GCF_000019725.1   MDDPTPADPRRRWDTLAQEARDACYDNTRAVADSASQVAARDAASA...  \n",
       "GCF_001190945.1   MSAQSPSSVAGSERTVPEGTTGTDRFGDDKKVVAMRVDGVLQDLFR...  \n",
       "GCF_000619905.2   MTIKSDKWIRRMAAEHRMIEPFEPNQVKHANGHRIVSYGTSSYGYD...  \n",
       "\n",
       "[175061 rows x 8 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outputPath = analysisCtermPath / 'overlapping_start_codon'\n",
    "speciesSetName = 'TGA_non_overlapping_start_codon_subsample'\n",
    "\n",
    "full_analysis_multispecies(refSeqFolder=str(refSeqPath),\n",
    "                           dataSingleSpeciesFolder=str(analysisCtermDataPath / 'Species'),\n",
    "                           dataSpeciesGroupFolder=str(outputPath),\n",
    "                           plotsFolder=str(outputPath),\n",
    "                           speciesSetName=speciesSetName,\n",
    "                           chooseRandomGenomes=False,\n",
    "                           nRandomGenomes=0,\n",
    "                           assemblyAccessionList=[],\n",
    "                           speciesList=[],\n",
    "                           multispeciesAllCDSDf_nr=multispeciesAllCDSDf_nr4,\n",
    "                           dfFormatOutput='csv',\n",
    "                           skipAnalysisExistingDataFiles=True,\n",
    "                           skipStatisticsAnalysis=False,\n",
    "                           method='allSeqDataframe',\n",
    "                           skipClustering=False,\n",
    "                           codonAnalysis=True\n",
    "                          )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## full analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### compute statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cterminal import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['R',\n",
       " 'K',\n",
       " 'H',\n",
       " 'D',\n",
       " 'E',\n",
       " 'S',\n",
       " 'T',\n",
       " 'N',\n",
       " 'Q',\n",
       " 'C',\n",
       " 'G',\n",
       " 'P',\n",
       " 'A',\n",
       " 'V',\n",
       " 'I',\n",
       " 'L',\n",
       " 'M',\n",
       " 'F',\n",
       " 'W',\n",
       " 'Y']"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aaTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "codonAnalysis = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "codonTableBio = Bio.Data.CodonTable.unambiguous_dna_by_id[11]\n",
    "refCodonTableDf = build_refCodonTable(codonTableBio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/users/lserrano/mweber/Analysis_data/C-terminal/Data/Taxonomy'"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotsFolder = str(analysisCtermDataPath / 'Taxonomy')\n",
    "dataSpeciesGroupFolder = str(analysisCtermDataPath / 'Taxonomy')\n",
    "dataSpeciesGroupFolder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "_startCodonVirtualIndex = int(-1e6)\n",
    "_stopCodonVirtualIndex = int(1e6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminus = 'C'\n",
    "statistical_test = 'contingency'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'taxaId_2'"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statName = 'all'\n",
    "dfFormatOutput = 'csv'\n",
    "speciesSetName = taxonSuffix\n",
    "outputFilePrefix = simplify_prefix(speciesSetName + '_subset_' + statName)\n",
    "plotTitle = '{} {} {}'.format(taxonRank, taxonName, re.sub(r'_', r' ', outputFilePrefix))\n",
    "speciesSetName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_statistics(allCDSDf_nr):\n",
    "    # Count the frequency of amino acids\n",
    "    bulkFreqAA, bulkRelFreqAA = count_bulk_aa_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize)\n",
    "    print('\\n#### Amino acid frequency bulk:\\n')\n",
    "    print(bulkFreqAA,'\\n')\n",
    "    print(sorted(bulkRelFreqAA, key=lambda item: item[1], reverse=True))\n",
    "    ctermFreqAA, ntermFreqAA = count_termina_aa_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize)\n",
    "    print('\\n#### Amino acid frequency c-terminal:\\n',ctermFreqAA[-1])\n",
    "    print('\\n#### Amino acid frequency n-terminal:\\n',ntermFreqAA[1])\n",
    "\n",
    "    if codonAnalysis:\n",
    "        # Count the frequency of codons\n",
    "        bulkFreqCodon, bulkRelFreqCodon = count_bulk_codon_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize)\n",
    "        ctermFreqCodon, ntermFreqCodon = count_termina_codon_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize)\n",
    "        print('Bulk codon frequency:\\n',bulkFreqCodon)\n",
    "    else:\n",
    "        bulkFreqCodon, bulkRelFreqCodon = None, None\n",
    "        ctermFreqCodon, ntermFreqCodon = None, None\n",
    "\n",
    "    # Count the frequency of amino acid pairs\n",
    "    subseqSizeAA = 2\n",
    "    bulkFreqAApair, bulkRelativeFreqAApair = count_bulk_subsequence_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize,\n",
    "                                                                              subseqSizeAA, seq_type='protein')\n",
    "    print('Bulk amino acid pairs frequency:\\n',bulkFreqAApair)\n",
    "    # Note: for the N-terminus we consider positions 2 and 3 (position 1 is always methionine)\n",
    "    ctermFreqAApair, ntermFreqAApair = count_termina_subseq_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize,\n",
    "                                                                      subseqSizeAA, seq_type='protein')\n",
    "    print('Amino acid pairs frequency c-terminal:\\n',ctermFreqAApair)\n",
    "\n",
    "    # Count the frequency of hexamers (codon pairs)\n",
    "    bulkFreqHexamer, bulkRelFreqHexamer = count_bulk_subsequence_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize,\n",
    "                                                                           subseqSize=6, seq_type='DNA')\n",
    "    print('Bulk frequency hexamers:\\n',bulkFreqHexamer)\n",
    "    ctermFreqHexamer, ntermFreqHexamer = count_termina_subseq_frequency(allCDSDf_nr, cTerminalSize, nTerminalSize,\n",
    "                                                                        subseqSize=6, seq_type='DNA')\n",
    "    print('Frequency c-terminal hexamers:\\n',ctermFreqHexamer)\n",
    "\n",
    "    statistics = (bulkFreqAA, ctermFreqAA, ntermFreqAA,\n",
    "                  bulkFreqCodon, ctermFreqCodon, ntermFreqCodon,\n",
    "                  bulkFreqAApair, ctermFreqAApair, ntermFreqAApair,\n",
    "                  bulkFreqHexamer, ctermFreqHexamer, ntermFreqHexamer)\n",
    "    return statistics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "multispeciesStatistics1 = compute_statistics(multispeciesAllCDSDf_nr_downsampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with (Path(plotsFolder) / speciesSetName / 'multispeciesStatistics1.json').open('w') as f:\n",
    "    json.dump(multispeciesStatistics1, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with (Path(plotsFolder) / speciesSetName / 'multispeciesStatistics1.json').open('r') as f:\n",
    "    multispeciesStatistics1 = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulkFreqAA, ctermFreqAA, ntermFreqAA,\\\n",
    "    bulkFreqCodon, ctermFreqCodon, ntermFreqCodon,\\\n",
    "    bulkFreqAApair, ctermFreqAApair, ntermFreqAApair,\\\n",
    "    bulkFreqHexamer, ctermFreqHexamer, ntermFreqHexamer = multispeciesStatistics1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = compute_odds_ratio(bulkFreqAA, ctermFreqAA, cTerminalSize, ntermFreqAA, nTerminalSize, computeMultipleTestsCorrection=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAADf.\" + dfFormatOutput)\n",
    "oddsRatioDf.to_csv(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAADf.\" + dfFormatOutput)\n",
    "oddsRatioDf = pd.read_csv(filename, index_col=[0,1,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAADf.\" + dfFormatOutput)\n",
    "oddsRatioDf = pd.read_csv(filename, index_col=[0,1])\n",
    "oddsRatioDf = oddsRatioDf.set_index('terminus', append=True).sort_index().reorder_levels([2,0,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "terminus  position from terminus  observable       \n",
       "C         -20                     BH_multiple_tests    object\n",
       "                                  count                object\n",
       "                                  log2OddsRatio        object\n",
       "                                  oddsRatio            object\n",
       "                                  pvalue               object\n",
       "dtype: object"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDf.T.copy()\n",
    "for col in df.columns:\n",
    "    df[col] = pd.to_numeric(df[col], errors='ignore')\n",
    "oddsRatioDf = df.T.copy()\n",
    "oddsRatioDf.T.dtypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>488780</td>\n",
       "      <td>47480</td>\n",
       "      <td>275797</td>\n",
       "      <td>333561</td>\n",
       "      <td>182998</td>\n",
       "      <td>352860</td>\n",
       "      <td>101092</td>\n",
       "      <td>285549</td>\n",
       "      <td>245132</td>\n",
       "      <td>498550</td>\n",
       "      <td>100170</td>\n",
       "      <td>162500</td>\n",
       "      <td>226555</td>\n",
       "      <td>180558</td>\n",
       "      <td>303537</td>\n",
       "      <td>273341</td>\n",
       "      <td>254372</td>\n",
       "      <td>357255</td>\n",
       "      <td>62759</td>\n",
       "      <td>139135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>-0.0362072</td>\n",
       "      <td>0.108959</td>\n",
       "      <td>0.00599361</td>\n",
       "      <td>0.164728</td>\n",
       "      <td>-0.0496154</td>\n",
       "      <td>-0.147099</td>\n",
       "      <td>-0.0124294</td>\n",
       "      <td>0.0146314</td>\n",
       "      <td>0.212663</td>\n",
       "      <td>-0.00695813</td>\n",
       "      <td>0.000852013</td>\n",
       "      <td>-0.0972637</td>\n",
       "      <td>-0.0141818</td>\n",
       "      <td>0.0223541</td>\n",
       "      <td>0.09804</td>\n",
       "      <td>-0.0542263</td>\n",
       "      <td>-0.0651725</td>\n",
       "      <td>-0.0113364</td>\n",
       "      <td>0.00265558</td>\n",
       "      <td>-0.0307824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>0.975215</td>\n",
       "      <td>1.07845</td>\n",
       "      <td>1.00416</td>\n",
       "      <td>1.12096</td>\n",
       "      <td>0.966194</td>\n",
       "      <td>0.903064</td>\n",
       "      <td>0.991422</td>\n",
       "      <td>1.01019</td>\n",
       "      <td>1.15882</td>\n",
       "      <td>0.995189</td>\n",
       "      <td>1.00059</td>\n",
       "      <td>0.934804</td>\n",
       "      <td>0.990218</td>\n",
       "      <td>1.01562</td>\n",
       "      <td>1.07032</td>\n",
       "      <td>0.963111</td>\n",
       "      <td>0.955831</td>\n",
       "      <td>0.992173</td>\n",
       "      <td>1.00184</td>\n",
       "      <td>0.978889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>2.22805e-62</td>\n",
       "      <td>1.16192e-58</td>\n",
       "      <td>0.0345097</td>\n",
       "      <td>0</td>\n",
       "      <td>1.53812e-47</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00677246</td>\n",
       "      <td>1.58879e-07</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0012718</td>\n",
       "      <td>0.852802</td>\n",
       "      <td>4.90348e-160</td>\n",
       "      <td>4.96715e-06</td>\n",
       "      <td>1.23397e-10</td>\n",
       "      <td>3.20197e-281</td>\n",
       "      <td>7.29844e-82</td>\n",
       "      <td>4.45522e-110</td>\n",
       "      <td>6.25352e-06</td>\n",
       "      <td>0.647519</td>\n",
       "      <td>3.98292e-15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             A            C  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -20                    BH_multiple_tests         True         True   \n",
       "                                count                   488780        47480   \n",
       "                                log2OddsRatio       -0.0362072     0.108959   \n",
       "                                oddsRatio             0.975215      1.07845   \n",
       "                                pvalue             2.22805e-62  1.16192e-58   \n",
       "\n",
       "                                                            D         E  \\\n",
       "terminus position from terminus observable                                \n",
       "C        -20                    BH_multiple_tests        True      True   \n",
       "                                count                  275797    333561   \n",
       "                                log2OddsRatio      0.00599361  0.164728   \n",
       "                                oddsRatio             1.00416   1.12096   \n",
       "                                pvalue              0.0345097         0   \n",
       "\n",
       "                                                             F         G  \\\n",
       "terminus position from terminus observable                                 \n",
       "C        -20                    BH_multiple_tests         True      True   \n",
       "                                count                   182998    352860   \n",
       "                                log2OddsRatio       -0.0496154 -0.147099   \n",
       "                                oddsRatio             0.966194  0.903064   \n",
       "                                pvalue             1.53812e-47         0   \n",
       "\n",
       "                                                            H            I  \\\n",
       "terminus position from terminus observable                                   \n",
       "C        -20                    BH_multiple_tests        True         True   \n",
       "                                count                  101092       285549   \n",
       "                                log2OddsRatio      -0.0124294    0.0146314   \n",
       "                                oddsRatio            0.991422      1.01019   \n",
       "                                pvalue             0.00677246  1.58879e-07   \n",
       "\n",
       "                                                          K           L  \\\n",
       "terminus position from terminus observable                                \n",
       "C        -20                    BH_multiple_tests      True        True   \n",
       "                                count                245132      498550   \n",
       "                                log2OddsRatio      0.212663 -0.00695813   \n",
       "                                oddsRatio           1.15882    0.995189   \n",
       "                                pvalue                    0   0.0012718   \n",
       "\n",
       "                                                             M             N  \\\n",
       "terminus position from terminus observable                                     \n",
       "C        -20                    BH_multiple_tests        False          True   \n",
       "                                count                   100170        162500   \n",
       "                                log2OddsRatio      0.000852013    -0.0972637   \n",
       "                                oddsRatio              1.00059      0.934804   \n",
       "                                pvalue                0.852802  4.90348e-160   \n",
       "\n",
       "                                                             P            Q  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -20                    BH_multiple_tests         True         True   \n",
       "                                count                   226555       180558   \n",
       "                                log2OddsRatio       -0.0141818    0.0223541   \n",
       "                                oddsRatio             0.990218      1.01562   \n",
       "                                pvalue             4.96715e-06  1.23397e-10   \n",
       "\n",
       "                                                              R            S  \\\n",
       "terminus position from terminus observable                                     \n",
       "C        -20                    BH_multiple_tests          True         True   \n",
       "                                count                    303537       273341   \n",
       "                                log2OddsRatio           0.09804   -0.0542263   \n",
       "                                oddsRatio               1.07032     0.963111   \n",
       "                                pvalue             3.20197e-281  7.29844e-82   \n",
       "\n",
       "                                                              T            V  \\\n",
       "terminus position from terminus observable                                     \n",
       "C        -20                    BH_multiple_tests          True         True   \n",
       "                                count                    254372       357255   \n",
       "                                log2OddsRatio        -0.0651725   -0.0113364   \n",
       "                                oddsRatio              0.955831     0.992173   \n",
       "                                pvalue             4.45522e-110  6.25352e-06   \n",
       "\n",
       "                                                            W            Y  \n",
       "terminus position from terminus observable                                  \n",
       "C        -20                    BH_multiple_tests       False         True  \n",
       "                                count                   62759       139135  \n",
       "                                log2OddsRatio      0.00265558   -0.0307824  \n",
       "                                oddsRatio             1.00184     0.978889  \n",
       "                                pvalue               0.647519  3.98292e-15  "
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Multiple test correction within the biases of C-terminal at all positions for all a.a.\n",
    "# df = oddsRatioDf[oddsRatioDf['terminus'] == terminus].drop('terminus', axis=1).copy()\n",
    "# # Drop bulk values\n",
    "# df = df[df.index.get_level_values('position from terminus') > _startCodonVirtualIndex]\n",
    "# df = df[df.index.get_level_values('position from terminus') < _stopCodonVirtualIndex]\n",
    "# df = df.loc[idx[:, 'pvalue'], :]\n",
    "# # We serialize all the values, and drop the NaN\n",
    "# df2 = df.stack().dropna().copy()\n",
    "# reject, pvals_corrected, alphacSidak, alphacBonf = \\\n",
    "#     multipletests(df2.values, alpha=family_wise_FDR, method='fdr_bh',\n",
    "#                   is_sorted=False, returnsorted=False)\n",
    "# df2 = pd.DataFrame(reject, index=df2.index)\n",
    "# # Stack again the values before merging\n",
    "# df2 = df2.unstack()\n",
    "# df2.columns = df2.columns.droplevel(0)\n",
    "# df2 = df2.rename(index={'pvalue':'BH_multiple_tests'})\n",
    "# df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfCodon = compute_odds_ratio_codons(bulkFreqCodon,\n",
    "                                             ctermFreqCodon, cTerminalSize,\n",
    "                                             ntermFreqCodon, nTerminalSize,\n",
    "                                             verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfCodon.\" + dfFormatOutput)\n",
    "oddsRatioDfCodon.to_csv(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCG</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-1000000</th>\n",
       "      <th>count</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               AAA  AAC  AAG  AAT  ACA  ACC  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count          NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                log2OddsRatio  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                oddsRatio      NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                pvalue         NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                terminus       NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                               ACG  ACT  AGA  AGC ...   TCG  \\\n",
       "terminus position from terminus observable                        ...         \n",
       "C        -1000000               count          NaN  NaN  NaN  NaN ...   NaN   \n",
       "                                log2OddsRatio  NaN  NaN  NaN  NaN ...   NaN   \n",
       "                                oddsRatio      NaN  NaN  NaN  NaN ...   NaN   \n",
       "                                pvalue         NaN  NaN  NaN  NaN ...   NaN   \n",
       "                                terminus       NaN  NaN  NaN  NaN ...   NaN   \n",
       "\n",
       "                                               TCT  TGA  TGC  TGG  TGT  TTA  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count          NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                log2OddsRatio  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                oddsRatio      NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                pvalue         NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "                                terminus       NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                               TTC  TTG  TTT  \n",
       "terminus position from terminus observable                    \n",
       "C        -1000000               count          NaN  NaN  NaN  \n",
       "                                log2OddsRatio  NaN  NaN  NaN  \n",
       "                                oddsRatio      NaN  NaN  NaN  \n",
       "                                pvalue         NaN  NaN  NaN  \n",
       "                                terminus       NaN  NaN  NaN  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfCodon.\" + dfFormatOutput)\n",
    "oddsRatioDfCodon = pd.read_csv(filename, index_col=[0,1])\n",
    "oddsRatioDfCodon = oddsRatioDfCodon.reset_index()\\\n",
    "                   .set_index(['terminus', 'position from terminus', 'observable']).sort_index()\n",
    "oddsRatioDfCodon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "terminus  position from terminus  observable   \n",
       "C         -1000000                count            object\n",
       "                                  log2OddsRatio    object\n",
       "                                  oddsRatio        object\n",
       "                                  pvalue           object\n",
       "                                  terminus         object\n",
       "dtype: object"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDfCodon.T.copy()\n",
    "for col in df.columns:\n",
    "    df[col] = pd.to_numeric(df[col], errors='ignore')\n",
    "oddsRatioDfCodon = df.T.copy()\n",
    "oddsRatioDfCodon.T.dtypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCG</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"10\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-1000000</th>\n",
       "      <th>count</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>135804</td>\n",
       "      <td>79078</td>\n",
       "      <td>109328</td>\n",
       "      <td>83421</td>\n",
       "      <td>43900</td>\n",
       "      <td>105021</td>\n",
       "      <td>65989</td>\n",
       "      <td>39462</td>\n",
       "      <td>28107</td>\n",
       "      <td>67929</td>\n",
       "      <td>...</td>\n",
       "      <td>52169</td>\n",
       "      <td>34787</td>\n",
       "      <td>332</td>\n",
       "      <td>29443</td>\n",
       "      <td>62428</td>\n",
       "      <td>18037</td>\n",
       "      <td>67030</td>\n",
       "      <td>96840</td>\n",
       "      <td>68510</td>\n",
       "      <td>86161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.251941</td>\n",
       "      <td>-0.150096</td>\n",
       "      <td>0.15285</td>\n",
       "      <td>-0.0412695</td>\n",
       "      <td>0.0483806</td>\n",
       "      <td>-0.145123</td>\n",
       "      <td>-0.0435355</td>\n",
       "      <td>0.0123124</td>\n",
       "      <td>0.233375</td>\n",
       "      <td>0.00908645</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.150612</td>\n",
       "      <td>-0.0406626</td>\n",
       "      <td>-0.0193254</td>\n",
       "      <td>0.0830465</td>\n",
       "      <td>0.00277434</td>\n",
       "      <td>0.149784</td>\n",
       "      <td>0.0649824</td>\n",
       "      <td>-0.0781173</td>\n",
       "      <td>0.0900188</td>\n",
       "      <td>-0.0144966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.19081</td>\n",
       "      <td>0.901191</td>\n",
       "      <td>1.11176</td>\n",
       "      <td>0.971799</td>\n",
       "      <td>1.0341</td>\n",
       "      <td>0.904302</td>\n",
       "      <td>0.970274</td>\n",
       "      <td>1.00857</td>\n",
       "      <td>1.17558</td>\n",
       "      <td>1.00632</td>\n",
       "      <td>...</td>\n",
       "      <td>0.900868</td>\n",
       "      <td>0.972208</td>\n",
       "      <td>0.986694</td>\n",
       "      <td>1.05925</td>\n",
       "      <td>1.00192</td>\n",
       "      <td>1.1094</td>\n",
       "      <td>1.04607</td>\n",
       "      <td>0.947293</td>\n",
       "      <td>1.06438</td>\n",
       "      <td>0.990002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0</td>\n",
       "      <td>9.0155e-191</td>\n",
       "      <td>2.18821e-253</td>\n",
       "      <td>2.16753e-16</td>\n",
       "      <td>3.78407e-12</td>\n",
       "      <td>1.0108e-234</td>\n",
       "      <td>1.15244e-14</td>\n",
       "      <td>0.0918914</td>\n",
       "      <td>2.24553e-152</td>\n",
       "      <td>0.103796</td>\n",
       "      <td>...</td>\n",
       "      <td>3.68131e-128</td>\n",
       "      <td>1.52379e-07</td>\n",
       "      <td>0.848375</td>\n",
       "      <td>2.0692e-22</td>\n",
       "      <td>0.63365</td>\n",
       "      <td>1.7966e-42</td>\n",
       "      <td>1.68828e-30</td>\n",
       "      <td>2.58543e-63</td>\n",
       "      <td>8.31532e-58</td>\n",
       "      <td>0.00348624</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        AAA          AAC  \\\n",
       "terminus position from terminus observable                                 \n",
       "C        -1000000               count                   NaN          NaN   \n",
       "                                log2OddsRatio           NaN          NaN   \n",
       "                                oddsRatio               NaN          NaN   \n",
       "                                pvalue                  NaN          NaN   \n",
       "                                terminus                NaN          NaN   \n",
       "         -20                    BH_multiple_tests      True         True   \n",
       "                                count                135804        79078   \n",
       "                                log2OddsRatio      0.251941    -0.150096   \n",
       "                                oddsRatio           1.19081     0.901191   \n",
       "                                pvalue                    0  9.0155e-191   \n",
       "\n",
       "                                                            AAG          AAT  \\\n",
       "terminus position from terminus observable                                     \n",
       "C        -1000000               count                       NaN          NaN   \n",
       "                                log2OddsRatio               NaN          NaN   \n",
       "                                oddsRatio                   NaN          NaN   \n",
       "                                pvalue                      NaN          NaN   \n",
       "                                terminus                    NaN          NaN   \n",
       "         -20                    BH_multiple_tests          True         True   \n",
       "                                count                    109328        83421   \n",
       "                                log2OddsRatio           0.15285   -0.0412695   \n",
       "                                oddsRatio               1.11176     0.971799   \n",
       "                                pvalue             2.18821e-253  2.16753e-16   \n",
       "\n",
       "                                                           ACA          ACC  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count                      NaN          NaN   \n",
       "                                log2OddsRatio              NaN          NaN   \n",
       "                                oddsRatio                  NaN          NaN   \n",
       "                                pvalue                     NaN          NaN   \n",
       "                                terminus                   NaN          NaN   \n",
       "         -20                    BH_multiple_tests         True         True   \n",
       "                                count                    43900       105021   \n",
       "                                log2OddsRatio        0.0483806    -0.145123   \n",
       "                                oddsRatio               1.0341     0.904302   \n",
       "                                pvalue             3.78407e-12  1.0108e-234   \n",
       "\n",
       "                                                           ACG        ACT  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -1000000               count                      NaN        NaN   \n",
       "                                log2OddsRatio              NaN        NaN   \n",
       "                                oddsRatio                  NaN        NaN   \n",
       "                                pvalue                     NaN        NaN   \n",
       "                                terminus                   NaN        NaN   \n",
       "         -20                    BH_multiple_tests         True      False   \n",
       "                                count                    65989      39462   \n",
       "                                log2OddsRatio       -0.0435355  0.0123124   \n",
       "                                oddsRatio             0.970274    1.00857   \n",
       "                                pvalue             1.15244e-14  0.0918914   \n",
       "\n",
       "                                                            AGA         AGC  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count                       NaN         NaN   \n",
       "                                log2OddsRatio               NaN         NaN   \n",
       "                                oddsRatio                   NaN         NaN   \n",
       "                                pvalue                      NaN         NaN   \n",
       "                                terminus                    NaN         NaN   \n",
       "         -20                    BH_multiple_tests          True       False   \n",
       "                                count                     28107       67929   \n",
       "                                log2OddsRatio          0.233375  0.00908645   \n",
       "                                oddsRatio               1.17558     1.00632   \n",
       "                                pvalue             2.24553e-152    0.103796   \n",
       "\n",
       "                                                      ...               TCG  \\\n",
       "terminus position from terminus observable            ...                     \n",
       "C        -1000000               count                 ...               NaN   \n",
       "                                log2OddsRatio         ...               NaN   \n",
       "                                oddsRatio             ...               NaN   \n",
       "                                pvalue                ...               NaN   \n",
       "                                terminus              ...               NaN   \n",
       "         -20                    BH_multiple_tests     ...              True   \n",
       "                                count                 ...             52169   \n",
       "                                log2OddsRatio         ...         -0.150612   \n",
       "                                oddsRatio             ...          0.900868   \n",
       "                                pvalue                ...      3.68131e-128   \n",
       "\n",
       "                                                           TCT        TGA  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -1000000               count                      NaN        NaN   \n",
       "                                log2OddsRatio              NaN        NaN   \n",
       "                                oddsRatio                  NaN        NaN   \n",
       "                                pvalue                     NaN        NaN   \n",
       "                                terminus                   NaN        NaN   \n",
       "         -20                    BH_multiple_tests         True      False   \n",
       "                                count                    34787        332   \n",
       "                                log2OddsRatio       -0.0406626 -0.0193254   \n",
       "                                oddsRatio             0.972208   0.986694   \n",
       "                                pvalue             1.52379e-07   0.848375   \n",
       "\n",
       "                                                          TGC         TGG  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -1000000               count                     NaN         NaN   \n",
       "                                log2OddsRatio             NaN         NaN   \n",
       "                                oddsRatio                 NaN         NaN   \n",
       "                                pvalue                    NaN         NaN   \n",
       "                                terminus                  NaN         NaN   \n",
       "         -20                    BH_multiple_tests        True       False   \n",
       "                                count                   29443       62428   \n",
       "                                log2OddsRatio       0.0830465  0.00277434   \n",
       "                                oddsRatio             1.05925     1.00192   \n",
       "                                pvalue             2.0692e-22     0.63365   \n",
       "\n",
       "                                                          TGT          TTA  \\\n",
       "terminus position from terminus observable                                   \n",
       "C        -1000000               count                     NaN          NaN   \n",
       "                                log2OddsRatio             NaN          NaN   \n",
       "                                oddsRatio                 NaN          NaN   \n",
       "                                pvalue                    NaN          NaN   \n",
       "                                terminus                  NaN          NaN   \n",
       "         -20                    BH_multiple_tests        True         True   \n",
       "                                count                   18037        67030   \n",
       "                                log2OddsRatio        0.149784    0.0649824   \n",
       "                                oddsRatio              1.1094      1.04607   \n",
       "                                pvalue             1.7966e-42  1.68828e-30   \n",
       "\n",
       "                                                           TTC          TTG  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count                      NaN          NaN   \n",
       "                                log2OddsRatio              NaN          NaN   \n",
       "                                oddsRatio                  NaN          NaN   \n",
       "                                pvalue                     NaN          NaN   \n",
       "                                terminus                   NaN          NaN   \n",
       "         -20                    BH_multiple_tests         True         True   \n",
       "                                count                    96840        68510   \n",
       "                                log2OddsRatio       -0.0781173    0.0900188   \n",
       "                                oddsRatio             0.947293      1.06438   \n",
       "                                pvalue             2.58543e-63  8.31532e-58   \n",
       "\n",
       "                                                          TTT  \n",
       "terminus position from terminus observable                     \n",
       "C        -1000000               count                     NaN  \n",
       "                                log2OddsRatio             NaN  \n",
       "                                oddsRatio                 NaN  \n",
       "                                pvalue                    NaN  \n",
       "                                terminus                  NaN  \n",
       "         -20                    BH_multiple_tests        True  \n",
       "                                count                   86161  \n",
       "                                log2OddsRatio      -0.0144966  \n",
       "                                oddsRatio            0.990002  \n",
       "                                pvalue             0.00348624  \n",
       "\n",
       "[10 rows x 64 columns]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCodon.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ update compute_odds_ratio_subseq\n",
    "+ update how we plot the results. include filter for BH_test boolean\n",
    "    + aa\n",
    "    + codon\n",
    "    + subseq pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfAApair, oddsRatioDfTableAApair = compute_odds_ratio_subseq(bulkFreqAApair, bulkFreqAA,\n",
    "                                                                      ctermFreqAApair, cTerminalSize,\n",
    "                                                                      ntermFreqAApair, nTerminalSize,\n",
    "                                                                      subseqSize=2, seq_type='protein',\n",
    "                                                                      computeTheoreticalFreq=True,\n",
    "                                                                      computeMultipleTestsCorrection=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfTableAApair.\" + dfFormatOutput)\n",
    "oddsRatioDfTableAApair.to_csv(filename)\n",
    "\n",
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfAApair.\" + dfFormatOutput)\n",
    "oddsRatioDfAApair.to_csv(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'taxaId_31969'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speciesSetName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfTableAApair.\" + dfFormatOutput)\n",
    "oddsRatioDfTableAApair = pd.read_csv(filename, index_col=[0,1,2,3])\n",
    "\n",
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfAApair.\" + dfFormatOutput)\n",
    "oddsRatioDfAApair = pd.read_csv(filename, index_col=[0,1,2,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "terminus  aa0  statistical_test  observable       \n",
       "C         A    contingency       BH_multiple_tests     object\n",
       "                                 NexpectedSubseq      float64\n",
       "                                 count                float64\n",
       "                                 log2OddsRatio        float64\n",
       "                                 oddsRatio            float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDfTableAApair.T.copy()\n",
    "for col in df.columns:\n",
    "    df[col] = pd.to_numeric(df[col], errors='ignore')\n",
    "oddsRatioDfTableAApairT = df.copy()\n",
    "oddsRatioDfTableAApair = oddsRatioDfTableAApairT.T.copy()\n",
    "oddsRatioDfTableAApairT.dtypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>J</th>\n",
       "      <th>...</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>aa0</th>\n",
       "      <th>statistical_test</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"10\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"10\" valign=\"top\">A</th>\n",
       "      <th rowspan=\"7\" valign=\"top\">contingency</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>41</td>\n",
       "      <td>108</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>16</td>\n",
       "      <td>96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>137</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>65</td>\n",
       "      <td>92</td>\n",
       "      <td>30</td>\n",
       "      <td>75</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>-1.479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.64793</td>\n",
       "      <td>-1.35531</td>\n",
       "      <td>-0.131338</td>\n",
       "      <td>-1.37677</td>\n",
       "      <td>-1.95414</td>\n",
       "      <td>-1.1425</td>\n",
       "      <td>-1.24814</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.159742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.305492</td>\n",
       "      <td>-0.246025</td>\n",
       "      <td>-0.752871</td>\n",
       "      <td>-2.14849</td>\n",
       "      <td>-0.75274</td>\n",
       "      <td>-1.29724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.561284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>0.358738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.319097</td>\n",
       "      <td>0.390852</td>\n",
       "      <td>0.912984</td>\n",
       "      <td>0.385079</td>\n",
       "      <td>0.258075</td>\n",
       "      <td>0.452973</td>\n",
       "      <td>0.420992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.895185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.809166</td>\n",
       "      <td>0.843217</td>\n",
       "      <td>0.593421</td>\n",
       "      <td>0.225549</td>\n",
       "      <td>0.593476</td>\n",
       "      <td>0.406905</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.677699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>5.84025e-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00110317</td>\n",
       "      <td>2.2651e-12</td>\n",
       "      <td>0.380493</td>\n",
       "      <td>3.07032e-14</td>\n",
       "      <td>2.08818e-22</td>\n",
       "      <td>0.000494619</td>\n",
       "      <td>1.13767e-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.208316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0952046</td>\n",
       "      <td>0.188424</td>\n",
       "      <td>7.27194e-08</td>\n",
       "      <td>1.05984e-26</td>\n",
       "      <td>1.29187e-06</td>\n",
       "      <td>0.00267262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00163039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>theoreticalProb</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">theoretical</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>119.895</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.5027</td>\n",
       "      <td>125.475</td>\n",
       "      <td>152.602</td>\n",
       "      <td>123.78</td>\n",
       "      <td>116.31</td>\n",
       "      <td>34.4395</td>\n",
       "      <td>209.533</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>165.776</td>\n",
       "      <td>65.2453</td>\n",
       "      <td>82.5303</td>\n",
       "      <td>67.1375</td>\n",
       "      <td>157.386</td>\n",
       "      <td>121.94</td>\n",
       "      <td>130.499</td>\n",
       "      <td>23.0403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93.2691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>41</td>\n",
       "      <td>108</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>16</td>\n",
       "      <td>96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>137</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>65</td>\n",
       "      <td>92</td>\n",
       "      <td>30</td>\n",
       "      <td>75</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           A    B           C  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        A   contingency      BH_multiple_tests         True  NaN        True   \n",
       "                              NexpectedSubseq            NaN  NaN         NaN   \n",
       "                              count                       47  NaN           6   \n",
       "                              log2OddsRatio           -1.479  NaN    -1.64793   \n",
       "                              oddsRatio             0.358738  NaN    0.319097   \n",
       "                              pvalue             5.84025e-17  NaN  0.00110317   \n",
       "                              theoreticalProb            NaN  NaN         NaN   \n",
       "             theoretical      BH_multiple_tests         True  NaN        True   \n",
       "                              NexpectedSubseq        119.895  NaN     14.5027   \n",
       "                              count                       47  NaN           6   \n",
       "\n",
       "                                                          D         E  \\\n",
       "terminus aa0 statistical_test observable                                \n",
       "C        A   contingency      BH_multiple_tests        True     False   \n",
       "                              NexpectedSubseq           NaN       NaN   \n",
       "                              count                      41       108   \n",
       "                              log2OddsRatio        -1.35531 -0.131338   \n",
       "                              oddsRatio            0.390852  0.912984   \n",
       "                              pvalue             2.2651e-12  0.380493   \n",
       "                              theoreticalProb           NaN       NaN   \n",
       "             theoretical      BH_multiple_tests        True      True   \n",
       "                              NexpectedSubseq       125.475   152.602   \n",
       "                              count                      41       108   \n",
       "\n",
       "                                                           F            G  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        A   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       46           32   \n",
       "                              log2OddsRatio         -1.37677     -1.95414   \n",
       "                              oddsRatio             0.385079     0.258075   \n",
       "                              pvalue             3.07032e-14  2.08818e-22   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq         123.78       116.31   \n",
       "                              count                       46           32   \n",
       "\n",
       "                                                           H            I  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        A   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       16           96   \n",
       "                              log2OddsRatio          -1.1425     -1.24814   \n",
       "                              oddsRatio             0.452973     0.420992   \n",
       "                              pvalue             0.000494619  1.13767e-22   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq        34.4395      209.533   \n",
       "                              count                       16           96   \n",
       "\n",
       "                                                   J     ...             N  \\\n",
       "terminus aa0 statistical_test observable                 ...                 \n",
       "C        A   contingency      BH_multiple_tests  NaN     ...         False   \n",
       "                              NexpectedSubseq    NaN     ...           NaN   \n",
       "                              count              NaN     ...           137   \n",
       "                              log2OddsRatio      NaN     ...     -0.159742   \n",
       "                              oddsRatio          NaN     ...      0.895185   \n",
       "                              pvalue             NaN     ...      0.208316   \n",
       "                              theoreticalProb    NaN     ...           NaN   \n",
       "             theoretical      BH_multiple_tests  NaN     ...          True   \n",
       "                              NexpectedSubseq    NaN     ...       165.776   \n",
       "                              count              NaN     ...           137   \n",
       "\n",
       "                                                       P          Q         R  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        A   contingency      BH_multiple_tests      NaN      False     False   \n",
       "                              NexpectedSubseq        NaN        NaN       NaN   \n",
       "                              count                    0         66        65   \n",
       "                              log2OddsRatio          NaN  -0.305492 -0.246025   \n",
       "                              oddsRatio              NaN   0.809166  0.843217   \n",
       "                              pvalue                 NaN  0.0952046  0.188424   \n",
       "                              theoreticalProb        NaN        NaN       NaN   \n",
       "             theoretical      BH_multiple_tests     True      False     False   \n",
       "                              NexpectedSubseq    65.2453    82.5303   67.1375   \n",
       "                              count                    0         66        65   \n",
       "\n",
       "                                                           S            T  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        A   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       92           30   \n",
       "                              log2OddsRatio        -0.752871     -2.14849   \n",
       "                              oddsRatio             0.593421     0.225549   \n",
       "                              pvalue             7.27194e-08  1.05984e-26   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq        157.386       121.94   \n",
       "                              count                       92           30   \n",
       "\n",
       "                                                           V           W    X  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        A   contingency      BH_multiple_tests         True        True  NaN   \n",
       "                              NexpectedSubseq            NaN         NaN  NaN   \n",
       "                              count                       75           9  NaN   \n",
       "                              log2OddsRatio         -0.75274    -1.29724  NaN   \n",
       "                              oddsRatio             0.593476    0.406905  NaN   \n",
       "                              pvalue             1.29187e-06  0.00267262  NaN   \n",
       "                              theoreticalProb            NaN         NaN  NaN   \n",
       "             theoretical      BH_multiple_tests         True        True  NaN   \n",
       "                              NexpectedSubseq        130.499     23.0403  NaN   \n",
       "                              count                       75           9  NaN   \n",
       "\n",
       "                                                          Y  \n",
       "terminus aa0 statistical_test observable                     \n",
       "C        A   contingency      BH_multiple_tests        True  \n",
       "                              NexpectedSubseq           NaN  \n",
       "                              count                      60  \n",
       "                              log2OddsRatio       -0.561284  \n",
       "                              oddsRatio            0.677699  \n",
       "                              pvalue             0.00163039  \n",
       "                              theoreticalProb           NaN  \n",
       "             theoretical      BH_multiple_tests        True  \n",
       "                              NexpectedSubseq       93.2691  \n",
       "                              count                      60  \n",
       "\n",
       "[10 rows x 23 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfTableAApair.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>aa1</th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>aa0</th>\n",
       "      <th>statistical_test</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"20\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"14\" valign=\"top\">R</th>\n",
       "      <th rowspan=\"7\" valign=\"top\">contingency</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>119</td>\n",
       "      <td>456</td>\n",
       "      <td>18</td>\n",
       "      <td>69</td>\n",
       "      <td>58</td>\n",
       "      <td>93</td>\n",
       "      <td>36</td>\n",
       "      <td>130</td>\n",
       "      <td>40</td>\n",
       "      <td>9</td>\n",
       "      <td>115</td>\n",
       "      <td>20</td>\n",
       "      <td>56</td>\n",
       "      <td>67</td>\n",
       "      <td>123</td>\n",
       "      <td>64</td>\n",
       "      <td>17</td>\n",
       "      <td>64</td>\n",
       "      <td>66</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.31738</td>\n",
       "      <td>2.01748</td>\n",
       "      <td>-0.208737</td>\n",
       "      <td>-0.0260725</td>\n",
       "      <td>-0.634304</td>\n",
       "      <td>0.282632</td>\n",
       "      <td>-0.853186</td>\n",
       "      <td>0.592484</td>\n",
       "      <td>-0.26928</td>\n",
       "      <td>0.229385</td>\n",
       "      <td>0.805524</td>\n",
       "      <td>-0.879883</td>\n",
       "      <td>-0.28537</td>\n",
       "      <td>-0.192589</td>\n",
       "      <td>0.0607836</td>\n",
       "      <td>-0.919574</td>\n",
       "      <td>-0.598269</td>\n",
       "      <td>-0.0780036</td>\n",
       "      <td>2.26959</td>\n",
       "      <td>0.103653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>2.49214</td>\n",
       "      <td>4.04876</td>\n",
       "      <td>0.865294</td>\n",
       "      <td>0.98209</td>\n",
       "      <td>0.644251</td>\n",
       "      <td>1.21641</td>\n",
       "      <td>0.553561</td>\n",
       "      <td>1.50784</td>\n",
       "      <td>0.829733</td>\n",
       "      <td>1.17233</td>\n",
       "      <td>1.74778</td>\n",
       "      <td>0.543411</td>\n",
       "      <td>0.820531</td>\n",
       "      <td>0.875034</td>\n",
       "      <td>1.04303</td>\n",
       "      <td>0.528665</td>\n",
       "      <td>0.660546</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>4.82186</td>\n",
       "      <td>1.07449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>5.80119e-18</td>\n",
       "      <td>8.69599e-129</td>\n",
       "      <td>0.659258</td>\n",
       "      <td>0.952277</td>\n",
       "      <td>0.000464047</td>\n",
       "      <td>0.066274</td>\n",
       "      <td>0.000138077</td>\n",
       "      <td>1.15786e-05</td>\n",
       "      <td>0.277891</td>\n",
       "      <td>0.584709</td>\n",
       "      <td>4.46075e-08</td>\n",
       "      <td>0.0036786</td>\n",
       "      <td>0.145126</td>\n",
       "      <td>0.301973</td>\n",
       "      <td>0.644</td>\n",
       "      <td>2.21827e-08</td>\n",
       "      <td>0.0920801</td>\n",
       "      <td>0.71445</td>\n",
       "      <td>3.80073e-24</td>\n",
       "      <td>0.546198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>theoreticalProb</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">theoretical</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>37.5948</td>\n",
       "      <td>120.263</td>\n",
       "      <td>19.285</td>\n",
       "      <td>70.2618</td>\n",
       "      <td>85.4521</td>\n",
       "      <td>88.1311</td>\n",
       "      <td>68.2825</td>\n",
       "      <td>92.8292</td>\n",
       "      <td>46.2143</td>\n",
       "      <td>8.12103</td>\n",
       "      <td>65.1297</td>\n",
       "      <td>36.5352</td>\n",
       "      <td>67.1375</td>\n",
       "      <td>73.0749</td>\n",
       "      <td>117.332</td>\n",
       "      <td>124.538</td>\n",
       "      <td>20.5124</td>\n",
       "      <td>69.3125</td>\n",
       "      <td>12.9018</td>\n",
       "      <td>52.2276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>119</td>\n",
       "      <td>456</td>\n",
       "      <td>18</td>\n",
       "      <td>69</td>\n",
       "      <td>58</td>\n",
       "      <td>93</td>\n",
       "      <td>36</td>\n",
       "      <td>130</td>\n",
       "      <td>40</td>\n",
       "      <td>9</td>\n",
       "      <td>115</td>\n",
       "      <td>20</td>\n",
       "      <td>56</td>\n",
       "      <td>67</td>\n",
       "      <td>123</td>\n",
       "      <td>64</td>\n",
       "      <td>17</td>\n",
       "      <td>64</td>\n",
       "      <td>66</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.66236</td>\n",
       "      <td>1.92285</td>\n",
       "      <td>-0.09948</td>\n",
       "      <td>-0.0261442</td>\n",
       "      <td>-0.559063</td>\n",
       "      <td>0.077579</td>\n",
       "      <td>-0.92352</td>\n",
       "      <td>0.48586</td>\n",
       "      <td>-0.208338</td>\n",
       "      <td>0.148262</td>\n",
       "      <td>0.820246</td>\n",
       "      <td>-0.869288</td>\n",
       "      <td>-0.261692</td>\n",
       "      <td>-0.125214</td>\n",
       "      <td>0.0680657</td>\n",
       "      <td>-0.960442</td>\n",
       "      <td>-0.270962</td>\n",
       "      <td>-0.115043</td>\n",
       "      <td>2.35489</td>\n",
       "      <td>0.20015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>3.16533</td>\n",
       "      <td>3.7917</td>\n",
       "      <td>0.933369</td>\n",
       "      <td>0.982041</td>\n",
       "      <td>0.678743</td>\n",
       "      <td>1.05525</td>\n",
       "      <td>0.527221</td>\n",
       "      <td>1.40042</td>\n",
       "      <td>0.865534</td>\n",
       "      <td>1.10823</td>\n",
       "      <td>1.76571</td>\n",
       "      <td>0.547417</td>\n",
       "      <td>0.834109</td>\n",
       "      <td>0.916868</td>\n",
       "      <td>1.04831</td>\n",
       "      <td>0.513899</td>\n",
       "      <td>0.828767</td>\n",
       "      <td>0.923355</td>\n",
       "      <td>5.11555</td>\n",
       "      <td>1.14882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>2.92763e-40</td>\n",
       "      <td>2.08833e-206</td>\n",
       "      <td>0.769773</td>\n",
       "      <td>0.880248</td>\n",
       "      <td>0.0029524</td>\n",
       "      <td>0.603645</td>\n",
       "      <td>9.23674e-05</td>\n",
       "      <td>0.000112408</td>\n",
       "      <td>0.3604</td>\n",
       "      <td>0.757727</td>\n",
       "      <td>6.24411e-10</td>\n",
       "      <td>0.00620463</td>\n",
       "      <td>0.173727</td>\n",
       "      <td>0.476932</td>\n",
       "      <td>0.600274</td>\n",
       "      <td>5.55557e-08</td>\n",
       "      <td>0.43792</td>\n",
       "      <td>0.523074</td>\n",
       "      <td>1.833e-49</td>\n",
       "      <td>0.281865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>theoreticalProb</th>\n",
       "      <td>0.000869243</td>\n",
       "      <td>0.00278064</td>\n",
       "      <td>0.000445895</td>\n",
       "      <td>0.00162455</td>\n",
       "      <td>0.00197577</td>\n",
       "      <td>0.00203771</td>\n",
       "      <td>0.00157879</td>\n",
       "      <td>0.00214634</td>\n",
       "      <td>0.00106854</td>\n",
       "      <td>0.000187769</td>\n",
       "      <td>0.00150589</td>\n",
       "      <td>0.000844745</td>\n",
       "      <td>0.00155231</td>\n",
       "      <td>0.00168959</td>\n",
       "      <td>0.00271287</td>\n",
       "      <td>0.00287949</td>\n",
       "      <td>0.000474275</td>\n",
       "      <td>0.0016026</td>\n",
       "      <td>0.000298308</td>\n",
       "      <td>0.00120757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">K</th>\n",
       "      <th rowspan=\"6\" valign=\"top\">contingency</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>349</td>\n",
       "      <td>2213</td>\n",
       "      <td>87</td>\n",
       "      <td>284</td>\n",
       "      <td>407</td>\n",
       "      <td>401</td>\n",
       "      <td>105</td>\n",
       "      <td>697</td>\n",
       "      <td>234</td>\n",
       "      <td>23</td>\n",
       "      <td>288</td>\n",
       "      <td>55</td>\n",
       "      <td>229</td>\n",
       "      <td>170</td>\n",
       "      <td>479</td>\n",
       "      <td>353</td>\n",
       "      <td>64</td>\n",
       "      <td>247</td>\n",
       "      <td>70</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.5464</td>\n",
       "      <td>2.57242</td>\n",
       "      <td>0.514118</td>\n",
       "      <td>0.206284</td>\n",
       "      <td>0.279758</td>\n",
       "      <td>0.768041</td>\n",
       "      <td>-1.25014</td>\n",
       "      <td>1.01527</td>\n",
       "      <td>0.579097</td>\n",
       "      <td>0.34705</td>\n",
       "      <td>0.816928</td>\n",
       "      <td>-1.03491</td>\n",
       "      <td>0.180619</td>\n",
       "      <td>-0.417392</td>\n",
       "      <td>0.294198</td>\n",
       "      <td>-0.0463664</td>\n",
       "      <td>-0.0342786</td>\n",
       "      <td>0.531734</td>\n",
       "      <td>0.739924</td>\n",
       "      <td>0.30973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>2.92087</td>\n",
       "      <td>5.94806</td>\n",
       "      <td>1.42812</td>\n",
       "      <td>1.15371</td>\n",
       "      <td>1.21399</td>\n",
       "      <td>1.70296</td>\n",
       "      <td>0.420408</td>\n",
       "      <td>2.02128</td>\n",
       "      <td>1.49391</td>\n",
       "      <td>1.27196</td>\n",
       "      <td>1.76165</td>\n",
       "      <td>0.488048</td>\n",
       "      <td>1.13337</td>\n",
       "      <td>0.748777</td>\n",
       "      <td>1.2262</td>\n",
       "      <td>0.968372</td>\n",
       "      <td>0.97652</td>\n",
       "      <td>1.44567</td>\n",
       "      <td>1.67009</td>\n",
       "      <td>1.23948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>2.06005e-64</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00161068</td>\n",
       "      <td>0.0178841</td>\n",
       "      <td>0.000152264</td>\n",
       "      <td>1.81989e-22</td>\n",
       "      <td>1.0073e-24</td>\n",
       "      <td>2.59129e-61</td>\n",
       "      <td>9.28166e-09</td>\n",
       "      <td>0.237806</td>\n",
       "      <td>2.06069e-18</td>\n",
       "      <td>2.78607e-09</td>\n",
       "      <td>0.0612804</td>\n",
       "      <td>9.34601e-05</td>\n",
       "      <td>1.7115e-05</td>\n",
       "      <td>0.579957</td>\n",
       "      <td>0.901485</td>\n",
       "      <td>4.82996e-08</td>\n",
       "      <td>7.18185e-05</td>\n",
       "      <td>0.00209735</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "aa1                                                        R             K  \\\n",
       "terminus aa0 statistical_test observable                                     \n",
       "C        R   contingency      BH_multiple_tests         True          True   \n",
       "                              NexpectedSubseq            NaN           NaN   \n",
       "                              count                      119           456   \n",
       "                              log2OddsRatio          1.31738       2.01748   \n",
       "                              oddsRatio              2.49214       4.04876   \n",
       "                              pvalue             5.80119e-18  8.69599e-129   \n",
       "                              theoreticalProb            NaN           NaN   \n",
       "             theoretical      BH_multiple_tests         True          True   \n",
       "                              NexpectedSubseq        37.5948       120.263   \n",
       "                              count                      119           456   \n",
       "                              log2OddsRatio          1.66236       1.92285   \n",
       "                              oddsRatio              3.16533        3.7917   \n",
       "                              pvalue             2.92763e-40  2.08833e-206   \n",
       "                              theoreticalProb    0.000869243    0.00278064   \n",
       "         K   contingency      BH_multiple_tests         True          True   \n",
       "                              NexpectedSubseq            NaN           NaN   \n",
       "                              count                      349          2213   \n",
       "                              log2OddsRatio           1.5464       2.57242   \n",
       "                              oddsRatio              2.92087       5.94806   \n",
       "                              pvalue             2.06005e-64             0   \n",
       "\n",
       "aa1                                                        H           D  \\\n",
       "terminus aa0 statistical_test observable                                   \n",
       "C        R   contingency      BH_multiple_tests        False       False   \n",
       "                              NexpectedSubseq            NaN         NaN   \n",
       "                              count                       18          69   \n",
       "                              log2OddsRatio        -0.208737  -0.0260725   \n",
       "                              oddsRatio             0.865294     0.98209   \n",
       "                              pvalue                0.659258    0.952277   \n",
       "                              theoreticalProb            NaN         NaN   \n",
       "             theoretical      BH_multiple_tests        False       False   \n",
       "                              NexpectedSubseq         19.285     70.2618   \n",
       "                              count                       18          69   \n",
       "                              log2OddsRatio         -0.09948  -0.0261442   \n",
       "                              oddsRatio             0.933369    0.982041   \n",
       "                              pvalue                0.769773    0.880248   \n",
       "                              theoreticalProb    0.000445895  0.00162455   \n",
       "         K   contingency      BH_multiple_tests         True        True   \n",
       "                              NexpectedSubseq            NaN         NaN   \n",
       "                              count                       87         284   \n",
       "                              log2OddsRatio         0.514118    0.206284   \n",
       "                              oddsRatio              1.42812     1.15371   \n",
       "                              pvalue              0.00161068   0.0178841   \n",
       "\n",
       "aa1                                                        E            S  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        R   contingency      BH_multiple_tests         True        False   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       58           93   \n",
       "                              log2OddsRatio        -0.634304     0.282632   \n",
       "                              oddsRatio             0.644251      1.21641   \n",
       "                              pvalue             0.000464047     0.066274   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests         True        False   \n",
       "                              NexpectedSubseq        85.4521      88.1311   \n",
       "                              count                       58           93   \n",
       "                              log2OddsRatio        -0.559063     0.077579   \n",
       "                              oddsRatio             0.678743      1.05525   \n",
       "                              pvalue               0.0029524     0.603645   \n",
       "                              theoreticalProb     0.00197577   0.00203771   \n",
       "         K   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                      407          401   \n",
       "                              log2OddsRatio         0.279758     0.768041   \n",
       "                              oddsRatio              1.21399      1.70296   \n",
       "                              pvalue             0.000152264  1.81989e-22   \n",
       "\n",
       "aa1                                                        T            N  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        R   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       36          130   \n",
       "                              log2OddsRatio        -0.853186     0.592484   \n",
       "                              oddsRatio             0.553561      1.50784   \n",
       "                              pvalue             0.000138077  1.15786e-05   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq        68.2825      92.8292   \n",
       "                              count                       36          130   \n",
       "                              log2OddsRatio         -0.92352      0.48586   \n",
       "                              oddsRatio             0.527221      1.40042   \n",
       "                              pvalue             9.23674e-05  0.000112408   \n",
       "                              theoreticalProb     0.00157879   0.00214634   \n",
       "         K   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                      105          697   \n",
       "                              log2OddsRatio         -1.25014      1.01527   \n",
       "                              oddsRatio             0.420408      2.02128   \n",
       "                              pvalue              1.0073e-24  2.59129e-61   \n",
       "\n",
       "aa1                                                        Q            C  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        R   contingency      BH_multiple_tests        False        False   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       40            9   \n",
       "                              log2OddsRatio         -0.26928     0.229385   \n",
       "                              oddsRatio             0.829733      1.17233   \n",
       "                              pvalue                0.277891     0.584709   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests        False        False   \n",
       "                              NexpectedSubseq        46.2143      8.12103   \n",
       "                              count                       40            9   \n",
       "                              log2OddsRatio        -0.208338     0.148262   \n",
       "                              oddsRatio             0.865534      1.10823   \n",
       "                              pvalue                  0.3604     0.757727   \n",
       "                              theoreticalProb     0.00106854  0.000187769   \n",
       "         K   contingency      BH_multiple_tests         True        False   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                      234           23   \n",
       "                              log2OddsRatio         0.579097      0.34705   \n",
       "                              oddsRatio              1.49391      1.27196   \n",
       "                              pvalue             9.28166e-09     0.237806   \n",
       "\n",
       "aa1                                                        G            P  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        R   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                      115           20   \n",
       "                              log2OddsRatio         0.805524    -0.879883   \n",
       "                              oddsRatio              1.74778     0.543411   \n",
       "                              pvalue             4.46075e-08    0.0036786   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq        65.1297      36.5352   \n",
       "                              count                      115           20   \n",
       "                              log2OddsRatio         0.820246    -0.869288   \n",
       "                              oddsRatio              1.76571     0.547417   \n",
       "                              pvalue             6.24411e-10   0.00620463   \n",
       "                              theoreticalProb     0.00150589  0.000844745   \n",
       "         K   contingency      BH_multiple_tests         True         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                      288           55   \n",
       "                              log2OddsRatio         0.816928     -1.03491   \n",
       "                              oddsRatio              1.76165     0.488048   \n",
       "                              pvalue             2.06069e-18  2.78607e-09   \n",
       "\n",
       "aa1                                                       A            V  \\\n",
       "terminus aa0 statistical_test observable                                   \n",
       "C        R   contingency      BH_multiple_tests       False        False   \n",
       "                              NexpectedSubseq           NaN          NaN   \n",
       "                              count                      56           67   \n",
       "                              log2OddsRatio        -0.28537    -0.192589   \n",
       "                              oddsRatio            0.820531     0.875034   \n",
       "                              pvalue               0.145126     0.301973   \n",
       "                              theoreticalProb           NaN          NaN   \n",
       "             theoretical      BH_multiple_tests       False        False   \n",
       "                              NexpectedSubseq       67.1375      73.0749   \n",
       "                              count                      56           67   \n",
       "                              log2OddsRatio       -0.261692    -0.125214   \n",
       "                              oddsRatio            0.834109     0.916868   \n",
       "                              pvalue               0.173727     0.476932   \n",
       "                              theoreticalProb    0.00155231   0.00168959   \n",
       "         K   contingency      BH_multiple_tests       False         True   \n",
       "                              NexpectedSubseq           NaN          NaN   \n",
       "                              count                     229          170   \n",
       "                              log2OddsRatio        0.180619    -0.417392   \n",
       "                              oddsRatio             1.13337     0.748777   \n",
       "                              pvalue              0.0612804  9.34601e-05   \n",
       "\n",
       "aa1                                                       I            L  \\\n",
       "terminus aa0 statistical_test observable                                   \n",
       "C        R   contingency      BH_multiple_tests       False         True   \n",
       "                              NexpectedSubseq           NaN          NaN   \n",
       "                              count                     123           64   \n",
       "                              log2OddsRatio       0.0607836    -0.919574   \n",
       "                              oddsRatio             1.04303     0.528665   \n",
       "                              pvalue                  0.644  2.21827e-08   \n",
       "                              theoreticalProb           NaN          NaN   \n",
       "             theoretical      BH_multiple_tests       False         True   \n",
       "                              NexpectedSubseq       117.332      124.538   \n",
       "                              count                     123           64   \n",
       "                              log2OddsRatio       0.0680657    -0.960442   \n",
       "                              oddsRatio             1.04831     0.513899   \n",
       "                              pvalue               0.600274  5.55557e-08   \n",
       "                              theoreticalProb    0.00271287   0.00287949   \n",
       "         K   contingency      BH_multiple_tests        True        False   \n",
       "                              NexpectedSubseq           NaN          NaN   \n",
       "                              count                     479          353   \n",
       "                              log2OddsRatio        0.294198   -0.0463664   \n",
       "                              oddsRatio              1.2262     0.968372   \n",
       "                              pvalue             1.7115e-05     0.579957   \n",
       "\n",
       "aa1                                                        M            F  \\\n",
       "terminus aa0 statistical_test observable                                    \n",
       "C        R   contingency      BH_multiple_tests        False        False   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       17           64   \n",
       "                              log2OddsRatio        -0.598269   -0.0780036   \n",
       "                              oddsRatio             0.660546     0.947368   \n",
       "                              pvalue               0.0920801      0.71445   \n",
       "                              theoreticalProb            NaN          NaN   \n",
       "             theoretical      BH_multiple_tests        False        False   \n",
       "                              NexpectedSubseq        20.5124      69.3125   \n",
       "                              count                       17           64   \n",
       "                              log2OddsRatio        -0.270962    -0.115043   \n",
       "                              oddsRatio             0.828767     0.923355   \n",
       "                              pvalue                 0.43792     0.523074   \n",
       "                              theoreticalProb    0.000474275    0.0016026   \n",
       "         K   contingency      BH_multiple_tests        False         True   \n",
       "                              NexpectedSubseq            NaN          NaN   \n",
       "                              count                       64          247   \n",
       "                              log2OddsRatio       -0.0342786     0.531734   \n",
       "                              oddsRatio              0.97652      1.44567   \n",
       "                              pvalue                0.901485  4.82996e-08   \n",
       "\n",
       "aa1                                                        W           Y  \n",
       "terminus aa0 statistical_test observable                                  \n",
       "C        R   contingency      BH_multiple_tests         True       False  \n",
       "                              NexpectedSubseq            NaN         NaN  \n",
       "                              count                       66          60  \n",
       "                              log2OddsRatio          2.26959    0.103653  \n",
       "                              oddsRatio              4.82186     1.07449  \n",
       "                              pvalue             3.80073e-24    0.546198  \n",
       "                              theoreticalProb            NaN         NaN  \n",
       "             theoretical      BH_multiple_tests         True       False  \n",
       "                              NexpectedSubseq        12.9018     52.2276  \n",
       "                              count                       66          60  \n",
       "                              log2OddsRatio          2.35489     0.20015  \n",
       "                              oddsRatio              5.11555     1.14882  \n",
       "                              pvalue               1.833e-49    0.281865  \n",
       "                              theoreticalProb    0.000298308  0.00120757  \n",
       "         K   contingency      BH_multiple_tests         True        True  \n",
       "                              NexpectedSubseq            NaN         NaN  \n",
       "                              count                       70         222  \n",
       "                              log2OddsRatio         0.739924     0.30973  \n",
       "                              oddsRatio              1.67009     1.23948  \n",
       "                              pvalue             7.18185e-05  0.00209735  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop the rare amino acid U and ambiguous letter X\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApairT.T.copy()\n",
    "oddsRatioDfTableAApair2.columns.name = 'aa1'\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair2[~oddsRatioDfTableAApair2.index.get_level_values('aa0')\\\n",
    "                                                  .str.contains(r'[XU]')]\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair2.loc[:, ~oddsRatioDfTableAApair2.columns.str.contains(r'[XUB]')]\n",
    "# Sort the aa columns\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair2[aaTable]\n",
    "\n",
    "# Sort the aa rows\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair2.copy()\n",
    "oddsRatioDfTableAApair3['sortby'] = oddsRatioDfTableAApair3.index.get_level_values(1).map(dict(zip(aaTable, range(len(aaTable)))))\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.reset_index(level=1)\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.dropna(subset=['sortby'])\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.set_index('sortby', append=True)\\\n",
    "                                                 .reorder_levels(['terminus', 'sortby', 'statistical_test', 'observable'])\\\n",
    "                                                 .sort_index()\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.reset_index(level='sortby').set_index('aa0', append=True)\\\n",
    "                                                 .reorder_levels(['terminus', 'aa0', 'statistical_test', 'observable'])\\\n",
    "                                                 .drop('sortby', axis=1)\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair3.copy()\n",
    "# print(oddsRatioDfTableAApair3.index)\n",
    "oddsRatioDfTableAApair3.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfHexamer, oddsRatioDfTableHexamer = compute_odds_ratio_subseq(bulkFreqHexamer, None,\n",
    "                                                                        ctermFreqHexamer, cTerminalSize,\n",
    "                                                                        ntermFreqHexamer, nTerminalSize,\n",
    "                                                                        subseqSize=6, seq_type='DNA',\n",
    "                                                                        computeTheoreticalFreq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfHexamer.\" + dfFormatOutput)\n",
    "oddsRatioDfHexamer.to_csv(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioDfHexamer.\" + dfFormatOutput)\n",
    "oddsRatioDfHexamer = pd.read_csv(filename, index_col=[0,1,2,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfHexamer.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfTableHexamer = oddsRatioDfHexamer.copy()\n",
    "# Filter out hexamers that contains ambiguous DNA letters\n",
    "oddsRatioDfTableHexamer = oddsRatioDfTableHexamer[~oddsRatioDfTableHexamer.index.get_level_values('subseq').str.contains(r'[^ATGC]')]\n",
    "\n",
    "oddsRatioDfTableHexamer['codon_-2'] = oddsRatioDfTableHexamer.index.get_level_values(level=1).map(lambda x: x[:3])\n",
    "oddsRatioDfTableHexamer['codon_-1'] = oddsRatioDfTableHexamer.index.get_level_values(level=1).map(lambda x: x[3:])\n",
    "oddsRatioDfTableHexamer = oddsRatioDfTableHexamer.set_index(['codon_-1', 'codon_-2'], append=True)\n",
    "oddsRatioDfTableHexamer.index = oddsRatioDfTableHexamer.index.droplevel(1)\n",
    "oddsRatioDfTableHexamer = oddsRatioDfTableHexamer.reorder_levels([3, 4, 0, 1, 2])\n",
    "oddsRatioDfTableHexamer = oddsRatioDfTableHexamer.unstack('codon_-1')\n",
    "# Remove the \"value\" empty level\n",
    "oddsRatioDfTableHexamer.columns = oddsRatioDfTableHexamer.columns.droplevel(0)\n",
    "\n",
    "# Convert to numeric values\n",
    "df = oddsRatioDfTableHexamer.T.copy()\n",
    "for col in df.columns:\n",
    "    df[col] = pd.to_numeric(df[col], errors='ignore')\n",
    "oddsRatioDfTableHexamerT = df\n",
    "print(oddsRatioDfTableHexamerT.dtypes.head())\n",
    "oddsRatioDfTableHexamer = oddsRatioDfTableHexamerT.T.copy()\n",
    "oddsRatioDfTableHexamer.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "type(oddsRatioDfTableAApair2.loc[('C', 'W', 'contingency', 'count'), 'A'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "type(oddsRatioDfTableAApair2.loc[('C', 'W', 'contingency', 'BH_multiple_tests'), 'A'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "type(oddsRatioDfTableAApair2.loc[('C', 'W', 'contingency', 'log2OddsRatio'), 'A'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plotData = compute_oddsratio_plot_data(oddsRatioDf, terminus)\n",
    "maskData = compute_oddsratio_mask_data(oddsRatioDf, pvalueThresholds, terminus)\n",
    "pvalAnnotTable = compute_pvalueAnnotation(oddsRatioDf, pvalueThresholds, terminus)\n",
    "maskDataSmallN = compute_smallN_mask_data(oddsRatioDf, pvalueThresholds, terminus)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bacteria kingdom."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>30</td>\n",
       "      <td>132</td>\n",
       "      <td>19</td>\n",
       "      <td>34</td>\n",
       "      <td>33</td>\n",
       "      <td>58</td>\n",
       "      <td>11</td>\n",
       "      <td>64</td>\n",
       "      <td>35</td>\n",
       "      <td>6</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>31</td>\n",
       "      <td>29</td>\n",
       "      <td>36</td>\n",
       "      <td>51</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>13</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.44127</td>\n",
       "      <td>1.47675</td>\n",
       "      <td>0.66263</td>\n",
       "      <td>0.0364875</td>\n",
       "      <td>-0.227941</td>\n",
       "      <td>0.503797</td>\n",
       "      <td>-1.9068</td>\n",
       "      <td>0.687241</td>\n",
       "      <td>-0.0300706</td>\n",
       "      <td>0.284783</td>\n",
       "      <td>-0.847342</td>\n",
       "      <td>-1.77249</td>\n",
       "      <td>-0.563203</td>\n",
       "      <td>-0.624863</td>\n",
       "      <td>-0.316902</td>\n",
       "      <td>-0.46533</td>\n",
       "      <td>-1.15427</td>\n",
       "      <td>-0.601396</td>\n",
       "      <td>0.818716</td>\n",
       "      <td>-0.0353682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.3578</td>\n",
       "      <td>2.78321</td>\n",
       "      <td>1.58297</td>\n",
       "      <td>1.02561</td>\n",
       "      <td>0.853853</td>\n",
       "      <td>1.41794</td>\n",
       "      <td>0.266683</td>\n",
       "      <td>1.6102</td>\n",
       "      <td>0.979372</td>\n",
       "      <td>1.21823</td>\n",
       "      <td>0.555808</td>\n",
       "      <td>0.292703</td>\n",
       "      <td>0.676798</td>\n",
       "      <td>0.648481</td>\n",
       "      <td>0.802792</td>\n",
       "      <td>0.724305</td>\n",
       "      <td>0.449294</td>\n",
       "      <td>0.659116</td>\n",
       "      <td>1.76384</td>\n",
       "      <td>0.975783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.105358</td>\n",
       "      <td>3.37292e-21</td>\n",
       "      <td>0.0573287</td>\n",
       "      <td>0.858281</td>\n",
       "      <td>0.45219</td>\n",
       "      <td>0.0162306</td>\n",
       "      <td>9.44371e-08</td>\n",
       "      <td>0.000673155</td>\n",
       "      <td>1</td>\n",
       "      <td>0.645399</td>\n",
       "      <td>0.00496324</td>\n",
       "      <td>0.000120532</td>\n",
       "      <td>0.0299081</td>\n",
       "      <td>0.0185353</td>\n",
       "      <td>0.212641</td>\n",
       "      <td>0.0250106</td>\n",
       "      <td>0.12341</td>\n",
       "      <td>0.0417409</td>\n",
       "      <td>0.0601406</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          R            K          H          D         E  \\\n",
       "observable                                                                 \n",
       "BH_multiple_tests     False         True      False      False     False   \n",
       "count                    30          132         19         34        33   \n",
       "log2OddsRatio       0.44127      1.47675    0.66263  0.0364875 -0.227941   \n",
       "oddsRatio            1.3578      2.78321    1.58297    1.02561  0.853853   \n",
       "pvalue             0.105358  3.37292e-21  0.0573287   0.858281   0.45219   \n",
       "\n",
       "                           S            T            N          Q         C  \\\n",
       "observable                                                                    \n",
       "BH_multiple_tests      False         True         True      False     False   \n",
       "count                     58           11           64         35         6   \n",
       "log2OddsRatio       0.503797      -1.9068     0.687241 -0.0300706  0.284783   \n",
       "oddsRatio            1.41794     0.266683       1.6102   0.979372   1.21823   \n",
       "pvalue             0.0162306  9.44371e-08  0.000673155          1  0.645399   \n",
       "\n",
       "                            G            P          A          V         I  \\\n",
       "observable                                                                   \n",
       "BH_multiple_tests       False         True      False      False     False   \n",
       "count                      21            7         31         29        36   \n",
       "log2OddsRatio       -0.847342     -1.77249  -0.563203  -0.624863 -0.316902   \n",
       "oddsRatio            0.555808     0.292703   0.676798   0.648481  0.802792   \n",
       "pvalue             0.00496324  0.000120532  0.0299081  0.0185353  0.212641   \n",
       "\n",
       "                           L         M          F          W          Y  \n",
       "observable                                                               \n",
       "BH_multiple_tests      False     False      False      False      False  \n",
       "count                     51         4         25         13         21  \n",
       "log2OddsRatio       -0.46533  -1.15427  -0.601396   0.818716 -0.0353682  \n",
       "oddsRatio           0.724305  0.449294   0.659116    1.76384   0.975783  \n",
       "pvalue             0.0250106   0.12341  0.0417409  0.0601406          1  "
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDf.loc['C'].loc[-1][aaTable]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tenericutes phylum."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oddsRatioDf.loc['C'].loc[-1]['P']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oddsRatioDf.loc['C'].loc[-1]['T']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mycoplasma genus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oddsRatioDf.loc['C'].loc[-1]['P']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oddsRatioDf.loc['C'].loc[-1]['T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(cterminal)\n",
    "from cterminal import plot_aa_composition_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.0"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FontProperties(size='medium').get_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5.8309999999999995"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FontProperties(size='small').get_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Analysis_data/C-terminal/Data/Taxonomy/taxaId_2104/taxaId_2104_subset_all_composition_bias_aa_Cterminal\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAEHCAYAAAB1Kw7rAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAH1RJREFUeJztnXl0FFXa/7/V6XSnO4nZIEzCElZhRkbmiMK8ID/lBT0jMy8CyhLQDMw5KiPCCyLIotFxZsBxiaMsDjJhcgSCCDpHHXlHARGQRVkDGgEBQxZCiB2ydDrp7uqq3x+xanqpul3dXZ104PmcwwFu3efeW/U8feupuzyXE0VRBEGoYOjoBhCxDRkIwYQMhGBCBkIwIQMhmJCBEEzIQAgmZCAEE2M4Qi6XC06nEwYD2VdnQBAEmM1mmEymkGXD0rDT6YTL5QpHlOgApB90OATtQVatWoXk5GScP38eS5cuRWJiIgwGAxISElBVVeWT9+abb8a5c+cC0gAopmvJK6U5P3vbJ6/5v/OiVlco8q4vtvqkm+6cGrW6ws3bvXt3+KOkVyU09SC5ubkYMmQILl++rCU70UnQoldNBlJUVISLFy9iwIABujWO6Hg06VUMwhtvvCG2traKr776qlhbWyuKoija7XbRbrcHEyViBCV9KelVCU4UQ5/ub25uBgDV9xYRW0Sir7A+cyWUHCS+5FPfCobcq5o31py5cOQ9Z/b7pMcNGqWa1330X3Ja/O2/aZMv/dxX/md3w/3VBz5p8cPuV22Xa98WnzTT/8sFANS8NE9OS5qzEuES1AdZtWoVHA4HVq5cCbvdHnZFRGyhVa+anNSCggI89NBDSEpK0q2BRMejRa+aDKR3797Ys2ePbg0jYgNNeg3mAUvebmFhofjNN9+oesVE7ML6ivHWqxJhf8XwPA+e5wEABoMBJpMJBoMBbrcbbrdbzmuxWGAwGODxeOByuSAIAoA2j1ryrr3lOY6T01nyFosFPM/D7XYz5aW6lOQNBgOam5uDylutVoiiCKfTyZR3uVywWCyK8nFxcWhubg4qbzab5ftiyXs/Q395qVwJjuNgNpvD+orR5KQ6nU75bwme52Gz2eBwOJCcnIy6ujpUVFRAFEXYbDafPxzHoa6uDrW1tXKa2+1WlJfSg8lXVlZCFMWg8g6HQ1Xe7XZrko+Pj4fH4wkqX1tbqyrvdrs1yUv3FUxeeoZK8t5tcDgcMBoDP1bV9BqygbCwWCzIyspCdXU1WlpaVK/zPM+8Hq68wWBoF/nq6mqfX25nk//hhx8CrmlFk4Fs2rQJJ0+e9EnjOK5DH47FYoHJZIpI3mg0apJXu95Z5NV6CCW9+qNpoOyhhx6Cw+HwSYuLi1NsnMFgCLi51NTUsG9OTT4rKwtOp7Nd5KX7ikQ+lPqNRmNE8sF+fBJKevUnbCfV5XIpvtvi4+PBcRx4npd/uXFxcfB4PHLjjUYjRFH0ccS85SWnVGqawWCQy+I4Ti6P47iA3sFgMGiWByA72qz2e9+D1H6e58FxnFxOMHnpnoPVL5UvCAK8VSOVGax+JfmWlhYkJCS071C7IAiorq72SbNYLMjMzMTVq1d9LFeyZu9fVmpqqqJ8t27dUFZWJqdFc6jd4XD4tEGt/QDQrVs3NDY2+vwypX9rke/RowcqKyuD1p+VlYWmpibU1tYG5K2vrw9af1ZWFtxuNyoqKuR0pfUgWtG8YKipqQlz585VzSc1rqWlJaJuX6lXaQ8i7baDyXMcp6n+6upqWK1WRXmtPltGRgazLkC7XjUvGMrJycGZM2cUr+vpkIbxxouY9nCI4+LiNNWv5jOo+Xxa5ZUIplcgxM9cpV9BtL31aKPVoXS5XBE5pP6+glZ5CY/HE7Xnx+rdgjqpq1atQlJSEhoaGjB//nwA/xlJNZlMssPFcggFQVB0JkVRDJA3Go1hrb4OB2nhdXs4xHFxcQH3pVQ/oPwMXC6XokPrX7+SvNJ6ECW9KhHRgqH2XLQcq+tBOkNdkpMala8YyZkxGo2YNGmSogNFdD606lXTZ25ubi7MZrOuDSQ6Hi161eSkbtu2De+9954ujSJiBy161eSkSl3Rvffei8zMTFq03MlQc1L99aoErWq/AYjaqvYdO3bg2rVruHTpEhYvXhww93K9efvRruuJ907JaasfuLXd7ktpqD2YbiWYPkhjYyNcLhdmzpypWgDROdGqW6aBTJo0Cffddx/WrVtH+3KvM7TqlumDFBcXA2gbEJs9ezaSk5MBkA/S2VDSl5pu/SEn9QYgprZedjbHsbPXNXub75LBv03+BQDg7tf2yWkfPzoU4cL0QT799FM88sgjKCoq8lnsQnR+tOqW2YPce++9OHv2LA21X4do1S1FoSPYBNu2J23RC7aVj4hd1PSlpFt/aD3IdVqXd9yR1pw7AERxPUh2djZcLhdGjRqFHj16hFwJEXto1asmH4RGUa9PtOhVk4EMHz4chw8fjrhBRGyhSa/BHBzJkZk2bZpYUVHBdHqI2IQVH8Rbr0rQUPsNAA21U13MvFHbevnZZ5/h/PnzcDgczL0TROdDq26ZTuqJEyfw6KOPIiMjAzU1Nbo3kug4tOqWaSDSDjFBEJh7S4nOh1bdMp3UvXv34uTJk+A4DvPm/Sdyr7T1Mj4+PkBG63ZEpe2YocjrsR1Sq7x3GdGW976ul7xSfBA13frD9EHuuusupKWl4V//+hcaGhqQkpIiX+N5PqBrkuJj1NTUhB0fRKt8S0sLunXr5tOGUOXV4nv4ywOhxQdRks/KytIc36OlpSXg2YQi73Q6faZBPB4PevXq5VMeS7feaBooS0pKkkM3q6HXLvhoBby70QPm+XPq1Cl8/vnnEARB8U0gEfQzd/fu3Zg/fz7cbrfqrnvvxt10000B10O5Of/AMrEgH0p8EDX5UEJg+D/DSGO6Ka330KJXQGMPIv4YT0zpnLpYiuYXLfnO3jN26dIlIB1g61UiqIHcd999WLt2Ld58882AgrRGA4zWw9EaDVCPbl+JaL02pGcYzTipLL16E1EobovFcl15+2ryoij6fAlFU14qQ/QLrhOJvMPhiF6UQ8mZqa+vx5IlS+T3ldFo9IlGCHTOIem/fnFRTpt/Z9+o1hVLQ+1qevUnqIHs2rULCxYswLVr1+gg5esIrXoNqnGO41BbW4sNGzaovouJzodWvQb1Qb799lvs3r0bra2tyM3NRffu3Wm6v5OhpC8lvSqh+3oQpWh8oUQujCSaX6jykTq0WusP1aHVO9JjTK0HaWpqgs1mk9Ok8NpaHSye5312erGGrzMyMnzqClU+0uHvHj16BDjq/iG31Ya/1epXuy8gNCf1f946JKe9M+NWhIumVe1JSUm4ePEili5dGlKUw0jDa2sJjx2pvNZxDKUR4lDCa4czwhtNtOpV02fJ9OnT8fOf/zwkJzXS8Npaw2NHIh/pCGsk4bWD1d8eaNGrJgN59913wfM8+vXrp6niSMNDhzL8HYl8pCO84YbX1jLC2x5o0asmH2TKlCmKEz7S+84bq9WKlJSUgIejlFcpLdRTDfzLCEU+lG5fafhbqX5vPybcib3ExESkpaUFXNP6DAHgo0f/S/63/yGLEmp69Ub3rZf19fUQRdHndMdQRl3dbjfsdrum0yGbm5tx7do1pjyg/XRItdMlLZa2EzY9Ho+PvNJ99enTJ6B+6R5aWlqY9RsMBiQmJqKxsRF1dXU+9wXEaCjuxYsXY8qUKbj99ts1FXj16lWf/6tZuBrx8fEBh+kAQHl5eUC5SnEt/OWBtq8CJXn/cqVzVkKZQtBSv1JdaveVlpbmYxzRRItumS+7mpoa3HHHHTh+/LjujSM6Fq26ZfYgn3zyCXiex9mzZ+FyudrtmA4i+mjWLWvL3ssvvyyKoiiePHlS/PTTT5lb+YjYRUlfarr1h7Ze3gDE1FB7Z143cb3WFcnWS6aTunLlSgDAvn37sHfv3rArIWIPrbplGkiPHj1QXV2NI0eOYOTIkfq2kOhQtOqWaSDjxo3Dxx9/DJPJRMH8rzM06zaYB7xw4UKxpKQkqFdMxC5q+lLSrT8U5fAGqCuSofag04bfffcdHnnkEfl0AOL6QKtegxrIgAEDcOutt+KBBx7QrXFEx6NVr7SPgWAS1Af57rvv8NJLL2H06NGYPn06ABpJ7Wwo6UtJr0rQUPsNQFSH2t9//30MGTJEcVna9ebtX691KQ21s/TqDfkgBBMyEIJJUB/k/fffR3l5Obp06YLJkyfLa0EB8kE6C0r6UtKrEhQfhOKDMIloBu56iwaoJi/dlxRRMRx57zaw5KW8FoslbPnu3bvDbDbL8kpRDrUS1AfJz8/HoUOH8Pjjj/usIvd4PNdlNMCOlg8lYF44OwYl1PQacD+qV36ka9euOHr0aEA3pvRmivbNdYaAd7EcMM8bNb36o+kV07NnT6SnpzPzeHf74YRqDLbRWuvDiZa81p4hWhvFIw3lqYQWvWr6zJ0wYQIsFovq9eslGmBHdfv+xuu/NzfSnkntCyWYXoEIv2J4ng/YzsjzvE+4B7XtjNHaDqm2nVFJPpTtkE6n02ezuJK8IAhyXf7ygiD4pKvJS/clPUM1eekZKsl734P0tWg0GqN36mVycjKMRiNyc3PlkAdGozEgVrvS0K/adsZobodU2s6oJO+fN1J5tfuSNmJrqUvpvpTk1epSKldpqF1Nr/5o8kHoaPbrEy161eSDbNq0CZs2bdKlUUTsoEmvwRa80tHsnR/WqZfBiMhJ9T9Ggoa/9ZFXOmwpEvmWlsADhbQS9lC7/4FCNPwd2vC3mrw0DuL/ARCKvP9hTVEban/ttdcAAFeuXMHmzZtV88X68HVnGwfRW17JEdWqW6aBGI1GFBUVobS0FH369FHM0xmGr2N5+Ls95JXOi9GiWyCIgcTHx8Nut+ODDz5QLKSjHw5NDIZ/Xkww3UowndQNGzagZ8+e2LRpE4qKiuTAsc3NzXC5XLBareB53qfxRqMRoij6OGwAAsJTx8fHg+O4AHlJRhr9k0Zr/ZvJcRwMBkPY9UvlC4KgyaGV2hpteemaFodU6f6lZyj+GNKb53k4HA5YrVYfJ1VNt/5E9BXT2NioKTy2w+FQDI999epVTetB/MNbS+Veu3YtIGCcUl618NhutxsVFRU+ZQLKi4NtNpt8X2ryUl4lee9nEI68Urt69uwZIA+orwfRfevl6tWrUV9fj6KiIly6dMnnGms9CItIu31vlLrdUMJjh/Hb6HB5b5Tkta4HYenWG6Y2f/vb32L16tUQRRE5OTmaGhfN8NjBsFgiC4+tpfyOlNdavpb1ICzdesM0kOTkZCQmJmLy5MmaGxet8Nha6w83PHYwIg28HywCdKQE+/H5o1W3QQfKjEZjwIipIAhwOp2yMsxmM1JSUlBeXo6UlBSf2cPm5mY4nU75utPpBAC0trYqyicnJwcoubW1NWBGUprO9ng8zPolef/6gbYRRv+2AoGzn9KyhGDyUl6lNJ7nUV1dHba8Ursk/8L7/p1OJxISEnyeoTS66o+Sbv0Jy0l1uVxwOp10hl0nQRAEmM3msOLchmUgxI0DdQEEEzIQggkZCMGEDIRgEvZ6EFEUcebMGQiCgEGDBiEuLg48z6O0tBQmkwkDBw4Ex3HySQJff/01evfujaSkJADA5cuXkZ2d7VNmc3MzrFYrSkpKMGDAAHlouLW1FSaTCSdPnsSgQYN8DuCz2WzywmaJuro6n/0eLpcLLpcLSUlJaGhoQEpKSkj3VVxcjHPnzqGpqQkjR47EpEmT8Prrr8urzxMTEzF37lwAwMMPP4xFixbh1lv/c9LkW2+9hczMTHzzzTdISUnBE088AQDYtm0b4uLicObMGWRkZOCxxx7D1q1bAQBfffUVhg0bhqlTp6K4uBitra2oqalBr169MGPGDABtC48bGhrwu9/9Drt27UJeXh6+//57n/thTcRpIewe5LXXXkNlZSWqqqqwZs0aAEBhYSEuXLiA7du3o7CwEADw4osvYuHChSgrK8OqVatk+aeffhqbNm2C3W6X09566y2sW7cOly9fxurVq+X0jRs34s0330RNTY1cV35+PgoKCrBo0SIUFBTIbXruueewfft2eb0D0BZ2ev369Th06BC2bNkCoO28tq1bt8p/WPfl8Xjw/PPPo0+fPvKEW3p6OtxuN5YsWYKEhARZfsSIEaisrMRf/vIX/O1vf2t7yAYDbDYbli9f7pP30qVLaG5uxrJly+QhA0EQkJycjN69e2PUqFFy3i5dumDp0qU+g2xWqxXLli1DcXExrly5AgBYs2YN9u7di5KSEpSUlGjSJYuwDcRsNuOOO+7A8OHD5UGY5uZmDBs2DElJSfJQfGpqKkwmE4YMGeIzKDNixAjcfffdKCoqwquvvgrgPzOUffr08flm//rrr2G32zFmzBg57cEHH0S/fv1w11134cknnwTQtj1g4sSJePTRR33mZG666SbMnz8f58+fx9mzZwG0DTJlZ2fjzjvvxJ133sm8r8TERPzxj39EfHy8vGstMzMTv/jFLzB69GifHiknJwfjxo1D//79kZeXBwC45ZZbUFNTg9GjR6Nv375y3sGDB8Nms2Ht2rVyubm5uUhNTcWmTZvkHjY1NRXjx49HcXGx3AMDbaOndXV1yM7Olk/6euGFFzB69GgIgoAJEyZo1qcqoSx+9aayslIsLCwU//GPf4jl5eWiKIpiaWmpuGHDBvHq1avi+fPnRVEUxbKyMrG8vFz85z//KZ48eVKW9/63RE1Njbhx40afMkVRFOvq6sQDBw6IdXV14qlTp+T0qqoqsaioSP6/dO3ChQvikSNH5PSjR4+KNptNFEVR3LJli5ze1NSkeF9///vfA9rgL8tKi1ZePeoKlbAHyqT336xZs7B79275/Sf+uPYBaHv/Se9EKV16Jyq9KyPNq/b+Zckr5ZUeiXcbblTCdlKtVivmzJmDgoICuStes2YNBg8eLHeXffr0UUyLVt5o1XVDE27Xs3nzZrG2tlbcvHmzuHDhQlEURbG5uVksKysT33vvPTmfUlq08karrhuZsJ3U6dOno0uXLjAYDHjllVcAtPUqOTk5PlP+SmnRyhutum5kaLKOYEIjqQST68JAnE4n6uvro1K2x+MJWBh9I9GpDaS2thbvvvsuduzYgdLSUmzfvj1gyyKLQ4cO4cSJE6ioqMDixYuxaNEiLFy40Ge08tixY/j8889DbltpaSn27NkTNN9TTz0Fnuexbt26kOtoD2LuILpVq1bJS+ymTp2K8+fPY8+ePXC5XFi2bBk2btyIpqYmJCQkYNq0aaiqqoLNZoPJZEJaWhpaW1uxevVqXLt2DaIoIj8/H5MmTcKYMWNQW1uL559/Xq5r586dyM/Px6JFi7B06VKkp6fj+PHjuHz5shzD/NixYzh58iSGDRuGDRs2QBAEjB07FqdOnUJVVRUeeOABvPHGGxg4cCBaW1uRmZkJo9GIoUOH4ocffkB+fj4yMzNRVlaGuXPn4sSJEzhx4gSuXLkinzxpNBqRnp6O0tJS/OxnP2v3Z84iJnuQadOmYc6cOdi6dSv+/e9/Y8WKFcjLy8OHH36Iqqoq9OrVC7/5zW/k/LfddhvGjh0LoG1i7+rVq3j22WeRlZWF0tJSZGdnY86cOT7zPvX19fL8B8/z8uTebbfd5hPgXir7nXfekdd+HjlyBEDbyvCsrCwMGjQIS5YsQWNjI+bNm4cLFy7I8i6XCzNnzsTEiRNx4sQJDBw4EL/61a+QnJyM8+fPy/kGDx6MAwcOROFpRkZMGogoinA6nQFbGARBQF5eHgYOHIgVK1bIA3Te8y6i10iuJCNNkPmf7igZiNlsRl1dHQDgiy++wIEDB7BgwQL8+c9/lssSBAFTp07F3LlzMWTIEACQZ5WtViuMRqM8f+S/NychIUHeRbdhwwbwPI/+/fsH7MiLxTW+MfeKAdpmdQHg8ccfx7fffov8/Hx4PB4sXLgQhYWFcDgc6Nevn/xAs7Oz8c4776Bv377yZuUXX3wRHMepdtmpqalobW0FAMyePRvPPvusvKL+hRdekM+SLSsrw9q1a5Gfn4+//vWvMBgMmDZtWtj3lpKSgn379qGiogL9+/eX00tKSmLybOKYGwdZtWoVxo8fz9zMoxdHjx4Fx3EYOnRo1OtiwfM8Nm7ciFmzZnVoO5SIOQMhYovYe+kRMQUZCMGEDIRgQgZCMCEDIZiQgRBMyEAIJmQgBBMyEIIJGQjBhAyEYEIGQjAhAyGYkIEQTMhACCZkIAQTMhCCCRkIwYQMhGBCBkIwIQMhmJCBEEzIQAgmZCAEEzIQggkZCMGEDIRgQgZCMCEDIZiQgRBMyEAIJmQgBBMyEIIJGQjBhAyEYEIGQjAhAyGYkIEQTMhACCZkIAQTMhCCCRkIwYQMhGBCBkIwIQMhmJCBEEzIQAgmZCAEEzIQggkZCMGEDIRgQgZCMCEDIZiQgRBMyEAIJmQgBBMyEIIJGQjBhAyEYEIGQjAhAyGYkIEQTMhACCZkIAQTMhCCibE9K/v4449RX18PAJgxY0Z7Vk2Eie49yNOmvnja1Be1tbWora31ufbrX/8aiYmJaGlpCbncuro62biI9kP3HiTZGAcA6Nq1a8C1CxcuYMKECVixYoXm8g4ePIidO3fCbDaD4zi4XC6MGzcOQ4cO1a3NhDq69yBp8QakxSsX++WXX2L9+vXo2bOn5vISExORl5cHj8cDt9uNmTNnIi4uTq/mEsEQdWZzl5+Km7v8VNcyn3rqKdFms4k2m01cvny5rmUTbHR/xSSZ9P91cxyH1NRUCIIAj8eje/mEOrobiDUtQe8iMX36dCxfvhwAMGXKFN3LJ9TR3UAsqeoGsnnzZjgcDrS0tGDevHmayrt69SoyMjLw+9//HkBbb0K0H7obiCnZBADyJ67318w999yDjIyMkL5i9u/fj+rqap+0J554QoeWElrQ/SsmIdWMhFQzunbtGvCpm5ycjIKCAsyaNUtTWQ0NDcjOzkZ9fT2GDh2KoUOH4uLFi3o3mWCgu4GYU60wp1oVrz3zzDMQBAG7du3SVJbFYkFNTQ2cTidqampQU1OD3NxcPZtLBIETRVHUs8BLS2YCAHJeLNKtzAsXLuD06dMQRRFNTU3Iy8vTrWyCjf49SEoizCmJupb5yiuv4OzZswCA06dP61o2wUZ/JzU1Se8ikZ6ejjFjxqC+vh4pKSm6l0+oo3sPYkpJhiklWdcys7OzkZ6ejsOHDyvO8RDRQ3cDibMmIc6q3os0Nzdj4cKFuHDhguYyKysr4XK5kJeXh3HjxunRTEIjur9iDEltrwClcRApffDgwSGV2b17d58vHxoHaT+iYCCpAJSn+wGgd+/eIfsRZBAdRxQM5Ca9iyQ6EN3HQYSLRwEAhr6361Ymz/MwGAw4d+4cunXrhrS0NN3KJtjo7qQKJisEk/JIargsWbIEx48fx/r167Fy5UpdyybY6G4gYrwFYrxF1zITExOxf/9+PPbYY7SarJ3R3QcRzexR1N27d+PKlSvy8sFgbNu2DT179gTHcfjyyy9DWq5IRE7UXjFKq9oB4Pjx45gxYwbq6uo0lZeTk4PTp0/j5ptvRt++fWk2t53R3UBcYhxcYpzidD8AGI1tnZYgCJrKGzZsGJKSkjBy5Ej88pe/hMWi7+uLYKP7K6aVb1O82sfu4MGDUVRUFNKQ+aRJk/Dss8+C4zhMnjxZh1YSWtH9M7fcZgcA9MrQZ9Ju27Zt4DgOUjM5jsODDz6oS9lEcHR/xbTwAlp4ba8PLeTk5GD//v34yU9+guzsbBw+fFi3song6G4grR4RrR79OiXyQToW3X2QxlZe7yLJB+lAdO9BGpxuNDjdmvLu2LEDr7/+etB86enpsFqtSEhIQHp6eqRNJEJA9x7E7mrb+aY23V9cXIwjR45AFEXMmjULpaWlQctcu3Ytli5dCgAoKCjAn/70J51bTaih/yvG2faK6dq1u+L16dOnY/r06fL/d+7cGbRM2nrZcehuIA0t2l4voeC99ZJ8kPZF93GQZ/7vWwDAn+77qZ7FEh2E7k7qD3YnfrA7dStv3bp1OHjwIBoaGmC327F3714UFhbqVj7BRvceZMo/vgIAvDtrmG5l7tmzBwcOHADHcRgxYgRGjx6tW9kEG/2dVI0+SFlZGYqLi8HzPGbPno3MzEzVvKNHj0a3bt1gt9sxbNgwlJSUYMiQIXo1mWCgu4E4HG0GEuwz1+Vy4ZlnnsEnn3wCu93ONBAAOHToEO655x5s3LgRPM+TgbQT+s/m/mggarO13p+5H374Ibp06YK+ffsGLfeWW25Br169MG3aNBQUFOjXYIKJ7k6q0+GG0xH8NbNv3z588MEHKC8vR0VFBTPvRx99JBuczWYDz/M4duyYLu0l2OjupA763w8AAGdev1+3MidOnIj7778fHMehqakJubm5ePvtt7FgwQLd6iCU0X82196CVnvogXJZDB8+HBMmTMC5c+dQWVkJq9VKs7rthO4+iMt+Te8iMWrUKBQWFuLhhx9GQ0MDXnrpJWRkZOheDxGI7gbidjTqXSREUURjYyO2bNmCsWPH4rnnntO9DkKZKPQg2uKpV1VV4aOPPoLNZsOSJUuY+10OHjyIP/zhDwCAl19+GaNGjdKlrURw9J+s29O28y3YOAgA5ObmYtu2bQjmJ7tcLjidTgiCALdb/8lAQh3dv2K0cvnyZaSlpeGzzz5Djx49VAe+Tp06BZfLhTVr1kAQBDz55JM0SNaOtOt5Md60tLTg7bffBsdxuPvuu1XzSYFmxo8fDwD4/vvvyUDakQ7rQYjOAR1JRjAhAyGYkIEQTMhACCZkIAQTMhCCyf8HvrzVkMKnONwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 108x180 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filenamePlot = os.path.join(plotsFolder, speciesSetName,\n",
    "                            outputFilePrefix + \"_composition_bias_aa_\" + terminus + \"terminal\")\n",
    "print(filenamePlot)\n",
    "fig = plot_aa_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN,\n",
    "                              speciesName=speciesSetName, terminus=terminus, width=1.5)\n",
    "fig.savefig(filenamePlot + \".svg\", dpi=800, bbox_inches=\"tight\")\n",
    "fig.savefig(filenamePlot + \".png\", dpi=800, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### aa pair epistatis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>aa-1</th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>aa0</th>\n",
       "      <th>statistical_test</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">R</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">contingency</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>119</td>\n",
       "      <td>456</td>\n",
       "      <td>18</td>\n",
       "      <td>69</td>\n",
       "      <td>58</td>\n",
       "      <td>93</td>\n",
       "      <td>36</td>\n",
       "      <td>130</td>\n",
       "      <td>40</td>\n",
       "      <td>9</td>\n",
       "      <td>115</td>\n",
       "      <td>20</td>\n",
       "      <td>56</td>\n",
       "      <td>67</td>\n",
       "      <td>123</td>\n",
       "      <td>64</td>\n",
       "      <td>17</td>\n",
       "      <td>64</td>\n",
       "      <td>66</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.31738</td>\n",
       "      <td>2.01748</td>\n",
       "      <td>-0.208737</td>\n",
       "      <td>-0.0260725</td>\n",
       "      <td>-0.634304</td>\n",
       "      <td>0.282632</td>\n",
       "      <td>-0.853186</td>\n",
       "      <td>0.592484</td>\n",
       "      <td>-0.26928</td>\n",
       "      <td>0.229385</td>\n",
       "      <td>0.805524</td>\n",
       "      <td>-0.879883</td>\n",
       "      <td>-0.28537</td>\n",
       "      <td>-0.192589</td>\n",
       "      <td>0.0607836</td>\n",
       "      <td>-0.919574</td>\n",
       "      <td>-0.598269</td>\n",
       "      <td>-0.0780036</td>\n",
       "      <td>2.26959</td>\n",
       "      <td>0.103653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>2.49214</td>\n",
       "      <td>4.04876</td>\n",
       "      <td>0.865294</td>\n",
       "      <td>0.98209</td>\n",
       "      <td>0.644251</td>\n",
       "      <td>1.21641</td>\n",
       "      <td>0.553561</td>\n",
       "      <td>1.50784</td>\n",
       "      <td>0.829733</td>\n",
       "      <td>1.17233</td>\n",
       "      <td>1.74778</td>\n",
       "      <td>0.543411</td>\n",
       "      <td>0.820531</td>\n",
       "      <td>0.875034</td>\n",
       "      <td>1.04303</td>\n",
       "      <td>0.528665</td>\n",
       "      <td>0.660546</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>4.82186</td>\n",
       "      <td>1.07449</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "aa-1                                                   R        K         H  \\\n",
       "terminus aa0 statistical_test observable                                      \n",
       "C        R   contingency      BH_multiple_tests     True     True     False   \n",
       "                              NexpectedSubseq        NaN      NaN       NaN   \n",
       "                              count                  119      456        18   \n",
       "                              log2OddsRatio      1.31738  2.01748 -0.208737   \n",
       "                              oddsRatio          2.49214  4.04876  0.865294   \n",
       "\n",
       "aa-1                                                     D         E  \\\n",
       "terminus aa0 statistical_test observable                               \n",
       "C        R   contingency      BH_multiple_tests      False      True   \n",
       "                              NexpectedSubseq          NaN       NaN   \n",
       "                              count                     69        58   \n",
       "                              log2OddsRatio     -0.0260725 -0.634304   \n",
       "                              oddsRatio            0.98209  0.644251   \n",
       "\n",
       "aa-1                                                    S         T         N  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        R   contingency      BH_multiple_tests     False      True      True   \n",
       "                              NexpectedSubseq         NaN       NaN       NaN   \n",
       "                              count                    93        36       130   \n",
       "                              log2OddsRatio      0.282632 -0.853186  0.592484   \n",
       "                              oddsRatio           1.21641  0.553561   1.50784   \n",
       "\n",
       "aa-1                                                    Q         C         G  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        R   contingency      BH_multiple_tests     False     False      True   \n",
       "                              NexpectedSubseq         NaN       NaN       NaN   \n",
       "                              count                    40         9       115   \n",
       "                              log2OddsRatio      -0.26928  0.229385  0.805524   \n",
       "                              oddsRatio          0.829733   1.17233   1.74778   \n",
       "\n",
       "aa-1                                                    P         A         V  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        R   contingency      BH_multiple_tests      True     False     False   \n",
       "                              NexpectedSubseq         NaN       NaN       NaN   \n",
       "                              count                    20        56        67   \n",
       "                              log2OddsRatio     -0.879883  -0.28537 -0.192589   \n",
       "                              oddsRatio          0.543411  0.820531  0.875034   \n",
       "\n",
       "aa-1                                                     I         L  \\\n",
       "terminus aa0 statistical_test observable                               \n",
       "C        R   contingency      BH_multiple_tests      False      True   \n",
       "                              NexpectedSubseq          NaN       NaN   \n",
       "                              count                    123        64   \n",
       "                              log2OddsRatio      0.0607836 -0.919574   \n",
       "                              oddsRatio            1.04303  0.528665   \n",
       "\n",
       "aa-1                                                    M          F        W  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        R   contingency      BH_multiple_tests     False      False     True   \n",
       "                              NexpectedSubseq         NaN        NaN      NaN   \n",
       "                              count                    17         64       66   \n",
       "                              log2OddsRatio     -0.598269 -0.0780036  2.26959   \n",
       "                              oddsRatio          0.660546   0.947368  4.82186   \n",
       "\n",
       "aa-1                                                    Y  \n",
       "terminus aa0 statistical_test observable                   \n",
       "C        R   contingency      BH_multiple_tests     False  \n",
       "                              NexpectedSubseq         NaN  \n",
       "                              count                    60  \n",
       "                              log2OddsRatio      0.103653  \n",
       "                              oddsRatio           1.07449  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfTableAApair3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1990</td>\n",
       "      <td>351</td>\n",
       "      <td>2376</td>\n",
       "      <td>3310</td>\n",
       "      <td>2282</td>\n",
       "      <td>1786</td>\n",
       "      <td>642</td>\n",
       "      <td>4016</td>\n",
       "      <td>4677</td>\n",
       "      <td>4150</td>\n",
       "      <td>747</td>\n",
       "      <td>3003</td>\n",
       "      <td>1200</td>\n",
       "      <td>1650</td>\n",
       "      <td>1283</td>\n",
       "      <td>2829</td>\n",
       "      <td>2211</td>\n",
       "      <td>2496</td>\n",
       "      <td>458</td>\n",
       "      <td>1793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>-0.204546</td>\n",
       "      <td>0.352227</td>\n",
       "      <td>-0.0045774</td>\n",
       "      <td>0.206388</td>\n",
       "      <td>-0.0453656</td>\n",
       "      <td>-0.321499</td>\n",
       "      <td>-0.0273615</td>\n",
       "      <td>0.0144548</td>\n",
       "      <td>0.21954</td>\n",
       "      <td>-0.0282585</td>\n",
       "      <td>0.104303</td>\n",
       "      <td>-0.0735224</td>\n",
       "      <td>-0.0477191</td>\n",
       "      <td>0.0768669</td>\n",
       "      <td>0.00913927</td>\n",
       "      <td>-0.0852095</td>\n",
       "      <td>-0.0706337</td>\n",
       "      <td>0.0107406</td>\n",
       "      <td>0.0664115</td>\n",
       "      <td>0.0181932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>0.867811</td>\n",
       "      <td>1.27653</td>\n",
       "      <td>0.996832</td>\n",
       "      <td>1.1538</td>\n",
       "      <td>0.969044</td>\n",
       "      <td>0.800238</td>\n",
       "      <td>0.981213</td>\n",
       "      <td>1.01007</td>\n",
       "      <td>1.16436</td>\n",
       "      <td>0.980603</td>\n",
       "      <td>1.07497</td>\n",
       "      <td>0.950315</td>\n",
       "      <td>0.967465</td>\n",
       "      <td>1.05473</td>\n",
       "      <td>1.00635</td>\n",
       "      <td>0.942648</td>\n",
       "      <td>0.95222</td>\n",
       "      <td>1.00747</td>\n",
       "      <td>1.04711</td>\n",
       "      <td>1.01269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>3.18088e-10</td>\n",
       "      <td>1.12143e-05</td>\n",
       "      <td>0.890865</td>\n",
       "      <td>9.59005e-15</td>\n",
       "      <td>0.145576</td>\n",
       "      <td>1.80951e-21</td>\n",
       "      <td>0.649919</td>\n",
       "      <td>0.543045</td>\n",
       "      <td>6.69893e-22</td>\n",
       "      <td>0.233037</td>\n",
       "      <td>0.0531688</td>\n",
       "      <td>0.00698063</td>\n",
       "      <td>0.266269</td>\n",
       "      <td>0.0345883</td>\n",
       "      <td>0.819799</td>\n",
       "      <td>0.00230246</td>\n",
       "      <td>0.0253448</td>\n",
       "      <td>0.716862</td>\n",
       "      <td>0.323821</td>\n",
       "      <td>0.601312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             A            C  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -20                    BH_multiple_tests         True         True   \n",
       "                                count                     1990          351   \n",
       "                                log2OddsRatio        -0.204546     0.352227   \n",
       "                                oddsRatio             0.867811      1.27653   \n",
       "                                pvalue             3.18088e-10  1.12143e-05   \n",
       "\n",
       "                                                           D            E  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -20                    BH_multiple_tests      False         True   \n",
       "                                count                   2376         3310   \n",
       "                                log2OddsRatio     -0.0045774     0.206388   \n",
       "                                oddsRatio           0.996832       1.1538   \n",
       "                                pvalue              0.890865  9.59005e-15   \n",
       "\n",
       "                                                           F            G  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -20                    BH_multiple_tests      False         True   \n",
       "                                count                   2282         1786   \n",
       "                                log2OddsRatio     -0.0453656    -0.321499   \n",
       "                                oddsRatio           0.969044     0.800238   \n",
       "                                pvalue              0.145576  1.80951e-21   \n",
       "\n",
       "                                                           H          I  \\\n",
       "terminus position from terminus observable                                \n",
       "C        -20                    BH_multiple_tests      False      False   \n",
       "                                count                    642       4016   \n",
       "                                log2OddsRatio     -0.0273615  0.0144548   \n",
       "                                oddsRatio           0.981213    1.01007   \n",
       "                                pvalue              0.649919   0.543045   \n",
       "\n",
       "                                                             K          L  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -20                    BH_multiple_tests         True      False   \n",
       "                                count                     4677       4150   \n",
       "                                log2OddsRatio          0.21954 -0.0282585   \n",
       "                                oddsRatio              1.16436   0.980603   \n",
       "                                pvalue             6.69893e-22   0.233037   \n",
       "\n",
       "                                                           M           N  \\\n",
       "terminus position from terminus observable                                 \n",
       "C        -20                    BH_multiple_tests      False        True   \n",
       "                                count                    747        3003   \n",
       "                                log2OddsRatio       0.104303  -0.0735224   \n",
       "                                oddsRatio            1.07497    0.950315   \n",
       "                                pvalue             0.0531688  0.00698063   \n",
       "\n",
       "                                                           P          Q  \\\n",
       "terminus position from terminus observable                                \n",
       "C        -20                    BH_multiple_tests      False       True   \n",
       "                                count                   1200       1650   \n",
       "                                log2OddsRatio     -0.0477191  0.0768669   \n",
       "                                oddsRatio           0.967465    1.05473   \n",
       "                                pvalue              0.266269  0.0345883   \n",
       "\n",
       "                                                            R           S  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -20                    BH_multiple_tests       False        True   \n",
       "                                count                    1283        2829   \n",
       "                                log2OddsRatio      0.00913927  -0.0852095   \n",
       "                                oddsRatio             1.00635    0.942648   \n",
       "                                pvalue               0.819799  0.00230246   \n",
       "\n",
       "                                                           T          V  \\\n",
       "terminus position from terminus observable                                \n",
       "C        -20                    BH_multiple_tests       True      False   \n",
       "                                count                   2211       2496   \n",
       "                                log2OddsRatio     -0.0706337  0.0107406   \n",
       "                                oddsRatio            0.95222    1.00747   \n",
       "                                pvalue             0.0253448   0.716862   \n",
       "\n",
       "                                                           W          Y  \n",
       "terminus position from terminus observable                               \n",
       "C        -20                    BH_multiple_tests      False      False  \n",
       "                                count                    458       1793  \n",
       "                                log2OddsRatio      0.0664115  0.0181932  \n",
       "                                oddsRatio            1.04711    1.01269  \n",
       "                                pvalue              0.323821   0.601312  "
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "freq1 = oddsRatioDf.loc[('C', -1, 'count')].copy()\n",
    "freq1 = freq1 / freq1.sum()\n",
    "freq2 = oddsRatioDf.loc[('C', -2, 'count')].copy()\n",
    "freq2 = freq2 / freq2.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A     0.0305896\n",
       "C    0.00467052\n",
       "D     0.0458035\n",
       "E     0.0794451\n",
       "F      0.048185\n",
       "G     0.0303353\n",
       "H     0.0111445\n",
       "I      0.076578\n",
       "K      0.247306\n",
       "L     0.0655723\n",
       "M    0.00769942\n",
       "N      0.101711\n",
       "P    0.00656647\n",
       "Q     0.0313988\n",
       "R     0.0392601\n",
       "S     0.0630058\n",
       "T     0.0189595\n",
       "V     0.0407861\n",
       "W     0.0138497\n",
       "Y     0.0371329\n",
       "Name: (C, -1, count), dtype: object"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RR     0.00152502\n",
       "RK     0.00960635\n",
       "RH    0.000432897\n",
       "RD     0.00177919\n",
       "RE     0.00308596\n",
       "dtype: object"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rowList = []\n",
    "for aa2 in aaTable:\n",
    "    row = (freq2.loc[aa2]*freq1).to_frame().T[aaTable]\n",
    "    row.index = [aa2]\n",
    "    rowList.append(row)\n",
    "\n",
    "df = pd.concat(rowList)\n",
    "df.index.name = 'aa-2'\n",
    "df.columns.name = 'aa-1'\n",
    "df = df.stack()\n",
    "df.index = df.index.map(lambda x: ''.join(x))\n",
    "freqRel12ProductDf = df\n",
    "freqRel12ProductDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RR    119\n",
       "RK    456\n",
       "RH     18\n",
       "RD     69\n",
       "RE     58\n",
       "dtype: object"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = oddsRatioDfTableAApair3.loc[idx['C', :, 'contingency', 'count']]\n",
    "df1.index = df1.index.droplevel([0, 2, 3])\n",
    "df1.index.name = 'aa-2'\n",
    "df1.columns.name = 'aa-1'\n",
    "\n",
    "df1 = df1.stack()\n",
    "df1.index = df1.index.map(lambda x: ''.join(x))\n",
    "ctermSubseqFreq = df1\n",
    "ctermSubseqFreq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctermSubseqFreq.get('KX', 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.019630657257290702"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binomial_chi_square_test(0.03*100, 100, 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.05"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "family_wise_FDR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:36: RuntimeWarning: divide by zero encountered in log2\n"
     ]
    }
   ],
   "source": [
    "subSeqList = ctermSubseqFreq.index\n",
    "computeMultipleTestsCorrection = True\n",
    "\n",
    "# This dataframe is organized with the list of subsequences as index\n",
    "terminiList = ['C']\n",
    "methodList = ['theoretical']\n",
    "obsList = ['count','theoreticalProb','NexpectedSubseq','log2OddsRatio','oddsRatio','pvalue']\n",
    "indexNameList = ['terminus','subseq','statistical_test','observable']\n",
    "multiIndex = pd.MultiIndex.from_product([terminiList, subSeqList, methodList, obsList],\n",
    "                                        names=indexNameList)\n",
    "oddsRatioDf = pd.DataFrame(index=multiIndex, columns=['value'])\n",
    "oddsRatioDf.sort_index(level=0, inplace=True)\n",
    "\n",
    "# For amino acid pairs: this dataframe is organized with the first amino acid\n",
    "# as row index and second amino acid as column index.\n",
    "multiIndex = pd.MultiIndex.from_product([terminiList, aaTable, methodList, obsList],\n",
    "                                        names=['terminus','aa0','statistical_test','observable'])\n",
    "oddsRatioPairsDf = pd.DataFrame(index=multiIndex, columns=aaTable)\n",
    "# It is better not to sort the dataframe because we want to keep the custom order for the aa table\n",
    "\n",
    "i = 0\n",
    "for terminus in ['C']:\n",
    "    terminalSize = None\n",
    "    observedFreq = None\n",
    "    if terminus == 'C':\n",
    "        terminalSize = cTerminalSize\n",
    "        observedFreq = ctermSubseqFreq\n",
    "\n",
    "    for subseq in subSeqList:\n",
    "\n",
    "        i += 1\n",
    "        NobservedSubseq = observedFreq.get(subseq, 0)\n",
    "        NobservedTot = sum(observedFreq.values)\n",
    "        theoreticalProb = freqRel12ProductDf.loc[subseq]    # !!!!!!!!!\n",
    "        oddsRatio = ((NobservedSubseq/NobservedTot) / theoreticalProb) if theoreticalProb > 0 else None\n",
    "        log2OddsRatio = log2(oddsRatio)\n",
    "        pvalue = binomial_chi_square_test(NobservedSubseq, NobservedTot, theoreticalProb)\n",
    "\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','count'] = NobservedSubseq\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','theoreticalProb'] = theoreticalProb\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','NexpectedSubseq'] = theoreticalProb*NobservedTot\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','oddsRatio'] = oddsRatio\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','log2OddsRatio'] = log2OddsRatio\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','pvalue'] = pvalue\n",
    "\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','count'),subseq[1]] = NobservedSubseq\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','theoreticalProb'),subseq[1]] = theoreticalProb\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','NexpectedSubseq'),subseq[1]] = theoreticalProb*NobservedTot\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','oddsRatio'),subseq[1]] = oddsRatio\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','log2OddsRatio'),subseq[1]] = log2OddsRatio\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','pvalue'),subseq[1]] = pvalue\n",
    "\n",
    "\n",
    "if computeMultipleTestsCorrection:\n",
    "    dfList = []\n",
    "    for terminus in ['C']:\n",
    "        # Multiple test correction within the biases of C-terminal at all positions for all a.a.\n",
    "        df = oddsRatioDf.xs(terminus, level='terminus', drop_level=False)\n",
    "        df = df.xs('pvalue', level='observable', drop_level=False)\n",
    "        # We serialize all the values, and drop the NaN\n",
    "        df2 = df.stack().dropna().copy()\n",
    "        reject, pvals_corrected, alphacSidak, alphacBonf = \\\n",
    "            multipletests(df2.values, alpha=family_wise_FDR, method='fdr_bh',\n",
    "                          is_sorted=False, returnsorted=False)\n",
    "        df2 = pd.DataFrame(reject, index=df2.index)\n",
    "        # Stack again the values before merging\n",
    "        df2 = df2.unstack()\n",
    "        df2.columns = df2.columns.droplevel(0)\n",
    "        df2 = df2.rename(index={'pvalue':'BH_multiple_tests'})\n",
    "        dfList.append(df2)\n",
    "    oddsRatioDf = pd.concat([oddsRatioDf] + dfList, axis=0, sort=True)\n",
    "    oddsRatioDf.sort_index(inplace=True)\n",
    "\n",
    "if computeMultipleTestsCorrection:\n",
    "    dfList = []\n",
    "    for terminus in ['C']:\n",
    "        # Multiple test correction within the biases of C-terminal at all positions for all a.a.\n",
    "        df = oddsRatioPairsDf.xs(terminus, level='terminus', drop_level=False)\n",
    "        df = df.xs('pvalue', level='observable', drop_level=False)\n",
    "        # We serialize all the values, and drop the NaN\n",
    "        df2 = df.stack().dropna().copy()\n",
    "        reject, pvals_corrected, alphacSidak, alphacBonf = \\\n",
    "            multipletests(df2.values, alpha=family_wise_FDR, method='fdr_bh',\n",
    "                          is_sorted=False, returnsorted=False)\n",
    "        df2 = pd.DataFrame(reject, index=df2.index)\n",
    "        # Stack again the values before merging\n",
    "        df2 = df2.unstack()\n",
    "        df2.columns = df2.columns.droplevel(0)\n",
    "        df2 = df2.rename(index={'pvalue':'BH_multiple_tests'})\n",
    "        dfList.append(df2)\n",
    "    oddsRatioPairsDf = pd.concat([oddsRatioPairsDf] + dfList, axis=0, sort=True)\n",
    "    oddsRatioPairsDf.sort_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>subseq</th>\n",
       "      <th>statistical_test</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">AA</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">theoretical</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>44.7526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.07069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.05022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      value\n",
       "terminus subseq statistical_test observable                \n",
       "C        AA     theoretical      BH_multiple_tests    False\n",
       "                                 NexpectedSubseq    44.7526\n",
       "                                 count                   47\n",
       "                                 log2OddsRatio      0.07069\n",
       "                                 oddsRatio          1.05022"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioAApairCtermNormDf = oddsRatioDf\n",
    "oddsRatioAApairCtermNormDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>aa0</th>\n",
       "      <th>statistical_test</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">A</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">theoretical</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>44.7526</td>\n",
       "      <td>6.83297</td>\n",
       "      <td>67.0105</td>\n",
       "      <td>116.228</td>\n",
       "      <td>70.4946</td>\n",
       "      <td>44.3805</td>\n",
       "      <td>16.3044</td>\n",
       "      <td>112.034</td>\n",
       "      <td>361.809</td>\n",
       "      <td>95.9322</td>\n",
       "      <td>11.2643</td>\n",
       "      <td>148.803</td>\n",
       "      <td>9.60675</td>\n",
       "      <td>45.9365</td>\n",
       "      <td>57.4375</td>\n",
       "      <td>92.1775</td>\n",
       "      <td>27.7378</td>\n",
       "      <td>59.6701</td>\n",
       "      <td>20.2621</td>\n",
       "      <td>54.3255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>47</td>\n",
       "      <td>6</td>\n",
       "      <td>41</td>\n",
       "      <td>108</td>\n",
       "      <td>46</td>\n",
       "      <td>32</td>\n",
       "      <td>16</td>\n",
       "      <td>96</td>\n",
       "      <td>458</td>\n",
       "      <td>70</td>\n",
       "      <td>9</td>\n",
       "      <td>137</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>65</td>\n",
       "      <td>92</td>\n",
       "      <td>30</td>\n",
       "      <td>75</td>\n",
       "      <td>9</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.07069</td>\n",
       "      <td>-0.187551</td>\n",
       "      <td>-0.708763</td>\n",
       "      <td>-0.105928</td>\n",
       "      <td>-0.615879</td>\n",
       "      <td>-0.471854</td>\n",
       "      <td>-0.0271909</td>\n",
       "      <td>-0.222826</td>\n",
       "      <td>0.340118</td>\n",
       "      <td>-0.45466</td>\n",
       "      <td>-0.323755</td>\n",
       "      <td>-0.119229</td>\n",
       "      <td>-inf</td>\n",
       "      <td>0.522825</td>\n",
       "      <td>0.178446</td>\n",
       "      <td>-0.0027801</td>\n",
       "      <td>0.113109</td>\n",
       "      <td>0.329882</td>\n",
       "      <td>-1.17079</td>\n",
       "      <td>0.143333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.05022</td>\n",
       "      <td>0.878095</td>\n",
       "      <td>0.611845</td>\n",
       "      <td>0.929207</td>\n",
       "      <td>0.652532</td>\n",
       "      <td>0.721038</td>\n",
       "      <td>0.981329</td>\n",
       "      <td>0.856885</td>\n",
       "      <td>1.26586</td>\n",
       "      <td>0.729682</td>\n",
       "      <td>0.798988</td>\n",
       "      <td>0.920679</td>\n",
       "      <td>0</td>\n",
       "      <td>1.43677</td>\n",
       "      <td>1.13166</td>\n",
       "      <td>0.998075</td>\n",
       "      <td>1.08156</td>\n",
       "      <td>1.25691</td>\n",
       "      <td>0.444178</td>\n",
       "      <td>1.10445</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       A         C         D  \\\n",
       "terminus aa0 statistical_test observable                                       \n",
       "C        A   theoretical      BH_multiple_tests    False     False      True   \n",
       "                              NexpectedSubseq    44.7526   6.83297   67.0105   \n",
       "                              count                   47         6        41   \n",
       "                              log2OddsRatio      0.07069 -0.187551 -0.708763   \n",
       "                              oddsRatio          1.05022  0.878095  0.611845   \n",
       "\n",
       "                                                        E         F         G  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        A   theoretical      BH_multiple_tests     False      True     False   \n",
       "                              NexpectedSubseq     116.228   70.4946   44.3805   \n",
       "                              count                   108        46        32   \n",
       "                              log2OddsRatio     -0.105928 -0.615879 -0.471854   \n",
       "                              oddsRatio          0.929207  0.652532  0.721038   \n",
       "\n",
       "                                                         H         I  \\\n",
       "terminus aa0 statistical_test observable                               \n",
       "C        A   theoretical      BH_multiple_tests      False     False   \n",
       "                              NexpectedSubseq      16.3044   112.034   \n",
       "                              count                     16        96   \n",
       "                              log2OddsRatio     -0.0271909 -0.222826   \n",
       "                              oddsRatio           0.981329  0.856885   \n",
       "\n",
       "                                                        K         L         M  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        A   theoretical      BH_multiple_tests      True      True     False   \n",
       "                              NexpectedSubseq     361.809   95.9322   11.2643   \n",
       "                              count                   458        70         9   \n",
       "                              log2OddsRatio      0.340118  -0.45466 -0.323755   \n",
       "                              oddsRatio           1.26586  0.729682  0.798988   \n",
       "\n",
       "                                                        N        P         Q  \\\n",
       "terminus aa0 statistical_test observable                                       \n",
       "C        A   theoretical      BH_multiple_tests     False     True      True   \n",
       "                              NexpectedSubseq     148.803  9.60675   45.9365   \n",
       "                              count                   137        0        66   \n",
       "                              log2OddsRatio     -0.119229     -inf  0.522825   \n",
       "                              oddsRatio          0.920679        0   1.43677   \n",
       "\n",
       "                                                        R          S  \\\n",
       "terminus aa0 statistical_test observable                               \n",
       "C        A   theoretical      BH_multiple_tests     False      False   \n",
       "                              NexpectedSubseq     57.4375    92.1775   \n",
       "                              count                    65         92   \n",
       "                              log2OddsRatio      0.178446 -0.0027801   \n",
       "                              oddsRatio           1.13166   0.998075   \n",
       "\n",
       "                                                        T         V         W  \\\n",
       "terminus aa0 statistical_test observable                                        \n",
       "C        A   theoretical      BH_multiple_tests     False     False      True   \n",
       "                              NexpectedSubseq     27.7378   59.6701   20.2621   \n",
       "                              count                    30        75         9   \n",
       "                              log2OddsRatio      0.113109  0.329882  -1.17079   \n",
       "                              oddsRatio           1.08156   1.25691  0.444178   \n",
       "\n",
       "                                                        Y  \n",
       "terminus aa0 statistical_test observable                   \n",
       "C        A   theoretical      BH_multiple_tests     False  \n",
       "                              NexpectedSubseq     54.3255  \n",
       "                              count                    60  \n",
       "                              log2OddsRatio      0.143333  \n",
       "                              oddsRatio           1.10445  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioAApairCtermNormDfTable = oddsRatioPairsDf\n",
    "oddsRatioAApairCtermNormDfTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAApairCtermNormDf.csv\")\n",
    "oddsRatioAApairCtermNormDf.to_csv(filename)\n",
    "\n",
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAApairCtermNormDfTable.csv\")\n",
    "oddsRatioAApairCtermNormDfTable.to_csv(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAApairCtermNormDf.csv\")\n",
    "oddsRatioAApairCtermNormDf = pd.read_csv(filename, index_col=[0,1,2,3])\n",
    "\n",
    "filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAApairCtermNormDfTable.csv\")\n",
    "oddsRatioAApairCtermNormDfTable = pd.read_csv(filename, index_col=[0,1,2,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>subseq</th>\n",
       "      <th>statistical_test</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">AA</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">theoretical</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NexpectedSubseq</th>\n",
       "      <td>44.7526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.07069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.05022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      value\n",
       "terminus subseq statistical_test observable                \n",
       "C        AA     theoretical      BH_multiple_tests    False\n",
       "                                 NexpectedSubseq    44.7526\n",
       "                                 count                   47\n",
       "                                 log2OddsRatio      0.07069\n",
       "                                 oddsRatio          1.05022"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioAApairCtermNormDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "43250.0"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDfTableAApair3.loc[idx['C', :, 'contingency', 'count'], :]\n",
    "countTot = df.sum().sum()\n",
    "countTot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.345460584738237"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioAApairCtermNormDfTable.loc['C'].loc[idx['M', 'theoretical', 'oddsRatio'], 'M']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminus = 'C'\n",
    "statistical_test = 'theoretical'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = compute_subseq_oddsratio_plot_data(oddsRatioAApairCtermNormDfTable, terminus, statistical_test)\n",
    "maskData = compute_subseq_oddsratio_mask_data(oddsRatioAApairCtermNormDfTable, pvalueThresholds,\n",
    "                                              terminus, statistical_test, subseq_type='aa')\n",
    "pvalAnnotTable = compute_subseq_pvalueAnnotation(oddsRatioAApairCtermNormDfTable, pvalueThresholds,\n",
    "                                                 terminus, statistical_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData.columns.name = 'aa1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         A      C      D      E      F      G      H      I      K      L  \\\n",
       "aa0                                                                         \n",
       "A    False  False  False  False  False  False  False  False  False  False   \n",
       "C    False  False  False  False  False  False  False  False  False  False   \n",
       "D    False  False  False  False  False  False  False  False  False  False   \n",
       "E    False  False  False  False  False  False  False  False  False  False   \n",
       "F    False  False  False  False  False  False  False  False  False  False   \n",
       "G    False  False  False  False  False  False  False  False  False  False   \n",
       "H    False  False  False  False  False  False  False  False  False  False   \n",
       "I    False  False  False  False  False  False  False  False  False  False   \n",
       "K    False  False  False  False  False  False  False  False  False  False   \n",
       "L    False  False  False  False  False  False  False  False  False  False   \n",
       "M    False  False  False  False  False  False  False  False  False  False   \n",
       "N    False  False  False  False  False  False  False  False  False  False   \n",
       "P    False  False  False  False  False  False  False  False  False  False   \n",
       "Q    False  False  False  False  False  False  False  False  False  False   \n",
       "R    False  False  False  False  False  False  False  False  False  False   \n",
       "S    False  False  False  False  False  False  False  False  False  False   \n",
       "T    False  False  False  False  False  False  False  False  False  False   \n",
       "V    False  False  False  False  False  False  False  False  False  False   \n",
       "W    False  False  False  False  False  False  False  False  False  False   \n",
       "Y    False  False  False  False  False  False  False  False  False  False   \n",
       "\n",
       "         M      N      P      Q      R      S      T      V      W      Y  \n",
       "aa0                                                                        \n",
       "A    False  False  False  False  False  False  False  False  False  False  \n",
       "C    False  False  False  False  False  False  False  False  False  False  \n",
       "D    False  False  False  False  False  False  False  False  False  False  \n",
       "E    False  False  False  False  False  False  False  False  False  False  \n",
       "F    False  False  False  False  False  False  False  False  False  False  \n",
       "G    False  False  False  False  False  False  False  False  False  False  \n",
       "H    False  False  False  False  False  False  False  False  False  False  \n",
       "I    False  False  False  False  False  False  False  False  False  False  \n",
       "K    False  False  False  False  False  False  False  False  False  False  \n",
       "L    False  False  False  False  False  False  False  False  False  False  \n",
       "M    False  False  False  False  False  False  False  False  False  False  \n",
       "N    False  False  False  False  False  False  False  False  False  False  \n",
       "P    False  False  False  False  False  False  False  False  False  False  \n",
       "Q    False  False  False  False  False  False  False  False  False  False  \n",
       "R    False  False  False  False  False  False  False  False  False  False  \n",
       "S    False  False  False  False  False  False  False  False  False  False  \n",
       "T    False  False  False  False  False  False  False  False  False  False  \n",
       "V    False  False  False  False  False  False  False  False  False  False  \n",
       "W    False  False  False  False  False  False  False  False  False  False  \n",
       "Y    False  False  False  False  False  False  False  False  False  False  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskDataSmallN = maskData.copy()\n",
    "maskDataSmallN.loc[:, :] = False\n",
    "maskDataSmallN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(cterminal)\n",
    "from cterminal import plot_seq_pair_composition_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Analysis_data/C-terminal/Data/Taxonomy/taxaId_31969/taxaId_31969_subset_all_composition_bias_aa_pairs_Cterminal_CtermNorm\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAJIAAAD2CAYAAAA9Ht7CAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJztnXtcVNX6/z8zzAx35KYRoqAoeOPiHVJfRpplec4xChUjoht1RDpiSdrvxJfSxKPmbUjLXiqvIgJBTEVOqHjjW5giIBJieBAQkYviIMNlBpj9/YMfGzbMFRaX4az3P7of19przTwf9zxrr7WexWMYhgGF0kf4g90ByvCAColCBCokChGokChEoEKiEIEKiUIEg8jIyMiBaEgsFuPWrVtISkrC3LlzIRKJIBaL4ebmhp07d2L69OkQiURsWU9PTxw4cACenp4QCARs/YyMDMydO7fHffPz8zFhwgQIhUKkpqbi6tWrSElJgbe3N/h8Ps6cOYNt27ahtrYWtra2sLCwYOuXl5fjxo0bsLKygoWFBc6fP4+zZ88iIyMDXl5eAIC8vDwkJCTg7NmzmDVrFoRCIceelpYGb29vGBgYQCwWo6CgAMnJyfDy8mLLhoeHw8rKCvb29mz/o6KisGDBAly+fBklJSVwcnICABQVFWHjxo2QSqVwc3MDACQnJ0MkEsHa2pqtHxERASMjI0RFRcHDw4PzuTq+l6lTp4LP52P37t3w9vZGZWUlTpw4AXd3d0RHR8PFxQXx8fGwsrKCpaVlr/w7oE8kf39/eHh4oKKigrXt2rULAQEBMDMz06q+o6MjCgsLe9gDAgJgYmICAHjy5AnkcjmCgoIgEAgAAEuWLIG7uzv8/f3h4ODAqd+1PwCQk5OD4OBg2NjYoKqqCgCQnp6O0NBQhIeHs+0AwLlz5xAaGoq1a9eCz+/8OlevXg03Nzc8ePAAAFBVVYXZs2cjOzub05aDgwMePHiAa9euYd68eax94sSJcHd3x6uvvqr2Oxk5ciSysrKg7HVgx/diYGAAABAIBIiJiUFBQQHGjRsHAHjzzTcRHR0NhmHg6Oioti11DKiQYmJiUFxcjIkTJ7I2JycnXLhwoUfZ2NhY5ObmKr0Pj8fjXCcmJuLYsWPsta+vL5YuXYpvv/22h0iUMXfuXFy5coW9VigU7J8dTgAAhmHQ0tICuVzO6UtNTQ0OHz7MigYAjh49itbWVjg7OwMA0tLSUF9fj9u3b3Pqv/TSSzh9+jREIhErel0ZM2YM5s+f38MeGxuL2NhY9looFEIqleLEiROskMzNzWFqago/P79etd3BgAopKCgIQqEQDx8+ZG1+fn5obW1FQUEBp2xAQAA8PT05tri4OBQXF8PV1ZVjb21tRVNTE6qrqwEASUlJOH/+PMzMzGBubq6xXx4eHmxdAJgzZw727t2Luro62NraAgCWLl2K/fv348CBAxwhvPjiizh69CgAcJ5IK1aswOrVq9nr6upqvP322wgMDMSlS5dYu42NDQoLC7Fw4UJOn4qKipCXl8f5DwIAp06dQmxsLGQyGWtbvnw5jI2Ne3yugIAABAQEsNcikQiurq6QSCSws7Nj7QKBgP357S08OkWinLy8PKSkpCAkJAQjRowY7O4MeaiQKESgw38KEXoX3emAXC6HTCbjxA+UoYtCoYChoSH7KkZb+t27MpmME5xShjYd//F1hdgTSSwWw9zcHHfu3MGmTZtgamoKoH0kU11dzRlGu7i4AAD+/PNPzj1cXFxQ19DEsY0wNe5RtqP+vH+d55T99ZPnlN5TVVuqyl4vl7C2mQ7tL+haKv/DKSu0c9apLfnjSo5NZNU+apI2dn5eM5Oen1VTX3Upm/zUVNbmW/WH0vqjR49Gd1T5titEn0jKXjhShgeafEtUSMpeOFKGBxp9yxBi3759THNzM/PVV18xNTU1rF0qlTJSqZRUM5R+Rpm/VPm2K/3+HqmhoQEAlP6uUoYevfVXvw//AaCsrExpsK34z1VOOb7zHLQ8vMexCW3HAFAebP9W8ohT9hknGyJBqbyucwpHNMJW5/q9Ldub+q33uRPYgtGTlH4vAHDxP52f61ln5Z9LWbCtDcRiJLFYjMbGRkRFRUEqlZK6LWUIoI1viQbbuiwJoegXmnxLVEiqloRQ9B+NviUV7XdE9ocOHWL++OMPtaMAytBF3aitu2+7MiCjNlXBtrIAcpOhM8cWJWt/o9yYtIO1mby2AQDQlp/OKWswbRHkkmqOTWQ5SmVbqgLYgB+yWFvsG7N0rq/Mrmxg0b1sR/2i6npO2YmjzHH3Idc2zrZ9nZWJ94cce2PmPp0C+64DCwBoEbS/Xdd11EY02JbJZOyflOGDNr6lU/IUIhAVkrp11hT9RqNvSQVpHQFZx5/qgjfK0EVdsN3dt12hUyQUDkN6iuTwtTI8aesctf2/RapHN7ernnBsrk+1b/hrPnOItRkteUdlfRLTFrKGzhGSoak5kbbu1HBHXRNG9ryv2vVMtdzlGyLr9k2WC7Zz3+1khPug9cYZjk3gsQQA0FrReV+BvfK2ejtFQnxhW319PUJDQ0ndljIE0Ma3xBe2KdsJS9F/NPlW7RNJJpPhp59+ws2bNyESiTB16lSsWrVK447Q7jthKcMHVb5VG2zHxMRg0aJFGDOmfSnH3bt3kZGRgcDAwB5lxWIxzMzMUFdXh3Xr1rF2GmzrF8r8pcq3XVErpLKysh4K7BCVLh17UNcEgaERa3OyaZ9B7uu6nZYrxzllhV6vkFmP9KictYlsHHSu7/npvzn23K1L+7we6VF9I8dmY26ic78+SOx8D/SNn6fS+h3BNtFR28mTJ3vY1q5dq7RsR0AmEAjg6+vLydhB0W+08a1aIa1duxa///47rl27BqDzsacKf39/GBoa9qHLlKGKJt9qHLX9+OOPMDQ0hKenJ5qamtSW7Z5ehjJ80ORbjW+2N27cCF9fX9y7dw/Z2dn48ssvlZbr+vhbsmQJRo1qX75Bg239QlWwrcy3XdEopMLCQlhZWeH48eOws7PD8uXL+9wxytCl36ZIysrK8P3334NhGJSWlqoUUmpqKh4/fozS0lKEh4dz3jXpsrBN1Yhj7bE81hb9qrvO9QdyZ8cfD7jTPFOftujzFElb6Q2OzcDRg8jnaiu7ySkLm/GcS3V+7YpGIV27dg2bN2/mCEEZyvI2UvQfbf2qMdhubm7G5cuXkZmZiczMTJXldM3bSNEPtPWrxhjp+PHjnJeSqn7a4uLiAAD379/HBx98wOZupDGSftHdX6r82h21Qtq/fz/u3r3bWZjHw/bt2/vUMcrQprf+0nphW0fKYF0zr+m6i0T+vwkcm2j+SgDAsZudj9VX3exV1lcVaJZuDOLYHbfF9NsUyUAG9qTb6pcpEgD45JNPwOPx2J+3qKgopeXOnDmDxMREzJs3D4sXL+6RFJ2in2jrV41CcnJywt///neNDS5ZsgS3b9+m0yTDDG39qlFIt2/fRmhoKIyMjHoVI1H+S9C0q2Dv3r1a70BQtsuA7iLRL9TtIlGHxmA7KCgIZmZm7NKB3ozaSLzZbiu4yNoMpjyrsn5L1V2OTfjUOJ3bGmoBMKm29vxvMWtbN3+80vr9Fmy//PLLWh14IhaLYW9vD7lcjgULFtBgexihjW81CumXX35BU1MT+0R67bXXVJatqKhgD4GhDC80+VbjS6H3338fkydPxtixYzWe59X9uCrK8EGjbzUFX4cOHWLWrVvHnDp1ijl8+LDKch0B2apVq5h79+6pDd4oQxd1wXZ333ZFY7C9efNmmJqaYv369YiIiMAXX3yhk5LpFIl+0W/rkfh8PkpKShAfH69xzbYqSIzaMoo7M7UuGN+epbXmCXdnxUgLE6VZXnVtqyF+K2szXfUpAOB/0rj3/fyFSf02ait7xE34OdbGTKe2zv7JTTb2vEv7isauSbkaM/cprd9vW7Zff/11lJaWQqFQIDg4WGW58+fP486dO2hsbFS594mif2jrV5XBdlVVFcLDw/Hhhx/i9OnTSE1Nxb/+9S+VN1J2oDBF/9HWryqfSE899RS2b9+O7OxszJgxA3V1dWqP5FR1oDBFv9HWr1ovI/noo4/w1Vdfqfz3S5cuITc3FzweDx9+2PlbTINt/aK7v1T5tTtaL64eO3as2n9fuHAhrKyskJKS0uPpRSLY1iZQ1HUqITzlD45t+7KpPcr297RFze71rG1k2C4AQNufv3LKGrjMU1k/04d7Orf3hUuo/4E7sjZ/I0Ll55KdO8IpC+8VnEt1fu2K2heSJSUluHjxIoB2hZaUlKgsm5eXh4sXL0KhUPT56G/K0EFbv6p9Im3bto1dyObr64vIyEjs27dPadn09HSsW7cOLS0tOp+HShm6aOtXtU8ka2trWFlZsX83NjZW2yjDMGhpaaFn2A4ztPGr2mD7iy++wOLFi+Hh4YHc3FykpKSoXGpbWFiIc+fOobm5GR988AF7+AkNtvWL7v5S5dfuqBWSXC5HXFwcewTlihUrdF5GS4WkX/TLFMnJkyfh5+fH3lQqlbK27nQEZRKJBBs3buT8npIYtQ3kSErbrLaqzgcZqp9Ll10kHajza1fUCsnHxwcHDx5ERUUFDAwMMGrUKLzzzjtKy547dw5hYWF4/PixzluWKEMXbf2qVkg2NjYICwvTqkEej4eamhp8//33WLlypc4pAilDE239Sizz/61bt5Ceno7m5mb4+/uzj0gaI+kX3f2lyq/dUSuk6upq5OTkwMfHByKRCNeuXcPs2bP71DHK0KZfgu0tW7bgr3/9KyIiIvDPf/4TiYmJOgsJ0C3YznnlBY5t+vE0ANrtgCAWbHeZNjBc/JbK+soO3etLvzpsX5y9zSkb8bwrkc9180Eda3N7un2qo3u2XKNehrdqhWRubo7FixdjxowZ2L59u9oD/TpyMRcXF2PTpk00q+0wQhvfqtWfl5cXLly4AGtrawQHB+PRo0fqimP16tVwc3PDgwcP+tZzypBDk2/VPpH+8pe/sH/PzMxEbGys2saOHj0KAwMDODs7qy1H0T80+lbb3QVhYWFq/13Vtl66i0S/6Lct2x0UFhZi0qRJOiu5oaEBrZJKmIg6H37Cke374/oalCo7VqF7LvCOiWb540qOXWRlp3TzAABOjqaO/EwD+bZZ2eL/rpsfgM4NEEMlP5LWMbo2IgoPD0dWVpbGchT9Qhu/EpvLqKqqwuzZs5GdnU3qlpQhgLZ+1bjU9vPPP0dDQwP7HkjVMpK0tDS0trbi9u3bkMvldHHbMEFrv2oKvvbv369VkLZjxw6GYRgmNzeXOXPmjNrgjTJ06e4vVX7tjsZg29fXFw4ODjAyaj9vjWa1Hd70a1bbW7duQaFQYMqUKTofM9rQ0IAKSSPn4D9S63YkUu6oy9LMpN9GUmYLPuKUlWZ8pVNbLZX/4diEdu3vY5obO0doRiZ9OxBRU9mWrJTO9mctU1q/30ZtW7ZswW+//YbMzEzs2LFDZbmO2Ony5cu4dOmSTp2gDF209atWo7Z33nkH7777LtQ9vBwcHPDgwQNcu3YN8+bN07G7lKGKtn7VOGqztLREZGQkeDwebGxsVJZ76aWXcPz4cYhEInqozTBCW79qXI/U3NyM5uZmMAwDExMTtSsfP/74YwQGBsLd3Z210WBbv1DmL2V+7Y5aIR07dqzHbK+qw5HVdYzE4n9tj3XoGrwCfQ9gO2yKYu6bXf74War7WveQYxeNsNVqowBbX8LNbySyHEUk2JYc3MTaLIOjlNYnHmzX1dXB3t4eEokEM2fOxMyZM1FcXKyqOIqKivDee++xp+lQhg/a+Fblj56xsTGqqqogl8vZvDj+/v4qbzRx4kS4u7vj1Vdf7UOXKUMRbXyrUkgikQgtLS2wtrZmk0eUlpb2aqktZfijNkZ6+PAh5HI5GIZhEy3Z29srLVtUVITt27fDx8cHq1evZu002NYvlPlLlW+7ovHN9q5du7Bs2TJ8+eWXmDBhAj777LM+d4wydOm3rLaNjY24evUqQkJCkJaWprJccnIyPDw8lC7F1GXUdqOijmPzsB/Ro6zagwO7jO6Avh/cx46karnnuYqs7VXWby0v4NgFDlMGdMu2slEjAMiknd+toVnP7xVQntVWnW870Phm28LCAvn5+TAzM9OY1oby34tGIQUEBGDSpEm4evUq3n333YHoE0UP0RgjhYeH44033gDDMEhMTMTmzZuVlktOTkZZWRlsbW3h5+fHpr+hMZJ+ocxfqnzbFY0xklAohJubGwAgKSlJZTlfX1+dO03RD7TxrUYh+fj4YMOGDWAYBi+++GKvOqJLsN1y5TjHJvR6BYB2a2lcXFxwp4Y7FTFhpO5rnxRFnacA8Sd66Vz/yRHuyNbirc39FmxfW7qIY5/973SVZe1e7cz/WXnsQ6Vt9fYICY0x0oQJE2Braws7Ozu1O0kiIiKQmZmJNWvWoLy8XGU5in6hrV81CikhIQEhISF46623sH//fpXlRo4ciaysLLVrlij6h7Z+1WrhUG1tLRQKBRiGwb1791QuJRkzZgysra117y1lSKONXzWO2qKjo3vYlC0lEYvFCA0NRXJyMubMmcOee0pHbfpFd3+p8mt3iGVsU9exoZqMtK3sJsdmMNZNZVtdA3CgPQhXeSyEkq3g31wp4dg+8HLqt881GFu2ia2JFYvFMDc3h0AggL+/Pz0haRihjW+JLq729/fXOQ83RT/Q5FuieYxjY2M15lCi6CcafUtopy/NjzRM6Pf8SL2Fjtr0i35bj0QCXUZtAT9wd2vEvjGrR9n+Ht385WAmazsV7A0AaL3HPSRQMGbqgI2kXFxc0JjMPb3TxPcjnduq2PJ31mb/zwNK6/fbFIm27N69GwBQWVmJH3/8kdRtKYOMtn4l9kQSCASIiYnB2LFjMW7cOFK3pQwy2vqV2BNJKBRCKpXixIkTVEjDCG39SizYPnz4MMaMGYPY2FjExMSw6W9osK1fdPeXKr92Z1CnSOJv3OeUXeUxWmWgWNslg621eXv2WWVlH0gaOLanLU1Vlr1eLuHYZjpYAlB+Xtu/x3tyyi4tzlW9Zfs37gJA0TOvDejUj7Lt5ara6rrOCwDkk30A9GNWW01ER0dDIpEgJiYGpaWlpG5LGWS09SsxIb355puIjo4GwzBwdHQkdVvKIKOtX4kJydzcHKampkqPKaXoL9r6lWiMJBaLERwczJncq6+vh1wuZ5OZUoY2zc3NEIlEMDc3Z23K/Nqdfg+25XI5ZDIZPedWT1AoFDA0NNQ5T3q/C4ny3wF9TFCIQIVEIYJBZGRk5EA0JBaLcevWLSQlJWHu3LkQiUQQi8Vwc3PDzp07MX36dPZ3WSwWw9PTEwcOHICnpycEAgFbPyMjA3Pnzu1x3/z8fEyYMAFCoRCpqam4evUqUlJS4O3tDT6fjzNnzmDbtm2ora2Fra0tLCws2Prl5eW4ceMGrKysYGFhgfPnz+Ps2bPIyMiAl1f7Bsm8vDwkJCTg7NmzmDVrFoRCIceelpYGb29vGBgYQCwWo6CgAMnJyfDy8mLLhoeHw8rKipNjKioqCgsWLMDly5dRUlICJycnAO05iTZu3AipVMrudE5OToZIJOLs6IiIiICRkRGioqLg4eHB+Vwd38vUqVPB5/Oxe/dueHt7o7KyEidOnIC7uzuio6Ph4uKC+Ph4WFlZwdLSslf+HdAnkr+/Pzw8PFBR0ZkiZteuXQgICICZmZlW9R0dHVFYWNjDHhAQwJ61+uTJE8jlcgQFBbEpfZcsWQJ3d3f4+/v32AnRtT8AkJOTg+DgYNjY2LBpD9PT0xEaGorw8HDOma7nzp1DaGgo1q5dyxlQdD+6U9UpQ6ryWGubSlHdvrOO76VjVqFjAragoICdNyP1/m9AhRQTE4Pi4mJMnDiRtTk5OeHChQs9ysbGxiI3N1fpfbrP9yQmJuLYsWPsta+vL5YuXYpvv/22h0iUMXfuXFy50rlLRKFQsH92ndphGAYtLS2Qy+WcvtTU1ODw4cOcDMBHjx5Fa2srm1MoLS0N9fX17ClDHbz00ks4ffp0n/KTjxkzBvPnz+9h7748VtkELKn3fwMqpKCgIAiFQjx82JkIys/PD62trSgo4CanCggIgKcnd34rLi4OxcXFcHV15dhbW1vR1NSE6ur2tMJJSUk4f/48zMzMOO9DVOHh4cHWBYA5c+Zg7969qKurg61te5KqpUuXYv/+/Thw4ABHCC+++CKOHj0KAJwn0ooVKzhp8qqrq/H2228jMDCQcxSDjY0NCgsLsXDhQk6fioqKkJeXx/kPAgCnTp1CbGws58Tz5cuXK81dFRAQgICAAPZaJBLB1dUVEokEdnZ2rF0gELA/v72FDv9VkJeXh5SUFISEhGDEiBGD3Z0hDxUShQh0+E8hAhUShQhUSBQiDHshyWQySCQSzQX7iY73UMOdYSukmpoaHD16FKmpqSgoKEBSUlKvnXry5Ens3buXY3v48CESExMhk8nw+eefY9OmTQgODu5x8M/OnTt71ebXX3+tsYxYLEZpaSmOHDmCpqamXrVDCr06oU8sFrOL01euXIk7d+7gwoULkMvl+PTTT/HDDz+gvr4eRkZGWLVqFe7fv49Hjx5BJBLBysoKzc3NiI6OxuPHj8EwDCIiIuDr64tFixahpqYGymaLcnJyUFZWxr6k7OCHH37Ae++9h59++gmLFi3C/PnzUVtbi9zcXIwfPx4AcP36dRQWFqK4uBhxcXFQKBQYPXo0pk2bhoMHD+KFF17A6dOnMX36dNy9exeurq64e/cuduzYgbt37yI5ORnZ2dkwMjLC7Nmz4ejoiCNHjqChoYHzfui5555DQkICgoKC+u2714TePZFWrVqFkJAQJCQk4JdffsHWrVsRGBiIkydP4v79+xg7diyWLVvGlp8xYwYWL14MoH3RVnV1NT777DM8/fTTKCgogL29PUJCQiCVSpW2N336dCxfvryHvby8HGZmZiguLmZfnFpbW+O5555jy8ycOROurq4oKytDVVUVTExMkJeXBwBYsGABVqxYAVNTU6xbtw6tra1Ys2YNeDweR7Q+Pj745JNPcPHiRYwYMQKvvPIK3N3dkZXVucDf0dERN29ycz0NNHonJIZhIJPJeuToUSgUCAwMhKurK7Zu3co6o+t0CsMwnGuFQsGu3Ow6PVFYWIiwsDAkJCSo7EdH+87OzuxUTm1tLb777jvs2bMHYWFhbGzW1tYGb29vfPzxx6yoO3ZpdMzbdbyZ5vP5HCEZGxtDKBSCYRj8/PPPKC8vx+TJk3vMrQ328a969dMGAAcPHgQArFmzBrdu3UJERATa2trw0Ucf4dChQ2hsbISzszM7XWFvb4/4+HiMHz8exsbGsLW1xbZt28Dj8TBlyhSlbUyaNIndqqyKUaNGob6+HitXrkRkZCROnjyJxsZGrF+/nv1pA9pjqdGjR+P06dPsZOmoUaN69dnt7Oxw9epVCIVCGBsbsxPdxcXFmDp1aq/uSYy+J0IZOPbt28eUlJQMdjcYhmGYx48fMwkJCYPdDYZhGObIkSOMTCYb1D7QKRIKEfQuRqIMTaiQKESgQqIQgQqJQgQqJAoRqJAoRKBCohCBColCBCokChGokChEoEKiEIEKiUIEKiQKEaiQKESgQqIQgQqJQgQqJAoRqJAoRKBCohCBColCBCokChGokChEoEKiEIEKiUIEKiQKEaiQKESgQqIQgQqJQgQqJAoRqJAoRKBCohCBColCBCokChGokChEoEKiEIEKiUIEKiQKEaiQKESgQqIQgQqJQgQqJAoRqJAoRKBCohCBColCBCokChGokChEoEKiEIEKiUIEKiQKEaiQKESgQqIQgQqJQgQqJAoRqJAoRKBCohCBColCBCokChGokChEoEKiEIEKiUIEKiQKEaiQKESgQqIQgQqJQgQqJAoRBIPVsNXzERCZWUFkZgWhiSmMzYxgZCKEoYkQRiZCWJiKYG0qgq25IaxNRbA0FsLCSAALQwFGGAphZmgAEwEfxgI+jAQ8GAn4MEQrePIG8GQN4MsbwZM1QCF9DEW9BAppHRRSCVobGiCTNKD58RPIJVLI6hrQLGlGs0SGZkkzGmub8aS5FY/lbXjcooCkpQ07W4oH62vSG+gTiUIEKiQKEaiQKETgMQzDDEbDNTU1GDlypF7ct7/6OpygTyQKEQbtiUQZXtAnEoUIVEgUIlAhUYgw4G+2U1NTUVRUhH/84x8AgJKSEpw4cQISiQQbN26EoaGhTvdLT09HZWUlWlpaEBQUhOPHj6OxsRGGhoZ47bXXet3PhoYGREREYM2aNXB2dkZiYiLkcjmamprw7rvv9vq+w5UBeSLFxcUhLCwM69atw+jRo9HS0sL+m4GBAR48eIC6ujoIhUKd752dnY3XX38dtbW1AIA7d+7g9ddfR1lZWZ/6XFNTg2nTprHXDQ0NqKioAJ9PH+LKGJBvZfXq1di9ezf27NkDDw8Pzr9lZWUhJCQEPj4+yM/P1/neAkH7Q1WhUCi97i1OTk4YMWIEe/3kyRNs2LABlZWVfbrvcGXQJm0BIC0tDU5OToiPj4dcLsezzz6r8z2mTZuGmJgYGBkZIT8/Hw4ODoiNjcW4ceOI9PHRo0doamqCUCjEoUOH8PTTTxO573CDvkeiEIH+4FOIQIVEIQIVEoUIVEgUIlAhUYhAhUQhAhXS/0cmk0EikQx2N/SWARdScnIycnNztS7PMAy++eYbtWV+//13JCYmqrzWhtTUVBQUFCApKQlVVVWcf8vNzcWvv/6q0/3+2xiUN9stLS3YsmULhEIhnnnmGfB4PGRkZOD333/HW2+9hb/97W9s2QsXLrBzXjk5OYiLi4NCoUBISAhqa2uRnJwMqVSKBQsWICsri72eNm0aIiIi0NDQgKioKIhEIgCAWCzG/fv34eDggIqKCtTW1uLjjz/G9evXIRKJYGVlhebm5h5t/fzzz5g3b95gfF16waAI6euvv8b+/fthYmKCiIgImJub4/3338dTTz2FKVOmIDMzE/n5+XjmmWfw22+/ISQkBACQkJCArVu34smTJzhw4AAaGhqwefNm/Prrr6iursaFdfWcAAABoElEQVSxY8fY66KiIlhaWmLZsmUwMDDgtP/mm29CJBLh0aNHSE9Px82bNzFjxgzY2dnh+vXrStvqKG9jYzPg35c+MCgxUltbG7rOzCxbtgw7d+4En8/HhAkT4O3tjeeffx5SqRRA50Rs1zoMw4DH4wEAK5Su10KhEC+//DLi4+Nx+/ZtTvsmJiaIi4tDXV0dXF1dOffqev+uf+fz+T3KUDoZlCfS+vXrsWfPHggEAixatAh1dXVobW1FfX09KisrYWNjg8uXLyMwMBBSqRTZ2dlYuHAhVqxYgYiICPB4PLz//vuora1FZGQkmpqa4OXlBT8/P/Z63LhxOHLkCPh8Puzs7Hr0wdbWFleuXIFEIoGbmxsmT56M+Ph4jB8/HgB6tBUTEwNra+uB/qr0hiExaRsZGYkNGzagqakJ33zzDQwMDDBq1CjMmDED06dPx3fffYf33ntv0PqXk5ODtrY2zJo1a9D6MNQZEkISi8UQCARobGzEwoULqcP0kCEhJIr+Q19IUohAhUQhAhUShQhUSBQiUCFRiECFRCECFRKFCP8Hu4n4V4XR1PgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 108x180 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plot_seq_pair_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus,\n",
    "                                    statistical_test, plotTitle=None, subseq_type='aa', refCodonTableDf=refCodonTableDf,\n",
    "                                    vmin1=-1.8, vmax1=1.8, width=1.5)\n",
    "filenamePlot = os.path.join(plotsFolder, speciesSetName, outputFilePrefix +\n",
    "                            \"_composition_bias_aa_pairs_\" + terminus + \"terminal\" + \"_CtermNorm\")\n",
    "print(filenamePlot)\n",
    "fig.savefig(filenamePlot + '.svg', dpi=800, bbox_inches=\"tight\")\n",
    "fig.savefig(filenamePlot + '.png', dpi=800, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCG</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">N</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">-1000000</th>\n",
       "      <th>count</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>221</td>\n",
       "      <td>0</td>\n",
       "      <td>2653</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>-10.8466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.55548</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.66342</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>0.000543054</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0850537</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.33531</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <th>-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           AAA    AAC   AAG  \\\n",
       "terminus position from terminus observable                                    \n",
       "N        -1000000               count                        2      0     0   \n",
       "                                log2OddsRatio         -10.8466    NaN   NaN   \n",
       "                                oddsRatio          0.000543054    NaN   NaN   \n",
       "                                pvalue                       0    NaN   NaN   \n",
       "NaN      -20                    BH_multiple_tests         True  False  True   \n",
       "\n",
       "                                                    AAT    ACA    ACC    ACG  \\\n",
       "terminus position from terminus observable                                     \n",
       "N        -1000000               count                 0      0      0      0   \n",
       "                                log2OddsRatio       NaN    NaN    NaN    NaN   \n",
       "                                oddsRatio           NaN    NaN    NaN    NaN   \n",
       "                                pvalue              NaN    NaN    NaN    NaN   \n",
       "NaN      -20                    BH_multiple_tests  True  False  False  False   \n",
       "\n",
       "                                                     ACT    AGA    AGC  ...   \\\n",
       "terminus position from terminus observable                              ...    \n",
       "N        -1000000               count                  0      0      0  ...    \n",
       "                                log2OddsRatio        NaN    NaN    NaN  ...    \n",
       "                                oddsRatio            NaN    NaN    NaN  ...    \n",
       "                                pvalue               NaN    NaN    NaN  ...    \n",
       "NaN      -20                    BH_multiple_tests  False  False  False  ...    \n",
       "\n",
       "                                                     TCG    TCT    TGA   TGC  \\\n",
       "terminus position from terminus observable                                     \n",
       "N        -1000000               count                  0      0      0     0   \n",
       "                                log2OddsRatio        NaN    NaN    NaN   NaN   \n",
       "                                oddsRatio            NaN    NaN    NaN   NaN   \n",
       "                                pvalue               NaN    NaN    NaN   NaN   \n",
       "NaN      -20                    BH_multiple_tests  False  False  False  True   \n",
       "\n",
       "                                                     TGG   TGT        TTA  \\\n",
       "terminus position from terminus observable                                  \n",
       "N        -1000000               count                  0     0        221   \n",
       "                                log2OddsRatio        NaN   NaN   -3.55548   \n",
       "                                oddsRatio            NaN   NaN  0.0850537   \n",
       "                                pvalue               NaN   NaN          0   \n",
       "NaN      -20                    BH_multiple_tests  False  True      False   \n",
       "\n",
       "                                                     TTC      TTG   TTT  \n",
       "terminus position from terminus observable                               \n",
       "N        -1000000               count                  0     2653     0  \n",
       "                                log2OddsRatio        NaN  2.66342   NaN  \n",
       "                                oddsRatio            NaN  6.33531   NaN  \n",
       "                                pvalue               NaN        0   NaN  \n",
       "NaN      -20                    BH_multiple_tests  False    False  True  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCodon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>9074</td>\n",
       "      <td>1618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.64056</td>\n",
       "      <td>1.26763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>3.11786</td>\n",
       "      <td>2.40766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0</td>\n",
       "      <td>1.78999e-203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   AAA           AAG\n",
       "observable                          \n",
       "count             9074          1618\n",
       "log2OddsRatio  1.64056       1.26763\n",
       "oddsRatio      3.11786       2.40766\n",
       "pvalue               0  1.78999e-203"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCodon.loc['C'].loc[-1][['AAA', 'AAG']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = compute_codon_oddsratio_plot_data(oddsRatioDfCodon, terminus)\n",
    "maskData = compute_codon_oddsratio_mask_data(oddsRatioDfCodon, pvalueThresholds, terminus)\n",
    "pvalAnnotTable = compute_codon_pvalueAnnotation(oddsRatioDfCodon, pvalueThresholds, terminus)\n",
    "maskDataSmallN = compute_codon_smallN_mask_data(oddsRatioDfCodon, pvalueThresholds, terminus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>position from terminus</th>\n",
       "      <th>-20</th>\n",
       "      <th>-19</th>\n",
       "      <th>-18</th>\n",
       "      <th>-17</th>\n",
       "      <th>-16</th>\n",
       "      <th>-15</th>\n",
       "      <th>-14</th>\n",
       "      <th>-13</th>\n",
       "      <th>-12</th>\n",
       "      <th>-11</th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>0.192697</td>\n",
       "      <td>0.250744</td>\n",
       "      <td>0.248346</td>\n",
       "      <td>0.263872</td>\n",
       "      <td>0.319666</td>\n",
       "      <td>0.321202</td>\n",
       "      <td>0.337629</td>\n",
       "      <td>0.341049</td>\n",
       "      <td>0.390464</td>\n",
       "      <td>0.335346</td>\n",
       "      <td>0.355414</td>\n",
       "      <td>0.467106</td>\n",
       "      <td>0.549823</td>\n",
       "      <td>0.508756</td>\n",
       "      <td>0.578371</td>\n",
       "      <td>0.653820</td>\n",
       "      <td>0.804875</td>\n",
       "      <td>0.639681</td>\n",
       "      <td>0.782256</td>\n",
       "      <td>1.640557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>-0.004471</td>\n",
       "      <td>-0.065445</td>\n",
       "      <td>0.063981</td>\n",
       "      <td>-0.025216</td>\n",
       "      <td>-0.035698</td>\n",
       "      <td>0.007840</td>\n",
       "      <td>-0.149319</td>\n",
       "      <td>-0.012734</td>\n",
       "      <td>0.005795</td>\n",
       "      <td>-0.004471</td>\n",
       "      <td>-0.002412</td>\n",
       "      <td>-0.010664</td>\n",
       "      <td>-0.129023</td>\n",
       "      <td>-0.122319</td>\n",
       "      <td>-0.095799</td>\n",
       "      <td>0.089347</td>\n",
       "      <td>0.201149</td>\n",
       "      <td>0.188528</td>\n",
       "      <td>0.375559</td>\n",
       "      <td>0.781252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>0.296917</td>\n",
       "      <td>0.302157</td>\n",
       "      <td>0.274000</td>\n",
       "      <td>0.163437</td>\n",
       "      <td>0.239863</td>\n",
       "      <td>0.270443</td>\n",
       "      <td>0.286384</td>\n",
       "      <td>0.366986</td>\n",
       "      <td>0.395107</td>\n",
       "      <td>0.430737</td>\n",
       "      <td>0.517804</td>\n",
       "      <td>0.602985</td>\n",
       "      <td>0.664554</td>\n",
       "      <td>0.706480</td>\n",
       "      <td>0.683631</td>\n",
       "      <td>0.960279</td>\n",
       "      <td>1.237702</td>\n",
       "      <td>1.048111</td>\n",
       "      <td>1.107773</td>\n",
       "      <td>1.267629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAT</th>\n",
       "      <td>-0.092366</td>\n",
       "      <td>-0.036811</td>\n",
       "      <td>-0.021465</td>\n",
       "      <td>-0.048418</td>\n",
       "      <td>-0.081762</td>\n",
       "      <td>-0.064023</td>\n",
       "      <td>-0.021465</td>\n",
       "      <td>-0.107061</td>\n",
       "      <td>-0.145772</td>\n",
       "      <td>-0.089045</td>\n",
       "      <td>-0.130040</td>\n",
       "      <td>-0.187553</td>\n",
       "      <td>-0.149212</td>\n",
       "      <td>-0.138229</td>\n",
       "      <td>-0.178408</td>\n",
       "      <td>-0.017650</td>\n",
       "      <td>0.115787</td>\n",
       "      <td>0.181735</td>\n",
       "      <td>0.590263</td>\n",
       "      <td>0.417231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>-0.039148</td>\n",
       "      <td>-0.097845</td>\n",
       "      <td>-0.086563</td>\n",
       "      <td>-0.212079</td>\n",
       "      <td>-0.054788</td>\n",
       "      <td>-0.078556</td>\n",
       "      <td>-0.148848</td>\n",
       "      <td>-0.054788</td>\n",
       "      <td>-0.172461</td>\n",
       "      <td>-0.005299</td>\n",
       "      <td>-0.153877</td>\n",
       "      <td>-0.138841</td>\n",
       "      <td>-0.243829</td>\n",
       "      <td>-0.270812</td>\n",
       "      <td>-0.231402</td>\n",
       "      <td>-0.127249</td>\n",
       "      <td>-0.238491</td>\n",
       "      <td>-0.157239</td>\n",
       "      <td>-0.186123</td>\n",
       "      <td>-2.038914</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "position from terminus       -20       -19       -18       -17       -16  \\\n",
       "AAA                     0.192697  0.250744  0.248346  0.263872  0.319666   \n",
       "AAC                    -0.004471 -0.065445  0.063981 -0.025216 -0.035698   \n",
       "AAG                     0.296917  0.302157  0.274000  0.163437  0.239863   \n",
       "AAT                    -0.092366 -0.036811 -0.021465 -0.048418 -0.081762   \n",
       "ACA                    -0.039148 -0.097845 -0.086563 -0.212079 -0.054788   \n",
       "\n",
       "position from terminus       -15       -14       -13       -12       -11  \\\n",
       "AAA                     0.321202  0.337629  0.341049  0.390464  0.335346   \n",
       "AAC                     0.007840 -0.149319 -0.012734  0.005795 -0.004471   \n",
       "AAG                     0.270443  0.286384  0.366986  0.395107  0.430737   \n",
       "AAT                    -0.064023 -0.021465 -0.107061 -0.145772 -0.089045   \n",
       "ACA                    -0.078556 -0.148848 -0.054788 -0.172461 -0.005299   \n",
       "\n",
       "position from terminus       -10       -9        -8        -7        -6   \\\n",
       "AAA                     0.355414  0.467106  0.549823  0.508756  0.578371   \n",
       "AAC                    -0.002412 -0.010664 -0.129023 -0.122319 -0.095799   \n",
       "AAG                     0.517804  0.602985  0.664554  0.706480  0.683631   \n",
       "AAT                    -0.130040 -0.187553 -0.149212 -0.138229 -0.178408   \n",
       "ACA                    -0.153877 -0.138841 -0.243829 -0.270812 -0.231402   \n",
       "\n",
       "position from terminus       -5        -4        -3        -2        -1   \n",
       "AAA                     0.653820  0.804875  0.639681  0.782256  1.640557  \n",
       "AAC                     0.089347  0.201149  0.188528  0.375559  0.781252  \n",
       "AAG                     0.960279  1.237702  1.048111  1.107773  1.267629  \n",
       "AAT                    -0.017650  0.115787  0.181735  0.590263  0.417231  \n",
       "ACA                    -0.127249 -0.238491 -0.157239 -0.186123 -2.038914  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>position from terminus</th>\n",
       "      <th>-20</th>\n",
       "      <th>-19</th>\n",
       "      <th>-18</th>\n",
       "      <th>-17</th>\n",
       "      <th>-16</th>\n",
       "      <th>-15</th>\n",
       "      <th>-14</th>\n",
       "      <th>-13</th>\n",
       "      <th>-12</th>\n",
       "      <th>-11</th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "position from terminus    -20    -19    -18    -17    -16    -15    -14  \\\n",
       "AAA                     False  False  False  False  False  False  False   \n",
       "AAC                     False  False  False  False  False  False  False   \n",
       "AAG                     False  False  False  False  False  False  False   \n",
       "AAT                     False  False  False  False  False  False  False   \n",
       "ACA                     False  False  False  False  False  False  False   \n",
       "\n",
       "position from terminus    -13    -12    -11    -10    -9     -8     -7   \\\n",
       "AAA                     False  False  False  False  False  False  False   \n",
       "AAC                     False  False  False  False  False  False  False   \n",
       "AAG                     False  False  False  False  False  False  False   \n",
       "AAT                     False  False  False  False  False  False  False   \n",
       "ACA                     False  False  False  False  False  False  False   \n",
       "\n",
       "position from terminus    -6     -5     -4     -3     -2     -1   \n",
       "AAA                     False  False  False  False  False  False  \n",
       "AAC                     False  False  False  False  False  False  \n",
       "AAG                     False  False  False  False  False  False  \n",
       "AAT                     False  False  False  False  False  False  \n",
       "ACA                     False  False  False  False  False  False  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskData = oddsRatioDfCodon.copy()\n",
    "maskData = maskData.xs(terminus, level='terminus')\n",
    "maskData = select_terminal_positions_in_index(maskData, terminus)\n",
    "# plotData = plotData.xs('BH_multiple_tests', level='observable')\n",
    "maskData = maskData.xs('log2OddsRatio', level='observable')\n",
    "maskData.iloc[:, :] = True\n",
    "# Mask both rejected cases and null values\n",
    "maskData = (maskData != True) & (maskData != 'True')\n",
    "maskData = maskData.transpose()\n",
    "maskData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>position from terminus</th>\n",
       "      <th>-20</th>\n",
       "      <th>-19</th>\n",
       "      <th>-18</th>\n",
       "      <th>-17</th>\n",
       "      <th>-16</th>\n",
       "      <th>-15</th>\n",
       "      <th>-14</th>\n",
       "      <th>-13</th>\n",
       "      <th>-12</th>\n",
       "      <th>-11</th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "position from terminus    -20    -19    -18    -17    -16    -15    -14  \\\n",
       "AAA                     False  False  False  False  False  False  False   \n",
       "AAC                     False  False  False  False  False  False  False   \n",
       "AAG                     False  False  False  False  False  False  False   \n",
       "AAT                     False  False  False  False  False  False  False   \n",
       "ACA                     False  False  False  False  False  False  False   \n",
       "\n",
       "position from terminus    -13    -12    -11    -10    -9     -8     -7   \\\n",
       "AAA                     False  False  False  False  False  False  False   \n",
       "AAC                     False  False  False  False  False  False  False   \n",
       "AAG                     False  False  False  False  False  False  False   \n",
       "AAT                     False  False  False  False  False  False  False   \n",
       "ACA                     False  False  False  False  False  False  False   \n",
       "\n",
       "position from terminus    -6     -5     -4     -3     -2     -1   \n",
       "AAA                     False  False  False  False  False  False  \n",
       "AAC                     False  False  False  False  False  False  \n",
       "AAG                     False  False  False  False  False  False  \n",
       "AAT                     False  False  False  False  False  False  \n",
       "ACA                     False  False  False  False  False  False  "
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskDataSmallN.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def sort_codon_index(df, refCodonTableDf, addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "#                      addEmptyRowBetweenGroups=False, addEmptyRowBetweenGroupsFillValue=np.nan, useULetter=True):\n",
    "#     \"\"\"Sort the codon index of the dataframe and group them by amino acid.\"\"\"\n",
    "    \n",
    "#     aaTable = refCodonTableDf.sort_values('aa_groups_sorted_index')['aa'].unique()\n",
    "#     # Sort codons following order defined in the reference codon table, grouping synomymous codons\n",
    "#     sort_order = {codon:i for i, codon in\n",
    "#                   enumerate(refCodonTableDf.sort_values(['aa_groups_sorted_index', 'alphabetical_sorted_index']).index)}\n",
    "#     if len(df.index.unique().difference(refCodonTableDf.index)) > 0:\n",
    "#         raise ValueError(\"DataFrame index contains a codon not found in the reference codon table index.\")\n",
    "    \n",
    "#     plotDf = df\n",
    "#     plotDf['sort_index'] = plotDf.index.get_level_values(0).map(lambda x: sort_order.get(x))\n",
    "#     plotDf = plotDf.sort_values('sort_index').drop('sort_index', axis=1)\n",
    "#     plotDf.index.name = 'codon'\n",
    "    \n",
    "#     aaDict = refCodonTableDf['aa'].to_dict()\n",
    "#     plotDf['aa'] = plotDf.index.map(lambda x: aaDict[x])\n",
    "#     if addAminoAcidLetterToIndex:\n",
    "#         plotDf.index = plotDf.index.map(lambda x: '{}/{}'.format(refCodonTableDf.loc[x, 'aa'], x))\n",
    "#     if useULetter:\n",
    "#         plotDf.index = plotDf.index.map(lambda x: x[:-3] + re.sub('T', 'U', x[-3:]))\n",
    "#     if addEmptyRowBetweenGroups:\n",
    "#         dfDict = {}\n",
    "#         for aa, group in plotDf.groupby('aa'):\n",
    "#             df = group.copy().append(pd.Series(addEmptyRowBetweenGroupsFillValue, index=group.columns, name=''))\n",
    "#             dfDict[aa] = df\n",
    "#         plotDf = pd.concat([dfDict[aa] for aa in aaTable if aa in dfDict.keys()])\n",
    "#     if not addAminoAcidLetterAsColumn:\n",
    "#         plotDf = plotDf.drop('aa', axis=1)\n",
    "\n",
    "#     return plotDf\n",
    "\n",
    "# df = oddsRatioDfCodon.T.iloc[:, 0:1].copy()\n",
    "# sort_codon_index(df, refCodonTableDf, addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "#                      addEmptyRowBetweenGroups=False, addEmptyRowBetweenGroupsFillValue=np.nan, useULetter=True).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(cterminal)\n",
    "from cterminal import plot_codon_composition_map\n",
    "reload(mwTools.bio)\n",
    "from mwTools.bio import sort_codon_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.0"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FontProperties(size='medium').get_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Research_cloud/Python_mwTools/mwTools/bio.py:847: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  plotDf['sort_index'] = plotDf.index.get_level_values(0).map(lambda x: sort_order.get(x))\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQQAAAPVCAYAAACA9ZSpAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzs3Xt4lNW9N/xvSCZDJueEBAiEU9CAVYLlUC+0p8fWAqIvWovF2qBppXkKYQPGFHrgcfu6de9ro1KDoPKUJ48HVLTdVtB3p00kSGutchCUkGiCIDEhCZAhB8hMJjPvHzezzJDkXitwr7kDfj/Xta+9t/PjPsywFmvd67d+d0QgEAiAiAjAELsvgIgGD3YIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYgEdghEJETZfQF1Le3SmNHJcfA11EjjokZOhPdknWlMdOpoAEDXiWOmcY5hmQAgPW/UyIkAgLNnz5rGxcTEwPdFlWkMAESNmgSvu0kaF52UDm/LcfOY5BEAgM6zZ0zjhsa4AABnJPfgiokBAHg62kzjnLHx8LS1mMYAgDM+WXrO4HlbO8zjEmKNa6tubDWNyx6eAAA4dsr8711mShwAYE+d2zRu2ugk1Dabfx8AkJUWL/3eAOO7sxNHCEQksEMgIkE6ZSguLkZ8fDxqamqwevVqHD9+HAcPHsS4ceNQUVGB1tZWrFy5En/4wx/gcDiQmZmJQ4cOobCwEMXFxVi8eDGcTmc47oXosqe7PSqNEBYuXIicnBzU19dj+/btuOWWW1BeXo6CggIUFRXB4/HA5XIhPz8fM2fOtOzmiag3ne1R6aFiSUkJ3G43hg0bhm9/+9uIiIgAAAQCAXR1dcHv98Pv96O8vBxHjx4Vn/v9fgwZwlkJkZV0tkel1nrPPffA4XDg4MGDmDp1KgBgzpw52LBhAzZu3Ain04nu7m40NTUhEAggOzsbJSUl8Hg8cDgcF3PvRHQene0xwu6KSVx2PO94XHbs87xcdgwP2zsEIho8OMEnIsH2TEVP+2lpjDMuEV3NR6VxjrSxysNjlSE+AHQdrzU/54gsI67xM/O44ePVpwIWxUUnpQOA8nBb9V5Vphaq96A6tVD9XU+1mcelxBtxTac7TOPSE2MBAPUt5nEZybFwt5ufEwCS4lzK92onjhCISNCSmJSZmRnymcvlCse9EF32dLdHpSnDwoUL8cYbb6C+vh5vvfUWli1bhnXr1mH58uXo6upCR0cHXC4Xfvazn6G5uRkvvPCC+Cw6OtqyL4OI9LZHLYlJPT8DwE6ByEI626OWxKSen3m93ou6eSIKpbM92p6HwFWGUFxl6I2rDOHDVQYiEmwfIRDR4MERAhEJtmcqyua3gDHHVd0YorLxBlDf3NR+xvz64lzn5t+SZxyOtLHKz0Fk1xa8PtmcNDgfVb1X76l607jolAwAavNqX/0npjEAEJVxpfK9ql6b6mY01WcS3tMnzM+bOEz52UD3kX3SuMhx10pjdOIIgYgELZmKN998M4qKirBgwQJMnz49HPdB9JWguz1qKaHW2NiIGTNmYO/evZZ8CUT0JZ3tUUumYlRUFHw+H6qrq+H1epmpSGQhne1RS6ZiU1MT8vLykJubi507d1rwFRBRkM72aHseAlcZ+ojjKkOv6+MqQ3hwlYGIBNtHCEQ0eHCEQESC7ZmKDW7z+SgAjEyKlc7lAWM+XyMpiT0xzXiG8NkJ87jxw849a1DYxQgAn580L+s9JjVOusMSMHZZqs5JPa0nzWMSUgGoPwdRLesu+81GJsVi+yHzYwHAvMkjUCUpmw4Ak4YnKD8vUbk2ADgp2RWZem5XpMqOUtlvDxi/v2yHJfDlLku7cIRARILWmoputxurVq1iHgKRRXS3Ry01FZ9//nmsWLECLS0tfLcjkcV0tkel1lpSUoLDhw+jrq5Omhm1bds2REREoLm5GZs3b0ZDQ8PFfwNEJOhsj1oyFWfPno2tW7caJ+AIgchSOtuj7XkIXGUIxVWG3rjKED7855uIBNtHCEQ0eNiemKRaht17sk4aF506Wno8Z1yicV7FIajqMFplA5Hq5ibZOYPnVdl4A6hvWpJ9x9GpowEo3qtFm5aC1+f7oso0JmrUJOPaFDaZDShOoTR99+cfmcYAQOSYa+Crq5TGRY2+ShqjE6cMRCSwQyAiQdohFBcXo6SkBL/97W/R0dGB2tpavPHGGwCAoqIi7N69W8Q9/fTTePPNN3t9RkTW0N0eL7imYs86bS0tLaypSBQmOtvjBddULC0tFXXaWFORKHx0tscLzlTsWadt7969rKlIFCY626PteQhcdjwvjsuOfV4flx3Dg6sMRCTYPkIgosHD9kxF1XLdZxQ2BrliYpTLa6sMBQH1oaXKFER1yKg6jFa9B1n572Dp7+6j+83jxuYAAHz7/2IaF5VzE7o/+btpDABEXnk9zm5/ShoXM28JTm38lWlMyv/8DwDAmdf+0zTOdccDAIDjjy4xjRux2riuE08WmsYNW7YWVT+/zTQGACb97/9C6//5nTQu4d7/VxqjE6cMRCSwQyAi4YJrKt56660hb5QtLi5mTUUizXS3xwuuqdgz+ykrK4s1FYnCRGd7vOCaiqWlpWhra+uVGcWaikR66WyPWjIVWVORSB+d7VE6ZSgoKAAArFy5MuS/FxYayzE5OcZS1Pe///2QzydPnqxyb0Q0ALrbI//5JiKBmYpEJNieqai8kUdxc5OVG3QA9TLsKnGyzELAyC5UjZNleUZlXGlcm2JGo6+hxvx4IycacccOmsdlfg3dn+0xjQGAyPHTpNmRgJEh6T9sXtxjyITpACDNkIy88noAQNf7fzaNc8z8f5SP17V7u2kMADimz8PRVfdI48b+e4k0RidOGYhIYIdARIJSTcUzZ87g0UcfRXt7e7813Pbs2SOedJ7/GRFZQ3d7VBohPP7447j77rsRFxfXZw03AHj//fcxa9YsNDY2sqYikUY626PSQ8Vx48Zhx44dyMzM7LOGm9frRU1NDZxOJ0pLSwGANRWJNNHZHpVGCD/60Y/Q2dmJysrKPjOjduzYgUWLFiEvLw9NTU2sqUikkc72aHseApcdz4vjsmPvOC47ho3tHQIRDR5cdiQiwfZMRdXS5MpxilMB5RLmVpYmV7wH2dAdMIbvqlMGWS3HYOlv1dqL/tr3TeOGZM1Ed9Uu0xgAiJz0TXR/XC6Pu/pGdFdWmMdc9R0AgOft50zjnP8jFwDQvuVh07i4u34LAOh862nTuKFz83H29XWmMQAQM3852v7v/5LGxS/6V2mMThwhEJGgJTHpwIEDePLJJ/HQQw/B6/VqvHyirxbd7VFLYlJZWRkKCgqwdOlSFkghspjO9qjUWoOJEDt27OizZFMwEcLtdqO0tJQl1Ig00tketSQmsYQakT4626NyCbXFixeH/PeeJZuCZZt6/neWUCOynu72yH++iUhgpiIRCRwhEJFge6aiu938bc0AkBTnQmuH/O3PCbEx6DzTbhoz1BUHANK3SbtiYgAAza3m15eWYLxN2tN+2jTOGZeI0wr3kBgbg5Nt8u8kNd6FxtMdpjHDE2MBAMdOmX8nmSnGd9LgNj/eyCTjeLL7SIyNkZ4zeF7ZOYPn/fyk+fHGpBr3UN3YahqXPTwBAFDXYn680cnq34nsjeOA8dZx2bX1vD67cIRARIJSpqLH4xH/O5gZFcx+evjhh3HmzJmQzKi1a9eG/Fkisobu9jjgKcP27duxbNkyrFu3DsuXL0dXVxeio6NDMqOIKDysbo9KHcILL7yADz/8EO+++67IjAKAQCCArq4uAAgp2RT83O/3MzGJyGI626NSh3D33Xfj9OnTqKysxJIlSwAAc+bMwYYNG9DZ2YmsrCwsWrQIU6ZMwdq1azFp0iSUlJTA4/HA4XBc8I0TUW8626PteQhcZQjFVYa+z8tVhvDgeJ6IBNtHCEQ0eHCEQESC7ZmKXneTNCY6KV065weMeb9qDcQmyfw7PVF9vgwAnrYW0zhnfLLyswFPR5s0zhkbL32uknDu2lTjZOd1xsYDAE5J7iMl3iV99gIYz1/az8h/1ziX+rMhK+8BgNJzGtmxgseTPbcAvnx2YReOEIhI0JKpeP5nRGQN3e1RS6ZieXl5yGdEpIfV7VFLpuL5n7FTILKOzvaoJVNx3rx54rP8/Hx2CEQW0tkebc9D4CpDKK4y9MZVhvDhKgMRCbaPEIho8LA9MemswlQgJiZGeROU6pDR6iGobIPL0BgXPK0nTWMAwJmQqrxZRjbcjnMZUwHVONmUJnUAw+j6FvmmpYzkWOVhtOpmpE+bzH+vK9LjBxRX02weNzEtXhoTjKtViMtKi5fG6MQpAxEJ7BCISFDKVCwpKUFxcTEA9MqMCr5RljUVifTT3R6VRggLFy7E2LFjUVVVJd42e/4bZVlTkSg8dLbHAU0ZPvjgA5EZ1fONsseOHev1tlmANRWJdNLRHpVa65YtW/Dpp5+itbVVvG225xtlP/zww5C3zWZnZ7OmIpEmOtuj8tufzzd58uQ+3ygbnLcQkfV0t0eO54lIYKYiEQmXTKai6sYg7+kTpjHRicMAqJVNB9QyEAHA23Lc/LzJI5Q3LVmdqaiabam6CUqlhL3qpiXZ7wAYv4Xq76Baml51c5NK9qZqtmXXiWPSOMewTGmMTpwyEJHADoGIBOmUobi4GPHx8Whra0NBQQFqa2tx8OBBjBs3DhUVFXC73Vi1ahU++ugjvPTSS1i7di0OHDiAiooKtLa2YuXKlXC5XOG4F6LLnu72qPQMYeHChSgtLUVVVRVKS0uxbNkyPPHEE1ixYgVaWlp6ZUaxpiKRPjrbo5ZMReDLGm5er/fC75yI+qWjPWrJVAzWd9u4cSM7BCKL6WyPtuchcNnxvPNy2bH39XHZMWy4ykBEgu0jBCIaPGzPVJQNtQFjuC0rrw4YJdZVhu4A0NX4mWmcY/h4I675qHlc2lgjTjIcdAzLlB4reDyrvpNgyXnV70Q57lS9eVxKhvLvpTqMlpXrj05KV742YAD3qnBeWQl+wCjDr/rKATtxykBEAjsEIhKUMxWjoqJw++23o6GhQZoZBQBFRUVYsGABpk+frv0miL4qdLdH5UxFp9MJ4Mu3zZplRgHAjBkzsHfvXnYIRBbT2R6VOoRXX30VMTExSElJ6ZUZ9dxzz+GHP/xhr7fN+nw+VFdXw+v1Mn2ZyEI626PSMwSfz4eOjg5UVlYqZUY1NTUhLy8Pubm52Llz58XePxH1oLM92p6HwGXH3sfjsmMoLjuGj+0dAhENHlx2JCLB/kxFyWYkwNiQpDy0VBxGK08ZFKYCgNrUQvkeJMNewBj6+r6oMo2JGjUJgPow2lf/ifnxMq404hTOK4sJxnUf2SeNixx3LfyfvmcaM+SK6wAA3VW7zI816ZsAAO87L5nGRX9rIQDg7OvrTONi5i+H92+vmMYAQPQNd0qvref12YUjBCIStCQmBUs2BT/jsiORNXS3Ry2JSWVlZSGfEZF1dLZHpdb66quv4o9//CN27Nih/HLJ4GcNDQ0X/w0QkaCzPWpJTOr5GUcIRNbS2R61vey1r8+I6OLobo/855uIBGYqEpHAEQIRCbZnKrrb5SXHk+LkJccBo6y3SolwQL28+mlJafLEc6XJVUqYy0qEA0aZ8OZW+XeSluBCk+R46edKjte3mMdlJA8srrbZvKx7Vlq8NCYY99kJedz4YfE4dsq81HlmShwASEuij0424j5tMj/vFelGaXqV70T195L9HQG+LHVvF44QiEgw7RDeeustvPjii3jkkUfg8/lQW1uLN954Q5RlKi4uhsfjwfr167Fp0ya8/vrrvT4jIuvobpOmHUJrayu8Xi/uueceREVFYfv27bjllltCYk6fPg2Xy4X77rsPc+fOvZh7JSIJ3W3S9BnC7bffjlOnTmHjxo249tprQ7KiAMDv9yMqKgrd3d0AgI6OjpDPmJREZC3dbdK0Q3jttdcAAE6nE3V1dZg/fz4AIDs7GyUlJfB4PEhJSUEgEEBJSQlSUlJCPnM4HBd390QUQnebNO0Q7rrrrj7/+7x580L+/8WLFyvfEBFdON1tkmN6IhKYqUhEAkcIRCTYnqnoaT8tjXHGJSrHqWYqyjIkk+KMTMXOM+aZb0NdRuabp/Wk+bUlpEqvLXh9snMGz2v1PTS4zbPyRiYZmYoqWYOyDE/AyPJUzVQ91WYelxI/sHs9K/ktYs79PVHJVFUtw+5rqJHGRY2cKI3RiSMEIhJMRwhvvfUWWlpacPToURQVFeHo0aM4ePAgPvnkExQWFqK4uBiLFy/Gpk2b4HQ6kZaWhgkTJqCiogKtra1YuXIlXC5XuO6F6LKnu01anqlYXl6OgoICFBUVsTMgspjuNml5piIABAIBdHV1AQArLhNZSHebtDxTcc6cOdiwYQM6OzuRn5/PDoHIQrrbpO15CFxl6H19XGUIxVWG8OEqAxEJto8QiGjwsD0xSTYUBIzhoOoQVFZqLc4Vo3Te4BBUtYSayvFkQ23AGG6rfieq5cJUpwI1krJnE9OMsmIHG1pN4742MkG5hJrsnMHzykqtjR9mXNvnJ82/kzGpAyuhpnK8jxrk09lrRiZi+6Hj0rh5k0dIY3TilIGIBHYIRCRYXlMRAIqKirB79279V0/0FaO7TVqeqdjY2IgZM2Zg7969F3rPRNQP3W3S8kzF0tJS+Hw+VFdXw+v1MjGJyEK626TlmYpNTU0oLCzE/v37sXPnTnz/+9+35IsgIv1tUltNxZycHPM7I6IB090mucpARAIzFYlIsD1TUTUDUTXLr7rRPIsue3gCAPVMRdVNMCovXlV92eueOrc0btroJOXsvXePmG+8mjUuFYB6pqLK8SqPm/8OAHDViATlTMUqye866dzvKvuOh597Aa7q3xOVjVz7vpD/XteOSsJbVY3SuLmThktjdOKUgYgEdghEJGirqeh2u7Fq1SrmIRBZSHebtDxTsaysDAUFBVi6dClf9kpkMd1t0vTT22+/HXPmzMEzzzyD119/XflNs83Nzdi8eTMaGhou5t6J6Dy626TlmYqzZ8/G1q1bAYAjBCKL6W6TWjIVJ0+ebHpSIrowutsk/wknIoGZikQk2J6pqFqaXFYrETDqJVpZXhsAWiVxCbHqZd1VS47Lzhk8r2qtxOZW8/OmJRhZmfUt5sfLSFYvw35EUosQAMalxilnoKrWj1S9V9XamyclGa2p8S5pCX7AKMPvPVUvjYtOyZDG6MQpAxEJ7BCISLigmopAaI224uJiPP3003jzzTd71XYjIuvobpMXlKnYs0ZbS0sLXC4X8vPzMXPmTCvumYj6obtNXlBNxZ412vx+P/x+P8rLy3H06NGQjCkmJhFZS3ebvKBMxZ412vbu3Yvu7m40NTUhEAiEZEw5HA4rvgMiOkd3m7Q9D4HLjr3juOzY+3hcdgwP2zsEIho8OMknIsH2TEVPu/zNuc64ROU41SmD9/QJ07joxGEA1KYCqnGeDnn9QGdsvPK9ql6bav1I1eOpTBlU32CtOj1SrZUoG5YHh+RdzUdN4xxpY5Xjuo/uN40BgMixOfDXvi+NG5Jl70odRwhEJLBDICLhgmoq3nrrrSgqKsKCBQswffp0FBcXw+FwIDMzE5mZmaioqEBraytWrlwJl8sVrnshuuzpbpOmHUJfWVHLli0LyYrKysqCy+XCz372MzQ3N+OFF17A8uXL0dXVxQKrRBbT3SYtz1QEgEAggK6uLgBgp0BkId1t0vJMxTlz5mDDhg3o7OxEfn4+OwQiC+luk7YnJnHZMRSXHXvjsmP42N4hENHgwWVHIhJsz1RUHTKqDkFVN7eobm5SjVMZlqu+6Vp2D4BxH6qbkVQ3QanGqUwZZJuRAGNDkurmJtlmqXGpxuamzyVxY87F7a83n5blZCQCgNKmKtW/w3/44HNp3M9mjJHG6MQRAhEJph3CX/7yF9x3330oKSlBXV1dv+Wa1q9fj02bNuH1119nCTUijXS3SdMO4aabbsKUKVOwcOFCjB49us9yTee/WJKI9NHdJpWfIezYsaPPJAifz9frxZIAS6gR6aajTSp1CF1dXaisrMSSJUsA9E6C6PliSZZQI9JPV5u0PQ+Bqwy9j8dVht7H4ypDeHBMT0SC7SMEIho8OEIgIsH2TEVZmWvAKHWtOv9WLa+tGqdahl1WTt4VEyOd8wPGvF+1XLvqRi6VUuKA+vxb5Xiy5xGA8Uzi0yb5hq8r0uOVfwfV5yCy816RHg8A+OyEedz4YfHS7w0wvrv/VVoljfvXH0ySxujEEQIRCaYjhL/85S949dVXcf311+N73/sePB5Pn+Wa1q9fD6fTibS0NEyYMIEl1Ig00d0mTTuEm266CdXV1Vi4cCGcTid+//vf9yrXdMUVV8DlciEvLw9erxdPPfUUS6gRaaK7TVqeqQiwhBpROOhok5ZnKrKEGpF+utqk7XkIXGUIxVWG3rjKED5cZSAiwfYRAhENHrYnJsmG2oAx3FadMqgOLZskVXzTz1XxtXJqobpBS3bO4HlVpwKqUwsrqzPLqiQDRqVk1amF6u+lWjlatYqzyneiOu2d9Wi5NO7d1TdKY3TilIGIBHYIRCRYXlPx/M+IyDq626TlNRV7fkZE1tLdJi3PVOz5mdfrZWISkSY62qTSM4RgVtTUqVMBGFlReXl5yM3NDcmK2rVrV8hnO3futOC2ieh8utqk7XkIXHYMxWXH3rjsGD5cZSAiwfYRAhENHrZnKqoOjzvPyDeQDHXFKZdhV9mMBKhvDFIp165acly1rLfqPaiWklcdRtc2m2/4yUqLR3Vjq2kMAGQPT1Au127176r6nahMVVR/15vW/00a95elN0hjdOKUgYgEdghEJGirqeh2u7Fq1SrmIRBZSHebtLym4vr167FixQq0tLTwZa9EFtPdJpVb7PlZUW1tbf2+aba5uRmbN29GQ0PDxd09EfVLR5u0vKbi7NmzsXXrVgDgCIFIE11t0vY8BC47huKyY29cdgwf/hNORILtIwQiGjxsz1S0angMGMNB1Y08VpdhVzme7NqC16e6kUt1M5LqPahuDFIZvqtu5FItOW/1Ri5PW4tpnDM+GQDgdTeZxkUnpcNf+75pDAAMyZoJ3xfyMuxRo1iGnYgGCXYIRCRIO4Ti4mL88Y9/xEsvvWRaw23Pnj0oLCwEAKxdu1b8WY/Ho+vaib5ydLdHpRFCfX29+L/7quEGAO+//z5mzZqFxsbGAd4iEQ2Ezvao9FDxG9/4Bl588UV0d3f3WcPN6/WipqYGTqcTpaWliIiIAAD4/X4mJxFZTGd7VGqtOTk5aGxsxOnTp/us4bZjxw4sWrQIeXl5aGpqQnZ2NkpKSuDxeOBwOC7y9omoJ53t0fY8BC479r4+Ljv2Pi+XHcPD9g6BiAYPTvCJSLA9U1F1GK26CcrTYb7xxhkbD0B9E4xqXOdZ8yHt0Bj1YbTqd6JyTkB9GG1lnOpUQDVOdTOSp/20aZwzLhEA4G05bhoXnTxCOc5X/4lpDABEZVyJrt3bpXGO6fOkMTpxhEBEAjsEIhK0ZCoeOHAATz75JB5++GGcOSMfEhKRGt3tUekZQn19PYYNGwbAyIw6v4bb9OnTQzKjysvLsXz5cnR1dbHIKpHFdLZHLZmKABAIBNDV1QUA7BSILKSzPWrJVJwzZw42bNiAjRs3wuv1Xuz9E1EPOtuj7YlJXHYMxWXHvuO47BgetncIRDR4cNmRiATbMxVlQ3zAGOarxg3mYbTs2oLXp1pyXnV4rBonO+9QVxwAtWG06u9l9ZRBdYqnupFLZVOV6nS2wW1e0h0ARibFSmN04giBiATTDuHtt9/Gs88+i3Xr1gGASII4P9GhZxIEEJogQUTW0d0mTTuEffv2YfHixUhNTUVjY6Mo11ReXo6CggIUFRXB5XKFJEGcX8qJiKyju02aPkPw+/3if7/77rsiCQIITXQ4PwmiZ4IEk5KIrKO7TUY++OCDD/b3oc/nw5///Ge0t7fj7NmzuPXWWwEAycnJeOWVV/Dee+/hxIkT+O53v4vZs2fjr3/9Kzo6OrBkyRKMGDEClZWVyMrKMr3B7i554lJUtFM5zufrMo85V0LK5/OZxjk0xMmuLXh9PpV7dUSj22teQTcqeigAKMfJzhvlMP4idXeaP3yMjIlT/r06vfLvZGi0A54u8+93aLTxO3Qp/l6y8waPd1YS53I64JVcGwBEOxxo75Tfa/xQ839AdbdJaR7CgQMHsH37dixZsgSJiYnSGxoorjL0vj6uMoTiKkMonW1Suuw4ZcoUTJkyxdKTEtGF09kmmalIRALzEIhIsD1T0Xv6hDQmOnGY+vxb8RmC6nxZuax360nTOGdCqvKzAdXvxMoNOoB6uXaVEuZW/66qZdNlG42iMq4EoPZ7AWrPX2TXFrw+78k6aVx06mhpjE4cIRCRYDpCePvtt1FTU4MzZ85g+fLlqK2txcGDBzFu3DhUVFSgtbUVK1euxKFDh/DSSy9h7dq1OHDgACoqKuB2u7Fq1SrmIRBZSHebNO0Q9u3bh/vvvx/PP/+8yIpatmwZ1q1bF1KSqWdWVFlZGVasWIGWlha+15HIYrrbpOmnKllRwXJNbrdbvFiyubkZmzdvRkNDg0VfAxEB+tuk6Qhh5syZ+P3vfw+fz4f29nbcdtttACBKMnV2diIrKwuLFi3ClClTsHbtWtx8883YunUrAHCEQGQx3W3StEP49re/jeTkZGzfvh0///nPxX+fNGkSJk3q/VLK4O6qyZMnD+wuiUiJ7jbJTEWiSwwzFYkoLGxPTOpqPiqNcaSNRfeRfdK4yHHXouvEMfNjDcs0znu81jxuhLEjzHuq3jQuOiUDgNqmGuWNXJJkGEAtISaYrCNLiAkmw6jGdR/dbxoXOTZH+XftavxMHjd8PDxl/8c0xvm9ewEA/sPmRUCGTJhuxH36nnncFdcZcbXvm8dlzcTnJ+UJZ2NS47C/Xv675mRYv4FwIPjUj4gEdghEJEg7hPNrtQVruK1duxaA8fJJj8eD4uJiPP3003jzzTd7fUZE1tDdHqUdwvm12oI13Hpyu90X8DyHAAAgAElEQVRwuVzIz8/HzJkzL/ReiUhCd3tUeqgYzIDatWuXyIwKZkf5/X44HA74/X6Ul5fj6NGjIZ8xOYnIWjrbo7RDCGZAtbW1ISEhAUuWLAEAZGdno6SkBB6PBykpKeju7kZTUxMCgUDIZ8EadkR08XS3R2mH0F8G1Lx5oS+lzM/PH8h9EdEF0N0eOZ4nIoGZikQk2J6pKMssA4zsMt++/08aF3XtHGlGY+S4awEAvmMHzY+V+TXj+hQz2lRKd3VX7TKNAYDISd+UlvcCjBJfsiw/x/DxxrXt/4v5teXcBADw/u0V07joG+404t59zTxu1h3ScwbP633nJWlc9LcWou35h0xj4n+6BgDQ/MRK07i0FY8DAE7/79+YxiX+/N8AAKc2/so0LuV//ofy303VTFU7ccpARAI7BCISpFOGYD22YK22hoYGHDx4EJ988gkKCwtRXFyMxYsX49lnn4XD4UBmZiYyMzND/ozL5QrHvRBd9nS3R2mHUF5eHlKrLVjD7bHHHhMxwcyon/3sZ2hubsYLL7wQ8meIyBq626OWTMWefwYAOwUiC+lsj1oyFXvWd8vPz2eHQGQR3e1RW6ZiX3+GiC6O7vbIVQYiEpipSESC7ZmKtc3y7K2stHgcUahbNy41DnUt5nGjk42XuDa3mr9kNC3BWJpRPZ7s+salxinX3lO918rjraYxV41IAABpLb9gHb+PGszjrhlpxFU3mp83e3iCcv3A946eksZdNzYF2w+Zv7B23mTjhbX/9bH5i0huu3okAOCtqkbTuLmThgMA/ru6yTRudna68t/hP35kXp8TAH54TYY0RidOGYhIYIdARIJypmLwzbHHjh0Ledts8L8/88wzIjPq0KFDIVlTTqczHPdCdNnT3R6lHcL5b44NZkY98cQT4r+3traGZEYdOnTI0i+BiAy626O0Qwi+Ofa5555Denp6SGZU8L/PmTOHNRWJwkB3e5S21tmzZ2Pr1q3o6upCa2srpk6dGvLfASApKYk1FYnCQHd7lI4QJk+e3OebY8//76ypSKSf7vbIxCQiEjjBJyLB9kzFY6fkWXmZKXFocHdI40YmxaLzrHkG4tAYIwNR9c3JpzvM3+qcGBsDQC3zUfaGaMB4S3TTafm9pifGov2M+fHiXMa1fdpknkl3RbpRx081K3PfF27TuGtHJeGzE/LsvfHD4pXjDjaYZ0d+baSRlakaV99i/h1nJMcCAN49Yl7fcta4VNQoZCpOTFO/VztxhEBEAjsEIhK0ZCqypiKRHrrbo5ZMRdZUJNJDd3vUkqkIsKYikQ6626OWTMVgDbeNGzfC6/Ve9JdARAbd7VFbpiJrKhJZT3d7ZKYiEQlcdiQiwfZMRVmGH2Bk+VkVF6yV6G43j0uKM+JkWWgT04zMMlm9xDEK9R4BIxuwUSFTcXhirDTLMzPFyCxUrakoqw2Yde5eVbIBVTNLZfUZAaNGo0rNSkA9A7HzjPnxhrqM48ne2OyMjVd+W/eeOvMMTwCYNjpJGqMTRwhEJGhJTLr55ptRVFSEBQsWYPr06eG4D6KvBN3tUUtiUmNjI2bMmIG9e/eyQyCykO72qCUxKSoqCj6fD9XV1fB6vUxMIrKI7vaoJTGpqakJeXl5yM3Nxc6dOy36KohId3vUWkItJydHdngiGgDd7ZGrDEQkMFORiASOEIhIsD1TUTWjTfXNybK6dcGadbJswOGJRkabavaeSkajau29k23yrMzUeJdyrUTVOoOqGY0qNRVVf1fVt0SrZoyqvJkaADzt5ud1xhn32iqpqZkQGyOtbQkY9S1l9RkBo0ajnThCICJBawm14GfMQyCyhu72qCVT8fnnnw/5M0RkDd3tUdpagxlQmzdvxpYtW3plRm3evBmNjY0iM2rbtm0hnzU0NFj2ZRB91eluj1oyFXt+xhECkXV0t0dtmYp9/Rkiuji62yP/+SYigZmKRCTYnpgkK2UGGOXMVJN1TkniUuKN0miqcaovj1U5nmppNNWXvaqWUFN9iatqspYs6WhkUqxyYlLlcXkJtatGJFj+wtoqSQLTpHMJTB81mCcwXTMyEdsPHTeNAYB5k0fgf/z+HWnc2//yLWmMTpwyEJHADoGIBOmUobi4GHFxcTh8+DBWr16NhoaGkMyo4AskN2/eDKfTibS0NNTU1KCwsBDFxcVYvHgxnE5nOO6F6LKnuz0qjRDuuusuXHPNNWhoaMD27dtxyy23oLy8HAUFBSgqKkJXVxdcLhfuu+8+zJ0717KbJ6LedLZHpYeKW7duRWRkJD7//HORGQV8+QJJn8+H7u5uAEBHR4f43O/3MzGJyGI626NSh7BgwQJERERg06ZNWLJkCQCIF0h2dnYiPz8fgUAAJSUlSElJQXZ2NkpKSuDxeOBwOC74xomoN53tUdohFBQUiP87eHLAeHlkzxdILl68eMA3RkQDo7s9cjxPRAIzFYlIsD1T8bSkRBUAJMbGKMfJylnFuWIAqGXbAZBmSKaey2hUyZBTfWGtt0We+RadPEI5U1H1HmTfcWKs8d2pvChX9cW2Z8/Kf9eYmBjll72qlHcD1DNVVUqtqZaB23qgXhq3YEqGNEYnThmISGCHQESClkzFCRMmhHzmcrnCcS9Elz3d7VHpGcJdd92FP//5zyIzatmyZVi3bh2WL1+Orq4unD17Fi6XC3l5efB6vXjqqafEZyywSmQtne1RS6Ziz88AsFMgspDO9qglU/H8z9ghEFlHZ3u0PQ+By46huOzYG5cdw4erDEQk2D5CIKLBw/ZMxc4z8qHlUFec9KWbgPHiTZWXcwLqQ0bVOoMqUxXV4bHV9SOtnh6pnFe1LqTqVFB1ylDfYn7ejGTjXlVfCiur+XjViATkv/qhaQwAPP2jqRj5o6ekcQ2vLpHG6MQpAxEJ7BCISNCSqTh//nwUFRVhwYIFmD59ejjug+grQXd71JKp2NjYiBkzZmDv3r3sEIgsprM9aslULC0thc/nQ3V1NbxeLxOTiCyksz0qPUNYsGAB7rjjDlRWVoq3zQaznzZu3Ain0ykyo3bt2oWmpibk5eUhNzcXO3fuvNj7J6IedLZH2/MQuOwYisuOvXHZMXxs7xCIaPDgsiMRCbZnKqoOGVU3BnlaT5rGOBNSAUB6vLQEY3isOoxWGZar3qvqcPuzE+ZvRB4/zHgjsuq9fi4Zlo85NyxXGb7LNl4BxuYr2RAfMIb5FbUnTGO+kzUMgPpUQHUTnMp38vR7R0xjACD/unGYvOINadyhJ26VxujEEQIRCewQiEjQWlPR7XZj1apVzEMgsoju9qglU3H9+vVYsWIFWlpa+LJXIovpbI9KrXXr1q3w+XwDettsc3MzNm/ejIaGhou8fSLqSWd71FJTcfbs2di6dSsAcIRAZDGd7VHb258nT56scGtENBC62yMzFYlI4HieiATbMxVlm4cAYwORakabamahalae6uYmWXZheqL82oLXp/qdqG5aUs1oVP3uVDIfa5vNzwkAWWnxynFVkgzESecyEGskx5uYZtyrarl2lQxJ2bGCx1Mt124njhCISJCOEM4vvVRbWxuSCBFMdnj22WdFIkRNTQ0KCwtRXFyMxYsXw+l0ar8Roq8C3e3RdITQs/RS0Pbt23HLLbegrKwMBQUFWLp0Kdrb2+FyuXDfffdh7ty5Ft06EfUUjvZo2iGUlpaira1NlF7asWOHSIQ4P9nh/EQIAPD7/cxDILJIONqj6ZShqakJhYWF2L9/P8rKyvDZZ5+Jtc+eyQ5JSUkhiRDZ2dkoKSmBx+OBw+G46C+CiMLTHk07hMLCQgBATk4OcnJyQj6bPHlySLLD+YkQRGStcLRHjueJSGCmIhEJHCEQkWB7pqJqBmJdi7xG3+jkOOWsPNUsP1mZ+KEuI6NRJXtPtVairMw5YJQ6V83ec7ebHy8pzjWgOJVS9582yTMQr0iXZyACxn2oZhYebDA/3tdGGt/JXz9pMo37/pXpAIAtH9aZxt01dTQee6fWNAYA7v9WFvbUyTMap41OksboxBECEQlaMhWDJZuC5ZxcLpf2GyH6KtDdHk07hL5eEhks2fTEE0+IskzBzKhgyaannnpKlHNiPUUia4SjPWrJVAS+LOfk9Xqt+C6IvvLC0R61ZCqeX86JowSiixeO9qgtU7FnOSciunjhaI9cZSAigZmKRCTYnpik+gJU2cs5AeMFnbKknpR465NwACidV/VeZecMnld2vMRz16b6YlMr71W1DJxqnCzR6Yp0I+FMluiWkWwknKmWWnvv6CnTuOvGpignV/3fPcekcYumZUpjdOKUgYgEdghEJEg7hKKiIuzevVv8/7W1tXjjjTdw4MABPPnkk3jooYfE++M2bdqE119/vc8/R0QXT3d71FJTsa8/R0QXJxztUUum4vl/joguXjjao2mH0NTUhLy8POTm5qKsrAyVlZWYOnUqgP4zo3bt2hXy53bu3GnFd0H0lReO9mh7HgKXHUNx2bHvOC47hgdXGYhIsH2EQESDh+2Ziqql0Y6dksdlpsQpD6NV4zwd5kNLZ6x6STbV0miq34ksbnSyUd5N5UW0gPqUwdNu/tJSZ1yi8lRAtayc6hBf9vckM8X4TmQvmc06d7wPPm8xjZsxJlm5NJrqS2HtxCkDEQnsEIhI0FpTMfgZC6QQWUN3e9RSU3H9+vXiM77slcga4WiP2t7+3PMzIrp44WiPWmoq9vyMIwQia4SjPdqeh8Blx1Bcduz7+rjsGB7855uIBNtHCEQ0eNieqai6aUk17sxZ8zhXzMCmDKrDaJXjqW5asuo7CW5aUn1hbedZ8ynN0Bhjw5fKd3xWEgMAMTEx0msLXp+VG68A9amFbOozPDFW+YXFsmkP8OXUxy6cMhCRwA6BiARph/Dpp5/ivvvuw5YtWwB8WcMN+LJOW11dHV566SUAwNq1a7F27VoAQHFxMTwej65rJ/rK0d0epR3CFVdcgSlTpuCHP/whgC9ruLFuIlH46W6PA5oy9MyM6pk1FRERgUAggO7ubkRGRiIiIgIA4Pf7mZhEpImO9ihdZfj0009x4MABvPzyy2hvbxeZUT2zpg4cOIDDhw/jmWeewXXXXYeTJ0+ipKQEHo8HDofDglsnIkB/e7Q9D4HLjr2Px2XH3tfHZcfw4HieiATbRwhENHjYnqmoOrT0tJlvMgEAZ3yy8mYkT+tJ87iEVACAt+W4aVx08gjjeJLrc8YnK9+D190kjYtOSle/V4VrG1CcwuYm2bUFr6+r+ag0zpE2Ft6TdaYx0amjAahNZwD1TWsqG8NUN6OpbuSyE6cMRCSwQyAiQUumYvBNtA8//DDOnJHXACAiNbrbo/QZQl+ZUcuWLQvJjJo7d27InykvL8fy5cvR1dXFAqtEFtLdHrVkKgJAIBBAV1cX3/5MpImO9qglUzE5ORkbNmxAZ2cn8vPzOUogsoju9mh7HgKXHXvHcdkxFJcdw8f2DoGIBg8uOxKRYHumovdUvTQmOiVDOnQHjOG7bLgdnZQOAOg6Xmsa5xiRZcQ1fmYeN3w8ALUpiOrwWDXO11BjGhM1ciIAwFf/iXlcxpVG3BdV5nGjJgFQ+066P//INAYAIsdcox53ZJ95zLhrAQDdn+0xjxs/zYiTnDdyzDXKcbLvDTC+O9Upo504QiAiQUtiUs/PiMg6utujlhJqLK9GpIfu9qglMen8l1ISkfV0tEelKUMwEaKyshJTp04FYCRC5OXlITc3t1ciRM/Pdu7cac3dE5H29mh7HgJXGc47HlcZ+o7jKkNYcJWBiATbRwhENHhwhEBEgu2Ziqolx1VLmKuW4VYtw656PJWS6LJzBs9rVbl2UYbdwvLqqsdT3dwk2ygFnNsspbgZTfVeVX9Xle9EVqodMMq1f/C5/BnCjDF8hkBEg4S0Q/jTn/6E2tovn8gHM6N6vkCytraWmYpEYaC7PQ54hBDMjDLDTEWi8LC6PQ7oGULPzKieL5B0uVy9MqN8Pp/IjGLFJCLr6WiPSiOEbdu24bnnngvJjMrOzhYvkExLS2OmIlGY6GyPtuchcJWh93m5ynBeHFcZwoarDEQk2D5CIKLBgyMEIhJsz1RUnmsq7hRTnVcrz5fPmJfYHuqKAyAvJx8TEyOd3wLn5t+q5dolO0WjUzIAqD3fANTLsKvsKFXdxSp7NgAYzwdU71W5vL7qrtgTx0zjHMMylZ8NbT8k37E7b/IIaYxOHCEQkaAlUzH4csmHHnqIFZOILKS7PWrJVCwrK0NBQQGWLl2KIUM4CCHSxer2eME1Fc0yoyIiItDc3IzNmzejoaFhIKcgIkU62qPSQ8Vt27YhJSUFbW1t4uWSZplRSUlJ2Lp1KwBwhEBkMZ3t0fY8BK4ynHc8rjL0Pi9XGcKG/3wTkWD7CIGIBg/bE5NUh9Gqw7ImyUaT9MRYAOqbmxrc5scbmWQcT2VTjepGLtV7lU23nLHxAAB3u/m1JcUZG3lUNwapTKOsvlfZhq+Ec7+XbKPR8HO/f3Or+b2mJRj3elKyCSo13qW8uamuxfx7A4DRyXHSGJ04ZSAigR0CEQkXnKkYzH56+OGH8emnn4ZkRvXMmvJ4PJouneirR3d7HPAzhO3bt2PZsmVYt24dli9fjq6uLjQ1mS/hEJEeVrfHC85UBIBAIICurq5eb5vtmTXFxCQiPXS0xwvOVJwzZw42bNiAzs5O5Ofnh2RGnTx5UmRNORyOi7lnIjqPzvZoex4Clx1DcdmxNy47hg/H80Qk2D5CIKLBw/ZMRdWS47LNSICxIcnqMuyqm2VUhrSqG3lUN3ypbLwB1Dc3qW74UdloJJumAMZURbUMu/dknfk5U0cDUJ8yKP/+kutzxiUqT3tUp8d24pSBiAR2CEQkSKcMf/rTn5CTk4OsrCwARmbUwYMHMW7cOFRUVKC1tRV33nkndu/ejYULF2Lt2rW46aabxGcrV66Ey2XvMIjocqG7PWrJVCwvLxef8UWvRPpY3R61ZCr2/IxVl4n00NEetWQqJicnh3zGUQKRdXS2R9vzELjs2Pt4XHYMxWXH8OEqAxEJto8QiGjwsD1TUVa+HDBKmKvGqQ7xVYeMyhuDFDb8yGKCcbKhO2AM31XLpivHWVjCXDbEB4xhvuq9qk5nVO9BdSOX9/QJ8/MmDlPeyFXTLJ8KTkyLl8boxCkDEQnsEIhI0FJTEQCKioqwe/duTZdN9NWkuz1e8Nufy8vLUVBQgKKiIsSce+1ZUGNjI2bMmIG9e/cO9PBENABWt8cBPVRUzYwqLS2Fz+dDdXU1vF4vE5OINNDRHpVGCNu2bcNzzz2HyspKTJ06FcCXmVEbN25EYmJiSGZUU1MT8vLykJubi507d1pw60QUpLM92p6HwGXH3nFcduzjeFx2DAvbOwQiGjy47EhEgu2ZirIhGWAMy5SH2y3HzY+VPMI4r2Kc6nBbZaOR6kYu5XtV3PDTdbzWNM4xwii2YeUwWvV3tSouOnEYAPUpg2oJe5XfVfZ3BDD+nnx2Qj5lGD+MmYpENEhoLaHmdruxatUqLjsSWUR3e9RSQq2srAwrVqxAS0sL3+1IpJHV7VHby16bm5uxefNmNDQ0DPAWiUiFjvaopYRaUlIStm7dCgAcIRBZTGd7tD0PgasMobjKcGFxXGWwBv/5JiLB9hECEQ0eHCEQkWB7pqLq/KuuRT6vHp0cp7xpSaVsOqC+CUql1LnqBi3VzTKq96D6bEB1Xq1y3pOScvgAkBrvUi45f+yU+e+fmRIHAGiSlGFPP1eGvb7FPC4j2YhTOe++L9ymMQBw7agkPL9XvuHrp18fLY3RiSMEIhKkHcKaNWvwj3/8A7/85S9RV1cnSjYBX5ZlqqurCynZFCzn9NBDD/FVbkQW0t0epR1CWloadu/ejeCzx2DJJrOyTGVlZSgoKMDSpUuZh0BkId3tUam1ZmZm4oYbbkBFRYXIjCotLUVbWxuqq6uZqUgURjrbo1KHMH/+fERFReH06dOiZFPPskwHDhwIyYyaPXs2MxWJNNHZHm3PQ+AqQyiuMvR9Xq4yhAf/+SYiwfYRAhENHhwhEJFge6ai6g5A1Xm1atn0U5I5bkq8Eaf6DKFRMncdnhiL5lb5vDotwSWdBwPGXFj2TCL4Bh+rn5eoxKn+XrLfATB+izOSe3XFqD/LASB9JjU62XgmIfvN0hJc+Pyk/PnWmNQ4NLjlv+vIpFhpjE4cIRCRIO0QiouLUVJSghdeeAHd3d0iMyr4Esni4mLU1taGZEb1/Mzj8Wi8fKKvFt3tUWmEsHDhQtx9992IjIwUmVFEZA+d7VHpGcILL7wAp9OJUaNGicyoYB03v98Pl8sVkhnl9/vFZ0xMIrKWzvao1FrvvvtuLFiwIOTlktnZ2SgpKYHH40FaWlpIZlTPzxwOx8XcOxGdR2d7tD0PgasMobjK0BtXGcKH43kiEmwfIRDR4GF7YpKsbDZglM72nqqXxkWnZCiX61be8GNhWW/V8uqqG4NUh9FWlpIH1ErYq04ZZMcKHk/1Hjztp83j4hIBQPr3KTolw4hT+PtU0yzfoDUxLV75Xu3EKQMRCewQiEiQThmKi4sRHx+PqKgoLFy4EEeOHMHBgwfxySefoLCwEMXFxZg7dy7ef//9Xm+bbW1txcqVK+FyucJxL0SXPd3tUekZwsKFC+F0OgF8+bbZxx57rN/48vJy8SZavgqeyFo626OWTEXgyzfRAmCnQGQhne1RqUO4++67ERERgU2bNom3zZplRiUnJ4e8iZYdApF1dLZH2/MQuOx43nm57Njn8bjsGB5cZSAiwfYRAhENHrZnKioPGRXLdXvdTebHSkoHoF5yvKv5qGmcI20sALVNMLJjBY8n2zwEGBuIVKcMsqnKUJexkcdXV2kaFzX6KiPuiyrzuFGT0HW81jQGABwjsqTHCh6v++h+05jIsTnGtSneQ3dlhfnxrvqOcbxjB82Pl/k16bUFr081zk6cMhCRwA6BiATTDuGJJ54AABw/fhwvvvgiAPRZw83j8aC4uBhPP/003nzzTb79mUiDcLRH0w4hKioKJSUlqKysxPjx4wGgzxpubrcbLpcL+fn5mDlzJt/+TKRBONqj6acOhwPt7e3485//jPHjx2PHjh19ZkY5HA74/X6Ul5dj27ZtfPszkQbhaI+mHUJ0dDSys7PhdruRmprabw23lJQUdHd3o6mpCYFAgG9/JtIgHO3R9jwELjv2Ph6XHXsfj8uO4cF/volIsH2EQESDh+2ZirJhL2AMfWWbkQBjQ5LqpiXVEuZWlv9WLTkuu7bg9Slv0FKcHqluIFKZglj9u6qWV5eVsE9PNMqcq5ZhVynX/2mTfDp7RXo8y7AT0aWFHQIRCaYdwvr16+F2u1FSUoKjR4/2mxW1Z88eFBYWAgDf/Eykke42adohLFq0COvXr0cgEMDYsWP7fdPs+++/j1mzZqGxsfGCbpKI1Ohuk6YPFePj4xEbG4sf/ehH/WZFRUREoKamBk6nE6WlpSGfMSmJyFq626S0xUZFGX1Gf1lRFRUVWLRoEfLy8tDU1MQ3PxNpprNN2p6HwGXHUFx27ON4XHYMG9s7BCIaPDjJJyLB9kxF1fLqvvpPpHFRGVcqlQgH1EqJAwMYbiuUa5eV9AaMst6ycuiAURJdtZS46gYtlU1LqsfzNdSYxgBA1MiJyhu+ZJulHCOyAKiXV1edHnlP1pkfL3U0qhtbTWMAIHt4gvRYwePZiSMEIhIsT0wKlmt6+OGHceaM/IEREanT3SYtT0wqLy9HQUEBioqK+NZnIovpbpOWJyYBfNErkS6626TliUlz5szBhg0bsHHjRlZcJtJAZ5u0PQ+BqwznnZerDH0ej6sM4cFVBiISbB8hENHgwRECEQm2ZyqqlhyXlTkHjFLn7nbzuKQ4Y9nlpGTTSmq8ESfbkBLcjCK7j8TYGOnGG8DYfCO7tuD1qZwTUNugAwCNkusbfm5jkMoGovoW+b1mJKvHqV5bTbP5M5+JacYznyrJvH/S8AQAQK3keFlp8Thy0nyjFACMS43DwQb5s4avjUyQxujEEQIRCaYdwqOPPgoAeOedd7Bz506RFdXz5ZGHDx/GSy+9BMAo1cQSakR6hKM9mnYIo0ePRkNDAz744ANcf/31Iiuq58sjIyMjL/Y+iUhBONqjaYcwd+5cvPnmm4iOjsauXbtCsqKCL49sbGxEIBBAd3c3IiMjWUKNSJNwtEfTiNTUVFRVVeFb3/pWSFZUz5dHjho1CocPH8YzzzyD6667jiXUiDQJR3u0PQ+BqwyhuMrQdxxXGcKDY3oiEmwfIRDR4MERAhEJtmcqynadAcbOs1aFZw0JsTHK5bqtnn+rlHVXfYYguwdAraz7QJ8NqM6/VUrTqz4bkpV0B4yy7rLfP+Hc76V6D6pl2FWeIak+85E93wK+fMZlF44QiEgwHSE8+uijWL16Nd555x0EAgGMHj0aBw8exLhx41BRUQG32427774b//znP7Fw4UKsXbsWN910EyoqKtDa2oqVK1eyjBqRRcLRHk07hJ6ZUf/yL/+Cp556CsuWLcMTTzyBFStWoKWlBW1toUO08vJyLF++HF1dXSyfRmShcLRHyzMVgS/rt7GEGpF1wtEeLc9UZE1FIj3C0R5tz0PgKkMorjL0xlWG8OEqAxEJto8QiGjwsD8xSVK+HDBKmHeelQ+3hsa4lKcMVsdZWYZdVvodMMq/q5aIl503OnEYAMDTftr8eHGJRpzCeVWngqpTBqvvwetuMj9eUrry8VT/bn52Qv67jh8WL43RiVMGIhLYIRCRYHlNRQAoKirC7t27NV860VdLONqj5TUVGxsbMWPGDOzdu3fAN0xE/QtHe7Q8U7G0tBRtbW2orq5mYhKRhcLRHi3PVGxqakJeXh5yc3Oxc+dOC74GIgLC0x5tz0PgsmMoLjv2xmXH8OEqAxEJto8QiGjwsD1T0bQxK9wAABbpSURBVHuyThoTnTpafRitOGRUngpYeDzVDT+ycwbPq1pyXnm4rTDtAdTuVbYBCjA2Qan+/iobqgBIpyBDXXEDivN9UWUaFzVqkvI9fNQg/12vGZkojdGJUwYiEtghEJGgraai2+3GqlWrWEaNyCLhaI+W11QsKysTn/Flr0TWCUd71PL25+BnDQ0NF/cNEJEQjvZoeaZiz884QiCyTjjao+15CFx2DMVlxz6uj8uOYcN/wolIsH2EQESDB0cIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYiEQdEhBAIBHDp0CAcPHkR3d3e/cT6fDwcOHEBVVRXM8qmC5aY//vhjtLf3n6JaX1+vdH0dHR0IBAL48MMP0dHR9+vBOzs74ff7sXfvXpw5Iy+6efKkvLgsAJw6dcr0c6/XK+7x9Gl5aqyM6m9Bl6dBkan4+OOP45prrkEgEEBVVRWWLVvWZ9wzzzyD9PR0HDx4ECNGjMDPf/7zPuMeeughnD59Gt/97nfx0UcfYfXq1X3G/fSnP8UPfvADzJ8/H3Fxcf1e3xNPPIGYmBiMGTMGH330EX71q1/1itm0aRO8Xi8mTJiAjz/+GA888ECfx1qzZg2SkpLw8ccf4+qrr8bKlSv7Pafb7caoUaPQ0dGBFStW9Bn3r//6r0hISMB1112H/fv3Iz8/v8+4rVu3hnSid955Z59xqr/Fli1b8Mknn6CtrQ3XX389br/99l4xv//97+Hz+dDV1YXY2FgUFBT0eayf/vSneOCBBzBlypQ+Pw969tlnxe+fmJiIpUuX9hn36quvIjIyElVVVUhNTcUvfvGLPuNeeeUVAMD777+PmTNn9vudbNmyBZ2dnWhsbMSYMWPwk5/8pFdMcXExTp8+jby8PJSVlSE3N7fPY3322Wch///48eP7vV87DIoRgtPpxIwZM/CNb3wDfr+/37iOjg7MnDkTcXFxpiOEpKQkREdHIycnBw6Ho9+4WbNm4Tvf+Q5KSkrw2GOP9RsXERGBIUOGYPz48f0WmAiORm688cZ+jwMAd9xxB7KysvDtb3+7384AAJKTk3Hbbbdh8eLFiIiI6DcuISEBy5cvR01NDaqrq/uNO3v2LDIyMnDDDTfghhtu6DdO9bfo7u7Ggw8+iPHjx8Pn8/UZk5KSgq6uLqxatQpDhw7t91izZs1CXV0d/uM//gNPP/10v3FDhgzByZMn8Zvf/Mb0eEePHkVHRwd+/etfm+7w8/v9iI+Px7hx4/DNb36z3zgAGDZsGFavXo2z/Wyycrlc+PWvf40tW7bg+PHj/R7nqaeews6dO7F//37s37/f9Jx2iHzwwQcftPsiRo0ahe3bt6O6uho333wzEhP73vGVlJSEnTt34o477kBGRgZSUlL6jBsxYgRmzpyJPXv2YNq0aRgxYkSfcdHR0bjyyisxc+ZMzJo1q9/rmzBhAj7//HMcO3YM8+bN6/P6rrvuOrhcLiQmJiI9PR3Dhw/v81jDhw9HQkICDh8+LLav9iUyMhI5OTk4fPgw0tLSkJGR0e89DB06FNdddx3cbjeuvvrqPuOmTp2K1NRUDBs2DAkJCf2ed9SoUdi2bZv0t6itrcVrr72GhIQEjBw5EllZWb1ijh8/jhEjRuC+++7DDTfcgK997Wt9Hqu5uRlz585FY2Mj7rjjjn47cY/Hg48//hhr1qzB/Pnz+/3XtbW1FceOHcPevXuRmpra73mvueYaNDc34z//8z9NO+fg775lyxY4nc4+v+OqqipkZGSgqakJNTU1uOmmm/o81g033ID09HTU19fjtttu6/ecdhkUU4bgcOvee+9FeXm56XArEAiIfzH7+wsRHJYFY2VxQRdzPCuPdSHHszIu+FfC7PqCXn75Zfz4xz++6JjLJc7qc4bboOgQ/vCHP+Dee+/F448/Dr/fj6Kioj7jCgsLcfXVVyMpKQkAMH/+/EETN5ivTUccXaYCg8CLL74YaG5uDrz44ouB+++/v9+4jo6OwJEjRwJ//OMfTY9nR9xgvjYdcXR5GhQPFe+66y4MGzYMQ4YMEe+074vL5cLYsWOlb7G1I24wX5uOOLo8DYopAxENDoNihEBEgwM7BCIS2CFcwjweD9xut5Zjd3d3o7m5WcuxafBih3AJam5uxtatW/HWW2+hsrISr732GhobG5X//D/+8Q/s27cPx44dQ1FRER544AHcf//9IVl4e/bsQUVFxYCvrbKyEjt27JDGFRYWwufz4ZlnnhnwOUgf218HT4bi4mKxcerOO+9ETU0NduzYAa/Xi1//+td4/vnn0dbWhqFDh+LHP/4xvvjiC5w8eRLR0dFITk5GZ2cn1q9fj5aWFgQCAaxZswa33347brzxRjQ3N6NnQupf//pXrFmzBg888ABWr16NlJQU7N27F/X19SLjcM+ePfjwww8xc+ZMbN68GX6/H9/73vdw4MABfPHFF/jhD3+IJ598EtnZ2ejs7ER6ejqioqIwbdo0nDhxAmvWrEF6ejqOHDmCgoIC7Nu3D/v27cPx48fx6KOPAgCioqKQkpKCyspKXHXVVWH/zqk3jhAGkR//+MdYsmQJXnnlFfz3f/83HnnkEeTm5uKNN97AF198gTFjxmDevHki/utf/zq+973vATB2WzY1NeF3v/sdRo4cicrKSmRkZGDJkiUhOz7dbrfI7/f5fCL9++tf/3pI+nHw2C+//DKcTicSExPxwQcfAAAWLVqEkSNHYtKkSVi1ahVaW1uxbNky1NbWij/v9Xpxzz334LbbbsO+ffuQnZ2N2bNnIz4+HjU1NSLu6quvxt///ncN3yZdCHYIg0ggEIDH40FkZGTIf/f7/cjNzUV2djYeeeQRsemo56anQI+U7uCfCW4AiooKHQgGOwSn0ym2V//tb3/D3//+d6xYsQL/9m//Jo7l9/tx5513oqCgADk5OQCMXIXg/46KihIbvs7fhDV06FBERUXB7/dj8+bN8Pl8mDhxYsjGtCFDhvCVf4MIpwyDyLPPPgsA+OUvf4lDhw5hzZo16O7uxv33348//OEPOHPmDLKyskQDysjIwMsvv4wJEyYgJiYGw4YNw7//+78jIiKi3yF4UlISOjs7AQD5+fn43e9+h/j4eHR3d+Ohhx7C9ddfDwA4cuQINmzYgDVr1mDdunUYMmTIReXeJyYm4p133sGxY8cwceJE8d/3798vzkn2Y2LSIFFcXIxbb70VY8eO1X6u3bt3IyIiAtOmTdN+LjM+nw/PP/887r33Xluvg77EDoGIBE7eiEhgh0BEAjsEIhLYIRCRwA6BiAR2CEQksEMgIoEdAhEJ7BCISGCHQEQCOwQiEtghEJHADoGIBHYIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYgEdghEJLBDICKBHQIRCewQiEhgh0BEAjsEIhLYIRCRwA6BiAR2CEQksEMgIoEdAhEJ7BCISGCHQEQCOwQiEtghEJHADoGIBHYIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYgEdghEJLBDICKBHQIRCewQiEhgh0BEAjsEIhLYIRCRwA6BiAR2CEQksEMgIoEdAhEJ7BCISGCHQEQCOwQiEtghEJHADoGIBHYIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYgEdghEJLBDICKBHQIRCewQiEhgh0BEAjsEIhLYIRCRwA6BiAR2CEQksEMgIoEdAhEJ7BCISGCHQEQCOwQiEtghEJHADoGIBHYIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYgEdghEJLBDICKBHQIRCewQiEhgh0BEAjsEIhLYIRCRwA6BiAR2CEQksEMgIoEdAhEJ7BCISGCHQEQCOwQiEtghEJHADoGIBHYIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYgEdghEJLBDICKBHQIRCewQiEhgh0BEAjsEIhLYIRCRwA6BiAR2CEQksEMgIoEdAhEJ7BCISGCHQEQCOwQiEtghEJHADoGIBHYIRCSwQyAigR0CEQnsEIhIYIdARAI7BCIS2CEQkcAOgYgEdghEJLBDICKBHQIRCewQiEhgh0BEAjsEIhLYIRCRwA6BiAR2CEQksEMgIiHK7guQefPNN+F2uwEAP/nJT2y+GqLLm7YOodAxQTl2xguP4pVXXhH//5133ok777wTAHDzzTfj9ddfx4kTJyy/RiIKNShGCD07gPPV1tZi/vz5eOSRR8J8VURfPdo6BMeQCEuO889//hNvv/02MjMzLTkeEfVPW4cQadFx7rrrLouOREQy2jqEaItGCEQUPto6hCER7BCILjUaRwi6jkxEuuh7hsARAtElR+OUQdeRiUgXfVMGjhCILjmcMhCRMOgTk4gofDSOEHQdmYh0YWISEQlcZSAiQeNeBvYIRJcafQ8V+RCB6JKjb8rAOQPRJUdbhxAVxc0MRJcabR1CBKcMRJccfSOEaKtKpBBRuOgbIXDKQHTJ0fdQkVMGokuOvjwEB6cMRJcaPlQkIoEjBCISOEIgIkFfh8BMRaJLjr4pA/MQiC45g36E8OKLL+LMmTM4e/Ysli1bZskxiahvGkcI6olJr7zySr9vf/7+97+P1NRUvuyVKAwiAoFAQMeB9932A+XYa/+rtN/Pzp49i/Xr12PhwoUYPXq0FZdGRP3Qll8cGT1E+X/M/Pa3v4Xf70dZWZmuSyWicwb9M4THHnvMkuMQkZzGPARubiK61DBTkYgEjhCISNC3/ZkjBKJLjsZ6CBwhEF1qNI4QtB2aiDTRuOzIEQLRpYYPFYlI4JSBiASOEIhI0Ljb0aHr0ESkCZcdiUjQOGVgYhLRpYYPFYlI4ENFIhK4l4GIBD5UJCKBzxCISOBeBiIS9P0zPoTPEIguNZwyEJHAKQMRCfr+GY/iXgaiS43GEQKfIRBdajhCICJB20Q/Ykik8v+Y6ejowP3334/a2lpdl0pE52hcdrTm7c/Nzc24+uqrLb88IupN3zOEAUwZenYA5xs3bhwSExOtuiwiMsHEJCISBsUIQeb222+37FhE1L9B8QyBiAYHjSOEaF2HJiJNOEIgIoEPFYlI0DdlcNg7ZTh16hSGDBmCpKQkW6+D6FJy2ZVhf/fdd/HXv/4VTqcTERER8Hq9mDt3LqZNm2bL9RBdSvRN9KOi1f/HQrGxscjNzUV3dze6urpwzz33IJLviCBSctk9VMzJycEDDzyA1atXAwAef/xxPPzww7ZcC9GlRt+UIcK+f5UjIiKQlJQEv9+P7u5u266D6FKjb4QQaV8Jtbvuugu/+c1vAAALFiyw7TqILjURgUAgoOPA3Z/tUY6NHG/dA7+mpiZ4PB4EbysiIgKZmZmWHZ/ocqbxGYI9I4Rdu3ahoaEh5L8tXbrUlmshutToe/IXMUT9fyxy+vRpZGRkwO12Y9q0aZg2bRoOHz5s2fGJLnfa/hkP2JCpGBMTg8bGRng8HjQ2NgIAFi5cGPbrILpUXVYjhOjoaMyfPx/33HMPACAQCODQoUOWHZ/ocndZdQhBa9euRXV1NQDgo48+svz4RJery2rKEJSSkoIbb7wRbreb5deIBuCyHCFkZGQgJSUF7733HtLS0iw/PtHl6rJMTKqrq4PX60Vubi4iIiJsuw6iS42+KYONDXHUqFEoKysT/z/zEIjUaMtU9LYcV46NTh6h4xKIaID0jes1PBsgIr0uy4eKPp8Pfr8fVVVVaGlpsfz4RJcrbR1CYEik8v9YbdWqVdi7dy82bdqERx991PLjE12uLssRQmxsLHbt2oVf/OIXrJZENACDPjGpvLwcx48f///bu3/QqLIojuO/F23MZhdMWA2uJuCCjUIKYZSEBMNaWIh/gkJELOxGS7FwQRGrBRUWGyFFRCLEDVPFgGixin8yKmIgChZCUKKxcBkhyUh0Ms7bQrhoo2+Gd3hzX74feN3j3tPk5My5977rPof2I7lcTuvWrVMQBHr06BFHn4Eq1EWFMDIyor6+Pvd8fRP0xMSEDh48qPfv30eatr29Xc+ePdOGDRu0fv16TjsCVbCrEBR9H8L3bn9evvxLiJVKJdJYmUxGo6Oj6urqUhiGunHjRuQ4gKXOLCF8jml3w6ZNm3T58uWqtiD39fXp1KlTCoJA+/fvjycQYAkw25g092Eh8ru//LQitnlzuZyCIPjmE2r79u2LbXwgzcx6CJUwjPzEqb29Xffu3VNra6vWrFmjhw8fxjo+kGaGCSH6E6dMJqOmpiZ1dXVp69atWrEivuoDSLu67yHUgh4CUBu7VQab1kQkzc3NamxsVBiGam5uTiwOwDd2W5ereOJ28eJFHTlyREePHtXAwIDBDEA62f1kiLZtwARXuQG1MUsIFZP//dF8fZUbPQQgOrN9CNOFYuR321qaLEIAUCW7ZccqnjgNDAwon89rdnZWxWJRd+7c0eDgYMyzAOlkViFM/Tcf+d3ff/051rlv376t8fFxBUGgzs5O9fb2xjo+kFaGh5uS09vbq9WrV6tYLCqTyWhyclIdHR0JRgT4wW7ZMYz+WHjw4IFaW1t15coVTUxM2EwCpIxZQvgchpEfCxs3blRbW5v6+/v17t07kzmAtEndWQZJGhsbc8elC4WCyuWynjx5Ev9EQMoYnmVIrotw6dIl7d69W/fv39f8/Lyy2ayGhoa0efPmxGICfJDKCmHLli3as2ePXrx4oTdv3qixsZFTj0AEZhVC2eIvPaLu7m4NDg7q0KFDmp2d1dmzZ9XS0pJYPIAv7LYuJ5gQwjDU3Nycrl69qu3bt+v06dOJxQL4JJU9hHw+rzNnzkiSzp07p+7u7sRiAXxilhAWE/xCSqlU0qdPn1SpVLS4uJhYHIBvUlchPH36VDt27FA2m1WlUtGxY8cSiQPwkWGFkMwHEaampiRJu3btkiS9fPmSbctARKn7puLevXuTmRhIAbtVhgSbigBqk8qmIoDaeHVRy/Xr13XhwgWrkIElz65CqGJj0sjIyDc3Pn99+evw8LAeP36sMAx1+PBhPX/+PPZYAXxh9sWkfyZnIr/b3/Fb5HfPnz+v48eP1xISgB+ghwDA8W6VgeoAsONdQgBgpy6aigDqQyqPPwOoDU1FAE7qTjsCqF3qTjsCqJ3hdfBUCIBvSAgAHLOEUCrzkwHwDRUCAMeuQqCpCHiHCgGAQ0IA4JglhI+Ln62GBmCECgGAw7IjAIcKAYBDhQDAoUIA4HD8GYDDTwYADp9QA+CYJYQyZxkA71AhAHDsmop8ZBXwjllCCGOuEF69eqXh4WGVy2Vls1mtWrUq1vEB1ElCiHL7c6lU0smTJ3Xz5k0Vi0USAmDA7Pbnzr/+jfxu/s8/Ir137do1NTQ0aOfOnbWGBeA7GqwGDith5CeKu3fvanR0VNPT03r9+rVV2MCS5k1TsaenRz09PbGOCeBbddFDAFAf7PYhsDEJ8A4bkwA4dj8ZyAeAd+wqBE47At6xSwiUCIB37JYdqRAA77DsCMAxXHYkIQC+MUwI3NwE+MZwHwI9BMA3hsuOVAiAbwybiiQEwDeGFcKi1dAAjFAhAHBICAAcfjIAcKgQADiGFULJamgARqgQADj0EAA4VAgAHBICAIefDAAcw9OOVAiAb7ypEGZmZjQ2NqZCoaATJ05o2bJlsY4PwDAhfBj/O/K7UW5/lqQDBw4ol8vJ6H5aYMkzu/05bm/fvtXKlSt169YtrV27Vh0dHUmHBKSOWYUQt4WFBQ0NDSkIAm3bti3pcIBU8qZCAGCvIekAANQPEgIAh4QAwCEhAHBICAAcEgIAh4QAwPkfxR/70bqQrHEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x600 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Draw codon biases for all positions using heatmap\n",
    "fig = plot_codon_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus,\n",
    "                                 plotTitle, refCodonTableDf)\n",
    "filenamePlot = outputFilePrefix + '_composition_bias_codon_' + terminus + 'terminal_heatmap'\n",
    "fig.savefig(os.path.join(plotsFolder, speciesSetName, filenamePlot + '.png'), dpi=600, bbox_inches=\"tight\")\n",
    "fig.savefig(os.path.join(plotsFolder, speciesSetName, filenamePlot + '.svg'), dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "# matplotlib.rcParams['plot.linewidth']\n",
    "matplotlib.rcParams['grid.linewidth'] = 0.5\n",
    "matplotlib.rcParams['axes.linewidth'] = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/colors.py:512: RuntimeWarning: invalid value encountered in less\n",
      "  xa[xa < 0] = -1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Analysis_data/C-terminal/Data/Taxonomy/taxaId_31969/taxaId_31969_subset_all_composition_bias_codon_Cterminal_barplot.svg\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQ4AAADMCAYAAACC5k2uAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3XtcFPX+P/DX7MJyj/tdgfTkDbzFN+ymSWp20ow8J1M0j3Tz1qNO6eNxPCc92TmPY528cNK0h5qC+ogyzBQDNS+VGIoJKhSiBFgBCoiC3Jbd2ZnfH/wY2WV2mdkbC7yfj8c+ct59Pvv5zGdm38zOfHaG4XmeByGEyKDo6Q4QQnofShyEENkocRBCZKPEQQiRjRIHIUQ2ShyEENkocRBCZKPEQQiRzcmejWVmZqK+vh4AMHfuXHs2TQixIrsmjmnTpuHAgQO4efOmyXJXr14FTWglxP5cXFwQFRXVbTm7Jo7S0lIkJCRgzZo1Jst5enoiLCzMTr0ihHSoqqqSVM6uiSM3NxcnT57EwIED7dksIcTK7Jo4EhMT7dkcIcRG6KoKIUQ2ShyEENkocRBCZKPEQQiRjRIHIUQ2WVdVbt26BYVCAR8fH7Ma+/TTT9HS0oLW1la8/vrrZr0HIaTnSUocOTk5OHbsGFxcXMAwDDQaDZ566inExsbKamzKlCnw9/fvdgKYWq2WPBGFEGJ/khKHh4cH5s+fj7S0NPA8jwULFqCurk52Y15eXtiwYQOSkpJMlnN1daWZo4T0AKl/sCWd4xg9ejS2bNmCxYsXY8mSJdi6dSvGjBkju1MrV64Ex3E4fvy47LqEEMch+RwHwzDw8fEBx3HQ6XRmNbZ+/Xqz6hFCHIvkxJGYmIi3334bADBr1iybdcgR8VoNGGeV0WVC+htJiaOmpgb+/v5YvHgxgPajj/6EcVbh1xULhOXI91N7rC+EOAJJiSM7OxvXr1/Xi7322ms26RAhxPF1e3K0oaEBYWFhqK+vR2xsLGJjY1FWVmaPvtkcz2pNLhNCxHV7xOHm5obq6mq0tbWhuroaADBnzhybd8weGCdn3ElZJSzfk/RvyXV5nRaM0tnoMiF9WbdHHCqVCgkJCViwYAEAgOd5XL582azGmpubsWzZMpSWlppV35EwSmc07npHeFHSIP2J5Ksq69atQ1RUFIYMGYLCwkKzGqutrUVMTEy35ew1c1RskplYu1Ino9FsV9JfSE4cfn5+mDRpEurr6+Ht7W1WY1FRUZLq9uTMUUvapdmupLez6sxRoP1D4efnh7NnzyIwMNDsjhFCej/JRxwVFRXQaDSYP3++RfM4Zs6caXZdQuyJ53RgFEqjy/2Z5MQRHh6u9xsTmsfRO/EcB0ahMLpM7mIUSmjPfy0sO//f9B7sjWORnDgcNVHQB0EeRqGArjhbWFYOG99jfaFt13vZ9fEItsAoFNBduyAsK6PGipbjeQ4MozC6TOyPUSjAlZ4TlhWD43qsL7R/yCM5cbAsC4VCgatXryI4OBi+vr627JfVMYwCbEWRsOw0YEQP9qZ36ssfLoZRgK0sFpadwof1YG8cn+StvmLFCuTn52P79u147733bNkn0g3D5+ra6zm7DKMAe/0X4SUnafRUn3tKX19fyVvew8MD2dnZWLhwIZRKOrPckxiGgfbm78LL2FUuR9p5GYaB5laV8Orrv7BmGAaaugrh1dfWV9JXlfT0dAwcOBAMwyA3N9fsZ7+eOHECN27cgFarFaawG8PzvN5gd+z0hrG+tkGsiWEYaOprhGWVT1AP9obIIbb/O9S+zkuQm5vLv/766/ypU6f406dP88uWLZNSrYsPPviA53meX79+vclypaWlPMuyejGWZcVjWq1+TKvltQbltEbKsW1q/Vibmr9z545e7M6dO7xW3ar/fupW6XW1Bn3RsnxlZaVoH6XWF1sXsbosa1CO1fKsVmNQV2OkrrSx1hjENEZilvSlTaNft02j5dUGMbVGy7eo9d+vRa3hW9oMYm1G1lds/9AY9E8jXlfq/iZWt7KyUnwbi4y/6NgYtNOm1Yq+n8YgptGyotupsrKSl0LSEUdcXBwOHjyIRx55BDzP48iRI2YlKSen9uY4jjNZztXVtcvXIbGvR6Kx/99GTUOzEAvy9jBSTn/1lSoXeKlc9GJeXl5d6jq5uHZ9PxN1r1TfEWJDg+8RpqZX3G4S4gN8Pbu01bFcWtsoxAYHdu2P0snJaF29csqum1vp5AwvL/0f6Hl5eXX5WqNQKLr8xVM6OUEJ4OvLN4TY9OEhAICs4moh9tSwYIv6AgA7fvxNiL30QAQA4J2jd09mvju1/WTmw++dEGI5f58EAHjio9NC7JvXHgVU4m0YrluXmLMzvJzF616vv7u/hfqI729i26gjdqmqQYiPDuv6swylUml0G39ZeHea+J9Gtu9b5Tfv7jP3BrSXy6uoF2KxA9ofcZJz7e5Nxx+O8u/SrjGSr6rMnDkTq1atAsMweO655yQ30FlMTAxSU1NtPmWd53m9ZME72mFeL2A4XjR+pDPJ5zgYhhHubF5SUoLRo0fLbmzKlCmy65ijN+z0PM9jgK+n3rIj9pN0j+d5vaOM/rAtJV1ViYyMRHZ2NkJCQhAWFoazZ8/aul99ntTkxvE8Bgd6CS+uj13W6wsc7Q8Vx/O4N8BLeNlin5GUOOLi4uDp6YlHHnkEDz74INzc3Kzekb6K43kMDb5HeMndiAqDndBwmRBD9thn7HqOoz+iDz7pDsfzeidEOZ53+P1E1o183N3dwfM8/Pz8bNknQvoVe/xx4XheuJLSsQxe/0oKx0k/GpY8c9TwEZCEkN5DLDkpFAYxhfSEZddHQGZlZaGkpARvvPGGWfWJ49FxvDB3o2MZ0J+7oeN4KGXslMTxmfUISDnnONLS0vDjjz+C53kkJSWhqKio2zr2ulkx6R2sfS/XvrRvBQQFCZO+AEDDsrhZU2OihnVIThxjxowx6wn1iYmJSExMFJaPHTvWbZ2evFkx6fv68r6lcnKyaP2kJlVJiWPr1q0YOXIkoqOjoVQqkZeXh19++QUvvfSS2R0kRCqW44Vp5h3LTvTVp0dJOjm6cOFCtLW1YdOmTfjwww/BcZzZSWP58uVm1SP9l2GSoKTR8yR/VYmPj0dwcDCampoQFxeHS5cumTXtnBBr0Oo44YdtHcvOyr5xN7LeQNZInzlzBiEhIdizZw/y8/Nt1SdCumWYJChp2Jes0Y6OjkZERARmz56NGjucuSWEOCbJiePQoUPCz+Hr6urAsizy8vJs1jFCiOOSfI5j586deOaZZ3D69Gk0NjZi0aJF2L17N2JjY23ZP0KIA5KcOMaNG4eEhASsXbsWOp0O7u7usn4le+3aNaSlpYFlWSxatAhBQXT/S2J7GpZrv+tXp2WVE50PsZTkxDF+/Hjs2LEDL7zwAhoaGvDBBx/A37/7W411zBzVaDRYuXIljh49iqamJkocxC4MkwQlDetgeMObSxpx+vRpYdbn5MmTMX68/EcHZmRkQKFQYPp008/gLCsrg6tr1/t6EmKKr38g3Fzu3hO0tU2L23W1Pdij3knKzFPJRxw5OTl49913AQBr166VnThOnTqFgwcPIjY2Fr///rvJRyzQlHNiDW4uznCj/UgWq045BwCNRoO2tjZwHAetViu7QxMmTMCECRNk1yOEOB5JiaOgoABPPvkkFi1aBI7j8NZbb9m6X4QQByYpcZSWlgIAZsyYAQAoLy+n6eaE9GOSEsezzz5r634QQnoRujZFCJGNEgchRDZKHIQQ2SRfjrVUZWUlDh06hLq6OqxYsUL0gdGEkN7B5omjY8o5AMyZMwfp6eldnoRuSKPR9KkbyhLSW2g0GknlJE85t1RVVRV8fX1x8uRJDBgwgC7nEtKL2e2rSmtrK3bv3g2GYTBx4kR7NUsIsQG7HXEQQvoOuqpCCJGNEgchRDZKHIQQ2ShxEEJko8RBCJHNIRNHVVUVqqqqUF1dLUwW6/gvx3FgWVavPMdxOHfunF5MrVZ3mcwitZwYc+rKXY/u6sqpb89+d9eGtcfBGvuCJW2IcYT9Q+64SOmjMQ6ZOAoLC/HTTz8hPz8f69atAwBs2LABLMtCrVbj448/BgAcPnwY6enpWL16Nfz8/AAA69evR0NDAzZu3IiNGzfKKifGkrpS10NqXTn17dFvqW1Yexws6YslbfTUOFt7XOT00Ri7TQCTi+d5MAwDjuOEWEFBAXQ6nbDc3NwMjuPAMIwwKAzDoLS0FA899BAuXLggq5yY7upevHhR0np0PmroWA+GMf3wZLG6Uutbq9/WGH+p72evvphqQ+52ssc4W1JXrH+GfexubMQ45ASwsrIyKJVK1NbWIjw8HKGhoWhubkZZWRmcnJzg4+OD0NBQaDQaKBQKXLlyBSUlJUhISEBTUxPKysrg7Owsu5wpdXV1yM7ORkJCArRaLS5fviz0NSEhQbROZWWlsNOdO3cOCQkJYFkWxcXF4DhOdl0Akutb0u/i4mJ4eXlBqVRCpVLBz88Pt27dQlVVFbRaLQYNGgRvb280NTXB09MTtbW1yM3NxfTp03Hr1i1UVFRAp9MhMjISfn5+krenPfpi7e1kyThbMi6WfEYAyN7/DTnkV5X09HT4+/vjxx9/xJdffgkA+Oijj3Dvvffiu+++E2LJyclQq9U4deoUKioqAACbN2/GoEGDzCon5vPPP0d6ejo2b96MyZMnAwC+/PJLXLlyBQcOHBBiYk6dOoUffvgBn3zyiVBu3759uHz5sll15dS3pN+ZmZkICQmBn58f0tLSAAApKSkYPnw4oqOjsWfPHgDA1q1bodPp4O3tjWvXrgnloqOjER0dLdSVuj3t0Rcx2dnZZm8nS8bZknGx5DMCSN//jXHIxOHh4YFvv/0WoaGhUKlUdouJCQgIEE42lZeXG42JCQwMtGpdOfUt6beLiwuysrKQlZUljI0lMUvGv/P7ubi4WByz9lhZUteScbF0P5fajjHK1atXr5Zdy8ZCQ0MxaNAghIaGIioqSjiUMjcWFxeHgIAAODk54fbt2xg1apRoTMygQYMQExMDf39/ZGdnIy4uTjQmxtnZGQ8++CDi4uJQXV2NsLAw0ZgYLy8vjBw5EgEBAQCAkJAQk3Ep9aXWraysRF1dHdzc3PDwww/D19fXopgl4x8XF4fAwEAolUq9uubGxPTUNq6srERtbS3c3d31xs8wJiYqKgohISFQKpWIiIhAcHCw5Jix+nI45BHH/v374e/vj9DQUBw9etTiWHJyMk6ePImTJ09i7ty5RmNiNmzYgPT0dJw8eRJLly41GjO2HgqFAp6ensjLyzMaE5OamgqlUonS0lLk5+d3G5dSX2rd+Ph4zJw5ExMmTEBmZqbFMUvGPzk5GSdOnOhS19yYmJ7axvHx8fjzn//cZfwMY2J27dqFgIAAlJaW4vz587JipuJSOeQRR3FxMQoKCnDhwgW4ublh1KhRFsXuuece1NTUoLq6Gi0tLfjDH/4gGhNjSV2p/RNTUlKCwsJCaLVaqFQqoZyxuJT6UuumpaWhuLgYV65cgUKhQGxsrEUxe4x/b9zGUsfPku1r6X5kFO+Azpw5Y9XY2bNnhX9XV1cbjYmxpK7U/kmta2l9a9d1pPHvK9vYHttITjvGOOQRR1tbGw4fPoyGhgZERkZaNTZixAijMWN9ycrK6lLXMGbJekita2l9OXU71q9zXWvEzBlDKXWttZ3M2T+suY2tXdfS/cgYhzzHUVxcjJaWFoSEhGDz5s12i4kpLS3FvHnz4OPjI5QTi4n5/vvv0dbWhsLCQqGcWEyMWq2GVqvFlStX9MoZixu6c+cO5s6di6CgIOGSpVjM2Do///zzuHTpkt462zompry8HPPmzUNkZKTQZ0tiYr7//ntoNBqUlJRg3LhxRmNi6urqoNFoUFxcjM2bNyMnJ0c0JubXX3/FrFmz4OvrK/RPLCbm+vXreOqpp1BQUCCUkxozFZfKIRNHdXU1XnnlFXzxxRfChBV7xMREREQgOTkZx48fR1JSktGYmBdffBEvvvgiPD09hRNsYjExZ86cwYIFC/Dyyy/rXS4zFjdUVFQEALhw4QIUCoXRmJja2locP34cEydOFCYGWSvWcSVHLCbmwIEDSElJwfHjx4WHnVsSEzNlyhR4enrC3d0dKSkpRmNiLl26hKSkJLzyyitgWRa///67aExMeXk5nJ2dERMTI8z+FIuJKSoqgq+vLx544AHhiQFSY6biUjnklPOGhgakpKTA3d0d33zzDZKSkuwSE5OTkwMPDw/cuXMHe/fuRVJSkmhMzM6dO8EwDG7fvo2UlBQkJSWJxsQ0NjYiJSUFPM+jqamp27ihmpoapKSkoL6+3mRMjKenJ2pqajBlyhSUlJTYLSZm2rRp+OMf/2i1mJht27YhIiICSqUSubm5GDFihGhMTHh4OFJTU8FxHIYPH44nnngCWVlZXWJiBg4ciF27doHneURFRRmNiYmMjERqaqrwbzkxU3GpHHLKOSHEsTnkVxVCiGOjxEEIkY0SByFENkochBDZKHEQQmSjxEEIkc2u8zgyMzOFOQSmfq1ICHFsdk0c06ZNw4EDB3Dz5k2T5a5evdrlPpuEENtzcXExOfGsg10TR2lpKRISErBmzRqT5Tw9PY3e/IQQYjtVVVWSytk1ceTm5uLkyZMYOHCgPZslhFiZXRNHYmKiPZsjhNgIXVUhhMhGiYMQIhslDkKIbJQ4CCGyUeIghMgm66rKrVu3oFAo4OPjY1Zjn376KVpaWtDa2orXX3/drPcghPQ8SYkjJycHx44dg4uLCxiGgUajwVNPPWX0mQ/GTJkyBf7+/t1OAFOr1ZInohBC7E9S4vDw8MD8+fORlpYGnuexYMEC1NXVyW7My8sLGzZsMHmDXwBwdXWlmaOE9ACpf7AlneMYPXo0tmzZgsWLF2PJkiXYunUrxowZI7tTK1euBMdxOH78uOy6hBDHIfkcB8Mw8PHxAcdx0Ol0ZjW2fv16s+oRQhyL5MSRmJiIt99+GwAwa9Ysm3WIEOL4JCWOmpoa+Pv7Y/HixQDajz4IIf2XpMSRnZ2N69ev68Vee+01m3SIEOL4uj052tDQgLCwMNTX1yM2NhaxsbEoKyuzR98IIQ6q2yMONzc3VFdXo62tDdXV1QCAOXPm2LxjhBDH1e0Rh0qlQkJCAhYsWAAA4Hkely9ftnW/CCEOTPJvVdatW4crV64AAAoLC81qrLm5GcuWLUNpaalZ9QkhjkHy5Vg/Pz9MmjQJ9fX18Pb2Nqux2tpaxMTEdFuOppwT4tgkJ46wsDD4+fnhyJEjCAwMNKuxqKgoSUmHppwT0jOsfrPiiooKaDQazJ8/n+ZxENLPSU4c4eHher8xMXcex8yZM82qRwhxHAzvgE8+qqqqoq8qhPQAqZ89ugMYcXg8z5lcJvYn+asKy7JQKBS4evUqgoOD4evra8t+ESJgGAV0v14SlpWRo3uwNwSQccSxYsUK5OfnY/v27Xjvvfds2SdCiIOTnDg8PDyQnZ2NhQsXQqlU2rJPpJ8wPL1mi9Nt9mijP5L0VSU9PR0DBw4EwzDIzc01+9mvJ06cwI0bN6DVaoUp7OQunuf1LnUbLvc1DMNAc/uGsKzyDbFNG3UVd9vwH2D1NvojSUcckZGRKCwsxJAhQzBo0CCzfx2bn5+PuXPn4tatWybLqdVqtGq0erFWjRYaltWLaVgWLWqNXqxFrUFLm37dljYtGhsb9WKNjY1gWf07mbGsDlqDmJbVQaM1aFfLitaV2kZVVRVYg7uosTodOE7/pB/HcaLvKdZHrcHYaFlWtA2pfZS6fmJjI9qGSF90OhYq3xDhpdOJ19WxLJSRo4WXjpXRBstC5T9AeBmtK7K+hne60+nEx0XqWInFqqqqJJe1R0wqSUcccXFxOHjwIB555BHwPI8jR45IbkCvMaf25gw/IIZcXV3hpnLG4x+eEmIn35gAAPii4O7MtlmjwqByAkKf2yzErqcvBQAMfzNDiF1OngG4OOu14eXlBQC4VNUgxEaHtc9qzauoF2KxA9ofBbEr73ch9pfY9iOuC5V3y40N9xHe07CNX2rvbpA/BHoJl7tqGpqFeJC3BwwplUqj7/njb7eF2AMR7Seqv75896/39OHtf70rbjcJsQG+npL7CADlN+/G7g3wMlp3T/7dv+gv3D8AKmfxctfr765vqI/Y+joZbUOvnJPxct22YaLuz9fvCLHo0Hu6fK1RKBRgGKbLuABA4fW7+9HIUG9J6+Hl1XVMTZW1R0xq8pB8VWXmzJlYtWoVGIbBc889J7WanpiYGKSmppo9Zb0v4XleL1n09a8lvZHh9qDtc5fkcxwMwwh3Ni8pKcHo0fIviU2ZMkV2nb7KHjslz/MY4Oupt9xTOz/P83pHALboiz3aIO0kn+PIzs5GSEgIwsLCcPbsWVv3i1iBJcmJ43ncG+AlvDgLr0bYI1Haow1rj0tvZddzHP0Rx/PCOYOOZYWD/RWU0kdH63NPoXFpZ9dzHI6G43nhhGjHsrV3hN6wo/WGPtoax/OIDr1Hb1nqOHA8j5Ghtt2PHI2sG/m4u7uD53n4+fnZsk92I/aB4XheuJICGN8JOJ7H2PDuy9kCx/PClRR7t91XWZI8+2PilTxz1PARkH2V1J2gJ3eW/rijEsdi10dAZmVloaSkBG+88YZZ9YlxOo4X5m50LCsVlFCIbZj1CEg55zjS0tLw448/gud5JCUloaioqNs6arVaNK5hWcwaFaa3rHKStgqW3MNU7P4EdE/UdjQ2/ZPkxDFmzBiznlCfmJiIxMREYfnYsWPd1nF1dRWNGyYJqUkDEN/BpdJxvDBbtGOZbjTUTsfxeOH+AXrLNDa9l9SkL+kcx9atW5GTk4OGhgY0NTXh+++/x44dOyzqYG9ieMhPXwHuorHpnyQljoULF6KtrQ2bNm3Chx9+CI7j8NJLL5nV4PLly82qRwhxHJKP9ePj4xEcHIympibExcXh0qVLZk07J4T0frLuOXrmzBmEhIRgz549yM/Pt1WfCCEOTlbiiI6ORkREBGbPno2amhpb9YkQ4uAkJ45Dhw4JP4evq6sDy7LIy8uzWccIIY5L8jmOnTt34plnnsHp06fR2NiIRYsWYffu3YiNjbVl/wghDkjyEce4ceOQkJCAq1evoqKiAu7u7nBzc7Nl3wghDkryEcf48eOxY8cOvPDCC2hoaMAHH3wAf39/yQ1du3YNaWlpYFkWixYtQlBQkMnyGpYTbhfYsaxyoudHEeIIJCcOnudx584dfPbZZ5g8eTLeeecdSfU6ppxrNBqsXLkSR48eRVNTk8nEoVarcbPmhtH/35nUWYo0DZoQ65GcOHJycvDuu+8CANauXYvx48dLqtd5ynlGRgYCAgIwaNAgk3VcXV2tPm2ZpkET0j2rTjkHAI1Gg7a2NrS2tkKr1XZfwcCpU6dw8OBB/Pbbb/j999+7r0AIcViSjjgKCgrw5JNPYtGiReA4Dm+99ZbshiZMmIAJEyZ0X9AK2rS69kcidFp2caanzxFiLZISR2lpKQBgxoz2D2N5eblDTzc3TBKUNAixLkmJ49lnn7V1PwghvQhd3ySEyCb9TjgOSq1hhcc+diy7qnr9ahHi0Hr9EYdhkqCkQYjt2e1TVllZiUOHDqGurg4rVqyAUkknLAnprWyeODpmjgLAnDlzkJ6e3uUp4IY0Gg3N9CSkB2g0GknlGL67T7GVVFVVwdfXFydPnsSAAQMc+nIuIcQ0u31VaW1txe7du8EwDCZOnGivZgkhNmC3Iw5CSN/R66+qEELsjxIHIUQ2ShyEENkocRBCZOu3iaOqqgpVVVWorq4W5pV0/JfjOLAsq1ee4zicO3dOL6ZWqyVf9yaWobF2LA6ZOOzxoS4sLMRPP/2E/Px8rFu3DgCwYcMGsCwLtVqNjz/+GABw+PBhpKenY/Xq1fDz8wMArF+/Hg0NDdi4cSM2btwoaz3MWRcx3a2zPRNjd+Xk9qUzS8Zaahv0R0Q+h0wc9vhQA+07B8Mw4DhOiBUUFODnn38Wlpubm8FxHBiGEdpgGAalpaV46KGHoFKpZK2HnHURI3Wd7TGGUstJ7YsYS8Zaahv22t/6Eof9RZixD7VOpxOWu/tQX7hwwej7R0ZGwsvLC0qlEnFxcQCApKQkVFVVQavVYv78+QCAJ598Ep6enqitrUVOTg6mT5+OBQsWoKKiAp6enpg3b16362K4HlLXRYzcdbblGEotJ7UvYl599VWUlZUhICAAzz33nMmyHftM56O7jjYYhrG4f5aMVV+jXL169eqe7oQhlmUxcOBAhIWFIS4uDm5ubhg2bBhaWloAABMnToSrqysiIiJw//33IyYmBmfPnsWQIUMwdOhQNDU1wdnZGRMmTDD67Jddu3Zh8uTJcHV1RWpqKuLi4rBlyxY8/fTTCAgIQEpKCuLi4rBp0yaMGzcOKpUKJ06cEMrNmDEDAQEBQl0xhYWFYFkWoaGhmDRpkl5ffvrpJ4SEhODxxx+Hq6srmpqa4Ovri2HDhuHy5csYMmSI6HuKleM4DoWFhRg4cCAee+wxuLm5GW1bSrti7ydGajmpfRGTnJyMqVOnIiMjAz///LPRsXZzc0NwcDCCg4OhVqsxbNgwxMXF4fbt21AqldDpdBg2bJhF/bNkrPoah/yqkpmZiZCQEPj5+SEtLQ0AkJKSguHDhyM6Ohp79uwBAGzduhU6nQ7e3t64du2aUC46OhrR0dFCXTEuLi7IyspCVlaWcAjcXczFxcVoOTH33Xcf7rvvPjg7O2Pnzp1C3NvbW0hG27dvBwB4eHjAw8MDKpUKxcXFRt9TrJyPj49wKN/RjljbUtsVez8xUstJ7Yux9f32228RGhpqcqxPnTqFH374AZ988gkmT54MANi3bx8uX76MAwcOCDFL+mfJWPU1DvlVpeODqdPpunyAzYmJmTVrFliWxY0bN4THNVgSE9NxAo1hGDg7Owvx3NzcLnGxmJjc3FzhsNvJyUkvxjCMEDt37pzRcoaxjnZNlTPWFyl97uiLWBudY2LCw8NRXV2NgIAATJkyxWi5wMBA4ZnG5eXlGDlyJAICArrETPUkk/RjAAAIbklEQVSv8/qaGmcAettNylj1NQ65ps888wzc3d2FRzJYGhOza9cuvPHGG/jhhx/Q2NiIUaNGWRQT8+KLLwJoPxQuKCgwGTdW1pCvry8AQKlUCj8W7Bx77LHHAABTpkxBeHg4WlpacPPmTYtjctbPVJ87+icWExMfHw+g/WtRWloaXnvtNdFyjY2NUKlUuP/++4XzEmIxMU8//TQ8PT1x5coVKBQKi2P9gUOu6f79++Hv74/Q0FAcPXrU4piYwMBApKWlobm5WfgKYklMTHJyMtLT03HixAnMnTvXZNxYWUPx8fGIj4/Ho48+iszMTKOx/fv3Q6FQwNPTU29szI3JWT9z+yxm//79+Oqrr3Dw4EGTYz1x4kTEx8djwoQJOH36tNGYmNTUVCiVSpSUlOD8+fMWx/oDhzzi8Pb2xu7du8FxHDw9PS2OiRk2bBgefPBBq8XEPPHEE/j555/R0tKCo0ePYurUqUbjxsoa2r9/v3Bo3PFBEovZYwyt3WcxI0aMkDTWX331VZf3E4uJ6fhD0LmcJbF+gXdAZ86csWpMzC+//MLv2bOHP3XqlFViYs6ePSv8u7q62mTcWFlD1h4bS8bQ2n0WI3WsLWnDHmPV1zjk5di2tjYcPnwYDQ0NiIyMtDgmxs/PD6NGjQLDMMjIyMDo0aMtihlbj6ysLDQ0NGDEiBEm48bK2npsLBnDzuWs0WcxcsbaGuthq7HqaxzyHMeVK1fQ2tqKoKAg4dKrJTFjysrKkJOTY7WYIZZlMW/ePERGRur1RSxurKyhwYMHY+7cuYiMjMTmzZvtFhOzb98+aDQalJSUmCxXV1eHuXPnws3NTVg3sZiYXbt2Ye/evcjPzzc5ieu7777D5MmT9cZPLGasrkqlQkREhF5dc2P9gUMmjqqqKrz88ss4f/48XnjhBYtjYl599VXk5+eD4zihnCUxMZs2bUJKSgqOHz+u99sGsbixssZERETAx8fH7rHOpk+fjl9//RUDBgzA0qVLjZYrKSkBABQVFQnjJRYT85e//AUsy+Ls2bOIiIgwWW79+vXIzMwUHoouFhPz0ksvYdq0afj6669x5swZi2P9gUPeOjA5ORk+Pj6or6+Hj48PkpKSLIqJ0el0yMvLQ0FBAVQqFebPn29RTExBQYFwqba7fxv7/4Z27twpTKtmWRavvvqqXWJiFi9ejCFDhsDb2xsMwxgd623btsHZ2RlarRbOzs5ISkoSjYnZtWsXnJycoNPpcPPmTaMPPN+2bRueeOIJREVFmYyJ+eSTT6BSqRATE4OxY8eCYRiLYv2BQyYOQohjc8ivKoQQx0aJgxAiGyWOfqStrQ319fU9/h6WqK6u7rG2yV2UOHqZ5cuXm103KysLRUVFZtWtra3FF198IbzHvn37zP4QZ2Rk4MMPP9SL3bx5E+np6Whra8O7776Lv//978K9ODrrfEMkOUxdLu6wadMm/Prrr0hJSUFra6tZ7fQXDjnlvD+6ePEi9u3bhzt37uCdd97Bli1bALRP/164cCH+9a9/wc3NDY2Njaivr8e///1vuLi4YOLEiWhqakJ+fj5cXV3xwAMPCFO/i4uLkZKSgubmZsybNw95eXlQqVR4+OGHAbR/UJqbmwEAzz//PH755Rd8++230Gg0+Mc//oE9e/agsbERrq6umD17NiorK1FXVweVSgVfX1+o1Wp89NFHuH37Nniexz//+U/MnDkTkyZNQm1tLcTmFl64cAG//fZblxsb7dmzB6+88go+++wzTJo0CY8++ihu3bqFixcvCr9AzsvLQ3FxMcrKypCWlgaO4xAeHo6YmBhs27YNU6dORWZmJsaOHYvy8nIMHToU5eXlWLt2LcrLy7F//369cYqMjNQbnw6PP/449u7diwULFlh7M/cZdMThIDpuSbdy5UqUlpYiKCgIq1atwu3bt3H06FE88sgjePvtt+Hi4oLMzEzMmzcPa9asweHDhwG0/2jsb3/7G7777jvhPb29vfHss89i1KhROH/+PO6///4u96WYPXs2li5dir179+LIkSNYs2YN5s+fj4yMDFRWViIiIgLTp08Xynd+D7VajZqaGqxatQqhoaEoKipCWFgYli5diqamJtH1HDt2LBISErrEO+6oVlZWhjFjxgBonzX6+OOPC2ViY2MxdOhQ/Pbbb6iuroa7u7vwq9zx48dj1qxZ8PDwwF//+lewLIslS5Z0uatX53EyHJ8OkZGRKCwslLTd+itKHA6CZVkwDIOWlhbcc889ovMBGIaBQqEQbl8H3L2prpubG5ydnfVum3fgwAFUVFRg+PDhwm31DPE8j7a2NiiVSr04x3GYP38+hg4dijVr1ggfvs7vYfieHMcJd/LqfG+K4uJivPnmm9i7d6/R9e9of/Dgwbh48SIA4NatW9i+fTv+97//4c033xTOreh0Ojz00ENYvny5kMQ8PDwAAO7u7sJ4ABDGq0PncTIcn8760701zEGj4yD+9Kc/YdWqVWhtbcWqVatQU1OD//73vwgMDMTUqVPxn//8BxcuXADLspg2bRref/99ZGRkYNq0aWhsbBR9z5CQEJw7dw7Ozs5wc3NDWFgYPv/8c+GrCtA+SQoAlixZgsuXL+Of//wndDodli1bhh07dqClpQWDBw8W7jXR8R6DBg2Cm5sbAgIC8P7774NhGKO/Vxk2bBiSk5NNrn9QUBAaGxvx/PPPY/Xq1cjIyEBLSwveeustvZsl3bx5E+Hh4cjMzERRURHuvfdeBAUFyRprY+PT8UvgsrIyREdHm/We/QVNAOvHNm3ahBkzZjjEj7Pq6+vxzTffYNasWT3dFaSmpiIxMdHkHeT6O0ochBDZ6BwHIUQ2ShyEENkocRBCZKPEQQiRjRIHIUS2/wfYOIChvCy/aQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 288x205.714 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "def plot_codon_composition_cterm(data, maskData, pvalAnnotTable, maskDataSmallN, terminus, speciesName, refCodonTableDf):\n",
    "    \n",
    "    addEmptyRowBetweenGroups = True\n",
    "    addEmptyRowBetweenGroupsFillValue = np.nan\n",
    "    plotData = data.copy()\n",
    "    nameIndex = plotData.index.name\n",
    "    nameCol = plotData.columns.name\n",
    "    stopCodonList = list(refCodonTableDf[refCodonTableDf['aa'] == '*'].index)\n",
    "\n",
    "    plotData = plotData[-1]\n",
    "    maskData = maskData[-1]\n",
    "    maskDataSmallN = maskDataSmallN[-1]\n",
    "    \n",
    "    # Remove values that are not significant using mask\n",
    "    plotData.loc[maskData[maskData].index] = np.nan\n",
    "    plotData = plotData.to_frame()\n",
    "    plotData = plotData.loc[plotData.index.map(lambda x: x not in stopCodonList)]\n",
    "    plotData = sort_codon_index(plotData, refCodonTableDf,\n",
    "                                addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                                addEmptyRowBetweenGroups=addEmptyRowBetweenGroups)\n",
    "    plotData.index.name = nameIndex\n",
    "    plotData.columns.name = nameCol\n",
    "    plotData = plotData[-1]\n",
    "#     print(\"plotData:\\n\", type(plotData), plotData)\n",
    "\n",
    "    ylabel = '$\\log_2$(odds ratio)'\n",
    "    ymin, ymax = -3, 3\n",
    "\n",
    "    n = 34\n",
    "    n = 44\n",
    "    plotDf1 = plotData.iloc[:n]\n",
    "    plotDf2 = plotData.iloc[n+1:]\n",
    "    m = n - len(plotDf2)\n",
    "    plotDf2 = plotDf2.copy().append(pd.Series(m*[np.nan], index=m*['']))\n",
    "\n",
    "    figsizeCoeff = 0.6\n",
    "    aspectratio = 1.4\n",
    "    imagesize = 4.0\n",
    "    linewidth1 = 0.5\n",
    "    figsize = (imagesize, imagesize/aspectratio)\n",
    "\n",
    "    fig, axes = plt.subplots(2, 1, figsize=figsize)\n",
    "    plt.subplots_adjust(hspace=0.5)\n",
    "    w = 0.8\n",
    "    \n",
    "    for i, ax in enumerate(axes):\n",
    "        if i == 0:\n",
    "            plotDf3 = plotDf1\n",
    "        elif i == 1:\n",
    "            plotDf3 = plotDf2\n",
    "        norm = matplotlib.colors.Normalize(vmin=-vmax, vmax=vmax)\n",
    "        colors = cmap(norm(plotDf3.values))\n",
    "        plotDf3.plot.bar(ax=ax, width=w, color=colors)\n",
    "        ax.set_ylabel(ylabel)\n",
    "        ax.set_ylim(ymin, ymax)\n",
    "        ax.xaxis.grid(False)\n",
    "        family = 'Liberation Mono'\n",
    "        ticks_font = FontProperties(family=family, size='x-small')\n",
    "        for tick in ax.get_xticklabels():\n",
    "            tick.set_fontproperties(ticks_font)\n",
    "            tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "        if i == 1:\n",
    "            ax.set_xlabel('codon at position -1 (C-terminal)')\n",
    "        ax.set_yticks(list(range(-3,4)))\n",
    "\n",
    "    \n",
    "    title = terminus + '-terminal, ' + speciesName\n",
    "#     axes[0].set_title(title)\n",
    "    return fig\n",
    "               \n",
    "#     ny = maskDataSmallN.shape[0]\n",
    "#     hatch = '////' if figsizeCoeff > 1.1 else '///'\n",
    "#     for (i,j), value in np.ndenumerate(maskDataSmallN.values):\n",
    "#         if value:\n",
    "#             ax.add_patch( matplotlib.patches.Rectangle((j, ny - 1 - i), 1, 1,\n",
    "#                                                        edgecolor='w', facecolor=colorSmallN, hatch=hatch,\n",
    "#                                                        linewidth=linewidth1) )\n",
    "\n",
    "\n",
    "\n",
    "fig = plot_codon_composition_cterm(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus,\n",
    "                             plotTitle, refCodonTableDf)\n",
    "filenamePlot = outputFilePrefix + '_composition_bias_codon_' + terminus + 'terminal_barplot'\n",
    "print(os.path.join(plotsFolder, speciesSetName, filenamePlot + '.svg'))\n",
    "fig.savefig(os.path.join(plotsFolder, speciesSetName, filenamePlot + '.svg'), dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot aa pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = get_divergent_color_map(name=\"RdBu_r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "C    0.00636876\n",
       "W      0.010118\n",
       "H     0.0151239\n",
       "M     0.0160864\n",
       "P      0.028652\n",
       "R      0.029483\n",
       "Q     0.0362426\n",
       "Y     0.0409585\n",
       "G     0.0510767\n",
       "A     0.0526513\n",
       "T     0.0535492\n",
       "F     0.0543569\n",
       "D     0.0551014\n",
       "V     0.0573075\n",
       "E     0.0670141\n",
       "S     0.0691151\n",
       "N     0.0727995\n",
       "I     0.0920151\n",
       "K     0.0943135\n",
       "L     0.0976665\n",
       "Name: count, dtype: object"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freqBulk = oddsRatioDf.loc['bulk'].loc[0].loc['count']\n",
    "freqBulk = freqBulk / freqBulk.sum()\n",
    "freqBulk.sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col0 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col1 {\n",
       "            background-color:  #f8f1ec;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col2 {\n",
       "            background-color:  #cae1ee;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col3 {\n",
       "            background-color:  #f9f0eb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col4 {\n",
       "            background-color:  #f5ab8a;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col5 {\n",
       "            background-color:  #6bacd1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col6 {\n",
       "            background-color:  #dfecf3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col7 {\n",
       "            background-color:  #f3f5f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col8 {\n",
       "            background-color:  #fbe5d9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col9 {\n",
       "            background-color:  #99c9e0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col10 {\n",
       "            background-color:  #edf2f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col11 {\n",
       "            background-color:  #e9f0f4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col12 {\n",
       "            background-color:  #fcdfce;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col13 {\n",
       "            background-color:  #fcd3bc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col14 {\n",
       "            background-color:  #f7b799;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col15 {\n",
       "            background-color:  #8fc3dd;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col16 {\n",
       "            background-color:  #e98c6f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col17 {\n",
       "            background-color:  #f7f5f3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col18 {\n",
       "            background-color:  #e6eff4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col19 {\n",
       "            background-color:  #ea8e71;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col0 {\n",
       "            background-color:  #a7d0e4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col1 {\n",
       "            background-color:  #f6af8f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col2 {\n",
       "            background-color:  #78b5d5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col3 {\n",
       "            background-color:  #db6b55;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col4 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col5 {\n",
       "            background-color:  #3783bb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col6 {\n",
       "            background-color:  #c03338;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col7 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col8 {\n",
       "            background-color:  #fddcc9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col9 {\n",
       "            background-color:  #2f79b6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col10 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col11 {\n",
       "            background-color:  #cee4ef;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col12 {\n",
       "            background-color:  #b7d8e9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col13 {\n",
       "            background-color:  #bad9e9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col14 {\n",
       "            background-color:  #e27d63;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col15 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col16 {\n",
       "            background-color:  #d5e7f1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col17 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col18 {\n",
       "            background-color:  #c7e0ed;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col19 {\n",
       "            background-color:  #e8896d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col0 {\n",
       "            background-color:  #fdddcb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col1 {\n",
       "            background-color:  #aed3e6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col2 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col3 {\n",
       "            background-color:  #87beda;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col4 {\n",
       "            background-color:  #b7d8e9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col5 {\n",
       "            background-color:  #f1f4f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col6 {\n",
       "            background-color:  #64a8ce;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col7 {\n",
       "            background-color:  #f0f4f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col8 {\n",
       "            background-color:  #b82431;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col9 {\n",
       "            background-color:  #dbeaf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col10 {\n",
       "            background-color:  #fbd0b9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col11 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col12 {\n",
       "            background-color:  #e6eff4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col13 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col14 {\n",
       "            background-color:  #e1ecf3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col15 {\n",
       "            background-color:  #cd4e44;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col16 {\n",
       "            background-color:  #9ecbe1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col17 {\n",
       "            background-color:  #e58368;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col18 {\n",
       "            background-color:  #f1f4f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col19 {\n",
       "            background-color:  #e8896d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col0 {\n",
       "            background-color:  #134a86;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col1 {\n",
       "            background-color:  #fbd0b9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col2 {\n",
       "            background-color:  #94c6df;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col3 {\n",
       "            background-color:  #a0cce2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col4 {\n",
       "            background-color:  #de725b;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col5 {\n",
       "            background-color:  #f8f3f0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col6 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col7 {\n",
       "            background-color:  #b3d6e7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col8 {\n",
       "            background-color:  #f9c3a8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col9 {\n",
       "            background-color:  #72b1d3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col10 {\n",
       "            background-color:  #cce3ef;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col11 {\n",
       "            background-color:  #fcdecc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col12 {\n",
       "            background-color:  #bfdceb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col13 {\n",
       "            background-color:  #deebf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col14 {\n",
       "            background-color:  #f3f5f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col15 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col16 {\n",
       "            background-color:  #327cb7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col17 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col18 {\n",
       "            background-color:  #f9efe9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col19 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col0 {\n",
       "            background-color:  #c2ddec;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col1 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col2 {\n",
       "            background-color:  #4c99c6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col3 {\n",
       "            background-color:  #185494;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col4 {\n",
       "            background-color:  #e58368;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col5 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col6 {\n",
       "            background-color:  #cce3ef;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col7 {\n",
       "            background-color:  #d05347;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col8 {\n",
       "            background-color:  #f8bb9e;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col9 {\n",
       "            background-color:  #4190c1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col10 {\n",
       "            background-color:  #246aad;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col11 {\n",
       "            background-color:  #1b5a9c;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col12 {\n",
       "            background-color:  #f7f6f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col13 {\n",
       "            background-color:  #fbe2d4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col14 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col15 {\n",
       "            background-color:  #fac7ae;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col16 {\n",
       "            background-color:  #f8f1ec;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col17 {\n",
       "            background-color:  #d5e7f1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col18 {\n",
       "            background-color:  #559ec9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col19 {\n",
       "            background-color:  #deebf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col0 {\n",
       "            background-color:  #8cc1dc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col1 {\n",
       "            background-color:  #f9c5ab;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col2 {\n",
       "            background-color:  #75b3d4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col3 {\n",
       "            background-color:  #e3eef3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col4 {\n",
       "            background-color:  #edf2f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col5 {\n",
       "            background-color:  #760521;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col6 {\n",
       "            background-color:  #edf2f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col7 {\n",
       "            background-color:  #fce1d2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col8 {\n",
       "            background-color:  #fbe5d9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col9 {\n",
       "            background-color:  #c5dfec;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col10 {\n",
       "            background-color:  #f7b99c;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col11 {\n",
       "            background-color:  #92c4de;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col12 {\n",
       "            background-color:  #81bad8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col13 {\n",
       "            background-color:  #307ab6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col14 {\n",
       "            background-color:  #edf2f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col15 {\n",
       "            background-color:  #f5ad8c;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col16 {\n",
       "            background-color:  #7bb6d6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col17 {\n",
       "            background-color:  #df755d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col18 {\n",
       "            background-color:  #f7f6f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col19 {\n",
       "            background-color:  #fac9b0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col0 {\n",
       "            background-color:  #99c9e0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col1 {\n",
       "            background-color:  #fac7ae;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col2 {\n",
       "            background-color:  #cce3ef;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col3 {\n",
       "            background-color:  #61a6cd;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col4 {\n",
       "            background-color:  #2f79b6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col5 {\n",
       "            background-color:  #e27d63;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col6 {\n",
       "            background-color:  #a51429;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col7 {\n",
       "            background-color:  #f3a583;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col8 {\n",
       "            background-color:  #92c4de;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col9 {\n",
       "            background-color:  #4997c5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col10 {\n",
       "            background-color:  #f5a987;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col11 {\n",
       "            background-color:  #cc4b43;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col12 {\n",
       "            background-color:  #a2cde3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col13 {\n",
       "            background-color:  #aed3e6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col14 {\n",
       "            background-color:  #e48065;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col15 {\n",
       "            background-color:  #d8e8f1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col16 {\n",
       "            background-color:  #a9d1e5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col17 {\n",
       "            background-color:  #f9eee7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col18 {\n",
       "            background-color:  #84bcd9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col19 {\n",
       "            background-color:  #f8f3f0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col0 {\n",
       "            background-color:  #0b3c72;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col1 {\n",
       "            background-color:  #df755d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col2 {\n",
       "            background-color:  #c2ddec;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col3 {\n",
       "            background-color:  #f7f6f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col4 {\n",
       "            background-color:  #fae7dc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col5 {\n",
       "            background-color:  #f9c5ab;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col6 {\n",
       "            background-color:  #2166aa;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col7 {\n",
       "            background-color:  #af182b;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col8 {\n",
       "            background-color:  #b82431;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col9 {\n",
       "            background-color:  #67aacf;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col10 {\n",
       "            background-color:  #78b5d5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col11 {\n",
       "            background-color:  #d86551;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col12 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col13 {\n",
       "            background-color:  #0b3c72;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col14 {\n",
       "            background-color:  #f7f5f3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col15 {\n",
       "            background-color:  #f6af8f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col16 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col17 {\n",
       "            background-color:  #f8bfa3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col18 {\n",
       "            background-color:  #fbe5d9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col19 {\n",
       "            background-color:  #c13639;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col0 {\n",
       "            background-color:  #f9c3a8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col1 {\n",
       "            background-color:  #860a24;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col2 {\n",
       "            background-color:  #276eb0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col3 {\n",
       "            background-color:  #3f8ec1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col4 {\n",
       "            background-color:  #e27d63;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col5 {\n",
       "            background-color:  #246aad;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col6 {\n",
       "            background-color:  #ce5046;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col7 {\n",
       "            background-color:  #d55e4c;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col8 {\n",
       "            background-color:  #d25849;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col9 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col10 {\n",
       "            background-color:  #5ba2cb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col11 {\n",
       "            background-color:  #84bcd9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col12 {\n",
       "            background-color:  #ed9676;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col13 {\n",
       "            background-color:  #78b5d5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col14 {\n",
       "            background-color:  #d25849;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col15 {\n",
       "            background-color:  #92c4de;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col16 {\n",
       "            background-color:  #fbe2d4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col17 {\n",
       "            background-color:  #256baf;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col18 {\n",
       "            background-color:  #a5cfe3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col19 {\n",
       "            background-color:  #94c6df;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col0 {\n",
       "            background-color:  #10447d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col1 {\n",
       "            background-color:  #99c9e0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col2 {\n",
       "            background-color:  #e48065;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col3 {\n",
       "            background-color:  #e98c6f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col4 {\n",
       "            background-color:  #fcd3bc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col5 {\n",
       "            background-color:  #890b24;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col6 {\n",
       "            background-color:  #a7d0e4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col7 {\n",
       "            background-color:  #10447d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col8 {\n",
       "            background-color:  #f9c3a8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col9 {\n",
       "            background-color:  #e98c6f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col10 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col11 {\n",
       "            background-color:  #fddac7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col12 {\n",
       "            background-color:  #f9f0eb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col13 {\n",
       "            background-color:  #cae1ee;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col14 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col15 {\n",
       "            background-color:  #f6af8f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col16 {\n",
       "            background-color:  #276eb0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col17 {\n",
       "            background-color:  #b11a2c;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col18 {\n",
       "            background-color:  #e5eef4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col19 {\n",
       "            background-color:  #f1f4f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col0 {\n",
       "            background-color:  #f9ece3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col1 {\n",
       "            background-color:  #dceaf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col2 {\n",
       "            background-color:  #f6f6f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col3 {\n",
       "            background-color:  #92c4de;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col4 {\n",
       "            background-color:  #75b3d4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col5 {\n",
       "            background-color:  #fddac7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col6 {\n",
       "            background-color:  #eb9173;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col7 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col8 {\n",
       "            background-color:  #d8e8f1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col9 {\n",
       "            background-color:  #c2ddec;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col10 {\n",
       "            background-color:  #cc4b43;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col11 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col12 {\n",
       "            background-color:  #df755d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col13 {\n",
       "            background-color:  #a51429;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col14 {\n",
       "            background-color:  #dd7059;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col15 {\n",
       "            background-color:  #81bad8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col16 {\n",
       "            background-color:  #e98c6f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col17 {\n",
       "            background-color:  #e8896d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col18 {\n",
       "            background-color:  #facbb3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col19 {\n",
       "            background-color:  #c2383a;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col0 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col1 {\n",
       "            background-color:  #f6f6f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col2 {\n",
       "            background-color:  #e1ecf3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col3 {\n",
       "            background-color:  #75b3d4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col4 {\n",
       "            background-color:  #ca4842;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col5 {\n",
       "            background-color:  #f6b394;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col6 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col7 {\n",
       "            background-color:  #f8bfa3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col8 {\n",
       "            background-color:  #f0f4f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col9 {\n",
       "            background-color:  #61a6cd;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col10 {\n",
       "            background-color:  #edf2f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col11 {\n",
       "            background-color:  #185494;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col12 {\n",
       "            background-color:  #78b5d5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col13 {\n",
       "            background-color:  #e48065;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col14 {\n",
       "            background-color:  #f8bb9e;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col15 {\n",
       "            background-color:  #3f8ec1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col16 {\n",
       "            background-color:  #e3eef3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col17 {\n",
       "            background-color:  #fcd9c4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col18 {\n",
       "            background-color:  #d0e5f0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col19 {\n",
       "            background-color:  #fcd7c2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col0 {\n",
       "            background-color:  #f2a07e;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col1 {\n",
       "            background-color:  #e0775f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col2 {\n",
       "            background-color:  #9ecbe1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col3 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col4 {\n",
       "            background-color:  #2064a9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col5 {\n",
       "            background-color:  #fbe2d4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col6 {\n",
       "            background-color:  #e27d63;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col7 {\n",
       "            background-color:  #edf2f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col8 {\n",
       "            background-color:  #e9f0f4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col9 {\n",
       "            background-color:  #f1f4f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col10 {\n",
       "            background-color:  #fbe3d5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col11 {\n",
       "            background-color:  #97c7df;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col12 {\n",
       "            background-color:  #d86551;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col13 {\n",
       "            background-color:  #b5d7e8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col14 {\n",
       "            background-color:  #c2383a;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col15 {\n",
       "            background-color:  #f5a987;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col16 {\n",
       "            background-color:  #f6b394;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col17 {\n",
       "            background-color:  #eff3f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col18 {\n",
       "            background-color:  #8ac0db;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col19 {\n",
       "            background-color:  #e3eef3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col0 {\n",
       "            background-color:  #67aacf;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col1 {\n",
       "            background-color:  #d8e8f1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col2 {\n",
       "            background-color:  #337eb8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col3 {\n",
       "            background-color:  #e27d63;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col4 {\n",
       "            background-color:  #fbe4d7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col5 {\n",
       "            background-color:  #de725b;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col6 {\n",
       "            background-color:  #f1f4f6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col7 {\n",
       "            background-color:  #cae1ee;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col8 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col9 {\n",
       "            background-color:  #e3eef3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col10 {\n",
       "            background-color:  #f8bda1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col11 {\n",
       "            background-color:  #eaf1f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col12 {\n",
       "            background-color:  #af182b;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col13 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col14 {\n",
       "            background-color:  #fac9b0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col15 {\n",
       "            background-color:  #d0e5f0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col16 {\n",
       "            background-color:  #e3eef3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col17 {\n",
       "            background-color:  #fcd9c4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col18 {\n",
       "            background-color:  #4f9bc7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col19 {\n",
       "            background-color:  #e1ecf3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col0 {\n",
       "            background-color:  #559ec9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col1 {\n",
       "            background-color:  #f6b192;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col2 {\n",
       "            background-color:  #2870b1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col3 {\n",
       "            background-color:  #db6b55;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col4 {\n",
       "            background-color:  #fae8de;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col5 {\n",
       "            background-color:  #ad162a;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col6 {\n",
       "            background-color:  #fbe3d5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col7 {\n",
       "            background-color:  #de725b;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col8 {\n",
       "            background-color:  #2870b1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col9 {\n",
       "            background-color:  #deebf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col10 {\n",
       "            background-color:  #deebf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col11 {\n",
       "            background-color:  #fcd7c2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col12 {\n",
       "            background-color:  #e58368;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col13 {\n",
       "            background-color:  #d0e5f0;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col14 {\n",
       "            background-color:  #fcd9c4;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col15 {\n",
       "            background-color:  #195696;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col16 {\n",
       "            background-color:  #7eb8d7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col17 {\n",
       "            background-color:  #fae7dc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col18 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col19 {\n",
       "            background-color:  #e1ecf3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col0 {\n",
       "            background-color:  #8ac0db;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col1 {\n",
       "            background-color:  #e0775f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col2 {\n",
       "            background-color:  #1d5ea1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col3 {\n",
       "            background-color:  #f7b799;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col4 {\n",
       "            background-color:  #f3a583;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col5 {\n",
       "            background-color:  #d76350;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col6 {\n",
       "            background-color:  #f6b192;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col7 {\n",
       "            background-color:  #f5a987;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col8 {\n",
       "            background-color:  #3986bc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col9 {\n",
       "            background-color:  #84bcd9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col10 {\n",
       "            background-color:  #e3eef3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col11 {\n",
       "            background-color:  #f7f5f3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col12 {\n",
       "            background-color:  #e27d63;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col13 {\n",
       "            background-color:  #fbceb6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col14 {\n",
       "            background-color:  #f8bfa3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col15 {\n",
       "            background-color:  #16508e;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col16 {\n",
       "            background-color:  #eff3f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col17 {\n",
       "            background-color:  #eff3f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col18 {\n",
       "            background-color:  #0b3c72;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col19 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col0 {\n",
       "            background-color:  #dbeaf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col1 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col2 {\n",
       "            background-color:  #f8f4f2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col3 {\n",
       "            background-color:  #84bcd9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col4 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col5 {\n",
       "            background-color:  #fbe5d9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col6 {\n",
       "            background-color:  #fcdfce;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col7 {\n",
       "            background-color:  #a9d1e5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col8 {\n",
       "            background-color:  #fddac7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col9 {\n",
       "            background-color:  #bad9e9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col10 {\n",
       "            background-color:  #d86551;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col11 {\n",
       "            background-color:  #a21328;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col12 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col13 {\n",
       "            background-color:  #b51e2e;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col14 {\n",
       "            background-color:  #d55e4c;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col15 {\n",
       "            background-color:  #fbe5d9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col16 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col17 {\n",
       "            background-color:  #f7b99c;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col18 {\n",
       "            background-color:  #3986bc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col19 {\n",
       "            background-color:  #4392c2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col0 {\n",
       "            background-color:  #144c88;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col1 {\n",
       "            background-color:  #f7f5f3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col2 {\n",
       "            background-color:  #09386d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col3 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col4 {\n",
       "            background-color:  #fae7dc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col5 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col6 {\n",
       "            background-color:  #a2cde3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col7 {\n",
       "            background-color:  #f6b192;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col8 {\n",
       "            background-color:  #256baf;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col9 {\n",
       "            background-color:  #e2edf3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col10 {\n",
       "            background-color:  #faebe2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col11 {\n",
       "            background-color:  #3986bc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col12 {\n",
       "            background-color:  #faebe2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col13 {\n",
       "            background-color:  #fbe6db;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col14 {\n",
       "            background-color:  #fcdfce;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col15 {\n",
       "            background-color:  #fdddcb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col16 {\n",
       "            background-color:  #acd2e5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col17 {\n",
       "            background-color:  #fce1d2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col18 {\n",
       "            background-color:  #f7f6f5;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col19 {\n",
       "            background-color:  #fcd3bc;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col0 {\n",
       "            background-color:  #5ba2cb;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col1 {\n",
       "            background-color:  #f8bda1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col2 {\n",
       "            background-color:  #246aad;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col3 {\n",
       "            background-color:  #fcd5bf;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col4 {\n",
       "            background-color:  #dceaf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col5 {\n",
       "            background-color:  #b7d8e9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col6 {\n",
       "            background-color:  #ed9676;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col7 {\n",
       "            background-color:  #c8433f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col8 {\n",
       "            background-color:  #aed3e6;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col9 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col10 {\n",
       "            background-color:  #b0d4e7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col11 {\n",
       "            background-color:  #1d5ea1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col12 {\n",
       "            background-color:  #61a6cd;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col13 {\n",
       "            background-color:  #8fc3dd;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col14 {\n",
       "            background-color:  #f3a583;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col15 {\n",
       "            background-color:  #246aad;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col16 {\n",
       "            background-color:  #f8bfa3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col17 {\n",
       "            background-color:  #df755d;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col18 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col19 {\n",
       "            background-color:  #fbe4d7;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col0 {\n",
       "            background-color:  #f8bb9e;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col1 {\n",
       "            background-color:  #f8f1ec;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col2 {\n",
       "            background-color:  #f8f2ee;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col3 {\n",
       "            background-color:  #cd4e44;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col4 {\n",
       "            background-color:  #faebe2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col5 {\n",
       "            background-color:  #fcd7c2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col6 {\n",
       "            background-color:  #337eb8;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col7 {\n",
       "            background-color:  #deebf2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col8 {\n",
       "            background-color:  #730421;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col9 {\n",
       "            background-color:  #f8f4f2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col10 {\n",
       "            background-color:  #e1ecf3;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col11 {\n",
       "            background-color:  #cee4ef;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col12 {\n",
       "            background-color:  #c7e0ed;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col13 {\n",
       "            background-color:  #4190c1;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col14 {\n",
       "            background-color:  #347fb9;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col15 {\n",
       "            background-color:  #760521;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col16 {\n",
       "            background-color:  #1c5c9f;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col17 {\n",
       "            background-color:  #cc4b43;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col18 {\n",
       "            background-color:  #fce1d2;\n",
       "        }    #T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col19 {\n",
       "            background-color:  #d55e4c;\n",
       "        }</style>  \n",
       "<table id=\"T_80d40414_379c_11ea_a970_e79fc7d1477b\" > \n",
       "<thead>    <tr> \n",
       "        <th class=\"index_name level0\" >aa1</th> \n",
       "        <th class=\"col_heading level0 col0\" >R</th> \n",
       "        <th class=\"col_heading level0 col1\" >K</th> \n",
       "        <th class=\"col_heading level0 col2\" >H</th> \n",
       "        <th class=\"col_heading level0 col3\" >D</th> \n",
       "        <th class=\"col_heading level0 col4\" >E</th> \n",
       "        <th class=\"col_heading level0 col5\" >S</th> \n",
       "        <th class=\"col_heading level0 col6\" >T</th> \n",
       "        <th class=\"col_heading level0 col7\" >N</th> \n",
       "        <th class=\"col_heading level0 col8\" >Q</th> \n",
       "        <th class=\"col_heading level0 col9\" >C</th> \n",
       "        <th class=\"col_heading level0 col10\" >G</th> \n",
       "        <th class=\"col_heading level0 col11\" >P</th> \n",
       "        <th class=\"col_heading level0 col12\" >A</th> \n",
       "        <th class=\"col_heading level0 col13\" >V</th> \n",
       "        <th class=\"col_heading level0 col14\" >I</th> \n",
       "        <th class=\"col_heading level0 col15\" >L</th> \n",
       "        <th class=\"col_heading level0 col16\" >M</th> \n",
       "        <th class=\"col_heading level0 col17\" >F</th> \n",
       "        <th class=\"col_heading level0 col18\" >W</th> \n",
       "        <th class=\"col_heading level0 col19\" >Y</th> \n",
       "    </tr>    <tr> \n",
       "        <th class=\"index_name level0\" >aa0</th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "        <th class=\"blank\" ></th> \n",
       "    </tr></thead> \n",
       "<tbody>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row0\" class=\"row_heading level0 row0\" >R</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col0\" class=\"data row0 col0\" >0.347353</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col1\" class=\"data row0 col1\" >-0.0831331</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col2\" class=\"data row0 col2\" >0.109164</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col3\" class=\"data row0 col3\" >-0.000113622</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col4\" class=\"data row0 col4\" >0.074173</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col5\" class=\"data row0 col5\" >-0.204501</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col6\" class=\"data row0 col6\" >-0.0713025</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col7\" class=\"data row0 col7\" >-0.105162</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col8\" class=\"data row0 col8\" >0.0606686</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col9\" class=\"data row0 col9\" >-0.0810783</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col10\" class=\"data row0 col10\" >0.0163642</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col11\" class=\"data row0 col11\" >0.0100345</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col12\" class=\"data row0 col12\" >0.0232694</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col13\" class=\"data row0 col13\" >0.0670559</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col14\" class=\"data row0 col14\" >0.00745139</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col15\" class=\"data row0 col15\" >-0.0427704</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col16\" class=\"data row0 col16\" >0.327016</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col17\" class=\"data row0 col17\" >-0.037158</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col18\" class=\"data row0 col18\" >0.087046</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow0_col19\" class=\"data row0 col19\" >0.0966359</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row1\" class=\"row_heading level0 row1\" >K</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col0\" class=\"data row1 col0\" >-0.00168052</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col1\" class=\"data row1 col1\" >0.0144942</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col2\" class=\"data row1 col2\" >-0.0172922</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col3\" class=\"data row1 col3\" >0.132474</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col4\" class=\"data row1 col4\" >0.296894</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col5\" class=\"data row1 col5\" >-0.254213</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col6\" class=\"data row1 col6\" >0.188533</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col7\" class=\"data row1 col7\" >0.22745</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col8\" class=\"data row1 col8\" >0.086001</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col9\" class=\"data row1 col9\" >-0.52257</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col10\" class=\"data row1 col10\" >-0.345671</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col11\" class=\"data row1 col11\" >-0.0544568</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col12\" class=\"data row1 col12\" >-0.0871469</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col13\" class=\"data row1 col13\" >-0.0440054</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col14\" class=\"data row1 col14\" >0.0606022</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col15\" class=\"data row1 col15\" >-0.12852</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col16\" class=\"data row1 col16\" >-0.00178092</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col17\" class=\"data row1 col17\" >-0.373451</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col18\" class=\"data row1 col18\" >0.0232092</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow1_col19\" class=\"data row1 col19\" >0.101793</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row2\" class=\"row_heading level0 row2\" >H</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col0\" class=\"data row2 col0\" >0.136689</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col1\" class=\"data row2 col1\" >-0.192592</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col2\" class=\"data row2 col2\" >0.730491</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col3\" class=\"data row2 col3\" >-0.121996</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col4\" class=\"data row2 col4\" >-0.170698</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col5\" class=\"data row2 col5\" >-0.0728093</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col6\" class=\"data row2 col6\" >-0.189167</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col7\" class=\"data row2 col7\" >-0.112277</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col8\" class=\"data row2 col8\" >0.312112</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col9\" class=\"data row2 col9\" >0.214965</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col10\" class=\"data row2 col10\" >0.128236</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col11\" class=\"data row2 col11\" >0.506316</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col12\" class=\"data row2 col12\" >-0.0396103</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col13\" class=\"data row2 col13\" >-0.200989</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col14\" class=\"data row2 col14\" >-0.12345</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col15\" class=\"data row2 col15\" >0.116805</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col16\" class=\"data row2 col16\" >-0.0912224</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col17\" class=\"data row2 col17\" >0.129667</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col18\" class=\"data row2 col18\" >0.111681</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow2_col19\" class=\"data row2 col19\" >0.0995729</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row3\" class=\"row_heading level0 row3\" >D</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col0\" class=\"data row3 col0\" >-0.152574</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col1\" class=\"data row3 col1\" >-0.0216736</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col2\" class=\"data row3 col2\" >0.0205548</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col3\" class=\"data row3 col3\" >-0.102968</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col4\" class=\"data row3 col4\" >0.137105</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col5\" class=\"data row3 col5\" >-0.0585567</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col6\" class=\"data row3 col6\" >-0.329425</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col7\" class=\"data row3 col7\" >-0.197447</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col8\" class=\"data row3 col8\" >0.122443</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col9\" class=\"data row3 col9\" >-0.217264</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col10\" class=\"data row3 col10\" >-0.0488788</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col11\" class=\"data row3 col11\" >0.128647</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col12\" class=\"data row3 col12\" >-0.0820632</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col13\" class=\"data row3 col13\" >-0.0147565</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col14\" class=\"data row3 col14\" >-0.0941702</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col15\" class=\"data row3 col15\" >0.165772</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col16\" class=\"data row3 col16\" >-0.262968</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col17\" class=\"data row3 col17\" >0.29373</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col18\" class=\"data row3 col18\" >0.147175</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow3_col19\" class=\"data row3 col19\" >0.248615</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row4\" class=\"row_heading level0 row4\" >E</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col0\" class=\"data row4 col0\" >0.0205313</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col1\" class=\"data row4 col1\" >0.206401</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col2\" class=\"data row4 col2\" >-0.0747136</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col3\" class=\"data row4 col3\" >-0.231935</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col4\" class=\"data row4 col4\" >0.118726</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col5\" class=\"data row4 col5\" >-0.338117</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col6\" class=\"data row4 col6\" >-0.0992213</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col7\" class=\"data row4 col7\" >0.114445</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col8\" class=\"data row4 col8\" >0.134836</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col9\" class=\"data row4 col9\" >-0.390555</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col10\" class=\"data row4 col10\" >-0.269124</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col11\" class=\"data row4 col11\" >-0.353658</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col12\" class=\"data row4 col12\" >-0.0177795</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col13\" class=\"data row4 col13\" >0.0478321</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col14\" class=\"data row4 col14\" >0.19317</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col15\" class=\"data row4 col15\" >0.0598483</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col16\" class=\"data row4 col16\" >0.109407</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col17\" class=\"data row4 col17\" >-0.100041</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col18\" class=\"data row4 col18\" >-0.115341</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow4_col19\" class=\"data row4 col19\" >-0.0839624</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row5\" class=\"row_heading level0 row5\" >S</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col0\" class=\"data row5 col0\" >-0.0227613</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col1\" class=\"data row5 col1\" >-0.0095323</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col2\" class=\"data row5 col2\" >-0.0215349</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col3\" class=\"data row5 col3\" >-0.0385939</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col4\" class=\"data row5 col4\" >-0.0846452</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col5\" class=\"data row5 col5\" >0.202558</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col6\" class=\"data row5 col6\" >-0.04915</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col7\" class=\"data row5 col7\" >-0.0482491</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col8\" class=\"data row5 col8\" >0.0593471</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col9\" class=\"data row5 col9\" >0.0978039</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col10\" class=\"data row5 col10\" >0.162005</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col11\" class=\"data row5 col11\" >-0.145434</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col12\" class=\"data row5 col12\" >-0.127802</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col13\" class=\"data row5 col13\" >-0.141982</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col14\" class=\"data row5 col14\" >-0.102859</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col15\" class=\"data row5 col15\" >0.0748319</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col16\" class=\"data row5 col16\" >-0.14101</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col17\" class=\"data row5 col17\" >0.142665</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col18\" class=\"data row5 col18\" >0.124917</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow5_col19\" class=\"data row5 col19\" >0.0335083</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row6\" class=\"row_heading level0 row6\" >T</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col0\" class=\"data row6 col0\" >-0.0134584</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col1\" class=\"data row6 col1\" >-0.0137532</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col2\" class=\"data row6 col2\" >0.112594</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col3\" class=\"data row6 col3\" >-0.14567</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col4\" class=\"data row6 col4\" >-0.334775</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col5\" class=\"data row6 col5\" >0.0775678</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col6\" class=\"data row6 col6\" >0.220247</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col7\" class=\"data row6 col7\" >0.0352718</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col8\" class=\"data row6 col8\" >-0.14982</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col9\" class=\"data row6 col9\" >-0.348406</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col10\" class=\"data row6 col10\" >0.184823</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col11\" class=\"data row6 col11\" >0.355625</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col12\" class=\"data row6 col12\" >-0.104848</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col13\" class=\"data row6 col13\" >-0.0537415</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col14\" class=\"data row6 col14\" >0.0568035</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col15\" class=\"data row6 col15\" >-0.00567016</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col16\" class=\"data row6 col16\" >-0.0728419</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col17\" class=\"data row6 col17\" >-0.0166233</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col18\" class=\"data row6 col18\" >-0.0681195</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow6_col19\" class=\"data row6 col19\" >-0.0369762</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row7\" class=\"row_heading level0 row7\" >N</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col0\" class=\"data row7 col0\" >-0.167804</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col1\" class=\"data row7 col1\" >0.0697035</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col2\" class=\"data row7 col2\" >0.0948484</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col3\" class=\"data row7 col3\" >-0.0118898</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col4\" class=\"data row7 col4\" >-0.0266122</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col5\" class=\"data row7 col5\" >0.0110793</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col6\" class=\"data row7 col6\" >-0.274172</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col7\" class=\"data row7 col7\" >0.170864</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col8\" class=\"data row7 col8\" >0.313153</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col9\" class=\"data row7 col9\" >-0.25291</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col10\" class=\"data row7 col10\" >-0.146736</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col11\" class=\"data row7 col11\" >0.320863</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col12\" class=\"data row7 col12\" >-0.259327</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col13\" class=\"data row7 col13\" >-0.195923</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col14\" class=\"data row7 col14\" >-0.0865563</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col15\" class=\"data row7 col15\" >0.0735839</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col16\" class=\"data row7 col16\" >-0.397154</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col17\" class=\"data row7 col17\" >0.0623471</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col18\" class=\"data row7 col18\" >0.179056</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow7_col19\" class=\"data row7 col19\" >0.171161</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row8\" class=\"row_heading level0 row8\" >Q</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col0\" class=\"data row8 col0\" >0.16277</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col1\" class=\"data row8 col1\" >0.188798</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col2\" class=\"data row8 col2\" >-0.171319</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col3\" class=\"data row8 col3\" >-0.170855</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col4\" class=\"data row8 col4\" >0.124862</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col5\" class=\"data row8 col5\" >-0.283221</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col6\" class=\"data row8 col6\" >0.163197</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col7\" class=\"data row8 col7\" >0.104516</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col8\" class=\"data row8 col8\" >0.256037</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col9\" class=\"data row8 col9\" >-0.855405</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col10\" class=\"data row8 col10\" >-0.176136</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col11\" class=\"data row8 col11\" >-0.16285</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col12\" class=\"data row8 col12\" >0.0909827</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col13\" class=\"data row8 col13\" >-0.0891506</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col14\" class=\"data row8 col14\" >0.0910876</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col15\" class=\"data row8 col15\" >-0.0419301</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col16\" class=\"data row8 col16\" >0.163457</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col17\" class=\"data row8 col17\" >-0.303332</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col18\" class=\"data row8 col18\" >-0.0242489</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow8_col19\" class=\"data row8 col19\" >-0.163594</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row9\" class=\"row_heading level0 row9\" >C</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col0\" class=\"data row9 col0\" >-0.159081</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col1\" class=\"data row9 col1\" >-0.21425</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col2\" class=\"data row9 col2\" >0.485746</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col3\" class=\"data row9 col3\" >0.109982</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col4\" class=\"data row9 col4\" >0.0196909</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col5\" class=\"data row9 col5\" >0.187528</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col6\" class=\"data row9 col6\" >-0.134617</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col7\" class=\"data row9 col7\" >-0.411852</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col8\" class=\"data row9 col8\" >0.121776</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col9\" class=\"data row9 col9\" >1.02177</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col10\" class=\"data row9 col10\" >0.415674</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col11\" class=\"data row9 col11\" >0.141038</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col12\" class=\"data row9 col12\" >-0.00682356</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col13\" class=\"data row9 col13\" >-0.0345076</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col14\" class=\"data row9 col14\" >-0.372207</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col15\" class=\"data row9 col15\" >0.0729878</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col16\" class=\"data row9 col16\" >-0.29315</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col17\" class=\"data row9 col17\" >0.233722</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col18\" class=\"data row9 col18\" >0.0840048</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow9_col19\" class=\"data row9 col19\" >-0.0533458</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row10\" class=\"row_heading level0 row10\" >G</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col0\" class=\"data row10 col0\" >0.108892</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col1\" class=\"data row10 col1\" >-0.141231</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col2\" class=\"data row10 col2\" >0.222863</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col3\" class=\"data row10 col3\" >-0.114242</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col4\" class=\"data row10 col4\" >-0.245551</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col5\" class=\"data row10 col5\" >-0.0096348</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col6\" class=\"data row10 col6\" >0.10779</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col7\" class=\"data row10 col7\" >-0.427385</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col8\" class=\"data row10 col8\" >-0.0541703</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col9\" class=\"data row10 col9\" >0.0886465</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col10\" class=\"data row10 col10\" >0.29234</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col11\" class=\"data row10 col11\" >-0.416116</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col12\" class=\"data row10 col12\" >0.115218</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col13\" class=\"data row10 col13\" >0.201467</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col14\" class=\"data row10 col14\" >0.0711956</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col15\" class=\"data row10 col15\" >-0.0491774</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col16\" class=\"data row10 col16\" >0.32688</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col17\" class=\"data row10 col17\" >0.124381</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col18\" class=\"data row10 col18\" >0.235346</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow10_col19\" class=\"data row10 col19\" >0.169884</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row11\" class=\"row_heading level0 row11\" >P</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col0\" class=\"data row11 col0\" >-0.172696</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col1\" class=\"data row11 col1\" >-0.0973528</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col2\" class=\"data row11 col2\" >0.16307</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col3\" class=\"data row11 col3\" >-0.133037</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col4\" class=\"data row11 col4\" >0.182551</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col5\" class=\"data row11 col5\" >0.0301245</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col6\" class=\"data row11 col6\" >0.262877</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col7\" class=\"data row11 col7\" >0.000195224</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col8\" class=\"data row11 col8\" >-0.00347877</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col9\" class=\"data row11 col9\" >-0.274187</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col10\" class=\"data row11 col10\" >0.0155804</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col11\" class=\"data row11 col11\" >-0.36558</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col12\" class=\"data row11 col12\" >-0.133413</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col13\" class=\"data row11 col13\" >0.128255</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col14\" class=\"data row11 col14\" >0.00303923</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col15\" class=\"data row11 col15\" >-0.075045</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col16\" class=\"data row11 col16\" >0.0354873</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col17\" class=\"data row11 col17\" >0.032095</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col18\" class=\"data row11 col18\" >0.0360384</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow11_col19\" class=\"data row11 col19\" >0.019659</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row12\" class=\"row_heading level0 row12\" >A</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col0\" class=\"data row12 col0\" >0.198943</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col1\" class=\"data row12 col1\" >0.0685175</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col2\" class=\"data row12 col2\" >0.0358662</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col3\" class=\"data row12 col3\" >-0.260527</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col4\" class=\"data row12 col4\" >-0.367749</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col5\" class=\"data row12 col5\" >-0.023834</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col6\" class=\"data row12 col6\" >0.121914</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col7\" class=\"data row12 col7\" >-0.115855</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col8\" class=\"data row12 col8\" >-0.0174852</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col9\" class=\"data row12 col9\" >0.374218</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col10\" class=\"data row12 col10\" >0.0892439</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col11\" class=\"data row12 col11\" >-0.135638</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col12\" class=\"data row12 col12\" >0.125151</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col13\" class=\"data row12 col13\" >-0.0480442</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col14\" class=\"data row12 col14\" >0.117665</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col15\" class=\"data row12 col15\" >0.0768116</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col16\" class=\"data row12 col16\" >0.261374</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col17\" class=\"data row12 col17\" >-0.053744</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col18\" class=\"data row12 col18\" >-0.059366</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow12_col19\" class=\"data row12 col19\" >-0.0761042</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row13\" class=\"row_heading level0 row13\" >V</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col0\" class=\"data row13 col0\" >-0.0479894</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col1\" class=\"data row13 col1\" >-0.148138</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col2\" class=\"data row13 col2\" >-0.136075</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col3\" class=\"data row13 col3\" >0.119442</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col4\" class=\"data row13 col4\" >-0.0177748</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col5\" class=\"data row13 col5\" >0.0873351</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col6\" class=\"data row13 col6\" >-0.0405469</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col7\" class=\"data row13 col7\" >-0.17542</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col8\" class=\"data row13 col8\" >-0.376899</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col9\" class=\"data row13 col9\" >0.27009</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col10\" class=\"data row13 col10\" >0.154263</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col11\" class=\"data row13 col11\" >0.0153997</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col12\" class=\"data row13 col12\" >0.181063</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col13\" class=\"data row13 col13\" >0.232274</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col14\" class=\"data row13 col14\" >-0.0123478</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col15\" class=\"data row13 col15\" >-0.0114113</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col16\" class=\"data row13 col16\" >0.0362393</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col17\" class=\"data row13 col17\" >0.0319725</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col18\" class=\"data row13 col18\" >-0.123012</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow13_col19\" class=\"data row13 col19\" >-0.081372</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row14\" class=\"row_heading level0 row14\" >I</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col0\" class=\"data row14 col0\" >-0.0601495</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col1\" class=\"data row14 col1\" >0.013694</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col2\" class=\"data row14 col2\" >-0.168476</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col3\" class=\"data row14 col3\" >0.132522</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col4\" class=\"data row14 col4\" >-0.0287538</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col5\" class=\"data row14 col5\" >0.160029</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col6\" class=\"data row14 col6\" >0.0102711</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col7\" class=\"data row14 col7\" >0.0823409</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col8\" class=\"data row14 col8\" >-0.289607</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col9\" class=\"data row14 col9\" >0.232549</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col10\" class=\"data row14 col10\" >-0.0172295</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col11\" class=\"data row14 col11\" >0.148735</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col12\" class=\"data row14 col12\" >0.105743</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col13\" class=\"data row14 col13\" >-0.0284908</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col14\" class=\"data row14 col14\" >-0.0281225</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col15\" class=\"data row14 col15\" >-0.110696</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col16\" class=\"data row14 col16\" >-0.137242</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col17\" class=\"data row14 col17\" >-0.00213543</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col18\" class=\"data row14 col18\" >-0.300632</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow14_col19\" class=\"data row14 col19\" >-0.0815973</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row15\" class=\"row_heading level0 row15\" >L</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col0\" class=\"data row15 col0\" >-0.0262128</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col1\" class=\"data row15 col1\" >0.0675021</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col2\" class=\"data row15 col2\" >-0.204046</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col3\" class=\"data row15 col3\" >0.0729803</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col4\" class=\"data row15 col4\" >0.0810259</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col5\" class=\"data row15 col5\" >0.0996609</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col6\" class=\"data row15 col6\" >0.0747734</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col7\" class=\"data row15 col7\" >0.0284554</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col8\" class=\"data row15 col8\" >-0.251793</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col9\" class=\"data row15 col9\" >-0.157187</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col10\" class=\"data row15 col10\" >-0.00546897</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col11\" class=\"data row15 col11\" >0.050326</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col12\" class=\"data row15 col12\" >0.109376</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col13\" class=\"data row15 col13\" >0.070076</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col14\" class=\"data row15 col14\" >-0.00337579</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col15\" class=\"data row15 col15\" >-0.114409</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col16\" class=\"data row15 col16\" >0.0681596</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col17\" class=\"data row15 col17\" >-0.054838</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col18\" class=\"data row15 col18\" >-0.290831</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow15_col19\" class=\"data row15 col19\" >-0.338498</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row16\" class=\"row_heading level0 row16\" >M</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col0\" class=\"data row16 col0\" >0.0481314</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col1\" class=\"data row16 col1\" >-0.400935</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col2\" class=\"data row16 col2\" >0.232985</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col3\" class=\"data row16 col3\" >-0.124057</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col4\" class=\"data row16 col4\" >-0.430303</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col5\" class=\"data row16 col5\" >-0.0314816</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col6\" class=\"data row16 col6\" >0.0182609</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col7\" class=\"data row16 col7\" >-0.208543</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col8\" class=\"data row16 col8\" >0.0888807</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col9\" class=\"data row16 col9\" >0.0596031</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col10\" class=\"data row16 col10\" >0.263409</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col11\" class=\"data row16 col11\" >0.444458</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col12\" class=\"data row16 col12\" >0.222589</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col13\" class=\"data row16 col13\" >0.188901</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col14\" class=\"data row16 col14\" >0.0859318</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col15\" class=\"data row16 col15\" >0.0375811</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col16\" class=\"data row16 col16\" >0.573782</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col17\" class=\"data row16 col17\" >0.0701895</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col18\" class=\"data row16 col18\" >-0.16354</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow16_col19\" class=\"data row16 col19\" >-0.227913</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row17\" class=\"row_heading level0 row17\" >F</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col0\" class=\"data row17 col0\" >-0.15159</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col1\" class=\"data row17 col1\" >-0.0930246</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col2\" class=\"data row17 col2\" >-0.280843</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col3\" class=\"data row17 col3\" >0.236301</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col4\" class=\"data row17 col4\" >-0.0241875</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col5\" class=\"data row17 col5\" >0.206632</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col6\" class=\"data row17 col6\" >-0.139226</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col7\" class=\"data row17 col7\" >0.0198979</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col8\" class=\"data row17 col8\" >-0.296749</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col9\" class=\"data row17 col9\" >0.2621</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col10\" class=\"data row17 col10\" >0.0692608</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col11\" class=\"data row17 col11\" >-0.267969</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col12\" class=\"data row17 col12\" >0.00387246</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col13\" class=\"data row17 col13\" >0.0413661</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col14\" class=\"data row17 col14\" >-0.0402823</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col15\" class=\"data row17 col15\" >0.0467455</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col16\" class=\"data row17 col16\" >-0.0681286</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col17\" class=\"data row17 col17\" >0.0144949</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col18\" class=\"data row17 col18\" >0.124543</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow17_col19\" class=\"data row17 col19\" >0.0240999</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row18\" class=\"row_heading level0 row18\" >W</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col0\" class=\"data row18 col0\" >-0.0564035</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col1\" class=\"data row18 col1\" >-0.00164825</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col2\" class=\"data row18 col2\" >-0.181255</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col3\" class=\"data row18 col3\" >0.045528</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col4\" class=\"data row18 col4\" >-0.118602</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col5\" class=\"data row18 col5\" >-0.142788</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col6\" class=\"data row18 col6\" >0.102414</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col7\" class=\"data row18 col7\" >0.127906</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col8\" class=\"data row18 col8\" >-0.113657</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col9\" class=\"data row18 col9\" >1.6662</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col10\" class=\"data row18 col10\" >-0.0838383</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col11\" class=\"data row18 col11\" >-0.347091</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col12\" class=\"data row18 col12\" >-0.14705</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col13\" class=\"data row18 col13\" >-0.0757202</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col14\" class=\"data row18 col14\" >0.0267994</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col15\" class=\"data row18 col15\" >-0.0997645</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col16\" class=\"data row18 col16\" >0.23799</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col17\" class=\"data row18 col17\" >0.143871</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col18\" class=\"data row18 col18\" >0.548381</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow18_col19\" class=\"data row18 col19\" >-0.00537147</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_80d40414_379c_11ea_a970_e79fc7d1477blevel0_row19\" class=\"row_heading level0 row19\" >Y</th> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col0\" class=\"data row19 col0\" >0.171084</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col1\" class=\"data row19 col1\" >-0.0836216</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col2\" class=\"data row19 col2\" >0.244288</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col3\" class=\"data row19 col3\" >0.153096</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col4\" class=\"data row19 col4\" >-0.0342249</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col5\" class=\"data row19 col5\" >-0.0041557</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col6\" class=\"data row19 col6\" >-0.246017</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col7\" class=\"data row19 col7\" >-0.14396</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col8\" class=\"data row19 col8\" >0.396572</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col9\" class=\"data row19 col9\" >0.430213</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col10\" class=\"data row19 col10\" >-0.0120142</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col11\" class=\"data row19 col11\" >-0.0525738</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col12\" class=\"data row19 col12\" >-0.0759934</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col13\" class=\"data row19 col13\" >-0.12039</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col14\" class=\"data row19 col14\" >-0.289749</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col15\" class=\"data row19 col15\" >0.164446</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col16\" class=\"data row19 col16\" >-0.326428</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col17\" class=\"data row19 col17\" >0.185125</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col18\" class=\"data row19 col18\" >0.190562</td> \n",
       "        <td id=\"T_80d40414_379c_11ea_a970_e79fc7d1477brow19_col19\" class=\"data row19 col19\" >0.138297</td> \n",
       "    </tr></tbody> \n",
       "</table> "
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fd342613048>"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDfTableAApair2.reorder_levels([0,2,3,1]).loc['bulk'].loc['theoretical'].loc['log2OddsRatio'].astype(float)\n",
    "df.style.background_gradient(cmap=cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>aa1</th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>1.272224</td>\n",
       "      <td>0.944005</td>\n",
       "      <td>1.078603</td>\n",
       "      <td>0.999921</td>\n",
       "      <td>1.052757</td>\n",
       "      <td>0.867839</td>\n",
       "      <td>0.951778</td>\n",
       "      <td>0.929701</td>\n",
       "      <td>1.042949</td>\n",
       "      <td>0.945351</td>\n",
       "      <td>1.011407</td>\n",
       "      <td>1.006980</td>\n",
       "      <td>1.016260</td>\n",
       "      <td>1.047577</td>\n",
       "      <td>1.005178</td>\n",
       "      <td>0.970789</td>\n",
       "      <td>1.254416</td>\n",
       "      <td>0.974573</td>\n",
       "      <td>1.062193</td>\n",
       "      <td>1.069277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>0.998836</td>\n",
       "      <td>1.010097</td>\n",
       "      <td>0.988085</td>\n",
       "      <td>1.096172</td>\n",
       "      <td>1.228497</td>\n",
       "      <td>0.838444</td>\n",
       "      <td>1.139604</td>\n",
       "      <td>1.170763</td>\n",
       "      <td>1.061424</td>\n",
       "      <td>0.696131</td>\n",
       "      <td>0.786942</td>\n",
       "      <td>0.962957</td>\n",
       "      <td>0.941383</td>\n",
       "      <td>0.969958</td>\n",
       "      <td>1.042901</td>\n",
       "      <td>0.914770</td>\n",
       "      <td>0.998766</td>\n",
       "      <td>0.771934</td>\n",
       "      <td>1.016217</td>\n",
       "      <td>1.073106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>1.099379</td>\n",
       "      <td>0.875032</td>\n",
       "      <td>1.659204</td>\n",
       "      <td>0.918916</td>\n",
       "      <td>0.888413</td>\n",
       "      <td>0.950785</td>\n",
       "      <td>0.877112</td>\n",
       "      <td>0.925127</td>\n",
       "      <td>1.241524</td>\n",
       "      <td>1.160676</td>\n",
       "      <td>1.092957</td>\n",
       "      <td>1.420419</td>\n",
       "      <td>0.972918</td>\n",
       "      <td>0.869954</td>\n",
       "      <td>0.917990</td>\n",
       "      <td>1.084330</td>\n",
       "      <td>0.938727</td>\n",
       "      <td>1.094042</td>\n",
       "      <td>1.080486</td>\n",
       "      <td>1.071456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.899644</td>\n",
       "      <td>0.985089</td>\n",
       "      <td>1.014349</td>\n",
       "      <td>0.931116</td>\n",
       "      <td>1.099696</td>\n",
       "      <td>0.960224</td>\n",
       "      <td>0.795853</td>\n",
       "      <td>0.872092</td>\n",
       "      <td>1.088576</td>\n",
       "      <td>0.860195</td>\n",
       "      <td>0.966687</td>\n",
       "      <td>1.093268</td>\n",
       "      <td>0.944706</td>\n",
       "      <td>0.989824</td>\n",
       "      <td>0.936811</td>\n",
       "      <td>1.121766</td>\n",
       "      <td>0.833372</td>\n",
       "      <td>1.225805</td>\n",
       "      <td>1.107399</td>\n",
       "      <td>1.188066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1.014333</td>\n",
       "      <td>1.153806</td>\n",
       "      <td>0.949531</td>\n",
       "      <td>0.851492</td>\n",
       "      <td>1.085775</td>\n",
       "      <td>0.791073</td>\n",
       "      <td>0.933537</td>\n",
       "      <td>1.082559</td>\n",
       "      <td>1.097968</td>\n",
       "      <td>0.762836</td>\n",
       "      <td>0.829823</td>\n",
       "      <td>0.782597</td>\n",
       "      <td>0.987752</td>\n",
       "      <td>1.033710</td>\n",
       "      <td>1.143273</td>\n",
       "      <td>1.042356</td>\n",
       "      <td>1.078785</td>\n",
       "      <td>0.933006</td>\n",
       "      <td>0.923164</td>\n",
       "      <td>0.943463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "aa1         R         K         H         D         E         S         T  \\\n",
       "aa0                                                                         \n",
       "R    1.272224  0.944005  1.078603  0.999921  1.052757  0.867839  0.951778   \n",
       "K    0.998836  1.010097  0.988085  1.096172  1.228497  0.838444  1.139604   \n",
       "H    1.099379  0.875032  1.659204  0.918916  0.888413  0.950785  0.877112   \n",
       "D    0.899644  0.985089  1.014349  0.931116  1.099696  0.960224  0.795853   \n",
       "E    1.014333  1.153806  0.949531  0.851492  1.085775  0.791073  0.933537   \n",
       "\n",
       "aa1         N         Q         C         G         P         A         V  \\\n",
       "aa0                                                                         \n",
       "R    0.929701  1.042949  0.945351  1.011407  1.006980  1.016260  1.047577   \n",
       "K    1.170763  1.061424  0.696131  0.786942  0.962957  0.941383  0.969958   \n",
       "H    0.925127  1.241524  1.160676  1.092957  1.420419  0.972918  0.869954   \n",
       "D    0.872092  1.088576  0.860195  0.966687  1.093268  0.944706  0.989824   \n",
       "E    1.082559  1.097968  0.762836  0.829823  0.782597  0.987752  1.033710   \n",
       "\n",
       "aa1         I         L         M         F         W         Y  \n",
       "aa0                                                              \n",
       "R    1.005178  0.970789  1.254416  0.974573  1.062193  1.069277  \n",
       "K    1.042901  0.914770  0.998766  0.771934  1.016217  1.073106  \n",
       "H    0.917990  1.084330  0.938727  1.094042  1.080486  1.071456  \n",
       "D    0.936811  1.121766  0.833372  1.225805  1.107399  1.188066  \n",
       "E    1.143273  1.042356  1.078785  0.933006  0.923164  0.943463  "
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDfTableAApair2.reorder_levels([0,2,3,1]).loc['bulk'].loc['theoretical'].loc['oddsRatio'].astype(float)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C 2.030412741220943\n",
      "H 1.6592042343850146\n",
      "M 1.4884201077275219\n",
      "W 1.4624440430925905\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[print(aa, df.loc[aa, aa]) for aa in ['C', 'H', 'M', 'W']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'taxaId_31969'"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speciesSetName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAFdCAYAAAAAF9ZZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzs3XdYFFf3B/DvUpauiIhdIyoWEAtiRSzBRkQEFQQFUQQ1toi995KIMSrWN/aKBVGjRF9QVIwitqgEC4gKKsUAKlJ22Z3fH/zYV0LZndmhbPZ8nmefh53dM/fOsJy93Jk5I2AYhgEhhJB/PY2q7gAhhJDKQQmfEELUBCV8QghRE5TwCSFETVDCJ4QQNUEJnxBC1AQlfEIIqQTTpk1DcnKy7HleXh58fX3h5uaGU6dOVUofKOETQkgFEolEmDJlCv78889iy2/cuIHOnTvj+PHjlPAJIeTfQCQSwcfHBz169Ci23NzcHGKxGAUFBdDW1q6UvmhVSiuEEFJNzdY2Vyre9vA6BAcHy567u7vD3d1d9tzQ0BC2trY4ffp0sThtbW1cuHABoaGhcHNzU6oPiqKETwghSvhnglfU4cOHERAQgP79+2Pq1Kl4+/YtGjZsWAE9/B9K+IQQtaYpqJp29fT0YGBgAA0NDRgaGiInJ6fC26Q5fEKIWtMQCJR6sLV7927Ex8fDy8sLe/bsgaenJ4yNjdGyZcsK2LriKOErafXq1XB2doazszOsrKwwcOBA2fO8vLwKb9/X1xcvX77kFPvmzRvMmDGD5x6VdPLkSXz//fesYjw8PBAeHl5BPfofZffBkSNH8Ouvv5b6Wrt27ZCSklJi+V9//YWOHTvKPifOzs548+ZNsfc8ePAAVlZW+PTpU4n4PXv2wNnZucRyhmEwadIkrFmzpsRrWVlZ6NevX7F9eufOHbi6usLZ2Rnu7u7466+/5G7vv5Gmkg9FrV+/Ho0aNYK/vz9atGgBU1NT7Nu3D0ePHsXChQv526By0JSOkhYvXiz7uV+/fggMDES7du0qrf09e/Zwjk1OTsarV6/464wKUnYfjB49mnXMgwcP4OzsjOXLl5f6+t9//40VK1ZALBaXeC0mJgb79u1D7dq1S7y2a9cuPHz4EI0bNy62XCqVYu7cucjOzi62fPbs2di4cSNsbW0RFhaG+fPn49y5c6y3h6gOGuFXsBcvXsDHx0c2kjpz5gwA4I8//oCnpydmz56NYcOGwcnJCQ8ePABQeBrX6tWr4eLigqFDh2LBggWyP1Z7e3vMnDkTgwcPxpUrV2Bvb4+4uDgAhSNpR0dHODk5wcfHB6mpqfjjjz+KjQaLnotEIixbtgyJiYnw8/MDANy9exceHh5wcXHB8OHDce3aNQBAamoqfHx84OLiAhcXF2zdurXEdhYUFKBdu3bYuHEjXFxcMHjw4GKjydTUVPj5+cHJyQmurq54+fIlkpKSYGNjI9s2hmHg4OCAFy9eyOJev36Nzp07l/r85MmTmDx5Mvz8/DBkyBD4+/sjLCwMXl5e6NWrF/bv31/u76a0fbBt2zaMGDECTk5OcHBwQEREBABgzpw5mDVrFgDg6dOn6N69O16+fIlNmzbJRtTR0dEYOnQohg0bhqVLl+LrW034+vrK9uf9+/fx4sULuLi4YMSIEcX2k0QiwZw5cxAQEFCiv2lpaVi9ejVmz55d4rVbt24hOjoaI0eOLPFaUFAQrKys0Lx582LLpVIpPn78CAD48uULdHR0yt1f/1aaAoFSD1VCCb8CicVizJgxA/Pnz0dISAgOHTqEXbt24fHjxwCAP//8E35+fggNDYWTkxN++eUXAMCOHTugq6uLkJAQnDt3DrVq1ZK9BgCtW7dGWFgY+vXrJ1v25MkTbNq0CXv37sX58+fRq1cv7Ny5s8y+CYVCrFixAs2aNcN//vMfZGZmYtGiRdi4cSPOnDmDbdu2YcmSJUhJSUFwcDDMzc1x5swZHD58GPHx8SVGi0BhAq1RowbOnDmDjRs3YsGCBcjMzARQOJJeunQpzp8/j/bt22P//v1o3LgxOnfujAsXLgAAbt68CTMzM1Zzmffu3cPq1atx7tw5vHnzBv/9739x8OBB7NixA7/88gvKu7/PP/dBUlISYmJicOTIEZw/fx7Tpk2TfbktW7YMjx8/RmhoKGbOnIklS5bA3Px/p/OJRCL88MMPWLRoEUJDQ9GpU6diI/Q9e/agd+/eAAADAwM4OTnh1KlTWLduHRYvXiz70t60aRM6depU4pztgoICzJo1CwsWLICZmVmx11JSUrBu3Tps2LABGhrF/6SvXbuGhw8fYurUqSW2f+3atZg1axbs7e2xZs0aLFmyRJFd/q+jKVDuoUpoSqcCJSQkICkpCfPmzZMtE4lEiIuLQ6NGjdCoUSO0atUKAGBpaYmLFy8CACIjI5GTk4MbN24AKPzi+PqP3MbGpkRbt27dgr29PerVqwegcEQJFI7oFXH//n2kp6dj8uTJsmUaGhp4/vw57O3tMXHiRCQnJ6NHjx6YO3cuDA0NS11P0RRH27ZtYW5ujnv37gEA2rdvL5tqaNOmjWy06+npiS1btsDd3R3BwcHw8PBQqL9FrK2tUbduXQBAw4YN0bNnTwgEAjRp0gS5ubkQiUQKj1wbN26MtWvX4ty5c3j9+jUePHggO3PC0NAQGzduxKhRozB8+HA4OjoWi42Li4Oenh66du0KABg2bFiZUzYrV66U/dyyZUsMHDgQV65cwdu3bxEXF4f//Oc/kEqlxWI2bNiAHj16oFu3bsV+pyKRCDNnzsTixYthampaLCY5ORmBgYHYv39/iS+CtLQ0LF++HEePHoWlpSUuXbqEqVOn4vLly9DV1VVof/1bcDnwqqoo4VcgqVQKY2NjnD17VrYsPT0dNWrUwL1794r9YQkEAtloVCKRYOnSpejZsycAIDs7u9ho0cDAoERbWlpaEHz1wc3NzUVKSkqx9QIodV64qK8WFhY4fvy4bFlqaipMTEygra2NiIgI3Lp1C7dv38aIESOwd+9etGnTptg6BAIBtLS0iq1TU1NT1r/SttXe3h6rV6/G7du38eDBA2zcuLHEOsvrv1AoLLEfuHr8+DGmTp2KcePGwc7ODjY2Nli7dq3s9cTERNSqVQuxsbEQi8Vyr44s2vavFRQUYPfu3fDx8YG+vj6AwqksbW1thISE4N27d3BxcZG9f/To0Vi3bh0uXLgAExMT/P7778jJyUFqaipcXV2xcOFCvHv3Tjat9OHDB0gkEuTn56Nx48bIzc3F+PHjARQeoF67di0yMzOhr6+Pxo0bw9LSEgAwcOBArFmzBomJiSV+r+Tfg6Z0KlCLFi2goaEhm7J4+/YthgwZgqdPn5YbZ2dnh0OHDkEsFkMikWDhwoXFpnRK061bN9y4cQPp6ekAgKNHj2Ljxo2oVasW3r59i4yMDDAMI+sLUJiQCgoKAAAdO3ZEQkKCbEQeGxuLgQMH4u+//8aPP/6I//znP+jfvz8WL16MZs2aFZtnL8IwjOzL7dGjR3jz5k2x+ffSCAQCeHh4YOHChXB2di6RwGvUqIG8vDzZmUi//fZbuetj6+t9cOfOHbRv3x4+Pj7o3LkzwsPDIZFIABQmy/Xr1+PAgQNo1KgRfv7552LradOmDUQiEaKiogAAly9fLnXaS0tLC+Hh4Th58iSAwlF4REQE+vfvj+3btyMsLAxnz56VXZV55MgRWFlZISoqCufOncPZs2dl01AhISHo3Lkzrl27hrNnz+Ls2bOy4w8rV66En58fwsPDZa+1bt0aCxcuxMiRI9G6dWs8ffoUr1+/BlD4H55YLEbTpk153b+qgKZ0CC+EQiF27NiBtWvXYufOnbJ52Pbt25c71TJt2jT8+OOPGDZsGCQSCSwtLTF37txy22rTpg1mzZolm8qpW7cu1q5dizp16mD48OEYPnw4TE1N0bt3bzx79gxA4XSChoaGbDply5YtWLduHUQiERiGQWBgIOrVqwcfHx/Mnz8fQ4YMgba2Ntq2bYtBgwaV2o+iOXCGYbB582YYGRnJ3U8uLi4IDAzEqFGjSrxmbGyMgIAA+Pr6wsTEBAMHDpS7vn+SSCRwdXXF+vXrS4xev94HW7duRXh4OBwdHSGVStGnTx9kZmbi8+fPCAgIgL+/P5o3b47ly5fDycmp2Dy7UCjEtm3bsHz5cmzYsAGWlpYwNjaWve7r6wtvb2/07t0bP//8M5YuXYpTp05BKpViyZIlaNasGevtUkbz5s2xdOlS2dy+vr4+tm3bJvuvQ52oWtJWhoAp76gWIQoqKCiApaUlYmJiUKNGDYXjGIbBuXPnEBYWVu5BZmVt2LABbm5uajmCJeXbaGShVPysz8956knFoxE+qVKenp7Izs7G9u3bK6wNhmHwzTffULInpdKgET4hhKiHTTWUG+HP/EQjfEIIUQmqdvGUMugsnSqwbNky9OvXD5s2beJlfRKJBBMnTsSHDx84xT9+/BjTp08v9bXw8HBs27at2LJPnz6VeoC1KixevBhPnjxhFZORkSG7/qE8cXFxcHBwgKura7Fb0ynr2rVrSv3uFy1aVOpB/6LPwd9//61M99ROZdXSqQ4o4VeB4OBgHD16FDNnzuRlfXv37kWXLl1KXHijqHbt2mHLli2lvubg4IC7d+/KrgQFgKtXr8Le3p5TW3z7448/yr2aVhkRERHo2rUrQkJC0KhRI17WmZ2djcDAQEyaNInzOtasWVPiSlyg8BTTCRMmYMWKFcp0Ue2oU2kFmtKpZJ6enmAYBn5+foiPj8fAgQPx7NkzBAQEwNraGitXrsT79+8hFovx3XffyRLDkSNHcODAARgZGcHOzg7nz5/HlStXkJubiwMHDuD8+fMAgK1bt+L169dISUlBeno6WrdujTVr1sDQ0BBXr17Frl27IBKJkJGRgWHDhuGHH35AdHQ0Vq1ahd9++w3z589HVlYWkpKS0KdPH8yZMwcjRoxAUFCQbKQfEREhO53vypUr2LFjB8RiMXR1dTFv3jx07NgRCxYsQE5ODjZv3owXL17A29sbhw8fxsWLF8vsX2pqapnbf/XqVfzyyy+QSqXQ19fHihUrEBYWhrS0NMyePRs//fQTzM3NsWbNGjx//hxisRjdu3fH3LlzoaWlhcuXL2PTpk3Q09ODlZWV3N/TuXPncOzYMUgkEuTl5aFnz544deoUcnNzYWhoiEOHDuHkyZM4duyY7AK7JUuWoHnz5khNTcX8+fORlpaGBg0aQFNTEwMGDICrqyuOHj0KOzs76OnpASi8ItnPzw83btxATk4OAgICMGDAAOTk5GD58uV4/fo1srKyYGBggMDAQJibm8PLywujR4+GlZUVRo8ejebNm+Pt27c4dOgQbG1tsWzZMsTFxdEFVApSp9MywZBKZ2Fhwfz9999M3759maCgINlyLy8vJiIigmEYhsnLy2O8vLyYCxcuMI8fP2a6d+/OpKWlMVKplFm8eDHTt29fhmEY5sqVK8yYMWNk69iyZQtjb2/PpKenMxKJhAkICGDWr1/PSKVSZsyYMUxiYiLDMAyTkpLCtGnThvn777+Z27dvM9999x3DMAwzb948ZuzYscX6m5mZyVhZWTG5ublMfn4+M3ToUIZhGCYxMZEZMmQIk5GRwTAMwzx//pzp2bMn8+XLF+bLly/MgAEDmJCQEOa7775jzp49W27/ytv+9PR0xsbGhomNjWUYhmEuXbrE+Pr6MgzDMH379mUePXrEMAzDzJ8/nzl48CDDMAxTUFDAzJ49m9m9e7cs/sWLFwzDMMzOnTsZCwsLub+nLVu2MCtWrGAYhmFOnz7N2NraMp8/f2YYhmGio6MZT09PJicnh2EYhrlx4wYzaNAghmEYZuLEicymTZsYhmGYly9fMu3bt2dOnz7NMAzDuLi4MLdv3y72WdixYwfDMAwTFxfH2NjYMH///TcTFhbGrFq1Sva+JUuWMCtXrmQYhmHGjBnDhIWFMUlJSYyFhQUTExNTrN+rVq1iNm/eLHf7SKHdxq2UeqgSGuFXsaIrUXNychATE4OPHz9i8+bNsmVPnz7F+/fv0bNnT9SpUwdA4X8JN2/eBAC8fPkSTZo0KbbOQYMGyaZ3RowYgbVr12LevHnYuXMnIiMj8dtvvyEhIQEMwyA3N7dEn/5Zq8fY2Bg6Ojp4+/YtkpOT0b17dwCFxc7S0tLg4+Mje69AIMCbN2/QunVrbNq0CW5ubhg6dCiGDh1abv+mTZtW5vZraWmhZcuWaNu2LQBgwIABGDBgQIl+R0ZG4vHjxzh16hQAyO5HcO/ePVhYWKBFixYACm9J988rZRXRqlUrWQ2hyMhIvH79utixjE+fPiErKwvR0dGy+ubNmjUrNv2SmJhY4vTQMWPGACgsimdhYYGYmBgMGjQIjRs3xqFDh/D69WvcuXMHHTt2LNEnLS0tdOjQodiyRo0a4c8//2S9feqKaumQSlN0ZaNUKgXDMDh+/Ljs3/2MjAzo6OggNDS02Dz11zVcBAJBiUJbX9dwkUql0NDQQE5ODlxcXODg4IDOnTtj+PDhCA8PL3X+u7SrLTU1NaGpqYnw8HBZ8pZKpejevXuxsg/v37+XFXpLTEyEsbEx4uLiIBKJZGUTSutfedt/+/btYnWCGIbBs2fP0Lp162J9lEql2Lx5s6wM8KdPnyAQCErM83Ott/P1fpFKpXB2dsacOXNkz9PS0lCzZk3o6Ogo9fvS1NTE0aNHceLECYwePRpOTk4wNjYu9cCxUCgssT1aWloliqWRsqnTlA59KqoJQ0NDdOjQAfv27QNQmKw8PDwQERGBHj164NatW3j37h0AICQkRBbXrFkzJCUlFVtXREQEPn/+DKlUihMnTqBv3754/fo1srOz8cMPP6Bfv36Ijo6GSCQqkXxK8/nzZ4hEIjRo0AAPHz5Ep06dAADdu3fHzZs3kZCQAKDw7JOhQ4ciLy8PycnJWLNmDfbu3Qtzc3MEBgaW27/ytr99+/ZISEiQ1e+JiIiQJdqva+HY2dlh//79YBgGIpEIkydPxuHDh2Fra4v4+HhZDaOv9x9XdnZ2uHDhAtLS0gAAx44dw9ixYwEAffr0kRWhS0lJwa1bt2Rx33zzTYm7W4WGhgIorF+UmJgIW1tbREVFwcXFBSNHjkSzZs1w5coVWV0feZKTk4uVbiblo1o6pEoEBgZi1apVcHJygkgkwpAhQ2Sj6aVLl2LixInQ1tYuVnelR48eWLRoET59+iQraWBqago/Pz9kZmbC1tYWkyZNglAoRJ8+fTB48GAIhULZFMfr169LFCz7p6ioKPTp0wd//fUXrKysZCPSFi1aYOXKlQgICADDMNDS0sKOHTugo6Mjq+tjYWGBpUuXFqs9U1r/5G1/YGAg5s2bB4lEAkNDQ9lpjf3798ecOXOwfPlyLFq0CGvWrIGTkxPEYjF69OiBCRMmQFtbG4GBgZg9eza0tbVha2ur9O/Kzs4Ofn5+GD9+PAQCAQwNDREUFASBQIAFCxZg2bJlcHJyQu3atVG/fn1Z3KBBg3Djxg1069ZNtuz+/fs4ceIEpFIpNm3ahJo1a2L8+PGyejsA0KFDBzx/rtgFPjdv3pRbbI+oJ7rSVgU9fvwYM2bMwJUrVwAAO3fuhKamJvz8/LB161ZkZmZi6dKlvLXn7e2NhQsXlphC4aIi+lfdTZw4EQMHDoSrqyuys7Ph5uaG06dPQ09PD61atcKtW7dgYmLCS1vR0dE4cuRImafZkpKO1mmrVLxnuurcC5hG+P8C48ePx+TJkzFs2DDe1/3f//4XnTt35iXZVze3b9/GunXrSn2ta9euFXJjaUNDQwQEBGDHjh2l3sZQGRKJBL/++mupNzEnZaNaOoQQoiZOmlkqFT8yLZannlQ8GuETQtSaqh14VQadpUMIIWqi2o3wb/XtzTqmUa+WrGNyUjJYxwCA0Ij9HYEk+SJObdVszr5+S17mJ9YxGprcvvdz07NYx+iYKH5zlK8VfMnjFMcWo8BpqqWRigtYx2jqlH92VFk0tNn/2abcfcU6pl7nb1jHAED+x5K3dpSHkXDb7612K3+KrYYaTeJXu4RPCCGViRI+IYSoCUr4hBCiJtQp4dNBW0IIURO8J3yxWIw//vgDMTExsmUfPnygmzIQQqolgaZAqYcq4X1KZ/LkyejQoQNSU1Nx//596Ovr4/jx45g4cSLfTRFCiNK4nqWminhP+Dk5ObK7IfXu3RsDBgxASEgIdHR0+G6KEEKUpqFio3Rl8J7wv67NXb9+fSxatIjvJgghhHDAe8JPT0+X3bDj48ePslrfACqkuBchhChDoEVTOpz5+/vL7vbz9c+EEFId0ZSOElxcXPheJSGEVBiBGt0Oki68IoSoNRrhV6F6nRqzjuFSWEvH2JB1DADkZXxmHfP5LfuCZgBQq3VT1jFZL96xjpGKuRWu0qvDfh/Wtm7Fqa2Cz+z3YVb8W9YxmkJufxISEfviaSIORcYAQKeWEesYs/bsC/GZtGH/+QOA9IcvWMeIK6k4nrqrdgmfEEIqk6pdPKUMSviEELUmUKNaOpTwCSFqTZ2utFWfLSWEEDXHe8LPyCj9TlKPHz/muylCCFGaOhVP4z3hT5gwAa9evZI9z8nJwerVq7Fy5Uq+myKEEKUJNARKPVQJ7wk/MDAQc+bMwf379xEeHo5Ro0ahefPmOHHiBN9NEUKI0jQ0BUo9VAnvB23Nzc2xe/duTJ8+Hfn5+Thw4ABq1arFdzOEEMILgRodtOU94RcVSxsyZAiCgoJw4MABfPPNNwCoeBohhFQl3r/aGIYBwzAQCoUICAhA06ZNZcsIIaS6qaw5/GnTpiE5OVn2/P379/Dy8sKoUaOwe/fuiti0Eqh4GiFErVX0PLxIJMLMmTNLnKl48OBBTJ06FV27dsXYsWPh6ekJQ0NuJV8URRdeEULUmrKnVgYHByM4OFj23N3dHe7u7rLnIpEIPj4+OH36dLG4iRMnyhK8RCKBpqamUv1QBCV8QohaU7Y88j8T/D8ZGhrC1ta2RMI3NjYGUPiF0bp1a+jp6SnVD0VUu4TPpRplbatmrGOOrLjIOgYA3Od+yzqmlgX7CqAAkJOWyTqmbmcL1jEFufmsYwAg+cYz1jF5GVc4tSU00mUdk3glkXVMW7cOrGMAQL+uCesYLT1u93n+9CqFdYxR4zqsY97eeMQ6BlCv2jTKOnv2LMLDw7Ft27ZKaa/aJXxCCKlMVXUu/Z9//onQ0FDs2LEDQqGwUtpUnxNQCSGkFJVdWmH37t2Ij4/Hjh078OHDB/j5+cHLywupqakVsHXF0QifEKLWKmsKav369QAK7/UNADt37qyUdr9GI3xCCFETvI/wY2JiynzN1taW7+YIIUQpVFpBCdHR0bKfz5w5U+xCLEr4hJDqRtUKoCmD94Q/depU2c/R0dHFnhNCSHWj7Hn4qqRCD9oKBOrzzUkIUU00paOEomqZDMMgPT1d9hygapmEEFKVeE/4X1fF9Pf3pyqZhJBqTZ2uDKZqmYQQtUZTOoQQoiY0KOFXHZM2TVnH5KRmsI5x9u3MOgYAGImEdUxG3DtObRk0qM06RqemAae2uKjRuCbrmNqW7AvdAUDuhyzWMc36sW9HQ8jtT+LL+w+sY7QN2BeEAwCpuIB1jG5t9r+rd7cTWMcAgGE99jXd87LyOLXFB3U6S0d9tpQQQtRctRvhE0JIZaI5fEIIURPqNKVDCZ8QotZohK+klJQU3LlzB2lpaahduzZ69uwJMzOzimiKEEKIgnj/agsNDcXKlSuhpaWFNm3aQCAQYMGCBcWuuCWEkOpCoKGh1EOV8D7CP3bsGI4dOwaNr3aEk5MTxowZQ6UVCCHVDk3pKEFHR6dYsgcATU1NaGtr890UIYQojRK+Ev5ZMA0orK/z4QP7C1MIIaSi0ZW2SiirYJqfnx/fTRFCCGGBiqcRQtSaQFOzqrtQaeg8fEKIWqMpnSrEpQhV9rtMTm3VatWQdYxOrRqsYwRJ6axjAKBms/qsY7Lfsm8r+y234yuMRMo6RtfEiFNbGtrsR2HGLdj/fjPiXrOOAQBJnoh1zOckbvtdrzb74mQfHrEvhNagW3PWMQC3EbOhSMypLT7QCF8NcEn2hJB/H3U6S0d9tpQQQtSc2o7wCSEEoOJphBCiNtRpSqfSEn5BQQG0tOj7hRBSvajTTcx5/2qbPn267OegoCDZz+PHj+e7KUIIISzwPuTOzPzfKZJ37tyR/Vza1beEEFLVaEpHCQJB6f8elbWcEEKqEl14pYSi4mkMwxT7mYqnEUKqIzpLRwlfF0/7+mcqnkYIqY5oSkcJVDyNEEKqJzpPkhCi1mhKhxBC1IUGFU+rMpo6QtYxNZrWYR3DpbohAHx5x/7gs0GD2pzaehPxgHVMTfN6rGP06hizjgEA0cds1jHvbj7m1JYkn301RZPWTVjHCGsYsI4BAI1a7KuA6tZmX3kVAHLTs1jHcKm8mvk8mXUMwG1OXGikx6ktPtAcPiGEqAl1mtJRny0lhBA1RyN8QohaU6cRPiV8Qoh6o4O2/BGJRJBKpdDV1a3opgghhD01GuHzvqUPHz6Ej48PCgoKcPLkSQwfPhwTJ07E77//zndThBBCWOB9hL9161Zs27YNmpqa2Lp1K06fPg0TExOMGzcOgwYN4rs5QghRioCmdJRjYGCAe/fuoXHjxqhTp/Ac+ZycnIpoihBClEMJn7smTZpg586diIyMxLhx45CdnY2jR4+idevWfDdFCCHKU6M5fN4T/uLFixEdHY2ePXuiXbt2SEhIgI6ODpYtW8Z3U4QQorSKnNIRiUSYMWMGMjMz4erqCjc3NwBAXl4eFixYgJSUFLRu3brS8iPvX22ampro0aMH2rVrBwBo3rw5xo4dC21tbb6bIoSQau3ixYvo1asXjh49it9//x35+fkAgBMnTqB37944duwYzM3NkZeXVyn9UZ//ZQghpDQaGso9yhEbGwsbGxtoaGjAwsICCQkJAIC7d+8iOTkZXl5eMDAwqLTT1qvdhVeMVFop7eRnsS/8BXArQsW1OBmXIllZ8e9Zxxg2qMU6BgAMG7IvWqdfz4RTW1wKySVefsL3e0NrAAAgAElEQVQ6pkmfVqxjAECUm886RvyZ24kMWvrsk4OYQ//06tRkHQMAmrrsCyBKRQWc2uKFklM6wcHBCA4Olj13d3eHu7s7AODLly/Q19cHAOjp6clOXvn06RPMzMywb98++Pr6ok+fPjAx4fa3wUa1S/iEEFKZlC2t8HWC/yd9fX3k5uYCAHJzc2FoaAgAqFGjBrp06QItLS20a9cOycnJlZLwaUqHEKLeNDSVe5TD0tISMTExYBgGcXFxaNasWbHlAPD06VPZ8opGCZ8QQiqIo6Mj/vjjD4wYMQIDBgzAgQMHEB8fDw8PD4SHh2PEiBGwtbWFkRH7+ylwQVM6hBD1VoHn4evo6GDbtm2lvrZr164Ka7csvG7p/Pnz+VwdIYRUOIGGplIPVcLrCP/du3d8ro4QQioeXWnLTVxcHLy9vcEwDAQCAQDIfj548CCfTRFCCGGJ14TfunVrSuyEENWiYtMyyuA14Tdo0IDP1RFCSIVTtXl4ZfCa8H/88Uc+V0cIIRWP5vAJIURNqNEIX32+2gghRM3RCJ8QotYEmuozwq92CV/PlH1lyc9vUljHCDS5/XOTl/mZdUz+xy+c2tLmUBXR1Lo565h3N2NZxwDAh7/Y7/fmTjac2jK1bMo6Jic9jnWMbm1uFSIl+WLWMZpCbn9+XKqo1m7XgnVMajT7aqMAkP0ug1NclVGjKZ1ql/AJIaQy0QifEELUhRqN8OmgLSGEqAka4RNC1JsajfB5T/ihoaFlvjZs2DC+myOEEOXQhVfcMQwj+1kgECAtLQ179uyBlZUVJXxCSLUjENAInzMXFxcAQEFBAfbt24fff/8da9euxbfffst3U4QQonbu3r2LAwcO4P79+9DS0oKmpiY6duwIT09P2NiUf9pzhczhX716FVu3bsXgwYNx/PhxCIXs72JPCCGVQoWmdFatWgVDQ0NMmzYNLVq0gMb/9/3Zs2c4d+4czp8/j+XLl5cZz3vC9/X1RWxsLMaPHw9TU1NcvHhR9hpN6RBCqh2B6iT8yZMnw9TUtMTyVq1aYc6cOfjw4UO58bwn/CFDhmDIkCGy51/P6RNCSHXD/P/NmlRBUbKXSCQ4fvw44uPj8c0338DDwwNCobDUL4OvyU34gYGBCAgIgIaGBj5//oxFixZhy5YtZb6/aA6fEEJUggqN8IssWbIERkZG6NmzJ+7cuYPFixfjp59+khsnN+ELhUL4+PjA29sbW7Zswbhx43jpMCGEEG5ev36NI0eOAAAcHBwwatQoheLkJvxp06Zh3rx5+OGHH7Bw4cIKH8H/HZvIOkYikrCOMevIvpgUV5nPkzjFmdlYsI5Juf0X6xhtfW4H1bX12RcMM7Rqz6mtj/fuso5p+q0V6xgunz8AMKhnwjrmSwq3ImM6NQ1ZxySFx7COadCT/f4DgI8v37OO0TWpwaktXqjgCD8/Px+5ubnQ09NDXl4eJBLFcqDcLR0zZgyMjY0RERGBGzduYMmSJUp3lhBCqg2BhnKPKuDt7Q1nZ2dMmTIFzs7OGDt2rEJxckf4fn5+6NOnDwBgx44ddJNyQsi/CqOCI/yhQ4fC3t4eSUlJaNSoEWrVqqVQnNyEX5Tsi3h7e3PqICGEEOVs374d33//PQICAiD4x9lFGzdulBtPxdMIIepNhU7L7NevHwAofJD2nyrtfxlFDyoQQkilUqE5/JYtW0IkEuHgwYPo2LEjOnToAGtrawQFBSkUL3eEHxcXh+DgYOTn58uWrVu3rsz3i8ViBAUFISoqCvr6+hAIBOjWrRtSU1Ph4eGB1q1bK9QxQgipFCo0h3/69Gns3LkTHz58wKBBg8AwDDQ1NeXW0CkiN+HPnz8fY8aMQb169RRa4Zo1a9ChQwfMnDlTtmzlypVISkqiZE8IqX5UKOG7ubnBzc0Np06dwogRI1jHy034pqamGDlypMIrTEhIKFG8p3bt2oiN5XajbEIIIcXZ2tpi165dEIsLr4VJS0vDypUr5cbJ/Wpr2LAhdu/ejRs3biAqKgpRUVHlvl8sFqOgoKDYMj8/P7kdIYSQqsAIBEo9qsK8efMAAPfv30dycjKysrIUipM7wheLxUhMTERi4v+uQLSzsyvz/d7e3pg6dSrGjRuHunXr4t27d9i3bx+VZCCEVE8qNKVTRFdXFxMnTsSrV6+wbt06eHp6KhQnN+GvW7cOz58/R3x8PJo1a4Y2bdqU+35HR0dYWFggLCwM6enpMDMzw7x589CiReWVMiCEEIWpUD38IgzDID09HTk5OcjJycHHjx8VipOb8A8dOoTffvsN1tbW2Lt3LwYPHgxfX99yY1q0aIFp06Yp1nNCCKlKKjjCnzp1KsLDwzF06FB8++23Ct9rRG7C/+2333DkyBFoaWlBLBZj1KhRchO+MmpbNmMdo6HN7fqxh7uusY5pYt+UdYzQSI91DAC8u8n+QLepJfv+GTSqyzoGAN5ee8A6ZlH/pZzaGtpJsbPEvlanbR3WMXp1arKOAYCCXBHrmPrdLDm1lRoTxzqmXre2rGNeX2b/+wUAvTrsi7sJNFUv6ValR48eyfIwm9vHyt3LDMNAS6swoWpra0NbW5tjF6sXLsmeEPLvo4oHba9du8bpYla5Q2MbGxtMnz4dNjY2uHfvHjp27Mipg4QQUi2p4JROZmYmevXqhUaNGkEgEEAgEOD48eNy4+Qm/Hnz5iEyMhIJCQkYPnw4evfuzUuHCSGkWlDBhL9z505OcWVu6dWrVwEAwcHBSE1NhaGhIVJSUhAcHMyth4QQQnjRsGHDEg9FlDnCLzqRPz09nXVnrl27hl69ekFDQwPPnz/H+/fv6T8DQkj1pIIjfK7KTPhFtzLU0NDA999/L1sur+bypk2bkJ6ejq5du0JXVxcGBgYICwvDn3/+ienTp/PUbUII4YmKJ3ypVAoNBa8lKDPhnzx5EqdOnUJCQgKuX78uW7FYLMasWbPKXGF0dHSxgwcNGzbEunXrMGrUKEr4hJBqp6rOtFFGWFgYpFIpRCIRfvrpJ0yYMEGh0+XLTPjOzs7o3r07du3ahUmTJgEoHO3Xrl27/BVqlVxl0VFkQgipdlRwhL93717s3r0bAQEBuHbtGsaPH69cwn/27BnatWuHAQMGFKujk5CQUG4tHVdXV8yaNQsjR46EmZkZUlJScOrUKTg5ObHcJEIIIaXR0dEBABgYGEAoFOLLly8KxZWZ8G/duoV27drh4sWLJV6Tl/Bbt26Ny5cvIyMjA2ZmZvDz85Nbg4cQQqqECo7wGzVqhOHDh2PJkiUICgqCtbW1QnFlJnx/f38AhcXTJBIJGIbBw4cPFVpx27Zt0bYt+0u5CSGk0lVgwheJRJgxYwYyMzPh6uoKNze3cpcrav369fjy5QsMDAxgZWWFOnUUKyMi98KrDRs2oHHjxnj37h1iY2NRp04drF+/nlXnCCGkuqrIg7YXL15Er169MGrUKEyYMAHOzs7Q0dEpc7k8CxYsKPO18m49W0TuV9u9e/cwatQoPHjwAHv27MH79+/lrpQQQlRGBd7EPDY2FjY2NtDQ0ICFhQUSEhLKXS6Po6MjHB0d8fHjR5ibm2PEiBFo1aoVRCLFivfJHeFLpVI8evQIjRo1gkgkQkZGhkIr5iorPpl1TM1m9VnH1G1vxjoGAIyasK8sqWPMvnogwK0K6MdE9l/Ihk3Z7z8A0K9bi3WM/4ROnNr6+CqTdUyjford2PlrXCqAAoD4Sz77mOwcTm0Ja7L/PL04Hc06puY37H+/AJCbns06piCH/f6rLoKDg4tVIHB3d4e7uzsA4MuXL9DX1wcA6OnpIScnp9zl8vTq1QsAsG/fPtmdBG1sbBS+wZTcjOLs7IxVq1Zh7dq12LBhA7y9vRVaMSGEqAIGyk3pfJ3g/0lfXx+5ubkAgNzcXBgaGpa7XFE5OTmyE2sePHggu7etPHKndEaPHo3du3cjNzcXkydPZnVDc0IIqe4YJR/lsbS0RExMDBiGQVxcHJo1a1buckWtWbMGx44dg5ubG06cOIFNmzYpFCd3hH/x4kVs3rwZzZs3x4sXLzB16lQ4Ozuz6hwhhFRXUnlZWwmOjo4ICAhASEgIXF1dceDAAfTr16/EckUO2H6tefPm2LJlC+v+yE34Bw4cQEhICAwMDJCdnY2xY8dSwieEEAXo6Ohg27Ztpb5W1vLyFF0DJRaLkZubi/r16yM1NRUmJia4cuWK3Hi5UzoCgQAGBgYAAENDQ7nfRC9fvpTVzHF0dIS3tzccHBwQHh4utzOEEFLZpIxyj8oUFRWFqKgo9OrVC5cuXZI9lL7wqkiTJk2wfv16dO7cGXfv3kWTJk3Kff/atWuxbNkyAEDt2rVx8OBBpKamYvbs2XBwcFCoU4QQUlkqOWfzIjk5GfXrF55dV7duXYVPl5eb8NeuXYvg4GDcunUL5ubm5VbKBID8/Hw0btwYAGT1c+rWrQtNTU2FOkQIIZWJYVQv5Tdv3hxz5syBtbU1Hjx4ABsbxU5BlpvwCwoKIBaLIRaLFd4xnz9/hpGRkexy4U+fPlG1TEJItVTZ0zJ8WLVqFa5fv474+HgMGTIE/fr1UyhO7hx+QEAAPnz4AHt7e7x7967cS3sBYNq0aZg0aRKOHj2KmzdvIiQkBFOmTMGMGTMU2xJCCCHlysnJQXR0NKKjo3H79m3ZHQrlkZvws7KyZPPvc+fOxbt378p9f5cuXbBt2zbo6+vjyZMnAIDNmzejQ4cOCnWIEEIqU0Weh19RFi5ciAYNGiAgIAANGzbE/PnzFYqTO6XTokUL3Lt3DzY2Nnj27BkaNGggm94RCoWlxhgbG2PYsGHstoAQQqqAKk7pZGZmwsvLCwDQpk0bXLp0SaE4uQn/3r17iIqKgra2tuzy3YEDB0IgECAiIkKJLhNCSNVTxYO2+fn5SE9PR506dfDhwwdIpVKF4uQm/AsXLijdOTaMWzRiHVOQm8c6RkuXfWEyANDQZF87+/3tOE5t6depwTrmS+pn1jHp956yjgEAAYd9YWrdnFNb2vpvWccwBYrVF/macfMGrGMAICuh/KnO0mjqlP4fsjxc9nuddty2iwuhkV6ltcUH1Uv3wIwZMzBq1CgYGRkhOzsbq1atUiiOW9YjhBBSZXr27ImIiAhkZGTAxMRE4TjVu7cXIYTwSJWutJ07dy4iIyMhkUgAQJbspVIpwsPDMXv27HLjaYRPCFFrqjSFv3r1ahw4cAAbN26EkZERTE1N8fHjR2RkZMDJyQlr1qwpN551wp8+fTqnKm2EEFIdSVVoFl8oFMLPzw9+fn549eoVMjMzUbt2bbklb4qwntJR5OBAbm6uwrfcIoQQopj8/HwcPnwYJ0+eRIMGDdCxY0c0adIEx48fVyhebsIvKCjAgwcPEBMTgzt37uDGjRvlvn/nzp3w9/eHr68vLl68qNhWEEJIFWEY5R6Vae7cuUhLS0NiYiI8PT3x8eNHAFA418qd0pk6dSrEYjHS0tIgkUhgZmaGIUOGlPn+a9eu4dixYxCJRJg4cSIcHR0V3BRCCKl8qjSHn5GRgc2bNwMA/vvf/2Ly5MnYv3+/wtcSyB3hZ2dnY8+ePbC2tkZISAjy88u/2XDR1bdCoVDhiwEIIaSqSJV8VCaxWIyMjAwAQP/+/TFgwADMnj1b4Xvayh3ha2kVviU3Nxe6urpyV5yWlobQ0FAwDCP7uQiVWyCEVDeqdKXtjBkzMHr0aBw6dAimpqbw8fFBbm6uQne7AhQY4ffv3x/btm1D69at4ebmJvfu6v7+/rIdWPRz0YMQQgh33bt3R1hYGExNTWXLJk+ejGvXrikUL3eEP3r0aDAMA4FAgN69e6Np06blvt/FxUWhhgkhpDpQxeJpAwYMkF18BRTOxNSvXx9z5syBpaVlmXEKnZZZdPOSVq1a4fz580p2lRBCqg9VLI/crVs3rFq1CmFhYVi7di3atWuHiRMnYvXq1eXGsT4PX09PtQojEUJIeVTptMwiiYmJ6NGjB4RCIbp27Yr09HR0794dGhrlp3TWV9oOHDiQcycVkfYgnnWMXu3yjyuURtfEiHUMAGgb6bOOafJtR05tZT5LYh2jX4f9vtCpxW1fiD59YR0Td+QPTm3VaVeXdUxOSgbrGC5VLwFAv24t1jE5qZmc2pLks7+o8VMS+7ZMrRqzjuGKSwVQvkhV8PiiUCjEsWPH0LFjRzx48ABCoRBPnjwpNs1TGrkJ/9ixY9i/fz8KCgrAMAy0tbUVLrZPCCGEf4GBgdi5cyeuXLmCli1b4qeffsKjR4+Ur6Vz8uRJHDp0CDt27MCgQYNw4MAB3jpNCCFVTfXG90CtWrXQvXt3mJqaolmzZqhVqxZ69+4tN07u/1G1atWCmZkZvnz5gq5du8ou5VVUdnY2Tp8+zSqGEEIqiyqVRy6yceNGhISEQFtbG6GhoVi/fr1CcXJH+EZGRggPD4dAIMDx48dlV3mVJz8/HxEREbh48SJSUlLQr18/hTpDCCGVTQWn8BETEyMrmDZ27Fi4ubkpFCc34a9evRpv3rzBrFmzsHfvXixfvrzM9167dg0XLlxAWloaevTogY8fP+LUqVOKbQEhhBCFFBQUQCqVQkNDQ3adlCLkJnxDQ0O0bdsWADB//vxy37tmzRp069YNM2fORPv27RETE6NQJwghpKqoUj38Io6OjvDw8ED79u3x6NEjhYtU8nrHq8uXL+PRo0f4/fff8eOPPyIlJQU3btxAt27doK2tzWdThBDCC1W80nb8+PGws7PDy5cvMWLECFhYWCgUx/stDq2trWFtbQ0AiI2NxaVLl7Bp0yaEhITw3RQhhChNlc7D37hxY4npm7/++gsAEBAQIDe+Qu9pa2lpCUtLS4U6QgghVUGVRvjm5uZKxdNNzAkhREUoW5ySEj4hRK1JVWmIryRK+IQQtSZRoTl8ZVW7hF/TvB7rGKFh5VXwTH/Ivrgb10Jtos85rGP0TGuyjtHS02EdAwBGjc1Yx2Q8e8+prYKc8m+tWZrcD+yuCgeAgjzFbhX3T5nP2Bddq2nOfv9xZWxeh3VMQU4ep7Z0a7P/DEpE3PY7H1TpoK2yql3CJ4SQyiRRo1tvV11NUkIIIZWKRviEELVGc/iEEKImJGp0lg6vUzoREREQidjfjYcQQqqKlGGUeqgSXkf4CQkJOHToEOrUqYOBAwfC3t4eQqGQzyYIIYRwxGvC9/f3h7+/P1JTUxEeHo4ffvgBNWrUgIODAxwcHPhsihBCeCFRrUG6UirkLJ26devCw8MD3t7e0NbWxvbt2yuiGUIIURpN6XCUnZ2N69ev48qVK7KboIwfPx7NmjXjsxlCCOGNqiVtZfCa8MePHw8HBwdMnz4dTZo04XPVhBBSIdTpwiteE/6JEyf4XB0hhBAe0ZW2hBC1Vtlz+JmZmfD29saoUaMQGRlZ7LWMjAz4+/vDzc0Nu3fv5mkL/+dfceEVlyJZAKBf14TnnpROoMHte5VLwTBtfV3WMVnxyaxjAEBLn33RNU1tTU5t6detxTomN53950Jbn9tpxEaNTDnFcaHNoVighib7zyDnvyszY9Yxn5PTObXFh8qewz969Ch8fX3RrVs3TJo0CX369JG9tmvXLvj7+6Nz58749ddfeW9bbUf4lZXsCSHVm4RR7sFWbGwsbGxsoKOjA0NDQ3z8+L8v1oSEBFy9ehVeXl5o2bIlj1tZSG0TPiGE8CE4OBiurq6yR3BwcLnv//LlCwwMDAAAenp6yMn5Xxn0169fo0ePHti5cye2b98OqZTfI8r/iikdQgjhStkpnVHu7nB3dy/z9YMHDyIsLEz2/PHjx8jJyYGBgQFyc3NlyR8AjIyMYGtrC6FQiAYNGiArKwsmJvzNRlDCJ4SotYqew/f29oa3t7fseVBQEO7evYtu3bohKysLNWrUkL1mZWWFu3fvokuXLkhLS+M12QM0pUMIUXMSqXIPtjw9PbF//354eHhg7NixAIA1a9YgKysLkyZNwo4dO+Dm5oaRI0fyvKUVMMIPCgoq87WpU6fy3RwhhCilss/SMTExwb59+4otW7RoEQDA2NgYhw4dqrC2eU/4Xbp04XuVhBBCeEAJnxCi1qiWDiGEqAk1uuEVJXxCiHqje9oSQoiakKrREJ9OyySEEDVBI3xCiFpTp1scVruEz3C4kkGSL2Ido6HNbdO1dLVZx+R/zObUlqYO+7a47AstXW4VIrn8rup1seDUVu6HLNYxQiP2VSUL8tjvPwAQZ+fIf9M/aHGobMpV9lv21SiFNQ05tSXl8Lmo0aQup7b4oE5TOtUu4RNCSGVSp4O2NIdPCCFqgkb4hBC1JqEpHUIIUQ/qdKVtpU3p3L17t7KaIoQQhUmkjFIPVVJpCX/9+vWV1RQhhCiMEj4hhJB/Hd7n8ENDQ0ssYxgGWVnsz6MmhJCKpmqjdGXwnvCZMg6ATJkyhe+mCCFEaZTwleDi4sL3KgkhpMJQwieEEDVBV9oSQgj516l2I/yCnDzWMXW7WrGOeXn2D9YxAFCnQzPWMTXMG3Jq6+9H8axjuBRPM2pSj3UMAPwdm8g6RkOT2xjjU1Im65iPbz6zjqltUYt1DAAINASsY3Q4FBkDAKm4gHXMk9N/sY6xGt6WdQwAiDkUXZNyLODHB5rSIYQQNUEJnxBC1IQ6JXyawyeEEDVBI3xCiForUKNbXvGe8GNiYsp8zdbWlu/mCCFEKQVSbgfPVRHvCX/q1KnQ09ODvb09zMzMir1GCZ8QUt2o0xw+7wn/1q1bePjwIa5fv47379+jbt266NOnD6ytrfluihBClEYJXwkaGhro1KkTOnXqBAB49uwZfvzxRzx//hxRUVF8N0cIIURBFXLQ9tGjR7h+/TqePHmCWrVqwdXVFXZ2dhXRFCGEKIVG+Ero1q0bDA0NYW9vjwEDBkAgEKCgoACRkZEYNmwY380RQohSKOErYd68eSWWlVUymRBCqholfCVQeWRCiCqhhF+FDOrXZh3z6WUy6xjTdk2gZ8a+UFZSxJ+sY8TZuaxjAOBjYirrGLMO5qxjJHn5yEljX5zMrGNL1jEvL3C7mX2Nxux/V037d+DU1qfE96xjuBQ009LXZR0DcCswaOPXjX07X9i3AwCZz5NYxwiN9Di1ZcopSn1Vu4RfWbgk+38rLsn+34pLsieqjUb4hBCiJtTpBiiU8Akhak2dRvhULZMQQtQEjfAJIWqNRvhK2Lt3r+znt2/f8r16QgjhlVTKKPVQJbwn/MjISNnPCxYs4Hv1hBDCK3VK+DSlQwhRa1I6S4e79PR0hIaGgmEY2c9FqJYOIYRUHd4Tvr+/v6x2ztc/E0JIdaRq0zLKoFo6hBC1pkZ3OKQ5fEKIelOnWQi68IoQotYYKaPUgwuxWAw3N7cSy58+fQoPDw+MHDmy2PFPvlS7Eb6IQ2VJSZ6IdYy2oT7rGAAwbGDMoS1ulQC5VL78yKH4F9cPbU5aFuuYoxcTOLU1Y0E/1jHaBuyrUeqZ1mQdAwDi3Hz2MZ9zOLWlW7sG65gPj9+wjhEa6bCOAYCE8ETWMVYe3CqbqqKsrCzMmDEDSUklq4pu374dGzZsQP369eHp6cn7iS40wieEqLXKHuHn5uZiwYIFaNGiRYnXli9fjkaNGhX2qwKmmqrdCJ8QQiqTsufhBwcHIzg4WPbc3d0d7u7uZb6/fv36qF+/fqmvmZiYAAB++eUXODo6KtWv0lDCJ4SoNa5TmkXkJfiDBw8iLCxM9nzSpEno3bt3me/fuXMnPnz4gICAAKX6VRpK+IQQUoG8vb3h7e2t0HsvXbqEFy9eYMOGDRAIBLz3hdc5/NmzZ+PYsWN49uwZn6slhJAKUxVn6fzTmjVrkJWVhaCgILx69Qpjx46Fl5cXJBIJL+svwusIf8aMGXj8+DFCQ0Px5s0bCIVCtGrVCp06dUKXLl34bIoQQnhRVafhHzp0SPbzokWLAADnz5+v0DZ5TfiNGzdG48aN0bdvXzx8+BBPnjxBZGQkbt++TQmfEFItUWkFjgIDAxEfHw8dHR106NABXbp0wbhx46ClRYcKCCHVE1/TMqqA1zn89+/fQygUokGDBmjatCmaNm1KyZ4QQqoJXrPxxo0bAQDJycm4d+8eNm3ahKysLBgbG2PFihV8NkUIIbxQp1o6vCb8goICPH/+HLGxsXjy5AkyMzNhaGiIli1b8tkMIYTwRp2mdHhN+D4+PmjXrh06dOiAKVOmwMzMjM/VE0II7+igLUeHDx9Weh2fXqexjqlpXo91zNuoONYxAKBTQ8g6RkuffREvAGA4FOo2tW7OOoZrEa8X5x6yjnFqX5dTW1wK5GXFv2UdI9DkdlhLU6jNOsbAojGntlJu/8U65ptBNqxjkiL/ZB0DAE3tGrGOyU3/yKktPqjRjA4VTyOEEHVBp9AQQtQazeETQoiaoIRPCCFqQtnyyKqE1zn8x48f87k6QgghPOJ1hL9r1y4YGxsjICBAVsifEEKqM5rS4SgoKAi3b9/GnDlz0LdvX/Tr97/7kDZo0IDPpgghhBd0Hr4SunTpgjdv3mDfvn2IjY2VLV+3bh3fTRFCiNJohM/RlStXEBQUhH79+uHMmTPQ1eV2wREhhFQWqYQSPidhYWHYsWMH6tbldjUlIYSQisNrwt+wYQOfqyOEkArHMOxLmKgqOg+fEKLWaA6fEELUhJRDkUJVJWDUqfo/IYT8QzO/E0rFJ/7HjaeeVDyqlkkIIWqCpnQIIWqNy30nVBUlfEKIWmOkkqruQqWhhE8IUWvqlPBpDp8QQtQEjfAJIWqN5vAJIURN0JRONRMSEoJBgwbBy8sLHh4e8PX1hUgkkhsTEhICAHj58iVcXV3x4cMHhWMAYOvWrYiOjmbVv5EjRxZbhyIxXl5emDp1qtyYgoICLF++HOPHj8e4cePw4MEDuTEAMHfuXHh5eaFz587w8vLC3LlzFeqfr69vseeKbHo9r2MAAAwkSURBVNfFixfh7u6O0aNHY+nSpXJ/T0VOnz6N0aNHw93dHT///LNCMXfu3MGYMWMwevRo+Pr6IiEhQW7M1/t91KhR2L59u0JtFenfvz+uXr0q930jR47E58+fZc937dqF8PBwhduJjo7G1q1bFX6/or+fr9/ftWtX2UVHEokEXbt2Zf159/Lywq+//lrme0ePHi37DMyaNQs//fQTAODNmzeYNWtWue3Mnj0bERERsudTpkzBo0eP5PaPLUYqUeqhSlRmhO/v7w9XV1cAhYn4zp07sLOzkxuXkZGBuXPnYsOGDTA1Na3w/kkkEkydOhXdunWTew+Ar7dJEdHR0TAzM8Py5cuRnZ2N77//HgcPHpQbV/RH5uXlhUOHDinc3pMnTxAREYFvv/1Wofe/ePECZ8+exaFDhyAUCnHmzBkEBQUhICCg3LjY2FjcvHkThw4dgoaGBhYsWIDIyEj06dOnzJjMzEz88ssv2L59O4yNjfHmzRts2LBBoST59X738vLC2LFjYWBgIDfuzz//RLdu3XDhwgX07du33Pc6ODjg+vXr+O677wAAf/zxB3x8fOS2UZkMDAzw6NEjdOjQAQ8fPoSOjo7CsYp+dq2srPD8+XNYWVlBLBYjPj4eQOFnq1OnTuXGzp49G5MnT0bv3r1x9+5d1KxZE9bW1gr3UVGqlrSVoRIj/H9KTU2Ftra23PeJRCJMnz4d8+fPR/PmzSuhZ4Cmpia+++473Llzh/d1GxkZ4dq1a7h79y709fXLHVnxYfz48QgKClJ4lB4WFgYfHx8IhUIAgIuLC6KiouTGXb58GSNHjoSGRuHHccmSJXK/zK9du4bvvvsOxsbGAIAmTZpgy5YtCvWziEgkglQqhZaWYuOe33//Ha6urkhLS0Nubm657x00aBCuXLkCAEhJSYGpqSmrhFoZ7O3tcf36dQDA9evX0bt3b97b6NixI2JjY5GUlISGDRtCV1cXOTk5ePz4MWxsbMqNrVevHvr164fTp09j165dcv8jIPKpTMLfvXs33N3dYW9vj4YNG6Jr165yY4KCgpCXl4fMzExW7RT9q3rmzBlOfTU2NkZaWhqrthRJVtbW1vDx8cHPP/8Me3t7hIaGcuqfourUqYPBgwdj//79Cr0/PT29RGlsPT09uXEZGRkwMzOTPdfX15ebhFNSUtCkSRMAQHBwMLy8vDB48GCF+lm03/38/ODp6alwIn748CE6dOiAnj17yp3Wadq0KdLS0lBQUIDw8HAMGjRIoTYqU7t27WQ3KUpPTy/2O5Dn689uUlJSme8rSvj37t2DjY0N2rdvj4cPH+LVq1ewsLCQ286ECROwb98+9OnTB7Vr11a4f2wwUqlSD1WiclM6q1evhr6+vkIxw4cPx9ixYzF27Fh0794dhoaGCrcDgNUc6tfS09MV+nCyndKJj4+HnZ0dBg8ejNTUVEyYMAFDhw6t0BvN+Pj4wN3dHYMHD5Y7JVanTh28f/8e5ubmsmVFo3ZF4or+C0tOTsaXL1/QqlWrcmOKvlTd3d3h7u4OLy8vRTaJ9X4HCqcg3r9/jwkTJiAvLw81a9aEo6NjuTHdu3dHTEwMbt68iU2bNrFqrzIIBAIYGxvj6dOnrO9hoeg+rFu3LrKysnD//n388MMPMDU1xZ07dyAUChX6bOjp6aFDhw4KTytyQVM61djs2bMRHBys0Ai6adOmMDExwahRo/DLL79UQu8KD3799ttvCh1fYCsqKgrHjh0DAJiamsLY2LjCK/0JhUJMmTIFe/bskfvewYMHY+/evcjPz8e2bduwYsUKhf4Tc3BwwKlTp8AwDBiGwdatW/Hu3btyY/r27YuQkBB8/PgRQOE039u3bxXbKA4uXbqEFStWYM+ePThy5AhSUlLw6dOncmMGDRqECxcuwMjIqNre/c3Ozg4//vhjhUznFDEwMEBWVhZMTExgaWmJyMjIcr/MK5s6HbRVuYSvq6sLf39/bNu2TeEYDw8PPHjwoEKO8Bcp+hd39OjRGDJkiEIjpq//Lfby8pI7V+7p6YnXr19j+PDh8PHxgYeHh8L/7SjDwcEBlpaWct/XsmVLDB8+HF5eXoiMjERiYiLS0tIgFovLjWvbti1sbW3h4eEBNzc31K9fX+5BURMTE8yaNQszZsyAt7c3AgICMG/ePFbbxUZUVBS6d+8ue96vXz9cunSp3Bhzc3PExsbCwcGhwvr1taCgILi6usLV1VXhY0h2dnZ4/vx5hRwMLdK2bVvZsRZtbW1oaWnJPWBLKgaVRyYVKi4uDm3atKnqbhBSJjPnQKXi087O5qknFU9l5vCJaqJkT6o7VZuWUQYlfEKIWlO1M22UQQmfEKLW1GmEr3IHbQkhhHBDI3xCiFqjET4h1cDu3bvx6NEj5Ofn4+TJkwAKC3d9XVCrqkgkEkyfPl1WmoCoLjoPn5BqwN/fH9bW1khPT5clfFdX1wq96lIRb968wZgxY/D48eMq7QfhB5VWIGopOzsbixYtwufPn5GZmYmRI0fC09Oz2HsOHz6My5cvo6CgAEZGRti6dausWNo/OTo6onPnznjx4gVq1qyJn3/+Gdra2li4cCGSkpIgkUgwbtw4ODo64siRIwgNDYWGhgY6der0f+3dX0iTCxjH8e9SB1MvGhNMSYImFd04xG48EE2QpD9eSGmjtiAQ1AIxAgcuKlAabBcDBxqr5USs14tuLKQRjIpDUFmXTqkLSS0U4vXijHRjnQvZS/Z6PHXU/pw9n7vt/bN378XDw/vn99DZ2Ynb7ebIkSNEo1HevHlDMBjk8+fPFBUV4XA48Hq9jI+PA3Ds2DHOnj2L2+3GaDQyOzvL/Pw8Xq931UtjsViMUCjE0NCQlrW0Vlx05rcPHjyoW5ZIJOju7iYUCm3kdItfxO/WpW+EdPhCMz09zdGjRwmHw/T39+tC09LpNKqqMjAwwPDwMKlUat0u99OnTxw/fpw7d+6we/duFEVBURTMZjN3797l9u3bBAIBPn78yL179+jq6kJRFMrKykilUtp+WlpaKC8vXzUzIBaLMTMzw8jICMPDw9y/f5/JyUkASktLuXXrFk6nE0VRVh2T3W5n//79dHZ28uLFC110cyQSwel08vTpU3w+H06nU9tvxr59+35Y+qoQm0k6fKEpKioiEokQjUYpLCxcVXRhJQgtLy+Pixcvkp+fz4cPH3TrfCk3N5cDBw4AUFlZyZMnT8jJyaG6uhqAwsJCrFYr79694/r164TDYfx+PzabjX97Afzt27dUVVVhMBjIy8ujoqJCG4CSedlrx44dvHr1Srdtc3MzdrudQCCgS+XMhO2t1+GL/5e0dPgiG4XDYWw2G36/n7q6Ol3RjcfjPHr0iEAgwOXLl0mn0+sW5lQqRTweB2B8fJzy8nKsVisvX74EVi4hTU1NsXPnTkZGRrh27RpDQ0NMTEysmua1bds2XUic1WrVLuckk0lev37Nrl27gJUUyPVcuXKFrq4uent7tfA1kb1+xk3bZDJJY2Oj7vuFhQXOnDnDyZMnefz48Ub/mo50+EJjt9u5evUqo6OjbN++nZycHJaXl3n48CGJRIL6+npMJhMNDQ0YjUYtolhVVTweD8FgULfPUCjE3NwcpaWldHR0ACsDThwOB0tLS1y4cAGLxcLevXs5ceIEZrOZ4uJiKioqtJF9FouFZDKJz+fTUiftdjvPnz+nqamJZDJJXV3dNwW8RSIRLBYLp0+fxmQy4fF41ozB9nq9GzmV4jfyo2+8qqpKe3v7mnMEHjx4QGNjI7W1tbS0tGx6iqmEp4ktU1NTw9jY2C836UmILxX80bGh7f/68/tmHbx//57FxUV6enp0I0efPXvG9PQ0hw8fxu12c+PGjQ0d29ekwxdCZLXvLdhfyzyMkJEZyPNPSkpKKCkpWXNZfn4+/f399PX1cenS5qdwSocvhBBbaHBwkLGxMe1z5lKN0+nUdfjnz5+ntbWVPXv2cO7cOW7evLmpw3OkwxdCiC3kcrlwuVzftK7JZKKgoACj0YjBYFj3Kbj/Qp7SEUKIn6ynpwdVVWlra8Pj8XDq1CkOHTr0TXO4v4dc0hFCiCwhHb4QQmQJKfhCCJElpOALIUSWkIIvhBBZQgq+EEJkCSn4QgiRJaTgCyFElpCCL4QQWeJvpJ0cP7k0xsMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x432 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set_fontsize_screen()\n",
    "fig, ax = plt.subplots(figsize=(6,6))\n",
    "sns.heatmap(df, cmap=cmap, vmin=-1.8, vmax=1.8, square=True, ax=ax, cbar_kws=dict(label='log2(odds ratio)'))\n",
    "ax.set_ylabel(\"a.a. position x\")\n",
    "ax.set_xlabel(\"a.a. position x + 1\")\n",
    "# ax.set_title(\"bacterial kingdom, taxid:2\\nfreq(pair)/expected_freq(pair)\")\n",
    "ax.set_title(\"Tenericutes phylum, taxid:544448\\nfreq(pair)/expected_freq(pair)\")\n",
    "plt.savefig(Path(plotsFolder) / speciesSetName / (outputFilePrefix +\n",
    "            \"_composition_bias_aa_pairs_bulk_theoretical.svg\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = compute_subseq_oddsratio_plot_data(oddsRatioDfTableAApair2, terminus, statistical_test)\n",
    "maskData = compute_subseq_oddsratio_mask_data(oddsRatioDfTableAApair2, pvalueThresholds,\n",
    "                                              terminus, statistical_test, subseq_type='aa')\n",
    "pvalAnnotTable = compute_subseq_pvalueAnnotation(oddsRatioDfTableAApair2, pvalueThresholds,\n",
    "                                                 terminus, statistical_test)\n",
    "maskDataSmallN = compute_subseq_smallN_mask_data(oddsRatioDfTableAApair2, pvalueThresholds,\n",
    "                                                 terminus, statistical_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20, 20)"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotData.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20, 20)"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskData.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20, 20)"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskDataSmallN.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(cterminal)\n",
    "from cterminal import plot_seq_pair_composition_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.0"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FontProperties(size='medium').get_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5.8309999999999995"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FontProperties(size='small').get_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:424: MatplotlibDeprecationWarning: \n",
      "Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warn_deprecated(\"2.2\", \"Passing one of 'on', 'true', 'off', 'false' as a \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAJIAAAD2CAYAAAA9Ht7CAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJztnXtc1FX+/19z+8BwUW4qKijewFTAa0rmuqa5Wu5mtFoaq3ajNrRHWpnuFmtZ4pplNaatfTfZIhJFSlPXu6W/8i6EhrgWIl4QkJsMw9w/vz+mz2fmM/P5zHwGPjBA5/lPzfGczznzOW/OvM95X46MpmkaBEILkft6AITOAREkgiQQQSJIAhEkgiQQQSJIAhEkgiQoVqxYsaItOtJoNLh48SJyc3MxduxYUBQFjUaD+Ph4rF27FiNGjABFUWzd4cOHY+PGjRg+fDiUSiXb/tixYxg7dqzLcy9cuICBAwdCpVJhz549OHXqFHbt2oWkpCTI5XLs378fq1evRk1NDSIiItClSxe2/fXr1/Hjjz8iNDQUXbp0weHDh3HgwAEcO3YM48aNAwAUFhYiJycHBw4cwOjRo6FSqTjl+/btQ1JSEhQKBTQaDYqKipCXl4dx48axdZcuXYrQ0FD06tWLHX9GRgYmTJiAo0ePorS0FDExMQCAy5cvY9myZdBqtYiPjwcA5OXlgaIohIWFse3T09Ph7++PjIwMJCYmcr4X816GDh0KuVyOdevWISkpCbdu3cKOHTuQkJCA9evXIzY2Flu2bEFoaChCQkKaNb9tuiLNmTMHiYmJuHnzJlv23nvvISUlBUFBQaLa9+3bF8XFxS7lKSkpCAgIAADcuXMHRqMRCxYsgFKpBABMnToVCQkJmDNnDqKiojjtHccDAPn5+UhNTUV4eDgqKioAAIcOHcKiRYuwdOlSth8AOHjwIBYtWoSFCxdCLre/zrlz5yI+Ph7l5eUAgIqKCowZMwbnzp3j9BUVFYXy8nKcPn0a48ePZ8sHDRqEhIQEPPLII27fSbdu3XDmzBnwHQcy70WhUAAAlEolMjMzUVRUhH79+gEA5s+fj/Xr14OmafTt29dtX+5oU0HKzMxESUkJBg0axJbFxMTgyJEjLnWzsrJQUFDA+xyZTMb5vG3bNmzfvp39nJycjOnTp+Nf//qXi5DwMXbsWJw4cYL9bLVa2f8ykwAANE3DZDLBaDRyxlJVVYVPP/2UFRoA2Lp1K8xmMwYMGAAA2LdvHxoaGnDp0iVO+wceeAC7d+8GRVGs0HtLdHQ07r33XpfyrKwsZGVlsZ9VKhW0Wi127NjBClJwcDACAwMxa9asZvXN0KaCtGDBAqhUKty+fZstmzVrFsxmM4qKijh1U1JSMHz4cE5ZdnY2SkpKEBcXxyk3m81oampCZWUlACA3NxeHDx9GUFAQgoODPY4rMTGRbQsAd999Nz744APU19cjIiICADB9+nRs2LABGzdu5AjCtGnTsHXrVgDgrEizZ8/G3Llz2c+VlZV48sknMW/ePHz33XdseXh4OIqLizFx4kTOmC5fvozCwkLOHwgAfPPNN8jKyoLBYGDLZs6cCbVa7fK9UlJSkJKSwn6mKApxcXGoq6tDZGQkW65UKtmf3+YiIyYSfgoLC7Fr1y6kpaWha9euvh5Ou4cIEkESyPafIAnN0+68wGg0wmAwcPQHQvvFarXCz8+PPYoRS6sLksFggNFohL+/v8e6ZrMZZrOZ/SyTyThbV1/hPC4GpVLp83EBgMViYbf/QmPi+w58dZmNhM8ESaPRIDg4GD///DOWL1+OwMBAALadTO2mDChr7buimDWfAQAKH53OecbAzdvZcxu1Wo2ePXuivLwc4eHhuH79OlsvNjYWAFB06w6n/ZDILvjf//7HKWPq7vjpFqf8oaGRsJSd55Qp+tgO/hyfERsbC51Ox46LQa1Wo0dIMGQ/HWTLVHc/BADYU8yt+8DgHoLjcu7LuYwp/+pCOafs4WE9YTQaUVFRgaamJrY8KiqK866EvoNarUaPHj1QWlrKqdu7d284IzS3jkj6e8N34NgcHIXI8SW1F5jxmfU6n47DYrE06/0w47dYLKLbeJpbSQWJ78DRWzqKEJWXl4O2uv7ctSV8G27nw1pnOOP3YsPuaW4l2/5rNBqkpqbio48+wrx589iDvMbGRgDgXQ6daWpqgkqlchGiqKgojlmirdHpdLh+/bqLkLeXcTEwP1d8+o1Op0N1dbXH8fPNl9DcOtLq50jeCJLRaITJZGJNFAwqlcpFWTcajYKKufOLNBqNsFgsUCgUrGKqVCp5X7jzcwGbnkfTNKc9U86MlenfarW6jF+ovdAYxKLX62Eymdg+lEolaJp26Z/5d7lc7nH8Wq0Wfn5+oubLkTbZctz+dwbq71Szn3u9thEAsK37UE69GVfP4MaNGy7to6OjYbr1C/tZFTkAZrOZ/Wt0XCnCAv2gaLArpoq+ibBYLC4rXVRUFIy5azj9BM19jfNcx/6vXbvGKYuNjUVjbRVuVNVy+g8JCeHY3ACbAuvn58e70pbV200dA7vZzDnm8p857ZU9B6K6gauPhQcHwGq1ory83GP/zEol/+kQq8soR0znXWm1Wi369OkDb5FMR9JoNNDpdMjIyIBWq23WM/gWR7VazTGc8v27J51KoVA0W+dqaf9qtRpKpbLVdD4x/btTrMW8PzFzK6my7Y1LiBikeAlA6+1uZHKFqElkjMou7T0oxp6QyWSi+hdSrD21d8TT3EoqSEIuIc1BypfQWrsbpYoSNYl8OounlU4M7lZaqVdqT3MrqY40a9YsfPHFFygqKsKQIUPY8oin+A+xZlX+xPms09n0AL6XoIocwKkr9BJk/kFQhCW6HSczieq5rwn+u3P/zGGhIxaZQtQkOnsdMv9uMplYvcgRZc+BLmXhwa67Q6GVVkiIlCO4B8DerNRCc8vQJrs2y+UTCJDZfx6UiVMBAPq9mzh15fctgPHWVagCu8DUeAe02QRFZAyUSiXnBDY2Nta2w2uo5rSXqbtCp9OhpqbGpe4Pv542+1MKdOsagOjwAJdTXaZuvd4Mrd4Ak8WKqK5ql/6ZulVVVRwbolqthlwuh9ZoRYPeCLPFthJFhwawu1e5XA6KoiCXy6FQKFy+FwCcL6/n9BXfsyvvabnRaMSRU1xv0YmjY2GxWGA0GtmVkNHT+N5hQ0MD9x3KZL7dtTFnDZs2bUJqair8/Py8fgZFUTAcyYLRoSxg/huCdRXVJZwyRex4lx0XU/evm/M5ZRfX/UnwuV+csj/jxXvDBcdbW1vL+cwIwidnr3DKF40PQ1lZGW/dlkBRFB5+5f84ZbrjHwqaY/jaV1dz/xiFTCSe5paY5AmSIKkgufOzJnRsPM4tLREffvghrdfr2f8yaLVaWqvVStUNoZXhmy+huXWkTZRts9ns4lzOZx7wxuzRWjiOwZ3ZA2i5iaOlML5DYswufKYfvrremLQ4z/KqdjOpbdTDqOXuOqKjo2HO/y+nzBw3kdfsER4ezru7sV4+wWkvHzQOy/dwo1EyHrBtVQOSXuCUCymljIlEjNlBLwvA71+xR3nc+DJVsK+Xdl7glL37p2EAgE2nrrJlqXfb4sqOlXAV4An9w7H3UiWnbFpcdwDg9Ufi22E6m34Ys4nzO+BTtsUguWNbQ0MDFi1a5LauFGaP1kTsOdDJIle7YFvS0hN7xuDrCTFzK7ljG18krCNSmT1aC2/MDnrjb8cfydPcul2RDAYDvvzyS5w/fx4URWHo0KF47LHHPPopC30Z558rvna+dmrzzuzQMhOH1LTFSi80t26V7czMTEyePBnR0dEAgCtXruDYsWOYN2+eS12NRoOgoCDU19fjxRdfZMsZZZuiKI/+QCaTCWaz2eUvpS2VWj6lVCaTwWq1QqlUwmw2s4o3n59UW8K3MQD4NyZiFXMhxza+uXXErSCVlZW5SCAjVGJpbGyEpfIKAlT2X1FllE0BrnyXO6juL70veCprKbWfTCtiRgAA9Pv/zanrP/UpwfbTNnzPKd/7/HgY/18Op4y691EArg75Qp6bId0iUd5kf32JvWwRuXzKNl/wAV9fzmVM+S9VXFPGgF9tdHx1WxJowCjbku7adu7c6VK2cOFC3rqMQqZUKpGcnOxTF1SpaYk/U2dAzNy6FaSFCxfi5MmTOH36NAD7sifEnDlzmmVja+8I7Y4UcjkA8ZEYHRlPc+tx1/bFF1/Az88Pw4cP9/gX6ZxeprPA9+uvVqvhR/kuOLKt8TS3Hk+2ly1bhuTkZFy7dg3nzp3D22+/zVvPcfmbOnUqune3HZg196S0PcEXrcGcw/Clk+nICCnbfHPriEdBKi4uRmhoKL766itERkZi5syZLR5YR4Nvd8REpfjSRNIatJqJpKysDJ999hlomsbVq1cFBWnPnj2ora3F1atXsXTpUs5Zk+V2Gcz19vMNZS/bjkGb/RbnGUFzX0P+jTpO2YjetlPltw/Zdxd/nyy8uxHasfCZTvhMLM7PjY2NBUVRvGaHUa/v5ZSdXTkNANB7Dtdh78aXqfjoOLd9WlIMACC7wL7SzR1uS0nI59h25TZ319YvwrZr+/wc1//qLyOj8LPTDo/xwryz+XW2rMsTKwEI79oY3M2rIx4F6fTp01i5cqVH/2K+vI2Ejo/YefWobOv1ehw9ehTHjx/H8ePHBet5m7eR0DEQO68edaSvvvqKcygp9NOWnZ0NALhx4waee+45NndjZ9CRfks4z5fQvDrjVpA2bNiAK1fs/scymQxr1qwRqi5qYATf0Nr+SKId2xj7kreZ1xobG1FWVsbRsRgFuHr9K5y64QvfEVSWt/xod9l4LNGmEF6r4UZ9RocF4Xottywq1BbQd7qM66g/pk8orCVnOGXy/qMBiDdbzP/iLKfsP4+PAgAkvX2QU37871Nw4H9cf6L7Y21b6B9K7b5H98TYDNl8yvahy1WcssmDugn2xadsCyWcKKnVc+pGB9nmSfJd26uvvgqZTMb+vGVkZPDW279/P7Zt24bx48djypQpLknRCe0Hb/yRxM6rR0GKiYnBX//6V48dTp06FZcuXeq0ZpLOgidXHmfEzqtHQbp06RIWLVoEf3//ZulIhPZDqzoNeooq+OCDD0RHIPBFGZAokvaBTqejTSYTXVZWRl+6dIm+dOkS3djY6FLPXRSJOzwq2wsWLEBQUBDrOtCcXZv11mUEKO1HCIq+ttj8sr8/yakb+Y+PXUKImXBjRclJe/vBEwAAurx3OXUDkl8SVKBXH7nMKV82aRCavn6f29dMm3+U+aZdsVb2imUTgDmHQTc2NnKibRnFfOH2Qs5z1z+S0OJkpELt+YIKzNe4ORWU0UPZNNWOmyaVSgVV6WlOXX30SACtoGw/+OCDoi480Wg06NWrF4xGIyZMmNAsZdtsNruEEDO7C19TU1Pj8nMgRsdoL1gsFpdIGLFzJGZuPQrS3r170dTUxK5If/7znwXr3rx5kze/YHNx3F340jtayB+pIyWh5/vh8SY/k6e59fgmnn32Wdx1113o06ePx/u8nK+ragnNjXZoDfj6V6vVHdry721+Jk9z63FFunDhAs6fP4/JkyejqqoKY8aMEaybmJiIdevWuZTLIwdBwfOb2+ftTzmf3eVHYvQiRwKSX3Lt61edyJllk1zT+jI6kTOMd4IQjislX6aP9Y8kuJQJZQThKxdbBtgDLR1RRg/lqWnHcfyU83sV8IIVmlsGj8r2ypUrERgYiCVLliA9PR1vvvmm20E64xyy7S7aQa/Xw2KxtLtoDb7ssWazGQqFwqfj8gbmO3gaf6v5I8nlcpSWlmLLli0efbaFMOsaUNGgczkM02X+g1PP7y/pgtlfHcOYJ/S3Kbne7G403zvnLOoHyxWuiUPRz2bicLxaQtEnnjd7bFNTE0K6ReJitd3EwPhO8YVXC43rp3L7NRhDe9ruoy2u4F6NMbhHF/z7NDe/0lNj+gi+gxNXazhl4/qGwWq1oq6uziXPtnPd+IjmHSZ7FKTHH38cV69ehdVqRWpqqmC9w4cP4+eff4ZOp+ONfWrt7LOtja8jgFtKc4NPPc0rg6CyXVFRgaVLl+KFF17A7t27sWfPHvzzn/8UfBDfhcL2XsR9CaHdUUuzv7aU9hAB3FKa+0fqdl4dEFyRevTogTVr1uDcuXMYOXIk6uvr3V7JKXShMAAola4/V3wI7Y5amv21pbTnlVIszU044W5eHRHtRvLSSy/h3XffFfz37777DgUFBZDJZHjhBXukqTf5kRyVWoB7LUJ7UbYd8fUmwBv4vgPf+J2VbaF5dUa0c7WnawUmTpyI0NBQ7Nq1y2X1unnzJq8/kvnH/ZxnWAfdy56+Oivmdy22R/0yiUSbG67sTd3Y2FhW2XZErVaDVndF6qd2hX3v8+MF+3r36C+cspd+N4C3r9b6Xs7fQWx+JHfz6ojbA8nS0lJ8++23AGwS6hxJ4UhhYSG+/fZbWK3WFl/93d4VW2Z81Vq958rtEG/uaxM7r25XpNWrV7OObMnJyVixYgU+/PBD3rqHDh3Ciy++aDvkasGJb0cRovLychjMHS9axlt/JLHz6nZFCgsLQ2hoKPv/nqJKaZpmreTNob3vjtq7kHuiueMXM69ule0333wTU6ZMQWJiIgoKCrBr1y5BV9vi4mIcPHgQer0ezz33HHv5ibf3tTnnRATax8V/zpfmtYdxeQNfah4xF/8JzaszbgXJaDQiOzubvYJy9uzZXrvRMrs2tVrtMcmTu92RmEvrLBYLmxTLEb5L77y9+E8mk3HMNswzaJpuVv+O38FTe8d/b257uVzuMn4xuzaxuBWk3NxcTJ8+nX2oVqvFf//7X17/JEYpq6urw7Jly9hJamxshF6vh06nczF7GLNWcp5BpbzucgWEWq1G9+7dUVlZyWnPrA7Oy3W3HpGoqrjlsb2726idoy342gNAePce0DXc4fTP/L+Y9j179nQxW/C179mzJ3RNetwqv9ns9gaDweVSRb534JxoS2henXGrLU6aNAmbNm1it+/du3fHU089xVv34MGDWLx4MWpra138dEwmU4uyrxoMhmZdJeWpfUv7V8jlLepfLrJ9eXk5gru4bru9ae+cmRcQZzFwN6+OuBWk8PBwLF682GNnzKCqqqrw2Wef4dFHH+WkCORLdu5N9lW+l9DaL9FTe7VaDX8/cfe1CbV3d7Ok8x+JsyB5254vxbMYi4G7eXVEsv3rtGnTsHXrVgCunoN+fn6cKyyZLxGSylXchfyR+F6C0EtUymWiX6KQj4/YPNtiV8qWtgcAhVzWovZ8/87nT+Xs4eFuXh1xqyNVVlYiPz8fkyZNAkVROH36tFvHNj6k8EdilFrH9kLXOoi9TZqvf4A/+yujaDPPYsbH3J7dnP6lUMwBuGwM+DYmjt/BJ/5Ib731Fv70pz8hPT0dr732GrZt2+a1IAHCJhLn4/no6GhBf6Txb3/Hfs5/23YT4otfn+e0f39mPCcEGrCHQY/+xz5O+Zk3/oBdF7mZZmfcFQmKovDtL7fZst8PiIBOp+NV7K9U6jn3pemO2w5rw/7AvWOuZt8/ePMYOX+H92fGAwASl+3h1P1x9QO8kSnMuBgYswdfLiedTsfrj/Tafy9y6i7/nfc3bAMeBCk4OBhTpkzByJEjsWbNGhgMBsG6TC7mkpISLF++vFnnK+3Zyt7eDyM9hWG35LBXzNy6PdkeN24cjhw5grCwMKSmprqECjkzd+5cxMfHuxg4xUL8kZqHmECJlv6ReppbtyvSH//4R/b/jx8/jqysLLedbd26FQqFAgMGDHBbTwjij+Q9rX1lPYOnuRXtj7RkyRK89957gv/O3HvqfPLtjfLWXrPHNjU14dq1ay7l0dHRPh+XGLMH4PpuAf7xC2W19XRPsWhBKi4uxuDBg8VUdRmYUH6kpp0aTl3FtGfZ/EZKhRyBfhQClLZdynPb7aHJn86xhRVfcnKSj+vRhaMoAzZlGQDuyTjEKf9h+WSU13G3uj1DbC/P2UdIr9dDq+fqh2qVEg0mGimf26+2OLhogkt75hmOyVQBe0JVx7xHTM4jPkf/Of/hhlZ/OX8MTCYTfiyrhsFkdwcZ078b3jzE9X16a/pdMBqNnGsoYsICoFQqcaGS+w5iQ2y761a7+E+MEC1duhSzZ8/G6NH8sWWeoCgK313jJsp6YnTzdhFSYrVaUVNp91dWq9UI6dkTlypr3LRqfVQqFV7Yyo3z/2E5/8V9FEUhr9iu4/59svhwczHzKlnMcUVFBcaMGYNz585J9ch2CdcfqfNfHyF2Xj2uSG+88QYaGxvZnyYhN5J9+/bBbDbj0qVLMBqNHTqcWQgXxVbGn5izMyF6Xj3l1dmwYYOnKjRN0/Q777xD0zRNFxQU0Pv372fLO0N+pMbGRrqsrExUfqGOjvN8Cc2rMx6V7eTkZERFRbFH6b/FrLZ6vZ7XH6kjRZHwXarY5lltL168CKvViiFDhnh9OOhu16bLfYdTN+DPrwheIeGY6TW+p80S7k20xYi/c2/0zn97umBd552UkOdm14hIlGkddkx9bG7JfBf/8d2cDQBnr9u/76ioEMHvJXTxn2N75hmXK7l1B3UP5j3CiIqKwuU67kn4wK42bcdbQfKobL/11lv44YcfcPz4cbzzzjuC9Rjd6ejRo/juu+8E63VEhPMj+fbE3Rv41gsxi4LYeRW1a3vqqafw9NNPu81TFBUVhfLycpw+fRrjx48X89gOA9/3VqvV8KdaFnblS8RaDMTOq8ddW0hICFasWAGZTOY2fOWBBx7AV199BYqiOv2lNszurVbb/kwmYvAmz7bYefXoj6TX66HX60HTNAICAtxejvzyyy9j3rx5SEiwJ5rqDMp2Z7ivjfkOnsbPN1988+qMW0Havn27i7VX6HJkIdwp2zUbX+XUDfvrP7H5DNc8wJxs84U2812VUFnPPfLv3tX2Qp7JyeeUf/LoCJhucU0JqkibQdLR9BLXo4tL/8wY+PIYAcCYFdxQ9NMrpgrmN3LMpTQtznatRIXTd+jRNZCTRwmw51Ka+X/cdHxfPz1OcBPhuLkJ+PMrvN+rubdsC+pI9fX16NWrF+rq6jBq1CiMGjUKJSUlgg+6fPkynnnmGfY2HULnQczcCv7oqdVqVFRUsFtfwHbTshCDBg1CQkICHnnkkRYMmdAeETO3goJEURRMJhPCwsJY182rV682y9WW0PlxqyPdvn0bRqMRNE2ziZZ69erFW/fy5ctYs2YNJk2ahLlz57LlnUHZ/i3BN19Cc+uIx5Pt9957DzNmzMDbb7+NgQMH4vXXX3dXXdTAOhp8YdwymQxKpbLFKXzaG62W1Van0+HUqVNIS0vDvn37BOvl5eUhMTGR1xVT0ESydTWnXsDsZdhayA1Lnp1gWwH5dlI/3uRekJfYq6vgjuXjE6Wc8ufGxcBYzfUYpMJtkR2bTl1ly1Lv7guz2czruVle14hvy+w7RyYyhO8yvv+c5Zon5o+yHaOITbQltGvb8RM3EuahoZGiHfYA7oWKAPDHga7BnO7mlsHjyXaXLl1w4cIFBAUF+dSttL3g6Epi5omr+63iUZBSUlIwePBgnDp1Ck8//XRbjKnd0t5DknyJRx1p6dKl+Mtf/gKaprFt2zasXLmSt15eXh7KysoQERGBWbNmsY7inUFH6gz5kcTCN19Cc+uIRx1JpVIhPt4WAZqbmytYLzk52etBdxTkcjm6d+8Og8GAkJAQNpbf17cjiTV7tBQxc+tRkCZNmoRXXnkFNE1j2rRpzRqIkLJtLv+ZO5ieA1t8QZ5QaPTwv3H9kQpWCfsjOfcll8t5/ZF0fiH4+FQx+/njWcMBAOP/eZhT7/tX7xPMautoOmHMJnyKNZ+JhbnfzjkMmy9km+97AXAJcU/s1jxHPY9/UgMHDkRERAQiIyPdRpKkp6fj+PHjeP75513ipzo6Qv5IlI8DN9siAljsvHoUpJycHKSlpeGJJ57Ahg0bBOt169YNZ86c8fndaq2BkD9SSKBvbxNviwhgsfMqynGopqYGVqsVNE3j2rVrgq4k0dHRCAsL8360HQxm9/a/W/WeK7ciLQ3DFouYefW4a1u/fr1LGZ8riUajwaJFi5CXl4e7776bvfe0s+zamGXdOU+1r7PtignD9gbn+RKaV2dEh2y3ZGB1dXXw9/eHXC4HRVGQy+VQKBRQON3wrIgdL6gsH/if3W/n/lib3443zv/TNnzPKd/7/HiYbxRzypS9B8NkMuFWvY49bOwR5AeZTAaj0ciOn0nGVaO3Yt33diX4/ZnxMJlMOHXxJvRGu0ll/LDeyL9WCz+lAuHB/qhu0GNwjyDI5XI06O25q4P8VLZMdHe44eF9woN474DT6/VodAglV8jlUPtR0Ol0nGgXtVoNuVyOSq0egX4UGg1GRATY5qHR4HD3i0wGOW1LdNZqIdue0Gg0CA4OhlKpxJw5czi7NJlMBp1O57K78K2qyo/ZbEZTjT08+3qdLXvsnTt3XBRbvX8ob/tu/gbAYfNjMpkQ5W9Gz57dUF5ejiBTE65ftyW9qq20mTjUajVCg8S5vzJYrVZU/xpK7pgOsKqqyqVu79690SskCOXl5dA3NeF6je17VVdw734pKytzudXc3dwySOpcPWfOHN7DKuYcpiOcCPMt0J4Sn7a0vbfXOrhr39LEp0LJ1ITmlkHSE7WsrCzeHErOqVc6Ep4moaXtW2p2EZt41Gw2t6h/obllkHRFSklJ4ZVamUzGn6c61jW8hdGJnGH0Ike8uY2auQLLEWVv9xlWPGWPHdQtiM372Jz2QitVn3DXaxoYf25HPJ0jtTTFtCNCc8uOpS2UbbEX/7UH+MwOQtcy8PkjedOe71oIb94LXxg2X/ZbmqY5mXXd9d/U1AR/f3/fKdvuEMpqq9+7iVPPf1qq4K7LdOIrtkw17mEAgDmfa/ZQjpjOqedYl++5pjO7uHVHzwBFURwTgztzjLMvz2OJvUFRFH635iin/MRrUwQTbbXk4j+VSoUrV66IqgsApUvnsWUxaz7j7cv54j+xSKYjrVu3DgBw69YtfPHFF1I9luBjxM6rZCuSUqlEZmYm+vTpg379+kkhjgLpAAAN1ElEQVT1WIKPETuvkq1IKpUKWq0WO3bsIILUiRA7r5Ip259++imio6ORlZWFzMxMNtNFZzCR8F0rAfh+w+CNP1Jz8yMJzaszbbJrE/JHMlVwFUVVj36CiiKficRSyg3DVsSMEMxDxHfdRGk1N/FpzK/bbsccTSN6hwjmR6q0BGLJtiL286n0+wEA497iOv+feG2KYFCDY9j5wF9zHn10vJRTNy0phjfnEZ/nptA9dEL5kTzd1yYWyX7a1q9fj7q6OmRmZuLq1aueG3QghKzsfqqO44/U3PxIYudVMkGaP38+1q9fD5qm0bdvX6ke2y4Q8keKCPZtVE1L/JHE5kcSO6+SCVJwcDACAwN5ryntbDAnxrcbfGvyaa4/EjN+5qoud4idV0l1JL6rBhoaGmA0GjtM0k4+9Ho9KittOpqfnx8iIiJw+/ZtdO3a1affy3FcDD169OA1ZTB1PY1fr9eDoigEB9tTP0t6hURzMRqNMBgMPo+4IIjDarXCz8/P691oqwsS4bcBWSYIkkAEiSAJihUrVqxoi440Gg0uXryI3NxcjB07FhRFQaPRID4+HmvXrsWIESPY32WNRoPhw4dj48aNGD58OJRKJdv+2LFjGDt2rMtzL1y4gIEDB0KlUmHPnj04deoUdu3ahaSkJMjlcuzfvx+rV69GTU0NIiIi0KVLF7b99evX8eOPPyI0NBRdunTB4cOHceDAARw7dgzjxo0DABQWFiInJwcHDhzA6NGjWfcRpnzfvn1ISkqCQqGARqNBUVER8vLyMG7cOLbu0qVLERoayskxlZGRgQkTJuDo0aMoLS1FTEwMAFtOomXLlkGr1bKRznl5eaAoihPRkZ6eDn9/f2RkZCAxMZHzvZj3MnToUMjlcqxbtw5JSUm4desWduzYgYSEBKxfv97mybBlC0JDQ3n9lsTQpivSnDlzkJiYiJs37ae87733HlJSUhAU5OrMxde+b9++KC4udilPSUlhIzru3LkDo9GIBQsWsCl9p06dioSEBMyZM8clEsJxPACQn5+P1NRUhIeHs2kPDx06hEWLFmHp0qWcyJGDBw9i0aJFWLhwIWdD4Xx1p9AtQ0J5rMWmUnQXd8a8F+a8iDHAFhUVsXYzqc7/2lSQMjMzUVJSgkGDBrFlMTExOHLkiEvdrKwsFBQU8D7H+UR227Zt2L59O/s5OTkZ06dPx7/+9S8XIeFj7NixOHHCnh2WcQBjstQx0DQNk8nE2t6YsVRVVeHTTz/lZADeunUrzGYzm1No3759aGhoYG8ZYnjggQewe/fuFuUnj46Oxr333utS7uwey2eAler8r00FacGCBVCpVLh9237L46xZs2A2m1FUVMSpm5KSguHDh3PKsrOzUVJSgri4OE4544/MnKnk5ubi8OHDCAoK4pyHCJGYmMg5j7n77rvxwQcfoL6+HhERthsop0+fjg0bNmDjxo0cQZg2bRq2bt0KgJtUYvbs2Zw0eZWVlXjyyScxb948zlUM4eHhKC4uxsSJEzljunz5MgoLCzl/IADwzTffICsri+OkP3PmTN5YtpSUFKSkpLCfKYpCXFwc6urqEBkZyZZLkXmObP8FKCwsxK5du5CWloauXbv6ejjtHiJIBEkg23+CJBBBIkgCESSCJHR6QTIYDKirq/NcsZVgzqE6O51WkKqqqrB161bs2bMHRUVFyM3Nbfak7ty5Ex988AGn7Pbt29i2bRsMBgPeeOMNLF++HKmpqS4X/6xdu7ZZfX700Uce62g0Gly9ehWbN2/2eTh8h7qhT6PRsM7pjz76KH7++WccOXIERqMRf/vb3/D555+joaEB/v7+eOyxx3Djxg1UV1eDoiiEhoZCr9dj/fr1qK2tBU3TSE9PR3JyMiZPnoyqqirwWYvy8/NRVlbmEin7+eef45lnnsGXX36JyZMn495770VNTQ0KCgrQv39/AMDZs2dRXFyMkpISZGdnw2q1onfv3hg2bBg2bdqEP/zhD9i9ezdGjBiBK1euIC4uDleuXME777yDK1euIC8vD+fOnYO/vz/GjBmDvn37YvPmzWhsbOScD913333IycnBggULWu3de6LDrUiPPfYY0tLSkJOTg71792LVqlWYN28edu7ciRs3bqBPnz6YMWMGW3/kyJGYMmUKALtz1+uvv46ePXuiqKgIvXr1QlpaGrRaLW9/I0aMwMyZM13Kr1+/jqCgIJSUlLAHp2FhYbjvvvvYOqNGjUJcXBzKyspQUVGBgIAAFBYWAgAmTJiA2bNnIzAwEC+++CLMZjOef/55l/DuSZMm4dVXX8W3336Lrl274uGHH0ZCQgLOnDnD1unbty/On+cGN7Q1HU6QaJqGwWBw8Te2Wq2YN28e4uLisGrVKnYyHM0pNE1zPlutVtZD0NE8UVxcjMWLFyMnJ0dwHEz/AwYMYE05NTU1+OSTT/D+++9j8eLFrG5msViQlJSEl19+mRVqJkqDsdsxJ9POMfpqtRoqlQo0TePrr7/G9evXcdddd7nY1nx9/WuH+mkDgE2bbPkCnn/+eVy8eBHp6emwWCx46aWX8O9//xs6nQ4DBgxgzRW9evXCli1b0L9/f5vDfkQEVq9eDZlMhiFDhvD2MXjwYDZUWYju3bujoaEBjz76KFasWIGdO3dCp9NhyZIl7E8bYNOlevfujd27d7PG0u7dXTOLiCEyMhKnTp2CSqWCWq1mDd0lJSUYOnRos54pGXQH4sMPP6RLS0t9PQyapmm6traWzsnJ8fUwaJqm6c2bN9MGg8GnYyAmEoIkdDgdidA+IYJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJIIJEkAQiSARJUPqq49D700EFhYIKCoUqIBDqIH/4B6jgF6CCf4AKXQIphAVSiAj2Q1gghRC1Cl38lejip0RXPxWC/BQIUMqhVsrhr5TBXymHH8yQGRshMzRCbtRBZmiEVVsLa0MdrNp6WLV1MDc2wlDXCH3tHRjrtDDUN0Jfp4e+zgB9nR66Gj3u6M2oNVpQa7KizmTBWlOJr15Th4GsSARJIIJEkAQiSARJkNE0Tfui46qqKnTr1q1DPLe1xtqZICsSQRJ8tiIROhdkRSJIAhEkgiQQQSJIgk9Otm/cuIFvvvkG1dXVWLZsGRQKRYufeejQIdy6dQsmkwkLFixo+SB/Zffu3airqwMAPP7445I9t7PRpitSdnY2Fi9ejLVr12LkyJGoq6uDVLr+uXPn8Pjjj6OmpkaS5zE8+OCDCAwMRFNTk6TP7Wy06Yo0d+5czJ07Fzdv3kRoaCiqqqrw008/ITExscXPViptX8Vqtbb4WY788ssvmDlzJlatWiXpczsbPvlpa2pqwmeffQaZTIbf//73kjxz2LBhyMzMlPzg8OTJkzh8+DCio6MlfW5ng5wjESSB7NoIkkAEiSAJRJAIkkAEiSAJRJAIkkAEiSAJRJB+xWAwsKYQgve0uSDl5eWhoKBAdH2apvHxxx+7rXPy5Els27ZN8LMY9uzZg6KiIuTm5qKiooLzbwUFBfj++++9et5vDZ+cbJtMJrz11ltQqVS45557IJPJcOzYMZw8eRJPPPEEHnroIbbukSNHMGzYMABAfn4+srOzYbVakZaWhpqaGuTl5UGr1WLChAk4c+YM+3nYsGFIT09HY2MjMjIyQFEUAECj0eDGjRuIiorCzZs3UVNTg5dffhlnz54FRVEIDQ2FXq936evrr7/G+PHjffG6OgQ+EaSPPvoIGzZsQEBAANLT0xEcHIxnn30WPXr0wJAhQ3D8+HFcuHAB99xzD3744QekpaUBAHJycrBq1SrcuXMHGzduRGNjI1auXInvv/8elZWV2L59O/v58uXLCAkJwYwZM1y8C+bPnw+KolBdXY1Dhw7h/PnzGDlyJCIjI3H27Fnevpj64eHhbf6+OgI+0ZEsFgvH6j9jxgysXbsWcrkcAwcORFJSEu6//35otVoAdoOsYxuapiGTyQCAFRTHzyqVCg8++CC2bNmCS5cucfoPCAhAdnY26uvrERcXx3mW4/Md/18ul7vUIdjxyYq0ZMkSvP/++1AqlZg8eTLq6+thNpvR0NCAW7duITw8HEePHsW8efOg1Wpx7tw5TJw4EbNnz0Z6ejpkMhmeffZZ1NTUYMWKFWhqasK4ceMwa9Ys9nO/fv2wefNmyOVyREZGuowhIiICJ06cQF1dHeLj43HXXXdhy5Yt6N+/PwC49JWZmYmwsLC2flUdhnZhtF2xYgVeeeUVNDU14eOPP4ZCoUD37t0xcuRIjBgxAp988gmeeeYZn40vPz8fFosFo0eP9tkY2jvtQpA0Gg2USiV0Oh0mTpxIJqwD0i4EidDxIQeSBEkggkSQBCJIBEkggkSQBCJIBEkggkSQBCJIBEn4/3Q4exudlMuoAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 108x180 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plot_seq_pair_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus,\n",
    "                                    statistical_test, plotTitle=None, subseq_type='aa', refCodonTableDf=refCodonTableDf,\n",
    "                                    width=1.5)\n",
    "statistical_test = 'contingency'\n",
    "filenamePlot = os.path.join(plotsFolder, speciesSetName, outputFilePrefix +\n",
    "                            \"_composition_bias_aa_pairs_\" + terminus + \"terminal\" + '_statistical_test_' +\n",
    "                            statistical_test)\n",
    "fig.savefig(filenamePlot + '.svg', dpi=600, bbox_inches=\"tight\")\n",
    "fig.savefig(filenamePlot + '.png', dpi=600, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot codon pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subseqOddsRatioDf = oddsRatioDfTableHexamer.copy()\n",
    "maskData = subseqOddsRatioDf.xs('BH_multiple_tests', level='observable')\n",
    "maskData = maskData.xs(terminus, level='terminus')\n",
    "maskData = maskData.xs(statistical_test, level='statistical_test')\n",
    "missingIndex = subseqOddsRatioDf.index.get_level_values(0).unique().difference(maskData.index)\n",
    "maskData = maskData.append(\n",
    "    [pd.Series(np.nan, index=subseqOddsRatioDf.columns, name=missingIndexName)\n",
    "     for missingIndexName in missingIndex])\n",
    "maskData = maskData.astype(bool)\n",
    "maskData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = compute_subseq_oddsratio_plot_data(oddsRatioDfTableHexamer, terminus, statistical_test)\n",
    "maskData = compute_subseq_oddsratio_mask_data(oddsRatioDfTableHexamer, pvalueThresholds,\n",
    "                                              terminus, statistical_test, subseq_type='codon')\n",
    "pvalAnnotTable = compute_subseq_pvalueAnnotation(oddsRatioDfTableHexamer, pvalueThresholds,\n",
    "                                                 terminus, statistical_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskDataSmallN = compute_subseq_smallN_mask_data(oddsRatioDfTableHexamer, pvalueThresholds,\n",
    "                                                 terminus, statistical_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for plotSynonymousCodonGroupSeparationLine in [True, False][:]:\n",
    "    fig = plot_seq_pair_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus,\n",
    "                                        statistical_test, plotTitle, subseq_type='codon',\n",
    "                                        plotSynonymousCodonGroupSeparationLine=plotSynonymousCodonGroupSeparationLine,\n",
    "                                        refCodonTableDf=refCodonTableDf);\n",
    "    if plotSynonymousCodonGroupSeparationLine:\n",
    "        suffix2 = '_withSynGroupLines'\n",
    "    else:\n",
    "        suffix2 = ''\n",
    "    filenamePlot = os.path.join(plotsFolder, speciesSetName, outputFilePrefix +\n",
    "                                \"_composition_bias_hexamer_\" + terminus + \"terminal\" + '_statistical_test_' +\n",
    "                                statistical_test + suffix2 + \".png\")\n",
    "    print(filenamePlot)\n",
    "    fig.savefig(filenamePlot, dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## multiple test correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha_family_wise_error_rate = 0.05\n",
    "reject, pvals_corrected, alphacSidak, alphacBonf = \\\n",
    "    multipletests(pvalueList, alpha=alpha_family_wise_error_rate, method='fdr_bh',\n",
    "                  is_sorted=False, returnsorted=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## COG category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "COGList = ['U', 'J', 'Y', 'P', 'T', 'A', 'M', 'F', 'W', 'V', 'K', 'G', 'D', 'H', 'E', 'Q', 'B', 'Z', 'N', 'S', 'I', 'O', 'L', 'R', 'C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggResDf3.rename(columns={'species_name_x':'species_name'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggResDf3.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggResDf3.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonRank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select genomes in the taxon group\n",
    "eggResDf4 = eggResDf3\n",
    "eggResDf4 = eggResDf4[eggResDf4[taxonRank + '_taxid'] == taxonId]\n",
    "taxonAssemblyAccessionList = list(eggResDf4['genome_accession'].unique())\n",
    "len(taxonAssemblyAccessionList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(eggResDf3))\n",
    "print(len(eggResDf4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groupIndex = 'J'\n",
    "groupCDSDf = eggResDf4[eggResDf4['COG cat'] == groupIndex]\n",
    "print(len(groupCDSDf))\n",
    "sel = eggResDf4['COG cat'].str.contains(groupIndex) == True\n",
    "groupCDSDf = eggResDf4[sel]\n",
    "groupComplementCDSDf = eggResDf4[~sel]\n",
    "print(len(groupCDSDf))\n",
    "print(len(groupComplementCDSDf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groupCDSDf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "refSeqPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "analysisCtermDataPath / 'Species'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "analysisEggnogPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonSuffix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "outputPath = analysisEggnogPath / 'Analysis_COG_groups_cterminal_bias' / 'COG_{}'.format(groupIndex)\n",
    "\n",
    "full_analysis_multispecies(refSeqFolder=str(refSeqPath),\n",
    "                           dataSingleSpeciesFolder=str(analysisCtermDataPath / 'Species'),\n",
    "                           dataSpeciesGroupFolder=str(outputPath),\n",
    "                           plotsFolder=str(outputPath),\n",
    "                           speciesSetName=taxonSuffix,\n",
    "                           chooseRandomGenomes=False,\n",
    "                           nRandomGenomes=0,\n",
    "                           assemblyAccessionList=[],\n",
    "                           speciesList=[],\n",
    "                           multispeciesAllCDSDf_nr=groupCDSDf,\n",
    "                           dfFormatOutput='csv',\n",
    "                           skipAnalysisExistingDataFiles=True,\n",
    "                           skipStatisticsAnalysis=False,\n",
    "                           method='allSeqDataframe',\n",
    "                           skipClustering=False,\n",
    "                           codonAnalysis=True\n",
    "                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "outputPath = analysisEggnogPath / 'Analysis_COG_groups_cterminal_bias' / 'COG_{}_complement'.format(groupIndex)\n",
    "\n",
    "full_analysis_multispecies(refSeqFolder=str(refSeqPath),\n",
    "                           dataSingleSpeciesFolder=str(analysisCtermDataPath / 'Species'),\n",
    "                           dataSpeciesGroupFolder=str(outputPath),\n",
    "                           plotsFolder=str(outputPath),\n",
    "                           speciesSetName=taxonSuffix,\n",
    "                           chooseRandomGenomes=False,\n",
    "                           nRandomGenomes=0,\n",
    "                           assemblyAccessionList=[],\n",
    "                           speciesList=[],\n",
    "                           multispeciesAllCDSDf_nr=groupComplementCDSDf,\n",
    "                           dfFormatOutput='csv',\n",
    "                           skipAnalysisExistingDataFiles=True,\n",
    "                           skipStatisticsAnalysis=False,\n",
    "                           method='allSeqDataframe',\n",
    "                           skipClustering=False,\n",
    "                           codonAnalysis=True\n",
    "                          )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stop codon context"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compute analysis downsampled data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multispeciesAllCDSDf_nr_downsampled = multispeciesAllCDSDf_nr.sample(frac=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(multispeciesAllCDSDf_nr_downsampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonSuffix + '_downsampled0.5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "outputPath = analysisCtermDataPath / 'Taxonomy'\n",
    "speciesSetName\n",
    "\n",
    "full_analysis_multispecies(refSeqFolder=str(refSeqPath),\n",
    "                           dataSingleSpeciesFolder=str(analysisCtermDataPath / 'Species'),\n",
    "                           dataSpeciesGroupFolder=str(outputPath),\n",
    "                           plotsFolder=str(outputPath),\n",
    "                           speciesSetName=(taxonSuffix + '_downsampled0.5'),\n",
    "                           chooseRandomGenomes=False,\n",
    "                           nRandomGenomes=0,\n",
    "                           assemblyAccessionList=[],\n",
    "                           speciesList=[],\n",
    "                           multispeciesAllCDSDf_nr=multispeciesAllCDSDf_nr_downsampled,\n",
    "                           dfFormatOutput='csv',\n",
    "                           skipAnalysisExistingDataFiles=True,\n",
    "                           skipStatisticsAnalysis=False,\n",
    "                           method='allSeqDataframe',\n",
    "                           skipClustering=False,\n",
    "                           codonAnalysis=True\n",
    "                          )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import all bias data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/users/lserrano/mweber/Analysis_data/C-terminal/Data')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysisCtermDataPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_index</th>\n",
       "      <th>rank_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_genomes_in_refseq</th>\n",
       "      <th>n_genomes_in_refseq_eggnog_common_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>phylum</td>\n",
       "      <td>1117.0</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>62</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rank_index rank_name  taxon_id     taxon_name  n_genomes_in_refseq  \\\n",
       "3           1    phylum    1117.0  Cyanobacteria                   62   \n",
       "\n",
       "   n_genomes_in_refseq_eggnog_common_species  \n",
       "3                                         31  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxaDf[taxaDf['taxon_id'] == 1117]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "oddsRatioAATaxonList = []\n",
    "oddsRatioCodonTaxonList = []\n",
    "# oddsRatioAAPairTaxonList = []\n",
    "\n",
    "subsetList = ['all', 'TAA', 'TAG', 'TGA']\n",
    "\n",
    "i = 0\n",
    "# for index, taxonGroup in taxaDf[:].iterrows():\n",
    "for index, taxonGroup in taxaDf[taxaDf['taxon_id'] == 1117].iterrows():\n",
    "    i += 1\n",
    "    taxonRank = taxonGroup['rank_name']\n",
    "    taxonName = taxonGroup['taxon_name']\n",
    "    taxonId = int(taxonGroup['taxon_id'])\n",
    "    taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "    print(i, \"/\", len(taxaDf), taxonSuffix)\n",
    "\n",
    "    taxonAssemblyDf = assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf[taxonRank + '_taxid'] == taxonId]\n",
    "    # Read nb of genomes for taxon\n",
    "    taxonNGenomes = len(taxonAssemblyDf)\n",
    "\n",
    "    taxonDirPath = analysisCtermDataPath / 'Taxonomy' / taxonSuffix\n",
    "    if not taxonDirPath.is_dir():\n",
    "        raise ValueError(\"taxon folder \\\"{}\\\" does not exist.\".format(taxonDirPath))\n",
    "\n",
    "        \n",
    "    for stopCodon in subsetList:\n",
    "        # Read total nb of sequences from statistics summary file for taxon\n",
    "        taxonNSeq = None\n",
    "        if stopCodon == 'all':\n",
    "            subsetSuffix = '_subset_all_'\n",
    "        else:\n",
    "            subsetSuffix = '_subset_stopCodon_{}_'.format(stopCodon)\n",
    "        multispeciesStatisticsSummaryDfPath = taxonDirPath / (taxonSuffix + subsetSuffix + 'multispeciesStatisticsSummary.csv')\n",
    "        if multispeciesStatisticsSummaryDfPath.is_file():\n",
    "            summaryDf = pd.read_csv(multispeciesStatisticsSummaryDfPath)\n",
    "            if not summaryDf.empty:\n",
    "                taxonNSeq = summaryDf['nSeq'].sum()\n",
    "\n",
    "        # Import AA odds ratio dataframe for taxon\n",
    "        oddsRatioDfPath = taxonDirPath / (taxonSuffix + subsetSuffix + 'oddsRatioAADf.csv')\n",
    "        if oddsRatioDfPath.is_file():\n",
    "            oddsRatioDf = pd.read_csv(oddsRatioDfPath)\n",
    "            oddsRatioDf2 = oddsRatioDf#.copy()\n",
    "            oddsRatioDf2['taxaid'] = taxonId\n",
    "            oddsRatioDf2['taxon_name'] = taxonName\n",
    "            oddsRatioDf2['taxon_rank'] = taxonRank\n",
    "            oddsRatioDf2['n_seq'] = taxonNSeq\n",
    "            oddsRatioDf2['n_genomes'] = taxonNGenomes\n",
    "            oddsRatioDf2['stop_codon'] = stopCodon\n",
    "            oddsRatioAATaxonList.append(oddsRatioDf2)\n",
    "        else:\n",
    "            print(\"ERROR: oddsRatioDfPath file not found\", oddsRatioDfPath.name)\n",
    "\n",
    "        # Import codon odds ratio dataframe for taxon\n",
    "        oddsRatioDfPath = taxonDirPath / (taxonSuffix + subsetSuffix + 'oddsRatioDfCodon.csv')\n",
    "        if oddsRatioDfPath.is_file():\n",
    "            oddsRatioDf = pd.read_csv(oddsRatioDfPath)\n",
    "            oddsRatioDf2 = oddsRatioDf#.copy()\n",
    "            oddsRatioDf2['taxaid'] = taxonId\n",
    "            oddsRatioDf2['taxon_name'] = taxonName\n",
    "            oddsRatioDf2['taxon_rank'] = taxonRank\n",
    "            oddsRatioDf2['n_seq'] = taxonNSeq\n",
    "            oddsRatioDf2['n_genomes'] = taxonNGenomes\n",
    "            oddsRatioDf2['stop_codon'] = stopCodon\n",
    "            oddsRatioCodonTaxonList.append(oddsRatioDf2)\n",
    "        else:\n",
    "            print(\"ERROR: oddsRatioDfPath file not found\", oddsRatioDfPath.name)\n",
    "\n",
    "\n",
    "oddsRatioAATaxon = pd.concat(oddsRatioAATaxonList, sort=True)\n",
    "oddsRatioAATaxon.set_index(['terminus', 'position from terminus', 'observable'], inplace=True)\n",
    "oddsRatioCodonTaxon = pd.concat(oddsRatioCodonTaxonList, sort=True)\n",
    "oddsRatioCodonTaxon.set_index(['terminus', 'position from terminus', 'observable'], inplace=True)\n",
    "oddsRatioAATaxon.sort_index(inplace=True)\n",
    "oddsRatioCodonTaxon.sort_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioCodonTaxonList[0]\n",
    "df[df['position from terminus'] == 1000000].iloc[:, 50:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(oddsRatioAATaxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioAATaxonList[0]\n",
    "df[df['position from terminus'] == 1000000].iloc[:, 50:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For some reason there is a missing `BH_multiple_tests` values for some phyla. for example, taxid_1117. The `terminus` value is corrupted. In these csv files, the terminus value is present as the last column. Maybe because of older results (they date from November 27, 2018). However, the `terminus` value is missing for the `BH_multiple_tests` row. This is the case in the CSV file (checked).\n",
    "\n",
    "possible solution: reassign `terminus` based on the `position from terminus`.\n",
    "\n",
    "Below, I remap the terminus values, and write again the full table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {x:'C' for x in list(range(-20,0))}\n",
    "d.update({x:'N' for x in list(range(1,21))})\n",
    "d.update({-1000000:'N', 0:'bulk', 1000000:'C'})\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon = oddsRatioCodonTaxon.reset_index()\n",
    "oddsRatioCodonTaxon['terminus'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(oddsRatioCodonTaxon))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon['terminus'] = oddsRatioCodonTaxon['position from terminus'].map(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon.set_index(['terminus', 'position from terminus', 'observable'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon['observable'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# drop rows that have NAN in every column for all codons.\n",
    "oddsRatioCodonTaxon = oddsRatioCodonTaxon.dropna(subset=list(refCodonTableDf.index), how='all')\n",
    "print(len(oddsRatioCodonTaxon))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df = oddsRatioCodonTaxon\n",
    "df = df[df['taxaid'] == 1224]\n",
    "df = df[df['position from terminus'] == -1]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    oddsRatioAATaxon.to_csv(analysisCtermDataPath / 'oddsRatioAATaxon.csv')\n",
    "    oddsRatioCodonTaxon.to_csv(analysisCtermDataPath / 'oddsRatioCodonTaxon.csv')\n",
    "#     oddsRatioAAPairTaxon.to_csv(analysisCtermDataPath / 'Taxonomy' / 'oddsRatioAAPairTaxon.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon = pd.read_csv(analysisCtermDataPath / 'oddsRatioCodonTaxon.csv', nrows=None, index_col=[0,1,2])\n",
    "oddsRatioCodonTaxon.sort_index(inplace=True)\n",
    "print(len(oddsRatioCodonTaxon))\n",
    "oddsRatioCodonTaxon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon.loc['C'].loc[-1];"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing a 2G csv file takes ages. We will store it in HDF5 format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.HDFStore(str(analysisCtermDataPath / 'oddsRatioCodonTaxon.h5'), mode='w') as store:\n",
    "    store['oddsRatioCodonTaxon'] = oddsRatioCodonTaxon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2352490\n"
     ]
    }
   ],
   "source": [
    "with pd.HDFStore(str(analysisCtermDataPath / 'oddsRatioCodonTaxon.h5'), mode='r') as store:\n",
    "    oddsRatioCodonTaxon = store['oddsRatioCodonTaxon']\n",
    "print(len(oddsRatioCodonTaxon))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon.set_index(['terminus', 'position from terminus', 'observable'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "      <th>n_genomes</th>\n",
       "      <th>n_seq</th>\n",
       "      <th>stop_codon</th>\n",
       "      <th>taxaid</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>taxon_rank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1576</td>\n",
       "      <td>4871983</td>\n",
       "      <td>all</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1576</td>\n",
       "      <td>1836672</td>\n",
       "      <td>TAA</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1576</td>\n",
       "      <td>936585</td>\n",
       "      <td>TAG</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1576</td>\n",
       "      <td>2098684</td>\n",
       "      <td>TGA</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>691</td>\n",
       "      <td>2165400</td>\n",
       "      <td>all</td>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    AAA   AAC   AAG    AAT  \\\n",
       "terminus position from terminus observable                                   \n",
       "C        -20                    BH_multiple_tests  True  True  True   True   \n",
       "                                BH_multiple_tests  True  True  True   True   \n",
       "                                BH_multiple_tests  True  True  True   True   \n",
       "                                BH_multiple_tests  True  True  True  False   \n",
       "                                BH_multiple_tests  True  True  True  False   \n",
       "\n",
       "                                                     ACA   ACC    ACG    ACT  \\\n",
       "terminus position from terminus observable                                     \n",
       "C        -20                    BH_multiple_tests   True  True   True  False   \n",
       "                                BH_multiple_tests  False  True   True   True   \n",
       "                                BH_multiple_tests   True  True   True  False   \n",
       "                                BH_multiple_tests   True  True   True   True   \n",
       "                                BH_multiple_tests   True  True  False   True   \n",
       "\n",
       "                                                    AGA    AGC      ...       \\\n",
       "terminus position from terminus observable                          ...        \n",
       "C        -20                    BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "\n",
       "                                                    TTA    TTC   TTG    TTT  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -20                    BH_multiple_tests  True   True  True   True   \n",
       "                                BH_multiple_tests  True  False  True   True   \n",
       "                                BH_multiple_tests  True   True  True  False   \n",
       "                                BH_multiple_tests  True   True  True   True   \n",
       "                                BH_multiple_tests  True   True  True  False   \n",
       "\n",
       "                                                  n_genomes    n_seq  \\\n",
       "terminus position from terminus observable                             \n",
       "C        -20                    BH_multiple_tests      1576  4871983   \n",
       "                                BH_multiple_tests      1576  1836672   \n",
       "                                BH_multiple_tests      1576   936585   \n",
       "                                BH_multiple_tests      1576  2098684   \n",
       "                                BH_multiple_tests       691  2165400   \n",
       "\n",
       "                                                  stop_codon taxaid  \\\n",
       "terminus position from terminus observable                            \n",
       "C        -20                    BH_multiple_tests        all      2   \n",
       "                                BH_multiple_tests        TAA      2   \n",
       "                                BH_multiple_tests        TAG      2   \n",
       "                                BH_multiple_tests        TGA      2   \n",
       "                                BH_multiple_tests        all   1224   \n",
       "\n",
       "                                                       taxon_name  \\\n",
       "terminus position from terminus observable                          \n",
       "C        -20                    BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests  Proteobacteria   \n",
       "\n",
       "                                                     taxon_rank  \n",
       "terminus position from terminus observable                       \n",
       "C        -20                    BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests        phylum  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioCodonTaxon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2335583\n"
     ]
    }
   ],
   "source": [
    "oddsRatioAATaxon = pd.read_csv(analysisCtermDataPath / 'oddsRatioAATaxon.csv', nrows=None)\n",
    "print(len(oddsRatioAATaxon))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioAATaxon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO: correct if needed oddsRatioAATaxon table, same way as the codon table above. Then write to HDF5 format."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot codon biases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# taxonGroup = taxaDf[taxaDf['taxon_id'] == 1280].iloc[0]\n",
    "# taxonGroup\n",
    "\n",
    "# print(\"taxonGroupRef:\", taxonGroup)\n",
    "# taxonRank = taxonGroup['rank_name']\n",
    "# taxonName = taxonGroup['taxon_name']\n",
    "# taxonId = int(taxonGroup['taxon_id'])\n",
    "\n",
    "# # Select genomes in the taxon group\n",
    "# taxonAssemblyDf = assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf[taxonRank + '_taxid'] == taxonId]\n",
    "# taxonAssemblyAccessionList = list(taxonAssemblyDf.index)\n",
    "\n",
    "# taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "# print(\"output path:\", str(analysisCtermDataPath / taxonSuffix))\n",
    "# suffix = taxonSuffix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plot codon bias for the 3 codon contexts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminus = 'C'\n",
    "statistical_test = 'contingency'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'taxaId_2'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statName = 'all'\n",
    "dfFormatOutput = 'csv'\n",
    "speciesSetName = taxonSuffix\n",
    "outputFilePrefix = simplify_prefix(speciesSetName + '_subset_' + statName)\n",
    "plotTitle = '{} {} {}'.format(taxonRank, taxonName, re.sub(r'_', r' ', outputFilePrefix))\n",
    "speciesSetName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "      <th>n_genomes</th>\n",
       "      <th>n_seq</th>\n",
       "      <th>stop_codon</th>\n",
       "      <th>taxaid</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>taxon_rank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1576</td>\n",
       "      <td>4871983</td>\n",
       "      <td>all</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1576</td>\n",
       "      <td>1836672</td>\n",
       "      <td>TAA</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1576</td>\n",
       "      <td>936585</td>\n",
       "      <td>TAG</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1576</td>\n",
       "      <td>2098684</td>\n",
       "      <td>TGA</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>691</td>\n",
       "      <td>2165400</td>\n",
       "      <td>all</td>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>phylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    AAA   AAC   AAG    AAT  \\\n",
       "terminus position from terminus observable                                   \n",
       "C        -20                    BH_multiple_tests  True  True  True   True   \n",
       "                                BH_multiple_tests  True  True  True   True   \n",
       "                                BH_multiple_tests  True  True  True   True   \n",
       "                                BH_multiple_tests  True  True  True  False   \n",
       "                                BH_multiple_tests  True  True  True  False   \n",
       "\n",
       "                                                     ACA   ACC    ACG    ACT  \\\n",
       "terminus position from terminus observable                                     \n",
       "C        -20                    BH_multiple_tests   True  True   True  False   \n",
       "                                BH_multiple_tests  False  True   True   True   \n",
       "                                BH_multiple_tests   True  True   True  False   \n",
       "                                BH_multiple_tests   True  True   True   True   \n",
       "                                BH_multiple_tests   True  True  False   True   \n",
       "\n",
       "                                                    AGA    AGC      ...       \\\n",
       "terminus position from terminus observable                          ...        \n",
       "C        -20                    BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "                                BH_multiple_tests  True  False      ...        \n",
       "\n",
       "                                                    TTA    TTC   TTG    TTT  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -20                    BH_multiple_tests  True   True  True   True   \n",
       "                                BH_multiple_tests  True  False  True   True   \n",
       "                                BH_multiple_tests  True   True  True  False   \n",
       "                                BH_multiple_tests  True   True  True   True   \n",
       "                                BH_multiple_tests  True   True  True  False   \n",
       "\n",
       "                                                  n_genomes    n_seq  \\\n",
       "terminus position from terminus observable                             \n",
       "C        -20                    BH_multiple_tests      1576  4871983   \n",
       "                                BH_multiple_tests      1576  1836672   \n",
       "                                BH_multiple_tests      1576   936585   \n",
       "                                BH_multiple_tests      1576  2098684   \n",
       "                                BH_multiple_tests       691  2165400   \n",
       "\n",
       "                                                  stop_codon taxaid  \\\n",
       "terminus position from terminus observable                            \n",
       "C        -20                    BH_multiple_tests        all      2   \n",
       "                                BH_multiple_tests        TAA      2   \n",
       "                                BH_multiple_tests        TAG      2   \n",
       "                                BH_multiple_tests        TGA      2   \n",
       "                                BH_multiple_tests        all   1224   \n",
       "\n",
       "                                                       taxon_name  \\\n",
       "terminus position from terminus observable                          \n",
       "C        -20                    BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests        Bacteria   \n",
       "                                BH_multiple_tests  Proteobacteria   \n",
       "\n",
       "                                                     taxon_rank  \n",
       "terminus position from terminus observable                       \n",
       "C        -20                    BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests  superkingdom  \n",
       "                                BH_multiple_tests        phylum  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioCodonTaxon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = oddsRatioCodonTaxon[oddsRatioCodonTaxon['taxaid'] == taxonId]\\\n",
    "           .drop(['taxaid', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes'], axis=1)\n",
    "plotData = plotData.xs(terminus, level='terminus')\n",
    "plotData = select_terminal_positions_in_index(plotData, terminus)\n",
    "plotDataTemp = plotData.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotData.to_csv(analysisCtermDataPath / 'oddsRatioCodonTaxon_taxaId_2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "      <th>stop_codon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>position from terminus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.45502</td>\n",
       "      <td>0.185648</td>\n",
       "      <td>0.779589</td>\n",
       "      <td>0.288478</td>\n",
       "      <td>-0.511378</td>\n",
       "      <td>-1.32289</td>\n",
       "      <td>-1.63726</td>\n",
       "      <td>-0.802509</td>\n",
       "      <td>1.24263</td>\n",
       "      <td>0.319846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.238899</td>\n",
       "      <td>0.39357</td>\n",
       "      <td>-0.109323</td>\n",
       "      <td>-0.213098</td>\n",
       "      <td>-0.0180203</td>\n",
       "      <td>0.102137</td>\n",
       "      <td>-0.0616251</td>\n",
       "      <td>-0.524969</td>\n",
       "      <td>0.0832374</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.55131</td>\n",
       "      <td>0.232382</td>\n",
       "      <td>1.26658</td>\n",
       "      <td>0.374393</td>\n",
       "      <td>-1.31151</td>\n",
       "      <td>-2.09353</td>\n",
       "      <td>-1.73563</td>\n",
       "      <td>-1.07142</td>\n",
       "      <td>1.06892</td>\n",
       "      <td>0.0165654</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0723837</td>\n",
       "      <td>0.309546</td>\n",
       "      <td>-0.405192</td>\n",
       "      <td>0.144922</td>\n",
       "      <td>-0.168143</td>\n",
       "      <td>-0.0792</td>\n",
       "      <td>0.1338</td>\n",
       "      <td>-0.402319</td>\n",
       "      <td>0.213132</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.6145</td>\n",
       "      <td>0.0166325</td>\n",
       "      <td>1.18059</td>\n",
       "      <td>0.297567</td>\n",
       "      <td>-0.862987</td>\n",
       "      <td>-1.80446</td>\n",
       "      <td>-1.43651</td>\n",
       "      <td>-0.820776</td>\n",
       "      <td>1.42268</td>\n",
       "      <td>0.214125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128262</td>\n",
       "      <td>0.681705</td>\n",
       "      <td>-0.531884</td>\n",
       "      <td>0.04579</td>\n",
       "      <td>0.102677</td>\n",
       "      <td>0.061261</td>\n",
       "      <td>-0.166365</td>\n",
       "      <td>-0.164196</td>\n",
       "      <td>0.23639</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.11719</td>\n",
       "      <td>0.213708</td>\n",
       "      <td>-0.181898</td>\n",
       "      <td>0.078626</td>\n",
       "      <td>0.617179</td>\n",
       "      <td>-0.897764</td>\n",
       "      <td>-1.66614</td>\n",
       "      <td>-0.274741</td>\n",
       "      <td>1.47539</td>\n",
       "      <td>0.53796</td>\n",
       "      <td>...</td>\n",
       "      <td>0.671197</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.154971</td>\n",
       "      <td>-0.688177</td>\n",
       "      <td>0.142402</td>\n",
       "      <td>0.641371</td>\n",
       "      <td>-0.150502</td>\n",
       "      <td>-0.975846</td>\n",
       "      <td>-0.368551</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            AAA        AAC       AAG       AAT       ACA  \\\n",
       "position from terminus                                                     \n",
       "-1                      1.45502   0.185648  0.779589  0.288478 -0.511378   \n",
       "-1                      1.55131   0.232382   1.26658  0.374393  -1.31151   \n",
       "-1                       1.6145  0.0166325   1.18059  0.297567 -0.862987   \n",
       "-1                      1.11719   0.213708 -0.181898  0.078626  0.617179   \n",
       "\n",
       "                             ACC      ACG       ACT      AGA        AGC  \\\n",
       "position from terminus                                                    \n",
       "-1                      -1.32289 -1.63726 -0.802509  1.24263   0.319846   \n",
       "-1                      -2.09353 -1.73563  -1.07142  1.06892  0.0165654   \n",
       "-1                      -1.80446 -1.43651 -0.820776  1.42268   0.214125   \n",
       "-1                     -0.897764 -1.66614 -0.274741  1.47539    0.53796   \n",
       "\n",
       "                          ...            TCT       TGA       TGC       TGG  \\\n",
       "position from terminus    ...                                                \n",
       "-1                        ...       0.238899   0.39357 -0.109323 -0.213098   \n",
       "-1                        ...      0.0723837  0.309546 -0.405192  0.144922   \n",
       "-1                        ...       0.128262  0.681705 -0.531884   0.04579   \n",
       "-1                        ...       0.671197       NaN  0.154971 -0.688177   \n",
       "\n",
       "                              TGT       TTA        TTC       TTG        TTT  \\\n",
       "position from terminus                                                        \n",
       "-1                     -0.0180203  0.102137 -0.0616251 -0.524969  0.0832374   \n",
       "-1                      -0.168143   -0.0792     0.1338 -0.402319   0.213132   \n",
       "-1                       0.102677  0.061261  -0.166365 -0.164196    0.23639   \n",
       "-1                       0.142402  0.641371  -0.150502 -0.975846  -0.368551   \n",
       "\n",
       "                       stop_codon  \n",
       "position from terminus             \n",
       "-1                            all  \n",
       "-1                            TAA  \n",
       "-1                            TAG  \n",
       "-1                            TGA  \n",
       "\n",
       "[4 rows x 65 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotData.xs('log2OddsRatio', level='observable').loc[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "      <th>stop_codon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAA    AAC   AAG   AAT   ACA   ACC   ACG   ACT   AGA    AGC  \\\n",
       "terminus                                                                 \n",
       "C         True   True  True  True  True  True  True  True  True   True   \n",
       "C         True   True  True  True  True  True  True  True  True  False   \n",
       "C         True  False  True  True  True  True  True  True  True   True   \n",
       "C         True   True  True  True  True  True  True  True  True   True   \n",
       "\n",
       "            ...       TCT   TGA   TGC   TGG    TGT   TTA   TTC   TTG   TTT  \\\n",
       "terminus    ...                                                              \n",
       "C           ...      True  True  True  True  False  True  True  True  True   \n",
       "C           ...      True  True  True  True   True  True  True  True  True   \n",
       "C           ...      True  True  True  True   True  True  True  True  True   \n",
       "C           ...      True   NaN  True  True   True  True  True  True  True   \n",
       "\n",
       "         stop_codon  \n",
       "terminus             \n",
       "C               all  \n",
       "C               TAA  \n",
       "C               TAG  \n",
       "C               TGA  \n",
       "\n",
       "[4 rows x 65 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioCodonTaxon[oddsRatioCodonTaxon['taxaid'] == taxonId]\\\n",
    "           .drop(['taxaid', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes'], axis=1)\\\n",
    "           .xs('BH_multiple_tests', level='observable')\\\n",
    "           .xs(-1, level='position from terminus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = plotData.xs('log2OddsRatio', level='observable')\n",
    "plotData = plotData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "plotData = plotData[plotData.columns].astype(float)\n",
    "plotData = plotData.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskData = plotDataTemp\n",
    "maskData = maskData.xs('BH_multiple_tests', level='observable')\n",
    "maskData = maskData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "# Mask both rejected cases and null values\n",
    "maskData = (maskData != True) & (maskData != 'True')\n",
    "maskData = maskData.transpose()\n",
    "maskData.sum().sum()\n",
    "# maskData.shape[0]*maskData.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>stop_codon</th>\n",
       "      <th>all</th>\n",
       "      <th>TAA</th>\n",
       "      <th>TAG</th>\n",
       "      <th>TGA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGC</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAA</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAG</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "stop_codon    all    TAA    TAG    TGA\n",
       "AAA         False  False  False  False\n",
       "AAC         False  False   True  False\n",
       "AAG         False  False  False  False\n",
       "AAT         False  False  False  False\n",
       "ACA         False  False  False  False\n",
       "ACC         False  False  False  False\n",
       "ACG         False  False  False  False\n",
       "ACT         False  False  False  False\n",
       "AGA         False  False  False  False\n",
       "AGC         False   True  False  False\n",
       "AGG         False  False  False  False\n",
       "AGT         False  False  False  False\n",
       "ATA         False  False  False  False\n",
       "ATC         False  False  False  False\n",
       "ATG         False  False  False  False\n",
       "ATT         False  False  False  False\n",
       "CAA         False  False  False  False\n",
       "CAC         False  False   True  False\n",
       "CAG         False  False  False  False\n",
       "CAT         False  False  False  False\n",
       "CCA         False  False   True  False\n",
       "CCC         False  False  False  False\n",
       "CCG         False  False  False  False\n",
       "CCT         False  False  False  False\n",
       "CGA         False  False  False  False\n",
       "CGC         False  False  False  False\n",
       "CGG         False  False  False  False\n",
       "CGT         False  False  False  False\n",
       "CTA         False  False  False  False\n",
       "CTC         False  False  False  False\n",
       "..            ...    ...    ...    ...\n",
       "GAG         False  False  False  False\n",
       "GAT         False  False  False  False\n",
       "GCA         False  False  False  False\n",
       "GCC         False  False  False  False\n",
       "GCG         False  False  False  False\n",
       "GCT         False  False  False  False\n",
       "GGA         False  False  False  False\n",
       "GGC         False  False  False  False\n",
       "GGG         False  False  False  False\n",
       "GGT         False  False  False  False\n",
       "GTA         False  False  False  False\n",
       "GTC         False  False  False  False\n",
       "GTG         False  False  False  False\n",
       "GTT         False  False  False  False\n",
       "TAA          True   True   True   True\n",
       "TAC         False  False  False  False\n",
       "TAG          True   True   True   True\n",
       "TAT         False  False  False  False\n",
       "TCA         False  False  False  False\n",
       "TCC         False  False  False  False\n",
       "TCG         False  False  False  False\n",
       "TCT         False  False  False  False\n",
       "TGA         False  False  False   True\n",
       "TGC         False  False  False  False\n",
       "TGG         False  False  False  False\n",
       "TGT          True  False  False  False\n",
       "TTA         False  False  False  False\n",
       "TTC         False  False  False  False\n",
       "TTG         False  False  False  False\n",
       "TTT         False  False  False  False\n",
       "\n",
       "[64 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskData"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the case of individual species, we decrease a lot the threshold of pvalue to plot the biases, otherwise we get very few significant biases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    maskData = plotDataTemp\n",
    "    maskData = maskData.xs('pvalue', level='observable')\n",
    "    maskData = maskData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "    # Mask both rejected cases and null values\n",
    "    # maskData = (maskData != True) & (maskData != 'True')\n",
    "    maskData = maskData < 1e-2\n",
    "    maskData = maskData.transpose()\n",
    "    maskData.sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "stop_codon\n",
       "TAA    0.232382\n",
       "TAG         NaN\n",
       "TGA    0.213708\n",
       "Name: N/AAC, dtype: float64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "addEmptyRowBetweenGroups = True\n",
    "addEmptyRowBetweenGroupsFillValue = np.nan\n",
    "plotData2 = plotData.copy()\n",
    "nameIndex = plotData2.index.name\n",
    "nameCol = plotData2.columns.name\n",
    "stopCodonList = list(refCodonTableDf[refCodonTableDf['aa'] == '*'].index)\n",
    "\n",
    "# Remove values that are not significant using mask\n",
    "plotData2 = plotData2.where(~maskData, other=np.nan)\n",
    "# plotData2 = plotData2.to_frame()\n",
    "plotData2 = plotData2.loc[plotData2.index.map(lambda x: x not in stopCodonList)]\n",
    "plotData2 = sort_codon_index(plotData2, refCodonTableDf,\n",
    "                             addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                             addEmptyRowBetweenGroups=addEmptyRowBetweenGroups)\n",
    "plotData2.index.name = nameIndex\n",
    "plotData2.columns.name = nameCol\n",
    "plotData2 = plotData2.drop('all', axis=1)\n",
    "plotData2.loc['N/AAC']\n",
    "# print(\"plotData:\\n\", type(plotData), plotData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UAA</th>\n",
       "      <th>UAG</th>\n",
       "      <th>UGA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R/CGU</th>\n",
       "      <td>0.362795</td>\n",
       "      <td>0.525103</td>\n",
       "      <td>0.349839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/CGC</th>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.351280</td>\n",
       "      <td>0.734030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/CGA</th>\n",
       "      <td>1.355864</td>\n",
       "      <td>1.743242</td>\n",
       "      <td>2.775695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/CGG</th>\n",
       "      <td>0.456227</td>\n",
       "      <td>0.710066</td>\n",
       "      <td>-0.055614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/AGA</th>\n",
       "      <td>1.068920</td>\n",
       "      <td>1.422678</td>\n",
       "      <td>1.475391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            UAA       UAG       UGA\n",
       "R/CGU  0.362795  0.525103  0.349839\n",
       "R/CGC  0.118014  0.351280  0.734030\n",
       "R/CGA  1.355864  1.743242  2.775695\n",
       "R/CGG  0.456227  0.710066 -0.055614\n",
       "R/AGA  1.068920  1.422678  1.475391"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotData2.columns = [re.sub('T', 'U', col) for col in plotData2.columns]\n",
    "plotData2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py:194: UserWarning: 'colors' is being deprecated. Please use 'color'instead of 'colors'\n",
      "  warnings.warn((\"'colors' is being deprecated. Please use 'color'\"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py:194: UserWarning: 'colors' is being deprecated. Please use 'color'instead of 'colors'\n",
      "  warnings.warn((\"'colors' is being deprecated. Please use 'color'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxaId_2_subset_all_composition_bias_codon_Cterminal_barplot_stopCodonContext\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/matplotlib/font_manager.py:1241: UserWarning: findfont: Font family ['cursive'] not found. Falling back to DejaVu Sans.\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATcAAADWCAYAAAC0TcWhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3XdYVGfaP/DvzEhHQYpSAlg2NrAkXGIvBIxubMQeUaJmE+vuG0t2MYpLyhI1KLFmsQ2GF9RgLCRoNCBRDGIBEVQgBFCD6CBtBJEZhpnfH/6YF/DMcKYxxftzXbmit89znvs8h7k558wpHJlMJgMhhJgYrr4TIIQQXaDiRggxSVTcCCEmiYobIcQkUXEjhJgkKm6EEJNExY0QYpKouBm5kydPYv78+Zg3bx62bNny0r9nZGSgoqJC7eU/efIE3377Lev27733HmO8pqYGly9fVjuPEydOICEhoVUsNDQU9+/fZ2wfHR2NefPmYeHChSgrK1N7XGK8qLgZsTt37uDChQuIjY3F0aNH8ezZM5w7d65Vm1OnTuHZs2dqj+Hs7Izly5drmioKCgpw/fp1jZfDRk1NDa5du4ajR49i1apVOHz4cIeMSwxLJ30nQNR3/vx5zJkzBzweDwCwfv16dOr0f5u0rKwMaWlpEAqFCA8Px/r169HQ0ICJEyfi/fffx/z588HlchEYGIiLFy+Cy+VCKpVixIgR+OWXXzB+/HhMnz4d33zzDWbPng0+nw+RSATgxZ7R7du3sWPHDjQ0NCAgIAAfffSRwly///57ZGVlYdKkSYiLi8PDhw/R0NCAb775BkePHoWbmxt8fHywZ88eREZG4l//+heEQiEAtNpzrKqqwj/+8Q906tQJT58+ZRyrc+fO2LZtGwBAIpHAwsJCs4kmRon23IxYVVUVunfvLv+7lZUVzMzM5H93c3PDmDFjEBoaiv379+Pvf/874uPjce3aNVRVVaGyshL79u1DYGAghEIhDhw4gB49eqBr1644evQoLl682Go8c3Nz8Pl8eHh4ICcnBw8ePMDu3bsRFxeHlJQUpbnOmTMH06ZNQ/fu3TF8+HDw+XxMmjQJV65cwbJly3DixAls3rwZGzduxKNHjzBr1izExMSgT58+uHPnjnw5P/zwA4KDg8Hn81uta0s8Hg/29vYQCoX45ptvMG/ePHWmlxg52nMzYs7Oznj06BH69OkDAHj8+DEqKyvx448/4s6dO5gyZYq87b179xAZGQkulwuhUIjHjx+ja9eusLa2RlVVFXr27AkOhwMHBwd4enqCx+PJ9wib9ezZUz6uWCyGs7MzwsLC0KVLF4jFYlY529raIjMzE2lpaaioqMDkyZNhYWGBkSNHoqCgAC4uLqipqUF0dDSSkpJw7949BAQEyPuXlZXB398fHA4H/fr1UzjO06dPsXTpUnz88cdwc3NjPafEdFBxM2KBgYGIjo7GmDFjwOVysXfvXgwfPhyhoaHyNtnZ2ZDJZPDw8MC8efPQt29fxMXFwd3dHRwOR96u5Z/Z2rFjB/73f/8XT58+RXBwsNK2zcu/ePEinJ2dsWLFCkREREAmk6GqqgrXr1+Hi4sLbt68iVu3bmHEiBGYPn06Vq5ciZbPdvDy8kJeXh7+8pe/oKioSOF469atw/LlyzFq1CiV14uYBipuRmzAgAHw8/PD/Pnz0dTUhKFDh+Kdd955qU14eDi++uorbNy4EbW1tfDx8YGdnZ3G448ZMwazZ8+GnZ0dbGxsIJFI5P/2z3/+E1u3bpX/3cPDAykpKRg9ejR+/vlnXLp0CXZ2dnBycsKWLVvw97//HZ6envj4448RGhqK9evX4+jRo7C0tERlZaV8OXPnzsWqVatw/PjxVuO1lJ2djRs3buD58+c4cOAARo4cqZUvRYhx4dAjj4guREdHY+nSpfpOg7zCqLgRnaioqICTk5POx7l06RL279/fKjZt2jTMnj1b52MTw0bFjRBikuhSEEKISaLiRggxSQb5bWl2djZdVU6IHohEIgwZMkTfaWiFQRY3CwsL9O/fX99pEPLKycvL03cKWmMyh6UpF3rrOwVCiAExyD03QohhqKmpwaNHj/SdBiNLS0u89tprCu8xpuJGCFGooqICPXr0gJWVlb5TaUUmk6GyshKlpaXye57bMpnDUkKI9jU2NsLS0lLfabyEw+HA0dERDQ0NCttQcSNaQec8TRfTQxUEAoEeMmmtvYc9UHEjhKgsMjJS68vctWuXwsfGq0Olc25VVVXgcrmwt7fXWgKEEMN35swZZGRk4NmzZ1i7di3y8/Nx5coVZGZmorq6GjKZDJs2bcLUqVPx7rvvQiwWY9myZQBeXDsXERGBxsZGjBw5EhYWFkhNTYVYLMann36KM2fO4P79+7h58yamTZuG3bt3t1rmjBkzEBAQgCdPniA8PJx1zqz23NLT0/HZZ59h3759iI6OxhdffIHMzEyVJyguLg779+/Hzp07Ve5LCNGfsrIy2NvbY+7cuejevTv69u2LgQMHory8HGFhYXB1dcXdu3fRp08fLFmyBDU1NairqwPw4iVFw4YNQ0REBF5//XX8/PPPiIiIQEhICBITE5GTk4MNGzZg2LBhaGhoeGmZbm5uWLlypXx5bLHac7OxsUFISAji4+Mhk8mwaNGiVs/YYmvChAlwdHRERESE0nYikUitiwlN6QJEY0Tzb7pGjx4NDoeD6Oho2NrayuMtz3tJpVL5n0UikfzfJBKJ/M9t30QmlUrB5b7Yx+LxeJDJZC8ts/kLjZbvB2GDVevBgwfjk08+wfr16wEA27dvx5dffqnSQMCLF3ds374dixcvVtpOnTsUyh6B7mrQI5p/06DoF1RhYSF+++03cLlcuLi4oEuXLrh8+TKcnJywefNmcDgcDBgwALdv38aWLVvQvXt32NjYAHhRGMPDw5GWlobRo0dj4sSJ2LRpE5qamrB27VqcO3cOX331FXJycjBz5syXlqk2GUuffPKJrKmpSdbY2CgLDQ1l262VNWvWyDZv3izj8/lK2929e1flZSen9FIrJ6IdNP+moe1nT9XP4tq1a7WZTruU5cd6P2/+/PnYsGEDgBdvMlJH8+vWtM0lNRtxGvR97G8aNwoTom/N36IKBIJWb2ZjUl5eDmdnZ7Xe38EGqy8UysvL4ejoiOXLl2P58uUd8oRVQ+SSmq3vFAgxCmwuFYmKikJTU5POcmC155aWlvbS/WWrVq3SSUIGK9wOGHex/XaEmKjs7GwcP34cT58+RdeuXdGpUyfY2dlhxowZ+PTTT+Hn54fGxkaMHTsW+fn5KC4uRnx8PKRSKdzd3dGnTx/k5eWBy+WiZ8+euHPnDpKTkzFp0iSd5NvunptQKISbmxtqamrg6+sLX19fFBcX6yQZXXNJzaYr6QlRU0JCAsLDw/H++++jW7duCAsLQ3V1NYRCIfr3749Vq1bh4cOH8PX1Rd++ffHgwQMIBAJYW1sjJycHY8aMQWlpKR4/foyAgAAMGDAAgYGBOsu33eJmZWUFgUAAkUgEgUAAgUCA9957T2cJqSxcg1fUadKXkFdM8yUdtra2rd4lK5PJGC/XaGpqwogRI7Bu3Tp5EXv+/Dlqa2sBqPeuXFW0e1hqbm6OoKAgDBw4ELm5uZDJZMjLy8PQoUN1mphRCLcDwoX6zoKQDjFz5kyEhYXh+fPn6NKlC7Zs2QJnZ2fGO5YqKirg7u6OpKQk3L17Fz179sTx48cxbtw4SKVS/PDDD/D09MSRI0ewcOFCneTL+tvSyMhI9OjRA3369EFubq5OktGXlAu9EfCW4reXK2tL37aSV4Wfnx/8/PwY/2316tUAgM2bNwMAYmJiALy4JlZfWN847+DggICAAHTu3FkrbyvvcHQISjoAndM1HKyLm5ubGxwcHJCRkQFnZ2dd5kQIIRpjfVhaWloKsViMkJAQnZ8IJIZNlcN4QvSFdXFzd3dHcnKy/O/6vM6NzYeLzoURokXqntZR8IXbrl27MG3aNHh5eeHq1at48OABXF1dsXv3bhw9elTebtOmTejRoweWLFmi8tCsD0tXrVrV6j/yajDWc0j6yNtY7mAx1G36008/YfTo0bh+/TqAF9fYAsCNGzdaXXrClkE/iddYflgMmSHOIdOHy1A/cKzRF1YqaXtqq7a2Fo2NjZg5c6Z8z+3HH3/E+PHj4ePjg8uXL6s8BuviJpFIIJVKkZ+fj+rqapUH6lBG9IPW4R9qPc+NIRbbVwnT/Otjm3Tu3BkVFRUAXtxkX19fj+rqanz77be4efMmKioq8NNPPyElJQVFRUWtDlXZYl3cQkNDkZWVhf379+Orr75SeSCV/f8PYdsPP304VKCgkGl7Dk1mmxjRL0VjN3XqVHz33XdYt24dUlJScPXqVfz3v//F559/jtDQUCQkJOCNN97Af/7zH2zbtg0cDuelB122h/UXCjY2NkhLS8PSpUvB5/NVXplXGtOdDK/a3Q3NhcPAHz7w0hdRRr6dtPbNtpbnwMHBATt27GD8t7fffvul2O7du1Ueg9WeW0JCAjw8PNClSxdcvXoVHh4eKg8EACkpKYiLi5NfvfyqMbbzSqz2yLSwt2MMe35qbTsV58Ykz0XqEavi5uXlhdzcXPTp0we9evVS+6kgWVlZCA4ORlVVlVr96bChA73Cc63NgsK2cHdIgX/FtilHxvI71g0bNuCLL76ATCZDeHg4vvjiC5UHi4qKwurVqxEZGYl169YpbJednQ0LCwuVl68vZY+mwM31J6Vt/B+LkOqi+jr1PzYcLuMuIk42Uz5Gc0yd5TVjk7OiXNoblylnTSmbh/bmRhf5qKK9ue5/bDjy5ma8FGO7jbX9M9LyXRh5eXkG/W4MZfmxPuc2Y8YMhIWFgcPhYPbs2Wol4uPjg5iYmHZv31LnBTH6xOblKI81XB2m8yaazJEmL3RRpZ8utiPTMtvGFI2rj58rNnOtjXy18TNiSm8wY1XcEhISwOFwMGTIixOthYWFGDx4sMqDTZgwQeU+xuBVuhWJ7V0fj/2HIOWCjpMxFVo4Wa/ru3HUPWxWlBebOxQKCgoQHR0NS0tLiMVirF27Fq6urqzHZlXcvLy8EBcXh1mzZoHL5eLkyZOYNWsW60GIBsKFeKzvHFSlw28XX6VfJK+alncoDB06FHv27EFkZCTMzc1RVVUFgUCgUnFj9YWCn58fbG1tMWrUKAwfPhxWVlZqrwDRAiO5NIEKkQaMZBuri80dChwOB+bm5khOTsbWrVtx/vx5lcbo0HNuRHvooQCmT9NtbMg/I813KHh5eSm8Q0EkEuHp06cIDAzE2LFjsWnTJpXGYF3cHBwcYG1tDZlMBgcHB5VXhpgAE9+bMFodsF20XSinTp2Kzz77DEeOHEFjYyOqqqpw8OBBmJub4/z580hISMDHH3+M0NBQWFhYQCKRqPxkENbFbe/evVi/fj2AF48O/vLLL1VbG2JQ6JCR6BPbOxT27t2r9hisixuHw4G9vT2kUqlOX6RKiCKGfJhlcGgvm31xmz9/PjZs2AAAdM6NGJS2e6FUBAmgQnEbMmSI/Do3Qgh7dApAP1gVt+joaAwcOBDe3t7g8XjIzMzEH3/8gQ8++EDX+RHyAh1m6ZW699sqKuxtL+K9f/8+fv/9dzx79gwSiQTjxo3DlClTAAAffPABZs+ejUmTJqk0NqvitnTpUqSmpmLXrl3gcDgYOXIkFTZCiNY0NDRg7NixGDt2LAAgPT0dAFBUVARvb28kJSWpXNxYP6zS398fM2bMwIQJE+Dv749bt26pNBAhHYL28IxC24t4t2zZgsGDB6OqqgqhoaH473//CwA4fvw4Jk+eDDs7OxQVqXZ4r9I7FK5cuQIXFxfExsYiKytLpYEIIaRZ28eML1++HOnp6XBwcMDmzZvh5OSExsZGpKamIi4uDhUVFThy5IhKY7D+QgEAvL294enpiXnz5mH79u0qDUSIPtGN/Ial7UW8X375JbZt24azZ88CAAYMGIDU1FQsXLgQwcHBkMlkCAkJwfPnz1nf/sm6uP34448YMGAAAKCyshISiQSZmZnw9fVVY9UIIcZE29/4Ml3EGx4errA9h8NBbGysSmOwLm6HDh3C9OnTcfnyZdTW1mLZsmX47rvvqLgRo0GXZLxaWJ9zGzZsGIKCgvD777+jtLQU1tbW9HQQonesLtilLxk0os4LkTtCe3mx3nMbM2YMDh48iIULF0IoFGLr1q1wdHRknci9e/cQHx8PiUSCZcuWoVu3bqz7EkL0w8zMDA0NDQa3IyOTyVBZWQlLS0uFbVgXN5lMhqdPn+LIkSMIDAzEv//9b1b94uPjcf36dYjFYmzcuBHnzp1DXV0dFTdCjICTkxPu3bun7zQYWVpa4rXXXlP476xfELN161b885//BAB8/fXX+OSTT1ROJjExEVwuV37lsSLG9oIYQkyJMb2/RBnWe25isRgikQhSqRSNjY0qD3Tp0iWcPn0avr6++PPPP5W++9TYXhBDiKl45V4Qk5OTg0mTJmHZsmWQSqVYs2aNygO1vLWCEEJ0jVVxa77tYdq0aQCAkpIStd5+RQghHYVVcXv33Xd1nQchhGiVSveWEkKIsaDiRggxSVTcCCEmiYobIcQkUXEjhJgklZ7n1lFEIpFJXUxIiLEQiUT6TkFrWN9+RQghxoQOSwkhJomKGyHEJFFxI4SYJCpuhBCTRMWNEGKSqLgRQkwSFTdCiEkyuOJWVlaGsrIyCASCVm+3af6zVCqFRCJp1UcqleLatWutYg0NDRCLxWq1Y6JOX6Z1UbYe2uqrLOf28tbF/HfkPHT0dlI2rq7H7oi5MWYGV9xyc3Nx+/ZtZGVlITIyUh7fvn07JBIJGhoa8O233wIAzp49i4SEBISHh8PBwQEAsG3bNgiFQuzcuRM7d+5UqR0TTfoyrQvTemi7L1PObPPWxfx3xDzoaztpMtfGMDfGzCBvv5LJZOBwOJBKpa3iOTk5aGpqkv/92bNnkEql4HA48o3G4XBQVFSEESNG4ObNmyq1Y9Je3+zsbFbr0nIvqHk9OByOTvoy5axK3tqef0XLbLs8VfLRJBdlY3T0XLNdP036ajo3xsrgbr8qLi4Gj8fDkydP4O7uDldXVwAvNlBxcTE6deoEe3t7uLq6QiwWg8vloqCgAIWFhQgKCkJdXR2Ki4thZmamcjtlKisrkZaWhqCgIDQ2NsrvfS0uLkZQUBBjn4cPH8o/GNeuXUNQUBAkEgny8/MhlUp11pcpZwCs8s7Pz0fnzp3B4/Fgbm4u/zBUVVWhrKwMjY2N6NWrF+zs7FBXVwdbW1s8efIEV69exZQpU1BVVYXS0lI0NTXBy8sLDg4OjNuUaXsyYcpHk1wMaa6b/03duWHbVxs//8bI4A5LExIS4OjoiOvXr+OHH36Qx3fv3o2ePXvi119/lcejoqLQ0NCAS5cuobS0FACwZ88e9OrVS612TI4ePYqEhATs2bMHgYGBAIAffvgBBQUFOHXqlDzG5NKlS/jtt99w4MABebvjx48jLy9Pp32Zcmabd1JSElxcXODg4ID4+Hh5nM/no3///vD29kZsbCwAIDo6Gk1NTbCzs5O/25LP58Pb2xve3t7y/kzblGl7ss1Hk1yYpKWl6WWuNZ0btn01+fk3ZgZX3GxsbJCamgpXV1eYm5srjWs7xsTJyUl+crakpERhjImzs7Ne+ipqx6a/hYUFzpw5gzNnzrSaF6Y425gm899yec3vstUkxna+OmKuNZ2bjvj5N2a88PDwcH0n0ZKrqyt69eoFV1dX9OjRA/b29grjbGN+fn5wcnJCp06dUF1djUGDBjHGmPTq1Qs+Pj5wdHREWloa/Pz8GGNMzMzMMHz4cPj5+UEgEMDNzY0xxqRz584YOHAgnJycAAAuLi6MMbZ9lcVbevjwISorK2FlZYWRI0eia9euCuNsY5rMv5+fH5ydncHj8Vr1VTfWEdtYWZxpvp88eQJra+tWc9g2xqRHjx5wcXEBj8eDp6cnunfvrlHM1BjcntuJEyfg6OgIV1dXnDt3TmmcbSwqKgoXLlzAhQsXEBwcrDDGZPv27UhISMCFCxewcuVKhTFF68LlcmFra4vMzEyFMSYxMTHg8XgoKipCVlaWwhjbvmz7+/v7Y8aMGRg7diySkpKUxtnGNJn/qKgopKSkvNRX3RgTbW9jZfG2/P39MWvWrJfmsG2MyeHDh+Hk5ISioiLcuHFD45ipMbg9t/z8fOTk5ODmzZuwsrKS/7ZlirONdenSBeXl5RAIBKivr8df/vIXxhgTTfqyzY9JYWEhcnNz0djYCHNzcwwaNIgxxravsnhL8fHxyM/PR0FBAbhcLnx9fRXG2cY6Yv4NaRsri7OZb0XbgM121iRmcmQG5sqVK6zjbGMZGRnyPwsEAoUxJpr0ZZtfR/Rl299Y59+QtrGyOJt22u6ryRjGzOD23EQiEc6ePQuhUAgvLy+lcXViAwYMUBhTlM+ZM2de6ts2xnZdFK2frvuy7d9y3dr2bRtXJ6bOHLLpq63tpM7Ph7pzraidtvtqMoYxM7hzbvn5+aivr4eLiwv27NmjNK7tGJOioiIsWLAA9vb28nZMMSYXL16ESCRCbm6uvB1TjElDQwMaGxtRUFAgb8cUY/L06VMEBwejW7du8ksllMXbru/cuXNx69atVmMwxbUdY1JSUoIFCxbAy8tLnrMmMSYXL16EWCxGYWEhhg0bpjDGpLKyEmKxGPn5+dizZw/S09OVxtu6f/8+5syZg65du8pzZIoxefToEd555x3k5OTI22kSMzUGV9wEAgE+/PBDfP/9960ufGSKazvGxNPTE1FRUUhOTsbixYsVxpgsWbIES5Ysga2trfykNFOMyZUrV7Bo0SL87W9/k39VzxRjcvfuXQDAzZs3weVy24239OTJEyQnJ2P8+PGtLuxkiqsTa/6GlinG5NSpU+Dz+UhOTkZjY6PGMSYTJkyAra0trK2twefzFcaY3Lp1C4sXL8aHH34IiUSCP//8U2m8rZKSEpiZmcHHx0d+lwFTjMndu3fRtWtXDB06FDweT+OYqTG426+EQiH4fD6sra1x/vx5efFgims7xiQ9PR02NjZ4+vQpjh07hsWLFzPGmBw6dAgcDgfV1dXg8/lYvHgxY4xJbW0t+Hw+ZDIZ6urqFMaYlJeXg8/no6amhlW8JVtbW5SXl2PChAkoLCxUGtd2jMnkyZPx17/+VWsxJvv27YOnpyd4PB6uXr2KAQMGMMaYuLu7IyYmBlKpFP3798fbb7+tNN6Wh4cHDh8+DJlMhh49eiiMMfHy8kJMTIz8z5rGTI3B3X5FCCHaYHCHpYQQog1U3AghJomKGyHEJFFxI4SYJCpuhBCTRMWNEGKSqLgRQkySwV3ECwDZ2dlKHy5ICNENkUiEIUOG6DsNrTDI4mZhYYH+/fvrOw1CXjnN730wBXRYSggxSVTcCCEmiYobIcQkUXEjhJgkKm6EEJNExY0QYpJMprilXOit7xQIIQZEpevcqqqqwOVy5S9KVlVcXBzq6+vx/Plz/OMf/1BrGYQQwgar4paeno5ffvkFFhYW4HA4EIvFeOeddxS+T1GRCRMmwNHREREREUrbiUQitS4mNKULEAkhmmFV3GxsbBASEoL4+HjIZDIsWrQIlZWVKg/WuXNnbN++XelLVQD17lAoewS6q4EQDZnSDgKrc26DBw/G3r17sXz5cqxYsQLR0dFq3X+2ceNGSKVSJCcnq9yXEEJUwfqcG4fDgb29PaRSKZqamtQabNu2bWr1MyQpF3oj4K0ifadBCGkH6+I2f/58bNiwAQAwZ84cnSVECCHawKq4lZeXw9HREcuXLwfwYi+OmB7aKyWmhFVxS0tLw6NHj1rFVq1apZOECCFEG9r9QkEoFMLNzQ01NTXw9fWFr68viouLOyI3QghRW7t7blZWVhAIBBCJRBAIBACA9957T+eJEUKIJtrdczM3N0dQUBAWLVoEAJDJZCZ1LQwhxDSxvrc0MjISBQUFAIDc3FydJUQIIdrAurg5ODggICAAnTt3hp2dnS5zIoQQjbEubm5ubnBwcEBGRgacnZ11mRMhhGiM9UW8paWlEIvFCAkJoevcCCEGj/Wem7u7O5KTk5GYmIjTp0/rMifCEj3DTr9o/g0b6z03U75o19CuzNdHPi6p2Yjr0BEJ0S2DfxIv02/HjviN+Ur9Vg6nL4iI6WFd3CQSCaRSKfLz81FdXa3LnAghRGOsi1toaCiysrKwf/9+fPXVV7rM6ZXmkpqt7xQIMQmsi5uNjQ3S0tKwdOlS8Hg8tQZLSUlBXFwcYmJi1OpPjFi4HRBuR8WbdBhWxS0hIQEeHh7o0qULrl69Cg8PD7UGy8rKQnBwMKqqqtTqr3Uqfth0/sFU4dwX21wMvpiwXecOOi9o8PNFWOPIZDJZe42uXbuGuLg4zJo1C1wuFydPnkRkZKTKg0VFRWH16tWIjIzEunXrFLbLzs7GkFPj4DLuIlJdLOTx/seGM8by5ma06t/cLk42E26uP6mUI9MY7bVtbxymdprmqMo6lz2aovK42m6nKB8A8H8sYt2/bV82+aiyLmzmVds/c4bGVN5FwupSED8/P5w+fRqjRo2CTCbDzz//rNZgPj4+iImJafcOBwuLF4XlsT/zexraTr6yjaHuhlK1H9v2TO1UzjFcCGU92i5P0ctzNMlZk3Yv5fM4m3V/VV4EpM46s/1Z0sXPnCEwpYdisL7ObcaMGQgLCwOHw8Hs2bPVGmzChAlq9TNY4ULAgA5jHvsPQcqFl+OGdA0fkxe/xAwgx3ChvjMgWsT6nFtJSQmGDBmCwYMHo7CwUNd5aUzRXh8xcR1RoKgIGgVWxc3LywtpaWlwcXGBm5sbMjIy2u+kKQU/QC8VLT3/oLEtojovtirMg74Lf0fsSWp7DEPf+yUvY1Xc/Pz8YGtri1GjRmH48OGwsrLSdV6mxZB+0xtSLh2MCtSrpUPPuRHD0e4H3cDOJ6qCce/eSNeFqI91cXNwcIC1tTVkMhkcHBx0mZPmNNw70fdhm1EJF+IxwPhFhi7Q3hdhi/UdCnv37sXy5cuxYsUKREdH6zInxVQsWh31QWA7jrEInCWRAAAPHUlEQVR9MKnIE2PGes+Nw+HA3t4eUqkUTU1NusyJEINCRd44sS5u8+fPx4YNGwDAtM+5GfkJd2PbO+woVKBePayL25AhQzBkCP2AaBsVIwPH9pddB597JO1jdc4tOjoa6enpEAqFqKurw8WLF3Hw4EFd52bajHwPkRBDx6q4LV26FCKRCLt27cKOHTsglUrxwQcf6Do3k0eHSoToDuvDUn9/f3Tv3h11dXXw8/PDrVu3MHjwYF3mRgghalPpHQpXrlyBi4sLYmNjkZWVpaucCCFEYyoVN29vb3h6emLevHkoLy/XVU6EEKIx1sXtxx9/lD+HrbKyEhKJBJmZmTpLjBgX+taXGBrW59wOHTqE6dOn4/Lly6itrcWyZcvw3XffwdfXV5f5EWJUqMgbDtZ7bsOGDUNQUBB+//13lJaWwtraWqWng9y7dw8RERH4/PPP6ZCWEKJzrPfcxowZg4MHD2LhwoUQCoXYunUrHB0d2+0XHx+P69evQywWY+PGjTh37hzq6urQrVs3jRInHYP2RIixYvWCGAC4fPkyfvnlFwBAYGAgxowZo/JgiYmJ4HK5mDJlitJ22dnZ8vcoEEI6ljG/A6Il1ntu6enp+OyzzwAAX3/9tcrF7dKlSzh9+jR8fX3x559/Kn09oIWFhclMMCHG5JV8QYxYLIZIJIJUKkVjY6PKA40dOxZjx45VuR8hhKiDVXHLycnBpEmTsGzZMkilUqxZs0bXeRFCiEZYFbeiohcnladNmwYAKCkpoVuvCCEGjVVxe/fdd3WdByGEaJVKt18RQoixoOJGCDFJVNwIISaJihshxCSxvs6tI4lEIpO6mJAQYyESifSdgtawvv2KEEKMCR2WEkJMEhU3QohJouJGCDFJVNwIISaJihshxCRRcSOEmKRXsriVlZWhrKwMAoEAzVfCNP9fKpVCIpG81EcqleLatWutYg0NDRCLxbpPmNBcE5UZXHFjKjyA8uKjauHJzc3F7du3kZWVhcjISADA9u3bIZFI0NDQgG+//Vbe9uzZs0hISEB4eDgcHBwAANu2bYNQKMTOnTuxc+dOldZF1fVgoqhdy3XWxthsC0p77dT5ZdJMV3OtSV/6RWccDK64MRUegLn4aFJ4ZDIZOBwOpFKpPJaTk4M7d+60avfs2TNIpVJwOBz5GBwOB0VFRRgxYgTMzc1VWhe268FEUTumddZkbLZzyLadKr9M2tL2XGvSV5O5Ih3PIG+/Yio8wIvi09TUJP97e4Xn5s2bjMv38vJC586dwePx4OfnBwBYvHgxysrK0NjYiJCQEHnbSZMmwdbWFk+ePEF6ejqmTJmCRYsWobS0FLa2tliwYEG768NURNtbDyaK2ilbZ3XGZjOHqrRjmwuTjz76CMXFxXBycsLs2bOVtm3+uWm5x988BofD0Tg/TeaKdDxeeHh4uL6TaEkikcDDwwNubm7w8/OTvxu1X79+qK+vBwCMHz8elpaW8PT0xJtvvgkfHx9kZGSgT58+6Nu3L+rq6mBmZoaxY8cyvlv18OHDCAwMhKWlJWJiYuDn54e9e/di6tSpcHJyAp/Plxe9Xbt2YdiwYTA3N0dKSoq87bRp0+Dk5CTvzyQ3NxcSiQSurq4ICAiQ53L79m24uLjgrbfegqWlJerq6tC1a1f069cPeXl56NOnD+PyFLWTSqXIzc2Fh4cHxo0bBysrK43GZloeE7bt2ObCJCoqChMnTkRiYiLu3LmjcK6trKzQvXt3dO/eHQ0NDejXrx/8/PxQXV0NHo+HpqYm9OvXT+vbie0ckI5ncIelSUlJcHFxgYODA+Lj4+VxPp+P/v37w9vbG7GxsQCA6OhoNDU1wc7ODvfu3ZO38/b2hre3d6v+LVlYWODMmTM4c+aM/FCHKdY23vy6QUVt23r99dfx+uuvw8zMDIcOHQIA2NnZyYvl/v37AQA2NjawsbGBubk58vPzFS5PUTt7e3v5YVvzOJqMzbQ8Jmzbsc1F0TqnpqbC1dVV6VxfunQJv/32Gw4cOIDAwEAAwPHjx5GXl4dTp07JY5rkp8lckY5ncIelzYWjqamJsci0jLONtTVnzhxIJBI8fvwYvXr1UhhTtW1bzSedORwOzMzMAABXr15lFWNy9epV+eFVp06dXopzOBx5/Nq1ay+1Zerfcmxl7RTlwybv5lyYxmgZY+Lu7g6BQAAnJydMmDBBYTtnZ2dUVlZCIpGgpKQEAwcOhJOT00sxZfm1XF9Fc9Uca7nt2MwV6XgGtzWmT58Oa2tr+asElcXZxto6fPgw/ud//ge//fYbamtrMWjQIMaYqm3bWrJkCYAXhz05OTkqxZh07doVAMDj8TB+/HjG+Lhx4wAAEyZMgLu7O+rr61FRUaFxjO36tZd3c35MMSb+/v4AXhwCx8fHY9WqVYztamtrYW5ujjfffFN+nowpxmTq1KmwtbVFQUEBuFyuxjFiGAxua5w4cQKOjo5wdXXFuXPnlMbZxtpydnZGfHw8nj17Jj/UZIqp2ratqKgoJCQkICUlBcHBwSrFmPj7+8Pf3x+jR49GUlKS0viJEyfA5XJha2vbam7UjbFdP7Z5K1qXtk6cOIGTJ0/i9OnTSud6/Pjx8Pf3x9ixY3H58mWFMSYxMTHg8XgoLCzEjRs3NI4Rw2Bwe252dnb47rvvIJVKYWtrqzTONtZWv379MHz48HZjqrZt6+2338adO3dQX1+Pc+fOYeLEiaxjTE6cOCE/BGr5QWeKazJfbOZQ0fqxzVvRurQ1YMAAVnN98uTJl5bHFGPS/MuqZTtNYsRAyAzMlStXWMfZxtr6448/ZLGxsbJLly4pjanatq2MjAz5nwUCgUoxJrqYG3XnUCbTLG+2Y7Cda03G6Ii5Ih3P4C4FEYlEOHv2LIRCIby8vJTG2cbacnBwwKBBg8DhcJCYmIjBgwczxlRty7QuZ86cgVAoxIABA1SKddTcqDuHbdupmjfbMVSZa22sh67minQ8gzvnVlBQgOfPn6Nbt27ySz4UxdnGmBQXFyM9Pb3dmKptW5JIJFiwYAG8vLzkubCNMenduzeCg4Ph5eWFPXv2KI1rO8bk+PHjEIvFKCwsVNqusrISwcHBsLKykq8fU4zJ4cOHcezYMWRlZSm9EPfXX39FYGBgqzlkiinqa25uDk9Pz1Z91Y0Rw2Bwxa2srAx/+9vfcOPGDSxcuFBpnG2srY8++ghZWVmQSqXyNkwxVdu2tWvXLvD5fCQnJ8svC2EbU8bT0xP29vas4tqOtTRlyhTcv38fr732GlauXKmwXWFhIQDg7t278vliijF5//33IZFIkJGRAU9PT6Xttm3bhqSkJDQ2NiqMMfnggw8wefJk/PTTT7hy5YrGMWIYDO4FMVFRUbC3t0dNTQ3s7e2xePFihXG2sbaampqQmZmJnJwcmJubIyQkhDGmatu2cnJy5JeJNP+ZbYzJoUOH5LcXSSQSfPTRRwrj2o4xWb58Ofr06QM7OztwOBzGuQaAffv2wczMDI2NjTAzM8PixYsZY0wOHz6MTp06oampCRUVFVizZo3CMd5++2306NFDaYzJgQMHYG5uDh8fH7zxxhvgcDgaxYhhMLjiRggh2mBwh6WEEKINVNxeISKRCDU1NXpfhiYEAoHexibGhYqbkVm3bp3afc+cOYO7d++q1ffJkyf4/vvv5cs4fvy42oUmMTERO3bsaBWrqKhAQkICRCIRPvvsM6xfv17+uKOWWj7jTxXKvs1ttmvXLty/fx98Ph/Pnz9XaxxiOAzuDoVXVXZ2No4fP46nT5/i3//+N/bu3Qvgxd0CS5cuxeeffw4rKyvU1taipqYGX3zxBSwsLDB+/HjU1dUhKysLlpaWGDp0qPxOgfz8fPD5fDx79gwLFixAZmYmzM3NMXLkSAAvPszPnj0DAMydOxd//PEHUlNTIRaL8emnnyI2Nha1tbWwtLTEvHnz8PDhQ1RWVsLc3Bxdu3ZFQ0MDdu/ejerqashkMmzatAkzZsxAQEAAnjx5AqZLKG/evIkHDx689Ky+2NhYfPjhhzhy5AgCAgIwevRoVFVVITs7W/5wgszMTOTn56O4uBjx8fGQSqVwd3eHj48P9u3bh4kTJyIpKQlvvPEGSkpK0LdvX5SUlODrr79GSUkJTpw40WqevLy8Ws1Ps7feegvHjh3DokWLtL2ZSQeiPTcD0fyE140bN6KoqAjdunVDWFgYqqurce7cOYwaNQobNmyAhYUFkpKSsGDBAkRERODs2bMAXtyr+a9//Qu//vqrfJl2dnZ49913MWjQINy4cQNvvvnmS4/+mTdvHlauXIljx47h559/RkREBEJCQpCYmIiHDx/C09MTU6ZMkbdvuYyGhgaUl5cjLCwMrq6uuHv3Ltzc3LBy5UrU1dUxrucbb7yBoKCgl+LND/8sLi7GkCFDALy4gPett96St/H19UXfvn3x4MEDCAQCWFtby2/YHzNmDObMmQMbGxt8/PHHkEgkWLFixUsPoGw5T23np5mXlxdyc3NZbTdiuKi4GQiJRAIOh4P6+np06dKF8ZICDocDLpcrfxos8H/P+reysoKZmVmrp9CeOnUKpaWl6N+/v/wptW3JZDKIRCLweLxWcalUipCQEPTt2xcRERHyAtFyGW2XKZVK5Q+dbPn4n/z8fKxevRrHjh1TuP7N4/fu3RvZ2dkAgKqqKuzfvx/ffPMNVq9eLT/X19TUhBEjRmDdunXyQmtjYwMAsLa2ls8HAPl8NWs5T23npyV6fJHxoy1oIGbOnImwsDA8f/4cYWFhKC8vx5YtW+Ds7IyJEyfiP//5D27evAmJRILJkydj8+bNSExMxOTJk1FbW8u4TBcXF1y7dg1mZmawsrKCm5sbjh49Kj8sBV5cCwYAK1asQF5eHjZt2oSmpiasXbsWBw8eRH19PXr37i1/nE/zMnr16gUrKys4OTlh8+bN4HA4Cm/B6tevH6KiopSuf7du3VBbW4u5c+ciPDwciYmJqK+vx5o1a1o9M6+iogLu7u5ISkrC3bt30bNnT3Tr1k2luVY0P80PCSguLoa3t7dayySGg65ze4Xt2rUL06ZNM4h7ImtqanD+/HnMmTNH36kgJiYG8+fPV/rkX2L4qLgRQkwSnXMjhJgkKm6EEJNExY0QYpKouBFCTBIVN0KISaLiRggxSf8PxsmNHB11nv4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x205.714 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ylabel = '$\\log_2$(odds ratio)'\n",
    "ymin, ymax = -3, 3\n",
    "\n",
    "n = 44\n",
    "plotDf1 = plotData2.iloc[:n]\n",
    "plotDf2 = plotData2.iloc[n+1:]\n",
    "m = n - len(plotDf2)\n",
    "emptyRow = plotData2.iloc[0].copy()\n",
    "emptyRow[:] = np.nan\n",
    "emptyRow.name = ''\n",
    "plotDf2 = plotDf2.copy().append(m*[emptyRow])\n",
    "\n",
    "set_fontsize_paper()\n",
    "figsizeCoeff = 0.6\n",
    "fontSize = 16\n",
    "aspectratio = 1.4\n",
    "imagesize = 4\n",
    "linewidth1 = 0.5\n",
    "figsize = (imagesize, imagesize/aspectratio)\n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(2, 1, figsize=figsize)\n",
    "plt.subplots_adjust(hspace=0.5)\n",
    "w = 0.6\n",
    "\n",
    "for i, ax in enumerate(axes):\n",
    "    if i == 0:\n",
    "        plotDf3 = plotDf1\n",
    "    elif i == 1:\n",
    "        plotDf3 = plotDf2\n",
    "#     norm = matplotlib.colors.Normalize(vmin=-vmax, vmax=vmax)\n",
    "#     colors = cmap(norm(plotDf3.values))\n",
    "    plotDf3.plot.bar(ax=ax, width=w, linewidth=0, colors=paletteCat)\n",
    "    ax.set_ylabel(ylabel)\n",
    "    ax.set_ylim(ymin, ymax)\n",
    "    ax.xaxis.grid(False)\n",
    "    family = 'Liberation Mono'\n",
    "    ticks_font = FontProperties(family=family, size='x-small')\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    if i == 0:\n",
    "        ax.legend(loc='upper left', bbox_to_anchor=(1, 1), title='stop codon\\ncontext')\n",
    "    if i == 1:\n",
    "        ax.set_xlabel('codon at position -1 (C-terminal)')\n",
    "        ax.legend_.remove()\n",
    "\n",
    "\n",
    "title = terminus + '-terminal, ' + speciesSetName\n",
    "axes[0].set_title(title)\n",
    "filenamePlot = outputFilePrefix + '_composition_bias_codon_' + terminus + 'terminal_barplot_stopCodonContext'\n",
    "print(filenamePlot)\n",
    "# fig.savefig(os.path.join(plotsFolder, speciesSetName, filenamePlot + '.svg'), dpi=400, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add excluded overlapping start codon NNA-UGA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/lserrano/mweber/Analysis_data/C-terminal/overlapping_start_codon/TGA_non_overlapping_start_codon/TGA_non_overlapping_start_codon_subset_stopCodon_TGA_oddsRatioDfCodon.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCG</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <th>-1000000</th>\n",
       "      <th>count</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            AAA  AAC  AAG  AAT  ACA  ACC  ACG  \\\n",
       "terminus position from terminus observable                                      \n",
       "C        -1000000               count       NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                            ACT  AGA  AGC ...   TCG  TCT  TGA  \\\n",
       "terminus position from terminus observable                ...                   \n",
       "C        -1000000               count       NaN  NaN  NaN ...   NaN  NaN  NaN   \n",
       "\n",
       "                                            TGC  TGG  TGT  TTA  TTC  TTG  TTT  \n",
       "terminus position from terminus observable                                     \n",
       "C        -1000000               count       NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "\n",
       "[1 rows x 64 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "speciesSetName = 'TGA_non_overlapping_start_codon'\n",
    "filename = (analysisCtermPath / 'overlapping_start_codon' / speciesSetName /\n",
    "            '{}_subset_stopCodon_TGA_oddsRatioDfCodon.csv'.format(speciesSetName))\n",
    "print(filename)\n",
    "oddsRatioDfCodonOverlapping = pd.read_csv(filename, index_col=[0,1])\n",
    "oddsRatioDfCodonOverlapping = oddsRatioDfCodonOverlapping.reset_index()\\\n",
    "                   .set_index(['terminus', 'position from terminus', 'observable']).sort_index()\n",
    "oddsRatioDfCodonOverlapping.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDfCodonOverlapping.T.copy()\n",
    "for col in df.columns:\n",
    "    df[col] = pd.to_numeric(df[col], errors='ignore')\n",
    "oddsRatioDfCodonOverlapping = df.T.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfCodonOverlapping['stop_codon'] = 'UGA no start overlap'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "      <th>stop_codon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"10\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-1000000</th>\n",
       "      <th>count</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>24813</td>\n",
       "      <td>27951</td>\n",
       "      <td>38542</td>\n",
       "      <td>16841</td>\n",
       "      <td>9199</td>\n",
       "      <td>46019</td>\n",
       "      <td>28412</td>\n",
       "      <td>7777</td>\n",
       "      <td>5524</td>\n",
       "      <td>27344</td>\n",
       "      <td>...</td>\n",
       "      <td>6929</td>\n",
       "      <td>0</td>\n",
       "      <td>12788</td>\n",
       "      <td>24724</td>\n",
       "      <td>4562</td>\n",
       "      <td>9638</td>\n",
       "      <td>40972</td>\n",
       "      <td>20425</td>\n",
       "      <td>18550</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.283761</td>\n",
       "      <td>-0.171512</td>\n",
       "      <td>0.112015</td>\n",
       "      <td>0.00262561</td>\n",
       "      <td>0.178584</td>\n",
       "      <td>-0.133748</td>\n",
       "      <td>-0.0282119</td>\n",
       "      <td>0.115167</td>\n",
       "      <td>0.361864</td>\n",
       "      <td>0.0173907</td>\n",
       "      <td>...</td>\n",
       "      <td>0.119672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0899205</td>\n",
       "      <td>0.0020634</td>\n",
       "      <td>0.253322</td>\n",
       "      <td>0.156387</td>\n",
       "      <td>-0.117476</td>\n",
       "      <td>0.131957</td>\n",
       "      <td>0.0382716</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.21736</td>\n",
       "      <td>0.887912</td>\n",
       "      <td>1.08074</td>\n",
       "      <td>1.00182</td>\n",
       "      <td>1.13177</td>\n",
       "      <td>0.91146</td>\n",
       "      <td>0.980635</td>\n",
       "      <td>1.0831</td>\n",
       "      <td>1.28508</td>\n",
       "      <td>1.01213</td>\n",
       "      <td>...</td>\n",
       "      <td>1.08649</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.06431</td>\n",
       "      <td>1.00143</td>\n",
       "      <td>1.19195</td>\n",
       "      <td>1.11449</td>\n",
       "      <td>0.921799</td>\n",
       "      <td>1.09578</td>\n",
       "      <td>1.02688</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>1.3614e-194</td>\n",
       "      <td>1.56538e-89</td>\n",
       "      <td>5.54769e-50</td>\n",
       "      <td>0.81282</td>\n",
       "      <td>4.91355e-31</td>\n",
       "      <td>6.25718e-88</td>\n",
       "      <td>0.00107341</td>\n",
       "      <td>4.4334e-12</td>\n",
       "      <td>3.38751e-71</td>\n",
       "      <td>0.048676</td>\n",
       "      <td>...</td>\n",
       "      <td>1.11372e-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.91827e-12</td>\n",
       "      <td>0.82219</td>\n",
       "      <td>1.44055e-30</td>\n",
       "      <td>2.17865e-25</td>\n",
       "      <td>5.98851e-61</td>\n",
       "      <td>2.00715e-37</td>\n",
       "      <td>0.000352808</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           AAA          AAC  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count                      NaN          NaN   \n",
       "                                log2OddsRatio              NaN          NaN   \n",
       "                                oddsRatio                  NaN          NaN   \n",
       "                                pvalue                     NaN          NaN   \n",
       "                                terminus                   NaN          NaN   \n",
       "         -20                    BH_multiple_tests         True         True   \n",
       "                                count                    24813        27951   \n",
       "                                log2OddsRatio         0.283761    -0.171512   \n",
       "                                oddsRatio              1.21736     0.887912   \n",
       "                                pvalue             1.3614e-194  1.56538e-89   \n",
       "\n",
       "                                                           AAG         AAT  \\\n",
       "terminus position from terminus observable                                   \n",
       "C        -1000000               count                      NaN         NaN   \n",
       "                                log2OddsRatio              NaN         NaN   \n",
       "                                oddsRatio                  NaN         NaN   \n",
       "                                pvalue                     NaN         NaN   \n",
       "                                terminus                   NaN         NaN   \n",
       "         -20                    BH_multiple_tests         True       False   \n",
       "                                count                    38542       16841   \n",
       "                                log2OddsRatio         0.112015  0.00262561   \n",
       "                                oddsRatio              1.08074     1.00182   \n",
       "                                pvalue             5.54769e-50     0.81282   \n",
       "\n",
       "                                                           ACA          ACC  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count                      NaN          NaN   \n",
       "                                log2OddsRatio              NaN          NaN   \n",
       "                                oddsRatio                  NaN          NaN   \n",
       "                                pvalue                     NaN          NaN   \n",
       "                                terminus                   NaN          NaN   \n",
       "         -20                    BH_multiple_tests         True         True   \n",
       "                                count                     9199        46019   \n",
       "                                log2OddsRatio         0.178584    -0.133748   \n",
       "                                oddsRatio              1.13177      0.91146   \n",
       "                                pvalue             4.91355e-31  6.25718e-88   \n",
       "\n",
       "                                                          ACG         ACT  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -1000000               count                     NaN         NaN   \n",
       "                                log2OddsRatio             NaN         NaN   \n",
       "                                oddsRatio                 NaN         NaN   \n",
       "                                pvalue                    NaN         NaN   \n",
       "                                terminus                  NaN         NaN   \n",
       "         -20                    BH_multiple_tests        True        True   \n",
       "                                count                   28412        7777   \n",
       "                                log2OddsRatio      -0.0282119    0.115167   \n",
       "                                oddsRatio            0.980635      1.0831   \n",
       "                                pvalue             0.00107341  4.4334e-12   \n",
       "\n",
       "                                                           AGA        AGC  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -1000000               count                      NaN        NaN   \n",
       "                                log2OddsRatio              NaN        NaN   \n",
       "                                oddsRatio                  NaN        NaN   \n",
       "                                pvalue                     NaN        NaN   \n",
       "                                terminus                   NaN        NaN   \n",
       "         -20                    BH_multiple_tests         True      False   \n",
       "                                count                     5524      27344   \n",
       "                                log2OddsRatio         0.361864  0.0173907   \n",
       "                                oddsRatio              1.28508    1.01213   \n",
       "                                pvalue             3.38751e-71   0.048676   \n",
       "\n",
       "                                                           ...           \\\n",
       "terminus position from terminus observable                 ...            \n",
       "C        -1000000               count                      ...            \n",
       "                                log2OddsRatio              ...            \n",
       "                                oddsRatio                  ...            \n",
       "                                pvalue                     ...            \n",
       "                                terminus                   ...            \n",
       "         -20                    BH_multiple_tests          ...            \n",
       "                                count                      ...            \n",
       "                                log2OddsRatio              ...            \n",
       "                                oddsRatio                  ...            \n",
       "                                pvalue                     ...            \n",
       "\n",
       "                                                           TCT  TGA  \\\n",
       "terminus position from terminus observable                            \n",
       "C        -1000000               count                      NaN  NaN   \n",
       "                                log2OddsRatio              NaN  NaN   \n",
       "                                oddsRatio                  NaN  NaN   \n",
       "                                pvalue                     NaN  NaN   \n",
       "                                terminus                   NaN  NaN   \n",
       "         -20                    BH_multiple_tests         True  NaN   \n",
       "                                count                     6929    0   \n",
       "                                log2OddsRatio         0.119672  NaN   \n",
       "                                oddsRatio              1.08649  NaN   \n",
       "                                pvalue             1.11372e-11  NaN   \n",
       "\n",
       "                                                           TGC        TGG  \\\n",
       "terminus position from terminus observable                                  \n",
       "C        -1000000               count                      NaN        NaN   \n",
       "                                log2OddsRatio              NaN        NaN   \n",
       "                                oddsRatio                  NaN        NaN   \n",
       "                                pvalue                     NaN        NaN   \n",
       "                                terminus                   NaN        NaN   \n",
       "         -20                    BH_multiple_tests         True      False   \n",
       "                                count                    12788      24724   \n",
       "                                log2OddsRatio        0.0899205  0.0020634   \n",
       "                                oddsRatio              1.06431    1.00143   \n",
       "                                pvalue             3.91827e-12    0.82219   \n",
       "\n",
       "                                                           TGT          TTA  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count                      NaN          NaN   \n",
       "                                log2OddsRatio              NaN          NaN   \n",
       "                                oddsRatio                  NaN          NaN   \n",
       "                                pvalue                     NaN          NaN   \n",
       "                                terminus                   NaN          NaN   \n",
       "         -20                    BH_multiple_tests         True         True   \n",
       "                                count                     4562         9638   \n",
       "                                log2OddsRatio         0.253322     0.156387   \n",
       "                                oddsRatio              1.19195      1.11449   \n",
       "                                pvalue             1.44055e-30  2.17865e-25   \n",
       "\n",
       "                                                           TTC          TTG  \\\n",
       "terminus position from terminus observable                                    \n",
       "C        -1000000               count                      NaN          NaN   \n",
       "                                log2OddsRatio              NaN          NaN   \n",
       "                                oddsRatio                  NaN          NaN   \n",
       "                                pvalue                     NaN          NaN   \n",
       "                                terminus                   NaN          NaN   \n",
       "         -20                    BH_multiple_tests         True         True   \n",
       "                                count                    40972        20425   \n",
       "                                log2OddsRatio        -0.117476     0.131957   \n",
       "                                oddsRatio             0.921799      1.09578   \n",
       "                                pvalue             5.98851e-61  2.00715e-37   \n",
       "\n",
       "                                                           TTT  \\\n",
       "terminus position from terminus observable                       \n",
       "C        -1000000               count                      NaN   \n",
       "                                log2OddsRatio              NaN   \n",
       "                                oddsRatio                  NaN   \n",
       "                                pvalue                     NaN   \n",
       "                                terminus                   NaN   \n",
       "         -20                    BH_multiple_tests         True   \n",
       "                                count                    18550   \n",
       "                                log2OddsRatio        0.0382716   \n",
       "                                oddsRatio              1.02688   \n",
       "                                pvalue             0.000352808   \n",
       "\n",
       "                                                             stop_codon  \n",
       "terminus position from terminus observable                               \n",
       "C        -1000000               count              UGA no start overlap  \n",
       "                                log2OddsRatio      UGA no start overlap  \n",
       "                                oddsRatio          UGA no start overlap  \n",
       "                                pvalue             UGA no start overlap  \n",
       "                                terminus           UGA no start overlap  \n",
       "         -20                    BH_multiple_tests  UGA no start overlap  \n",
       "                                count              UGA no start overlap  \n",
       "                                log2OddsRatio      UGA no start overlap  \n",
       "                                oddsRatio          UGA no start overlap  \n",
       "                                pvalue             UGA no start overlap  \n",
       "\n",
       "[10 rows x 65 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCodonOverlapping.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = oddsRatioCodonTaxon[oddsRatioCodonTaxon['taxaid'] == taxonId]\\\n",
    "           .drop(['taxaid', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes'], axis=1)\n",
    "plotData = pd.concat([plotData, oddsRatioDfCodonOverlapping])\n",
    "plotData = plotData.xs(terminus, level='terminus')\n",
    "plotData = select_terminal_positions_in_index(plotData, terminus)\n",
    "plotDataTemp = plotData.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAC</th>\n",
       "      <th>AAG</th>\n",
       "      <th>AAT</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACG</th>\n",
       "      <th>ACT</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>TGC</th>\n",
       "      <th>TGG</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TTT</th>\n",
       "      <th>stop_codon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>position from terminus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.45502</td>\n",
       "      <td>0.185648</td>\n",
       "      <td>0.779589</td>\n",
       "      <td>0.288478</td>\n",
       "      <td>-0.511378</td>\n",
       "      <td>-1.32289</td>\n",
       "      <td>-1.63726</td>\n",
       "      <td>-0.802509</td>\n",
       "      <td>1.24263</td>\n",
       "      <td>0.319846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.238899</td>\n",
       "      <td>0.39357</td>\n",
       "      <td>-0.109323</td>\n",
       "      <td>-0.213098</td>\n",
       "      <td>-0.0180203</td>\n",
       "      <td>0.102137</td>\n",
       "      <td>-0.0616251</td>\n",
       "      <td>-0.524969</td>\n",
       "      <td>0.0832374</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.55131</td>\n",
       "      <td>0.232382</td>\n",
       "      <td>1.26658</td>\n",
       "      <td>0.374393</td>\n",
       "      <td>-1.31151</td>\n",
       "      <td>-2.09353</td>\n",
       "      <td>-1.73563</td>\n",
       "      <td>-1.07142</td>\n",
       "      <td>1.06892</td>\n",
       "      <td>0.0165654</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0723837</td>\n",
       "      <td>0.309546</td>\n",
       "      <td>-0.405192</td>\n",
       "      <td>0.144922</td>\n",
       "      <td>-0.168143</td>\n",
       "      <td>-0.0792</td>\n",
       "      <td>0.1338</td>\n",
       "      <td>-0.402319</td>\n",
       "      <td>0.213132</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.6145</td>\n",
       "      <td>0.0166325</td>\n",
       "      <td>1.18059</td>\n",
       "      <td>0.297567</td>\n",
       "      <td>-0.862987</td>\n",
       "      <td>-1.80446</td>\n",
       "      <td>-1.43651</td>\n",
       "      <td>-0.820776</td>\n",
       "      <td>1.42268</td>\n",
       "      <td>0.214125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128262</td>\n",
       "      <td>0.681705</td>\n",
       "      <td>-0.531884</td>\n",
       "      <td>0.04579</td>\n",
       "      <td>0.102677</td>\n",
       "      <td>0.061261</td>\n",
       "      <td>-0.166365</td>\n",
       "      <td>-0.164196</td>\n",
       "      <td>0.23639</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>1.11719</td>\n",
       "      <td>0.213708</td>\n",
       "      <td>-0.181898</td>\n",
       "      <td>0.078626</td>\n",
       "      <td>0.617179</td>\n",
       "      <td>-0.897764</td>\n",
       "      <td>-1.66614</td>\n",
       "      <td>-0.274741</td>\n",
       "      <td>1.47539</td>\n",
       "      <td>0.53796</td>\n",
       "      <td>...</td>\n",
       "      <td>0.671197</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.154971</td>\n",
       "      <td>-0.688177</td>\n",
       "      <td>0.142402</td>\n",
       "      <td>0.641371</td>\n",
       "      <td>-0.150502</td>\n",
       "      <td>-0.975846</td>\n",
       "      <td>-0.368551</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-1</th>\n",
       "      <td>0.46374</td>\n",
       "      <td>0.45814</td>\n",
       "      <td>-0.204468</td>\n",
       "      <td>0.311231</td>\n",
       "      <td>-0.387048</td>\n",
       "      <td>-0.652191</td>\n",
       "      <td>-1.78815</td>\n",
       "      <td>-0.0196467</td>\n",
       "      <td>0.986246</td>\n",
       "      <td>0.787171</td>\n",
       "      <td>...</td>\n",
       "      <td>0.89399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.398294</td>\n",
       "      <td>-0.52298</td>\n",
       "      <td>0.413462</td>\n",
       "      <td>-0.221984</td>\n",
       "      <td>0.0947938</td>\n",
       "      <td>-0.888877</td>\n",
       "      <td>-0.111094</td>\n",
       "      <td>UGA no start overlap</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            AAA        AAC       AAG       AAT       ACA  \\\n",
       "position from terminus                                                     \n",
       "-1                      1.45502   0.185648  0.779589  0.288478 -0.511378   \n",
       "-1                      1.55131   0.232382   1.26658  0.374393  -1.31151   \n",
       "-1                       1.6145  0.0166325   1.18059  0.297567 -0.862987   \n",
       "-1                      1.11719   0.213708 -0.181898  0.078626  0.617179   \n",
       "-1                      0.46374    0.45814 -0.204468  0.311231 -0.387048   \n",
       "\n",
       "                             ACC      ACG        ACT       AGA        AGC  \\\n",
       "position from terminus                                                      \n",
       "-1                      -1.32289 -1.63726  -0.802509   1.24263   0.319846   \n",
       "-1                      -2.09353 -1.73563   -1.07142   1.06892  0.0165654   \n",
       "-1                      -1.80446 -1.43651  -0.820776   1.42268   0.214125   \n",
       "-1                     -0.897764 -1.66614  -0.274741   1.47539    0.53796   \n",
       "-1                     -0.652191 -1.78815 -0.0196467  0.986246   0.787171   \n",
       "\n",
       "                                ...                 TCT       TGA       TGC  \\\n",
       "position from terminus          ...                                           \n",
       "-1                              ...            0.238899   0.39357 -0.109323   \n",
       "-1                              ...           0.0723837  0.309546 -0.405192   \n",
       "-1                              ...            0.128262  0.681705 -0.531884   \n",
       "-1                              ...            0.671197       NaN  0.154971   \n",
       "-1                              ...             0.89399       NaN  0.398294   \n",
       "\n",
       "                             TGG        TGT       TTA        TTC       TTG  \\\n",
       "position from terminus                                                       \n",
       "-1                     -0.213098 -0.0180203  0.102137 -0.0616251 -0.524969   \n",
       "-1                      0.144922  -0.168143   -0.0792     0.1338 -0.402319   \n",
       "-1                       0.04579   0.102677  0.061261  -0.166365 -0.164196   \n",
       "-1                     -0.688177   0.142402  0.641371  -0.150502 -0.975846   \n",
       "-1                      -0.52298   0.413462 -0.221984  0.0947938 -0.888877   \n",
       "\n",
       "                              TTT            stop_codon  \n",
       "position from terminus                                   \n",
       "-1                      0.0832374                   all  \n",
       "-1                       0.213132                   TAA  \n",
       "-1                        0.23639                   TAG  \n",
       "-1                      -0.368551                   TGA  \n",
       "-1                      -0.111094  UGA no start overlap  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotData.xs('log2OddsRatio', level='observable').loc[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = plotData.xs('log2OddsRatio', level='observable')\n",
    "plotData = plotData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "plotData = plotData[plotData.columns].astype(float)\n",
    "plotData = plotData.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDataCopy = plotData.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskData = plotDataTemp\n",
    "maskData = maskData.xs('BH_multiple_tests', level='observable')\n",
    "maskData = maskData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "# Mask both rejected cases and null values\n",
    "maskData = (maskData != True) & (maskData != 'True')\n",
    "maskData = maskData.transpose()\n",
    "maskData.sum().sum()\n",
    "# maskData.shape[0]*maskData.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskDataCopy = maskData.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>stop_codon</th>\n",
       "      <th>all</th>\n",
       "      <th>TAA</th>\n",
       "      <th>TAG</th>\n",
       "      <th>TGA</th>\n",
       "      <th>UGA no start overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACG</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACT</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGA</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGC</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "stop_codon    all    TAA    TAG    TGA  UGA no start overlap\n",
       "AAA         False  False  False  False                 False\n",
       "AAC         False  False   True  False                 False\n",
       "AAG         False  False  False  False                 False\n",
       "AAT         False  False  False  False                 False\n",
       "ACA         False  False  False  False                 False\n",
       "ACC         False  False  False  False                 False\n",
       "ACG         False  False  False  False                 False\n",
       "ACT         False  False  False  False                  True\n",
       "AGA         False  False  False  False                 False\n",
       "AGC         False   True  False  False                 False"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maskData.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statannot import add_stat_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p-value annotation legend:\n",
      "ns: 5.00e-02 < p <= 1.00e+00\n",
      "*: 1.00e-02 < p <= 5.00e-02\n",
      "**: 1.00e-03 < p <= 1.00e-02\n",
      "***: 1.00e-04 < p <= 1.00e-03\n",
      "****: p <= 1.00e-04\n",
      "\n",
      "$^{\\dagger}$UGA_False v.s. $^{\\dagger}$UGA_True: t-test independent samples, P_val=1.183e-01 stat=-1.585e+00\n",
      "UGA_False v.s. UGA_True: t-test independent samples, P_val=2.190e-08 stat=-6.462e+00\n",
      "TGA_non_overlapping_start_codon_comparison_bias_NNA_codons\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUQAAADVCAYAAAAvp4mUAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3XlcVOX+B/APIDOA4IIYoYgKaLhcF8xwSbuaBqGZIm4oZIlmN7kqoYFLmsqNDHLBXEqNxQU1yTQXXqlFpbkl5YbKpkAuxCKGA8MI5/cH4/wyGIZl5swAn/dfwjlzni8z+OE55zzneYwEQRBAREQw1ncBRESGgoFIRKTEQCQiUmIgEhEpMRCJiJQYiERESgxEIiIlBiIRkRIDkYhIqZm+C6jKb7/9BqlUqu8yiCqRy+Xo06ePvssgHTHIQJRKpejWrZu+yyCqJDk5Wd8lkA7xlJmISImBSESkxECsxt27d7Fz506cOnUKly5dQkREBFJSUvDw4UNkZWUhJCQEOTk5kMvlyMrKQnBwMLKzs/VdNhHVkUFeQzQUaWlp2Lt3L1q0aIEhQ4bg8OHDyMjIQFBQEE6dOoUffvgBRkZGCAwMxIEDB3Dy5Em0bNkS8+bNg7m5ub7LJ5EoFApkZ2ejpKRE36VQDZiZmcHe3h6mpqaVtjEQq9G9e3eMHj0aBQUFGDp0KFq2bIlbt26hU6dOuHv3LpYuXYqUlBTY2NigQ4cO+Pjjj3H79m2GYROTnZ0NKysrdOrUCUZGRvouh6ohCALy8vKQnZ2Nzp07V9puZIgTxCYnJ/MuMxmkqn43k5OT4eLiwjBsIARBwPXr16vMGF5DJNIChmHDUd1nxUAkIlJiIBLpQEpKCmbNmgVfX1+MHz8e69evR1VXp86fP4/r169rrd3Q0FDcuXOnVq85e/Ys5s+fr5X2f/zxRwQHB9frGMOHD0dMTIzq67S0NPj6+gIAgoODMWfOnKf2Hzx48FNfL1u2DGPHjq1T2wxEIi17+PAhAgMDsWjRIsTGxmLv3r24efMm4uLiKu27f/9+5OTkaK3txYsXo127dlo7nr5ERUUhPT29ym2//vorDhw4UOW24uJiXLx4EU5OTjh79myt222Sd5nj4+Nx7NgxfZcBDw8PeHl56bsM0rITJ07Azc0NnTp1AgCYmJjg448/rjTM48qVK/jpp59w9epVODs74/fff0dUVBSMjY3Rr18/BAUFITIyEklJSZDJZAgNDUVwcDDs7OyQnZ2NUaNGISUlBdeuXcO///1vBAYGwtfXF8uXL8eRI0eQnZ2NvLw83LlzByEhIRgyZAiOHTuGnTt3qmpYt25dlT9DeXk5Vq1ahUuXLkGhUCAgIAAjRoxAWFgYfv31VwDA6NGj8cYbbyAtLQ2LFi2Cubk5zM3N0bJlSwDAwYMHER0dDYlEgk6dOmHFihU4dOgQEhMTUVJSgszMTMycObPK/wPBwcEIDg7G7t27K2177733EBkZiQEDBuDZZ599atvRo0cxcOBADB06FDt37oSbm1vNPzg00R7isWPHcPPmTb3WcPPmTYMIZdK+nJwcdOjQ4anvNW/eHBKJ5Knv9ezZE0OGDMGCBQtgYWGByMhIREVFYffu3bh//z5OnToFAHB0dERcXBykUimysrIQGhqKLVu2YN26dQgODsa+ffvw1VdfVapDIpFg69atWLx4MaKiogAAt27dwueff47Y2Fh07twZP//8c5U/w4kTJ1BQUICvvvoKW7duxeXLl/H9998jOzsbe/fuxa5du/Dtt9/ixo0bWLduHf773/8iKioKffv2BQAUFBQgMjIS0dHR2L17N6ysrLBnzx4AQFFREbZs2YJNmzbh888/r7L9l156CV27dsUXX3xRadszzzyDuXPnYvHixZW27du3DxMmTMCgQYNw7do13L9/v8rjq9Mke4gA0LVrV7UfhhhmzZqlt7ZJt9q1a4dr16499b2srCzcu3cPN27cQEJCAgAgPDxctT0zMxP5+fmq34tHjx4hKysLAJ4aL9ehQwdYWVlBIpHAxsYGrVq1AlD1ndMnw0qeffZZlJaWAgDatGmD999/H82bN0d6erramXsyMjJU29q2bYv58+dj69ateP7552FkZARTU1P07t0baWlpSElJQa9evQAArq6uSE9PR1ZWFpydnWFpaQkA6N+/P37++Wf07t0bLi4uAAA7OztVXVUJDg7G+PHj4eDgUGnbmDFjcPz4cezatUv1vSe1hIWFqd6T3bt3Y968eWrb+CdReohFRUWYMWMGJk6ciEOHDonRJJHeDBs2DD/99BMyMzMBVDzJEhYWhps3b2LatGmIjY1FbGwsbG1tYWRkBEEQYG9vDzs7O2zfvh2xsbGYNm0aevfuDQAwNv7//6a1Gd7zz33/+usvrF+/HmvWrMGqVasglUqrvNEDVPRKL1++rHrdjBkz4OTkpDpdVigUSEpKQseOHeHo6IikpCQAFZcBAMDe3h5paWmQyWQAgHPnzqmCvaY/g6WlJVasWIHQ0NAqty9fvhzbt2/Ho0ePAFT0DufPn49t27Zh27ZtiI6Oxv79+6sN3X8SpYf47bffYtSoURg3bhz8/Pzw2muvidGsWmPGjNFr+4ZSA+mGpaUlwsLCsGTJEgiCgEePHmHYsGHw8fGptG/v3r0RHh6OtWvXYvr06fD19UVZWRnat2+PV199Vet1ubq6Yty4cbCwsECLFi2Qk5MDe3v7Svu+/PLL+OWXXzBlyhSUlZXh3XffxUsvvYRz585h0qRJUCgU8PDwQI8ePbBs2TJVEFlbW0MqlcLa2hoBAQHw8/ODsbExHBwcEBQUhMOHD9eqZjc3N4waNarKadesra0RHByMd999F6WlpTh8+DC++eYb1fZ27drBxcUFCQkJNc4c0Z5UKSsrQ3FxMWbPno0dO3ZUuy+fVCFDpe5JFf6+NizqPjPRriHm5uZi8uTJcHd317ivXC7nRJxEJDrRAtHW1hYnT55EQEAA0tPT4ejoqHZfzphNhop/qBs3UW6qxMbG4tSpUzAyMoJUKuVzn0RkkEQJRA8PD3zxxRfw8fGBg4NDldPuEBHpmyinzG3btlUNDCUiMlRNdmA2UUM1J3ABcnLztXa8Z2ysseHTT7R2vIaMgagjCQkJOHLkiNrtBQUFAIDWrVur3cfT07NGd+WpacnJzUea7UvaO+D9RI27ZGdnY8yYMejRo4fqe25ubpVmngEqnjDx9PTE0KFDtVejSBiIepKXlweg+kAkMiTOzs6IjY3Vdxk6xUDUEXd392p7d3PnzgWgfrYRIkNXVlaGDz74APfu3VOtO/T354YzMjIQEhKCZs2awcTEBKtXr4atrS0iIiJw/vx5CIKA6dOna/2JnPpgIBJRjaSmpqomagWAefPmoU+fPpgwYQLkcnmlQDx9+jR69OiB4OBgXLhwAYWFhbh+/Tqys7MRFxcHuVyOiRMnYvDgwWjRooU+fqRKGIhEVCP/PGUuKirCN998gzNnzsDS0rLSJAre3t744osv4O/vDysrK8yfPx83b97E1atXVcH6+PFj3Llzx2ACsUnOh0hE9RcfHw8rKytERETgrbfeQklJyVOz55w4cQL9+vVDdHQ0PDw8sHXrVjg6OsLNzQ2xsbGIjo7Gq6++WuXkEvrCHiJRA/OMjXWN7gzX6nh1MHDgQAQGBuLXX3+Fubk5Onbs+NRyCD179sSCBQsQGRkJY2NjhISEoHv37jh37hx8fHwgk8kwYsQI1ZyJhoDrMusJb6o0TJztpnFQ95nxlJmISImBSESkxEAkIlJiIBIRKTEQiYiUOOyGqIFZ9N4cFObWbr3h6rS0scX/IjZo7XgNGQORqIEpzL2P952ua+14H6dVvz0sLAxXr17Fn3/+iZKSEnTo0AGtW7fG+vXrtVaDoWAgkl7Ex8cjMTERJSUlyMzMxMyZM1FcXIwDBw7A2NgYrq6ueP/99/VdJqFiOi+g4jNLT09HUFCQnivSHQYi6U1RURG2bduGW7duYfbs2WjevDmWLl2KPn36YNeuXXj8+DGaNeOvqCE6e/YswsPDYWpqiokTJ2L9+vU4evQopFIpwsPD4ejoCC8vL4Oe2aYqvKlCeuPi4gIAsLOzQ2lpKT766CPExcVh2rRpuHPnDgzwISr6G7lcjl27dmHs2LFVbk9MTFTNbBMTE4PNmzfj4cOHIldZO/zzS3rzz9UX9+7diw8//BBSqRQzZsxAUlISXnjhBT1VR5qoWyzuyR8yQ5/ZpiqiBOLdu3excOFCKBQKDB8+HLNmzRKjWWpgnnvuOXh7e6N169awtbVF79699V0SVcPY+P9PMCUSCXJycmBvb4/r16/DyclJNbPNypUrUV5ejo0bNxrUzDZVESUQY2JiMGfOHLi5ueGNN96Aj4+PQc1wQeLz8vJS/VsqleLkyZMAgAkTJuirpAajpY2txjvDtT1effn7+2PWrFlo3769qgc4fPhwg57ZpiqiBOLbb7+teiPKyspgYmIiRrMkkvj4eBw7dkzfZcDDw+OpoG2s9DVm8O/vrZubG9zc3FRfe3t7w9vbu9JrQkJCRKlNW0QJxFatWgEA9uzZAxcXF5ibm1e7v1wuR3Jyshil6Y1MJgOARvFzxsfHIzMzEw4ODnqrITMzEzKZjNNwUb2IdlPlm2++wfHjx/HZZ59p3FcqlTb6X2wLCwsAaBQ/p4WFBVxcXPD555/rrYYn16V1/X42hj9gpF61gXjhwgVER0fj4sWLqpWz+vbtCx8fH/Tr16/Gjfz+++84cOAANm3aBIlEUu+iiQyNIAiV7pqTYapuOJfaQFy5ciUsLS0REBAAZ2dn1R2lGzdu4ODBgzh06BCWL19eowI2bdqE3NxczJw5EwAQHh4OW9v6X8glwzBmzBh9l6DXGszMzJCXl4c2bdowFA2cIAjIy8uDmZlZldvVLiGQm5sLGxsbtQfWtL0+msKU7FxCoGGq6ndToVAgOzsbJSUleqqKasPMzAz29vYwNTWttE1tD/FJ2JWVlSEuLg6pqano1KkTpkyZAolEorMwJGpoTE1N1Q5SpoZF46N7S5cuRWZmJgYPHow//vgDS5YsEaMuIiLRabzLfPv2bezcuRMAMGLECEyePFnnRRER6YPGQJTL5SguLoa5uTlKSkpQVlYmRl0GLzIyEqmpqXV+/ZPXPrmWWFfOzs4ICAio1zGIqILGQPTz88Prr7+OLl26IDU1lf/5lFJTU5FyNQkOlnX7A9FCqLgbKb99oc41ZBbxiR8ibdIYiGPGjMHQoUORlZUFe3t7tG7dWoy6GgQHyzIsctXfdEb/u2i4s4YQNURqA3Hjxo34z3/+g8DAwEpjqyIiInReGNHfJSQk4MiRI2q3FxQUAEC1f7A9PT3h7u6u9dqo8VAbiMOHDwcA3kShBiEvLw9A9YFIpInaQOzSpQtKS0sRExODNWvWQBAElJeXY9asWYiJiRGzRiK4u7tX27vjQHfSBrWBuH//fmzevBm5ubnw8PCAIAgwMTGp1TPMREQNidpAnDhxIiZOnIivvvqqynnOiIgaG413mfv3748tW7ZAoVAAAHJycrBixQqdF0ZEJDaNj+49WRv34sWLyM7OxoMHD3ReFBGRPmgMRDMzM7z99tuwtbVFWFgYcnNzxaiLiEh0Gk+ZBUHAn3/+CZlMBplMhsLCQjHqIgOhafwfwDGA1Hho7CHOmTMHx48fx5gxY/Dyyy9j6NChYtRFDUheXp5qHCBRQ6axh3jp0iXMmDEDAPDyyy/rvCAyLJrG/wEcA0iNh8YeYmJiIme4IaImQWMPsaCgAEOGDIG9vT2MjIxgZGSEuLg4MWqrMz73SkR1oTEQN2/erNUGAwIC8P7778Pe3l6rx60NPvdKRFXRGIjt27fXSkOlpaWYP38+Ll++rJXjVUeM517z8/OR+5eJXqfguv2XCWzy8/XWPlFjo/EaoraUlpZi+vTpGDRokFhNEhHVisYe4t+Vl5er1meuLUtLS/Tv3x/79+/XuK9cLkdycnKd2qkJmUwGAPVqw8zMDB2t9D9BbJmZmU7fq5rQxvvZGGqghk9jIB49ehTl5eUoLS3F6tWr4e/vrxqGoytSqVSn6zJbWFgAQL3asLCwgFxbBdWDhYWF3tew1sb72VBqYOA2bhq7e9u3b8egQYNw8OBBJCYm4vvvvxejLiIi0WkMRKlUCgBo3rw5JBIJHj16pPOiiIj0QWMg2tvbY/z48Rg/fjw2bNiAXr161avBsLAwvQ65ISJSR+M1xLCwMDx69AjNmzdHz5490bZtWzHqIiISndpADAkJUfuijz76SCfFEBHpk9pTZk9PT3h6eqKwsBCOjo7w9vbGc889h9LSUjHrIyISjdoe4pAhQwAAX375JWbOnAkA6NevH958801xKqMmJTIyEqmpqXV+/ZPXPnkKqS6cnZ0REBBQ59dTw6fxGqJMJsMvv/yCf/3rX0hKSlKtrUJAZlHdH90rLDUCALSUCPVqv0udX21YUlNTkXI1CQ6WdZtZqYVQ8X7Kb1+o0+szi0zq9DpqXDQGYmhoKNatW4eVK1fCyckJa9asEaMug+fs7Fyv1z9U9mie6Vj343TRQh2GxMFSf0/+6POZdDIcGgPRyckJ69evF6OWBqW+p1acVJXI8KgNxBdffBEAoFAoUFxcDDs7O9y/fx/W1tY4efKkaAWSbtX32h2gnet3f/zxB2zqVQVR/akNxJ9//hkAEBQUhPfee08ViBxy07ikpqbityvJKLOwrvMxjMoqfo1+Tb9fp9ebyPJhaWYKmNa5BCKt0HjKnJ2dDTs7OwCAra0t7t69q/OiSFxlFtYodvHUW/vm148A5X/prX2iJ2p0DXHBggXo1asXkpKS0K9fPzHqIiISncZAXLlyJX788UekpqZi9OjRGD58uBh1ERGJrkbjEM+ePYvU1FTk5OTA1dUVrVq1EqM2akIUCgVul+hvSQYux0BADQJx0aJF6N+/P8aMGYNz584hODhY6wtP1Yah3BXlUw1EjU+NliH19fUFUDEbcUJCgs6Lqo6h3BUl7TI1NUUHi0d6HZgtta777xQ1DhoDUS6X488//0Tbtm2Rm5uL8vJyMeqqlkHcFSWiRkdjIM6dOxeTJ0+GlZUVioqKsHLlSjHqIiISncZAHDx4ME6cOIH8/HxY85SCiBoxtfMhLly4ED/88APKyipmH3kShuXl5Th+/DiCgoLEqZCISCRqe4irVq1CdHQ0IiIiYGVlBRsbGxQWFiI/Px+vvfYaQkNDa9xIaWkp5s6di4KCAnh5eWHixIlaKZ7qLz8/HyayPL1eFzWR5UFhYsRH90jv1AaiRCLBzJkzMXPmTNy6dQsFBQVo06YNHBwcat3IkSNHMGTIEEyePBn+/v54/fXXVav5EREZCrWBKJfLsW/fPkilUrz++uvo1KkTACAuLg6TJ0+uVSNXr16Ft7c3jI2N0bVrV6SlpaF79+71Kpy0w9raGhkPFHq/a2/KZ5nJAFR7DTEnJwcZGRnw8fFBYWEhgIreXm09evQIFhYWAABzc3PIZLI6lktEpDtqe4j5+fmqyUu/++47vPPOO4iKioIg1H7KewsLCxQXFwMAiouLYWlpWe3+crkcycnJVW4zlDCVyWRqawSAM2fO4PTp02q3Z2dnA4BqvZqqDBo0CAMGDKh7kTVgKO+nIYxv1fSZUuOnNhAVCoVqqM3IkSPxxx9/ICgoqE5rqvTo0QPnz59Hly5dkJycjPnz51e7v1QqRbdu3arcVtHT1P/plYWFhdoaASAzM1PVK67Kk/Wtq9unXbt21bahDYbyfhobqz1ZEY2mzxQAA7ORUxuIc+fOxdSpUxEbGwsbGxtMnz4dxcXFdZot29PTE4GBgYiPj4eXl1eTuKHi7u4Od3d3fZdBRLWgNhAHDhyIo0ePPvW9d955p05DZqRSKT777LPaV1cFQxkmkp/PMSJEjY3GJ1VeeeUV1eBsAGjWrBns7OywYMEC9OjRQ6fFUdOiz2VdG9OSrlR3GgNxwIAB8PDwwPPPP4+kpCTs27cP48ePx6pVq7B7924xanyKoQwT4WOM2mVubo729VhStb7Luja2JV2pbjQGYkZGBgYNGgQAcHNzw8aNGzFw4EBs2LBB58VR09G+fft6LcnKZV1JGzQGokQiwe7du9G3b18kJSVBIpHgypUrT51GExE1BhrHOoSHh+PWrVuIiIhAVlYWVq9ejby8vFo9y0xE1BBo7CG2bt0aAwcOhI2NDTp37ozWrVvjpZdeEqM2IpWEhIRqn5KqybIQnp6eHApF1dLYQ4yIiEB8fDxMTU1x4MABhIWFiVEXUa20adMGbdq00XcZ1MBp7CGeP38ecXFxAIA33niDU3eRXnCgO4lBYw/x8ePHqudMBUGAkZGRzosiItIHjT1ET09PTJkyBb1798alS5fg6am/8X+kGyay/Ho9+WOkqJi4QzA1r3P7gG2d2yfSFo2B+NZbb+HFF19Eeno6vL290bVrVzHqIpFoYzDykxsazo51DTVbDoomg6A2ECMiIiqdHl+7dg0AEBgYqNuqSDQBAQH1PgYHRVNjoTYQHR0dxayDiEjv1AbiuHHjxKyjVnjNi4h0QeM1REPDa15EpCsNLhB5zYuIdEX/87YTERkIBiIRkRIDkYhIiYFIRKQkWiAqFApODEFEBk2UQHzw4AH8/f2RlZUlRnNERHUiSiAWFxcjJCSEY/eIyKCJMg7Rzs4OdnZ2Nd5fLpcjOTlZZ/XIZDIA0GkbTQnfT2osdBKIMTExTy1yP3v27FotOyCVStGtWzddlAYAsLCwAACdttGUNKX3k6HfuOkkEP38/ODn56eLQxMR6QyH3RARKYkaiLGxsWI2R0RUK+whEhEpMRCJiJQYiERESgxEIiIlBiIRkRIDkYhIiYFIRKTEQCQiUmIgEhEpMRCJiJQYiERESgxEIiIlBiIRkRIDkYhIiYFIRKTEQCQiUmIgEhEpMRCJiJQYiERESqKsy3z9+nV8+OGHePz4MaZOnYqxY8fqtL2EhAQcOXJE7fbU1FQAwNy5c9Xu4+npCXd3d63XRkSGS5RA3LhxIz755BPY2dnBx8dH54GoSZs2bfTaPhEZJlECcfny5bC2tkZZWRkEQdB5e+7u7uzdEVGtiRKI1tbWAIC1a9fC09NT4/5yuRzJycm6Lou0RCaTAQA/M2rwdBKIMTExOHr0qOrr2bNnIzk5Gbm5uQgMDNT4eqlUim7duumiNKolTddjAeDOnTsAgM2bN6vdp7Fck2XoN246CUQ/Pz/4+fmpvk5ISEBKSgo++eQTGBkZ6aJJ0iNek6XGwkgQ4aLea6+9BolEAgsLCwBAVFQUTExM1O6fnJzMHiIZJP5uNm6iXEM8dOiQGM0QEdULB2YTESkxEImIlBiIRERKolxDrC2OQyRDJZfL9V0C6ZAod5mJiBoCnjITESkxEImIlBiIRERKDEQiIiUGIhGREgORiEjJIMchGoL4+HgAgJeXFwAgMjISL7zwAh48eIAvv/wSzZo1Q8uWLfHBBx/A1tYWALBu3Trcvn0bn376qd7qJqK6Yw+xFtLT07Fv3z5ERUVhx44deOutt7Bp0ybV9itXrqCkpARFRUV6rNLwxMTE6LsEohphINZCXl4epk2bBjMzMwBAv379sGzZMgAVC1fZ2dlh8ODBOH78uD7LNChnz55Fbm4uLl68iPj4eFXPG6jodZ89exYAcO7cOUybNg1Tp07FjBkzkJaWptpv5MiR+P7770WvnZoeBmItnD17Fg4ODgCADRs2wNfXVzUR7rFjxzBy5EiMGDFC4wzTTYmbmxvatWsHV1dXtfsUFBRg7dq12LBhA3bu3Illy5Zh7dq1AIDff/8dAwYMwOHDh8UqmZowBqIaUqkUCoVC9XVxcTEkEglycnIAAHPmzEFsbKxq+8mTJ7F9+3YsWrQIly5dQn5+vug1G6r79+9Xuz0xMRGjRo1Cq1atAAAODg5Yv349gIo/NF5eXsjJyUFxcbHOa6WmjYGohpOTE86cOQMAKC8vx+XLl/Huu+9i+/btqgf809LSUFhYiIyMDHTs2BFffvkltm3bhlmzZuHYsWP6LN+gVLf+tZGREe7du6fqee/Zswe+vr549dVXAQC//fYb+vTpg8GDB/O0mXSOd5nVcHFxQfv27eHl5QUTExOMHTsWrq6umDRpEvz9/VVrwyxevBgJCQkYNmyY6rXu7u5YuHAhfHx89FW+QZJKpU/dcHrS627btq2q5z1p0iRMmjQJvr6+uHLlCu7evQt/f3+UlJSgZcuWNVq1kajOBCKRJCcnC/PmzRMEQRDKysqEadOmCQUFBUJeXp7g4+MjPHjwQBAEQbh3754wbNgwITw8XPjhhx9Urx83bpxQWFiol9qpaeD0XySq8PBwnD59WtXrnjp1KgDg4sWLWL9+PcrLy1FWVgY/Pz9s3rwZe/bsgUQiAVBxI8vW1hYTJkzQ549AjRgDkYhIiTdViIiUGIhEREoMRCIiJQYiEZESA5GISImBSESkxEA0MDt27NB3CZg/f75qFhqipoSBaGD+Pr8iEYmLzzLrUUZGBkJCQtCsWTOYmJhgwIABKCwsxPLly7F48WIsWrQIWVlZKCsrw5tvvglPT0/4+vqic+fOyMjIgCAIWLNmDdq2bVvl8W/duoUlS5ZAoVDAzMwMa9asgUwmw+LFi/H48WMYGRlhyZIlcHFxwc6dO7Fv3z60bdsWeXl5AACFQqG2BhcXF6SkpKCoqAjr1q1D+/btxXzriHRDrw8ONnE7duwQVqxYIZSWlgqnT58Wbty4IQwaNEgQBEGIjY0VQkNDBUEQhL/++ksYOXKkkJeXJ0ybNk34+uuvVa9fuXKl2uPPnj1bSExMFARBEA4fPiz89NNPQkBAgPDdd98JgiAI165dE8aNGyc8fPhQeOWVVwS5XC6UlpYKo0ePFs6cOVNtDQcPHhQEQRA+/fRTYcuWLbp5g4hExlNmPfL29kbr1q3h7++PnTt3wsTERLUtLS0N/fv3BwBYWlrCyckJWVlZAIABAwYAAFxdXZGRkaH2+BkZGejbty8AwNPTEy+++OJTx+3WrRvu3buH9PR0ODs7QyKRwNTUFL169dLCF9x9AAABVklEQVRYQ/fu3QEAzz77rGo6NKKGjoGoRydOnEC/fv0QHR0NDw8PbN26FYLy0XInJydcuHABAFBUVISbN2/C3t4eQMXaLUDFhAjOzs5qj+/k5ITLly8DAA4ePIjY2NinjpucnAwbGxt06NABqampKCkpQVlZGZKTkzXWQNQY8RqiHvXs2RMLFixAZGQkjI2NERISguzsbAQFBeF///sfli5diilTpkAul2POnDlo06YNAODrr79GVFQUzM3NsXr1arXHX7hwIT744ANs2rQJZmZm+OSTTzBs2DAsXboU27dvx+PHjxEaGgpra2vMnTsXkydPhrW1NczNzQEAEydOVFsDUWPE2W4aGF9fXyxfvhxOTk76LoWo0WEPsYErLS3FjBkzKn2/c+fOWLFihR4qImq42EMkIlLiTRUiIiUGIhGREgORiEiJgUhEpMRAJCJSYiASESn9H2ToOpj27CBuAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 216x216 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set_fontsize_screen()\n",
    "df = plotData.copy()\n",
    "df['NNA'] = df.index.map(lambda x: x[-1] == 'A')\n",
    "df = df.rename(columns={'UGA no start overlap':'$^{\\dagger}$UGA', 'TGA':'UGA'})\n",
    "df = df[['UGA', '$^{\\dagger}$UGA', 'NNA']]\n",
    "df = df.melt(id_vars=['NNA'])#.reset_index(level=1)\n",
    "df = df.rename(columns={'value':'log2(odds ratio)'})\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3,3))\n",
    "x = 'stop_codon'\n",
    "y = 'log2(odds ratio)'\n",
    "hue = 'NNA'\n",
    "sns.boxplot(ax=ax, data=df, x=x, y=y, hue=hue)\n",
    "ax, test_results = add_stat_annotation(ax, data=df, x=x, y=y, hue=hue,\n",
    "                                   box_pairs=[(('UGA', True), ('UGA', False)),\n",
    "                                              (('$^{\\dagger}$UGA', True), ('$^{\\dagger}$UGA', False))],\n",
    "                                   test='t-test_ind', text_format='star', loc='inside',\n",
    "                                        comparisons_correction=None, \n",
    "                                       verbose=2)\n",
    "ax.legend(loc='upper left', bbox_to_anchor=(1, 1), title='C-terminal codon NNA')\n",
    "\n",
    "speciesSetName = 'TGA_non_overlapping_start_codon'\n",
    "outputFilePrefix = speciesSetName\n",
    "plotsFolder = analysisCtermPath / 'overlapping_start_codon' / speciesSetName\n",
    "filenamePlot = '{}_comparison_bias_NNA_codons'.format(speciesSetName)\n",
    "print(filenamePlot)\n",
    "fig.savefig(plotsFolder / (filenamePlot + '.svg'), dpi=400, bbox_inches=\"tight\")\n",
    "fig.savefig(plotsFolder / (filenamePlot + '.png'), dpi=400, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "stop_codon\n",
       "TAA                     0.232382\n",
       "TAG                          NaN\n",
       "TGA                     0.213708\n",
       "UGA no start overlap    0.458140\n",
       "Name: N/AAC, dtype: float64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "addEmptyRowBetweenGroups = True\n",
    "addEmptyRowBetweenGroupsFillValue = np.nan\n",
    "plotData2 = plotData.copy()\n",
    "nameIndex = plotData2.index.name\n",
    "nameCol = plotData2.columns.name\n",
    "stopCodonList = list(refCodonTableDf[refCodonTableDf['aa'] == '*'].index)\n",
    "\n",
    "# Remove values that are not significant using mask\n",
    "plotData2 = plotData2.where(~maskData, other=np.nan)\n",
    "# plotData2 = plotData2.to_frame()\n",
    "plotData2 = plotData2.loc[plotData2.index.map(lambda x: x not in stopCodonList)]\n",
    "plotData2 = sort_codon_index(plotData2, refCodonTableDf,\n",
    "                             addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                             addEmptyRowBetweenGroups=addEmptyRowBetweenGroups)\n",
    "plotData2.index.name = nameIndex\n",
    "plotData2.columns.name = nameCol\n",
    "plotData2 = plotData2.drop('all', axis=1)\n",
    "plotData2.loc['N/AAC']\n",
    "# print(\"plotData:\\n\", type(plotData), plotData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UAA</th>\n",
       "      <th>UAG</th>\n",
       "      <th>UGA</th>\n",
       "      <th>UGA no start overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R/CGU</th>\n",
       "      <td>0.362795</td>\n",
       "      <td>0.525103</td>\n",
       "      <td>0.349839</td>\n",
       "      <td>0.538713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/CGC</th>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.351280</td>\n",
       "      <td>0.734030</td>\n",
       "      <td>0.993814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/CGA</th>\n",
       "      <td>1.355864</td>\n",
       "      <td>1.743242</td>\n",
       "      <td>2.775695</td>\n",
       "      <td>1.558219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/CGG</th>\n",
       "      <td>0.456227</td>\n",
       "      <td>0.710066</td>\n",
       "      <td>-0.055614</td>\n",
       "      <td>-0.166855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R/AGA</th>\n",
       "      <td>1.068920</td>\n",
       "      <td>1.422678</td>\n",
       "      <td>1.475391</td>\n",
       "      <td>0.986246</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            UAA       UAG       UGA  UGA no start overlap\n",
       "R/CGU  0.362795  0.525103  0.349839              0.538713\n",
       "R/CGC  0.118014  0.351280  0.734030              0.993814\n",
       "R/CGA  1.355864  1.743242  2.775695              1.558219\n",
       "R/CGG  0.456227  0.710066 -0.055614             -0.166855\n",
       "R/AGA  1.068920  1.422678  1.475391              0.986246"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotData2.columns = [re.sub('T', 'U', col) for col in plotData2.columns]\n",
    "plotData2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py:194: UserWarning: 'colors' is being deprecated. Please use 'color'instead of 'colors'\n",
      "  warnings.warn((\"'colors' is being deprecated. Please use 'color'\"\n",
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py:194: UserWarning: 'colors' is being deprecated. Please use 'color'instead of 'colors'\n",
      "  warnings.warn((\"'colors' is being deprecated. Please use 'color'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGA_non_overlapping_start_codon_composition_bias_codon_Cterminal_barplot_stopCodonContext\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWkAAADWCAYAAAATkHNaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3XlcVPX++PHXgKyiIIuyCLiUint5wyzXxOyXS2bX5aqY1i2XvN1r2tctjcrrinLLpdQKjERNr5m5pGmmlOKGhKkouSWSqKDIIgPMzO8PLxPLGTysM8j7+Xj0CN98Pue8z2eG9znzmbNoDAaDASGEEBbJytwJCCGEME2KtBBCWDAp0kIIYcGkSAshhAWTIi2EEBZMirQQQliwWlukv/76a0aMGMHw4cNZuHBhid/HxMRw69atci//5s2bfPzxx6rb/+1vf1OML1iwgODgYP7yl78watQogoODATh//jxjx45lxIgRjBw5ktjYWGOfo0eP8tRTT5Gbm1vu/KuTqW1Xo6zjrMaOHTse2ObatWucPHmyUtZXke0XD79aWaRPnz7NDz/8QGRkJBs2bCArK4vdu3cXabN161aysrLKvQ4PDw8mTJhQ0VSZPn06kZGRBAQEEBERQWRkJHl5ecyePZu5c+cSFRXFsmXL+Oijj4x9du/eTVBQENHR0RVev6WrrHEu7Msvv3xgmyNHjvDbb79V6nqFUFLH3AmYw549exg6dCjW1tYAzJgxgzp1/hyK5ORkoqOjSU9PJyQkhBkzZpCTk0Pfvn15+eWXGTFiBFZWVgQFBXHgwAGsrKzQ6/V06dKF77//np49e/LCCy/wn//8hyFDhhAeHo5WqwVg1apV/Prrr3z44Yfk5OTQu3dvXn/99TLlHxcXx+OPP46Pjw8Arq6uhIeHA2AwGEhISGDu3LksW7aM3r17Ky5j+vTp2Nvb8+uvvzJgwABefvll1q5dy86dO7GxsSEkJIT4+HgOHDjA7du3cXNzIywsTHFZixYt4uTJkzg4ODB//nxmz55NaGgo9evXZ+zYsXz66afMnDmT5ORkfHx8mDdvHjNnzuTOnTsEBAQYlxMWFkZ8fDyZmZnMnTvXuDO9c+cOXl5eLF68mPHjx+Pi4sK5c+eYMGECrVu3NjnO58+f591338XV1ZWkpCR27dpVIvfExERmz56NwWBg+PDh5Obmcv78ebZs2cK9e/f4/vvvuXv3Lm+//TZWVlaEhoZiY2ODTqfjzp079OzZEw8PjyLLvHr1KtOmTSMnJ4fhw4fTv39/3nrrLTIzM/Hx8WH+/Pns3buXTz75hBYtWpgcx6+++opr165x5coVOnbsyLRp08rwLhEPi1p5JJ2WlkajRo2M/3ZwcMDGxsb4b29vb7p168b06dNZs2YN//jHP4iKiuLo0aOkpaWRmprK6tWrCQoKIj09nU8//ZQmTZrQoEEDNmzYwIEDB4qsz9bWlvDwcHx9fYmPj+f3339n+fLlrFu3jn379pU5/+vXr9O4cWMA9u3bR3BwMAMGDODatWucOHGC9u3b07RpU65du0Z2drbJ5fTo0YOoqCi+/vprUlNTOXDgABs2bGD+/PksXbrUOBZffPEFd+/e5fr16yWWcfr0aW7fvs369euZOHGicVwOHjzI1atX8fX1Zd++fTRr1ozIyEgeeeQR9u/fD0BwcDD//Oc/AcjLy8PT05Pw8HDGjRvH3r17jev/8ssvcXR05JdffiErK4sxY8awbt06IiIiSh3nTz75hOXLlxMWFkZaWpriGBw6dIgXXniBL7/8Ejs7O4YNG0aLFi0YPHgwOp2OiIgI5s2bx7fffguAn58fUVFRDBs2jL///e8lCjTA2rVrmTp1Kps3byYlJYWNGzfSt29fvvzyS/z8/Pj+++9Zu3YtUVFRvPbaaybHEaBt27asX7+en376Cbk4uHaqlUXaw8ODP/74w/jv69evc/r0aeP878aNG42/u3z5MqGhoQQHB3P16lWuX79OgwYNcHR0BKBp06ZoNBpcXV3x8/PD2traeIReoGnTpsb15ubm4uHhwezZs3n//ffLNW/s4eHBzZs3AejduzeRkZG0bduW/Px8vvvuO06cOMGrr77K7du3jcVOSdOmTbG1tcXBwYGkpCRatmyJRqPB19eXu3fvlsi94Ci1sCtXrtC2bVsA2rVrx+XLlwkKCuLHH39kz5499O3bl8uXL7N9+3aCg4PZvXs3KSkpAMZPAgB16tThjz/+4O233+abb75Br9cD0LFjRwACAgJISkrCzs6OVq1a4ejo+MBxLjgCd3JyokmTJopj8Ne//pULFy7w8ssvG7e5gE6nY+rUqURERBjzKdg5lubKlSsEBARgZWXFP/7xDy5fvlxijKysrLC3t6d58+bY29srjmPhbXJxcUGn0z1w3eLhUyuLdFBQEFu2bDH+4a1cuZIrV64Y53+HDRuGRqPBYDDg6+vLnDlzjHEfHx80Go1xWYV/VuvDDz9k0aJFvPnmm6Ue6Zry+OOPc+TIEeOOJiMjg99++w2NRkNsbCwbNmzgs88+Y8WKFWzfvt3kcgrn7u3tzfnz54H7H9cLdkIP2r7GjRtz+vRpAOLj4/Hx8cHV1ZXc3FyOHTtG586d8fX1ZejQoURGRvLmm2/Srl27Ess+e/YsqampLF68mL/85S/Go8bExETj75s1a4ZWq+Xq1atkZ2eXKNLF1atXj5SUFLKzs7ly5YpimwMHDjB48GDWrVtHVFSUMZ6ens6BAwcIDQ0lKCioxFFswftDiY+PD4mJiRgMBiZPnoyvr2+RMWrcuDEajYasrCyuXr2KVqtVHMfiYyRqp1o5J926dWsCAwMZMWIEOp2OJ554gueff75Em5CQEObPn88777xDRkYGbdu2xdnZucLr79atG0OGDMHZ2Zm6deuSn59v/N3//d//sWjRolL729raMnfuXObMmYNWqyUvL48XX3yRW7duGY/gAB599FGuX79OWloarq6upS7Tw8ODp59+muHDh2MwGFi4cGGRM0ZMad++Pdu3b2fEiBHUqVOH//znPwA8/fTT/PLLL9SpU4c+ffowY8YMRo4ciZ2dnbFNYX5+fvz2228MHz4cd3d3XF1dady4MTExMRw7doxHHnnEOH+9ZMkSrl69yr/+9a9Sc5s0aRKTJk3Czc3NuNMp7tFHH2XatGnUrVuXfv36AWBtbc327dupU6cOQ4cOxdPTs8QnnubNmzNr1iy6dOmCr69vkd+99tprzJgxA61Wy9ChQ3n22Wd5++23+eqrr/D392fixIl4eHgQHBzMI488gr29veI4rlu37oHjLx5+GrkLnmVZtWoV48aNM3caFmHLli3odDqGDBlijAUHBxMeHl7ki15Ttm7dSp8+fYD7p7lt27atynIVoqrUyiNpS/bSSy9V+jInT55c4pzviIiIB04XFHfw4EHWrFlTJDZw4MAiRdSS1KtXj+DgYPLz83nzzTdZvXp1idMSp02bZpwLLo/KGlshTJEjaSGEsGC18otDIYSoKaRICyGEBZMiLYQQFswivziMi4vDzs7O3GkIUetotVrjBUTCMlhkkbazsytyTwchRPU4e/asuVMQxch0hxBCWDCLPJIWQliGO3fuFLnPjaga9vb2NG7cuMiN3gpIkRZCmHTr1i2aNGmCg4ODuVN5aBkMBlJTU0lKSjLeUKswme4QQpiUl5eHvb29udN4qGk0Gtzc3MjJyVH8vRRpUSn2/dDc3CmIKmLOO/EV3Nb2YVfaGEuRFkJYrNDQ0Epf5rJly0zeutYSlWlOOi0tDSsrK1xcXKoqHyFELbZz505iYmLIyspiypQpJCQkcPjwYU6cOMHt27cxGAzMmTOHAQMG8OKLL5Kbm8v48eOB++d4z5s3j7y8PJ566ins7OzYv38/ubm5zJw5k507d3LlyhVOnjzJwIEDWb58eZFlDh48mN69e3Pz5k1CQkLMOxCFqDqSPnToEO+99x6rV69m1apVfPDBB5w4caLMK1u3bh1r1qwp8tBUIYQokJycjIuLC8OGDaNRo0a0bNmSdu3acePGDWbPno2XlxdnzpyhRYsWvPLKK9y5c4fMzEwAYmJi6Ny5M/PmzePRRx/lu+++Y968eYwePZpt27YRHx/PrFmz6Ny5Mzk5OSWW6e3tzRtvvGFcnqVQdSRdt25dRo8eTVRUFAaDgTFjxpCamlrmlfXp0wc3NzfmzZtXajutVisn1ddA8pqJiuratSsajYZVq1bh5ORkjBeesy14ohLcrxUFv8vPzzf+nJycXGS5er3e+DAMa2trDAZDiWUWfEGq5l7l1UlVNh06dODtt99mxowZACxdupS5c+eWeWX16tVj6dKljB07ttR2csVhzZP8B/KaPQTMvaNNTEzk559/xsrKCk9PT+rXr89PP/2Eu7s7CxYsQKPR0Lp1a3799VcWLlxIo0aNqFu3LnC/wIeEhBAdHU3Xrl3p27cvc+bMQafTMWXKFHbv3s38+fOJj4/npZdeKrFMi2VQ6e233zbodDpDXl6eYfr06Wq7FfHWW28ZFixYYAgPDy+13ZkzZ8q1fGE+e/c1M3cKohIU/9uz1L/FKVOmmDuFSmdqrFUf148YMYJZs2YBMHTo0HLtEJYsWVKufg/iuT+O673Kd1OYivQVQphHaGgoKSkpNGrUqNR2N27cwMPDo0Y/0FfVF4c3btzAzc2NCRMmMGHCBNzd3as6L4vjuT/O3CkIIQpRc3peWFgYOp2uGrKpOqqOpKOjo0tcvz9p0qQqSai8kqZH03hBN3OnIYSogLi4ODZv3szdu3dp0KABderUwdnZmcGDBzNz5kwCAwPJy8uje/fuJCQkcPHiRaKiotDr9fj4+NCiRQvOnj2LlZUVTZs25fTp0+zdu5fnnnvO3JtWbg88kk5PT8fb25s7d+7QqVMnOnXqxMWLF6sjt0rnuT9OrowTwoJt2rSJkJAQXn75ZRo2bMjs2bO5ffs26enpBAQEMGnSJK5du0anTp1o2bIlv//+OykpKTg6OhIfH0+3bt1ISkri+vXr9O7dm9atWxMUFGTuzaqQBxZpBwcHUlJS0Gq1pKSkkJKSwt/+9rfqyE2dEGfz9BVCVLqC0+icnJwwFHpGtsFgUDxFTqfT0aVLF6ZOnWosxvfu3SMjIwMw7yXtleWB0x22trYMGjSIdu3acerUKQwGA2fPnuWJJ56ojvyEELXISy+9xOzZs7l37x7169dn4cKFeHh4KF7lfOvWLXx8fNixYwdnzpyhadOmbN68mR49eqDX6/nvf/+Ln58f69evJzg42AxbUzlUn90RGhpKkyZNaNGiBadOnarKnKrdvh+a0/uZC2VuJ2eGCFG5AgMDCQwMVPzd5MmTAViwYAEAERERwP3rNh5mqm+w5OrqSu/evalXrx7OzjVwmkCmNoQQNZDqIu3t7Y2rqysxMTF4eHhUZU4PlDQ92qzrF0KI6qK6SCclJZGbm8vo0aN5/vnnqzInYeEqeoaM7GSFUE/1nLSPjw979+41/tvSzpMurqbMF8v53ffJOFi48k4XhqQrhpctW8bAgQPx9/fnyJEj/P7773h5ebF8+XI2bNhgbDdnzhyaNGnCK6+8Ur71PwRUH0lPmjSpyH+idqiJ55WbK+eacFWqJb+e27dvp2vXrhw7dgy4f40GwPHjx4ucjlfbWPSTWTz3x8lH4wqytMKhVCQsuXCoZqFfTFvq30/x85czMjLIy8vjpZdeMh5Jf/vtt/Ts2ZO2bdvy008/mSNNi6C6SOfn56PX60lISOD27dtVmVPFVeEfTEXe9MX7mqU4mbGYWNoOoywstdiVhdL4m+s1qVevHrdu3QLuP8cwOzub27dv8/HHH3Py5Elu3brF9u3b2bdvHxcuXCgyBVLbqC7S06dPJzY2ljVr1jB//vyqzOk+E8WkJv+hV7tqGsOH5jWx0KPhymYJO5wBAwbwxRdfMHXqVPbt28eRI0f45JNPeP/995k+fTqbNm3iscce49///jdLlixBo9GUuJF/baH6i8O6desSHR3NuHHjCA8Pr8qcaocQZ+hei+4mGOIMPQ6YO4sH8twfx/HdGX9+iRnibPLLL0u374fmtCSichZWyWPg6urKhx9+qPi7Z599tkRs+fLllbr+mkTVkfSmTZvw9fWlfv36HDlyBF9f33KtbN++faxbt854pVBtUhVTG1V9RKTqCLmCR58PzVG4kjKOTUXn6y3hCFlUPlVF2t/fn1OnTtGiRQuaNWtW7rvgxcbGMnLkSNLS0srVv7Z8HFVS7X+AtXisK6L466R2J2TunZUUeMulMag8t2XWrFl88MEHGAwGQkJC+OCDD8q8srCwMCZPnkxoaChTp0412S4uLg47O7syL98ckv/oT8s9EWS8bHrqotd1Lfs9y7c9ARufZF93d7y9theJnR0Wo3oZ9dbeKpGfUkxNLp49DjxwWwrarTO8VCTviuh1XVtieQVjU3j8C8am8PYptatOyX/0L3UclF5PtWMNymNTofdcoWdVnj17Vp5dWU1MjbXqOenBgwcze/ZsNBoNQ4YMKVcSbdu2JSIi4oGXldekB9Em/+9ZCKXle72Cm6J086eyjE8S0SXaK8XUKku/Snsdr8eVurzC8YCAAJPbZ473lZqH9KrZLpOux5V8j1yPK9e2mvtBtKIkVUV606ZNaDQaOna8fwVfYmIiHTp0KPPK+vTpU+Y+lq73MxfgGXNnUX3UXsV5vVdH9v1Qxck8LCr4pVx1XFlb3ukYU7mpueLw3LlzrFq1Cnt7e3Jzc5kyZQpeXl7l3oaaSvWcdHR0NJ6ennh7exMTo/6jtqigmnhmQRXlrOZ2sqLmKn7F4YoVK1iwYAHz5s1j5syZ3Llzx8wZmoeqIh0YGIiTkxNPP/00Tz75JA4ODlWdlyhNGYugue6JYc6iWuPvA1ITd85loOaKQ41Gg62tLXv37mXRokXs2bPHHKmaXbXOSQvLUeOLWC1QkWkMS7+5WMEVh/7+/iavONRqtdy9e5egoCC6d+/OnDlzzJ22Wagu0q6urjg6OmIwGHB1da3KnISlesiP7oRplV30BwwYwHvvvcf69evJy8sjLS2Nzz77DFtbW/bs2cOmTZv417/+xfTp07GzsyM/P7/W3glPdZFeuXIlM2bMAO4/rmbu3LlVlpQQZSGfCmoetVccrly5srpSsliqi7RGo8HFxQW9Xo9Op6vKnIQowdI/vgtRVVQX6REjRjBr1iwAmZMWFkPO+BAPO9VFumPHjsbzpIUQ6smORFSEqiK9atUq2rVrR5s2bbC2tubEiRP89ttvvPrqq1WdnxDCApT3BmGmdlDFL2a5cuUK58+fJysri/z8fHr06EH//v0BePXVVxkyZAjPPfdcufOvyVQV6XHjxrF//36WLVuGRqPhqaeekgIthKg0OTk5dO/ene7duwNw6NAhAC5cuECbNm3YsWNHrS3Sqm/636tXLwYPHkyfPn3o1asXv/zyS1XmJUT5yGmCNULxi1kWLlxIhw4dSEtLY/r06XzyyScAbN68mX79+uHs7MyFC7Vz2qhMzzg8fPgwnp6eREZGEhsbW1U5CSEecsUfnzVhwgQOHTqEq6srCxYswN3dnby8PPbv38+6deu4desW69evN3PW5qH6i0OANm3a4Ofnx/Dhw1m6dGlV5SSEeMgVv5hl7ty5LFmyhF27dgHQunVr9u/fT3BwMCNHjsRgMDB69Gju3btX625LobpIf/vtt7Ru3RqA1NRU8vPzOXHiBJ06daqy5IQQlqGyz1BRupglJCTEZHuNRkNkZGSl5lBTqJ7u+Pzzz4mOjmbt2rVs3ryZ8ePHc/DgwarMTYhKI6fBiZpKdZHu3LkzgwYN4vz58yQlJeHo6FjrPnaIGup/XybK5eOiJlI93dGtWzc+++wzgoODSU9PZ9GiRbi5uale0eXLl4mKiiI/P5/x48fTsGHDciUshBC1ieoibTAYuHv3LuvXrycoKIh3331XVb+oqCiOHTtGbm4u77zzDrt37yYzM1OKtBDCLFJSUmjUqJHJ32dkZGBtbY2jo2M1ZmWa6iJ96NAh3nvvPQAWL15Mt27qPjqOGDGCESNGALBt2zbc3d1p1qxZqX20Wq08a0082LAYkPdJtSjv08RNTTEpPT5r0KBBLFy4sMqvOgwNDWXx4sUmf7927VoGDBiAv79/hdYzdepUQkNDK7QMKEORzs3NRavVotfrycvLK/OKDh48yDfffEOnTp24evUqvr6+JtvWpAfRCvEwMefB0VdffVWmqw63bNlCbGws9vb2PPHEE7Rq1YqPPvoIa2trhg0bZjzzbOfOncTExJCVlcWUKVNISEjg8OHDHDp0iIyMDO7du8fChQsZMGAAXbt2JSEhAQcHB+NV1d9//z379+83Pmfx448/5v333+fdd99l4sSJfPzxxwC0b98egB9//NG4c0lISCA8PJysrCxGjRrFsWPHyMrKAmDYsGE0bdr0geOiqkjHx8fz3HPPMX78ePR6PW+99ZaabkUUHnwhRO1W/IpDuP/g2f79+5OWlsaiRYtITk7mqaeeYvPmzQwaNIjIyEguXLhA8+Z/3kekV69exqe2xMXFMW3aNBo0aMDMmTONRTo5ORkXFxf69+9Po0aNaNmyJV26dMHGxoaMjAyWL18OgK+vL9OmTWPZsmUEBQUZ1/Hdd9+xZMkS4uPj2b17N1ZWVmRlZaHRaNi7dy9arRY3Nzfi4uLo2LEjAwcOJCgoiD179uDs7MyLL75IfHw8x48fB2D48OG4ubmxYsUKpk+f/sCxUnV2x4ULF7h69SoDBw5k0KBBXLp0SU03IYRQVPyKQ2dnZx555JEyX3Xo4OCAjY0NBoMBg8GARqNBo9FgMBiMbbp27Uq/fv3YsGED586dAyA7O5uIiAhcXFyM348VzEEr7UAK6PV6unfvTlhYGF27dkWn0/Hcc88xdepUAgMDiywHYOvWrSQlJREQEGDMyWAwoNVqsba2VjVWqo6kX3zxRVULE0IINYpfcTh//nzq1KnDvHnzyn3V4bBhw1i8eDF2dnaMHDnSGE9MTOTnn3/GysoKT09P6tevz48//ohOp+O7774jJSWFu3fvGts3btyYDRs2MG3aNAD69u3LnDlz0Ol0TJkyBScnJxYsWMD//d//kZmZSUhICNu3b6dHjx4lcvL09OTo0aPY2Njg4OCAk5MTq1evBmDixImqxkpjKLzLsRBnz56VOWkhzKD43578LVauwl+YFmdqrMt07w4hhBDl949//KPMfcp0FzwhRO1jgR+2HzqljbEUaSGESTY2NuTk5Jg7jYeawWAgNTUVe3t7xd/LdIcQwiR3d3cuX75s7jQeevb29jRu3FjxdxZZpOWKQyHMQ6vVFvm3i4sLLi4uZspGgIWe3SGEEOI+mZMWQggLJkVaCCEsmBRpIYSwYFKkhRDCgkmRFkIICyZFWgghLJgUaSGEsGAWV6STk5NJTk4mJSWlyPXsBT/r9Xry8/OL9NHr9Rw9erRILCcnh9zc3HK1U1KevkrbUtp2VFbf0nJ+UN5VMf7VOQ7V/TqVtt6qXnd1jI0wP4sr0qdOneLXX38lNja2yPPBli5dSn5+Pjk5OcbH1ezatYtNmzYREhKCq6srAEuWLCE9PZ2PPvqIjz76qEztlFSkr9K2KG1HZfdVyllt3lUx/tUxDuZ6nSoy1jVhbIT5WeRl4QVPWNDr9UXi8fHx6HQ647+zsrLQ6/VoNBrjm0+j0XDhwgW6dOnCyZMny9ROyYP6xsXFqdqWwkelBdtR2hMgKtJXKeey5F3Z429qmcWXV5Z8KpJLaeuo7rFWu30V6VvRsRHmZXGXhV+8eBFra2tu3ryJj48PXl5ewP032sWLF6lTpw4uLi54eXmRm5uLlZUV586dIzExkUGDBpGZmcnFixexsbEpc7vSpKamEh0dzaBBg8jLyzPeW+TixYsMGjRIsc+1a9eMf+BHjx5l0KBB5Ofnk5CQgF6vr7K+SjkDqvJOSEigXr16WFtbY2tra/yjTktLIzk5mby8PJo1a4azszOZmZk4OTlx8+ZNjhw5Ynw+XVJSEjqdDn9/f1xdXRVfU6XXU4lSPhXJxZLGuuB35R0btX0r4/0vzMfipjs2bdqEm5sbx44d47///a8xvnz5cpo2bcqPP/5ojIeFhZGTk8PBgwdJSkoCYMWKFTRr1qxc7ZRs2LCBTZs2sWLFCuPDKf/73/9y7tw5tm7dWuSBlcUdPHiQn3/+mU8//dTYbvPmzZw9e7ZK+yrlrDbvHTt24OnpiaurK1FRUcZ4eHg4AQEBtGnThsjISABWrVqFTqfD2dnZeKe08PBw2rRpQ5s2bYz9lV5TpddTbT4VyUVJdHS0Wca6omOjtm9F3v/C/CyuSNetW5f9+/fj5eWFra1tqfHKjilxd3c3fglT8ABepZgSDw8Ps/Q11U5Nfzs7O3bu3MnOnTuLjItSXG2sIuNfeHl2dnYVjqkdr+oY64qOTXW8/4X5WYeEhISYO4nCvLy8aNasGV5eXjRp0sR4m0SluNpYYGAg7u7u1KlTh9u3b9O+fXvFmJJmzZrRtm1b3NzciI6OJjAwUDGmxMbGhieffJLAwEBSUlLw9vZWjCmpV68e7dq1w93dHbj/QEulmNq+pcULu3btGqmpqTg4OPDUU0/RoEEDk3G1sYqMf2BgIB4eHlhbWxfpW95YdbzGpcWVxvvmzZs4OjoWGcPiMSVNmjTB09MTa2tr/Pz8aNSoUYViwjJZ3JH0li1bcHNzw8vLi927d5caVxsLCwvjhx9+4IcffjA+RVgppmTp0qVs2rSJH374gTfeeMNkzNS2WFlZ4eTkxIkTJ0zGlERERGBtbc2FCxeIjY01GVPbV23/Xr16MXjwYLp3786OHTtKjauNVWT8w8LC2LdvX4m+5Y0pqezXuLR4cb169eKvf/1riTEsHlOydu1a3N3duXDhAsePH69wTFgmizuSTkhIID4+npMnT+Lg4GA8+lGKq43Vr1+fGzdukJKSQnZ2No888ohiTElF+qrNT0liYiKnTp0iLy8PW1tb2rdvrxhT27e0eGFRUVEkJCRw7tw5rKys6NSpk8m42lh1jL8lvcalxdUqoVNjAAAXqElEQVSMt6nXQM3rXJGYsFAGC3P48GHVcbWxmJgY488pKSkmY0oq0ldtftXRV23/mjr+lvQalxZX066y+1ZkHcL8LO5IWqvVsmvXLtLT0/H39y81Xp5Y69atTcZM5bNz584SfYvH1G6Lqe2r6r5q+xfetuJ9i8fLEyvPGKrpW1mvU3neH+Uda1PtKrtvRdYhzM/i5qQTEhLIzs7G09OTFStWlBqv7JiSCxcuMGrUKFxcXIztlGJKDhw4gFar5dSpU8Z2SjElOTk55OXlce7cOWM7pZiSu3fvMnLkSBo2bGg8Ra20ePHtHTZsGL/88kuRdSjFKzum5NKlS4waNQp/f39jzhWJKTlw4AC5ubkkJibSuXNnkzElqamp5ObmkpCQwIoVKzh06FCp8eKuXLnC0KFDadCggTFHpZiSP/74g+eff574+Hhju4rEhGWyuCKdkpLCa6+9xldffVXkAgCleGXHlPj5+REWFsbevXsZO3asyZiSV155hVdeeQUnJyfjl09KMSWHDx9mzJgx/P3vfzeeIqUUU3LmzBkATp48iZWV1QPjhd28eZO9e/fSs2fPIhc4KMXLEys4o0QppmTr1q2Eh4ezd+9e8vLyKhxT0qdPH5ycnHB0dCQ8PNxkTMkvv/zC2LFjee2118jPz+fq1aulxou7dOkSNjY2tG3b1njVoFJMyZkzZ2jQoAFPPPEE1tbWFY4Jy2Rxl4Wnp6cTHh6Oo6Mje/bsMRZBpXhlx5QcOnSIunXrcvfuXTZu3MjYsWMVY0o+//xzNBoNt2/fJjw8nLFjxyrGlGRkZBAeHo7BYCAzM9NkTMmNGzcIDw/nzp07quKFOTk5cePGDfr06UNiYmKp8cqOKenXrx//7//9v0qLKVm9ejV+fn5YW1tz5MgRWrdurRhT4uPjQ0REBHq9noCAAJ599tlS48X5+vqydu1aDAYDTZo0MRlT4u/vT0REhPHnisaEZbK4y8KFEEL8yeKmO4QQQvxJirQQQlgwKdJCCGHBpEgLIYQFkyIthBAWTIq0EEJYMCnSQghhwSzuYhaAuLi4Um/SLoSoGlqtlo4dO5o7DVGIRRZpOzs7AgICzJ2GELVOwXMZheWQ6Q4hhLBgUqSFEMKCSZEWQggLJkVaCCEsmBRpIYSwYFKkhRDCgkmRFkIIC1am86TT0tKwsrLCxcWlXCtbt24d2dnZ3Lt3jzfffLNcyxBCiNpEVZE+dOgQ33//PXZ2dmg0GnJzc3n++efp1KlTmVbWp08f3NzcmDdvXqnttFqtnFQvhBCoLNJ169Zl9OjRREVFYTAYGDNmDKmpqWVeWb169Vi6dGmpD28FueJQCHORgyPLo2pOukOHDqxcuZIJEyYwceJEVq1aVa7r+9955x30ej179+4tc18hhKiNVM9JazQaXFxc0Ov16HS6cq1syZIl5epnKfb90Jzez1wwdxpCiFpEdZEeMWIEs2bNAmDo0KFVlpAQQog/qSrSN27cwM3NjQkTJgD3j6qFEEJUPVVFOjo6mj/++KNIbNKkSVWSkBBCiD898IvD9PR0vL29uXPnDp06daJTp05cvHixOnITQoha74FH0g4ODqSkpKDVaklJSQHgb3/7W5UnJoQQQsWRtK2tLYMGDWLMmDEAGAwGOZdSCCGqiep7d4SGhnLu3DkATp06VWUJCSGE+JPqIu3q6krv3r2pV68ezs7OVZmTEEKI/1FdpL29vXF1dSUmJgYPD4+qzEkIIcT/qC7SSUlJ5ObmMnr0aJ5//vmqzOnBuUyPVhUTQoiaTnWR9vHxYe/evWzbto1vvvmmKnMqQoqvsn0/NDd3CkKIaqD6svCH+eIVS7onh7ly8dwfx/VeZb9plhCiasmTWUyoVUeqIfJFsBCWSnWRzs/PR6/Xk5CQwO3bt6syJ4v2ME+/PMzbJkRNpbpIT58+ndjYWNasWcP8+fOrMichzEp2VsKSqC7SdevWJTo6mnHjxmFtbV2ule3bt49169YRERFRrv6iBgtxxnN/nLmzEKLGUVWkN23ahK+vL/Xr1+fIkSP4+vqWa2WxsbGMHDmStLS0cvWvdGUoHNVSYKpgbrimFEZLOXqtKeMlag+NwWAwPKjR0aNHWbduHX/961+xsrLi66+/JjQ0tMwrCwsLY/LkyYSGhjJ16lST7eLi4ui4tQeePQ6w39POGA/Y+KRi7OywmCL9C9qtM7yEt9f2MuWotA5Tel3Xss7wEi33RJDxsnuZ2gVsfJJ93d0V+9Zbe6vU5RVsc+F2BXkf353xwL5lafegMSzrWCf/0b/INheMTXn6msqn+LiWNtbF+5p6LxUer4qMf00gzxe1LKpOwQsMDOSbb77h6aefxmAw8N1335VrZW3btiUiIuKBVyza2d0vkKZOCSv+JirtTVXeN5yqftf/POoqtf0D2pWILXjAekPSCQCSiFa3PBPM0S75D2i8oNufgf+Njdq+Fc3nQX3VvpcCAgIqPP6WSG6eZnlUnyc9ePBgZs+ejUajYciQIeVaWZ8+fcrVTygrUuy4v1NL2m0Z0wZq3d8RV/554cXHxlSsiJD0Mi1PiOqgek760qVLdOzYkQ4dOpCYmFjVeVWYXJhRuoe26JRSaCt7HQ/tGAqLoqpI+/v7Ex0djaenJ97e3sTExDy4U0WZ+GMrUXyr44/SBLU7gmrZYZRhHMy5A6uOqykt5epRISqDqiIdGBiIk5MTTz/9NE8++SQODg5VnddDySKOvMy4U6sM5R1DKdyipqrWOWlRg4Skw8NyOloN3zGJ2k11kXZ1dcXR0RGDwYCrq2tV5lRxFfyjlPnsMghJ5zqw74eqX5UcDYvaSPUVhytXrmTChAlMnDiRVatWVWVOppWx+FrS/GdNLDCysxLC/FQfSWs0GlxcXNDr9eh0uqrMqcapyFxzZc9TW8S890NGdlbCnFQX6REjRjBr1iyAh3tOWuYvhRAWRHWR7tixIx07yhFFZaqJUyC1jtqd9v/m5mvaxUTC8qmak161ahWHDh0iPT2dzMxMDhw4wGeffVbVuT3c5IhdCKGCqiI9btw4tFoty5Yt48MPP0Sv1/Pqq69WdW5CCFHrqZ7u6NWrF40aNSIzM5PAwEB++eUXOnToUJW5CSFErVemZxwePnwYT09PIiMjiY2NraqchBBC/E+ZinSbNm3w8/Nj+PDh3Lhxo6pyEkII8T+qi/S3335rvA90amoq+fn5nDhxosoSEzWLnKlyn5ynLiqb6jnpzz//nBdeeIGffvqJjIwMxo8fzxdffEGnTp2qMj8hhKjVVB9Jd+7cmUGDBnH+/HmSkpJwdHQs093wLl++zLx583j//fdlqkQIIVRSfSTdrVs3PvvsM4KDg0lPT2fRokW4ubk9sF9UVBTHjh0jNzeXd955h927d5OZmUnDhg0rlLioHjKNIYR5qXoQLcBPP/3E999/D0BQUBDdupV97m3btm1YWVnRv3//UtvFxcUZn3MohKheNfkZjQ8j1UfShw4d4r333gNg8eLFZS7SBw8e5JtvvqFTp05cvXoVX19fk23t7OzkjSKEGciDaC2P6iKdm5uLVqtFr9eTl5dX5hV1796d7t27l7mfEELUZqqKdHx8PM899xzjx49Hr9fz1ltvVXVeQgghUFmkL1y4/+XRwIEDAbh06ZJcEi6EENVAVZF+8cUXqzoPIYQQCsp0WbgQQojqJUVaCCEsmBRpIYSwYFKkhRDCgqk+T7o6abVaOaleCDPQarXmTkEUo/qycCGEENVPpjuEEMKCSZEWQggLJkVaCCEsmBRpIYSwYFKkhRDCgkmRFkIIC1Yri3RycjLJycmkpKRQcAZiwf/1ej35+fkl+uj1eo4ePVoklpOTQ25ubtUnLGSsRa1lcUVaqYBC6UW0rAX01KlT/Prrr8TGxhIaGgrA0qVLyc/PJycnh48//tjYdteuXWzatImQkBBcXV0BWLJkCenp6Xz00Ud89NFHZdqWsm6HElPtCm9zZaxbbWF8ULvy7BQLVNVYV6Sv7LBFdbK4Iq1UQEG5iFakgBoMBjQaDXq93hiLj4/n9OnTRdplZWWh1+vRaDTGdWg0Gi5cuECXLl2wtbUt07ao3Q4lptopbXNF1q12DNW2K8tOsbjKHuuK9K3IWAlRXhZ5WbhSAYX7RVSn0xn//aACevLkScXl+/v7U69ePaytrQkMDARg7NixJCcnk5eXx+jRo41tn3vuOZycnLh58yaHDh2if//+jBkzhqSkJJycnBg1atQDt0dpZ/Cg7VBiql1p21yedasZw7K0U5uLktdff52LFy/i7u7OkCFDSm1b8L4p/AmsYB0ajabC+VVkrIQoL+uQkJAQcydRWH5+Pr6+vnh7exMYGIiDgwMArVq1Ijs7G4CePXtib2+Pn58fjz/+OG3btiUmJoYWLVrQsmVLMjMzsbGxoXv37sb+ha1du5agoCDs7e2JiIggMDCQlStXMmDAANzd3QkPDzcW72XLltG5c2dsbW3Zt2+fse3AgQNxd3c39ldy6tQp8vPz8fLyonfv3sZcfv31Vzw9PXnmmWewt7cnMzOTBg0a0KpVK86ePUuLFi0Ul2eqnV6v59SpU/j6+tKjRw8cHBwqtG6l5SlR205tLkrCwsLo27cv27Zt4/Tp0ybH2sHBgUaNGtGoUSNycnJo1aoVgYGB3L59G2tra3Q6Ha1atar010ntGAhRXhY33bFjxw48PT1xdXUlKirKGA8PDycgIIA2bdoQGRkJwKpVq9DpdDg7O3P58mVjuzZt2tCmTZsi/Quzs7Nj586d7Ny50/gRWilWPG5nZ1dq2+IeffRRHn30UWxsbPj8888BcHZ2Nhb9NWvWAFC3bl3q1q2Lra0tCQkJJpdnqp2Li4txOqBgPRVZt9LylKhtpzYXU9u8f/9+vLy8Sh3rgwcP8vPPP/Ppp58SFBQEwObNmzl79ixbt241xiqSX0XGSojysrjpjoICqNPpFItl4bjaWHFDhw4lPz+f69ev06xZM5OxsrYtruDLJY1Gg42NDQBHjhxRFVNy5MgR48f2OnXqlIhrNBpj/OjRoyXaKvUvvO7S2pnKR03eBbkoraNwTImPjw8pKSm4u7vTp08fk+08PDxITU0lPz+fS5cu0a5dO9zd3UvESsuv8PaaGquCWOHXTs1YCVFeFveueuGFF3B0dCQ3N7fIbROV4mpjxa1du5Z//vOf/Pzzz2RkZNC+fXvFWFnbFvfKK68A9z9Ox8fHlymmpEGDBgBYW1vTs2dPxXiPHj0A6NOnDz4+PmRnZ3Pr1q0Kx9Ru34PyLshPKaakV69ewP2plaioKCZNmqTYLiMjA1tbWx5//HHjPLJSTMmAAQNwcnLi3LlzWFlZVTgmRGWyuHfVli1bcHNzw8vLi927d5caVxsrzsPDg6ioKLKysoxTGEqxsrYtLiwsjE2bNrFv3z5GjhxZppiSXr160atXL7p27cqOHTtKjW/ZsgUrKyucnJyKjE15Y2q3T23epraluC1btvD111/zzTfflDrWPXv2pFevXnTv3p2ffvrJZExJREQE1tbWJCYmcvz48QrHhKhMFnck7ezszBdffIFer8fJyanUuNpYca1ateLJJ598YKysbYt79tlnOX36NNnZ2ezevZu+ffuqjinZsmWL8aN14YKlFK/IeKkZQ1PbpzZvU9tSXOvWrVWN9ddff11ieUoxJQU73cLtKhITolIZLMzhw4dVx9XGivvtt98MkZGRhoMHD5YaK2vb4mJiYow/p6SklCmmpCrGprxjaDBULG+161A71hVZR3WMlRDlZXGn4Gm1Wnbt2kV6ejr+/v6lxtXGinN1daV9+/ZoNBq2bdtGhw4dFGNlbau0LTt37iQ9PZ3WrVuXKVZdY1PeMSzerqx5q11HWca6MrajqsZKiPKyuDnpc+fOce/ePRo2bGg81c5UXG1MycWLFzl06NADY2VtW1h+fj6jRo3C39/fmIvamJLmzZszcuRI/P39WbFiRanxyo4p2bx5M7m5uSQmJpbaLjU1lZEjR+Lg4GDcPqWYkrVr17Jx40ZiY2NLvSDlxx9/JCgoqMgYKsVM9bW1tcXPz69I3/LGhKhMFlekk5OT+fvf/87x48cJDg4uNa42Vtzrr79ObGwser3e2EYpVta2xS1btozw8HD27t1rPB1Pbaw0fn5+uLi4qIpXdqyw/v37c+XKFRo3bswbb7xhsl1iYiIAZ86cMY6XUkzJyy+/TH5+PjExMfj5+ZXabsmSJezYsYO8vDyTMSWvvvoq/fr1Y/v27Rw+fLjCMSEqk8U9iDYsLAwXFxfu3LmDi4sLY8eONRlXGytOp9Nx4sQJ4uPjsbW1ZfTo0YqxsrYtLj4+3nh6XsHPamNKPv/8c+Nlz/n5+bz++usm45UdUzJhwgRatGiBs7MzGo1GcawBVq9ejY2NDXl5edjY2DB27FjFmJK1a9dSp04ddDodt27d4q233jK5jmeffZYmTZqUGlPy6aefYmtrS9u2bXnsscfQaDQViglRmSyuSAshhPiTxU13CCGE+JMU6VpEq9Vy584dsy+jIlJSUsy2biHMQYp0DTN16tRy9925cydnzpwpV9+bN2/y1VdfGZexefPmchfMbdu28eGHHxaJ3bp1i02bNqHVannvvfeYMWOG8TalhRW+x3hZlHb2SYFly5Zx5coVwsPDuXfvXrnWI0Rls7grDmuruLg4Nm/ezN27d3n33XdZuXIlcP/qv3HjxvH+++/j4OBARkYGd+7c4YMPPsDOzo6ePXuSmZlJbGws9vb2PPHEE8Yr/xISEggPDycrK4tRo0Zx4sQJbG1teeqpp4D7RSkrKwuAYcOG8dtvv7F//35yc3OZOXMmkZGRZGRkYG9vz/Dhw7l27RqpqanY2trSoEEDcnJyWL58Obdv38ZgMDBnzhwGDx5M7969uXnzJkqn4J88eZLff/+9xL3CIyMjee2111i/fj29e/ema9eupKWlERcXZ7yJ1YkTJ0hISODixYtERUWh1+vx8fGhbdu2rF69mr59+7Jjxw4ee+wxLl26RMuWLbl06RKLFy/m0qVLbNmypcg4+fv7FxmfAs888wwbN25kzJgxlf0yC1FmciRtIQqe+PHOO+9w4cIFGjZsyOzZs7l9+za7d+/m6aefZtasWdjZ2bFjxw5GjRrFvHnz2LVrF3D/XhjTpk3jxx9/NC7T2dmZF198kfbt23P8+HEef/zxErfsHD58OG+88QYbN27ku+++Y968eYwePZpt27Zx7do1/Pz86N+/v7F94WXk5ORw48YNZs+ejZeXF2fOnMHb25s33niDzMxMxe187LHHGDRoUIl4wUMULl68SMeOHYH7F7I888wzxjadOnWiZcuW/P7776SkpODo6Gi8sVO3bt0YOnQodevW5V//+hf5+flMnDixxI38C49T8fEp4O/vz6lTp1S9bkJUNSnSFiI/Px+NRkN2djb169dXPJVLo9FgZWVlfDoI/PksPgcHB2xsbIo8lWTr1q0kJSUREBBgfGpJcQaDAa1Wi7W1dZG4Xq9n9OjRtGzZknnz5hkLXeFlFF+mXq833ry/8G07ExISmDx5Mhs3bjS5/QXrb968OXFxcQCkpaWxZs0a/vOf/zB58mTjXLhOp6NLly5MnTrVuMOoW7cuAI6OjsbxAIzjVaDwOBUfn8LktqPCUsg70UK89NJLzJ49m3v37jF79mxu3LjBwoUL8fDwoG/fvvz73//m5MmT5Ofn069fPxYsWMC2bdvo168fGRkZisv09PTk6NGj2NjY4ODggLe3Nxs2bDBOd8D9c4kBJk6cyNmzZ5kzZw46nY4pU6bw2WefkZ2dTfPmzY234SxYRrNmzXBwcMDd3Z0FCxag0WhMXhreqlUrwsLCSt3+hg0bkpGRwbBhwwgJCWHbtm1kZ2fz1ltvFbln961bt/Dx8WHHjh2cOXOGpk2b0rBhwzKNtanxKbiZ1MWLF2nTpk25lilEZZPzpGuxZcuWMXDgQIu458SdO3fYs2cPQ4cONXcqREREMGLEiFKfBCNEdZEiLYQQFkzmpIUQwoJJkRZCCAsmRVoIISyYFGkhhLBgUqSFEMKCSZEWQggL9v8BLMeiAGBmUCUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x205.714 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ylabel = '$\\log_2$(odds ratio)'\n",
    "ymin, ymax = -3, 3\n",
    "\n",
    "n = 44\n",
    "plotDf1 = plotData2.iloc[:n]\n",
    "plotDf2 = plotData2.iloc[n+1:]\n",
    "m = n - len(plotDf2)\n",
    "emptyRow = plotData2.iloc[0].copy()\n",
    "emptyRow[:] = np.nan\n",
    "emptyRow.name = ''\n",
    "plotDf2 = plotDf2.copy().append(m*[emptyRow])\n",
    "\n",
    "set_fontsize_paper()\n",
    "figsizeCoeff = 0.6\n",
    "fontSize = 16\n",
    "aspectratio = 1.4\n",
    "imagesize = 4\n",
    "linewidth1 = 0.5\n",
    "figsize = (imagesize, imagesize/aspectratio)\n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(2, 1, figsize=figsize)\n",
    "plt.subplots_adjust(hspace=0.5)\n",
    "w = 0.6\n",
    "\n",
    "for i, ax in enumerate(axes):\n",
    "    if i == 0:\n",
    "        plotDf3 = plotDf1\n",
    "    elif i == 1:\n",
    "        plotDf3 = plotDf2\n",
    "#     norm = matplotlib.colors.Normalize(vmin=-vmax, vmax=vmax)\n",
    "#     colors = cmap(norm(plotDf3.values))\n",
    "    plotDf3.plot.bar(ax=ax, width=w, linewidth=0, colors=paletteCat)\n",
    "    ax.set_ylabel(ylabel)\n",
    "    ax.set_ylim(ymin, ymax)\n",
    "    ax.xaxis.grid(False)\n",
    "    family = 'Liberation Mono'\n",
    "    ticks_font = FontProperties(family=family, size='x-small')\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    if i == 0:\n",
    "        ax.legend(loc='upper left', bbox_to_anchor=(1, 1), title='stop codon\\ncontext')\n",
    "    if i == 1:\n",
    "        ax.set_xlabel('codon at position -1 (C-terminal)')\n",
    "        ax.legend_.remove()\n",
    "\n",
    "\n",
    "title = terminus + '-terminal, ' + speciesSetName\n",
    "axes[0].set_title(title)\n",
    "\n",
    "speciesSetName = 'TGA_non_overlapping_start_codon'\n",
    "outputFilePrefix = speciesSetName\n",
    "plotsFolder = analysisCtermPath / 'overlapping_start_codon' / speciesSetName\n",
    "filenamePlot = outputFilePrefix + '_composition_bias_codon_' + terminus + 'terminal_barplot_stopCodonContext'\n",
    "print(filenamePlot)\n",
    "fig.savefig(plotsFolder / (filenamePlot + '.svg'), dpi=400, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Heatmap version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- adapt code to make 1 block for each synonymous group.\n",
    "- 2 rows of codons, as in the barplot.\n",
    "- internal rows shows stop codon context, UAA, UAG, UGA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData2 = plotDataCopy.copy()\n",
    "maskData = maskDataCopy.copy()\n",
    "\n",
    "nameIndex = plotData2.index.name\n",
    "nameCol = plotData2.columns.name\n",
    "stopCodonList = list(refCodonTableDf[refCodonTableDf['aa'] == '*'].index)\n",
    "\n",
    "# Remove values that are not significant using mask\n",
    "plotData2 = plotData2.where(~maskData, other=np.nan)\n",
    "plotData2 = plotData2.loc[plotData2.index.map(lambda x: x not in stopCodonList)]\n",
    "plotData2 = sort_codon_index(plotData2, refCodonTableDf,\n",
    "                             addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=True,\n",
    "                             addEmptyRowBetweenGroups=False)\n",
    "plotData2.index.name = nameIndex\n",
    "plotData2.columns.name = nameCol\n",
    "plotData2 = plotData2.drop('all', axis=1)\n",
    "plotData2.loc['N/AAC']\n",
    "# print(\"plotData:\\n\", type(plotData), plotData)\n",
    "plotData2.columns = [re.sub('T', 'U', col) for col in plotData2.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData2 = plotData2.rename(columns={'UGA no start overlap':'$^{\\dagger}$UGA'})\n",
    "cols = list(maskData.columns)\n",
    "maskData.columns = cols[:-1] + ['$^{\\dagger}$UGA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['R', 'K', 'H', 'D', 'E', 'S', 'T', 'N', 'Q', 'C', 'G']"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aaTable1 = aaTable[:11]\n",
    "aaTable2 = aaTable[11:]\n",
    "aaTable1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34\n",
      "34 34\n",
      "27 34\n",
      "34 34\n",
      "TGA_non_overlapping_start_codon_composition_bias_codon_Cterminal_stopCodonContext_heatmap\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnQAAAC7CAYAAADsSAOHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3XlYVGX/P/D3DDBsiiyyKKKiJmKilpqlLT6JiVu2uKWitlkqT0USiLlQ4RqIZfYYZmK4QZalYi6oYX2tXFMfdxMVUBZZBNlm/f3Bj3kkZ86cgQPO1Pt1XVyXzrznvu9zn0PenTnnc2Q6nU4HIiIiIrJa8vs9ACIiIiJqGC7oiIiIiKwcF3REREREVo4LOiIiIiIrxwUdERERkZXjgo6IiIjIynFB9zezdetWjB8/HuPGjcOSJUvuef+3337DrVu36t1+QUEB/vOf/4jOv/TSSwZfX7x4MUJDQ9G7d29MnDgRoaGhAICLFy/i5Zdfxvjx4zFhwgQcP35c/5nDhw+jX79+UCqV9R5/UzK27WKYO89ipKWlmczk5OTgxIkTkvTXkO0nIiLzcEH3N3LmzBns378fycnJ2Lx5M8rLy7F79+46me+//x7l5eX17sPT0xPTpk1r6FAxa9YsJCcnIzAwEElJSUhOToZKpcLcuXMRGxuLjRs3YsWKFfj000/1n9m9ezeCg4Px888/N7h/SyfVPN9t/fr1JjO///47Ll++LGm/RETU+Gzv9wBIOnv27MGYMWNgY2MDAIiOjoat7f928Y0bN/Dzzz/j9u3biImJQXR0NKqqqjB48GBMnjwZ48ePh1wuR3BwMDIyMiCXy6HVavHYY49h7969GDBgAEaOHInly5dj9OjRWLt2LaqrqwEAX3zxBf773//ik08+QVVVFQYOHIipU6eaNf4//vgDDz/8MHx9fQEA7u7uWLt2LQBAp9Ph/PnziI2NxYoVKzBw4ECDbcyaNQsODg7473//ixEjRmDy5MlYt24ddu7cCTs7O8TExODUqVPIyMhAcXExPDw8kJCQYLCtpUuX4sSJE3B0dMSiRYswd+5cxMXFwcXFBS+//DK+/PJLzJ49Gzdu3ICvry8WLlyI2bNno6SkBIGBgfp2EhIScOrUKdy5cwexsbH6hXdJSQlatWqFjz/+GG+++SZcXV1x4cIFTJs2DV27djU6zxcvXsT8+fPh7u6O7Oxs/Pjjj/eM/dKlS5g7dy50Oh3GjRsHpVKJixcv4rvvvkNlZSX27t2L0tJSvPfee5DL5YiLi4OdnR00Gg1KSkowYMAAeHp61mkzKysLUVFRqKqqwrhx4zB8+HC8++67uHPnDnx9fbFo0SKkp6dj1apV6Ny5s9F5TE1NRU5ODq5du4aePXsiKirKjKOEiIgM4Rm6v5GioiJ4e3vr/+7o6Ag7Ozv931u3bo0nnngCs2bNwurVq/Hvf/8bGzduxOHDh1FUVITCwkIkJiYiODgYt2/fxpdffon27dvDzc0NmzdvRkZGRp3+FAoF1q5dCz8/P5w6dQrXr1/HZ599hg0bNmDfvn1mjz83Nxdt2rQBAOzbtw+hoaEYMWIEcnJycOzYMXTv3h3+/v7IyclBRUWF0XaeeuopbNy4EVu3bkVhYSEyMjKwefNmLFq0CMuWLdPPxddff43S0lLk5ube08aZM2dQXFyMTZs2Yfr06fp5OXjwILKysuDn54d9+/ahQ4cOSE5ORqdOnXDgwAEAQGhoKN5++20AgEqlgo+PD9auXYs33ngD6enp+v7Xr18PJycnnDx5EuXl5ZgyZQo2bNiApKQkwXletWoVPvvsMyQkJKCoqMjgHBw6dAgjR47E+vXrYW9vj7Fjx6Jz58544YUXoNFokJSUhIULF2L79u0AgLZt22Ljxo0YO3YsXnvttXsWcwCwbt06REREYMuWLcjLy0NKSgoGDx6M9evXo23btti7dy/WrVuHjRs34vXXXzc6jwDQrVs3bNq0Cb/88gv4sBoioobjgu5vxNPTEzdv3tT/PTc3F2fOnNFfr5aSkqJ/7+rVq4iLi0NoaCiysrKQm5sLNzc3ODk5AQD8/f0hk8ng7u6Otm3bwsbGRn/mr5a/v7++X6VSCU9PT8ydOxcffvhhva5z8/T0REFBAQBg4MCBSE5ORrdu3aBWq7Fr1y4cO3YMr776KoqLi/ULI0P8/f2hUCjg6OiI7OxsBAQEQCaTwc/PD6WlpfeMvfbs192uXbuGbt26AQCCgoJw9epVBAcH46effsKePXswePBgXL16FTt27EBoaCh2796NvLw8ANCfYQQAW1tb3Lx5E++99x5++OEHaLVaAEDPnj0BAIGBgcjOzoa9vT26dOkCJycnk/Nce2avWbNmaN++vcE5GDVqFP78809MnjxZv821NBoNIiIikJSUpB9P7UJayLVr1xAYGAi5XI5///vfuHr16j1zJJfL4eDggI4dO8LBwcHgPN69Ta6urtBoNCb7JiIiYVzQ/Y0EBwfju+++0/8j/fnnn+PatWv669XGjh0LmUwGnU4HPz8/zJs3T/+6r68vZDKZvq27/yzWJ598gqVLl+Ktt94SPINmzMMPP4zff/9dvygtKyvD5cuXIZPJcPz4cWzevBlr1qzBypUrsWPHDqPt3D321q1b4+LFiwBqvjKsXbCa2r42bdrgzJkzAIBTp07B19cX7u7uUCqVOHLkCPr27Qs/Pz+MGTMGycnJeOuttxAUFHRP2+fOnUNhYSE+/vhj9O7dW3826tKlS/r3O3TogOrqamRlZaGiouKeBd1fNW/eHHl5eaioqMC1a9cMZjIyMvDCCy9gw4YN2Lhxo/7127dvIyMjA3FxcQgODr7n7Fjt8WGIr68vLl26BJ1Oh/DwcPj5+dWZozZt2kAmk6G8vBxZWVmorq42OI9/nSMiImo4XkP3N9K1a1c88sgjGD9+PDQaDfr06YOhQ4fek4mJicGiRYswZ84clJWVoVu3bmjRokWD+3/iiScwevRotGjRAs7OzlCr1fr3IiMjsXTpUsHPKxQKxMbGYt68eaiuroZKpcLzzz+PW7du6c8MAcADDzyA3NxcFBUVwd3dXbBNT09P9O/fH+PGjYNOp8OSJUvq3DlrTPfu3bFjxw6MHz8etra2WL58OQCgf//+OHnyJGxtbTFo0CBER0djwoQJsLe312fu1rZtW1y+fBnjxo1Dy5Yt4e7ujjZt2uC3337DkSNH0KlTJ/31dvHx8cjKysI777wjOLawsDCEhYXBw8NDv0D9qwceeABRUVFwdnbGsGHDAAA2NjbYsWMHbG1tMWbMGPj4+NxzJrVjx454//338dhjj8HPz6/Oe6+//jqio6NRXV2NMWPG4JlnnsF7772H1NRUtGvXDtOnT4enpydCQ0PRqVMnODg4GJzHDRs2mJx/IiIyj0zHC1ioCXzxxRd444037vcwLMJ3330HjUaD0aNH618LDQ3F2rVr69zEYsz333+PQYMGAagpDbJt27ZGGysREVkHnqGjJvHiiy9K3mZ4ePg9NfWSkpJMfmX5VwcPHsTq1avrvPbss8/WWXBZkubNmyM0NBRqtRpvvfUWEhMT7ynlEhUVpb92rT6kmlsiImoaPENHREREZOV4UwQRERGRleOCjoiIiMjKcUFHREREZOWs+qYI9U3hZ07atuqE6tJCwYy9iwdKEqMFM65TF9X0l3XGeF9+DwIAKnesFGzLcfgMlK6dK5hxefkj5Hwg/Ngs3/mJKFs3XzDTfPIHAABVgeFaZQBg59kOAKAszDaaUXjUFJ29WSL8DNhWrs64WnhHMNPeoxnUOecFM7a+XQAA6uyzxjNtugIAqu/cNpqxb1ZTiuVsbqnRDAB09XGBsiRfMKNw9cKNYuHtb+3mDNWtLMGMXcuaUiC1tfEMqX1sVvJx4/sEAEIfboNHPtwrmDk8bxA+zhD+PXnvqU547svfBDPfv/YoAOHfOdtWnQAAyqIbgm0p3FtDe+WoYEbeoTeUt28JZhQtWqKwTLjeoUfzmrIumbfKjGb8WzYHAJy5afw4ebCVCwBg36UCwf4GPuApeEwCNcdlRWWlYMbJ0VHwfSKiv+IZOiIiIiIrxwUdERERkZUzuaD74IMPEBkZiejoaERHC381KVZcXBwAYMWKFTh27BgiIiL070VGRuLoUeGvY4iIiIjof0xeQ+fl5YVp06Y12gAOHz6Mfv366R9s3qdPHxw/fhy9e/dutD6JiIiI/k5MLuj27t2Lc+fOwcHBAQBMPo9TjNoHc6tUKmRmZsLe3h67d+8GAKjValy4cAFKpRIKhaLBfRERERH93Zlc0H333Xc4d+4ctFotunbtKkmnAQEBSEpKwu7du/Hxxx+je/fu+q9hIyIicPLkSWRkZOifV0lERERExplc0MXGxqJVq1bQ6XRIS0tDZGRkgzsdPnw4AGDKlCn61+6+jq5Hjx4N7oOIiIjon8Lks1xjY2MxZ84cAMCSJUsQFRXVJAMjIiIiInFMnqFzdXVFTEwMZDIZPDw89K9rtVqkp6cjODgYALBp0yZMmDCh8UZqgObqCcH3bdo/BFXun4IZO5+OuBw2RjDT6bPUmv7+u894X90GAgBufRphNAMALd+Kw5Xw8YKZDgkbcf39VwQzbRd8hYrUxYIZpzGzAACl5caLmLo41xQwFVPoVkxh3Uv5xgu4AsADXs2h/fOwYEbe8REAENx3dj4dAUCwcLS9S83xKqb4rJjjpKBUuB1PFydorp8WzNi0DQIgXHxX4d4aALD3onCx40GdvTB01SHBzM43+2HXBeF2QgK8EL3TeBFnAFg0tOZyCzFFqoWOJaDmeNKc/1kwY9PlCVFFwavLioUzzd0AAOfzjBcN7uJdUzQ4/7bx49urhTMAiDq+xRwDYotrExGJJbigy8rKwsiRI1FZWQmdTgcnp5qq6wUFBVi2bBnOnj2LAwcOAADat2/f6IMlIiIionsJLuh++OEHHDx4EH379oVcLsfRo0exYcMGeHp6YtGiRdi/fz+efvppADU3TxARERFR0xMsLBwWFoaePXti5syZCA8PR1BQUJ33MzMzsWDBAkRHR5tVYsRYYeFTp07h008/xYcffgilUmnuthARERH9I5m8hq5du3aYNWsWZDIZevbsWec9rVaL6upqaDQayOX1e4rY3YWF09PTER4ejuLi4nq3R0RERPRPY3JBN2TIEKhUKuh0OgwePLjOe4GBgXj99dcBmPeVq7HCwjKZDAUFBfj6668xduxY+Pn5mbMtRERERP9IJk+DLV68GL169UKvXr0QHx9f572LFy8iPDwcaWlpuH37tuhO7y4sPGXKFLzyyivIz89HSEgIUlNr7ijlGToiIiIicUyeobOzs9NfO7dly5Y67+Xk5MDPzw/Dhg3DvHnzRHcqVFg4MDBQdDtEREREJKKwcHp6Onbt2gUACAkJ0dedA4AFCxYgLy8P/fr1w5EjR+45g0dEREREjc/kgm7nzp3o1KkTbty4gaqqKoSEhOjfu3r1Kq5duwatVovWrVsjICCg0Qd8NzHFYNUn9whmbHs8g/jmnQUzM8tqCqXeqTBeoLeZU02B3vz4dwTb8pq5HJXfLxfMOD73jqiMmIK5gLgitpprJ41mbNrVPIqtyER/7s2dUFFpfI4AwMnRUXAegf/N5ZHrxovG9mlbUzBWqECtTZcnAADKX1IE+1M8PhY7/LoLZoZnnULmLeGisv4tm0N75ahgRt6hNwDhotg27R8CANwsES7k3MrVGTvO5Qpmhgf6QHvpN+ExPfAoXlp3RDCzaXIfAED1HeOXVtg3awFA+HgDao65qj1rBDMOz7yKzSdzBDPjevhCefuWcF8tWtaMSSBXm7ktUIC7xf8vwK3KyxTsz87bX1TRZDG/J0RE5jD5leuhQ4cwdOhQdO7cGXPmzEFISAjy8vIQHx+P8+fP6xdxhYWF+Oqrrxp9wERERERUl8kFna2tLX788UfIZDJotVoAgLe3N5YuXYrjx4/j4YcfbvRBEhEREZFxJhd0s2fPxm+//Qa1Wo2YmJg67x08eBCbN2+GjY0NAGDRokWiOo2Li0NERARWrFiB6upqNGvWTH9zRWRkJMaMGYPevXubvzVERERE/0AmF3QKhQJPPvmk0feWLl1a785LSkrg4eGBN998EwUFBcjLy0OfPn1w/PhxLuiIiIiIRDK5oAOA8vJyyOVyOP7lQt3ff/8dZWVlcHBwAAC8/fbbojqtLSwsk8lga2uLffv24dq1a7C1tYVarcaFCxegVCrNepwYERER0T+VyQXdF198gZKSElRXV8PHxwdTp07Vv7dw4UJ9QeEWLVqI7rS2sLBCoYBWq0V+fj50Oh3y8/MRERGBkydPIiMjA4MGDarHJhERERH9s5hc0Nnb2yMqKgoA7vl6devWrSgqKgIAeHl54c033xTVaW1hYWN69Oghqh0iIiIiElGH7tlnn0VAQAB0Oh1Onz6NHj166Bd2sbGxmDNnDoCaxV5kZGTjj5iIiIiI6jB5hm7btm04d+4ctFotunbtqr/+DQDUajX+85//QKvVotJEoUwiIiIiahwmF3SxsbFo1aoVdDod0tLS6pyFKy4uxsiRI+Hs7IwvvviiUQdqiFC1faCm4r6YivQ5H0wVzPjOTwQAwScz1D6VQUx/ZckfCmaah86DOuuMYMbW70EUlAo/ucHTpWZM6uyzxttp07Umc+JH45mHhgAAqsuMP7kBAOybu0FZLPzkAoWbD64W3hHMtPdoBgA4m1tqNNPVxwUAoDy4yXhfT74EAKj4TviRdE4vzBS13w5dLRTM9GvvAdXhHwQzdo+MBACoju4wnuldc0nCxYsXBdvq3LkzTuSUCGYe8nUVNd9DVx0SzOx8sx8A4Scl2Hn7A4CoY1fMsbTuWJZgZnIvP6hzzgv35dulZkw3jM+lbeuaJ8WIeZqEJvOYYH82/r1EPcGmulT4WLJ38RB8n4jor0Td5frqq68CAJYsWVLndRcXF/31brV3uhIRERFR0zK5oHN1dUVMTAxkMhk8POr+X+OoUaMwb948aDQavPjii4Lt/Prrr8jPz8fIkSMbNmIiIiIiqsPkgq5Pnz4ICgqCVqvFoUN1v5bp0aOH6DtSvb29YWNjU+cpEY899hgOHTqEkpISzJo1CwqFgk+KICIiIjKT0QVdXl4e4uPjce7cOQQGBgKo+Yr1mWeeqVdHCoXinkLBe/fuxaxZs1BcXAy5XM4nRRARERHVg9EFnbe3N5YuXYqbN2/C29sbFy5cgI+PT707atOmDdq0aYN9+/YBALRaLRwcHFBQUICvv/4aY8eOxYEDB/ikCCIiIiIzyU0F4uPjcfz4cXz55ZdYtGhRgzusfUpEdXU1QkJCkJqaWjMQuRz5+fl45ZVXMGnSJGRkZDS4LyIiIqJ/ApOFhefPnw9XV1cMGzYMa9eulWRRR0RERETSMXmGrm/fvnBxcYFOp0P//v2bYkxEREREZAaTZ+iIiIiIyLKZPENHRERERJaNCzoiIiIiK8cFHREREZGV44KOiIiIyMpxQUdERERk5bigIyIiIrJyXNARERERWTmrX9AplUoolUpoNBqjmaqqKgCASqVCRUWFYHtqtRrp6ekG31OpVNBoNMjOzsbNmzdNjk2qthrajpg5AsTNk6XNt9TjsaQxSbnfalnSsdSUY7LW+SYiEsvqF3SJiYlITEzEypUrERcXZzCzbNkyqFQqqFQqrFq1ymDm2LFjWL16NWbNmgUvLy+DmSVLlqCsrAwpKSnYtGmT0TFJ1ZZU7YiZI0DcPFnafEs1Hksck5T7zRKPJc63uP+eEBGJYXu/B9BQYWFh+j8nJSUZzDRv3hx2dnawtbWFvb29wUxqair69OmDBx98EN27dzeYadGiBVxdXTFp0iSkpKQYHZNUbUnVjpg5AsTNk6XNt1TjuXtM3bp1s4gxSbnfmnLbzB2TpRwDlrhtRERiWf2jv+Lj4yGTySCTyRAQEIChQ4fekzl9+jT27dsHmUyGkJAQBAQEGGwrPz8fu3fvRkFBAd59912D7aSnp0Mul2Pw4MHo0qWL0XFJ1ZYU7YiZo9q2TM2Tpc33smXLAECS8Ug1JqnmSMy2mbN9lnYsNeWYpDq2m3rbiIjEsvqvXAMDAzFp0iS4uroiLy/PYCYrKwsTJ06Ei4sLDh06ZDCzc+dOyOVyqNVquLm5GW0nNDQULi4u+PXXX42OScq2atfbxtbdQUFBGD9+PNzd3bFz506DmZkzZ2LChAlo2bIlzp8/b7SvoKAgvPTSS3B3d0daWlq92xLTjlTb1rFjR4SFheHll1+GVqs1mMnMzMT06dMxefJkXLp0yeh4tm3bBjc3N4wYMQIPPPCA0bZmzJiByZMn4/LlywYzUh2TYrZN7PZJtW1SHkuANMeAVPMtph2xbUn53wAiIjGsfkGXnZ2Nli1bol27dnBycrrvGSnbeu+99/Drr79Co9Fg5syZBjORkZH6jKGzAGIzUrYlJiPVtt26dQsKhQKurq4oKioymCkoKDCZqc3Z2dnB1dUVxcXF9W5Lqv0vZtvEjkmqbZPyWJLqGLDW32+x/RERiWH119AplUqsXbsWGo3G6FmMpsxI2dZLL72Ew4cP49KlSzh9+jSCgoLuyYwbN06SjJRticlItW2VlZVISkqCWq02Oo9iMlK2JdX+v7svoSsjxOSkykh5LEl1DNw9l8buULXE32+x/RERiaKzchUVFbri4mJdYWGhrry8/L5natXm9u7d26CMTqfTKZVK3Z49e+qdqays1Ge2bt1qtA0xOaVSqVOr1brMzEyLyFRXV+vy8/N1RUVFRrdLTEbqtqTa/7dv3xZ1jIjJSZXR6cQdk6ZyYo43MZnq6mpdbm6uyf1mKqNUKnW3bt3SnT17Vnfjxo0G5aTKEBGJZfVfuSYkJMDZ2RkODg5GSwg0ZQaoKVmwZcsWLFy40GjJArGZL7/8EtHR0fD29q535u4yC1euXDE6bjG52lIL3377rUVkEhMTkZKSguTkZKOlJsRkpGxLyv2fmpoqmBGbkzJj6ngTmxNzvInJJCYm4ptvvjG530xllixZAhsbG+zcuVOwjIiYnFQZIiKxrH5BV1tCwNHR0WgJgabMADUlC1xdXQVLRDRlRuy4xeTuLrWgUCjueyYsLAxhYWF466230LJly3pnpGyrKfdtU/cn5Zik+p2Tar+JOd7E5qTKEBGJZfVlS8SUEGjKTC1TJQuaMiN23GLnwFSphabMiCntIbb8h5RtNeX+b+r+pBqTVL9zUu03sWVEmvL4JiIS7X5/59tQ27dvt6iMJY7p7zzubdu2SZKRsq1/+n5r6v6acr+JzUk5l0REYtjExMTE3O9FZUPY2tpi/fr1cHR0NHqtTlNmpO4vOTlZkkxTj7upMnZ2dpJkpGxLyv1mKtPU/VnisSTlfrO0bSMiEsvqv3KtrKxETk4OsrOzcebMGcyYMeO+ZqRsq7y8HM7Ozvjpp59w7tw5TJs2rV4ZpVIJADh06JDRjDm5Cxcu4MiRIygrK6t3pqqqCg4ODoLjFpPRarWQy+X4/vvvkZOTY3AexWSkbEuqzI0bN9C6dWvB7RebkzKj0+nw008/obS0VHBMYnL5+fk4dOgQbt68iYkTJ6J58+ZmZ2qPk/379+PChQuCx5JQRqVSQS6XY+PGjSgtLTV6nIjJSZUhIhLL6m+KWLRoEU6ePInff/9dvyC5nxkp21qxYgWys7ORlZUFW1vDJQPFZGofOr5+/XpUVVUZHbeY3GeffYZLly6hRYsWDcp8++23AICioiKD/4iLzSxduhS5ubn4448/UFZWVu+MlG1JlUlPT4dSqUS/fv0QGhpqdNxiclJlDh48iF9++QVHjhzBwYMHjY5JTO7TTz/F3r17UVpaisrKSqxcubJemdo7YR999FGUlpYa7EtMpvau6vz8fNy5c8fotonJSZUhIhLL6gsLd+jQASNGjEBSUhKmTJly3zNStvX2228jPj4e/v7+mDx5cr0zU6dOBQCsX78eEydONDpuMbnmzZvjmWeeAQAUFhbWO6PRaKBUKlFZWWm0qKqYjJ+fH3bt2oVp06Zh//799c5I2ZZUmV9++aXO4uPuh8ebm5Mqo1AocOPGDUycOFHwLlcxOQcHB0yYMAFAzcItODi4XpnaO2FtbW3RrFkzg32Jydx912lKSorRbROTkypDRCSW1S/o7ty5g8TERAA1Z5gM/SPUlBkp21qzZo3+UU0NydT2Y2rcYnJlZWV1coaIyZSWlkqScXd3R2lpKdzd3RuUkbItqTKjR4/GoEGDBMcsNidV5saNGwCA3377Db/99pvRY0lM7oEHHsCyZcug0+nwyCOPGFz4ickMGDAAy5cv198Ja4jYTEJCgv6uU2PE5KTKEBGJZfXX0BERERFZktjYWNy6dQsAsHz58ibpkws6IiIiIhOiFB0Mvv7mhX0AAH9//zqvz5s3D7m5uSa/aZKK1d8UQURERNTYHORygz/+/v73LOYOHDiADz/8EPn5+U02Pqu/ho6IiIiosTnYyERnt23bho0bN6Jt27aNOKK6+JUrERERkQkJLp0Nvh5eerGJR2IYz9ARERERmaCQiT9Ddz9wQUdERERkgoONZd92wAUdERERkQmOZlxDdz9wQUdERERkgoOcCzoiIiIiq6ZQ2NzvIQjigo6IiIjIBDsHy14yWfboiIiIiCyAnZNlL5kse3REREREFsCWZ+iIiIiIrBsXdERERERWzpwF3TvvvIPy8nJUVFRgw4YNjTiq/+GCjoiIiMgEWwfDd7lmZmYCAPz9/fWvTZ06FZ06dcKoUaOaZGwAF3REREREJhk7Q3f3Qq6Wj48PJkyYgHnz5jX2sPQs+zkWRERERBbA1sHW4I8hL774IrRaLb766qumG1+T9URERERkpWzs7URnDxw40IgjMYwLOiIiIiITbB0U93sIgrigIyIiIjLBhgs6IiIiIuvGBR0RERGRlbNztG/0Pq5cuQIbGxu0a9fO7M9yQUdERERkgo1D4y3oVq9ejR9++AEKhQIymQwqlQrjxo3D+PHjRbfBsiVEREREJtg42Bn8kYKbmxuioqKg1WqhUqnw/vvvw9bWvHNuPENHREREZEIerBitAAAdeElEQVRj3uU6atQoDBw4EKtXr4ZOp8OMGTOwa9cus9rggo6IiIjIBLl9419D17ZtW6jVami1WrM/a9ULuj8LygTf7+jZHD9fKRTMPNHBA1cL7whm2ns0AwCcyCkxmnnI1xUAUF1WLNiWfXM3/HAmVzAz8kEfVFRWCmacHB1RWFYhmPFo7gQAuJRvfJ4e8GoOAFj561WjmRmPtQcgbttKy4XH7eLsiMxbwvvNv2XNmC5evGg007lzZwDitk2ddUawP1u/B3H65m3BTFCrFvjtWpFg5tF27lAW3RDMKNxbAwCqS40fl/YuHgAA5e1bwm21aIkbxeWCmdZuzqK2Lf+2cDteLZwBAMnHs41mQh9uAwDILhb+fWrj1gzrjmUJZib38hPc/0DNMSAmA4g7lhqaMWdMYsdNRJZDrmjcBd0bb7yBIUOGAABee+01sz9v1Qs6IiIioqYgUziIzhYUFGDMmDGYP38+BgwYYDJ/9uxZtGvXDh999FFNXzKZ2ePjgo6IiIjIBJmRM3SZmZkAAH9/f/1rFy9eRJcuXUS3vX79emRn1/0GpG/fvmaNz+Rdrh988AEiIyMRHR2N6Ohosxo3Ji4uDgCwYsUKHDt2DBEREfr3IiMjcfToUUn6ISIiIpKCTOFg8Mff37/OYg4A+vfvD3d3d1HtZmVloWPHjigpKcFTTz2Fp556CllZwpelGGLyDJ2XlxemTZtmdsNiHT58GP369UNeXh4AoE+fPjh+/Dh69+7daH0SERERmcPYGbqGcnNzw9WrV6FSqXD16lXIZDK88sorZrdjckG3d+9enDt3Dg4ONd8dL1261PzR/kXtd8MqlQqZmZmwt7fH7t27AQBqtRoXLlyAUqmEQmHZj9kgIiKifwaZrXlrkgULFojKNWvWDB999BEOHDiAAwcOQKvV4vDhwwgNDTWrP5MLuu+++w7nzp2DVqtF165dzWrcmICAACQlJWH37t34+OOP0b17d/3XsBERETh58iQyMjIwaNAgSfojIiIiagiZvfibIupjzpw58PHxQWBgIM6ePWv2500u6GJjY9GqVSvodDqkpaUhMjKyXgO92/DhwwEAU6ZM0b9293V0PXr0aHAfRERERFKR2Ts2avvNmjXD888/j7y8PDRr1szsz8t0Op1OKBAbG4s5c+YAAJYsWYKoqKj6jZSIiIjISmkv/WbwdfkDj0rS/uTJk/Hqq68iLi4Onp6eWLNmjVmfN3mGztXVFTExMZDJZPDw8NC/rtVqkZ6ejuDgYADApk2bMGHCBDOH3zC3TRSxbeHsiKwi4SKnfu7NRBf5FCri2tqtpvCqUMFYoKZo7LFs4wWKAaBXG1dU3xEuBmvfrIXogsjXBXJt/39m+S9XjGbeebwDAEB1S/iuG7uWfqguFy4abO/cXHRBZKFCxvbN3QBAsEhxbYFi7RXhu6blHXrjpz+Fi/gO6NhSXGHhQuOFdwFA4VFTfFdoDmq3v6pCeP86ODXD2dxSwUxXHxd8e1q42PGLQa1F7TcA2HUh32gmJMALAEQVxd74h/A8je/ZhoWFicii6GyleW6rMTdu3EBFRQVmz54tfR26rKwsjBw5EpWVldDpdHByqvmHpqCgAMuWLcPZs2dx4MABAED79u3NHz0RERGRNbBp3Bs1fXx8sGHDBgA1N4+aW4dOcEH3ww8/4ODBg+jbty/kcjmOHj2KDRs2wNPTE4sWLcL+/fvx9NNPA6i5eYKIiIjo70hn07hn6JKTkxv0ecHCwmFhYejZsydmzpyJ8PBwBAUF1Xk/MzMTCxYsQHR0tFklRowVFj516hQ+/fRTfPjhh1AqleZuCxEREVGj0NnYGfyxFCavoWvXrh1mzZoFmUyGnj171nlPq9WiuroaGo0GcrnJh04YdHdh4fT0dISHh6O4uLje7RERERFJzsw6dOaqqqqCra0t9u7di8DAQLMvZTO5oBsyZAhUKhV0Oh0GDx5c573AwEC8/vrrAMz7ytVYYWGZTIaCggJ8/fXXGDt2LPz8/MzZFiIiIqJG0dhn44YOHYpXXnkFa9asgVwux759+8z6vMnTYIsXL0avXr3Qq1cvxMfH13nv4sWLCA8PR1paGm7fFr4r8253FxaeMmUKXnnlFeTn5yMkJASpqak1A+MZOiIiIrIQjf2Vq6OjI9LS0jB79ux6rYFMnqGzs7PTXzu3ZcuWOu/l5OTAz88Pw4YNw7x580R3KlRYODAwUHQ7RERERE3CjLtcExIScO3aNSiVSnz++ecm81FRUfD19YVMJsP27dvh6+tr9vBMFhZOT0/Hrl27AAAhISH6unNAzXPK8vLy0K9fPxw5cuSeM3hEREREfweVRmps5ubmAgD8/f31r40ePRrffPMNxo4di5SUFJNtr1u3Dl999RXGjRsHGxsbbN68Gfv37zdrfCbP0CmVSkydOhU3btxAVVVVnfcmTJiAa9euQavVYurUqWZ1LAV11hnB9239HoTm+mnBjE3bIPzf408IZvr/8jMACBafHdCxJQBAk3lMuD//XtBcPSGcaf8QroSPF8x0SNgoqhguABy6arzYcb/2NcWi1x0zXjR4cq+aaxm1fx4W7E/e8RFoLv6fYMamc39RhW4BIP+28ULOXi1qCjkLFdbt6uMCAMj5QPjY9J2fiAEJBwUzP4U/ickbhPftugm9BIs4A/8r5FwkUFjY/f8XFj5y3XhhZQDo09ZNsLAyUFNc+VK+cOYBr+aiihgDIgs5G6mmXkv+wKMYkfirYGb71MdYWJiILIpKa/j1uxdytWxta5ZXJs6Z6U2ePBkbN27E66+/Dq1Wi61bt5o9PpMLukOHDmHo0KHo3Lkz5syZg5CQEOTl5SE+Ph7nz59HQEAAAKCwsBBfffWV2QMgIiIisnRqrbjFGQD06tUL06dPh6enp+jPTJ48GcOGDYNMJsNrr71m9vhMLuhsbW3x448/QiaTQautWZ56e3tj6dKlOH78OB5++GGzOyUiIiKyJuYs6GrvCxArKiqqTnm4o0ePYtSoUWa1YfI2itmzZ8PZ2RkKhQIxMTF13jt48CAiIyMRHR2N6Oho0Z3eXVg4Li4Oq1atQlpaGgAgMjISR48KP3uTiIiIqCmptYZ/pNC1a1f8+uuvaN++PTp27IjDh4UvcTLE5Bk6hUKBJ5980uh7S5cuNbvTWiUlJfDw8MCbb76JgoIC5OXloU+fPjh+/Dh69+5d73aJiIiIpGTOGTpzNck1dABQXl4OuVwOR0fHOq///vvvKCsrg4ODAwDg7bffFtVpbWFhmUwGW1tb7Nu3D9euXYOtrS3UajUuXLgApVJp1uPEiIiIiBqLqhEXdEATXEP3xRdfoKSkBNXV1fDx8alzN+vChQv1BYVbtGghutPawsIKhQJarRb5+fnQ6XTIz89HREQETp48iYyMDAwaNMjsDSIiIiKSWmOeoQMAPz8/ODo6QqvV1utJWSYXdPb29oiKigKAe75e3bp1K4qKigAAXl5eePPNN0V1WltY2JgePXqIaoeIiIioKUh1vZwxMTExWL16NXQ6HWbMmKGvASyWycLCzz77LAICAqDT6XD69Gn06NFDv7CLjY3FnDlzANQs9iIjI+u5GURERESW60ROicHXH/J1laT9gQMHYvfu3VCr1Xj22WexZ88esz5v8gzdtm3bcO7cOWi1WnTt2lV//RsAqNVq/Oc//4FWqzVaQbkxac7/LPi+TZcnoD4pPCG2PZ4RlQGA5OPZRjOhD7cBAKiO7hBsy673cHGZ3D+FMz4dsfN8nmBmaBdvAMDlAuPFYDt51hSD3Xsx32hmUGcvAMLFcIGagrhCxYCBmoLAYsctVOy2ttDtjWLj/bV2qyk+rL55WbA/21ad8NiCdMHMr+8HI3KHcCHrpcMfxO1y4d+DFs4116FWlxkvGmzf3A0AcD7PeNFkAOji7SJqn+SZ2CfeLZyhyssUzNh51xTOFFMQObtYuEhxG7dmooo0s7AwEVmS6kY+RffGG29gyJAhAIBXX33V7M+bXNDFxsaiVatW0Ol0SEtLq3MWrri4GCNHjoSzszO++OILszsnIiIisgZVjbygGzNmDMaMGVPvz4u6y7V2pbhkyZI6r7u4uOivd6u905WIiIjo76YxF3SvvvoqHn30UYSEhEChUOCbb77B6dOnzTpZZnJB5+rqipiYGMhkMnh4eNR5b9SoUZg3bx40Gg1efPFFwXZ+/fVX5OfnY+TIkaIHR0RERGQJ6vuV64IFC3D27Fls2LDBaGbNmjX45JNPEBYWBplMhn/9619mf/NpckHXp08fBAUFQavV4tChQ3Xe69Gjh+g7Ur29vWFjY4O4uDhERERgxYoVeOyxx3Do0CGUlJRg1qxZUCgUiIyMxJgxY1hYmIiIiCxGtcbwgi4zs+YaZH9/f/1r4eHhOHHiBHQ6Hf7973/j5MmTJtt/++230bFjRxQWFmLy5MnYsmWLWY//Mrqgy8vLQ3x8PM6dO4fAwEAANV+xPvPMM6Ibv5tCobinUPDevXsxa9YsFBcXQy6X80kRREREZJHUWsMLursXcrUSEhLq/D01NVVUH9u3b8fLL7+Mt956CyqVSpoFnbe3N5YuXYqbN2/C29sbFy5cgI+Pj+iG/6pNmzZo06YN9u3bBwDQarVwcHBAQUEBvv76a4wdOxYHDhzgkyKIiIjI4jT2Xa4A8Oijj+LRRx9F9+7dERoaatZn5aYC8fHxOH78OL788kssWrSo3oOsVfuUiOrqaoSEhOhXrXK5HPn5+XjllVcwadIkZGRkNLgvIiIiIilUq7UGf8QQc4bugw8+QPv27QEAly9fhkajwcaNG0WPz2Rh4fnz58PV1RXDhg3D2rVrJVnUEREREVmT5b9cMfj6O493kKT9Hj16oH///pDJZCgrK8PcuXMxd+5cbN68WdTnTZ6h69u3L1xcXKDT6dC/f/8GD5iIiIjI2lSpNAZ/pNKxY0dERUXh8uXLuHnzJjw8PODk5CT68ybvch06dKj+zwEBAfUbJREREZEVq1RKt3gzJCQkBJGRkQgPD0dubi4mTZoEd3d30Z8XVViYiIiI6J+sUsKzcYZotVrcuXMHK1aswPDhw7Fjh/BjQv/K5FeuRERERP90FdUagz9SSU9PR1paGtLS0rB//36zP88zdEREREQmVCjVjdq+SqVCaWkp1Go1VCqV2Z/ngo6IiIjIhMa8hu7bb7/F6NGjERwcDJ1Oh6ioKLPb4IKOiIiIyITGvIbuxIkTAKB/GtfJkyfNekoEwAUdERERkUlVjXiGLjY2tsFt8KYIiSmVSiiVSmg0xnd8VVUVgJrvyysqKky2qVarkZ6ebvA9lUoFjUaD7Oxs3Lx5s36DJovEfUtEZDmUKo3BH0th9Qs6qRdQDV08JSYmIjExEStXrkRcXJzBzLJly6BSqaBSqbBq1SqjYzl27BhWr16NWbNmwcvLy2BmyZIlKCsrQ0pKCjZt2mS0raacJzEZsYsVS9u/YjJixixm3E29by1tromILIlapTX4I+T//u//8Nxzz2HEiBE4e/Zso47P6hd0Ui2gpFo8hYWFISwsDG+99RZatmxpMNO8eXPY2dnB0dER9vb2RrctNTUVbm5u6NatG7p3724w06JFC7i6umLSpElQKBRG22rKeZJyIWpp+1dMRsyYxYy7qfetpc01EZElUas0Bn8yMzORmZlZJxseHo4BAwYgKioKH3zwAfz8/HDr1q1GHZ/VX0MXFham/3NSUpLBTO0CytbW1ugCKjU1FX369BG9eEpJSTGYWbZsGQBAJpMZfbLGgAEDsHz5cshkMoSEhBjbNCxZsgT5+fnYvXs3li1bhnfffddgWwkJCZDL5Rg8eLDRtqSepwcffNDoPInJiJnLph63mDGJyYgZs5hxN/W+tbS5JiKyJMbOxvn7+9/zWkJCgv7PMTEx8PHxwZNPPtloYwMAmU6n0zVqD43srwuoux9VVuv06dPYt2+ffgFlbKFVu3gqKCgwuHg6ffo00tPT9f/AdunS5Z7MDz/8gCFDhqCiogK//PILhg8ffk9m27ZtCAkJEczU5oYMGYLy8nL8/PPPGDFiRL3bEjtPtcUMGzJPYjJi5rKpxy1mTGIyYsYsZtw7d+7EI488gu3btwMAXn75ZYPtxMfHQyaTNfh3wNLmmojIkvSau8vg68c+Mn5i5vPPP8e6desQGBiIadOmoW/fvo01POs/Q9exY8c6CyhDMjMzMX36dH3G0D9CtYunESNG4OeffzbazowZM/TtGPpH6NatW1AoFFAoFCgqKjLYTkFBgclMbc7Ozg6urq4oLi5uUFu163a5XA4PDw+DmT179gAAbGxsUFJSYjATHx+vz/Tv37/emdq+ZDIZbt++bRHjFjMmMRkxYxYz7uzsbAwdOhTt2rVDQUGB0XZmzpyJvLw87NmzB+fPnze4EAsKCoKXlxf27NmDtLQ0g78DljbXRESWxNT1coZMnz4d06dPb4TR3MvqF3RSLaCkWjxVVlYiKSkJarUaxk5+islI3dbMmTMB1DwrLjEx0eD/JVhaxhLH1JTbplQqsXbtWmg0Gmi1xv9DEhkZiX79+kGj0Rg9YyomY4lzRERkKdQWdEerIVa/oLt7QWPsH72mzLzxxhsoKytDbm4u3Nzc6p2Ruq3NmzdDJpNBq9UaXYw2RqawsLDeGSnbaoxMQ7ZfTG7q1KkoLi6GQqEQ3Lfjxo3D4cOHcenSJZw+fRpBQUH1ytSOR6fTmdz+psgQEVkSlYTPbW0MVn8NXWVlJVQqFdRqNRwcHODk5HRfM7GxsQgLC8OaNWsgk8kMng0Rk5G6rZycHMhkMgDAqVOnDN6MIVXmypUrkMlkcHJygqenJ+Tye2+mFpMBgMuXL8PR0REymQyurq4G57wpM1Jtv5jcZ599BqDmLJZSqURERITBdpRKpf7PBw8eRHBwcL0ymZmZsLGxQX5+Pnx9fdGqVav7miEisiQd3vjG4OtXvhjdxCMxzOrLliQkJMDR0REODg5Gyyg0ZUZMqQmx5SikbCs3NxdpaWlYtmwZWrdu3aiZnTt3Ii0tDSkpKfoL7euTAWrumPTy8oKrq6vROW/KjFTbLyYnpgQO8L9yI6tWrcIff/xR78yGDRvg4uKCjIwMo6VEmjJDRGRJNEqlwR9LYfVfuYopo9CUGTGlJsSWo5CyLTFlWaTKiCl/IVVpj6bOSLX9YnJiSuCI7U9MRqqyLVJliIgsiUZZdb+HIMjqv3IVU46hKTM7duwwWjrEnIzUbQHSlBsRkxFT/sKc0h6WtH+l2n4xue3btxssVVOf/sSWJJGibItUGSIiS9Jq9EqDr9/8ZkYTj8QInZXbvn27RWX+/PNP3aeffqo7efJkgzJSt9WUc7Bt2zZJMlKOydK2X0xO7L6Var4tbR6JiCyJ57MfG/yxFDYxMTEx93tR2RC2trZYv349HB0d4e3tfd8zbm5u6Nu3L86dO4cff/wRffr0qVdG6rZsbW2RnJxscvukyNjZ2ZmcJzGZ2v4saf9Ktf1icmL3rVTzbWlzTURkSRZ/vR86jfqen6iJjfsECNHu94qyoSorK3WXL1/WHThwQPf555/f90wtpVKpS0tLa3BG6rbEjN3SMpY0ppycHF12drZu/fr1RjNarVZUX6Zy69ev1x06dEiXm5urUyqVRtu5fv26yf7EZA4ePKjTaDQWkyEisiQtnn7f4I+lsPqbIhYuXIigoCBoNBq89tpr9z1z7NgxHD9+HBcuXMCkSZPqnZG6rc2bN+OFF15Av379kJWV1aiZWgMGDJAkI2VbDc0sWLAAAPD4448bfZrCxo0bMWHCBOTk5Bg9TsTkgoODcfbsWSxatAi3b9/G2rVrDbazf/9+6HQ6jBkzBr179653BgCWL18OhUIh+MzBpswQEVkKjdr8O1qPHTuGlStXoqioCKmpqYLVKBrK6suWdOjQASNHjoRGozH6pISmzKSmpsLNzc3knaKmMlK3tX//fn3pCmOPSJMqk5mZqf8x9tQNMRkp25IqExoaisDAQDg7O2PMmDEGMxqNBkqlEkqlUvDpHaZy3t7euH79Ojp16oQJEyYYbee5555DTk4OEhMTjd7BKibzxBNPYPz48XB0dMT69evve4aIyJKUZSw1+FP778bdwsPDMWDAAMycORNPPfUUysrKBJ/4IwWrv8u1tvhqrbvLM9yPDCDd3aRStnX58mV06tTpnj83Rmbr1q11/v7888/XKyNlW1Jl5s+fD09PT/3fG3KcmMolJCTAzs4OdnZ2sLe3x5QpUwy2s3LlSkyePBnNmjUz+L7YzOLFi+Hn54chQ4bA3d39vmeIiKzdiRMn0K5dO3z++efo0qULRo0a1Wh9Wf2CjoiIiMgS/fTTT/j0008hk8mwatWqOicEpMYFHREREZGVs/pr6IiIiIj+6bigI6tXXV2NkpKS+95GQ+Tl5d23vomIyPpxQUcWISIiot6f3blzJ86ePVuvzxYUFCA1NVXfxpYtW+q9uNq2bRs++eSTOq/dunUL33zzDaqrq/HBBx8gOjoaU6dOxZUrV+rk4uLi6tXnypWGH0VztxUrVuDatWtYu3YtKisr69UPERFZNquvQ0eW7Y8//sCWLVtQWlqK+fPn4/PPPwdQ83D2N954Ax9++CEcHR1RVlaGkpISfPTRR7C3t8eAAQNw584dHD9+HA4ODujTpw8GDx4MADh//jzWrl2L8vJyTJw4EceOHYNCoUC/fv0A1CxgysvLAQBjx47F5cuXceDAASiVSsyePRvJyckoKyuDg4MDxo0bh5ycHBQWFkKhUMDNzQ1VVVX47LPPUFxcDJ1Oh3nz5uGFF17AwIEDUVBQAEMPVzlx4gSuX79+z23pycnJeP3117Fp0yYMHDgQjz/+OIqKivDHH3+gQ4cOAGrqFJ0/fx5XrlzBxo0bodVq4evri27duiExMRGDBw9GWloaHnroIWRmZiIgIACZmZn4+OOPkZmZie+++67OPLVr167O/NR6+umnkZKSYvRuWSIisl48Q0eN6ptvvkFMTAzmzJmDP//8E15eXpg7dy6Ki4uxe/du9O/fH++//z7s7e2RlpaGiRMnYuHChfjxxx8BAP/6178QFRWFn376Sd9mixYt8Pzzz6N79+44evQoHn74YQQHB9fpd9y4cZgxYwZSUlKwa9cuLFy4EJMmTcK2bduQk5ODtm3bYvjw4fr83W1UVVUhPz8fc+fORatWrXD27Fm0bt0aM2bMwJ07dwxu50MPPYTnnnvuntezs7PRrFkzXLlyBT179gQAuLu74+mnn9ZnevXqhYCAAFy/fh15eXlwcnLCqVOnANTUaxszZgycnZ3xzjvvQK1WY/r06ZDJZHUWj3fP01/np1a7du1w+vRpUfuNiIisCxd01KjUajVkMhkqKirg4uICmUx2T0Ymk0Eul0Or1erfr7352tHREXZ2dnUK737//ffIzs5GYGAgdDqdwTZ1Oh2qq6thY2NT53WtVotJkyYhICAACxcu1C+K7m7jr21qtVo4ODgAqHkGaa3z588jPDwcKSkpRre/tv+OHTvijz/+AAAUFRVh9erVWL58OcLDw/XX7mk0Gjz22GOIiIjQLy6dnZ0BAE5OTvr5AKCfr1p3z9Nf5+dud4+fiIj+Pvhfd2pUL774IubOnYvKykrMnTsX+fn5WLJkCTw9PTF48GAsWLAAJ06cgFqtxrBhw7B48WJs27YNw4YNQ1lZmcE2fXx8cPjwYdjZ2cHR0RGtW7fG5s2b9V+5AkBiYiIAYPr06Th37hzmzZsHjUaDmTNnYs2aNaioqEDHjh0hl9f8P01tGx06dICjoyNatmyJxYsXQyaToWvXrgbH0aVLFyQkJAhuv5eXF8rKyjB27FjExMRg27ZtqKiowLvvvqv/yhWoudbO19cXaWlpOHv2LPz9/eHl5WXWXBubn9piwleuXMGDDz5YrzaJiMiysQ4d/e2sWLECzz77LNq1a3e/h4KSkhLs2bPH6GPCmlJSUhLGjx/fqM8SJCKi+4MLOiIiIiIrx2voiIiIiKwcF3REREREVo4LOiIiIiIrxwUdERERkZXjgo6IiIjIyv0/JszCEu/0FAYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 504x216 with 4 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fontSizeLabels = FontProperties(size='small').get_size()\n",
    "\n",
    "linewidth1 = 0.5\n",
    "aspectratio = 1.5\n",
    "figsize = (7, 3)\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "nmax = max(len(aaTable1), len(aaTable2))\n",
    "nmax = max(plotData2['aa'].map(lambda x: x in aaTable1).sum(), plotData2['aa'].map(lambda x: x in aaTable2).sum())\n",
    "print(nmax)\n",
    "fig = plt.figure(figsize=figsize)\n",
    "gsOuter = gridspec.GridSpec(2, 1, height_ratios = [1, 1])\n",
    "\n",
    "axes = []\n",
    "# for i, codon in enumerate(codonList):\n",
    "for i, aaTable9 in enumerate([aaTable1, aaTable2]):\n",
    "\n",
    "    gs = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=gsOuter[i])\n",
    "    ax = fig.add_subplot(gs[0, 0])\n",
    "    axes.append(ax)\n",
    "    ax.set_facecolor(colorBackground1)\n",
    "    plotData4 = plotData2[plotData2['aa'].map(lambda x: x in aaTable9)].drop('aa', axis=1)\n",
    "    maskData4 = maskData\n",
    "    maskData4.index = [re.sub('T', 'U', c) for c in maskData4.index]\n",
    "    maskData4 = maskData.loc[plotData4.index.map(lambda x: x[2:])]\n",
    "    maskData4.index = plotData4.index\n",
    "    maskData4.columns = [re.sub('T', 'U', c) for c in maskData4.columns]\n",
    "    maskData4 = maskData4[plotData4.columns]\n",
    "    print(len(plotData4), nmax)\n",
    "    if len(plotData4) < nmax:\n",
    "        ndummy = nmax - len(plotData4)\n",
    "        s = plotData4.iloc[:ndummy, :].copy()\n",
    "        s.iloc[:] = np.nan\n",
    "        s.index = ndummy*['']\n",
    "        plotData4 = pd.concat([plotData4, s])\n",
    "        s = maskData4.iloc[:ndummy, :].copy()\n",
    "        s.iloc[:] = True\n",
    "        s.index = ndummy*['']\n",
    "        maskData4 = pd.concat([maskData4, s])\n",
    "        print(len(plotData4), nmax)\n",
    "    plotData4 = plotData4.T\n",
    "    maskData4 = maskData4.T\n",
    "\n",
    "    # Create axes for the colorbar for more precise positioning\n",
    "    # vertical bar\n",
    "    cbar_aspect = 1\n",
    "    cbar_length = 0.4\n",
    "    cbarLabel = '$\\log_2$(odds ratio)'\n",
    "    ax = seaborn.heatmap(plotData4, square=True, mask=maskData4, ax=ax, cmap=cmap,\n",
    "                         cbar=True, vmin=-vmax, vmax=vmax, cbar_kws={\"label\":cbarLabel, 'aspect':15},\n",
    "                         xticklabels=True, yticklabels=True,\n",
    "                         linewidth=linewidth1, linecolor=colorHeatmapLine)\n",
    "    cbar_ax = fig.axes[-1]\n",
    "    cbar_ax.set_position([1.2, 0.5 - cbar_length, cbar_length/cbar_aspect, cbar_length])\n",
    "    tickLength = FontProperties(size='small').get_size()/4\n",
    "#         cbar_ax.xaxis.set_ticks([-vmax, 0, vmax1])\n",
    "    cbar_ax.tick_params(axis='y', length=tickLength, color=colorAxis, labelsize='small')\n",
    "    if i == 1:\n",
    "        ax.set_xlabel('codon at position -1 (C-terminal)')\n",
    "    family = 'Liberation Mono'\n",
    "    ticks_font = FontProperties(family=family, size=fontSizeLabels)\n",
    "    ax.set_ylabel('stop codon\\ncontext')\n",
    "    ax.tick_params(which='y', left=True, labelleft=True)\n",
    "    for tick in ax.get_yticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "#     stopCodonIndex = plotData4.columns.get_level_values(0)\n",
    "#     tform = transforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "#     for yi, stopCodon in enumerate(stopCodonIndex):\n",
    "#         ax.annotate(stopCodon, xy=(yi + 0.5, 1.0), xycoords=tform,\n",
    "#                     xytext=(0, 6), textcoords='offset points', ha='center', va='bottom',\n",
    "#                     rotation=90, bbox=dict(pad=0, facecolor='none', edgecolor='none'),\n",
    "#                     fontproperties=FontProperties(family=family, size=fontSizeLabels))\n",
    "#     ax.tick_params(labelbottom=False)\n",
    "#     ax.annotate(codon, xy=(0.5, 0.0), xycoords='axes fraction',\n",
    "#                 xytext=(0, -6), textcoords='offset points', ha='center', va='top',\n",
    "#                 rotation=0, bbox=dict(pad=0, facecolor='none', edgecolor='none'),\n",
    "#                 fontproperties=FontProperties(family=family, size=fontSizeLabels))\n",
    "\n",
    "\n",
    "# ax.annotate('codon at ' + terminus + '-terminal', xy=(0.5, 0.06), xycoords='figure fraction', ha='center', va='top')\n",
    "# ax.annotate('stop codon context', xy=(0.5, 0.95), xycoords='figure fraction', ha='center', va='bottom')\n",
    "\n",
    "title = terminus + '-terminal, ' + speciesSetName\n",
    "axes[0].set_title(title)\n",
    "\n",
    "speciesSetName = 'TGA_non_overlapping_start_codon'\n",
    "outputFilePrefix = speciesSetName\n",
    "outputPath = analysisCtermPath / 'overlapping_start_codon' / speciesSetName\n",
    "filenamePlot = outputFilePrefix + '_composition_bias_codon_' + terminus + 'terminal_stopCodonContext_heatmap'\n",
    "print(filenamePlot)\n",
    "fig.savefig(outputPath / (filenamePlot + \".png\"), dpi=500, bbox_inches=\"tight\")\n",
    "fig.savefig(outputPath / (filenamePlot + \".svg\"), dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot codon biases for all phyla"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plot codon bias for the 3 codon contexts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for taxonId in phylaList[:]:\n",
    "\n",
    "    terminus = 'C'\n",
    "    statistical_test = 'contingency'\n",
    "    \n",
    "    taxonGroup = taxaDf[taxaDf['taxon_id'] == taxonId].iloc[0]\n",
    "    print(\"taxonGroupRef:\", taxonGroup)\n",
    "    taxonRank = taxonGroup['rank_name']\n",
    "    taxonName = taxonGroup['taxon_name']\n",
    "    taxonId = int(taxonId)\n",
    "    taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "    print(\"output path:\", str(analysisCtermDataPath / taxonSuffix))\n",
    "    suffix = taxonSuffix\n",
    "\n",
    "    statName = 'all'\n",
    "    dfFormatOutput = 'csv'\n",
    "    speciesSetName = taxonSuffix\n",
    "    outputFilePrefix = simplify_prefix(speciesSetName + '_subset_' + statName)\n",
    "    plotTitle = '{} {} {}'.format(taxonRank, taxonName, re.sub(r'_', r' ', outputFilePrefix))\n",
    "\n",
    "    plotData = oddsRatioCodonTaxon[oddsRatioCodonTaxon['taxaid'] == taxonId]\\\n",
    "               .drop(['taxaid', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes'], axis=1)\n",
    "    plotData = plotData.xs(terminus, level='terminus')\n",
    "    plotData = select_terminal_positions_in_index(plotData, terminus)\n",
    "    plotDataTemp = plotData.copy()\n",
    "\n",
    "    plotData.xs('log2OddsRatio', level='observable').loc[-1]\n",
    "\n",
    "    plotData.xs('log2OddsRatio', level='observable').loc[-1]\n",
    "\n",
    "    plotData = plotData.xs('log2OddsRatio', level='observable')\n",
    "    plotData = plotData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "    plotData = plotData[plotData.columns].astype(float)\n",
    "    plotData = plotData.transpose()\n",
    "\n",
    "    plotDataTemp.xs('BH_multiple_tests', level='observable')\n",
    "\n",
    "    maskData = plotDataTemp\n",
    "    maskData = maskData.xs('BH_multiple_tests', level='observable')\n",
    "    maskData = maskData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "    # Mask both rejected cases and null values\n",
    "    maskData = (maskData != True) & (maskData != 'True')\n",
    "    maskData = maskData.transpose()\n",
    "\n",
    "    addEmptyRowBetweenGroups = True\n",
    "    addEmptyRowBetweenGroupsFillValue = np.nan\n",
    "    plotData2 = plotData.copy()\n",
    "    nameIndex = plotData2.index.name\n",
    "    nameCol = plotData2.columns.name\n",
    "    stopCodonList = list(refCodonTableDf[refCodonTableDf['aa'] == '*'].index)\n",
    "\n",
    "    # Remove values that are not significant using mask\n",
    "    plotData2 = plotData2.where(~maskData, other=np.nan)\n",
    "    # plotData2 = plotData2.to_frame()\n",
    "    plotData2 = plotData2.loc[plotData2.index.map(lambda x: x not in stopCodonList)]\n",
    "    plotData2 = sort_codon_index(plotData2, refCodonTableDf,\n",
    "                                 addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                                 addEmptyRowBetweenGroups=addEmptyRowBetweenGroups)\n",
    "    plotData2.index.name = nameIndex\n",
    "    plotData2.columns.name = nameCol\n",
    "    plotData2 = plotData2.drop('all', axis=1)\n",
    "    plotData2.loc['N/AAC']\n",
    "    # print(\"plotData:\\n\", type(plotData), plotData)\n",
    "\n",
    "    ylabel = '$\\log_2$(odds ratio)'\n",
    "    ymin, ymax = -3, 3\n",
    "\n",
    "    n = 44\n",
    "    plotDf1 = plotData2.iloc[:n]\n",
    "    plotDf2 = plotData2.iloc[n+1:]\n",
    "    m = n - len(plotDf2)\n",
    "    emptyRow = plotData2.iloc[0].copy()\n",
    "    emptyRow[:] = np.nan\n",
    "    emptyRow.name = ''\n",
    "    plotDf2 = plotDf2.copy().append(m*[emptyRow])\n",
    "\n",
    "    figsizeCoeff = 0.6\n",
    "    fontSize = 16\n",
    "    aspectratio = 1.4\n",
    "    imagesize = 12\n",
    "    linewidth1 = 1.0\n",
    "    figsize = (figsizeCoeff*imagesize, figsizeCoeff*(imagesize/aspectratio))\n",
    "\n",
    "    fig, axes = plt.subplots(2, 1, figsize=figsize)\n",
    "    plt.subplots_adjust(hspace=0.5)\n",
    "    w = 0.6\n",
    "\n",
    "    for i, ax in enumerate(axes):\n",
    "        if i == 0:\n",
    "            plotDf3 = plotDf1\n",
    "        elif i == 1:\n",
    "            plotDf3 = plotDf2\n",
    "    #     norm = matplotlib.colors.Normalize(vmin=-vmax, vmax=vmax)\n",
    "    #     colors = cmap(norm(plotDf3.values))\n",
    "        plotDf3.plot.bar(ax=ax, width=w, linewidth=0, colors=paletteCat)\n",
    "        ax.set_ylabel(ylabel)\n",
    "        ax.set_ylim(ymin, ymax)\n",
    "        ax.xaxis.grid(False)\n",
    "        family = 'Liberation Mono'\n",
    "        ticks_font = FontProperties(family=family, size=figsizeCoeff*fontSize)\n",
    "        for tick in ax.get_xticklabels():\n",
    "            tick.set_fontproperties(ticks_font)\n",
    "            tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "        if i == 0:\n",
    "            ax.legend(loc='upper left', bbox_to_anchor=(1, 1), title='stop codon\\ncontext')\n",
    "        if i == 1:\n",
    "            ax.set_xlabel('codon at position -1 (C-terminal)')\n",
    "            ax.legend_.remove()\n",
    "\n",
    "\n",
    "    title = speciesSetName + ' ' + taxonName\n",
    "    axes[0].set_title(title)\n",
    "\n",
    "    filenamePlot = outputFilePrefix + '_composition_bias_codon_' + terminus + 'terminal_barplot_stopCodonContext'\n",
    "#     fig.savefig(os.path.join(plotsFolder, speciesSetName, filenamePlot + '.png'), dpi=400, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine plots in a grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.cbook as cbook\n",
    "from math import ceil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aspect = image.shape[1]/image.shape[0]\n",
    "aspect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "phylaList2 = phylaList[:]\n",
    "n = len(phylaList2)\n",
    "ncol = 3\n",
    "nrow = ceil(n / ncol)\n",
    "w = 2.5\n",
    "h = w/aspect\n",
    "fig = plt.figure(figsize=(ncol*w, nrow*h))\n",
    "gs = gridspec.GridSpec(nrow, ncol, wspace=0, hspace=0)\n",
    "    \n",
    "for i, taxonId in enumerate(phylaList2):\n",
    "    irow = i // ncol\n",
    "    icol = i % ncol\n",
    "    ax = fig.add_subplot(gs[irow, icol])\n",
    "\n",
    "    terminus = 'C'\n",
    "    statistical_test = 'contingency'\n",
    "    \n",
    "    taxonGroup = taxaDf[taxaDf['taxon_id'] == taxonId].iloc[0]\n",
    "    taxonRank = taxonGroup['rank_name']\n",
    "    taxonName = taxonGroup['taxon_name']\n",
    "    taxonId = int(taxonId)\n",
    "    taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "    suffix = taxonSuffix\n",
    "    statName = 'all'\n",
    "    dfFormatOutput = 'csv'\n",
    "    speciesSetName = taxonSuffix\n",
    "    outputFilePrefix = simplify_prefix(speciesSetName + '_subset_' + statName)\n",
    "    filenamePlot = outputFilePrefix + '_composition_bias_codon_' + terminus + 'terminal_barplot_stopCodonContext'\n",
    "    filePath = Path(plotsFolder) / speciesSetName / (filenamePlot + '.png')\n",
    "    print(filePath)\n",
    "    \n",
    "    # A sample image\n",
    "    with cbook.get_sample_data(filePath) as image_file:\n",
    "        image = plt.imread(image_file)\n",
    "\n",
    "    ax.imshow(image)\n",
    "    ax.axis('off')\n",
    "\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_\" + terminus + \"terminal\" +\n",
    "                                                       '_stop_codon_context_all_grid' + \".png\")\n",
    "fig.tight_layout()\n",
    "fig.savefig(filenamePlot, dpi=800)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare codon biases for all phyla"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Build phyla biases dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfList = []\n",
    "for taxonId in phylaList[:]:\n",
    "\n",
    "    terminus = 'C'\n",
    "\n",
    "    taxonGroup = taxaDf[taxaDf['taxon_id'] == taxonId].iloc[0]\n",
    "    taxonId = int(taxonId)\n",
    "\n",
    "    plotData = oddsRatioCodonTaxon[oddsRatioCodonTaxon['taxaid'] == taxonId]\\\n",
    "               .drop(['taxaid', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes'], axis=1)\n",
    "    plotData = plotData.xs(terminus, level='terminus')\n",
    "    plotData = select_terminal_positions_in_index(plotData, terminus)\n",
    "    plotDataTemp = plotData.copy()\n",
    "\n",
    "    plotData = plotData.xs('log2OddsRatio', level='observable')\n",
    "    plotData = plotData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "    plotData = plotData[plotData.columns].astype(float)\n",
    "    plotData = plotData.transpose()\n",
    "\n",
    "    maskData = plotDataTemp\n",
    "    maskData = maskData.xs('BH_multiple_tests', level='observable')\n",
    "    maskData = maskData.loc[-1].reset_index(drop=True).set_index('stop_codon')\n",
    "    # Mask both rejected cases and null values\n",
    "    maskData = (maskData != True) & (maskData != 'True')\n",
    "    maskData = maskData.transpose()\n",
    "\n",
    "    addEmptyRowBetweenGroups = False\n",
    "    addEmptyRowBetweenGroupsFillValue = np.nan\n",
    "    plotData2 = plotData.copy()\n",
    "    nameIndex = plotData2.index.name\n",
    "    nameCol = plotData2.columns.name\n",
    "    stopCodonList = list(refCodonTableDf[refCodonTableDf['aa'] == '*'].index)\n",
    "\n",
    "    # Remove values that are not significant using mask\n",
    "    plotData2 = plotData2.where(~maskData, other=np.nan)\n",
    "    # plotData2 = plotData2.to_frame()\n",
    "    plotData2 = plotData2.loc[plotData2.index.map(lambda x: x not in stopCodonList)]\n",
    "\n",
    "    plotData2 = sort_codon_index(plotData2, refCodonTableDf,\n",
    "                                 addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                                 addEmptyRowBetweenGroups=addEmptyRowBetweenGroups)\n",
    "    plotData2.index.name = nameIndex\n",
    "    plotData2.columns.name = nameCol\n",
    "    plotData2 = plotData2.drop('all', axis=1)\n",
    "\n",
    "    plotData2['taxaid'] = taxonId\n",
    "    plotData2\n",
    "    dfList.append(plotData2)\n",
    "\n",
    "oddsRatioCodonPhyla = pd.concat(dfList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonPhyla.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonPhyla.index.name = 'codon'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonPhyla2 = oddsRatioCodonPhyla.set_index('taxaid', append=True).reorder_levels(['taxaid', 'codon']).sort_index()\n",
    "oddsRatioCodonPhyla2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select phyla with enough sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We select phyla for which we have at least X protein sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxaid</th>\n",
       "      <th>taxon_name</th>\n",
       "      <th>n_seq</th>\n",
       "      <th>stop_codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>2165400</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>756756</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>368993</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1224</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1039649</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1239</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>756567</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1239</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>451749</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1239</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>147985</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1239</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>156833</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>201174</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>912659</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>201174</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>101831</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>201174</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>196824</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>201174</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>613990</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>976</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>339458</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>976</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>217626</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>976</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>65490</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>976</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>56342</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>544448</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>45030</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>1117</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>226172</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>1117</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>109177</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>1117</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>57139</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>1117</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>59856</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>544448</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>34269</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>544448</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>9217</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>544448</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>1518</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>203691</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>70812</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>203691</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>35109</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>203691</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>13861</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>203691</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>21842</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>1297</td>\n",
       "      <td>Deinococcus-Thermus</td>\n",
       "      <td>51606</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>1297</td>\n",
       "      <td>Deinococcus-Thermus</td>\n",
       "      <td>11585</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4872</th>\n",
       "      <td>74152</td>\n",
       "      <td>Elusimicrobia</td>\n",
       "      <td>420</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4873</th>\n",
       "      <td>74152</td>\n",
       "      <td>Elusimicrobia</td>\n",
       "      <td>389</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6265</th>\n",
       "      <td>200938</td>\n",
       "      <td>Chrysiogenetes</td>\n",
       "      <td>2549</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6266</th>\n",
       "      <td>200938</td>\n",
       "      <td>Chrysiogenetes</td>\n",
       "      <td>735</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6267</th>\n",
       "      <td>200938</td>\n",
       "      <td>Chrysiogenetes</td>\n",
       "      <td>425</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6268</th>\n",
       "      <td>200938</td>\n",
       "      <td>Chrysiogenetes</td>\n",
       "      <td>1389</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6269</th>\n",
       "      <td>142182</td>\n",
       "      <td>Gemmatimonadetes</td>\n",
       "      <td>3913</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>142182</td>\n",
       "      <td>Gemmatimonadetes</td>\n",
       "      <td>471</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>142182</td>\n",
       "      <td>Gemmatimonadetes</td>\n",
       "      <td>700</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6272</th>\n",
       "      <td>142182</td>\n",
       "      <td>Gemmatimonadetes</td>\n",
       "      <td>2742</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6273</th>\n",
       "      <td>65842</td>\n",
       "      <td>Fibrobacteres</td>\n",
       "      <td>3066</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6274</th>\n",
       "      <td>65842</td>\n",
       "      <td>Fibrobacteres</td>\n",
       "      <td>1879</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6275</th>\n",
       "      <td>65842</td>\n",
       "      <td>Fibrobacteres</td>\n",
       "      <td>785</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6276</th>\n",
       "      <td>65842</td>\n",
       "      <td>Fibrobacteres</td>\n",
       "      <td>402</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9363</th>\n",
       "      <td>67814</td>\n",
       "      <td>Caldiserica</td>\n",
       "      <td>1463</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9364</th>\n",
       "      <td>67814</td>\n",
       "      <td>Caldiserica</td>\n",
       "      <td>776</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9365</th>\n",
       "      <td>67814</td>\n",
       "      <td>Caldiserica</td>\n",
       "      <td>298</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9366</th>\n",
       "      <td>67814</td>\n",
       "      <td>Caldiserica</td>\n",
       "      <td>389</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9367</th>\n",
       "      <td>67819</td>\n",
       "      <td>Armatimonadetes</td>\n",
       "      <td>4576</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9368</th>\n",
       "      <td>67819</td>\n",
       "      <td>Armatimonadetes</td>\n",
       "      <td>1423</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9369</th>\n",
       "      <td>67819</td>\n",
       "      <td>Armatimonadetes</td>\n",
       "      <td>1783</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9370</th>\n",
       "      <td>67819</td>\n",
       "      <td>Armatimonadetes</td>\n",
       "      <td>1370</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9491</th>\n",
       "      <td>456828</td>\n",
       "      <td>Candidatus Cloacimonetes</td>\n",
       "      <td>1669</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9492</th>\n",
       "      <td>456828</td>\n",
       "      <td>Candidatus Cloacimonetes</td>\n",
       "      <td>909</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9493</th>\n",
       "      <td>456828</td>\n",
       "      <td>Candidatus Cloacimonetes</td>\n",
       "      <td>267</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9494</th>\n",
       "      <td>456828</td>\n",
       "      <td>Candidatus Cloacimonetes</td>\n",
       "      <td>493</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9777</th>\n",
       "      <td>1134404</td>\n",
       "      <td>Ignavibacteriae</td>\n",
       "      <td>5865</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9778</th>\n",
       "      <td>1134404</td>\n",
       "      <td>Ignavibacteriae</td>\n",
       "      <td>3437</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9779</th>\n",
       "      <td>1134404</td>\n",
       "      <td>Ignavibacteriae</td>\n",
       "      <td>786</td>\n",
       "      <td>TAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9780</th>\n",
       "      <td>1134404</td>\n",
       "      <td>Ignavibacteriae</td>\n",
       "      <td>1642</td>\n",
       "      <td>TGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       taxaid                taxon_name    n_seq stop_codon\n",
       "4        1224            Proteobacteria  2165400        all\n",
       "5        1224            Proteobacteria   756756        TAA\n",
       "6        1224            Proteobacteria   368993        TAG\n",
       "7        1224            Proteobacteria  1039649        TGA\n",
       "8        1239                Firmicutes   756567        all\n",
       "9        1239                Firmicutes   451749        TAA\n",
       "10       1239                Firmicutes   147985        TAG\n",
       "11       1239                Firmicutes   156833        TGA\n",
       "16     201174            Actinobacteria   912659        all\n",
       "17     201174            Actinobacteria   101831        TAA\n",
       "18     201174            Actinobacteria   196824        TAG\n",
       "19     201174            Actinobacteria   613990        TGA\n",
       "36        976             Bacteroidetes   339458        all\n",
       "37        976             Bacteroidetes   217626        TAA\n",
       "38        976             Bacteroidetes    65490        TAG\n",
       "39        976             Bacteroidetes    56342        TGA\n",
       "40     544448               Tenericutes    45030        all\n",
       "125      1117             Cyanobacteria   226172        all\n",
       "126      1117             Cyanobacteria   109177        TAA\n",
       "127      1117             Cyanobacteria    57139        TAG\n",
       "128      1117             Cyanobacteria    59856        TGA\n",
       "130    544448               Tenericutes    34269        TAA\n",
       "131    544448               Tenericutes     9217        TAG\n",
       "132    544448               Tenericutes     1518        TGA\n",
       "153    203691              Spirochaetes    70812        all\n",
       "154    203691              Spirochaetes    35109        TAA\n",
       "155    203691              Spirochaetes    13861        TAG\n",
       "156    203691              Spirochaetes    21842        TGA\n",
       "278      1297       Deinococcus-Thermus    51606        all\n",
       "279      1297       Deinococcus-Thermus    11585        TAA\n",
       "...       ...                       ...      ...        ...\n",
       "4872    74152             Elusimicrobia      420        TAG\n",
       "4873    74152             Elusimicrobia      389        TGA\n",
       "6265   200938            Chrysiogenetes     2549        all\n",
       "6266   200938            Chrysiogenetes      735        TAA\n",
       "6267   200938            Chrysiogenetes      425        TAG\n",
       "6268   200938            Chrysiogenetes     1389        TGA\n",
       "6269   142182          Gemmatimonadetes     3913        all\n",
       "6270   142182          Gemmatimonadetes      471        TAA\n",
       "6271   142182          Gemmatimonadetes      700        TAG\n",
       "6272   142182          Gemmatimonadetes     2742        TGA\n",
       "6273    65842             Fibrobacteres     3066        all\n",
       "6274    65842             Fibrobacteres     1879        TAA\n",
       "6275    65842             Fibrobacteres      785        TAG\n",
       "6276    65842             Fibrobacteres      402        TGA\n",
       "9363    67814               Caldiserica     1463        all\n",
       "9364    67814               Caldiserica      776        TAA\n",
       "9365    67814               Caldiserica      298        TAG\n",
       "9366    67814               Caldiserica      389        TGA\n",
       "9367    67819           Armatimonadetes     4576        all\n",
       "9368    67819           Armatimonadetes     1423        TAA\n",
       "9369    67819           Armatimonadetes     1783        TAG\n",
       "9370    67819           Armatimonadetes     1370        TGA\n",
       "9491   456828  Candidatus Cloacimonetes     1669        all\n",
       "9492   456828  Candidatus Cloacimonetes      909        TAA\n",
       "9493   456828  Candidatus Cloacimonetes      267        TAG\n",
       "9494   456828  Candidatus Cloacimonetes      493        TGA\n",
       "9777  1134404           Ignavibacteriae     5865        all\n",
       "9778  1134404           Ignavibacteriae     3437        TAA\n",
       "9779  1134404           Ignavibacteriae      786        TAG\n",
       "9780  1134404           Ignavibacteriae     1642        TGA\n",
       "\n",
       "[120 rows x 4 columns]"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioCodonTaxon[['taxaid', 'taxon_name', 'n_seq', 'stop_codon']].reset_index(drop=True)\n",
    "df = df[df['taxaid'].map(lambda x: x in phylaList)].drop_duplicates()\n",
    "phylaDf = df\n",
    "phylaDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1224,\n",
       " 1239,\n",
       " 201174,\n",
       " 976,\n",
       " 544448,\n",
       " 1117,\n",
       " 203691,\n",
       " 1297,\n",
       " 200918,\n",
       " 1090,\n",
       " 200795,\n",
       " 57723,\n",
       " 203682,\n",
       " 74201]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylaDfBig = phylaDf[phylaDf['n_seq'] > 2e4]\n",
    "phylaDfBigList = list(phylaDfBig['taxaid'].unique())\n",
    "print(len(phylaDfBigList))\n",
    "phylaDfBigList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Proteobacteria', 'Firmicutes', 'Actinobacteria', 'Bacteroidetes',\n",
       "       'Tenericutes', 'Cyanobacteria', 'Spirochaetes',\n",
       "       'Deinococcus-Thermus', 'Thermotogae', 'Chlorobi', 'Chloroflexi',\n",
       "       'Acidobacteria', 'Planctomycetes', 'Verrucomicrobia'], dtype=object)"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylaNameBigList = phylaDf.set_index('taxaid').loc[phylaDfBigList, 'taxon_name'].unique()\n",
    "phylaNameBigList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonPhyla3 = oddsRatioCodonPhyla2.loc[phylaDfBigList]\n",
    "oddsRatioCodonPhyla3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioCodonTaxon.index.get_level_values('observable').unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (oddsRatioCodonTaxon\n",
    "#     .drop(['taxaid', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes'], axis=1)\\\n",
    "    .loc['C'].loc[-1].loc[['BH_multiple_tests', 'oddsRatio']][['CCT', 'CCC', 'CCA', 'CCG', 'taxaid', 'n_seq', 'taxon_name', 'stop_codon']]\n",
    ")\n",
    "df = df[df['taxaid'].map(lambda x: x in phylaDfBigList)]\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['stop_codon'] == 'TGA'].sort_values('taxon_name');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['stop_codon'] != 'all'].reset_index().sort_values(['taxon_name', 'stop_codon', 'observable'])\n",
    "df.set_index('taxon_name').loc[['Firmicutes', 'Tenericutes', 'Thermotogae']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "['Actinobacteria', 'Proteobacteria', 'Deinococcus-Thermus', 'Spirochaetes', 'Verrucomicrobia']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (oddsRatioCodonTaxon\n",
    "#     .drop(['taxaid', 'taxon_name', 'taxon_rank', 'n_seq', 'n_genomes'], axis=1)\\\n",
    "    .loc['C'].loc[-1].loc[['BH_multiple_tests', 'log2OddsRatio']][['CCT', 'CCC', 'CCA', 'CCG', 'taxaid', 'n_seq', 'taxon_name', 'stop_codon']]\n",
    ")\n",
    "df = df[df['taxaid'].map(lambda x: x in phylaDfBigList)]\n",
    "len(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['stop_codon'] == 'TGA'].sort_values('taxon_name');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Correlation of TAG biases across phyla"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioCodonTaxon.copy()\n",
    "terminus = 'C'\n",
    "terminusPos = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf[oddsRatioDf['taxon_rank'] == 'phylum'].xs(terminusPos, level='position from terminus').xs(terminus, level='terminus')\n",
    "oddsRatioDf = oddsRatioDf.sort_values('n_seq', ascending=False)\n",
    "oddsRatioDf = oddsRatioDf.drop(['n_genomes', 'n_seq', 'taxaid', 'taxon_rank'], axis=1)\\\n",
    "                         .set_index('taxon_name', append=True).sort_index()\n",
    "oddsRatioDf = oddsRatioDf.loc[oddsRatioDf.index.get_level_values('taxon_name').map(lambda x: x in phylaNameBigList)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf.set_index('stop_codon', append=True).loc[idx[:, :, ['TAA', 'TAG', 'TGA']], :].unstack(level=2)\n",
    "oddsRatioDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDf.loc[idx['log2OddsRatio', :], idx[['AAG', 'ACA'], 'TGA']]\n",
    "df.columns = df.columns.droplevel(1)\n",
    "df = df.astype(float)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "fig, ax = plt.subplots(figsize=(3, 3))\n",
    "# ax.set_aspect('equal')\n",
    "# ax.set_xlim(-2, 2)\n",
    "# ax.set_ylim(-2, 2)\n",
    "# df.plot.scatter(ax=ax, x='AAG', y='ACA')\n",
    "sns.regplot(data=df, x='AAG', y='ACA')\n",
    "pearson = df.corr().iloc[0, 1]\n",
    "print(pearson)\n",
    "# ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.05, 0.92),\n",
    "#             xycoords='axes fraction', ha='left', va='top', fontsize='medium',\n",
    "#             bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "ax.set_xlabel('log2(odds ratio)\\n(K)AAG-TGA')\n",
    "ax.set_ylabel('log2(odds ratio)\\n(T)ACA-TGA')\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_C-terminal\" +\n",
    "                                                       '_stop_codon_context.correlation_AAG-ACA')\n",
    "print(filenamePlot)\n",
    "# fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Fig. S5B. Correlation between codon biases in the TGA stop codon context for (K)AAG-TGA and (T)ACA-TGA across the 14 phyla with more than 20k sequences. Biases are expressed as log2(odds ratio). Phyla that favor the codon AAG at the C-terminal in the TGA stop codon context also disfavor the codon ACA, and vice-versa. The pearson correlation is -0.87.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We take all phyla, not only the big ones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioCodonTaxon.copy()\n",
    "terminus = 'C'\n",
    "terminusPos = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf[oddsRatioDf['taxon_rank'] == 'phylum'].xs(terminusPos, level='position from terminus').xs(terminus, level='terminus')\n",
    "oddsRatioDf = oddsRatioDf.sort_values('n_seq', ascending=False)\n",
    "oddsRatioDf = oddsRatioDf.drop(['n_genomes', 'n_seq', 'taxaid', 'taxon_rank'], axis=1)\\\n",
    "                         .set_index('taxon_name', append=True).sort_index()\n",
    "# oddsRatioDf = oddsRatioDf.loc[oddsRatioDf.index.get_level_values('taxon_name').map(lambda x: x in phylaNameBigList)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf.set_index('stop_codon', append=True).loc[idx[:, :, ['TAA', 'TAG', 'TGA']], :].unstack(level=2)\n",
    "oddsRatioDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDf.loc[idx['log2OddsRatio', :], idx[['AAG', 'ACA'], 'TGA']]\n",
    "df.columns = df.columns.droplevel(1)\n",
    "df = df.astype(float)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "fig, ax = plt.subplots(figsize=(4, 3))\n",
    "# df.plot.scatter(ax=ax, x='AAG', y='ACA')\n",
    "sns.regplot(data=df, x='AAG', y='ACA')\n",
    "ax.set_xlabel('log2(odds ratio)\\n(K)AAG-TGA')\n",
    "ax.set_ylabel('log2(odds ratio)\\n(T)ACA-TGA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Correlation seem more noisy..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioCodonTaxon.copy()\n",
    "terminus = 'C'\n",
    "terminusPos = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf[oddsRatioDf['taxon_rank'] == 'phylum'].xs(terminusPos, level='position from terminus').xs(terminus, level='terminus')\n",
    "oddsRatioDf = oddsRatioDf.sort_values('n_seq', ascending=False)\n",
    "oddsRatioDf = oddsRatioDf.drop(['n_genomes', 'n_seq', 'taxaid', 'taxon_rank'], axis=1)\\\n",
    "                         .set_index('taxon_name', append=True).sort_index()\n",
    "oddsRatioDf = oddsRatioDf.loc[oddsRatioDf.index.get_level_values('taxon_name').map(lambda x: x in phylaNameBigList)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf.set_index('stop_codon', append=True).loc[idx[:, :, ['TAA', 'TAG', 'TGA']], :].unstack(level=2)\n",
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDf.loc[idx['log2OddsRatio', :], :].copy()\n",
    "# df.columns = df.columns.droplevel(1)\n",
    "df = df.astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove all biases that are not significant\n",
    "maskData = oddsRatioDf.loc['BH_multiple_tests']\n",
    "maskData = (maskData != True) & (maskData != 'True')\n",
    "\n",
    "maskData = ~(oddsRatioDf.loc[idx['pvalue', :], :] < 1e-2)\n",
    "\n",
    "# or not\n",
    "df = df.where(~maskData.values, other=np.nan)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exclude the stop codons at position -1\n",
    "print(len(df.columns))\n",
    "df = df[[col for col in df.columns if not col[0] in ['TAA', 'TAG', 'TGA']]]\n",
    "print(len(df.columns))\n",
    "df.columns.names = ['codon', 'stop_codon']\n",
    "df.index = df.index.droplevel(0)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy().T\n",
    "df2 = df2.reset_index().set_index('codon')\n",
    "print(len(df2.index))\n",
    "df2 = sort_codon_index(df2, refCodonTableDf,\n",
    "                       addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                       addEmptyRowBetweenGroups=False)\n",
    "df2 = df2.set_index('stop_codon', append=True).T\n",
    "df2.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will also filter our correlation map to only include codons that show fairly strong biases. At least one bias that is stronger than 1 log2, and at least 6 data points with non-null values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colValid = df2.apply(lambda col: ((col.max() > 0.6) or (col.min() < -0.6)) & ((~col.isnull()).sum() >= 6), axis=0)\n",
    "colValid.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colValid.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We could include or not the codon which are not valid, however the clustermap only works with non-null values.\n",
    "df3 = df2.loc[:, colValid]\n",
    "# df3 = df2.copy()\n",
    "# for col, valid in colValid.iteritems():\n",
    "#     if not valid:\n",
    "#         print(col)\n",
    "#         df3.loc[:, col] = np.nan\n",
    "df3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "61*3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df3.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df3['C/TGT']['TAA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df3['R/CGT']['TGA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df3.corr().loc[df3.corr().isnull().apply(any, axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "data = df3.corr()\n",
    "sns.clustermap(data, method='average', metric='euclidean', z_score=None, standard_scale=None,\n",
    "               row_cluster=True, col_cluster=True, figsize=(16, 16), xticklabels=1, yticklabels=1)\n",
    "fig = plt.gcf()\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_C-terminal\" +\n",
    "                                                       '_stop_codon_context.correlation_clustermap')\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot only TGA stop codon context,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "data = df2.loc[:, idx[:, 'TGA']].corr()\n",
    "sns.clustermap(data, method='average', metric='euclidean', z_score=None, standard_scale=None,\n",
    "               row_cluster=True, col_cluster=True, figsize=(16, 16), xticklabels=1, yticklabels=1)\n",
    "fig = plt.gcf()\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_C-terminal\" +\n",
    "                                                       '_stop_codon_context.correlation_clustermap.only_TGA')\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Fig. S5C. Correlation map between codon biases in the TGA stop codon context across the 14 phyla with more than 20k sequences. Biases were expressed as log2(odds ratio). Pearson correlation was computed for biases in the 14 phyla for each codon pair. Codons were clustered using euclidean metric with the average method to compute distances between clusters. Correlations values ranged from -0.90 to 0.91.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddf = data.values\n",
    "print(ddf[np.triu_indices(ddf.shape[0], k=1)].max())\n",
    "print(ddf[np.triu_indices(ddf.shape[0], k=1)].min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifList = []\n",
    "for c in df.columns[:]:\n",
    "    seq = ''.join(c)\n",
    "    motif = dict()\n",
    "    for m in ['ATG', 'GTG', 'CTG']:\n",
    "        motif[m] = bool(re.search(m, seq))\n",
    "        print(m, seq, bool(re.search(m, seq)))\n",
    "    for N in 'GCAT':\n",
    "        for pos in range(3):\n",
    "            motif['{}{:1d}'.format(N, pos + 1)] = seq[pos] == N\n",
    "    motifList.append(motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.index = df.index.droplevel(0)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3 = pd.concat([df.T, pd.DataFrame(motifList, index=df.columns)], axis=1)\n",
    "df3 = df.T\n",
    "df3.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "ymin, ymax = -4, +4\n",
    "phylaList1 = ['Bacteroidetes', 'Acidobacteria', 'Cyanobacteria', 'Tenericutes', 'Firmicutes', 'Thermotogae',\n",
    "              'Verrucomicrobia', 'Spirochaetes', 'Proteobacteria', 'Actinobacteria', 'Deinococcus-Thermus'][:]\n",
    "nPhyla = len(phylaList1)\n",
    "fig, axes = plt.subplots(nPhyla, 3, figsize=(7,3.7*nPhyla), gridspec_kw=dict(hspace=0.4))\n",
    "for i, phylum in enumerate(phylaList1):\n",
    "    df4 = df3[[phylum, 'ATG', 'GTG', 'CTG']]\n",
    "    for j, motif in enumerate(['ATG', 'GTG', 'CTG']):\n",
    "        ax = axes[i, j]\n",
    "        ax.set_ylim(ymin, ymax)\n",
    "        sns.boxplot(data=df4, y=phylum, x=motif, ax=ax)\n",
    "        ax, testResultList = add_stat_annotation(ax, data=df4, y=phylum, x=motif,\n",
    "                                         boxPairList=[(False, True)],\n",
    "                                         test='t-test_ind', textFormat='star', loc='outside',\n",
    "                                         lineYOffsetAxesCoord=0.010, lineYOffsetToBoxAxesCoord=0.005,\n",
    "                                         textYOffsetPoints=0, verbose=2, linewidth=1)\n",
    "        ax.set_xlabel(\"NNN-Stop contains {}\".format(motif))\n",
    "        if j == 0:\n",
    "            ax.set_ylabel(\"{}\\n\\nlog2(odds ratio)\".format(phylum))\n",
    "        else:\n",
    "            ax.set_ylabel(\"\")\n",
    "\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_C-terminal\" +\n",
    "                                                       '_stop_codon_context.boxplot_motif_ATGallPhyla')\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "ymin, ymax = -4, +4\n",
    "phylaList1 = ['Bacteroidetes', 'Acidobacteria', 'Cyanobacteria', 'Tenericutes', 'Firmicutes', 'Thermotogae',\n",
    "              'Verrucomicrobia', 'Spirochaetes', 'Proteobacteria', 'Actinobacteria', 'Deinococcus-Thermus'][:]\n",
    "nPhyla = len(phylaList1)\n",
    "motifList = ['G1', 'C1', 'A1', 'T1', 'G2', 'C2', 'A2', 'T2', 'G3', 'C3', 'A3', 'T3']\n",
    "nMotifs = len(motifList)\n",
    "\n",
    "fig, axes = plt.subplots(nPhyla, nMotifs, figsize=(2*nMotifs, 3.5*nPhyla), gridspec_kw=dict(hspace=0.35))\n",
    "for i, phylum in enumerate(phylaList1):\n",
    "    for j, motif in enumerate(motifList):\n",
    "        ax = axes[i, j]\n",
    "        ax.set_ylim(ymin, ymax)\n",
    "        sns.boxplot(data=df3, y=phylum, x=motif, ax=ax)\n",
    "        ax, testResultList = add_stat_annotation(ax, data=df3, y=phylum, x=motif,\n",
    "                                         boxPairList=[(False, True)],\n",
    "                                         test='t-test_ind', textFormat='star', loc='outside',\n",
    "                                         lineYOffsetAxesCoord=0.010, lineYOffsetToBoxAxesCoord=0.005,\n",
    "                                         textYOffsetPoints=0, verbose=2, linewidth=1)\n",
    "        ax.set_xlabel(\"NNN-Stop contains {}\".format(motif))\n",
    "        pos = int(motif[1]) - 1\n",
    "        label = 'NNN'\n",
    "        label = label[:pos] + motif[0] + label[pos + 1:] + '-Stop'\n",
    "        print(label)\n",
    "        ax.set_xlabel(label)\n",
    "        if j == 0:\n",
    "            ax.set_ylabel(\"{}\\n\\nlog2(odds ratio)\".format(phylum))\n",
    "        else:\n",
    "            ax.set_ylabel(\"\")\n",
    "\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_C-terminal\" +\n",
    "                                                       '_stop_codon_context.boxplot_motif_NNNallPhyla')\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df4 = df3.iloc[:, :-3].stack().reset_index().set_index(['codon', 'stop_codon'])\n",
    "df4.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifList = []\n",
    "for c in df4.index[:]:\n",
    "    seq = ''.join(c)\n",
    "    motif = dict()\n",
    "    for m in ['ATG', 'GTG', 'CTG']:\n",
    "        motif[m] = bool(re.search(m, seq))\n",
    "    for N in 'GCAT':\n",
    "        for pos in range(3):\n",
    "            motif['{}{:1d}'.format(N, pos + 1)] = seq[pos] == N\n",
    "    motifList.append(motif)\n",
    "motifList\n",
    "df5 = pd.concat([df4, pd.DataFrame(motifList, index=df4.index)], axis=1)\n",
    "df5 = df5.rename(columns={'level_2':'phylum', 0:'value'})\n",
    "df5.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "ymin, ymax = -4, +4\n",
    "\n",
    "motifList = df5.columns[2:]\n",
    "motifList = ['ATG', 'GTG', 'CTG', 'G1', 'C1', 'A1', 'T1', 'G2', 'C2', 'A2', 'T2', 'G3', 'C3', 'A3', 'T3']\n",
    "nMotifs = len(motifList)\n",
    "fig, axes = plt.subplots(4, 4, figsize=(2*4,3.5*4), gridspec_kw=dict(hspace=0.35))\n",
    "\n",
    "for j, motif in enumerate(motifList):\n",
    "    print()\n",
    "    if motif in ['ATG', 'GTG', 'CTG']:\n",
    "        i = 0\n",
    "        j1 = j\n",
    "    elif re.search('1', motif):\n",
    "        i = 1\n",
    "        j1 = j - 3\n",
    "    elif re.search('2', motif):\n",
    "        i = 2\n",
    "        j1 = j - 3 - 4\n",
    "    elif re.search('3', motif):\n",
    "        i = 3\n",
    "        j1 = j - 3 - 4 - 4\n",
    "    ax = axes[i, j1]\n",
    "    sns.boxplot(data=df5, y='value', x=motif, ax=ax)\n",
    "    ax, testResultList = add_stat_annotation(ax, data=df5, y='value', x=motif,\n",
    "                                         boxPairList=[(False, True)],\n",
    "                                         test='t-test_ind', textFormat='star', loc='outside',\n",
    "                                         lineYOffsetAxesCoord=0.010, lineYOffsetToBoxAxesCoord=0.005,\n",
    "                                         textYOffsetPoints=0, verbose=2, linewidth=1)\n",
    "    if i == 0:\n",
    "        ax.set_xlabel(\"NNN-Stop contains {}\".format(motif))\n",
    "    else:\n",
    "        print(motif[1])\n",
    "        print(motif[0])\n",
    "        pos = int(motif[1]) - 1\n",
    "        label = 'NNN'\n",
    "        label = label[:pos] + motif[0] + label[pos + 1:] + '-Stop'\n",
    "        print(label)\n",
    "        ax.set_xlabel(label)\n",
    "    ax.set_ylim(ymin, ymax)\n",
    "    if j1 == 0:\n",
    "        ax.set_ylabel(\"all phyla\\n\\nlog2(odds ratio)\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "axes[0, 3].axis('off')\n",
    "\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_C-terminal\" +\n",
    "                                                       '_stop_codon_context.boxplot_motif_allphyla')\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df5.loc[idx[:, 'TGA'], :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(mwTools.plot)\n",
    "reload(statannot)\n",
    "from mwTools.plot import seaborn_boxplot_add_number_samples\n",
    "from statannot import add_stat_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df6 = df5.loc[idx[:, 'TGA'], :]\n",
    "print(len(df6[df6['A3']].index.get_level_values(0).unique()))\n",
    "print(len(df6[df6['G3']].index.get_level_values(0).unique()))\n",
    "\n",
    "data1 = df6.loc[df6['A3'], 'value']\n",
    "data2 = df6.loc[df6['G3'], 'value']\n",
    "\n",
    "stats.ttest_ind(a=data1, b=data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data1.mean() - data2.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.distplot(data1)\n",
    "sns.distplot(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data1B = 2**data1\n",
    "data2B = 2**data2\n",
    "stats.ttest_ind(a=data1B, b=data2B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data1B.mean() / data2B.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.distplot(data1)\n",
    "sns.distplot(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "ymin, ymax = -4, +4\n",
    "\n",
    "for stop in ['TAA', 'TAG', 'TGA'][:]:\n",
    "    print(\"#######\", stop)\n",
    "    \n",
    "    df6 = df5.loc[idx[:, stop], :]\n",
    "\n",
    "    motifList = df6.columns[2:]\n",
    "    includeATGMotifs = False\n",
    "    if includeATGMotifs:\n",
    "        motifList = ['ATG', 'GTG', 'CTG', 'G1', 'C1', 'A1', 'T1', 'G2', 'C2', 'A2', 'T2', 'G3', 'C3', 'A3', 'T3']\n",
    "        nRows = 4\n",
    "    else:\n",
    "        motifList = ['G1', 'C1', 'A1', 'T1', 'G2', 'C2', 'A2', 'T2', 'G3', 'C3', 'A3', 'T3']\n",
    "        nRows = 3\n",
    "    nMotifs = len(motifList)\n",
    "    nCols = 4\n",
    "    fig, axes = plt.subplots(nRows, nCols, figsize=(2*nCols,3.5*nRows), gridspec_kw=dict(hspace=0.35))\n",
    "\n",
    "    for j, motif in enumerate(motifList):\n",
    "        print(\"##\", motif)\n",
    "        if motif in ['ATG', 'GTG', 'CTG']:\n",
    "            i = 0\n",
    "            j1 = j\n",
    "        elif re.search('1', motif):\n",
    "            i = 1 if includeATGMotifs else 0\n",
    "            j1 = j - 3 if includeATGMotifs else j\n",
    "        elif re.search('2', motif):\n",
    "            i = 2 if includeATGMotifs else 1\n",
    "            j1 = j - 3 - 4 if includeATGMotifs else j - 4\n",
    "        elif re.search('3', motif):\n",
    "            i = 3 if includeATGMotifs else 2\n",
    "            j1 = j - 3 - 4 - 4 if includeATGMotifs else j - 4 - 4\n",
    "        ax = axes[i, j1]\n",
    "        ax.set_ylim(ymin, ymax)\n",
    "        print(df6.head())\n",
    "        sns.boxplot(data=df6, y='value', x=motif, ax=ax, fliersize=2)\n",
    "        ax, testResultList = add_stat_annotation(ax, data=df6, y='value', x=motif,\n",
    "                                             box_pairs=[(False, True)],\n",
    "                                             test='t-test_ind', text_format='star', loc='outside',\n",
    "                                             line_offset=0.010, line_offset_to_box=0.005,\n",
    "                                             text_offset=0, verbose=2, linewidth=1)\n",
    "#         seaborn_boxplot_add_number_samples(ax, df=df6, xlabel=motif, ylabel='value', order=None, position='top', yPosAxes=0.9,\n",
    "#                                            fontsize='medium', color='black', text_rotation=0, verbose=1)\n",
    "        if i == 0 and includeATGMotifs:\n",
    "            ax.set_xlabel(\"NNN-Stop contains {}\".format(motif))\n",
    "        else:\n",
    "            pos = int(motif[1]) - 1\n",
    "            label = 'NNN'\n",
    "            label = label[:pos] + motif[0] + label[pos + 1:] + '-Stop'\n",
    "            print(label)\n",
    "            ax.set_xlabel(label)\n",
    "        if j1 == 0:\n",
    "            ax.set_ylabel(\"all phyla\\n\\nlog2(odds ratio)\")\n",
    "        else:\n",
    "            ax.set_ylabel(\"\")\n",
    "    if includeATGMotifs:\n",
    "        axes[0, 3].axis('off')\n",
    "\n",
    "    filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_C-terminal\" +\n",
    "                                                           '_stop_codon_context.boxplot_motif_allphyla_stopCodon_{}'.format(stop))\n",
    "    print(filenamePlot)\n",
    "#     fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import phylo tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ete3\n",
    "from ete3 import PhyloTree, Tree, TreeStyle, TreeNode, TreeFace, NodeStyle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phyloDataPath = p.databasesPath / 'Phylogenetic_tree_Lang2013'\n",
    "phyloTreeName = 'Lang2013_pruned_phyla_tree'\n",
    "phyloDataPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with (phyloDataPath / '{}.txt'.format(phyloTreeName)).open() as f:\n",
    "    phyloTree5 = Tree(f.read())\n",
    "phyloTree5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = TreeStyle()\n",
    "ts.scale = 300\n",
    "ts.show_branch_length = False\n",
    "ts.show_branch_support = False\n",
    "phyloTree5.render(\"%%inline\", tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "phyloPhylaList = [n.name for n in phyloTree5.get_leaves()]\n",
    "len(phyloPhylaList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phyloPhylaList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotPhylaSet = set(phylaNameBigList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(plotPhylaSet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotPhylaSet - set(phyloPhylaList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(phyloPhylaList) - plotPhylaSet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove the 4 phyla that are in the phyloTree but not in the plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(phyloTree5.get_leaves())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for n in phyloTree5.get_leaves():\n",
    "    if n.name not in plotPhylaSet:\n",
    "        print(n.name)\n",
    "        n.delete()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = TreeStyle()\n",
    "ts.scale = 300\n",
    "ts.show_branch_length = False\n",
    "ts.show_branch_support = False\n",
    "phyloTree5.render(\"%%inline\", tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(phyloTree5.get_leaves())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phyloTree5Filename = str(phyloDataPath / 'phyloTree5_codon_bias.txt')\n",
    "phyloTree5.write(outfile=phyloTree5Filename, format=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = Phylo.read(phyloTree5Filename, 'newick')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The parser cannot understand names with whitespace, so we correct them..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# node = [n for n in tree.get_terminals() if n.name == 'Cloacimonetes'][0]\n",
    "# node.name = 'Candidatus Cloacimonetes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "leaves = tree.get_terminals()\n",
    "leavesXPos = [tree.distance(n) for n in leaves]\n",
    "nLeaves = len(leaves)\n",
    "nLeaves"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "leaves"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Biopython PhyloTree package chooses a different order for the sister leaves sometimes. We re-adjust the order of the phyla."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phyloPhylaPlotList = [n.name for n in leaves]\n",
    "len(phyloPhylaPlotList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(set(phyloPhylaPlotList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(plotPhylaSet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotPhylaSet - set(phyloPhylaPlotList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(phyloPhylaPlotList) - plotPhylaSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plotPhylaList = (\n",
    "    phyloPhylaPlotList +\n",
    "    list(plotPhylaSet - set(phyloPhylaPlotList)))\n",
    "print(len(plotPhylaList))\n",
    "plotPhylaList"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "check that the tree leaves correspond to the correct phyla in the heatmap plot rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "fig, ax = plt.subplots(figsize=(12,8))\n",
    "\n",
    "# Set line width of the tree, based on the default linewidth\n",
    "lw = 0.7\n",
    "tree.root.width = lw / plt.rcParams['lines.linewidth']\n",
    "\n",
    "Phylo.draw(tree, axes=ax, do_show=False, show_confidence=False)\n",
    "# Phylo.draw(tree, axes=ax, do_show=False, show_confidence=False, label_func=lambda x: None)\n",
    "nLeaves = len(leavesXPos)\n",
    "for i, x in enumerate(leavesXPos):\n",
    "    y = i + 1\n",
    "    ax.plot([x + 0.02, 1], [y, y], lw=lw, ls='--', c='0.4')\n",
    "    ax.annotate(plotData.index[i], xy=(1.05, y), ha='left', va='center')\n",
    "    \n",
    "for j in range(nLeaves, len(plotPhylaList)):\n",
    "    print(j)\n",
    "    y = j + 1\n",
    "    ax.annotate(plotData.index[j], xy=(1.05, y), ha='left', va='center')\n",
    "\n",
    "ax.set_xlim(0, 1.2)\n",
    "ax.set_ylim(0.5, len(plotPhylaList) + 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot heatmap codon biases for phyla"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonNSeqDf = oddsRatioCodonTaxon[oddsRatioCodonTaxon['taxon_rank'] == 'phylum'].sort_values('taxon_name')\\\n",
    "              [['taxaid', 'taxon_name', 'n_seq', 'n_genomes', 'stop_codon']]\\\n",
    "              .drop_duplicates().reset_index(drop=True)\n",
    "taxonNSeqDf = taxonNSeqDf.set_index(['taxon_name', 'stop_codon']).loc[plotPhylaList]\n",
    "print(len(taxonNSeqDf))\n",
    "taxonNSeqDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to compute the small N, we have to apply the method to every case: taxaid, stop codon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioCodonTaxon[oddsRatioCodonTaxon['taxaid'] == taxonId]\\\n",
    ".drop(['n_genomes', 'n_seq', 'taxaid', 'taxon_rank', 'taxon_name'], axis=1).set_index('stop_codon', append=True)\n",
    "df = df.xs('TAA', level='stop_codon')\n",
    "compute_codon_smallN_mask_data(df, pvalueThresholds, terminus);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioCodonTaxon.copy()\n",
    "terminus = 'C'\n",
    "terminusPos = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf[oddsRatioDf['taxon_rank'] == 'phylum'].xs(terminusPos, level='position from terminus').xs(terminus, level='terminus')\n",
    "oddsRatioDf = oddsRatioDf.sort_values('n_seq', ascending=False)\n",
    "oddsRatioDf = oddsRatioDf.drop(['n_genomes', 'n_seq', 'taxaid', 'taxon_rank'], axis=1)\\\n",
    "                         .set_index('taxon_name', append=True).sort_index()\n",
    "oddsRatioDf = oddsRatioDf.loc[oddsRatioDf.index.get_level_values('taxon_name').map(lambda x: x in phylaNameBigList)]\n",
    "oddsRatioDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf.set_index('stop_codon', append=True).loc[idx[:, :, ['TAA', 'TAG', 'TGA']], :].unstack(level=2)\n",
    "oddsRatioDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = oddsRatioDf.copy()\n",
    "plotData = plotData.xs('log2OddsRatio', level='observable')\n",
    "plotData = plotData[plotData.columns].astype(float)\n",
    "print(len(plotData))\n",
    "plotData = plotData.loc[plotPhylaList]\n",
    "plotData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskData = oddsRatioDf.xs('BH_multiple_tests', level='observable')\n",
    "# Mask both rejected cases and null values\n",
    "maskData = (maskData != True) & (maskData != 'True')\n",
    "maskData = maskData.loc[plotPhylaList]\n",
    "maskData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData.columns.names = ['codon', 'stop_codon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData2 = plotData.T.reset_index(level=1)\n",
    "maskData2 = maskData.T.reset_index(level=1)\n",
    "plotData2 = sort_codon_index(plotData2, refCodonTableDf,\n",
    "                             addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                             addEmptyRowBetweenGroups=False)\n",
    "plotData2 = plotData2.set_index('stop_codon', append=True)\n",
    "plotData2 = plotData2.loc[idx[:, ['TAA', 'TAG', 'TGA']], :]\n",
    "maskData2 = sort_codon_index(maskData2, refCodonTableDf,\n",
    "                             addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False,\n",
    "                             addEmptyRowBetweenGroups=False)\n",
    "maskData2 = maskData2.set_index('stop_codon', append=True)\n",
    "maskData2 = maskData2.loc[idx[:, ['TAA', 'TAG', 'TGA']], :]\n",
    "plotData2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codonSortedList = plotData2.index.get_level_values(0).drop_duplicates()\n",
    "codonSortedList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "codonSortedDict = dict(zip(codonSortedList, range(len(codonSortedList))))\n",
    "stopSortedDict = {'TAA':0, 'TAG':1, 'TGA':2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData12 = plotData2.reset_index()\n",
    "plotData12['sort_codon'] = plotData12['codon'].map(codonSortedDict)\n",
    "plotData12['sort_stop'] = plotData12['stop_codon'].map(stopSortedDict)\n",
    "plotData12 = plotData12.sort_values(['sort_codon', 'sort_stop']).set_index(['codon', 'stop_codon'])\n",
    "plotData12.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskData12 = maskData2.loc[plotData12.index]\n",
    "maskData12.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData3 = plotData12.T\n",
    "maskData3 = maskData12.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codonList = ['R/CGA', 'R/AGA', 'R/AGG', 'K/AAA', 'K/AAG', 'T/ACU', 'T/ACC', 'T/ACA', 'T/ACG',\n",
    "             'P/CCU', 'P/CCC', 'P/CCA', 'P/CCG']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Order taxa following nb of sequences in the all category."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# taxaOrder = taxonNSeqDf.xs('all', level='stop_codon').sort_values('n_seq', ascending=False).index\n",
    "taxaOrder = plotPhylaList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "taxonNSeqDf2 = taxonNSeqDf.loc[idx[:, 'all'], :].reset_index(level=1, drop=True).reset_index().set_index('taxon_name').loc[taxaOrder]\n",
    "taxonNSeqDf2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData3 = plotData3.loc[taxaOrder, codonList]\n",
    "maskData3 = maskData3.loc[taxaOrder, codonList]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData3.columns.set_levels(['UAA', 'UAG', 'UGA'], level='stop_codon', inplace=True)\n",
    "maskData3.columns.set_levels(['UAA', 'UAG', 'UGA'], level='stop_codon', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.transforms as transforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FontProperties(size='medium').get_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FontProperties(size='small').get_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vmax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fontSizeLabels = FontProperties(size='small').get_size()\n",
    "\n",
    "linewidth1 = 0.5\n",
    "aspectratio = 1.5\n",
    "n = len(codonList)\n",
    "figsize = (7, 1.45)\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "fig = plt.figure(figsize=figsize)\n",
    "gsOuter = gridspec.GridSpec(1, 3, width_ratios = [0.5, n, 1], wspace=0.1)\n",
    "gs = gridspec.GridSpecFromSubplotSpec(1, n, subplot_spec=gsOuter[1], width_ratios=n*[1], wspace=0.2)\n",
    "\n",
    "\n",
    "ax = fig.add_subplot(gsOuter[0])\n",
    "# Phylo.draw(tree, axes=ax, do_show=False, show_confidence=False)\n",
    "Phylo.draw(tree, axes=ax, do_show=False, show_confidence=False, label_func=lambda x: None)\n",
    "nLeaves = len(leavesXPos)\n",
    "for i, x in enumerate(leavesXPos):\n",
    "    y = i + 1\n",
    "    ax.plot([x + 0.02, 1], [y, y], lw=lw, ls='--', c='0.4')\n",
    "#         ax.annotate(plotData.index[i], xy=(1.05, y), ha='left', va='center')\n",
    "ax.set_ylim(0.5, len(plotPhylaList) + 0.5)\n",
    "ax.set_xlim(0, 0.4)\n",
    "ax.axis('off')\n",
    "ax.annotate('phylum', xy=(0, 0.5), xytext=(-5, 0), xycoords='axes fraction', textcoords='offset points',\n",
    "            ha='right', va='center', rotation=90)\n",
    "\n",
    "\n",
    "for i, codon in enumerate(codonList):\n",
    "        \n",
    "    ax = fig.add_subplot(gs[0, i])\n",
    "    ax.set_facecolor(colorBackground1)\n",
    "    plotData4 = plotData3.loc[:, [codon]][::-1]\n",
    "    maskData4 = maskData3.loc[:, [codon]][::-1]\n",
    "    if i == 0:\n",
    "        # Create axes for the colorbar for more precise positioning\n",
    "        # vertical bar\n",
    "        cbar_aspect = 1\n",
    "        cbar_length = 0.4\n",
    "        cbarLabel = '$\\log_2$(odds ratio)'\n",
    "        ax = seaborn.heatmap(plotData4, square=True, mask=maskData4, ax=ax, cmap=cmap,\n",
    "                             cbar=cbar_ax, vmin=-vmax, vmax=vmax, cbar_kws={\"label\":cbarLabel, 'aspect':15},\n",
    "                             xticklabels=True, yticklabels=True,\n",
    "                             linewidth=linewidth1, linecolor=colorHeatmapLine)\n",
    "        cbar_ax = fig.axes[-1]\n",
    "        cbar_ax.set_position([0.5, -0.3 - cbar_length, cbar_length/cbar_aspect, cbar_length])\n",
    "        tickLength = FontProperties(size='small').get_size()/4\n",
    "#         cbar_ax.xaxis.set_ticks([-vmax, 0, vmax1])\n",
    "        cbar_ax.tick_params(axis='y', length=tickLength, color=colorAxis, labelsize='small')\n",
    "    else:\n",
    "        ax = seaborn.heatmap(plotData4, square=True, mask=maskData4, ax=ax, cmap=cmap,\n",
    "                             cbar=False, vmin=-vmax, vmax=vmax,\n",
    "                             xticklabels=True, yticklabels=True,\n",
    "                             linewidth=linewidth1, linecolor=colorHeatmapLine)\n",
    "    ax.set_xlabel('')\n",
    "    if i == 0:\n",
    "        ax.set_ylabel('')\n",
    "#         ax.tick_params(which='y', reset=True, left=True, labelleft=True)\n",
    "        for tick in ax.get_yticklabels():\n",
    "            tick.set_fontproperties(FontProperties(size=0.8*fontSizeLabels))\n",
    "            tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    else:\n",
    "        ax.set_ylabel('')\n",
    "        ax.tick_params(labelleft=False)\n",
    "    family = 'Liberation Mono'\n",
    "    ticks_font = FontProperties(family=family, size=fontSizeLabels)\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    stopCodonIndex = plotData4.columns.get_level_values(1)\n",
    "    tform = transforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "    for yi, stopCodon in enumerate(stopCodonIndex):\n",
    "        ax.annotate(stopCodon, xy=(yi + 0.5, 1.0), xycoords=tform,\n",
    "                    xytext=(0, 6), textcoords='offset points', ha='center', va='bottom',\n",
    "                    rotation=90, bbox=dict(pad=0, facecolor='none', edgecolor='none'),\n",
    "                    fontproperties=FontProperties(family=family, size=fontSizeLabels))\n",
    "    ax.tick_params(labelbottom=False)\n",
    "    ax.annotate(codon, xy=(0.5, 0.0), xycoords='axes fraction',\n",
    "                xytext=(0, -6), textcoords='offset points', ha='center', va='top',\n",
    "                rotation=0, bbox=dict(pad=0, facecolor='none', edgecolor='none'),\n",
    "                fontproperties=FontProperties(family=family, size=fontSizeLabels))\n",
    "\n",
    "\n",
    "ax.annotate('codon at ' + terminus + '-terminal', xy=(0.5, 0.06), xycoords='figure fraction', ha='center', va='top')\n",
    "ax.annotate('stop codon context', xy=(0.5, 0.95), xycoords='figure fraction', ha='center', va='bottom')\n",
    "\n",
    "ax = fig.add_subplot(gsOuter[-1])\n",
    "colorNBar = sns.color_palette()[7]\n",
    "colorNBar = '0.5'\n",
    "print(len(taxonNSeqDf))\n",
    "with sns.axes_style('ticks'):\n",
    "    df = taxonNSeqDf2['n_seq']\n",
    "    ax.barh(range(len(df)), df.values, height=0.7, color=[colorNBar for _ in range(len(taxonNSeqDf2))])\n",
    "    ax.set_ylim(-0.5, len(df) - 0.5)\n",
    "    ax.set_xlabel('nb of protein\\nsequences')\n",
    "    ax.set_xscale('log')\n",
    "    ax.set_xlim(1e3, 2.5e6)\n",
    "#         ax.set_xticks([1e3, 1e4, 1e5, 1e6])\n",
    "    ax.set_xticks([1e3, 1e6])\n",
    "    ax.set_xticks([1e4, 1e5], minor = True)\n",
    "    ax.tick_params(axis='x', bottom=True, color='0.8', direction='in', which='both', size=fontSizeLabels)\n",
    "    ax.tick_params(axis='x', which='minor', labelbottom=False)\n",
    "    plt.setp(ax.get_yticklabels(), visible=False)\n",
    "    ax.set_ylabel('')\n",
    "#     sns.despine(ax=ax)\n",
    "    ax.grid(False)\n",
    "\n",
    "# fig.tight_layout()\n",
    "filenamePlot = analysisCtermPath / 'Taxonomic_tree' / (\"phyla_composition_bias_codon_\" + terminus + \"terminal\" +\n",
    "                                                       '_stop_codon_context')\n",
    "print(filenamePlot)\n",
    "\n",
    "# plotTitle = 'C-terminal biases by phyla'\n",
    "\n",
    "# fig.savefig(str(filenamePlot) + \".png\", dpi=500, bbox_inches=\"tight\")\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3rd nucleotide frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioCodonTaxon.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioDf[oddsRatioDf['taxon_rank'] == 'phylum']\\\n",
    "              .xs(0, level='position from terminus').xs('bulk', level='terminus')\n",
    "oddsRatioDf = oddsRatioDf.sort_values('n_seq', ascending=False)\n",
    "oddsRatioDf = oddsRatioDf.drop(['n_genomes', 'n_seq', 'taxaid', 'taxon_rank'], axis=1)\\\n",
    "                         .set_index('taxon_name', append=True).sort_index()\n",
    "oddsRatioDf = oddsRatioDf.loc[oddsRatioDf.index.get_level_values('taxon_name').map(lambda x: x in phylaNameBigList)]\n",
    "oddsRatioDf = oddsRatioDf[oddsRatioDf['stop_codon'] == 'all'].drop('stop_codon', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDf.T\n",
    "df['N3'] = df.index.map(lambda x: x[-1])\n",
    "df = df.groupby('N3').sum()\n",
    "df = df/df.sum()\n",
    "baseFreqDf = df.T.loc['count']\n",
    "baseFreqDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df6 = df5.loc[idx[:, 'TGA'], :]\n",
    "df6 = df6.set_index('phylum', append=True)\n",
    "df7 = df6[df6['G3']]['value']\n",
    "df7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df8 = df7.reorder_levels([2,0,1]).sort_index().reset_index()\n",
    "df8.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df8['phylum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseFreqDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.barplot(data=df8, x='phylum', y='value')\n",
    "\n",
    "n = len(baseFreqDf)\n",
    "ax.plot(np.arange(0, n), baseFreqDf['G'], markersize=5)\n",
    "ax.set_title('stop codon TGA, NNG codons')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for nuc in 'ATGC':\n",
    "\n",
    "    for stop in ['TAA', 'TAG', 'TGA'][:]:\n",
    "        print(\"#######\", stop)\n",
    "\n",
    "        df6 = df5.loc[idx[:, stop], :]\n",
    "        df6 = df6.set_index('phylum', append=True)\n",
    "        df7 = df6[df6['{}3'.format(nuc)]]['value']\n",
    "        df8 = df7.reorder_levels([2,0,1]).sort_index().reset_index()\n",
    "        fig, ax = plt.subplots()\n",
    "        sns.barplot(data=df8, x='phylum', y='value', ax=ax)\n",
    "        n = len(baseFreqDf)\n",
    "        ax.plot(np.arange(0, n), baseFreqDf[nuc], markersize=5)\n",
    "        ax.set_title('stop codon {}, NN{} codons'.format(stop, nuc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biases for COG categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 302,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COGList = ['J', 'U', 'Y', 'P', 'T', 'A', 'M', 'F', 'W', 'V', 'K', 'G', 'D', 'H', 'E', 'Q', 'B', 'Z', 'N', 'S', 'I', 'O', 'L', 'R', 'C']\n",
    "len(COGList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/users/lserrano/mweber/Analysis_data/C-terminal/Eggnog')"
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysisEggnogPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error with COG Y\n",
      "error with COG R\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>terminus</th>\n",
       "      <th>position from terminus</th>\n",
       "      <th>observable</th>\n",
       "      <th>COG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">-20</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <th>J</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <th>J</th>\n",
       "      <td>25334</td>\n",
       "      <td>1356</td>\n",
       "      <td>12928</td>\n",
       "      <td>18203</td>\n",
       "      <td>8290</td>\n",
       "      <td>17969</td>\n",
       "      <td>4224</td>\n",
       "      <td>14922</td>\n",
       "      <td>19524</td>\n",
       "      <td>21792</td>\n",
       "      <td>5006</td>\n",
       "      <td>7626</td>\n",
       "      <td>8482</td>\n",
       "      <td>8265</td>\n",
       "      <td>14087</td>\n",
       "      <td>11239</td>\n",
       "      <td>14108</td>\n",
       "      <td>23001</td>\n",
       "      <td>1576</td>\n",
       "      <td>4805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <th>J</th>\n",
       "      <td>0.20365834512229747</td>\n",
       "      <td>-0.7628573739339086</td>\n",
       "      <td>-0.19570543030927692</td>\n",
       "      <td>0.01136549014058734</td>\n",
       "      <td>-0.06990521152832739</td>\n",
       "      <td>-0.06118697239794936</td>\n",
       "      <td>-0.4002303331157669</td>\n",
       "      <td>0.06969359559989818</td>\n",
       "      <td>0.4510295332278455</td>\n",
       "      <td>-0.014653159257426655</td>\n",
       "      <td>-0.008660460531598245</td>\n",
       "      <td>-0.15326181812043072</td>\n",
       "      <td>-0.40239376644703795</td>\n",
       "      <td>-0.05923945539447665</td>\n",
       "      <td>-0.20838965508851515</td>\n",
       "      <td>-0.10376733046245995</td>\n",
       "      <td>0.19285938042580753</td>\n",
       "      <td>0.34235340723685004</td>\n",
       "      <td>-0.6854295899859104</td>\n",
       "      <td>-0.513072907100442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <th>J</th>\n",
       "      <td>1.1516148879385883</td>\n",
       "      <td>0.5893279624926393</td>\n",
       "      <td>0.8731458519860368</td>\n",
       "      <td>1.007909070201383</td>\n",
       "      <td>0.9527005906666536</td>\n",
       "      <td>0.9584752126400036</td>\n",
       "      <td>0.7577372972388677</td>\n",
       "      <td>1.0494937649543479</td>\n",
       "      <td>1.3670154356555946</td>\n",
       "      <td>0.9898946100396573</td>\n",
       "      <td>0.99401500804788</td>\n",
       "      <td>0.8992151091372954</td>\n",
       "      <td>0.7566018628237838</td>\n",
       "      <td>0.9597699472833596</td>\n",
       "      <td>0.8655027717538291</td>\n",
       "      <td>0.9305997273596478</td>\n",
       "      <td>1.1430269190802806</td>\n",
       "      <td>1.2678230540285567</td>\n",
       "      <td>0.6218206401145754</td>\n",
       "      <td>0.700728312094303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <th>J</th>\n",
       "      <td>2.5105342858031816e-96</td>\n",
       "      <td>8.424573116105584e-100</td>\n",
       "      <td>1.0442320486160008e-52</td>\n",
       "      <td>0.30823451763308235</td>\n",
       "      <td>1.3428899929755472e-05</td>\n",
       "      <td>4.096080431258017e-08</td>\n",
       "      <td>1.0553133977827732e-77</td>\n",
       "      <td>1.4673692990941998e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.15437172453467912</td>\n",
       "      <td>0.6842100946146605</td>\n",
       "      <td>2.099594701671073e-20</td>\n",
       "      <td>1.4801401656654558e-152</td>\n",
       "      <td>0.0002374153166798272</td>\n",
       "      <td>1.6711281507470546e-64</td>\n",
       "      <td>6.065931656897627e-14</td>\n",
       "      <td>2.1506747123962817e-51</td>\n",
       "      <td>3.8051634403992606e-240</td>\n",
       "      <td>6.31821944189036e-92</td>\n",
       "      <td>5.77917087493818e-147</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                            A  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                     25334   \n",
       "                                log2OddsRatio     J       0.20365834512229747   \n",
       "                                oddsRatio         J        1.1516148879385883   \n",
       "                                pvalue            J    2.5105342858031816e-96   \n",
       "\n",
       "                                                                            C  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                      1356   \n",
       "                                log2OddsRatio     J       -0.7628573739339086   \n",
       "                                oddsRatio         J        0.5893279624926393   \n",
       "                                pvalue            J    8.424573116105584e-100   \n",
       "\n",
       "                                                                            D  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                     12928   \n",
       "                                log2OddsRatio     J      -0.19570543030927692   \n",
       "                                oddsRatio         J        0.8731458519860368   \n",
       "                                pvalue            J    1.0442320486160008e-52   \n",
       "\n",
       "                                                                         E  \\\n",
       "terminus position from terminus observable        COG                        \n",
       "C        -20                    BH_multiple_tests J                  False   \n",
       "                                count             J                  18203   \n",
       "                                log2OddsRatio     J    0.01136549014058734   \n",
       "                                oddsRatio         J      1.007909070201383   \n",
       "                                pvalue            J    0.30823451763308235   \n",
       "\n",
       "                                                                            F  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                      8290   \n",
       "                                log2OddsRatio     J      -0.06990521152832739   \n",
       "                                oddsRatio         J        0.9527005906666536   \n",
       "                                pvalue            J    1.3428899929755472e-05   \n",
       "\n",
       "                                                                           G  \\\n",
       "terminus position from terminus observable        COG                          \n",
       "C        -20                    BH_multiple_tests J                     True   \n",
       "                                count             J                    17969   \n",
       "                                log2OddsRatio     J     -0.06118697239794936   \n",
       "                                oddsRatio         J       0.9584752126400036   \n",
       "                                pvalue            J    4.096080431258017e-08   \n",
       "\n",
       "                                                                            H  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                      4224   \n",
       "                                log2OddsRatio     J       -0.4002303331157669   \n",
       "                                oddsRatio         J        0.7577372972388677   \n",
       "                                pvalue            J    1.0553133977827732e-77   \n",
       "\n",
       "                                                                            I  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                     14922   \n",
       "                                log2OddsRatio     J       0.06969359559989818   \n",
       "                                oddsRatio         J        1.0494937649543479   \n",
       "                                pvalue            J    1.4673692990941998e-08   \n",
       "\n",
       "                                                                        K  \\\n",
       "terminus position from terminus observable        COG                       \n",
       "C        -20                    BH_multiple_tests J                  True   \n",
       "                                count             J                 19524   \n",
       "                                log2OddsRatio     J    0.4510295332278455   \n",
       "                                oddsRatio         J    1.3670154356555946   \n",
       "                                pvalue            J                   0.0   \n",
       "\n",
       "                                                                           L  \\\n",
       "terminus position from terminus observable        COG                          \n",
       "C        -20                    BH_multiple_tests J                    False   \n",
       "                                count             J                    21792   \n",
       "                                log2OddsRatio     J    -0.014653159257426655   \n",
       "                                oddsRatio         J       0.9898946100396573   \n",
       "                                pvalue            J      0.15437172453467912   \n",
       "\n",
       "                                                                           M  \\\n",
       "terminus position from terminus observable        COG                          \n",
       "C        -20                    BH_multiple_tests J                    False   \n",
       "                                count             J                     5006   \n",
       "                                log2OddsRatio     J    -0.008660460531598245   \n",
       "                                oddsRatio         J         0.99401500804788   \n",
       "                                pvalue            J       0.6842100946146605   \n",
       "\n",
       "                                                                           N  \\\n",
       "terminus position from terminus observable        COG                          \n",
       "C        -20                    BH_multiple_tests J                     True   \n",
       "                                count             J                     7626   \n",
       "                                log2OddsRatio     J     -0.15326181812043072   \n",
       "                                oddsRatio         J       0.8992151091372954   \n",
       "                                pvalue            J    2.099594701671073e-20   \n",
       "\n",
       "                                                                             P  \\\n",
       "terminus position from terminus observable        COG                            \n",
       "C        -20                    BH_multiple_tests J                       True   \n",
       "                                count             J                       8482   \n",
       "                                log2OddsRatio     J       -0.40239376644703795   \n",
       "                                oddsRatio         J         0.7566018628237838   \n",
       "                                pvalue            J    1.4801401656654558e-152   \n",
       "\n",
       "                                                                           Q  \\\n",
       "terminus position from terminus observable        COG                          \n",
       "C        -20                    BH_multiple_tests J                     True   \n",
       "                                count             J                     8265   \n",
       "                                log2OddsRatio     J     -0.05923945539447665   \n",
       "                                oddsRatio         J       0.9597699472833596   \n",
       "                                pvalue            J    0.0002374153166798272   \n",
       "\n",
       "                                                                            R  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                     14087   \n",
       "                                log2OddsRatio     J      -0.20838965508851515   \n",
       "                                oddsRatio         J        0.8655027717538291   \n",
       "                                pvalue            J    1.6711281507470546e-64   \n",
       "\n",
       "                                                                           S  \\\n",
       "terminus position from terminus observable        COG                          \n",
       "C        -20                    BH_multiple_tests J                     True   \n",
       "                                count             J                    11239   \n",
       "                                log2OddsRatio     J     -0.10376733046245995   \n",
       "                                oddsRatio         J       0.9305997273596478   \n",
       "                                pvalue            J    6.065931656897627e-14   \n",
       "\n",
       "                                                                            T  \\\n",
       "terminus position from terminus observable        COG                           \n",
       "C        -20                    BH_multiple_tests J                      True   \n",
       "                                count             J                     14108   \n",
       "                                log2OddsRatio     J       0.19285938042580753   \n",
       "                                oddsRatio         J        1.1430269190802806   \n",
       "                                pvalue            J    2.1506747123962817e-51   \n",
       "\n",
       "                                                                             V  \\\n",
       "terminus position from terminus observable        COG                            \n",
       "C        -20                    BH_multiple_tests J                       True   \n",
       "                                count             J                      23001   \n",
       "                                log2OddsRatio     J        0.34235340723685004   \n",
       "                                oddsRatio         J         1.2678230540285567   \n",
       "                                pvalue            J    3.8051634403992606e-240   \n",
       "\n",
       "                                                                          W  \\\n",
       "terminus position from terminus observable        COG                         \n",
       "C        -20                    BH_multiple_tests J                    True   \n",
       "                                count             J                    1576   \n",
       "                                log2OddsRatio     J     -0.6854295899859104   \n",
       "                                oddsRatio         J      0.6218206401145754   \n",
       "                                pvalue            J    6.31821944189036e-92   \n",
       "\n",
       "                                                                           Y  \n",
       "terminus position from terminus observable        COG                         \n",
       "C        -20                    BH_multiple_tests J                     True  \n",
       "                                count             J                     4805  \n",
       "                                log2OddsRatio     J       -0.513072907100442  \n",
       "                                oddsRatio         J        0.700728312094303  \n",
       "                                pvalue            J    5.77917087493818e-147  "
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfList = []\n",
    "for COG in COGList[:]:\n",
    "    try:\n",
    "        df = pd.read_csv(analysisEggnogPath / 'Analysis_COG_groups_cterminal_bias' / 'COG_{}'.format(COG) /\n",
    "                         'taxaId_2' / 'taxaId_2_subset_all_oddsRatioAADf.csv', index_col=[0,1,2])\n",
    "    except:\n",
    "        print(\"error with COG\", COG)\n",
    "        df = pd.DataFrame()\n",
    "        pass\n",
    "    df['COG'] = COG\n",
    "    df = df.set_index('COG', append=True)\n",
    "    dfList.append(df)\n",
    "\n",
    "oddsRatioDfCOG = pd.concat(dfList)\n",
    "oddsRatioDfCOG.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Good enough because these 2 categories are not relevant for bacteria:\n",
    "```\n",
    "Y\tNuclear structure\n",
    "R\tGeneral Functional Prediction only\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "bulk frequency in each of the COG cat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mweber/.local/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:1472: PerformanceWarning: indexing past lexsort depth may impact performance.\n",
      "  return self._getitem_tuple(key)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "COG\n",
       "J     57542001\n",
       "U     22159626\n",
       "P     82951198\n",
       "T     77575303\n",
       "A       122440\n",
       "M     84970370\n",
       "F     31648274\n",
       "W       475381\n",
       "V     30742187\n",
       "K     62572015\n",
       "G     84868513\n",
       "D     12702385\n",
       "H     40956460\n",
       "E    106427917\n",
       "Q     32051238\n",
       "B       430735\n",
       "Z       326212\n",
       "N     12280031\n",
       "S    310117333\n",
       "I     38414867\n",
       "O     48493762\n",
       "L     78310641\n",
       "C     89383583\n",
       "dtype: int64"
      ]
     },
     "execution_count": 305,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDfCOG.loc[idx['bulk', 0, 'count'], aaTable].astype(int)\n",
    "countTot = df.apply(lambda x: x.sum(), axis=1)\n",
    "countTot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R    0.058572\n",
       "K    0.043620\n",
       "H    0.021039\n",
       "D    0.056267\n",
       "E    0.061573\n",
       "S    0.057636\n",
       "T    0.054250\n",
       "N    0.035320\n",
       "Q    0.036420\n",
       "C    0.008988\n",
       "G    0.079886\n",
       "P    0.046751\n",
       "A    0.102642\n",
       "V    0.074226\n",
       "I    0.058388\n",
       "L    0.103084\n",
       "M    0.020806\n",
       "F    0.038770\n",
       "W    0.012752\n",
       "Y    0.029008\n",
       "dtype: float64"
      ]
     },
     "execution_count": 306,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aaFreq = df.sum(axis=0)\n",
    "aaFreq = aaFreq / aaFreq.sum()\n",
    "aaFreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>COG</th>\n",
       "      <th>J</th>\n",
       "      <th>U</th>\n",
       "      <th>P</th>\n",
       "      <th>T</th>\n",
       "      <th>A</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>V</th>\n",
       "      <th>K</th>\n",
       "      <th>...</th>\n",
       "      <th>E</th>\n",
       "      <th>Q</th>\n",
       "      <th>B</th>\n",
       "      <th>Z</th>\n",
       "      <th>N</th>\n",
       "      <th>S</th>\n",
       "      <th>I</th>\n",
       "      <th>O</th>\n",
       "      <th>L</th>\n",
       "      <th>C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>0.066453</td>\n",
       "      <td>0.053370</td>\n",
       "      <td>0.049685</td>\n",
       "      <td>0.067162</td>\n",
       "      <td>0.069414</td>\n",
       "      <td>0.057463</td>\n",
       "      <td>0.055864</td>\n",
       "      <td>0.019944</td>\n",
       "      <td>0.056161</td>\n",
       "      <td>0.076650</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053374</td>\n",
       "      <td>0.062814</td>\n",
       "      <td>0.061855</td>\n",
       "      <td>0.038073</td>\n",
       "      <td>0.048114</td>\n",
       "      <td>0.057326</td>\n",
       "      <td>0.058245</td>\n",
       "      <td>0.055744</td>\n",
       "      <td>0.072145</td>\n",
       "      <td>0.053945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>0.060137</td>\n",
       "      <td>0.045387</td>\n",
       "      <td>0.036304</td>\n",
       "      <td>0.040656</td>\n",
       "      <td>0.043368</td>\n",
       "      <td>0.046297</td>\n",
       "      <td>0.046316</td>\n",
       "      <td>0.058713</td>\n",
       "      <td>0.043001</td>\n",
       "      <td>0.042045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041827</td>\n",
       "      <td>0.029094</td>\n",
       "      <td>0.030425</td>\n",
       "      <td>0.024941</td>\n",
       "      <td>0.043528</td>\n",
       "      <td>0.042559</td>\n",
       "      <td>0.040594</td>\n",
       "      <td>0.050442</td>\n",
       "      <td>0.054888</td>\n",
       "      <td>0.043665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>0.022838</td>\n",
       "      <td>0.013736</td>\n",
       "      <td>0.016904</td>\n",
       "      <td>0.021835</td>\n",
       "      <td>0.033126</td>\n",
       "      <td>0.019702</td>\n",
       "      <td>0.023709</td>\n",
       "      <td>0.008507</td>\n",
       "      <td>0.016554</td>\n",
       "      <td>0.023700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022572</td>\n",
       "      <td>0.024923</td>\n",
       "      <td>0.038903</td>\n",
       "      <td>0.018304</td>\n",
       "      <td>0.012867</td>\n",
       "      <td>0.020347</td>\n",
       "      <td>0.022591</td>\n",
       "      <td>0.019082</td>\n",
       "      <td>0.023296</td>\n",
       "      <td>0.022773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.060528</td>\n",
       "      <td>0.053614</td>\n",
       "      <td>0.048341</td>\n",
       "      <td>0.058118</td>\n",
       "      <td>0.057971</td>\n",
       "      <td>0.055921</td>\n",
       "      <td>0.060294</td>\n",
       "      <td>0.065350</td>\n",
       "      <td>0.050419</td>\n",
       "      <td>0.057607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.055184</td>\n",
       "      <td>0.062216</td>\n",
       "      <td>0.068269</td>\n",
       "      <td>0.065583</td>\n",
       "      <td>0.055966</td>\n",
       "      <td>0.056882</td>\n",
       "      <td>0.055326</td>\n",
       "      <td>0.059719</td>\n",
       "      <td>0.058992</td>\n",
       "      <td>0.054205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.074446</td>\n",
       "      <td>0.057041</td>\n",
       "      <td>0.048252</td>\n",
       "      <td>0.069882</td>\n",
       "      <td>0.073661</td>\n",
       "      <td>0.055949</td>\n",
       "      <td>0.067486</td>\n",
       "      <td>0.040990</td>\n",
       "      <td>0.053979</td>\n",
       "      <td>0.074549</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060166</td>\n",
       "      <td>0.059338</td>\n",
       "      <td>0.059340</td>\n",
       "      <td>0.032770</td>\n",
       "      <td>0.062060</td>\n",
       "      <td>0.059380</td>\n",
       "      <td>0.059791</td>\n",
       "      <td>0.068536</td>\n",
       "      <td>0.072061</td>\n",
       "      <td>0.060812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.049583</td>\n",
       "      <td>0.065849</td>\n",
       "      <td>0.061714</td>\n",
       "      <td>0.059264</td>\n",
       "      <td>0.056117</td>\n",
       "      <td>0.060622</td>\n",
       "      <td>0.052802</td>\n",
       "      <td>0.088327</td>\n",
       "      <td>0.061900</td>\n",
       "      <td>0.057485</td>\n",
       "      <td>...</td>\n",
       "      <td>0.055238</td>\n",
       "      <td>0.054989</td>\n",
       "      <td>0.046553</td>\n",
       "      <td>0.077906</td>\n",
       "      <td>0.072762</td>\n",
       "      <td>0.060552</td>\n",
       "      <td>0.053328</td>\n",
       "      <td>0.056767</td>\n",
       "      <td>0.055060</td>\n",
       "      <td>0.052344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>0.051220</td>\n",
       "      <td>0.060635</td>\n",
       "      <td>0.057594</td>\n",
       "      <td>0.051269</td>\n",
       "      <td>0.053063</td>\n",
       "      <td>0.055383</td>\n",
       "      <td>0.053192</td>\n",
       "      <td>0.095736</td>\n",
       "      <td>0.056669</td>\n",
       "      <td>0.050489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053560</td>\n",
       "      <td>0.057377</td>\n",
       "      <td>0.044844</td>\n",
       "      <td>0.093320</td>\n",
       "      <td>0.063347</td>\n",
       "      <td>0.055863</td>\n",
       "      <td>0.051632</td>\n",
       "      <td>0.053330</td>\n",
       "      <td>0.050694</td>\n",
       "      <td>0.053203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.034815</td>\n",
       "      <td>0.041664</td>\n",
       "      <td>0.035710</td>\n",
       "      <td>0.034287</td>\n",
       "      <td>0.029231</td>\n",
       "      <td>0.040271</td>\n",
       "      <td>0.035194</td>\n",
       "      <td>0.086878</td>\n",
       "      <td>0.035927</td>\n",
       "      <td>0.029731</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034242</td>\n",
       "      <td>0.029806</td>\n",
       "      <td>0.029603</td>\n",
       "      <td>0.047310</td>\n",
       "      <td>0.048153</td>\n",
       "      <td>0.037233</td>\n",
       "      <td>0.030921</td>\n",
       "      <td>0.035960</td>\n",
       "      <td>0.034288</td>\n",
       "      <td>0.032743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>0.035426</td>\n",
       "      <td>0.043274</td>\n",
       "      <td>0.033005</td>\n",
       "      <td>0.042043</td>\n",
       "      <td>0.036303</td>\n",
       "      <td>0.038073</td>\n",
       "      <td>0.032779</td>\n",
       "      <td>0.031873</td>\n",
       "      <td>0.039290</td>\n",
       "      <td>0.039008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032933</td>\n",
       "      <td>0.034221</td>\n",
       "      <td>0.031862</td>\n",
       "      <td>0.035419</td>\n",
       "      <td>0.048359</td>\n",
       "      <td>0.037447</td>\n",
       "      <td>0.031603</td>\n",
       "      <td>0.039183</td>\n",
       "      <td>0.040775</td>\n",
       "      <td>0.031852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.009442</td>\n",
       "      <td>0.004852</td>\n",
       "      <td>0.005930</td>\n",
       "      <td>0.007823</td>\n",
       "      <td>0.010797</td>\n",
       "      <td>0.007588</td>\n",
       "      <td>0.011288</td>\n",
       "      <td>0.000886</td>\n",
       "      <td>0.005807</td>\n",
       "      <td>0.008987</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010042</td>\n",
       "      <td>0.009365</td>\n",
       "      <td>0.014775</td>\n",
       "      <td>0.017302</td>\n",
       "      <td>0.004298</td>\n",
       "      <td>0.008674</td>\n",
       "      <td>0.010478</td>\n",
       "      <td>0.008513</td>\n",
       "      <td>0.009852</td>\n",
       "      <td>0.013411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0.076987</td>\n",
       "      <td>0.082830</td>\n",
       "      <td>0.084712</td>\n",
       "      <td>0.068853</td>\n",
       "      <td>0.062863</td>\n",
       "      <td>0.079133</td>\n",
       "      <td>0.082159</td>\n",
       "      <td>0.121145</td>\n",
       "      <td>0.076488</td>\n",
       "      <td>0.067856</td>\n",
       "      <td>...</td>\n",
       "      <td>0.086523</td>\n",
       "      <td>0.086197</td>\n",
       "      <td>0.086891</td>\n",
       "      <td>0.127693</td>\n",
       "      <td>0.076550</td>\n",
       "      <td>0.078021</td>\n",
       "      <td>0.087628</td>\n",
       "      <td>0.081730</td>\n",
       "      <td>0.068621</td>\n",
       "      <td>0.087921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>0.045671</td>\n",
       "      <td>0.042692</td>\n",
       "      <td>0.044532</td>\n",
       "      <td>0.041772</td>\n",
       "      <td>0.047223</td>\n",
       "      <td>0.047127</td>\n",
       "      <td>0.045256</td>\n",
       "      <td>0.018141</td>\n",
       "      <td>0.040371</td>\n",
       "      <td>0.043680</td>\n",
       "      <td>...</td>\n",
       "      <td>0.047526</td>\n",
       "      <td>0.055465</td>\n",
       "      <td>0.055819</td>\n",
       "      <td>0.047803</td>\n",
       "      <td>0.041033</td>\n",
       "      <td>0.048415</td>\n",
       "      <td>0.047893</td>\n",
       "      <td>0.045876</td>\n",
       "      <td>0.045178</td>\n",
       "      <td>0.050002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.091890</td>\n",
       "      <td>0.099040</td>\n",
       "      <td>0.102861</td>\n",
       "      <td>0.098746</td>\n",
       "      <td>0.099886</td>\n",
       "      <td>0.099208</td>\n",
       "      <td>0.101230</td>\n",
       "      <td>0.123770</td>\n",
       "      <td>0.096001</td>\n",
       "      <td>0.102826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.107452</td>\n",
       "      <td>0.112161</td>\n",
       "      <td>0.113339</td>\n",
       "      <td>0.104377</td>\n",
       "      <td>0.100950</td>\n",
       "      <td>0.102413</td>\n",
       "      <td>0.117324</td>\n",
       "      <td>0.100913</td>\n",
       "      <td>0.093499</td>\n",
       "      <td>0.107360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>0.076266</td>\n",
       "      <td>0.075818</td>\n",
       "      <td>0.081158</td>\n",
       "      <td>0.072089</td>\n",
       "      <td>0.073187</td>\n",
       "      <td>0.074547</td>\n",
       "      <td>0.078070</td>\n",
       "      <td>0.075356</td>\n",
       "      <td>0.077963</td>\n",
       "      <td>0.071196</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076862</td>\n",
       "      <td>0.076834</td>\n",
       "      <td>0.067761</td>\n",
       "      <td>0.078572</td>\n",
       "      <td>0.073410</td>\n",
       "      <td>0.071264</td>\n",
       "      <td>0.076977</td>\n",
       "      <td>0.074410</td>\n",
       "      <td>0.065649</td>\n",
       "      <td>0.077212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>0.058745</td>\n",
       "      <td>0.064632</td>\n",
       "      <td>0.065884</td>\n",
       "      <td>0.061120</td>\n",
       "      <td>0.055211</td>\n",
       "      <td>0.058696</td>\n",
       "      <td>0.062949</td>\n",
       "      <td>0.053372</td>\n",
       "      <td>0.069066</td>\n",
       "      <td>0.055539</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060232</td>\n",
       "      <td>0.047883</td>\n",
       "      <td>0.049720</td>\n",
       "      <td>0.037574</td>\n",
       "      <td>0.062776</td>\n",
       "      <td>0.054828</td>\n",
       "      <td>0.056539</td>\n",
       "      <td>0.057660</td>\n",
       "      <td>0.056942</td>\n",
       "      <td>0.058490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>0.090609</td>\n",
       "      <td>0.102231</td>\n",
       "      <td>0.115078</td>\n",
       "      <td>0.115061</td>\n",
       "      <td>0.104329</td>\n",
       "      <td>0.099145</td>\n",
       "      <td>0.095161</td>\n",
       "      <td>0.057991</td>\n",
       "      <td>0.115628</td>\n",
       "      <td>0.111218</td>\n",
       "      <td>...</td>\n",
       "      <td>0.099267</td>\n",
       "      <td>0.102823</td>\n",
       "      <td>0.097777</td>\n",
       "      <td>0.071864</td>\n",
       "      <td>0.098610</td>\n",
       "      <td>0.103231</td>\n",
       "      <td>0.097264</td>\n",
       "      <td>0.097813</td>\n",
       "      <td>0.104130</td>\n",
       "      <td>0.095447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.020745</td>\n",
       "      <td>0.022080</td>\n",
       "      <td>0.023592</td>\n",
       "      <td>0.021004</td>\n",
       "      <td>0.024036</td>\n",
       "      <td>0.018492</td>\n",
       "      <td>0.021972</td>\n",
       "      <td>0.009123</td>\n",
       "      <td>0.023281</td>\n",
       "      <td>0.019603</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022836</td>\n",
       "      <td>0.018842</td>\n",
       "      <td>0.017939</td>\n",
       "      <td>0.009328</td>\n",
       "      <td>0.024128</td>\n",
       "      <td>0.018683</td>\n",
       "      <td>0.023620</td>\n",
       "      <td>0.019820</td>\n",
       "      <td>0.018123</td>\n",
       "      <td>0.023999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.035787</td>\n",
       "      <td>0.036806</td>\n",
       "      <td>0.045792</td>\n",
       "      <td>0.034737</td>\n",
       "      <td>0.030725</td>\n",
       "      <td>0.039070</td>\n",
       "      <td>0.036912</td>\n",
       "      <td>0.019136</td>\n",
       "      <td>0.043407</td>\n",
       "      <td>0.032195</td>\n",
       "      <td>...</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>0.036324</td>\n",
       "      <td>0.039421</td>\n",
       "      <td>0.023181</td>\n",
       "      <td>0.033774</td>\n",
       "      <td>0.040366</td>\n",
       "      <td>0.038312</td>\n",
       "      <td>0.036921</td>\n",
       "      <td>0.035715</td>\n",
       "      <td>0.038978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>0.010400</td>\n",
       "      <td>0.009122</td>\n",
       "      <td>0.014405</td>\n",
       "      <td>0.009649</td>\n",
       "      <td>0.013117</td>\n",
       "      <td>0.013069</td>\n",
       "      <td>0.008775</td>\n",
       "      <td>0.003984</td>\n",
       "      <td>0.010587</td>\n",
       "      <td>0.009350</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>0.013376</td>\n",
       "      <td>0.012028</td>\n",
       "      <td>0.024199</td>\n",
       "      <td>0.006571</td>\n",
       "      <td>0.015364</td>\n",
       "      <td>0.013529</td>\n",
       "      <td>0.010202</td>\n",
       "      <td>0.011213</td>\n",
       "      <td>0.013044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>0.028012</td>\n",
       "      <td>0.025328</td>\n",
       "      <td>0.028547</td>\n",
       "      <td>0.024631</td>\n",
       "      <td>0.026372</td>\n",
       "      <td>0.034242</td>\n",
       "      <td>0.028591</td>\n",
       "      <td>0.020777</td>\n",
       "      <td>0.027499</td>\n",
       "      <td>0.026286</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029314</td>\n",
       "      <td>0.025952</td>\n",
       "      <td>0.032876</td>\n",
       "      <td>0.024481</td>\n",
       "      <td>0.022744</td>\n",
       "      <td>0.031152</td>\n",
       "      <td>0.026404</td>\n",
       "      <td>0.027379</td>\n",
       "      <td>0.028880</td>\n",
       "      <td>0.028593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "COG         J         U         P         T         A         M         F  \\\n",
       "R    0.066453  0.053370  0.049685  0.067162  0.069414  0.057463  0.055864   \n",
       "K    0.060137  0.045387  0.036304  0.040656  0.043368  0.046297  0.046316   \n",
       "H    0.022838  0.013736  0.016904  0.021835  0.033126  0.019702  0.023709   \n",
       "D    0.060528  0.053614  0.048341  0.058118  0.057971  0.055921  0.060294   \n",
       "E    0.074446  0.057041  0.048252  0.069882  0.073661  0.055949  0.067486   \n",
       "S    0.049583  0.065849  0.061714  0.059264  0.056117  0.060622  0.052802   \n",
       "T    0.051220  0.060635  0.057594  0.051269  0.053063  0.055383  0.053192   \n",
       "N    0.034815  0.041664  0.035710  0.034287  0.029231  0.040271  0.035194   \n",
       "Q    0.035426  0.043274  0.033005  0.042043  0.036303  0.038073  0.032779   \n",
       "C    0.009442  0.004852  0.005930  0.007823  0.010797  0.007588  0.011288   \n",
       "G    0.076987  0.082830  0.084712  0.068853  0.062863  0.079133  0.082159   \n",
       "P    0.045671  0.042692  0.044532  0.041772  0.047223  0.047127  0.045256   \n",
       "A    0.091890  0.099040  0.102861  0.098746  0.099886  0.099208  0.101230   \n",
       "V    0.076266  0.075818  0.081158  0.072089  0.073187  0.074547  0.078070   \n",
       "I    0.058745  0.064632  0.065884  0.061120  0.055211  0.058696  0.062949   \n",
       "L    0.090609  0.102231  0.115078  0.115061  0.104329  0.099145  0.095161   \n",
       "M    0.020745  0.022080  0.023592  0.021004  0.024036  0.018492  0.021972   \n",
       "F    0.035787  0.036806  0.045792  0.034737  0.030725  0.039070  0.036912   \n",
       "W    0.010400  0.009122  0.014405  0.009649  0.013117  0.013069  0.008775   \n",
       "Y    0.028012  0.025328  0.028547  0.024631  0.026372  0.034242  0.028591   \n",
       "\n",
       "COG         W         V         K    ...            E         Q         B  \\\n",
       "R    0.019944  0.056161  0.076650    ...     0.053374  0.062814  0.061855   \n",
       "K    0.058713  0.043001  0.042045    ...     0.041827  0.029094  0.030425   \n",
       "H    0.008507  0.016554  0.023700    ...     0.022572  0.024923  0.038903   \n",
       "D    0.065350  0.050419  0.057607    ...     0.055184  0.062216  0.068269   \n",
       "E    0.040990  0.053979  0.074549    ...     0.060166  0.059338  0.059340   \n",
       "S    0.088327  0.061900  0.057485    ...     0.055238  0.054989  0.046553   \n",
       "T    0.095736  0.056669  0.050489    ...     0.053560  0.057377  0.044844   \n",
       "N    0.086878  0.035927  0.029731    ...     0.034242  0.029806  0.029603   \n",
       "Q    0.031873  0.039290  0.039008    ...     0.032933  0.034221  0.031862   \n",
       "C    0.000886  0.005807  0.008987    ...     0.010042  0.009365  0.014775   \n",
       "G    0.121145  0.076488  0.067856    ...     0.086523  0.086197  0.086891   \n",
       "P    0.018141  0.040371  0.043680    ...     0.047526  0.055465  0.055819   \n",
       "A    0.123770  0.096001  0.102826    ...     0.107452  0.112161  0.113339   \n",
       "V    0.075356  0.077963  0.071196    ...     0.076862  0.076834  0.067761   \n",
       "I    0.053372  0.069066  0.055539    ...     0.060232  0.047883  0.049720   \n",
       "L    0.057991  0.115628  0.111218    ...     0.099267  0.102823  0.097777   \n",
       "M    0.009123  0.023281  0.019603    ...     0.022836  0.018842  0.017939   \n",
       "F    0.019136  0.043407  0.032195    ...     0.039266  0.036324  0.039421   \n",
       "W    0.003984  0.010587  0.009350    ...     0.011583  0.013376  0.012028   \n",
       "Y    0.020777  0.027499  0.026286    ...     0.029314  0.025952  0.032876   \n",
       "\n",
       "COG         Z         N         S         I         O         L         C  \n",
       "R    0.038073  0.048114  0.057326  0.058245  0.055744  0.072145  0.053945  \n",
       "K    0.024941  0.043528  0.042559  0.040594  0.050442  0.054888  0.043665  \n",
       "H    0.018304  0.012867  0.020347  0.022591  0.019082  0.023296  0.022773  \n",
       "D    0.065583  0.055966  0.056882  0.055326  0.059719  0.058992  0.054205  \n",
       "E    0.032770  0.062060  0.059380  0.059791  0.068536  0.072061  0.060812  \n",
       "S    0.077906  0.072762  0.060552  0.053328  0.056767  0.055060  0.052344  \n",
       "T    0.093320  0.063347  0.055863  0.051632  0.053330  0.050694  0.053203  \n",
       "N    0.047310  0.048153  0.037233  0.030921  0.035960  0.034288  0.032743  \n",
       "Q    0.035419  0.048359  0.037447  0.031603  0.039183  0.040775  0.031852  \n",
       "C    0.017302  0.004298  0.008674  0.010478  0.008513  0.009852  0.013411  \n",
       "G    0.127693  0.076550  0.078021  0.087628  0.081730  0.068621  0.087921  \n",
       "P    0.047803  0.041033  0.048415  0.047893  0.045876  0.045178  0.050002  \n",
       "A    0.104377  0.100950  0.102413  0.117324  0.100913  0.093499  0.107360  \n",
       "V    0.078572  0.073410  0.071264  0.076977  0.074410  0.065649  0.077212  \n",
       "I    0.037574  0.062776  0.054828  0.056539  0.057660  0.056942  0.058490  \n",
       "L    0.071864  0.098610  0.103231  0.097264  0.097813  0.104130  0.095447  \n",
       "M    0.009328  0.024128  0.018683  0.023620  0.019820  0.018123  0.023999  \n",
       "F    0.023181  0.033774  0.040366  0.038312  0.036921  0.035715  0.038978  \n",
       "W    0.024199  0.006571  0.015364  0.013529  0.010202  0.011213  0.013044  \n",
       "Y    0.024481  0.022744  0.031152  0.026404  0.027379  0.028880  0.028593  \n",
       "\n",
       "[20 rows x 23 columns]"
      ]
     },
     "execution_count": 307,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.T / countTot\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.069414</td>\n",
       "      <td>0.043368</td>\n",
       "      <td>0.033126</td>\n",
       "      <td>0.057971</td>\n",
       "      <td>0.073661</td>\n",
       "      <td>0.056117</td>\n",
       "      <td>0.053063</td>\n",
       "      <td>0.029231</td>\n",
       "      <td>0.036303</td>\n",
       "      <td>0.010797</td>\n",
       "      <td>0.062863</td>\n",
       "      <td>0.047223</td>\n",
       "      <td>0.099886</td>\n",
       "      <td>0.073187</td>\n",
       "      <td>0.055211</td>\n",
       "      <td>0.104329</td>\n",
       "      <td>0.024036</td>\n",
       "      <td>0.030725</td>\n",
       "      <td>0.013117</td>\n",
       "      <td>0.026372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.061855</td>\n",
       "      <td>0.030425</td>\n",
       "      <td>0.038903</td>\n",
       "      <td>0.068269</td>\n",
       "      <td>0.059340</td>\n",
       "      <td>0.046553</td>\n",
       "      <td>0.044844</td>\n",
       "      <td>0.029603</td>\n",
       "      <td>0.031862</td>\n",
       "      <td>0.014775</td>\n",
       "      <td>0.086891</td>\n",
       "      <td>0.055819</td>\n",
       "      <td>0.113339</td>\n",
       "      <td>0.067761</td>\n",
       "      <td>0.049720</td>\n",
       "      <td>0.097777</td>\n",
       "      <td>0.017939</td>\n",
       "      <td>0.039421</td>\n",
       "      <td>0.012028</td>\n",
       "      <td>0.032876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.053945</td>\n",
       "      <td>0.043665</td>\n",
       "      <td>0.022773</td>\n",
       "      <td>0.054205</td>\n",
       "      <td>0.060812</td>\n",
       "      <td>0.052344</td>\n",
       "      <td>0.053203</td>\n",
       "      <td>0.032743</td>\n",
       "      <td>0.031852</td>\n",
       "      <td>0.013411</td>\n",
       "      <td>0.087921</td>\n",
       "      <td>0.050002</td>\n",
       "      <td>0.107360</td>\n",
       "      <td>0.077212</td>\n",
       "      <td>0.058490</td>\n",
       "      <td>0.095447</td>\n",
       "      <td>0.023999</td>\n",
       "      <td>0.038978</td>\n",
       "      <td>0.013044</td>\n",
       "      <td>0.028593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.061425</td>\n",
       "      <td>0.046376</td>\n",
       "      <td>0.017580</td>\n",
       "      <td>0.054787</td>\n",
       "      <td>0.070385</td>\n",
       "      <td>0.058835</td>\n",
       "      <td>0.051646</td>\n",
       "      <td>0.034087</td>\n",
       "      <td>0.042502</td>\n",
       "      <td>0.007620</td>\n",
       "      <td>0.077000</td>\n",
       "      <td>0.045629</td>\n",
       "      <td>0.100550</td>\n",
       "      <td>0.075106</td>\n",
       "      <td>0.062956</td>\n",
       "      <td>0.107190</td>\n",
       "      <td>0.022023</td>\n",
       "      <td>0.032255</td>\n",
       "      <td>0.008692</td>\n",
       "      <td>0.023354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.053374</td>\n",
       "      <td>0.041827</td>\n",
       "      <td>0.022572</td>\n",
       "      <td>0.055184</td>\n",
       "      <td>0.060166</td>\n",
       "      <td>0.055238</td>\n",
       "      <td>0.053560</td>\n",
       "      <td>0.034242</td>\n",
       "      <td>0.032933</td>\n",
       "      <td>0.010042</td>\n",
       "      <td>0.086523</td>\n",
       "      <td>0.047526</td>\n",
       "      <td>0.107452</td>\n",
       "      <td>0.076862</td>\n",
       "      <td>0.060232</td>\n",
       "      <td>0.099267</td>\n",
       "      <td>0.022836</td>\n",
       "      <td>0.039266</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>0.029314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.055864</td>\n",
       "      <td>0.046316</td>\n",
       "      <td>0.023709</td>\n",
       "      <td>0.060294</td>\n",
       "      <td>0.067486</td>\n",
       "      <td>0.052802</td>\n",
       "      <td>0.053192</td>\n",
       "      <td>0.035194</td>\n",
       "      <td>0.032779</td>\n",
       "      <td>0.011288</td>\n",
       "      <td>0.082159</td>\n",
       "      <td>0.045256</td>\n",
       "      <td>0.101230</td>\n",
       "      <td>0.078070</td>\n",
       "      <td>0.062949</td>\n",
       "      <td>0.095161</td>\n",
       "      <td>0.021972</td>\n",
       "      <td>0.036912</td>\n",
       "      <td>0.008775</td>\n",
       "      <td>0.028591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0.048469</td>\n",
       "      <td>0.039277</td>\n",
       "      <td>0.020770</td>\n",
       "      <td>0.054014</td>\n",
       "      <td>0.052796</td>\n",
       "      <td>0.057077</td>\n",
       "      <td>0.055228</td>\n",
       "      <td>0.035660</td>\n",
       "      <td>0.031507</td>\n",
       "      <td>0.008685</td>\n",
       "      <td>0.089594</td>\n",
       "      <td>0.047782</td>\n",
       "      <td>0.106729</td>\n",
       "      <td>0.075614</td>\n",
       "      <td>0.059063</td>\n",
       "      <td>0.102179</td>\n",
       "      <td>0.023170</td>\n",
       "      <td>0.043738</td>\n",
       "      <td>0.017085</td>\n",
       "      <td>0.031563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>0.061780</td>\n",
       "      <td>0.040588</td>\n",
       "      <td>0.024356</td>\n",
       "      <td>0.058710</td>\n",
       "      <td>0.062971</td>\n",
       "      <td>0.051354</td>\n",
       "      <td>0.051452</td>\n",
       "      <td>0.030493</td>\n",
       "      <td>0.034374</td>\n",
       "      <td>0.012119</td>\n",
       "      <td>0.081629</td>\n",
       "      <td>0.048859</td>\n",
       "      <td>0.109120</td>\n",
       "      <td>0.076442</td>\n",
       "      <td>0.058034</td>\n",
       "      <td>0.104692</td>\n",
       "      <td>0.021050</td>\n",
       "      <td>0.034971</td>\n",
       "      <td>0.011299</td>\n",
       "      <td>0.025709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>0.058245</td>\n",
       "      <td>0.040594</td>\n",
       "      <td>0.022591</td>\n",
       "      <td>0.055326</td>\n",
       "      <td>0.059791</td>\n",
       "      <td>0.053328</td>\n",
       "      <td>0.051632</td>\n",
       "      <td>0.030921</td>\n",
       "      <td>0.031603</td>\n",
       "      <td>0.010478</td>\n",
       "      <td>0.087628</td>\n",
       "      <td>0.047893</td>\n",
       "      <td>0.117324</td>\n",
       "      <td>0.076977</td>\n",
       "      <td>0.056539</td>\n",
       "      <td>0.097264</td>\n",
       "      <td>0.023620</td>\n",
       "      <td>0.038312</td>\n",
       "      <td>0.013529</td>\n",
       "      <td>0.026404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>J</th>\n",
       "      <td>0.066453</td>\n",
       "      <td>0.060137</td>\n",
       "      <td>0.022838</td>\n",
       "      <td>0.060528</td>\n",
       "      <td>0.074446</td>\n",
       "      <td>0.049583</td>\n",
       "      <td>0.051220</td>\n",
       "      <td>0.034815</td>\n",
       "      <td>0.035426</td>\n",
       "      <td>0.009442</td>\n",
       "      <td>0.076987</td>\n",
       "      <td>0.045671</td>\n",
       "      <td>0.091890</td>\n",
       "      <td>0.076266</td>\n",
       "      <td>0.058745</td>\n",
       "      <td>0.090609</td>\n",
       "      <td>0.020745</td>\n",
       "      <td>0.035787</td>\n",
       "      <td>0.010400</td>\n",
       "      <td>0.028012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>0.076650</td>\n",
       "      <td>0.042045</td>\n",
       "      <td>0.023700</td>\n",
       "      <td>0.057607</td>\n",
       "      <td>0.074549</td>\n",
       "      <td>0.057485</td>\n",
       "      <td>0.050489</td>\n",
       "      <td>0.029731</td>\n",
       "      <td>0.039008</td>\n",
       "      <td>0.008987</td>\n",
       "      <td>0.067856</td>\n",
       "      <td>0.043680</td>\n",
       "      <td>0.102826</td>\n",
       "      <td>0.071196</td>\n",
       "      <td>0.055539</td>\n",
       "      <td>0.111218</td>\n",
       "      <td>0.019603</td>\n",
       "      <td>0.032195</td>\n",
       "      <td>0.009350</td>\n",
       "      <td>0.026286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>0.072145</td>\n",
       "      <td>0.054888</td>\n",
       "      <td>0.023296</td>\n",
       "      <td>0.058992</td>\n",
       "      <td>0.072061</td>\n",
       "      <td>0.055060</td>\n",
       "      <td>0.050694</td>\n",
       "      <td>0.034288</td>\n",
       "      <td>0.040775</td>\n",
       "      <td>0.009852</td>\n",
       "      <td>0.068621</td>\n",
       "      <td>0.045178</td>\n",
       "      <td>0.093499</td>\n",
       "      <td>0.065649</td>\n",
       "      <td>0.056942</td>\n",
       "      <td>0.104130</td>\n",
       "      <td>0.018123</td>\n",
       "      <td>0.035715</td>\n",
       "      <td>0.011213</td>\n",
       "      <td>0.028880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.057463</td>\n",
       "      <td>0.046297</td>\n",
       "      <td>0.019702</td>\n",
       "      <td>0.055921</td>\n",
       "      <td>0.055949</td>\n",
       "      <td>0.060622</td>\n",
       "      <td>0.055383</td>\n",
       "      <td>0.040271</td>\n",
       "      <td>0.038073</td>\n",
       "      <td>0.007588</td>\n",
       "      <td>0.079133</td>\n",
       "      <td>0.047127</td>\n",
       "      <td>0.099208</td>\n",
       "      <td>0.074547</td>\n",
       "      <td>0.058696</td>\n",
       "      <td>0.099145</td>\n",
       "      <td>0.018492</td>\n",
       "      <td>0.039070</td>\n",
       "      <td>0.013069</td>\n",
       "      <td>0.034242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.048114</td>\n",
       "      <td>0.043528</td>\n",
       "      <td>0.012867</td>\n",
       "      <td>0.055966</td>\n",
       "      <td>0.062060</td>\n",
       "      <td>0.072762</td>\n",
       "      <td>0.063347</td>\n",
       "      <td>0.048153</td>\n",
       "      <td>0.048359</td>\n",
       "      <td>0.004298</td>\n",
       "      <td>0.076550</td>\n",
       "      <td>0.041033</td>\n",
       "      <td>0.100950</td>\n",
       "      <td>0.073410</td>\n",
       "      <td>0.062776</td>\n",
       "      <td>0.098610</td>\n",
       "      <td>0.024128</td>\n",
       "      <td>0.033774</td>\n",
       "      <td>0.006571</td>\n",
       "      <td>0.022744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <td>0.055744</td>\n",
       "      <td>0.050442</td>\n",
       "      <td>0.019082</td>\n",
       "      <td>0.059719</td>\n",
       "      <td>0.068536</td>\n",
       "      <td>0.056767</td>\n",
       "      <td>0.053330</td>\n",
       "      <td>0.035960</td>\n",
       "      <td>0.039183</td>\n",
       "      <td>0.008513</td>\n",
       "      <td>0.081730</td>\n",
       "      <td>0.045876</td>\n",
       "      <td>0.100913</td>\n",
       "      <td>0.074410</td>\n",
       "      <td>0.057660</td>\n",
       "      <td>0.097813</td>\n",
       "      <td>0.019820</td>\n",
       "      <td>0.036921</td>\n",
       "      <td>0.010202</td>\n",
       "      <td>0.027379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>0.049685</td>\n",
       "      <td>0.036304</td>\n",
       "      <td>0.016904</td>\n",
       "      <td>0.048341</td>\n",
       "      <td>0.048252</td>\n",
       "      <td>0.061714</td>\n",
       "      <td>0.057594</td>\n",
       "      <td>0.035710</td>\n",
       "      <td>0.033005</td>\n",
       "      <td>0.005930</td>\n",
       "      <td>0.084712</td>\n",
       "      <td>0.044532</td>\n",
       "      <td>0.102861</td>\n",
       "      <td>0.081158</td>\n",
       "      <td>0.065884</td>\n",
       "      <td>0.115078</td>\n",
       "      <td>0.023592</td>\n",
       "      <td>0.045792</td>\n",
       "      <td>0.014405</td>\n",
       "      <td>0.028547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>0.062814</td>\n",
       "      <td>0.029094</td>\n",
       "      <td>0.024923</td>\n",
       "      <td>0.062216</td>\n",
       "      <td>0.059338</td>\n",
       "      <td>0.054989</td>\n",
       "      <td>0.057377</td>\n",
       "      <td>0.029806</td>\n",
       "      <td>0.034221</td>\n",
       "      <td>0.009365</td>\n",
       "      <td>0.086197</td>\n",
       "      <td>0.055465</td>\n",
       "      <td>0.112161</td>\n",
       "      <td>0.076834</td>\n",
       "      <td>0.047883</td>\n",
       "      <td>0.102823</td>\n",
       "      <td>0.018842</td>\n",
       "      <td>0.036324</td>\n",
       "      <td>0.013376</td>\n",
       "      <td>0.025952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.057326</td>\n",
       "      <td>0.042559</td>\n",
       "      <td>0.020347</td>\n",
       "      <td>0.056882</td>\n",
       "      <td>0.059380</td>\n",
       "      <td>0.060552</td>\n",
       "      <td>0.055863</td>\n",
       "      <td>0.037233</td>\n",
       "      <td>0.037447</td>\n",
       "      <td>0.008674</td>\n",
       "      <td>0.078021</td>\n",
       "      <td>0.048415</td>\n",
       "      <td>0.102413</td>\n",
       "      <td>0.071264</td>\n",
       "      <td>0.054828</td>\n",
       "      <td>0.103231</td>\n",
       "      <td>0.018683</td>\n",
       "      <td>0.040366</td>\n",
       "      <td>0.015364</td>\n",
       "      <td>0.031152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>0.067162</td>\n",
       "      <td>0.040656</td>\n",
       "      <td>0.021835</td>\n",
       "      <td>0.058118</td>\n",
       "      <td>0.069882</td>\n",
       "      <td>0.059264</td>\n",
       "      <td>0.051269</td>\n",
       "      <td>0.034287</td>\n",
       "      <td>0.042043</td>\n",
       "      <td>0.007823</td>\n",
       "      <td>0.068853</td>\n",
       "      <td>0.041772</td>\n",
       "      <td>0.098746</td>\n",
       "      <td>0.072089</td>\n",
       "      <td>0.061120</td>\n",
       "      <td>0.115061</td>\n",
       "      <td>0.021004</td>\n",
       "      <td>0.034737</td>\n",
       "      <td>0.009649</td>\n",
       "      <td>0.024631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U</th>\n",
       "      <td>0.053370</td>\n",
       "      <td>0.045387</td>\n",
       "      <td>0.013736</td>\n",
       "      <td>0.053614</td>\n",
       "      <td>0.057041</td>\n",
       "      <td>0.065849</td>\n",
       "      <td>0.060635</td>\n",
       "      <td>0.041664</td>\n",
       "      <td>0.043274</td>\n",
       "      <td>0.004852</td>\n",
       "      <td>0.082830</td>\n",
       "      <td>0.042692</td>\n",
       "      <td>0.099040</td>\n",
       "      <td>0.075818</td>\n",
       "      <td>0.064632</td>\n",
       "      <td>0.102231</td>\n",
       "      <td>0.022080</td>\n",
       "      <td>0.036806</td>\n",
       "      <td>0.009122</td>\n",
       "      <td>0.025328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>0.056161</td>\n",
       "      <td>0.043001</td>\n",
       "      <td>0.016554</td>\n",
       "      <td>0.050419</td>\n",
       "      <td>0.053979</td>\n",
       "      <td>0.061900</td>\n",
       "      <td>0.056669</td>\n",
       "      <td>0.035927</td>\n",
       "      <td>0.039290</td>\n",
       "      <td>0.005807</td>\n",
       "      <td>0.076488</td>\n",
       "      <td>0.040371</td>\n",
       "      <td>0.096001</td>\n",
       "      <td>0.077963</td>\n",
       "      <td>0.069066</td>\n",
       "      <td>0.115628</td>\n",
       "      <td>0.023281</td>\n",
       "      <td>0.043407</td>\n",
       "      <td>0.010587</td>\n",
       "      <td>0.027499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>0.019944</td>\n",
       "      <td>0.058713</td>\n",
       "      <td>0.008507</td>\n",
       "      <td>0.065350</td>\n",
       "      <td>0.040990</td>\n",
       "      <td>0.088327</td>\n",
       "      <td>0.095736</td>\n",
       "      <td>0.086878</td>\n",
       "      <td>0.031873</td>\n",
       "      <td>0.000886</td>\n",
       "      <td>0.121145</td>\n",
       "      <td>0.018141</td>\n",
       "      <td>0.123770</td>\n",
       "      <td>0.075356</td>\n",
       "      <td>0.053372</td>\n",
       "      <td>0.057991</td>\n",
       "      <td>0.009123</td>\n",
       "      <td>0.019136</td>\n",
       "      <td>0.003984</td>\n",
       "      <td>0.020777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z</th>\n",
       "      <td>0.038073</td>\n",
       "      <td>0.024941</td>\n",
       "      <td>0.018304</td>\n",
       "      <td>0.065583</td>\n",
       "      <td>0.032770</td>\n",
       "      <td>0.077906</td>\n",
       "      <td>0.093320</td>\n",
       "      <td>0.047310</td>\n",
       "      <td>0.035419</td>\n",
       "      <td>0.017302</td>\n",
       "      <td>0.127693</td>\n",
       "      <td>0.047803</td>\n",
       "      <td>0.104377</td>\n",
       "      <td>0.078572</td>\n",
       "      <td>0.037574</td>\n",
       "      <td>0.071864</td>\n",
       "      <td>0.009328</td>\n",
       "      <td>0.023181</td>\n",
       "      <td>0.024199</td>\n",
       "      <td>0.024481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            R         K         H         D         E         S         T  \\\n",
       "COG                                                                         \n",
       "A    0.069414  0.043368  0.033126  0.057971  0.073661  0.056117  0.053063   \n",
       "B    0.061855  0.030425  0.038903  0.068269  0.059340  0.046553  0.044844   \n",
       "C    0.053945  0.043665  0.022773  0.054205  0.060812  0.052344  0.053203   \n",
       "D    0.061425  0.046376  0.017580  0.054787  0.070385  0.058835  0.051646   \n",
       "E    0.053374  0.041827  0.022572  0.055184  0.060166  0.055238  0.053560   \n",
       "F    0.055864  0.046316  0.023709  0.060294  0.067486  0.052802  0.053192   \n",
       "G    0.048469  0.039277  0.020770  0.054014  0.052796  0.057077  0.055228   \n",
       "H    0.061780  0.040588  0.024356  0.058710  0.062971  0.051354  0.051452   \n",
       "I    0.058245  0.040594  0.022591  0.055326  0.059791  0.053328  0.051632   \n",
       "J    0.066453  0.060137  0.022838  0.060528  0.074446  0.049583  0.051220   \n",
       "K    0.076650  0.042045  0.023700  0.057607  0.074549  0.057485  0.050489   \n",
       "L    0.072145  0.054888  0.023296  0.058992  0.072061  0.055060  0.050694   \n",
       "M    0.057463  0.046297  0.019702  0.055921  0.055949  0.060622  0.055383   \n",
       "N    0.048114  0.043528  0.012867  0.055966  0.062060  0.072762  0.063347   \n",
       "O    0.055744  0.050442  0.019082  0.059719  0.068536  0.056767  0.053330   \n",
       "P    0.049685  0.036304  0.016904  0.048341  0.048252  0.061714  0.057594   \n",
       "Q    0.062814  0.029094  0.024923  0.062216  0.059338  0.054989  0.057377   \n",
       "S    0.057326  0.042559  0.020347  0.056882  0.059380  0.060552  0.055863   \n",
       "T    0.067162  0.040656  0.021835  0.058118  0.069882  0.059264  0.051269   \n",
       "U    0.053370  0.045387  0.013736  0.053614  0.057041  0.065849  0.060635   \n",
       "V    0.056161  0.043001  0.016554  0.050419  0.053979  0.061900  0.056669   \n",
       "W    0.019944  0.058713  0.008507  0.065350  0.040990  0.088327  0.095736   \n",
       "Z    0.038073  0.024941  0.018304  0.065583  0.032770  0.077906  0.093320   \n",
       "\n",
       "            N         Q         C         G         P         A         V  \\\n",
       "COG                                                                         \n",
       "A    0.029231  0.036303  0.010797  0.062863  0.047223  0.099886  0.073187   \n",
       "B    0.029603  0.031862  0.014775  0.086891  0.055819  0.113339  0.067761   \n",
       "C    0.032743  0.031852  0.013411  0.087921  0.050002  0.107360  0.077212   \n",
       "D    0.034087  0.042502  0.007620  0.077000  0.045629  0.100550  0.075106   \n",
       "E    0.034242  0.032933  0.010042  0.086523  0.047526  0.107452  0.076862   \n",
       "F    0.035194  0.032779  0.011288  0.082159  0.045256  0.101230  0.078070   \n",
       "G    0.035660  0.031507  0.008685  0.089594  0.047782  0.106729  0.075614   \n",
       "H    0.030493  0.034374  0.012119  0.081629  0.048859  0.109120  0.076442   \n",
       "I    0.030921  0.031603  0.010478  0.087628  0.047893  0.117324  0.076977   \n",
       "J    0.034815  0.035426  0.009442  0.076987  0.045671  0.091890  0.076266   \n",
       "K    0.029731  0.039008  0.008987  0.067856  0.043680  0.102826  0.071196   \n",
       "L    0.034288  0.040775  0.009852  0.068621  0.045178  0.093499  0.065649   \n",
       "M    0.040271  0.038073  0.007588  0.079133  0.047127  0.099208  0.074547   \n",
       "N    0.048153  0.048359  0.004298  0.076550  0.041033  0.100950  0.073410   \n",
       "O    0.035960  0.039183  0.008513  0.081730  0.045876  0.100913  0.074410   \n",
       "P    0.035710  0.033005  0.005930  0.084712  0.044532  0.102861  0.081158   \n",
       "Q    0.029806  0.034221  0.009365  0.086197  0.055465  0.112161  0.076834   \n",
       "S    0.037233  0.037447  0.008674  0.078021  0.048415  0.102413  0.071264   \n",
       "T    0.034287  0.042043  0.007823  0.068853  0.041772  0.098746  0.072089   \n",
       "U    0.041664  0.043274  0.004852  0.082830  0.042692  0.099040  0.075818   \n",
       "V    0.035927  0.039290  0.005807  0.076488  0.040371  0.096001  0.077963   \n",
       "W    0.086878  0.031873  0.000886  0.121145  0.018141  0.123770  0.075356   \n",
       "Z    0.047310  0.035419  0.017302  0.127693  0.047803  0.104377  0.078572   \n",
       "\n",
       "            I         L         M         F         W         Y  \n",
       "COG                                                              \n",
       "A    0.055211  0.104329  0.024036  0.030725  0.013117  0.026372  \n",
       "B    0.049720  0.097777  0.017939  0.039421  0.012028  0.032876  \n",
       "C    0.058490  0.095447  0.023999  0.038978  0.013044  0.028593  \n",
       "D    0.062956  0.107190  0.022023  0.032255  0.008692  0.023354  \n",
       "E    0.060232  0.099267  0.022836  0.039266  0.011583  0.029314  \n",
       "F    0.062949  0.095161  0.021972  0.036912  0.008775  0.028591  \n",
       "G    0.059063  0.102179  0.023170  0.043738  0.017085  0.031563  \n",
       "H    0.058034  0.104692  0.021050  0.034971  0.011299  0.025709  \n",
       "I    0.056539  0.097264  0.023620  0.038312  0.013529  0.026404  \n",
       "J    0.058745  0.090609  0.020745  0.035787  0.010400  0.028012  \n",
       "K    0.055539  0.111218  0.019603  0.032195  0.009350  0.026286  \n",
       "L    0.056942  0.104130  0.018123  0.035715  0.011213  0.028880  \n",
       "M    0.058696  0.099145  0.018492  0.039070  0.013069  0.034242  \n",
       "N    0.062776  0.098610  0.024128  0.033774  0.006571  0.022744  \n",
       "O    0.057660  0.097813  0.019820  0.036921  0.010202  0.027379  \n",
       "P    0.065884  0.115078  0.023592  0.045792  0.014405  0.028547  \n",
       "Q    0.047883  0.102823  0.018842  0.036324  0.013376  0.025952  \n",
       "S    0.054828  0.103231  0.018683  0.040366  0.015364  0.031152  \n",
       "T    0.061120  0.115061  0.021004  0.034737  0.009649  0.024631  \n",
       "U    0.064632  0.102231  0.022080  0.036806  0.009122  0.025328  \n",
       "V    0.069066  0.115628  0.023281  0.043407  0.010587  0.027499  \n",
       "W    0.053372  0.057991  0.009123  0.019136  0.003984  0.020777  \n",
       "Z    0.037574  0.071864  0.009328  0.023181  0.024199  0.024481  "
      ]
     },
     "execution_count": 308,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.sort_index(axis=1)\n",
    "df = df.loc[aaTable]\n",
    "df = df.T\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df / aaFreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = df\n",
    "\n",
    "set_fontsize_paper()\n",
    "width = 3\n",
    "\n",
    "linewidth1 = 0.5\n",
    "aspectratio = 0.7\n",
    "figsizeCoeff = 0.5\n",
    "figsize = (width, (width/aspectratio))\n",
    "\n",
    "fig, (ax,cbar_ax) = plt.subplots(1, 2, figsize=figsize)\n",
    "# main axes\n",
    "ax.set_position([0, 0, 1, 1])\n",
    "ax.set_facecolor(colorBackground1)\n",
    "\n",
    "# Create axes for the colorbar for more precise positioning\n",
    "# vertical bar\n",
    "cbar_aspect = 10\n",
    "cbar_length = 0.3\n",
    "cbar_ax.set_position([0.2, -0.3 - cbar_length, cbar_length/cbar_aspect, cbar_length])\n",
    "\n",
    "cbarLabel = 'normalized AA freq. in bulk'\n",
    "ax = seaborn.heatmap(plotData, square=True, ax=ax, cmap=cmap,\n",
    "                     cbar_kws={\"label\":cbarLabel}, cbar_ax=cbar_ax,\n",
    "                     xticklabels=True, yticklabels=True,\n",
    "                     linewidth=linewidth1, linecolor=colorHeatmapLine\n",
    "                     )\n",
    "for _, spine in ax.spines.items():\n",
    "    spine.set_visible(True)\n",
    "    # spine.set_color('0.8')\n",
    "    spine.set_linewidth(0.4)\n",
    "tickLength = FontProperties(size='small').get_size()/4\n",
    "cbar_ax.tick_params(axis='y', length=tickLength, color=colorAxis)\n",
    "ax.tick_params(axis='y', labelleft='on', labelright='on')\n",
    "if len(ax.xaxis.get_ticklabels()[0].get_text()) > 2:\n",
    "    ax.xaxis.set_tick_params(rotation=90)\n",
    "ax.yaxis.set_tick_params(rotation=0)\n",
    "family = 'Liberation Mono'\n",
    "ticks_font = FontProperties(family=family, size='small')\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontproperties(ticks_font)\n",
    "    tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "\n",
    "ax.set_xlabel('amino acid at C-terminal')\n",
    "ax.set_ylabel('COG category')\n",
    "\n",
    "filenamePlot = analysisEggnogPath / ('taxaId_2_subset_all_oddsRatioAADf_COG_groups_heatmap_bulk_aa_freq')\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + \".png\", dpi=500, bbox_inches=\"tight\")\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "COG\n",
       "J    0.060137\n",
       "U    0.045387\n",
       "P    0.036304\n",
       "T    0.040656\n",
       "A    0.043368\n",
       "M    0.046297\n",
       "F    0.046316\n",
       "W    0.058713\n",
       "V    0.043001\n",
       "K    0.042045\n",
       "G    0.039277\n",
       "D    0.046376\n",
       "H    0.040588\n",
       "E    0.041827\n",
       "Q    0.029094\n",
       "B    0.030425\n",
       "Z    0.024941\n",
       "N    0.043528\n",
       "S    0.042559\n",
       "I    0.040594\n",
       "O    0.050442\n",
       "L    0.054888\n",
       "C    0.043665\n",
       "Name: K, dtype: float64"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc['K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9830513927486966"
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = df.loc['K']\n",
    "s[s.index != 'J'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "observable\n",
       "BH_multiple_tests                  True\n",
       "count                             39369\n",
       "log2OddsRatio        1.5971686771787748\n",
       "oddsRatio             3.025489708430977\n",
       "pvalue                              0.0\n",
       "Name: K, dtype: object"
      ]
     },
     "execution_count": 311,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = oddsRatioDfCOG.loc[idx['C', -1], aaTable]\n",
    "df = df.reorder_levels([1,0]).sort_index()\n",
    "df.loc['J', 'K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>21072</td>\n",
       "      <td>39369</td>\n",
       "      <td>2793</td>\n",
       "      <td>10626</td>\n",
       "      <td>19862</td>\n",
       "      <td>13572</td>\n",
       "      <td>3387</td>\n",
       "      <td>8834</td>\n",
       "      <td>8807</td>\n",
       "      <td>1226</td>\n",
       "      <td>19215</td>\n",
       "      <td>4843</td>\n",
       "      <td>29756</td>\n",
       "      <td>14233</td>\n",
       "      <td>8240</td>\n",
       "      <td>16462</td>\n",
       "      <td>3221</td>\n",
       "      <td>7559</td>\n",
       "      <td>4719</td>\n",
       "      <td>4942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.4173329187219579</td>\n",
       "      <td>1.5971686771787748</td>\n",
       "      <td>-1.0056543942787919</td>\n",
       "      <td>-0.4929854492675295</td>\n",
       "      <td>0.1478992300355653</td>\n",
       "      <td>0.18296527416323338</td>\n",
       "      <td>-1.9316873209172318</td>\n",
       "      <td>0.06631012502722017</td>\n",
       "      <td>0.03573513727743984</td>\n",
       "      <td>-0.9090323498970772</td>\n",
       "      <td>0.04355566556826341</td>\n",
       "      <td>-1.233136196314422</td>\n",
       "      <td>0.46541104215658496</td>\n",
       "      <td>-0.4065513145407548</td>\n",
       "      <td>-0.8287376193890541</td>\n",
       "      <td>-0.45370190573855435</td>\n",
       "      <td>-0.655602036577335</td>\n",
       "      <td>-0.2075731945220512</td>\n",
       "      <td>0.9157099036931593</td>\n",
       "      <td>-0.47168339542708826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.335456439904414</td>\n",
       "      <td>3.025489708430977</td>\n",
       "      <td>0.4980441715436644</td>\n",
       "      <td>0.7105531880513064</td>\n",
       "      <td>1.1079549562626105</td>\n",
       "      <td>1.1352147770660113</td>\n",
       "      <td>0.26212242320409285</td>\n",
       "      <td>1.0470353308957898</td>\n",
       "      <td>1.0250790275332748</td>\n",
       "      <td>0.5325421602421835</td>\n",
       "      <td>1.0306508406221837</td>\n",
       "      <td>0.4253917041403423</td>\n",
       "      <td>1.3807106793862334</td>\n",
       "      <td>0.7544246317401716</td>\n",
       "      <td>0.5630216795988365</td>\n",
       "      <td>0.7301668591524552</td>\n",
       "      <td>0.6348105295828163</td>\n",
       "      <td>0.865992722059996</td>\n",
       "      <td>1.8864971278691423</td>\n",
       "      <td>0.7211226713093073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.483360671559279e-287</td>\n",
       "      <td>3.1284780164297066e-42</td>\n",
       "      <td>6.74784029502009e-45</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.6104264269772995e-05</td>\n",
       "      <td>0.02308253745883346</td>\n",
       "      <td>2.1771996200802676e-132</td>\n",
       "      <td>6.595386157817943e-05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.6011649535014696e-252</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.7586272233067254e-167</td>\n",
       "      <td>3.4768348486210835e-36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0355789940084485e-126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    R                   K  \\\n",
       "observable                                                  \n",
       "BH_multiple_tests                True                True   \n",
       "count                           21072               39369   \n",
       "log2OddsRatio      0.4173329187219579  1.5971686771787748   \n",
       "oddsRatio           1.335456439904414   3.025489708430977   \n",
       "pvalue                            0.0                 0.0   \n",
       "\n",
       "                                     H                       D  \\\n",
       "observable                                                       \n",
       "BH_multiple_tests                 True                    True   \n",
       "count                             2793                   10626   \n",
       "log2OddsRatio      -1.0056543942787919     -0.4929854492675295   \n",
       "oddsRatio           0.4980441715436644      0.7105531880513064   \n",
       "pvalue                             0.0  1.483360671559279e-287   \n",
       "\n",
       "                                        E                     S  \\\n",
       "observable                                                        \n",
       "BH_multiple_tests                    True                  True   \n",
       "count                               19862                 13572   \n",
       "log2OddsRatio          0.1478992300355653   0.18296527416323338   \n",
       "oddsRatio              1.1079549562626105    1.1352147770660113   \n",
       "pvalue             3.1284780164297066e-42  6.74784029502009e-45   \n",
       "\n",
       "                                     T                       N  \\\n",
       "observable                                                       \n",
       "BH_multiple_tests                 True                    True   \n",
       "count                             3387                    8834   \n",
       "log2OddsRatio      -1.9316873209172318     0.06631012502722017   \n",
       "oddsRatio          0.26212242320409285      1.0470353308957898   \n",
       "pvalue                             0.0  2.6104264269772995e-05   \n",
       "\n",
       "                                     Q                        C  \\\n",
       "observable                                                        \n",
       "BH_multiple_tests                 True                     True   \n",
       "count                             8807                     1226   \n",
       "log2OddsRatio      0.03573513727743984      -0.9090323498970772   \n",
       "oddsRatio           1.0250790275332748       0.5325421602421835   \n",
       "pvalue             0.02308253745883346  2.1771996200802676e-132   \n",
       "\n",
       "                                       G                   P  \\\n",
       "observable                                                     \n",
       "BH_multiple_tests                   True                True   \n",
       "count                              19215                4843   \n",
       "log2OddsRatio        0.04355566556826341  -1.233136196314422   \n",
       "oddsRatio             1.0306508406221837  0.4253917041403423   \n",
       "pvalue             6.595386157817943e-05                 0.0   \n",
       "\n",
       "                                     A                        V  \\\n",
       "observable                                                        \n",
       "BH_multiple_tests                 True                     True   \n",
       "count                            29756                    14233   \n",
       "log2OddsRatio      0.46541104215658496      -0.4065513145407548   \n",
       "oddsRatio           1.3807106793862334       0.7544246317401716   \n",
       "pvalue                             0.0  1.6011649535014696e-252   \n",
       "\n",
       "                                     I                     L  \\\n",
       "observable                                                     \n",
       "BH_multiple_tests                 True                  True   \n",
       "count                             8240                 16462   \n",
       "log2OddsRatio      -0.8287376193890541  -0.45370190573855435   \n",
       "oddsRatio           0.5630216795988365    0.7301668591524552   \n",
       "pvalue                             0.0                   0.0   \n",
       "\n",
       "                                         M                       F  \\\n",
       "observable                                                           \n",
       "BH_multiple_tests                     True                    True   \n",
       "count                                 3221                    7559   \n",
       "log2OddsRatio           -0.655602036577335     -0.2075731945220512   \n",
       "oddsRatio               0.6348105295828163       0.865992722059996   \n",
       "pvalue             1.7586272233067254e-167  3.4768348486210835e-36   \n",
       "\n",
       "                                    W                        Y  \n",
       "observable                                                      \n",
       "BH_multiple_tests                True                     True  \n",
       "count                            4719                     4942  \n",
       "log2OddsRatio      0.9157099036931593     -0.47168339542708826  \n",
       "oddsRatio          1.8864971278691423       0.7211226713093073  \n",
       "pvalue                            0.0  1.0355789940084485e-126  "
      ]
     },
     "execution_count": 312,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df.loc['J']\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "242738\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "R    0.086810\n",
       "K    0.162187\n",
       "H    0.011506\n",
       "D    0.043776\n",
       "E    0.081825\n",
       "S    0.055912\n",
       "T    0.013953\n",
       "N    0.036393\n",
       "Q    0.036282\n",
       "C    0.005051\n",
       "G    0.079159\n",
       "P    0.019952\n",
       "A    0.122585\n",
       "V    0.058635\n",
       "I    0.033946\n",
       "L    0.067818\n",
       "M    0.013269\n",
       "F    0.031141\n",
       "W    0.019441\n",
       "Y    0.020359\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 313,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tot = df2.loc['count'].astype(int).sum()\n",
    "print(tot)\n",
    "df2.loc['count'].astype(float) / tot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">A</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>31</td>\n",
       "      <td>44</td>\n",
       "      <td>4</td>\n",
       "      <td>45</td>\n",
       "      <td>55</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "      <td>58</td>\n",
       "      <td>32</td>\n",
       "      <td>77</td>\n",
       "      <td>51</td>\n",
       "      <td>70</td>\n",
       "      <td>108</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>-0.693271051904881</td>\n",
       "      <td>0.5585519267797228</td>\n",
       "      <td>-2.58187599893849</td>\n",
       "      <td>0.15227088164812616</td>\n",
       "      <td>0.09408620111261075</td>\n",
       "      <td>-0.4154815281994051</td>\n",
       "      <td>-1.3618759538736174</td>\n",
       "      <td>0.3510027643641223</td>\n",
       "      <td>-0.030886775495811626</td>\n",
       "      <td>-1.3487013105157886</td>\n",
       "      <td>0.4227917307014406</td>\n",
       "      <td>-0.05559612951644007</td>\n",
       "      <td>0.14852651881926163</td>\n",
       "      <td>-0.013681018395559259</td>\n",
       "      <td>0.9201937281855587</td>\n",
       "      <td>0.6399756962616083</td>\n",
       "      <td>-2.522705564148908</td>\n",
       "      <td>-0.4447574285317583</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.908497334160046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>0.6184500347379803</td>\n",
       "      <td>1.4727901945823296</td>\n",
       "      <td>0.16702361502625615</td>\n",
       "      <td>1.1113173709619817</td>\n",
       "      <td>1.067389112696242</td>\n",
       "      <td>0.749769202473172</td>\n",
       "      <td>0.3890760404623243</td>\n",
       "      <td>1.2754468355741662</td>\n",
       "      <td>0.9788184662757863</td>\n",
       "      <td>0.392645342554571</td>\n",
       "      <td>1.340519058568072</td>\n",
       "      <td>0.9621967759545658</td>\n",
       "      <td>1.1084368036489123</td>\n",
       "      <td>0.9905618621387978</td>\n",
       "      <td>1.8923693878123335</td>\n",
       "      <td>1.5583029078235333</td>\n",
       "      <td>0.1740163098878695</td>\n",
       "      <td>0.7347078383621707</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5327396872766956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.007137213670721523</td>\n",
       "      <td>0.015791826658973416</td>\n",
       "      <td>2.028341221679015e-06</td>\n",
       "      <td>0.4667586992491468</td>\n",
       "      <td>0.6125740828737359</td>\n",
       "      <td>0.13811753074782185</td>\n",
       "      <td>4.305732143214525e-05</td>\n",
       "      <td>0.2164643876832293</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.09896812200450078</td>\n",
       "      <td>0.03549844277465371</td>\n",
       "      <td>0.9288822431478668</td>\n",
       "      <td>0.3777368989010897</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.9477769043096295e-06</td>\n",
       "      <td>5.493094170602914e-05</td>\n",
       "      <td>9.22763654415142e-05</td>\n",
       "      <td>0.271290040827016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.04366404711602853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">B</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>132</td>\n",
       "      <td>122</td>\n",
       "      <td>26</td>\n",
       "      <td>123</td>\n",
       "      <td>53</td>\n",
       "      <td>113</td>\n",
       "      <td>43</td>\n",
       "      <td>29</td>\n",
       "      <td>32</td>\n",
       "      <td>5</td>\n",
       "      <td>151</td>\n",
       "      <td>55</td>\n",
       "      <td>150</td>\n",
       "      <td>52</td>\n",
       "      <td>57</td>\n",
       "      <td>205</td>\n",
       "      <td>11</td>\n",
       "      <td>82</td>\n",
       "      <td>6</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.5599817011709203</td>\n",
       "      <td>1.5069209139683486</td>\n",
       "      <td>-1.1885418114752087</td>\n",
       "      <td>0.29631992894676673</td>\n",
       "      <td>-0.7743384731918935</td>\n",
       "      <td>0.749073302909063</td>\n",
       "      <td>-0.6598443041556618</td>\n",
       "      <td>-0.6199940495152496</td>\n",
       "      <td>-0.5844625005760626</td>\n",
       "      <td>-2.1550706957054917</td>\n",
       "      <td>0.24497270815231342</td>\n",
       "      <td>-0.625227522825236</td>\n",
       "      <td>-0.19145788193205807</td>\n",
       "      <td>-1.0072336237084454</td>\n",
       "      <td>-0.39546819313465836</td>\n",
       "      <td>0.5578435660865309</td>\n",
       "      <td>-1.2963419785405552</td>\n",
       "      <td>0.5049887681728684</td>\n",
       "      <td>-1.590345057686168</td>\n",
       "      <td>-0.19400670601663597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.4742505181006464</td>\n",
       "      <td>2.8420282870472633</td>\n",
       "      <td>0.43874609452831226</td>\n",
       "      <td>1.2280079741276693</td>\n",
       "      <td>0.5846566486213345</td>\n",
       "      <td>1.680712898823263</td>\n",
       "      <td>0.6329466009823453</td>\n",
       "      <td>0.6506736114589072</td>\n",
       "      <td>0.666897755836301</td>\n",
       "      <td>0.22452209080116056</td>\n",
       "      <td>1.1850703524369361</td>\n",
       "      <td>0.648317525177379</td>\n",
       "      <td>0.8757203352132595</td>\n",
       "      <td>0.49749929153304434</td>\n",
       "      <td>0.7602426175158838</td>\n",
       "      <td>1.47206723472285</td>\n",
       "      <td>0.40715725694427124</td>\n",
       "      <td>1.4191123079291144</td>\n",
       "      <td>0.3320920156637018</td>\n",
       "      <td>0.8741745570133416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>5.112321091746052e-05</td>\n",
       "      <td>6.401363726487133e-22</td>\n",
       "      <td>3.0702616474021627e-06</td>\n",
       "      <td>0.03473055753940365</td>\n",
       "      <td>4.4719144417806113e-05</td>\n",
       "      <td>7.313342920223233e-07</td>\n",
       "      <td>0.0020548789250997308</td>\n",
       "      <td>0.02111095268512736</td>\n",
       "      <td>0.0215782120450621</td>\n",
       "      <td>3.230438461596552e-05</td>\n",
       "      <td>0.05288050032066993</td>\n",
       "      <td>0.0010087512411233082</td>\n",
       "      <td>0.12974062907164013</td>\n",
       "      <td>4.964229369662906e-08</td>\n",
       "      <td>0.04208589977126963</td>\n",
       "      <td>9.310525489793519e-07</td>\n",
       "      <td>0.0008315055280774621</td>\n",
       "      <td>0.003265964700197696</td>\n",
       "      <td>0.002565907736147211</td>\n",
       "      <td>0.4233723598626794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">C</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>23106</td>\n",
       "      <td>25113</td>\n",
       "      <td>8818</td>\n",
       "      <td>10685</td>\n",
       "      <td>15793</td>\n",
       "      <td>16764</td>\n",
       "      <td>4370</td>\n",
       "      <td>8220</td>\n",
       "      <td>8215</td>\n",
       "      <td>1351</td>\n",
       "      <td>21717</td>\n",
       "      <td>10186</td>\n",
       "      <td>32091</td>\n",
       "      <td>14492</td>\n",
       "      <td>10732</td>\n",
       "      <td>21263</td>\n",
       "      <td>3779</td>\n",
       "      <td>9740</td>\n",
       "      <td>3307</td>\n",
       "      <td>7459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.7916222644087325</td>\n",
       "      <td>1.2447933027441127</td>\n",
       "      <td>0.6073564939163778</td>\n",
       "      <td>-0.4029836214881138</td>\n",
       "      <td>0.014858919247313768</td>\n",
       "      <td>0.3360448936228425</td>\n",
       "      <td>-1.7009258524560966</td>\n",
       "      <td>-0.03612050432611706</td>\n",
       "      <td>0.0040998130130443225</td>\n",
       "      <td>-1.3641565303445202</td>\n",
       "      <td>-0.06386048410749744</td>\n",
       "      <td>-0.35208298462724663</td>\n",
       "      <td>0.24522926634052344</td>\n",
       "      <td>-0.4867893656199353</td>\n",
       "      <td>-0.5127104710899388</td>\n",
       "      <td>-0.22754844256589463</td>\n",
       "      <td>-0.7215260390380607</td>\n",
       "      <td>-0.043308238520626104</td>\n",
       "      <td>-0.021012887894985947</td>\n",
       "      <td>0.021008278769817514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.7310198450755296</td>\n",
       "      <td>2.3698459858854775</td>\n",
       "      <td>1.5234651456274968</td>\n",
       "      <td>0.7562925845706923</td>\n",
       "      <td>1.0103526395479507</td>\n",
       "      <td>1.262291311340743</td>\n",
       "      <td>0.3075886443494499</td>\n",
       "      <td>0.9752739961862124</td>\n",
       "      <td>1.0028458154976616</td>\n",
       "      <td>0.388461484591066</td>\n",
       "      <td>0.9567006714024515</td>\n",
       "      <td>0.7834521210481914</td>\n",
       "      <td>1.1852811153005856</td>\n",
       "      <td>0.7136114329699866</td>\n",
       "      <td>0.7009043722252475</td>\n",
       "      <td>0.8540849975447006</td>\n",
       "      <td>0.6064556124583887</td>\n",
       "      <td>0.9704271113427921</td>\n",
       "      <td>0.9855405328614022</td>\n",
       "      <td>1.0146683691431886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.035806101121986e-290</td>\n",
       "      <td>6.372275643304935e-191</td>\n",
       "      <td>0.21065443795911107</td>\n",
       "      <td>1.890859400953751e-174</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0257240740872553</td>\n",
       "      <td>0.8002632622552507</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.7083969581006864e-10</td>\n",
       "      <td>2.70657628815546e-138</td>\n",
       "      <td>1.229186442651746e-170</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.61063597e-316</td>\n",
       "      <td>6.270191890179256e-112</td>\n",
       "      <td>8.453867583693156e-241</td>\n",
       "      <td>0.003638674377470797</td>\n",
       "      <td>0.41335734087507</td>\n",
       "      <td>0.2156699395087413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">D</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>4100</td>\n",
       "      <td>3289</td>\n",
       "      <td>772</td>\n",
       "      <td>2384</td>\n",
       "      <td>3159</td>\n",
       "      <td>2984</td>\n",
       "      <td>902</td>\n",
       "      <td>1664</td>\n",
       "      <td>1500</td>\n",
       "      <td>165</td>\n",
       "      <td>2711</td>\n",
       "      <td>1449</td>\n",
       "      <td>3797</td>\n",
       "      <td>1967</td>\n",
       "      <td>1272</td>\n",
       "      <td>2362</td>\n",
       "      <td>331</td>\n",
       "      <td>1549</td>\n",
       "      <td>300</td>\n",
       "      <td>826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.9081690449579762</td>\n",
       "      <td>0.9839560055881013</td>\n",
       "      <td>0.23282608092587243</td>\n",
       "      <td>0.22875348424617994</td>\n",
       "      <td>0.28161014302851295</td>\n",
       "      <td>0.4684644430693385</td>\n",
       "      <td>-1.143124395580278</td>\n",
       "      <td>0.39659314640860466</td>\n",
       "      <td>-0.09061404907631158</td>\n",
       "      <td>-0.7962705329974455</td>\n",
       "      <td>-0.09762684236731466</td>\n",
       "      <td>-0.24968523648575092</td>\n",
       "      <td>0.011905892256402376</td>\n",
       "      <td>-0.5521092526263878</td>\n",
       "      <td>-0.9355664644147603</td>\n",
       "      <td>-0.8360744611794162</td>\n",
       "      <td>-1.3377642119508932</td>\n",
       "      <td>0.3710868401085711</td>\n",
       "      <td>-0.12011204015699223</td>\n",
       "      <td>-0.08571065431888877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.8766622743352566</td>\n",
       "      <td>1.9778815171314255</td>\n",
       "      <td>1.1751346556164124</td>\n",
       "      <td>1.1718220353436615</td>\n",
       "      <td>1.2155507626290614</td>\n",
       "      <td>1.3836359876621893</td>\n",
       "      <td>0.4527779492980159</td>\n",
       "      <td>1.3163956337394351</td>\n",
       "      <td>0.9391229486306008</td>\n",
       "      <td>0.5758358307845073</td>\n",
       "      <td>0.9345690448614863</td>\n",
       "      <td>0.8410798998983331</td>\n",
       "      <td>1.0082866816877327</td>\n",
       "      <td>0.682022267199494</td>\n",
       "      <td>0.5228371411339936</td>\n",
       "      <td>0.5601656947673631</td>\n",
       "      <td>0.39563330596343155</td>\n",
       "      <td>1.2933267787328366</td>\n",
       "      <td>0.9201161913332649</td>\n",
       "      <td>0.9423202458951588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>1.3911844489485582e-268</td>\n",
       "      <td>9.226188475020506e-254</td>\n",
       "      <td>1.4874242289118781e-05</td>\n",
       "      <td>2.677086501323572e-13</td>\n",
       "      <td>1.8560604209980214e-24</td>\n",
       "      <td>2.3368217327433813e-59</td>\n",
       "      <td>6.604862759430035e-158</td>\n",
       "      <td>7.426267526800048e-26</td>\n",
       "      <td>0.01707593313416668</td>\n",
       "      <td>1.0631792552618966e-14</td>\n",
       "      <td>0.0006595578576446023</td>\n",
       "      <td>3.8381210110522214e-11</td>\n",
       "      <td>0.6300787136420689</td>\n",
       "      <td>2.493591981345758e-68</td>\n",
       "      <td>5.492346805074204e-140</td>\n",
       "      <td>2.5440844451241457e-193</td>\n",
       "      <td>1.8587178670206804e-86</td>\n",
       "      <td>1.6808442795532283e-21</td>\n",
       "      <td>0.1552489357082094</td>\n",
       "      <td>0.0932352645169315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">E</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>27666</td>\n",
       "      <td>29291</td>\n",
       "      <td>8542</td>\n",
       "      <td>14005</td>\n",
       "      <td>18266</td>\n",
       "      <td>22097</td>\n",
       "      <td>6789</td>\n",
       "      <td>11835</td>\n",
       "      <td>11928</td>\n",
       "      <td>3310</td>\n",
       "      <td>24396</td>\n",
       "      <td>11415</td>\n",
       "      <td>38870</td>\n",
       "      <td>18559</td>\n",
       "      <td>13624</td>\n",
       "      <td>28904</td>\n",
       "      <td>2794</td>\n",
       "      <td>8179</td>\n",
       "      <td>1801</td>\n",
       "      <td>5797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.807261988810817</td>\n",
       "      <td>1.2671872661743806</td>\n",
       "      <td>0.304432140722072</td>\n",
       "      <td>-0.2943932532721293</td>\n",
       "      <td>-0.022436190108087586</td>\n",
       "      <td>0.4022735346995401</td>\n",
       "      <td>-1.3284762629345244</td>\n",
       "      <td>0.17223507369395322</td>\n",
       "      <td>0.24215022134324016</td>\n",
       "      <td>0.09853846495788161</td>\n",
       "      <td>-0.13929956428598805</td>\n",
       "      <td>-0.3748679715483922</td>\n",
       "      <td>0.262293009774181</td>\n",
       "      <td>-0.37722149963666746</td>\n",
       "      <td>-0.4700670377808065</td>\n",
       "      <td>-0.09006574322495384</td>\n",
       "      <td>-1.3524401012863967</td>\n",
       "      <td>-0.5835572706593932</td>\n",
       "      <td>-0.9947688663316094</td>\n",
       "      <td>-0.6550661326299726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.7498872748992287</td>\n",
       "      <td>2.406918457009515</td>\n",
       "      <td>1.2349324596146503</td>\n",
       "      <td>0.8154151952147913</td>\n",
       "      <td>0.9845687195009779</td>\n",
       "      <td>1.321588954914631</td>\n",
       "      <td>0.398188576306867</td>\n",
       "      <td>1.1268028155436554</td>\n",
       "      <td>1.1827541487003177</td>\n",
       "      <td>1.0706882426866253</td>\n",
       "      <td>0.9079598674165801</td>\n",
       "      <td>0.7711759837598088</td>\n",
       "      <td>1.199383482520604</td>\n",
       "      <td>0.7699189580803857</td>\n",
       "      <td>0.7219310509764991</td>\n",
       "      <td>0.939479936390537</td>\n",
       "      <td>0.3916291068467965</td>\n",
       "      <td>0.6673163371852309</td>\n",
       "      <td>0.5018162636232523</td>\n",
       "      <td>0.6350463803009097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.095915758755146e-77</td>\n",
       "      <td>1.8722082386538957e-130</td>\n",
       "      <td>0.041998678320581996</td>\n",
       "      <td>1.714e-321</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.740844297897081e-36</td>\n",
       "      <td>1.5758337513150615e-68</td>\n",
       "      <td>0.00011179819005556807</td>\n",
       "      <td>1.4293440809754403e-48</td>\n",
       "      <td>1.5486498791078382e-176</td>\n",
       "      <td>1.2899573428233083e-234</td>\n",
       "      <td>2.4607994358863487e-281</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.6554544757685694e-24</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.4475983413583616e-236</td>\n",
       "      <td>1.926260459508698e-297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">F</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>8981</td>\n",
       "      <td>10731</td>\n",
       "      <td>3624</td>\n",
       "      <td>5579</td>\n",
       "      <td>7943</td>\n",
       "      <td>7641</td>\n",
       "      <td>2188</td>\n",
       "      <td>4185</td>\n",
       "      <td>3955</td>\n",
       "      <td>898</td>\n",
       "      <td>8437</td>\n",
       "      <td>3183</td>\n",
       "      <td>11376</td>\n",
       "      <td>6240</td>\n",
       "      <td>4412</td>\n",
       "      <td>8269</td>\n",
       "      <td>827</td>\n",
       "      <td>3459</td>\n",
       "      <td>680</td>\n",
       "      <td>2418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.66023871948596</td>\n",
       "      <td>1.228539282306384</td>\n",
       "      <td>0.5574692689933536</td>\n",
       "      <td>-0.19373045225378085</td>\n",
       "      <td>0.17700411579587577</td>\n",
       "      <td>0.49312335936762897</td>\n",
       "      <td>-1.4008273942665523</td>\n",
       "      <td>0.18611961910979938</td>\n",
       "      <td>0.2074372505835611</td>\n",
       "      <td>-0.4047429981280542</td>\n",
       "      <td>-0.035307631028491035</td>\n",
       "      <td>-0.6008798534703859</td>\n",
       "      <td>0.10903399813208889</td>\n",
       "      <td>-0.42287557133284365</td>\n",
       "      <td>-0.6153899710993346</td>\n",
       "      <td>-0.299364796838504</td>\n",
       "      <td>-1.501100377141249</td>\n",
       "      <td>-0.17041354362679084</td>\n",
       "      <td>-0.4419147446496162</td>\n",
       "      <td>-0.3207422931494484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.58034409805755</td>\n",
       "      <td>2.343296127549005</td>\n",
       "      <td>1.4716853667589496</td>\n",
       "      <td>0.874341963933432</td>\n",
       "      <td>1.130533792365929</td>\n",
       "      <td>1.4074887191891707</td>\n",
       "      <td>0.37871188579485404</td>\n",
       "      <td>1.1376995547831663</td>\n",
       "      <td>1.154635310217394</td>\n",
       "      <td>0.7553708426514654</td>\n",
       "      <td>0.9758236602174326</td>\n",
       "      <td>0.6593517152617993</td>\n",
       "      <td>1.078505847150366</td>\n",
       "      <td>0.7459363456267706</td>\n",
       "      <td>0.6527534256348065</td>\n",
       "      <td>0.8126101011203983</td>\n",
       "      <td>0.3532838299930198</td>\n",
       "      <td>0.8885879339479696</td>\n",
       "      <td>0.7361569326288107</td>\n",
       "      <td>0.8006578184100672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.027510916939543e-102</td>\n",
       "      <td>3.039333144756162e-23</td>\n",
       "      <td>5.7663889748090505e-25</td>\n",
       "      <td>5.539976325692798e-165</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.1499471244400097e-15</td>\n",
       "      <td>4.601772053029021e-18</td>\n",
       "      <td>2.5298508574540015e-18</td>\n",
       "      <td>0.031564859392375244</td>\n",
       "      <td>1.2254625002299093e-134</td>\n",
       "      <td>5.219347339080129e-14</td>\n",
       "      <td>3.2821941671066666e-121</td>\n",
       "      <td>2.2219442228348963e-192</td>\n",
       "      <td>2.5384205271519286e-77</td>\n",
       "      <td>1.9933457864233387e-280</td>\n",
       "      <td>4.01120882003429e-12</td>\n",
       "      <td>6.950097593186476e-17</td>\n",
       "      <td>5.957787617881626e-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">S</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>131264</td>\n",
       "      <td>117304</td>\n",
       "      <td>32094</td>\n",
       "      <td>64651</td>\n",
       "      <td>84662</td>\n",
       "      <td>91939</td>\n",
       "      <td>32553</td>\n",
       "      <td>55003</td>\n",
       "      <td>54336</td>\n",
       "      <td>10455</td>\n",
       "      <td>96477</td>\n",
       "      <td>66737</td>\n",
       "      <td>123076</td>\n",
       "      <td>55868</td>\n",
       "      <td>45118</td>\n",
       "      <td>92903</td>\n",
       "      <td>11932</td>\n",
       "      <td>53166</td>\n",
       "      <td>16195</td>\n",
       "      <td>27838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.9307756265924941</td>\n",
       "      <td>1.2030351862296575</td>\n",
       "      <td>0.3274114171733928</td>\n",
       "      <td>-0.16152940803356572</td>\n",
       "      <td>0.1859655517210798</td>\n",
       "      <td>0.2838740369997345</td>\n",
       "      <td>-1.1618815964103053</td>\n",
       "      <td>0.234901979892448</td>\n",
       "      <td>0.20791241106941483</td>\n",
       "      <td>-0.06874002640524308</td>\n",
       "      <td>-0.03379448177889477</td>\n",
       "      <td>0.13221261773249768</td>\n",
       "      <td>-0.08037736811775897</td>\n",
       "      <td>-0.7300858654553622</td>\n",
       "      <td>-0.6476086165854361</td>\n",
       "      <td>-0.5366038970921188</td>\n",
       "      <td>-0.9979187117715504</td>\n",
       "      <td>0.062421901150843</td>\n",
       "      <td>-0.26519243145612875</td>\n",
       "      <td>-0.5132887384911861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.9063005924184837</td>\n",
       "      <td>2.302235131726489</td>\n",
       "      <td>1.2547599767375561</td>\n",
       "      <td>0.8940767532416478</td>\n",
       "      <td>1.1375780647692728</td>\n",
       "      <td>1.2174597165206624</td>\n",
       "      <td>0.4469292596556334</td>\n",
       "      <td>1.1768267781124788</td>\n",
       "      <td>1.155015659090754</td>\n",
       "      <td>0.9534703451306142</td>\n",
       "      <td>0.976847675276544</td>\n",
       "      <td>1.0959732740259998</td>\n",
       "      <td>0.9458102172306004</td>\n",
       "      <td>0.6028680316408265</td>\n",
       "      <td>0.6383375331610841</td>\n",
       "      <td>0.6893918282817189</td>\n",
       "      <td>0.5007218400859038</td>\n",
       "      <td>1.0442172532435954</td>\n",
       "      <td>0.8320877406828754</td>\n",
       "      <td>0.7006234889338285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.9640768245235985e-174</td>\n",
       "      <td>7.33051763415298e-276</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.6078508543627972e-289</td>\n",
       "      <td>6.6873530102970335e-226</td>\n",
       "      <td>1.0917249718887638e-06</td>\n",
       "      <td>2.560293136277123e-12</td>\n",
       "      <td>6.56905891961142e-114</td>\n",
       "      <td>8.76269781027807e-78</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.643661967793528e-22</td>\n",
       "      <td>2.839980130830935e-126</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">T</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>18444</td>\n",
       "      <td>13692</td>\n",
       "      <td>4928</td>\n",
       "      <td>10445</td>\n",
       "      <td>14330</td>\n",
       "      <td>18448</td>\n",
       "      <td>6659</td>\n",
       "      <td>7807</td>\n",
       "      <td>7544</td>\n",
       "      <td>1769</td>\n",
       "      <td>17335</td>\n",
       "      <td>10662</td>\n",
       "      <td>24920</td>\n",
       "      <td>10244</td>\n",
       "      <td>7774</td>\n",
       "      <td>13761</td>\n",
       "      <td>1723</td>\n",
       "      <td>5406</td>\n",
       "      <td>1069</td>\n",
       "      <td>2845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.49827267831894273</td>\n",
       "      <td>0.7957351225136622</td>\n",
       "      <td>0.17997731754269963</td>\n",
       "      <td>-0.16176527163992882</td>\n",
       "      <td>0.040299674486403106</td>\n",
       "      <td>0.6912727274784456</td>\n",
       "      <td>-0.6484143758082919</td>\n",
       "      <td>0.19568531114636448</td>\n",
       "      <td>-0.16158097002786603</td>\n",
       "      <td>0.17998176569766158</td>\n",
       "      <td>0.3615290954434676</td>\n",
       "      <td>0.37084821340717977</td>\n",
       "      <td>0.37910848980226647</td>\n",
       "      <td>-0.5236667917727693</td>\n",
       "      <td>-0.685316479829271</td>\n",
       "      <td>-0.8138047574848921</td>\n",
       "      <td>-1.3024893914568423</td>\n",
       "      <td>-0.3719238735797329</td>\n",
       "      <td>-0.8569756468163463</td>\n",
       "      <td>-0.8059098358146063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.4125213544612398</td>\n",
       "      <td>1.7359617049649632</td>\n",
       "      <td>1.1328660739166625</td>\n",
       "      <td>0.8939305941961536</td>\n",
       "      <td>1.0283274071255593</td>\n",
       "      <td>1.614707364896384</td>\n",
       "      <td>0.637981114976312</td>\n",
       "      <td>1.1452680583451735</td>\n",
       "      <td>0.8940447994639311</td>\n",
       "      <td>1.1328695668042608</td>\n",
       "      <td>1.2847869066527868</td>\n",
       "      <td>1.2931128757442645</td>\n",
       "      <td>1.3005379425450196</td>\n",
       "      <td>0.6956016250847676</td>\n",
       "      <td>0.6218693939989759</td>\n",
       "      <td>0.5688795962261581</td>\n",
       "      <td>0.40542602571809433</td>\n",
       "      <td>0.7727513241972696</td>\n",
       "      <td>0.5521087432865142</td>\n",
       "      <td>0.572001233875067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.466484630393436e-17</td>\n",
       "      <td>9.880630371790567e-30</td>\n",
       "      <td>0.001341658912458188</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.4277481880991333e-30</td>\n",
       "      <td>3.2822997804463726e-22</td>\n",
       "      <td>2.999964139536694e-07</td>\n",
       "      <td>2.3926824967976223e-203</td>\n",
       "      <td>4.0357047805682455e-136</td>\n",
       "      <td>8.282111777787279e-308</td>\n",
       "      <td>4.90666410428e-312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.6881089293780015e-84</td>\n",
       "      <td>8.353694011767148e-102</td>\n",
       "      <td>4.8373853804914224e-231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">U</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>7722</td>\n",
       "      <td>7977</td>\n",
       "      <td>1550</td>\n",
       "      <td>2938</td>\n",
       "      <td>4537</td>\n",
       "      <td>4585</td>\n",
       "      <td>1367</td>\n",
       "      <td>3365</td>\n",
       "      <td>4231</td>\n",
       "      <td>473</td>\n",
       "      <td>4570</td>\n",
       "      <td>3932</td>\n",
       "      <td>4936</td>\n",
       "      <td>2672</td>\n",
       "      <td>2317</td>\n",
       "      <td>4240</td>\n",
       "      <td>830</td>\n",
       "      <td>2821</td>\n",
       "      <td>1115</td>\n",
       "      <td>1081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.2019716525633766</td>\n",
       "      <td>1.5008978035260778</td>\n",
       "      <td>0.7601934718050505</td>\n",
       "      <td>-0.3106246008379005</td>\n",
       "      <td>0.25797420459741727</td>\n",
       "      <td>0.05353993692488491</td>\n",
       "      <td>-1.6375514425219184</td>\n",
       "      <td>0.2766472726623286</td>\n",
       "      <td>0.5696040826014372</td>\n",
       "      <td>0.5385802138634729</td>\n",
       "      <td>-0.30898796277120383</td>\n",
       "      <td>0.47747371195059163</td>\n",
       "      <td>-0.4729651898786316</td>\n",
       "      <td>-0.987677882610586</td>\n",
       "      <td>-0.9536026903335857</td>\n",
       "      <td>-0.759135055676845</td>\n",
       "      <td>-0.8536718188029068</td>\n",
       "      <td>0.19618943998286903</td>\n",
       "      <td>0.8727064692884771</td>\n",
       "      <td>-0.6698054488360393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>2.3005385830992453</td>\n",
       "      <td>2.8301878309772452</td>\n",
       "      <td>1.693717744559022</td>\n",
       "      <td>0.8062926070754913</td>\n",
       "      <td>1.1957984156966635</td>\n",
       "      <td>1.0378082696128745</td>\n",
       "      <td>0.3214014978151635</td>\n",
       "      <td>1.2113764524187987</td>\n",
       "      <td>1.4841162301632849</td>\n",
       "      <td>1.4525423366358108</td>\n",
       "      <td>0.807207809472501</td>\n",
       "      <td>1.3923034825600868</td>\n",
       "      <td>0.7204822583924111</td>\n",
       "      <td>0.5042888098415563</td>\n",
       "      <td>0.5163414455380826</td>\n",
       "      <td>0.5908504593213497</td>\n",
       "      <td>0.5533745426928136</td>\n",
       "      <td>1.145668325570486</td>\n",
       "      <td>1.8310947804169737</td>\n",
       "      <td>0.6285914487310336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.4435112508564133e-79</td>\n",
       "      <td>5.500571875217949e-32</td>\n",
       "      <td>1.0697753490346545e-29</td>\n",
       "      <td>0.01580715125723551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.521806928760941e-26</td>\n",
       "      <td>1.1139051319452184e-121</td>\n",
       "      <td>2.379697881370379e-14</td>\n",
       "      <td>6.14461766769338e-47</td>\n",
       "      <td>6.718480110167459e-82</td>\n",
       "      <td>4.964857776757042e-119</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.320663664001555e-264</td>\n",
       "      <td>1.2648923818534465e-280</td>\n",
       "      <td>6.415230133034139e-78</td>\n",
       "      <td>4.4303515042228544e-12</td>\n",
       "      <td>3.3270251532525784e-74</td>\n",
       "      <td>7.233730857511791e-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">V</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>6588</td>\n",
       "      <td>5283</td>\n",
       "      <td>1623</td>\n",
       "      <td>3765</td>\n",
       "      <td>7901</td>\n",
       "      <td>5972</td>\n",
       "      <td>1957</td>\n",
       "      <td>2983</td>\n",
       "      <td>3131</td>\n",
       "      <td>480</td>\n",
       "      <td>5128</td>\n",
       "      <td>2786</td>\n",
       "      <td>9774</td>\n",
       "      <td>4350</td>\n",
       "      <td>2993</td>\n",
       "      <td>4618</td>\n",
       "      <td>651</td>\n",
       "      <td>1688</td>\n",
       "      <td>680</td>\n",
       "      <td>1086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>0.7278982002553338</td>\n",
       "      <td>0.7866938363203918</td>\n",
       "      <td>0.4250934851211317</td>\n",
       "      <td>0.02539119641814632</td>\n",
       "      <td>1.0792141469167034</td>\n",
       "      <td>0.42386980380986267</td>\n",
       "      <td>-1.1336973452354941</td>\n",
       "      <td>0.1841424661115443</td>\n",
       "      <td>0.12289812117594323</td>\n",
       "      <td>0.17178529499196454</td>\n",
       "      <td>-0.14179014339168672</td>\n",
       "      <td>-0.09334261518687745</td>\n",
       "      <td>0.5317615766375354</td>\n",
       "      <td>-0.42536534713944446</td>\n",
       "      <td>-0.8041744952860088</td>\n",
       "      <td>-0.9623022758771723</td>\n",
       "      <td>-1.4141100050912598</td>\n",
       "      <td>-0.9476782658629279</td>\n",
       "      <td>-0.19524694366115125</td>\n",
       "      <td>-0.9136672361629999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>1.656224451694865</td>\n",
       "      <td>1.7251165520547498</td>\n",
       "      <td>1.3426595028849682</td>\n",
       "      <td>1.0177556259414604</td>\n",
       "      <td>2.1128848541920546</td>\n",
       "      <td>1.3415211536789693</td>\n",
       "      <td>0.4557462386948595</td>\n",
       "      <td>1.1361414531970981</td>\n",
       "      <td>1.088920116327693</td>\n",
       "      <td>1.1264515750534199</td>\n",
       "      <td>0.906393774091394</td>\n",
       "      <td>0.9373484659305528</td>\n",
       "      <td>1.4456933559310483</td>\n",
       "      <td>0.7446501329741373</td>\n",
       "      <td>0.5726896774599389</td>\n",
       "      <td>0.5132372272718297</td>\n",
       "      <td>0.3752411612546872</td>\n",
       "      <td>0.5184661602953369</td>\n",
       "      <td>0.8734233807235</td>\n",
       "      <td>0.5308340295150962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>5.882060476147684e-291</td>\n",
       "      <td>1.8157773228716033e-273</td>\n",
       "      <td>3.8779588658541333e-29</td>\n",
       "      <td>0.29130194712981117</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0086431966899315e-96</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.218594582743417e-11</td>\n",
       "      <td>4.183753451443084e-06</td>\n",
       "      <td>0.010667271639970399</td>\n",
       "      <td>6.4813257681652435e-12</td>\n",
       "      <td>0.0007525073982417184</td>\n",
       "      <td>4.170122480997239e-230</td>\n",
       "      <td>3.7837190122916964e-86</td>\n",
       "      <td>1.286243502384413e-232</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.9031573175713766e-192</td>\n",
       "      <td>7.147564801425799e-194</td>\n",
       "      <td>0.0003747158233188343</td>\n",
       "      <td>3.301441142235305e-117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">W</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>21</td>\n",
       "      <td>67</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>152</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.4604110200711362</td>\n",
       "      <td>1.707033883420407</td>\n",
       "      <td>-0.16858781065153247</td>\n",
       "      <td>-1.161822305498847</td>\n",
       "      <td>-0.4518051248840108</td>\n",
       "      <td>-2.428678905988094</td>\n",
       "      <td>-3.3750932494336148</td>\n",
       "      <td>0.15786564927803912</td>\n",
       "      <td>-0.5014898673203665</td>\n",
       "      <td>1.5140642640304793</td>\n",
       "      <td>-2.937376782540565</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.7762640580572135</td>\n",
       "      <td>-2.997611085480533</td>\n",
       "      <td>-4.477060934484629</td>\n",
       "      <td>-0.5480224882777285</td>\n",
       "      <td>-0.2703993646040214</td>\n",
       "      <td>1.8487356143797082</td>\n",
       "      <td>7.282937897586357</td>\n",
       "      <td>-0.7305167587498651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>2.7518675245227295</td>\n",
       "      <td>3.264888863527241</td>\n",
       "      <td>0.8897131552917903</td>\n",
       "      <td>0.44694762763149426</td>\n",
       "      <td>0.7311274761367136</td>\n",
       "      <td>0.1857354487688384</td>\n",
       "      <td>0.09638194668344366</td>\n",
       "      <td>1.1156354241346194</td>\n",
       "      <td>0.7063769308219346</td>\n",
       "      <td>2.856135181454644</td>\n",
       "      <td>0.13054537192484236</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.14596920497763735</td>\n",
       "      <td>0.12520715513405764</td>\n",
       "      <td>0.044902483950213234</td>\n",
       "      <td>0.6839569905611388</td>\n",
       "      <td>0.8290900066356066</td>\n",
       "      <td>3.601843791686597</td>\n",
       "      <td>155.7337580279398</td>\n",
       "      <td>0.6026879984422204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>6.432906013465564e-05</td>\n",
       "      <td>8.729890617319468e-15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002978426533356537</td>\n",
       "      <td>0.37254540595871544</td>\n",
       "      <td>7.386123167643203e-09</td>\n",
       "      <td>1.1810555250760989e-12</td>\n",
       "      <td>0.5314498544116163</td>\n",
       "      <td>0.3884580081320247</td>\n",
       "      <td>0.29639722215446435</td>\n",
       "      <td>3.3518012005100014e-14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.047481055172473e-13</td>\n",
       "      <td>2.9135543452439157e-09</td>\n",
       "      <td>1.637327931690171e-08</td>\n",
       "      <td>0.1606301914626022</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.618975963465429e-08</td>\n",
       "      <td>3.333561283747108e-250</td>\n",
       "      <td>0.373554222014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Z</th>\n",
       "      <th>BH_multiple_tests</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>61</td>\n",
       "      <td>47</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>53</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>31</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>85</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log2OddsRatio</th>\n",
       "      <td>1.7600860956546691</td>\n",
       "      <td>1.9700527690970622</td>\n",
       "      <td>0.8696084828482719</td>\n",
       "      <td>-1.8283955320522849</td>\n",
       "      <td>1.756458914803617</td>\n",
       "      <td>-1.9556083536932056</td>\n",
       "      <td>-1.6854436078656045</td>\n",
       "      <td>0.36415395654959776</td>\n",
       "      <td>0.7493984101734025</td>\n",
       "      <td>-1.590083894245885</td>\n",
       "      <td>-2.103277077390549</td>\n",
       "      <td>1.9739253671631096</td>\n",
       "      <td>-1.135109069240139</td>\n",
       "      <td>-2.632374777910147</td>\n",
       "      <td>-1.3098047141850495</td>\n",
       "      <td>-0.22787484779422126</td>\n",
       "      <td>-0.269349329933785</td>\n",
       "      <td>1.327264034244864</td>\n",
       "      <td>0.4581352461403243</td>\n",
       "      <td>-0.07588259206317033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oddsRatio</th>\n",
       "      <td>3.3871833802268583</td>\n",
       "      <td>3.917824489017198</td>\n",
       "      <td>1.827166978404976</td>\n",
       "      <td>0.2815775986839796</td>\n",
       "      <td>3.3786781210917534</td>\n",
       "      <td>0.2578120580610827</td>\n",
       "      <td>0.31090729912620724</td>\n",
       "      <td>1.2871265951058148</td>\n",
       "      <td>1.681091685419191</td>\n",
       "      <td>0.3321521379636192</td>\n",
       "      <td>0.23272900380042283</td>\n",
       "      <td>3.928355155933316</td>\n",
       "      <td>0.4553004943056256</td>\n",
       "      <td>0.16127840956344383</td>\n",
       "      <td>0.4033754775513141</td>\n",
       "      <td>0.853891785350294</td>\n",
       "      <td>0.8296936616825501</td>\n",
       "      <td>2.5092635988762573</td>\n",
       "      <td>1.3737650116549287</td>\n",
       "      <td>0.9487615231416884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pvalue</th>\n",
       "      <td>1.2279849723881646e-14</td>\n",
       "      <td>6.465579761701443e-14</td>\n",
       "      <td>0.020126740241883833</td>\n",
       "      <td>1.755235607278212e-06</td>\n",
       "      <td>5.429560817089812e-13</td>\n",
       "      <td>3.8538607806917204e-08</td>\n",
       "      <td>5.1509290475440594e-08</td>\n",
       "      <td>0.1762877734965421</td>\n",
       "      <td>0.008434368326188001</td>\n",
       "      <td>0.04104973895146187</td>\n",
       "      <td>1.3982674118791923e-13</td>\n",
       "      <td>5.085114868598065e-23</td>\n",
       "      <td>1.8425447123661718e-05</td>\n",
       "      <td>6.863375008025015e-11</td>\n",
       "      <td>0.005120321210862592</td>\n",
       "      <td>0.44227479617950816</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.178087348096782e-05</td>\n",
       "      <td>0.19494140122611237</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             R                        K  \\\n",
       "COG observable                                                            \n",
       "A   BH_multiple_tests                     True                     True   \n",
       "    count                                   31                       44   \n",
       "    log2OddsRatio           -0.693271051904881       0.5585519267797228   \n",
       "    oddsRatio               0.6184500347379803       1.4727901945823296   \n",
       "    pvalue                0.007137213670721523     0.015791826658973416   \n",
       "B   BH_multiple_tests                     True                     True   \n",
       "    count                                  132                      122   \n",
       "    log2OddsRatio           0.5599817011709203       1.5069209139683486   \n",
       "    oddsRatio               1.4742505181006464       2.8420282870472633   \n",
       "    pvalue               5.112321091746052e-05    6.401363726487133e-22   \n",
       "C   BH_multiple_tests                     True                     True   \n",
       "    count                                23106                    25113   \n",
       "    log2OddsRatio           0.7916222644087325       1.2447933027441127   \n",
       "    oddsRatio               1.7310198450755296       2.3698459858854775   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "D   BH_multiple_tests                     True                     True   \n",
       "    count                                 4100                     3289   \n",
       "    log2OddsRatio           0.9081690449579762       0.9839560055881013   \n",
       "    oddsRatio               1.8766622743352566       1.9778815171314255   \n",
       "    pvalue             1.3911844489485582e-268   9.226188475020506e-254   \n",
       "E   BH_multiple_tests                     True                     True   \n",
       "    count                                27666                    29291   \n",
       "    log2OddsRatio            0.807261988810817       1.2671872661743806   \n",
       "    oddsRatio               1.7498872748992287        2.406918457009515   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "F   BH_multiple_tests                     True                     True   \n",
       "    count                                 8981                    10731   \n",
       "    log2OddsRatio             0.66023871948596        1.228539282306384   \n",
       "    oddsRatio                 1.58034409805755        2.343296127549005   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "...                                        ...                      ...   \n",
       "S   BH_multiple_tests                     True                     True   \n",
       "    count                               131264                   117304   \n",
       "    log2OddsRatio           0.9307756265924941       1.2030351862296575   \n",
       "    oddsRatio               1.9063005924184837        2.302235131726489   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "T   BH_multiple_tests                     True                     True   \n",
       "    count                                18444                    13692   \n",
       "    log2OddsRatio          0.49827267831894273       0.7957351225136622   \n",
       "    oddsRatio               1.4125213544612398       1.7359617049649632   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "U   BH_multiple_tests                     True                     True   \n",
       "    count                                 7722                     7977   \n",
       "    log2OddsRatio           1.2019716525633766       1.5008978035260778   \n",
       "    oddsRatio               2.3005385830992453       2.8301878309772452   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "V   BH_multiple_tests                     True                     True   \n",
       "    count                                 6588                     5283   \n",
       "    log2OddsRatio           0.7278982002553338       0.7866938363203918   \n",
       "    oddsRatio                1.656224451694865       1.7251165520547498   \n",
       "    pvalue              5.882060476147684e-291  1.8157773228716033e-273   \n",
       "W   BH_multiple_tests                     True                     True   \n",
       "    count                                   21                       67   \n",
       "    log2OddsRatio           1.4604110200711362        1.707033883420407   \n",
       "    oddsRatio               2.7518675245227295        3.264888863527241   \n",
       "    pvalue               6.432906013465564e-05    8.729890617319468e-15   \n",
       "Z   BH_multiple_tests                     True                     True   \n",
       "    count                                   61                       47   \n",
       "    log2OddsRatio           1.7600860956546691       1.9700527690970622   \n",
       "    oddsRatio               3.3871833802268583        3.917824489017198   \n",
       "    pvalue              1.2279849723881646e-14    6.465579761701443e-14   \n",
       "\n",
       "                                            H                        D  \\\n",
       "COG observable                                                           \n",
       "A   BH_multiple_tests                    True                    False   \n",
       "    count                                   4                       45   \n",
       "    log2OddsRatio           -2.58187599893849      0.15227088164812616   \n",
       "    oddsRatio             0.16702361502625615       1.1113173709619817   \n",
       "    pvalue              2.028341221679015e-06       0.4667586992491468   \n",
       "B   BH_multiple_tests                    True                    False   \n",
       "    count                                  26                      123   \n",
       "    log2OddsRatio         -1.1885418114752087      0.29631992894676673   \n",
       "    oddsRatio             0.43874609452831226       1.2280079741276693   \n",
       "    pvalue             3.0702616474021627e-06      0.03473055753940365   \n",
       "C   BH_multiple_tests                    True                     True   \n",
       "    count                                8818                    10685   \n",
       "    log2OddsRatio          0.6073564939163778      -0.4029836214881138   \n",
       "    oddsRatio              1.5234651456274968       0.7562925845706923   \n",
       "    pvalue             5.035806101121986e-290   6.372275643304935e-191   \n",
       "D   BH_multiple_tests                    True                     True   \n",
       "    count                                 772                     2384   \n",
       "    log2OddsRatio         0.23282608092587243      0.22875348424617994   \n",
       "    oddsRatio              1.1751346556164124       1.1718220353436615   \n",
       "    pvalue             1.4874242289118781e-05    2.677086501323572e-13   \n",
       "E   BH_multiple_tests                    True                     True   \n",
       "    count                                8542                    14005   \n",
       "    log2OddsRatio           0.304432140722072      -0.2943932532721293   \n",
       "    oddsRatio              1.2349324596146503       0.8154151952147913   \n",
       "    pvalue              4.095915758755146e-77  1.8722082386538957e-130   \n",
       "F   BH_multiple_tests                    True                     True   \n",
       "    count                                3624                     5579   \n",
       "    log2OddsRatio          0.5574692689933536     -0.19373045225378085   \n",
       "    oddsRatio              1.4716853667589496        0.874341963933432   \n",
       "    pvalue             1.027510916939543e-102    3.039333144756162e-23   \n",
       "...                                       ...                      ...   \n",
       "S   BH_multiple_tests                    True                     True   \n",
       "    count                               32094                    64651   \n",
       "    log2OddsRatio          0.3274114171733928     -0.16152940803356572   \n",
       "    oddsRatio              1.2547599767375561       0.8940767532416478   \n",
       "    pvalue                                0.0  2.9640768245235985e-174   \n",
       "T   BH_multiple_tests                    True                     True   \n",
       "    count                                4928                    10445   \n",
       "    log2OddsRatio         0.17997731754269963     -0.16176527163992882   \n",
       "    oddsRatio              1.1328660739166625       0.8939305941961536   \n",
       "    pvalue              2.466484630393436e-17    9.880630371790567e-30   \n",
       "U   BH_multiple_tests                    True                     True   \n",
       "    count                                1550                     2938   \n",
       "    log2OddsRatio          0.7601934718050505      -0.3106246008379005   \n",
       "    oddsRatio               1.693717744559022       0.8062926070754913   \n",
       "    pvalue             1.4435112508564133e-79    5.500571875217949e-32   \n",
       "V   BH_multiple_tests                    True                    False   \n",
       "    count                                1623                     3765   \n",
       "    log2OddsRatio          0.4250934851211317      0.02539119641814632   \n",
       "    oddsRatio              1.3426595028849682       1.0177556259414604   \n",
       "    pvalue             3.8779588658541333e-29      0.29130194712981117   \n",
       "W   BH_multiple_tests                   False                     True   \n",
       "    count                                   3                       12   \n",
       "    log2OddsRatio        -0.16858781065153247       -1.161822305498847   \n",
       "    oddsRatio              0.8897131552917903      0.44694762763149426   \n",
       "    pvalue                                1.0     0.002978426533356537   \n",
       "Z   BH_multiple_tests                   False                     True   \n",
       "    count                                  17                       10   \n",
       "    log2OddsRatio          0.8696084828482719      -1.8283955320522849   \n",
       "    oddsRatio               1.827166978404976       0.2815775986839796   \n",
       "    pvalue               0.020126740241883833    1.755235607278212e-06   \n",
       "\n",
       "                                            E                       S  \\\n",
       "COG observable                                                          \n",
       "A   BH_multiple_tests                   False                   False   \n",
       "    count                                  55                      30   \n",
       "    log2OddsRatio         0.09408620111261075     -0.4154815281994051   \n",
       "    oddsRatio               1.067389112696242       0.749769202473172   \n",
       "    pvalue                 0.6125740828737359     0.13811753074782185   \n",
       "B   BH_multiple_tests                    True                    True   \n",
       "    count                                  53                     113   \n",
       "    log2OddsRatio         -0.7743384731918935       0.749073302909063   \n",
       "    oddsRatio              0.5846566486213345       1.680712898823263   \n",
       "    pvalue             4.4719144417806113e-05   7.313342920223233e-07   \n",
       "C   BH_multiple_tests                   False                    True   \n",
       "    count                               15793                   16764   \n",
       "    log2OddsRatio        0.014858919247313768      0.3360448936228425   \n",
       "    oddsRatio              1.0103526395479507       1.262291311340743   \n",
       "    pvalue                0.21065443795911107  1.890859400953751e-174   \n",
       "D   BH_multiple_tests                    True                    True   \n",
       "    count                                3159                    2984   \n",
       "    log2OddsRatio         0.28161014302851295      0.4684644430693385   \n",
       "    oddsRatio              1.2155507626290614      1.3836359876621893   \n",
       "    pvalue             1.8560604209980214e-24  2.3368217327433813e-59   \n",
       "E   BH_multiple_tests                    True                    True   \n",
       "    count                               18266                   22097   \n",
       "    log2OddsRatio       -0.022436190108087586      0.4022735346995401   \n",
       "    oddsRatio              0.9845687195009779       1.321588954914631   \n",
       "    pvalue               0.041998678320581996              1.714e-321   \n",
       "F   BH_multiple_tests                    True                    True   \n",
       "    count                                7943                    7641   \n",
       "    log2OddsRatio         0.17700411579587577     0.49312335936762897   \n",
       "    oddsRatio               1.130533792365929      1.4074887191891707   \n",
       "    pvalue             5.7663889748090505e-25  5.539976325692798e-165   \n",
       "...                                       ...                     ...   \n",
       "S   BH_multiple_tests                    True                    True   \n",
       "    count                               84662                   91939   \n",
       "    log2OddsRatio          0.1859655517210798      0.2838740369997345   \n",
       "    oddsRatio              1.1375780647692728      1.2174597165206624   \n",
       "    pvalue              7.33051763415298e-276                     0.0   \n",
       "T   BH_multiple_tests                    True                    True   \n",
       "    count                               14330                   18448   \n",
       "    log2OddsRatio        0.040299674486403106      0.6912727274784456   \n",
       "    oddsRatio              1.0283274071255593       1.614707364896384   \n",
       "    pvalue               0.001341658912458188                     0.0   \n",
       "U   BH_multiple_tests                    True                    True   \n",
       "    count                                4537                    4585   \n",
       "    log2OddsRatio         0.25797420459741727     0.05353993692488491   \n",
       "    oddsRatio              1.1957984156966635      1.0378082696128745   \n",
       "    pvalue             1.0697753490346545e-29     0.01580715125723551   \n",
       "V   BH_multiple_tests                    True                    True   \n",
       "    count                                7901                    5972   \n",
       "    log2OddsRatio          1.0792141469167034     0.42386980380986267   \n",
       "    oddsRatio              2.1128848541920546      1.3415211536789693   \n",
       "    pvalue                                0.0  1.0086431966899315e-96   \n",
       "W   BH_multiple_tests                   False                    True   \n",
       "    count                                  12                       7   \n",
       "    log2OddsRatio         -0.4518051248840108      -2.428678905988094   \n",
       "    oddsRatio              0.7311274761367136      0.1857354487688384   \n",
       "    pvalue                0.37254540595871544   7.386123167643203e-09   \n",
       "Z   BH_multiple_tests                    True                    True   \n",
       "    count                                  53                      11   \n",
       "    log2OddsRatio           1.756458914803617     -1.9556083536932056   \n",
       "    oddsRatio              3.3786781210917534      0.2578120580610827   \n",
       "    pvalue              5.429560817089812e-13  3.8538607806917204e-08   \n",
       "\n",
       "                                            T                        N  \\\n",
       "COG observable                                                           \n",
       "A   BH_multiple_tests                    True                    False   \n",
       "    count                                  15                       26   \n",
       "    log2OddsRatio         -1.3618759538736174       0.3510027643641223   \n",
       "    oddsRatio              0.3890760404623243       1.2754468355741662   \n",
       "    pvalue              4.305732143214525e-05       0.2164643876832293   \n",
       "B   BH_multiple_tests                    True                     True   \n",
       "    count                                  43                       29   \n",
       "    log2OddsRatio         -0.6598443041556618      -0.6199940495152496   \n",
       "    oddsRatio              0.6329466009823453       0.6506736114589072   \n",
       "    pvalue              0.0020548789250997308      0.02111095268512736   \n",
       "C   BH_multiple_tests                    True                     True   \n",
       "    count                                4370                     8220   \n",
       "    log2OddsRatio         -1.7009258524560966     -0.03612050432611706   \n",
       "    oddsRatio              0.3075886443494499       0.9752739961862124   \n",
       "    pvalue                                0.0       0.0257240740872553   \n",
       "D   BH_multiple_tests                    True                     True   \n",
       "    count                                 902                     1664   \n",
       "    log2OddsRatio          -1.143124395580278      0.39659314640860466   \n",
       "    oddsRatio              0.4527779492980159       1.3163956337394351   \n",
       "    pvalue             6.604862759430035e-158    7.426267526800048e-26   \n",
       "E   BH_multiple_tests                    True                     True   \n",
       "    count                                6789                    11835   \n",
       "    log2OddsRatio         -1.3284762629345244      0.17223507369395322   \n",
       "    oddsRatio               0.398188576306867       1.1268028155436554   \n",
       "    pvalue                                0.0    8.740844297897081e-36   \n",
       "F   BH_multiple_tests                    True                     True   \n",
       "    count                                2188                     4185   \n",
       "    log2OddsRatio         -1.4008273942665523      0.18611961910979938   \n",
       "    oddsRatio             0.37871188579485404       1.1376995547831663   \n",
       "    pvalue                                0.0   1.1499471244400097e-15   \n",
       "...                                       ...                      ...   \n",
       "S   BH_multiple_tests                    True                     True   \n",
       "    count                               32553                    55003   \n",
       "    log2OddsRatio         -1.1618815964103053        0.234901979892448   \n",
       "    oddsRatio              0.4469292596556334       1.1768267781124788   \n",
       "    pvalue                                0.0  1.6078508543627972e-289   \n",
       "T   BH_multiple_tests                    True                     True   \n",
       "    count                                6659                     7807   \n",
       "    log2OddsRatio         -0.6484143758082919      0.19568531114636448   \n",
       "    oddsRatio               0.637981114976312       1.1452680583451735   \n",
       "    pvalue                                0.0   1.4277481880991333e-30   \n",
       "U   BH_multiple_tests                    True                     True   \n",
       "    count                                1367                     3365   \n",
       "    log2OddsRatio         -1.6375514425219184       0.2766472726623286   \n",
       "    oddsRatio              0.3214014978151635       1.2113764524187987   \n",
       "    pvalue                                0.0    6.521806928760941e-26   \n",
       "V   BH_multiple_tests                    True                     True   \n",
       "    count                                1957                     2983   \n",
       "    log2OddsRatio         -1.1336973452354941       0.1841424661115443   \n",
       "    oddsRatio              0.4557462386948595       1.1361414531970981   \n",
       "    pvalue                                0.0    2.218594582743417e-11   \n",
       "W   BH_multiple_tests                    True                    False   \n",
       "    count                                   4                       38   \n",
       "    log2OddsRatio         -3.3750932494336148      0.15786564927803912   \n",
       "    oddsRatio             0.09638194668344366       1.1156354241346194   \n",
       "    pvalue             1.1810555250760989e-12       0.5314498544116163   \n",
       "Z   BH_multiple_tests                    True                    False   \n",
       "    count                                  16                       31   \n",
       "    log2OddsRatio         -1.6854436078656045      0.36415395654959776   \n",
       "    oddsRatio             0.31090729912620724       1.2871265951058148   \n",
       "    pvalue             5.1509290475440594e-08       0.1762877734965421   \n",
       "\n",
       "                                             Q                       C  \\\n",
       "COG observable                                                           \n",
       "A   BH_multiple_tests                    False                   False   \n",
       "    count                                   25                       3   \n",
       "    log2OddsRatio        -0.030886775495811626     -1.3487013105157886   \n",
       "    oddsRatio               0.9788184662757863       0.392645342554571   \n",
       "    pvalue                                 1.0     0.09896812200450078   \n",
       "B   BH_multiple_tests                     True                    True   \n",
       "    count                                   32                       5   \n",
       "    log2OddsRatio          -0.5844625005760626     -2.1550706957054917   \n",
       "    oddsRatio                0.666897755836301     0.22452209080116056   \n",
       "    pvalue                  0.0215782120450621   3.230438461596552e-05   \n",
       "C   BH_multiple_tests                    False                    True   \n",
       "    count                                 8215                    1351   \n",
       "    log2OddsRatio        0.0040998130130443225     -1.3641565303445202   \n",
       "    oddsRatio               1.0028458154976616       0.388461484591066   \n",
       "    pvalue                  0.8002632622552507                     0.0   \n",
       "D   BH_multiple_tests                     True                    True   \n",
       "    count                                 1500                     165   \n",
       "    log2OddsRatio         -0.09061404907631158     -0.7962705329974455   \n",
       "    oddsRatio               0.9391229486306008      0.5758358307845073   \n",
       "    pvalue                 0.01707593313416668  1.0631792552618966e-14   \n",
       "E   BH_multiple_tests                     True                    True   \n",
       "    count                                11928                    3310   \n",
       "    log2OddsRatio          0.24215022134324016     0.09853846495788161   \n",
       "    oddsRatio               1.1827541487003177      1.0706882426866253   \n",
       "    pvalue              1.5758337513150615e-68  0.00011179819005556807   \n",
       "F   BH_multiple_tests                     True                    True   \n",
       "    count                                 3955                     898   \n",
       "    log2OddsRatio           0.2074372505835611     -0.4047429981280542   \n",
       "    oddsRatio                1.154635310217394      0.7553708426514654   \n",
       "    pvalue               4.601772053029021e-18  2.5298508574540015e-18   \n",
       "...                                        ...                     ...   \n",
       "S   BH_multiple_tests                     True                    True   \n",
       "    count                                54336                   10455   \n",
       "    log2OddsRatio          0.20791241106941483    -0.06874002640524308   \n",
       "    oddsRatio                1.155015659090754      0.9534703451306142   \n",
       "    pvalue             6.6873530102970335e-226  1.0917249718887638e-06   \n",
       "T   BH_multiple_tests                     True                    True   \n",
       "    count                                 7544                    1769   \n",
       "    log2OddsRatio         -0.16158097002786603     0.17998176569766158   \n",
       "    oddsRatio               0.8940447994639311      1.1328695668042608   \n",
       "    pvalue              3.2822997804463726e-22   2.999964139536694e-07   \n",
       "U   BH_multiple_tests                     True                    True   \n",
       "    count                                 4231                     473   \n",
       "    log2OddsRatio           0.5696040826014372      0.5385802138634729   \n",
       "    oddsRatio               1.4841162301632849      1.4525423366358108   \n",
       "    pvalue             1.1139051319452184e-121   2.379697881370379e-14   \n",
       "V   BH_multiple_tests                     True                    True   \n",
       "    count                                 3131                     480   \n",
       "    log2OddsRatio          0.12289812117594323     0.17178529499196454   \n",
       "    oddsRatio                1.088920116327693      1.1264515750534199   \n",
       "    pvalue               4.183753451443084e-06    0.010667271639970399   \n",
       "W   BH_multiple_tests                    False                   False   \n",
       "    count                                    9                       1   \n",
       "    log2OddsRatio          -0.5014898673203665      1.5140642640304793   \n",
       "    oddsRatio               0.7063769308219346       2.856135181454644   \n",
       "    pvalue                  0.3884580081320247     0.29639722215446435   \n",
       "Z   BH_multiple_tests                     True                   False   \n",
       "    count                                   30                       3   \n",
       "    log2OddsRatio           0.7493984101734025      -1.590083894245885   \n",
       "    oddsRatio                1.681091685419191      0.3321521379636192   \n",
       "    pvalue                0.008434368326188001     0.04104973895146187   \n",
       "\n",
       "                                             G                        P  \\\n",
       "COG observable                                                            \n",
       "A   BH_multiple_tests                    False                    False   \n",
       "    count                                   58                       32   \n",
       "    log2OddsRatio           0.4227917307014406     -0.05559612951644007   \n",
       "    oddsRatio                1.340519058568072       0.9621967759545658   \n",
       "    pvalue                 0.03549844277465371       0.9288822431478668   \n",
       "B   BH_multiple_tests                    False                     True   \n",
       "    count                                  151                       55   \n",
       "    log2OddsRatio          0.24497270815231342       -0.625227522825236   \n",
       "    oddsRatio               1.1850703524369361        0.648317525177379   \n",
       "    pvalue                 0.05288050032066993    0.0010087512411233082   \n",
       "C   BH_multiple_tests                     True                     True   \n",
       "    count                                21717                    10186   \n",
       "    log2OddsRatio         -0.06386048410749744     -0.35208298462724663   \n",
       "    oddsRatio               0.9567006714024515       0.7834521210481914   \n",
       "    pvalue              3.7083969581006864e-10    2.70657628815546e-138   \n",
       "D   BH_multiple_tests                     True                     True   \n",
       "    count                                 2711                     1449   \n",
       "    log2OddsRatio         -0.09762684236731466     -0.24968523648575092   \n",
       "    oddsRatio               0.9345690448614863       0.8410798998983331   \n",
       "    pvalue               0.0006595578576446023   3.8381210110522214e-11   \n",
       "E   BH_multiple_tests                     True                     True   \n",
       "    count                                24396                    11415   \n",
       "    log2OddsRatio         -0.13929956428598805      -0.3748679715483922   \n",
       "    oddsRatio               0.9079598674165801       0.7711759837598088   \n",
       "    pvalue              1.4293440809754403e-48  1.5486498791078382e-176   \n",
       "F   BH_multiple_tests                     True                     True   \n",
       "    count                                 8437                     3183   \n",
       "    log2OddsRatio        -0.035307631028491035      -0.6008798534703859   \n",
       "    oddsRatio               0.9758236602174326       0.6593517152617993   \n",
       "    pvalue                0.031564859392375244  1.2254625002299093e-134   \n",
       "...                                        ...                      ...   \n",
       "S   BH_multiple_tests                     True                     True   \n",
       "    count                                96477                    66737   \n",
       "    log2OddsRatio         -0.03379448177889477      0.13221261773249768   \n",
       "    oddsRatio                0.976847675276544       1.0959732740259998   \n",
       "    pvalue               2.560293136277123e-12    6.56905891961142e-114   \n",
       "T   BH_multiple_tests                     True                     True   \n",
       "    count                                17335                    10662   \n",
       "    log2OddsRatio           0.3615290954434676      0.37084821340717977   \n",
       "    oddsRatio               1.2847869066527868       1.2931128757442645   \n",
       "    pvalue             2.3926824967976223e-203  4.0357047805682455e-136   \n",
       "U   BH_multiple_tests                     True                     True   \n",
       "    count                                 4570                     3932   \n",
       "    log2OddsRatio         -0.30898796277120383      0.47747371195059163   \n",
       "    oddsRatio                0.807207809472501       1.3923034825600868   \n",
       "    pvalue                6.14461766769338e-47    6.718480110167459e-82   \n",
       "V   BH_multiple_tests                     True                     True   \n",
       "    count                                 5128                     2786   \n",
       "    log2OddsRatio         -0.14179014339168672     -0.09334261518687745   \n",
       "    oddsRatio                0.906393774091394       0.9373484659305528   \n",
       "    pvalue              6.4813257681652435e-12    0.0007525073982417184   \n",
       "W   BH_multiple_tests                     True                      NaN   \n",
       "    count                                    7                        0   \n",
       "    log2OddsRatio           -2.937376782540565                      NaN   \n",
       "    oddsRatio              0.13054537192484236                      NaN   \n",
       "    pvalue              3.3518012005100014e-14                      NaN   \n",
       "Z   BH_multiple_tests                     True                     True   \n",
       "    count                                   17                       85   \n",
       "    log2OddsRatio           -2.103277077390549       1.9739253671631096   \n",
       "    oddsRatio              0.23272900380042283        3.928355155933316   \n",
       "    pvalue              1.3982674118791923e-13    5.085114868598065e-23   \n",
       "\n",
       "                                             A                        V  \\\n",
       "COG observable                                                            \n",
       "A   BH_multiple_tests                    False                    False   \n",
       "    count                                   77                       51   \n",
       "    log2OddsRatio          0.14852651881926163    -0.013681018395559259   \n",
       "    oddsRatio               1.1084368036489123       0.9905618621387978   \n",
       "    pvalue                  0.3777368989010897                      1.0   \n",
       "B   BH_multiple_tests                    False                     True   \n",
       "    count                                  150                       52   \n",
       "    log2OddsRatio         -0.19145788193205807      -1.0072336237084454   \n",
       "    oddsRatio               0.8757203352132595      0.49749929153304434   \n",
       "    pvalue                 0.12974062907164013    4.964229369662906e-08   \n",
       "C   BH_multiple_tests                     True                     True   \n",
       "    count                                32091                    14492   \n",
       "    log2OddsRatio          0.24522926634052344      -0.4867893656199353   \n",
       "    oddsRatio               1.1852811153005856       0.7136114329699866   \n",
       "    pvalue              1.229186442651746e-170                      0.0   \n",
       "D   BH_multiple_tests                    False                     True   \n",
       "    count                                 3797                     1967   \n",
       "    log2OddsRatio         0.011905892256402376      -0.5521092526263878   \n",
       "    oddsRatio               1.0082866816877327        0.682022267199494   \n",
       "    pvalue                  0.6300787136420689    2.493591981345758e-68   \n",
       "E   BH_multiple_tests                     True                     True   \n",
       "    count                                38870                    18559   \n",
       "    log2OddsRatio            0.262293009774181     -0.37722149963666746   \n",
       "    oddsRatio                1.199383482520604       0.7699189580803857   \n",
       "    pvalue             1.2899573428233083e-234  2.4607994358863487e-281   \n",
       "F   BH_multiple_tests                     True                     True   \n",
       "    count                                11376                     6240   \n",
       "    log2OddsRatio          0.10903399813208889     -0.42287557133284365   \n",
       "    oddsRatio                1.078505847150366       0.7459363456267706   \n",
       "    pvalue               5.219347339080129e-14  3.2821941671066666e-121   \n",
       "...                                        ...                      ...   \n",
       "S   BH_multiple_tests                     True                     True   \n",
       "    count                               123076                    55868   \n",
       "    log2OddsRatio         -0.08037736811775897      -0.7300858654553622   \n",
       "    oddsRatio               0.9458102172306004       0.6028680316408265   \n",
       "    pvalue                8.76269781027807e-78                      0.0   \n",
       "T   BH_multiple_tests                     True                     True   \n",
       "    count                                24920                    10244   \n",
       "    log2OddsRatio          0.37910848980226647      -0.5236667917727693   \n",
       "    oddsRatio               1.3005379425450196       0.6956016250847676   \n",
       "    pvalue              8.282111777787279e-308       4.90666410428e-312   \n",
       "U   BH_multiple_tests                     True                     True   \n",
       "    count                                 4936                     2672   \n",
       "    log2OddsRatio          -0.4729651898786316       -0.987677882610586   \n",
       "    oddsRatio               0.7204822583924111       0.5042888098415563   \n",
       "    pvalue              4.964857776757042e-119                      0.0   \n",
       "V   BH_multiple_tests                     True                     True   \n",
       "    count                                 9774                     4350   \n",
       "    log2OddsRatio           0.5317615766375354     -0.42536534713944446   \n",
       "    oddsRatio               1.4456933559310483       0.7446501329741373   \n",
       "    pvalue              4.170122480997239e-230   3.7837190122916964e-86   \n",
       "W   BH_multiple_tests                     True                     True   \n",
       "    count                                    8                        4   \n",
       "    log2OddsRatio          -2.7762640580572135       -2.997611085480533   \n",
       "    oddsRatio              0.14596920497763735      0.12520715513405764   \n",
       "    pvalue               1.047481055172473e-13   2.9135543452439157e-09   \n",
       "Z   BH_multiple_tests                     True                     True   \n",
       "    count                                   26                        7   \n",
       "    log2OddsRatio           -1.135109069240139       -2.632374777910147   \n",
       "    oddsRatio               0.4553004943056256      0.16127840956344383   \n",
       "    pvalue              1.8425447123661718e-05    6.863375008025015e-11   \n",
       "\n",
       "                                             I                        L  \\\n",
       "COG observable                                                            \n",
       "A   BH_multiple_tests                     True                     True   \n",
       "    count                                   70                      108   \n",
       "    log2OddsRatio           0.9201937281855587       0.6399756962616083   \n",
       "    oddsRatio               1.8923693878123335       1.5583029078235333   \n",
       "    pvalue              2.9477769043096295e-06    5.493094170602914e-05   \n",
       "B   BH_multiple_tests                    False                     True   \n",
       "    count                                   57                      205   \n",
       "    log2OddsRatio         -0.39546819313465836       0.5578435660865309   \n",
       "    oddsRatio               0.7602426175158838         1.47206723472285   \n",
       "    pvalue                 0.04208589977126963    9.310525489793519e-07   \n",
       "C   BH_multiple_tests                     True                     True   \n",
       "    count                                10732                    21263   \n",
       "    log2OddsRatio          -0.5127104710899388     -0.22754844256589463   \n",
       "    oddsRatio               0.7009043722252475       0.8540849975447006   \n",
       "    pvalue                     1.61063597e-316   6.270191890179256e-112   \n",
       "D   BH_multiple_tests                     True                     True   \n",
       "    count                                 1272                     2362   \n",
       "    log2OddsRatio          -0.9355664644147603      -0.8360744611794162   \n",
       "    oddsRatio               0.5228371411339936       0.5601656947673631   \n",
       "    pvalue              5.492346805074204e-140  2.5440844451241457e-193   \n",
       "E   BH_multiple_tests                     True                     True   \n",
       "    count                                13624                    28904   \n",
       "    log2OddsRatio          -0.4700670377808065     -0.09006574322495384   \n",
       "    oddsRatio               0.7219310509764991        0.939479936390537   \n",
       "    pvalue                                 0.0   2.6554544757685694e-24   \n",
       "F   BH_multiple_tests                     True                     True   \n",
       "    count                                 4412                     8269   \n",
       "    log2OddsRatio          -0.6153899710993346       -0.299364796838504   \n",
       "    oddsRatio               0.6527534256348065       0.8126101011203983   \n",
       "    pvalue             2.2219442228348963e-192   2.5384205271519286e-77   \n",
       "...                                        ...                      ...   \n",
       "S   BH_multiple_tests                     True                     True   \n",
       "    count                                45118                    92903   \n",
       "    log2OddsRatio          -0.6476086165854361      -0.5366038970921188   \n",
       "    oddsRatio               0.6383375331610841       0.6893918282817189   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "T   BH_multiple_tests                     True                     True   \n",
       "    count                                 7774                    13761   \n",
       "    log2OddsRatio           -0.685316479829271      -0.8138047574848921   \n",
       "    oddsRatio               0.6218693939989759       0.5688795962261581   \n",
       "    pvalue                                 0.0                      0.0   \n",
       "U   BH_multiple_tests                     True                     True   \n",
       "    count                                 2317                     4240   \n",
       "    log2OddsRatio          -0.9536026903335857       -0.759135055676845   \n",
       "    oddsRatio               0.5163414455380826       0.5908504593213497   \n",
       "    pvalue              4.320663664001555e-264  1.2648923818534465e-280   \n",
       "V   BH_multiple_tests                     True                     True   \n",
       "    count                                 2993                     4618   \n",
       "    log2OddsRatio          -0.8041744952860088      -0.9623022758771723   \n",
       "    oddsRatio               0.5726896774599389       0.5132372272718297   \n",
       "    pvalue              1.286243502384413e-232                      0.0   \n",
       "W   BH_multiple_tests                     True                    False   \n",
       "    count                                    1                       16   \n",
       "    log2OddsRatio           -4.477060934484629      -0.5480224882777285   \n",
       "    oddsRatio             0.044902483950213234       0.6839569905611388   \n",
       "    pvalue               1.637327931690171e-08       0.1606301914626022   \n",
       "Z   BH_multiple_tests                     True                    False   \n",
       "    count                                    8                       32   \n",
       "    log2OddsRatio          -1.3098047141850495     -0.22787484779422126   \n",
       "    oddsRatio               0.4033754775513141        0.853891785350294   \n",
       "    pvalue                0.005120321210862592      0.44227479617950816   \n",
       "\n",
       "                                             M                       F  \\\n",
       "COG observable                                                           \n",
       "A   BH_multiple_tests                     True                   False   \n",
       "    count                                    3                      16   \n",
       "    log2OddsRatio           -2.522705564148908     -0.4447574285317583   \n",
       "    oddsRatio               0.1740163098878695      0.7347078383621707   \n",
       "    pvalue                9.22763654415142e-05       0.271290040827016   \n",
       "B   BH_multiple_tests                     True                    True   \n",
       "    count                                   11                      82   \n",
       "    log2OddsRatio          -1.2963419785405552      0.5049887681728684   \n",
       "    oddsRatio              0.40715725694427124      1.4191123079291144   \n",
       "    pvalue               0.0008315055280774621    0.003265964700197696   \n",
       "C   BH_multiple_tests                     True                    True   \n",
       "    count                                 3779                    9740   \n",
       "    log2OddsRatio          -0.7215260390380607   -0.043308238520626104   \n",
       "    oddsRatio               0.6064556124583887      0.9704271113427921   \n",
       "    pvalue              8.453867583693156e-241    0.003638674377470797   \n",
       "D   BH_multiple_tests                     True                    True   \n",
       "    count                                  331                    1549   \n",
       "    log2OddsRatio          -1.3377642119508932      0.3710868401085711   \n",
       "    oddsRatio              0.39563330596343155      1.2933267787328366   \n",
       "    pvalue              1.8587178670206804e-86  1.6808442795532283e-21   \n",
       "E   BH_multiple_tests                     True                    True   \n",
       "    count                                 2794                    8179   \n",
       "    log2OddsRatio          -1.3524401012863967     -0.5835572706593932   \n",
       "    oddsRatio               0.3916291068467965      0.6673163371852309   \n",
       "    pvalue                                 0.0                     0.0   \n",
       "F   BH_multiple_tests                     True                    True   \n",
       "    count                                  827                    3459   \n",
       "    log2OddsRatio           -1.501100377141249    -0.17041354362679084   \n",
       "    oddsRatio               0.3532838299930198      0.8885879339479696   \n",
       "    pvalue             1.9933457864233387e-280    4.01120882003429e-12   \n",
       "...                                        ...                     ...   \n",
       "S   BH_multiple_tests                     True                    True   \n",
       "    count                                11932                   53166   \n",
       "    log2OddsRatio          -0.9979187117715504       0.062421901150843   \n",
       "    oddsRatio               0.5007218400859038      1.0442172532435954   \n",
       "    pvalue                                 0.0   3.643661967793528e-22   \n",
       "T   BH_multiple_tests                     True                    True   \n",
       "    count                                 1723                    5406   \n",
       "    log2OddsRatio          -1.3024893914568423     -0.3719238735797329   \n",
       "    oddsRatio              0.40542602571809433      0.7727513241972696   \n",
       "    pvalue                                 0.0  2.6881089293780015e-84   \n",
       "U   BH_multiple_tests                     True                    True   \n",
       "    count                                  830                    2821   \n",
       "    log2OddsRatio          -0.8536718188029068     0.19618943998286903   \n",
       "    oddsRatio               0.5533745426928136       1.145668325570486   \n",
       "    pvalue               6.415230133034139e-78  4.4303515042228544e-12   \n",
       "V   BH_multiple_tests                     True                    True   \n",
       "    count                                  651                    1688   \n",
       "    log2OddsRatio          -1.4141100050912598     -0.9476782658629279   \n",
       "    oddsRatio               0.3752411612546872      0.5184661602953369   \n",
       "    pvalue             6.9031573175713766e-192  7.147564801425799e-194   \n",
       "W   BH_multiple_tests                    False                    True   \n",
       "    count                                    3                      26   \n",
       "    log2OddsRatio          -0.2703993646040214      1.8487356143797082   \n",
       "    oddsRatio               0.8290900066356066       3.601843791686597   \n",
       "    pvalue                                 1.0   8.618975963465429e-08   \n",
       "Z   BH_multiple_tests                    False                    True   \n",
       "    count                                    4                      29   \n",
       "    log2OddsRatio           -0.269349329933785       1.327264034244864   \n",
       "    oddsRatio               0.8296936616825501      2.5092635988762573   \n",
       "    pvalue                                 1.0   2.178087348096782e-05   \n",
       "\n",
       "                                             W                        Y  \n",
       "COG observable                                                           \n",
       "A   BH_multiple_tests                      NaN                    False  \n",
       "    count                                    0                       10  \n",
       "    log2OddsRatio                          NaN       -0.908497334160046  \n",
       "    oddsRatio                              NaN       0.5327396872766956  \n",
       "    pvalue                                 NaN      0.04366404711602853  \n",
       "B   BH_multiple_tests                     True                    False  \n",
       "    count                                    6                       43  \n",
       "    log2OddsRatio           -1.590345057686168     -0.19400670601663597  \n",
       "    oddsRatio               0.3320920156637018       0.8741745570133416  \n",
       "    pvalue                0.002565907736147211       0.4233723598626794  \n",
       "C   BH_multiple_tests                    False                    False  \n",
       "    count                                 3307                     7459  \n",
       "    log2OddsRatio        -0.021012887894985947     0.021008278769817514  \n",
       "    oddsRatio               0.9855405328614022       1.0146683691431886  \n",
       "    pvalue                    0.41335734087507       0.2156699395087413  \n",
       "D   BH_multiple_tests                    False                    False  \n",
       "    count                                  300                      826  \n",
       "    log2OddsRatio         -0.12011204015699223     -0.08571065431888877  \n",
       "    oddsRatio               0.9201161913332649       0.9423202458951588  \n",
       "    pvalue                  0.1552489357082094       0.0932352645169315  \n",
       "E   BH_multiple_tests                     True                     True  \n",
       "    count                                 1801                     5797  \n",
       "    log2OddsRatio          -0.9947688663316094      -0.6550661326299726  \n",
       "    oddsRatio               0.5018162636232523       0.6350463803009097  \n",
       "    pvalue             1.4475983413583616e-236   1.926260459508698e-297  \n",
       "F   BH_multiple_tests                     True                     True  \n",
       "    count                                  680                     2418  \n",
       "    log2OddsRatio          -0.4419147446496162      -0.3207422931494484  \n",
       "    oddsRatio               0.7361569326288107       0.8006578184100672  \n",
       "    pvalue               6.950097593186476e-17    5.957787617881626e-29  \n",
       "...                                        ...                      ...  \n",
       "S   BH_multiple_tests                     True                     True  \n",
       "    count                                16195                    27838  \n",
       "    log2OddsRatio         -0.26519243145612875      -0.5132887384911861  \n",
       "    oddsRatio               0.8320877406828754       0.7006234889338285  \n",
       "    pvalue              2.839980130830935e-126                      0.0  \n",
       "T   BH_multiple_tests                     True                     True  \n",
       "    count                                 1069                     2845  \n",
       "    log2OddsRatio          -0.8569756468163463      -0.8059098358146063  \n",
       "    oddsRatio               0.5521087432865142        0.572001233875067  \n",
       "    pvalue              8.353694011767148e-102  4.8373853804914224e-231  \n",
       "U   BH_multiple_tests                     True                     True  \n",
       "    count                                 1115                     1081  \n",
       "    log2OddsRatio           0.8727064692884771      -0.6698054488360393  \n",
       "    oddsRatio               1.8310947804169737       0.6285914487310336  \n",
       "    pvalue              3.3270251532525784e-74    7.233730857511791e-60  \n",
       "V   BH_multiple_tests                     True                     True  \n",
       "    count                                  680                     1086  \n",
       "    log2OddsRatio         -0.19524694366115125      -0.9136672361629999  \n",
       "    oddsRatio                  0.8734233807235       0.5308340295150962  \n",
       "    pvalue               0.0003747158233188343   3.301441142235305e-117  \n",
       "W   BH_multiple_tests                     True                    False  \n",
       "    count                                  152                        5  \n",
       "    log2OddsRatio            7.282937897586357      -0.7305167587498651  \n",
       "    oddsRatio                155.7337580279398       0.6026879984422204  \n",
       "    pvalue              3.333561283747108e-250           0.373554222014  \n",
       "Z   BH_multiple_tests                    False                    False  \n",
       "    count                                   17                       12  \n",
       "    log2OddsRatio           0.4581352461403243     -0.07588259206317033  \n",
       "    oddsRatio               1.3737650116549287       0.9487615231416884  \n",
       "    pvalue                 0.19494140122611237                      1.0  \n",
       "\n",
       "[115 rows x 20 columns]"
      ]
     },
     "execution_count": 314,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCOG = oddsRatioDfCOG.loc[idx['C', -1], aaTable]\n",
    "oddsRatioDfCOG = oddsRatioDfCOG.reorder_levels([1,0]).sort_index()\n",
    "oddsRatioDfCOG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Revision: We examine the COG categories A, W, Z."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     20.000000\n",
       "mean      35.150000\n",
       "std       28.593844\n",
       "min        0.000000\n",
       "25%       13.750000\n",
       "50%       30.500000\n",
       "75%       52.000000\n",
       "max      108.000000\n",
       "Name: (A, count), dtype: float64"
      ]
     },
     "execution_count": 315,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCOG.loc[idx['A', 'count'], :].astype(int).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     20.000000\n",
       "mean      19.800000\n",
       "std       34.926389\n",
       "min        0.000000\n",
       "25%        3.750000\n",
       "50%        7.500000\n",
       "75%       17.250000\n",
       "max      152.000000\n",
       "Name: (W, count), dtype: float64"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCOG.loc[idx['W', 'count'], :].astype(int).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    20.00000\n",
       "mean     25.80000\n",
       "std      21.37288\n",
       "min       3.00000\n",
       "25%      10.75000\n",
       "50%      17.00000\n",
       "75%      31.25000\n",
       "max      85.00000\n",
       "Name: (Z, count), dtype: float64"
      ]
     },
     "execution_count": 317,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCOG.loc[idx['Z', 'count'], :].astype(int).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count       20.00000\n",
       "mean     12136.90000\n",
       "std       9866.52636\n",
       "min       1226.00000\n",
       "25%       4812.00000\n",
       "50%       8820.50000\n",
       "75%      17150.25000\n",
       "max      39369.00000\n",
       "Name: (J, count), dtype: float64"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCOG.loc[idx['J', 'count'], :].astype(int).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>N</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>P</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG</th>\n",
       "      <th>observable</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <th>count</th>\n",
       "      <td>31</td>\n",
       "      <td>44</td>\n",
       "      <td>4</td>\n",
       "      <td>45</td>\n",
       "      <td>55</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "      <td>58</td>\n",
       "      <td>32</td>\n",
       "      <td>77</td>\n",
       "      <td>51</td>\n",
       "      <td>70</td>\n",
       "      <td>108</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <th>count</th>\n",
       "      <td>132</td>\n",
       "      <td>122</td>\n",
       "      <td>26</td>\n",
       "      <td>123</td>\n",
       "      <td>53</td>\n",
       "      <td>113</td>\n",
       "      <td>43</td>\n",
       "      <td>29</td>\n",
       "      <td>32</td>\n",
       "      <td>5</td>\n",
       "      <td>151</td>\n",
       "      <td>55</td>\n",
       "      <td>150</td>\n",
       "      <td>52</td>\n",
       "      <td>57</td>\n",
       "      <td>205</td>\n",
       "      <td>11</td>\n",
       "      <td>82</td>\n",
       "      <td>6</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <th>count</th>\n",
       "      <td>23106</td>\n",
       "      <td>25113</td>\n",
       "      <td>8818</td>\n",
       "      <td>10685</td>\n",
       "      <td>15793</td>\n",
       "      <td>16764</td>\n",
       "      <td>4370</td>\n",
       "      <td>8220</td>\n",
       "      <td>8215</td>\n",
       "      <td>1351</td>\n",
       "      <td>21717</td>\n",
       "      <td>10186</td>\n",
       "      <td>32091</td>\n",
       "      <td>14492</td>\n",
       "      <td>10732</td>\n",
       "      <td>21263</td>\n",
       "      <td>3779</td>\n",
       "      <td>9740</td>\n",
       "      <td>3307</td>\n",
       "      <td>7459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <th>count</th>\n",
       "      <td>4100</td>\n",
       "      <td>3289</td>\n",
       "      <td>772</td>\n",
       "      <td>2384</td>\n",
       "      <td>3159</td>\n",
       "      <td>2984</td>\n",
       "      <td>902</td>\n",
       "      <td>1664</td>\n",
       "      <td>1500</td>\n",
       "      <td>165</td>\n",
       "      <td>2711</td>\n",
       "      <td>1449</td>\n",
       "      <td>3797</td>\n",
       "      <td>1967</td>\n",
       "      <td>1272</td>\n",
       "      <td>2362</td>\n",
       "      <td>331</td>\n",
       "      <td>1549</td>\n",
       "      <td>300</td>\n",
       "      <td>826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <th>count</th>\n",
       "      <td>27666</td>\n",
       "      <td>29291</td>\n",
       "      <td>8542</td>\n",
       "      <td>14005</td>\n",
       "      <td>18266</td>\n",
       "      <td>22097</td>\n",
       "      <td>6789</td>\n",
       "      <td>11835</td>\n",
       "      <td>11928</td>\n",
       "      <td>3310</td>\n",
       "      <td>24396</td>\n",
       "      <td>11415</td>\n",
       "      <td>38870</td>\n",
       "      <td>18559</td>\n",
       "      <td>13624</td>\n",
       "      <td>28904</td>\n",
       "      <td>2794</td>\n",
       "      <td>8179</td>\n",
       "      <td>1801</td>\n",
       "      <td>5797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <th>count</th>\n",
       "      <td>8981</td>\n",
       "      <td>10731</td>\n",
       "      <td>3624</td>\n",
       "      <td>5579</td>\n",
       "      <td>7943</td>\n",
       "      <td>7641</td>\n",
       "      <td>2188</td>\n",
       "      <td>4185</td>\n",
       "      <td>3955</td>\n",
       "      <td>898</td>\n",
       "      <td>8437</td>\n",
       "      <td>3183</td>\n",
       "      <td>11376</td>\n",
       "      <td>6240</td>\n",
       "      <td>4412</td>\n",
       "      <td>8269</td>\n",
       "      <td>827</td>\n",
       "      <td>3459</td>\n",
       "      <td>680</td>\n",
       "      <td>2418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <th>count</th>\n",
       "      <td>22388</td>\n",
       "      <td>28258</td>\n",
       "      <td>5583</td>\n",
       "      <td>10609</td>\n",
       "      <td>15215</td>\n",
       "      <td>16106</td>\n",
       "      <td>5410</td>\n",
       "      <td>9879</td>\n",
       "      <td>9825</td>\n",
       "      <td>1946</td>\n",
       "      <td>19147</td>\n",
       "      <td>9371</td>\n",
       "      <td>26446</td>\n",
       "      <td>11435</td>\n",
       "      <td>8489</td>\n",
       "      <td>17683</td>\n",
       "      <td>2081</td>\n",
       "      <td>6004</td>\n",
       "      <td>1953</td>\n",
       "      <td>4466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <th>count</th>\n",
       "      <td>13301</td>\n",
       "      <td>11508</td>\n",
       "      <td>3357</td>\n",
       "      <td>6458</td>\n",
       "      <td>8254</td>\n",
       "      <td>11008</td>\n",
       "      <td>3723</td>\n",
       "      <td>5141</td>\n",
       "      <td>4745</td>\n",
       "      <td>1500</td>\n",
       "      <td>13771</td>\n",
       "      <td>7310</td>\n",
       "      <td>16657</td>\n",
       "      <td>7726</td>\n",
       "      <td>6440</td>\n",
       "      <td>12251</td>\n",
       "      <td>1018</td>\n",
       "      <td>4900</td>\n",
       "      <td>1652</td>\n",
       "      <td>4161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <th>count</th>\n",
       "      <td>12181</td>\n",
       "      <td>10414</td>\n",
       "      <td>2299</td>\n",
       "      <td>5428</td>\n",
       "      <td>7439</td>\n",
       "      <td>9509</td>\n",
       "      <td>2997</td>\n",
       "      <td>4131</td>\n",
       "      <td>4389</td>\n",
       "      <td>999</td>\n",
       "      <td>9820</td>\n",
       "      <td>4711</td>\n",
       "      <td>17212</td>\n",
       "      <td>7340</td>\n",
       "      <td>4512</td>\n",
       "      <td>10391</td>\n",
       "      <td>2027</td>\n",
       "      <td>5020</td>\n",
       "      <td>1400</td>\n",
       "      <td>2260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>J</th>\n",
       "      <th>count</th>\n",
       "      <td>21072</td>\n",
       "      <td>39369</td>\n",
       "      <td>2793</td>\n",
       "      <td>10626</td>\n",
       "      <td>19862</td>\n",
       "      <td>13572</td>\n",
       "      <td>3387</td>\n",
       "      <td>8834</td>\n",
       "      <td>8807</td>\n",
       "      <td>1226</td>\n",
       "      <td>19215</td>\n",
       "      <td>4843</td>\n",
       "      <td>29756</td>\n",
       "      <td>14233</td>\n",
       "      <td>8240</td>\n",
       "      <td>16462</td>\n",
       "      <td>3221</td>\n",
       "      <td>7559</td>\n",
       "      <td>4719</td>\n",
       "      <td>4942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <th>count</th>\n",
       "      <td>27845</td>\n",
       "      <td>19652</td>\n",
       "      <td>5937</td>\n",
       "      <td>16320</td>\n",
       "      <td>20882</td>\n",
       "      <td>23418</td>\n",
       "      <td>9073</td>\n",
       "      <td>11102</td>\n",
       "      <td>9767</td>\n",
       "      <td>2560</td>\n",
       "      <td>23944</td>\n",
       "      <td>13304</td>\n",
       "      <td>32120</td>\n",
       "      <td>12617</td>\n",
       "      <td>10717</td>\n",
       "      <td>19385</td>\n",
       "      <td>2281</td>\n",
       "      <td>5361</td>\n",
       "      <td>1460</td>\n",
       "      <td>3774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <th>count</th>\n",
       "      <td>16790</td>\n",
       "      <td>17667</td>\n",
       "      <td>4718</td>\n",
       "      <td>10715</td>\n",
       "      <td>14314</td>\n",
       "      <td>15791</td>\n",
       "      <td>6169</td>\n",
       "      <td>8240</td>\n",
       "      <td>6209</td>\n",
       "      <td>1781</td>\n",
       "      <td>16684</td>\n",
       "      <td>7035</td>\n",
       "      <td>22836</td>\n",
       "      <td>10372</td>\n",
       "      <td>9656</td>\n",
       "      <td>16109</td>\n",
       "      <td>2204</td>\n",
       "      <td>9923</td>\n",
       "      <td>1826</td>\n",
       "      <td>3522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <th>count</th>\n",
       "      <td>26846</td>\n",
       "      <td>25914</td>\n",
       "      <td>4915</td>\n",
       "      <td>10428</td>\n",
       "      <td>16337</td>\n",
       "      <td>16797</td>\n",
       "      <td>4869</td>\n",
       "      <td>12598</td>\n",
       "      <td>11809</td>\n",
       "      <td>1682</td>\n",
       "      <td>16020</td>\n",
       "      <td>11447</td>\n",
       "      <td>18445</td>\n",
       "      <td>8826</td>\n",
       "      <td>7877</td>\n",
       "      <td>17094</td>\n",
       "      <td>2164</td>\n",
       "      <td>13842</td>\n",
       "      <td>2496</td>\n",
       "      <td>6245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <th>count</th>\n",
       "      <td>5487</td>\n",
       "      <td>3108</td>\n",
       "      <td>607</td>\n",
       "      <td>1308</td>\n",
       "      <td>2632</td>\n",
       "      <td>2920</td>\n",
       "      <td>930</td>\n",
       "      <td>1556</td>\n",
       "      <td>3002</td>\n",
       "      <td>132</td>\n",
       "      <td>3726</td>\n",
       "      <td>1218</td>\n",
       "      <td>4091</td>\n",
       "      <td>2687</td>\n",
       "      <td>2341</td>\n",
       "      <td>3026</td>\n",
       "      <td>731</td>\n",
       "      <td>1975</td>\n",
       "      <td>201</td>\n",
       "      <td>716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <th>count</th>\n",
       "      <td>12422</td>\n",
       "      <td>17640</td>\n",
       "      <td>4735</td>\n",
       "      <td>6567</td>\n",
       "      <td>11093</td>\n",
       "      <td>12169</td>\n",
       "      <td>3099</td>\n",
       "      <td>7453</td>\n",
       "      <td>7742</td>\n",
       "      <td>1436</td>\n",
       "      <td>12668</td>\n",
       "      <td>5977</td>\n",
       "      <td>19050</td>\n",
       "      <td>7210</td>\n",
       "      <td>5473</td>\n",
       "      <td>12384</td>\n",
       "      <td>2307</td>\n",
       "      <td>5065</td>\n",
       "      <td>1428</td>\n",
       "      <td>3191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <th>count</th>\n",
       "      <td>21496</td>\n",
       "      <td>23554</td>\n",
       "      <td>8836</td>\n",
       "      <td>9749</td>\n",
       "      <td>13996</td>\n",
       "      <td>17255</td>\n",
       "      <td>5172</td>\n",
       "      <td>8703</td>\n",
       "      <td>9245</td>\n",
       "      <td>1383</td>\n",
       "      <td>19519</td>\n",
       "      <td>8551</td>\n",
       "      <td>25502</td>\n",
       "      <td>10288</td>\n",
       "      <td>7569</td>\n",
       "      <td>16480</td>\n",
       "      <td>2286</td>\n",
       "      <td>17244</td>\n",
       "      <td>2673</td>\n",
       "      <td>5005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <th>count</th>\n",
       "      <td>9184</td>\n",
       "      <td>6170</td>\n",
       "      <td>1696</td>\n",
       "      <td>3330</td>\n",
       "      <td>4173</td>\n",
       "      <td>6026</td>\n",
       "      <td>2189</td>\n",
       "      <td>2693</td>\n",
       "      <td>2808</td>\n",
       "      <td>565</td>\n",
       "      <td>7367</td>\n",
       "      <td>4114</td>\n",
       "      <td>10862</td>\n",
       "      <td>4324</td>\n",
       "      <td>2748</td>\n",
       "      <td>5302</td>\n",
       "      <td>735</td>\n",
       "      <td>1880</td>\n",
       "      <td>1608</td>\n",
       "      <td>1018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <th>count</th>\n",
       "      <td>131264</td>\n",
       "      <td>117304</td>\n",
       "      <td>32094</td>\n",
       "      <td>64651</td>\n",
       "      <td>84662</td>\n",
       "      <td>91939</td>\n",
       "      <td>32553</td>\n",
       "      <td>55003</td>\n",
       "      <td>54336</td>\n",
       "      <td>10455</td>\n",
       "      <td>96477</td>\n",
       "      <td>66737</td>\n",
       "      <td>123076</td>\n",
       "      <td>55868</td>\n",
       "      <td>45118</td>\n",
       "      <td>92903</td>\n",
       "      <td>11932</td>\n",
       "      <td>53166</td>\n",
       "      <td>16195</td>\n",
       "      <td>27838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <th>count</th>\n",
       "      <td>18444</td>\n",
       "      <td>13692</td>\n",
       "      <td>4928</td>\n",
       "      <td>10445</td>\n",
       "      <td>14330</td>\n",
       "      <td>18448</td>\n",
       "      <td>6659</td>\n",
       "      <td>7807</td>\n",
       "      <td>7544</td>\n",
       "      <td>1769</td>\n",
       "      <td>17335</td>\n",
       "      <td>10662</td>\n",
       "      <td>24920</td>\n",
       "      <td>10244</td>\n",
       "      <td>7774</td>\n",
       "      <td>13761</td>\n",
       "      <td>1723</td>\n",
       "      <td>5406</td>\n",
       "      <td>1069</td>\n",
       "      <td>2845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U</th>\n",
       "      <th>count</th>\n",
       "      <td>7722</td>\n",
       "      <td>7977</td>\n",
       "      <td>1550</td>\n",
       "      <td>2938</td>\n",
       "      <td>4537</td>\n",
       "      <td>4585</td>\n",
       "      <td>1367</td>\n",
       "      <td>3365</td>\n",
       "      <td>4231</td>\n",
       "      <td>473</td>\n",
       "      <td>4570</td>\n",
       "      <td>3932</td>\n",
       "      <td>4936</td>\n",
       "      <td>2672</td>\n",
       "      <td>2317</td>\n",
       "      <td>4240</td>\n",
       "      <td>830</td>\n",
       "      <td>2821</td>\n",
       "      <td>1115</td>\n",
       "      <td>1081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <th>count</th>\n",
       "      <td>6588</td>\n",
       "      <td>5283</td>\n",
       "      <td>1623</td>\n",
       "      <td>3765</td>\n",
       "      <td>7901</td>\n",
       "      <td>5972</td>\n",
       "      <td>1957</td>\n",
       "      <td>2983</td>\n",
       "      <td>3131</td>\n",
       "      <td>480</td>\n",
       "      <td>5128</td>\n",
       "      <td>2786</td>\n",
       "      <td>9774</td>\n",
       "      <td>4350</td>\n",
       "      <td>2993</td>\n",
       "      <td>4618</td>\n",
       "      <td>651</td>\n",
       "      <td>1688</td>\n",
       "      <td>680</td>\n",
       "      <td>1086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <th>count</th>\n",
       "      <td>21</td>\n",
       "      <td>67</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>152</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z</th>\n",
       "      <th>count</th>\n",
       "      <td>61</td>\n",
       "      <td>47</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>53</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>31</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>85</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     R       K      H      D      E      S      T      N  \\\n",
       "COG observable                                                             \n",
       "A   count           31      44      4     45     55     30     15     26   \n",
       "B   count          132     122     26    123     53    113     43     29   \n",
       "C   count        23106   25113   8818  10685  15793  16764   4370   8220   \n",
       "D   count         4100    3289    772   2384   3159   2984    902   1664   \n",
       "E   count        27666   29291   8542  14005  18266  22097   6789  11835   \n",
       "F   count         8981   10731   3624   5579   7943   7641   2188   4185   \n",
       "G   count        22388   28258   5583  10609  15215  16106   5410   9879   \n",
       "H   count        13301   11508   3357   6458   8254  11008   3723   5141   \n",
       "I   count        12181   10414   2299   5428   7439   9509   2997   4131   \n",
       "J   count        21072   39369   2793  10626  19862  13572   3387   8834   \n",
       "K   count        27845   19652   5937  16320  20882  23418   9073  11102   \n",
       "L   count        16790   17667   4718  10715  14314  15791   6169   8240   \n",
       "M   count        26846   25914   4915  10428  16337  16797   4869  12598   \n",
       "N   count         5487    3108    607   1308   2632   2920    930   1556   \n",
       "O   count        12422   17640   4735   6567  11093  12169   3099   7453   \n",
       "P   count        21496   23554   8836   9749  13996  17255   5172   8703   \n",
       "Q   count         9184    6170   1696   3330   4173   6026   2189   2693   \n",
       "S   count       131264  117304  32094  64651  84662  91939  32553  55003   \n",
       "T   count        18444   13692   4928  10445  14330  18448   6659   7807   \n",
       "U   count         7722    7977   1550   2938   4537   4585   1367   3365   \n",
       "V   count         6588    5283   1623   3765   7901   5972   1957   2983   \n",
       "W   count           21      67      3     12     12      7      4     38   \n",
       "Z   count           61      47     17     10     53     11     16     31   \n",
       "\n",
       "                    Q      C      G      P       A      V      I      L  \\\n",
       "COG observable                                                            \n",
       "A   count          25      3     58     32      77     51     70    108   \n",
       "B   count          32      5    151     55     150     52     57    205   \n",
       "C   count        8215   1351  21717  10186   32091  14492  10732  21263   \n",
       "D   count        1500    165   2711   1449    3797   1967   1272   2362   \n",
       "E   count       11928   3310  24396  11415   38870  18559  13624  28904   \n",
       "F   count        3955    898   8437   3183   11376   6240   4412   8269   \n",
       "G   count        9825   1946  19147   9371   26446  11435   8489  17683   \n",
       "H   count        4745   1500  13771   7310   16657   7726   6440  12251   \n",
       "I   count        4389    999   9820   4711   17212   7340   4512  10391   \n",
       "J   count        8807   1226  19215   4843   29756  14233   8240  16462   \n",
       "K   count        9767   2560  23944  13304   32120  12617  10717  19385   \n",
       "L   count        6209   1781  16684   7035   22836  10372   9656  16109   \n",
       "M   count       11809   1682  16020  11447   18445   8826   7877  17094   \n",
       "N   count        3002    132   3726   1218    4091   2687   2341   3026   \n",
       "O   count        7742   1436  12668   5977   19050   7210   5473  12384   \n",
       "P   count        9245   1383  19519   8551   25502  10288   7569  16480   \n",
       "Q   count        2808    565   7367   4114   10862   4324   2748   5302   \n",
       "S   count       54336  10455  96477  66737  123076  55868  45118  92903   \n",
       "T   count        7544   1769  17335  10662   24920  10244   7774  13761   \n",
       "U   count        4231    473   4570   3932    4936   2672   2317   4240   \n",
       "V   count        3131    480   5128   2786    9774   4350   2993   4618   \n",
       "W   count           9      1      7      0       8      4      1     16   \n",
       "Z   count          30      3     17     85      26      7      8     32   \n",
       "\n",
       "                    M      F      W      Y  \n",
       "COG observable                              \n",
       "A   count           3     16      0     10  \n",
       "B   count          11     82      6     43  \n",
       "C   count        3779   9740   3307   7459  \n",
       "D   count         331   1549    300    826  \n",
       "E   count        2794   8179   1801   5797  \n",
       "F   count         827   3459    680   2418  \n",
       "G   count        2081   6004   1953   4466  \n",
       "H   count        1018   4900   1652   4161  \n",
       "I   count        2027   5020   1400   2260  \n",
       "J   count        3221   7559   4719   4942  \n",
       "K   count        2281   5361   1460   3774  \n",
       "L   count        2204   9923   1826   3522  \n",
       "M   count        2164  13842   2496   6245  \n",
       "N   count         731   1975    201    716  \n",
       "O   count        2307   5065   1428   3191  \n",
       "P   count        2286  17244   2673   5005  \n",
       "Q   count         735   1880   1608   1018  \n",
       "S   count       11932  53166  16195  27838  \n",
       "T   count        1723   5406   1069   2845  \n",
       "U   count         830   2821   1115   1081  \n",
       "V   count         651   1688    680   1086  \n",
       "W   count           3     26    152      5  \n",
       "Z   count           4     29     17     12  "
      ]
     },
     "execution_count": 319,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsRatioDfCOG.loc[idx[:, 'count'], :].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggResDf3 = pd.read_csv('~/multiCDS_eggNOG_taxa.csv.gz', compression='gzip', nrows=None, index_col=0)\n",
    "eggResDf3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['proteinGI', 'species_name_x', 'genome_accession', 'DNASeq',\n",
       "       'stopCodon', 'DNA3UTRSeq', 'codonTableBio', 'proteinSeq',\n",
       "       'seed_eggNOG_ortholog', 'seed_ortholog_evalue', 'seed_ortholog_score',\n",
       "       'predicted_gene_name', 'GO_terms', 'KEGG_pathways',\n",
       "       'Annotation_tax_scope', 'OGs', 'COG cat', 'eggNOG annot', 'taxid_x',\n",
       "       'bestOG', 'bestOG_evalue', 'bestOG_score', 'taxid_y', 'organism_name',\n",
       "       'compressedGenomeFilename', 'class_name', 'class_taxid', 'family_name',\n",
       "       'family_taxid', 'genus_name', 'genus_taxid', 'order_name',\n",
       "       'order_taxid', 'phylum_name', 'phylum_taxid', 'species group_name',\n",
       "       'species group_taxid', 'species subgroup_name',\n",
       "       'species subgroup_taxid', 'species_name_y', 'species_taxid',\n",
       "       'subspecies_name', 'subspecies_taxid', 'superkingdom_name',\n",
       "       'superkingdom_taxid'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 320,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggResDf3.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['S', 'K', 'E', 'G', 'L', 'O', 'C', 'M', 'H', 'T', 'I', 'V', 'K, T',\n",
       "       'J', 'D', 'P', 'Q', 'F', 'N', nan, 'U', 'N, U', 'E, P', 'O, U',\n",
       "       'E, H', 'E, G', 'N, T', 'Q, T', 'E, T', 'H, J', 'G, M', 'N, O, U',\n",
       "       'K, O', 'O, P', 'C, O', 'F, J', 'M, U', 'F, G', 'E, I', 'F, P',\n",
       "       'C, H', 'C, P', 'I, Q', 'B, Q', 'E, G, P', 'A', 'F, L', 'G, K',\n",
       "       'M, P', 'H, I', 'C, Q', 'N, P, T, U', 'D, J', 'O, T', 'F, H',\n",
       "       'C, K', 'E, Q', 'B', 'U, W', 'K, L, T', 'K, L', 'E, M', 'G, T',\n",
       "       'C, G', 'P, T', 'N, O, T', 'I, M', 'P, Q', 'G, K, T', 'O, P, U',\n",
       "       'L, P', 'J, M', 'C, E, H', 'B, K', 'E, F', 'H, P', 'L, U', 'I, J',\n",
       "       'C, E', 'D, Z', 'I, O', 'G, O', 'G, H', 'E, K', 'Z', 'G, J, M',\n",
       "       'C, T', 'O, Q', 'D, M', 'K, Q', 'G, V', 'H, Q', 'I, P', 'C, I',\n",
       "       'C, E, K', 'F, Q', 'H, I, Q', 'K, P', 'G, U', 'D, T, Z', 'G, I',\n",
       "       'E, J', 'E, J, M', 'C, F, H', 'H, K'], dtype=object)"
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggResDf3['COG cat'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "396"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = eggResDf3[~eggResDf3['COG cat'].isnull()]\n",
    "df = df[df['COG cat'].str.contains('W')]\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "516"
      ]
     },
     "execution_count": 322,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = eggResDf3[~eggResDf3['COG cat'].isnull()]\n",
    "df = df[df['COG cat'].str.contains('Z')]\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "703"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = eggResDf3[~eggResDf3['COG cat'].isnull()]\n",
    "df = df[df['COG cat'].str.contains('A')]\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1490"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = eggResDf3[~eggResDf3['COG cat'].isnull()]\n",
    "df = df[df['COG cat'].str.contains('B')]\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "242738"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = eggResDf3[~eggResDf3['COG cat'].isnull()]\n",
    "df = df[df['COG cat'].str.contains('J')]\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "143"
      ]
     },
     "execution_count": 323,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df['seed_eggNOG_ortholog'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seed_eggNOG_ortholog\n",
       "1032480.MLP_24980                regulator of chromosome condensation, RCC1\n",
       "1032480.MLP_48790                         Beta-lactamase inhibitory protein\n",
       "1034807.FBFL15_1456              regulator of chromosome condensation, RCC1\n",
       "1036673.KNP414_02375             regulator of chromosome condensation, RCC1\n",
       "110662.Syncc9605_0770        signal transduction protein with EFhand domain\n",
       "1112204.GPOL_c31870                       Beta-lactamase inhibitory protein\n",
       "134676.ACPL_3856                 regulator of chromosome condensation, RCC1\n",
       "159087.Daro_1010                 regulator of chromosome condensation, RCC1\n",
       "177416.FTT_0243                  regulator of chromosome condensation, RCC1\n",
       "177437.HRM2_09590                    ATP-dependent protease La (LON) domain\n",
       "187272.Mlg_2687                  regulator of chromosome condensation, RCC1\n",
       "187272.Mlg_2688                  regulator of chromosome condensation, RCC1\n",
       "189518.LA_2498                   regulator of chromosome condensation, RCC1\n",
       "204773.HEAR2211                  regulator of chromosome condensation, RCC1\n",
       "205922.Pfl01_1556         Inherit from COG: regulator of chromosome cond...\n",
       "226900.BC1772                             Beta-lactamase inhibitory protein\n",
       "227377.CBU_1217                  regulator of chromosome condensation, RCC1\n",
       "243231.GSU1251                   regulator of chromosome condensation, RCC1\n",
       "243231.GSU2272                   regulator of chromosome condensation, RCC1\n",
       "246197.MXAN_0889                 regulator of chromosome condensation, RCC1\n",
       "264462.Bd1137                    regulator of chromosome condensation, RCC1\n",
       "264462.Bd1921                    regulator of chromosome condensation, RCC1\n",
       "264462.Bd2144                        ATP-dependent protease La (LON) domain\n",
       "264462.Bd2644                    regulator of chromosome condensation, RCC1\n",
       "264462.Bd3390                    regulator of chromosome condensation, RCC1\n",
       "269798.CHU_0430                  regulator of chromosome condensation, RCC1\n",
       "269798.CHU_1101                  regulator of chromosome condensation, RCC1\n",
       "272563.CD1374                    regulator of chromosome condensation, RCC1\n",
       "272624.lpg2587            Inherit from COG: regulator of chromosome cond...\n",
       "281090.Lxx21640                  regulator of chromosome condensation, RCC1\n",
       "                                                ...                        \n",
       "637905.SVI_1082           Inherit from COG: regulator of chromosome cond...\n",
       "637910.ROD_41561                 regulator of chromosome condensation, RCC1\n",
       "661367.LLO_1727           Inherit from COG: regulator of chromosome cond...\n",
       "663278.Ethha_2466                regulator of chromosome condensation, RCC1\n",
       "665571.STHERM_c20190             regulator of chromosome condensation, RCC1\n",
       "676032.FN3523_0062               regulator of chromosome condensation, RCC1\n",
       "680198.SCAB_3491                          Beta-lactamase inhibitory protein\n",
       "700598.Niako_3726                regulator of chromosome condensation, RCC1\n",
       "720554.Clocl_0711                         Beta-lactamase inhibitory protein\n",
       "720554.Clocl_2046                regulator of chromosome condensation, RCC1\n",
       "720554.Clocl_3248                regulator of chromosome condensation, RCC1\n",
       "749927.AMED_1770                          Beta-lactamase inhibitory protein\n",
       "755732.Fluta_1725                                Protein metal binding site\n",
       "756067.MicvaDRAFT_1915       signal transduction protein with EFhand domain\n",
       "760011.Spico_0052                regulator of chromosome condensation, RCC1\n",
       "760011.Spico_0060                regulator of chromosome condensation, RCC1\n",
       "760011.Spico_0332                regulator of chromosome condensation, RCC1\n",
       "760011.Spico_1525                regulator of chromosome condensation, RCC1\n",
       "765420.OSCT_2502                 regulator of chromosome condensation, RCC1\n",
       "856793.MICA_1467                 regulator of chromosome condensation, RCC1\n",
       "886377.Murru_2512                regulator of chromosome condensation, RCC1\n",
       "886882.PPSC2_c5325               regulator of chromosome condensation, RCC1\n",
       "926562.Oweho_0316                regulator of chromosome condensation, RCC1\n",
       "926562.Oweho_2413         Inherit from COG: regulator of chromosome cond...\n",
       "944547.ABLL_2339                 regulator of chromosome condensation, RCC1\n",
       "96561.Dole_2171           Inherit from COG: regulator of chromosome cond...\n",
       "979556.MTES_0431                 regulator of chromosome condensation, RCC1\n",
       "979556.MTES_3083                 regulator of chromosome condensation, RCC1\n",
       "979556.MTES_3324                 regulator of chromosome condensation, RCC1\n",
       "985665.HPL003_12310              regulator of chromosome condensation, RCC1\n",
       "Name: eggNOG annot, Length: 143, dtype: object"
      ]
     },
     "execution_count": 324,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('seed_eggNOG_ortholog')['eggNOG annot'].first()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df['proteinSeq'].map(lambda x: x[-15:]);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "regulator of chromosome condensation, RCC1                419\n",
       "Beta-lactamase inhibitory protein                          39\n",
       "Protein metal binding site                                 32\n",
       "Inherit from COG: regulator of chromosome condensation     13\n",
       "ATP-dependent protease La (LON) domain                     10\n",
       "signal transduction protein with EFhand domain              3\n",
       "Name: eggNOG annot, dtype: int64"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['eggNOG annot'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "85\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seed_eggNOG_ortholog</th>\n",
       "      <th>eggNOG annot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_953321.1</th>\n",
       "      <td>243231.GSU2272</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012685496.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_020477836.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011551010.1</th>\n",
       "      <td>246197.MXAN_0889</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011670363.1</th>\n",
       "      <td>355276.LBL_1814</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011905864.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011955209.1</th>\n",
       "      <td>357808.RoseRS_0421</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_049818895.1</th>\n",
       "      <td>351605.Gura_1475</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012181743.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012181742.1</th>\n",
       "      <td>391037.Sare_1538</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_041262720.1</th>\n",
       "      <td>391625.PPSIR1_08012</td>\n",
       "      <td>Inherit from COG: regulator of chromosome cond...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012632399.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012825436.1</th>\n",
       "      <td>502025.Hoch_0168</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012238144.1</th>\n",
       "      <td>448385.sce5513</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012241089.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_044965556.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012236818.1</th>\n",
       "      <td>448385.sce4185</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012998295.1</th>\n",
       "      <td>680198.SCAB_3491</td>\n",
       "      <td>Beta-lactamase inhibitory protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_712679.1</th>\n",
       "      <td>189518.LA_2498</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_002613990.1</th>\n",
       "      <td>378806.STAUR_0858</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_002611633.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_002615425.1</th>\n",
       "      <td>378806.STAUR_7770</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_037583321.1</th>\n",
       "      <td>378806.STAUR_0333</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_015719079.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_043922671.1</th>\n",
       "      <td>545695.TREAZ_2091</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_013935945.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_049789105.1</th>\n",
       "      <td>886377.Murru_2512</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_014371271.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_014371068.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_052312925.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_053235389.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_052429679.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_052138843.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_052138758.1</th>\n",
       "      <td>378806.STAUR_4644</td>\n",
       "      <td>Protein metal binding site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_045478279.1</th>\n",
       "      <td>448385.sce4185</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_045094448.1</th>\n",
       "      <td>439235.Dalk_3426</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_045105817.1</th>\n",
       "      <td>439235.Dalk_3426</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_053066110.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_053066142.1</th>\n",
       "      <td>483219.LILAB_10430</td>\n",
       "      <td>Protein metal binding site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_047857536.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_063796971.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_047855383.1</th>\n",
       "      <td>378806.STAUR_4644</td>\n",
       "      <td>Protein metal binding site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_053067032.1</th>\n",
       "      <td>483219.LILAB_10430</td>\n",
       "      <td>Protein metal binding site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_063796878.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_047861464.1</th>\n",
       "      <td>246197.MXAN_0889</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_063796966.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_053066201.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050434564.1</th>\n",
       "      <td>246197.MXAN_0889</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050434191.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050430781.1</th>\n",
       "      <td>246197.MXAN_0889</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_052594852.1</th>\n",
       "      <td>383372.Rcas_0264</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050410728.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050726805.1</th>\n",
       "      <td>448385.sce4185</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050725498.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050725976.1</th>\n",
       "      <td>448385.sce4185</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050726232.1</th>\n",
       "      <td>448385.sce4185</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_051905493.1</th>\n",
       "      <td>483219.LILAB_04265</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_054294260.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_003045689.1</th>\n",
       "      <td>498848.TaqDRAFT_5368</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_003046056.1</th>\n",
       "      <td>483219.LILAB_18585</td>\n",
       "      <td>regulator of chromosome condensation, RCC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seed_eggNOG_ortholog  \\\n",
       "refSeqProteinId                         \n",
       "NP_953321.1            243231.GSU2272   \n",
       "WP_012685496.1     483219.LILAB_18585   \n",
       "WP_020477836.1     483219.LILAB_18585   \n",
       "WP_011551010.1       246197.MXAN_0889   \n",
       "WP_011670363.1        355276.LBL_1814   \n",
       "WP_011905864.1     483219.LILAB_18585   \n",
       "WP_011955209.1     357808.RoseRS_0421   \n",
       "WP_049818895.1       351605.Gura_1475   \n",
       "WP_012181743.1     483219.LILAB_18585   \n",
       "WP_012181742.1       391037.Sare_1538   \n",
       "WP_041262720.1    391625.PPSIR1_08012   \n",
       "WP_012632399.1     483219.LILAB_18585   \n",
       "WP_012825436.1       502025.Hoch_0168   \n",
       "WP_012238144.1         448385.sce5513   \n",
       "WP_012241089.1     483219.LILAB_04265   \n",
       "WP_044965556.1     483219.LILAB_18585   \n",
       "WP_012236818.1         448385.sce4185   \n",
       "WP_012998295.1       680198.SCAB_3491   \n",
       "NP_712679.1            189518.LA_2498   \n",
       "WP_002613990.1      378806.STAUR_0858   \n",
       "WP_002611633.1     483219.LILAB_18585   \n",
       "WP_002615425.1      378806.STAUR_7770   \n",
       "WP_037583321.1      378806.STAUR_0333   \n",
       "WP_015719079.1     483219.LILAB_04265   \n",
       "WP_043922671.1      545695.TREAZ_2091   \n",
       "WP_013935945.1     483219.LILAB_04265   \n",
       "WP_049789105.1      886377.Murru_2512   \n",
       "WP_014371271.1     483219.LILAB_18585   \n",
       "WP_014371068.1     483219.LILAB_18585   \n",
       "WP_052312925.1     483219.LILAB_18585   \n",
       "...                               ...   \n",
       "WP_053235389.1     483219.LILAB_18585   \n",
       "WP_052429679.1     483219.LILAB_04265   \n",
       "WP_052138843.1     483219.LILAB_04265   \n",
       "WP_052138758.1      378806.STAUR_4644   \n",
       "WP_045478279.1         448385.sce4185   \n",
       "WP_045094448.1       439235.Dalk_3426   \n",
       "WP_045105817.1       439235.Dalk_3426   \n",
       "WP_053066110.1     483219.LILAB_18585   \n",
       "WP_053066142.1     483219.LILAB_10430   \n",
       "WP_047857536.1     483219.LILAB_04265   \n",
       "WP_063796971.1     483219.LILAB_04265   \n",
       "WP_047855383.1      378806.STAUR_4644   \n",
       "WP_053067032.1     483219.LILAB_10430   \n",
       "WP_063796878.1     483219.LILAB_18585   \n",
       "WP_047861464.1       246197.MXAN_0889   \n",
       "WP_063796966.1     483219.LILAB_18585   \n",
       "WP_053066201.1     483219.LILAB_18585   \n",
       "WP_050434564.1       246197.MXAN_0889   \n",
       "WP_050434191.1     483219.LILAB_18585   \n",
       "WP_050430781.1       246197.MXAN_0889   \n",
       "WP_052594852.1       383372.Rcas_0264   \n",
       "WP_050410728.1     483219.LILAB_18585   \n",
       "WP_050726805.1         448385.sce4185   \n",
       "WP_050725498.1     483219.LILAB_18585   \n",
       "WP_050725976.1         448385.sce4185   \n",
       "WP_050726232.1         448385.sce4185   \n",
       "WP_051905493.1     483219.LILAB_04265   \n",
       "WP_054294260.1     483219.LILAB_18585   \n",
       "WP_003045689.1   498848.TaqDRAFT_5368   \n",
       "WP_003046056.1     483219.LILAB_18585   \n",
       "\n",
       "                                                      eggNOG annot  \n",
       "refSeqProteinId                                                     \n",
       "NP_953321.1             regulator of chromosome condensation, RCC1  \n",
       "WP_012685496.1          regulator of chromosome condensation, RCC1  \n",
       "WP_020477836.1          regulator of chromosome condensation, RCC1  \n",
       "WP_011551010.1          regulator of chromosome condensation, RCC1  \n",
       "WP_011670363.1          regulator of chromosome condensation, RCC1  \n",
       "WP_011905864.1          regulator of chromosome condensation, RCC1  \n",
       "WP_011955209.1          regulator of chromosome condensation, RCC1  \n",
       "WP_049818895.1          regulator of chromosome condensation, RCC1  \n",
       "WP_012181743.1          regulator of chromosome condensation, RCC1  \n",
       "WP_012181742.1          regulator of chromosome condensation, RCC1  \n",
       "WP_041262720.1   Inherit from COG: regulator of chromosome cond...  \n",
       "WP_012632399.1          regulator of chromosome condensation, RCC1  \n",
       "WP_012825436.1          regulator of chromosome condensation, RCC1  \n",
       "WP_012238144.1          regulator of chromosome condensation, RCC1  \n",
       "WP_012241089.1          regulator of chromosome condensation, RCC1  \n",
       "WP_044965556.1          regulator of chromosome condensation, RCC1  \n",
       "WP_012236818.1          regulator of chromosome condensation, RCC1  \n",
       "WP_012998295.1                   Beta-lactamase inhibitory protein  \n",
       "NP_712679.1             regulator of chromosome condensation, RCC1  \n",
       "WP_002613990.1          regulator of chromosome condensation, RCC1  \n",
       "WP_002611633.1          regulator of chromosome condensation, RCC1  \n",
       "WP_002615425.1          regulator of chromosome condensation, RCC1  \n",
       "WP_037583321.1          regulator of chromosome condensation, RCC1  \n",
       "WP_015719079.1          regulator of chromosome condensation, RCC1  \n",
       "WP_043922671.1          regulator of chromosome condensation, RCC1  \n",
       "WP_013935945.1          regulator of chromosome condensation, RCC1  \n",
       "WP_049789105.1          regulator of chromosome condensation, RCC1  \n",
       "WP_014371271.1          regulator of chromosome condensation, RCC1  \n",
       "WP_014371068.1          regulator of chromosome condensation, RCC1  \n",
       "WP_052312925.1          regulator of chromosome condensation, RCC1  \n",
       "...                                                            ...  \n",
       "WP_053235389.1          regulator of chromosome condensation, RCC1  \n",
       "WP_052429679.1          regulator of chromosome condensation, RCC1  \n",
       "WP_052138843.1          regulator of chromosome condensation, RCC1  \n",
       "WP_052138758.1                          Protein metal binding site  \n",
       "WP_045478279.1          regulator of chromosome condensation, RCC1  \n",
       "WP_045094448.1          regulator of chromosome condensation, RCC1  \n",
       "WP_045105817.1          regulator of chromosome condensation, RCC1  \n",
       "WP_053066110.1          regulator of chromosome condensation, RCC1  \n",
       "WP_053066142.1                          Protein metal binding site  \n",
       "WP_047857536.1          regulator of chromosome condensation, RCC1  \n",
       "WP_063796971.1          regulator of chromosome condensation, RCC1  \n",
       "WP_047855383.1                          Protein metal binding site  \n",
       "WP_053067032.1                          Protein metal binding site  \n",
       "WP_063796878.1          regulator of chromosome condensation, RCC1  \n",
       "WP_047861464.1          regulator of chromosome condensation, RCC1  \n",
       "WP_063796966.1          regulator of chromosome condensation, RCC1  \n",
       "WP_053066201.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050434564.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050434191.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050430781.1          regulator of chromosome condensation, RCC1  \n",
       "WP_052594852.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050410728.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050726805.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050725498.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050725976.1          regulator of chromosome condensation, RCC1  \n",
       "WP_050726232.1          regulator of chromosome condensation, RCC1  \n",
       "WP_051905493.1          regulator of chromosome condensation, RCC1  \n",
       "WP_054294260.1          regulator of chromosome condensation, RCC1  \n",
       "WP_003045689.1          regulator of chromosome condensation, RCC1  \n",
       "WP_003046056.1          regulator of chromosome condensation, RCC1  \n",
       "\n",
       "[85 rows x 2 columns]"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df[df['proteinSeq'].map(lambda x: x[-1] in 'P')]\n",
    "print(len(df2))\n",
    "df2[['seed_eggNOG_ortholog', 'eggNOG annot']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df2['seed_eggNOG_ortholog'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seed_eggNOG_ortholog\n",
       "398527.Bphyt_3373         36\n",
       "887327.HMPREF0476_1808     8\n",
       "667128.HMPREF0621_1251     8\n",
       "205914.HS_0209             6\n",
       "937774.TEQUI_0952          5\n",
       "dtype: int64"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.groupby('seed_eggNOG_ortholog').size().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfCOG.to_csv(analysisEggnogPath / 'taxaId_2_subset_all_oddsRatioAADf_COG_groups.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfCOG.loc[idx[:, 'log2OddsRatio'], :].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = oddsRatioDfCOG.copy()\n",
    "plotData = plotData.xs('log2OddsRatio', level='observable')\n",
    "plotData = plotData.astype(float)\n",
    "print(len(plotData))\n",
    "plotData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskData = oddsRatioDfCOG.xs('BH_multiple_tests', level='observable')\n",
    "# Mask both rejected cases and null values\n",
    "maskData = (maskData != True) & (maskData != 'True')\n",
    "maskData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vmax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colorHeatmapLine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colorAxis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_id'] == 2].iloc[0]\n",
    "taxonGroup\n",
    "\n",
    "print(\"taxonGroupRef:\", taxonGroup)\n",
    "taxonRank = taxonGroup['rank_name']\n",
    "taxonName = taxonGroup['taxon_name']\n",
    "taxonId = int(taxonGroup['taxon_id'])\n",
    "\n",
    "# Select genomes in the taxon group\n",
    "taxonAssemblyDf = assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf[taxonRank + '_taxid'] == taxonId]\n",
    "taxonAssemblyAccessionList = list(taxonAssemblyDf.index)\n",
    "\n",
    "taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "print(\"output path:\", str(analysisCtermDataPath / taxonSuffix))\n",
    "suffix = taxonSuffix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "width = 3\n",
    "\n",
    "linewidth1 = 0.5\n",
    "aspectratio = 0.7\n",
    "figsizeCoeff = 0.5\n",
    "figsize = (width, (width/aspectratio))\n",
    "\n",
    "fig, (ax,cbar_ax) = plt.subplots(1, 2, figsize=figsize)\n",
    "# main axes\n",
    "ax.set_position([0, 0, 1, 1])\n",
    "ax.set_facecolor(colorBackground1)\n",
    "\n",
    "# Create axes for the colorbar for more precise positioning\n",
    "# vertical bar\n",
    "cbar_aspect = 10\n",
    "cbar_length = 0.3\n",
    "cbar_ax.set_position([0.2, -0.3 - cbar_length, cbar_length/cbar_aspect, cbar_length])\n",
    "\n",
    "cbarLabel = '$\\log_2$(odds ratio)'\n",
    "ax = seaborn.heatmap(plotData, square=True, mask=maskData, ax=ax, cmap=cmap,\n",
    "                     cbar_ax=cbar_ax,\n",
    "                     cbar_kws={\"label\":cbarLabel},\n",
    "                     vmin=-vmax, vmax=vmax,\n",
    "                     xticklabels=True, yticklabels=True,\n",
    "                     linewidth=linewidth1, linecolor=colorHeatmapLine\n",
    "                     )\n",
    "for _, spine in ax.spines.items():\n",
    "    spine.set_visible(True)\n",
    "    # spine.set_color('0.8')\n",
    "    spine.set_linewidth(0.4)\n",
    "tickLength = FontProperties(size='small').get_size()/4\n",
    "cbar_ax.tick_params(axis='y', length=tickLength, color=colorAxis)\n",
    "ax.tick_params(axis='y', labelleft='on', labelright='on')\n",
    "if len(ax.xaxis.get_ticklabels()[0].get_text()) > 2:\n",
    "    ax.xaxis.set_tick_params(rotation=90)\n",
    "ax.yaxis.set_tick_params(rotation=0)\n",
    "family = 'Liberation Mono'\n",
    "ticks_font = FontProperties(family=family, size='small')\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontproperties(ticks_font)\n",
    "    tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "\n",
    "ax.set_xlabel('amino acid at C-terminal')\n",
    "ax.set_ylabel('COG category')\n",
    "\n",
    "filenamePlot = analysisEggnogPath / ('taxaId_2_subset_all_oddsRatioAADf_COG_groups_heatmap')\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + \".png\", dpi=500, bbox_inches=\"tight\")\n",
    "fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p.eggNOGPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(str(p.eggNOGPath / 'COG_functional_categories.txt')) as f:\n",
    "    text = f.read()\n",
    "    COGCategoriesDict = dict(re.findall(r'\\[([A-Z])\\]\\s+(.*)\\s+[$\\n]', text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "COGCategoriesDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted(COGCategoriesDict.items())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "width = 3\n",
    "aspectratio = 0.7\n",
    "figsize = (width, (width/aspectratio))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=figsize)\n",
    "ax.axis('off')\n",
    "\n",
    "cellText = sorted(COGCategoriesDict.items())\n",
    "the_table = plt.table(cellText=cellText,\n",
    "                      colLabels=['', 'description'],\n",
    "                      colWidths=[0.1, 0.9],\n",
    "                      loc='top')\n",
    "\n",
    "filenamePlot = p.analysisEggnogPath / ('COG_groups_description')\n",
    "print(filenamePlot)\n",
    "# fig.savefig(str(filenamePlot) + \".png\", dpi=500, bbox_inches=\"tight\")\n",
    "# fig.savefig(str(filenamePlot) + \".svg\", dpi=500, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinGI</th>\n",
       "      <th>species_name_x</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>DNASeq</th>\n",
       "      <th>stopCodon</th>\n",
       "      <th>DNA3UTRSeq</th>\n",
       "      <th>codonTableBio</th>\n",
       "      <th>proteinSeq</th>\n",
       "      <th>seed_eggNOG_ortholog</th>\n",
       "      <th>seed_ortholog_evalue</th>\n",
       "      <th>...</th>\n",
       "      <th>species group_name</th>\n",
       "      <th>species group_taxid</th>\n",
       "      <th>species subgroup_name</th>\n",
       "      <th>species subgroup_taxid</th>\n",
       "      <th>species_name_y</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>subspecies_name</th>\n",
       "      <th>subspecies_taxid</th>\n",
       "      <th>superkingdom_name</th>\n",
       "      <th>superkingdom_taxid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>refSeqProteinId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_417846.1</th>\n",
       "      <td>16131265</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>GCF_000005845.2</td>\n",
       "      <td>ATGAAGAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGCAAGTATC...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TTCTCAAGGAGAAGCGGATGAAACAGTATT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...</td>\n",
       "      <td>155864.Z4740</td>\n",
       "      <td>4.600000e-195</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_232254.1</th>\n",
       "      <td>15642621</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>GCF_000006745.1</td>\n",
       "      <td>ATGAAAAAGCAGCGCGCCTTTCTGAAATGGGCTGGGGGTAAATATA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TCTTGTGGGCTGAATGGTGGATTTCCTCTC</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKQRAFLKWAGGKYSLVEDIQRHLPEARELVEPFVGAGSVFLNTD...</td>\n",
       "      <td>243277.VC2626</td>\n",
       "      <td>1.800000e-192</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>666.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_709160.1</th>\n",
       "      <td>24114650</td>\n",
       "      <td>Shigella flexneri 2a str. 301</td>\n",
       "      <td>GCF_000006925.2</td>\n",
       "      <td>ATGAAGAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGCAAGTATC...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TTCTCAAGGAGAAGCGGATGAAACAGTATT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...</td>\n",
       "      <td>198214.SF3405</td>\n",
       "      <td>3.600000e-195</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shigella flexneri</td>\n",
       "      <td>623.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_461657.1</th>\n",
       "      <td>16766042</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>GCF_000006945.1</td>\n",
       "      <td>ATGAGCACCATCCTGAAATGGGCGGGTAATAAAACCGCCATCATGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GTAAATTCACATTACATAATGCACCAACCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSTILKWAGNKTAIMAELKKHLPEGPRLVEPFAGSCAVMIETDYPH...</td>\n",
       "      <td>99287.STM2730</td>\n",
       "      <td>7.100000e-202</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901.0</td>\n",
       "      <td>Salmonella enterica subsp. enterica</td>\n",
       "      <td>59201.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_462387.1</th>\n",
       "      <td>16766772</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>GCF_000006945.1</td>\n",
       "      <td>ATGAAAAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGAAAATACC...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TTCTCAAGGAGAAGCGGATGAAACAGTATT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...</td>\n",
       "      <td>220341.STY4312</td>\n",
       "      <td>4.300000e-194</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901.0</td>\n",
       "      <td>Salmonella enterica subsp. enterica</td>\n",
       "      <td>59201.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_720943.1</th>\n",
       "      <td>24378988</td>\n",
       "      <td>Streptococcus mutans UA159</td>\n",
       "      <td>GCF_000007465.2</td>\n",
       "      <td>ATGTCAGAACAGATAAAATTAGCAAATACTACTTTGCAACCTTTTA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>CGAATATAAGTATGGGGGTATTTTTATGAC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSEQIKLANTTLQPFTKWTGGKRQLLPILRSYMPEKYNCYFEPFIG...</td>\n",
       "      <td>210007.SMU_504</td>\n",
       "      <td>1.300000e-193</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptococcus mutans</td>\n",
       "      <td>1309.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011098694.1</th>\n",
       "      <td>499411227</td>\n",
       "      <td>Clostridium tetani E88</td>\n",
       "      <td>GCF_000007625.1</td>\n",
       "      <td>ATGAAAAAGAATAAAATAATATTAGCAGCACCTGTATTGAAATGGG...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>GAATAAAAATGATGTTGCATGGGAAACCCT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNKIILAAPVLKWVGGKRQLMGEIKKVLPKTYTAYYEPFIGGGA...</td>\n",
       "      <td>212717.CTC00387</td>\n",
       "      <td>2.400000e-187</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Clostridium tetani</td>\n",
       "      <td>1513.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011135659.1</th>\n",
       "      <td>499448195</td>\n",
       "      <td>Chromobacterium violaceum ATCC 12472</td>\n",
       "      <td>GCF_000007705.1</td>\n",
       "      <td>ATGTTACAAATCCCAGCATCTCCCATCGTGCCCTGGATTGGCGGCA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CAAGCTTCGTGAAACAGCAAAGGGGCCAAG</td>\n",
       "      <td>11</td>\n",
       "      <td>MLQIPASPIVPWIGGKRRLAKHILPLFPEHTCYVEPFCGAAALYFL...</td>\n",
       "      <td>243365.CV_2111</td>\n",
       "      <td>3.500000e-182</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chromobacterium violaceum</td>\n",
       "      <td>536.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011201125.1</th>\n",
       "      <td>499514485</td>\n",
       "      <td>Mannheimia succiniciproducens MBEL55E</td>\n",
       "      <td>GCF_000007745.1</td>\n",
       "      <td>ATGTCGCATTCTGGCAAAACTAAACATGGCTTAAAACACCGCTCAT...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>TCGGTATTGCGATAAATTATAAAAATGAAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSHSGKTKHGLKHRSFLKWAGGKYRLTDNINNLFPKRRKCLVEPFV...</td>\n",
       "      <td>221988.MS1967</td>\n",
       "      <td>1.600000e-194</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Mannheimia] succiniciproducens</td>\n",
       "      <td>157673.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_793171.1</th>\n",
       "      <td>28870552</td>\n",
       "      <td>[Pseudomonas syringae] pv. tomato str. DC3000</td>\n",
       "      <td>GCF_000007805.1</td>\n",
       "      <td>ATGTCCACACCCATCATCCCTTGGATGGGCGGCAAACGCCGCCTGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GACTTTTACCCTAGAGGGGGGGGTGGTAGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MSTPIIPWMGGKRRLADRLIPLFPPHECYVEVFAGGAALYFMRPQA...</td>\n",
       "      <td>223283.PSPTO_3388</td>\n",
       "      <td>1.100000e-189</td>\n",
       "      <td>...</td>\n",
       "      <td>Pseudomonas syringae group</td>\n",
       "      <td>136849.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas syringae group genomosp. 3</td>\n",
       "      <td>251701.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011115082.1</th>\n",
       "      <td>499427618</td>\n",
       "      <td>Helicobacter hepaticus ATCC 51449</td>\n",
       "      <td>GCF_000007905.1</td>\n",
       "      <td>ATGAATGCGATAGTCTCACTCAAAACCCTTATCCCACCGCTTAAGA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GTTGGATTCAGATTAATTTCAATCCTCCCC</td>\n",
       "      <td>11</td>\n",
       "      <td>MNAIVSLKTLIPPLKIQGIKSKIIPHIRDIFQTWNKKGIYFEPFMG...</td>\n",
       "      <td>235279.HH0238</td>\n",
       "      <td>9.100000e-201</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Helicobacter hepaticus</td>\n",
       "      <td>32025.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_010944438.1</th>\n",
       "      <td>499246898</td>\n",
       "      <td>Haemophilus ducreyi 35000HP</td>\n",
       "      <td>GCF_000007945.1</td>\n",
       "      <td>ATGAACAATAACCAACGTGCTTTTTTAAAATGGGCGGGCGGAAAAT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TCATTACTGAACTAAAATTTATTAAAATAT</td>\n",
       "      <td>11</td>\n",
       "      <td>MNNNQRAFLKWAGGKYRLSIEIRKFLPPQSTQGVCFVEPFVGAGSV...</td>\n",
       "      <td>233412.HD0420</td>\n",
       "      <td>1.200000e-192</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Haemophilus] ducreyi</td>\n",
       "      <td>730.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011284350.1</th>\n",
       "      <td>499603616</td>\n",
       "      <td>Mycoplasma hyopneumoniae J</td>\n",
       "      <td>GCF_000008205.1</td>\n",
       "      <td>ATGTCAAAATCTGAATTAAAGCCTTTTGTAAAATGGGTTGGTGGGA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>ATTTTTGTTTTAGTTTAATTATTCGGGCTT</td>\n",
       "      <td>4</td>\n",
       "      <td>MSKSELKPFVKWVGGKTQLINVILSLLPKNFNSYIEPFLGGGALFL...</td>\n",
       "      <td>262719.MHJ_0623</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycoplasma hyopneumoniae</td>\n",
       "      <td>2099.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011264865.1</th>\n",
       "      <td>499584082</td>\n",
       "      <td>Mycoplasma mobile 163K</td>\n",
       "      <td>GCF_000008365.1</td>\n",
       "      <td>ATGAACATTTTATTGGAAAATTTTGGAACTAGATTAAAGGAACTAA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>ATTGCTATTTGTAATTTAAAAAAAGGACAA</td>\n",
       "      <td>4</td>\n",
       "      <td>MNILLENFGTRLKELRKSKQISQKELSESTGIVREQISRIENGLVN...</td>\n",
       "      <td>267748.MMOB3450</td>\n",
       "      <td>1.200000e-239</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycoplasma mobile</td>\n",
       "      <td>2118.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011235699.1</th>\n",
       "      <td>499554916</td>\n",
       "      <td>Idiomarina loihiensis L2TR</td>\n",
       "      <td>GCF_000008465.1</td>\n",
       "      <td>ATGAAGAAGAATCGGGCATTTTTAAAATGGGCAGGCGGTAAGTACT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>ACTGGAATTTAAAAGCATTTAACGACCTTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYSLIEPITARLPQGKKLIEPFVGAGSVFLNTD...</td>\n",
       "      <td>283942.IL2475</td>\n",
       "      <td>6.100000e-200</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Idiomarina loihiensis</td>\n",
       "      <td>135577.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YP_096088.1</th>\n",
       "      <td>52842289</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila str....</td>\n",
       "      <td>GCF_000008485.1</td>\n",
       "      <td>ATGATAAAAAATAGTGAAAAGGAATCAATAATCCGCATTAGACCAT...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>GGGTGTTGCCATGAAGCAGGGTGGTTCATA</td>\n",
       "      <td>11</td>\n",
       "      <td>MIKNSEKESIIRIRPFLKWAGGKFQIINKIRASLPKGARLIEPFVG...</td>\n",
       "      <td>272624.lpg2075</td>\n",
       "      <td>2.100000e-190</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Legionella pneumophila</td>\n",
       "      <td>446.0</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila</td>\n",
       "      <td>91891.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YP_094690.1</th>\n",
       "      <td>52840891</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila str....</td>\n",
       "      <td>GCF_000008485.1</td>\n",
       "      <td>ATGGTAAAAATCAGACCTTTTCTGAAATGGGCAGGAAGTAAGTACA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CAAAGATAAGGAAGAGTCTTTACCAGGATT</td>\n",
       "      <td>11</td>\n",
       "      <td>MVKIRPFLKWAGSKYNCLNEIISSLPTGQRLIEPFAGSGVVFMNTS...</td>\n",
       "      <td>272624.lpg0654</td>\n",
       "      <td>1.600000e-191</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Legionella pneumophila</td>\n",
       "      <td>446.0</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila</td>\n",
       "      <td>91891.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_312256.1</th>\n",
       "      <td>15833483</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>GCF_000008865.1</td>\n",
       "      <td>ATGAAGAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGCAAGTATC...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TTCTCAAGGAGAAGCGGATGAAACAGTATT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...</td>\n",
       "      <td>155864.Z4740</td>\n",
       "      <td>4.600000e-195</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YP_002345240.1</th>\n",
       "      <td>218927365</td>\n",
       "      <td>Yersinia pestis CO92</td>\n",
       "      <td>GCF_000009065.1</td>\n",
       "      <td>ATGAAGAAAAACCGCGCTTTTTTAAAATGGGCTGGTGGGAAATATC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GTGAAATTGCTGTCCGGTAATCGCTCAGAC</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLVDDIRRHLPAGDCLIEPFVGAGSVFLNTE...</td>\n",
       "      <td>187410.y3937</td>\n",
       "      <td>4.600000e-190</td>\n",
       "      <td>...</td>\n",
       "      <td>Yersinia pseudotuberculosis complex</td>\n",
       "      <td>1649845.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>632.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YP_001008117.1</th>\n",
       "      <td>123444147</td>\n",
       "      <td>Yersinia enterocolitica subsp. enterocolitica ...</td>\n",
       "      <td>GCF_000009345.1</td>\n",
       "      <td>ATGAAGAAAAACCGCGCTTTTTTAAAATGGGCTGGTGGGAAATATC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGGGGACTGCCGCTCAGTAAAGATTCAAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLVDDIRRHLPAGDCLIEPFVGAGSVFLNTE...</td>\n",
       "      <td>393305.YE3972</td>\n",
       "      <td>7.300000e-190</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yersinia enterocolitica</td>\n",
       "      <td>630.0</td>\n",
       "      <td>Yersinia enterocolitica subsp. enterocolitica</td>\n",
       "      <td>150052.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_010994238.1</th>\n",
       "      <td>499303463</td>\n",
       "      <td>Nostoc sp. PCC 7120</td>\n",
       "      <td>GCF_000009705.1</td>\n",
       "      <td>ATGGTAAGCCAAATCCCTAAAGAAACTTGCCCACGTCCATTTTTAA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>AATATCGGCATTAATGAATTAATTAATATA</td>\n",
       "      <td>11</td>\n",
       "      <td>MVSQIPKETCPRPFLKWAGGKSRLIPQYLSHLPKNYRTYHEPFLGG...</td>\n",
       "      <td>103690.all0061</td>\n",
       "      <td>1.800000e-193</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nostoc sp. PCC 7120</td>\n",
       "      <td>103690.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011151274.1</th>\n",
       "      <td>499464633</td>\n",
       "      <td>Vibrio vulnificus YJ016</td>\n",
       "      <td>GCF_000009745.1</td>\n",
       "      <td>ATGAAAAAGCAGCGAGCCTTTCTAAAATGGGCAGGCGGAAAATATG...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>CCCCATCGATTCATTCTCTACATTCACACT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKQRAFLKWAGGKYGLVEDIQKHLPPARKLVEPFVGAGSVFLNTD...</td>\n",
       "      <td>216895.VV1_1385</td>\n",
       "      <td>5.700000e-192</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Vibrio vulnificus</td>\n",
       "      <td>672.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011160764.1</th>\n",
       "      <td>499474124</td>\n",
       "      <td>Onion yellows phytoplasma OY-M</td>\n",
       "      <td>GCF_000009845.1</td>\n",
       "      <td>TTGTCTTTTTATTTTATTTATGGATTAAGGAATACAAAAATGCAAA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>AAAATCAATTTAAATTTTTGACCTCAACAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSFYFIYGLRNTKMQKFINWVGCKKKLLPHLLPLIPKKFKTYYEPF...</td>\n",
       "      <td>262768.PAM_409</td>\n",
       "      <td>1.400000e-200</td>\n",
       "      <td>...</td>\n",
       "      <td>Candidatus Phytoplasma asteris</td>\n",
       "      <td>85620.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Onion yellows phytoplasma</td>\n",
       "      <td>100379.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_034172244.1</th>\n",
       "      <td>735683320</td>\n",
       "      <td>Onion yellows phytoplasma OY-M</td>\n",
       "      <td>GCF_000009845.1</td>\n",
       "      <td>ATGGATTTTCGTCAAATTATCGCCAATAGTCAACAAGGCGATTTTC...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>AATTAAAATTTTCACTATCGTTTCACATCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MDFRQIIANSQQGDFLFIDPPYYYEGIKDKDFYQKPFSFTDHQQLA...</td>\n",
       "      <td>59748.PAa_0250</td>\n",
       "      <td>3.500000e-36</td>\n",
       "      <td>...</td>\n",
       "      <td>Candidatus Phytoplasma asteris</td>\n",
       "      <td>85620.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Onion yellows phytoplasma</td>\n",
       "      <td>100379.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_041624742.1</th>\n",
       "      <td>753869130</td>\n",
       "      <td>Onion yellows phytoplasma OY-M</td>\n",
       "      <td>GCF_000009845.1</td>\n",
       "      <td>ATGGATTTTTGTAAAGTTATCGCCAATAGTCAAAAAGGCGATTTTC...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>AATTAAAATTTTTACCATTGTTTTACATCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MDFCKVIANSQKGDFLFVDPPYYYEGIKDKDFYQKPFSFEDHKLLA...</td>\n",
       "      <td>59748.PAa_0250</td>\n",
       "      <td>5.200000e-36</td>\n",
       "      <td>...</td>\n",
       "      <td>Candidatus Phytoplasma asteris</td>\n",
       "      <td>85620.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Onion yellows phytoplasma</td>\n",
       "      <td>100379.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011160755.1</th>\n",
       "      <td>499474115</td>\n",
       "      <td>Onion yellows phytoplasma OY-M</td>\n",
       "      <td>GCF_000009845.1</td>\n",
       "      <td>ATGAAAATCAATTGGATTGGTTGTAAGAAAAAACTTTTGCCTCAAA...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>GTCAGCTATCTTTTATGTTTTAAATAAATA</td>\n",
       "      <td>11</td>\n",
       "      <td>MKINWIGCKKKLLPQIISHLPKQFKTYYEPFLGSGVLYFYLKSSKA...</td>\n",
       "      <td>262768.PAM_318</td>\n",
       "      <td>5.000000e-64</td>\n",
       "      <td>...</td>\n",
       "      <td>Candidatus Phytoplasma asteris</td>\n",
       "      <td>85620.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Onion yellows phytoplasma</td>\n",
       "      <td>100379.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_043743234.1</th>\n",
       "      <td>760060369</td>\n",
       "      <td>Magnetospirillum magneticum AMB-1</td>\n",
       "      <td>GCF_000009985.1</td>\n",
       "      <td>ATGACCATCAACCTCGATCACCGTTCCGTCACCCCCGTCGATCCCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>ACCCGTTCACCGTGGGCCAAACCTAAGTGT</td>\n",
       "      <td>11</td>\n",
       "      <td>MTINLDHRSVTPVDPAVGYIGGKRNLAGAIVPLIQSIPHTCYAEPF...</td>\n",
       "      <td>342108.amb0481</td>\n",
       "      <td>1.600000e-189</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Magnetospirillum magneticum</td>\n",
       "      <td>84159.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_043743981.1</th>\n",
       "      <td>760061119</td>\n",
       "      <td>Magnetospirillum magneticum AMB-1</td>\n",
       "      <td>GCF_000009985.1</td>\n",
       "      <td>ATGACCATAAACCCCCACCATGAACGCATTGCCCCGGTCAAGCCGG...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>ATGGGGAGATCAAGGCCGATTCAAAAGTCT</td>\n",
       "      <td>11</td>\n",
       "      <td>MTINPHHERIAPVKPAAAYIGGKRNLSRDLIQLIQSTPHDVYAEAF...</td>\n",
       "      <td>342108.amb1784</td>\n",
       "      <td>6.800000e-188</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Magnetospirillum magneticum</td>\n",
       "      <td>84159.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_000664222.1</th>\n",
       "      <td>446586876</td>\n",
       "      <td>Sodalis glossinidius str. 'morsitans'</td>\n",
       "      <td>GCF_000010085.1</td>\n",
       "      <td>ATGAAAGAACAATCTTTACCCATCGTCCCATGGATTGGCGGTAAAC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>AAGAGGTAGTTAAAGGTGGTGGTGCATCTG</td>\n",
       "      <td>11</td>\n",
       "      <td>MKEQSLPIVPWIGGKRRLAKHILPLFPAHTCYVEPFCGAAALYFLK...</td>\n",
       "      <td>343509.SG0828</td>\n",
       "      <td>5.600000e-183</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sodalis glossinidius</td>\n",
       "      <td>63612.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011412114.1</th>\n",
       "      <td>499731380</td>\n",
       "      <td>Sodalis glossinidius str. 'morsitans'</td>\n",
       "      <td>GCF_000010085.1</td>\n",
       "      <td>ATGAAGAAGAACCGCGCATTTTTGAAATGGGCCGGCGGCAAGTATC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGGGGCCGCCCCGCCGCAGACTGACCGTA</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLIEDIRRHLPEGDCLIEPFVGAGSVFLNTD...</td>\n",
       "      <td>343509.SG2306</td>\n",
       "      <td>1.100000e-188</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sodalis glossinidius</td>\n",
       "      <td>63612.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050692982.1</th>\n",
       "      <td>914732334</td>\n",
       "      <td>Aggregatibacter aphrophilus</td>\n",
       "      <td>GCF_001262035.1</td>\n",
       "      <td>ATGCCTGAACCCGCCAAACCCGTTGTGAAAACCAAAAGCCGCCCTT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TCCGAAAATCGACCGCACTTTTACGTTAAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MPEPAKPVVKTKSRPFLKWAGGKFRLTDEINRLLPKRKQCLVEPFV...</td>\n",
       "      <td>634176.NT05HA_2228</td>\n",
       "      <td>5.800000e-196</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aggregatibacter aphrophilus</td>\n",
       "      <td>732.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050715369.1</th>\n",
       "      <td>914772646</td>\n",
       "      <td>Ottowia sp. oral taxon 894</td>\n",
       "      <td>GCF_001262075.1</td>\n",
       "      <td>ATGCCCACACCCATCATCCCGTGGATCGGCGGCAAACGCCGCCTGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CACGGCGTGCAGCGTGTGCTATATTTCGCG</td>\n",
       "      <td>11</td>\n",
       "      <td>MPTPIIPWIGGKRRLAEHLIPRFPRHTCYVEVFAGGAALYFLKNPS...</td>\n",
       "      <td>397945.Aave_1638</td>\n",
       "      <td>7.100000e-140</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ottowia sp. oral taxon 894</td>\n",
       "      <td>1658672.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_053101747.1</th>\n",
       "      <td>921155650</td>\n",
       "      <td>Oblitimonas alkaliphila</td>\n",
       "      <td>GCF_001267175.1</td>\n",
       "      <td>ATGGCAAAACCTATAATCCCCTGGATGGGCGGCAAACGTCGCCTAG...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>AAAGCACCTATCTGTAGACCATTCTGCTGA</td>\n",
       "      <td>11</td>\n",
       "      <td>MAKPIIPWMGGKRRLVKQILPKFPEHQSYIEPFAGGAAMLFMKDPS...</td>\n",
       "      <td>637616.MDMS009_1001</td>\n",
       "      <td>3.400000e-113</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oblitimonas alkaliphila</td>\n",
       "      <td>1697053.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_053549854.1</th>\n",
       "      <td>924871148</td>\n",
       "      <td>Desulfuromonas sp. WTL</td>\n",
       "      <td>GCF_001278055.1</td>\n",
       "      <td>ATGCACAGACCAATTATCCCCTGGATCGGCGGCAAACGCCGCCTGG...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>GAGGCGGTTATGTGGCCAGGTTATATTGAG</td>\n",
       "      <td>11</td>\n",
       "      <td>MHRPIIPWIGGKRRLAKHLLPLFPKHTCYVEPFCGAAALFFLKPPA...</td>\n",
       "      <td>398767.Glov_2979</td>\n",
       "      <td>1.500000e-129</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Desulfuromonas soudanensis</td>\n",
       "      <td>1603606.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_062152675.1</th>\n",
       "      <td>1011233529</td>\n",
       "      <td>Beggiatoa leptomitiformis</td>\n",
       "      <td>GCF_001305575.2</td>\n",
       "      <td>ATGACAAACAAACACAGCGCGACCGCTAGCCACATTAAGCCATTTA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TATTAACGTACTAATTAATTTAGCTATTCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MTNKHSATASHIKPFIKWVGGKRGLLDELLKRLPQHFNHYFEPFVG...</td>\n",
       "      <td>572480.Arnit_2986</td>\n",
       "      <td>1.900000e-103</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Beggiatoa leptomitiformis</td>\n",
       "      <td>319938.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_062147837.1</th>\n",
       "      <td>1011228689</td>\n",
       "      <td>Beggiatoa leptomitiformis</td>\n",
       "      <td>GCF_001305575.2</td>\n",
       "      <td>ATGAAATTACCACATCCGATACCTTATCAAGGTAGTAAACGCAGTC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTTGTTATAATCATTAATGAAAAAAATAGT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKLPHPIPYQGSKRSLAIHILQYFPSHISRLIEPFAGSAAITIASA...</td>\n",
       "      <td>679189.HMPREF9019_0346</td>\n",
       "      <td>7.800000e-60</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Beggiatoa leptomitiformis</td>\n",
       "      <td>319938.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_054724231.1</th>\n",
       "      <td>938944685</td>\n",
       "      <td>Algibacter alginolytica</td>\n",
       "      <td>GCF_001310225.1</td>\n",
       "      <td>TTGAAAACAGCAAAACCGTTTCTTAAATGGGCTGGTGGAAAAAGCC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TGATTTCTTAAACGAATTAAGCGAAACTAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MKTAKPFLKWAGGKSQLINEIENSLPQDFKKKEFTFIEPFVGSGAV...</td>\n",
       "      <td>992406.RIA_0345</td>\n",
       "      <td>4.800000e-136</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Algibacter alginolytica</td>\n",
       "      <td>1736674.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_004315709.1</th>\n",
       "      <td>490444781</td>\n",
       "      <td>Bacteroides thetaiotaomicron</td>\n",
       "      <td>GCF_001314975.1</td>\n",
       "      <td>ATGGAGAAACGTACCTACTACAATGAAGGAAACCCGAATAACATCA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>ATGCTTTCCAACTCCGATCCCCGGCAAAAG</td>\n",
       "      <td>11</td>\n",
       "      <td>MEKRTYYNEGNPNNITRAALFIFFMRTCYNGIYSVNHSGKLSVTFG...</td>\n",
       "      <td>679190.HMPREF0650_1951</td>\n",
       "      <td>1.900000e-34</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteroides thetaiotaomicron</td>\n",
       "      <td>818.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_004298811.1</th>\n",
       "      <td>490426646</td>\n",
       "      <td>Bacteroides ovatus</td>\n",
       "      <td>GCF_001314995.1</td>\n",
       "      <td>ATGTCATCAAGAAGAACTAATAAACTTGTAGTTCCCTTTGTAAAGT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TGCAAGTAAGATAGACACTAATTGGACGAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MSSRRTNKLVVPFVKWVGGKRQILSEIEKYLPSKFATYYEPFIGGG...</td>\n",
       "      <td>411476.BACOVA_02878</td>\n",
       "      <td>8.700000e-192</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteroides ovatus</td>\n",
       "      <td>28116.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_027787034.1</th>\n",
       "      <td>654271882</td>\n",
       "      <td>Burkholderia cepacia ATCC 25416</td>\n",
       "      <td>GCF_001411495.1</td>\n",
       "      <td>ATGGCAAATCCCATCATTCCTTGGATCGGCGGTAAGCGCCGTCTTG...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>CGGATGGCTCGCCCGCAGGACAGCTCCTGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MANPIIPWIGGKRRLADHLIPRFPAHDCYVEVFAGGAALYFMRPPA...</td>\n",
       "      <td>339670.Bamb_1850</td>\n",
       "      <td>4.700000e-160</td>\n",
       "      <td>...</td>\n",
       "      <td>Burkholderia cepacia complex</td>\n",
       "      <td>87882.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Burkholderia cepacia</td>\n",
       "      <td>292.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_055732932.1</th>\n",
       "      <td>944521354</td>\n",
       "      <td>Agarivorans gilvus</td>\n",
       "      <td>GCF_001420915.1</td>\n",
       "      <td>ATGAAAAAAACACGTGCTTTTTTAAAGTGGGCCGGGGGTAAATATG...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>ACGCACTAATAAAACAATGCTGAGCACAAA</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKTRAFLKWAGGKYGLIEDIQQQLPEGDLLIEPFVGAASVFLNTN...</td>\n",
       "      <td>58051.PE36_00684</td>\n",
       "      <td>1.300000e-116</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Agarivorans gilvus</td>\n",
       "      <td>680279.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_057900243.1</th>\n",
       "      <td>951586723</td>\n",
       "      <td>Aneurinibacillus sp. XH2</td>\n",
       "      <td>GCF_001439585.2</td>\n",
       "      <td>TTGTTCTCTGATCAGAATGGTAAGGAGGAAACCCCTTTGAAGCCGT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TGTGTAAATGTATAACAAATAGCTCATCAC</td>\n",
       "      <td>11</td>\n",
       "      <td>MFSDQNGKEETPLKPFLKWAGGKYKLVPHIKELLPKTRRLIEPFVG...</td>\n",
       "      <td>521098.Aaci_2656</td>\n",
       "      <td>5.100000e-118</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aneurinibacillus sp. XH2</td>\n",
       "      <td>1450761.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_057179500.1</th>\n",
       "      <td>950123378</td>\n",
       "      <td>Colwellia sp. MT41</td>\n",
       "      <td>GCF_001444365.1</td>\n",
       "      <td>ATGAATAATAAACATCGAGCGTTTTTGAAGTGGGCTGGCGGTAAAT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>GCTGGTCCGAATTAATTTGGTATAACTAAG</td>\n",
       "      <td>11</td>\n",
       "      <td>MNNKHRAFLKWAGGKYSLCDVISKMLPNERRLIEPFVGAGSVFLNT...</td>\n",
       "      <td>298386.PBPRA0283</td>\n",
       "      <td>3.300000e-124</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Colwellia sp. MT41</td>\n",
       "      <td>58049.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_058031670.1</th>\n",
       "      <td>953560398</td>\n",
       "      <td>Pseudoalteromonas phenolica</td>\n",
       "      <td>GCF_001444405.1</td>\n",
       "      <td>ATCAATCGCCGTTCCCCATTAAAATGGGCAGGCGGCAAAAAGCGGA...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTTTTATACAAGGTTTGAAAGAGCAATCCT</td>\n",
       "      <td>11</td>\n",
       "      <td>MNRRSPLKWAGGKKRIIEEIKANLPLKGKQRLIEPFVGGGSVFLNI...</td>\n",
       "      <td>167879.CPS_0475</td>\n",
       "      <td>1.100000e-130</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudoalteromonas phenolica</td>\n",
       "      <td>161398.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_050667307.1</th>\n",
       "      <td>914676883</td>\n",
       "      <td>Aeromonas schubertii</td>\n",
       "      <td>GCF_001447335.1</td>\n",
       "      <td>ATGAAAAAAACACGCGCCTTCCTGAAATGGGCCGGGGGCAAGTTCT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>TTCGGCCTCGTCTCTGTCCTGGCTTTGCGG</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKTRAFLKWAGGKFSLIEEIAERLPEGRVLLEPFVGAGSVFLNTD...</td>\n",
       "      <td>380703.AHA_3186</td>\n",
       "      <td>2.300000e-169</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aeromonas schubertii</td>\n",
       "      <td>652.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_008115051.1</th>\n",
       "      <td>495390350</td>\n",
       "      <td>Pseudoalteromonas issachenkonii</td>\n",
       "      <td>GCF_001455325.1</td>\n",
       "      <td>ATGCAGCAAAAGAGTAGAGCCTTCCTTAAATGGGCTGGCGGAAAAT...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>CCAGTACAAATAATACAAACAGTACTACGG</td>\n",
       "      <td>11</td>\n",
       "      <td>MQQKSRAFLKWAGGKYSLVEDINARLNQASKHADTLVEPFVGAGSV...</td>\n",
       "      <td>234831.PSM_A0313</td>\n",
       "      <td>1.400000e-196</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudoalteromonas issachenkonii</td>\n",
       "      <td>152297.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_062483258.1</th>\n",
       "      <td>1011632908</td>\n",
       "      <td>Lacimicrobium alkaliphilum</td>\n",
       "      <td>GCF_001466725.1</td>\n",
       "      <td>ATGGATAATAAGAAAAGCCGGGCCTTTCTGAAGTGGGCCGGTGGCA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TGGCTACAAGCAGGCAAATTTGCACGCCAG</td>\n",
       "      <td>11</td>\n",
       "      <td>MDNKKSRAFLKWAGGKYSLVSEIRSLLPDGERLIEPFVGAGSVFLN...</td>\n",
       "      <td>983545.Glaag_0729</td>\n",
       "      <td>1.400000e-125</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lacimicrobium alkaliphilum</td>\n",
       "      <td>1526571.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_010386213.1</th>\n",
       "      <td>498072057</td>\n",
       "      <td>Pseudoalteromonas rubra</td>\n",
       "      <td>GCF_001482385.1</td>\n",
       "      <td>ATGCAAAAAAAGACCAGAGCTTTCTTAAAATGGGCAGGTGGAAAAT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TCGCGTTGAATGGCTGGCACGTCATAATAC</td>\n",
       "      <td>11</td>\n",
       "      <td>MQKKTRAFLKWAGGKYALVEEITKRLQAANEEAETLVEPFVGAGSV...</td>\n",
       "      <td>234831.PSM_A0313</td>\n",
       "      <td>1.100000e-160</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudoalteromonas rubra</td>\n",
       "      <td>43658.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_059307157.1</th>\n",
       "      <td>976148944</td>\n",
       "      <td>Leclercia adecarboxylata</td>\n",
       "      <td>GCF_001518835.1</td>\n",
       "      <td>ATGAGCACCATCCTGAAATGGGCAGGCAATAAAACAGCCGTCATGC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCCATCACGACGTTAAAAAGCACGGCGGTG</td>\n",
       "      <td>11</td>\n",
       "      <td>MSTILKWAGNKTAVMPELIKHLPAGPRLVEPFAGSCAVMMATDYPH...</td>\n",
       "      <td>220341.STY4636</td>\n",
       "      <td>4.100000e-140</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leclercia adecarboxylata</td>\n",
       "      <td>83655.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_032614792.1</th>\n",
       "      <td>695639588</td>\n",
       "      <td>Leclercia adecarboxylata</td>\n",
       "      <td>GCF_001518835.1</td>\n",
       "      <td>ATGAAAAAAAATCGCGCTTTTCTGAAATGGGCAGGGGGGAAATACC...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCCGTTTCGCCCGCCGTTAAACAAAGTTCA</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKYPLLDDIKQRLPAGECLVEPFVGAGSVFLNTD...</td>\n",
       "      <td>500639.ENTCAN_08385</td>\n",
       "      <td>2.200000e-180</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leclercia adecarboxylata</td>\n",
       "      <td>83655.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_010655730.1</th>\n",
       "      <td>498341574</td>\n",
       "      <td>Pseudomonas fragi</td>\n",
       "      <td>GCF_001543265.1</td>\n",
       "      <td>ATGTCTTCTCCAATCATTCCATGGATGGGCGGTAAGCGGCGTTTGG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CTCAGATCGCGAGTGAAAAGCTTTGTATAT</td>\n",
       "      <td>11</td>\n",
       "      <td>MSSPIIPWMGGKRRLADRLIPLFPPHECYVEVFAGGAALYFLRPQG...</td>\n",
       "      <td>1114970.PSF113_4284</td>\n",
       "      <td>1.600000e-174</td>\n",
       "      <td>...</td>\n",
       "      <td>Pseudomonas chlororaphis group</td>\n",
       "      <td>136842.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas fragi</td>\n",
       "      <td>296.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_005554455.1</th>\n",
       "      <td>491706821</td>\n",
       "      <td>Aggregatibacter actinomycetemcomitans NUM4039</td>\n",
       "      <td>GCF_001547775.1</td>\n",
       "      <td>ATGCCTGAACCCGCTAAACCCGCTACGCCAGCCAAAAGCCGCCCGT...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>CCGCACTTTTAGTGCCACTAAAACACTGTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MPEPAKPATPAKSRPFLKWAGGKYRLMDEINRLLPKRKQCLVEPFV...</td>\n",
       "      <td>694569.D7S_1130</td>\n",
       "      <td>6.600000e-198</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aggregatibacter actinomycetemcomitans</td>\n",
       "      <td>714.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_060836405.1</th>\n",
       "      <td>985596904</td>\n",
       "      <td>Rhodovulum sulfidophilum</td>\n",
       "      <td>GCF_001548075.1</td>\n",
       "      <td>ATGCGAAAAACCGTCCCCGCAACCCCCGTCGCCCCCTGGATGGGCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GTTCCCCCGGCCCGGTCGGCGTTTAACCGG</td>\n",
       "      <td>11</td>\n",
       "      <td>MRKTVPATPVAPWMGGKKALHRRIIERIEAIPHRTYVEPFLGMGGV...</td>\n",
       "      <td>272943.RSP_3769</td>\n",
       "      <td>2.300000e-132</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rhodovulum sulfidophilum</td>\n",
       "      <td>35806.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_062332635.1</th>\n",
       "      <td>1011421816</td>\n",
       "      <td>Moraxella osloensis</td>\n",
       "      <td>GCF_001553955.1</td>\n",
       "      <td>ATGTCTCAACTTAAACCTTTTCTAAAATGGGCTGGCGGTAAATCAC...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TCTCTAAACAATTTTATCGGCTTTGATAGT</td>\n",
       "      <td>11</td>\n",
       "      <td>MSQLKPFLKWAGGKSQLIGDIQARLPNYTANQDFCMIEPFLGGGAV...</td>\n",
       "      <td>553217.ENHAE0001_1514</td>\n",
       "      <td>1.300000e-201</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moraxella osloensis</td>\n",
       "      <td>34062.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_005501354.1</th>\n",
       "      <td>491643827</td>\n",
       "      <td>Grimontia hollisae</td>\n",
       "      <td>GCF_001558255.1</td>\n",
       "      <td>ATGAAAAAGCATCGTGCCTTTTTGAAGTGGGCGGGGGGTAAGTATT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>GGCACCATACTGGATTGCTGAAAATCCCGT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKHRAFLKWAGGKYSLVEDIRRHLPEGKKLIEPFVGAGSVFLNTD...</td>\n",
       "      <td>675812.VHA_000396</td>\n",
       "      <td>2.000000e-210</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Grimontia hollisae</td>\n",
       "      <td>673.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_013359279.1</th>\n",
       "      <td>503124618</td>\n",
       "      <td>Pantoea agglomerans</td>\n",
       "      <td>GCF_001558735.1</td>\n",
       "      <td>ATGAAAAAAAATCGCGCTTTCCTGAAATGGGCAGGGGGAAAATTTC...</td>\n",
       "      <td>TAG</td>\n",
       "      <td>TCTGTTTCGTCCGCGACCAACGCATTACGC</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKNRAFLKWAGGKFPLVEEIHRHLPQGDCLVEPFVGAGSVFLNTE...</td>\n",
       "      <td>712898.Pvag_2935</td>\n",
       "      <td>5.300000e-190</td>\n",
       "      <td>...</td>\n",
       "      <td>Pantoea agglomerans group</td>\n",
       "      <td>1654067.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pantoea agglomerans</td>\n",
       "      <td>549.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_061063426.1</th>\n",
       "      <td>992369549</td>\n",
       "      <td>Pantoea agglomerans</td>\n",
       "      <td>GCF_001558735.1</td>\n",
       "      <td>ATGAAAACCATCCTGAAATGGGCCGGCAGCAAGTCCGGCCTGATGC...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TGACTCAGGTCTACGCATATCCCTGGAATG</td>\n",
       "      <td>11</td>\n",
       "      <td>MKTILKWAGSKSGLMPELIKHLPAGNRLVEPFAGSCAVMMNTDYPA...</td>\n",
       "      <td>706191.PANA_3169</td>\n",
       "      <td>8.900000e-120</td>\n",
       "      <td>...</td>\n",
       "      <td>Pantoea agglomerans group</td>\n",
       "      <td>1654067.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pantoea agglomerans</td>\n",
       "      <td>549.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_057794374.1</th>\n",
       "      <td>951421099</td>\n",
       "      <td>Alteromonas stellipolaris</td>\n",
       "      <td>GCF_001562115.1</td>\n",
       "      <td>ATGATGCAGAAAAAGAACCGGGCCTTTTTAAAGTGGGCCGGTGGCA...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>CTAATAACCGGTTAGCCACTTCACAAACAT</td>\n",
       "      <td>11</td>\n",
       "      <td>MMQKKNRAFLKWAGGKYSLVEHIQARLPQANKLIEPFVGAGSVFLN...</td>\n",
       "      <td>715451.ambt_16195</td>\n",
       "      <td>6.400000e-198</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alteromonas stellipolaris</td>\n",
       "      <td>233316.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_024708114.1</th>\n",
       "      <td>639625290</td>\n",
       "      <td>Martelella sp. AD-3</td>\n",
       "      <td>GCF_001578105.1</td>\n",
       "      <td>ATGATCTCGGTGCGGCCTGTGTCGCCGCCGGCCGCCTATCTTGGCG...</td>\n",
       "      <td>TGA</td>\n",
       "      <td>GCGGGTGTTCATCCGAGGTGACGCCTGATC</td>\n",
       "      <td>11</td>\n",
       "      <td>MISVRPVSPPAAYLGGKKRLAKRVSSLIEQTPHDLYAEPFVGMGGV...</td>\n",
       "      <td>411684.HPDFL43_05935</td>\n",
       "      <td>5.600000e-128</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Martelella sp. AD-3</td>\n",
       "      <td>686597.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_064335703.1</th>\n",
       "      <td>1033019523</td>\n",
       "      <td>Aeromonas veronii</td>\n",
       "      <td>GCF_001593245.1</td>\n",
       "      <td>ATGAAAAAAACACGCGCTTTTTTAAAATGGGCCGGGGGAAAATACT...</td>\n",
       "      <td>TAA</td>\n",
       "      <td>CAGACCATGCGGCGCCCAAGGGCGCCGTTT</td>\n",
       "      <td>11</td>\n",
       "      <td>MKKTRAFLKWAGGKYSLVEEIAERLPAGRVLLEPFVGAGSVFLNTD...</td>\n",
       "      <td>998088.B565_0972</td>\n",
       "      <td>1.800000e-202</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aeromonas veronii</td>\n",
       "      <td>654.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>364 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  proteinGI  \\\n",
       "refSeqProteinId               \n",
       "NP_417846.1        16131265   \n",
       "NP_232254.1        15642621   \n",
       "NP_709160.1        24114650   \n",
       "NP_461657.1        16766042   \n",
       "NP_462387.1        16766772   \n",
       "NP_720943.1        24378988   \n",
       "WP_011098694.1    499411227   \n",
       "WP_011135659.1    499448195   \n",
       "WP_011201125.1    499514485   \n",
       "NP_793171.1        28870552   \n",
       "WP_011115082.1    499427618   \n",
       "WP_010944438.1    499246898   \n",
       "WP_011284350.1    499603616   \n",
       "WP_011264865.1    499584082   \n",
       "WP_011235699.1    499554916   \n",
       "YP_096088.1        52842289   \n",
       "YP_094690.1        52840891   \n",
       "NP_312256.1        15833483   \n",
       "YP_002345240.1    218927365   \n",
       "YP_001008117.1    123444147   \n",
       "WP_010994238.1    499303463   \n",
       "WP_011151274.1    499464633   \n",
       "WP_011160764.1    499474124   \n",
       "WP_034172244.1    735683320   \n",
       "WP_041624742.1    753869130   \n",
       "WP_011160755.1    499474115   \n",
       "WP_043743234.1    760060369   \n",
       "WP_043743981.1    760061119   \n",
       "WP_000664222.1    446586876   \n",
       "WP_011412114.1    499731380   \n",
       "...                     ...   \n",
       "WP_050692982.1    914732334   \n",
       "WP_050715369.1    914772646   \n",
       "WP_053101747.1    921155650   \n",
       "WP_053549854.1    924871148   \n",
       "WP_062152675.1   1011233529   \n",
       "WP_062147837.1   1011228689   \n",
       "WP_054724231.1    938944685   \n",
       "WP_004315709.1    490444781   \n",
       "WP_004298811.1    490426646   \n",
       "WP_027787034.1    654271882   \n",
       "WP_055732932.1    944521354   \n",
       "WP_057900243.1    951586723   \n",
       "WP_057179500.1    950123378   \n",
       "WP_058031670.1    953560398   \n",
       "WP_050667307.1    914676883   \n",
       "WP_008115051.1    495390350   \n",
       "WP_062483258.1   1011632908   \n",
       "WP_010386213.1    498072057   \n",
       "WP_059307157.1    976148944   \n",
       "WP_032614792.1    695639588   \n",
       "WP_010655730.1    498341574   \n",
       "WP_005554455.1    491706821   \n",
       "WP_060836405.1    985596904   \n",
       "WP_062332635.1   1011421816   \n",
       "WP_005501354.1    491643827   \n",
       "WP_013359279.1    503124618   \n",
       "WP_061063426.1    992369549   \n",
       "WP_057794374.1    951421099   \n",
       "WP_024708114.1    639625290   \n",
       "WP_064335703.1   1033019523   \n",
       "\n",
       "                                                    species_name_x  \\\n",
       "refSeqProteinId                                                      \n",
       "NP_417846.1              Escherichia coli str. K-12 substr. MG1655   \n",
       "NP_232254.1           Vibrio cholerae O1 biovar El Tor str. N16961   \n",
       "NP_709160.1                          Shigella flexneri 2a str. 301   \n",
       "NP_461657.1      Salmonella enterica subsp. enterica serovar Ty...   \n",
       "NP_462387.1      Salmonella enterica subsp. enterica serovar Ty...   \n",
       "NP_720943.1                             Streptococcus mutans UA159   \n",
       "WP_011098694.1                              Clostridium tetani E88   \n",
       "WP_011135659.1                Chromobacterium violaceum ATCC 12472   \n",
       "WP_011201125.1               Mannheimia succiniciproducens MBEL55E   \n",
       "NP_793171.1          [Pseudomonas syringae] pv. tomato str. DC3000   \n",
       "WP_011115082.1                   Helicobacter hepaticus ATCC 51449   \n",
       "WP_010944438.1                         Haemophilus ducreyi 35000HP   \n",
       "WP_011284350.1                          Mycoplasma hyopneumoniae J   \n",
       "WP_011264865.1                              Mycoplasma mobile 163K   \n",
       "WP_011235699.1                          Idiomarina loihiensis L2TR   \n",
       "YP_096088.1      Legionella pneumophila subsp. pneumophila str....   \n",
       "YP_094690.1      Legionella pneumophila subsp. pneumophila str....   \n",
       "NP_312256.1                    Escherichia coli O157:H7 str. Sakai   \n",
       "YP_002345240.1                                Yersinia pestis CO92   \n",
       "YP_001008117.1   Yersinia enterocolitica subsp. enterocolitica ...   \n",
       "WP_010994238.1                                 Nostoc sp. PCC 7120   \n",
       "WP_011151274.1                             Vibrio vulnificus YJ016   \n",
       "WP_011160764.1                      Onion yellows phytoplasma OY-M   \n",
       "WP_034172244.1                      Onion yellows phytoplasma OY-M   \n",
       "WP_041624742.1                      Onion yellows phytoplasma OY-M   \n",
       "WP_011160755.1                      Onion yellows phytoplasma OY-M   \n",
       "WP_043743234.1                   Magnetospirillum magneticum AMB-1   \n",
       "WP_043743981.1                   Magnetospirillum magneticum AMB-1   \n",
       "WP_000664222.1               Sodalis glossinidius str. 'morsitans'   \n",
       "WP_011412114.1               Sodalis glossinidius str. 'morsitans'   \n",
       "...                                                            ...   \n",
       "WP_050692982.1                         Aggregatibacter aphrophilus   \n",
       "WP_050715369.1                          Ottowia sp. oral taxon 894   \n",
       "WP_053101747.1                             Oblitimonas alkaliphila   \n",
       "WP_053549854.1                              Desulfuromonas sp. WTL   \n",
       "WP_062152675.1                           Beggiatoa leptomitiformis   \n",
       "WP_062147837.1                           Beggiatoa leptomitiformis   \n",
       "WP_054724231.1                             Algibacter alginolytica   \n",
       "WP_004315709.1                        Bacteroides thetaiotaomicron   \n",
       "WP_004298811.1                                  Bacteroides ovatus   \n",
       "WP_027787034.1                     Burkholderia cepacia ATCC 25416   \n",
       "WP_055732932.1                                  Agarivorans gilvus   \n",
       "WP_057900243.1                            Aneurinibacillus sp. XH2   \n",
       "WP_057179500.1                                  Colwellia sp. MT41   \n",
       "WP_058031670.1                         Pseudoalteromonas phenolica   \n",
       "WP_050667307.1                                Aeromonas schubertii   \n",
       "WP_008115051.1                     Pseudoalteromonas issachenkonii   \n",
       "WP_062483258.1                          Lacimicrobium alkaliphilum   \n",
       "WP_010386213.1                             Pseudoalteromonas rubra   \n",
       "WP_059307157.1                            Leclercia adecarboxylata   \n",
       "WP_032614792.1                            Leclercia adecarboxylata   \n",
       "WP_010655730.1                                   Pseudomonas fragi   \n",
       "WP_005554455.1       Aggregatibacter actinomycetemcomitans NUM4039   \n",
       "WP_060836405.1                            Rhodovulum sulfidophilum   \n",
       "WP_062332635.1                                 Moraxella osloensis   \n",
       "WP_005501354.1                                  Grimontia hollisae   \n",
       "WP_013359279.1                                 Pantoea agglomerans   \n",
       "WP_061063426.1                                 Pantoea agglomerans   \n",
       "WP_057794374.1                           Alteromonas stellipolaris   \n",
       "WP_024708114.1                                 Martelella sp. AD-3   \n",
       "WP_064335703.1                                   Aeromonas veronii   \n",
       "\n",
       "                genome_accession  \\\n",
       "refSeqProteinId                    \n",
       "NP_417846.1      GCF_000005845.2   \n",
       "NP_232254.1      GCF_000006745.1   \n",
       "NP_709160.1      GCF_000006925.2   \n",
       "NP_461657.1      GCF_000006945.1   \n",
       "NP_462387.1      GCF_000006945.1   \n",
       "NP_720943.1      GCF_000007465.2   \n",
       "WP_011098694.1   GCF_000007625.1   \n",
       "WP_011135659.1   GCF_000007705.1   \n",
       "WP_011201125.1   GCF_000007745.1   \n",
       "NP_793171.1      GCF_000007805.1   \n",
       "WP_011115082.1   GCF_000007905.1   \n",
       "WP_010944438.1   GCF_000007945.1   \n",
       "WP_011284350.1   GCF_000008205.1   \n",
       "WP_011264865.1   GCF_000008365.1   \n",
       "WP_011235699.1   GCF_000008465.1   \n",
       "YP_096088.1      GCF_000008485.1   \n",
       "YP_094690.1      GCF_000008485.1   \n",
       "NP_312256.1      GCF_000008865.1   \n",
       "YP_002345240.1   GCF_000009065.1   \n",
       "YP_001008117.1   GCF_000009345.1   \n",
       "WP_010994238.1   GCF_000009705.1   \n",
       "WP_011151274.1   GCF_000009745.1   \n",
       "WP_011160764.1   GCF_000009845.1   \n",
       "WP_034172244.1   GCF_000009845.1   \n",
       "WP_041624742.1   GCF_000009845.1   \n",
       "WP_011160755.1   GCF_000009845.1   \n",
       "WP_043743234.1   GCF_000009985.1   \n",
       "WP_043743981.1   GCF_000009985.1   \n",
       "WP_000664222.1   GCF_000010085.1   \n",
       "WP_011412114.1   GCF_000010085.1   \n",
       "...                          ...   \n",
       "WP_050692982.1   GCF_001262035.1   \n",
       "WP_050715369.1   GCF_001262075.1   \n",
       "WP_053101747.1   GCF_001267175.1   \n",
       "WP_053549854.1   GCF_001278055.1   \n",
       "WP_062152675.1   GCF_001305575.2   \n",
       "WP_062147837.1   GCF_001305575.2   \n",
       "WP_054724231.1   GCF_001310225.1   \n",
       "WP_004315709.1   GCF_001314975.1   \n",
       "WP_004298811.1   GCF_001314995.1   \n",
       "WP_027787034.1   GCF_001411495.1   \n",
       "WP_055732932.1   GCF_001420915.1   \n",
       "WP_057900243.1   GCF_001439585.2   \n",
       "WP_057179500.1   GCF_001444365.1   \n",
       "WP_058031670.1   GCF_001444405.1   \n",
       "WP_050667307.1   GCF_001447335.1   \n",
       "WP_008115051.1   GCF_001455325.1   \n",
       "WP_062483258.1   GCF_001466725.1   \n",
       "WP_010386213.1   GCF_001482385.1   \n",
       "WP_059307157.1   GCF_001518835.1   \n",
       "WP_032614792.1   GCF_001518835.1   \n",
       "WP_010655730.1   GCF_001543265.1   \n",
       "WP_005554455.1   GCF_001547775.1   \n",
       "WP_060836405.1   GCF_001548075.1   \n",
       "WP_062332635.1   GCF_001553955.1   \n",
       "WP_005501354.1   GCF_001558255.1   \n",
       "WP_013359279.1   GCF_001558735.1   \n",
       "WP_061063426.1   GCF_001558735.1   \n",
       "WP_057794374.1   GCF_001562115.1   \n",
       "WP_024708114.1   GCF_001578105.1   \n",
       "WP_064335703.1   GCF_001593245.1   \n",
       "\n",
       "                                                            DNASeq stopCodon  \\\n",
       "refSeqProteinId                                                                \n",
       "NP_417846.1      ATGAAGAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGCAAGTATC...       TAA   \n",
       "NP_232254.1      ATGAAAAAGCAGCGCGCCTTTCTGAAATGGGCTGGGGGTAAATATA...       TGA   \n",
       "NP_709160.1      ATGAAGAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGCAAGTATC...       TAA   \n",
       "NP_461657.1      ATGAGCACCATCCTGAAATGGGCGGGTAATAAAACCGCCATCATGG...       TGA   \n",
       "NP_462387.1      ATGAAAAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGAAAATACC...       TAA   \n",
       "NP_720943.1      ATGTCAGAACAGATAAAATTAGCAAATACTACTTTGCAACCTTTTA...       TAA   \n",
       "WP_011098694.1   ATGAAAAAGAATAAAATAATATTAGCAGCACCTGTATTGAAATGGG...       TAA   \n",
       "WP_011135659.1   ATGTTACAAATCCCAGCATCTCCCATCGTGCCCTGGATTGGCGGCA...       TGA   \n",
       "WP_011201125.1   ATGTCGCATTCTGGCAAAACTAAACATGGCTTAAAACACCGCTCAT...       TAG   \n",
       "NP_793171.1      ATGTCCACACCCATCATCCCTTGGATGGGCGGCAAACGCCGCCTGG...       TGA   \n",
       "WP_011115082.1   ATGAATGCGATAGTCTCACTCAAAACCCTTATCCCACCGCTTAAGA...       TGA   \n",
       "WP_010944438.1   ATGAACAATAACCAACGTGCTTTTTTAAAATGGGCGGGCGGAAAAT...       TAA   \n",
       "WP_011284350.1   ATGTCAAAATCTGAATTAAAGCCTTTTGTAAAATGGGTTGGTGGGA...       TAA   \n",
       "WP_011264865.1   ATGAACATTTTATTGGAAAATTTTGGAACTAGATTAAAGGAACTAA...       TAA   \n",
       "WP_011235699.1   ATGAAGAAGAATCGGGCATTTTTAAAATGGGCAGGCGGTAAGTACT...       TGA   \n",
       "YP_096088.1      ATGATAAAAAATAGTGAAAAGGAATCAATAATCCGCATTAGACCAT...       TAG   \n",
       "YP_094690.1      ATGGTAAAAATCAGACCTTTTCTGAAATGGGCAGGAAGTAAGTACA...       TGA   \n",
       "NP_312256.1      ATGAAGAAAAATCGCGCTTTTTTGAAGTGGGCAGGGGGCAAGTATC...       TAA   \n",
       "YP_002345240.1   ATGAAGAAAAACCGCGCTTTTTTAAAATGGGCTGGTGGGAAATATC...       TGA   \n",
       "YP_001008117.1   ATGAAGAAAAACCGCGCTTTTTTAAAATGGGCTGGTGGGAAATATC...       TGA   \n",
       "WP_010994238.1   ATGGTAAGCCAAATCCCTAAAGAAACTTGCCCACGTCCATTTTTAA...       TAA   \n",
       "WP_011151274.1   ATGAAAAAGCAGCGAGCCTTTCTAAAATGGGCAGGCGGAAAATATG...       TAA   \n",
       "WP_011160764.1   TTGTCTTTTTATTTTATTTATGGATTAAGGAATACAAAAATGCAAA...       TAA   \n",
       "WP_034172244.1   ATGGATTTTCGTCAAATTATCGCCAATAGTCAACAAGGCGATTTTC...       TAA   \n",
       "WP_041624742.1   ATGGATTTTTGTAAAGTTATCGCCAATAGTCAAAAAGGCGATTTTC...       TAG   \n",
       "WP_011160755.1   ATGAAAATCAATTGGATTGGTTGTAAGAAAAAACTTTTGCCTCAAA...       TAG   \n",
       "WP_043743234.1   ATGACCATCAACCTCGATCACCGTTCCGTCACCCCCGTCGATCCCG...       TGA   \n",
       "WP_043743981.1   ATGACCATAAACCCCCACCATGAACGCATTGCCCCGGTCAAGCCGG...       TAA   \n",
       "WP_000664222.1   ATGAAAGAACAATCTTTACCCATCGTCCCATGGATTGGCGGTAAAC...       TGA   \n",
       "WP_011412114.1   ATGAAGAAGAACCGCGCATTTTTGAAATGGGCCGGCGGCAAGTATC...       TGA   \n",
       "...                                                            ...       ...   \n",
       "WP_050692982.1   ATGCCTGAACCCGCCAAACCCGTTGTGAAAACCAAAAGCCGCCCTT...       TAA   \n",
       "WP_050715369.1   ATGCCCACACCCATCATCCCGTGGATCGGCGGCAAACGCCGCCTGG...       TGA   \n",
       "WP_053101747.1   ATGGCAAAACCTATAATCCCCTGGATGGGCGGCAAACGTCGCCTAG...       TAA   \n",
       "WP_053549854.1   ATGCACAGACCAATTATCCCCTGGATCGGCGGCAAACGCCGCCTGG...       TAG   \n",
       "WP_062152675.1   ATGACAAACAAACACAGCGCGACCGCTAGCCACATTAAGCCATTTA...       TAA   \n",
       "WP_062147837.1   ATGAAATTACCACATCCGATACCTTATCAAGGTAGTAAACGCAGTC...       TGA   \n",
       "WP_054724231.1   TTGAAAACAGCAAAACCGTTTCTTAAATGGGCTGGTGGAAAAAGCC...       TGA   \n",
       "WP_004315709.1   ATGGAGAAACGTACCTACTACAATGAAGGAAACCCGAATAACATCA...       TGA   \n",
       "WP_004298811.1   ATGTCATCAAGAAGAACTAATAAACTTGTAGTTCCCTTTGTAAAGT...       TAA   \n",
       "WP_027787034.1   ATGGCAAATCCCATCATTCCTTGGATCGGCGGTAAGCGCCGTCTTG...       TAG   \n",
       "WP_055732932.1   ATGAAAAAAACACGTGCTTTTTTAAAGTGGGCCGGGGGTAAATATG...       TAA   \n",
       "WP_057900243.1   TTGTTCTCTGATCAGAATGGTAAGGAGGAAACCCCTTTGAAGCCGT...       TAA   \n",
       "WP_057179500.1   ATGAATAATAAACATCGAGCGTTTTTGAAGTGGGCTGGCGGTAAAT...       TAA   \n",
       "WP_058031670.1   ATCAATCGCCGTTCCCCATTAAAATGGGCAGGCGGCAAAAAGCGGA...       TGA   \n",
       "WP_050667307.1   ATGAAAAAAACACGCGCCTTCCTGAAATGGGCCGGGGGCAAGTTCT...       TGA   \n",
       "WP_008115051.1   ATGCAGCAAAAGAGTAGAGCCTTCCTTAAATGGGCTGGCGGAAAAT...       TAG   \n",
       "WP_062483258.1   ATGGATAATAAGAAAAGCCGGGCCTTTCTGAAGTGGGCCGGTGGCA...       TAA   \n",
       "WP_010386213.1   ATGCAAAAAAAGACCAGAGCTTTCTTAAAATGGGCAGGTGGAAAAT...       TAA   \n",
       "WP_059307157.1   ATGAGCACCATCCTGAAATGGGCAGGCAATAAAACAGCCGTCATGC...       TGA   \n",
       "WP_032614792.1   ATGAAAAAAAATCGCGCTTTTCTGAAATGGGCAGGGGGGAAATACC...       TGA   \n",
       "WP_010655730.1   ATGTCTTCTCCAATCATTCCATGGATGGGCGGTAAGCGGCGTTTGG...       TGA   \n",
       "WP_005554455.1   ATGCCTGAACCCGCTAAACCCGCTACGCCAGCCAAAAGCCGCCCGT...       TGA   \n",
       "WP_060836405.1   ATGCGAAAAACCGTCCCCGCAACCCCCGTCGCCCCCTGGATGGGCG...       TGA   \n",
       "WP_062332635.1   ATGTCTCAACTTAAACCTTTTCTAAAATGGGCTGGCGGTAAATCAC...       TAA   \n",
       "WP_005501354.1   ATGAAAAAGCATCGTGCCTTTTTGAAGTGGGCGGGGGGTAAGTATT...       TAA   \n",
       "WP_013359279.1   ATGAAAAAAAATCGCGCTTTCCTGAAATGGGCAGGGGGAAAATTTC...       TAG   \n",
       "WP_061063426.1   ATGAAAACCATCCTGAAATGGGCCGGCAGCAAGTCCGGCCTGATGC...       TAA   \n",
       "WP_057794374.1   ATGATGCAGAAAAAGAACCGGGCCTTTTTAAAGTGGGCCGGTGGCA...       TAA   \n",
       "WP_024708114.1   ATGATCTCGGTGCGGCCTGTGTCGCCGCCGGCCGCCTATCTTGGCG...       TGA   \n",
       "WP_064335703.1   ATGAAAAAAACACGCGCTTTTTTAAAATGGGCCGGGGGAAAATACT...       TAA   \n",
       "\n",
       "                                     DNA3UTRSeq  codonTableBio  \\\n",
       "refSeqProteinId                                                  \n",
       "NP_417846.1      TTCTCAAGGAGAAGCGGATGAAACAGTATT             11   \n",
       "NP_232254.1      TCTTGTGGGCTGAATGGTGGATTTCCTCTC             11   \n",
       "NP_709160.1      TTCTCAAGGAGAAGCGGATGAAACAGTATT             11   \n",
       "NP_461657.1      GTAAATTCACATTACATAATGCACCAACCA             11   \n",
       "NP_462387.1      TTCTCAAGGAGAAGCGGATGAAACAGTATT             11   \n",
       "NP_720943.1      CGAATATAAGTATGGGGGTATTTTTATGAC             11   \n",
       "WP_011098694.1   GAATAAAAATGATGTTGCATGGGAAACCCT             11   \n",
       "WP_011135659.1   CAAGCTTCGTGAAACAGCAAAGGGGCCAAG             11   \n",
       "WP_011201125.1   TCGGTATTGCGATAAATTATAAAAATGAAA             11   \n",
       "NP_793171.1      GACTTTTACCCTAGAGGGGGGGGTGGTAGC             11   \n",
       "WP_011115082.1   GTTGGATTCAGATTAATTTCAATCCTCCCC             11   \n",
       "WP_010944438.1   TCATTACTGAACTAAAATTTATTAAAATAT             11   \n",
       "WP_011284350.1   ATTTTTGTTTTAGTTTAATTATTCGGGCTT              4   \n",
       "WP_011264865.1   ATTGCTATTTGTAATTTAAAAAAAGGACAA              4   \n",
       "WP_011235699.1   ACTGGAATTTAAAAGCATTTAACGACCTTT             11   \n",
       "YP_096088.1      GGGTGTTGCCATGAAGCAGGGTGGTTCATA             11   \n",
       "YP_094690.1      CAAAGATAAGGAAGAGTCTTTACCAGGATT             11   \n",
       "NP_312256.1      TTCTCAAGGAGAAGCGGATGAAACAGTATT             11   \n",
       "YP_002345240.1   GTGAAATTGCTGTCCGGTAATCGCTCAGAC             11   \n",
       "YP_001008117.1   GCGGGGACTGCCGCTCAGTAAAGATTCAAA             11   \n",
       "WP_010994238.1   AATATCGGCATTAATGAATTAATTAATATA             11   \n",
       "WP_011151274.1   CCCCATCGATTCATTCTCTACATTCACACT             11   \n",
       "WP_011160764.1   AAAATCAATTTAAATTTTTGACCTCAACAA             11   \n",
       "WP_034172244.1   AATTAAAATTTTCACTATCGTTTCACATCA             11   \n",
       "WP_041624742.1   AATTAAAATTTTTACCATTGTTTTACATCA             11   \n",
       "WP_011160755.1   GTCAGCTATCTTTTATGTTTTAAATAAATA             11   \n",
       "WP_043743234.1   ACCCGTTCACCGTGGGCCAAACCTAAGTGT             11   \n",
       "WP_043743981.1   ATGGGGAGATCAAGGCCGATTCAAAAGTCT             11   \n",
       "WP_000664222.1   AAGAGGTAGTTAAAGGTGGTGGTGCATCTG             11   \n",
       "WP_011412114.1   GCGGGGCCGCCCCGCCGCAGACTGACCGTA             11   \n",
       "...                                         ...            ...   \n",
       "WP_050692982.1   TCCGAAAATCGACCGCACTTTTACGTTAAA             11   \n",
       "WP_050715369.1   CACGGCGTGCAGCGTGTGCTATATTTCGCG             11   \n",
       "WP_053101747.1   AAAGCACCTATCTGTAGACCATTCTGCTGA             11   \n",
       "WP_053549854.1   GAGGCGGTTATGTGGCCAGGTTATATTGAG             11   \n",
       "WP_062152675.1   TATTAACGTACTAATTAATTTAGCTATTCA             11   \n",
       "WP_062147837.1   TTTGTTATAATCATTAATGAAAAAAATAGT             11   \n",
       "WP_054724231.1   TGATTTCTTAAACGAATTAAGCGAAACTAA             11   \n",
       "WP_004315709.1   ATGCTTTCCAACTCCGATCCCCGGCAAAAG             11   \n",
       "WP_004298811.1   TGCAAGTAAGATAGACACTAATTGGACGAA             11   \n",
       "WP_027787034.1   CGGATGGCTCGCCCGCAGGACAGCTCCTGC             11   \n",
       "WP_055732932.1   ACGCACTAATAAAACAATGCTGAGCACAAA             11   \n",
       "WP_057900243.1   TGTGTAAATGTATAACAAATAGCTCATCAC             11   \n",
       "WP_057179500.1   GCTGGTCCGAATTAATTTGGTATAACTAAG             11   \n",
       "WP_058031670.1   TTTTTATACAAGGTTTGAAAGAGCAATCCT             11   \n",
       "WP_050667307.1   TTCGGCCTCGTCTCTGTCCTGGCTTTGCGG             11   \n",
       "WP_008115051.1   CCAGTACAAATAATACAAACAGTACTACGG             11   \n",
       "WP_062483258.1   TGGCTACAAGCAGGCAAATTTGCACGCCAG             11   \n",
       "WP_010386213.1   TCGCGTTGAATGGCTGGCACGTCATAATAC             11   \n",
       "WP_059307157.1   GCCATCACGACGTTAAAAAGCACGGCGGTG             11   \n",
       "WP_032614792.1   GCCGTTTCGCCCGCCGTTAAACAAAGTTCA             11   \n",
       "WP_010655730.1   CTCAGATCGCGAGTGAAAAGCTTTGTATAT             11   \n",
       "WP_005554455.1   CCGCACTTTTAGTGCCACTAAAACACTGTT             11   \n",
       "WP_060836405.1   GTTCCCCCGGCCCGGTCGGCGTTTAACCGG             11   \n",
       "WP_062332635.1   TCTCTAAACAATTTTATCGGCTTTGATAGT             11   \n",
       "WP_005501354.1   GGCACCATACTGGATTGCTGAAAATCCCGT             11   \n",
       "WP_013359279.1   TCTGTTTCGTCCGCGACCAACGCATTACGC             11   \n",
       "WP_061063426.1   TGACTCAGGTCTACGCATATCCCTGGAATG             11   \n",
       "WP_057794374.1   CTAATAACCGGTTAGCCACTTCACAAACAT             11   \n",
       "WP_024708114.1   GCGGGTGTTCATCCGAGGTGACGCCTGATC             11   \n",
       "WP_064335703.1   CAGACCATGCGGCGCCCAAGGGCGCCGTTT             11   \n",
       "\n",
       "                                                        proteinSeq  \\\n",
       "refSeqProteinId                                                      \n",
       "NP_417846.1      MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...   \n",
       "NP_232254.1      MKKQRAFLKWAGGKYSLVEDIQRHLPEARELVEPFVGAGSVFLNTD...   \n",
       "NP_709160.1      MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...   \n",
       "NP_461657.1      MSTILKWAGNKTAIMAELKKHLPEGPRLVEPFAGSCAVMIETDYPH...   \n",
       "NP_462387.1      MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...   \n",
       "NP_720943.1      MSEQIKLANTTLQPFTKWTGGKRQLLPILRSYMPEKYNCYFEPFIG...   \n",
       "WP_011098694.1   MKKNKIILAAPVLKWVGGKRQLMGEIKKVLPKTYTAYYEPFIGGGA...   \n",
       "WP_011135659.1   MLQIPASPIVPWIGGKRRLAKHILPLFPEHTCYVEPFCGAAALYFL...   \n",
       "WP_011201125.1   MSHSGKTKHGLKHRSFLKWAGGKYRLTDNINNLFPKRRKCLVEPFV...   \n",
       "NP_793171.1      MSTPIIPWMGGKRRLADRLIPLFPPHECYVEVFAGGAALYFMRPQA...   \n",
       "WP_011115082.1   MNAIVSLKTLIPPLKIQGIKSKIIPHIRDIFQTWNKKGIYFEPFMG...   \n",
       "WP_010944438.1   MNNNQRAFLKWAGGKYRLSIEIRKFLPPQSTQGVCFVEPFVGAGSV...   \n",
       "WP_011284350.1   MSKSELKPFVKWVGGKTQLINVILSLLPKNFNSYIEPFLGGGALFL...   \n",
       "WP_011264865.1   MNILLENFGTRLKELRKSKQISQKELSESTGIVREQISRIENGLVN...   \n",
       "WP_011235699.1   MKKNRAFLKWAGGKYSLIEPITARLPQGKKLIEPFVGAGSVFLNTD...   \n",
       "YP_096088.1      MIKNSEKESIIRIRPFLKWAGGKFQIINKIRASLPKGARLIEPFVG...   \n",
       "YP_094690.1      MVKIRPFLKWAGSKYNCLNEIISSLPTGQRLIEPFAGSGVVFMNTS...   \n",
       "NP_312256.1      MKKNRAFLKWAGGKYPLLDDIKRHLPKGECLVEPFVGAGSVFLNTD...   \n",
       "YP_002345240.1   MKKNRAFLKWAGGKYPLVDDIRRHLPAGDCLIEPFVGAGSVFLNTE...   \n",
       "YP_001008117.1   MKKNRAFLKWAGGKYPLVDDIRRHLPAGDCLIEPFVGAGSVFLNTE...   \n",
       "WP_010994238.1   MVSQIPKETCPRPFLKWAGGKSRLIPQYLSHLPKNYRTYHEPFLGG...   \n",
       "WP_011151274.1   MKKQRAFLKWAGGKYGLVEDIQKHLPPARKLVEPFVGAGSVFLNTD...   \n",
       "WP_011160764.1   MSFYFIYGLRNTKMQKFINWVGCKKKLLPHLLPLIPKKFKTYYEPF...   \n",
       "WP_034172244.1   MDFRQIIANSQQGDFLFIDPPYYYEGIKDKDFYQKPFSFTDHQQLA...   \n",
       "WP_041624742.1   MDFCKVIANSQKGDFLFVDPPYYYEGIKDKDFYQKPFSFEDHKLLA...   \n",
       "WP_011160755.1   MKINWIGCKKKLLPQIISHLPKQFKTYYEPFLGSGVLYFYLKSSKA...   \n",
       "WP_043743234.1   MTINLDHRSVTPVDPAVGYIGGKRNLAGAIVPLIQSIPHTCYAEPF...   \n",
       "WP_043743981.1   MTINPHHERIAPVKPAAAYIGGKRNLSRDLIQLIQSTPHDVYAEAF...   \n",
       "WP_000664222.1   MKEQSLPIVPWIGGKRRLAKHILPLFPAHTCYVEPFCGAAALYFLK...   \n",
       "WP_011412114.1   MKKNRAFLKWAGGKYPLIEDIRRHLPEGDCLIEPFVGAGSVFLNTD...   \n",
       "...                                                            ...   \n",
       "WP_050692982.1   MPEPAKPVVKTKSRPFLKWAGGKFRLTDEINRLLPKRKQCLVEPFV...   \n",
       "WP_050715369.1   MPTPIIPWIGGKRRLAEHLIPRFPRHTCYVEVFAGGAALYFLKNPS...   \n",
       "WP_053101747.1   MAKPIIPWMGGKRRLVKQILPKFPEHQSYIEPFAGGAAMLFMKDPS...   \n",
       "WP_053549854.1   MHRPIIPWIGGKRRLAKHLLPLFPKHTCYVEPFCGAAALFFLKPPA...   \n",
       "WP_062152675.1   MTNKHSATASHIKPFIKWVGGKRGLLDELLKRLPQHFNHYFEPFVG...   \n",
       "WP_062147837.1   MKLPHPIPYQGSKRSLAIHILQYFPSHISRLIEPFAGSAAITIASA...   \n",
       "WP_054724231.1   MKTAKPFLKWAGGKSQLINEIENSLPQDFKKKEFTFIEPFVGSGAV...   \n",
       "WP_004315709.1   MEKRTYYNEGNPNNITRAALFIFFMRTCYNGIYSVNHSGKLSVTFG...   \n",
       "WP_004298811.1   MSSRRTNKLVVPFVKWVGGKRQILSEIEKYLPSKFATYYEPFIGGG...   \n",
       "WP_027787034.1   MANPIIPWIGGKRRLADHLIPRFPAHDCYVEVFAGGAALYFMRPPA...   \n",
       "WP_055732932.1   MKKTRAFLKWAGGKYGLIEDIQQQLPEGDLLIEPFVGAASVFLNTN...   \n",
       "WP_057900243.1   MFSDQNGKEETPLKPFLKWAGGKYKLVPHIKELLPKTRRLIEPFVG...   \n",
       "WP_057179500.1   MNNKHRAFLKWAGGKYSLCDVISKMLPNERRLIEPFVGAGSVFLNT...   \n",
       "WP_058031670.1   MNRRSPLKWAGGKKRIIEEIKANLPLKGKQRLIEPFVGGGSVFLNI...   \n",
       "WP_050667307.1   MKKTRAFLKWAGGKFSLIEEIAERLPEGRVLLEPFVGAGSVFLNTD...   \n",
       "WP_008115051.1   MQQKSRAFLKWAGGKYSLVEDINARLNQASKHADTLVEPFVGAGSV...   \n",
       "WP_062483258.1   MDNKKSRAFLKWAGGKYSLVSEIRSLLPDGERLIEPFVGAGSVFLN...   \n",
       "WP_010386213.1   MQKKTRAFLKWAGGKYALVEEITKRLQAANEEAETLVEPFVGAGSV...   \n",
       "WP_059307157.1   MSTILKWAGNKTAVMPELIKHLPAGPRLVEPFAGSCAVMMATDYPH...   \n",
       "WP_032614792.1   MKKNRAFLKWAGGKYPLLDDIKQRLPAGECLVEPFVGAGSVFLNTD...   \n",
       "WP_010655730.1   MSSPIIPWMGGKRRLADRLIPLFPPHECYVEVFAGGAALYFLRPQG...   \n",
       "WP_005554455.1   MPEPAKPATPAKSRPFLKWAGGKYRLMDEINRLLPKRKQCLVEPFV...   \n",
       "WP_060836405.1   MRKTVPATPVAPWMGGKKALHRRIIERIEAIPHRTYVEPFLGMGGV...   \n",
       "WP_062332635.1   MSQLKPFLKWAGGKSQLIGDIQARLPNYTANQDFCMIEPFLGGGAV...   \n",
       "WP_005501354.1   MKKHRAFLKWAGGKYSLVEDIRRHLPEGKKLIEPFVGAGSVFLNTD...   \n",
       "WP_013359279.1   MKKNRAFLKWAGGKFPLVEEIHRHLPQGDCLVEPFVGAGSVFLNTE...   \n",
       "WP_061063426.1   MKTILKWAGSKSGLMPELIKHLPAGNRLVEPFAGSCAVMMNTDYPA...   \n",
       "WP_057794374.1   MMQKKNRAFLKWAGGKYSLVEHIQARLPQANKLIEPFVGAGSVFLN...   \n",
       "WP_024708114.1   MISVRPVSPPAAYLGGKKRLAKRVSSLIEQTPHDLYAEPFVGMGGV...   \n",
       "WP_064335703.1   MKKTRAFLKWAGGKYSLVEEIAERLPAGRVLLEPFVGAGSVFLNTD...   \n",
       "\n",
       "                   seed_eggNOG_ortholog  seed_ortholog_evalue  \\\n",
       "refSeqProteinId                                                 \n",
       "NP_417846.1                155864.Z4740         4.600000e-195   \n",
       "NP_232254.1               243277.VC2626         1.800000e-192   \n",
       "NP_709160.1               198214.SF3405         3.600000e-195   \n",
       "NP_461657.1               99287.STM2730         7.100000e-202   \n",
       "NP_462387.1              220341.STY4312         4.300000e-194   \n",
       "NP_720943.1              210007.SMU_504         1.300000e-193   \n",
       "WP_011098694.1          212717.CTC00387         2.400000e-187   \n",
       "WP_011135659.1           243365.CV_2111         3.500000e-182   \n",
       "WP_011201125.1            221988.MS1967         1.600000e-194   \n",
       "NP_793171.1           223283.PSPTO_3388         1.100000e-189   \n",
       "WP_011115082.1            235279.HH0238         9.100000e-201   \n",
       "WP_010944438.1            233412.HD0420         1.200000e-192   \n",
       "WP_011284350.1          262719.MHJ_0623          0.000000e+00   \n",
       "WP_011264865.1          267748.MMOB3450         1.200000e-239   \n",
       "WP_011235699.1            283942.IL2475         6.100000e-200   \n",
       "YP_096088.1              272624.lpg2075         2.100000e-190   \n",
       "YP_094690.1              272624.lpg0654         1.600000e-191   \n",
       "NP_312256.1                155864.Z4740         4.600000e-195   \n",
       "YP_002345240.1             187410.y3937         4.600000e-190   \n",
       "YP_001008117.1            393305.YE3972         7.300000e-190   \n",
       "WP_010994238.1           103690.all0061         1.800000e-193   \n",
       "WP_011151274.1          216895.VV1_1385         5.700000e-192   \n",
       "WP_011160764.1           262768.PAM_409         1.400000e-200   \n",
       "WP_034172244.1           59748.PAa_0250          3.500000e-36   \n",
       "WP_041624742.1           59748.PAa_0250          5.200000e-36   \n",
       "WP_011160755.1           262768.PAM_318          5.000000e-64   \n",
       "WP_043743234.1           342108.amb0481         1.600000e-189   \n",
       "WP_043743981.1           342108.amb1784         6.800000e-188   \n",
       "WP_000664222.1            343509.SG0828         5.600000e-183   \n",
       "WP_011412114.1            343509.SG2306         1.100000e-188   \n",
       "...                                 ...                   ...   \n",
       "WP_050692982.1       634176.NT05HA_2228         5.800000e-196   \n",
       "WP_050715369.1         397945.Aave_1638         7.100000e-140   \n",
       "WP_053101747.1      637616.MDMS009_1001         3.400000e-113   \n",
       "WP_053549854.1         398767.Glov_2979         1.500000e-129   \n",
       "WP_062152675.1        572480.Arnit_2986         1.900000e-103   \n",
       "WP_062147837.1   679189.HMPREF9019_0346          7.800000e-60   \n",
       "WP_054724231.1          992406.RIA_0345         4.800000e-136   \n",
       "WP_004315709.1   679190.HMPREF0650_1951          1.900000e-34   \n",
       "WP_004298811.1      411476.BACOVA_02878         8.700000e-192   \n",
       "WP_027787034.1         339670.Bamb_1850         4.700000e-160   \n",
       "WP_055732932.1         58051.PE36_00684         1.300000e-116   \n",
       "WP_057900243.1         521098.Aaci_2656         5.100000e-118   \n",
       "WP_057179500.1         298386.PBPRA0283         3.300000e-124   \n",
       "WP_058031670.1          167879.CPS_0475         1.100000e-130   \n",
       "WP_050667307.1          380703.AHA_3186         2.300000e-169   \n",
       "WP_008115051.1         234831.PSM_A0313         1.400000e-196   \n",
       "WP_062483258.1        983545.Glaag_0729         1.400000e-125   \n",
       "WP_010386213.1         234831.PSM_A0313         1.100000e-160   \n",
       "WP_059307157.1           220341.STY4636         4.100000e-140   \n",
       "WP_032614792.1      500639.ENTCAN_08385         2.200000e-180   \n",
       "WP_010655730.1      1114970.PSF113_4284         1.600000e-174   \n",
       "WP_005554455.1          694569.D7S_1130         6.600000e-198   \n",
       "WP_060836405.1          272943.RSP_3769         2.300000e-132   \n",
       "WP_062332635.1    553217.ENHAE0001_1514         1.300000e-201   \n",
       "WP_005501354.1        675812.VHA_000396         2.000000e-210   \n",
       "WP_013359279.1         712898.Pvag_2935         5.300000e-190   \n",
       "WP_061063426.1         706191.PANA_3169         8.900000e-120   \n",
       "WP_057794374.1        715451.ambt_16195         6.400000e-198   \n",
       "WP_024708114.1     411684.HPDFL43_05935         5.600000e-128   \n",
       "WP_064335703.1         998088.B565_0972         1.800000e-202   \n",
       "\n",
       "                       ...                           species group_name  \\\n",
       "refSeqProteinId        ...                                                \n",
       "NP_417846.1            ...                                          NaN   \n",
       "NP_232254.1            ...                                          NaN   \n",
       "NP_709160.1            ...                                          NaN   \n",
       "NP_461657.1            ...                                          NaN   \n",
       "NP_462387.1            ...                                          NaN   \n",
       "NP_720943.1            ...                                          NaN   \n",
       "WP_011098694.1         ...                                          NaN   \n",
       "WP_011135659.1         ...                                          NaN   \n",
       "WP_011201125.1         ...                                          NaN   \n",
       "NP_793171.1            ...                   Pseudomonas syringae group   \n",
       "WP_011115082.1         ...                                          NaN   \n",
       "WP_010944438.1         ...                                          NaN   \n",
       "WP_011284350.1         ...                                          NaN   \n",
       "WP_011264865.1         ...                                          NaN   \n",
       "WP_011235699.1         ...                                          NaN   \n",
       "YP_096088.1            ...                                          NaN   \n",
       "YP_094690.1            ...                                          NaN   \n",
       "NP_312256.1            ...                                          NaN   \n",
       "YP_002345240.1         ...          Yersinia pseudotuberculosis complex   \n",
       "YP_001008117.1         ...                                          NaN   \n",
       "WP_010994238.1         ...                                          NaN   \n",
       "WP_011151274.1         ...                                          NaN   \n",
       "WP_011160764.1         ...               Candidatus Phytoplasma asteris   \n",
       "WP_034172244.1         ...               Candidatus Phytoplasma asteris   \n",
       "WP_041624742.1         ...               Candidatus Phytoplasma asteris   \n",
       "WP_011160755.1         ...               Candidatus Phytoplasma asteris   \n",
       "WP_043743234.1         ...                                          NaN   \n",
       "WP_043743981.1         ...                                          NaN   \n",
       "WP_000664222.1         ...                                          NaN   \n",
       "WP_011412114.1         ...                                          NaN   \n",
       "...                    ...                                          ...   \n",
       "WP_050692982.1         ...                                          NaN   \n",
       "WP_050715369.1         ...                                          NaN   \n",
       "WP_053101747.1         ...                                          NaN   \n",
       "WP_053549854.1         ...                                          NaN   \n",
       "WP_062152675.1         ...                                          NaN   \n",
       "WP_062147837.1         ...                                          NaN   \n",
       "WP_054724231.1         ...                                          NaN   \n",
       "WP_004315709.1         ...                                          NaN   \n",
       "WP_004298811.1         ...                                          NaN   \n",
       "WP_027787034.1         ...                 Burkholderia cepacia complex   \n",
       "WP_055732932.1         ...                                          NaN   \n",
       "WP_057900243.1         ...                                          NaN   \n",
       "WP_057179500.1         ...                                          NaN   \n",
       "WP_058031670.1         ...                                          NaN   \n",
       "WP_050667307.1         ...                                          NaN   \n",
       "WP_008115051.1         ...                                          NaN   \n",
       "WP_062483258.1         ...                                          NaN   \n",
       "WP_010386213.1         ...                                          NaN   \n",
       "WP_059307157.1         ...                                          NaN   \n",
       "WP_032614792.1         ...                                          NaN   \n",
       "WP_010655730.1         ...               Pseudomonas chlororaphis group   \n",
       "WP_005554455.1         ...                                          NaN   \n",
       "WP_060836405.1         ...                                          NaN   \n",
       "WP_062332635.1         ...                                          NaN   \n",
       "WP_005501354.1         ...                                          NaN   \n",
       "WP_013359279.1         ...                    Pantoea agglomerans group   \n",
       "WP_061063426.1         ...                    Pantoea agglomerans group   \n",
       "WP_057794374.1         ...                                          NaN   \n",
       "WP_024708114.1         ...                                          NaN   \n",
       "WP_064335703.1         ...                                          NaN   \n",
       "\n",
       "                species group_taxid species subgroup_name  \\\n",
       "refSeqProteinId                                             \n",
       "NP_417846.1                     NaN                   NaN   \n",
       "NP_232254.1                     NaN                   NaN   \n",
       "NP_709160.1                     NaN                   NaN   \n",
       "NP_461657.1                     NaN                   NaN   \n",
       "NP_462387.1                     NaN                   NaN   \n",
       "NP_720943.1                     NaN                   NaN   \n",
       "WP_011098694.1                  NaN                   NaN   \n",
       "WP_011135659.1                  NaN                   NaN   \n",
       "WP_011201125.1                  NaN                   NaN   \n",
       "NP_793171.1                136849.0                   NaN   \n",
       "WP_011115082.1                  NaN                   NaN   \n",
       "WP_010944438.1                  NaN                   NaN   \n",
       "WP_011284350.1                  NaN                   NaN   \n",
       "WP_011264865.1                  NaN                   NaN   \n",
       "WP_011235699.1                  NaN                   NaN   \n",
       "YP_096088.1                     NaN                   NaN   \n",
       "YP_094690.1                     NaN                   NaN   \n",
       "NP_312256.1                     NaN                   NaN   \n",
       "YP_002345240.1            1649845.0                   NaN   \n",
       "YP_001008117.1                  NaN                   NaN   \n",
       "WP_010994238.1                  NaN                   NaN   \n",
       "WP_011151274.1                  NaN                   NaN   \n",
       "WP_011160764.1              85620.0                   NaN   \n",
       "WP_034172244.1              85620.0                   NaN   \n",
       "WP_041624742.1              85620.0                   NaN   \n",
       "WP_011160755.1              85620.0                   NaN   \n",
       "WP_043743234.1                  NaN                   NaN   \n",
       "WP_043743981.1                  NaN                   NaN   \n",
       "WP_000664222.1                  NaN                   NaN   \n",
       "WP_011412114.1                  NaN                   NaN   \n",
       "...                             ...                   ...   \n",
       "WP_050692982.1                  NaN                   NaN   \n",
       "WP_050715369.1                  NaN                   NaN   \n",
       "WP_053101747.1                  NaN                   NaN   \n",
       "WP_053549854.1                  NaN                   NaN   \n",
       "WP_062152675.1                  NaN                   NaN   \n",
       "WP_062147837.1                  NaN                   NaN   \n",
       "WP_054724231.1                  NaN                   NaN   \n",
       "WP_004315709.1                  NaN                   NaN   \n",
       "WP_004298811.1                  NaN                   NaN   \n",
       "WP_027787034.1              87882.0                   NaN   \n",
       "WP_055732932.1                  NaN                   NaN   \n",
       "WP_057900243.1                  NaN                   NaN   \n",
       "WP_057179500.1                  NaN                   NaN   \n",
       "WP_058031670.1                  NaN                   NaN   \n",
       "WP_050667307.1                  NaN                   NaN   \n",
       "WP_008115051.1                  NaN                   NaN   \n",
       "WP_062483258.1                  NaN                   NaN   \n",
       "WP_010386213.1                  NaN                   NaN   \n",
       "WP_059307157.1                  NaN                   NaN   \n",
       "WP_032614792.1                  NaN                   NaN   \n",
       "WP_010655730.1             136842.0                   NaN   \n",
       "WP_005554455.1                  NaN                   NaN   \n",
       "WP_060836405.1                  NaN                   NaN   \n",
       "WP_062332635.1                  NaN                   NaN   \n",
       "WP_005501354.1                  NaN                   NaN   \n",
       "WP_013359279.1            1654067.0                   NaN   \n",
       "WP_061063426.1            1654067.0                   NaN   \n",
       "WP_057794374.1                  NaN                   NaN   \n",
       "WP_024708114.1                  NaN                   NaN   \n",
       "WP_064335703.1                  NaN                   NaN   \n",
       "\n",
       "                species subgroup_taxid  \\\n",
       "refSeqProteinId                          \n",
       "NP_417846.1                        NaN   \n",
       "NP_232254.1                        NaN   \n",
       "NP_709160.1                        NaN   \n",
       "NP_461657.1                        NaN   \n",
       "NP_462387.1                        NaN   \n",
       "NP_720943.1                        NaN   \n",
       "WP_011098694.1                     NaN   \n",
       "WP_011135659.1                     NaN   \n",
       "WP_011201125.1                     NaN   \n",
       "NP_793171.1                        NaN   \n",
       "WP_011115082.1                     NaN   \n",
       "WP_010944438.1                     NaN   \n",
       "WP_011284350.1                     NaN   \n",
       "WP_011264865.1                     NaN   \n",
       "WP_011235699.1                     NaN   \n",
       "YP_096088.1                        NaN   \n",
       "YP_094690.1                        NaN   \n",
       "NP_312256.1                        NaN   \n",
       "YP_002345240.1                     NaN   \n",
       "YP_001008117.1                     NaN   \n",
       "WP_010994238.1                     NaN   \n",
       "WP_011151274.1                     NaN   \n",
       "WP_011160764.1                     NaN   \n",
       "WP_034172244.1                     NaN   \n",
       "WP_041624742.1                     NaN   \n",
       "WP_011160755.1                     NaN   \n",
       "WP_043743234.1                     NaN   \n",
       "WP_043743981.1                     NaN   \n",
       "WP_000664222.1                     NaN   \n",
       "WP_011412114.1                     NaN   \n",
       "...                                ...   \n",
       "WP_050692982.1                     NaN   \n",
       "WP_050715369.1                     NaN   \n",
       "WP_053101747.1                     NaN   \n",
       "WP_053549854.1                     NaN   \n",
       "WP_062152675.1                     NaN   \n",
       "WP_062147837.1                     NaN   \n",
       "WP_054724231.1                     NaN   \n",
       "WP_004315709.1                     NaN   \n",
       "WP_004298811.1                     NaN   \n",
       "WP_027787034.1                     NaN   \n",
       "WP_055732932.1                     NaN   \n",
       "WP_057900243.1                     NaN   \n",
       "WP_057179500.1                     NaN   \n",
       "WP_058031670.1                     NaN   \n",
       "WP_050667307.1                     NaN   \n",
       "WP_008115051.1                     NaN   \n",
       "WP_062483258.1                     NaN   \n",
       "WP_010386213.1                     NaN   \n",
       "WP_059307157.1                     NaN   \n",
       "WP_032614792.1                     NaN   \n",
       "WP_010655730.1                     NaN   \n",
       "WP_005554455.1                     NaN   \n",
       "WP_060836405.1                     NaN   \n",
       "WP_062332635.1                     NaN   \n",
       "WP_005501354.1                     NaN   \n",
       "WP_013359279.1                     NaN   \n",
       "WP_061063426.1                     NaN   \n",
       "WP_057794374.1                     NaN   \n",
       "WP_024708114.1                     NaN   \n",
       "WP_064335703.1                     NaN   \n",
       "\n",
       "                                         species_name_y species_taxid  \\\n",
       "refSeqProteinId                                                         \n",
       "NP_417846.1                            Escherichia coli         562.0   \n",
       "NP_232254.1                             Vibrio cholerae         666.0   \n",
       "NP_709160.1                           Shigella flexneri         623.0   \n",
       "NP_461657.1                         Salmonella enterica       28901.0   \n",
       "NP_462387.1                         Salmonella enterica       28901.0   \n",
       "NP_720943.1                        Streptococcus mutans        1309.0   \n",
       "WP_011098694.1                       Clostridium tetani        1513.0   \n",
       "WP_011135659.1                Chromobacterium violaceum         536.0   \n",
       "WP_011201125.1          [Mannheimia] succiniciproducens      157673.0   \n",
       "NP_793171.1      Pseudomonas syringae group genomosp. 3      251701.0   \n",
       "WP_011115082.1                   Helicobacter hepaticus       32025.0   \n",
       "WP_010944438.1                    [Haemophilus] ducreyi         730.0   \n",
       "WP_011284350.1                 Mycoplasma hyopneumoniae        2099.0   \n",
       "WP_011264865.1                        Mycoplasma mobile        2118.0   \n",
       "WP_011235699.1                    Idiomarina loihiensis      135577.0   \n",
       "YP_096088.1                      Legionella pneumophila         446.0   \n",
       "YP_094690.1                      Legionella pneumophila         446.0   \n",
       "NP_312256.1                            Escherichia coli         562.0   \n",
       "YP_002345240.1                          Yersinia pestis         632.0   \n",
       "YP_001008117.1                  Yersinia enterocolitica         630.0   \n",
       "WP_010994238.1                      Nostoc sp. PCC 7120      103690.0   \n",
       "WP_011151274.1                        Vibrio vulnificus         672.0   \n",
       "WP_011160764.1                Onion yellows phytoplasma      100379.0   \n",
       "WP_034172244.1                Onion yellows phytoplasma      100379.0   \n",
       "WP_041624742.1                Onion yellows phytoplasma      100379.0   \n",
       "WP_011160755.1                Onion yellows phytoplasma      100379.0   \n",
       "WP_043743234.1              Magnetospirillum magneticum       84159.0   \n",
       "WP_043743981.1              Magnetospirillum magneticum       84159.0   \n",
       "WP_000664222.1                     Sodalis glossinidius       63612.0   \n",
       "WP_011412114.1                     Sodalis glossinidius       63612.0   \n",
       "...                                                 ...           ...   \n",
       "WP_050692982.1              Aggregatibacter aphrophilus         732.0   \n",
       "WP_050715369.1               Ottowia sp. oral taxon 894     1658672.0   \n",
       "WP_053101747.1                  Oblitimonas alkaliphila     1697053.0   \n",
       "WP_053549854.1               Desulfuromonas soudanensis     1603606.0   \n",
       "WP_062152675.1                Beggiatoa leptomitiformis      319938.0   \n",
       "WP_062147837.1                Beggiatoa leptomitiformis      319938.0   \n",
       "WP_054724231.1                  Algibacter alginolytica     1736674.0   \n",
       "WP_004315709.1             Bacteroides thetaiotaomicron         818.0   \n",
       "WP_004298811.1                       Bacteroides ovatus       28116.0   \n",
       "WP_027787034.1                     Burkholderia cepacia         292.0   \n",
       "WP_055732932.1                       Agarivorans gilvus      680279.0   \n",
       "WP_057900243.1                 Aneurinibacillus sp. XH2     1450761.0   \n",
       "WP_057179500.1                       Colwellia sp. MT41       58049.0   \n",
       "WP_058031670.1              Pseudoalteromonas phenolica      161398.0   \n",
       "WP_050667307.1                     Aeromonas schubertii         652.0   \n",
       "WP_008115051.1          Pseudoalteromonas issachenkonii      152297.0   \n",
       "WP_062483258.1               Lacimicrobium alkaliphilum     1526571.0   \n",
       "WP_010386213.1                  Pseudoalteromonas rubra       43658.0   \n",
       "WP_059307157.1                 Leclercia adecarboxylata       83655.0   \n",
       "WP_032614792.1                 Leclercia adecarboxylata       83655.0   \n",
       "WP_010655730.1                        Pseudomonas fragi         296.0   \n",
       "WP_005554455.1    Aggregatibacter actinomycetemcomitans         714.0   \n",
       "WP_060836405.1                 Rhodovulum sulfidophilum       35806.0   \n",
       "WP_062332635.1                      Moraxella osloensis       34062.0   \n",
       "WP_005501354.1                       Grimontia hollisae         673.0   \n",
       "WP_013359279.1                      Pantoea agglomerans         549.0   \n",
       "WP_061063426.1                      Pantoea agglomerans         549.0   \n",
       "WP_057794374.1                Alteromonas stellipolaris      233316.0   \n",
       "WP_024708114.1                      Martelella sp. AD-3      686597.0   \n",
       "WP_064335703.1                        Aeromonas veronii         654.0   \n",
       "\n",
       "                                               subspecies_name  \\\n",
       "refSeqProteinId                                                  \n",
       "NP_417846.1                                                NaN   \n",
       "NP_232254.1                                                NaN   \n",
       "NP_709160.1                                                NaN   \n",
       "NP_461657.1                Salmonella enterica subsp. enterica   \n",
       "NP_462387.1                Salmonella enterica subsp. enterica   \n",
       "NP_720943.1                                                NaN   \n",
       "WP_011098694.1                                             NaN   \n",
       "WP_011135659.1                                             NaN   \n",
       "WP_011201125.1                                             NaN   \n",
       "NP_793171.1                                                NaN   \n",
       "WP_011115082.1                                             NaN   \n",
       "WP_010944438.1                                             NaN   \n",
       "WP_011284350.1                                             NaN   \n",
       "WP_011264865.1                                             NaN   \n",
       "WP_011235699.1                                             NaN   \n",
       "YP_096088.1          Legionella pneumophila subsp. pneumophila   \n",
       "YP_094690.1          Legionella pneumophila subsp. pneumophila   \n",
       "NP_312256.1                                                NaN   \n",
       "YP_002345240.1                                             NaN   \n",
       "YP_001008117.1   Yersinia enterocolitica subsp. enterocolitica   \n",
       "WP_010994238.1                                             NaN   \n",
       "WP_011151274.1                                             NaN   \n",
       "WP_011160764.1                                             NaN   \n",
       "WP_034172244.1                                             NaN   \n",
       "WP_041624742.1                                             NaN   \n",
       "WP_011160755.1                                             NaN   \n",
       "WP_043743234.1                                             NaN   \n",
       "WP_043743981.1                                             NaN   \n",
       "WP_000664222.1                                             NaN   \n",
       "WP_011412114.1                                             NaN   \n",
       "...                                                        ...   \n",
       "WP_050692982.1                                             NaN   \n",
       "WP_050715369.1                                             NaN   \n",
       "WP_053101747.1                                             NaN   \n",
       "WP_053549854.1                                             NaN   \n",
       "WP_062152675.1                                             NaN   \n",
       "WP_062147837.1                                             NaN   \n",
       "WP_054724231.1                                             NaN   \n",
       "WP_004315709.1                                             NaN   \n",
       "WP_004298811.1                                             NaN   \n",
       "WP_027787034.1                                             NaN   \n",
       "WP_055732932.1                                             NaN   \n",
       "WP_057900243.1                                             NaN   \n",
       "WP_057179500.1                                             NaN   \n",
       "WP_058031670.1                                             NaN   \n",
       "WP_050667307.1                                             NaN   \n",
       "WP_008115051.1                                             NaN   \n",
       "WP_062483258.1                                             NaN   \n",
       "WP_010386213.1                                             NaN   \n",
       "WP_059307157.1                                             NaN   \n",
       "WP_032614792.1                                             NaN   \n",
       "WP_010655730.1                                             NaN   \n",
       "WP_005554455.1                                             NaN   \n",
       "WP_060836405.1                                             NaN   \n",
       "WP_062332635.1                                             NaN   \n",
       "WP_005501354.1                                             NaN   \n",
       "WP_013359279.1                                             NaN   \n",
       "WP_061063426.1                                             NaN   \n",
       "WP_057794374.1                                             NaN   \n",
       "WP_024708114.1                                             NaN   \n",
       "WP_064335703.1                                             NaN   \n",
       "\n",
       "                subspecies_taxid  superkingdom_name superkingdom_taxid  \n",
       "refSeqProteinId                                                         \n",
       "NP_417846.1                  NaN           Bacteria                2.0  \n",
       "NP_232254.1                  NaN           Bacteria                2.0  \n",
       "NP_709160.1                  NaN           Bacteria                2.0  \n",
       "NP_461657.1              59201.0           Bacteria                2.0  \n",
       "NP_462387.1              59201.0           Bacteria                2.0  \n",
       "NP_720943.1                  NaN           Bacteria                2.0  \n",
       "WP_011098694.1               NaN           Bacteria                2.0  \n",
       "WP_011135659.1               NaN           Bacteria                2.0  \n",
       "WP_011201125.1               NaN           Bacteria                2.0  \n",
       "NP_793171.1                  NaN           Bacteria                2.0  \n",
       "WP_011115082.1               NaN           Bacteria                2.0  \n",
       "WP_010944438.1               NaN           Bacteria                2.0  \n",
       "WP_011284350.1               NaN           Bacteria                2.0  \n",
       "WP_011264865.1               NaN           Bacteria                2.0  \n",
       "WP_011235699.1               NaN           Bacteria                2.0  \n",
       "YP_096088.1              91891.0           Bacteria                2.0  \n",
       "YP_094690.1              91891.0           Bacteria                2.0  \n",
       "NP_312256.1                  NaN           Bacteria                2.0  \n",
       "YP_002345240.1               NaN           Bacteria                2.0  \n",
       "YP_001008117.1          150052.0           Bacteria                2.0  \n",
       "WP_010994238.1               NaN           Bacteria                2.0  \n",
       "WP_011151274.1               NaN           Bacteria                2.0  \n",
       "WP_011160764.1               NaN           Bacteria                2.0  \n",
       "WP_034172244.1               NaN           Bacteria                2.0  \n",
       "WP_041624742.1               NaN           Bacteria                2.0  \n",
       "WP_011160755.1               NaN           Bacteria                2.0  \n",
       "WP_043743234.1               NaN           Bacteria                2.0  \n",
       "WP_043743981.1               NaN           Bacteria                2.0  \n",
       "WP_000664222.1               NaN           Bacteria                2.0  \n",
       "WP_011412114.1               NaN           Bacteria                2.0  \n",
       "...                          ...                ...                ...  \n",
       "WP_050692982.1               NaN           Bacteria                2.0  \n",
       "WP_050715369.1               NaN           Bacteria                2.0  \n",
       "WP_053101747.1               NaN           Bacteria                2.0  \n",
       "WP_053549854.1               NaN           Bacteria                2.0  \n",
       "WP_062152675.1               NaN           Bacteria                2.0  \n",
       "WP_062147837.1               NaN           Bacteria                2.0  \n",
       "WP_054724231.1               NaN           Bacteria                2.0  \n",
       "WP_004315709.1               NaN           Bacteria                2.0  \n",
       "WP_004298811.1               NaN           Bacteria                2.0  \n",
       "WP_027787034.1               NaN           Bacteria                2.0  \n",
       "WP_055732932.1               NaN           Bacteria                2.0  \n",
       "WP_057900243.1               NaN           Bacteria                2.0  \n",
       "WP_057179500.1               NaN           Bacteria                2.0  \n",
       "WP_058031670.1               NaN           Bacteria                2.0  \n",
       "WP_050667307.1               NaN           Bacteria                2.0  \n",
       "WP_008115051.1               NaN           Bacteria                2.0  \n",
       "WP_062483258.1               NaN           Bacteria                2.0  \n",
       "WP_010386213.1               NaN           Bacteria                2.0  \n",
       "WP_059307157.1               NaN           Bacteria                2.0  \n",
       "WP_032614792.1               NaN           Bacteria                2.0  \n",
       "WP_010655730.1               NaN           Bacteria                2.0  \n",
       "WP_005554455.1               NaN           Bacteria                2.0  \n",
       "WP_060836405.1               NaN           Bacteria                2.0  \n",
       "WP_062332635.1               NaN           Bacteria                2.0  \n",
       "WP_005501354.1               NaN           Bacteria                2.0  \n",
       "WP_013359279.1               NaN           Bacteria                2.0  \n",
       "WP_061063426.1               NaN           Bacteria                2.0  \n",
       "WP_057794374.1               NaN           Bacteria                2.0  \n",
       "WP_024708114.1               NaN           Bacteria                2.0  \n",
       "WP_064335703.1               NaN           Bacteria                2.0  \n",
       "\n",
       "[364 rows x 45 columns]"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggResDf3\n",
    "'05DFE'\n",
    "'ECDH10B_3562'\n",
    "df = eggResDf3[eggResDf3['bestOG'] == '05DFE']\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Membrane protein analysis 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import PSORTdb localization database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "database for subcellular localization in bacteria: \n",
    "\n",
    "PSORTdb\n",
    "\n",
    "https://db.psort.org/browse/genome?id=7576\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTPath = p.databasesPath / 'PSORTdb'\n",
    "PSORTPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputPath = analysisCtermPath / 'Membrane'\n",
    "outputPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nrows = None\n",
    "cols1 = ['SeqID', 'Final_Localization', 'Final_Localization_2',\n",
    "       'Final_Localization_Comments', 'Secondary_Localization', 'Final_Score']\n",
    "df1 = pd.read_csv(PSORTPath / 'Computed-Gram_positive-PSORTdb-3.00.tab.gz', sep='\\t', nrows=nrows, usecols=cols1)\n",
    "df2 = pd.read_csv(PSORTPath / 'Computed-Gram_negative-PSORTdb-3.00.tab.gz', sep='\\t', nrows=nrows, usecols=cols1)\n",
    "df = pd.concat([df1, df2]).reset_index(drop=True)\n",
    "# df = df[cols1 + [col for col in df.columns if col not in cols1]]\n",
    "PSORTDf = df\n",
    "print(len(PSORTDf))\n",
    "PSORTDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(PSORTDf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# too slow... the Series creation and concatenation is very slow.\n",
    "\n",
    "def extract_ids(s):\n",
    "    sList = s.split('|')\n",
    "    gi = None\n",
    "    ref = None\n",
    "    for i in range(len(sList)):\n",
    "        if sList[i] == 'gi':\n",
    "            gi = int(sList[i+1])\n",
    "        elif sList[i] == 'ref':\n",
    "            ref = sList[i+1]\n",
    "    return pd.Series({'refSeqProteinId':ref, 'proteinGI':gi})\n",
    "            \n",
    "\n",
    "# PSORTDf2 = pd.concat([PSORTDf['SeqID'][:10000].apply(extract_ids), PSORTDf], axis=1)\n",
    "%timeit -n1 -r1 PSORTDf['SeqID'][:10000].apply(extract_ids)\n",
    "# PSORTDf2.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = re.compile(r'^gi\\|(\\d+)\\|ref\\|(.+)\\|.+$')\n",
    "\n",
    "def extract_ids(s):\n",
    "    m = reg.search(s)\n",
    "    if m:\n",
    "        gi = int(m.group(1))\n",
    "        ref = m.group(2)\n",
    "    else:\n",
    "        gi = None\n",
    "        ref = None\n",
    "    return (ref, gi)\n",
    "            \n",
    "\n",
    "idList = PSORTDf['SeqID'][:].map(extract_ids)\n",
    "PSORTDf2 = PSORTDf\n",
    "PSORTDf2['refSeqProteinId'] = [t[0] for t in idList]\n",
    "PSORTDf2['proteinGI'] = [t[1] for t in idList]\n",
    "PSORTDf2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols0 = ['refSeqProteinId', 'proteinGI', 'SeqID']\n",
    "PSORTDf2 = PSORTDf2[cols0 + [c for c in PSORTDf2.columns if c not in cols0]]\n",
    "PSORTDf2.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are many duplicated proteins in the PSORT table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf2['refSeqProteinId'].duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf2['proteinGI'].duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf2[PSORTDf2['refSeqProteinId'].duplicated(keep=False)].sort_values('refSeqProteinId')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(PSORTDf2))\n",
    "PSORTDf2 = PSORTDf2.drop_duplicates(subset=['refSeqProteinId'])\n",
    "print(len(PSORTDf2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multispeciesAllCDSDf_nr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf3 = multispeciesAllCDSDf_nr.set_index('refSeqProteinId').join(PSORTDf2.set_index('refSeqProteinId'), rsuffix='_PSORT')\n",
    "PSORTDf3.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf3.to_csv(outputPath / 'PSORTDf3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf3 = pd.read_csv(outputPath / 'PSORTDf3.csv')\n",
    "PSORTDf3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multispeciesAllCDSDf_nr['refSeqProteinId'].duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(PSORTDf3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(PSORTDf2))\n",
    "print(PSORTDf2['Final_Score'].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(len(multispeciesAllCDSDf_nr))\n",
    "print(len(PSORTDf3))\n",
    "print((~PSORTDf3['Final_Score'].isnull()).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(PSORTDf3[(~PSORTDf3['Final_Score'].isnull())]['species_name'].unique()))\n",
    "print(len(PSORTDf3['species_name'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(PSORTDf3[(PSORTDf3['Final_Score'].isnull())]['species_name'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf3[PSORTDf3['species_name'].str.contains('Helicobacter')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf2[PSORTDf2['refSeqProteinId'] == 'NP_109689.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf2[PSORTDf2['refSeqProteinId'] == 'WP_010874358.1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I checked on the website of PSORT for the MPN M129 organism, they indeed have all the predictions for all proteins. Downloaded a file for myucoplasma.\n",
    "\n",
    "We get exactly the same intersection between the two sets when joining on the proteinGI.\n",
    "\n",
    "So either\n",
    "\n",
    "1. the database dump table that I downloaded is not updated, or\n",
    "2. I did not extract the ids correctly in some cases.\n",
    "\n",
    "I checked that the 1st protein in MPN is not in the dump tabular files.\n",
    "\n",
    "I think this is because the 2 tabular files contain only the *classical gram-positive* and *classical gram-negative* species, not the new non-classical membrane species. Although they refer to them in the paper on the website, and we can search for these species on the website, I did not find the dump files for them.\n",
    "\n",
    "**So we will select the species that have at least 80% of their protein with a predicted localization.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv(PSORTPath / 'psortb-results.txt', sep='\\t').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf[PSORTDf['SeqID'].str.contains('NP_109689.1')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf[PSORTDf['SeqID'].str.contains('NP_109693')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "it seems that the table dump does not contains the mycoplasma proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "multispeciesAllCDSDf_nr[multispeciesAllCDSDf_nr['species_name'] == 'Mycoplasma pneumoniae M129'].head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTSpeciesDf = PSORTDf3.groupby('species_name')['Final_Localization'].agg(['count', 'size'])\n",
    "PSORTSpeciesDf.columns = ['n_proteins_with_localization', 'n_proteins_tot']\n",
    "PSORTSpeciesDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTSpeciesDf['ratio_proteins_with_localization'] = \\\n",
    "    PSORTSpeciesDf['n_proteins_with_localization'] / PSORTSpeciesDf['n_proteins_tot']\n",
    "PSORTSpeciesDf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTSpeciesDf['ratio_proteins_with_localization'].hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTSpeciesList = PSORTSpeciesDf[PSORTSpeciesDf['ratio_proteins_with_localization'] > 0.8].index.tolist()\n",
    "print(len(PSORTSpeciesList))\n",
    "PSORTSpeciesList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf4 = PSORTDf3.reset_index().set_index('species_name').loc[PSORTSpeciesList].reset_index().set_index('refSeqProteinId')\n",
    "print(len(PSORTDf3))\n",
    "print(len(PSORTDf4))\n",
    "PSORTDf4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "PSORTDf4['Final_Localization'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "PSORTDf4['Final_Localization'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf4['Final_Localization'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"cytoplasmic\", 524093 / 1156640)\n",
    "print(\"unknown\", (70498 + 257899) / 1156640)\n",
    "print(\"cytoplasmic membrane + outer membrane\", (255800 + 14161) / 1156640)\n",
    "print(\"Periplasmic\", (18827) / 1156640)\n",
    "print(\"Cellwall\", (3007) / 1156640)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "1156640 - 70498 - 257899"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "828243 / 1156640"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only keep cytoplasmic, and cytoplasmic membrane + outer membrane together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf4['localization'] = PSORTDf4['Final_Localization'].replace({'Unknown':np.nan, 'Periplasmic':np.nan,\n",
    "                                               'Extracellular':np.nan, 'Cellwall':np.nan,\n",
    "                                               'Cytoplasmic Membrane':'membrane', 'Outer Membrane':'membrane',\n",
    "                                               'Cytoplasmic':'cytoplasmic'\n",
    "                                              })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf4.to_csv(outputPath / 'PSORTDf4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf4 = pd.read_csv(outputPath / 'PSORTDf4.csv')\n",
    "PSORTDf4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(PSORTDf4))\n",
    "PSORTDf5 = PSORTDf4.dropna(subset=['localization'])\n",
    "print(len(PSORTDf5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf5.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf5.to_csv(outputPath / 'PSORTDf5.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PSORTDf5 = pd.read_csv(outputPath / 'PSORTDf5.csv')\n",
    "PSORTDf5.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute biases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(cterminal)\n",
    "from cterminal import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "full_analysis_multispecies(refSeqFolder=str(refSeqPath),\n",
    "                           dataSingleSpeciesFolder=str(outputPath),\n",
    "                           dataSpeciesGroupFolder=str(outputPath),\n",
    "                           plotsFolder=str(outputPath),\n",
    "                           speciesSetName='membrane',\n",
    "                           chooseRandomGenomes=False,\n",
    "                           nRandomGenomes=0,\n",
    "                           assemblyAccessionList=[],\n",
    "                           speciesList=[],\n",
    "                           multispeciesAllCDSDf_nr=PSORTDf5,\n",
    "                           dfFormatOutput='csv',\n",
    "                           skipAnalysisExistingDataFiles=True,\n",
    "                           skipStatisticsAnalysis=False,\n",
    "                           method='allSeqDataframe',\n",
    "                           skipClustering=True,\n",
    "                           codonAnalysis=False,\n",
    "                           categoryAnalysisColumn='localization',\n",
    "                           GOanalysis=False, amiGODf=None, membraneGOTermsDf=None\n",
    "                          )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import biases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "gather all 3 categories plots of C-terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName = 'membrane'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "list((outputPath / speciesSetName).glob('*oddsRatioAADf*'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "groupList = ['all', 'membrane', 'cytoplasmic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import analysis_summarize_results\n",
    "reload(analysis_summarize_results)\n",
    "from analysis_summarize_results import gather_cterm_bias_group_results\n",
    "\n",
    "oddsRatioMembraneDf = \\\n",
    "    gather_cterm_bias_group_results(path=(outputPath / speciesSetName), groupList=groupList, suffix=speciesSetName,\n",
    "                                   onlyCtermPosition=False)\n",
    "\n",
    "columns = list(oddsRatioMembraneDf.columns)\n",
    "columns = columns[-3:] + columns[:-3]\n",
    "oddsRatioMembraneDf = oddsRatioMembraneDf[columns]\n",
    "oddsRatioMembraneDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf.to_csv(outputPath / 'oddsRatioMembraneDf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf['group'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioMembraneDf[['group', 'n_seq', 'n_genomes']].drop_duplicates()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['n_seq'] / df['n_seq'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf2 = oddsRatioMembraneDf.copy().drop(['n_seq', 'n_genomes'], axis=1)\\\n",
    "                        .rename(columns={'group':'localization'})\\\n",
    "                        .set_index(['terminus', 'position from terminus', 'observable', 'localization'])\n",
    "oddsRatioMembraneDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf2.to_csv(outputPath / 'oddsRatioMembraneDf2.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare the average odds ratio over the last 10 residues between the membrane and non-membrane proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf2.loc['C'].sort_index().loc[-10:-1].loc[idx[:, 'oddsRatio', 'membrane'], :].astype(float).mean().loc[aaTable]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf2.loc['C'].sort_index().loc[-10:-1].loc[idx[:, 'oddsRatio', 'cytoplasmic'], :].astype(float).mean().loc[aaTable]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aaHydrophobic = list('AVILMFWY')\n",
    "aaHydrophobic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf2.loc['C'].sort_index().loc[-1].loc[idx['oddsRatio', 'cytoplasmic'], :].astype(float).loc[aaHydrophobic].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioMembraneDf2.loc['C'].sort_index().loc[-1].loc[idx['oddsRatio', 'membrane'], :].astype(float).loc[aaHydrophobic].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot aa biases map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminus = 'C'\n",
    "statName = 'membrane'\n",
    "statName = 'cytoplasmic'\n",
    "statName = 'all'\n",
    "dfFormatOutput = 'csv'\n",
    "outputFilePrefix = simplify_prefix(speciesSetName + '_subset_' + str(statName))\n",
    "plotTitle = outputFilePrefix\n",
    "oddsRatioDf = oddsRatioMembraneDf2.xs(statName, level='localization')\n",
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plotData = compute_oddsratio_plot_data(oddsRatioDf, terminus)\n",
    "maskData = compute_oddsratio_mask_data(oddsRatioDf, pvalueThresholds, terminus)\n",
    "pvalAnnotTable = compute_pvalueAnnotation(oddsRatioDf, pvalueThresholds, terminus)\n",
    "maskDataSmallN = compute_smallN_mask_data(oddsRatioDf, pvalueThresholds, terminus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(cterminal)\n",
    "from cterminal import plot_aa_composition_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotTitle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filenamePlot = os.path.join(outputPath, outputFilePrefix + \"_composition_bias_aa_\" + terminus + \"terminal\")\n",
    "print(filenamePlot)\n",
    "fig = plot_aa_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus, plotTitle, width=1.5)\n",
    "fig.savefig(filenamePlot + \".svg\", dpi=800, bbox_inches=\"tight\")\n",
    "fig.savefig(filenamePlot + \".png\", dpi=800, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot biases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groupList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDf2 = oddsRatioMembraneDf2.copy()\n",
    "position = -1\n",
    "observable = 'log2OddsRatio'\n",
    "aaLabel = 'C-terminal amino acid at position {:d}'.format(position)\n",
    "plotDf2.columns.name = aaLabel\n",
    "mask = plotDf2.loc[idx['C', position, 'BH_multiple_tests']].T.copy()\n",
    "# convert to boolean\n",
    "mask = mask == 'True'\n",
    "print(mask.head())\n",
    "plotDf2 = plotDf2.loc[idx['C', position, observable]].T\n",
    "for col in plotDf2.columns:\n",
    "    plotDf2[col] = pd.to_numeric(plotDf2[col])\n",
    "    \n",
    "# We assign NAN where the BH test is negative\n",
    "plotDf2[~mask] = np.nan\n",
    "\n",
    "plotDf2 = plotDf2.loc[aaTable]\n",
    "# plotDf2.rename(columns={'all':groupList[2], True:groupList[0], False:groupList[1]}, inplace=True)\n",
    "plotDf2 = plotDf2[groupList]\n",
    "plotDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4.5, 2))\n",
    "plotDf2.plot.bar(ax=ax, width=0.65, color=[paletteMembraneCat[c] for c in plotDf2.columns], alpha=1)\n",
    "ax.set_ylim(-2.2, 2.2)\n",
    "ax.set_ylabel('log2(odds ratio)')\n",
    "ax.xaxis.grid(False)\n",
    "ax.tick_params(axis='x', rotation=0)\n",
    "ax.legend_.set_title(False)\n",
    "family = 'Liberation Mono'\n",
    "ticks_font = FontProperties(family=family)\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontproperties(ticks_font)\n",
    "filepath = outputPath / ('membrane_groups_cterminal{:d}_bias_barplot'.format(position))\n",
    "plt.savefig(str(filepath) + '.svg', dpi=800)\n",
    "plt.savefig(str(filepath) + '.png', dpi=800)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDf2 = oddsRatioMembraneDf2.copy()\n",
    "position = -3\n",
    "observable = 'log2OddsRatio'\n",
    "aaLabel = 'C-terminal amino acid at position {:d}'.format(position)\n",
    "plotDf2.columns.name = aaLabel\n",
    "mask = plotDf2.loc[idx['C', position, 'BH_multiple_tests']].T.copy()\n",
    "# convert to boolean\n",
    "mask = mask == 'True'\n",
    "print(mask.head())\n",
    "plotDf2 = plotDf2.loc[idx['C', position, observable]].T\n",
    "for col in plotDf2.columns:\n",
    "    plotDf2[col] = pd.to_numeric(plotDf2[col])\n",
    "    \n",
    "# We assign NAN where the BH test is negative\n",
    "plotDf2[~mask] = np.nan\n",
    "\n",
    "plotDf2 = plotDf2.loc[aaTable]\n",
    "# plotDf2 = plotDf2[groupList]\n",
    "plotDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4.5, 2))\n",
    "plotDf2.plot.bar(ax=ax, width=0.65, color=[paletteMembraneCat[c] for c in plotDf2.columns], alpha=1)\n",
    "ax.set_ylim(-2.2, 2.2)\n",
    "ax.set_ylabel('log2(odds ratio)')\n",
    "ax.xaxis.grid(False)\n",
    "ax.tick_params(axis='x', rotation=0)\n",
    "ax.legend_.set_title(False)\n",
    "family = 'Liberation Mono'\n",
    "ticks_font = FontProperties(family=family)\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontproperties(ticks_font)\n",
    "\n",
    "filepath = outputPath / ('membrane_groups_cterminal{:d}_bias_barplot'.format(position))\n",
    "plt.savefig(str(filepath) + '.svg', dpi=800)\n",
    "plt.savefig(str(filepath) + '.png', dpi=800)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bulk frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDf2 = oddsRatioMembraneDf2.copy()\n",
    "position = 0\n",
    "observable = 'count'\n",
    "aaLabel = 'amino acid'\n",
    "plotDf2.columns.name = aaLabel\n",
    "plotDf2 = plotDf2.loc[idx['bulk', position, observable]].T\n",
    "for col in plotDf2.columns:\n",
    "    plotDf2[col] = pd.to_numeric(plotDf2[col])\n",
    "plotDf2 = plotDf2.loc[aaTable]\n",
    "# plotDf2.rename(columns={'all':groupList[2], True:groupList[0], False:groupList[1]}, inplace=True)\n",
    "plotDf2 = plotDf2[groupList]\n",
    "plotDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "countTot = plotDf2.sum(axis=0)\n",
    "print(countTot)\n",
    "plotDf3 = plotDf2 / countTot\n",
    "plotDf3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4.5, 2))\n",
    "(100*plotDf3).plot.bar(ax=ax, width=0.65, color=[paletteMembraneCat[c] for c in plotDf2.columns])\n",
    "ax.set_ylabel('bulk frequency (%)')\n",
    "ax.xaxis.grid(False)\n",
    "ax.legend_.set_title(False)\n",
    "ax.tick_params(axis='x', rotation=0)\n",
    "family = 'Liberation Mono'\n",
    "ticks_font = FontProperties(family=family)\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontproperties(ticks_font)\n",
    "\n",
    "filepath = outputPath / ('membrane_groups_bulk_aa_frequency_barplot')\n",
    "plt.savefig(str(filepath) + '.svg', dpi=800)\n",
    "plt.savefig(str(filepath) + '.png', dpi=800)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### aa pair epistatis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "statName = 'membrane'\n",
    "# statName = 'cytoplasmic'\n",
    "dfFormatOutput = 'csv'\n",
    "outputFilePrefix = simplify_prefix(speciesSetName + '_subset_' + statName)\n",
    "print(outputFilePrefix)\n",
    "\n",
    "filename = outputPath / 'membrane' / (outputFilePrefix + \"_oddsRatioDfTableAApair.csv\")\n",
    "print(filename)\n",
    "oddsRatioDfTableAApair = pd.read_csv(filename, index_col=[0,1,2,3])\n",
    "\n",
    "filename = outputPath / 'membrane' / (outputFilePrefix + \"_oddsRatioDfAApair.csv\")\n",
    "print(filename)\n",
    "oddsRatioDfAApair = pd.read_csv(filename, index_col=[0,1,2,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDfTableAApair.T.copy()\n",
    "for col in df.columns:\n",
    "    df[col] = pd.to_numeric(df[col], errors='ignore')\n",
    "oddsRatioDfTableAApairT = df.copy()\n",
    "oddsRatioDfTableAApair = oddsRatioDfTableAApairT.T.copy()\n",
    "oddsRatioDfTableAApairT.dtypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDfTableAApair.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop the rare amino acid U and ambiguous letter X\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApairT.T.copy()\n",
    "oddsRatioDfTableAApair2.columns.name = 'aa1'\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair2[~oddsRatioDfTableAApair2.index.get_level_values('aa0')\\\n",
    "                                                  .str.contains(r'[XU]')]\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair2.loc[:, ~oddsRatioDfTableAApair2.columns.str.contains(r'[XUB]')]\n",
    "# Sort the aa columns\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair2[aaTable]\n",
    "\n",
    "# Sort the aa rows\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair2.copy()\n",
    "oddsRatioDfTableAApair3['sortby'] = oddsRatioDfTableAApair3.index.get_level_values(1).map(dict(zip(aaTable, range(len(aaTable)))))\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.reset_index(level=1)\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.dropna(subset=['sortby'])\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.set_index('sortby', append=True)\\\n",
    "                                                 .reorder_levels(['terminus', 'sortby', 'statistical_test', 'observable'])\\\n",
    "                                                 .sort_index()\n",
    "oddsRatioDfTableAApair3 = oddsRatioDfTableAApair3.reset_index(level='sortby').set_index('aa0', append=True)\\\n",
    "                                                 .reorder_levels(['terminus', 'aa0', 'statistical_test', 'observable'])\\\n",
    "                                                 .drop('sortby', axis=1)\n",
    "oddsRatioDfTableAApair2 = oddsRatioDfTableAApair3.copy()\n",
    "# print(oddsRatioDfTableAApair3.index)\n",
    "oddsRatioDfTableAApair3.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioMembraneDf[oddsRatioMembraneDf['group'] == statName]\\\n",
    "              .set_index(['terminus', 'position from terminus', 'observable']).sort_index()\\\n",
    "              .drop(['group', 'n_seq', 'n_genomes'], axis=1)\n",
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "freq1 = oddsRatioDf.loc[('C', -1, 'count')].astype(np.int64).copy()\n",
    "freq1 = freq1 / freq1.sum()\n",
    "freq2 = oddsRatioDf.loc[('C', -2, 'count')].astype(np.int64).copy()\n",
    "freq2 = freq2 / freq2.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "freq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rowList = []\n",
    "for aa2 in aaTable:\n",
    "    row = (freq2.loc[aa2]*freq1).to_frame().T[aaTable]\n",
    "    row.index = [aa2]\n",
    "    rowList.append(row)\n",
    "\n",
    "df = pd.concat(rowList)\n",
    "df.index.name = 'aa-2'\n",
    "df.columns.name = 'aa-1'\n",
    "df = df.stack()\n",
    "df.index = df.index.map(lambda x: ''.join(x))\n",
    "freqRel12ProductDf = df\n",
    "freqRel12ProductDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df1 = oddsRatioDfTableAApair3.loc[idx['C', :, 'contingency', 'count']]\n",
    "df1.index = df1.index.droplevel([0, 2, 3])\n",
    "df1.index.name = 'aa-2'\n",
    "df1.columns.name = 'aa-1'\n",
    "\n",
    "df1 = df1.stack()\n",
    "df1.index = df1.index.map(lambda x: ''.join(x))\n",
    "ctermSubseqFreq = df1\n",
    "ctermSubseqFreq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctermSubseqFreq.get('KX', 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "binomial_chi_square_test(0.03*100, 100, 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "family_wise_FDR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subSeqList = ctermSubseqFreq.index\n",
    "computeMultipleTestsCorrection = True\n",
    "\n",
    "# This dataframe is organized with the list of subsequences as index\n",
    "terminiList = ['C']\n",
    "methodList = ['theoretical']\n",
    "obsList = ['count','theoreticalProb','NexpectedSubseq','log2OddsRatio','oddsRatio','pvalue']\n",
    "indexNameList = ['terminus','subseq','statistical_test','observable']\n",
    "multiIndex = pd.MultiIndex.from_product([terminiList, subSeqList, methodList, obsList],\n",
    "                                        names=indexNameList)\n",
    "oddsRatioDf = pd.DataFrame(index=multiIndex, columns=['value'])\n",
    "oddsRatioDf.sort_index(level=0, inplace=True)\n",
    "\n",
    "# For amino acid pairs: this dataframe is organized with the first amino acid\n",
    "# as row index and second amino acid as column index.\n",
    "multiIndex = pd.MultiIndex.from_product([terminiList, aaTable, methodList, obsList],\n",
    "                                        names=['terminus','aa0','statistical_test','observable'])\n",
    "oddsRatioPairsDf = pd.DataFrame(index=multiIndex, columns=aaTable)\n",
    "# It is better not to sort the dataframe because we want to keep the custom order for the aa table\n",
    "\n",
    "i = 0\n",
    "for terminus in ['C']:\n",
    "    terminalSize = None\n",
    "    observedFreq = None\n",
    "    if terminus == 'C':\n",
    "        terminalSize = cTerminalSize\n",
    "        observedFreq = ctermSubseqFreq\n",
    "\n",
    "    for subseq in subSeqList:\n",
    "\n",
    "        i += 1\n",
    "        NobservedSubseq = observedFreq.get(subseq, 0)\n",
    "        NobservedTot = sum(observedFreq.values)\n",
    "        theoreticalProb = freqRel12ProductDf.loc[subseq]    # !!!!!!!!!\n",
    "        oddsRatio = ((NobservedSubseq/NobservedTot) / theoreticalProb) if theoreticalProb > 0 else None\n",
    "        log2OddsRatio = log2(oddsRatio)\n",
    "        pvalue = binomial_chi_square_test(NobservedSubseq, NobservedTot, theoreticalProb)\n",
    "\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','count'] = NobservedSubseq\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','theoreticalProb'] = theoreticalProb\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','NexpectedSubseq'] = theoreticalProb*NobservedTot\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','oddsRatio'] = oddsRatio\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','log2OddsRatio'] = log2OddsRatio\n",
    "        oddsRatioDf.loc[terminus,subseq,'theoretical','pvalue'] = pvalue\n",
    "\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','count'),subseq[1]] = NobservedSubseq\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','theoreticalProb'),subseq[1]] = theoreticalProb\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','NexpectedSubseq'),subseq[1]] = theoreticalProb*NobservedTot\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','oddsRatio'),subseq[1]] = oddsRatio\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','log2OddsRatio'),subseq[1]] = log2OddsRatio\n",
    "        oddsRatioPairsDf.loc[(terminus,subseq[0],'theoretical','pvalue'),subseq[1]] = pvalue\n",
    "\n",
    "\n",
    "if computeMultipleTestsCorrection:\n",
    "    dfList = []\n",
    "    for terminus in ['C']:\n",
    "        # Multiple test correction within the biases of C-terminal at all positions for all a.a.\n",
    "        df = oddsRatioDf.xs(terminus, level='terminus', drop_level=False)\n",
    "        df = df.xs('pvalue', level='observable', drop_level=False)\n",
    "        # We serialize all the values, and drop the NaN\n",
    "        df2 = df.stack().dropna().copy()\n",
    "        reject, pvals_corrected, alphacSidak, alphacBonf = \\\n",
    "            multipletests(df2.values, alpha=family_wise_FDR, method='fdr_bh',\n",
    "                          is_sorted=False, returnsorted=False)\n",
    "        df2 = pd.DataFrame(reject, index=df2.index)\n",
    "        # Stack again the values before merging\n",
    "        df2 = df2.unstack()\n",
    "        df2.columns = df2.columns.droplevel(0)\n",
    "        df2 = df2.rename(index={'pvalue':'BH_multiple_tests'})\n",
    "        dfList.append(df2)\n",
    "    oddsRatioDf = pd.concat([oddsRatioDf] + dfList, axis=0, sort=True)\n",
    "    oddsRatioDf.sort_index(inplace=True)\n",
    "\n",
    "if computeMultipleTestsCorrection:\n",
    "    dfList = []\n",
    "    for terminus in ['C']:\n",
    "        # Multiple test correction within the biases of C-terminal at all positions for all a.a.\n",
    "        df = oddsRatioPairsDf.xs(terminus, level='terminus', drop_level=False)\n",
    "        df = df.xs('pvalue', level='observable', drop_level=False)\n",
    "        # We serialize all the values, and drop the NaN\n",
    "        df2 = df.stack().dropna().copy()\n",
    "        reject, pvals_corrected, alphacSidak, alphacBonf = \\\n",
    "            multipletests(df2.values, alpha=family_wise_FDR, method='fdr_bh',\n",
    "                          is_sorted=False, returnsorted=False)\n",
    "        df2 = pd.DataFrame(reject, index=df2.index)\n",
    "        # Stack again the values before merging\n",
    "        df2 = df2.unstack()\n",
    "        df2.columns = df2.columns.droplevel(0)\n",
    "        df2 = df2.rename(index={'pvalue':'BH_multiple_tests'})\n",
    "        dfList.append(df2)\n",
    "    oddsRatioPairsDf = pd.concat([oddsRatioPairsDf] + dfList, axis=0, sort=True)\n",
    "    oddsRatioPairsDf.sort_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioAApairCtermNormDf = oddsRatioDf\n",
    "oddsRatioAApairCtermNormDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioAApairCtermNormDfTable = oddsRatioPairsDf\n",
    "oddsRatioAApairCtermNormDfTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(outputFilePrefix)\n",
    "\n",
    "filename = outputPath / 'membrane' / (outputFilePrefix + \"_oddsRatioAApairCtermNormDf.csv\")\n",
    "print(filename)\n",
    "oddsRatioAApairCtermNormDf.to_csv(filename)\n",
    "\n",
    "filename = outputPath / 'membrane' / (outputFilePrefix + \"_oddsRatioAApairCtermNormDfTable.csv\")\n",
    "print(filename)\n",
    "oddsRatioAApairCtermNormDfTable.to_csv(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAApairCtermNormDf.csv\")\n",
    "# oddsRatioAApairCtermNormDf = pd.read_csv(filename, index_col=[0,1,2,3])\n",
    "\n",
    "# filename = os.path.join(dataSpeciesGroupFolder, speciesSetName, outputFilePrefix + \"_oddsRatioAApairCtermNormDfTable.csv\")\n",
    "# oddsRatioAApairCtermNormDfTable = pd.read_csv(filename, index_col=[0,1,2,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioAApairCtermNormDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDfTableAApair3.loc[idx['C', :, 'contingency', 'count'], :]\n",
    "countTot = df.sum().sum()\n",
    "countTot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioAApairCtermNormDfTable.loc['C'].loc[idx['M', 'theoretical', 'oddsRatio'], 'M']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminus = 'C'\n",
    "statistical_test = 'theoretical'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData = compute_subseq_oddsratio_plot_data(oddsRatioAApairCtermNormDfTable, terminus, statistical_test)\n",
    "maskData = compute_subseq_oddsratio_mask_data(oddsRatioAApairCtermNormDfTable, pvalueThresholds,\n",
    "                                              terminus, statistical_test, subseq_type='aa')\n",
    "pvalAnnotTable = compute_subseq_pvalueAnnotation(oddsRatioAApairCtermNormDfTable, pvalueThresholds,\n",
    "                                                 terminus, statistical_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotData.columns.name = 'aa1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maskDataSmallN = maskData.copy()\n",
    "maskDataSmallN.loc[:, :] = False\n",
    "maskDataSmallN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(cterminal)\n",
    "from cterminal import plot_seq_pair_composition_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig = plot_seq_pair_composition_map(plotData, maskData, pvalAnnotTable, maskDataSmallN, terminus,\n",
    "                                    statistical_test, plotTitle=None, subseq_type='aa', refCodonTableDf=refCodonTableDf,\n",
    "                                    vmin1=-1.8, vmax1=1.8, width=1.5)\n",
    "filenamePlot = outputPath / 'membrane' / (outputFilePrefix +\n",
    "                                      \"_composition_bias_aa_pairs_\" + terminus + \"terminal\" + \"_CtermNorm\")\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + '.svg', dpi=800, bbox_inches=\"tight\")\n",
    "fig.savefig(str(filenamePlot) + '.png', dpi=800, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### epistasis in the bulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioMembraneDf[oddsRatioMembraneDf['group'] == statName]\\\n",
    "              .set_index(['terminus', 'position from terminus', 'observable']).sort_index()\\\n",
    "              .drop(['group', 'n_seq', 'n_genomes'], axis=1)\n",
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "freqBulk = oddsRatioDf.loc['bulk'].loc[0].loc['count'].astype(np.int64)\n",
    "freqBulk = freqBulk / freqBulk.sum()\n",
    "freqBulk.sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = oddsRatioDfTableAApair2.reorder_levels([0,2,3,1]).loc['bulk'].loc['theoretical'].loc['log2OddsRatio'].astype(float)\n",
    "df.style.background_gradient(cmap=cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[print(aa, df.loc[aa, aa]) for aa in ['C', 'H', 'M', 'W']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "speciesSetName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "fig, ax = plt.subplots(figsize=(6,6))\n",
    "sns.heatmap(df, cmap=cmap, vmin=-1.8, vmax=1.8, square=True, ax=ax, cbar_kws=dict(label='log2(odds ratio)'))\n",
    "ax.set_ylabel(\"a.a. position x\")\n",
    "ax.set_xlabel(\"a.a. position x + 1\")\n",
    "\n",
    "filenamePlot = outputPath / 'membrane' / (outputFilePrefix +\n",
    "                                      \"_composition_bias_aa_pairs_bulk_theoretical\")\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + '.svg', dpi=800, bbox_inches=\"tight\")\n",
    "fig.savefig(str(filenamePlot) + '.png', dpi=800, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Disordered protein regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "References:\n",
    "\n",
    "+ Unusual biophysics of intrinsically disordered proteins - ScienceDirect https://www.sciencedirect.com/science/article/pii/S1570963912002889#bb0185\n",
    "+ Classification of Intrinsically Disordered Regions and Proteins | Chemical Reviews https://pubs.acs.org/doi/full/10.1021/cr400525m\n",
    "+ D2P2: database of disordered protein predictions https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3531159/\n",
    "+ Sequence patterns associated with disordered regions in proteins - Lise - 2005 - Proteins: Structure, Function, and Bioinformatics - Wiley Online Library https://onlinelibrary.wiley.com/doi/full/10.1002/prot.20279\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disorderedPath = analysisCtermPath / 'Disordered_regions'\n",
    "disorderedPath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We cannot be sure that the 4th column is the % consensus between predictors..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disConsensus = pd.read_table(disorderedPath / 'consensus_ranges.tsv.gz', nrows=None, header=None)\n",
    "disConsensus.columns = ['protein_internal_id', 'start', 'end', 'consensus']\n",
    "disConsensus.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disConsensus['consensus'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(disConsensus))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disProtein = pd.read_table(disorderedPath / 'genomes.protein.conv', nrows=None, header=None, sep='\\t',\n",
    "                           error_bad_lines=False, engine='c', encoding='utf-8')\n",
    "disProtein.columns = ['genome_internal_id', 'protein_ids', 'protein_internal_id', 'function']\n",
    "print(len(disProtein))\n",
    "disProtein.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disCons2 = pd.merge(disConsensus, disProtein, on='protein_internal_id', how='left')\n",
    "disCons2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disProteinSeq = list(SeqIO.parse(disorderedPath / 'genomes.fa', format='fasta'))\n",
    "print(len(disProteinSeq))\n",
    "disProteinSeq[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disProteinSeq2 = pd.DataFrame([(seq.id, str(seq.seq)) for seq in disProteinSeq[:]], columns=['protein_internal_id', 'protein_seq'])\n",
    "disProteinSeq2['protein_internal_id'] = disProteinSeq2['protein_internal_id'].astype(np.int64)\n",
    "disProteinSeq2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disCons3 = pd.merge(disCons2, disProteinSeq2, on='protein_internal_id', how='left')\n",
    "disCons3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = disCons3.iloc[3].protein_seq\n",
    "seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We extract the disordered regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_disordered_regions(row):\n",
    "    return row['protein_seq'][row['start']-1 : row['end']]\n",
    "    \n",
    "    \n",
    "disCons3['disordered_region'] = disCons3[:].apply(get_disordered_regions, axis=1)\n",
    "disCons3.head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select only bacterial genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disGenomes = pd.read_table(disorderedPath / 'genomes.tab', nrows=None, header=None,\n",
    "                           names=['name', 'genome_internal_id', 2, 'kingdom', 'lineage', 'genome_taxid', 'ftp_path', 'ensembl_page'])\n",
    "# disGenomes.columns = ['protein_internal_id', 'start', 'end', 'consensus']\n",
    "disGenomes.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disGenomes['kingdom'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disCons4 = pd.merge(disCons3, disGenomes[['genome_internal_id', 'kingdom', 'genome_taxid']], on='genome_internal_id', how='left')\n",
    "disCons4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disCons5.to_csv(disorderedPath / 'disCons5.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(disCons4))\n",
    "disCons5 = disCons4[disCons4['kingdom'] == 'B']\n",
    "print(len(disCons5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disRegionAll = ''.join(disCons5['disordered_region'].values)\n",
    "print(len(disRegionAll))\n",
    "disRegionAll[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disFreq = Counter(disRegionAll)\n",
    "disFreq.pop('X')\n",
    "disFreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disFreqRel = {aa: count / sum(disFreq.values()) for aa, count in disFreq.items()}\n",
    "disFreqRel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disFreqRel = pd.Series(disFreqRel, name='disordered regions')\n",
    "disFreqRel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare to the aa freq of the C-terminal at each position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "\n",
    "for pos in range(-20, 0):\n",
    "    df = oddsRatioDf.loc['C', pos, 'count'].copy()\n",
    "    df = df / df.sum()\n",
    "    df.name = 'C-terminal position {:d}'.format(pos)\n",
    "    df = df.astype(np.float)\n",
    "    dfs.append(df)\n",
    "\n",
    "disCtermDf = pd.concat([disFreqRel] + dfs, axis=1)\n",
    "disCtermDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xlabel = 'disordered regions'\n",
    "ylabel = 'C-terminal position -1'\n",
    "fig, ax = plt.subplots(figsize=(4,4))\n",
    "disCtermDf[[xlabel, ylabel]].plot.scatter(x=xlabel, y=ylabel, ax=ax, s=0)\n",
    "for aa, row in disCtermDf.iterrows():\n",
    "    x = row[xlabel]\n",
    "    y = row[ylabel]\n",
    "    ax.annotate(aa, xy=(x, y), xytext=(0, 0), xycoords='data', textcoords='offset points',\n",
    "                ha='center', va='center', fontsize='large', fontweight='bold')\n",
    "\n",
    "ax.axis('equal')\n",
    "ax.set_xlim(0, 0.14)\n",
    "ax.set_ylim(0, 0.14)\n",
    "labelFreq = 'a.a. frequency\\n'\n",
    "ax.set_xlabel(labelFreq + xlabel)\n",
    "ax.set_ylabel(labelFreq + ylabel)\n",
    "\n",
    "pearson = disCtermDf[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.06, 0.94),\n",
    "            xycoords='axes fraction', ha='left', va='top', fontsize='medium',\n",
    "            bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "\n",
    "filename = 'aa_freq_corr_disordered_regions_vs_cterm-1'\n",
    "# fig.savefig(disorderedPath / (filename + '.svg'), dpi=300, bbox_inches='tight')\n",
    "# fig.savefig(disorderedPath / (filename + '.png'), dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xlabel = 'disordered regions'\n",
    "ylabel = 'C-terminal position -2'\n",
    "fig, ax = plt.subplots(figsize=(4,4))\n",
    "disCtermDf[[xlabel, ylabel]].plot.scatter(x=xlabel, y=ylabel, ax=ax, s=0)\n",
    "for aa, row in disCtermDf.iterrows():\n",
    "    x = row[xlabel]\n",
    "    y = row[ylabel]\n",
    "    ax.annotate(aa, xy=(x, y), xytext=(0, 0), xycoords='data', textcoords='offset points',\n",
    "                ha='center', va='center', fontsize='large', fontweight='bold')\n",
    "\n",
    "ax.axis('equal')\n",
    "ax.set_xlim(0, 0.14)\n",
    "ax.set_ylim(0, 0.14)\n",
    "labelFreq = 'a.a. frequency\\n'\n",
    "ax.set_xlabel(labelFreq + xlabel)\n",
    "ax.set_ylabel(labelFreq + ylabel)\n",
    "\n",
    "pearson = disCtermDf[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.06, 0.94),\n",
    "            xycoords='axes fraction', ha='left', va='top', fontsize='medium',\n",
    "            bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "\n",
    "filename = 'aa_freq_corr_disordered_regions_vs_cterm-2'\n",
    "# fig.savefig(disorderedPath / (filename + '.svg'), dpi=300, bbox_inches='tight')\n",
    "# fig.savefig(disorderedPath / (filename + '.png'), dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctermFreq = oddsRatioDf.loc['C', -1, 'count']\n",
    "ctermFreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disFreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([ctermFreq, pd.Series(disFreq)], axis=1, sort=True)\n",
    "df.columns = ['cterm', 'disordered']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats.chi2_contingency(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = (df / 50000).astype(np.int)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats.chi2_contingency(df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats.chi2_contingency(df.T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Structured regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "superFam = pd.read_table(disorderedPath / 'superfamily' / 'ass_09-Feb-2014.tab.gz', nrows=None, header=None)\n",
    "# disConsensus.columns = ['protein_internal_id', 'start', 'end', 'consensus']\n",
    "superFam.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "superFam.columns = ['genome_internal_id', 'protein_ids'] + list(range(2, 11))\n",
    "superFam.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "superFamGenome = pd.read_table(disorderedPath / 'superfamily' / 'genome.tab', nrows=None, header=None,\n",
    "                               names=['genome_internal_id', 'name', 'kingdom', 3, 4, 'ftp_path', 6])\n",
    "superFamGenome.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "superFamGenome.set_index('genome_internal_id').loc['1K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "superFam.set_index('protein_ids').loc['gi|253795548|ref|YP_003038644.1|']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "superFam2 = pd.merge(superFam, superFamGenome, on='genome_internal_id', how='left')\n",
    "superFam2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Protein sequences are not included in the flat files provided to download at the superfamily website.\n",
    "\n",
    "I am downloading the full SQL dump of the database and installing it on simba.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evolution of release factors, EF-P, etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggResDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggOGAnnotDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggOGAnnotDf[eggOGAnnotDf['eggNOG annot'].astype(str).str.contains('release factor', flags=re.I)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we don't find the class II release factor RF2, gene name should be prfC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = eggOGAnnotDf[eggOGAnnotDf['predicted_gene_name'].astype(str).str.contains('prfc', flags=re.I)]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.iloc[0]['eggNOG annot']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Study the distribution of prfC across the taxonomy.\n",
    "\n",
    "For each taxon, group all the genomes in our database. Count how many species do have the homolog, and how many do not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RF3OGid = '05C8A'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonGroup = taxaDf[taxaDf['taxon_name'] == 'Bacillus'].iloc[0]\n",
    "print(\"taxonGroupRef:\", taxonGroup)\n",
    "taxonRank = taxonGroup['rank_name']\n",
    "taxonName = taxonGroup['taxon_name']\n",
    "taxonId = int(taxonGroup['taxon_id'])\n",
    "\n",
    "# Select genomes in the taxon group\n",
    "taxonAssemblyDf = assemblySummaryRepTaxonomyDf[assemblySummaryRepTaxonomyDf[taxonRank + '_taxid'] == taxonId]\n",
    "taxonAssemblyAccessionList = list(taxonAssemblyDf.index)\n",
    "\n",
    "taxonSuffix = simplify_prefix('taxaId_{:d}'.format(taxonId).strip(\"'\"))\n",
    "print(\"output path:\", str(analysisCtermDataPath / taxonSuffix))\n",
    "suffix = taxonSuffix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = taxonAssemblyDf.join(eggResDf.set_index('genome_accession'), rsuffix='_egg')\n",
    "df.index.name = 'genome_accession'\n",
    "# df[['taxid', 'taxid_egg']]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['bestOG'] == RF3OGid]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.reset_index().groupby('genome_accession')['bestOG'].value_counts().unstack(fill_value=0)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2[RF3OGid]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggTaxaDf = assemblySummaryRepTaxonomyDf.join(eggResDf.set_index('genome_accession'), rsuffix='_egg')\n",
    "eggTaxaDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggTaxaDf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "taxonRF3DistList = []\n",
    "for ii, taxonGroup in taxaDf[taxaDf['rank_name'] == 'family'].loc[:].iterrows():\n",
    "    taxonId = int(taxonGroup['taxon_id'])\n",
    "    taxonRank = taxonGroup['rank_name']\n",
    "    taxonName = taxonGroup['taxon_name']\n",
    "    df = eggTaxaDf[eggTaxaDf['{}_name'.format(taxonRank)] == taxonName]\n",
    "    if df['bestOG'].count() > 0:\n",
    "        df.index.name = 'genome_accession'\n",
    "        df2 = df.reset_index().groupby('genome_accession')['bestOG'].value_counts().unstack(fill_value=0)\n",
    "        if RF3OGid not in df2.columns:\n",
    "            nSpeciesContainsRF3 = 0\n",
    "            nSpeciesTot = len(df2)\n",
    "            nSpeciesNotContainsRF3 = len(df2)\n",
    "        else:\n",
    "            df3 = df2[RF3OGid]\n",
    "            nSpeciesTot = len(df2)\n",
    "            nSpeciesContainsRF3 = (df3 >= 1).sum()\n",
    "            nSpeciesNotContainsRF3 = (df3 == 0).sum()\n",
    "        print(ii, taxonName, nSpeciesTot, nSpeciesContainsRF3, nSpeciesNotContainsRF3)\n",
    "        taxonRF3Dist = {'taxid':taxonId, 'rank_name':taxonRank, 'taxon_name':taxonName,\n",
    "                        'nSpeciesTot':nSpeciesTot, 'nSpeciesContainsRF3':nSpeciesContainsRF3,\n",
    "                        'nSpeciesNotContainsRF3':nSpeciesNotContainsRF3, 'ratio':nSpeciesContainsRF3/nSpeciesTot}\n",
    "        taxonRF3DistList.append(taxonRF3Dist)\n",
    "    \n",
    "taxonRF3DistDf = pd.DataFrame(taxonRF3DistList)\n",
    "taxonRF3DistDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonRF3DistDf.to_csv(analysisCtermDataPath / 'taxonRF3DistDf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select taxa with at least 4 genomes\n",
    "taxonRF3DistDf2 = taxonRF3DistDf[taxonRF3DistDf['nSpeciesTot'] >= 4]\n",
    "len(taxonRF3DistDf2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonRF3DistDf2['taxon_name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioAATaxon = pd.read_csv(analysisCtermDataPath / 'oddsRatioAATaxon.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioAATaxon.set_index(['terminus', 'position from terminus', 'observable', 'stop_codon'], inplace=True)\n",
    "oddsRatioAATaxon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf = oddsRatioAATaxon\n",
    "terminus = 'C'\n",
    "terminusPos = -1\n",
    "oddsRatioDf = oddsRatioDf\\\n",
    "              .xs('all', level='stop_codon')\\\n",
    "              .xs(terminusPos, level='position from terminus')\\\n",
    "              .xs(terminus, level='terminus')\n",
    "oddsRatioDf = oddsRatioDf.sort_values('n_seq', ascending=False)\n",
    "# oddsRatioDf = oddsRatioDf.drop(['n_genomes', 'n_seq', 'taxaid', 'taxon_rank'], axis=1)\\\n",
    "oddsRatioDf = oddsRatioDf.set_index('taxon_name', append=True).reorder_levels([1,0]).sort_index()\n",
    "oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oddsRatioDf.index.unique(level=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oddsRatioDf = oddsRatioDf.loc['log2OddsRatio', ['R', 'K', 'T', 'P']].astype(float)\n",
    "# oddsRatioDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonRF3DistDf3 = taxonRF3DistDf2.set_index('taxon_name').join(oddsRatioDf)\n",
    "taxonRF3DistDf3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = taxonRF3DistDf3.loc[idx[:, 'log2OddsRatio'], ['ratio', 'R', 'K', 'T', 'P']].astype(float)\n",
    "df = taxonRF3DistDf3.loc[idx[:, 'log2OddsRatio'], ['ratio'] + aaTable].astype(float)\n",
    "print(len(df))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df, x='ratio', y='P')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df, x='ratio', y='T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df, x='ratio', y='K')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy()\n",
    "labelClass = 'RF3 homolog in >50% of species'\n",
    "df2[labelClass] = np.nan\n",
    "df2.loc[df2['ratio'] < 0.5, labelClass] = False\n",
    "df2.loc[df2['ratio'] >= 0.5, labelClass] = True\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = df2.reset_index(drop=False).drop(['ratio', 'observable'], axis=1)\\\n",
    "      .melt(id_vars=[labelClass, 'taxon_name'], var_name='amino acid at C-terminal', value_name='log2(odds ratio)')\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotsFolder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxPairList = [((aa, True), (aa, False)) for aa in aaTable]\n",
    "boxPairList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.text import Text\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import transforms, lines\n",
    "import matplotlib.transforms as mtransforms\n",
    "from matplotlib.font_manager import FontProperties\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from seaborn.utils import remove_na\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "def stat_test(box_data1, box_data2, test):\n",
    "    testShortName = ''\n",
    "    formattedOutput = None\n",
    "    if test == 'Mann-Whitney':\n",
    "        u_stat, pval = stats.mannwhitneyu(box_data1, box_data2, alternative='two-sided')\n",
    "        testShortName = 'M.W.W.'\n",
    "        formattedOutput = \"MWW RankSum two-sided P_val={:.3e} U_stat={:.3e}\".format(pval, u_stat)\n",
    "    elif test == 't-test_ind':\n",
    "        stat, pval = stats.ttest_ind(a=box_data1, b=box_data2)\n",
    "        testShortName = 't-test_ind'\n",
    "        formattedOutput = \"t-test independent samples, P_val={:.3e} stat={:.3e}\".format(pval, stat)\n",
    "    elif test == 't-test_paired':\n",
    "        stat, pval = stats.ttest_rel(a=box_data1, b=box_data2)\n",
    "        testShortName = 't-test_rel'\n",
    "        formattedOutput = \"t-test paired samples, P_val={:.3e} stat={:.3e}\".format(pval, stat)\n",
    "\n",
    "    return pval, formattedOutput, testShortName\n",
    "\n",
    "\n",
    "\n",
    "def pvalAnnotation_text(x, pvalueThresholds):\n",
    "    singleValue = False\n",
    "    if type(x) is np.array:\n",
    "        x1 = x\n",
    "    else:\n",
    "        x1 = np.array([x])\n",
    "        singleValue = True\n",
    "    # Sort the threshold array\n",
    "    pvalueThresholds = pd.DataFrame(pvalueThresholds).sort_values(by=0, ascending=False).values\n",
    "    xAnnot = pd.Series([\"\" for _ in range(len(x1))])\n",
    "    for i in range(0, len(pvalueThresholds)):\n",
    "        if (i < len(pvalueThresholds)-1):\n",
    "            condition = (x1 <= pvalueThresholds[i][0]) & (pvalueThresholds[i+1][0] < x1)\n",
    "            xAnnot[condition] = pvalueThresholds[i][1]\n",
    "        else:\n",
    "            condition = x1 < pvalueThresholds[i][0]\n",
    "            xAnnot[condition] = pvalueThresholds[i][1]\n",
    "\n",
    "    return xAnnot if not singleValue else xAnnot.iloc[0]\n",
    "\n",
    "\n",
    "def add_stat_annotation(ax,\n",
    "                        data=None, x=None, y=None, hue=None, order=None, hue_order=None,\n",
    "                        boxPairList=None,\n",
    "                        test='Mann-Whitney', textFormat='star', loc='inside', stack=True,\n",
    "                        pvalueThresholds=[[1,\"ns\"], [0.05,\"*\"], [1e-2,\"**\"], [1e-3,\"***\"], [1e-4,\"****\"]],\n",
    "                        useFixedOffset=False, lineYOffsetToBoxAxesCoord=None, lineYOffsetAxesCoord=None,\n",
    "                        lineHeightAxesCoord=0.02, textYOffsetPoints=1,\n",
    "                        color='0.2', linewidth=1.5, fontsize='medium', verbose=1):\n",
    "    \"\"\"\n",
    "    User should use the same argument for the data, x, y, hue, order, hue_order as the seaborn boxplot function.\n",
    "\n",
    "    boxPairList can be of either form:\n",
    "    For non-grouped boxplot: [(cat1, cat2), (cat3, cat4)]\n",
    "    For boxplot grouped by hue: [((cat1, hue1), (cat2, hue2)), ((cat3, hue3), (cat4, hue4))]\n",
    "    \"\"\"\n",
    "\n",
    "    def find_x_position_box(boxPlotter, boxName):\n",
    "        \"\"\"\n",
    "        boxName can be either a name \"cat\" or a tuple (\"cat\", \"hue\")\n",
    "        \"\"\"\n",
    "        if boxPlotter.plot_hues is None:\n",
    "            cat = boxName\n",
    "            hueOffset = 0\n",
    "        else:\n",
    "            cat = boxName[0]\n",
    "            hue = boxName[1]\n",
    "            hueOffset = boxPlotter.hue_offsets[boxPlotter.hue_names.index(hue)]\n",
    "\n",
    "        groupPos = boxPlotter.group_names.index(cat)\n",
    "        boxPos = groupPos + hueOffset\n",
    "        return boxPos\n",
    "\n",
    "\n",
    "    def get_box_data(boxPlotter, boxName):\n",
    "        \"\"\"\n",
    "        boxName can be either a name \"cat\" or a tuple (\"cat\", \"hue\")\n",
    "\n",
    "        Here we really have to duplicate seaborn code, because there is not direct access to the\n",
    "        box_data in the BoxPlotter class.\n",
    "        \"\"\"\n",
    "        if boxPlotter.plot_hues is None:\n",
    "            cat = boxName\n",
    "        else:\n",
    "            cat = boxName[0]\n",
    "            hue = boxName[1]\n",
    "\n",
    "        i = boxPlotter.group_names.index(cat)\n",
    "        group_data = boxPlotter.plot_data[i]\n",
    "\n",
    "        if boxPlotter.plot_hues is None:\n",
    "            # Draw a single box or a set of boxes\n",
    "            # with a single level of grouping\n",
    "            box_data = remove_na(group_data)\n",
    "        else:\n",
    "            hue_level = hue\n",
    "            hue_mask = boxPlotter.plot_hues[i] == hue_level\n",
    "            box_data = remove_na(group_data[hue_mask])\n",
    "\n",
    "        return box_data\n",
    "\n",
    "    fig = plt.gcf()\n",
    "\n",
    "    validList = ['inside', 'outside']\n",
    "    if loc not in validList:\n",
    "        raise ValueError(\"loc value should be one of the following: {}.\".format(', '.join(validList)))\n",
    "    validList = ['t-test_ind', 't-test_paired', 'Mann-Whitney']\n",
    "    if test not in validList:\n",
    "        raise ValueError(\"test value should be one of the following: {}.\".format(', '.join(validList)))\n",
    "\n",
    "    if verbose >= 1 and textFormat == 'star':\n",
    "        print(\"pvalue annotation legend:\")\n",
    "        pvalueThresholds = pd.DataFrame(pvalueThresholds).sort_values(by=0, ascending=False).values\n",
    "        for i in range(0, len(pvalueThresholds)):\n",
    "            if (i < len(pvalueThresholds)-1):\n",
    "                print('{}: {:.2e} < p <= {:.2e}'.format(pvalueThresholds[i][1], pvalueThresholds[i+1][0], pvalueThresholds[i][0]))\n",
    "            else:\n",
    "                print('{}: p <= {:.2e}'.format(pvalueThresholds[i][1], pvalueThresholds[i][0]))\n",
    "        print()\n",
    "\n",
    "    # Create the same BoxPlotter object as seaborn's boxplot\n",
    "    boxPlotter = sns.categorical._BoxPlotter(x, y, hue, data, order, hue_order,\n",
    "                                             orient=None, width=.8, color=None, palette=None, saturation=.75,\n",
    "                                             dodge=True, fliersize=5, linewidth=None)\n",
    "    plotData = boxPlotter.plot_data\n",
    "\n",
    "    xtickslabels = [t.get_text() for t in ax.xaxis.get_ticklabels()]\n",
    "    ylim = ax.get_ylim()\n",
    "    yRange = ylim[1] - ylim[0]\n",
    "\n",
    "    if lineYOffsetAxesCoord is None:\n",
    "        if loc == 'inside':\n",
    "            lineYOffsetAxesCoord = 0.05\n",
    "            if lineYOffsetToBoxAxesCoord is None:\n",
    "                lineYOffsetToBoxAxesCoord = 0.06\n",
    "        elif loc == 'outside':\n",
    "            lineYOffsetAxesCoord = 0.03\n",
    "            lineYOffsetToBoxAxesCoord = lineYOffsetAxesCoord\n",
    "    else:\n",
    "        if loc == 'inside':\n",
    "            if lineYOffsetToBoxAxesCoord is None:\n",
    "                lineYOffsetToBoxAxesCoord = 0.06\n",
    "        elif loc == 'outside':\n",
    "            lineYOffsetToBoxAxesCoord = lineYOffsetAxesCoord\n",
    "    yOffset = lineYOffsetAxesCoord*yRange\n",
    "    yOffsetToBox = lineYOffsetToBoxAxesCoord*yRange\n",
    "\n",
    "    yStack = []\n",
    "    annList = []\n",
    "    testResultList = []\n",
    "    for box1, box2 in boxPairList:\n",
    "\n",
    "        valid = None\n",
    "        groupNames = boxPlotter.group_names\n",
    "        hueNames = boxPlotter.hue_names\n",
    "        if boxPlotter.plot_hues is None:\n",
    "            cat1 = box1\n",
    "            cat2 = box2\n",
    "            hue1 = None\n",
    "            hue2 = None\n",
    "            label1 = '{}'.format(cat1)\n",
    "            label2 = '{}'.format(cat2)\n",
    "            valid = cat1 in groupNames and cat2 in groupNames\n",
    "        else:\n",
    "            cat1 = box1[0]\n",
    "            hue1 = box1[1]\n",
    "            cat2 = box2[0]\n",
    "            hue2 = box2[1]\n",
    "            label1 = '{}_{}'.format(cat1, hue1)\n",
    "            label2 = '{}_{}'.format(cat2, hue2)\n",
    "            valid = cat1 in groupNames and cat2 in groupNames and hue1 in hueNames and hue2 in hueNames\n",
    "\n",
    "\n",
    "        if valid:\n",
    "            # Get position of boxes\n",
    "            x1 = find_x_position_box(boxPlotter, box1)\n",
    "            x2 = find_x_position_box(boxPlotter, box2)\n",
    "            box_data1 = get_box_data(boxPlotter, box1)\n",
    "            box_data2 = get_box_data(boxPlotter, box2)\n",
    "            ymax1 = box_data1.max()\n",
    "            ymax2 = box_data2.max()\n",
    "\n",
    "            pval, formattedOutput, testShortName = stat_test(box_data1, box_data2, test)\n",
    "            testResultList.append({'pvalue':pval, 'testShortName':testShortName,\n",
    "                                   'formattedOutput':formattedOutput, 'box1':box1, 'box2':box2\n",
    "                                  })\n",
    "            if verbose >= 2: print (\"{} v.s. {}: {}\".format(label1, label2, formattedOutput))\n",
    "\n",
    "            if textFormat == 'full':\n",
    "                text = \"{} p < {:.2e}\".format(testShortName, pval)\n",
    "            elif textFormat is None:\n",
    "                text = None\n",
    "            elif textFormat is 'star':\n",
    "                text = pvalAnnotation_text(pval, pvalueThresholds)\n",
    "\n",
    "            if loc == 'inside':\n",
    "                yRef = max(ymax1, ymax2)\n",
    "            elif loc == 'outside':\n",
    "                yRef = ylim[1]\n",
    "\n",
    "            if stack:\n",
    "                if len(yStack) > 0:\n",
    "                    yRef2 = max(yRef, max(yStack))\n",
    "                else:\n",
    "                    yRef2 = yRef\n",
    "            else:\n",
    "                yRef2 = yRef\n",
    "\n",
    "            if len(yStack) == 0:\n",
    "                y = yRef2 + yOffsetToBox\n",
    "            else:\n",
    "                y = yRef2 + yOffset\n",
    "            h = lineHeightAxesCoord*yRange\n",
    "            lineX, lineY = [x1, x1, x2, x2], [y, y + h, y + h, y]\n",
    "            if loc == 'inside':\n",
    "                ax.plot(lineX, lineY, lw=linewidth, c=color)\n",
    "            elif loc == 'outside':\n",
    "                line = lines.Line2D(lineX, lineY, lw=linewidth, c=color, transform=ax.transData)\n",
    "                line.set_clip_on(False)\n",
    "                ax.add_line(line)\n",
    "\n",
    "            if text is not None:\n",
    "                ann = ax.annotate(text, xy=(np.mean([x1, x2]), y + h),\n",
    "                                  xytext=(0, textYOffsetPoints), textcoords='offset points',\n",
    "                                  xycoords='data', ha='center', va='bottom', fontsize=fontsize,\n",
    "                                  clip_on=False, annotation_clip=False)\n",
    "                annList.append(ann)\n",
    "\n",
    "            ax.set_ylim((ylim[0], 1.1*(y + h)))\n",
    "\n",
    "            if text is not None:\n",
    "                plt.draw()\n",
    "                yTopAnnot = None\n",
    "                gotMatplotlibError = False\n",
    "                if not useFixedOffset:\n",
    "                    try:\n",
    "                        bbox = ann.get_window_extent()\n",
    "                        bbox_data = bbox.transformed(ax.transData.inverted())\n",
    "                        yTopAnnot = bbox_data.ymax\n",
    "                    except RuntimeError:\n",
    "                        gotMatplotlibError = True\n",
    "\n",
    "                if useFixedOffset or gotMatplotlibError:\n",
    "                    if verbose >= 1:\n",
    "                        print(\"Warning: cannot get the text bounding box. Falling back to a fixed y offset. Layout may be not optimal.\")\n",
    "                    # We will apply a fixed offset in points, based on the font size of the annotation.\n",
    "                    fontsizePoints = FontProperties(size='medium').get_size_in_points()\n",
    "                    offsetTrans = mtransforms.offset_copy(ax.transData, fig=fig,\n",
    "                                                          x=0, y=1.0*fontsizePoints + textYOffsetPoints, units='points')\n",
    "                    yTopDisplay = offsetTrans.transform((0, y + h))\n",
    "                    yTopAnnot = ax.transData.inverted().transform(yTopDisplay)[1]\n",
    "            else:\n",
    "                yTopAnnot = y + h\n",
    "\n",
    "            yStack.append(yTopAnnot)\n",
    "        else:\n",
    "            raise ValueError(\"boxPairList contains an unvalid box pair.\")\n",
    "            pass\n",
    "\n",
    "\n",
    "    yStackMax = max(yStack)\n",
    "    if loc == 'inside':\n",
    "        ax.set_ylim((ylim[0], 1.03*yStackMax))\n",
    "    elif loc == 'outside':\n",
    "        ax.set_ylim((ylim[0], ylim[1]))\n",
    "\n",
    "    return ax, testResultList\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(6, 3))\n",
    "\n",
    "sns.boxplot(data=df3, x='amino acid at C-terminal', y='log2(odds ratio)', hue=labelClass, fliersize=2, ax=ax)\n",
    "\n",
    "add_stat_annotation(ax=ax, data=df3, x='amino acid at C-terminal', y='log2(odds ratio)', hue=labelClass,\n",
    "                    boxPairList=boxPairList,\n",
    "                    test='t-test_ind', textFormat='star', loc='inside', stack=False, verbose=2)\n",
    "#                     lineYOffsetAxesCoord=0.010, lineYOffsetToBoxAxesCoord=0.005,\n",
    "#                     textYOffsetPoints=0, verbose=2, linewidth=1)\n",
    "\n",
    "filenamePlot = analysisCtermDataPath / 'RF3_composition_bias_aa_Cterminal_taxa_family'\n",
    "print(filenamePlot)\n",
    "fig.savefig(str(filenamePlot) + '.svg', dpi=800, bbox_inches=\"tight\")\n",
    "fig.savefig(str(filenamePlot) + '.png', dpi=800, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ELISA assay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = ctermELISAPath / 'Cterminal_AA_project data.xlsx'\n",
    "df = pd.read_excel(filepath)\n",
    "df = df.iloc[2:22, 0:3]\n",
    "df.columns = ['aa', 'rep1', 'rep2']\n",
    "elisaDf = df.set_index('aa').stack().reset_index()\n",
    "xlabel = 'C-terminal amino acid'\n",
    "ylabel = 'CAT (ng/ml)'\n",
    "elisaDf.columns = [xlabel, 'replicate', ylabel]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elisaDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "width = 2.8\n",
    "fig, ax = plt.subplots(figsize=(width, width/1.5))\n",
    "sns.barplot(ax=ax, data=elisaDf, x=xlabel, y=ylabel, ci=\"sd\", color='0.6', errwidth=1.5)\n",
    "ticks_font = FontProperties(family='Liberation Mono')\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontproperties(ticks_font)\n",
    "filename = 'elisa_quantification'\n",
    "fig.savefig(ctermELISAPath / (filename + '.svg'), dpi=600)\n",
    "fig.savefig(ctermELISAPath / (filename + '.png'), dpi=600)\n",
    "filepath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ELM-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqUp = 'AACAACAAGCGCATCTGGCGGAGATCGCCGAAGGTCTGGTTGAGCGCCATATTCCAGTGCTGATCTCCAATCACGATACGATGTTAACGCGTGAGTGGTATCAGCGCGCAAAATTGCATGTCGTCAAAGTTCGACGCAGTATAAGCAGCAACGGCGGCACACGTAAAAAGGTGGACGAACTGCTGGCTTTGTACAAACCAGGAGTCGTTTCACCCGCGAAAAAA'\n",
    "seqSpacer = 'CAGGCCTTGAAGATA'\n",
    "seqGATC = 'GATCTCGATCTCGATCTCGATC'\n",
    "term517 = 'CGATTAAAGAAGCACTGGCACTAGCTGGTGCTTTTTTAATTGTTT'\n",
    "pMT85 = 'CCTCATAAAGGCCAAGAAGGGCGGA'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import data and compute DAMratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa</th>\n",
       "      <th>i</th>\n",
       "      <th>j</th>\n",
       "      <th>alphabetical_sorted_index</th>\n",
       "      <th>aa_groups_sorted_index</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>codon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CGT</th>\n",
       "      <td>R</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGC</th>\n",
       "      <td>R</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGA</th>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGG</th>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>55</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGA</th>\n",
       "      <td>R</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>58</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGG</th>\n",
       "      <td>R</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>59</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>K</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>K</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>43</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAT</th>\n",
       "      <td>H</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAC</th>\n",
       "      <td>H</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAT</th>\n",
       "      <td>D</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>44</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAC</th>\n",
       "      <td>D</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>45</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAA</th>\n",
       "      <td>E</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>46</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAG</th>\n",
       "      <td>E</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>47</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCT</th>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCC</th>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCA</th>\n",
       "      <td>S</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCG</th>\n",
       "      <td>S</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGT</th>\n",
       "      <td>S</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>56</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGC</th>\n",
       "      <td>S</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACT</th>\n",
       "      <td>T</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>T</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>T</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACG</th>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAT</th>\n",
       "      <td>N</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>40</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>N</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>41</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAA</th>\n",
       "      <td>Q</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAG</th>\n",
       "      <td>Q</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>39</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT</th>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>48</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGC</th>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>49</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCT</th>\n",
       "      <td>P</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCC</th>\n",
       "      <td>P</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCA</th>\n",
       "      <td>P</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCG</th>\n",
       "      <td>P</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCT</th>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCC</th>\n",
       "      <td>A</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>29</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA</th>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCG</th>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>31</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATT</th>\n",
       "      <td>I</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATC</th>\n",
       "      <td>I</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATA</th>\n",
       "      <td>I</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTA</th>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTG</th>\n",
       "      <td>L</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTT</th>\n",
       "      <td>L</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTC</th>\n",
       "      <td>L</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTA</th>\n",
       "      <td>L</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTG</th>\n",
       "      <td>L</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATG</th>\n",
       "      <td>M</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT</th>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC</th>\n",
       "      <td>F</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGA</th>\n",
       "      <td>W</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>50</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGG</th>\n",
       "      <td>W</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>51</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAT</th>\n",
       "      <td>Y</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>32</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAC</th>\n",
       "      <td>Y</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>33</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTT</th>\n",
       "      <td>V</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTC</th>\n",
       "      <td>V</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTA</th>\n",
       "      <td>V</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTG</th>\n",
       "      <td>V</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAA</th>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>34</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAG</th>\n",
       "      <td>*</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>35</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      aa   i  j alphabetical_sorted_index  aa_groups_sorted_index\n",
       "codon                                                            \n",
       "CGT    R   5  7                        52                       0\n",
       "CGC    R   6  7                        53                       1\n",
       "CGA    R   7  7                        54                       2\n",
       "CGG    R   8  7                        55                       3\n",
       "AGA    R  11  7                        58                       4\n",
       "AGG    R  12  7                        59                       5\n",
       "AAA    K  11  5                        42                       6\n",
       "AAG    K  12  5                        43                       7\n",
       "CAT    H   5  5                        36                       8\n",
       "CAC    H   6  5                        37                       9\n",
       "GAT    D  13  5                        44                      10\n",
       "GAC    D  14  5                        45                      11\n",
       "GAA    E  15  5                        46                      12\n",
       "GAG    E  16  5                        47                      13\n",
       "TCT    S   1  3                        16                      14\n",
       "TCC    S   2  3                        17                      15\n",
       "TCA    S   3  3                        18                      16\n",
       "TCG    S   4  3                        19                      17\n",
       "AGT    S   9  7                        56                      18\n",
       "AGC    S  10  7                        57                      19\n",
       "ACT    T   9  3                        24                      20\n",
       "ACC    T  10  3                        25                      21\n",
       "ACA    T  11  3                        26                      22\n",
       "ACG    T  12  3                        27                      23\n",
       "AAT    N   9  5                        40                      24\n",
       "AAC    N  10  5                        41                      25\n",
       "CAA    Q   7  5                        38                      26\n",
       "CAG    Q   8  5                        39                      27\n",
       "TGT    C   1  7                        48                      28\n",
       "TGC    C   2  7                        49                      29\n",
       "...   ..  .. ..                       ...                     ...\n",
       "CCT    P   5  3                        20                      34\n",
       "CCC    P   6  3                        21                      35\n",
       "CCA    P   7  3                        22                      36\n",
       "CCG    P   8  3                        23                      37\n",
       "GCT    A  13  3                        28                      38\n",
       "GCC    A  14  3                        29                      39\n",
       "GCA    A  15  3                        30                      40\n",
       "GCG    A  16  3                        31                      41\n",
       "ATT    I   9  1                         8                      42\n",
       "ATC    I  10  1                         9                      43\n",
       "ATA    I  11  1                        10                      44\n",
       "TTA    L   3  1                         2                      45\n",
       "TTG    L   4  1                         3                      46\n",
       "CTT    L   5  1                         4                      47\n",
       "CTC    L   6  1                         5                      48\n",
       "CTA    L   7  1                         6                      49\n",
       "CTG    L   8  1                         7                      50\n",
       "ATG    M  12  1                        11                      51\n",
       "TTT    F   1  1                         0                      52\n",
       "TTC    F   2  1                         1                      53\n",
       "TGA    W   3  7                        50                      54\n",
       "TGG    W   4  7                        51                      55\n",
       "TAT    Y   1  5                        32                      56\n",
       "TAC    Y   2  5                        33                      57\n",
       "GTT    V  13  1                        12                      58\n",
       "GTC    V  14  1                        13                      59\n",
       "GTA    V  15  1                        14                      60\n",
       "GTG    V  16  1                        15                      61\n",
       "TAA    *   3  5                        34                      62\n",
       "TAG    *   4  5                        35                      63\n",
       "\n",
       "[64 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codonTableBio = Bio.Data.CodonTable.unambiguous_dna_by_id[4]\n",
    "refCodonTableDf = build_refCodonTable(codonTableBio)\n",
    "refCodonTableDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = ctermELMseqPath / 'C-term.weak.txt'\n",
    "with filepath.open() as f:\n",
    "    line = next(f)\n",
    "columns = line.strip(\"# \\n\").split(', ')\n",
    "print(columns)\n",
    "df = pd.read_table(filepath, nrows=None, sep=' ', header=None, skiprows=1)\n",
    "df.columns = columns\n",
    "df['promoter'] = 'weak'\n",
    "\n",
    "filepath = ctermELMseqPath / 'C-term.strong.txt'\n",
    "with filepath.open() as f:\n",
    "    line = next(f)\n",
    "columns = line.strip(\"# \\n\").split(', ')\n",
    "print(columns)\n",
    "df2 = pd.read_table(filepath, nrows=None, sep=' ', header=None, skiprows=1)\n",
    "df2.columns = columns\n",
    "df2['promoter'] = 'strong'\n",
    "\n",
    "damDf = pd.concat([df, df2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 30000\n",
    "damDf.iloc[i:i+10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDf.loc[30003]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['MboI_ratio']/df['DpnI_ratio']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.log10(df['MboI_ratio']/df['DpnI_ratio'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The \"odds ratio\" is actually the log10 of the DAM ratio = CPM_MboI / CPM_DpnI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf['DAMratio'] = damDf['MboI_ratio']/damDf['DpnI_ratio']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf = damDf.rename(columns={'odd_ratio':'log10DAMratio'})\n",
    "damDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf.groupby('promoter')[['Uncut_cnt', 'DpnI_cnt', 'MboI_cnt', 'DpnI_ratio', 'MboI_ratio']].sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The actual nb of total reads is probably larger since there are unmapped reads in the sequencing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "(damDf['MboI_ratio']*58351299).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, the counts seems not to be normalized by a constant factor, but is slightly varying... ???!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(damDf['MboI_cnt']/damDf['MboI_ratio']).head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Whatever...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(damDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codonDict = refCodonTableDf['aa'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDf2 = damDf.copy()\n",
    "damDf2['codon_-2'] = damDf['seq'].map(lambda x: x[:3])\n",
    "damDf2['codon_-1'] = damDf['seq'].map(lambda x: x[3:6])\n",
    "damDf2['codon_1'] = damDf['seq'].map(lambda x: x[9:12])\n",
    "damDf2['codon_2'] = damDf['seq'].map(lambda x: x[12:])\n",
    "damDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf2['aa_-2'] = damDf2['codon_-2'].map(codonDict)\n",
    "damDf2['aa_-1'] = damDf2['codon_-1'].map(codonDict)\n",
    "damDf2['aa_1'] = damDf2['codon_1'].map(codonDict)\n",
    "damDf2['aa_2'] = damDf2['codon_2'].map(codonDict)\n",
    "damDf2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    " GACGGATAAATTAAA\n",
    "-654321012345678\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(-6,9):\n",
    "    damDf2['n_{:d}'.format(i)] = damDf['seq'].map(lambda x: x[i + 6])\n",
    "damDf2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "nReadCutoff = 30\n",
    "subFolder = 'nread_cutoff_{:d}'.format(nReadCutoff)\n",
    "(ctermELMseqPath / subFolder).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read cutoff\n",
    "damDf3 = damDf2[(damDf2['DpnI_cnt'] >= nReadCutoff) |\n",
    "                (damDf2['MboI_cnt'] >= nReadCutoff)]\n",
    "print(damDf2.groupby('promoter').size())\n",
    "print(damDf3.groupby('promoter').size())\n",
    "print()\n",
    "print(len(damDf3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Filter out sequences that contain the GATC motif**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3['seq_extended'] = damDf3['seq'].map(lambda x: seqUp[-3:] + x + seqSpacer[:3])\n",
    "damDf3['contains_GATC'] = damDf3['seq_extended'].str.contains('GATC')\n",
    "damDf3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3.groupby('contains_GATC').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(damDf3))\n",
    "damDf3 = damDf3[~damDf3['contains_GATC']]\n",
    "print(len(damDf3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {},
   "outputs": [],
   "source": [
    "nReadCutoff1 = 60\n",
    "nReadCutoff2 = 8\n",
    "subFolder = 'nread_cutoff_{:d}_{:d}'.format(nReadCutoff1, nReadCutoff2)\n",
    "(ctermELMseqPath / subFolder).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read cutoff\n",
    "damDf3 = damDf2[((damDf2['DpnI_cnt'] >= nReadCutoff1) & (damDf2['MboI_cnt'] >= nReadCutoff2)) |\n",
    "                ((damDf2['MboI_cnt'] >= nReadCutoff1) & (damDf2['DpnI_cnt'] >= nReadCutoff2))]\n",
    "print(damDf2.groupby('promoter').size())\n",
    "print(damDf3.groupby('promoter').size())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3.to_csv(ctermELMseqPath / 'damDf3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>seq</th>\n",
       "      <th>Uncut_cnt</th>\n",
       "      <th>DpnI_cnt</th>\n",
       "      <th>MboI_cnt</th>\n",
       "      <th>DpnI_ratio</th>\n",
       "      <th>MboI_ratio</th>\n",
       "      <th>promoter</th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>codon_-2</th>\n",
       "      <th>...</th>\n",
       "      <th>n_1</th>\n",
       "      <th>n_2</th>\n",
       "      <th>n_3</th>\n",
       "      <th>n_4</th>\n",
       "      <th>n_5</th>\n",
       "      <th>n_6</th>\n",
       "      <th>n_7</th>\n",
       "      <th>n_8</th>\n",
       "      <th>seq_extended</th>\n",
       "      <th>contains_GATC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.052750</td>\n",
       "      <td>GACGGATAAATTAAA</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>129</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>1.902999e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>112.914694</td>\n",
       "      <td>GAC</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAGACGGATAAATTAAACAG</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.825298</td>\n",
       "      <td>CACCGGTAAAGGTGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>1.127161e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>66.880242</td>\n",
       "      <td>CAC</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACACCGGTAAAGGTGTCAG</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.794629</td>\n",
       "      <td>GGTTTCTAATCGCGT</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>286</td>\n",
       "      <td>6.741369e-08</td>\n",
       "      <td>4.201236e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>62.320226</td>\n",
       "      <td>GGT</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAGGTTTCTAATCGCGTCAG</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.709659</td>\n",
       "      <td>CGGACGTAACGGGCA</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>8.636688e-07</td>\n",
       "      <td>weak</td>\n",
       "      <td>51.245900</td>\n",
       "      <td>CGG</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>AAACGGACGTAACGGGCACAG</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.686995</td>\n",
       "      <td>ACTAGTTAATGAAGT</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>8.197534e-07</td>\n",
       "      <td>weak</td>\n",
       "      <td>48.640176</td>\n",
       "      <td>ACT</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAACTAGTTAATGAAGTCAG</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   log10DAMratio              seq  Uncut_cnt  DpnI_cnt  MboI_cnt  \\\n",
       "0       2.052750  GACGGATAAATTAAA          2         0       129   \n",
       "1       1.825298  CACCGGTAAAGGTGT          0         0        76   \n",
       "2       1.794629  GGTTTCTAATCGCGT          3         3       286   \n",
       "3       1.709659  CGGACGTAACGGGCA         10         0        58   \n",
       "4       1.686995  ACTAGTTAATGAAGT         23         0        55   \n",
       "\n",
       "     DpnI_ratio    MboI_ratio promoter    DAMratio codon_-2      ...      n_1  \\\n",
       "0  1.685342e-08  1.902999e-06     weak  112.914694      GAC      ...        A   \n",
       "1  1.685342e-08  1.127161e-06     weak   66.880242      CAC      ...        A   \n",
       "2  6.741369e-08  4.201236e-06     weak   62.320226      GGT      ...        A   \n",
       "3  1.685342e-08  8.636688e-07     weak   51.245900      CGG      ...        A   \n",
       "4  1.685342e-08  8.197534e-07     weak   48.640176      ACT      ...        A   \n",
       "\n",
       "  n_2 n_3 n_4 n_5 n_6 n_7 n_8           seq_extended contains_GATC  \n",
       "0   A   A   T   T   A   A   A  AAAGACGGATAAATTAAACAG         False  \n",
       "1   A   A   G   G   T   G   T  AAACACCGGTAAAGGTGTCAG         False  \n",
       "2   A   T   C   G   C   G   T  AAAGGTTTCTAATCGCGTCAG         False  \n",
       "3   A   C   G   G   G   C   A  AAACGGACGTAACGGGCACAG         False  \n",
       "4   A   T   G   A   A   G   T  AAAACTAGTTAATGAAGTCAG         False  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDf3 = pd.read_csv(ctermELMseqPath / 'damDf3.csv', index_col=0)\n",
    "damDf3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare correlations for cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for subFolder in [ctermELMseqPath / 'nread_cutoff_{:d}'.format(nReadCutoff) for nReadCutoff in [10, 30, 50, 100, 200]]:\n",
    "    filename = 'ELM_seq_codon_pair_correlation_weak_strong.png'\n",
    "    print(subFolder)\n",
    "    display(Image(filename=str(ctermELMseqPath / subFolder / filename), width=300))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for subFolder in [ctermELMseqPath / 'nread_cutoff_{:d}'.format(nReadCutoff) for nReadCutoff in [10, 30, 50, 100, 200]]:\n",
    "    filename = 'ELM_seq_dodecamer_correlation_weak_strong.png'\n",
    "    print(subFolder)\n",
    "    display(Image(filename=str(ctermELMseqPath / subFolder / filename), width=300))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Color palette"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sns.choose_diverging_palette()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABHEAAABLCAYAAAD3RhuZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAA/ZJREFUeJzt27+r1XUcx/H393uO3G6JCA5aYCCJQ5AcXASXyj/A0qWpQaJWEQWniEj/AB3EUVCEFifJwU1o8F6CSw2iYA2lhSCpQXi955xPi/v1Bh8+vr8+HvNneH34cn7wPJyulFICAAAAgFda33oAAAAAAOsTcQAAAAASEHEAAAAAEhBxAAAAABIQcQAAAAASEHEAAAAAEhBxAAAAABIQcQAAAAASEHEAAAAAEhhv5PDS8lKsTae1tjS3aTSO6XzWekYV434Us4HeLSJiNBrHfMj360cxL/PWM6rp+z5KKa1nVNN3fZQY5v266CK61ivq6YZ8uRe6brh3LKUM/n7DfQGWGOjbZkS8uNqA7xdRYsAf6xFdxIC/lkXXR5TZMB9gN+piPh3uw+tHXcymw3x2ERH9uIvZ2nCf3+JbCzGZTNY9t6GIszadxqGvv/rfo151F46fjm8unW09o4pvPz8W53+40npGNSePfBGXb15rPaOaowcPx/Wff2w9o5pD+z6M5d9vt55RzYFdH8S9p3+1nlHFe1t2xL9vtF5Rz7ZYjPG2za1nVNM9WY0du3e2nlHNwzv3493tu1vPqObXP+7E1vE7rWdU8Xj6IMrfW1rPqOb54qP457dNrWdU8+bOZ3H/p+H+8Lt9bx+3bzxpPaOaPR9tjlvf/9l6RhX7P3s7rp+723pGNR9/uSuunvml9YxqPjn1flw8cav1jGq+u/HpS53zdyoAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABLpSSnnZwysrK7GwsFBzDwAAAMBrZXV1NSaTybrnNhRxAAAAAGjD36kAAAAAEhBxAAAAABIQcQAAAAASEHEAAAAAEhBxAAAAABIQcQAAAAASEHEAAAAAEhBxAAAAABIQcQAAAAAS+A/YOoGzNOml1wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1440x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.palplot(sns.diverging_palette(145, 280, s=85, l=25, n=20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABHEAAABLCAYAAAD3RhuZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAA/BJREFUeJzt2z2IVFcYx+H33t11VyIJSvwIYWOECGI1RSohaJVK0qSzkIQgBgJikdqA2AgWVhoINhaBFGmFEAQLK7fICkGwCLpgPohBWEnh7s7MsYj9uoHD8b0+D0x3iv9hGO7MD6YrpZQAAAAA4JXWtx4AAAAAwOZEHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABGa3cvjO0lKsj8e1tjQ3NzsTG5Np6xlVzM30MZ6W1jOqmev7GJfh3m+272My5Pt1XQzzk/efvutiqO9e9+I1VF1ERDfgG5YS3YDvV16L+7VeUUcpEUN+MpSIGO6TIaKULmLI71/pIsqk9Yx6upmI6UB/8/WzUSYbrVfUM/D7dTNzMRnw/Ra274jRaLTpuS1FnPXxOD66+M3/HvWqu376bJz88fvWM6q4/umJ+Pr2zdYzqrny8fE4f3ep9YxqLnx4JK79/lvrGdV89f6h+HnjaesZ1RzfvivuLbReUcfhZxHP9uxsPaOat5/8G9ve3dd6RjXTP/6OffsXW8+o5vHKShx4b2/rGdU8WHkUB97Z0le5NB78OY7FNx+3nlHNw3/eiN39r61nVPPX+sHYsfpT6xnVrC4cjXLv29Yzqpl+8Hms3jrXekYVbx07H/d/ONN6RjWLn1yKO9dOtZ5Rzeiz7+LG5S9az6jmy6u/vNQ5f6cCAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASEDEAQAAAEhAxAEAAABIQMQBAAAASKArpZSXPby8vBzz8/M19wAAAAC8VtbW1mI0Gm16bksRBwAAAIA2/J0KAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAgAREHAAAAIAERBwAAACABEQcAAAAggee0NoSzOesqbAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1440x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.palplot(sns.color_palette(\"BrBG_r\", 20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmapName = \"BrBG_r\"\n",
    "vmaxNuc = 0.1\n",
    "vmaxCodon = 0.2\n",
    "vmaxAApair = 0.3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nucleotide level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "position = -3\n",
    "group = 'n_{:d}'.format(position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfGroup = damDf3.groupby(['promoter', group])[['DAMratio']].mean()\n",
    "damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "damDfGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meanDAM = damDfGroup['log10DAMratio'].groupby(level=0).mean()\n",
    "print(meanDAM)\n",
    "damDfGroup['log10DAMratio_norm'] = damDfGroup['log10DAMratio'] - meanDAM\n",
    "damDfGroup['position'] = position\n",
    "damDfGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.barplot(data=damDfGroup['log10DAMratio_norm'].reset_index(), x=group, y='log10DAMratio_norm', hue='promoter')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We apply the same approach for all the positions, except that we then use the overall average of all DAMratios over all the positions for a specific experiment as the reference for the normalized values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfList = []\n",
    "\n",
    "for position in range(-6,9):\n",
    "    group = 'n_{:d}'.format(position)\n",
    "    damDfGroup = damDf3.groupby(['promoter', group])[['DAMratio']].mean()\n",
    "    damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "    damDfGroup['position'] = position\n",
    "    dfList.append(damDfGroup)\n",
    "    \n",
    "damDfNucl = pd.concat(dfList)\n",
    "damDfNucl.sort_index(inplace=True)\n",
    "damDfNucl.index.names = ['promoter', 'nucleotide']\n",
    "damDfNucl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meanDAM = damDfNucl['DAMratio'].groupby(level=0).mean()\n",
    "log10(meanDAM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfNucl['log10DAMratio_norm'] = damDfNucl['log10DAMratio'] - log10(meanDAM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfNucl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfNucl.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dfList = []\n",
    "for promoter in ['weak', 'strong']:\n",
    "    damPlotDf = damDfNucl.loc[promoter][['position', 'log10DAMratio_norm']].reset_index().pivot(index='nucleotide', columns='position', values='log10DAMratio_norm').copy()\n",
    "    damPlotDf[0] = np.nan\n",
    "    damPlotDf[1] = np.nan\n",
    "    damPlotDf[2] = np.nan\n",
    "    cols = list(damPlotDf.columns)\n",
    "    cols[6] = 'T'\n",
    "    cols[7] = 'A'\n",
    "    cols[8] = 'A'\n",
    "    damPlotDf.columns = cols\n",
    "    damPlotDf.columns.name = 'position'\n",
    "    damPlotDf['promoter'] = promoter\n",
    "    dfList.append(damPlotDf)\n",
    "\n",
    "damPlotDf = pd.concat(dfList).reset_index().set_index(['promoter', 'nucleotide'])\n",
    "damPlotDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "for promoter in ['weak', 'strong']:\n",
    "    aspect = 2\n",
    "    width = 3.0\n",
    "    figsize = (width, width/aspect)\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "    cbarLabel = 'log10(norm.\\nDAMRatio)'\n",
    "    axes = sns.heatmap(data=damPlotDf.loc[promoter], cmap=cmapName, vmin=-vmaxNuc, vmax=vmaxNuc, square=True,\n",
    "                       cbar=True, cbar_kws=dict(label=cbarLabel, shrink=0.5, ticks=[-0.1, 0, 0.1], aspect=10))\n",
    "    cbar_ax = fig.axes[1]\n",
    "    print(cbar_ax)\n",
    "    cbar_ax.tick_params(labelsize='small')\n",
    "    ticks_font = FontProperties(family='Liberation Mono')\n",
    "    for tick in ax.get_yticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "    filename = 'ELM_seq_nucleotide_promoter_{}'.format(promoter)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.png'), dpi=800)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.svg'), dpi=800)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Amino acid level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDfGroup = damDf3.groupby(['promoter', 'aa_-1'])[['DAMratio']].mean()\n",
    "damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "promoter\n",
      "strong    0.992946\n",
      "weak      0.141025\n",
      "Name: log10DAMratio, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>log10DAMratio_norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>promoter</th>\n",
       "      <th>aa_-1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">strong</th>\n",
       "      <th>*</th>\n",
       "      <td>10.169228</td>\n",
       "      <td>1.007288</td>\n",
       "      <td>0.014342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>8.915522</td>\n",
       "      <td>0.950147</td>\n",
       "      <td>-0.042799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>8.023959</td>\n",
       "      <td>0.904389</td>\n",
       "      <td>-0.088557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>11.815256</td>\n",
       "      <td>1.072443</td>\n",
       "      <td>0.079497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>12.086276</td>\n",
       "      <td>1.082292</td>\n",
       "      <td>0.089347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 DAMratio  log10DAMratio  log10DAMratio_norm\n",
       "promoter aa_-1                                              \n",
       "strong   *      10.169228       1.007288            0.014342\n",
       "         A       8.915522       0.950147           -0.042799\n",
       "         C       8.023959       0.904389           -0.088557\n",
       "         D      11.815256       1.072443            0.079497\n",
       "         E      12.086276       1.082292            0.089347"
      ]
     },
     "execution_count": 446,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meanDAM = damDfGroup['log10DAMratio'].groupby(level=0).mean()\n",
    "print(meanDAM)\n",
    "damDfGroup['log10DAMratio_norm'] = damDfGroup['log10DAMratio'] - meanDAM\n",
    "damDfGroup.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fd1bd8dde48>"
      ]
     },
     "execution_count": 388,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAEICAYAAACqMQjAAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3Xl0TPf/P/DnZJmZTCYLSRtJUCGUpB9b+AqCWj75RBQVQoKJ2Peq2hraEkSlqBKU0hIqStFFhfZQtZ5WqDWlmqBEak2CLJN1fn/0Z0rlTu7MZJYkz8c5PUfuzHverztJ5znve9/3fSUajUYDIiKicthYugAiIrJeDAkiIhLEkCAiIkEMCSIiEsSQICIiQQwJIiISxJAgIiJBDAkiIhLEkCAiIkF2li7AWNevX4dUKrV0GUREVUpRUREaNGhQ4fOqfEhIpVJ4eXlZugwioiolMzNT1PN4uImIiAQxJIiISBBDgoiIBFX5cxJERBUpLi5GRkYG1Gq1pUsxO7lcjrp168Le3t6g9gwJIqr2MjIy4OTkhAYNGkAikVi6HLPRaDR48OABMjIy4OPjY9Br8HATEVV7arUabm5uNSogAEAikcDNzc2oERRDgohqhJoWEE8Yu98MCSIiEsSQINJBU1Ko13YiQ6SkpODy5cuWLqNcPHFNpIPEToYb8//z3Pb6712wQDVUXe3atQuhoaFo2rSppUt5DkOCiOhfdu/ejYMHDyI3NxfZ2dmYOHEiEhIS0KBBA0ilUsybNw8zZsxAbm4uSktLMWXKFLRv3x69e/dGmzZtcOXKFfj4+MDNzQ2nTp2CVCrFJ598goKCgufaOTk54ejRo0hNTYWvry/OnTuHTZs2wcbGBgEBAZg+fToSEhJw5swZ5OfnIy4uDo0aNTLbe8GQICIqR35+PjZu3IisrCyEh4ejtLQUEyZMgJ+fH+Lj49GhQwcMGzYMd+7cQWRkJA4cOIC8vDy89tprCAgIQEhICGJiYjB16lQMHToUaWlp+Pbbb8tt16lTJ4SGhkKhUCAhIQG7du2Cg4MDZsyYgePHjwMAGjZsiHfeecfs7wNDgoioHG3btoWNjQ3c3d3h7OyM9PR07bUG6enp6N27NwDAw8MDSqUSWVlZAAB/f38AgLOzs/Ybv7OzMwoLC3W2A4AbN24gKysLY8aMAQDk5eXh5s2bAGDwdQ7G4olrIqJypKamAgDu37+P3NxcuLm5wcbm74/MRo0a4dSpUwCAO3fu4NGjR3B1dQWge8qpUDuJRAKNRoO6devC09MTn332GbZs2YKhQ4eiRYsWAKDt29w4kiAiKsf9+/cxbNgwPH78GHPnzsW8efO0j40dOxazZ8/G999/D7Vajfnz58POruKPU6F2LVq0wNKlS/HRRx8hOjoaKpUKpaWl8Pb2Rs+ePU24lxWTaDQajUUrMFJmZibvJ0EmxdlNVd+lS5fQrFkz0c/fvXs3rl69iunTp5uwKvMpb//FfnbycBMREQkyS0gUFRVh/PjxiIiIwI4dO557fPLkycjIyAAAbN68Gf369YNKpcKGDRvMUR4R0TPCwsKqzSjCWGY5J5GcnIxOnTohIiICo0aNQt++fSGTyVBUVISpU6fiwoV/hu5paWlYs2YNPD09zVEaERHpYJaRRGpqKgICAmBjY4MmTZogPT0dwN8jjOjoaHTo0EH73GvXrmHBggUYMWKEduoXERFZhllGEnl5eVAoFAAABwcH5OfnAwCUSiXatm2LXbt2aZ/brVs3hIeHIyMjA0uWLMHKlSt1vrZarRZ9Q28ifek6sce/u6qjtLQURUVFli7DYkpLSw3+ezVLSCgUChQUFAAACgoKoFQqBZ8bHh4OpVKJpk2bIjs7u8LXlsvlnN1EFsG/u6rj4cOHkEql2p8Li0shs7ettNev7NerbLa2ts/9vYoNDbOEhL+/P1JSUtC4cWNcunQJU6dOLfd5Go0GKpUKO3bsQEZGBtzc3MxRHhHVMDJ7WwTM2Fxpr3d6SZTebT7//HMMHTq00mowFbOckwgNDcWJEycwYMAABAcHIzExEWlpac89TyKRYMSIERg8eDDee+89wTAhIqrqPv74Y0uXIIpZRhIymQyrV68WfHzx4sXaf/fu3Vu7tgkRUXVw7do1xMTEwM7ODra2tggMDMTDhw8xb948NG/eHLt27UJZWRneeOMN3Lt3D4mJiZBKpWjQoAHmz5+PPXv24PDhw1Cr1bhx4wZGjx6NsLAwnD9/HrGxsXB0dISbmxtkMtkzn6eVgRfTERGZ2IkTJ+Dv74+NGzdi3Lhx6N69O1xcXLRLfTg7O2Pbtm1o2rQpEhISkJiYiG3btsHJyQnbt28HAOTm5mLdunX4+OOP8cknnwAA5s6di8WLF2Pz5s2oX7++SWpnSBARmdiAAQNQq1YtjBo1Clu3boWt7bMnuZ+s8Hrz5k34+vpqJ/e0bdsWf/zxBwBob0jk6empnal19+5dNG7cGAAQEBBgktoZEkREJnbw4EEEBAQgMTERISEh2LBhA55eNu/JCq9169ZFenq69jKBkydPagOkvNVl69Spoz2/e+7cOZPUzlVgiajGKSwuNWhGkq7X0zUF9pVXXsGMGTOQkJAAGxsbxMTEICMjA9OnT3/mYuLatWtj8uTJiIqKgo2NDerXr4/p06dj79695b7u3LlzMXv2bCgUCtjb28PDw6PS9ukJrgJLVAGuAlv16bsKbFWxdetW9OzZE7Vr18by5cthb2+PSZMmPfc8Y1aB5UiCiKiKcnNzw4gRI6BQKODk5FTpM5sAhgQRUZUVEhKCkJAQk/bBE9dERCSIIUFERIIYEkREJIghQUQ1jqak0Kpfz5rwxDUR1TgSO1m5U5sNZa4p0R07dsTx48fN0tcTHEkQEZEghgQRkYn169cPDx48QHFxMVq3bo3ffvtNuz0xMRGDBg1CREQENm/++x4XV65cwYgRIxAdHY2wsDD8+uuvz7zehx9+iNjYWJjjWmgebiIiMrHu3bvj6NGjqFOnDurWrYvjx49DKpWifv362L9/P5KSkiCRSBAdHY2goCCkpaVh1qxZePnll7Fnzx7s3r0brVu3BgDEx8dDIpFg7ty5ZqmdIUE1nrXfepKqvuDgYKxduxaenp6YOnUqtmzZAo1Gg//973+Ij49HdHQ0gL9vs3rjxg28+OKLWLNmDeRyOfLy8rSrwt6/fx+///67yZYFLw9Dgmo8XbeyrMxF4KjmatKkCTIyMnDv3j1MmzYN69atw8GDBxEbGwtfX19s2LABEokEmzZtQpMmTTBx4kQsXboUjRo1wsqVK3Hr1i0AgLu7Oz799FOoVCocOXIEnTt3NnntDAkiqnE0JYWVOiNJU1IIiZ1M53Patm2LjIwM2NjYoG3btkhLS0PTpk3Rvn17REZGoqioCM2bN4eHhwf69OmDCRMmwM3NDXXq1EF2drb2dSQSCRYtWoSRI0dix44dqFWrVqXtR3m4CiwRoHMkwVVgq77qugqsWMasAsvZTUREJIghQUREghgSRFQjVPEj6wYzdr8ZEkRU7cnlcjx48KDGBYVGo8GDBw8gl8sNfg3ObiKiaq9u3braKag1jVwuR926dQ1ub5aQKCoqwpQpU5CdnY2wsDAMHDjwmccnT56MWbNmoW7dujhz5gwWLVoEuVyORYsWoV69euYo0WrpmlonZtodEQH29vbw8fGxdBlVkllCIjk5GZ06dUJERARGjRqFvn37QiaToaioCFOnTsWFC/9MJ1y7di3Wr1+P+/fvY/369Zg/f745SrRaular5DRMIjI1s5yTSE1NRUBAAGxsbNCkSROkp6cD+HuEER0djQ4dOmifq1ar4erqCl9fX1y/ft0c5RERkQCzhEReXh4UCgUAwMHBAfn5+QAApVKJtm3bPvPcsrIy7b9r2kkmUxC6GUp1vkkKEVUesxxuUigUKCgoAAAUFBRoF6sqj0Qi0f7bxqbiDFOr1cjMzDS+SCtV0RWRFe27l5eX4BXD1fl904ehV+zz/aOawCwh4e/vj5SUFDRu3BiXLl3C1KlTBZ8rl8uRlZWFrKwseHt7V/jacrm8Ri/LYcy+1+T3rTLw/aOqTOyXHLMcbgoNDcWJEycwYMAABAcHIzExEWlpaeU+d8KECRg3bhzmzJmDsWPHmqM8IiISYJaRhEwmw+rVqwUfX7x4sfbfLVu2xI4dO8xRFhERVYBXXBORVdI1uYITL8yHV1wTkVXiNULWgSMJIiISxJCoJgqLSy1dAhFVQzzcVE0I3aeZ92gmImOICombN2/i0KFDKCz852TR6NGjTVYUERFZB1GHmyZMmICHDx9CKpVq/yMioupP1EjC09MTkydPNnUtRERkZUSFRNeuXbF06VL4+vpqt73++usmK4qIiKyDqJBITk5Gw4YNtUt8P70IH1WOwuJSyOxtLV0GEdEzRIWEVCpFbGysqWup0Tg7iYiskaiQ8PLywrp16+Dn56cdRQQFBZm0MCIisjxRIVFSUoLr168/c6c4hgQRUfUnKiScnZ0RExNj6lqIiMjKiLpOIj09HY8ePTJ1LUREZGVEjSTS09PRrl071KpVS3tL0WPHjpm0MCIisjxRIXHo0CFT10FERFZI1OGm33//Hf3790dQUBBef/11/Pbbb6aui4iIrICokcTChQsRFxeHpk2b4tKlS4iNjcUXX3xh6tqIyEiakkJI7GSitxP9m6iQ0Gg0aNq0KQCgWbNmsLPjCuNkHvyQM47Q3d14ZzcSS9SnvZ2dHQ4dOoQ2bdogJSWFq8CS2fBDjsiyRJ2TiIuLw1dffYXIyEh88803WLBgganrIiIiKyBqJOHt7Y2VK1eauhYiIrIyokJi7dq12LBhA+RyuXYbr5MgIqr+RIXEvn37cPToUTg4OJi6HiIisiKizkl4e3s/M4rQV1FREcaPH4+IiAjs2LFD5/bNmzejX79+UKlU2LBhg8F9EhGR8USNJIqLi9G7d280adJEu1T4smXLRHeSnJyMTp06ISIiAqNGjULfvn0hk8nK3Z6WloY1a9bA09PTsD0iIqJKIyokRo8eXe72W7duwdvbu8L2qampGDBgAGxsbNCkSROkp6fDz8+v3O3Xrl3DggULoFarERsbi3r16um3R0REVGlEhcT//d//lbs9JiYGmzc/fze1f8vLy4NCoQAAODg4ID8/X3B7t27dEB4ejoyMDCxZsqTCWVVqtRqZmZlidsOqeXl5GdTuyb4b0r4qvG+69quy6jf2vbdm5nj/TKWi34u1119dGHXptEajEfU8hUKBgoICAEBBQQGUSqXg9vDwcCiVSjRt2hTZ2dkVvrZcLjf4f/LqwJh9r+rvm6Xrt3T/xmL9NZvYkBV14lrIk/MTFfH390dKSgo0Gg0uXboEHx+fcrc3aNAAKpUKxcXFuHbtGtzc3Iwpj4gsSFNSKOqxwuJSc5RDBjLLIkyhoaF46623sHv3boSFhSExMRHdunV7brtcLseIESMwePBgyOVyLFy40BzlEZEJCC2pAjy7rIrM3hYBM54/bH16SZTJaiPxzHK4SSaTYfXq1eU+9u/tvXv3Ru/evY0pi6hGKSwuhcze1tJlUDWlV0hkZWXB1dVVe3e6wMBAkxRFROIJfRMH+G2cjCfqnMTPP/+M7t27Y/jw4ejRoweOHz8OAJg4caJJi6OqT+i4tK7j1URkPUSNJFasWIGkpCR4eHjgzp07mDRpEjp27Gjq2qga4FLfRFWbqJGEra0tPDw8AAAeHh6QyXizFyKimkDUSEKpVGLLli1o27YtUlJS4OLiYuq6iIiMouvuhWLubGhs++pCVEgsWbIEa9aswfLly9GoUSMsWrTI1HURERlF7BRcU7WvLnSGxO3bt1GnTh3cv38fAwcO1G7PysriaIKIqAbQGRIbN25ETEwM3nvvPUgkEu11ERKJRNSaTURica4/kXXSGRIxMTEAgOHDh6Nbt27a7cnJyaatimocXnVLZJ10hsShQ4fw66+/Yu/evTh79iwAoKysDAcPHkRoaKhZCiQiMiWOYnXTGRJNmzZFTk4OZDKZdlE+iUSCXr16maU4IiJT4yhWN50h4enpiX79+qFv377apTgA4O7duyYvjKoWfhurufi7r95ETYFdtWoVkpKSUFxcDLVajQYNGmDv3r2mro2qEH4bq7n4u6/eRF1xfeTIERw5cgS9e/dGcnKy9uprIiKq3kSFhKurK6RSKfLy8vDSSy9p7yZHRETVm6iQqFOnDnbu3AkHBwcsW7YMubm5pq6LiIisgKhzEjNmzEBubi5CQkLw1VdfYfny5aaui4iIrICokcS4cePg7e0NpVIJlUoFX19fU9dFRCQK75FtWqJGEi4uLkhMTISPj492KmxQUJBJCyMiEoOzq0xLVEjUqlULly9fxuXLl7XbrC0khOZqcw43EZHhRIXE+++/b+o6jMZvE0REla/CZTlcXFxgb2//3GPHjh0zWVFERGQddIbErFmzcOTIEdSrVw99+vRBmzZtzFUXERFZAZ0hMXz4cAwfPhxXr17Ft99+i4SEBLRq1Qp9+vRBw4YNzVUjERFZiKgpsA0bNsSbb76J+Ph4XL9+HX379jV1XUREZAUqDImcnBxs27YNUVFRmDlzJgIDA3H06FG9OikqKsL48eMRERGBHTt26Nx+5swZhIeHQ6VS4ebNm3ruDhERVSadITFmzBhERUUhOzsbCxcuxIYNGxAWFgaFQqFXJ8nJyejUqROSkpKwf/9+FBYWCm5fu3Yt1q9fj7lz52L9+vWG7xkRERlN5zmJtLQ0AMDOnTuxa9cuAIBGo4FEIsHBgwdFd5KamooBAwbAxsYGTZo0QXp6Ovz8/Mrdrlar4erqCldXV1y/ft3wPSMiIqNJNBqNRt9GZ86cQatWrUQ/f/bs2Rg/fjzq1auHFStWoGPHjmjTpk2521esWIEtW7YAAFQqlfbfQq5evQq5XI5abu5wkEmfe7y4pBT2duVfTPf48WM8fvwYAATblxWrYWMvL7d9SWE+7j7IqZLtn25b1dsLtWV7y7a31r99tv+nvZeXV7mPP03UxXTA3+cP9uzZg61bt6KoqAjfffed2KZQKBTa5cULCgqgVCoFt0skEm27p++GJ0Qul+vcUaGAAAAnJyc4OTlpfxa6GO/G/P+U277+exfg5fXPoTdj25dH6JcMAHYyhcHty2urT/1iate3f2Pal1c7YL76je2/qtevT9/l9V/V21v6s8OQ9kBOuY/9W4WfwhkZGYiPj0dwcDBWrlyJd955R6+AAAB/f3+kpKRAo9Hg0qVL2vtll7ddLpcjKysLaWlp8Pb21qsfIiKqXDpDYvz48Zg1axYaNmyI7777Do0bN0br1q317iQ0NBQnTpzAgAEDEBwcjMTERKSlpT23XSaTYcKECRg3bhzmzJmDsWPHGrxjRERkPJ2HmzQaDezs7KBWq1FWVvbMoSB9yGQyrF69utzH/r29ZcuWz0yTJSIiy9E5kli7di3i4+Px8OFDhIeH4/Llyzhy5AjKysrMVR8REVlQheck6tSpg0mTJmH//v2Ii4vDzp070a1bN3PURkREFiZqdtPNmzeRnZ2Nl19+GStXrkRWVpap6yIiIhEKi0tNeksEnSGRkZGBN998E/b29nBzc0NmZiYcHBx4j2siIith6puq6QyJxYsX4+23335mifDjx49j/vz5WLVqlUkLIyLL05QU/v859c9vl9jJLFARmZvOcxJZWVnP3UOiY8eOyM3NNWlRRGQdhIKAAVFz6AwJO7vyBxqc3UREVDPoPNyUk5Pz3G1KNRoNHj58aNKiiMh8dJ34LCwuNfkxb7JuOkPC398fe/fufW67n5+fyQoiIvPSFQIMCNIZEu+//z4AIDs7G7m5uXBycoKrq6tZCiMiIsvTGRIXLlxAbGwsysrK4OjoiNzcXGg0GsydO1evpcKJiKhq0hkSixYtQkJCAjw9PbXbMjMzMWXKFHz55ZcmL46IiAwjNH35yWNi6ZzdVFJS8kxAAICnp6fBC/0REZF56JqmrM8UZp0jiS5duiA6OhodO3aEk5MT8vLycOzYMXTu3Fl8pUREVGXpDIlJkybht99+w+nTp5GdnQ2lUonp06fD39/fXPUREZEFVbjAn5+fH6e8EhHVUDpD4t8X0j0tKCio0oshIiLrojMkduzYgYsXL6Jdu3bPPcaQICKq/nSGxPLly6FSqTB69Gg0bNjQXDUREZGV0DkF1tbWFvHx8SgqKjJXPUREZEUqPHFdr149c9RBRERWSNTtS4ODg1FaWvpPIzs7eHp6YsaMGZwOS0Q1mqlvH2ppokIiMDAQISEhaNOmDc6cOYMvv/wS/fv3x8KFC7Ft2zZT10hEZLWq+0q5Os9JPHHt2jV06NABUqkU7dq1w71799C+fXvY2IhqTkREVZSokYRUKsW2bdvQqlUrnDlzBlKpFBcvXnzmEFRVV92HjEREhhAVEkuXLsXatWvx448/onHjxvjggw9w/vx5xMXFieokOzsbU6ZMQVFREcaNG4dXX31V52Pvv/8+zp07B3t7e/Tv3x+vv/66QTunj+o+ZCQi62TtX1BFhUStWrXQvn17uLu7w8fHB7Vq1UKXLl1Ed5KUlISRI0ciMDDwuZAo77EbN27g888/F7zHNhFRdWHtX1BFfQovW7YMf/75J1q3bo2vv/4ap06dwttvvy26k9TUVAwbNgwymQxKpRIPHz6Ei4uL4GOZmZkYO3Ys5HI5FixYgNq1axu2d0REVVRF94PQZ7lvY4gKiZSUFHzxxRcAgGHDhmHgwIF6dZKXlwdHR0cAgIODA/Lz87UhUd5jvXr1wvDhw3Hs2DGsXbsWs2fPFnxttVqNzMxMverRh5eXl87Hn/Rd0fMqam9phtRflWsHKq9+Y/u3dP1kGmI/O6ydqJAoKSlBWVkZbGxsUFZWVuFNhzZv3ox9+/Zpf75w4QLy8/Ph6OiIgoICbSgAgEKheOYxhUKBIUOGwN7eHh06dEBSUpLOvuRyucH/k1UGY/u2ZO3Gqsq1A5avvyb/7ZDlf39iQ0pUSPTq1QuRkZFo0aIFzp8/j9DQUJ3Pj4qKQlTUPydiVq1ahVOnTiEwMBA5OTlwdnbWPubv7//MYxKJBKNHj0ZSUhJOnz4NX19fUTtCRESVT2dILFu2TDtq8PDwwKFDh9CsWTNkZWXp1cngwYMxbdo0rFixAhMmTAAAxMXFYeLEic895uzsjO7du2PQoEFwcnLCBx98YOCuVQ5rOS5IRGQJOkPi6ZVffXx80LVrV4M6qV27NjZu3PjMtjlz5mj//e/HRo4ciZEjRxrUV2WrrPvEEhFVRTpDol+/fuaqg4iIrBAvRCAA1n9Bjy5VuXYia8fFlwiA9V/Qo0tVrp3I2nEkQURkAtVl0gtHEkREJlBdJr0wJIiISBBDgoiIBDEkiIhIEEOCiIgEMSSIiEgQQ4KIiAQxJIiISBBDgoiIBPGKayIL49pTZM04kiCyMK49RdaMIUFERIIYEkREJIghQUREghgSREQkiCFBRESCOAWWdBK6cUpVumkKERmOIVFJqutcd6EgYEAQ1QwMiUrCue41V3X9gkAE8JwEkdH4BYGqM4YEEREJMktIZGdnIyoqChEREfjpp5+ee7y4uBgDBw7U/nzgwAEMGDAAo0ePRk5OjjlKJCKicpglJJKSkjBy5EgkJiYiMTHxmcdycnIwatQo3Lx5U7stMTERSUlJGDp0KLZv326OEomIqBxmCYnU1FQEBARAJpNBqVTi4cOH2scKCgoQExMDX19fAMCjR4/g6uoKqVSKgIAAnD171hwlEhFROcwyuykvLw+Ojo4AAAcHB+Tn58PFxQUA4OnpCU9Pz2eeq1AoAAAKhQL5+fk6X1utViMzM9NElRvPy8tL5+PWXHt1oOv9N8d7b2z/lq6fyCQhsXnzZuzbt0/784ULF5Cfnw9HR0cUFBRoA6M8CoUCBQUFAID8/HwolUqdfcnl8go/iK1ZVa69qrP0e29s/5aun6o2sV8yTBISUVFRiIr6Z974qlWrcOrUKQQGBiInJwfOzs6CbV1cXJCVlYXCwkKcPn0afn5+piiRiIhEMMs5icGDB2PTpk2IjIzEsGHDAABxcXGCM5eio6MxdOhQbNiwAYMHDzZHiUREVA6znJOoXbs2Nm7c+My2OXPmPPPzli1btP/u0aMHevToYY7SiIhIBy7LQWTFuMAiWRqvuCayYlxgkSyNIUFERIJ4uImqNR6uITIORxJUrfFwDZFxGBJERCSIIUFERIIYEkREJIghQUREgji7ycSEZtc8eYwnUInImnEkYWK6QoABQUTWjiFBRESCGBJERCSIIUFERIIYEkREJIghQUREghgSREQkiNdJEJkQV6Glqo4jCSIT4iq0VNUxJIiISBBDgoiIBDEkiIhIEEOCiIgEMSSIiEgQQ4KIiAQxJIiISBBDgoiIBFX5K66LioqQmZlp6TKIiKqUoqIiUc+TaDQajYlrISKiKoqHm4iISBBDgoiIBDEkiIhIEEOCiIgEMSSIiEgQQ6IC//3vf3Ho0CGD2u7atQtDhgzBoEGD8OGHH+rVdvfu3QgJCYFKpUJ4eDh2795tcHuVSoVJkyYZ1X7Dhg16tT958iSGDh2KIUOGYOTIkUhPT9e7/6f3OSEhAb/88oter/HLL78gISFBrzbl9X/16lWEhYXh/v37BrU3pO927dqhrKwMAFBaWop27dqJ3v/du3dj5MiRBtfy9O8+IiICa9as0W8HjPR0/5GRkRg5cqTo6ZoAUFJSgnnz5mHEiBEYPnw4zpw5o1f/M2fOhEqlQps2baBSqTBz5ky92oeHh+Px48fan9etW4cDBw5U2G7IkCHa/Zw2bRo++OADAMCNGzcwbdo0UX1Pnz4dBw8e1P48ceJEnD9/Xp/yn1Plr5PQ5fbt29i+fTsGDRqEOnXq6N3+3LlzCAwMxN69e9G1a1e92qampuL48ePYsmULbGxsEBMTg59++gmvvvqq6NcYM2YMwsLCUFpaikmTJiEwMBBeXl56tzeUoe2zs7Px0UcfYc2aNXB1dcWNGzewZMkSgz+wLSkrKwszZ87EkiVL4O7ubrZ+HR0dcf78ebRs2RJnz56FTKZhvm69AAAHPElEQVTfTYouXryIgwcPonv37gb1//TvXqVSYdiwYXB0dBTdPjk5GYmJibCzs0OjRo3wzjvvQCqVGtR/QkICTp48iaCgIFFtf/nlF7z44ouYN28ecnNzMWHCBGzevFl0308+nFUqFbZs2SK63RM9evTAkSNH0KtXLwDAiRMnEB0dXWG7V155BVeuXMErr7yC4uJipKWlAfj7d9m6dWtRfU+fPh3jx49Hly5dcOrUKbi4uKB58+Z678PTqvVI4sKFC/jyyy/x9ddfG9R+//79CAsLw927d1FQUKBX2x9++AHh4eGwsfn7LX733XdF/5H/m62tLXr16oWTJ08a1N7cDh8+jF69esHV1RUAUL9+faxcudLCVemvqKgIb7zxBt5++200atTIrH137twZR44cAQAcOXIEXbp00av9iBEjsGrVKr2+gZenqKgIZWVlsLMT/33yjz/+wDfffIMtW7Zg69ataNWqFVatWmVwDXfu3IG9vb3o5zs5OeHw4cM4deoUFAqF3qNgY4WEhODHH38E8PcXVXd3d1Eh36pVK6SmpuLmzZvw9vaGXC5Hfn4+Lly4gICAAFF916lTB926dcOuXbuwbt060SMQXap1SNStWxdTpkzR69v3086ePYuWLVuiY8eOeh9yysrKwosvvqj9WaFQ6PU/2r+5urri7t27erX55JNPtIeLDPmQfrr9zZs3Rbe7ffs26tevDwDYvn07VCoVevbsaVT/X331ld7tjbVq1Sqo1WpkZ2ebve///Oc/SE1NBQDcu3fvmb8lMV544QX07NkTmzZtMqj/J+/96NGjMXjwYL1GMvv27UN0dLR25NCvXz8cO3ZM7/4HDRqEzp07w9vbG+3atRPdtnnz5oiOjsaHH36Izp07G/wl0VAvvfQS7t69i5KSEhw4cAAhISGi2j0JidOnTyMgIAAtWrTA2bNncf36dTRp0kR0/6NGjcLGjRvx6quvws3NzdDd0KrWh5uaNWuGZs2aGdT24sWL+OuvvzBq1Cio1Wq4uLggNDRUdPsXXngBf/31l/YbaEZGBvLy8vDyyy8bVM+9e/f0/oVb6nDTCy+8oA20QYMGYdCgQVCpVEb1b4lDVf3798ewYcMwbNgwtG/fHkql0mx9SyQSuLq64vLly/Dw8DDoNaKjozFo0CD07NlT70Nlxvzt3Lt377maHRwcDOp/4cKFUCgUerVNS0tDUFAQevbsiTt37mDUqFHo06cP5HK5Xq9jjPbt2yMlJQXHjx/H8uXLRbXx8PBATk4Ofv31V7z55ptwd3fHyZMnIZVKtUckxHBwcEDLli0NPtT4b9V6JGGM77//HrGxsfj000+xdetW3L59G48ePRLdvkePHti5cyc0Gg00Gg0SEhIMXmOqtLQU3333ncGHq8yta9eu2L17Nx4+fAjg78MFt27dsnBV+nvppZdQu3ZtRERE4KOPPjJ7/0FBQYiPj9f7UNMTUqkUEydOxKefflrJlen25AvS0/T5kHva9OnTsX37dr1G0ceOHcO2bdsAAO7u7nB1ddVOAjCXkJAQ7N27F05OTnqFk6OjI3JyclC7dm34+/vjp59+MviLZWVhSAg4duwY2rdvr/25W7du+P7770W39/PzQ9u2bREZGYmBAwfC09NT75PfT4b8Q4YMwWuvvab3N8qnD9eoVCqjj0+LVbt2bUybNg1TpkxBVFQU3nrrLcyaNcssfZtCZGQkzpw5o/cskVWrViEsLAxhYWEGnU8KCgrClStXjDrx2KNHD/j7+xvc3hA9e/bEZ599hsLCQqxevRqxsbF6HS56mlwux5gxY7B69WrRbQYPHow///wT/fv3R3R0NCIjI/UejRirYcOGSE1NRY8ePfRq5+fnpz2XZ29vDzs7O9EnrU2FC/wRUaVLTk7Gpk2boNFo4OjoiPr16+Pdd9/V6wQ0WQeGBBGZ3KVLlww+P0iWxZAgIiJBPCdBRESCGBJERCSIIUFERIIYEkREJIghQWRBWVlZCA4ORmFhoaVLISoXQ4LIQo4ePYoRI0botQQ5kblV67WbiCpTbm4u5syZg8ePHyM7Oxvh4eHw9fXVrnCqVqsRHx8PHx8fUa9nY2ODjRs3on///qYsm8govE6CSKTU1FTcunULwcHBuHPnjvY+Cz169ICHhwfWrl0LjUaD8ePH6/W63bp1w759+/S+ZwSROXAkQSSSu7s7EhMT8cMPP0CpVKKkpAQeHh6Ii4uDQqHAnTt3dK6zM2fOHNy4cQO1atWqkvfXoJqJIUEk0meffYaWLVti8ODB+Pnnn3H48GG88847OHDgAJRKJWbNmgVdA/O4uDgzVktUORgSRCJ17doV8+bNw549e+Dq6gpbW1v06dMHAwcOhLOzM9zd3fW+MRSRteM5CSIiEsSRBFElmzRpkvaGS08olUp8/PHHFqqIyHAcSRARkSBeTEdERIIYEkREJIghQUREghgSREQkiCFBRESC/h9hTyCYUcYPjwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.barplot(data=damDfGroup['log10DAMratio_norm'].reset_index(), x='aa_-1', y='log10DAMratio_norm', hue='promoter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>promoter</th>\n",
       "      <th>amino acid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">strong</th>\n",
       "      <th>*</th>\n",
       "      <td>11.889748</td>\n",
       "      <td>1.075173</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>10.169228</td>\n",
       "      <td>1.007288</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>9.715327</td>\n",
       "      <td>0.987457</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>9.860408</td>\n",
       "      <td>0.993895</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>9.173361</td>\n",
       "      <td>0.962528</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      DAMratio  log10DAMratio  position\n",
       "promoter amino acid                                    \n",
       "strong   *           11.889748       1.075173        -2\n",
       "         *           10.169228       1.007288        -1\n",
       "         *            9.715327       0.987457         1\n",
       "         *            9.860408       0.993895         2\n",
       "         A            9.173361       0.962528        -2"
      ]
     },
     "execution_count": 447,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfList = []\n",
    "\n",
    "for position in [-2, -1, 1, 2]:\n",
    "    group = 'aa_{:d}'.format(position)\n",
    "    damDfGroup = damDf3.groupby(['promoter', group])[['DAMratio']].mean()\n",
    "    damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "    damDfGroup['position'] = position\n",
    "    dfList.append(damDfGroup)\n",
    "    \n",
    "damDfAA = pd.concat(dfList)\n",
    "damDfAA.sort_index(inplace=True)\n",
    "damDfAA.index.names = ['promoter', 'amino acid']\n",
    "damDfAA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nread_cutoff_30'"
      ]
     },
     "execution_count": 448,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subFolder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAA.to_csv(ctermELMseqPath / subFolder / 'damDfAA.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amino acid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>1.018660</td>\n",
       "      <td>0.008029</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>1.048772</td>\n",
       "      <td>0.020681</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>1.070085</td>\n",
       "      <td>0.029418</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>1.079216</td>\n",
       "      <td>0.033108</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>1.129595</td>\n",
       "      <td>0.052923</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>1.202162</td>\n",
       "      <td>0.079963</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1.219617</td>\n",
       "      <td>0.086223</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>1.240023</td>\n",
       "      <td>0.093430</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>1.312273</td>\n",
       "      <td>0.118024</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>1.340805</td>\n",
       "      <td>0.127366</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>1.385000</td>\n",
       "      <td>0.141450</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1.504686</td>\n",
       "      <td>0.177446</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>1.507042</td>\n",
       "      <td>0.178125</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>1.524369</td>\n",
       "      <td>0.183090</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>1.545939</td>\n",
       "      <td>0.189192</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>1.602763</td>\n",
       "      <td>0.204869</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>1.610542</td>\n",
       "      <td>0.206972</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>1.716010</td>\n",
       "      <td>0.234520</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1.746299</td>\n",
       "      <td>0.242119</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>1.762645</td>\n",
       "      <td>0.246165</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>2.034273</td>\n",
       "      <td>0.308409</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            DAMratio  log10DAMratio  position\n",
       "amino acid                                   \n",
       "L           1.018660       0.008029        -1\n",
       "C           1.048772       0.020681        -1\n",
       "I           1.070085       0.029418        -1\n",
       "M           1.079216       0.033108        -1\n",
       "F           1.129595       0.052923        -1\n",
       "V           1.202162       0.079963        -1\n",
       "A           1.219617       0.086223        -1\n",
       "W           1.240023       0.093430        -1\n",
       "Y           1.312273       0.118024        -1\n",
       "P           1.340805       0.127366        -1\n",
       "*           1.385000       0.141450        -1\n",
       "S           1.504686       0.177446        -1\n",
       "T           1.507042       0.178125        -1\n",
       "G           1.524369       0.183090        -1\n",
       "Q           1.545939       0.189192        -1\n",
       "H           1.602763       0.204869        -1\n",
       "N           1.610542       0.206972        -1\n",
       "R           1.716010       0.234520        -1\n",
       "E           1.746299       0.242119        -1\n",
       "D           1.762645       0.246165        -1\n",
       "K           2.034273       0.308409        -1"
      ]
     },
     "execution_count": 449,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDfAA[damDfAA['position'] == -1].loc['weak'].sort_values('log10DAMratio')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amino acid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>7.985168</td>\n",
       "      <td>0.902284</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>8.023959</td>\n",
       "      <td>0.904389</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>8.255760</td>\n",
       "      <td>0.916757</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>8.600441</td>\n",
       "      <td>0.934521</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>8.606447</td>\n",
       "      <td>0.934824</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>8.869556</td>\n",
       "      <td>0.947902</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>8.915522</td>\n",
       "      <td>0.950147</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>9.006816</td>\n",
       "      <td>0.954571</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>9.454486</td>\n",
       "      <td>0.975638</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>9.598168</td>\n",
       "      <td>0.982188</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>10.169228</td>\n",
       "      <td>1.007288</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>10.236489</td>\n",
       "      <td>1.010151</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>10.441145</td>\n",
       "      <td>1.018748</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>10.450425</td>\n",
       "      <td>1.019134</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>10.526743</td>\n",
       "      <td>1.022294</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>10.609195</td>\n",
       "      <td>1.025682</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>10.904656</td>\n",
       "      <td>1.037612</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>10.973148</td>\n",
       "      <td>1.040331</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>11.815256</td>\n",
       "      <td>1.072443</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>12.086276</td>\n",
       "      <td>1.082292</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>12.961766</td>\n",
       "      <td>1.112664</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             DAMratio  log10DAMratio  position\n",
       "amino acid                                    \n",
       "L            7.985168       0.902284        -1\n",
       "C            8.023959       0.904389        -1\n",
       "M            8.255760       0.916757        -1\n",
       "F            8.600441       0.934521        -1\n",
       "I            8.606447       0.934824        -1\n",
       "W            8.869556       0.947902        -1\n",
       "A            8.915522       0.950147        -1\n",
       "V            9.006816       0.954571        -1\n",
       "P            9.454486       0.975638        -1\n",
       "Y            9.598168       0.982188        -1\n",
       "*           10.169228       1.007288        -1\n",
       "S           10.236489       1.010151        -1\n",
       "Q           10.441145       1.018748        -1\n",
       "G           10.450425       1.019134        -1\n",
       "T           10.526743       1.022294        -1\n",
       "H           10.609195       1.025682        -1\n",
       "N           10.904656       1.037612        -1\n",
       "R           10.973148       1.040331        -1\n",
       "D           11.815256       1.072443        -1\n",
       "E           12.086276       1.082292        -1\n",
       "K           12.961766       1.112664        -1"
      ]
     },
     "execution_count": 452,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDfAA[damDfAA['position'] == -1].loc['strong'].sort_values('log10DAMratio')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (damDfAA[damDfAA['position'] == -1].loc['weak'] + damDfAA[damDfAA['position'] == -1].loc['strong'])/2\n",
    "df.sort_values('log10DAMratio_norm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "promoter\n",
       "strong    0.989979\n",
       "weak      0.137875\n",
       "Name: DAMratio, dtype: float64"
      ]
     },
     "execution_count": 453,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meanDAM = damDfAA['DAMratio'].groupby(level=0).mean()\n",
    "meanDAMAA = meanDAM\n",
    "log10(meanDAM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAA['log10DAMratio_norm'] = damDfAA['log10DAMratio'] - log10(meanDAM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "      <th>TAA</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>promoter</th>\n",
       "      <th>amino acid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">weak</th>\n",
       "      <th>*</th>\n",
       "      <td>0.089720</td>\n",
       "      <td>0.003575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.010792</td>\n",
       "      <td>-0.007642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>-0.040141</td>\n",
       "      <td>-0.051652</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.001208</td>\n",
       "      <td>-0.001216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-0.108736</td>\n",
       "      <td>-0.117194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.014977</td>\n",
       "      <td>-0.017937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.076543</td>\n",
       "      <td>0.108290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.007491</td>\n",
       "      <td>-0.002145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.010773</td>\n",
       "      <td>0.104244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000754</td>\n",
       "      <td>0.000830</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "position                   -2        -1  TAA         1         2\n",
       "promoter amino acid                                             \n",
       "weak     *           0.089720  0.003575  NaN -0.010792 -0.007642\n",
       "         A          -0.040141 -0.051652  NaN -0.001208 -0.001216\n",
       "         C          -0.108736 -0.117194  NaN -0.014977 -0.017937\n",
       "         D           0.076543  0.108290  NaN -0.007491 -0.002145\n",
       "         E           0.010773  0.104244  NaN -0.000754  0.000830"
      ]
     },
     "execution_count": 455,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfList = []\n",
    "for promoter in ['weak', 'strong']:\n",
    "    damPlotDf = damDfAA.loc[promoter][['position', 'log10DAMratio_norm']].reset_index().pivot(index='amino acid', columns='position', values='log10DAMratio_norm').copy()\n",
    "    damPlotDf.columns.name = 'position'\n",
    "    damPlotDf['TAA'] = np.nan\n",
    "    damPlotDf = damPlotDf[[-2, -1, 'TAA', 1, 2]]\n",
    "    damPlotDf['promoter'] = promoter\n",
    "    dfList.append(damPlotDf)\n",
    "\n",
    "damPlotDf = pd.concat(dfList).reset_index().set_index(['promoter', 'amino acid'])\n",
    "damPlotDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAF4AAAEBCAYAAADxfRL4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAADtlJREFUeJztnW1MXNW6x/97ZhiYAnJAkNOX9EUUFXuop6KNbyERbI5Wq9EaUBG1Nk2kaltIa9IUQqyi0tI0gH6YWptUODVqGquk3NY2JW2uVy5CbQGrVFIqQg9gBygwwLzt+6FXbLWzWdtZwyN7nt8nyDx99tNfFmvWXnutvRRVVVUwU46JuoBQhcUTweKJYPFEsHgiWDwRLJ4IFk8EiyfCQnHRs8c2C8cOd58Tjv1H9odCcSNOp3BOt88nHAsAf4uKEorjFk8EiyeCxRMhXbzX65Wd0pBI+XKtrKzEzTffDJfLhUOHDmHHjh0y0hoaKS3earWitrYWAFi6INK6mhtuuAGZmZlwuVyyUhoaKeJdLhe8Xi/sdjvsdruMlIZHSh//8ssvy0gTUigUz1xX3hUuHDs7xiwcu+W/xO5IB4aHhXOGmfR1CpEzZgjF8TieCBZPBIsngsUTweKJkCK+sbFx4uePP/5YRkrDI2UcX19fD5PJhIMHD+Kuu+6SkdLwSGnxeXl5qKmpQWZmJtLT02WkNDxSxJeVlcFms+HAgQPYvn27jJSGR0pXU1BQICNNSEHysPvdz/9bOHawq1769X0QnyXxBWlGhYeTRLB4Ilg8ESyeCBZPhJRRTVlZGfr6+pCQkABFUZCfny8jraGRNo632+1YvXq1jHQhgbSH3T6fDy6Xi1cZCCJFvN1uh8fj4VUGOuBVBkSQTBn0/XhQOPZ841Hh2L/ftubPlEMCDyeJYPFEsHgiWDwRLJ4IKaOaw4cPo6WlBW63GwsXLsSDDz4oI62hkdLiW1pasG7dOmzYsAFnzpyRkdLwSGnxFotlYqrApHN1bagi7QaKpwr0wVMGRJBMGcQvXCUc29fytfTrh5vENzsoOlYk6IE7ZCJYPBEsnggWT4QU8U1NTejo6AAAHD9+XEZKwyNFfH19Pb788ksAQGtrq4yUhkfKcNLr9UJRFLhcLng8HhkpDY+0Pv7Xh92MGHznSgSPaoggmTJwmyKFY1OWvyb9+qM+8bdIeVR9b+ETe5MBt3gyWDwRLJ4IFk9EwOJ/+uknDAwM4NNPP71iaz2jTcDiq6qq0NnZCZPJhGPHjsmoKSQIWPw111yDxsZG3H///bBYSEan05KATd14440YHh5GT08PkpKSZNQUEpC8DG5wRPxlbNbhb4VjbYn3CsX9MjwknFPvDdTfo2OE4nhUQwRJpxzmG6G47G/XVxQdseIrEvTALZ4IFk8EiyeCxRMRsPjTp0/LqCPkCFj8kSNHJnZ0865ucQIeTjY0NFzxOz9/FSNg8UuXLsUzzzwjo5aQgmTKwDnUIxyrOMW39ohOGeiZstBLTCSffPaXhmTKYKBtb1DyirZ4t44/cj3TC3rgFk8EiyeCxRPB4omQ9hY+RVHg8Xhw+vRp7N69W0ZaQyOlxRcUFOCJJ56A0+nE5s3ipxOHMlJa/CeffILz589j48aNvNJAECktvqenByaTCR988AFvThCENyYQQdIvxC74l3Csa/AH6de3QPxu1KIEZ+DHw0kiWDwRLJ4IFk8EiydC2qimu7sbbrcb1157LcxmM2w2m6zUhsSv+A0bNkBRFJhMJiiKArPZjDfeeOOqsXv27EF4eDhMJhPq6upQXl4etIKNgt+uZuvWrYiNjUVxcTGKi4sRGel/b6rT6URWVhaWL18Os9kMszk4Cz2NhGZXMzAwgLa2NgDAxYsX/cZ5vZc27IaHh+OWW26RWJ5x0Vxl0N/fj9raWiiKgoceeggxMVdfdH/u3Dns378fiqLg0Ucfxdy5czUvOur4XrhAPXeuMQseFYr7K5xS71f8e++9h46ODvz6saIoKC0t1VWEP5bt/UA41jEkLul/Vr8qFNd1cUA4p0fn6pd5MbFCcX67mry8PLS3tyMyMhK9vb2Ijo7WVQCjjebfUWVlJaKiojBnzhxUVlZOVU0hgaZ4m802sRDVarVOSUGhguao5qWXXsLu3bvh8XiwapX421GZydFs8SdPnsSFCxcwMDCA999/f6pqCgk0xdfV1SEtLQ3PPfccIiIirhpTW1uL8+fPo7GxEfv27QtKkUZEU7zZbMacOXNQV1fn9wbqzJkzmDlzJm6//XZ0dnYGpUgjotnH5+fnw2q1oqenB6mpqZMmI1jxPW3RFJ+YmAgAeOSRR/zGLFu2DOXl5fB6vZpxzJUEPC2clJSEV18Vu2NkfkNTfHV19cTLf5YsWYKsrCwpF/10aZpwrHesV8o1LyfWEiZ+/SD1nprif/zxR2zfvh0A8M477wSnghBFU/zY2BgOHDgAABgcHJySgkIFzWlht9uNr776Cqqq4p577kFYmPifqBajF04Jx+rpaqJmZwrFOZ3ibw/R29VEazwwuhzNFv/uu++iq6sLFosFBw8exFtvvaWvCsYvmuKtViu2bt0qlMjhcMBqtSIqSmy7YaijKb6+vh5DQ0MT0wVr1679Q8zOnTuxZMkS7N+/H5GRkcjPzw9OpQZDU3xJScnEz4qfbYdmsxmpqano7e2dOK6CmRy/4n999Hc5V3v053Q6AQAZGRkTQ09mcjQf/TkcDrS1tWnOwSxduhQ7duyAqqpYsWJFUIo0IppdTXFxMdLT0ycWNV2N5ORkJCcnB6U4I6MpPiUlBbfddpv0x35q+GzhWKd1vnCs6Hjqolf8AJlos5x7l9+jKb65uRnj4+MTK8N4y408NMUnJyfjxRdfhM1m89vVMH8OTfH9/f148803J37nO1d5aIp/9tln0dzcDFVVMTQk/j5eZnI0n7lu27YNP/xwae1ic3PzlBQUKmiKj4uLQ0ZGBqKjo/0uWAWAmpoaFBUVTawsZiZHU/yyZcsQFxeHr7/+GgkJCX7j2traUFhYiEOHDkkv0Kho9vF33303AEz6Ygir1QpVVWE2m+FyuXi5nwDSNp/Z7XZ4vV5+l4EgNK8/HL4gHDusit85Xhd9jVDcf4bEH2PqvXMV3ZhA8i6DU6PjwrG9ow7h2OWC4oM1DaAH3udKBIsngsUTIU18Q0MDtmzZIiud4Qn4y3Xfvn04d+4cHA4HiouLJZQUGgQsvqurC5GRkbBYLLyjWwcBdzWvvPIKnn76aSQkJFyxKoHRRso4PioqCtnZ2TJShQw8qiGCxRNBMmVgM4l/CY+45Z+0o+fEynGfvtMtxdYKc4sng8UTweKJYPFEBCy+rKwMJ0+elFFLSBGw+PXr18PhcKCyspLfZaCDgMWbTCbMnj0bFosFv/zyi4yaQoKAx/FbtmxBamoqVq5cyasLdBCw+MLCQhl1hBw8qiGCZMogPixcOLbqf+uFY59KulUobtTnFc4ZoQTnGQO3eCJYPBEsnggWT4SUL9ffn07P4/nJkSL+9yuEeXfg5PDJZ0RwH08EiyeCxRNBMmXQOtQvHNv+c5f065t0HLLo0rEiQV8NDAksnggWTwSLJ0K6+F27dslOaUiki3e73bJTGhIpw8mysjIoigJVVTE6OiojpeGRIr6goEBGmpCCv1yJILlzbXH0CMfePH+e9OtbdLxfLTj3rdziyWDxRLB4Ilg8ESyeCCmjms8+++yK3x977DEZaQ2NlBZ/9uxZnDp1CpGRkVi0aJGMlIZHivj169ejsLAQIyMjV7wSl/GPtK7m559/xnXXXYeysjIZKQ2PlBbf1NSE8fFxdHZ28rSwIFJa/Ouvv64r/t/HvxKOzfjnP/SWMymD3uBNXccLxvFwkggWTwSLJ4LFEyFFPB+Srh8p4mtqangYqRMp4r1eL9xuN1wu1x92hzBXR1of7/F4YLfb+f3xgvCOECJ4VEMEySqDky0twrHbHpd/YqaeF/eP6jgXUA/c4olg8USweCJYPBEBiz927BjGx8VPuWEuEfCoJjw8HHv27IGqqoiJiUFWVpaMugxPwOItFgt8///i44iIiIALChUCFr93716kpKTg3nvv5UPTdRBwH79t2zasWLECX3zxBZ588kkZNYUEAbf40tJSxMTEID09HWvXrpVRU0gQsPiNGzfq/jdLFi8Wjp1ni9KdfzIuesRXGYQF6ZB4HscTweKJYPFEsHgiWDwRAYuvqqqSUUfIEbD4tLQ0lJaW4vvvv8fZs2dl1BQSBDyOj4+Ph8vlwrfffouIiAgsWLBARl2GJ2Dxdrsd69atQ1SU/BsdIxOw+E2bNsmoI+RQVFVVqYsIRXg4SQSLJ4LFE8HiiWDxRLB4Ilg8EX8J8dXV1di5cyfKy8snjR0ZGUFBQQHa29v/8FlVVRVyc3NRUVGBhoYG5OTkwOu9dNjWN998g/z8/D+VN5B6/fGXEP/AAw9g5cqVGBwcnDS2r68PCxcuvOpnOTk5uOOOO7B69WqcOHECeXl5OHLkCADg+PHjyMzMRGtrq+68gdTrD5L18Zezd+9e+Hw+dHd344UXXpg0NiwsDDExMZpxHo8Hp06dgtVqRWtrK+688060t7cjOjoaJ06cwK23/vFouvnz50+a91eio6Oxffv2SevVgrzFP/XUU2hqaoLP58Phw4cnjV2xYvKNCkePHkVOTg6ef/55LFq0CNXV1XjttdewatUqzJgxAw6HI6CaN2/eLFSvFjxXQwR5iw9VWDwRLJ6IkBA/Pj6OgYEB6jKugHw4GUz6+vpw9OhRhIWFITExEd3d3bjvvvuQmJhIXdr0FF9RUYGRkREAQFxcHDo6OuByubBp0yZ8+OGHGBoaQkREBLKzs9HV1YULFy7AarUiNjYWY2NjqKysRH9/P1RVRVFRER5//HFkZGSgr68PxcXFU/J/mLZdTXZ2NtasWYPm5maUlJQgNzcXn3/+Obq6ujB37lw8/PDDE7GLFy9GZmYmAGBsbAy9vb0oLCzEzJkz8d1332HWrFlYs2YNhoeHp6z+aSteVVWMj49PbAMCAJ/Ph9zcXNx0000oKSmZ+Ey5bKm1qqpX/O7z+Sa2EFksU9cBTMuuBvjtDNm0tDQUFRXB6/WioKAAu3btgtPpRFJSEkymS+1q1qxZ+Oijj3D99dfDZrMhPj4eb7/9NhRFQUpKCkn90/LOtaKiAsuXL8e8efJPU5gqpqV4IzBt+/jpDosngsUTweKJYPFEsHgi/g+owoMZc/YvEQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 288x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAF4AAAEBCAYAAADxfRL4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAADvBJREFUeJztnX9QlOXax7/P7rKwLEgQxBE9alFU5GCTpGM/XuYENVOWNWQDFVGZ40xYanCsGUcY5rWoUJgGqD/WzBmDsaleJ4sXRtORdE4TQ2CKZGEcMQIPoPwQ2GB/nj887Yl0n72f9l4v2b0+f8FweT0XH2/uvZ/7x/MobrfbDeaKo6MuIFRh8USweCJYPBEsnggWTwSLJ4LFE8HiiTBQXHS894BwrNM2IRwbc/2jQnETVqtwTq2YIyOF4rjFE8HiiWDxREgX73Q6ZacMSqR8uNbU1OCWW26BzWbD/v378c4778hIG9RIafFGoxGNjY0AwNIFkdbV3HjjjcjKyoLNZpOVMqiRIt5ms8HpdMJiscBischIGfQoFEt/fANFJP7zjTcKx0aawoVjs/63QyjOOn5eOKfWTiEyKjYAWRlpsHgiWDwRLJ4IFk+EFPGtra2erz/++GMZKYMeKXM1zc3N0Ol02LdvH5YtWyYjZdAjpcUXFBSgvr4eWVlZyMjIkJEy6JEivqKiAiaTCQ0NDaisrJSRMuiR0tUUFRXJSBNSkEwZWEdOC8cq9kHhWFPCEqG4M6PDwjnDddo6hb9ExwjF8XCSCBZPBIsngsUTweKJkDKcrKiowODgIBISEqAoCgoLC2WkDWqkDSctFgvWrFkjFMvDSYmL3S6XCzabjXcZCCJFvMVigcPh4F0GGuA7Vx8Eqqsh2R8/1iO+vWPi3D+FY2/4m5j4BIP4r93vsAvHaoGHk0SweCJYPBEsnggWT4SUUc2BAwdw4sQJ2O12LFy4EA8++KCMtEGNlBZ/4sQJbNiwARs3bsSpU6dkpAx6pLR4g8HgmSrQabzhCFWk3UDxVIE2pIh/6aWXZKQJKUimDGbN/Ztw7LnOb6Rff9wlPj0VZzBKvz7Aw0kyWDwRLJ4IFk+EFPFtbW3o7u4GABw5ckRGyqBHivjm5mZ8+eWXAICODrEjj6GOlOGk0+mEoiiw2WxwOBwyUgY90vr43xa7GTH4zpUIHtUQQbO9Y2JUPNh+Tjg08ppkobixCfEHU9jdLuFYAIiLihaK4xZPBIsngsUTweKJ8Fv8zz//jJGREXz66afTjtYz6vgtvra2Fj09PdDpdDh8+LCMmkICv8XPmjULra2tuO+++2DQsBk01PHb1E033YTx8XH09/cjOVlsHM3wDZTvy/MNVHBB0ynbh4RDFYeGvw5BnBr+yCMUbS1eFG7xRLB4Ilg8ESyeCL/Fnzx5UkYdIYff4g8ePOg50c2nusXxezjZ0tIy7XtefxXDb/EPPPAAnn76aRm1hBRX/5F6h/jxd1P8HUJxI+PjwjmNira3/ESa+WFwVzUkUwa280eFY3VhEcKxJsE4I6aEc7oV0aza4BZPBIsngsUTweKJkPYUPkVR4HA4cPLkSezcuVNG2qBGSosvKirC448/DqvVis2bN8tIGfRIafGffPIJzp49i1dffZV3GggipcX39/dDp9Phgw8+4MMJgvDBBCJI+gVj/GLh2EAsdlshfkw+MPetPJwkg8UTweKJYPFEsHgipI1q+vr6YLfbce2110Kv18NkCtR4IDjwKn7jxo1QFAU6nQ6KokCv1+P111+/bOyuXbsQHh4OnU6HpqYmVFVVBazgYMFrV7N161bExsaitLQUpaWlMJvNXpNYrVbk5ORgxYoV0Ov10Ov1ASk2mFDtakZGRtDZ2QkAuHDhgtc4p/PignB4eDhuvfVWieUFL6q7DIaHh9HY2AhFUfDQQw8hJubyK+hnzpzB3r17oSgKHn30UcybN0/1otbRM+IFarhzNV2bJhR3bnxMPKdO21+vOTJSKM6r+Pfeew/d3d347ceKoqC8vFxTEd5469jXwrHDVvHTG28vu18o7uwF8f9MRREOBSDhjQkFBQXo6uqC2WzGwMAAoqPFjpgwYqiO42tqahAVFYW5c+eipqbmStUUEqiKN5lMno2oRmNgHnwZqqiOal588UXs3LkTDocDq1evvlI1hQSqLf7YsWM4f/48RkZG8P7771+pmkICVfFNTU1IT0/Hs88+i4iIy2+la2xsxNmzZ9Ha2oo9e/YEpMhgRFW8Xq/H3Llz0dTU5PUG6tSpU5g9ezYWL16Mnp6egBQZjKj28YWFhTAajejv70damu+bE4Id3zMWVfGJiYkAgEceecRrzPLly1FVVQWn06kax0zH72nh5ORkrFu3TkYtIYWq+Lq6Os/Df5YuXYqcnBwpF1234K/CsVpOhIgSqWH21KAEZq1IVfxPP/2EyspKAMDbb78dkAJCFVXxk5OTaGhoAACMjsrf3xLKqE4L2+12fP3113C73bj77rsRFhYm5aLWUfFhp6bDZ4LTwqMT4ofPtHY1otPCqi3+3XffRW9vLwwGA/bt24c333xTUxGMd1TFG41GbN26VSjR0NAQjEYjoqKipBQW7KiKb25uxtjYmGe6YP369ZfEbN++HUuXLsXevXthNptRWFgYmEqDDFXxZWVlnq8VL0sxer0eaWlpGBgY8LyugvGNV/G/Lf39nsst/VmtVgBAZmamZ+jJ+EZ1VDM0NITOzk7PHMyyZcsuiens7ERDQwPcbjeys7Mxf/58nxflUY2Prqa0tBQZGRmeTU2XIyUlBSkpKZqKY3yIT01Nxe233y5/2U8nfkx+KEx9q8jvmSN6eYhvHZhyaXuIhPdtX9NRFd/e3o6pqSnPzjA+ciMPVfEpKSl44YUXYDKZvHY1zJ9DVfzw8DDeeOMNz/d85yoPVfHPPPMM2tvb4Xa7MTYmvu2N8Y3qWGnbtm348ccfAVzs7xl5qIqPi4tDZmYmoqOjvW5YBYD6+nqUlJR4dhYzvlEVv3z5csTFxeGbb75BQkKC17jOzk4UFxdj//790gsMVlT7+LvuugsAfD4Ywmg0wu12Q6/Xw2az8XY/AaQtKFosFjidTn6WgSA0jz8cGxSOHXaLr3rNmXWNUNzV8MYEkmcZdNvFYy84rcKxcyAmftwp/s5Zh1bxgnF8zpUIFk8EiydCmviWlhZs2bJFVrqgx+8P1z179uDMmTMYGhpCaWmphJJCA7/F9/b2wmw2w2Aw8IluDfjd1bz88st46qmnkJCQMG1XAqOOlHF8VFQUcnNzZaQKGXhUQwSLJ4JkysCgYf02LAAHA7QcTBh3BmYqi1s8ESyeCBZPBIsnwm/xFRUVOHbsmIxaQgq/V6BcLhe++uordHR0ICkpCdnZ2T7/Teew+Hu4xzQsWiyO/4tQnJbdwloWTQDxVTC/W7xOp8OcOXNgMBhw7py40FDH73H8li1bkJaWhlWrVvHuAg2QLHZzV8OjGjJIpgziw8KFY4+Piv91iLb4MIgfNojVB2Z7Ord4Ilg8ESyeCBZPhJQP1z++nZ7H876RIv6PO4T5dKBv+M1nRHAfTwSLJ4LFE0EyZfAv26Rw7D9+Pi0cu3LBLUJxVrd4ewsL0Il2bvFEsHgiWDwRLJ4I6eJ37NghO2VQIl283a7hLGUII2U4WVFRAUVR4Ha78euvv8pIGfRIEV9UVCQjTUjBH65EkNy5to+IP8vgr7Gx0q9vd4kfkzcZ5DxB/I9wiyeCxRPB4olg8USweCKkjGo+++yzad8/9thjMtIGNVJa/OnTp3H8+HGYzWYsWrRIRsqgR4r4V155BcXFxZiYmJj2SFzGO9K6ml9++QXXXXcdKioqZKQMeqS0+La2NkxNTaGnp4enhQUhORGyrun/hWP/Z0GycKzwYveEhgdUazzSHxkp9uh+Hk4SweKJYPFEsHgipIjnl6RrR4r4+vp6HkZqRIp4p9MJu90Om812yekQ5vJI6+MdDgcsFgs/P14QPhFCBI9qiCDZZfB/hw4Lx+asvV369a1aJkk0Prhf7N2W3OLJYPFEsHgiWDwRfos/fPgwpqamZNQSUvg9qgkPD8euXbvgdrsRExODnJwcGXUFPX6LNxgMcP1nE2hEhPgroUMdv8Xv3r0bqampuOeee/il6Rrwu4/ftm0bVq5ciS+++AJPPPGEjJpCAr9bfHl5OWJiYpCRkYH169fLqCkkINllsKn5oHDs329bIhwr+oJDLS9Z1PoQPt5lcJXD4olg8USweCJYPBF+i6+trZVRR8jht/j09HSUl5fjhx9+wOnT4k9TCnX8voGKj4+HzWbDd999h4iICFx//fUy6gp6/BZvsViwYcMGREVFyagnZPBb/KZNm2TUEXKQTBkwPJwkg8UTweKJYPFEsHgiWDwRLJ6Iq0J8XV0dtm/fjqqqKp+xExMTKCoqQldX1yU/q62tRX5+Pqqrq9HS0oK8vDw4nRdftvXtt9+isLDwT+X1p15vXBXi77//fqxatQqjo6M+YwcHB7Fw4cLL/iwvLw933nkn1qxZg6NHj6KgoAAHD15c3z1y5AiysrLQ0dGhOa8/9XqDZH/879m9ezdcLhf6+vrw/PPP+4wNCwtDTEyMapzD4cDx48dhNBrR0dGBJUuWoKurC9HR0Th69Chuu+22S/7NggULfOb9jejoaFRWVvqsVw3yFv/kk0+ira0NLpcLBw4c8Bm7cuVKnzkPHTqEvLw8PPfcc1i0aBHq6urw2muvYfXq1YiMjMTQ0JBfNW/evFmoXjV4roYI8hYfqrB4Ilg8ESEhfmpqCiMjI9RlTIN8OBlIBgcHcejQIYSFhSExMRF9fX249957kZiYSF3azBRfXV2Nif9sPI2Li0N3dzdsNhs2bdqEDz/8EGNjY4iIiEBubi56e3tx/vx5GI1GxMbGYnJyEjU1NRgeHobb7UZJSQmys7ORmZmJwcFBlJaWXpHfYcZ2Nbm5uVi7di3a29tRVlaG/Px8fP755+jt7cW8efPw8MMPe2LvuOMOZGVlAQAmJycxMDCA4uJizJ49G99//z2SkpKwdu1ajI+Lv73eX2aseLfbjampKc8xIABwuVzIz8/HzTffjLKyMs/PlN+9vcztdk/73uVyeY4QGQxXrgOYkV0N8N93yKanp6OkpAROpxNFRUXYsWMHrFYrkpOTodNdbFdJSUn46KOPcMMNN8BkMiE+Ph5vvfUWFEVBamoqSf0z8s61uroaK1aswPz586lL+dPMSPHBwIzt42c6LJ4IFk8EiyeCxRPB4on4N8A54RMSEi16AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 288x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set_fontsize_paper()\n",
    "for promoter in ['weak', 'strong']:\n",
    "    figsize = (4, 4)\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "    cbarLabel = 'log10(normalized DAMRatio)'\n",
    "    sns.heatmap(data=damPlotDf.loc[promoter].loc[aaTable], cmap=cmapName,\n",
    "                vmin=-vmaxCodon, vmax=vmaxCodon, square=True, cbar=False, cbar_kws={\"label\":cbarLabel})\n",
    "    filename = 'ELM_seq_AA_promoter_{}.png'.format(promoter)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / filename, dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Codon level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfList = []\n",
    "\n",
    "for position in [-2, -1, 1, 2]:\n",
    "    group = 'codon_{:d}'.format(position)\n",
    "    damDfGroup = damDf3.groupby(['promoter', group])[['DAMratio']].mean()\n",
    "    damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "    damDfGroup['position'] = position\n",
    "    dfList.append(damDfGroup)\n",
    "    \n",
    "damDfCodon = pd.concat(dfList)\n",
    "damDfCodon.sort_index(inplace=True)\n",
    "damDfCodon.index.names = ['promoter', 'codon']\n",
    "damDfCodon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meanDAM = damDfCodon['DAMratio'].groupby(level=0).mean()\n",
    "meanDAMCodon = meanDAM\n",
    "meanDAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodon['log10DAMratio_norm'] = damDfCodon['log10DAMratio'] - log10(meanDAM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodon.to_csv(str(ctermELMseqPath / 'damDfCodon.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfList = []\n",
    "for promoter in ['weak', 'strong']:\n",
    "    damPlotDf = damDfCodon.loc[promoter][['position', 'log10DAMratio_norm']].reset_index().pivot(index='codon', columns='position', values='log10DAMratio_norm').copy()\n",
    "    damPlotDf.columns.name = 'position'\n",
    "    damPlotDf['TAA'] = np.nan\n",
    "    damPlotDf = damPlotDf[[-2, -1, 'TAA', 1, 2]]\n",
    "    damPlotDf['promoter'] = promoter\n",
    "    dfList.append(damPlotDf)\n",
    "\n",
    "damPlotDf = pd.concat(dfList).reset_index().set_index(['promoter', 'codon'])\n",
    "damPlotDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damPlotDf.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damPlotDf.loc[idx[:, ['AGA', 'AGG', 'CGT', 'CGC', 'CGA', 'CGG']], :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# fig = plt.figure(figsize=figsize)\n",
    "# gs = gridspec.GridSpec(1, 3, width_ratios=[1, 1, 0.2])\n",
    "# cbar_ax = fig.add_subplot(gs[0,2])\n",
    "# ax = fig.add_subplot(gs[0, 0])\n",
    "# ax = fig.add_subplot(gs[0, 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "for promoter in ['weak', 'strong'][:]:\n",
    "    width = 1.7\n",
    "    height = 4\n",
    "    figsize = (width, height)\n",
    "\n",
    "    cbarLabel = 'log10(norm.\\nDAMRatio)'\n",
    "    plotDf = damPlotDf.loc[promoter]\n",
    "    plotDf = sort_codon_index(plotDf, refCodonTableDf,\n",
    "                              addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False)\n",
    "\n",
    "    # split in 2\n",
    "    fig, axes = plt.subplots(1, 2, figsize=figsize, gridspec_kw=dict(wspace=0.7))\n",
    "#     fig, axes = plt.subplots(1, 2, figsize=figsize, gridspec_kw=dict(left=0, right=0.85))\n",
    "    n = 34\n",
    "    plotDf1 = plotDf.iloc[:n]\n",
    "    plotDf2 = plotDf.iloc[n:]\n",
    "    m = n - len(plotDf2)\n",
    "    row = plotDf.iloc[0].copy()\n",
    "    row[:] = np.nan\n",
    "    row.name = ''\n",
    "    plotDf2 = pd.concat([pd.DataFrame(m*[row]), plotDf2])\n",
    "    plotDfList = [plotDf1, plotDf2]\n",
    "    cbar_ax = fig.add_axes([1.0, 0.4, 0.035, 0.15])\n",
    "\n",
    "    for i, ax in enumerate(axes):\n",
    "        if i == len(axes) - 1:\n",
    "            cbar = True\n",
    "            cbar_kws = dict(label=cbarLabel,\n",
    "                            ticks=[-vmaxCodon, 0, vmaxCodon], aspect=15)\n",
    "            cbar_ax2 = cbar_ax\n",
    "        else:\n",
    "            cbar = False\n",
    "            cbar_kws = None\n",
    "            cbar_ax2 = None\n",
    "        sns.heatmap(ax=ax, data=plotDfList[i], square=True, cmap=cmapName, vmin=-vmaxCodon, vmax=vmaxCodon,\n",
    "                    yticklabels=True,\n",
    "                    cbar=cbar, cbar_kws=cbar_kws, cbar_ax=cbar_ax2)\n",
    "        ax.tick_params(axis='both', pad=0)\n",
    "        ticks_font = FontProperties(family='Liberation Mono', size='x-small')\n",
    "        for tick in ax.get_yticklabels():\n",
    "            tick.set_fontproperties(ticks_font)\n",
    "            tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "        if i == 1:\n",
    "            ax.set_ylabel('')\n",
    "#     axes[0].annotate('{} promoter'.format(promoter), xy=(0.5, 1), xycoords='figure fraction',\n",
    "#                      xytext=(0, 0), ha='center', va='top')\n",
    "#     fig.tight_layout()\n",
    "    fig.suptitle('{} promoter'.format(promoter))\n",
    "#     fig.patches.extend([plt.Rectangle((0, 0), 1, 1,\n",
    "#                                   color='g', alpha=0.2, zorder=1000,\n",
    "#                                   transform=fig.transFigure, figure=fig)])\n",
    "    \n",
    "    filename = 'ELM_seq_codon_promoter_{}'.format(promoter)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.png'), dpi=800, bbox_inches='tight')\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.svg'), dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correlation weak - strong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = damDfCodon[damDfCodon['position'] == -1]\n",
    "# df = df[['log10DAMratio_norm']].unstack('promoter')\n",
    "df = df[['DAMratio']].unstack('promoter')\n",
    "# df.columns = df.columns.droplevel(0)\n",
    "df.columns = df.columns.map('{0[0]} {0[1]} promoter'.format)\n",
    "xlabel = df.columns[0]\n",
    "ylabel = df.columns[1]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ci : int in [0, 100] or None, optional\n",
    "\n",
    "Size of the confidence interval for the regression estimate. This will be drawn using translucent bands around the regression line. The confidence interval is estimated using a bootstrap; for large datasets, it may be advisable to avoid that computation by setting this parameter to None.\n",
    "\n",
    "n_boot : int, optional\n",
    "\n",
    "Number of bootstrap resamples used to estimate the ci. The default value attempts to balance time and stability; you may want to increase this value for “final” versions of plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "width = 1.5\n",
    "fig, ax = plt.subplots(figsize=(width, width))\n",
    "pearson = df[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.05, 0.92),\n",
    "            xycoords='axes fraction', ha='left', va='top', fontsize='medium',\n",
    "            bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "plt.tight_layout()\n",
    "sns.regplot(ax=ax, data=df, x=xlabel, y=ylabel, ci=95, n_boot=1000, fit_reg=True,\n",
    "            scatter_kws=dict(alpha=1, s=20, edgecolors='none'))\n",
    "ax.set_title('codon position -1')\n",
    "ax.set_xlabel('DAMRatio strong promoter')\n",
    "ax.set_ylabel('DAMRatio nweak promoter')\n",
    "filename = 'ELM_seq_codon_pos-1_correlation_weak_strong'\n",
    "fig.savefig(ctermELMseqPath / subFolder / (filename + '.svg'), dpi=800, bbox_inches='tight')\n",
    "fig.savefig(ctermELMseqPath / subFolder / (filename + '.png'), dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relative differences min and max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'weak'\n",
    "df = damDfCodon.loc[promoter, 'DAMratio']\n",
    "mean = df.mean()\n",
    "print(df.idxmax())\n",
    "print(damDfCodon.loc[promoter, df.idxmax()])\n",
    "print((df.max() - mean)/mean)\n",
    "\n",
    "print(df.idxmin())\n",
    "print(damDfCodon.loc[promoter, df.idxmin()])\n",
    "print((df.min() - mean)/mean)\n",
    "\n",
    "print((df.max() - df.min())/df.min())\n",
    "print(df.max()/df.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meanDAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for promoter in ['weak', 'strong']:\n",
    "    for pos in [-2, -1]:\n",
    "        df = damDfCodon.loc[promoter]\n",
    "        df = df[df['position'] == pos]\n",
    "#         mean = df['DAMratio'].mean()\n",
    "        mean = meanDAM.loc[promoter]\n",
    "        print(mean)\n",
    "        for aa in aaTable + ['*']:\n",
    "            codonList = refCodonTableDf[refCodonTableDf['aa'] == aa].index\n",
    "            fold = df.loc[codonList]['DAMratio'].mean()/mean\n",
    "            print(promoter, pos, aa, '{:.2f}'.format(fold))\n",
    "\n",
    "        # compared K at position -1 to the mean of the other a.a..\n",
    "#         df.loc[codonList]['DAMratio'].mean()/df.loc[df.index.map(lambda x: x not in codonList)]['DAMratio'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.idxmax())\n",
    "print(damDfCodon.loc[promoter, df.idxmax()])\n",
    "print((df.max() - mean)/mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'strong'\n",
    "df = damDfCodon.loc[promoter, 'DAMratio']\n",
    "mean = df.mean()\n",
    "print(df.idxmax())\n",
    "print(damDfCodon.loc[promoter, df.idxmax()])\n",
    "print((df.max() - mean)/mean)\n",
    "\n",
    "print(df.idxmin())\n",
    "print(damDfCodon.loc[promoter, df.idxmin()])\n",
    "print((df.min() - mean)/mean)\n",
    "\n",
    "print((df.max() - df.min())/df.min())\n",
    "print(df.max()/df.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damPlotDf2 = 10**damPlotDf\n",
    "damPlotDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damPlotDf2[[1, 2]].stack().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damPlotDf3 = sort_codon_index(damPlotDf2.reorder_levels([1,0]).reset_index(level=1), refCodonTableDf,\n",
    "                          addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=True)\n",
    "damPlotDf3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = df[[-2, -1]].stack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damPlotDf3[damPlotDf3['aa'] == 'K']\n",
    "df = df[df['promoter'] == 'weak']\n",
    "df.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damPlotDf3[damPlotDf3['aa'] == 'K']\n",
    "df = df[df['promoter'] == 'strong']\n",
    "df.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damPlotDf3[damPlotDf3['aa'] == 'C']\n",
    "\n",
    "df = df[df['promoter'] == 'weak']\n",
    "df = df[[-2, -1]].stack()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damPlotDf3[damPlotDf3['aa'].map(lambda x: x in ['A', 'I', 'L', 'M', 'V', 'F', 'W', 'Y'])]\n",
    "\n",
    "df = df[df['promoter'] == 'weak']\n",
    "df = df[[-2, -1]].stack()\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damPlotDf3[damPlotDf3['aa'] == '*']\n",
    "\n",
    "df = df[df['promoter'] == 'weak']\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Amino acid pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDf3.loc[:, 'aa_pair'] = damDf3['aa_-2'].str.cat(damDf3['aa_-1'])\n",
    "damDf3.head();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 405,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDfGroup = damDf3.groupby(['promoter', 'aa_-1', 'aa_-2'])[['DAMratio']].mean()\n",
    "damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "damDfGroup;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "promoter\n",
      "strong    0.993079\n",
      "weak      0.143181\n",
      "Name: log10DAMratio, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "meanDAM = damDfGroup['log10DAMratio'].groupby(level=0).mean()\n",
    "print(meanDAM)\n",
    "damDfGroup['log10DAMratio_norm'] = damDfGroup['log10DAMratio'] - meanDAM\n",
    "damDfGroup;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApair = damDfGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16.12791399398891"
      ]
     },
     "execution_count": 408,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDfAApair.max().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.4066934412486249"
      ]
     },
     "execution_count": 409,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDfAApair.min().min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>log10DAMratio_norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_-1</th>\n",
       "      <th>aa_-2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">*</th>\n",
       "      <th>*</th>\n",
       "      <td>6.839328</td>\n",
       "      <td>0.649270</td>\n",
       "      <td>0.081140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>5.264588</td>\n",
       "      <td>0.539964</td>\n",
       "      <td>-0.028166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>4.508715</td>\n",
       "      <td>0.460728</td>\n",
       "      <td>-0.107402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>7.270627</td>\n",
       "      <td>0.691335</td>\n",
       "      <td>0.123205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>7.035366</td>\n",
       "      <td>0.671121</td>\n",
       "      <td>0.102991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             DAMratio  log10DAMratio  log10DAMratio_norm\n",
       "aa_-1 aa_-2                                             \n",
       "*     *      6.839328       0.649270            0.081140\n",
       "      A      5.264588       0.539964           -0.028166\n",
       "      C      4.508715       0.460728           -0.107402\n",
       "      D      7.270627       0.691335            0.123205\n",
       "      E      7.035366       0.671121            0.102991"
      ]
     },
     "execution_count": 434,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = (damDfAApair.loc['weak'] + damDfAApair.loc['strong'])/2\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              2.599727\n",
       "log10DAMratio         0.202173\n",
       "log10DAMratio_norm   -0.365957\n",
       "Name: (L, W), dtype: float64"
      ]
     },
     "execution_count": 436,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['DAMratio'].idxmin()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              9.259017\n",
       "log10DAMratio         0.792999\n",
       "log10DAMratio_norm    0.224869\n",
       "Name: (K, N), dtype: float64"
      ]
     },
     "execution_count": 437,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['DAMratio'].idxmax()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfAApair.loc['weak']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              0.545114\n",
       "log10DAMratio        -0.263513\n",
       "log10DAMratio_norm   -0.406693\n",
       "Name: (L, W), dtype: float64"
      ]
     },
     "execution_count": 426,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['DAMratio'].idxmin()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              2.408301\n",
       "log10DAMratio         0.381711\n",
       "log10DAMratio_norm    0.238530\n",
       "Name: (Q, K), dtype: float64"
      ]
     },
     "execution_count": 427,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['DAMratio'].idxmax()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfAApair.loc['strong']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 429,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              4.590017\n",
       "log10DAMratio         0.661814\n",
       "log10DAMratio_norm   -0.331265\n",
       "Name: (M, M), dtype: float64"
      ]
     },
     "execution_count": 429,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['DAMratio'].idxmin()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              16.127914\n",
       "log10DAMratio          1.207578\n",
       "log10DAMratio_norm     0.214499\n",
       "Name: (K, N), dtype: float64"
      ]
     },
     "execution_count": 430,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['DAMratio'].idxmax()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApair.to_csv(ctermELMseqPath / subFolder / 'damDfAApair.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAMsAAACkCAYAAAA5Wdq8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJztnXdcVNe2x38DzAxdQClSRAmiYImiJmg0uQaN13rVqKBgLKjBcjUaLBHFhjUQn896LbHFJGAh+tSoiJprL0gEFfRClGaGNoDUocx5f3DnxGHvw4zJCAL7+/nwAdYs9uw5zJp91t6riDiO48BgMDSi19ATYDAaC8xYGAwtYcbCYGgJMxYGQ0uYsTAYWsKMhcHQEmYsWnL79m1s3ry5QZ77zJkzDfK8DHWYsTQCvvvuu4aeAgOAQUNPoD6ZNGkSDh48iN27d6OkpATz58/H9OnTERoaiq+++grl5eUYNGgQJk2ahCNHjiA6OhovX77EwoULoadX87kSHx+PHTt2YOvWrRCLxQCAwMBAWFhY4MmTJ5g5cyY8PDwwe/ZsSKVSbN68GWvWrEFxcTEcHBywfv16zJkzB5aWlnjy5AlGjBiBq1evorCwEIcOHUJ8fDzCw8MBAFOmTEFhYSGePn2KEydOwMjICAcPHoRYLMbKlSuRm5uLsLAwiMViHDhwABKJpMGubbOAa0asXr2ae/bsGbd48WJu3rx5nEwm41asWMGtWbOGu3//PsdxHDdr1iwuLy+PO3jwIMdxHJeYmMh99dVX3K1bt7igoCBu0qRJXFFRkdq4/v7+XGJiIldSUsKNHz+eS09P53x8fDiO47hvv/2WO3HiBMdxHLdt2zbu3LlznL+/P5eUlMQ9ePCAmzBhAsdxHD+HCRMmcC9fvuQqKiq4sWPHclVVVZyvry//vbKyksvKyuLmzp3L3bp1i1uwYEF9Xb5mT7NaWXr37o27d+8CADiOw7Vr1/DBBx8gIiICYWFh0NPTQ2FhIWQyGaqrqxEUFAQDgz8uUWxsLGxsbNRkACCVStGxY0cAgL6+PgDAwcEBAPD8+XP4+/sDALp06YLExEQAgIuLC7KysuDk5AQAMDExQUVFBQDAzMyMHyMvLw8AkJ+fj+fPn2PKlCkAwK9qjo6OurxEjDpoVj6Ll5cXjh8/DhcXF3Ts2BFHjhyBl5cXnJycEBISgsOHD8PHxwcODg745ZdfEBYWhgEDBoD7b/jc8OHDMW7cOOzdu1dtXIVCgfT0dJSWlvLGIhKJAABOTk549OgRgJpbONWbW/V4bZRKJYqLi1FZWYm0tDRYWloCACwtLeHm5oZDhw5h+/bt+Pjjj3V/gRh10qyMxdTUFADg6emJnj17QiKRwMzMDDNmzMCmTZswbtw4pKSkwMzMDAYGBhg3bhxOnTqFwsJCfoyRI0fi+vXrePHihdrY4eHhmDhxIgICAtTkPj4+OHfuHCZMmIDMzEwMHjy4zjnOnTsXAQEB8PX1xeeffw6xWAx9fX1ERkZizJgxmDBhAiZNmgRnZ2cdXRWGtog4jkUd/1UmTpyI/fv3E7dnjKZFs1pZGIy/AltZGAwtYSsLg6ElzFgYDC2pd4/0yuquhEyeX0zVde3lReomP6Tqiv67Zfsqtl16U3U5ZRUhUxQVUjSB6vJSqtxg9F5Cpn9pMVXXZXA4KVRW0J9P35yQFVTRdcuU1YTMicum6qJSThWXF/1OqpYXUXXzU+Oo8g5Dt9Kfs4nBVhYGQ0uYsTAYWqITY4mIiEBQUBAeP36MFy9e4KuvvtLFsAzGW4VOfJYnT55g06ZNmD59OpydnTF//nxdDMtgvFXoxFgqKyuRmZkJBwcHTJkyBQqFQlA3N+8lIZNlF1B1i69cJmRWlmZU3aoq0tm1LKaPm3TzOiGzaGFC1X2YmEaV91NOIWSluaSzDABJ13oQMqHXIc8nnWv7tm2oui9zcghZtgW5QQAABlJjqrxKQW5gFBeQ/yMAUHJKqrzDUKq4yaGT2zBbW1ucPHkSNjY2OHnyJE6ePKmLYRmMtwqdrCxz5sxR+50FBTCaIjpZWb799lv+59TUVMybN08XwzIYbxU6WVneeecdrFq1Cq6urvj111+xatUqXQzLYLxV6GRlcXFxgbe3N6KiovD555+juJh+Is9gNGZ0EnW8bds2Qlbbj1Fx/9tRhKyylB5eYdOVDHepEAhLKXiWSMiceg+j6kqMWxIy2cMYqq5pa3qSlazD54SsU3kCVVdp9QEhK64mQ24AoLCqkpA9yKeHsLysJHcde7ayp+omv6SHuwy3dSKFHD28pkp+myo3bzOEKm9qvBEHn8FoirBwFwZDS3SystDCW9avX6+LoRmMtwadGEtoaCgAYMOGDViyZIkuhmQw3jp0YiyqSiclJSX8z6p6WLUxtScd5ufXLlB1XQd6ELJs+RWqroHUiJAV5zyn6hpbkQ6shbMHClIfk88Xf4s6RteuMwjZpbB/0nVHTiJk1h0+pepaich/iZNdW6ru7xVlVLmjREzIzPRtqLqoIkOCRNUlVNXM+z9R5czBfw1U4S329vb8z43N6acZSmOEZigM3cB2wxgMLdGJsaSlpWHfvn1QKBQwNTXF5MmTWVlRRpNDJ8ayY8cOBAcHw8zMDPn5+Vi7di3CwsJ0MTSD8dagE2MRiUQoKChAQUGNs6gqWs1gNCXqNJbs7GzExcWhf//+kEgkuHv3Lnr16kXoOTg4qOWwqCrI03DqMYGQPTx/SusJv8z4jSrPf5FJyBy9RlB1zR37EzIz6w5UXUkLunxv+gtCNs5/IVW3oiSPkBWJ6WEpOZXlhCz86jmqbmIa+ZrH9n6Pqqsnop8/x2ekE7J1Hwyg6prZu1LluiQmJgYymQyVlZWYPHkyIiMjUV5eDktLSxgbG6s9Vt/UaSyhoaEYMWIEQkJCsGzZMhw9epRqLMzBZwgxtjvZYGnMksOIiIjgf/fx8YGPjw8A4P79+1i4cCG++eYbAMC4ceOwZ88eKJVK4rH6pk5jMTMzw4ABA+Dp6YlNmzbVmS7MYNAwEpOtNV41jtqoiqsrlTUpzCkpKZg+fTrWrVsHIyMjtcfqmzpjw7y8vHD58mVYWVlhxowZfGOd2vzf//0fAODevXsAgEOHDul4mozGisSA/KqLzp0748CBAzA0NMTDhw8RHR2NPXv2oFOnTvxj1tbW9TP5WtQ59eHDh/M/37x5U7AR6MOHDzF8+HBER0ejZ8+eyMwk76UZzRPaylIXAwcOVPu9c+fOupzOX0LrqOObN28KPlZUVIT09HS17wwGUGMstb8aK1pvHc+YQcZCqVCFubz6XYiXsgeEzK4N/QCTlqQllIxFSxQzNKXHQ3EKcuWrpuxCAUBpzn2qvKSqLSETG9HLEGW0Hk7IWlSTSV4AkFlGj8uiYWpkSMjyy+nxYuZSUhcAAjzJ61YtkA9Y3H6q1nNTITVovMZRG62NRdVglAbbDWMIYdiEjtxYXzfGG6VZrSyrVq1CSUkJ34WXltQVFRUFT09PPH/+HJGRkfDy8sLEiRN1P1tGo8OwCS0tGo3FxsYGM2fOrFPn0aNHGDVqFHbu3Int27dj3bp1Opsgo3EjETedmxeNryQ6OhqJiYkwNKxxEDdt2kTolJWV4d///jef8FVSIuykVpWTdXQriug1iRXFZFUTA0N6zd6SrAxC1qpdP6quSJ+sa6ynT/8ElJrRk9jM86SUudHrF7tLyDrMORz59wDQ2oicm52lBVXX1JB02i0NySQ4AHC3pG92WIvJeYgFQmOcK59S5UArAXkzW1lOnDiBxMREKJVKeHiQmYsAMH78eNy7dw9Tp05FbGwsPvnkE51PlNE4MZSS4S6NFY3GEhoaitatW4PjOJw5cwaLFi0idDp37swfHvXoQVaMZzRfJJqO7BsRWh1KBgQEYNq0aYIFv7dubR49BRmvj6FUQnw1VjSavYWFBVauXAmRSISWLclDQgCIi4vDsWPH1GRjxozRzQwZjZqmtLJozGcZMWIEysvLwXEcjI3pzrWpqSmcnZ1ZqwkGgUTgPdMYqdNYrl69it9/V+9mRTut79WrFzXPhUbqNbLR0a8Jz6i6rXvcIWRC3bW4anLHKffZVaquxMiSkJXkkUlQgHBr74kDNxOyvDh6EputEbkTZQN6mLm1mIyoXdTpfaqujBKiY6ZP/5cqBMLaabthBmVPqLpViteP+RPqONYYEfRZCgsLYW9vj4KCAvTo0QM9evTAb7/RsxTZASRDCLGxKfFVFzExMThy5AgOHDgAAJDL5Zg8eTIUCgXi4+OxfPly7Nu3rx5mTiK4shgZGSErKwsVFRXIysoCULNFzGC8DmIT8uwpIiJC60zJnJwcdOrUCQCQlJQEOzs7iEQNE0IjaCwSiQSVlZWwsrLC8+fPAdR09aLdbr1a61j1QtgpPgMA9CmVQn0+1T5TskOHDvyB+Icffgg7Ozvs2bMHMpkMdnZ2b2jWdOr0Wfr374+KigpwHAelUsnHh9UmNDQUlZWVuHDhAuLi4uDq+uYLGzAaB7SVpS5qZ0q+mvz16NEjHD16FBUVFa+dLbl58+a/3HJeYzOjb775BsOGDcPatWvh6uqK5cuXUyfy7NkzeHt7w9PTE4BwrePn11YQsqfnjlJ1e0xcRsjSbp2g6lq0I1MIhMJdinPIsA2RHt19MzRrTZWfF5HPN1iP7tMZtCQd9JegnzeUUjYqIlOTqLq5lCQ7P/duVF2FkhwXAGzEZMiMnQH9LVGZc40qb9HuH1Q5APx2meyw4NK/fjosVFVVQSaT8bu0Qu9JbdG4CV5aWoo7d+5g9uzZOH/+PFVHLBbDzc0N6enpSE+v2VViOS4MADAwrNuhf5MsWLAArq6u0PvvB6HqPZmTk4NDhw7hxYsXMDAwgJ2dHQICAmBuTk/eU6HRWMzNzfHw4UN4enry1TVqwwyDIYS+AT1Dsz5wd3fHlClTIJGor+JXrlxBYGAgTExqglaLi4tx7tw5jQfpGo3F398fp06dwp07dzBt2jSqTlxcHCIjIwEA1dXVsLGxwZw5c7B3715mSM0cAwkZRV1fJCYmIiQkhMjFGjt2LPLz83Hs2DFwHIeRI0dqFXGi0Vg2bNiAiRMnguM4hIeHY82aNYROZGQk1q1bx++E/frrr/D39+f9F0bzRb8BjcXFxQVffPEF9bH169fD398fAATf17XRaCxisRhdunQBACL+S0V1dTWSk5Ph7OyMkpISxMbG4rPPPsOIEfTyqYzmgz5lA6G+SE1NxZw5c/gwrVdzscRiMbp27QpA+H1dG43G0r9/fyxcuBAcx+Hvf/87VWfJkiU4fvw40tPTYWxsDC8vL/ztb3+j6tq6jyVkykp6K2nOgayK0oYsRgIAMLYmd4Ey9WypullSshZVd4EYphdV9DCRORtDCVli8Gqq7g8Z/yFkh6/doOqmpKYSMtlNetgOKBVCwwWKsosc6VVxrJ3aELKh/fpQdR1aWlHla9rRpwc07Mri4eGB6dOnUx/T5n1dG43G4urqitjYWOjr6wtWeLGyshKcFKN5o29A3xSqD2JiYpCbm8uvLPPmzQNQ4+Dn5+fjvfdqiqiruj9oQqOxREREYPbs2VAoFAgPD2cn84zXQq8BjSU8PByFhYUAgBYtWvByW1tbVFdX81vF2dlk+joNrZIN5HI5lEolOI5Deno6cbhz8eJF2NjYoGvXroiJiQHHcRgwgN62gNHM0Gu4ZK+oqCjI5XIANYVXAgMDkZ2djfPnz+PevXvo1asXOI5DbGwshg4dqnE8jZmSJiYmOHXqFE6fPk30YVERExPDO0ve3t64dOnS674uRlNFz5D8qicKCgoQEhKCkJAQvHxZUyjFxsYG48aNg4+PD8aOHQsfHx/s3LlTq/E0rizanJMolUqkpqaidevWSE9PrzMJTPboR0KWdIEewtKjlQshK80nHWAAyE+NJWSmNuTfA4BtSzJ2LT/1IVXXkKOHifz8JVmLQCKnO+Kf2bQlZBMnTKHqvqggc1ROpdPbgJdVVxGyIU7tqbpFVfRNFDMD8pPfVeDUveQ/B6hygGwOpYLTo1exqS92794NkUiEioo/Xr+9vT3OnDmDdevWwcDAAOXl5di1a5fGsXSS87lgwQIcPXoU2dnZcHR0RFBQkC6GZTQFRA13G7Zs2TIkJyeD4zi0b6/+IZKamoohQ4bA1dVVLV2gLgSNpbCwEI8fP0ZMTAy8vb0B1OxHh4eHE7q2trbspJ5Bpx5vu2qzatUqFBcX82H/r1ZTraqqgrW1Nf8hrw0ak78UCgWf/OXr60vVPXv2LDp27IjIyEjI5XK4urrWWXWf0Yx4TQe/dk/JqKgolJaWQiqVon379rh79y7Ky8u1+nC2sbHBihVklDsArFy5EkqlEnp6elonkwk6+BKJBCNHjsSwYcOQkZGBjIwMQcf9l19+gYuLC4yNjbFp0yatLZXR9KmGAfEVERGB0aNH81+v3gbdv38ffn5+/C5WcnIy/Pz8kJaWhujoaEybNg3GxsZanY1ER0dj7ty5WLRokVq9uwMHDiA4OBgPHjzAy5cvsWPHDq1ei0af5ciRI+jSpQvc3d1x8eJFqo5IJEJaWhoCAwORnJzMmhkxeCo5MgLidXpKvvr7q8mHmvpKFhcX48QJ9Y2joqIimJmZ4cmTJwgLC4Ovry969eqllXMPaGEsxsbGePfdd5Geno7KSnoDnvnz5+Po0aPIycmBg4MDFi9eLDheZWkxOQmBDExaM6Lyl/TqLgbm5A6OebthVF2uhEzSMjCk5zKYWNFjOXakPCZky9+hp7kel5P/2H7W9N0pGj1t6K3SkwvlhOxeLtlyHABMKbteAFBM2SWrtKK/js5thwhNURChqjJC1M6UdHR0xHfffYd27dqhXbt22LVrF5RKJays6KE3Kk6ePInCwkK0b98e+vr6ePToEezs7DB27FhIpVKIRCI4ODi81maURmOZPHkyLCwscP/+fcFyR8zBZwhRLrD1LoSmnpLaRrKrbuXu378PpVIJPz8/3sDi4+OxZcsWpKSkYMuWLQD+CIWpC43GoooHCwwM1GqSDMarVDZQG+4rV67g9OnTAACO4yCVStGvX02a+bZt2wD8UTVVWwdfJ+csUVFRuHHjBiZPnsyXrWEwAEBB8Vnqg0uXLiEsLIz/fdmyZbyx+Pn5oVOnTmoVVmmtVGpTp7EsXLgQIpGI317T19dHaCgZmp6QkIANGzZg7dq1zFgYalQ00MpSWVmJqKgoWFtbIycnR+2xf/3rX7h06RLKysrQtWtXfPjhh1qNWaexfP3111i/fj3mz58PkUjEFz6jTezFixcoKyvjC1YIVdIws3mHkNl06ELVlZiSjq3EKI2qS20kxJHhIABQZt6TkOmb94RJaQI5hBE9fOTs7f2EbPk79OBRdwuy2c+zMvqO4f0c0kH//io99yW71ptAhbkZeS1sBBxi9zbkNTanlHQFAOuWdMe/jnQWlAtUlXnTrF69Grdv30Z+fj4cHBwwcuRI/jE3Nze4ubmhqqoKe/bswddff41z585pHFPjbVhBQQGePq0pHaQKRquNra0tTp48qRZo2dgcfpqhNEZohtKQlFfRd1DfNDKZDM7OznB2rkl6y8jI4D/Ab9y4gatXr6Kqqgp9+vRBQECAVmNqNJYlS5bg559/hkgkwpIlS6g6tQ2juJjcHmY0T0oayFjWr18PY2NjtG3blpep3qcrVqyAu7s7pFIpzp49i59//vmv+yw7duzA8+fP+SjiuLg4wUEVCgViYmIQExMDiURC7WrMaH6UCKSMv2k2b96MH374AZMnTyYei46O5n/Oy8vD0aP0Io+1qdNYZs2ahZSUFJiYmCA7OxtmAkv8okWLoK+vj379+qFVq1ZqtY8ZzRtaGkF9IJVKqYai4t69e/jxxx8RFxen9ftVY/LXtm3bYGpqCkdHR35/ujZdu3aFQqFAQkICn8bJYAA1K0vtr/pi2LBhCAwMRGBgIIYNG4bdu3cDqAkIvnPnDoKCgtC3b1+ts3o1+ixGRkZ84kztyn4q/P394e/vj8zMTJw+fRr//Oc/BftMGrUgd1+yn9Cd6w7DyBAUoWoh1MIIYnrLaSMRJbzGoO7widroUWojc0bkTh8AOCnJQ68ESpIXAJhJyJ2oTm3pO4vOtmQ4ULXAVu2HHd2o8v725F6WVKC1d2oZ3Rdt10L42pU30MoCAF5eXli2rKZe9saNG3lfes6cOThz5gzCw8Mhl8tRVVXFx6DVhUaNmTNnYv/+/aiqqhKsSKnCwcEBn3/+uTavg9FMaCifBag5md+zZw9EIhHKy8v5QMz33nsPz549Q0pKCpydnREQEICDBw9qHE+jsTx48AB5eXngOA579+6lOu4KhQJSacOmjzLeTsoqGs5YgoOD8euvv/If9Ko7pJCQEPj7+8PHxwcJCQl8/yFNaDSWK1euoE+fPvDw8BBMv1y+fDmkUim8vb3Rt29frZY0RvOguIx+u6mJly9fYv/+/aisrMSMGTNgbm4OjuNw8OBBGBoawtfXF4sWLYKrqyu8vb3xzjvkLXBoaChat24NjuNw6dIlPqelurqaD9Ds0aMHfvyRrAtBQ+O7Wl9fH46Ojrhy5YrgoeSmTZtQVlaGK1euYN26ddDX10dwcLBWE2A0bYrLX89Yvv/+e9y9exfnzp1DREQE9PT0cOvWLXzyyScoLi6Gi4sLMjMzUV5ejtLSUpSVlcHUVLitherAcePGjbystLSUjzThOA5lZWVazU2jsSxYsAASiQRZWVl8uSMaKSkpSExMhEKh4GsjMxhlCvJQsq6ekhMmTMCECRMwePBg6Ovr813nAMDMzIw3Fj09PaxevRocxyEiIgKzZs1Se4709HRYWFhg6tSpMDAwgIWFBf+Yu7u7WkkvoUqrtdFoLLa2NfWChw8n6w6rmDt3Ljw9PeHn58frvw6WTm3pD1SRqaNChaY5Wt6EiP7yRBVZFKHALrqS/snYt7M7ISvl6GOYUJLbzATir/pY0xO9aOSVk5+IWQJb9+8KxHW5GJK7i3qgh6xfz8mgyv9mR9ZLVlFCWVl8xk8VzJRU4eXlhV27dkEsFiMgIADnz5/HoEGD+McrKiqwZcsWtGzZkrr1qzKGDh06gOM4tdXnz4Zi6cS5+N///V9dDMNogpQp/pyDb25urpY3rzIUR0dHvmv2qlWrBP/+TcQmapWDHxtbU8Du/fffp34i3Lx5E7179xb8ndF8eV2f5W1Go7EkJyfzofmvOkmvsnfvXmRmZgKocZjOnTvHjIUBAGqVIBs7Go2lvLwcZ8+eBQDBUBYTExM4OzvzAZeqXn0MRjmlh0xjRaOxrF69Gjdu3ADHcYL3iLNmzVLbUair66vUzJ6Q6Qs4u2VGpBNtbCHwSWVAPmeqgr4laEMJgzHk6LcLWUp65ZmzN+8Qsi+60VdTOUe+YY4n02srf+JMJpsVKOhzS8/LI2RPM+nVb7q1pm8c/E6pttPO3JKq+yRHoB5cB7oYqPmwbSpoNJbt27cjMzMTBgYGOH/+PPUEv/bWm7ZbcYymj6I53YZJJBJ8/fXX9TEXRhOkWa0st2/fRlFREQwNa843tKmvdPz4cXz6Kb1NAqN5USFQmLExotFYXm2Lp219JVXOPoPRbFYWVVrxq9DSio8ePcobEsdxSE5O1t0MGY0argFD9HWNxrRiuVyOp0+f1tnNS1VBQ0WdOS2UFgRpj+Kpqu4KsstXfsY9qq4ppZtXtTk9xEOfskJWi+itvW1RSpX/lkjWOlaCfo1oxbFl+fQq8HuzbxGyiP9WVtQGS6uWVPnDp2R7cQDo59mdkI3q+i5Vd5gruTupkea0dbxy5Up89NFHdfaxULVIZjAIKpqRsXh4eKBbt26CKcUMRl1wzcnBT0hIgEKh4FMyaQFqU6dOhYeHB79jJqTHaIYIHKg2RjQai5ubGwICAmBkZCR4G7ZixQpcvHgRubm5GDNmDDVrjdFM+ZMrCy1TMj4+HrGxscjPz8e0adOIx980Go0lPz8fa9eu5X+nneBfu3YNcrkczs7OGvNZOIMWhKzH2OlU3QrDtoTMwlXg/IaycSArojvnOZTKKp1N6SEeuVX0Dwju90xCVihQfXHX/ZuE7Nx10pEHQE3J5rIp+TcAIM8lRWav96ap7kY68788o+9m9m37+h+CXDRZp7mu5K+6MiW7du2K6upqREVF4fr16/j4448hEon4x980Go1l4sSJSEhIAMdxgu3v8vPzYWRkhNzcXBw4cAAAuw1jCFNXm7y6MiVTUlLQvXt33L17FxzHQU9PT+3xN41GYwkLC0Pbtm3h5uaGhAR6fS9mGAxdQ8uUtLKywqlTp6BUKtGnTx/s3r2bf7w+0GgsVlZW8Pb2RkFBAVq0IG+hAOCnn37CyJEjER0dTbQ5YzD+DEKZkq+2anz18fpAY/nWoUOHwsrKCrdu3YK1tTVVJzExEUBN/VgGo6micWXp06cPAPBlMGn8/vvvuHnzJv8dAMuUZDQ5dFKwon///pDJZPz3ugIuq8VkCEplGT0D83hmCiEb60CGtQCAPiV563E+vTNWXjm5S1ZhR+9QdSubXtEElMoqV2X0rmQOLclawNn/vkQf14pSn/l1HNgiem03IV7kke3BU7PoSV4Tu5Id05oTOjGWUaNG6WIYBuOtRqPPog2ZmZl1BloyGE0Bnawsx44dQ35+PgwMDGBoaAiRSIQvv/xSF0MzGG8NOllZ5s2bh08//RQmJiYwNTVlOfiMJonOyt3b2NjAxsYGGRkZKCoqwtChQ+lPWE467YVp9FyLXu+2Jv++jK4LJem0v29DbzptRCmnKtSvfW57Mt8DAFaIxYRsoH1bqm7oVdKZF7Wh60JJ3s5yGWRejxCidvQNEAi0BPHp40XIbqf8RtXtZEw/Z2su6MRYlixZArFYDLFYDCsrK7UOsQxGU0EnxrJhwwZdDMNgvNXoxGcJCgpiRSoYTR6drCyhoaHYv38/oqKi+FyWMWPG6GJoBuOtQSfGUlhYiJycHLRp0wZt2gj36mAwGjMiTgeniWvWrMHs2bNhZaW5PXZWwh5CZmJF37UqKyRDTaSmZDtrAHgpI2sHG5qRu2kAIDEiE71K5M+oumZ5AK3cAAAJLUlEQVS29Iom1RUlhKw4l540JTEhK65UldNzg2jXgnYdAMDQnHx9EhN68l11Jb0tN+11i2itzwFIKa8DAFq0+wdV3tTQycqyfPlyXQzDYLzV6MTBZzCaAzozlrKyMpSUkLcmDEZTQWfGcv36daxdu5YlgDGaLDozllatWsHY2LjeigcwGPWNTnbDAKC6uhrV1dWsciWjyaIzY2EwmjpsN4zB0BJmLAyGluivXLlyZX0/6datW3Hq1ClcvXoVMpkMnTt35uV2dnZ48OABLl68CE9PT15mamqKpUuXYsCAAbzuTz/9hAsXLqB9+/awsLDgZbdv30Z+fj6fhJaQkIAtW7bgxo0byM3NhYeHBw4dOoR9+/YhMjISqamp8PLy4se9fv06evXqpfZ8eXl5WLFiBWJiYtCyZUu0bt0aMpkMq1evxsWLFwGAj4uTyWRYs2YNLly4gIqKCri5uam9ZrlcDnf3msiAw4cP47vvvlMrPzp37lwMHjwYALBw4UL069cPYrEYSUlJWL16NSIjI2Ftbc2HFt2+fRs3b95Ep06d1K7xf/7zH5ibm2P+/PkYOXIkAODEiRPYsWMH4uLikJubC3d3dyxevJiv9/Y///M/cHBwwJYtW+Dm5gYDAwMsWbKEr9vVnNFZ8tfrMmPGDDg6OmLRokXw9fXl5Wlpabh06ZJgG/FXCQwMhIWFBcLDw7F69WpeVru50r179zBo0CB89NFHePasJrzjs88+Q4cOHZCWloaxY8eq6f/222+orFXQ+vjx45g2bRpcXV0RHBwMT09PHDt2DDNnzkS7du3w4MEDXlclb9u2LVJS/kh2mzFjBhwcHLB48WKMHj0aAHD//n1IJBKUlZXByMgIANCtWzfEx8fD1dUVEomEl3fs2BF+fn5IS0vDBx98oPH6ZGZm4uLFi7xhqhg/fjxf4gqoqa8sk8mQnZ0NuVwOe3t7zJ49G9u2bYO1tTWmT1evRV1VVYWEhATk5uaiT58+MDEx0TiXpkCDGcu+ffuQkZGBBQsWqMk3b97Mv5FU7Nq1S/Af0qJFC7W+hbt27YKpqSkGDhzIN1ny8fHBt99+izNnzmDw4MFo144ei6ZiwIABuHDhgposIyMDzs7OfP1dAMjOzoa9vT0A4N13/yiwnZ2dDTs7O2zZsgUlJSVYunQpPzepVIrJkyfzY9ra2qJz5864fPkyhgwZAgAYMmQIjhw5goyMDHh7e9c517pwcnJCWloace1++OEHXL58GT179sSgQYMgkUgQExOD7OxsiMVi6Ovro2XLlmjTpg1kMhmRJq5UKnH69GmUlpY2q/pwDeazBAQEoE2bNrC0VA9q3LBhA+Li4tR6WQYGBmLJkiXUceRyOczMzNR0g4OD1bqRPX78GLNnz8bGjRtx9OhRjXPr3r074uLi1OqfGRoaorS0JnVZdZZkb2+PjIyaIEfVrRgAODg4ID09HfPmzYNc/kddrsDAQKxcuRJdunQBAJw+fRpyuRwPHjxAdHQ0r2dnZ4e8vDzcuXMHffv21ThfIYYNG0ZNlRg/fjyCg4P5W6uWLVsiIyMDUqkU0lfSjz08PODqSqYpSyQSDBw4EKNGjYKpqemfnl9jo8FWFgCYOXMmvvnmG2zcuJGXSaVSBAcHIyQkBFu2bKnz73fu3InKykp88cUXvEy1sjg4OPCf4MXFxQgODoaxsTHef/99rebWvn17PHz4RySzl5cXTpw4gR49evBlbMeMGYOwsDCIRCJ8/PHHvO6YMWOwceNGmJqaon379oLPkZCQgO3btwMAFi9ejKKiIt7wu3TpgpSUFLVzq6SkJBw5cgRFRUWwt7dXuxU7e/Ysnj59Cjc3N/620tzcXKu+Jc7Oznj8+DGcnJyQlJSkzeXhfbzmBDtneQ1OnjyJQ4cO4fDhwzA2pjdsZTRd2Nbxa/CPf/wD3bt3x969ext6KowGgK0sDIaWsJWFwdASZiwMhpYwY2EwtIQZyxtGoVCgoKCgwcdg/HWajbGozjPexFhBQUGCumfPnsXjx4//1PPk5OQgMjKSH+PYsWPIyhJo862BU6dOaTy3YtRNgx5K6oLr16/jzJkzkMvlCAoKwuHDh2FpaYmnT5+iS5cuePz4MVauXIlnz54hIyMDS5cuxXvvvYfKykrMnj0bK1asgEQigZubG/z8/AAAWVlZCA8PR3V1NT766CN069YNO3fuRGlpKebOnYtnz55BLpdjw4YNsLCwUJtPUlIS9u/fj5KSEvj7+yM2NhYSiYSPxdq6dStfq8DHxwfJycm4fPkyKioqsHTpUhw+fBhFRUUwNDSEr68vMjMzkZeXB4lEAktLS5SXl2Pbtm3Iz88Hx3EICQnB6NGj4e3tjZycHNDiYuPi4pCWlsayWP8ijX5lsbOzw8iRI+Hs7IyEhAQoFApMnToV7du3x4ABA/Dhhx+qlZZ1d3fHnDlzkJmZiRs3buCDDz7AqlWr1GoHiMVijB49Gr1798bNmzcRERGBL7/8EmvWrOEPI3/++Wf4+fnhiy++gJ7eH5exRYsWGDVqFLp27Yp79+7B09OTj1xW4evri9mzZyMiIgLnzp3DunXr8Nlnn+HUqVPIzMxEmzZtMGzYMF7/1THKy8uRnZ2N5cuXo3Xr1nj8+DEf+FhcTK8N1r17dz7qmPHnafTG8v3336OwsBAdOnTgAxyNjY1hYGAAiUSiFvgI1MR4ATWRtkqlko//elXn2rVriIuL40Peq6qqoKenh6qqKuTkqPep1NPTUzOWn376CRkZGXB3dwfHcdT+mhzHQaFQQL9W6wulUslHQ69bt45fCV4do/aYSqVS7TWpSEpKwvz58xEREaHxGjK0o9HfhrVq1Qq3bt1CQUEBH6CoLX379sWKFSsQHx+vFjPWsmVLXLlyBTKZDEqlEmPHjsWmTZugUCgwZ84cADWRwZs2bYKpqamasdjZ2eHOnTsQi8UwMjKCvb09fvzxR7WQ+N27dwMAZs2ahcTERISEhKC6uhpffvkl9u3bh9LSUrzzzjv8uKoxXFxcYGRkhFatWmHDhg0QiUTw8PCgvraOHTti8+bNr3U9GHXDTvDrma1bt2LEiBFEzg3j7YcZC4OhJY3eZ2Ew6gtmLAyGljBjYTC0hBkLg6ElzFgYDC35fy1kuvOcpRPeAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 201.6x201.6 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAMsAAACkCAYAAAA5Wdq8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJztnXdYVMf6x7/Lsrt0EaUughJAwRZRE2OMxqBJLPGqsQb1Glswem0x6k8UsWOP13otsRewEIkdUROjxroqisaAIEWRsnTYhWXP7w8uJywzW8xdQWA+z7MP8J6XOXPO7rtzZuYtAo7jODAYDL2Y1HQHGIzaAjMWBsNAmLEwGAbCjIXBMBBmLAyGgTBjYTAMpF4aS0lJCS5cuFDT3XgtcnJy8Ntvv9V0N+o19dJY0tPTcfbs2Zruxmvxxx9/4NatWzXdjXqNaU13oDq4fv06fvjhB6jVakydOhU3b97EtWvX8Ouvv2L58uVo0KAB5s6di/379+PFixdo0KABVq5cicWLF8PMzAwPHz7EF198gX/+85/YtGkTfv31V0ilUojFYoSGhvLn+fLLL2Fra4ucnBysXr0aMpkMx44dg7W1Nb777jssWLAAarUavXv3xqhRozBkyBA0atQIcrkcvXr1woULF9CsWTMsXrwYe/bswenTpyESiRASEoLw8HDcvXsXn3/+OSIjIxETEwNbW1usWLECP/74I2QyGaRSKRYvXlyDd7puUy9GlkuXLmHSpEnYvn07VCoVhgwZgs6dO6Nr164oKirCwYMHkZaWBjc3N+zfvx89e/ZEWFgYAKBbt244ePAgIiIikJ2dDZlMhrCwMHz88cfEedLT07Fp0yYEBwdj9+7dAIAOHTpg69atWLduHVasWIFDhw7hl19+QUZGBl6+fIkffvgBPXv2RE5ODvbv34/Y2FhkZWXhl19+weHDh7F8+XKsXbsWQ4YMQb9+/aBWq1FYWIj9+/ejT58+OHbsGADg888/Z4byhqkXxjJ+/HicOXMGkydPRmlpqcYxZ2dnmJiYIDExEa1atQIAtG7dGomJiQCAZs2aQSwWw9zcHMnJyfD29gYAXrcyzZs3h5mZGXx8fJCSkgIAkEqlAIDc3Fw0adIEAoGAPy6VSiGRSGBnZwc3NzcAgFgsRkpKCpo3bw6BQIAmTZogLy+PP8fz589x48YNjBw5Env27EFWVpbGeRhvjnphLGfPnsW0adOwbds2bN++HQKBABUucQKBAADQpEkTPHr0CAAQExMDV1dXjeNAuWHFxcUBAJ48eUKcJzExESqVCo8fP4aHh4fG/1tbWyMlJQUcxyE2NhYuLi4abVfGxcUFT58+BQAkJyfDwsKC13V1dUX37t2xb98+BAUFoUOHDkQ/GW+GejFn8fLywoQJE2BlZYXhw4fDzs4OcXFxGitiPXv2xOzZsxEQEAAbGxusW7cOISEhGu3Y29ujbdu2GD58OGxtbWFnZ6dxXK1WY8KECVAoFFi1ahVu3LjBH5s2bRpmzZqF0tJS9O/fH46Ojlr7a29vjw8//BDDhg0Dx3FYsWIFJBIJoqOj0b17d3AchxEjRkClUmHNmjW4d++ecW4UQycC5nVsOCUlJYiMjMSgQYPw888/49mzZ5g6dSp/fPjw4Th06FAN9pDxJqkXI4uxEIvFkMlkOHz4MCQSCTZu3FjTXWJUI2xkYTAMpF5M8BkMY8CMhcEwkGqfs2TE7iZkzy7up+q2+YqcE+QkRlF11WUlhKyBS1uqrlDSmJAl3yb7BQAOXl2p8nQ7Um6XdoaqK3fqReqKJFRdFacmZKVqUqYNW1MxVW6qSKDKFdmPCZlQbEnVFQiEVLl1k88M7F3tho0sDIaBMGNhMAzEKMYSFhaGmTNnIjY2Fi9evMD//d//GaNZBuOtwihzlj/++AMrV67E+PHj4e7ujunTpxujWQbjrcIoxlJaWorU1FRIpVJ8/fXXUCqVWnUVeWmEzOvT8VRd2Z4JhMzczoGq69S6GyHLfXGfqpvx5CYhs2/xHlU3P+MpVS5M/J1sNz+Hqlv0Wzghs3z3Y6quqZkNIct7Sp4LACzsnQlZ7AO6rn3LjlR5TgLp42Yioi8SFGW+pMo7T2cTfINxdHTEiRMn4ODggBMnTuDEiRPGaJbBeKswysgyefJkjb+ZUwCjLmKUkeXHH3/kf3/+/LmGcyGDUVcwysjyzjvvYOHChfD09MS9e/ewcOFCYzTLYLxVGGVk8fDwgL+/PyIiIvDNN9+goKDAGM0yGG8VRvE6prmqV53HVJCXeJKQJd+JoOo26xFMkdLtW/HqGiEzb/gOVVelyCJkZSWFVF2OK6Ofz+FTQtZALafqwsSMEL1Q01eczE1Il5JcVSlFk04DUxFVnqospsqbSCxI3ZIiqq5n7nWq3PadgQb2rnbzRib4DEZdhLm7MBgGYpSRhebesnz5cmM0zWC8NRjFWJYsWQIACA0NxZw5c4zRJIPx1mEUY3nx4gUAoLCwkP+9SZMmVF3ahLlh09ZU3bI80hXjz6itVN28lHhC9t6EDVTd4txUqvzp6f8QsvZfb6LqmgjJibgy/TZVV0VZPHBpOpiqm//nPkLW1L0nVTfXhIzLAQAbyiTfriyTqltoYkXIWuIVVTcrK44qt6Wvo9Q5jGIsFe4tLi4u/O+1bdJPM5TaCM1QGMaBrYYxGAZiFGNJSkrCzp07oVQqYWVlhdGjR/MZHRmMuoJRjGXz5s0ICgqCtbU1srOzsXTpUqxevdoYTTMYbw1GMRaBQICcnBzk5JTHc4hE7LmZUffQaSzp6emQyWTo3r07xGIxbt26hY4dySAiqVSqEcOiK6O72IZcOsl8dpWqa2HrRsi4MhVVt8mHnxMygQndaK2lXQhZ68GNqLolBfSVs12vyIwr3zi6U3WVeSmkkKNfh5nnCEJ2M5/uRpNXkk7IWtnQryNJSXevKSkiV8l8rOjXIZaQgWnGJjo6GmlpaSgtLcXo0aMRHh4OhUKBhg0bwsLCQuNYdaPTWJYsWYJ+/fohODgY8+bNw5EjR6jGwib4DG0Mbkca6aA5+/j6NwAwdOhQDB06FABw9+5dfP/991i7di0AYMiQIdi+fTvUajVxrLrRaSzW1tbo0aMH/Pz8sHLlSp3hwgwGDXMRWQqjsnFUxdS0/COp/m+utPj4eIwfPx7Lli2Dubm5xrHqRqdvWKdOnXDp0iXY2dlhwoQJfOGcqvz8888AgNu3yzfl9u7da+RuMmorYlPypYtWrVph9+7dfHnCqKgobN++HS1btuSP2dvbV0/nq6Cz61988QX/+/Xr17F/Pz1zZEXNxaioKHTo0AGpqfTnfEb9gzay6KJnT01vBVqFtZrCYK/j69fpsQwAkJ+fj+TkZI2fDAZQbixVX7UVg5eOJ0wg0xJVUOHmUvmnNtTKDEJWoiWFkNj+A0LW0JNe3lpk3oAUCkm/p/JOkHmRXze/r1hBytWleRRNII2SF9mJo7f7pJhsw0xIf5tiC8l76WROv2YRJagMAEwo5fUstZxP4jWKKteFxLT2GkdVDDaWFi1aaD3GVsMY2jCrQ1turPIX441Sr0aWhQsXorCwEML/uqTTgroiIiLg5+eHxMREhIeHo1OnThg5cqTxe8uodZjVoaFFr7E4ODhg4sSJOnUePXqEAQMGYMuWLdi0aROWLVtmtA4yajdiUd15eNF7JVFRUXj8+DHMzMozlKxcuZLQKS4uxq+//soHfBUW0jOlAIDAhFyAU5fRM6iocmSEzN7zY6puZsIVUlimJSWTgNxVlifdoCgCNg4+VHkza9IlhOOyqbpuYvLbNfE1sq0kK+nZVmjE59FdY3ppccVJUJLvlRnIBRAAKHn5C/2kOrK71KuR5fjx43j8+DHUajV8fX2pOsOHD8ft27cxZswY3LlzB59+SqYJYtRPzCR0n7TaiF5jWbJkCZydncFxHE6dOoVZs2YROq1ateI3j9q3b2/8XjJqLWJ9W/a1CIM2JceOHYtx48ZpTfi9YQM91p3BMJOIiVdtRa/Z29raIiQkBAKBAI0a0d2/ZTIZjh49qiEbNGiQcXrIqNXUpZFFbzxLv379oFAowHEcLCzIyScAWFlZwd3dnZWaYBCItXxmaiM6jeXKlSt4+VKz2hNtt75jx47UOBcatDREd86dpuo6tfI3qE0A4Ggraib0N6q0gAzGMrMmq2gBgImQ/tjwsSUZvFWUTgZjAYC5Nbma5Vb8iH4+GzItlJ0pfTWsiZQMpJOr6CtZQopbCwA0pzwWFafSS5RbNGxKlevClLK6V1vROmfJzc2Fi4sLcnJy0L59e7Rv3x7Pnj2j6rINSIY2RBZWxEsX0dHROHDgAHbv3g0AkMvlGD16NJRKJR48eID58+dj586d1dBzEq0ji7m5OV69eoWSkhK8elWedG348OHV1jFG3UBkaU3IwsLCDI6UzMjIQMuWLQEAT548gZOTEwRaRsk3jVZjEYvFKC0thZ2dHRITEwGUV/WiPW5VznVccSFsF58BAEKJOSEb+qXhkZLNmzfnN8S7du0KJycnbN++HWlpaXBycnpDvaajc87SvXt3lJSUgOM4qNVq3j+sKkuWLEFpaSnOnz8PmUwGT0/PN9JZRu2DNrLoomqkZOXgr0ePHuHIkSMoKSl57WjJdevW/c8l5/UWM1q7di369u2LpUuXwtPTE/Pnz6d2JCEhAf7+/vDz8wOgPddx1h8HCVlMWChV970plCJHQnqGEVU2GeciMqfnAs5/9ZDUNaO/qWZ29DzMj9Rk/EwLJX3SrmjwPiG7n0cP0fa2JNtNLaG7xjynxAE5apkTeFnQ71t8MRmo10gkoepKs3+jynUVM3p2iayw4NG9eiosqFQqpKWl8au02j6ThqJ3EbyoqAg3b97EpEmTcO7cOaqOSCSCt7c3kpOTkZycDIDFuDDKMTXTPaF/k8yYMQOenp4w+a8/YsVnMiMjA3v37sWLFy9gamoKJycnjB07FjY2ulM96TUWGxsbPHz4EH5+fnx2jaoww2BoQ2hKlgisLnx8fPD1119DLNZcHr98+TICAwNhaVkeHVtQUICzZ8/q3UjXaywjRoxAZGQkbt68iXHjxlF1ZDIZwsPDAQBlZWVwcHDA5MmTsWPHDmZI9RxTLeHa1cHjx48RHBxMxGINHjwY2dnZOHr0KDiOQ//+/Q3yONFrLKGhoRg5ciQ4jsOaNWuwePFiQic8PBzLli3jV8Lu3buHESNG8PMXRv1FW26D6sDDwwPTpk2jHlu+fDlGjCjP/qntc10VvcYiEonQunX5JLeq/1cFZWVliIuLg7u7OwoLC3Hnzh2MGjUK/fr109sBRt1GKKq5x7Dnz59j8uTJvJtW5VgskUiENm3aAND+ua6KXmPp3r07vv/+e3Ach88/J/MJA8CcOXNw7NgxJCcnw8LCAp06dcLHH39M1TV37kbIOox1oJ9cRMoFxWSFLwAQWXsTsgSOkvEFQFOPloQsUUF3KXEQ09/sIzG/E7JvfOjhCYIysjR3aPR5et8cySXRF1n0oLJHceS9sLKkf5N7SOkZd3zcyLzUxUq6y0x7KX1L4J9UaTk1ObL4+vpi/Pjx1GOGfK6rotdYPD09cefOHQiFQq0ZXuzs7LR2ilG/EZrSF4Wqg+joaGRmZvIjy9SpUwGUT/Czs7Px3nvvAQBf/UEfeo0lLCwMkyZNglKpxJo1a9jOPOO1MKlBY1mzZg1yc3MBAA0a/PWU4ejoiLKyMn6pOF2LA2xVDAo2kMvlUKvV4DgOycnJxObOhQsX4ODggDZt2iA6Ohocx6FHjx4GdYBRxzGpuWCviIgIyOXlOQkcHBwQGBiI9PR0nDt3Drdv30bHjh3BcRzu3LmDPn366G1Pb6SkpaUlIiMjcfLkSaIOSwXR0dH8ZMnf3x8XL1583eti1FVMzMhXNZGTk4Pg4GAEBwcjL68806eDgwOGDBmCoUOHYvDgwRg6dCi2bNliUHt6RxZD9knUajWeP38OZ2dnJCcn6wwC4wqeEjJF3kuKJmBpTsa+FGXFUnUL5QmErJmXlm8LJRljIc19TFUVU7K4AMAw7zaEzLnsBVWXMyPbODHgS6pufAkZlyME3cs2pSuZvcZWy4KEuRa/PnNKWlcXIb2kgzrvPlWuC86E7jpTXWzbtg0CgQAlJX8tWri4uODUqVNYtmwZTE1NoVAosHUrvWR8ZYwS8zljxgwcOXIE6enpcHV1xcyZM43RLKMuQEk7VV3MmzcPcXFx4DgOXl5eGseeP3+O3r17w9PTUyNcQBdajSU3NxexsbGIjo6Gv395xOLRo0exZs0aQtfR0ZHt1DPoVONjV1UWLlyIgoIC3u2/cjZVlUoFe3t7/kveEPQGfymVSj74a9iwYVTd06dPo0WLFggPD4dcLoenp6fOrPuMesRrTvCr1pSMiIhAUVERJBIJvLy8cOvWLSgUCoO+nB0cHLBgwQLqsZCQEKjVapiYmBgcTKZ1gi8Wi9G/f3/07dsXKSkpSElJ0Tpx/+WXX+Dh4QELCwusXLnSYEtl1H3KYEq8wsLCMHDgQP5V+THo7t27CAgI4Fex4uLiEBAQgKSkJERFRWHcuHGwsLAwaG8kKioKU6ZMwaxZszTy3e3evRtBQUG4f/8+8vLysHnzZoOuRe+c5cCBA2jdujV8fHxw4cIFqo5AIEBSUhICAwMRFxfHihkxeEo5crHgdWpKVv67cvChvrqSBQUFOH78uIYsPz8f1tbW+OOPP7B69WoMGzYMHTt2NGhyDxhgLBYWFmjbti2Sk5NRWkq6bQDA9OnTceTIEWRkZEAqlWL27Nla2zMxcyRkihR6gSILiiuGmTV95ez51SOEzL4F/Q3hTMlAL7El2S8AUBW/ospfgIzTaGFJH6izVGQmGGshfZXoHdNcQpampm/siSkrXEUU1xoAMNHypPEwmyyIpC0vstiqOb0RHShfs1hq1UhJV1dX7N+/H82aNUOzZs2wdetWqNVq2NnZ6WznxIkTyM3NhZeXF4RCIR49egQnJycMHjwYEokEAoEAUqn0tRaj9BrL6NGjYWtri7t372pNd8Qm+AxtKDh60ndt6Kspaagne8Wj3N27d6FWqxEQEMAb2IMHD7B+/XrEx8dj/fr1AP5yhdGFXmOp8AcLDAw0qJMMRmVKa6gM9+XLl3Hy5EkAAMdxkEgk+OijjwAAGzduBPBX1lRDJ/hG2WeJiIjAtWvXMHr0aD5tDYMBAErKnKU6uHjxIlavXs3/PW/ePN5YAgIC0LJlS40Mq7RSKlXRaSzff/89BAIBv7wmFAqxZMkSQi8mJgahoaFYunQpMxaGBiU1NLKUlpYiIiIC9vb2yMjQnJf95z//wcWLF1FcXIw2bdqga1eyQC4NncayatUqLF++HNOnT4dAIOATn9E69uLFCxQXF/MJK7Rm0igli+3kv4ijqto4kQWKtLl8+/QnM4YUiuiTdosysg+cxBkwtSXkpiJyAgwAx26Qrh+fdPSiaAL5ZeQEn5ZVBQDKKB+uKy/p9ye3kB6DQ6u25WxNj+2hpXV9TilwBAC2pvTkE/TlgHIU6tebsxiLRYsW4caNG8jOzoZUKkX//v35Y97e3vD29oZKpcL27duxatUqnD17Vm+beh/DcnJy8PRpuT9XhTNaVRwdHXHixAkNR8taN+GnGEpt5G0rS6dQ0Vfn3jRpaWlwd3eHu3u5KaekpPBf4NeuXcOVK1egUqnQuXNnjB071qA29d7ZOXPm4MyZMxAIBJgzZw5Vp6phFBRoKU/HqHcU1pCxLF++HBYWFmjatCkvq/icLliwAD4+PpBIJDh9+jTOnDnzv89ZNm/ejMTERN6LWCaTaW1UqVQiOjoa0dHREIvF1KrGjPpHYSk9RPlNs27dOhw6dAijR48mjkVFRfG/Z2Vl4cgRco+Ohk5j+fbbbxEfHw9LS0ukp6fD2pqetXHWrFkQCoX46KOP0LhxY43cx4z6TTFlvlYdSCQSqqFUcPv2bRw+fBgymczgz6ve4K+NGzfCysoKrq6u/Pp0Vdq0aQOlUomYmBg+jJPBAMpHlqqv6qJv374IDAxEYGAg+vbti23btgEodwi+efMmZs6ciS5duhgc1at3zmJubs4HzlTN7FfBiBEjMGLECKSmpuLkyZP417/+pbXOJGdF5g62dqFnDTG1IctqC5RkQBgAcCWkW4qFhHSXKW+EctmUVToAKBHTM7W3caWsqJmTxYUAQFyiIGTWpvR7yYEMnPNsQC9PWGbTkJD9Iaev3tFcYwDgY0c3UpdSfh0wsABpFRQ1NLIAQKdOnTBv3jwAwIoVK/i59OTJk3Hq1CmsWbMGcrkcKpWK90HThV6NiRMnYteuXVCpVFozUlYglUrxzTffGHIdjHpCTc1ZgPKd+e3bt0MgEEChUPCOmO+99x4SEhIQHx8Pd3d3jB07Fnv27NHbnl5juX//PrKyssBxHHbs2EGduCuVSkgkNRs+yng7KS6pOWMJCgrCvXv3+C/6iiek4OBgjBgxAkOHDkVMTAxff0gfeo3l8uXL6Ny5M3x9fbWGX86fPx8SiQT+/v7o0qWLQUMao35QUEw+ghpCXl4edu3ahdLSUkyYMAE2NjbgOA579uyBmZkZhg0bhlmzZsHT0xP+/v545x3yEXjJkiVwdnYGx3G4ePEiH9NSVlbGO2i2b98ehw8fNqhPej/VQqEQrq6uuHz5stZNyZUrV6K4uBiXL1/GsmXLIBQKERQUZFAHGHWbAsXrGcvBgwdx69YtnD17FmFhYTAxMcHvv/+OTz/9FAUFBfDw8EBqaioUCgWKiopQXFwMKyvtZS0qNhxXrFjBy4qKinhPE47jUFxMr39TFb3GMmPGDIjFYrx69YpPd0QjPj4ejx8/hlKp5HMjMxjFSnJTUldNya+++gpfffUVevXqBaFQyFedAwBra2veWExMTLBo0SJwHIewsDB8++23GudITk6Gra0txowZA1NTU9ja/uWh4ePjo5HSS1um1aroNRZHx3L/qi+++EKrzpQpU+Dn54eAgABeXxuCoj8JWfrD61RdO28yXRAt5REAWDQkV3XAaXleppX8VtO/AV9RVrIAoLEZ2YZCi9OgpQl5m3NA75tUTLZrYUd/mxIKyWX6Elt6UFSHxvRcxyYU3zBtqZASSl/fz6uQMrIMHT5Ga6RkBZ06dcLWrVshEokwduxYnDt3Dp999hl/vKSkBOvXr0ejRo2oS78VxtC8eXNwHKcx+vxdVyyjTC7+/e9/G6MZRh1EW5JxfdjY2GjEzVcYiqurK181e+HChVr//034JhoUg3/nzh0AwPvvv0/9Rrh+/To++OADrX8z6i+vO2d5m9FrLHFxcbxrfuVJUmV27NiB1NTyzUKO43D27FlmLAwA0MgEWdvRaywKhQKnT58GAK2uLJaWlnB3d+cdLi211Ahh1D8USmVNd8Fo6DWWRYsW4dq1a+A4Tusz4rfffquxoqCr6isnJov1ePbQUttFSC4JWjaiB1iVKkj3kxLQXUpo+X3zTOhlwG20hGffeZFMyHo70cOgRBT3kcQC+hePnS25uastNDeTUoBJW8y7UksQVm4p+WEu5ehZJLMpuvpQ1KfHsE2bNiE1NRWmpqY4d+4cdQe/6tKboUtxjLqPsj49honFYqxatao6+sKog9SrkeXGjRvIz8+HmVn50GxIfqVjx47hyy/pJRUY9YsSLYkZayN6jaVyWTxD8ytVxOwzGPVmZKkIK64MLaz4yJEjvCFxHIe4OHo2Ekb9g6tBF31jozesWC6X4+nTpzqreVVk0KhAZ0wLZYUr48/LVNXstu3Ifz+zmqIJeHQdScjkWlZvrIUiQqamBF0BgEhLyNMvDx4RstntOlN1syj98LImA7cAYPNDMu/z1dg/qLoNLEnXmHNbN1F1vw6aR5XffkK6H50bM5GqG5uTSZXrpD4tHYeEhKBbt24661hUlEhmMAhK6pGx+Pr64t1339UaUsxg6IKrTxP8mJgYKJVKPiST5qA2ZswY+Pr68itm2vQY9RBlPZngA+WpLseOHQtzc3Otj2ELFizAhQsXkJmZiUGDBlGj1hj1lL85stAiJR88eIA7d+4gOzsb48aNI46/afQaS3Z2NpYuXcr/TdvB/+233yCXy+Hu7q43noWWWcXakb7j39icUnSoXyhVtzCNjInJeA33jMYieg6BBAU97+/9hw8JWaqSHnF3I4PMSHPlKTmxBgBzCbn4YK7lETj66jVC1uzzvlTdc9dvUuW+nuQXW1gSfUGhnys9C48uuCiyj7qCv3RFSrZp0wZlZWWIiIjA1atX8cknn0AgEPDH3zR6jWXkyJGIiYkBx3Fay99lZ2fD3NwcmZmZ2L17NwD2GMbQjq4yeboiJePj49GuXTvcunULHMfBxMRE4/ibRq+xrF69Gk2bNoW3tzdiYmKoOswwGMaGFilpZ2eHyMhIqNVqdO7cGdu2beOPVwd6jcXOzg7+/v7IyclBgwb0sgU//fQT+vfvj6ioKKLMGYPxd9AWKVm5VGPl49WB3iSDffr0gZ2dHX7//XfY25Pu9QDw+PFjAOX5YxmMuorekaVz5/Jd6Yo0mDRevnyJ69ev8z8BsEhJRp3DKAkrunfvjrS0NP6nLofLx5QcTS3smlF18ykBS2ItWVhMxWR05pWXz6m6NmJy5etDB1eq7rF40q0FALjMdEL2y8tEqi4tV/HhDT/Q+9a2PSFTa3E1UsXICFnCn0+outpwk5L5oGVJ9PvW3VlXja+6j1GMZcCAAcZohsF4q/k7idEJUlNTdTpaMhh1AaOMLEePHkV2djZMTU1hZmYGgUCA7777zhhNMxhvDUYZWaZOnYovv/wSlpaWsLKyYjH4jDqJ0dLdOzg4wMHBASkpKcjPz0efPn2oej5ictKe85wskw0AYpsOhIwT0kt7S6zJ9KSf2dIXDppLDPegntGSHn6w3pVMF9tVywT45HMyctS9mz9Vt4UH2efMXHpC9ruZZOGir4YMouo+iKOnvR350YeE7FkWPW7FS1RzhYneBoxiLHPmzIFIJIJIJIKdnZ1GhVgGo65gFGMJDaU7NzIYdQmjzFlmzpzJklQw6jxGGVmWLFmCXbt2ISIigo9lGTSI/uzMYNRWjGLZc9sKAAAJcklEQVQsubm5yMjIgJubG9zcKHVSGIw6gIAzwm7i4sWLMWnSJNjZ0QvpVCYn/jghS713iqrr0eVrQpb7kr5yJhCSK1w2Tm2pusq8FEImsaUHNmUnXqTKGzb7nBSq6OXBISLzKOcknKOqCrSU1aYhsXIgZKky+r109etHlZcqyZU2ZQHpygMAZtbOVHljHzKzTl3EKCPL/PnzjdEMg/FWY5QJPoNRHzCasRQXF6OwkB6vzmDUBYxmLFevXsXSpUtZABijzmI0Y2ncuDEsLCyqLXkAg1HdGGU1DADKyspQVlbGMlcy6ixGMxYGo67DVsMYDANhxsJgGIgwJCQkpLpPumHDBkRGRuLKlStIS0tDq1ateLmTkxPu37+PCxcuwM/Pj5dZWVlh7ty56NGjB6/7008/4fz58/Dy8oKtrS0vu3HjBrKzs/kgtJiYGKxfvx7Xrl1DZmYmfH19sXfvXuzcuRPh4eF4/vw5OnXqxLd79epVdOzYUeN8WVlZWLBgAaKjo9GoUSM4OzsjLS0NixYtwoULFwCA94tLS0vD4sWLcf78eZSUlMDb21vjmuVyOXx8fAAA+/btw/79+zXSj06ZMgW9evUCAHz//ff46KOPIBKJ8OTJEyxatAjh4eGwt7fnXYtu3LiB69evo2XLlhr3+M8//4SNjQ2mT5+O/v37AwCOHz+OzZs3QyaTITMzEz4+Ppg9ezaf7+2HH36AVCrF+vXr4e3tDVNTU8yZM4fP21WfMVrw1+syYcIEuLq6YtasWRg2bBgvT0pKwsWLF7WWEa9MYGAgbG1tsWbNGixatIiXVS2udPv2bXz22Wfo1q0bEhLKg6BGjRqF5s2bIykpCYMHD9bQf/bsGUqrJLQ+duwYxo0bB09PTwQFBcHPzw9Hjx7FxIkT0axZM9y//5cbToW8adOmiI+P17hmqVSK2bNnY+DAgQCAu3fvQiwWo7i4GObm5YFt7777Lh48eABPT0+IxWJe3qJFCwQEBCApKQkffkgGbVUlNTUVFy5c4A2zguHDh/MprgDA1NQUaWlpSE9Ph1wuh4uLCyZNmoSNGzfC3t4e48drll5XqVSIiYlBZmYmOnfuDEtLMrNOXaTGjGXnzp1ISUnBjBkzNOTr1q3jP0gVbN26Vesb0qBBA426hVu3boWVlRV69uzJF1kaOnQofvzxR5w6dQq9evVCs2b0CMoKevTogfPnz2vIUlJS4O7uzuffBYD09HS4uJRHaLZt+5cfWnp6OpycnLB+/XoUFhZi7ty5fN8kEglGjx7Nt+no6IhWrVrh0qVL6N27NwCgd+/eOHDgAFJSUuDvT4+oNIQmTZogKSmJuHeHDh3CpUuX0KFDB3z22WcQi8WIjo5Geno6RCIRhEIhGjVqBDc3N6SlpRFh4mq1GidPnkRRUVG9yg9XY3OWsWPHws3NDQ0bapaLCw0NhUwm06hlGRgYiDlz5lDbkcvlsLa21tANCgrSqEYWGxuLSZMmYcWKFThy5IjevrVr1w4ymUwj/5mZmRmKiooAgN9LcnFxQUpKuVNmxaMYAEilUiQnJ2Pq1KmQy/9yrgwMDERISAhat24NADh58iTkcjnu37+PqKgoXs/JyQlZWVm4efMmunTpore/2ujbty81VGL48OEICgriH60aNWqElJQUSCQSSCR/5VTz9fWFpyfpYCoWi9GzZ08MGDAAVlZk2cO6So2NLAAwceJErF27FitWrOBlEokEQUFBCA4Oxvr163X+/5YtW1BaWopp06bxsoqRRSqV8t/gBQUFCAoKgoWFBd5//32D+ubl5YWHlcpKdOrUCcePH0f79u35NLaDBg3C6tWrIRAI8Mknn/C6gwYNwooVK2BlZQUvLy+t54iJicGmTeU1IGfPno38/Hze8Fu3bo34+HiNfasnT57gwIEDyM/Ph4uLi8aj2OnTp/H06VN4e3vzj5U2NjYG1S1xd3dHbGwsmjRpgidPDEvSVzHHq0+wfZbX4MSJE9i7dy/27dsHCwuy+CmjbsOWjl+Df/zjH2jXrh127NhR011h1ABsZGEwDISNLAyGgTBjYTAMhBkLg2EgzFjeMEqlEjk5OTXeBuN/p94YS8V+xptoa+bMmVp1T58+jdjY2L91noyMDISHh/NtHD16FK9evfpbbUVGRurdt2LopkY3JY3B1atXcerUKcjlcsycORP79u1Dw4YN8fTpU7Ru3RqxsbEICQlBQkICUlJSMHfuXLz33nsoLS3FpEmTsGDBAojFYnh7eyMgIAAA8OrVK6xZswZlZWXo1q0b3n33XWzZsgVFRUWYMmUKEhISIJfLERoaCltbW43+PHnyBLt27UJhYSFGjBiBO3fuQCwW875YGzZs4HMVDB06FHFxcbh06RJKSkowd+5c7Nu3D/n5+TAzM8OwYcOQmpqKrKwsiMViNGzYEAqFAhs3bkR2djY4jkNwcDAGDhwIf39/ZGRkgOYXK5PJkJSUxKJY/0dq/cji5OSE/v37w93dHTExMVAqlRgzZgy8vLzQo0cPdO3aVSO1rI+PDyZPnozU1FRcu3YNH374IRYuXKiRO0AkEmHgwIH44IMPcP36dYSFheG7777D4sWL+c3IM2fOICAgANOmTYNJpVxfDRo0wIABA9CmTRvcvn0bfn5+vOdyBcOGDcOkSZMQFhaGs2fPYtmyZRg1ahQiIyORmpoKNzc39O3bl9ev3IZCoUB6ejrmz58PZ2dnxMbG8o6PBQUF1HvUrl073uuY8fep9cZy8OBB5Obmonnz5ryDo4WFBUxNTSEWizUcH4FyHy+g3NNWrVbz/l+VdX777TfIZDLe5V2lUsHExAQqlQoZGZplHkxMTDSM5aeffkJKSgp8fHzAcRy1vibHcVAqlRAKhRpytVrNe0MvW7aMHwkqt1G1TbVarXFNFTx58gTTp09HWFiY3nvIMIxa/xjWuHFj/P7778jJyeEdFA2lS5cuWLBgAR48eKDhM9aoUSNcvnwZaWlpUKvVGDx4MFauXAmlUonJkycDKPcMXrlyJaysrDSMxcnJCTdv3oRIJIK5uTlcXFxw+PBhDZf4bdu2AQC+/fZbPH78GMHBwSgrK8N3332HnTt3oqioCO+88w7fbkUbHh4eMDc3R+PGjREaGgqBQABfX1/qtbVo0QLr1q17rfvB0A3bwa9mNmzYgH79+hExN4y3H2YsDIaB1Po5C4NRXTBjYTAMhBkLg2EgzFgYDANhxsJgGMj/A1p1NsfifB0jAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 201.6x201.6 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set_fontsize_paper()\n",
    "aaTable2 = aaTable + ['*']\n",
    "for promoter in ['weak', 'strong']:\n",
    "# promoter = 'weak'\n",
    "    plotDf = damDfAApair.loc[promoter][['log10DAMratio_norm']].unstack(level=0)\n",
    "    plotDf.columns = plotDf.columns.droplevel(0)\n",
    "    plotDf = plotDf[aaTable2].loc[aaTable2]\n",
    "    width = 2.8\n",
    "    height = width\n",
    "    fig, ax = plt.subplots(figsize=(width, height))\n",
    "    cbarLabel = 'log10(norm.\\nDAMRatio)'\n",
    "    sns.heatmap(data=plotDf, cmap=cmapName, vmin=-vmaxAApair, vmax=vmaxAApair, square=True,\n",
    "                xticklabels=True, yticklabels=True,\n",
    "                cbar=True, cbar_kws=dict(label=cbarLabel, shrink=0.4,\n",
    "                                         ticks=[-vmaxAApair, -0.15, 0, 0.15, vmaxAApair], aspect=15))\n",
    "    ax.tick_params(axis='both', pad=0)\n",
    "    ticks_font = FontProperties(family='Liberation Mono')\n",
    "    for tick in ax.get_yticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "    ax.set_ylabel('amino acid at position -2')\n",
    "    ax.set_xlabel('amino acid at position -1')\n",
    "    ax.set_title('{} promoter'.format(promoter))\n",
    "    fig.tight_layout()\n",
    "    filename = 'ELM_seq_AApair_promoter_{}'.format(promoter)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.svg'), dpi=800)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.png'), dpi=800)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "metadata": {},
   "outputs": [],
   "source": [
    "AAhydrophobic = ['A','V','I','L','M','F','W','Y']\n",
    "AAhydrophobicNonAromatic = ['A','V','I','L','M']\n",
    "AAaromatic = ['F','W','Y']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 412,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    30.000000\n",
       "mean      0.567478\n",
       "std       0.083743\n",
       "min       0.392018\n",
       "25%       0.513809\n",
       "50%       0.556393\n",
       "75%       0.620797\n",
       "max       0.719277\n",
       "Name: log10DAMratio_norm, dtype: float64"
      ]
     },
     "execution_count": 412,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDfAApair.loc[idx[:, AAhydrophobicNonAromatic, AAaromatic], :]['log10DAMratio_norm'].map(lambda x: 10**x).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    30.000000\n",
       "mean      0.837421\n",
       "std       0.102748\n",
       "min       0.614963\n",
       "25%       0.782413\n",
       "50%       0.855573\n",
       "75%       0.906406\n",
       "max       1.028260\n",
       "Name: log10DAMratio_norm, dtype: float64"
      ]
     },
     "execution_count": 413,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDfAApair.loc[idx[:, AAaromatic, AAhydrophobicNonAromatic], :]['log10DAMratio_norm'].map(lambda x: 10**x).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              4.590017\n",
       "log10DAMratio         0.661814\n",
       "log10DAMratio_norm   -0.331265\n",
       "Name: (strong, M, M), dtype: float64"
      ]
     },
     "execution_count": 414,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDfAApair.loc[damDfAApair.loc[['strong'], 'log10DAMratio_norm'].idxmin()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 415,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'weak'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>log10DAMratio_norm</th>\n",
       "      <th>DAMratio_norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_-1</th>\n",
       "      <th>aa_-2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <th>W</th>\n",
       "      <td>0.545114</td>\n",
       "      <td>-0.263513</td>\n",
       "      <td>-0.406693</td>\n",
       "      <td>0.392018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <th>W</th>\n",
       "      <td>0.596341</td>\n",
       "      <td>-0.224505</td>\n",
       "      <td>-0.367686</td>\n",
       "      <td>0.428859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <th>W</th>\n",
       "      <td>0.604485</td>\n",
       "      <td>-0.218615</td>\n",
       "      <td>-0.361796</td>\n",
       "      <td>0.434715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <th>W</th>\n",
       "      <td>0.661317</td>\n",
       "      <td>-0.179590</td>\n",
       "      <td>-0.322771</td>\n",
       "      <td>0.475586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <th>Y</th>\n",
       "      <td>0.673143</td>\n",
       "      <td>-0.171893</td>\n",
       "      <td>-0.315073</td>\n",
       "      <td>0.484091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             DAMratio  log10DAMratio  log10DAMratio_norm  DAMratio_norm\n",
       "aa_-1 aa_-2                                                            \n",
       "L     W      0.545114      -0.263513           -0.406693       0.392018\n",
       "C     W      0.596341      -0.224505           -0.367686       0.428859\n",
       "M     W      0.604485      -0.218615           -0.361796       0.434715\n",
       "A     W      0.661317      -0.179590           -0.322771       0.475586\n",
       "L     Y      0.673143      -0.171893           -0.315073       0.484091"
      ]
     },
     "execution_count": 416,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = damDfAApair.loc[promoter].sort_values('log10DAMratio_norm')\n",
    "df['DAMratio_norm'] = df['log10DAMratio_norm'].map(lambda x: 10**x)\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DAMratio              1.127299\n",
       "log10DAMratio         0.052039\n",
       "log10DAMratio_norm   -0.091142\n",
       "DAMratio_norm         0.810697\n",
       "Name: P, dtype: float64"
      ]
     },
     "execution_count": 417,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = damDfAApair.loc[promoter].sort_values('log10DAMratio_norm', ascending=False)\n",
    "df['DAMratio_norm'] = df['log10DAMratio_norm'].map(lambda x: 10**x)\n",
    "df.loc['P'].loc['P']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.417977113166153"
      ]
     },
     "execution_count": 418,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['DAMratio_norm'].max()/df['DAMratio_norm'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'strong'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>log10DAMratio_norm</th>\n",
       "      <th>DAMratio_norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_-1</th>\n",
       "      <th>aa_-2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <td>4.590017</td>\n",
       "      <td>0.661814</td>\n",
       "      <td>-0.331265</td>\n",
       "      <td>0.466375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <th>W</th>\n",
       "      <td>4.654340</td>\n",
       "      <td>0.667858</td>\n",
       "      <td>-0.325221</td>\n",
       "      <td>0.472911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <th>W</th>\n",
       "      <td>5.050952</td>\n",
       "      <td>0.703373</td>\n",
       "      <td>-0.289706</td>\n",
       "      <td>0.513209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <th>W</th>\n",
       "      <td>5.072931</td>\n",
       "      <td>0.705259</td>\n",
       "      <td>-0.287820</td>\n",
       "      <td>0.515442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <th>W</th>\n",
       "      <td>5.074589</td>\n",
       "      <td>0.705401</td>\n",
       "      <td>-0.287678</td>\n",
       "      <td>0.515610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <th>W</th>\n",
       "      <td>5.343075</td>\n",
       "      <td>0.727791</td>\n",
       "      <td>-0.265288</td>\n",
       "      <td>0.542890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <th>Y</th>\n",
       "      <td>5.524534</td>\n",
       "      <td>0.742296</td>\n",
       "      <td>-0.250783</td>\n",
       "      <td>0.561328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <th>F</th>\n",
       "      <td>5.704757</td>\n",
       "      <td>0.756237</td>\n",
       "      <td>-0.236842</td>\n",
       "      <td>0.579640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <th>F</th>\n",
       "      <td>5.728392</td>\n",
       "      <td>0.758033</td>\n",
       "      <td>-0.235046</td>\n",
       "      <td>0.582041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <th>Y</th>\n",
       "      <td>5.823906</td>\n",
       "      <td>0.765214</td>\n",
       "      <td>-0.227865</td>\n",
       "      <td>0.591746</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             DAMratio  log10DAMratio  log10DAMratio_norm  DAMratio_norm\n",
       "aa_-1 aa_-2                                                            \n",
       "M     M      4.590017       0.661814           -0.331265       0.466375\n",
       "L     W      4.654340       0.667858           -0.325221       0.472911\n",
       "A     W      5.050952       0.703373           -0.289706       0.513209\n",
       "F     W      5.072931       0.705259           -0.287820       0.515442\n",
       "M     W      5.074589       0.705401           -0.287678       0.515610\n",
       "C     W      5.343075       0.727791           -0.265288       0.542890\n",
       "L     Y      5.524534       0.742296           -0.250783       0.561328\n",
       "C     F      5.704757       0.756237           -0.236842       0.579640\n",
       "L     F      5.728392       0.758033           -0.235046       0.582041\n",
       "C     Y      5.823906       0.765214           -0.227865       0.591746"
      ]
     },
     "execution_count": 420,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = damDfAApair.loc[promoter].sort_values('log10DAMratio_norm')\n",
    "df['DAMratio_norm'] = df['log10DAMratio_norm'].map(lambda x: 10**x)\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>log10DAMratio_norm</th>\n",
       "      <th>DAMratio_norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_-1</th>\n",
       "      <th>aa_-2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <th>N</th>\n",
       "      <td>16.127914</td>\n",
       "      <td>1.207578</td>\n",
       "      <td>0.214499</td>\n",
       "      <td>1.638699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <th>D</th>\n",
       "      <td>14.891839</td>\n",
       "      <td>1.172948</td>\n",
       "      <td>0.179869</td>\n",
       "      <td>1.513106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <th>Q</th>\n",
       "      <td>14.861378</td>\n",
       "      <td>1.172059</td>\n",
       "      <td>0.178980</td>\n",
       "      <td>1.510011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <th>D</th>\n",
       "      <td>14.667739</td>\n",
       "      <td>1.166363</td>\n",
       "      <td>0.173284</td>\n",
       "      <td>1.490336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <th>K</th>\n",
       "      <td>14.513302</td>\n",
       "      <td>1.161766</td>\n",
       "      <td>0.168687</td>\n",
       "      <td>1.474644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <th>K</th>\n",
       "      <td>14.469443</td>\n",
       "      <td>1.160452</td>\n",
       "      <td>0.167373</td>\n",
       "      <td>1.470187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <th>R</th>\n",
       "      <td>14.333364</td>\n",
       "      <td>1.156348</td>\n",
       "      <td>0.163269</td>\n",
       "      <td>1.456361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <th>K</th>\n",
       "      <td>14.322758</td>\n",
       "      <td>1.156027</td>\n",
       "      <td>0.162948</td>\n",
       "      <td>1.455283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <th>K</th>\n",
       "      <td>14.109227</td>\n",
       "      <td>1.149503</td>\n",
       "      <td>0.156424</td>\n",
       "      <td>1.433587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <td>14.096574</td>\n",
       "      <td>1.149114</td>\n",
       "      <td>0.156034</td>\n",
       "      <td>1.432302</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              DAMratio  log10DAMratio  log10DAMratio_norm  DAMratio_norm\n",
       "aa_-1 aa_-2                                                             \n",
       "K     N      16.127914       1.207578            0.214499       1.638699\n",
       "E     D      14.891839       1.172948            0.179869       1.513106\n",
       "K     Q      14.861378       1.172059            0.178980       1.510011\n",
       "N     D      14.667739       1.166363            0.173284       1.490336\n",
       "H     K      14.513302       1.161766            0.168687       1.474644\n",
       "*     K      14.469443       1.160452            0.167373       1.470187\n",
       "K     R      14.333364       1.156348            0.163269       1.456361\n",
       "M     K      14.322758       1.156027            0.162948       1.455283\n",
       "P     K      14.109227       1.149503            0.156424       1.433587\n",
       "D     E      14.096574       1.149114            0.156034       1.432302"
      ]
     },
     "execution_count": 421,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = damDfAApair.loc[promoter].sort_values('log10DAMratio_norm', ascending=False)\n",
    "df['DAMratio_norm'] = df['log10DAMratio_norm'].map(lambda x: 10**x)\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.513694000445431"
      ]
     },
     "execution_count": 422,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['DAMratio_norm'].max()/df['DAMratio_norm'].min()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Epistasis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDfAA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAA2 = damDfAA\n",
    "damDfAA2 = damDfAA2.set_index('position', append=True)\n",
    "damDfAA2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aaMin2 = 'K'\n",
    "aaMin1 = 'R'\n",
    "promoter = 'weak'\n",
    "(damDfAA2.loc[(promoter, aaMin1, -1), 'log10DAMratio_norm'] +\n",
    " damDfAA2.loc[(promoter, aaMin2, -2), 'log10DAMratio_norm'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApairExpected = damDfAApair.copy().drop(['DAMratio', 'log10DAMratio'], axis=1)\n",
    "damDfAApairExpected.loc[:, :] = np.nan\n",
    "damDfAApairExpected.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApairExpected.index.unique(level=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for promoter in ['weak', 'strong']:\n",
    "    for aaMin1 in damDfAApairExpected.index.unique(level='aa_-1'):\n",
    "        for aaMin2 in damDfAApairExpected.index.unique(level='aa_-2'):\n",
    "            damDfAApairExpected.loc[(promoter, aaMin1, aaMin2), 'log10DAMratio_norm'] = \\\n",
    "                (damDfAA2.loc[(promoter, aaMin1, -1), 'log10DAMratio_norm'] +\n",
    "                 damDfAA2.loc[(promoter, aaMin2, -2), 'log10DAMratio_norm'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApairExpected.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApair.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApairExpected['Delta_log10DAMratio_norm'] = damDfAApair['log10DAMratio_norm'] - damDfAApairExpected['log10DAMratio_norm']\n",
    "damDfAApairExpected.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApairExpected['Delta_log10DAMratio_norm'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApairExpected['Delta_log10DAMratio_norm'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vmaxAApairEpistatis = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the epistatis, we drop the stop codon which is confusing to interprete."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "aaTable2 = aaTable + ['*']\n",
    "aaTable2 = aaTable\n",
    "for promoter in ['weak', 'strong']:\n",
    "\n",
    "    plotDf = damDfAApairExpected.loc[promoter][['Delta_log10DAMratio_norm']].unstack(level=0)\n",
    "    plotDf.columns = plotDf.columns.droplevel(0)\n",
    "    plotDf = plotDf[aaTable2].loc[aaTable2]\n",
    "    width = 2.8\n",
    "    height = width\n",
    "    fig, ax = plt.subplots(figsize=(width, height))\n",
    "    cbarLabel = '$\\Delta$ log10(norm.\\nDAMRatio)'\n",
    "    sns.heatmap(data=plotDf, cmap=cmapName, vmin=-vmaxAApairEpistatis, vmax=vmaxAApairEpistatis, square=True,\n",
    "                xticklabels=True, yticklabels=True,\n",
    "                cbar=True, cbar_kws=dict(label=cbarLabel, shrink=0.4,\n",
    "                                         ticks=[-vmaxAApair, -0.15, 0, 0.15, vmaxAApair], aspect=15))\n",
    "    ax.tick_params(axis='both', pad=0)\n",
    "    ticks_font = FontProperties(family='Liberation Mono')\n",
    "    for tick in ax.get_yticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "    ax.set_ylabel('amino acid at position -2')\n",
    "    ax.set_xlabel('amino acid at position -1')\n",
    "    ax.set_title('{} promoter'.format(promoter))\n",
    "    fig.tight_layout()\n",
    "    filename = 'ELM_seq_AApairEpistatis_promoter_{}'.format(promoter)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.svg'), dpi=800)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.png'), dpi=800)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDfAApairExpected.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = 10**damDfAApairExpected.loc[idx[:, ['A','V','I','L','M','F'], ['M','F','W','Y']], 'Delta_log10DAMratio_norm']\n",
    "df = df.loc['weak']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = 10**damDfAApairExpected\n",
    "df.loc[idx[:, 'K', 'K'], :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correlation weak - strong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = damDfAApair\n",
    "# df = df[['log10DAMratio_norm']].unstack('promoter')\n",
    "df = df[['DAMratio']].unstack('promoter')\n",
    "# df.columns = df.columns.droplevel(0)\n",
    "df.columns = df.columns.map('{0[0]} {0[1]} promoter'.format)\n",
    "xlabel = df.columns[0]\n",
    "ylabel = df.columns[1]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ci : int in [0, 100] or None, optional\n",
    "\n",
    "Size of the confidence interval for the regression estimate. This will be drawn using translucent bands around the regression line. The confidence interval is estimated using a bootstrap; for large datasets, it may be advisable to avoid that computation by setting this parameter to None.\n",
    "\n",
    "n_boot : int, optional\n",
    "\n",
    "Number of bootstrap resamples used to estimate the ci. The default value attempts to balance time and stability; you may want to increase this value for “final” versions of plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "width = 1.5\n",
    "fig, ax = plt.subplots(figsize=(width, width))\n",
    "pearson = df[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.06, 0.93),\n",
    "            xycoords='axes fraction', ha='left', va='top', fontsize='medium',\n",
    "            bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "plt.tight_layout()\n",
    "sns.regplot(ax=ax, data=df, x=xlabel, y=ylabel, ci=95, n_boot=1000, fit_reg=True,\n",
    "            scatter_kws=dict(alpha=0.4, s=20, edgecolors='none'))\n",
    "ax.set_title('amino acid pair')\n",
    "ax.set_xlabel('DAMRatio strong promoter')\n",
    "ax.set_ylabel('DAMRatio nweak promoter')\n",
    "filename = 'ELM_seq_AApair_correlation_weak_strong'\n",
    "fig.savefig(ctermELMseqPath / subFolder / (filename + '.svg'), dpi=800, bbox_inches='tight')\n",
    "fig.savefig(ctermELMseqPath / subFolder / (filename + '.png'), dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relative differences min and max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApair.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'weak'\n",
    "df = damDfAApair.loc[promoter, 'DAMratio']\n",
    "mean = df.mean()\n",
    "print(df.idxmax())\n",
    "print(damDfAApair.loc[promoter].loc[df.idxmax()])\n",
    "print((df.max() - mean)/mean)\n",
    "\n",
    "print(df.idxmin())\n",
    "print(damDfAApair.loc[promoter].loc[df.idxmin()])\n",
    "print((df.min() - mean)/mean)\n",
    "\n",
    "print((df.max() - df.min())/df.min())\n",
    "print(df.max()/df.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'strong'\n",
    "df = damDfAApair.loc[promoter, 'DAMratio']\n",
    "mean = df.mean()\n",
    "print(df.idxmax())\n",
    "print(damDfAApair.loc[promoter].loc[df.idxmax()])\n",
    "print((df.max() - mean)/mean)\n",
    "\n",
    "print(df.idxmin())\n",
    "print(damDfAApair.loc[promoter].loc[df.idxmin()])\n",
    "print((df.min() - mean)/mean)\n",
    "\n",
    "print((df.max() - df.min())/df.min())\n",
    "print(df.max()/df.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfAApair.loc[('strong', 'K', 'K'), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = 10**damDfAApair\n",
    "df.loc[idx[:, 'K', 'K'], :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = 10**damDfAApair\n",
    "df.loc[idx[:, 'K', 'N'], :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = 10**damDfAApair\n",
    "df.loc[idx['strong', 'K', :], :].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Codon pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDfGroup = damDf3.groupby(['promoter', 'codon_-1', 'codon_-2'])[['DAMratio']].mean()\n",
    "damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "damDfGroup;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meanDAM = damDfGroup['log10DAMratio'].groupby(level=0).mean()\n",
    "print(meanDAM)\n",
    "damDfGroup['log10DAMratio_norm'] = damDfGroup['log10DAMratio'] - meanDAM\n",
    "damDfGroup;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpair = damDfGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpair.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### looking for patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpair.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpair.min()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "List the best and worst hexamers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodonpair.loc['weak'].sort_values('log10DAMratio_norm')\n",
    "df['seq'] = df.index.map(lambda x: 'AAAAAA-' + ''.join(x[::-1]) + '-TAACAG')\n",
    "df['seq2'] = df.index.map(lambda x: 'AAAAAA' + ''.join(x[::-1]) + 'TAACAG')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "worst weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "best weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy()\n",
    "df2.loc[df2['seq2'].str.contains('TATAAC'), 'has_motif'] = True\n",
    "df2.loc[~df2['seq2'].str.contains('TATAAC'), 'has_motif'] = False\n",
    "sns.boxplot(data=df2, x='has_motif', y='log10DAMratio_norm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy()\n",
    "df2.loc[df2['seq2'].str.contains('TATAA.')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy()\n",
    "df2.loc[df2['seq'].str.contains('GATC'), 'has_GATC'] = True\n",
    "df2.loc[~df2['seq'].str.contains('GATC'), 'has_GATC'] = False\n",
    "sns.boxplot(data=df2, x='has_GATC', y='log10DAMratio_norm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodonpair.loc['strong'].sort_values('log10DAMratio_norm')\n",
    "df['seq'] = df.index.map(lambda x: 'AAAAAA-' + ''.join(x[::-1]) + '-TAACAG')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "worst strong promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "best strong promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*effect of GATC motif*\n",
    "\n",
    "In the strong promoter:\n",
    "\n",
    "The hexamers `D/GAT-CNN`, where -1 codon can be `L/CTN, P/CCN, H/CAN, Q/CAN, R/CGN`, do not show particularly higher or lower levels. Only higher for 3 hexamers `D/GAT-P/CCT,CCA,CCG`, but not the others.\n",
    "\n",
    "the 4 hexamers `R/AGA-S/TCN` show higher expression.\n",
    "\n",
    "the expression of hexamers `X/NNN-I/ATC` are slightly higher when the G at 3rd position in `X/NNG`.\n",
    "\n",
    "In the weak promoter:\n",
    "\n",
    "the 4 hexamers `D/GAT-R/CGN` and the isolated hexamer `D/GAT-P/CCC` show *lower* expression.\n",
    "\n",
    "the 4 hexamers `R/AGA-S/TCN` show slightly lower expression.\n",
    "\n",
    "the expression of hexamers `X/NNN-I/ATC` do not seem to depend on the G at 3rd position in `X/NNG`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy()\n",
    "df2.loc[df2['seq'].str.contains('GATC'), 'has_GATC'] = True\n",
    "df2.loc[~df2['seq'].str.contains('GATC'), 'has_GATC'] = False\n",
    "sns.boxplot(data=df2, x='has_GATC', y='log10DAMratio_norm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the weak promoter, we have\n",
    "\n",
    "`R/CGC-G/GGC` with very high expression, and `R/CGC-G/GGA` with very low expression, with only 1 nucleotide difference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(damDf3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# damDf4[(damDf4['codon_-2'] == 'CGC')* (damDf4['codon_-1'] == 'GGA')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'weak'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodonpair.loc[promoter].sort_values('log10DAMratio_norm')\n",
    "df['seq_extended'] = df.index.map(lambda x: 'AAAAAA-' + ''.join(x[::-1]) + '-TAACAG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3B = damDf3[damDf3['promoter'] == promoter].set_index(['codon_-2', 'codon_-1']).sort_index()\n",
    "damDf3B.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "worst weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "best weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We look if the different randomized sequences for upstream region present large variation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[0].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[-1].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "df2 = df.copy()\n",
    "df2.loc[df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = True\n",
    "df2.loc[~df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = False\n",
    "width = 2\n",
    "aspect = 0.5\n",
    "fig, ax = plt.subplots(figsize=(width, width/aspect))\n",
    "sns.boxplot(ax=ax, data=df2, x='has_GATC', y='log10DAMratio_norm')\n",
    "ax.set_title(promoter + ' promoter')\n",
    "fig.savefig(ctermELMseqPath / subFolder / 'ELM_seq_codon_pair_GATC_boxplot_promoter_{}.png'.format(promoter),\n",
    "            dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### strong promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'strong'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodonpair.loc[promoter].sort_values('log10DAMratio_norm')\n",
    "df['seq_extended'] = df.index.map(lambda x: 'AAAAAA-' + ''.join(x[::-1]) + '-TAACAG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3B = damDf3[damDf3['promoter'] == promoter].set_index(['codon_-1', 'codon_-2']).sort_index()\n",
    "damDf3B.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "worst weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "best weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We look if the different randomized sequences for upstream region present large variation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[0].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[-1].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "df2 = df.copy()\n",
    "df2.loc[df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = True\n",
    "df2.loc[~df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = False\n",
    "width = 2\n",
    "aspect = 0.5\n",
    "fig, ax = plt.subplots(figsize=(width, width/aspect))\n",
    "sns.boxplot(ax=ax, data=df2, x='has_GATC', y='log10DAMratio_norm')\n",
    "ax.set_title(promoter + ' promoter')\n",
    "fig.savefig(ctermELMseqPath / subFolder / 'ELM_seq_codon_pair_GATC_boxplot_promoter_{}.png'.format(promoter),\n",
    "            dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vmax = 0.6\n",
    "\n",
    "# for promoter in ['weak', 'strong']:\n",
    "\n",
    "#     subseq_type='codon'\n",
    "#     plotSynonymousCodonGroupSeparationLine=True\n",
    "#     plotTitle = '{} promoter'.format(promoter)\n",
    "\n",
    "#     plotData = damDfCodonpair.loc[promoter, ['log10DAMratio_norm']].unstack(level=0).copy()\n",
    "#     plotData.columns = plotData.columns.droplevel(0)\n",
    "#     nameIndex = plotData.index.name\n",
    "#     nameCol = plotData.columns.name\n",
    "\n",
    "#     plotData = sort_codon_index(plotData.T, refCodonTableDf,\n",
    "#                                 addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False)\n",
    "#     plotData = sort_codon_index(plotData.T, refCodonTableDf,\n",
    "#                                 addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False)\n",
    "#     plotData.index.name = nameIndex\n",
    "#     plotData.columns.name = nameCol\n",
    "#     linewidth1 = 0.7\n",
    "#     aspectratio = 1\n",
    "#     imagesize = 5.2\n",
    "#     figsize = (imagesize, imagesize/aspectratio)\n",
    "#     plotData.columns.name = re.sub(r'_', r' ', plotData.columns.name)\n",
    "#     plotData.index.name = re.sub(r'_', r' ', plotData.index.name)\n",
    "\n",
    "#     fig, (ax,cbar_ax) = plt.subplots(1, 2, figsize=figsize)\n",
    "#     # main axes\n",
    "#     ax.set_position([0, 0, 1, 1])\n",
    "#     ax.set_facecolor(colorBackground1)\n",
    "\n",
    "#     # Create axes for the colorbar for more precise positioning\n",
    "#     cbar_ax.set_position([1.1, 0.45, 0.12/15, 0.12])\n",
    "\n",
    "#     cbarLabel = 'log10(normalized DAMRatio)'\n",
    "#     ax = seaborn.heatmap(plotData, square=True, ax=ax, cmap=cmapName,\n",
    "#                          cbar_ax=cbar_ax,\n",
    "#                          cbar_kws=dict(label=cbarLabel, ticks=[-vmax, 0, vmax], aspect=15),\n",
    "#                          xticklabels=True, yticklabels=True,\n",
    "#                          vmin=-vmax, vmax=vmax,\n",
    "#                          linewidth=linewidth1, linecolor=colorHeatmapLine)\n",
    "#     for _, spine in ax.spines.items():\n",
    "#         spine.set_visible(True)\n",
    "#         spine.set_color('0.3')\n",
    "#     ax.tick_params(axis='y', which='both', labelleft=True, labelright=True, labelsize='small', pad=1)\n",
    "#     ax.tick_params(axis='x', which='both', labeltop=True, labelbottom=True, labelsize='small', pad=1)\n",
    "#     if len(ax.xaxis.get_ticklabels()[0].get_text()) > 2:\n",
    "#         ax.xaxis.set_tick_params(rotation=90)\n",
    "#     ax.yaxis.set_tick_params(rotation=0)\n",
    "#     family = 'Liberation Mono'\n",
    "#     ticks_font = FontProperties(family=family, size=VERY_SMALL_SIZE)\n",
    "#     for tick in ax.get_xticklabels():\n",
    "#         tick.set_fontproperties(ticks_font)\n",
    "#         tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "#     for tick in ax.get_yticklabels():\n",
    "#         tick.set_fontproperties(ticks_font)\n",
    "#         tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "\n",
    "#     ax.set_xlabel('codon at position -1')\n",
    "#     ax.set_ylabel('codon at position -2')\n",
    "#     ax.set_title(plotTitle, y=1.05)\n",
    "\n",
    "#     # fig.tight_layout()\n",
    "\n",
    "#     if subseq_type == 'codon' and plotSynonymousCodonGroupSeparationLine:\n",
    "#         synCodonLineWidth = 0.7*linewidth1\n",
    "#         synCodonLineColor = '0.3'\n",
    "#         synCodonLineAlpha = 1\n",
    "#         synCodonLineStyle = '--'\n",
    "#         refCodonTableDf.sort_values('aa_groups_sorted_index')\n",
    "#         # Finding last rows of grouped dataframe on a multiindex column\n",
    "#         groupCol = 'aa'\n",
    "#         df = refCodonTableDf\n",
    "#         colDf = pd.DataFrame(df[groupCol].copy()).reset_index(drop=True).reset_index()\n",
    "#         lastS = colDf.groupby(by=groupCol).last().sort_values('index')['index']\n",
    "#         for aa, i in lastS[:-1].iteritems():\n",
    "#             ax.axvline(i + 1, ls=synCodonLineStyle, lw=synCodonLineWidth, c=synCodonLineColor, alpha=synCodonLineAlpha)\n",
    "#             ax.axhline(i + 1, ls=synCodonLineStyle, lw=synCodonLineWidth, c=synCodonLineColor, alpha=synCodonLineAlpha)\n",
    "\n",
    "#     # fig.tight_layout()\n",
    "#     filename = 'ELM_seq_codon_pair_promoter_{}.png'.format(promoter)\n",
    "#     fig.savefig(ctermELMseqPath / filename, dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "vmaxCodonPair = 0.6\n",
    "displayType = 'paper'\n",
    "if displayType == 'screen':\n",
    "    imageWidth = 10\n",
    "    set_fontsize_screen()\n",
    "if displayType == 'paper':\n",
    "    imageWidth = 5.2\n",
    "    set_fontsize_paper()\n",
    "highlightGATC = False\n",
    "\n",
    "for promoter in ['weak', 'strong'][:]:\n",
    "\n",
    "    subseq_type='codon'\n",
    "    plotSynonymousCodonGroupSeparationLine=True\n",
    "    plotTitle = '{} promoter'.format(promoter)\n",
    "\n",
    "    plotData = damDfCodonpair.loc[promoter, ['log10DAMratio_norm']].unstack(level=0).copy()\n",
    "    plotData.columns = plotData.columns.droplevel(0)\n",
    "    \n",
    "    nameIndex = plotData.index.name\n",
    "    nameCol = plotData.columns.name\n",
    "    plotData = sort_codon_index(plotData.T, refCodonTableDf,\n",
    "                                addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False)\n",
    "    plotData = sort_codon_index(plotData.T, refCodonTableDf,\n",
    "                                addAminoAcidLetterToIndex=True, addAminoAcidLetterAsColumn=False)\n",
    "    plotData.index.name = nameIndex\n",
    "    plotData.columns.name = nameCol\n",
    "    linewidth1 = 0.7\n",
    "    aspectratio = 1\n",
    "    figsize = (imageWidth, imageWidth/aspectratio)\n",
    "\n",
    "    fig, (ax,cbar_ax) = plt.subplots(1, 2, figsize=figsize)\n",
    "    # main axes\n",
    "    ax.set_position([0, 0, 1, 1])\n",
    "    ax.set_facecolor(colorBackground1)\n",
    "    # Create axes for the colorbar for more precise positioning\n",
    "    cbar_ax.set_position([1.1, 0.45, 0.12/15, 0.12])\n",
    "\n",
    "    cbarLabel = 'log10(normalized DAMRatio)'\n",
    "    ax = seaborn.heatmap(plotData, square=True, ax=ax, cmap=cmapName,\n",
    "                         cbar_ax=cbar_ax,\n",
    "                         cbar_kws=dict(label=cbarLabel, ticks=[-vmax, 0, vmax], aspect=15),\n",
    "                         xticklabels=True, yticklabels=True,\n",
    "                         vmin=-vmaxCodonPair, vmax=vmaxCodonPair,\n",
    "                         linewidth=linewidth1, linecolor=colorHeatmapLine)\n",
    "    for _, spine in ax.spines.items():\n",
    "        spine.set_visible(True)\n",
    "        spine.set_color('0.3')\n",
    "    ax.tick_params(axis='y', which='both', labelleft=True, labelright=True, labelsize='small', pad=1)\n",
    "    ax.tick_params(axis='x', which='both', labeltop=True, labelbottom=True, labelsize='small', pad=1)\n",
    "    if len(ax.xaxis.get_ticklabels()[0].get_text()) > 2:\n",
    "        ax.xaxis.set_tick_params(rotation=90)\n",
    "    ax.yaxis.set_tick_params(rotation=0)\n",
    "    family = 'Liberation Mono'\n",
    "    ticks_font = FontProperties(family=family)\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    for tick in ax.get_yticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "\n",
    "    ax.set_xlabel('codon at position -1')\n",
    "    ax.set_ylabel('codon at position -2')\n",
    "    ax.set_title(plotTitle, y=1.05)\n",
    "\n",
    "    if subseq_type == 'codon' and plotSynonymousCodonGroupSeparationLine:\n",
    "        synCodonLineWidth = 0.7*linewidth1\n",
    "        synCodonLineColor = '0.3'\n",
    "        synCodonLineAlpha = 1\n",
    "        synCodonLineStyle = '--'\n",
    "        refCodonTableDf.sort_values('aa_groups_sorted_index')\n",
    "        # Finding last rows of grouped dataframe on a multiindex column\n",
    "        groupCol = 'aa'\n",
    "        df = refCodonTableDf\n",
    "        colDf = pd.DataFrame(df[groupCol].copy()).reset_index(drop=True).reset_index()\n",
    "        lastS = colDf.groupby(by=groupCol).last().sort_values('index')['index']\n",
    "        for aa, i in lastS[:-1].iteritems():\n",
    "            ax.axvline(i + 1, ls=synCodonLineStyle, lw=synCodonLineWidth, c=synCodonLineColor, alpha=synCodonLineAlpha)\n",
    "            ax.axhline(i + 1, ls=synCodonLineStyle, lw=synCodonLineWidth, c=synCodonLineColor, alpha=synCodonLineAlpha)\n",
    "            \n",
    "    if highlightGATC:\n",
    "        # Highlight hexamers with GATC motif\n",
    "        if plotData.index.name == 'codon_-2':\n",
    "            index = plotData.index.map(lambda x: x[:] if len(x) == 3 else x[2:])\n",
    "            columns = plotData.columns.map(lambda x: x[:] if len(x) == 3 else x[2:])\n",
    "            for i, trimer1 in enumerate(index):\n",
    "                for j, trimer2 in enumerate(columns):\n",
    "                    if re.search('GATC', trimer1 + trimer2) or motifWithSpacer:\n",
    "                        # draw rectangle\n",
    "                        ax.add_patch(Rectangle((j, i), 1, 1, ec='red', fc='none', alpha=0.6, lw=linewidth1, ls='--'))\n",
    "\n",
    "    # fig.tight_layout()\n",
    "    filename = 'ELM_seq_codon_pair_promoter_{}{}.png'.format(promoter, '_GATC' if highlightGATC else '')\n",
    "    if displayType == 'paper':\n",
    "        fig.savefig(ctermELMseqPath / subFolder / filename, dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correlation weak - strong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = damDfCodonpair.copy()\n",
    "# df = df[['log10DAMratio_norm']].unstack('promoter')\n",
    "df = df[['DAMratio']].unstack('promoter')\n",
    "# df.columns = df.columns.droplevel(0)\n",
    "df.columns = df.columns.map('{0[0]} {0[1]} promoter'.format)\n",
    "xlabel = df.columns[0]\n",
    "ylabel = df.columns[1]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_paper()\n",
    "width = 1.5\n",
    "fig, ax = plt.subplots(figsize=(width, width))\n",
    "pearson = df[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.95, 0.97),\n",
    "            xycoords='axes fraction', ha='right', va='top', fontsize='medium',\n",
    "            bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "plt.tight_layout()\n",
    "sns.regplot(ax=ax, data=df, x=xlabel, y=ylabel, ci=95, n_boot=1000, fit_reg=True,\n",
    "            scatter_kws=dict(alpha=0.3, s=20, edgecolors='none'))\n",
    "ax.set_title('codon pair')\n",
    "ax.set_xlabel('DAMRatio strong promoter')\n",
    "ax.set_ylabel('DAMRatio nweak promoter')\n",
    "filename = 'ELM_seq_codon_pair_correlation_weak_strong.png'\n",
    "fig.savefig(ctermELMseqPath / subFolder / filename, dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hexamers downstream"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDfGroup = damDf3.groupby(['promoter', 'codon_1', 'codon_2'])[['DAMratio']].mean()\n",
    "damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "damDfGroup;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meanDAM = damDfGroup['log10DAMratio'].groupby(level=0).mean()\n",
    "print(meanDAM)\n",
    "damDfGroup['log10DAMratio_norm'] = damDfGroup['log10DAMratio'] - meanDAM\n",
    "damDfGroup;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpairDown = damDfGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpairDown.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### looking for patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpairDown.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpairDown.min()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "List the best and worst hexamers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqSpacer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'weak'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodonpairDown.loc[promoter].sort_values('log10DAMratio_norm')\n",
    "df['seq_extended'] = df.index.map(lambda x: 'TAA' + ''.join(x[::+1]) + 'CAGGCC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3B = damDf3[damDf3['promoter'] == promoter].set_index(['codon_1', 'codon_2']).sort_index()\n",
    "damDf3B.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "fig, ax = plt.subplots(figsize=(5,3))\n",
    "ax.set_title(promoter + ' promoter')\n",
    "\n",
    "sns.distplot(df['log10DAMratio_norm'], ax=ax, label='downstream')\n",
    "sns.distplot(damDfCodonpair.loc[promoter, 'log10DAMratio_norm'], ax=ax, color=sns.color_palette()[1], label='upstream')\n",
    "plt.legend()\n",
    "ax.set_xlim(-0.6, 0.6)\n",
    "fig.savefig(ctermELMseqPath / subFolder / 'ELM_seq_codon_pair_downstream_distribution_promoter_{}.png'.format(promoter),\n",
    "            dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "worst weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "best weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We look if the different randomized sequences for upstream region present large variation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[3].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[-1].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "df2 = df.copy()\n",
    "df2.loc[df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = True\n",
    "df2.loc[~df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = False\n",
    "width = 2\n",
    "aspect = 0.5\n",
    "fig, ax = plt.subplots(figsize=(width, width/aspect))\n",
    "sns.boxplot(ax=ax, data=df2, x='has_GATC', y='log10DAMratio_norm')\n",
    "ax.set_title(promoter + ' promoter')\n",
    "fig.savefig(ctermELMseqPath / subFolder / 'ELM_seq_codon_pair_downstream_GATC_boxplot_promoter_{}.png'.format(promoter),\n",
    "            dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### strong promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter = 'strong'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodonpairDown.loc[promoter].sort_values('log10DAMratio_norm')\n",
    "df['seq_extended'] = df.index.map(lambda x: 'TAA' + ''.join(x[::+1]) + 'CAGGCC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3B = damDf3[damDf3['promoter'] == promoter].set_index(['codon_1', 'codon_2']).sort_index()\n",
    "damDf3B.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "fig, ax = plt.subplots(figsize=(5,3))\n",
    "ax.set_title(promoter + ' promoter')\n",
    "\n",
    "sns.distplot(df['log10DAMratio_norm'], ax=ax, label='downstream')\n",
    "sns.distplot(damDfCodonpair.loc[promoter, 'log10DAMratio_norm'], ax=ax, color=sns.color_palette()[1], label='upstream')\n",
    "plt.legend()\n",
    "ax.set_xlim(-0.6, 0.6)\n",
    "fig.savefig(ctermELMseqPath / subFolder / 'ELM_seq_codon_pair_downstream_distribution_promoter_{}.png'.format(promoter),\n",
    "            dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "worst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "best weak promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We look if the different randomized sequences for upstream region present large variation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[3].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "damDf3B.loc[df.iloc[-1].name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_fontsize_screen()\n",
    "df2 = df.copy()\n",
    "df2.loc[df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = True\n",
    "df2.loc[~df2['seq_extended'].str.contains('GATC'), 'has_GATC'] = False\n",
    "width = 2\n",
    "aspect = 0.5\n",
    "fig, ax = plt.subplots(figsize=(width, width/aspect))\n",
    "sns.boxplot(ax=ax, data=df2, x='has_GATC', y='log10DAMratio_norm')\n",
    "ax.set_title(promoter + ' promoter')\n",
    "fig.savefig(ctermELMseqPath / subFolder / 'ELM_seq_codon_pair_downstream_GATC_boxplot_promoter_{}.png'.format(promoter),\n",
    "            dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfCodonpairDown.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Detect GATC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hexamerList = damDfCodonpairDown.index.map(lambda x: ''.join(x[1:3])).unique()\n",
    "hexamerHasGATCMask = [re.search('GATC', h) is not None for h in hexamerList]\n",
    "damDfCodonpairDown.loc['strong'].iloc[hexamerHasGATCMask].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vmaxCodonPair = 0.6\n",
    "displayType = 'paper'\n",
    "if displayType == 'screen':\n",
    "    imageWidth = 10\n",
    "    set_fontsize_screen()\n",
    "if displayType == 'paper':\n",
    "    imageWidth = 5.2\n",
    "    set_fontsize_paper()\n",
    "highlightGATC = False\n",
    "\n",
    "for promoter in ['weak', 'strong'][:]:\n",
    "\n",
    "    subseq_type='codon'\n",
    "    plotSynonymousCodonGroupSeparationLine=True\n",
    "    plotTitle = '{} promoter'.format(promoter)\n",
    "\n",
    "    plotData = damDfCodonpairDown.loc[promoter, ['log10DAMratio_norm']].unstack(level=0).copy()\n",
    "    plotData.columns = plotData.columns.droplevel(0)\n",
    "    plotData = plotData.T\n",
    "    \n",
    "    nameIndex = plotData.index.name\n",
    "    nameCol = plotData.columns.name\n",
    "    plotData = sort_codon_index(plotData.T, refCodonTableDf,\n",
    "                                addAminoAcidLetterToIndex=False, addAminoAcidLetterAsColumn=False)\n",
    "    plotData = sort_codon_index(plotData.T, refCodonTableDf,\n",
    "                                addAminoAcidLetterToIndex=False, addAminoAcidLetterAsColumn=False)\n",
    "    plotData.index.name = nameIndex\n",
    "    plotData.columns.name = nameCol\n",
    "    linewidth1 = 0.7\n",
    "    aspectratio = 1\n",
    "    figsize = (imageWidth, imageWidth/aspectratio)\n",
    "#     plotData.columns.name = re.sub(r'_', r' ', plotData.columns.name)\n",
    "#     plotData.index.name = re.sub(r'_', r' ', plotData.index.name)\n",
    "\n",
    "    fig, (ax,cbar_ax) = plt.subplots(1, 2, figsize=figsize)\n",
    "    # main axes\n",
    "    ax.set_position([0, 0, 1, 1])\n",
    "    ax.set_facecolor(colorBackground1)\n",
    "    # Create axes for the colorbar for more precise positioning\n",
    "    cbar_ax.set_position([1.1, 0.45, 0.12/15, 0.12])\n",
    "\n",
    "    cbarLabel = 'log10(normalized DAMRatio)'\n",
    "    ax = seaborn.heatmap(plotData, square=True, ax=ax, cmap=cmapName,\n",
    "                         cbar_ax=cbar_ax,\n",
    "                         cbar_kws=dict(label=cbarLabel, ticks=[-vmax, 0, vmax], aspect=15),\n",
    "                         xticklabels=True, yticklabels=True,\n",
    "                         vmin=-vmaxCodonPair, vmax=vmaxCodonPair,\n",
    "                         linewidth=linewidth1, linecolor=colorHeatmapLine)\n",
    "    for _, spine in ax.spines.items():\n",
    "        spine.set_visible(True)\n",
    "        spine.set_color('0.3')\n",
    "    ax.tick_params(axis='y', which='both', labelleft=True, labelright=True, labelsize='small', pad=1)\n",
    "    ax.tick_params(axis='x', which='both', labeltop=True, labelbottom=True, labelsize='small', pad=1)\n",
    "    if len(ax.xaxis.get_ticklabels()[0].get_text()) > 2:\n",
    "        ax.xaxis.set_tick_params(rotation=90)\n",
    "    ax.yaxis.set_tick_params(rotation=0)\n",
    "    family = 'Liberation Mono'\n",
    "    ticks_font = FontProperties(family=family)\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "    for tick in ax.get_yticklabels():\n",
    "        tick.set_fontproperties(ticks_font)\n",
    "        tick.set_bbox(dict(pad=0, facecolor='none', edgecolor='none'))\n",
    "\n",
    "    ax.set_xlabel('trimer at position +2')\n",
    "    ax.set_ylabel('trimer at position +1')\n",
    "    ax.set_title(plotTitle, y=1.05)\n",
    "\n",
    "    if subseq_type == 'codon' and plotSynonymousCodonGroupSeparationLine:\n",
    "        synCodonLineWidth = 0.7*linewidth1\n",
    "        synCodonLineColor = '0.3'\n",
    "        synCodonLineAlpha = 1\n",
    "        synCodonLineStyle = '--'\n",
    "        refCodonTableDf.sort_values('aa_groups_sorted_index')\n",
    "        # Finding last rows of grouped dataframe on a multiindex column\n",
    "        groupCol = 'aa'\n",
    "        df = refCodonTableDf\n",
    "        colDf = pd.DataFrame(df[groupCol].copy()).reset_index(drop=True).reset_index()\n",
    "        lastS = colDf.groupby(by=groupCol).last().sort_values('index')['index']\n",
    "        for aa, i in lastS[:-1].iteritems():\n",
    "            ax.axvline(i + 1, ls=synCodonLineStyle, lw=synCodonLineWidth, c=synCodonLineColor, alpha=synCodonLineAlpha)\n",
    "            ax.axhline(i + 1, ls=synCodonLineStyle, lw=synCodonLineWidth, c=synCodonLineColor, alpha=synCodonLineAlpha)\n",
    "            \n",
    "    if highlightGATC:\n",
    "        # Highlight hexamers with GATC motif\n",
    "        if plotData.index.name == 'codon_1':\n",
    "            index = plotData.index.map(lambda x: x[:] if len(x) == 3 else x[2:])\n",
    "            columns = plotData.columns.map(lambda x: x[:] if len(x) == 3 else x[2:])\n",
    "            for i, trimer1 in enumerate(index):\n",
    "                for j, trimer2 in enumerate(columns):\n",
    "                    # Also add the case of the GAT at position +2, with the first nucleotide of the spacer is C\n",
    "                    motifWithSpacer = trimer2 == 'GAT'\n",
    "                    if re.search('GATC', trimer1 + trimer2) or motifWithSpacer:\n",
    "                        # draw rectangle\n",
    "                        ax.add_patch(Rectangle((j, i), 1, 1, ec='red', fc='none', alpha=0.6, lw=linewidth1, ls='--'))\n",
    "\n",
    "    # fig.tight_layout()\n",
    "    filename = 'ELM_seq_codon_pair_downstream_promoter_{}{}.png'.format(promoter, '_GATC' if highlightGATC else '')\n",
    "    if displayType == 'paper':\n",
    "        fig.savefig(ctermELMseqPath / subFolder / filename, dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Full sequences dodecamers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "damDfGroup = damDf3.groupby(['promoter', 'seq'])[['DAMratio']].mean()\n",
    "damDfGroup['log10DAMratio'] = log10(damDfGroup['DAMratio'])\n",
    "damDfGroup;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meanDAM = damDfGroup['log10DAMratio'].groupby(level=0).mean()\n",
    "print(meanDAM)\n",
    "damDfGroup['log10DAMratio_norm'] = damDfGroup['log10DAMratio'] - meanDAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfDodecamer = damDfGroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDfDodecamer.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correlation weak - strong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for obs in ['DAMratio', 'log10DAMratio_norm']:\n",
    "\n",
    "    df = damDfDodecamer.copy()\n",
    "    df = df[[obs]].unstack('promoter')\n",
    "    df.columns = df.columns.map('{0[0]} {0[1]} promoter'.format)\n",
    "    xlabel = df.columns[0]\n",
    "    ylabel = df.columns[1]\n",
    "\n",
    "    set_fontsize_paper()\n",
    "    width = 1.5\n",
    "    fig, ax = plt.subplots(figsize=(width, width))\n",
    "    pearson = df[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "    ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.95, 0.97),\n",
    "                xycoords='axes fraction', ha='right', va='top', fontsize='medium',\n",
    "                bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "    plt.tight_layout()\n",
    "    sns.regplot(ax=ax, data=df, x=xlabel, y=ylabel, ci=95, n_boot=1000, fit_reg=True,\n",
    "                scatter_kws=dict(alpha=0.3, s=20, edgecolors='none'))\n",
    "    ax.set_title('dodecamer')\n",
    "#     ax.set_xlabel('log10(norm. DAMRatio)\\nstrong promoter')\n",
    "#     ax.set_ylabel('log10(norm. DAMRatio)\\nweak promoter')\n",
    "#     ax.set_xlabel('DAMRatio strong promoter')\n",
    "#     ax.set_ylabel('DAMRatio nweak promoter')\n",
    "    filename = 'ELM_seq_dodecamer_correlation_weak_strong_{}.png'.format(obs)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / filename, dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Folding energy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    GACGGATAAATTAAA\n",
       "1    CACCGGTAAAGGTGT\n",
       "2    GGTTTCTAATCGCGT\n",
       "3    CGGACGTAACGGGCA\n",
       "4    ACTAGTTAATGAAGT\n",
       "Name: seq, dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damDf3['seq'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqUp[-10:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "6 + 3 + 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(seqSpacer[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(seqSpacer + seqGATC[:1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqUp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Seq.Seq(seqUp[-3*30:], alphabet=Bio.Alphabet.IUPAC.ambiguous_dna).translate(table=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(seqSpacer + seqGATC))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\n"
     ]
    }
   ],
   "source": [
    "print(len(seqSpacer))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "15+3+6+30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# damDf4, region length ~65:\n",
    "# damDf3['region_seq'] = damDf3['seq'].map(lambda x: seqUp[-30:] + x + seqSpacer + seqGATC)\n",
    "\n",
    "# damDf4, region length 39:\n",
    "# damDf3['region_seq'] = damDf3['seq'].map(lambda x: seqUp[-12:] + x + seqSpacer[:12])\n",
    "# print(len(damDf3['region_seq'].iloc[0]))\n",
    "\n",
    "# region length 54:\n",
    "damDf3['region_seq'] = damDf3['seq'].map(lambda x: seqUp[-30:] + x + seqSpacer)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TAA*GAT*GAT*CAGGCCTTGAAGATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>seq</th>\n",
       "      <th>Uncut_cnt</th>\n",
       "      <th>DpnI_cnt</th>\n",
       "      <th>MboI_cnt</th>\n",
       "      <th>DpnI_ratio</th>\n",
       "      <th>MboI_ratio</th>\n",
       "      <th>promoter</th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>codon_-2</th>\n",
       "      <th>...</th>\n",
       "      <th>n_4</th>\n",
       "      <th>n_5</th>\n",
       "      <th>n_6</th>\n",
       "      <th>n_7</th>\n",
       "      <th>n_8</th>\n",
       "      <th>seq_extended</th>\n",
       "      <th>contains_GATC</th>\n",
       "      <th>region_seq</th>\n",
       "      <th>folding_energy</th>\n",
       "      <th>folding_struct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.052750</td>\n",
       "      <td>GACGGATAAATTAAA</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>129</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>weak</td>\n",
       "      <td>112.914694</td>\n",
       "      <td>GAC</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAGACGGATAAATTAAACAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAGACGGATAAATTAAAC...</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>....((((.(((((((...((((.......).))).......))))...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.825298</td>\n",
       "      <td>CACCGGTAAAGGTGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>weak</td>\n",
       "      <td>66.880242</td>\n",
       "      <td>CAC</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACACCGGTAAAGGTGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...</td>\n",
       "      <td>-12.5</td>\n",
       "      <td>..(((.((.((..((((......))))..)).))))).(((((......</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.825298</td>\n",
       "      <td>CACCGGTAAAGGTGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>weak</td>\n",
       "      <td>66.880242</td>\n",
       "      <td>CAC</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACACCGGTAAAGGTGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>....(((((((..((((......))))...........)))))))....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.381081</td>\n",
       "      <td>ACAAATTAATTCCTG</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>298</td>\n",
       "      <td>1.297763e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>strong</td>\n",
       "      <td>240.481383</td>\n",
       "      <td>ACA</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AAAACAAATTAATTCCTGCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...</td>\n",
       "      <td>-12.5</td>\n",
       "      <td>..(((.((.((..((((......))))..)).))))).(((((......</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.381081</td>\n",
       "      <td>ACAAATTAATTCCTG</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>298</td>\n",
       "      <td>1.297763e-08</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>strong</td>\n",
       "      <td>240.481383</td>\n",
       "      <td>ACA</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AAAACAAATTAATTCCTGCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>....(((((((..((((......))))...........)))))))....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   log10DAMratio              seq  Uncut_cnt  DpnI_cnt  MboI_cnt  \\\n",
       "0       2.052750  GACGGATAAATTAAA          2         0       129   \n",
       "1       1.825298  CACCGGTAAAGGTGT          0         0        76   \n",
       "1       1.825298  CACCGGTAAAGGTGT          0         0        76   \n",
       "1       2.381081  ACAAATTAATTCCTG         61         0       298   \n",
       "1       2.381081  ACAAATTAATTCCTG         61         0       298   \n",
       "\n",
       "     DpnI_ratio  MboI_ratio promoter    DAMratio codon_-2  \\\n",
       "0  1.685342e-08    0.000002     weak  112.914694      GAC   \n",
       "1  1.685342e-08    0.000001     weak   66.880242      CAC   \n",
       "1  1.685342e-08    0.000001     weak   66.880242      CAC   \n",
       "1  1.297763e-08    0.000003   strong  240.481383      ACA   \n",
       "1  1.297763e-08    0.000003   strong  240.481383      ACA   \n",
       "\n",
       "                         ...                         n_4 n_5 n_6 n_7 n_8  \\\n",
       "0                        ...                           T   T   A   A   A   \n",
       "1                        ...                           G   G   T   G   T   \n",
       "1                        ...                           G   G   T   G   T   \n",
       "1                        ...                           T   C   C   T   G   \n",
       "1                        ...                           T   C   C   T   G   \n",
       "\n",
       "            seq_extended contains_GATC  \\\n",
       "0  AAAGACGGATAAATTAAACAG         False   \n",
       "1  AAACACCGGTAAAGGTGTCAG         False   \n",
       "1  AAACACCGGTAAAGGTGTCAG         False   \n",
       "1  AAAACAAATTAATTCCTGCAG         False   \n",
       "1  AAAACAAATTAATTCCTGCAG         False   \n",
       "\n",
       "                                          region_seq folding_energy  \\\n",
       "0  AAACCAGGAGTCGTTTCACCCGCGAAAAAAGACGGATAAATTAAAC...           -8.4   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...          -12.5   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...           -6.5   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...          -12.5   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...           -6.5   \n",
       "\n",
       "                                      folding_struct  \n",
       "0  ....((((.(((((((...((((.......).))).......))))...  \n",
       "1  ..(((.((.((..((((......))))..)).))))).(((((......  \n",
       "1  ....(((((((..((((......))))...........)))))))....  \n",
       "1  ..(((.((.((..((((......))))..)).))))).(((((......  \n",
       "1  ....(((((((..((((......))))...........)))))))....  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def dam_fold(s):\n",
    "    f = RNA.fold(s)\n",
    "    return pd.Series({'folding_energy':f[1], 'folding_struct':f[0]})\n",
    "\n",
    "damDf5 = damDf3.join(damDf3['region_seq'].apply(dam_fold))\n",
    "damDf5.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>seq</th>\n",
       "      <th>Uncut_cnt</th>\n",
       "      <th>DpnI_cnt</th>\n",
       "      <th>MboI_cnt</th>\n",
       "      <th>DpnI_ratio</th>\n",
       "      <th>MboI_ratio</th>\n",
       "      <th>promoter</th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>codon_-2</th>\n",
       "      <th>...</th>\n",
       "      <th>n_5</th>\n",
       "      <th>n_6</th>\n",
       "      <th>n_7</th>\n",
       "      <th>n_8</th>\n",
       "      <th>seq_extended</th>\n",
       "      <th>contains_GATC</th>\n",
       "      <th>region_seq</th>\n",
       "      <th>folding_energy</th>\n",
       "      <th>folding_struct</th>\n",
       "      <th>folding_energy_bucket</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.052750</td>\n",
       "      <td>GACGGATAAATTAAA</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>129</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>1.902999e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>112.914694</td>\n",
       "      <td>GAC</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAGACGGATAAATTAAACAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAGACGGATAAATTAAAC...</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>....((((.(((((((...((((.......).))).......))))...</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.825298</td>\n",
       "      <td>CACCGGTAAAGGTGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>1.127161e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>66.880242</td>\n",
       "      <td>CAC</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACACCGGTAAAGGTGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...</td>\n",
       "      <td>-12.5</td>\n",
       "      <td>..(((.((.((..((((......))))..)).))))).(((((......</td>\n",
       "      <td>(-13.504, -8.872]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.825298</td>\n",
       "      <td>CACCGGTAAAGGTGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>1.127161e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>66.880242</td>\n",
       "      <td>CAC</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACACCGGTAAAGGTGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>....(((((((..((((......))))...........)))))))....</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.381081</td>\n",
       "      <td>ACAAATTAATTCCTG</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>298</td>\n",
       "      <td>1.297763e-08</td>\n",
       "      <td>3.120878e-06</td>\n",
       "      <td>strong</td>\n",
       "      <td>240.481383</td>\n",
       "      <td>ACA</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AAAACAAATTAATTCCTGCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...</td>\n",
       "      <td>-12.5</td>\n",
       "      <td>..(((.((.((..((((......))))..)).))))).(((((......</td>\n",
       "      <td>(-13.504, -8.872]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.381081</td>\n",
       "      <td>ACAAATTAATTCCTG</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>298</td>\n",
       "      <td>1.297763e-08</td>\n",
       "      <td>3.120878e-06</td>\n",
       "      <td>strong</td>\n",
       "      <td>240.481383</td>\n",
       "      <td>ACA</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AAAACAAATTAATTCCTGCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>....(((((((..((((......))))...........)))))))....</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.794629</td>\n",
       "      <td>GGTTTCTAATCGCGT</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>286</td>\n",
       "      <td>6.741369e-08</td>\n",
       "      <td>4.201236e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>62.320226</td>\n",
       "      <td>GGT</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAGGTTTCTAATCGCGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...</td>\n",
       "      <td>-9.2</td>\n",
       "      <td>....((((.(((........(((((..............))))).....</td>\n",
       "      <td>(-13.504, -8.872]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.794629</td>\n",
       "      <td>GGTTTCTAATCGCGT</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>286</td>\n",
       "      <td>6.741369e-08</td>\n",
       "      <td>4.201236e-06</td>\n",
       "      <td>weak</td>\n",
       "      <td>62.320226</td>\n",
       "      <td>GGT</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAGGTTTCTAATCGCGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>...((..((((.((((......(........).....)))).))))...</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.303350</td>\n",
       "      <td>AGAAATTAAATTACT</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>249</td>\n",
       "      <td>1.297763e-08</td>\n",
       "      <td>2.609430e-06</td>\n",
       "      <td>strong</td>\n",
       "      <td>201.071391</td>\n",
       "      <td>AGA</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAAGAAATTAAATTACTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAAGAAATTAAATTACTC...</td>\n",
       "      <td>-9.2</td>\n",
       "      <td>....((((.(((........(((((..............))))).....</td>\n",
       "      <td>(-13.504, -8.872]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.303350</td>\n",
       "      <td>AGAAATTAAATTACT</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>249</td>\n",
       "      <td>1.297763e-08</td>\n",
       "      <td>2.609430e-06</td>\n",
       "      <td>strong</td>\n",
       "      <td>201.071391</td>\n",
       "      <td>AGA</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAAGAAATTAAATTACTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAAGAAATTAAATTACTC...</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>...((..((((.((((......(........).....)))).))))...</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.709659</td>\n",
       "      <td>CGGACGTAACGGGCA</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>8.636688e-07</td>\n",
       "      <td>weak</td>\n",
       "      <td>51.245900</td>\n",
       "      <td>CGG</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>AAACGGACGTAACGGGCACAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAACGGACGTAACGGGCAC...</td>\n",
       "      <td>-10.1</td>\n",
       "      <td>...........((((.(..(((........)))..).))))(((.....</td>\n",
       "      <td>(-13.504, -8.872]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   log10DAMratio              seq  Uncut_cnt  DpnI_cnt  MboI_cnt  \\\n",
       "0       2.052750  GACGGATAAATTAAA          2         0       129   \n",
       "1       1.825298  CACCGGTAAAGGTGT          0         0        76   \n",
       "1       1.825298  CACCGGTAAAGGTGT          0         0        76   \n",
       "1       2.381081  ACAAATTAATTCCTG         61         0       298   \n",
       "1       2.381081  ACAAATTAATTCCTG         61         0       298   \n",
       "2       1.794629  GGTTTCTAATCGCGT          3         3       286   \n",
       "2       1.794629  GGTTTCTAATCGCGT          3         3       286   \n",
       "2       2.303350  AGAAATTAAATTACT         49         0       249   \n",
       "2       2.303350  AGAAATTAAATTACT         49         0       249   \n",
       "3       1.709659  CGGACGTAACGGGCA         10         0        58   \n",
       "\n",
       "     DpnI_ratio    MboI_ratio promoter    DAMratio codon_-2  \\\n",
       "0  1.685342e-08  1.902999e-06     weak  112.914694      GAC   \n",
       "1  1.685342e-08  1.127161e-06     weak   66.880242      CAC   \n",
       "1  1.685342e-08  1.127161e-06     weak   66.880242      CAC   \n",
       "1  1.297763e-08  3.120878e-06   strong  240.481383      ACA   \n",
       "1  1.297763e-08  3.120878e-06   strong  240.481383      ACA   \n",
       "2  6.741369e-08  4.201236e-06     weak   62.320226      GGT   \n",
       "2  6.741369e-08  4.201236e-06     weak   62.320226      GGT   \n",
       "2  1.297763e-08  2.609430e-06   strong  201.071391      AGA   \n",
       "2  1.297763e-08  2.609430e-06   strong  201.071391      AGA   \n",
       "3  1.685342e-08  8.636688e-07     weak   51.245900      CGG   \n",
       "\n",
       "           ...          n_5 n_6 n_7 n_8           seq_extended contains_GATC  \\\n",
       "0          ...            T   A   A   A  AAAGACGGATAAATTAAACAG         False   \n",
       "1          ...            G   T   G   T  AAACACCGGTAAAGGTGTCAG         False   \n",
       "1          ...            G   T   G   T  AAACACCGGTAAAGGTGTCAG         False   \n",
       "1          ...            C   C   T   G  AAAACAAATTAATTCCTGCAG         False   \n",
       "1          ...            C   C   T   G  AAAACAAATTAATTCCTGCAG         False   \n",
       "2          ...            G   C   G   T  AAAGGTTTCTAATCGCGTCAG         False   \n",
       "2          ...            G   C   G   T  AAAGGTTTCTAATCGCGTCAG         False   \n",
       "2          ...            T   A   C   T  AAAAGAAATTAAATTACTCAG         False   \n",
       "2          ...            T   A   C   T  AAAAGAAATTAAATTACTCAG         False   \n",
       "3          ...            G   G   C   A  AAACGGACGTAACGGGCACAG         False   \n",
       "\n",
       "                                          region_seq folding_energy  \\\n",
       "0  AAACCAGGAGTCGTTTCACCCGCGAAAAAAGACGGATAAATTAAAC...           -8.4   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...          -12.5   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...           -6.5   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...          -12.5   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAAACAAATTAATTCCTGC...           -6.5   \n",
       "2  AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...           -9.2   \n",
       "2  AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...           -5.0   \n",
       "2  AAACCAGGAGTCGTTTCACCCGCGAAAAAAAGAAATTAAATTACTC...           -9.2   \n",
       "2  AAACCAGGAGTCGTTTCACCCGCGAAAAAAAGAAATTAAATTACTC...           -5.0   \n",
       "3  AAACCAGGAGTCGTTTCACCCGCGAAAAAACGGACGTAACGGGCAC...          -10.1   \n",
       "\n",
       "                                      folding_struct folding_energy_bucket  \n",
       "0  ....((((.(((((((...((((.......).))).......))))...       (-8.872, -4.24]  \n",
       "1  ..(((.((.((..((((......))))..)).))))).(((((......     (-13.504, -8.872]  \n",
       "1  ....(((((((..((((......))))...........)))))))....       (-8.872, -4.24]  \n",
       "1  ..(((.((.((..((((......))))..)).))))).(((((......     (-13.504, -8.872]  \n",
       "1  ....(((((((..((((......))))...........)))))))....       (-8.872, -4.24]  \n",
       "2  ....((((.(((........(((((..............))))).....     (-13.504, -8.872]  \n",
       "2  ...((..((((.((((......(........).....)))).))))...       (-8.872, -4.24]  \n",
       "2  ....((((.(((........(((((..............))))).....     (-13.504, -8.872]  \n",
       "2  ...((..((((.((((......(........).....)))).))))...       (-8.872, -4.24]  \n",
       "3  ...........((((.(..(((........)))..).))))(((.....     (-13.504, -8.872]  \n",
       "\n",
       "[10 rows x 38 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out, bins = pd.cut(x=damDf5['folding_energy'], bins=5, retbins=True)\n",
    "damDf5['folding_energy_bucket'] = out\n",
    "damDf5.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "damDf5.to_csv(ctermELMseqPath / 'damDf5.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>folding_energy_bucket</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(-27.423, -22.768]</th>\n",
       "      <td>29.0</td>\n",
       "      <td>-0.050674</td>\n",
       "      <td>0.690162</td>\n",
       "      <td>-1.626450</td>\n",
       "      <td>-0.158103</td>\n",
       "      <td>0.203839</td>\n",
       "      <td>0.381166</td>\n",
       "      <td>0.744987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-22.768, -18.136]</th>\n",
       "      <td>814.0</td>\n",
       "      <td>-0.013153</td>\n",
       "      <td>0.586510</td>\n",
       "      <td>-1.918525</td>\n",
       "      <td>-0.164221</td>\n",
       "      <td>0.219634</td>\n",
       "      <td>0.361923</td>\n",
       "      <td>1.430169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-18.136, -13.504]</th>\n",
       "      <td>19800.0</td>\n",
       "      <td>-0.000464</td>\n",
       "      <td>0.554648</td>\n",
       "      <td>-2.106516</td>\n",
       "      <td>-0.137878</td>\n",
       "      <td>0.203145</td>\n",
       "      <td>0.348982</td>\n",
       "      <td>1.562056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-13.504, -8.872]</th>\n",
       "      <td>208571.0</td>\n",
       "      <td>-0.003266</td>\n",
       "      <td>0.550930</td>\n",
       "      <td>-2.368689</td>\n",
       "      <td>-0.151370</td>\n",
       "      <td>0.197085</td>\n",
       "      <td>0.346855</td>\n",
       "      <td>1.825298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(-8.872, -4.24]</th>\n",
       "      <td>563545.0</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.547138</td>\n",
       "      <td>-2.993667</td>\n",
       "      <td>-0.153507</td>\n",
       "      <td>0.199634</td>\n",
       "      <td>0.348722</td>\n",
       "      <td>2.052750</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          count      mean       std       min       25%  \\\n",
       "folding_energy_bucket                                                     \n",
       "(-27.423, -22.768]         29.0 -0.050674  0.690162 -1.626450 -0.158103   \n",
       "(-22.768, -18.136]        814.0 -0.013153  0.586510 -1.918525 -0.164221   \n",
       "(-18.136, -13.504]      19800.0 -0.000464  0.554648 -2.106516 -0.137878   \n",
       "(-13.504, -8.872]      208571.0 -0.003266  0.550930 -2.368689 -0.151370   \n",
       "(-8.872, -4.24]        563545.0  0.001225  0.547138 -2.993667 -0.153507   \n",
       "\n",
       "                            50%       75%       max  \n",
       "folding_energy_bucket                                \n",
       "(-27.423, -22.768]     0.203839  0.381166  0.744987  \n",
       "(-22.768, -18.136]     0.219634  0.361923  1.430169  \n",
       "(-18.136, -13.504]     0.203145  0.348982  1.562056  \n",
       "(-13.504, -8.872]      0.197085  0.346855  1.825298  \n",
       "(-8.872, -4.24]        0.199634  0.348722  2.052750  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoter = 'weak'\n",
    "damFoldDf = damDf5[damDf5['promoter'] == promoter]\n",
    "damFoldDf.groupby('folding_energy_bucket')['log10DAMratio'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQkAAADUCAYAAACPv9YhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3Xl8TPf+x/HXzGQhxBK5QYOmoREpba56aFAeLWItutGiGkpRS4urbpc8XFyll6CXLkIXSWMptzRpm4jYG721VIVa2oSGkAYjEib7Mr8/cjM/Ic6MzCRnJvk8/2lPcvI9H5OT93znnO/5fjVGo9GIEELchVbtAoQQ9k1CQgihSEJCCKFIQkIIoUhCQgihSEJCCKFIQkIAcPDgQVasWKHKsb///ntVjissIyEhVBcVFaV2CUKBk9oFCNsJCQkhIiKCNWvWkJOTw8yZM3n11VdZuHAhb7/9Nvn5+fTv35+QkBDWr19PQkICN27c4M0330SrLXu/OH78OB9//DGrVq3C2dkZgMmTJ9OkSRN+++03XnvtNQICApg6dSqurq6sWLGCf/7znxgMBry9vVm8eDHTpk2jadOm/PbbbwwdOpQffviB7OxsIiMjOX78OMuWLQNg3LhxZGdn8/vvv7N161bq169PREQEzs7OzJs3D71eT1hYGM7Ozqxbtw4XFxfVXtu6THoStUi7du1ITU3l3LlznD9/nsuXL+Pt7c3atWuZPn06GzZs4NChQ2RmZlJSUsK6detYtGgR3377LQDp6eksX77c9IdZLicnh7Fjx7J+/XrWrVsHQP369dm8eTM7duygf//+REVF0aZNGxISErh58yYvv/wyc+fOJT4+nrVr1/Lwww9z+vRp/v3vf/Ppp58SFRXF559/zvPPP4+fnx/Dhg0jMjKSqKgoli1bxsqVKwFo06YNGzZskIBQkfQkapFu3bpx+PBhAIxGI4mJifTo0YOvvvqKsLAwtFot2dnZZGRkUFJSwuzZs3Fy+v9T4Oeff8bLy6vC1wBcXV3x9/cHQKfTAeDt7Q1AamoqL730EgCdOnXi9OnTAPj6+nL58mVat24NQIMGDSgsLATA3d3d1Ma1a9cAuH79OqmpqYwbNw7AFFKtWrWy5UskqkB6ErVIUFAQX3/9Nb6+vvj7+7N+/XqCgoJo3bo1c+fO5csvv+SFF17A29ubffv2ERYWRt++fSl/fGfIkCGMGDGCTz/9tEK7BQUFpKWlkZubawoJjUYDQOvWrTl58iRQ9lGl/I+6/Pu3Ky0txWAwUFRUxIULF2jatCkATZs2xc/Pj8jISD766CN69+5t+xdIVImERC3SsGFDADp37kyXLl1wcXHB3d2diRMnsmTJEkaMGMHZs2dxd3fHycmJESNGEBMTQ3Z2tqmNp59+mgMHDpCenl6h7WXLljFmzBjGjx9f4esvvPAC27dvZ9SoUVy6dImBAwcq1vj6668zfvx4XnzxRSZNmoSzszM6nY7Nmzfz/PPPM2rUKEJCQrj//vtt9KoIa2nkKVBhzpgxY/jiiy/u+Bgi6gbpSQghFElPQgihSHoSQghFEhJCCEUSEkIIRaperj527Biurq5qliBEnVRQUEBgYKBF+6oaEq6urnTo0EHNEoSok8pHxlpCPm4IIRRJSAhhp/R6PdOnTzc936IWCQkh7FRERATHjx8nIiJC1TokJISwQ3q9nri4OIxGI3Fxcar2JiQkhLBDERERpqdzS0tLVe1NVEtIrF+/nrVr15omDhFC3JuEhASKiooAKCoqYseOHarVUi0hERwczCuvvFLhEWQhhOWCg4NNE+84OzvTr18/1WqplnES7u7uLF++3DTL0N0UFBTc0/1aIeqKHj16EBsba9ru3r27an8r1RISoaGheHl5sXPnTsaOHXvX/WQwlagper2e+fPnM2/ePJo1a6Z2ORbp3bs38fHx9OnTh6CgIJu2fS+BUy0hUT4bshD2Ijw8nKSkJMLDw3nnnXfULsciBQUFFf6rFrm7IWo9vV5PQkICADt27FB9cJIl9Ho9+/fvB2Dfvn1yC1SI6hQeHk5paSlQdjsxPDxc5YrMs6eaJSRErbdz584K2+W9CntmTzVLSAghFElIiFqvefPmitv2yJ5qlpAQtd6ff/6puG2Pbl/35PbtmiQhIarEXh5jtsTtE8LLBPH3RkJCVMmt4w5E7SYhIe6ZXq8nPj4egO3btztEb0JUnYSEHXCkrjtwR+9BehO1m4SEHbCXGYgsdftjy+W9ClE7SUiozJ5mILKUXAisWyQkVGZPMxAJURkJCZXZ0wxEQlRGQkJl9jQDkRCVkZBQWUhICBqNBgCtVktISIjKFZnn4uKiuC1qFwkJlXl6ejJw4EA0Gg0DBw50iFmTnnzyyQrbvXv3VqkSURMkJOzAkCFDcHNzY+jQoWqXYpHbZ0pSe+YkUb0kJOzAt99+S25uLjExMWqXYpG9e/dW2N6zZ486hYgaISGhMkccJyHqFgkJlck4CWHvJCRUJuMkhL2rlin1heWCg4OJjY2lqKhIxklU0fbt2yssZGOJ119/vdKvDxo0iAEDBtiiLEWOVLOEhMpCQkKIi4sD7GechCOdwKL6SUiorHycRExMjMOMk7A3AwYMUAyiXr163fE1tRezdqSaqy0kcnJymDt3LlOmTKFt27bVdRi7Z8m78tWrV3FxcSE5Ofmu78hQc+/K5k7gQ4cOMXv2bNP2ihUrePTRR6u9rqqqX78+eXl5pu0GDRqoWI3jqbaQuHr1Kh07dlTcpy4sGJyenk5ubq7iPpmZmWi1WoqKikwXMe/Wlj28Xu7u7qb/1+l0uLm52UVdd/PBBx8wadIk0/by5cvtul4om8jn1prDw8Nr14LBAD4+PjRu3Fhxn7qwYHCHDh145ZVXFPcp7z2o3QW+F76+vpw7d46wsDCH+h02aNDAoeotZ+uaVV8wWNR+jRo1IjAw0K4/ZtwqMDAQcKwgtpeaqzUknn322epsXghRA2QwlRBCkYSEEEKRhIQQQpHDXbi0ZNxBZmYmAB4eHnfdR0YCCmEZsyFx5swZEhMT0el0dO/enfbt29dEXVYpf9xaKSSEEJZRDIk1a9bg4eHB448/jlarJSkpif/+97+MHTu2hsq7k7nRgOCY4w6EsFeKITF+/Hh0Oh03btzAaDQyfPhwSkpKaqo2IYQdUAwJnU7Hhx9+aPqM7+XlxeTJk2ukMCGEfTB7TSIrK4u5c+cCsGTJkmovSAhhXyy6u7FmzRo0Gg2FhYXVXY8Qws4ohoTRaOTtt98mOTkZnU5Xpx/5FqKuMnt3IzU1tcLXFi9eXJ31CCHszF1DIj8/n4EDB7Jr1y769u2L0Whk06ZNNVmbEMIO3DUkiouLiYmJ4ejRo+Tk5KDRaPD396/WYlauXElKSorV7SQnJwN3n3fRUu3atbO6DSEc3V1DomHDhkybNo2zZ89y4sQJjEYjWVlZ1VpMSkoKv5w4RambdSMlNSVl/6yfz2ZUuQ1tbqbZfewt1MB8sNXVmh2tXqjZmpWYvbsRFhaGj48Pfn5+nDhxokoHuRelbh7kBzxV7ccxp96p78zuk5KSwu+/HqVNQ+sGmDUylq0qnp962Kp2Lhh0ZvdJSUnhl5O/QBOrDmV6NPCXS79Y144F7zspKSmcOXaMFlYcpn754Y4ds6IVsORtJyUlhZMnTtPEzcuqY2lLXAG4dNa6Vd2ycq9Y9fNmQ8LDw4M+ffqQlZVldjq6uqhNwxJCuxjULgOAhUcaWrZjEyh9orR6i7GQdq9lDyK3AMajqd5iLPAZRov2a+LmxZP+L1ZzNZbZc8a6a4lmf0ODBw/Gw8ODn376ib/85S9WHUwI4XgUexJ//PEHycnJPPTQQ8yYMYPly5dXazGZmZloc69Z1NWvbtrca2Rmuijuk5mZydWbOsvfwavZ+Zs6/pKpfC0lMzMTsix/B692WZBZ3/z1H6EexZBYtmwZEyZMYP78+RgMBqZMmVJTdQlhkpmZyWUs7+pXpz+BUguCOCv3itXdfFvJyr1C/cyqf1RTDAlPT08CAwP54IMPWLVqVYX1FqqDh4cHf1wvtJsLl+bmo/Dw8MDtxlm7uiZRz4Kaz+edt6trEjLvh31TDAmDwUBaWhqNGjUiKyuLrKwsWrduXVO1CQGUBZv2wgW7uXDZxIIgzrtutKsLl9YEseIHUx8fH6Kjo/Hz8yM6Opro6OgqH0gI4ZgUexLFxcVoNGXpbTSq/3lQCFHzFEOisLCQK1eu0KVLF3r06IFOZ36wDsCuXbvIyMigqKhI1anuhBDWUwyJOXPmAHDkyBG++eYbbt68yTvvvGO20aNHj/Lmm2+avWV6+4LB5hbWrWm5ubmKaybm5uba3ZoEltRsbxytZkerF8zXrMTsiMujR4+yd+9e8vLy6NKli2WNOpU1W1qqfAX99gWD3dzcgBsWHaMmuLm5KS7U6ubmRn4N1mMJS2rmeg0WZAFLaran6Y4sqfc6eTVYkXm312yzBYN79+5N69atefLJJ2nQoAE5OTkWNdqxY0fWrVsnIzSFqAUUQ+L99983/b9GoyE9Pd2iRoODg62rSlQvW4y4LO9C1bO+FrytbENUK8WQ6Nq1KwCJiYls2rQJrVbLsGHDqrUgbW6m1cOyNUVlXT2jc30zeyrXgQXPHV4wWD8sO7uw7A5SYxfr7iBdMOjwM7NPu3btrDpGufLHmB/0ftC6hrwtqykD60Zclg93s3YAfQaWPUBrixGX+UVlPfd6zg2saicr9wreNKvyzyuGxLp16zhy5AgPPfQQ9erVIywsrMoHsoTNT+C21jxc3MJsPbaqN+1/9Tb3se4Pzg/zNdlqEp2aXADJFq/z1f+9xq0etO41bmJBPbY7j8uGf3u3bWNVO940s6omxZA4d+4cHh4e+Pv7W/xRwxqOdgI7Wr2Oyhavc02+xrXtvFD8YLpgwQJmz55Neno6ycnJLF26tKbqEkLYCbO3QI1GI4888giDBg3izz//rImahBB2RDEkNm7cyB9//IGvry9btmyhbdu2BAQE1FRtQgg7oBgSR44cYdmyZabtmTNn8vLLL1d7UUII+6F4TeL2ZzWcnZ2rtRghhP0xO31d+fMbRqOR8+fP10hRQgj7oRgSW7Zsqak6hBB2yt4eYhRC2BkJCSGEonsKiRUrVlRXHUIIO2V2MFVxcTEZGRkYjUaef/75mqhJCGFHzIbErFmzaNeuHVptWadj2rRp1V6UEMJ+mA2JDh06MG7cOFxclFezEkLUTmZD4vTp08ydO9c0sGrx4sXVXpQQwn6YDQlfX19mzJhRE7UIIeyQ2ZA4f/4806ZN+98ktbBkyZJqL0rYv2PHjgHQq1cv9u/fr3I1ojqZDYmAgABeffXVmqjFItu3byc2NlZxn/KZqZQm/xg0aBADBgywaW3CfjliqNlLzWZDYteuXej1elNP4o033qj2oqzVrFnV5/NTQ1FREampqVy7ds0hau/Vq9cd247yhyfuncZoZv2+S5cukZ2dDUDjxo3x9rbd1ManT59WXL+grujXrx/5+fkEBASwevVqtcsx21srf4e7VWBgYKX72kOP7fZQA+w+1Kq75nv52zPbk9i2bRuZmWUTcnp5eTF58mTrqhMV6PV68vPL5qc/deqUw/Qm7IklH0Fvd7ePojUVao5Us9mQyMrKYu7cuYBctKwKcyfDmTNnKmyPHDkSf3//SvetqRN4wIABisep7F1O7claRfUxGxIAa9asQaPRUFho2WJrsbGxJCcnO8T1C7WV9yLuti3Mc8RQc6SazYZEaGgoKSkpGI1GHlRYs2DDhg0cPnwYo9HIuHHjOHXqlE0LdVSOdDIIURmzITF//nwMBoNpEeC7jbgcNWoUo0aNMm0nJCSYPfjtq4qLMo74mjhazY5WL6hXs9mQ8PLy4h//+Ee1HPz2VcVFGUd8TRytZkerF2xbs81WFYeyHsHp06epV69sZVhLL17Onj3b4iKEEPZLMSQMBgNbt26t8LWbN2/i7u5erUUJIeyHYkhER0eTnZ3Ngw8+iE6n4+TJk7Ro0YLhw4fXVH1VotfrmT9/PvPmzZMxB0JYSXH6utGjR/Piiy+i0WgoLS1l9OjRdh8QAOHh4SQlJREeHq52KUI4PMWexN69e/nuu++AsnU3XF1d6dmzZ40UVlV6vd50Z2XHjh1MmjRJehNCWEGxJ7F7927CwsIICwtj2bJlxMfH11RdVRYeHk5paSkApaWl0psQwkqKIVFUVMS2bdtITExk27ZtNVWTVXbu3Flh25LxGkKIu1P8uLFgwQIOHjzI9evX8fb25umnn66puqpMo9Eobgsh7o1iTyIjI4P777+fwMBAWrZsycWLF2uqrirr06dPhe2+ffuqVIkQtYNiT2Lx4sW4ubnh4+Nj+pq9T6k/adIkEhISKC0tRavVMmnSJLVLEsKhKYbEihUr2LhxI2PHjq2hcqzn6elJcHAw8fHx9OvXT+5sCGElxZBwdXV1qIAoN2nSJDIyMqQXIYQNmH1246mnnqJVq1YAXLx4kaFDhzJx4sRqL8wamZmZpKSkcP36delJCGElswsGBwUFsXr1alavXk3Pnj0xGAw1UZdVFi5cSE5ODgsWLFC7FCEcntmehEajYe3atWg0GvLz800redmr33//ndTUVABSU1NJSUmhXbt26hYlhAMzGxLvvvsux44do7i4mAkTJlg8hZ1aFi5cWGF7wYIFREZGqlSNEI7PbEgsXLiQli1bYjQa2b17N3PmzKmJuqqsvBdxt20hxL0xe00CYPz48UyYMAEzS3TYhVvHdFS2LYS4N2Z7Ek2aNGHevHloNBqHuFMQGhrKhAkTTNvlywEIIapGMSTS0tIYNmwYeXl5GI1G01J/9szPzw8fHx9SU1Px8fGx+4uWrq6uFBQUVNgWwp6YnZnqdvY+LBvKehNvvPGGQ/Qibg2IyraFUJtiSDhCIFTGz8+PuLg4tcsQolaw6MKlEKLukpBQmZeXV4Xt5s2bq1SJEJWTkFBZQECA4rYQapOQUNmhQ4cqbB88eFClSoSonM1DIjU1lUWLFrFgwQKuXLli6+ZrneDgYNPzMDqdjn79+qlckRAVmR1MZanyVcULCwsJDQ0lPj4eg8Fwx2fuW8mCwdCjRw9iY2MpKSlBq9XSvXt3h3xNHK1mR6sX7HjBYEvduqp4TEwMnp6e+Pr6Kv6MLBhcZtCgQcTExDB48GCCgoLULqdKHO336Gj1gnoLBtv848b+/fuJjo7mwoULpKWl2br5WikkJISHH36YkJAQtUsR4g4260mU69WrF7169bJ1s7Wap6cnq1atUrsMISoldzfEPZMnbesWCQlxz0JDQytsO8IzMqLqJCTEPSt/0hZwiCdthXUkJESVhIaG0qBBA+lF1AE2v3Ap6gZ50rbukJ6EEEKRhIQQduj2GcrUnLFMQkLUes7Ozorb9ui9996rsP3++++rVImEhKgDBg0aVGF78ODBKlViua5du5p6D66urjz66KOq1SIhIWq9kJCQCk/aOsrw9/feew+tVqtqLwIkJEQd4OnpyVNPPYVGo2HIkCEOsTQElPUm9u7dq2ovAuQWqKgjQkJCSE1NdZhehD2RkBB1gjxEV3XycUMIoUhCQgihSEJCCKFIQkIIoUhCQgihSEJCCKFIQkIIoUhCQgihSEJCCDul1+uZPn06165dU7UOCQkh7FRERATHjx8nIiJC1TokJISwQ3q9nri4OIxGI3Fxcar2JmweEpcuXWL16tW89957lJSU2Lp5IeqEiIgIjEYjAKWlpar2JmwWEhs2bGDmzJmEhYXRuXNnsrKyTP9IIcS9SUhIoKioCICioiJ27NihWi0ao43/ktPT02natCm7d++mVatWPPLII3fd99ixY6rO3SeEvVq/fj0HDhygpKQEnU7H448/blqQ21YsXYDY5o+K5+XlERkZiUaj4YknnlDcV1YVF6JyM2bM4KeffqKkpAQnJydmzJhh08ly7mVVcZuHRNu2bXnrrbds3awQdYqnpycDBw4kJiaGgQMHqjqblkw6I4SdspfZtCQkhLBT9jKbloyTEEIokpAQQihS9eNGQUHBPV1lFULYRkFBgcX72nychBCidpGPG0IIRRISQghFEhJCCEUSEkIIRRISQghFunnz5s1T48A3b94kISEBPz8/oOypt4MHD3LgwAFcXV3ZunUrK1eupGvXrjRq1AiAffv2sXXrVlq2bMnHH3/Mnj178PPzIyoqisTERLy8vPDw8LjjWLe2rdPp2LRpE/Hx8QQFBeHk5MQPP/zA999/z/LlyxkwYAArV67kyJEjeHt7ExERwdGjR8nNzeXDDz8kLS2Nzp0727RdHx+fO2ouKioiLCyM0tJSGjduzKpVq9i7dy/t27enYcOGGI1GIiIiOHPmDD4+PqxZs4Y9e/YQEBDAl19+yYkTJ7h+/TofffRRhZpzcnJ4++23adeuHTdu3CAiIoL4+Hi6deuGk5NTheM6OTmxYcMG4uPj6dGjBytXriQpKQkXFxeaN29+R823tn3x4kViY2PZs2cP3bp1AyAuLo5vvvmGS5cu0bhxYyIjI9m/fz+PPfYYWVlZvPbaa/Tr149p06bRrVs3du/ebTo/bNH2008/Xem5uH37dn788Uf27NlDUFAQAFu2bOHYsWMcOnSInJwc0++xU6dOpte6Q4cOzJkzh8GDB5s9v605RoMGDSqte+nSpZSUlHD//fcDsGTJEk6ePElaWhr+/v5s2LCBX3/9lby8PDZu3Ej37t0rbccc1XoS0dHR9OrVy7QdHBzMK6+8QnZ2NoGBgQwZMoTg4GBatWoFwKlTpzAajZSWltKsWTNmzZpFmzZtyMvLY/To0Vy9ehWdTlfpsW5t+4EHHmDGjBm4u7tTWFgIQM+ePQkKCmLMmDEcOnSIkpISNBoNTZs2Ra/Xc+nSJdzd3QHo0qVLtbR7u8zMTPz9/QHQarVkZGSg1+txc3MDwGAw4Ovri9FopGHDhjz33HPo9Xry8vJISkri5s2btGnT5o6ar169SseOHQHQ6XT8+eefZGdn4+zsfMdxvb296du3L1lZWaSkpHD58mUKCgpo0aJFpTXf2vbDDz9MYGAgWVlZpu+fOXMGLy8vtFot27dvZ9SoUXTq1ImTJ0+ydetWAgICcHNzo2PHjnz33XcVzg9btH03RqORixcvmuZvgLJQSk9PR6vVVvg93nfffaZzz2Aw8MADD1Ta5u3ntzXHqExsbKwpHMqNHTuWYcOGkZ6ezoEDB2jcuDEAXbt2veu/3RKqhURGRgaNGjVi48aN/Oc//8Hd3Z3ly5czbtw4oOxFfuaZZ0z77969m9TUVH799VecnJzYsWMHnp6etGzZEoPBwMyZM++YmKOytps1a0ZUVBRBQUGmHgqUTfLRv39/ioqK6Ny5Mz179mTXrl20atWKf/zjH+zcuZMbN27QsGFDm7d7q++//54vvviC5s2bm/4YT58+zfDhwxkzZgyJiYkAuLu74+vrC5T9Ybu6ujJ69GiSkpJo2bIlU6ZMYfPmzXfU7OPjYzp5jhw5wtSpU3nyySf59ddfASocNz09HW9vb5544gny8/Np3749I0eOZOvWrZW+zre2ffbsWf7617/SqlUr9Ho9ACNHjiQkJITffvsNrVaLVlt2+iUmJlJYWMipU6c4ceIEAJcvX65wftiy7dvrTk5O5p133qG4uNg0UdKNGzd48803ycjIqPB7dHd3N5175a9/ZW4/v211jDVr1hAbG8vPP//MmTNnOHz4sOl7+fn5fPnll4wfP57ExERSUlIqfL+qVBtx6enpSWZmJiNHjgTgb3/7G15eXuzcuZOxY8eSl5dnepeNj49n2rRpQNncf4cPHyY6OppHH32UjIwMoqKicHd3Z8CAAcTHx9O/f3+AStt2dXXl559/xmAwEBAQwKFDh+jfvz9arRadTkdQUBCrV68GYMyYMaxdu5bPPvuMLl26kJGRQXZ2ts3bvbXmyrquLVu2JDIyEmdnZ0aPHl1h/3IfffQR9erV4+WXXyY5OZnPPvuMbt26cfXq1Qo138rHx4dNmzZRWFjIE088cUe7t88Nsn37dtavX0/fvn0rfZ1vlZmZSUxMDKWlpbi6unLgwAEuXrxIdnY2Hh4e9O3bly+++AKNRsPUqVNxdnYmNzeXTp06sXfvXpo1a1bh/LBV25XVvWnTJtasWYObmxvp6elcv34dZ2dnPvvsM1q2bAlg+j3u37/fdO6lpaXdUVu5289va45x6tQpU80TJ04EYNCgQRw8eJCCggJOnTpFvXr1mD17Nn369GH//v38/e9/5+LFi/zwww93rdFiRpXk5OQYv/76a5u3e/ToUZu3WW727NnGTz75xObtSs3/Lycnxzhx4kRjZmam3Z8fV65cMU6ZMqXS79ny/La25oMHDxqXLl1a5Z+XYdlCCEVyC1QIoUhCQgihSEJCCKFIQqKW+/HHH5k8eTKXL1+u8PW33nqL4uJi0/bs2bO5evUqmzdvrukShZ2TOS5ruS1bttCkSRPS09NNIymnTp0KlA3wWbhwIRqNhrS0NAoKCrh06RKrVq0iLy+P7OxsnnvuOZycnIiIiMDFxYX77ruP6dOnA2XrpkRHR5Ofn8+oUaPYsmULXl5epKamMn36dM6dO0diYiIGg4Hp06ezdOlSGjRowIsvvsjnn3+Op6cnBoOBRo0aMWHCBC5cuMDFixfvOjJSqENCopbr2bMnLVq04KuvvmLRokXcuHGDTz75BCgblOTl5cXEiRNNf/jlhg4dSr169fjuu+/IyMjgX//6F6dOnWLfvn2mfaKiomjRooVpjEhhYSFjx44lKSmJX375hbi4OB588EFKS0s5duwYRUVFLFq0iLCwMN599110Oh1Llizh2WefJSYmhoyMDGbNmlWjr48wT0Kijrj1Tndld73LRyiWq1+/PjqdDqPRaBpKrNFoKuxTUlLC5MmTyc/PJzk5mTNnzlCvXj2cnJwoLS0FYNasWfzxxx/k5+ebhpQXFxdXaMvf35/IyEjc3d1N+wj7ISFRR4wYMYK5c+ei0WiYNGkSn3zyCW3btmXbtm0sXrzYNLy5MsOHDyc0NJSioiLat29v+vpLL72fCKCFAAAAlklEQVREaGgoxcXFvPHGG3f83HPPPcecOXPIy8vj1ucIhw8fzvvvv0/Dhg1xcXEBoFWrVjz22GO2+wcLm5HBVMKsnTt3kpSURFZWFhMmTLjjwaJ7dfToUXbt2oXBYOCZZ54hKyuL+Ph4Fi9ebKOKhS1JSAghFMktUCGEIgkJIYQiCQkhhCIJCSGEIgkJIYQiCQkhhKL/A/D9pU6fb0/2AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 288x216 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQkAAADUCAYAAACPv9YhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3XtYVNX6B/DvzDCACEIDIgoa3kBJ09LHNI+kJ1G009X08ZKhaViKRyOPncqjmD6SipeDefLBLMdILUuDjItIKpE93gBFkQINA3GEEQccLs4Ms39/EPNzAPcMzDB775n38885SzZrfxuGl7X3rL2WiGEYBoQQ8hBirgMQQviNigQhhBUVCUIIKyoShBBWVCQIIayoSBBCWFGRsEMajQbHjx/nOka7qFQqZGdncx2DtIGKhB2qqKhAWloa1zHa5bfffsO5c+e4jkHa4MR1AGK5X3/9Fdu3b4der8eyZctw9uxZnD59GllZWYiNjYWnpyc++OADJCYmory8HJ6enti0aRPWrVsHV1dXXL58Gc8//zwiIiKwc+dOZGVlwd/fH87Ozvj4448N55k2bRq8vLygUqkQFxeH3NxcfPfdd/Dw8MC7776LNWvWQK/XY+rUqXj99dcxY8YMeHt7o6qqClOmTMHx48fRt29frFu3DnK5HCkpKZBKpYiJicE333yDnJwchIeHIzk5Gfn5+fDy8sLGjRvx+eefIzc3F/7+/li3bh2Hr7RjopGEHThx4gSWLFmC3bt3Q6fTYcaMGXj66acRGhqKuro67N+/HwqFAn369EFiYiLCwsLw9ddfAwCeeeYZ7N+/H0eOHMHdu3eRm5uLr7/+GuPHj291noqKCuzcuROrV6/G3r17AQAjR47Erl27sG3bNmzcuBEHDhzAqVOnUFlZiVu3bmH79u0ICwuDSqVCYmIiCgoKcOfOHZw6dQoHDx5EbGwstm7dihkzZuCFF16AXq9HbW0tEhMT8dxzz+G7774DAISHh1OB4AgVCTvw5ptvIjU1FVFRUdBqtUZf69mzJ8RiMUpKSjBkyBAAwNChQ1FSUgIA6Nu3L5ydndGlSxeUlpYiKCgIAAzHPig4OBiurq4YPHgwysrKAAD+/v4AgOrqavTu3RsikcjwdX9/f7i4uEAmk6FPnz4AAGdnZ5SVlSE4OBgikQi9e/dGTU2N4Rw3btzAmTNnMHfuXMjlcty5c8foPMT2qEjYgbS0NCxfvhwJCQnYvXs3RCIRmh/JEYlEAIDevXvjypUrAID8/HwEBAQYfR1oKijFxcUAgMLCwlbnKSkpgU6nw9WrV9GvXz+j7/fw8EBZWRkYhkFBQQF69epl1PeDevXqhd9//x0AUFpaCjc3N8OxAQEBmDBhAr788kt8+OGHGDlyZKucxLbonoQdGDhwICIjI+Hu7o5Zs2ZBJpOhuLjY6BOOsLAwvPfee5gzZw66deuGbdu2ISYmxqif7t27Y9iwYZg1axa8vLwgk8mMvq7X6xEZGYmGhgZs3rwZZ86cMXxt+fLlWLlyJbRaLV566SX06NHjoXm7d++OsWPHYubMmWAYBhs3boSLiwsyMzMxYcIEMAyD1157DTqdDlu2bEFeXp51XijSISJ6CpQ002g0SE5OxquvvooffvgB169fx7JlywxfnzVrFg4cOMBhQsIFGkkQA2dnZ+Tm5uLgwYNwcXHBJ598wnUkwgM0kiCEsKIbl4QQVlQkCCGsqEgQQlhxeuMyLy8PLi4uXEYgxCHdv38fw4cPN+tYTouEi4sLBg8ezGUEQhzS1atXzT6WLjcIIayoSBCHoFQqsXTpUsOzIELAl8xUJIhDkMvluHTpEuRyOddRzMaXzFQkiN1TKpVISUkBwzBISUnh/C+zOfiUmYoEsXtyudzwCL1Go+H8L7M5+JSZigSxe8eOHTNqp6enc5TEfHzKTEWCdAhfbqqZw8vLi7XNR3zKTEWCdAhfbqqZ49atW6xtPuJTZioSpN2USiVSU1PBMAxSU1MFMZogHUdFgrSbXC6HXq8HADQ2NgpiNEE6jooEabeMjAzodDoAgE6na3WTjdgXKhI8IKSbgAAwbtw4o3ZoaChHSYgtUJHgASHdBARgtAR+W21iX6hIcEyINwF//fVXo/bp06c5SkJsgYoEx+gmIOG7TikSX331FXbv3o34+PjO6N6u0E1AwnedsuhMWFgYvL29sWHDBtbj7t+/367FL+zR0KFDjTa5GTZsmCBfE6FlFlpegLvMnVIkPDw8sHXrVsyfP5/1OFqZqvV0W09PT0G+JkLLLLS8gHUzc74y1apVq6DX6422mSNty8rKMmqfOnWKoySEtK1TRhJbtmzpjG7tUo8ePQw7fDe3if1LS0tDSkpKu77nn//8Z5v/PnXqVISHh1sjVptomz+OKRQK1jYXhPQGBoSXV2ioSHDMz8/PaCTh5+fHXRhiM+Hh4azFqK1ZrFx9WkhFgmN8HEmYegNHRUXh0qVLhvaTTz6J7du32yJam4T0CydENJmKYy1HDkIYSbTcbZzLAmGOwMBAo/aAAQO4CdIOLW9ot2zbEhUJjt2+fZu1zVddunQB0DSK4Lt9+/YZtT///HOOkggTXW50MlM31dzc3FBfX2/UFsJNteDgYAD8H0U0c3Z2hkajEcQoolnzNnxcXxpRkeCYn5+f4aEukUgkiMsNIQoJCQHA/S+cENllkVAqlVi7di1iYmLg7e3NaRZTN9UA4OWXX8adO3fw4osvIjo62kbJCDGPXd6TENr6DH5+fujatSsiIiK4jkJIK3ZXJPi085G5pFIpBg4cyPmoh5C22F2RkMvlhkevtVqtYEYThPCV3RWJY8eOgWEYAADDMILYrYkQPrO7ItHyASl6YIoQy9hdkRDq5CRC+MruisSkSZMgEokANM07mDx5MseJCBE2wc2TMDWDUavVGu5JAEBRUVGbMxj5NHuRED6zu5GEVCqFk1NT7ZPJZJBKpRwnIkTYBDeSMGcG49tvv42SkhJ89tlnNPeAEAvZ3UgCoMlJhFiTXRYJQoj1UJEghLCiIkEIYSW4G5d8Eh8fj+LiYov7KSoqAvDwFZzbY8CAAaz9OGpmoeUFbJuZDa+KBN9eXFMvbHFxMX6/nIM+7o0Wnacb0zT5q6HknEX9/KmWmDymuLgYuVdyAS+Th7L7awyaezPXsn5Upg8pLi5GYV4eLFmOp0vz6fLyLOgFMGeZ4uLiYlzJvwovN1+LziVudAEA3Lxm2ZPMqroKi77fZJEoLCxEdnY2JBIJnn76acOyZZ2huLgYufkF0LvJLOpH1Nj0n3XhWsdXnhbXVZl1XB/3Rqwaqe7weaxp/Xl38w70AvTj9Z0bxkzik+Zd8foBWABR54Yxwx4wpg8C4OXmiwmDZnZyGvOcKDxo0fez/oQSEhJw+fJl/O1vf8PYsWNx6dIl7N2716yOa2tr8e677+LatWtmh6mqqgLM/CGwYaRdwEi7mD6QvZe/8hDi2FhHEgsWLIBEIkFNTQ0YhsH06dPR2Gje0LqyshJDhgxhPablruIajcasvm1Fo9GwbqyqUChQfU9i/l/wTnbjngSeCgVr5rq6OhsmMk9dXZ2gMgstL2A6MxvWIiGRSPDJJ58Y/qL6+vrirbfeMqvjwMBAeHp6sh7TcldxPz8/3KwFGkL+YdY5OpNrwVH4+fmx7uTs7Oxsw0TmcXZ2Zs3c0NAAqMwf5nc6FdDQpcFk5tswf6jfmW4B6NFgOq+qrsLiYb61qOoq0KWhh1Hm9hQMk/ckVCoVVq9eDQDYtGlTByK2j7iuCq4FRy3qQ6RtWqLekkuOpnsS7LfKZDIZ3Gqu8eqehKvMsvs5hLRk1qcbCQkJEIlE7b4ceOWVV9p1vLX2RGj+dGNgf0vuh/sJao8Gc8lkMtyov8GrG5cyE4VNJpNB/OefvLlx6WVG3vq7DK9uXJp6jdmwFgmGYfD++++jqKgIEokE/fv37/CJzGGNz4Mf7If2WCDEcqxFIiEhwWjHawCIjY3tzDyEEJ55aJFoaGjAlClTkJmZiYkTJ4JhGBw8yI8bMYQQ23lokdDpdEhOTkZOTg5qa2shEokwaNAgW2YjhPDAQ4uEu7s7oqKicO3aNeTn54NhGKhUZsyh7WSmlq8DzJuWba3l6/5UWz5PolrTdEPO09myj/j+VEsQZFEPhLRm8tONuLg4BAYGIigoCPn5+bbIZDFbLTZjrU8/Sv8qaj0CB1rUTxCsl4mQZiaLhEwmw7PPPguVSmVycpQtmLN8na3QpzHEEZgsEs899xxkMhnS0tLQvXt3W2QipBUFLJtx2TzdzdIJ9ApY/gCt0LAWiT/++ANFRUV47LHHsHz5cmzdutVWuUhnssa07Ia//tfV8izwZz/EGpdQlX9d0gUMtOySzsvMPNaYlt2grQUAuEq7WtSPqq4C/uj4JThrkdiyZQsWLlyItWvXQq1WY/HixR0+kS0plUqsXbsWMTExtBhuC1af1epv2S8d/E1nssZlnS0v6az3Gjc9M+Xfv49F/fjD26JMrEXCx8cHw4cPx/bt27Fjxw54eHh0+ES2JJfLcenSJcjlckRHR3Mdh1foPkrns7fXmHXMqVarUVpaim7dukGlUqG0tNRWuTpMqVQiNTUVDMMgNTUVd+5YtqoPIY6OtUgEBgYiKSkJQUFBSEpKQlJSkq1ydZhcLjds86fX6yGXyzlORIiwsRYJnU6HxsZGNDY2QqfTQafT2SpXh2VkZECr1QJo2hf02LFjHCciRNhY70loNBpUVFRg5MiRGDt2LCQS0wutci0sLAwpKSnQarWQSqWYNGkS15EIETTWkcTKlSsRFxeHAQMG4Pvvvzd7fUsuRUREQCRqmuYsFosRERHBcSJChM3kh+U5OTk4efIkVCoVnnjiCVtksoiPjw+mTJkCkUiEKVOm0EeghFiI9XLj73//O3r37o0JEyaga9euqK2ttVUui0RERKCkpIRGEYRYAWuR+Pjjjw3/XyQSoby8vNMDWYOPjw927NjBdQxC7ALr5caoUaMwatQoaDQayOVyZGZm2ioXIYQnWEcSe/fuxfnz5/HYY4/B1dUVcXFxtspFCOEJ1iJx/fp1yGQyDBo0SDCXGnxjrUVyAOstlENIe7Bebnz00UdYsWIFysvLUVRUhM2bN9sql0Px9vamT2EIb5lcT4JhGAwbNgxTp07FrVu3bJHJrpizSE7zU6tr1qyhYkF4h7VIHDhwAH/88Qf69euHQ4cOoX///ggJCTHZaWZmJhQKBbRaLebNm2etrHYrPj4eFy9eRHx8PNauXct1HEKMsF5unD9/Hh988AFmzpyJtWvXIjc316xOc3JyMGfOHNqV2wxKpRInT54EAJw4cYKeWiW8Y3LD4AdJpVLzOnVq6lavZ99KruWu4o5o165dRu3169cjMjKSozTma945Wyg/P6HlBfiT2eTydStXrgTQdG/ixo0bZnU6ZMgQ7N271+SamC13FXdELUdnFy5cEMxrUlxcDKlUKogVun///XcAwKJFi5CVlcVxGvO4ubkBQKe8H6y2q/ihQ4c6FCAsLKxD30eEo/mX7o033hDML53Q5OXlAQBCQ0M5fY0tXA2VWMrV1ZW1zUfNBaJZcXExR0nMExoaytom7Ex+BEo6V0NDA2ubC6YmgDX/hWv2xhtvYPjw4W0ea4sJYOZMWGvpYRPX+DJhra3CxtVool1FYtu2bXjnnXc6KwshDkNIhc1kkdDpdFAoFGAYBq+++mqnBXFUvXr1Mpry3qtXLw7TNDE1Aayt4TqXKzoLLa/QmCwS0dHRGDBgAMTiptsXUVFRnR7KkQQFBRkVieDgYA7TEFsRUmEzWSQGDx6M+fPnw9nZ2RZ5HM7Zs2eN2mfOnOEoCSFtM1kkrl69itWrVxsmVsXGxnZ6KEcSFhZmtFUBLdxL+MZkkejXrx+WL19uiywOKSIiwqhI0JJ7hG9MFokbN24gKirKMPtr06ZNnR7KkVy/ft2oXVJSQk+CEl4xWSRCQkLw5ptv2iKLQ4qJiTFq/+c//2n3R2OEdCaTRSIzMxNKpdIwkli2bFmnh3IkarWatU0I10wWiS1btqC6uhoA4Onp2emBHI27u7tRYXB3d+cwDSGtmXx248iRI/j222/x7bff4ocffrBFJofS8nJj3bp13AQh5CFMjiRUKhVWr14NgG5adoZRo0YZRhPu7u4YMWIE15EIMWLWU6AJCQnYvXs3NBpNZ+dxSDExMRCLxTSKILxkciSxatUqFBcXg2EYDBw40BaZHM6oUaMMS9gRwjcmi8TatWuhVqsNS9LRjEvra14tOyYmhuZIEN4xWSR8fX2xZs0aW2RxWHK5HJcuXYJcLkd0dDTXcQgxYrJIZGRk4OrVq4YVk+jmpXUplUqkpKSAYRikpKQgIiKCRhOEV1iLhFqtxuHDh43+7d69e/Dw8OjUUI5ELpdDq9UCgGFjZhpNED5hLRJJSUmorq7GwIEDIZFIcOXKFfj5+WH69Om2ymf3jh07ZtROT0+nIkF4hbVING+wk5OTA71ejzlz5kAmk9kqm0Pw8vJCfX29UZsQPmEtEidPnsTRo0cBNO274eLignHjxtkkmKNoub8q7bdK+Ia1SPz000+Ii4sztFetWkVFghAHw1oktFotjhw5gu7du6OystJWmRyKRCJBY2OjUZsQPmGdlv3RRx+he/fuuHv3Lvz9/c2eNpySkoL//ve/Vglo7x4sEG21CeEa60hCoVDg0UcfxaOPPgoAKCsrQ+/evds8dv/+/Th37hwYhsH8+fNRUFBg8uS0YXDbhPiaCC2z0PIC3GVmLRKxsbFwc3NDYGCg4d8etqT+7NmzMXv2bEM7IyPD5Mlpw+C2CfE1EVpmoeUFrJvZahsGb9u2DQcOHMC8efMszUQIESjWexIuLi4dLhArVqzo0PdZg1KpxNKlS3Hnzh3OMhBiL0w+u/GPf/wDAQEBAJruSbzwwguIjIzs9GCWoAemCLEek4vOjB49Grt27cKuXbswbtw43i/UqlQqkZqaCoZhkJqaSqMJQixkskiIRCLs3r0bn332GRoaGqDX622Rq8PkcjkYhgEA6PV6yOVyjhMRImwmLzc+/PBD5OXlQafTYeHChbxfwi4jI8PwVKVWq8WxY8fokoMQC5gcSaxfvx4XLlxAXl4eNm3axPuNg8PCwiCVSgEAUqmU9tYkxEJmLYS7YMECLFy40DCM57OIiAiIRCIAgFgspr01CbGQycsNLy8vxMTEQCQSCWLFJB8fH0yZMgXJycmYMmWKIDITwmesRaK0tBQvvvgi6uvrwTCMYas/vouIiEBJSQmNIgixApMrU7X0sGnZfOLj44MdO3ZwHYMQu8BaJIRQEAghncusG5eEEMdFRYJjLde0fOSRRzhKQkjbqEhw7N69e0btmpoajpIQ0jYqEhxrntPxsDYhXKMiwbFnn33WqD1x4kSOkhDSNioSHFu0aBHE4qYfg1gsxqJFizhORIgxKhIc8/HxQVhYGABg0qRJNEOU8A4VCR6YPn06unbtihkzZnAdhZBWqEjwwA8//IC6ujokJydzHYWQVqhIcIxW0iJ8R0WCY7SSFuE7KhIca2slLUL4hIoEx2glLcJ3VCQ4RitpEb6zepEoKSnBhg0b8NFHH6GiosLa3dud5pW0RCIRraRFeMnk8nXmat4wWKPRYNWqVUhPT4darYavr6+1TmG3aCUtwmdWKxIPbhicnJwMHx8f9OvXj/V7aFfx/7d48WJUVFQIdvQltJ+j0PICPN1VvCOysrKQlJSEESNGoLS0FL17937osbSruDB16dIF9fX1Rm2h/RyFlhfg6a7iHREaGorQ0FBrd0t4pOXWCkLYaoF0HH26wQNC2wV98uTJRu3w8HCOkhBboCLBAw/ugi4E48aNM2o/88wzHCUhtkBFgmNCfHZj+/btRu0tW7ZwlITYAhUJjgnx2Y2ysjLWNrEvVCQ4JsRnN2hdTsdCRYJjQnx2Y/To0UbtMWPGcJSE2AIVCY4J8dkNFxcX1jaxL1QkOCbEZzd+/vlno3ZWVhZHSYgtWH0yFWk/oT27QfckHAuNJHigeRd0IYwiANorxNFQkSDtRnuFOBYqEqTdaK8Qx0JFgnTIokWLMGzYMEGMIt555x2j9ooVKzhKIkxUJEiHCOk+SstnTcaOHctREmGiIkHsnlwuh0QiAQBIJBJBTH3nEyoSxO5lZGSgsbERANDY2CiIqe98QkWC2D0hTn13d3dnbdsSFQli94Q49T0mJsaovW7dOm6CgIoEcQBCnPo+atQow+jB3d0dI0aM4CwLFQniECIiIvD4448LYhTRLCYmBmKxmNNRBACIGA5XMb169aogVy0mROja87tHIwlCCCsqEoQQVpw+Kk47eBHCjfv375t9LKf3JAgh/EeXG4QQVlQkCCGsqEgQQlhRkSCEsKIiQQhhJYlp+SSJjdy7dw8ZGRkICgoCAHz11Vc4c+YMfvnlF7i4uODw4cOIj4/HqFGj0K1bNwDAqVOncPjwYfTs2RP/+9//cOLECQQFBSExMRHZ2dnw9fWFTCZrda4H+5ZIJDh48CDS09MxevRoODk54eeff8aPP/6IrVu3Ijw8HPHx8Th//jz8/f0hl8uRk5ODuro6fPLJJygtLcWTTz5p1X4DAwNbZdZqtYiLi4Ner4enpyd27NiBkydPIjg4GO7u7mAYBnK5HIWFhQgMDERCQgJOnDiBkJAQfPnll8jPz8fdu3exc+dOo8y1tbV4//33MWDAANTU1EAulyM9PR1jxoyBk5OT0XmdnJywf/9+pKenY+zYsYiPj8fFixfh7OyMHj16tMr8YN9lZWVISUnBiRMnDJv3pKam4vvvv8fNmzfh6emJffv2ISsrC0899RRUKhXefvttTJo0CVFRURgzZgx++uknw/vDGn2/9NJLbb4X09LScPr0aZw4ccKw8dChQ4eQl5eHs2fPora21vBzHDp0qOG1Hjx4MFauXInnnnvO5PvbknN07dq1zdybN29GY2MjHn30UQDApk2bcOXKFZSWlmLQoEHYv38/Ll++jPr6ehw4cABPP/10m/2YwtlIIikpCaGhoYZ2WFgY3njjDVRXV2P48OF4/vnnERYWhoCAAABAQUEBGIaBXq+Ht7c3oqOj0adPH9TX12POnDmorKw0LCzS0oN99+3bF8uXL4eHhwc0Gg2AppWLRo8ejblz5+Ls2bNobGyESCTCI488AqVSiZs3b8LDwwMAMHLkyE7pt6WqqioMGjQIQNOTiwqFAkqlEm5ubgAAtVqNfv36gWEYuLu7Y9q0aVAqlaivr8fFixdx79499OnTp1XmyspKDBkyBEDTAiy3bt1CdXW14VHqB8/r7++PiRMnQqVSobi4GLdv38b9+/fh5+fXZuYH+3788ccxfPhwqFQqw9cLCwvh6+sLsViMtLQ0zJ49G0OHDsWVK1dw+PBhhISEwM3NDUOGDMHRo0eN3h/W6PthGIZBWVmZYbtFoKkolZeXQywWG/0ce/XqZXjvqdVq9O3bt80+W76/LTlHW1JSUgzFodm8efPw4osvory8HL/88gs8PT0BND0sZgnOioRCoUC3bt1w4MABfPvtt/Dw8MDWrVsxf/58AE0v8ssvv2w4/qeffkJJSQkuX74MJycnHDt2DD4+PujZsyfUajXeeeedVouJtNW3t7c3EhMTMXr0aMMIBWhamGTy5MnQarV48sknMW7cOGRmZiIgIABr1qzB8ePHUVNTA3d3d6v3+6Aff/wRX3zxBXr06GH4Zbx69SqmT5+OuXPnIjs7GwDg4eGBfv36AWj6xXZxccGcOXNw8eJF9OzZE4sXL8Y333zTKnNgYKDhzXP+/HksWbIEEyZMwOXLlwHA6Lzl5eXw9/fH+PHj0dDQgODgYMyaNQuHDx9u83V+sO9r167hiSeeQEBAAJRKJQBg1qxZiIiIwG+//QaxWGxYcTs7OxsajQYFBQXIz88HANy+fdvo/WHNvlvmLioqwgcffACdTmfYvLmmpgb/+te/oFAojH6OHh4ehvde8+vflpbvb2udIyEhASkpKbhw4QIKCwtx7tw5w9caGhrw5ZdfYsGCBcjOzkZxcbHR1zuKsxmXPj4+qKqqwqxZswAA7777Lnx9fXH8+HHMmzcP9fX1hr+y6enpiIqKAgAolUqcO3cOSUlJGDFiBBQKBRITE+Hh4YHw8HCkp6dj8uTJANBm3y4uLrhw4QLUajVCQkJw9uxZTJ48GWKxGBKJBKNHj8auXbsAAHPnzsXu3buxZ88ejBw5EgqFAtXV1Vbv98HMbQ1de/bsiX379kEqlWLOnDlGxzfbuXMnXF1d8frrr6OoqAh79uzBmDFjUFlZaZT5QYGBgTh48CA0Gg3Gjx/fqt/6+nrs27cPIpEI48ePR1paGr766itMnDixzdf5QVVVVUhOToZer4eLiwt++eUXlJWVobq6GjKZDBMnTsQXX3wBkUiEJUuWQCqVoq6uDkOHDsXJkyfh7e1t9P6wVt9t5T548CASEhLg5uaG8vJy3L17F1KpFHv27EHPnj0BwPBzzMrKMrz3SktLW2Vr1vL9bck5CgoKDJkjIyMBAFOnTsWZM2dw//59FBQUwNXVFStWrMCzzz6LrKwsvPfeeygrK2u121qHMBypra1lvvvuO6v3m5OTY/U+m61YsYL59NNPrd4vZf5/tbW1TGRkJFNVVcX790dFRQWzePHiNr9mzfe3pZnPnDnDbN68ucPfT9OyCSGs6CNQQggrKhKEEFZUJAghrKhI2LnTp0/jrbfewu3bt43+/d///jd0Op2hvWLFClRWVuKbb76xdUTCc5wuOkM636FDh+Dl5YXy8nLDTMolS5YAaJrgs379eohEIpSWluL+/fu4efMmduzYgfr6elRXV2PatGlwcnKCXC6Hs7MzevXqhaVLlwIA8vLykJSUhIaGBsyePRuHDh2Cr68vSkpKsHTpUly/fh3Z2dlQq9VYunQpNm/ejK5du2LmzJn4/PPP4ePjA7VajW7dumHhwoX4888/UVZW9tCZkYQbVCTs3Lhx4+Dn54evv/4aGzZsQE1NDT799FMATZOSfH19ERkZafjFb/bCCy/A1dUVR48ehUKhwMaNG1FQUIBTp06cKIIEAAABeklEQVQZjklMTISfn59hjohGo8G8efNw8eJF5ObmIjU1FQMHDoRer0deXh60Wi02bNiAuLg4fPjhh5BIJNi0aRNeeeUVJCcnQ6FQIDo62qavDzGNioSDePCT7rY+9W6eodisS5cukEgkYBjGMJW4eYObZo2NjXjrrbfQ0NCAoqIiFBYWwtXVFU5OTtDr9QCA6Oho/PHHH2hoaDBMKdfpdEZ9DRo0CPv27YOHh4fhGMIfVCQcxIwZM7B69WqIRCIsWrQIn376Kfr3748jR44gNjbWML25LdOnT8eqVaug1WoRHBxs+PfXXnsNq1atgk6nw7Jly1p937Rp07By5UrU19cb7Ug1ffp0fPzxx3B3d4ezszMAICAgAE899ZT1/oOJ1dBkKmLS8ePHcfHiRahUKixcuLDVg0XtlZOTg8zMTKjVarz88stQqVRIT09HbGyslRITa6IiQQhhRR+BEkJYUZEghLCiIkEIYUVFghDCiooEIYQVFQlCCKv/AwZVcJbSojTCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 288x216 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for promoter in ['weak', 'strong']:\n",
    "\n",
    "    damFoldDf = damDf5[damDf5['promoter'] == promoter]\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(4, 3))\n",
    "\n",
    "    xlabel = 'folding_energy_bucket'\n",
    "    ylabel = 'log10DAMratio'\n",
    "    sns.boxplot(ax=ax, data=damFoldDf, x=xlabel, y=ylabel)\n",
    "\n",
    "    ax.set_title('{} promoter'.format(promoter))\n",
    "    ax.set_xlabel('folding energy')\n",
    "    ax.set_ylabel('log10(norm. DAMRatio)')\n",
    "\n",
    "    filename = 'ELM_seq_folding_energy_boxplot_promoter_{}_regionLength54'.format(promoter)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.png'), dpi=300)\n",
    "    fig.savefig(ctermELMseqPath / subFolder / (filename + '.svg'), dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the amino acid is the main determinant, then the synonymous codon choice could be related to the secondary structure.\n",
    "\n",
    "Anyway, the folding energy change is very small upon 1 nucleotide (synonymous codon)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>seq</th>\n",
       "      <th>Uncut_cnt</th>\n",
       "      <th>DpnI_cnt</th>\n",
       "      <th>MboI_cnt</th>\n",
       "      <th>DpnI_ratio</th>\n",
       "      <th>MboI_ratio</th>\n",
       "      <th>promoter</th>\n",
       "      <th>DAMratio</th>\n",
       "      <th>codon_-2</th>\n",
       "      <th>...</th>\n",
       "      <th>n_5</th>\n",
       "      <th>n_6</th>\n",
       "      <th>n_7</th>\n",
       "      <th>n_8</th>\n",
       "      <th>seq_extended</th>\n",
       "      <th>contains_GATC</th>\n",
       "      <th>region_seq</th>\n",
       "      <th>folding_energy</th>\n",
       "      <th>folding_struct</th>\n",
       "      <th>folding_energy_bucket</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.052750</td>\n",
       "      <td>GACGGATAAATTAAA</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>129</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>weak</td>\n",
       "      <td>112.914694</td>\n",
       "      <td>GAC</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAGACGGATAAATTAAACAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAGACGGATAAATTAAAC...</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>....((((.(((((((...((((.......).))).......))))...</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.825298</td>\n",
       "      <td>CACCGGTAAAGGTGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>weak</td>\n",
       "      <td>66.880242</td>\n",
       "      <td>CAC</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACACCGGTAAAGGTGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...</td>\n",
       "      <td>-12.5</td>\n",
       "      <td>..(((.((.((..((((......))))..)).))))).(((((......</td>\n",
       "      <td>(-13.504, -8.872]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.825298</td>\n",
       "      <td>CACCGGTAAAGGTGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>1.685342e-08</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>weak</td>\n",
       "      <td>66.880242</td>\n",
       "      <td>CAC</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACACCGGTAAAGGTGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>....(((((((..((((......))))...........)))))))....</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.794629</td>\n",
       "      <td>GGTTTCTAATCGCGT</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>286</td>\n",
       "      <td>6.741369e-08</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>weak</td>\n",
       "      <td>62.320226</td>\n",
       "      <td>GGT</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAGGTTTCTAATCGCGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...</td>\n",
       "      <td>-9.2</td>\n",
       "      <td>....((((.(((........(((((..............))))).....</td>\n",
       "      <td>(-13.504, -8.872]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.794629</td>\n",
       "      <td>GGTTTCTAATCGCGT</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>286</td>\n",
       "      <td>6.741369e-08</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>weak</td>\n",
       "      <td>62.320226</td>\n",
       "      <td>GGT</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAGGTTTCTAATCGCGTCAG</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>...((..((((.((((......(........).....)))).))))...</td>\n",
       "      <td>(-8.872, -4.24]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   log10DAMratio              seq  Uncut_cnt  DpnI_cnt  MboI_cnt  \\\n",
       "0       2.052750  GACGGATAAATTAAA          2         0       129   \n",
       "1       1.825298  CACCGGTAAAGGTGT          0         0        76   \n",
       "1       1.825298  CACCGGTAAAGGTGT          0         0        76   \n",
       "2       1.794629  GGTTTCTAATCGCGT          3         3       286   \n",
       "2       1.794629  GGTTTCTAATCGCGT          3         3       286   \n",
       "\n",
       "     DpnI_ratio  MboI_ratio promoter    DAMratio codon_-2  \\\n",
       "0  1.685342e-08    0.000002     weak  112.914694      GAC   \n",
       "1  1.685342e-08    0.000001     weak   66.880242      CAC   \n",
       "1  1.685342e-08    0.000001     weak   66.880242      CAC   \n",
       "2  6.741369e-08    0.000004     weak   62.320226      GGT   \n",
       "2  6.741369e-08    0.000004     weak   62.320226      GGT   \n",
       "\n",
       "           ...          n_5 n_6 n_7 n_8           seq_extended contains_GATC  \\\n",
       "0          ...            T   A   A   A  AAAGACGGATAAATTAAACAG         False   \n",
       "1          ...            G   T   G   T  AAACACCGGTAAAGGTGTCAG         False   \n",
       "1          ...            G   T   G   T  AAACACCGGTAAAGGTGTCAG         False   \n",
       "2          ...            G   C   G   T  AAAGGTTTCTAATCGCGTCAG         False   \n",
       "2          ...            G   C   G   T  AAAGGTTTCTAATCGCGTCAG         False   \n",
       "\n",
       "                                          region_seq folding_energy  \\\n",
       "0  AAACCAGGAGTCGTTTCACCCGCGAAAAAAGACGGATAAATTAAAC...           -8.4   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...          -12.5   \n",
       "1  AAACCAGGAGTCGTTTCACCCGCGAAAAAACACCGGTAAAGGTGTC...           -6.5   \n",
       "2  AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...           -9.2   \n",
       "2  AAACCAGGAGTCGTTTCACCCGCGAAAAAAGGTTTCTAATCGCGTC...           -5.0   \n",
       "\n",
       "                                      folding_struct folding_energy_bucket  \n",
       "0  ....((((.(((((((...((((.......).))).......))))...       (-8.872, -4.24]  \n",
       "1  ..(((.((.((..((((......))))..)).))))).(((((......     (-13.504, -8.872]  \n",
       "1  ....(((((((..((((......))))...........)))))))....       (-8.872, -4.24]  \n",
       "2  ....((((.(((........(((((..............))))).....     (-13.504, -8.872]  \n",
       "2  ...((..((((.((((......(........).....)))).))))...       (-8.872, -4.24]  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoter = 'weak'\n",
    "damFoldDf = damDf5[damDf5['promoter'] == promoter]\n",
    "damFoldDf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>folding_energy</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>aa</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>codon_-1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGA</th>\n",
       "      <td>-7.031174</td>\n",
       "      <td>0.117257</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGG</th>\n",
       "      <td>-8.448908</td>\n",
       "      <td>0.112235</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGA</th>\n",
       "      <td>-8.059680</td>\n",
       "      <td>0.126879</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGC</th>\n",
       "      <td>-9.040299</td>\n",
       "      <td>0.109238</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGG</th>\n",
       "      <td>-9.405348</td>\n",
       "      <td>0.124067</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGT</th>\n",
       "      <td>-8.411085</td>\n",
       "      <td>0.110119</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>-6.818759</td>\n",
       "      <td>0.176716</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>-7.145855</td>\n",
       "      <td>0.189810</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAC</th>\n",
       "      <td>-7.224731</td>\n",
       "      <td>0.086105</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAT</th>\n",
       "      <td>-7.213986</td>\n",
       "      <td>0.076400</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAC</th>\n",
       "      <td>-7.764309</td>\n",
       "      <td>0.132475</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAT</th>\n",
       "      <td>-7.499135</td>\n",
       "      <td>0.129465</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAA</th>\n",
       "      <td>-7.442969</td>\n",
       "      <td>0.120992</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAG</th>\n",
       "      <td>-7.893703</td>\n",
       "      <td>0.128402</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGC</th>\n",
       "      <td>-7.639914</td>\n",
       "      <td>0.059098</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGT</th>\n",
       "      <td>-7.133666</td>\n",
       "      <td>0.073754</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCA</th>\n",
       "      <td>-7.251367</td>\n",
       "      <td>0.047420</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCC</th>\n",
       "      <td>-8.011850</td>\n",
       "      <td>0.050991</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCG</th>\n",
       "      <td>-9.701874</td>\n",
       "      <td>0.058918</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCT</th>\n",
       "      <td>-7.473017</td>\n",
       "      <td>0.050396</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>-6.854227</td>\n",
       "      <td>0.062007</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>-6.948114</td>\n",
       "      <td>0.066979</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACG</th>\n",
       "      <td>-7.772748</td>\n",
       "      <td>0.057062</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACT</th>\n",
       "      <td>-6.987373</td>\n",
       "      <td>0.056351</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>-6.820962</td>\n",
       "      <td>0.094960</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAT</th>\n",
       "      <td>-6.770742</td>\n",
       "      <td>0.089334</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAA</th>\n",
       "      <td>-7.130419</td>\n",
       "      <td>0.070778</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAG</th>\n",
       "      <td>-7.499543</td>\n",
       "      <td>0.071279</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGC</th>\n",
       "      <td>-8.266496</td>\n",
       "      <td>-0.124071</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT</th>\n",
       "      <td>-7.752825</td>\n",
       "      <td>-0.127701</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGG</th>\n",
       "      <td>-11.143907</td>\n",
       "      <td>0.066998</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGT</th>\n",
       "      <td>-9.276683</td>\n",
       "      <td>0.065466</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCA</th>\n",
       "      <td>-7.402899</td>\n",
       "      <td>-0.009470</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCC</th>\n",
       "      <td>-7.557002</td>\n",
       "      <td>0.011770</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCG</th>\n",
       "      <td>-8.298247</td>\n",
       "      <td>0.003266</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCT</th>\n",
       "      <td>-7.760669</td>\n",
       "      <td>-0.011102</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA</th>\n",
       "      <td>-8.223831</td>\n",
       "      <td>-0.046712</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCC</th>\n",
       "      <td>-8.825812</td>\n",
       "      <td>-0.031086</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCG</th>\n",
       "      <td>-9.878484</td>\n",
       "      <td>-0.056845</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCT</th>\n",
       "      <td>-8.364808</td>\n",
       "      <td>-0.045332</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTA</th>\n",
       "      <td>-7.760370</td>\n",
       "      <td>-0.064119</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTC</th>\n",
       "      <td>-7.956008</td>\n",
       "      <td>-0.039379</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTG</th>\n",
       "      <td>-9.210331</td>\n",
       "      <td>-0.035104</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTT</th>\n",
       "      <td>-7.894414</td>\n",
       "      <td>-0.054824</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATA</th>\n",
       "      <td>-6.588586</td>\n",
       "      <td>-0.113839</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATC</th>\n",
       "      <td>-6.618100</td>\n",
       "      <td>-0.103730</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATT</th>\n",
       "      <td>-6.724692</td>\n",
       "      <td>-0.107362</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTA</th>\n",
       "      <td>-7.350248</td>\n",
       "      <td>-0.137830</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTC</th>\n",
       "      <td>-7.892456</td>\n",
       "      <td>-0.123785</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTG</th>\n",
       "      <td>-8.780064</td>\n",
       "      <td>-0.125346</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTT</th>\n",
       "      <td>-8.024942</td>\n",
       "      <td>-0.137120</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTA</th>\n",
       "      <td>-6.817202</td>\n",
       "      <td>-0.163402</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTG</th>\n",
       "      <td>-8.703653</td>\n",
       "      <td>-0.144427</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATG</th>\n",
       "      <td>-7.081958</td>\n",
       "      <td>-0.110384</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC</th>\n",
       "      <td>-7.544201</td>\n",
       "      <td>-0.075927</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT</th>\n",
       "      <td>-7.216686</td>\n",
       "      <td>-0.083180</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGA</th>\n",
       "      <td>-7.747629</td>\n",
       "      <td>-0.061969</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGG</th>\n",
       "      <td>-9.743693</td>\n",
       "      <td>-0.021271</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAC</th>\n",
       "      <td>-6.858673</td>\n",
       "      <td>-0.004837</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAT</th>\n",
       "      <td>-6.839183</td>\n",
       "      <td>-0.000669</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          folding_energy  log10DAMratio aa\n",
       "codon_-1                                  \n",
       "AGA            -7.031174       0.117257  R\n",
       "AGG            -8.448908       0.112235  R\n",
       "CGA            -8.059680       0.126879  R\n",
       "CGC            -9.040299       0.109238  R\n",
       "CGG            -9.405348       0.124067  R\n",
       "CGT            -8.411085       0.110119  R\n",
       "AAA            -6.818759       0.176716  K\n",
       "AAG            -7.145855       0.189810  K\n",
       "CAC            -7.224731       0.086105  H\n",
       "CAT            -7.213986       0.076400  H\n",
       "GAC            -7.764309       0.132475  D\n",
       "GAT            -7.499135       0.129465  D\n",
       "GAA            -7.442969       0.120992  E\n",
       "GAG            -7.893703       0.128402  E\n",
       "AGC            -7.639914       0.059098  S\n",
       "AGT            -7.133666       0.073754  S\n",
       "TCA            -7.251367       0.047420  S\n",
       "TCC            -8.011850       0.050991  S\n",
       "TCG            -9.701874       0.058918  S\n",
       "TCT            -7.473017       0.050396  S\n",
       "ACA            -6.854227       0.062007  T\n",
       "ACC            -6.948114       0.066979  T\n",
       "ACG            -7.772748       0.057062  T\n",
       "ACT            -6.987373       0.056351  T\n",
       "AAC            -6.820962       0.094960  N\n",
       "AAT            -6.770742       0.089334  N\n",
       "CAA            -7.130419       0.070778  Q\n",
       "CAG            -7.499543       0.071279  Q\n",
       "TGC            -8.266496      -0.124071  C\n",
       "TGT            -7.752825      -0.127701  C\n",
       "...                  ...            ... ..\n",
       "GGG           -11.143907       0.066998  G\n",
       "GGT            -9.276683       0.065466  G\n",
       "CCA            -7.402899      -0.009470  P\n",
       "CCC            -7.557002       0.011770  P\n",
       "CCG            -8.298247       0.003266  P\n",
       "CCT            -7.760669      -0.011102  P\n",
       "GCA            -8.223831      -0.046712  A\n",
       "GCC            -8.825812      -0.031086  A\n",
       "GCG            -9.878484      -0.056845  A\n",
       "GCT            -8.364808      -0.045332  A\n",
       "GTA            -7.760370      -0.064119  V\n",
       "GTC            -7.956008      -0.039379  V\n",
       "GTG            -9.210331      -0.035104  V\n",
       "GTT            -7.894414      -0.054824  V\n",
       "ATA            -6.588586      -0.113839  I\n",
       "ATC            -6.618100      -0.103730  I\n",
       "ATT            -6.724692      -0.107362  I\n",
       "CTA            -7.350248      -0.137830  L\n",
       "CTC            -7.892456      -0.123785  L\n",
       "CTG            -8.780064      -0.125346  L\n",
       "CTT            -8.024942      -0.137120  L\n",
       "TTA            -6.817202      -0.163402  L\n",
       "TTG            -8.703653      -0.144427  L\n",
       "ATG            -7.081958      -0.110384  M\n",
       "TTC            -7.544201      -0.075927  F\n",
       "TTT            -7.216686      -0.083180  F\n",
       "TGA            -7.747629      -0.061969  W\n",
       "TGG            -9.743693      -0.021271  W\n",
       "TAC            -6.858673      -0.004837  Y\n",
       "TAT            -6.839183      -0.000669  Y\n",
       "\n",
       "[62 rows x 3 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfList = []\n",
    "for aa in aaTable:\n",
    "    damFoldDf2 = damFoldDf[damFoldDf['aa_-1'] == aa]\n",
    "    df = damFoldDf2.groupby('codon_-1').mean()[['folding_energy', 'log10DAMratio']]\n",
    "    df['aa'] = aa\n",
    "    dfList.append(df)\n",
    "    \n",
    "damFoldSyn = pd.concat(dfList)\n",
    "damFoldSyn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>folding_energy</th>\n",
       "      <th>log10DAMratio</th>\n",
       "      <th>aa</th>\n",
       "      <th>log10DAMratio_norm_syn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>codon_-1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA</th>\n",
       "      <td>-8.223831</td>\n",
       "      <td>-0.046712</td>\n",
       "      <td>A</td>\n",
       "      <td>1.038196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCC</th>\n",
       "      <td>-8.825812</td>\n",
       "      <td>-0.031086</td>\n",
       "      <td>A</td>\n",
       "      <td>0.690886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCG</th>\n",
       "      <td>-9.878484</td>\n",
       "      <td>-0.056845</td>\n",
       "      <td>A</td>\n",
       "      <td>1.263394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCT</th>\n",
       "      <td>-8.364808</td>\n",
       "      <td>-0.045332</td>\n",
       "      <td>A</td>\n",
       "      <td>1.007524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGC</th>\n",
       "      <td>-8.266496</td>\n",
       "      <td>-0.124071</td>\n",
       "      <td>C</td>\n",
       "      <td>0.985582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGT</th>\n",
       "      <td>-7.752825</td>\n",
       "      <td>-0.127701</td>\n",
       "      <td>C</td>\n",
       "      <td>1.014418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAC</th>\n",
       "      <td>-7.764309</td>\n",
       "      <td>0.132475</td>\n",
       "      <td>D</td>\n",
       "      <td>1.011492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAT</th>\n",
       "      <td>-7.499135</td>\n",
       "      <td>0.129465</td>\n",
       "      <td>D</td>\n",
       "      <td>0.988508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAA</th>\n",
       "      <td>-7.442969</td>\n",
       "      <td>0.120992</td>\n",
       "      <td>E</td>\n",
       "      <td>0.970289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAG</th>\n",
       "      <td>-7.893703</td>\n",
       "      <td>0.128402</td>\n",
       "      <td>E</td>\n",
       "      <td>1.029711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC</th>\n",
       "      <td>-7.544201</td>\n",
       "      <td>-0.075927</td>\n",
       "      <td>F</td>\n",
       "      <td>0.954411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTT</th>\n",
       "      <td>-7.216686</td>\n",
       "      <td>-0.083180</td>\n",
       "      <td>F</td>\n",
       "      <td>1.045589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGA</th>\n",
       "      <td>-8.630133</td>\n",
       "      <td>0.045989</td>\n",
       "      <td>G</td>\n",
       "      <td>0.769949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGC</th>\n",
       "      <td>-9.170678</td>\n",
       "      <td>0.060468</td>\n",
       "      <td>G</td>\n",
       "      <td>1.012353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGG</th>\n",
       "      <td>-11.143907</td>\n",
       "      <td>0.066998</td>\n",
       "      <td>G</td>\n",
       "      <td>1.121673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGT</th>\n",
       "      <td>-9.276683</td>\n",
       "      <td>0.065466</td>\n",
       "      <td>G</td>\n",
       "      <td>1.096025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAC</th>\n",
       "      <td>-7.224731</td>\n",
       "      <td>0.086105</td>\n",
       "      <td>H</td>\n",
       "      <td>1.059724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAT</th>\n",
       "      <td>-7.213986</td>\n",
       "      <td>0.076400</td>\n",
       "      <td>H</td>\n",
       "      <td>0.940276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATA</th>\n",
       "      <td>-6.588586</td>\n",
       "      <td>-0.113839</td>\n",
       "      <td>I</td>\n",
       "      <td>1.051041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATC</th>\n",
       "      <td>-6.618100</td>\n",
       "      <td>-0.103730</td>\n",
       "      <td>I</td>\n",
       "      <td>0.957710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATT</th>\n",
       "      <td>-6.724692</td>\n",
       "      <td>-0.107362</td>\n",
       "      <td>I</td>\n",
       "      <td>0.991248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAA</th>\n",
       "      <td>-6.818759</td>\n",
       "      <td>0.176716</td>\n",
       "      <td>K</td>\n",
       "      <td>0.964277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG</th>\n",
       "      <td>-7.145855</td>\n",
       "      <td>0.189810</td>\n",
       "      <td>K</td>\n",
       "      <td>1.035723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTA</th>\n",
       "      <td>-7.350248</td>\n",
       "      <td>-0.137830</td>\n",
       "      <td>L</td>\n",
       "      <td>0.994072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTC</th>\n",
       "      <td>-7.892456</td>\n",
       "      <td>-0.123785</td>\n",
       "      <td>L</td>\n",
       "      <td>0.892775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTG</th>\n",
       "      <td>-8.780064</td>\n",
       "      <td>-0.125346</td>\n",
       "      <td>L</td>\n",
       "      <td>0.904034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTT</th>\n",
       "      <td>-8.024942</td>\n",
       "      <td>-0.137120</td>\n",
       "      <td>L</td>\n",
       "      <td>0.988957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTA</th>\n",
       "      <td>-6.817202</td>\n",
       "      <td>-0.163402</td>\n",
       "      <td>L</td>\n",
       "      <td>1.178505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTG</th>\n",
       "      <td>-8.703653</td>\n",
       "      <td>-0.144427</td>\n",
       "      <td>L</td>\n",
       "      <td>1.041657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATG</th>\n",
       "      <td>-7.081958</td>\n",
       "      <td>-0.110384</td>\n",
       "      <td>M</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCA</th>\n",
       "      <td>-7.402899</td>\n",
       "      <td>-0.009470</td>\n",
       "      <td>P</td>\n",
       "      <td>6.843780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCC</th>\n",
       "      <td>-7.557002</td>\n",
       "      <td>0.011770</td>\n",
       "      <td>P</td>\n",
       "      <td>-8.506209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCG</th>\n",
       "      <td>-8.298247</td>\n",
       "      <td>0.003266</td>\n",
       "      <td>P</td>\n",
       "      <td>-2.360591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCT</th>\n",
       "      <td>-7.760669</td>\n",
       "      <td>-0.011102</td>\n",
       "      <td>P</td>\n",
       "      <td>8.023020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAA</th>\n",
       "      <td>-7.130419</td>\n",
       "      <td>0.070778</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.996468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAG</th>\n",
       "      <td>-7.499543</td>\n",
       "      <td>0.071279</td>\n",
       "      <td>Q</td>\n",
       "      <td>1.003532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGA</th>\n",
       "      <td>-7.031174</td>\n",
       "      <td>0.117257</td>\n",
       "      <td>R</td>\n",
       "      <td>1.005351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGG</th>\n",
       "      <td>-8.448908</td>\n",
       "      <td>0.112235</td>\n",
       "      <td>R</td>\n",
       "      <td>0.962297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGA</th>\n",
       "      <td>-8.059680</td>\n",
       "      <td>0.126879</td>\n",
       "      <td>R</td>\n",
       "      <td>1.087855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGC</th>\n",
       "      <td>-9.040299</td>\n",
       "      <td>0.109238</td>\n",
       "      <td>R</td>\n",
       "      <td>0.936601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGG</th>\n",
       "      <td>-9.405348</td>\n",
       "      <td>0.124067</td>\n",
       "      <td>R</td>\n",
       "      <td>1.063743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGT</th>\n",
       "      <td>-8.411085</td>\n",
       "      <td>0.110119</td>\n",
       "      <td>R</td>\n",
       "      <td>0.944153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGC</th>\n",
       "      <td>-7.639914</td>\n",
       "      <td>0.059098</td>\n",
       "      <td>S</td>\n",
       "      <td>1.041135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGT</th>\n",
       "      <td>-7.133666</td>\n",
       "      <td>0.073754</td>\n",
       "      <td>S</td>\n",
       "      <td>1.299335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCA</th>\n",
       "      <td>-7.251367</td>\n",
       "      <td>0.047420</td>\n",
       "      <td>S</td>\n",
       "      <td>0.835411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCC</th>\n",
       "      <td>-8.011850</td>\n",
       "      <td>0.050991</td>\n",
       "      <td>S</td>\n",
       "      <td>0.898317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCG</th>\n",
       "      <td>-9.701874</td>\n",
       "      <td>0.058918</td>\n",
       "      <td>S</td>\n",
       "      <td>1.037969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCT</th>\n",
       "      <td>-7.473017</td>\n",
       "      <td>0.050396</td>\n",
       "      <td>S</td>\n",
       "      <td>0.887833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACA</th>\n",
       "      <td>-6.854227</td>\n",
       "      <td>0.062007</td>\n",
       "      <td>T</td>\n",
       "      <td>1.023216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC</th>\n",
       "      <td>-6.948114</td>\n",
       "      <td>0.066979</td>\n",
       "      <td>T</td>\n",
       "      <td>1.105269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACG</th>\n",
       "      <td>-7.772748</td>\n",
       "      <td>0.057062</td>\n",
       "      <td>T</td>\n",
       "      <td>0.941624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACT</th>\n",
       "      <td>-6.987373</td>\n",
       "      <td>0.056351</td>\n",
       "      <td>T</td>\n",
       "      <td>0.929891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTA</th>\n",
       "      <td>-7.760370</td>\n",
       "      <td>-0.064119</td>\n",
       "      <td>V</td>\n",
       "      <td>1.325967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTC</th>\n",
       "      <td>-7.956008</td>\n",
       "      <td>-0.039379</td>\n",
       "      <td>V</td>\n",
       "      <td>0.814346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTG</th>\n",
       "      <td>-9.210331</td>\n",
       "      <td>-0.035104</td>\n",
       "      <td>V</td>\n",
       "      <td>0.725937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTT</th>\n",
       "      <td>-7.894414</td>\n",
       "      <td>-0.054824</td>\n",
       "      <td>V</td>\n",
       "      <td>1.133750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGA</th>\n",
       "      <td>-7.747629</td>\n",
       "      <td>-0.061969</td>\n",
       "      <td>W</td>\n",
       "      <td>1.488929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGG</th>\n",
       "      <td>-9.743693</td>\n",
       "      <td>-0.021271</td>\n",
       "      <td>W</td>\n",
       "      <td>0.511071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAC</th>\n",
       "      <td>-6.858673</td>\n",
       "      <td>-0.004837</td>\n",
       "      <td>Y</td>\n",
       "      <td>1.757137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAT</th>\n",
       "      <td>-6.839183</td>\n",
       "      <td>-0.000669</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.242863</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          folding_energy  log10DAMratio aa  log10DAMratio_norm_syn\n",
       "codon_-1                                                          \n",
       "GCA            -8.223831      -0.046712  A                1.038196\n",
       "GCC            -8.825812      -0.031086  A                0.690886\n",
       "GCG            -9.878484      -0.056845  A                1.263394\n",
       "GCT            -8.364808      -0.045332  A                1.007524\n",
       "TGC            -8.266496      -0.124071  C                0.985582\n",
       "TGT            -7.752825      -0.127701  C                1.014418\n",
       "GAC            -7.764309       0.132475  D                1.011492\n",
       "GAT            -7.499135       0.129465  D                0.988508\n",
       "GAA            -7.442969       0.120992  E                0.970289\n",
       "GAG            -7.893703       0.128402  E                1.029711\n",
       "TTC            -7.544201      -0.075927  F                0.954411\n",
       "TTT            -7.216686      -0.083180  F                1.045589\n",
       "GGA            -8.630133       0.045989  G                0.769949\n",
       "GGC            -9.170678       0.060468  G                1.012353\n",
       "GGG           -11.143907       0.066998  G                1.121673\n",
       "GGT            -9.276683       0.065466  G                1.096025\n",
       "CAC            -7.224731       0.086105  H                1.059724\n",
       "CAT            -7.213986       0.076400  H                0.940276\n",
       "ATA            -6.588586      -0.113839  I                1.051041\n",
       "ATC            -6.618100      -0.103730  I                0.957710\n",
       "ATT            -6.724692      -0.107362  I                0.991248\n",
       "AAA            -6.818759       0.176716  K                0.964277\n",
       "AAG            -7.145855       0.189810  K                1.035723\n",
       "CTA            -7.350248      -0.137830  L                0.994072\n",
       "CTC            -7.892456      -0.123785  L                0.892775\n",
       "CTG            -8.780064      -0.125346  L                0.904034\n",
       "CTT            -8.024942      -0.137120  L                0.988957\n",
       "TTA            -6.817202      -0.163402  L                1.178505\n",
       "TTG            -8.703653      -0.144427  L                1.041657\n",
       "ATG            -7.081958      -0.110384  M                1.000000\n",
       "...                  ...            ... ..                     ...\n",
       "CCA            -7.402899      -0.009470  P                6.843780\n",
       "CCC            -7.557002       0.011770  P               -8.506209\n",
       "CCG            -8.298247       0.003266  P               -2.360591\n",
       "CCT            -7.760669      -0.011102  P                8.023020\n",
       "CAA            -7.130419       0.070778  Q                0.996468\n",
       "CAG            -7.499543       0.071279  Q                1.003532\n",
       "AGA            -7.031174       0.117257  R                1.005351\n",
       "AGG            -8.448908       0.112235  R                0.962297\n",
       "CGA            -8.059680       0.126879  R                1.087855\n",
       "CGC            -9.040299       0.109238  R                0.936601\n",
       "CGG            -9.405348       0.124067  R                1.063743\n",
       "CGT            -8.411085       0.110119  R                0.944153\n",
       "AGC            -7.639914       0.059098  S                1.041135\n",
       "AGT            -7.133666       0.073754  S                1.299335\n",
       "TCA            -7.251367       0.047420  S                0.835411\n",
       "TCC            -8.011850       0.050991  S                0.898317\n",
       "TCG            -9.701874       0.058918  S                1.037969\n",
       "TCT            -7.473017       0.050396  S                0.887833\n",
       "ACA            -6.854227       0.062007  T                1.023216\n",
       "ACC            -6.948114       0.066979  T                1.105269\n",
       "ACG            -7.772748       0.057062  T                0.941624\n",
       "ACT            -6.987373       0.056351  T                0.929891\n",
       "GTA            -7.760370      -0.064119  V                1.325967\n",
       "GTC            -7.956008      -0.039379  V                0.814346\n",
       "GTG            -9.210331      -0.035104  V                0.725937\n",
       "GTT            -7.894414      -0.054824  V                1.133750\n",
       "TGA            -7.747629      -0.061969  W                1.488929\n",
       "TGG            -9.743693      -0.021271  W                0.511071\n",
       "TAC            -6.858673      -0.004837  Y                1.757137\n",
       "TAT            -6.839183      -0.000669  Y                0.242863\n",
       "\n",
       "[62 rows x 4 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupList = []\n",
    "for aa, group in damFoldSyn.groupby(['aa']):\n",
    "    group2 = group.copy()\n",
    "    group2['log10DAMratio_norm_syn'] = group2['log10DAMratio']/group2['log10DAMratio'].mean()\n",
    "    groupList.append(group2)\n",
    "damFoldSyn2 = pd.concat(groupList)\n",
    "damFoldSyn2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fdf9a917ef0>"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAECCAYAAAD5OrxGAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHYZJREFUeJzt3Xt0VPW9NvBn75nMTMJwDUZygcCRHhqInByLirJKUS4taEO7grpA33fJpRa5iMUACpF6IECFULWlVNIusC5MPVxKUUE4gh59lWMWlXcSMUNLuJQkFgohCUySue/zx5hJZjIze4fMnslkP5+/SNiX73wn8+zf/PaePYIkSRKIiEgzxHgXQEREscXgJyLSGAY/EZHGMPiJiDSGwU9EpDEMfiIijWHwExFpDIOfiEhjGPxERBrD4Cci0hh9vAsIxWKxwGg0qroPh8Oh+j4SBXvhwz60Yy/aJVIvHA4H8vLyZJfrkcFvNBqRk5Oj6j6sVqvq+0gU7IUP+9COvWiXSL2wWq2KluNUDxGRxjD4iYg0hsFPRKQxDH4iIo1h8BMRaQyDn4iiqt7mQEVNI+ptjniXQmH0yMs5iSgxHbTUYdX+SiSJIlxeLzYXjEV+Xma8y6IgHPETUVTU2xxYtb8SdpcXNx1u2F1erNxfyZF/D8TgJ6KoqG1oRZIYGClJoojahtY4VUThMPiJKCqyBibD5fUG/M7l9SJrYHKcKqJwGPxEFBWpZiM2F4yFKUlEX6MepiQRmwvGItWcGPe50RKe3CWiqMnPy8SEkYNR29CKrIHJDP0eSrXgP3z4MC5evIgTJ05g9+7d2LlzJ5xOJ4YNG4YZM2aotVsiirNUs5GB38OpNtUzY8YMZGdno7CwEABQV1cHk8kEg8Gg1i6JiEgBQZIkSa2Nb9q0CS+88AIA4PLlyxgyZAiKi4tRVFQUcb1Y3I/fbrfDZDKpuo9EwV74sA/t2It2idYLJbeQVm2qp7m5Gf3790ddXR0aGhpw4sQJiKKIESNGyK7L+/HHFnvhwz60Yy/aJVIvlN6PX7Xg79OnDxYtWgQAyMzMRG5urlq7IiKiLuDlnEREGsPgJyLSGAY/ESUU3v2z+/gBLiJKGLz7Z3RwxE9ECYF3/4weBj8RJQTe/TN6GPxElBB498/oYfATUULg3T+jhyd3iShh8O6f0cHgJ6KEwrt/dh+neoiINIbBT0SkMQx+IiKNYfATEWkMg5+ISGMY/EREGsPgJyLSGAY/EZHGMPiJiDSGwU9EpDEMfiIijWHwExFpDIOfiEhjGPxERBrD4Cci0hjV7sf//vvv4/Tp08jOzsajjz6KAwcOoKWlBUajEbNmzVJrt0REJEO1Ef+ZM2eQlpYG8ZsvR66ursbjjz+OS5cuqbVLIiJSQLUR/+zZszFkyBBs2LABkiRBr/ftyhv0ZcmhOBwOWK1WtUoDANjtdtX3kSjYCx/2oR170a439kK14P/444/R1NSE/v3746uvvkJWVhZ2796NESNGyK5rNBqRk5OjVmkAAKvVqvo+EgV74cM+tGMv2iVSL5QeoFQL/sceeyzg59zcXLV2RUREXcCreoiINIbBT0SkMQx+IiKNYfATEWkMg5+ISGMY/EREGsPgJyLSGAY/EZHGMPiJiDSGwU9EpDEMfiIijWHwExFpDIOfiEhjGPxERBrD4Cci0hgGPxGRxjD4iYg0hsFPRKQxDH4iIo1h8BMRaQyDn4hIY/RyC7z66quoq6uDXu9bdNOmTaoXRURE6pENfoPBgC1btsSiFiIiigHZ4C8vL8fNmzdhMpkAAMuWLVO9KCIiUo9s8G/cuDEWdRARUYzIBv+7776Lr7/+GnfffTemTZumeMNvvfUWWlpa0NraimeeeQY7d+6E0+nEsGHDMGPGjG4VTUREt072qp6FCxdi1qxZ2L9/P1asWIGdO3cq2vDUqVMxb948NDU1AQDq6upgMplgMBi6VzEREXWLIEmSFGmB5cuX46677sLMmTPRt29f/P73v8eCBQtkN9za2opt27Zh9uzZyMrKwuXLlzFkyBAUFxejqKgo4roWiwVGo7Frj6SL7Ha7/7yF1rEXPuxDO/aiXaL1IicnR3YZ2eCvrq5Geno6/vSnP2HMmDG46667FO38ueeeQ1paGvr06YNJkybhxIkTEEURycnJePzxxyOua7VaFRXfHbHYR6JgL3zYh3bsRbtE6oXSWmXn+Hfs2IHp06fDZrNh7969ioN/69atAT/n5uYqWo+IiNQlO8fv9Xrx3nvvYcqUKWhubo5FTUREpCLZEf/SpUvR0NAAs9nsv4b/3LlzuOOOO1QvjoiIok82+IcPH47hw4cH/O6DDz5g8BMRAai3OVDb0IqsgclINat7UUq0yAZ/KDLng4mINOGgpQ6r9lciSRTh8nqxuWAs8vMy412WLN6dk4joFtTbHFi1vxJ2lxc3HW7YXV6s3F+Jepsj3qXJuqXgz8rKinYdREQJpbahFUliYIQmiSJqG1rjVJFyslM9GzZsQENDAwRBgCAI2Lx5M374wx/GojYiSlCJOO/dVVkDk+HyegN+5/J6kTUwOU4VKScb/KmpqVizZk0saiGiXiBR5727KtVsxOaCsVgZ9FgT4UAnG/xnz57F0qVLYTKZ/CN+0g4tjNwoeuptDqzcVwmH2ws7fKPhlfsrMWHk4B739xONv+38vExMGDk44V4jssE/dOhQPPvss7GohXoYrYzcKHreKr8Ehztw+qNt3rsnhWI0/7ZTzUb/Y0uUgZJs8H/88ceorq5GSkoKAPToEX+iND0RdLxioSeP3Picd19Xehhp2XqbA7/56GyndZyenjXvfat/23J9SqSBkmzwL1q0CFOnTo1FLd2SSE1PBG1XLLS9MABlI7doBnF3X2jxPCiouW8l21a6/668bg5a6rByXyV0ogCPV8KWWYHL1ja0wqDTweF2B6y35IGRATVEs34lqq/chKWmEXlDB2Dk7X1R29AKvSgELNPxbzvUviP1qd7mwFdfN2HFXgucHtzyQCmWf6+KvohFFEX/iP++++5TtaBbkSij00RyK1csRPPgqyTUIz3nSmvp+GKLlmj1Ibi22oZWnK5rwvpDVWEDSG6Z4O0H97BwXyVGp/cLuWzh3gq4PO0f3nxub0XAayxrYDJanJ6A9UQBmHPvMEW96Vj/uveqAg4w4ebR5cJy7Z+/xJufX/L//H/vG4ZRt/eDzRFYp9PjRVOrC299/veA3r340GhIthas/+8LcLilTn9rn1Zfw6r9lfB6gaCHDp0g4KMz/8QD307zP3/h3im9VX4Jv/noLAw6XUwGrrLB/8ADD+DmzZu4efOmakV0162OTim8rl6xEM2Dr5JtRXrOASiqJTiElo1Pxa3cfTc4oKPRh461tbrcEAQBRr3oD6xwAaQXhbDLBO8/VA+dbi9+8NonePruwF589XVTQOgDgMsj4auvmzDxX33Btu+LWniCPtXvlYCGZqd/JB2uN2316wQBzUEJ+ux/WqAXAaNeHxCKcgfYv1yoDwh9AHjzfy7BqO/88SWn24und3/h33dbfWv+fBpGnQBH0GNPEkV89XWT//GE0uz04KV3v8ILB76EJElITtJ3qrPtXVTbeZG2d0tqD1xlg99sNuP999+HJEl46KGHVCmiuxL5etqerCtXLETz4KtkW5GecyXrhwqhV09cw6Pfc9xyQLu8XiyeNDLiAUlJL0PVBkhweTydlvUF0I2IARTueQjVQwBwe4Ffl9djSMbf8YMxQ1Db0Iobre5Oy/kI/ppL/uuvIZew1DT6p1hC9UYuQL2SbzTt9LSH4uj0frLv+J7bUxFye6FuOSMBnQ44bYJDH8A3fRM6PZ5gHd9Z3HQEhjrgGyQEnwwH1B+4Kjq5+8tf/hIAUFxcjClTpqhSSHck8vW0PV3HKxYiiebBV8m2wj3nANDU6oQzKCSD1w8VQjoRil9svnndG50uXdz20Vm0hWHHfZ+ua8Jjpf+jaPonVG3h+PokRVw+3PPQ1sPCfZVwhgiftX8+jXXvVsGgE+H0hN52Rn+Tv2aDTgx5cMobOgBA+OdVSYB2lCSKsNQ0yr7jc3tD31Ms+F1JV6Qk6eCFhM0FYzEmo1/IA6ecjnWGe9xqD1xlg99ut6Oiwnfk7MnTPYl6PW1vEc2Dr9JtBT/nn1Zfw4SXP0SSKMIrAXoRAW+vO64fKoQ8Xih6sbWN8kUInUZrBp0OT038F/zmv6vb54kfHo3171Upnv4JNxLvqI9BB4/UFkD9Qy7fcZlwz0N+XiZGp/fDD177BMHZ75EAj9sbckQKAKYk0T9KzhqYHDJQHx2XiZG39wUQ/nntaoC6vF7kDR3QpXd8bXSigFU/GIWNh89E3EeKQUSLM3B9o17E6//nOxiT0c/fz7bHI0kI26dQ9bf9nYV63Ea9oPrAVTb4X3zxRRw6dAiSJOGFF15QrZBoUDo6pehqm+OeMHIwPlv1YFQOvkoP5G3PeajpEaNexG8evyvghdpxveAQWjY+VdHljJGmJVxeL+bcOwxz7h3mr72r02DBtbXN8Zv0Ov+BJDejf0Bfgh9LqGXCGXl7X/xHfi7W/Pl0xOVCaQuwjjWLggCXx4sV00bhqe8F3r493PPasX672wOPRwo5/m8LxZG39404OAgVqAadgJJH/g35eZnoY9TjP96tQpJOgNPtDZiDf/Gh0cjN7I/TXzdh/XtVECHBC99+J/7rbWEfT/n5epT811+RpBPhcHv8z5nd7ek0xx/qcTs9Xix5YCTm3DtM9RyT/c7d3bt3o6KiAnq97xixadMmVQsC+J27sdadXvSUy2grahrxxO/L/fOoANDXqMfuBffi376Zagil44nZf9acl+1DqP0AQIpBB+83o+vgx19vc2DCyx8GHCxMSSI+W/Wg4ktjAfnzA929HPCt8r/7w9Dl8kAShIATukk6AaKAiFeedKeG4MdbVn4J2z6q9k81hQrFcPt7x1InG6hK+ltvc+D//f8qfPffRyueAgy1zXDb727PgkXtO3evXr2KLVu2dKsY6p160mW0t3qOoeO7xH/e4n6MegGvP3EXxmT0VzSCVzoNFvwOtqvLd9Xj92b7T+a2XKvFVWFgp5rl3oV1p4bgdZdO/lbAO6eu7E/JO0Yl/U01GzFqsEnxY4q0Tbl3rbEkG/wXLlzA7t27YTL5TuLMmjVL9aIoMfSky2hjdYI/3H7aLmkMJ1HOQbWFkNX2D9yXE7rmWNYezQMJtZMN/ieeeKLT7/iduwT0vMtoYxWut7qfRAyiRKw5HhLt1iGywX/PPfd0+h2/c5eAnnkZbayCioFIbXrKea6u4HfuUrckyhQGkRp60nmurril4CfqiKNf0qqedJ6rKxQF/40bNyBJEvr37w9A2XfuHj9+HJcvX4bL5cKTTz6JAwcOoKWlBUajkSeIiahX6GnnuZSSDf5t27bh+vXrAIC0tDQsXLhQ0Xfunjp1CitWrPDf7qG6ujrg50gcDgesVqvsct1ht9tV30eiYC982Id27EU7uV4sG5+KV09cg070ffp72fhU/LPmvKLLg+NFNvgbGxuxdu1aAF37Epa2D3x5vzkaBv8cidFo5Ae4Yoi98GEf2rEX7eR6kZMDPPq9nnFVj9KDtaKpntLSUgiCAKfTqbiA3NxcvPHGGzCZTDh9+jSysrKwe/dujBgxQvE2iIgSQaKd54oY/G335zl79ix0Ol2XLuEM/tau3NzcW6uQiIiiKmLwl5aW4uLFiwG/i8W9eoiISD1hg99ut2P69Ok4fvw4pkyZAkmS8Pbbb8eyNiIiUkHY4He73XjnnXdw6tQpNDc3QxAEfPvb345lbUREpIKwwW82m7FkyRKcO3cOX37p+87IxsbGWNZGREQq6Pytw0FKSkrw17/6vkvzyy+/VL0gIiJSl2zwDxo0CJMnT0bfvn39n9wlIqLEJRv8Dz30EAYNGoTPP/8ct912m9ziRETUw8l+gOvIkSNYt24dioqKYlEPERGpTDb4L168iCVLliAlJQVA127bQEREPU/E4K+pqcGGDRt4/30iol4kYvDPnj0bI0aMwD333ANBEAAAS5YsiUlhRESkjogndz/55BM8/fTT8Hg8EAQBEydOjFVdRESkkojBL4oi7r//fkybNg12ux2lpaWxqouIiFQScarnV7/6Fa5du4bRo0dj7ty5vI6fiKgXiBj8//jHPwAAFRUVqKiogCAI2LhxY0wKIyIidUQM/rZbMEuShOPHj2Pfvn0xKYqIiNQTMfivXbuGPXv24PPPP0dLSwtef/31WNVFREQqiXhyt6CgACaTCbt27cKYMWMwePDgWNVFREQqiRj8hw4dQnJyMp5++mmcOnUK1dXVsaqLiIhUEjH4zWYzZs+ejdLSUmzZsgV79uyJVV1ERKSSiHP858+fxxtvvAGPxwNJkjBv3rxY1UVERCqRvY7/F7/4BUwmE2w2G55//nls27YtVrUREZEKIk71GAwGmEwmAL5pn7Y7dBJRz1Fvc6CiphH1Nke8S6EEEXHEf+nSJf+1+5Ik4dKlSzEpioiUOWipw6r9lUgSRbi8XmwuGIv8vMx4l0U9XMTgX758uf+unG0/K3Hx4kWUlZXB7XZj4cKFSEtLw/Xr1/HSSy8hNzcXTzzxBN89EHVTvc2BVfsrYXd5YYcXALByfyUmjByMVLMxztVRTxYx+Jubm3Ho0CH/zzNnzoy4sbKyMpw8eRJOpxNFRUU4evQobDYb0tLSYLVakZGRAafTCYPBEJ3qiTSstqEVSaLoD30ASBJF1Da0MvgpIkGK8C0ra9euxbp16/w/FxUVobi4WNGG33nnHYiiiIcffhgAcP36dZjNZhw7dgxpaWkYN25c2HUtFguMRnX/cO12u//8hdaxFz6J1odGuwdP7rsEh6f9JWzUCXhj1jAMMOm6te1E64WaEq0XOTk5sstEHPG7XC4cOHAAt912G65evap4x5988gkOHjyI73znO6ipqUFVVRXS09Nx5MgRSJKExYsXR1zfaDQqKr47rFar6vtIFOyFTyL2YYswECuD5vjvi8IcfyL2Qi2J1Aur1apouYjBv27dOpSXl6OhoQGZmZn40Y9+pGijEydODPjSlqFDhwIAxo4dq2h9IlImPy8TE0YORm1DK7IGJnOKhxSJGPyXL19GdnY2srOzAQC1tbX+ECeiniHVbGTgU5fI3pY5JSUFw4cP9/+O37lLRJTYIgb/K6+8gj/+8Y948sknY1QOERGpLeInd41GI0OfiKiXiTjiB4CHH34YWVlZAHxz/Pn5+XjqqadUL4yIiNQhG/zjx49HUVERAODll1+GzWZTvSgiIlKPbPALgoDf/e53EAQBdrsdOl33PhhCRETxJRv8a9asgcVigdvtxoIFC+B0OmNRFxERqUQ2+IuLi5Geng5JkvDhhx9i5cqVsaiLiIhUEvGqnjbz58/HggULEOG2PkRElCAijvhramowYMAAzJs3D3q9HgMGDIhVXUREpJKIwX/w4EEAwKhRoyBJEsxmc0yKIiIi9UQMft6egYio91E0x09ERL0Hg5+ISGMY/EREGsPgJyLSGAY/EZHGMPiJiDSGwU9EpDEMfiIijWHwExFpDIOfiEhjGPxERBrD4Cci0hhVgl+SJCxevBilpaW4cuUKAN8XtW/btg0lJSXweDxq7JaIiBSQ/QaurigrK8PJkyfhdrvRr18/NDc3o0+fPgCAI0eOYM6cOTh58iTOnDmDMWPGRHPXRESkkCCp8LVaNpsNAHD+/HlcuHABM2fOxK5du/DjH/8Y5eXlyMjIwJ133hl2fYvFAqPRGO2yAtjtdphMJlX3kSjYCx/2oR170S7RepGTkyO7TFRH/G2ampqwc+dOJCcnY86cOTh69CgmT56MXbt2QRAEPPjggxHXNxqNiorvDqvVqvo+EgV74cM+tGMv2iVSL6xWq6LlVAn+zMxMvPjii/6fMzIyAAA/+9nP1NgdERF1Aa/qISLSGAY/EZHGMPiJiDSGwU9EpDEMfiIijWHwExFpDIOfiEhjGPxERBrD4Cci0hgGPxGRxjD4iYg0hsFPRKQxDH4iIo1h8BMRaQyDn4hIYxj8REQaw+AnItIYBj8RkcYw+ImINIbBT0SkMQx+IiKNYfATEWkMg5+ISGMY/EREGsPgJyLSGFWC/9y5c9i+fTvmz58Pi8UCALh+/TqeeeYZlJaWoqWlRY3dEhGRAvpobqysrAwnT54EABQWFsLtdiMvLw8AYLVakZGRAafTCYPBEM3dEhFRFwiSJElqbLisrAwTJkxAdnY2AN+I32w249ixY0hLS8O4cePCrmuxWGA0GtUoy89ut8NkMqm6j0TBXviwD+3Yi3aJ1oucnBzZZaI64u+orq7OH/pHjx5Feno6jhw5AkmSsHjx4ojrGo1GRcV3h9VqVX0fiYK98GEf2rEX7RKpF1arVdFyqgX/ihUr/P/+/ve/DwAYO3asWrsjIiKFeFUPEZHGMPiJiDSGwU9EpDEMfiIijWHwExFpDIOfiEhjGPxEFHf1NgcqahpRb3PEuxRNUO06fiIiJQ5a6rBqfyWSRBEurxebC8YiPy8z3mX1ahzxE1Hc1NscWLW/EnaXFzcdbthdXqzcX8mRv8oY/EQUN7UNrUgSA2MoSRRR29Aap4q0gcFPRHGTNTAZLq834HcurxdZA5PjVJE2MPiJKG5SzUZsLhgLU5KIvkY9TEkiNheMRapZ3bvzah1P7hJRXOXnZWLCyMGobWhF1sBkhn4MMPiJKO5SzUYGfgxxqoeISGMY/EREGsPgJyLSGAY/EZHGMPiJiDRGkCRJincRwSwWC4xGnuEnIuoKh8OBvLw82eV6ZPATEZF6ONVDRKQxDH4iIo1h8BMRaQyDn4hIYxj8REQao7ngP3z4MF577TUAwJkzZ7Bs2bI4VxQ/HXuxY8cOvP7667BYLHGuKj4uX76MDRs2YPv27Whqaop3OXHV2NiILVu2YMOGDbhy5Uq8y4mrc+fOYfv27Zg/f36vem1o4u6cZWVlOHnyJCRJwty5c1FVVQXAd83r0KFD41xdbIXqxfXr16HX6zF//ny88soriq4D7i3a+nH06FFs3boVgwYNwmeffYYZM2bEu7SYa+vFjRs3oNPp0KdPH6SkpMS7rLho6wUAFBYWwu1296rXhSav4y8pKUFhYWGnf2tRSUkJFixYgAMHDmDu3LnYunUrnnvuuXiXFXMtLS3YsWMHdDodcnJyMHXq1HiXFDfl5eWQJAkGgwFXrlzB9OnT411SXJWVlWHChAnIzs6OdylRo7mpHupswIABuHnzJkpLSzF58uR4lxMXdrsdOp0OBoMB999/f7zLiav09HQcO3YMH3zwAe688854lxN3dXV1vSr0AY2O+ImItIwjfiIijWHwExFpDIOfiEhjGPxERBrD4Cci0hgGP/U6J06cwMKFCwM+dfr888/D7Xb7fy4sLMTVq1exZ8+eeJRIFFea+OQuacvevXvhcrlQUlICr9eLxYsXAwAkSUJxcTEEQUBNTQ0cDgfq6urw61//Gq2trWhqakJBQQH0ej3+8Ic/wGAwICMjA0uXLg3Y/kcffYRPP/0UNpsNS5cuxZYtW/Ctb30L58+fx89//nMcP34cVVVVaG5uxurVq7Fo0SIMHz4cjzzySMB2a2trsX79ehw+fBiZmZm4++6749Eu0iCO+KnX+e53v4uzZ89i48aNWL16NcrKygD47ruSlpaGNWvWIC0tLWCd/Px8/OQnP8GJEyewZ88evPzyy5g9e3bI7b/99ttITk5GSkoKLBYLPB4PlixZgvHjx+Nvf/sb9uzZA5PJBFEUUVVVhb59+2L9+vWdtjtt2jQcP34cf/nLXzBu3Dh1m0LUAUf81Ct1/FxiqM8oimLgmCc5ORk6nQ6SJMHlcgEABEEIu/3ly5fjwoULsNvtMJlMAAC9Xg+v1wuz2YzCwkJYLBakpqb673cTvN1Jkybh2WefxZgxYyLuiyjaGPzUKz322GNYu3YtBEHAT3/6U/z2t7/FHXfcgQMHDmDTpk24du1a2HUfeeQRFBUVweVyYdSoUZ3+v6CgACtXrkRrayteeumlTv8/adIkrF69Gi0tLdi0aVPY7ep0OgwcOBD5+flRecxESvGWDURBjh07hoqKCjQ2NmLBggVRu09L8Ha/+OILnD17FqtWrYrK9omUYvATRWCz2fDmm2/6f9br9XjqqafiWBFR9zH4iYg0hlf1EBFpDIOfiEhjGPxERBrD4Cci0hgGPxGRxvwvnCYWHzOQFyoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "damFoldSyn2.plot.scatter(x='folding_energy', y='log10DAMratio_norm_syn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>folding_energy</th>\n",
       "      <th>log10DAMratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_-1</th>\n",
       "      <th>codon_-1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [folding_energy, log10DAMratio]\n",
       "Index: []"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "damFoldSyn.groupby(['aa', 'codon_-1']).mean()[['folding_energy', 'log10DAMratio']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare ELISA to ELM-seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We compare the ELISA results to the DAM screening.\n",
    "\n",
    "First, we take the exact -1 codon that was used.\n",
    "\n",
    "Then, we take the exact codon pair (-2 and -1) that was used.\n",
    "\n",
    "Note that we should compare the linear scale of the normalized cat to the linear scale of the DAMratio (not the log!)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elisaCodons = {\n",
    "    'A':'GCA',\n",
    "    'C':'TGT',\n",
    "    'D':'GAT',\n",
    "    'E':'GAA',\n",
    "    'F':'TTT',\n",
    "    'G':'GGA',\n",
    "    'H':'CAT',\n",
    "    'I':'ATT',\n",
    "    'K':'AAA',\n",
    "    'L':'TTA',\n",
    "    'M':'ATG',\n",
    "    'N':'AAT',\n",
    "    'P':'CCA',\n",
    "    'Q':'CAA',\n",
    "    'R':'AGA',\n",
    "    'S':'TCA',\n",
    "    'T':'ACA',\n",
    "    'V':'GTT',\n",
    "    'W':'TGG',\n",
    "    'Y':'TAT'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "elisaDf2 = elisaDf.rename(columns={'C-terminal amino acid':'aa', 'CAT (ng/ml)':'CAT'})\n",
    "elisaDf2['CAT'] = pd.to_numeric(elisaDf2['CAT'])\n",
    "# elisaDf2.groupby('aa')['CAT'].mean()\n",
    "elisaDf2 = elisaDf2.pivot(index='aa', columns='replicate', values='CAT')\n",
    "elisaDf2.columns = ['CAT rep1', 'CAT rep2']\n",
    "elisaDf2['CAT'] = elisaDf2.mean(axis=1)\n",
    "elisaDf2['codon'] = elisaDf2.index.map(elisaCodons)\n",
    "elisaDf2 = elisaDf2.reset_index().set_index('codon')\n",
    "elisaDf2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodon[damDfCodon['position'] == -1].loc['strong'][['DAMratio']].copy()\n",
    "df.columns = ['DAMratio strong promoter']\n",
    "df1 = df\n",
    "df = damDfCodon[damDfCodon['position'] == -1].loc['weak'][['DAMratio']].copy()\n",
    "df.columns = ['DAMratio weak promoter']\n",
    "df2 = df\n",
    "df3 = df1.join(df2)\n",
    "df3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elisaDf3 = elisaDf2.join(df3).sort_values('CAT')\n",
    "elisaDf3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elisaDf3.to_csv(ctermELMseqPath / 'elisaDf3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AAhydrophobic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AApositive = ['K', 'R']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elisaDf3['amino acid group'] = \"others\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elisaDf3.loc[elisaDf3['aa'].map(lambda x: x in AApositive), 'amino acid group'] = 'positively charged'\n",
    "elisaDf3.loc[elisaDf3['aa'].map(lambda x: x in AAhydrophobic), 'amino acid group'] = 'hydrophobic'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for ylabel in ['CAT']:\n",
    "    set_fontsize_paper()\n",
    "    width = 2.5\n",
    "    fig, ax = plt.subplots(figsize=(width, width))\n",
    "    df = elisaDf3\n",
    "    sns.boxplot(data=elisaDf3, x='amino acid group', y=ylabel,\n",
    "                order=['hydrophobic', 'positively charged', 'others'], fliersize=0)\n",
    "    sns.swarmplot(data=elisaDf3, x='amino acid group',\n",
    "                  order=['hydrophobic', 'positively charged', 'others'], y=ylabel, size=5, color='0.2')\n",
    "    ylim = ax.get_ylim()\n",
    "    add_stat_annotation(ax, data=elisaDf3, x='amino acid group', y=ylabel,\n",
    "                        order=['hydrophobic', 'positively charged', 'others'], \n",
    "                        boxPairList=[('hydrophobic', 'positively charged')],\n",
    "                        test='t-test_ind', textFormat='star', loc='outside',\n",
    "                        lineYOffsetAxesCoord=0.010, lineYOffsetToBoxAxesCoord=0.005,\n",
    "                        textYOffsetPoints=0, verbose=2, linewidth=1)\n",
    "    ax.set_ylabel(ylabel + ' ng/ml')\n",
    "    filename = 'ELM_seq_ELISA_{}_aa_group'.format(ylabel)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(ctermELMseqPath / (filename + '.svg'), dpi=300)\n",
    "    fig.savefig(ctermELMseqPath / (filename + '.png'), dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for promoter in ['weak', 'strong']:\n",
    "    for xlabel in ['CAT', 'CAT rep1', 'CAT rep2']:\n",
    "        set_fontsize_paper()\n",
    "        width = 1.5\n",
    "        fig, ax = plt.subplots(figsize=(width, width))\n",
    "        df = elisaDf3\n",
    "        ylabel = 'DAMratio {} promoter'.format(promoter)\n",
    "        pearson = df[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "        ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.06, 0.93),\n",
    "                    xycoords='axes fraction', ha='left', va='top', fontsize='medium',\n",
    "                    bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "        plt.tight_layout()\n",
    "        sns.regplot(ax=ax, data=df, x=xlabel, y=ylabel, ci=False, n_boot=1000, fit_reg=True,\n",
    "                    scatter_kws=dict(alpha=1, s=20, edgecolors='none'))\n",
    "        ylim = list(ax.get_ylim())\n",
    "        ylim[1] = ylim[0] + 1.2*(ylim[1] - ylim[0])\n",
    "        ax.set_ylim(ylim)\n",
    "        ax.set_ylabel('DAMratio {} promoter'.format(promoter))\n",
    "        ax.set_xlabel(xlabel + ' ng/ml')\n",
    "        ax.set_title('{} promoter'.format(promoter))\n",
    "        filename = 'ELM_seq_ELISA_{}_correlation_promoter_{}.png'.format(xlabel, promoter)\n",
    "        fig.tight_layout()\n",
    "        fig.savefig(ctermELMseqPath / filename, dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the cat construct, the -2 codon is A/GCG. If we take into account the -2 codon, the correlation is worse in general."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = damDfCodonpair.loc[idx['strong', :, 'GCG'], :][['DAMratio']].copy()\n",
    "df.columns = ['DAMratio strong promoter']\n",
    "df1 = df.reset_index(level=0, drop=True)\n",
    "df = damDfCodonpair.loc[idx['weak', :, 'GCG'], :][['DAMratio']].copy()\n",
    "df.columns = ['DAMratio weak promoter']\n",
    "df2 = df.reset_index(level=0, drop=True)\n",
    "df3 = df1.join(df2)\n",
    "df3.index = df3.index.droplevel('codon_-2')\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "elisaDf3 = elisaDf2.join(df3)\n",
    "elisaDf3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for promoter in ['weak', 'strong']:\n",
    "    for xlabel in ['CAT', 'CAT rep1', 'CAT rep2']:\n",
    "        set_fontsize_paper()\n",
    "        width = 1.5\n",
    "        fig, ax = plt.subplots(figsize=(width, width))\n",
    "        df = elisaDf3\n",
    "        ylabel = 'DAMratio {} promoter'.format(promoter)\n",
    "        pearson = df[[xlabel, ylabel]].corr().iloc[0, 1]\n",
    "        ax.annotate('pearson r={:.2f}'.format(pearson), xy=(0.06, 0.93),\n",
    "                    xycoords='axes fraction', ha='left', va='top', fontsize='medium',\n",
    "                    bbox=dict(boxstyle=\"round\", fc='1', ec='0.5'),)\n",
    "        plt.tight_layout()\n",
    "        sns.regplot(ax=ax, data=df, x=xlabel, y=ylabel, ci=False, n_boot=1000, fit_reg=True,\n",
    "                    scatter_kws=dict(alpha=1, s=20, edgecolors='none'))\n",
    "        ylim = list(ax.get_ylim())\n",
    "        ylim[1] = ylim[0] + 1.2*(ylim[1] - ylim[0])\n",
    "        ax.set_ylim(ylim)\n",
    "        ax.set_ylabel('DAMratio {} promoter'.format(promoter))\n",
    "        ax.set_xlabel(xlabel + ' ng/ml')\n",
    "        ax.set_title('{} promoter'.format(promoter))\n",
    "        filename = 'ELM_seq_ELISA_{}_correlation_promoter_{}_consider_codon-2.png'.format(xlabel, promoter)\n",
    "        fig.tight_layout()\n",
    "        fig.savefig(ctermELMseqPath / filename, dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "455px",
    "width": "252px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "897px",
    "left": "0px",
    "right": "1238px",
    "top": "136px",
    "width": "309px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
